From a4f90cb227cbe458d13f09e69e550ea27054285b Mon Sep 17 00:00:00 2001
From: Jiang Bin <bin.jiang@windriver.com>
Date: Mon, 9 Jul 2012 16:21:42 +0800
Subject: [PATCH 234/236] SE: Fix octeon_edac_pci has unknown symbols in module.

octeon_edac_pci module requires exported cvmx_pcie_cfgx_write and
cvmx_pcie_cfgx_read symbols.

Signed-off-by: Jiang Bin <bin.jiang@windriver.com>
---
 simple_exec/cvmx-pcie.c |    4 ++--
 1 files changed, 2 insertions(+), 2 deletions(-)

diff --git a/simple_exec/cvmx-pcie.c b/simple_exec/cvmx-pcie.c
index 493197b..b0de664 100644
--- a/simple_exec/cvmx-pcie.c
+++ b/simple_exec/cvmx-pcie.c
@@ -1413,7 +1413,7 @@ uint32_t cvmx_pcie_cfgx_read(int pcie_port, uint32_t cfg_offset)
         return pemx_cfg_rd.s.data;
     }
 }
-
+EXPORT_SYMBOL(cvmx_pcie_cfgx_read);
 
 /**
  * Write a PCIe config space register indirectly. This is used for
@@ -1442,7 +1442,7 @@ void cvmx_pcie_cfgx_write(int pcie_port, uint32_t cfg_offset, uint32_t val)
         cvmx_write_csr(CVMX_PEMX_CFG_WR(pcie_port), pemx_cfg_wr.u64);
     }
 }
-
+EXPORT_SYMBOL(cvmx_pcie_cfgx_write);
 
 /**
  * Initialize a PCIe port for use in target(EP) mode.
-- 
1.7.0

