From c10c4df96eb7e861c1e093c0d2d7bee3b5c0a4f7 Mon Sep 17 00:00:00 2001
From: Jiang Bin <bin.jiang@windriver.com>
Date: Tue, 17 Apr 2012 13:52:10 +0800
Subject: [PATCH 233/236] SE: Get MII PHY address associated with given IPD port.

Boot cn61xx and cn66xx boards, there are following warning message:
cvmx_helper_board_get_mii_address: Unknown board type 46
cvmx_helper_board_get_mii_address: Unknown board type 46

cn61xx and cn66xx boards have same ports with cn63xx board.

Signed-off-by: Jiang Bin <bin.jiang@windriver.com>
---
 simple_exec/cvmx-helper-board.c |    2 ++
 1 files changed, 2 insertions(+), 0 deletions(-)

diff --git a/simple_exec/cvmx-helper-board.c b/simple_exec/cvmx-helper-board.c
index 56eaf39..b764f6f 100644
--- a/simple_exec/cvmx-helper-board.c
+++ b/simple_exec/cvmx-helper-board.c
@@ -461,7 +461,9 @@ int cvmx_helper_board_get_mii_address(int ipd_port)
                 else
                     return -1;
             }
+        case CVMX_BOARD_TYPE_EBB6100:
         case CVMX_BOARD_TYPE_EBB6300:
+        case CVMX_BOARD_TYPE_EBB6600:
 #ifdef CONFIG_WRHV
 	case CVMX_BOARD_TYPE_CUST_WRHV_OCTEON:
 #endif
-- 
1.7.0

