From da687fe269a13793cac60a1e4f3d3c83cadbbe6f Mon Sep 17 00:00:00 2001
From: Yanjiang Jin <yanjiang.jin@windriver.com>
Date: Fri, 1 Nov 2013 13:16:29 +0800
Subject: [PATCH] MIPS: add gcc option -fno-omit-frame-pointer

commit b732d439cb("ftrace: Remove selecting FRAME_POINTER with FUNCTION_TRACER")
removes the gcc option -fno-omit-frame-pointer.
But MIPS still needs it, else ftrace will insert the jump instruction
"INSN_B_1F", 0x10000004 or 0x10000005, into a wrong place. It brings the
below errors during installing some modules, such as tipc, ext4.

    Reserved instruction in kernel code[#1]:
    CPU: 3 PID: 616 Comm: modprobe Not tainted 3.10.10-WR6.0.0.0_standard #5
    task: c0000000ff2ce900 ti: c0000000fe880000 task.ti: c0000000fe880000
    $ 0   : 0000000000000000 ffffffffc16e0258 ffffffffe002c420 0000000000016c4f
    $ 4   : 0000000000000001 0000000000000001 0000000000000001 0000000000000003
    $ 8   : 0000000000000000 ffffffffd0450001 ffffffffd0450002 0000000000000000
    $12   : c0000000fe88fb90 ffffffffc13fcd68 0000000000000001 c0000001005bffff
    $16   : ffffffffe003c198 0000000000000001 0000000000000003 ffffffffd0450002
    $20   : ffffffffd0450001 0000000000000001 0000000000000001 ffffffffc1920000
    $24   : 0000000000010000 00000000f5c8079f
    $28   : c0000000fe880000 c0000000fe88fb28 0000000000000000 ffffffffe002cb48
    Hi    : 000000000002fece
    Lo    : 000000000000ff9a
    epc   : ffffffffe002c420 nametbl_find_seq+0x0/0x78 [tipc]
        Not tainted
    ra    : ffffffffe002cb48 tipc_nametbl_insert_publ+0x68/0x5d0 [tipc]
    Status: 7010f8e3    KX SX UX KERNEL EXL IE
    Cause : 80800028
    PrId  : 000c1201 (XLP208 Rev A1)
    Modules linked in: tipc(+)
    Process modprobe (pid: 616, threadinfo=c0000000fe880000)
    Stack : 0000000000000001 ffffffffc1148d0c c0000000ff596040 ffffffffc1051594
          ffffffffe0040000 ffffffffc16e0198 ffffffffe003c198 ffffffffd0450002
          ffffffffe0040000 ffffffffd0450001 0000000000000001 0000000000000001
          ffffffffe003ab38 ffffffffc1920000 ffffffffc1920000 ffffffffe002d9a8
          0000000000000001 0000000000000001 0000000000000003 ffffffffc16e0198
          c0000000fea64c00 ffffffffd0450001 c0000000fe88fc38 0000000000000003
          ffffffffc16d5128 ffffffffe0030f70 ffffffffc16d5128 0000000000000000
          ffffffffe0040000 ffffffffe0040000 ffffffffe0040000 ffffffffe002c2b4
          0000000000000000 ffffffffe003c170 0000000100000001 00000001e003ab38
          ffffffffc1920000 0000000000000000 ffffffffc16d5128 ffffffffe00600d4
          ...
    Call Trace:
    [<ffffffffe002c420>] nametbl_find_seq+0x0/0x78 [tipc]
    [<ffffffffe002cb48>] tipc_nametbl_insert_publ+0x68/0x5d0 [tipc]
    [<ffffffffe002d9a8>] tipc_nametbl_publish+0xa0/0x118 [tipc]
    [<ffffffffe0030f70>] tipc_publish+0xc0/0x110 [tipc]
    [<ffffffffe002c2b4>] tipc_subscr_start+0xb4/0x108 [tipc]
    [<ffffffffe00600d4>] tipc_init+0xd4/0x178 [tipc]
    [<ffffffffc10004f0>] do_one_initcall+0xf0/0x140

    Code: 67bd0040  00000000  00000000 <10000004> 10000004

Signed-off-by: Yanjiang Jin <yanjiang.jin@windriver.com>
---
 Makefile | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/Makefile b/Makefile
index 89e20e1..65b77c7 100644
--- a/Makefile
+++ b/Makefile
@@ -614,6 +614,10 @@ else
 # -fomit-frame-pointer with FUNCTION_TRACER.
 ifndef CONFIG_FUNCTION_TRACER
 KBUILD_CFLAGS	+= -fomit-frame-pointer
+else
+ifeq ($(ARCH),mips)
+KBUILD_CFLAGS	+= -fno-omit-frame-pointer
+endif
 endif
 endif
 
-- 
1.8.4.93.g57e4c17

