From dafb9d4767517a755d55575b1d6e2d57c70fe774 Mon Sep 17 00:00:00 2001
From: Jianchuan Wang <jianchuan.wang@windriver.com>
Date: Tue, 25 Oct 2016 07:06:38 +0000
Subject: [PATCH] Remove the __init since the function is called by kernel

Signed-off-by: Yue Tao <Yue.Tao@windriver.com>
Signed-off-by: Jianchuan Wang <jianchuan.wang@windriver.com>
---
 arch/mips/pci/fixup-malta.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/arch/mips/pci/fixup-malta.c b/arch/mips/pci/fixup-malta.c
index b41a06d..83cf3ec 100644
--- a/arch/mips/pci/fixup-malta.c
+++ b/arch/mips/pci/fixup-malta.c
@@ -142,7 +142,7 @@ DECLARE_PCI_FIXUP_HEADER(PCI_VENDOR_ID_INTEL, PCI_DEVICE_ID_INTEL_82371AB,
 	 malta_piix_func1_fixup);
 
 /* Use the io resource allocated by YAMON */
-static void __init malta_pcnet32_ioports_fixup(struct pci_dev *pdev)
+static void  malta_pcnet32_ioports_fixup(struct pci_dev *pdev)
 {
 	u32 check = 0, new = 0x1060;
 
-- 
2.0.2

