From a1c3e46f2ee4446a848ff1e5c711a5fe4230986d Mon Sep 17 00:00:00 2001
From: Ralf Baechle <ralf@linux-mips.org>
Date: Thu, 27 Jun 2013 02:04:25 +0200
Subject: MIPS: tlbex: Fix size of area to be flushed.

commit 3f90b82df110ef9cb33761b56ca85ae0d0372d4a upstream.

Signed-off-by: Ralf Baechle <ralf@linux-mips.org>
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/mm/tlbex.c b/arch/mips/mm/tlbex.c
index bef2ad0..d27bc88 100644
--- a/arch/mips/mm/tlbex.c
+++ b/arch/mips/mm/tlbex.c
@@ -2182,7 +2182,7 @@ static void __cpuinit flush_tlb_handlers(void)
 			   (unsigned long)handle_tlbm + sizeof(handle_tlbm));
 #ifdef CONFIG_MIPS_PGD_C0_CONTEXT
 	local_flush_icache_range((unsigned long)tlbmiss_handler_setup_pgd_array,
-			   (unsigned long)tlbmiss_handler_setup_pgd_array + sizeof(handle_tlbm));
+			   (unsigned long)tlbmiss_handler_setup_pgd_array + sizeof(tlbmiss_handler_setup_pgd_array));
 #endif
 }
 
-- 
1.7.1

