From c6bef165f773ff47f570d422487784106facab70 Mon Sep 17 00:00:00 2001
From: Kamlakant Patel <kamlakant.patel@broadcom.com>
Date: Sun, 19 Oct 2014 18:05:10 +0530
Subject: sd/mmc: fix simultaneous transfer issue

Signed-off-by: Kamlakant Patel <kamlakant.patel@broadcom.com>
[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/drivers/mmc/host/sdhci.c b/drivers/mmc/host/sdhci.c
index 8a66737..fefba27 100644
--- a/drivers/mmc/host/sdhci.c
+++ b/drivers/mmc/host/sdhci.c
@@ -180,8 +180,13 @@ static void sdhci_reset(struct sdhci_host *host, u8 mask)
 
 	if (host->quirks & SDHCI_QUIRK_NO_CARD_NO_RESET) {
 		if (!(sdhci_readl(host, SDHCI_PRESENT_STATE) &
-			SDHCI_CARD_PRESENT))
+			SDHCI_CARD_PRESENT)) {
+#ifdef CONFIG_CPU_XLP
+			sdhci_writeb(host, mask, SDHCI_SOFTWARE_RESET);
+			udelay(100);
+#endif
 			return;
+		}
 	}
 
 	if (host->quirks & SDHCI_QUIRK_RESTORE_IRQS_AFTER_RESET)
@@ -1305,6 +1310,16 @@ static int sdhci_set_power(struct sdhci_host *host, unsigned short power)
  *                                                                           *
 \*****************************************************************************/
 
+/**
+ * On Netlogic XLP platform, while doing transfers on both the slots
+ * simultaneously, hardware generates CRC error. This is a software workaround
+ * to avoid the CRC error.
+ */
+
+#ifdef CONFIG_CPU_XLP
+static DEFINE_MUTEX(sdhci_cmd_inprogress);
+#endif
+
 static void sdhci_request(struct mmc_host *mmc, struct mmc_request *mrq)
 {
 	struct sdhci_host *host;
@@ -1312,6 +1327,10 @@ static void sdhci_request(struct mmc_host *mmc, struct mmc_request *mrq)
 	unsigned long flags;
 	u32 tuning_opcode;
 
+#ifdef CONFIG_CPU_XLP
+	mutex_lock(&sdhci_cmd_inprogress);
+#endif
+
 	host = mmc_priv(mmc);
 
 	sdhci_runtime_pm_get(host);
@@ -2150,6 +2169,10 @@ static void sdhci_tasklet_finish(unsigned long param)
 	mmiowb();
 	spin_unlock_irqrestore(&host->lock, flags);
 
+#ifdef CONFIG_CPU_XLP
+	mutex_unlock(&sdhci_cmd_inprogress);
+#endif
+
 	mmc_request_done(host->mmc, mrq);
 	sdhci_runtime_pm_put(host);
 }
-- 
1.7.1

