From ede1571b13e873365d94bedd7d97eb2af5a50d48 Mon Sep 17 00:00:00 2001
From: Subhendu Sekhar Behera <sbehera@broadcom.com>
Date: Wed, 3 Sep 2014 17:57:30 +0530
Subject: dts: add cpld entry for each board

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/netlogic/dts/xlp_evp.dts b/arch/mips/netlogic/dts/xlp_evp.dts
index cf0d901..96f5103 100644
--- a/arch/mips/netlogic/dts/xlp_evp.dts
+++ b/arch/mips/netlogic/dts/xlp_evp.dts
@@ -36,6 +36,14 @@
 			interrupt-parent = <&pic>;
 			interrupts = <18>;
 		};
+		cpld@0 {
+			compatible = "netlogic,xlp-cpld";
+			reg = <1 0x1000000 0x100000>;
+			reg-io-width = <2>;
+			version-reg = <0>;
+			reset-reg = <1>;
+			little-endian;
+		};
 		i2c0: ocores@32000 {
 			compatible = "opencores,i2c-ocores";
 			#address-cells = <1>;
diff --git a/arch/mips/netlogic/dts/xlp_fvp.dts b/arch/mips/netlogic/dts/xlp_fvp.dts
index 4e61019..8cecc77 100644
--- a/arch/mips/netlogic/dts/xlp_fvp.dts
+++ b/arch/mips/netlogic/dts/xlp_fvp.dts
@@ -36,6 +36,14 @@
 			interrupt-parent = <&pic>;
 			interrupts = <18>;
 		};
+		cpld@0 {
+			compatible = "netlogic,xlp-cpld";
+			reg = <1 0x1000000 0x100000>;
+			reg-io-width = <2>;
+			version-reg = <0>;
+			reset-reg = <1>;
+			little-endian;
+		};
 		i2c0: ocores@37100 {
 			compatible = "opencores,i2c-ocores";
 			#address-cells = <1>;
diff --git a/arch/mips/netlogic/dts/xlp_gvp.dts b/arch/mips/netlogic/dts/xlp_gvp.dts
index c5c37ff..eb38657 100644
--- a/arch/mips/netlogic/dts/xlp_gvp.dts
+++ b/arch/mips/netlogic/dts/xlp_gvp.dts
@@ -33,6 +33,14 @@
 			reg = <0 0x110000 0x200>;
 			interrupt-controller;
 		};
+		cpld@0 {
+			compatible = "netlogic,xlp-cpld";
+			reg = <1 0x1000000 0x100000>;
+			reg-io-width = <2>;
+			version-reg = <0>;
+			reset-reg = <1>;
+			little-endian;
+		};
 		gpio: xlp_gpio@114000 {
 			compatible = "netlogic,xlp-gpio";
 			reg = <0 0x114100 0x1000>;
diff --git a/arch/mips/netlogic/dts/xlp_rvp.dts b/arch/mips/netlogic/dts/xlp_rvp.dts
index 6849f49..ae8ae20 100644
--- a/arch/mips/netlogic/dts/xlp_rvp.dts
+++ b/arch/mips/netlogic/dts/xlp_rvp.dts
@@ -36,6 +36,14 @@
 			interrupt-parent = <&pic>;
 			interrupts = <18>;
 		};
+		cpld@0 {
+			compatible = "netlogic,xlp-cpld";
+			reg = <1 0x1000000 0x100000>;
+			reg-io-width = <2>;
+			version-reg = <0>;
+			reset-reg = <1>;
+			little-endian;
+		};
 		pic: pic@11000 {
 			interrupt-controller;
 			#address-cells = <0>;
diff --git a/arch/mips/netlogic/dts/xlp_svp.dts b/arch/mips/netlogic/dts/xlp_svp.dts
index 662e0c8..1f2b92a 100644
--- a/arch/mips/netlogic/dts/xlp_svp.dts
+++ b/arch/mips/netlogic/dts/xlp_svp.dts
@@ -36,6 +36,14 @@
 			interrupt-parent = <&pic>;
 			interrupts = <18>;
 		};
+		cpld@0 {
+			compatible = "netlogic,xlp-cpld";
+			reg = <1 0x1000000 0x100000>;
+			reg-io-width = <2>;
+			version-reg = <0>;
+			reset-reg = <1>;
+			little-endian;
+		};
 		i2c0: ocores@32000 {
 			compatible = "opencores,i2c-ocores";
 			#address-cells = <1>;
-- 
1.7.1

