From 6d1f816480c044fb3f7b434ff6c2361183bf0634 Mon Sep 17 00:00:00 2001
From: Jayachandran C <jchandra@broadcom.com>
Date: Wed, 11 Jun 2014 13:29:26 +0530
Subject: initrd passed in fw_arg1

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/netlogic/xlp/setup.c b/arch/mips/netlogic/xlp/setup.c
index 9c24f99..e4c76e9 100644
--- a/arch/mips/netlogic/xlp/setup.c
+++ b/arch/mips/netlogic/xlp/setup.c
@@ -98,6 +98,21 @@ void __init plat_mem_setup(void)
 	_machine_halt	= nlm_linux_exit;
 	pm_power_off	= nlm_linux_exit;
 
+#ifdef CONFIG_BLK_DEV_INITRD
+	if (fw_arg1) {
+		 extern unsigned long initrd_start, initrd_end;
+		 initrd_start = (unsigned long)fw_arg1;
+#if defined(CONFIG_32BIT) && defined (CONFIG_MAPPED_KERNEL)
+		 initrd_start = CKSEG2ADDR(initrd_start);
+#else
+		 initrd_start = CKSEG1ADDR(initrd_start);
+#endif
+		 initrd_end = initrd_start + (unsigned long)fw_arg2;
+		 printk("initrd_start: 0x%lx initrd_end: 0x%lx\n",
+			 initrd_start, initrd_end);
+	}
+#endif
+
 	/* memory and bootargs from DT */
 	early_init_devtree(initial_boot_params);
 
-- 
1.7.1

