From b42f7ec2caf0aaa54440f06374372faa07e3f286 Mon Sep 17 00:00:00 2001
From: Kamlakanth Patel <kpatel@broadcom.com>
Date: Mon, 15 Sep 2014 22:33:53 +0530
Subject: Cpuinfo: Netlogic: Added @ xxxMHz to cpuinfo

Signed-off-by: Prem Mallappa <pmallapp@broadcom.com>
[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/kernel/proc.c b/arch/mips/kernel/proc.c
index 57b524f..19ff18d 100644
--- a/arch/mips/kernel/proc.c
+++ b/arch/mips/kernel/proc.c
@@ -7,6 +7,7 @@
 #include <linux/kernel.h>
 #include <linux/sched.h>
 #include <linux/seq_file.h>
+#include <linux/cpufreq.h>
 #include <asm/bootinfo.h>
 #include <asm/cpu.h>
 #include <asm/cpu-features.h>
@@ -14,6 +15,7 @@
 #include <asm/mipsregs.h>
 #include <asm/processor.h>
 #include <asm/prom.h>
+#include <asm/time.h>
 
 unsigned int vced_count, vcei_count;
 
@@ -22,6 +24,7 @@ static int show_cpuinfo(struct seq_file *m, void *v)
 	unsigned long n = (unsigned long) v - 1;
 	unsigned int version = cpu_data[n].processor_id;
 	unsigned int fp_vers = cpu_data[n].fpu_id;
+	unsigned int freq;
 	char fmt [64];
 	int cpu,i;
 
@@ -41,8 +44,19 @@ static int show_cpuinfo(struct seq_file *m, void *v)
 	}
 
 	seq_printf(m, "processor\t\t: %ld\n", n);
-	sprintf(fmt, "cpu model\t\t: %%s V%%d.%%d%s\n",
-		      cpu_data[n].options & MIPS_CPU_FPU ? "  FPU V%d.%d" : "");
+#ifdef CONFIG_MIPS_CPUFREQ
+	freq = cpufreq_quick_get(n)/1000;
+#else
+	freq = mips_hpt_frequency / (1000 * 1000);
+#endif
+	if (freq)
+		sprintf(fmt, "cpu model\t\t: %%s V%%d.%%d%s\t@ %uMHz\n",
+			cpu_data[n].options & MIPS_CPU_FPU ? "  FPU V%d.%d" : "",
+			freq);
+	else
+		sprintf(fmt, "cpu model\t\t: %%s V%%d.%%d%s\n",
+			cpu_data[n].options & MIPS_CPU_FPU ? "  FPU V%d.%d" : "");
+
 	cpu = get_cpu();
 	seq_printf(m, fmt, __cpu_name[cpu],
 		      (version >> 4) & 0x0f, version & 0x0f,
-- 
1.7.1

