From ec5fc4a8a7eb02e4db06ac186ac9a6091fbd6a48 Mon Sep 17 00:00:00 2001
From: Saswat Kumar Dash <saswat.dash@broadcom.com>
Date: Wed, 17 Dec 2014 02:02:53 -0800
Subject: FMN spill: handling error condition

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/netlogic/xlp/dt.c b/arch/mips/netlogic/xlp/dt.c
index b5e0250..c50e6b0 100644
--- a/arch/mips/netlogic/xlp/dt.c
+++ b/arch/mips/netlogic/xlp/dt.c
@@ -150,8 +150,10 @@ static int __init xlp_reserve_memory(unsigned long node, const char *uname,
 
 	/* Reserve memory for fmn spill */
 	p = of_get_flat_dt_prop(node, "netlogic,spill-mem", &size);
-	if (p == NULL)
+	if (p == NULL) {
+		spill_mem_size = 0xffffffff;
 		return 0;
+	}
 
 	val = be32_to_cpup(p);
 
-- 
1.7.1

