From d57eb7a6c99addb171677c083825c7b24f41b36e Mon Sep 17 00:00:00 2001
From: Venu Vadapalli <vvadapalli@netlogicmicro.com>
Date: Fri, 10 Sep 2010 23:07:02 -0700
Subject: Integrated nae-config and ucore configuration to build system, Dummy soc node for nae-config

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/boot/dts/netos.dts b/arch/mips/boot/dts/netos.dts
index 830efe7..3ea4de2 100644
--- a/arch/mips/boot/dts/netos.dts
+++ b/arch/mips/boot/dts/netos.dts
@@ -62,4 +62,8 @@
 	chosen {
 		bootargs = " xlp_noi2c console=ttyS0,115200 rdinit=/sbin/init";
 	};
+
+	soc {
+
+	};
 };
-- 
1.7.1

