From fb9b92d9e6fe24406f31789b357e59591cbd4141 Mon Sep 17 00:00:00 2001
From: Venu Vadapalli <vvadapalli@netlogicmicro.com>
Date: Thu, 23 Sep 2010 13:15:49 -0700
Subject: Use dram map < 2GB

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/boot/dts/multi-os.dts b/arch/mips/boot/dts/multi-os.dts
index 9136a46..9eb6010 100644
--- a/arch/mips/boot/dts/multi-os.dts
+++ b/arch/mips/boot/dts/multi-os.dts
@@ -14,8 +14,8 @@
 		hypervisor-name = "Xen";
 		cpu_possible_map = <0xff>;
 		is_dom0_hvm = <1>;
-		bootargs = "ncores=2 dom0_loadaddr=0xe0800000 dom0_size=0x1c000000 dom0_cpumask=0x0f -- ";
-		domain_heap = <0x80000000 0x20000000>;
+		bootargs = "ncores=1 dom0_loadaddr=0x78000000 dom0_size=0x1c000000 dom0_cpumask=0x0f -- ";
+		domain_heap = <0x50000000 0x20000000>;
 	};
 	doms {
 		#address-cells = <1>;
@@ -55,7 +55,7 @@
 			#size-cells = <1>;
 
 			app-bootcmd = "ampload";
-			app-loadaddr = "0x60000000";
+			app-loadaddr = "0x74000000";
 
 			kernel {
 				/* Kernel Memory allocated from xen domain-heap */
@@ -92,7 +92,7 @@
 			#size-cells = <1>;
 
 			app-bootcmd = "smpload";
-			app-loadaddr = "0x60000000";
+			app-loadaddr = "0x74000000";
 
 			kernel {
 				/* Kernel Memory allocated from xen domain-heap */
-- 
1.7.1

