From 2b81342868d60390c65a067b9735b9331e1931fa Mon Sep 17 00:00:00 2001
From: Siva Pochiraju <sivap@netlogicmicro.com>
Date: Fri, 1 Oct 2010 21:52:33 -0700
Subject: RX calender slot support for xaui & sgmii

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/boot/dts/nae.dts b/arch/mips/boot/dts/nae.dts
index 35b1304..cf49955 100644
--- a/arch/mips/boot/dts/nae.dts
+++ b/arch/mips/boot/dts/nae.dts
@@ -8,6 +8,7 @@
 	compatible = "NETL,XLP8XX_A0";
 	#address-cells = <1>;
 	#size-cells = <1>;
+	frequency = <250>;
 
 	ucore {
 		src@1 {
-- 
1.7.1

