From cfaf7e2a0cdb029bbd039440d02c3beded74bc8e Mon Sep 17 00:00:00 2001
From: Siva Pochiraju <sivap@netlogicmicro.com>
Date: Wed, 6 Oct 2010 12:57:28 -0700
Subject: Added VFBID programming

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/boot/dts/nae.dts b/arch/mips/boot/dts/nae.dts
index 3e142d7..02f17c1 100644
--- a/arch/mips/boot/dts/nae.dts
+++ b/arch/mips/boot/dts/nae.dts
@@ -10,6 +10,11 @@
 	#size-cells = <1>;
 	frequency = <250>;
 
+	// VFBID MAP: Upto 127 entries
+	// (each entry is a pair of (vfbid , dest-vc)
+	// Legal range: (vfbid (0 - 126), dest-vc (0 - 4095))
+	vfbid-map = <0 0 1 1 2 2 3 3 4 4
+		     5 5 6 6 7 7 8 8 9 9 >;
 	ucore {
 		src@1 {
 			file = "poe_bp_fwd_dest_portb.c";
-- 
1.7.1

