From 33ed3b4125732ba646d17fa5df99748fabfbe7d1 Mon Sep 17 00:00:00 2001
From: Siva Pochiraju <sivap@netlogicmicro.com>
Date: Wed, 6 Oct 2010 16:42:35 -0700
Subject: NAE: Added POE distribution vectors

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/boot/dts/nae.dts b/arch/mips/boot/dts/nae.dts
index 02f17c1..ce08d17 100644
--- a/arch/mips/boot/dts/nae.dts
+++ b/arch/mips/boot/dts/nae.dts
@@ -22,6 +22,33 @@
 		};
 	};
 
+	// Packet Ordering Engine (POE)
+	poe {	
+	
+		// Each vector is 512 bit with masb indicating vc 512 and lsb indicating vc 0
+		// Format : 512-bit-vector is specified as 16 32-bit words
+		// Left most word has the vc range 511-479 right most word has vc range 31 - 0
+		// Each word has the MSB select higer vc number and LSB select lower vc num
+		distribution_vectors {
+				     dv0  = <0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0x1>;
+				     dv1  = <0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0>;
+				     dv2  = <0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0>;
+				     dv3  = <0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0>;
+				     dv4  = <0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0>;
+				     dv5  = <0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0>;
+				     dv6  = <0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0>;
+				     dv7  = <0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0>;
+				     dv8  = <0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0>;
+				     dv9  = <0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0>;
+				     dv10 = <0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0>;
+				     dv11 = <0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0>;
+				     dv12 = <0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0>;
+				     dv13 = <0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0>;
+				     dv14 = <0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0>;
+				     dv15 = <0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0>;
+		};
+	};
+
 	complex@0 {
 		device_type = "nae-complex";
 		mode = "sgmii";
-- 
1.7.1

