From 7606ed3507e54388853e872cc8248ee7f13c91d0 Mon Sep 17 00:00:00 2001
From: Siva Pochiraju <sivap@netlogicmicro.com>
Date: Thu, 7 Oct 2010 16:19:57 -0700
Subject: NAE: Flow base config and Flow CRC poly config

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/boot/dts/nae.dts b/arch/mips/boot/dts/nae.dts
index 67ba532..f6bb412 100644
--- a/arch/mips/boot/dts/nae.dts
+++ b/arch/mips/boot/dts/nae.dts
@@ -15,6 +15,10 @@
 	// Legal range: (vfbid (0 - 126), dest-vc (0 - 4095))
 	vfbid-map = <0 0 1 1 2 2 3 3 4 4
 		     5 5 6 6 7 7 8 8 9 9 >;
+
+        // FLOW Table Generation: CRC16 polynomial
+  	flow_crc_poly = <0xffff>;
+
 	ucore {
 		src@1 {
 			file = "default.c";
-- 
1.7.1

