From ee6eaf02c9b429dcfc5355800cbefe468ec55a39 Mon Sep 17 00:00:00 2001
From: Krishnamurthy D V <kmurthy@netlogicmicro.com>
Date: Wed, 29 Sep 2010 17:25:29 +0530
Subject: Changes for running hyperexec standalone app on XLR/XLS

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/boot/dts/xlr.dts b/arch/mips/boot/dts/xlr.dts
index 70d4823..4b72df7 100644
--- a/arch/mips/boot/dts/xlr.dts
+++ b/arch/mips/boot/dts/xlr.dts
@@ -17,15 +17,18 @@
 			shared-mem = <0x0 0x80000000 0x0 0x20000000>;
 			#address-cells = <1>;
 			#size-cells = <1>;
+
+			app-bootcmd = "ampload";
+			app-loadaddr = "0x2000000";
+
 			cpu {
-				onlinemask = <0xffffffff>;
+				onlinemask = <0xff>;
 			};
 			uart {
 				id = <0>;
 			};
 			memory {
-				reg = <0x00100000 0x0FF00000	// 255M at 1M
-					   0x20000000 0x20000000>;  // 512M at 512M
+				reg = <0x20000000 0x40000000>; // 512M at 512M
 			};
 			pic {
 			};
@@ -39,8 +42,8 @@
 
 	system {
 		iobase = <0xbef00000>;
-		boardmajor = <2>;
-		boardminor = <1>;
+		boardmajor = <8>;
+		boardminor = <3>;
 		macaddress = [ 00 0f 30 00 09 6a ];
 		cpufrequency = <1190000000>;
 		globalshmem = <0x800ff000>;
-- 
1.7.1

