From ea23266815b63109f26826057b9d5a5825af136c Mon Sep 17 00:00:00 2001
From: Venu Vadapalli <vvadapalli@netlogicmicro.com>
Date: Tue, 7 Dec 2010 21:11:56 -0800
Subject: Use 1c4t hyperexec domain on linux

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/boot/dts/helinux.dts b/arch/mips/boot/dts/helinux.dts
index f8970b5..d92cb36 100644
--- a/arch/mips/boot/dts/helinux.dts
+++ b/arch/mips/boot/dts/helinux.dts
@@ -76,7 +76,7 @@
 			};
 
 			cpu {
-				onlinemask = <0xff>;
+				onlinemask = <0xf>;
                                 nae-rx-vc = <0>;
                                 nae-fb-vc = <3>;
 			};
-- 
1.7.1

