From 9019faae5c1e164fedcb3ec35ddcb92ad41d2c29 Mon Sep 17 00:00:00 2001
From: Siva Pochiraju <sivap@netlogicmicro.com>
Date: Thu, 9 Dec 2010 13:05:23 -0800
Subject: Changed Frequency to 500

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/boot/dts/nae.dts b/arch/mips/boot/dts/nae.dts
index 5261a8d..590581e 100644
--- a/arch/mips/boot/dts/nae.dts
+++ b/arch/mips/boot/dts/nae.dts
@@ -8,7 +8,7 @@
 	compatible = "NETL,XLP8XX_A0";
 	#address-cells = <1>;
 	#size-cells = <1>;
-	frequency = <250>;
+	frequency = <500>;
 
 	// VFBID MAP: Upto 127 entries
 	// (each entry is a pair of (vfbid , dest-vc)
-- 
1.7.1

