From 973266db8dffb6b8cbaf8f2af2a589f4af35387b Mon Sep 17 00:00:00 2001
From: Siva Pochiraju <sivap@netlogicmicro.com>
Date: Fri, 10 Dec 2010 12:24:48 -0800
Subject: Moved Free Fifo Carving to Global regs

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/boot/dts/nae.dts b/arch/mips/boot/dts/nae.dts
index 590581e..76e2428 100644
--- a/arch/mips/boot/dts/nae.dts
+++ b/arch/mips/boot/dts/nae.dts
@@ -72,7 +72,7 @@
 
 		// Max available descriptors are 1024  (across all complexes).
 		// Per port num_free_descriptors must be even number
-		num_free_descs = <512>;
+		num_free_descs = <350>;
 		free_desc_sizes = <2048>;
 		ucore_mask = <0xffff>;
 	};
@@ -95,7 +95,7 @@
 
 		// Max available descriptors are 1024  (across all complexes).
 		// Per port num_free_descriptors must be even number
-		num_free_descs = <256>;
+		num_free_descs = <350>;
 		free_desc_sizes = <2048>;
 		ucore_mask = <0xffff>;
 	};
@@ -120,7 +120,7 @@
 
 		// Max available descriptors are 1024 (across all complexes).
 		// Per port num_free_descriptors must be even number
-		num_free_descs = <128 128>;
+		num_free_descs = <66 66>;
 		free_desc_sizes = <2048 2048>;
 		ucore_mask = <0xffff 0xffff>;
 	};
-- 
1.7.1

