From eeaa662fa88aab18e52cefe6ca0cd3e4e5b03e53 Mon Sep 17 00:00:00 2001
From: Siva Pochiraju <sivap@netlogicmicro.com>
Date: Tue, 14 Dec 2010 18:02:02 -0800
Subject: Fixed iface fifo config

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/boot/dts/nae.dts b/arch/mips/boot/dts/nae.dts
index c8dde9d..e425036 100644
--- a/arch/mips/boot/dts/nae.dts
+++ b/arch/mips/boot/dts/nae.dts
@@ -14,7 +14,7 @@
 	// (each entry is a pair of (vfbid , dest-vc)
 	// Legal range: (vfbid (0 - 126), dest-vc (0 - 4095))
 	vfbid-map = <0 0 1 1 2 2 3 3 4 4
-		     5 5 6 6 7 7 8 8 9 9 >;
+		     5 5 6 6 7 7 8 8 9 1000 >;
 
         // FLOW Table Generation: CRC16 polynomial
   	flow_crc_poly = <0xffff>;
@@ -72,7 +72,7 @@
 
 		// Max available descriptors are 1024  (across all complexes).
 		// Per port num_free_descriptors must be even number
-		num_free_descs = <350>;
+		num_free_descs = <450>;
 		free_desc_sizes = <2048>;
 		ucore_mask = <0xffff>;
 	};
@@ -95,7 +95,7 @@
 
 		// Max available descriptors are 1024  (across all complexes).
 		// Per port num_free_descriptors must be even number
-		num_free_descs = <350>;
+		num_free_descs = <450>;
 		free_desc_sizes = <2048>;
 		ucore_mask = <0xffff>;
 	};
@@ -120,7 +120,7 @@
 
 		// Max available descriptors are 1024 (across all complexes).
 		// Per port num_free_descriptors must be even number
-		num_free_descs = <66 66>;
+		num_free_descs = <62 62>;
 		free_desc_sizes = <2048 2048>;
 		ucore_mask = <0xffff 0xffff>;
 	};
diff --git a/arch/mips/boot/dts/netos.dts b/arch/mips/boot/dts/netos.dts
index bd39dd6..4ad833c 100644
--- a/arch/mips/boot/dts/netos.dts
+++ b/arch/mips/boot/dts/netos.dts
@@ -47,7 +47,7 @@
 			cpu {
 				onlinemask = <0xf>;
                                 nae-rx-vc = <0>;
-                                nae-fb-vc = <1>;
+                                nae-fb-vc = <9>;
 			};
 			uart {
 				id = <0>;
-- 
1.7.1

