From 7b6270b89d1181205ac88fca85fbc5b5777dd968 Mon Sep 17 00:00:00 2001
From: Siva Pochiraju <sivap@netlogicmicro.com>
Date: Tue, 14 Dec 2010 18:43:12 -0800
Subject: Added xaui complexes 1 3

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/boot/dts/nae.dts b/arch/mips/boot/dts/nae.dts
index e425036..c9c58e1 100644
--- a/arch/mips/boot/dts/nae.dts
+++ b/arch/mips/boot/dts/nae.dts
@@ -59,20 +59,43 @@
 		loopback_mode = <0>;
 
 		// Max ingress fifo size 256 units (size of one unit is 64 byte)
-		iface_fifo_sizes = <110>;
+		iface_fifo_sizes = <55>;
 
 		// Max parser sequence fifo size 1024 packets 
 		// (if 1588 Timestamp is not required, then max size increases to 2048)
 		num_channels = <1>;
-		parser_sequence_fifo_sizes = <450>;
+		parser_sequence_fifo_sizes = <225>;
 
 		// Since rx_buffer is per context, we can have a few different ways
 		//  to populate the operand
-		rx_buffer_sizes = <1888>;
+		rx_buffer_sizes = <944>;
 
 		// Max available descriptors are 1024  (across all complexes).
 		// Per port num_free_descriptors must be even number
-		num_free_descs = <450>;
+		num_free_descs = <225>;
+		free_desc_sizes = <2048>;
+		ucore_mask = <0xffff>;
+	};
+	complex@1 {
+		device_type = "nae-complex";
+		mode = "xaui";
+		loopback_mode = <0>;
+
+		// Max ingress fifo size 256 units (size of one unit is 64 byte)
+		iface_fifo_sizes = <55>;
+
+		// Max parser sequence fifo size 1024 packets 
+		// (if 1588 Timestamp is not required, then max size increases to 2048)
+		num_channels = <1>;
+		parser_sequence_fifo_sizes = <225>;
+
+		// Since rx_buffer is per context, we can have a few different ways
+		//  to populate the operand
+		rx_buffer_sizes = <944>;
+
+		// Max available descriptors are 1024  (across all complexes).
+		// Per port num_free_descriptors must be even number
+		num_free_descs = <225>;
 		free_desc_sizes = <2048>;
 		ucore_mask = <0xffff>;
 	};
@@ -82,20 +105,43 @@
 		loopback_mode = <0>;
 
 		// Max ingress fifo size 256 units (size of one unit is 64 byte)
-		iface_fifo_sizes = <110>;
+		iface_fifo_sizes = <55>;
+
+		// Max parser sequence fifo size 1024 packets 
+		// (if 1588 Timestamp is not required, then max size increases to 2048)
+		num_channels = <1>;
+		parser_sequence_fifo_sizes = <225>;
+
+		// Since rx_buffer is per context, we can have a few different ways
+		//  to populate the operand
+		rx_buffer_sizes = <944>;
+
+		// Max available descriptors are 1024  (across all complexes).
+		// Per port num_free_descriptors must be even number
+		num_free_descs = <225>;
+		free_desc_sizes = <2048>;
+		ucore_mask = <0xffff>;
+	};
+	complex@3 {
+		device_type = "nae-complex";
+		mode = "xaui";
+		loopback_mode = <0>;
+
+		// Max ingress fifo size 256 units (size of one unit is 64 byte)
+		iface_fifo_sizes = <55>;
 
 		// Max parser sequence fifo size 1024 packets 
 		// (if 1588 Timestamp is not required, then max size increases to 2048)
 		num_channels = <1>;
-		parser_sequence_fifo_sizes = <450>;
+		parser_sequence_fifo_sizes = <225>;
 
 		// Since rx_buffer is per context, we can have a few different ways
 		//  to populate the operand
-		rx_buffer_sizes = <1888>;
+		rx_buffer_sizes = <944>;
 
 		// Max available descriptors are 1024  (across all complexes).
 		// Per port num_free_descriptors must be even number
-		num_free_descs = <450>;
+		num_free_descs = <225>;
 		free_desc_sizes = <2048>;
 		ucore_mask = <0xffff>;
 	};
-- 
1.7.1

