From 6aa6a4ef8b104596ea2398e15035c2454916e259 Mon Sep 17 00:00:00 2001
From: Siva Pochiraju <sivap@netlogicmicro.com>
Date: Wed, 15 Dec 2010 18:31:38 -0800
Subject: Changed VFBID map and free fifo carving

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/boot/dts/nae.dts b/arch/mips/boot/dts/nae.dts
index c9c58e1..01147c4 100644
--- a/arch/mips/boot/dts/nae.dts
+++ b/arch/mips/boot/dts/nae.dts
@@ -13,8 +13,23 @@
 	// VFBID MAP: Upto 127 entries
 	// (each entry is a pair of (vfbid , dest-vc)
 	// Legal range: (vfbid (0 - 126), dest-vc (0 - 4095))
-	vfbid-map = <0 0 1 1 2 2 3 3 4 4
-		     5 5 6 6 7 7 8 8 9 1000 >;
+	vfbid-map = <
+		 	/* NAE -> CPU mapping */ 
+		      0   0  1   4  2   8  3  12
+		      4  16  5  20  6  24  7  28
+		      8  32  9  36 10  40 11  44
+	 	     12  48 13  52 14  56 15  60
+		     16  64 17  68 18  72 19  76
+		     20  80 21  84 22  88 23  92
+		     24  96 25 100 26 104 27 108
+		     28 112 29 116 30 120 31 124
+			/* NAE -> NAE mapping */
+		     32	1000 33 1001 34 1002 35 1003
+		     36	1004 37 1005 38 1006 39 1007
+		     40	1008 41 1009 42 1010 43 1011
+		     44	1012 45 1013 46 1014 47 1015
+		     48	1016 49 1017 50 1018 51 1019
+		>;
 
         // FLOW Table Generation: CRC16 polynomial
   	flow_crc_poly = <0xffff>;
@@ -72,7 +87,7 @@
 
 		// Max available descriptors are 1024  (across all complexes).
 		// Per port num_free_descriptors must be even number
-		num_free_descs = <225>;
+		num_free_descs = <150>;
 		free_desc_sizes = <2048>;
 		ucore_mask = <0xffff>;
 	};
@@ -95,7 +110,7 @@
 
 		// Max available descriptors are 1024  (across all complexes).
 		// Per port num_free_descriptors must be even number
-		num_free_descs = <225>;
+		num_free_descs = <150>;
 		free_desc_sizes = <2048>;
 		ucore_mask = <0xffff>;
 	};
@@ -118,7 +133,7 @@
 
 		// Max available descriptors are 1024  (across all complexes).
 		// Per port num_free_descriptors must be even number
-		num_free_descs = <225>;
+		num_free_descs = <150>;
 		free_desc_sizes = <2048>;
 		ucore_mask = <0xffff>;
 	};
@@ -141,7 +156,7 @@
 
 		// Max available descriptors are 1024  (across all complexes).
 		// Per port num_free_descriptors must be even number
-		num_free_descs = <225>;
+		num_free_descs = <150>;
 		free_desc_sizes = <2048>;
 		ucore_mask = <0xffff>;
 	};
@@ -166,7 +181,7 @@
 
 		// Max available descriptors are 1024 (across all complexes).
 		// Per port num_free_descriptors must be even number
-		num_free_descs = <62 62>;
+		num_free_descs = <52 52>;
 		free_desc_sizes = <2048 2048>;
 		ucore_mask = <0xffff 0xffff>;
 	};
diff --git a/arch/mips/boot/dts/netos.dts b/arch/mips/boot/dts/netos.dts
index 4ad833c..2ef1c16 100644
--- a/arch/mips/boot/dts/netos.dts
+++ b/arch/mips/boot/dts/netos.dts
@@ -45,9 +45,9 @@
 			};
 
 			cpu {
-				onlinemask = <0xf>;
+				onlinemask = <0x3>;
                                 nae-rx-vc = <0>;
-                                nae-fb-vc = <9>;
+                                nae-fb-vc = <3>;
 			};
 			uart {
 				id = <0>;
-- 
1.7.1

