From d913bcdfcab672431b0f5ec3f6948055e0bf06cd Mon Sep 17 00:00:00 2001
From: Siva Pochiraju <sivap@netlogicmicro.com>
Date: Wed, 15 Dec 2010 23:36:49 -0800
Subject: Fixed VFBID mapping for NULL_FBID to detect NAE TX under-runs

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/boot/dts/nae.dts b/arch/mips/boot/dts/nae.dts
index 01147c4..7fe200e 100644
--- a/arch/mips/boot/dts/nae.dts
+++ b/arch/mips/boot/dts/nae.dts
@@ -29,6 +29,9 @@
 		     40	1008 41 1009 42 1010 43 1011
 		     44	1012 45 1013 46 1014 47 1015
 		     48	1016 49 1017 50 1018 51 1019
+
+		     /* NULL FBID Should map to cpu0 to detect NAE send message errors */
+		     127 0
 		>;
 
         // FLOW Table Generation: CRC16 polynomial
diff --git a/arch/mips/boot/dts/netos.dts b/arch/mips/boot/dts/netos.dts
index 2ef1c16..f704230 100644
--- a/arch/mips/boot/dts/netos.dts
+++ b/arch/mips/boot/dts/netos.dts
@@ -45,7 +45,7 @@
 			};
 
 			cpu {
-				onlinemask = <0x3>;
+				onlinemask = <0xf>;
                                 nae-rx-vc = <0>;
                                 nae-fb-vc = <3>;
 			};
-- 
1.7.1

