From 2ce7e7581006fbb6553ce9d3a7d1ed1ea34ac214 Mon Sep 17 00:00:00 2001
From: Siva Pochiraju <sivap@netlogicmicro.com>
Date: Thu, 16 Dec 2010 12:08:31 -0800
Subject: Chnaged NAE RX cal slot config from contiguous to interleaved

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/boot/dts/netos.dts b/arch/mips/boot/dts/netos.dts
index f704230..1149711 100644
--- a/arch/mips/boot/dts/netos.dts
+++ b/arch/mips/boot/dts/netos.dts
@@ -45,7 +45,7 @@
 			};
 
 			cpu {
-				onlinemask = <0xf>;
+				onlinemask = <0x1>;
                                 nae-rx-vc = <0>;
                                 nae-fb-vc = <3>;
 			};
-- 
1.7.1

