From 99d10ee609bbaec5de29017bf28bd3d674fd251f Mon Sep 17 00:00:00 2001
From: Siva Pochiraju <sivap@netlogicmicro.com>
Date: Mon, 20 Dec 2010 19:04:16 -0800
Subject: commented the vfbid mapping

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/boot/dts/nae.dts b/arch/mips/boot/dts/nae.dts
index 7fe200e..564e572 100644
--- a/arch/mips/boot/dts/nae.dts
+++ b/arch/mips/boot/dts/nae.dts
@@ -14,7 +14,9 @@
 	// (each entry is a pair of (vfbid , dest-vc)
 	// Legal range: (vfbid (0 - 126), dest-vc (0 - 4095))
 	vfbid-map = <
-		 	/* NAE -> CPU mapping */ 
+                     0 0 1 1 2 2 3 3 4 4 
+/*
+		     // NAE -> CPU mapping 
 		      0   0  1   4  2   8  3  12
 		      4  16  5  20  6  24  7  28
 		      8  32  9  36 10  40 11  44
@@ -23,15 +25,16 @@
 		     20  80 21  84 22  88 23  92
 		     24  96 25 100 26 104 27 108
 		     28 112 29 116 30 120 31 124
-			/* NAE -> NAE mapping */
+		     // NAE -> NAE mapping 
 		     32	1000 33 1001 34 1002 35 1003
 		     36	1004 37 1005 38 1006 39 1007
 		     40	1008 41 1009 42 1010 43 1011
 		     44	1012 45 1013 46 1014 47 1015
 		     48	1016 49 1017 50 1018 51 1019
 
-		     /* NULL FBID Should map to cpu0 to detect NAE send message errors */
+		     // NULL FBID Should map to cpu0 to detect NAE send message errors 
 		     127 0
+*/
 		>;
 
         // FLOW Table Generation: CRC16 polynomial
-- 
1.7.1

