From d7008612b7e2629fd144868bd749c3055d3e4356 Mon Sep 17 00:00:00 2001
From: Siva Pochiraju <sivap@netlogicmicro.com>
Date: Tue, 22 Feb 2011 16:21:35 -0800
Subject: Removed overlapping memory map

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/boot/dts/helinux.dts b/arch/mips/boot/dts/helinux.dts
index adc5589..fe0643a 100644
--- a/arch/mips/boot/dts/helinux.dts
+++ b/arch/mips/boot/dts/helinux.dts
@@ -100,7 +100,7 @@
 	};
 
 	chosen {
-		bootargs = " xlp_noi2c console=ttyS0,115200 rdinit=/sbin/init mem=512m@3584m ";
+		bootargs = " xlp_noi2c console=ttyS0,115200 rdinit=/sbin/init ";
 	};
 
 	/* These binaries are downloaded at the resp physical memory locations
-- 
1.7.1

