From 55c6e971edf11e1796218d673f0637b67be49987 Mon Sep 17 00:00:00 2001
From: Siva Pochiraju <sivap@netlogicmicro.com>
Date: Fri, 25 Feb 2011 01:23:52 -0800
Subject: Moved shared memeory for hyperxec in helinux.dts from 3.5G to 1G

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/boot/dts/helinux.dts b/arch/mips/boot/dts/helinux.dts
index 49c59a9..8b48f45 100644
--- a/arch/mips/boot/dts/helinux.dts
+++ b/arch/mips/boot/dts/helinux.dts
@@ -56,7 +56,7 @@
 			os = "netos";
 
 			heap-size = <0x0 0x2000000>; // 32M
-			shared-mem = <0x0 0xe0000000 0x0 0x10000000>; // 128M @ 3.5GB
+			shared-mem = <0x0 0x40000000 0x0 0x10000000>; // 128M @ 1GB
 
 			#address-cells = <1>;
 			#size-cells = <1>;
@@ -76,7 +76,7 @@
 			};
 
 			cpu {
-				onlinemask = <0xf>;
+				onlinemask = <0xff>;
                                 nae-rx-vc = <0>;
                                 nae-fb-vc = <3>;
                                 /* Setting this to '1' makes NAE driver to skip buf alloction and initial
@@ -87,7 +87,7 @@
 				id = <0>;
 			};
 			memory {
-				reg = <0x40000000 0x80000000>;  // 2048M @ 1024M
+				reg = <0x60000000 0x60000000>;  // 1.5GB @ 1.5GB
 			};
 
                         fmn {
-- 
1.7.1

