From c79bdb133abf79daf66b91246838c3e93959b077 Mon Sep 17 00:00:00 2001
From: Siva Pochiraju <sivap@netlogicmicro.com>
Date: Thu, 17 Mar 2011 20:13:52 -0700
Subject: Fixed typo

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/boot/dts/nae.dts b/arch/mips/boot/dts/nae.dts
index 70aba9d..45436a6 100644
--- a/arch/mips/boot/dts/nae.dts
+++ b/arch/mips/boot/dts/nae.dts
@@ -35,7 +35,7 @@
 
 		     // NULL FBID Should map to cpu0 to detect NAE send message errors 
 		     127 0
-/*
+*/
 		>;
 
         // FLOW Table Generation: CRC16 polynomial
-- 
1.7.1

