From 60b8e6d73970fa655166424c9d620e4576f1cac6 Mon Sep 17 00:00:00 2001
From: Mehul <vmehul@netlogicmicro.com>
Date: Wed, 23 Mar 2011 18:19:38 +0530
Subject: Added hybrid-nae.dts and multi-os-hybrid.dts for hybrid nae application.

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/boot/dts/hybrid-nae.dts b/arch/mips/boot/dts/hybrid-nae.dts
new file mode 100644
index 0000000..f0f18f1
--- /dev/null
+++ b/arch/mips/boot/dts/hybrid-nae.dts
@@ -0,0 +1,199 @@
+/* XLP8XX Device Tree Source 
+ *
+ */
+
+/dts-v1/;
+/ {
+	model = "MIPS,XLP8XX NAE CFG";
+	compatible = "NETL,XLP8XX_A0";
+	#address-cells = <1>;
+	#size-cells = <1>;
+	frequency = <500>;
+
+	// VFBID MAP: Upto 127 entries
+	// (each entry is a pair of (vfbid , dest-vc)
+	// Legal range: (vfbid (0 - 126), dest-vc (0 - 4095))
+	vfbid-map = <
+		     // NAE -> CPU mapping 
+		      0   1  1   5  2  9   3  13
+		      4  17  5  21  6  25  7  29
+		      8  33  9  37 10  41 11  45
+	 	     12  49 13  53 14  57 15  61
+		     16  65 17  69 18  73 19  77
+		     20  81 21  85 22  89 23  93
+		     24  97 25 101 26 105 27 109
+		     28 113 29 117 30 121 31 125
+		     // NAE -> NAE mapping 
+		     32	1000 33 1001 34 1002 35 1003
+		     36	1004 37 1005 38 1006 39 1007
+		     40	1008 41 1009 42 1010 43 1011
+		     44	1012 45 1013 46 1014 47 1015
+		     48	1016 49 1017 50 1018 51 1019
+
+		     // NULL FBID Should map to cpu0 to detect NAE send message errors 
+		     127 0
+		>;
+
+        // FLOW Table Generation: CRC16 polynomial
+  	flow_crc_poly = <0xffff>;
+
+	ucore {
+		src@1 {
+			file = "default.c";
+			mask = <0xffff>;
+		};
+	};
+
+	// Packet Ordering Engine (POE)
+	poe {	
+	
+		// Each vector is 512 bit with masb indicating vc 512 and lsb indicating vc 0
+		// Format : 512-bit-vector is specified as 16 32-bit words
+		// Left most word has the vc range 511-479 right most word has vc range 31 - 0
+		// Each word has the MSB select higer vc number and LSB select lower vc num
+		distribution_vectors {
+				     dv0  = <0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0x1>;
+				     dv1  = <0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0>;
+				     dv2  = <0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0>;
+				     dv3  = <0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0>;
+				     dv4  = <0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0>;
+				     dv5  = <0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0>;
+				     dv6  = <0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0>;
+				     dv7  = <0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0>;
+				     dv8  = <0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0>;
+				     dv9  = <0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0>;
+				     dv10 = <0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0>;
+				     dv11 = <0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0>;
+				     dv12 = <0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0>;
+				     dv13 = <0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0>;
+				     dv14 = <0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0>;
+				     dv15 = <0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0>;
+		};
+	};
+
+	complex@0 {
+		device_type = "nae-complex";
+		mode = "xaui";
+		loopback_mode = <0>;
+
+		// Max ingress fifo size 256 units (size of one unit is 64 byte)
+		iface_fifo_sizes = <55>;
+
+		// Max parser sequence fifo size 1024 packets 
+		// (if 1588 Timestamp is not required, then max size increases to 2048)
+		parser_sequence_fifo_sizes = <225>;
+
+		// Since rx_buffer is per context, we can have a few different ways
+		//  to populate the operand
+//		num_channels = <8>;
+//		rx_buffer_sizes = <118 118 118 118 118 118 118 118>;
+		num_channels = <1>;
+		rx_buffer_sizes = <944>;
+
+		// Max available descriptors are 1024  (across all complexes).
+		// Per port num_free_descriptors must be even number
+		num_free_descs = <150>;
+		free_desc_sizes = <2048>;
+		ucore_mask = <0xffff>;
+	};
+	complex@1 {
+		device_type = "nae-complex";
+		mode = "xaui";
+		loopback_mode = <0>;
+
+		// Max ingress fifo size 256 units (size of one unit is 64 byte)
+		iface_fifo_sizes = <55>;
+
+		// Max parser sequence fifo size 1024 packets 
+		// (if 1588 Timestamp is not required, then max size increases to 2048)
+		parser_sequence_fifo_sizes = <225>;
+
+		// Since rx_buffer is per context, we can have a few different ways
+		//  to populate the operand
+//		num_channels = <8>;
+//		rx_buffer_sizes = <118 118 118 118 118 118 118 118>;
+		num_channels = <1>;
+		rx_buffer_sizes = <944>;
+
+		// Max available descriptors are 1024  (across all complexes).
+		// Per port num_free_descriptors must be even number
+		num_free_descs = <150>;
+		free_desc_sizes = <2048>;
+		ucore_mask = <0xffff>;
+	};
+	complex@2 {
+		device_type = "nae-complex";
+		mode = "xaui";
+		loopback_mode = <0>;
+
+		// Max ingress fifo size 256 units (size of one unit is 64 byte)
+		iface_fifo_sizes = <55>;
+
+		// Max parser sequence fifo size 1024 packets 
+		// (if 1588 Timestamp is not required, then max size increases to 2048)
+		parser_sequence_fifo_sizes = <225>;
+
+		// Since rx_buffer is per context, we can have a few different ways
+		//  to populate the operand
+//		num_channels = <8>;
+//		rx_buffer_sizes = <118 118 118 118 118 118 118 118>;
+		num_channels = <1>;
+		rx_buffer_sizes = <944>;
+
+		// Max available descriptors are 1024  (across all complexes).
+		// Per port num_free_descriptors must be even number
+		num_free_descs = <150>;
+		free_desc_sizes = <2048>;
+		ucore_mask = <0xffff>;
+	};
+	complex@3 {
+		device_type = "nae-complex";
+		mode = "xaui";
+		loopback_mode = <0>;
+
+		// Max ingress fifo size 256 units (size of one unit is 64 byte)
+		iface_fifo_sizes = <55>;
+
+		// Max parser sequence fifo size 1024 packets 
+		// (if 1588 Timestamp is not required, then max size increases to 2048)
+		parser_sequence_fifo_sizes = <225>;
+
+		// Since rx_buffer is per context, we can have a few different ways
+		//  to populate the operand
+//		num_channels = <8>;
+//		rx_buffer_sizes = <118 118 118 118 118 118 118 118>;
+		num_channels = <1>;
+		rx_buffer_sizes = <944>;
+
+		// Max available descriptors are 1024  (across all complexes).
+		// Per port num_free_descriptors must be even number
+		num_free_descs = <150>;
+		free_desc_sizes = <2048>;
+		ucore_mask = <0xffff>;
+	};
+	complex@4 {
+		device_type = "nae-complex";
+		mode = "sgmii";
+
+		loopback_mode = <0 0>;
+
+		// Max ingress fifo size 256 units (size of one unit is 64 byte)
+		iface_fifo_sizes = <13 13>;
+
+		// Max parser sequence fifo size 1024 packets 
+		// (if 1588 Timestamp is not required, then max size increases to 2048)
+		num_channels = <1 1>;
+		parser_sequence_fifo_sizes = <62 62>;
+
+		// Since rx_buffer is per context, we can have a few different ways
+		//  to populate the operand
+		rx_buffer_sizes = <160 160>;
+				   
+
+		// Max available descriptors are 1024 (across all complexes).
+		// Per port num_free_descriptors must be even number
+		num_free_descs = <52 52>;
+		free_desc_sizes = <2048 2048>;
+		ucore_mask = <0xffff 0xffff>;
+	};
+};
diff --git a/arch/mips/boot/dts/multi-os-hybrid.dts b/arch/mips/boot/dts/multi-os-hybrid.dts
new file mode 100644
index 0000000..9b0761f
--- /dev/null
+++ b/arch/mips/boot/dts/multi-os-hybrid.dts
@@ -0,0 +1,209 @@
+/* XLP8XX Device Tree Source
+ *
+ */
+
+/dts-v1/;
+/ {
+	model = "MIPS,XLP8XX";
+	compatible = "NETL,XLP8XX_A0";
+
+	#address-cells = <1>;
+	#size-cells = <1>;
+
+	hypervisor {
+		hypervisor-name = "Xen";
+		alloc_dom0_memory = <0>;
+		bootargs = "ncores=8 dom0_loadaddr=0xf8000000 dom0_size=0x1c000000 dom0_cpumask=0x0f -- ";
+		domain_heap = <0x30000000 0x20000000>; /* 0.75 GB -> 1.25 GB */
+	};
+	doms {
+		#address-cells = <1>;
+		#size-cells = <1>;
+		dom@0 {
+			device_type = "domain";
+			os = "linux";
+
+			#address-cells = <1>;
+			#size-cells = <1>;
+
+			cpu {
+				onlinemask = <0x0f>;
+                                nae-rx-vc = <0>;
+                                nae-fb-vc = <1>;
+			};
+			uart {
+				id = <0>;
+				owner = <1>;
+				sharedcfg = <0x1f000000>;
+			};
+			memory {
+				reg = <0x04000000 0x08000000	// 128M at 64M
+					0x20000000 0x20000000>;  // 512M at 512M
+			};
+			pic {
+			};
+			app-param{
+				shmem = <0x1 0x40000000 		// START ADDR <MSB LSB> 5GB
+					0x0 0x80000000>;  // SIZE <MSB LSB> 2GB
+			};
+		};
+		dom@1 {
+			device_type = "domain";
+			os = "netos";
+
+			heap-size = <0x0 0x2000000>;                   // 32M
+			shared-mem = <0x0 0xe0000000 0x0 0x04000000>;  // 3584MB -> 3648 MB (64MB)
+
+			#address-cells = <1>;
+			#size-cells = <1>;
+
+			app-bootcmd = "ampload";
+			app-loadaddr = "0xf4000000"; // 3.75 GB + 64 MB
+
+			kernel {
+				/* Kernel Memory allocated from xen domain-heap */
+				size = <0x02000000>;  // 32M
+
+				/* Location in Dom0 ramdisk */
+				location = "/tmp/xen/netos.elf";
+
+				/* Domain destroy currently doesn't work with hyperexec */
+				dom_destroy_hack = <1>;
+			};
+
+			cpu {
+				onlinemask = <0xf0>;
+                                nae-rx-vc = <0>;
+                                nae-fb-vc = <1>;
+			};
+			uart {
+				id = <0>;
+				sharedcfg = <0x1f000000>;
+			};
+			memory {
+				reg = <0x50000000 0x10000000>;  // 1280M -> 1536M (256M)
+			};
+		};
+		dom@2 {
+			device_type = "domain";
+			os = "netos";
+
+			heap-size = <0x0 0x2000000>;                   // 32M
+			shared-mem = <0x0 0xe4000000 0x0 0x04000000>;  // 3648MB -> 3712 (64MB)
+
+			#address-cells = <1>;
+			#size-cells = <1>;
+
+			app-bootcmd = "smpload";
+			app-loadaddr = "0xf4000000"; // 3.75 GB + 64 MB
+
+		       	kernel {
+				/* Kernel Memory allocated from xen domain-heap */
+				size = <0x02000000>;  // 32M
+
+				/* Location in Dom0 ramdisk */
+				location = "/tmp/xen/netos.elf";
+
+				/* Domain destroy currently doesn't work with hyperexec */
+				dom_destroy_hack = <0>;
+			};
+
+			cpu {
+				onlinemask = <0xfffffff0>;
+                                nae-rx-vc = <0>;
+                                nae-fb-vc = <1>;
+				/* Setting this to '1' makes NAE driver to skip buf alloction and initial
+                        	   sending of packet buffers*/
+				nae-buf-alloc-by-app = <1>; 
+			};
+			uart {
+				id = <0>;
+				sharedcfg = <0x1f000000>;
+			};
+			memory {
+				reg = <0x60000000 0x60000000>;  // 1536M -> 3072M (1536M)
+			};
+			nae{
+				owner = <0>;
+			};
+			app-param{
+				shmem = <0x1 0x40000000 	// START ADDR <MSB LSB> 5GB
+					0x0 0x80000000>;  // SIZE <MSB LSB> 2GB
+				linux-mem = <0x00000000 0x04000000	// START ADDR <MSB LSB> 
+					0x00000000 0x3C000000>;  // SIZE <MSB LSB>
+				/* CP -> control plane
+				   FP - Fast path
+				*/
+				nae-rx-vc = <0>;//VC for FP and CP for nae rx 
+				sae-rx-vc = <1>; //VC for FP and CP for sae rx
+				io-fb-vc = <2>;//IO freeeback VC for FP and CP
+				cp-fp-vc = <3>;//VC for receiving msgs from cp/fp
+				cp-to-fp-dst = <19>;//FP cpu 1 will process
+				fp-to-cp-dst = <3>;//CP cpu 0 will process
+			};
+		};
+		dom@3 {
+			device_type = "domain";
+			os = "netos";
+
+			heap-size = <0x0 0x2000000>;                   // 32M
+			shared-mem = <0x0 0xe8000000 0x0 0x04000000>;  // 3712MB -> 3776 MB (64MB)
+
+			#address-cells = <1>;
+			#size-cells = <1>;
+
+			app-bootcmd = "ampload";
+			app-loadaddr = "0xf4000000"; // 3.75 GB + 64 MB
+
+			kernel {
+				/* Kernel Memory allocated from xen domain-heap */
+				size = <0x02000000>;  // 32M
+
+				/* Location in Dom0 ramdisk */
+				location = "/tmp/xen/netos.elf";
+
+				/* Domain destroy currently doesn't work with hyperexec */
+				dom_destroy_hack = <0>;
+			};
+
+			cpu {
+				onlinemask = <0xfff00000>;
+                                nae-rx-vc = <0>;
+                                nae-fb-vc = <1>;
+			};
+			uart {
+				id = <0>;
+				sharedcfg = <0x1f000000>;
+			};
+			memory {
+				reg = <0x90000000 0x30000000>;  // 2304M -> 3072M (768MB)
+			};
+		};
+
+	};
+
+	chosen {
+		bootargs = " xlp_noi2c console=ttyS0,115200 rdinit=/sbin/init";
+	};
+
+	/* These binaries are downloaded at the resp physical memory locations
+	 * by their corresponding bootstrapping loaders. For example,
+	 * u-boot is loaded by x-load at 193M, sysconfig is loaded by u-boot
+	 * at 1M, hypervisor by u-boot hypervisor at 136M.
+	 * On simulator all these binaries are pre-loaded by using '-F' option
+	 * for convenience. On silicon, they will be loaded by their corresponding
+	 * temporary download locations the stage-1/stage-2 firmware
+	 */
+	firmware-download {
+		u-boot        = <0x0c100000 0x3f00000>;  /* 63M @ 193M */
+		sysconfig     = <0x00100000 0x00100000>; /* 1M @ 1M */
+		hypervisor    = <0x08800000 0x00800000>; /* 8M @ 136M */
+		dom0          = <0x09000000 0x03000000>; /* 48M @ 144M */
+		domU-ramdisk  = <0x60000000 0x60800000>; /* 8M @ 1536M */
+	};
+
+	soc {
+
+	};
+};
+
-- 
1.7.1

