From 5a185938f014ee6c2ec45559eb4ff0ff577f5cd6 Mon Sep 17 00:00:00 2001
From: Yonghong Song <ysong@netlogicmicro.com>
Date: Mon, 4 Apr 2011 09:57:04 -0700
Subject: Add nae:owner=0 to multi-os.dts and multi-os-hybrid.dts

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/boot/dts/multi-os-hybrid.dts b/arch/mips/boot/dts/multi-os-hybrid.dts
index 9b0761f..72df72e 100644
--- a/arch/mips/boot/dts/multi-os-hybrid.dts
+++ b/arch/mips/boot/dts/multi-os-hybrid.dts
@@ -83,6 +83,9 @@
 			memory {
 				reg = <0x50000000 0x10000000>;  // 1280M -> 1536M (256M)
 			};
+			nae{
+				owner = <0>;
+			};
 		};
 		dom@2 {
 			device_type = "domain";
@@ -178,6 +181,9 @@
 			memory {
 				reg = <0x90000000 0x30000000>;  // 2304M -> 3072M (768MB)
 			};
+			nae{
+				owner = <0>;
+			};
 		};
 
 	};
diff --git a/arch/mips/boot/dts/multi-os.dts b/arch/mips/boot/dts/multi-os.dts
index 16a8035..93c95c4 100644
--- a/arch/mips/boot/dts/multi-os.dts
+++ b/arch/mips/boot/dts/multi-os.dts
@@ -79,6 +79,9 @@
 			memory {
 				reg = <0x50000000 0x10000000>;  // 1280M -> 1536M (256M)
 			};
+			nae {
+				owner = <0>;
+			};
 		};
 		dom@2 {
 			device_type = "domain";
@@ -116,6 +119,9 @@
 			memory {
 				reg = <0x60000000 0x30000000>;  // 1536M -> 2304M (768M)
 			};
+			nae {
+				owner = <0>;
+			};
 		};
 		dom@3 {
 			device_type = "domain";
@@ -153,6 +159,9 @@
 			memory {
 				reg = <0x90000000 0x30000000>;  // 2304M -> 3072M (768MB)
 			};
+			nae {
+				owner = <0>;
+			};
 		};
 	};
 
-- 
1.7.1

