From 084f27b5e415350ef5c825e414df888490c5ad4d Mon Sep 17 00:00:00 2001
From: Siva Pochiraju <sivap@netlogicmicro.com>
Date: Thu, 7 Apr 2011 09:29:55 -0700
Subject: Fixed typo in multi-os.dts

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/boot/dts/multi-os.dts b/arch/mips/boot/dts/multi-os.dts
index a4dc9c7..8749f51 100644
--- a/arch/mips/boot/dts/multi-os.dts
+++ b/arch/mips/boot/dts/multi-os.dts
@@ -37,7 +37,7 @@
 				sharedcfg = <0x1f000000>;
 			};
 			memory {
-				reg = <0x00100000 0x0F000000	// 240M at 16M
+				reg = <0x01000000 0x0F000000	// 240M at 16M
 				       0x20000000 0x10000000>;  // 256M at 512M
 			};
 			pic {
-- 
1.7.1

