From 3d44838ed2046314393632812c54cf62b702e037 Mon Sep 17 00:00:00 2001
From: P. Sadik <psadik@netlogicmicro.com>
Date: Sat, 9 Apr 2011 12:50:18 -0700
Subject: xlr console's baud rate should be 38400.

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/boot/dts/xlr.dts b/arch/mips/boot/dts/xlr.dts
index 084beec..60a1f03 100644
--- a/arch/mips/boot/dts/xlr.dts
+++ b/arch/mips/boot/dts/xlr.dts
@@ -47,7 +47,7 @@
 	};
 
 	chosen {
-		bootargs = "console=ttyS0,115200 rdinit=/sbin/init mem=240m@16m";
+		bootargs = "console=ttyS0,38400 rdinit=/sbin/init mem=240m@16m";
 		bootenv = "board=arizona";
 	};
 
-- 
1.7.1

