From a2a9979df97f9a7ea9c68f8dd4f39c1990220a2b Mon Sep 17 00:00:00 2001
From: Mehul <vmehul@netlogicmicro.com>
Date: Thu, 21 Apr 2011 14:22:36 +0530
Subject: Define poe configuration mode.

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/boot/dts/perf-nae.dts b/arch/mips/boot/dts/perf-nae.dts
index b7d8df6..ebac977 100644
--- a/arch/mips/boot/dts/perf-nae.dts
+++ b/arch/mips/boot/dts/perf-nae.dts
@@ -46,6 +46,8 @@
 
 	// Packet Ordering Engine (POE)
 	poe {	
+		// Supported modes are 'enque', 'serial', 'bypass'
+		mode = "bypass";
 	
 		// Each vector is 512 bit with masb indicating vc 512 and lsb indicating vc 0
 		// Format : 512-bit-vector is specified as 16 32-bit words
-- 
1.7.1

