From 268de04bb12ca13f4167a2e0a976fe2975f1f10f Mon Sep 17 00:00:00 2001
From: jayanthi <jayanthia@rmicorp.com>
Date: Sat, 13 Aug 2011 11:04:14 +0530
Subject: Fix: Max calendar slots to support SGMII card

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/boot/dts/helinux.dts b/arch/mips/boot/dts/helinux.dts
index 9e920cc..f0c68b7 100644
--- a/arch/mips/boot/dts/helinux.dts
+++ b/arch/mips/boot/dts/helinux.dts
@@ -76,7 +76,7 @@
 			};
 
 			cpu {
-				onlinemask = <0xffffffff>;
+				onlinemask = <0xffff>;
                                 nae-rx-vc = <0>;
                                 nae-fb-vc = <3>;
                                 /* Setting this to '1' makes NAE driver to skip buf alloction and initial
@@ -90,7 +90,7 @@
 				reg = <0x60000000 0x60000000>;  // 1.5GB @ 1.5GB
 			};
 			app-param{
-				onlinemask = <0xffffffff>;
+				onlinemask = <0xffff>;
 			};
 
                         fmn {
-- 
1.7.1

