From bd07ceb2054a88a1568ec6d222a86a16618f53b1 Mon Sep 17 00:00:00 2001
From: Vikas Gupta <vikasg@netlogicmicro.com>
Date: Thu, 8 Sep 2011 15:57:25 +0530
Subject: Online mask changed according to xlp3XX.

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/boot/dts/README.storm b/arch/mips/boot/dts/README.storm
new file mode 100644
index 0000000..885ae88
--- /dev/null
+++ b/arch/mips/boot/dts/README.storm
@@ -0,0 +1,2 @@
+the default storm-*dts files support XLP316. 
+online mask need to be changed for XLP308 and XLP304 accordingly
diff --git a/arch/mips/boot/dts/storm-helinux.dts b/arch/mips/boot/dts/storm-helinux.dts
index 81e0e91..8cdca44 100644
--- a/arch/mips/boot/dts/storm-helinux.dts
+++ b/arch/mips/boot/dts/storm-helinux.dts
@@ -13,7 +13,7 @@
 	hypervisor {
 		hypervisor-name = "Xen";
 		alloc_dom0_memory = <0>;
-		bootargs = "ncores=8 dom0_loadaddr=0x72000000 dom0_size=0x1c000000 dom0_cpumask=0xffffffff -- ";
+		bootargs = "ncores=4 dom0_loadaddr=0x72000000 dom0_size=0x1c000000 dom0_cpumask=0xffff -- ";
 		domain_heap = <0x80000000 0x20000000>;
 	};
 	doms {
diff --git a/arch/mips/boot/dts/storm-linux.dts b/arch/mips/boot/dts/storm-linux.dts
index 2ac7aa4..ccf7101 100644
--- a/arch/mips/boot/dts/storm-linux.dts
+++ b/arch/mips/boot/dts/storm-linux.dts
@@ -13,7 +13,7 @@
 	hypervisor {
 		hypervisor-name = "Xen";
 		alloc_dom0_memory = <0>;
-		bootargs = "ncores=8 dom0_loadaddr=0x72000000 dom0_size=0x1c000000 dom0_cpumask=0xffffffff -- ";
+		bootargs = "ncores=4 dom0_loadaddr=0x72000000 dom0_size=0x1c000000 dom0_cpumask=0xffff -- ";
 		domain_heap = <0x80000000 0x20000000>;
 	};
 	doms {
diff --git a/arch/mips/boot/dts/storm-multi-os.dts b/arch/mips/boot/dts/storm-multi-os.dts
index d4a4233..7192be0 100644
--- a/arch/mips/boot/dts/storm-multi-os.dts
+++ b/arch/mips/boot/dts/storm-multi-os.dts
@@ -13,7 +13,7 @@
 	hypervisor {
 		hypervisor-name = "Xen";
 		alloc_dom0_memory = <0>;
-		bootargs = "ncores=8 dom0_loadaddr=0x72000000 dom0_size=0x1c000000 dom0_cpumask=0x0f -- ";
+		bootargs = "ncores=4 dom0_loadaddr=0x72000000 dom0_size=0x1c000000 dom0_cpumask=0x0f -- ";
 		domain_heap = <0x30000000 0x20000000>; /* 0.75 GB -> 1.25 GB */
 	};
 	doms {
@@ -108,7 +108,7 @@
 			};
 
 			cpu {
-				onlinemask = <0x000fff00>;
+				onlinemask = <0xf00>;
                                 nae-rx-vc = <0>;
                                 nae-fb-vc = <1>;
 			};
@@ -148,7 +148,7 @@
 			};
 
 			cpu {
-				onlinemask = <0xfff00000>;
+				onlinemask = <0xf000>;
                                 nae-rx-vc = <0>;
                                 nae-fb-vc = <1>;
 			};
-- 
1.7.1

