From fd6effe2f511570593391b572ea0fd82e3c544ad Mon Sep 17 00:00:00 2001
From: Rahul Jain <rajain@netlogicmicro.com>
Date: Wed, 21 Sep 2011 13:05:55 +0530
Subject: added dom2

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/boot/dts/helinux.dts b/arch/mips/boot/dts/helinux.dts
index f0c68b7..92d4c0f 100644
--- a/arch/mips/boot/dts/helinux.dts
+++ b/arch/mips/boot/dts/helinux.dts
@@ -27,7 +27,7 @@
 			#size-cells = <1>;
 
 			cpu {
-				onlinemask = <0xffff>;
+				onlinemask = <0xffffffff>;
                                 nae-rx-vc = <1>;
                                 nae-fb-vc = <2>;
 
@@ -92,6 +92,62 @@
 			app-param{
 				onlinemask = <0xffff>;
 			};
+			nae{
+				owner = <1>;
+			};
+
+                        fmn {
+                                node_0_vc_mask = <0x99999999 0x99999999 0x99999999 0x99999999>;
+                                node_1_vc_mask = <0x99999999 0x99999999 0x99999999 0x99999999>;
+                                node_2_vc_mask = <0x99999999 0x99999999 0x99999999 0x99999999>;
+                                node_3_vc_mask = <0x99999999 0x99999999 0x99999999 0x99999999>;
+                        };
+		};
+
+		dom@2 {
+			device_type = "domain";
+			os = "netos";
+
+			heap-size = <0x0 0x8000000>; // 128M
+			shared-mem = <0x0 0x48000000 0x0 0x10000000>; // 128M @ 1152Mb
+
+			#address-cells = <1>;
+			#size-cells = <1>;
+
+			app-bootcmd = "smpload";
+			app-loadaddr = "0xf0000000"; // 3.75GB
+
+			kernel {
+				/* Kernel Memory allocated from xen domain-heap */
+				size = <0x02000000>;  // 32M
+
+				/* Location in Dom0 ramdisk */
+				location = "/tmp/xen/netos.elf";
+
+				/* Domain destroy currently doesn't work with hyperexec */
+				dom_destroy_hack = <0>;
+			};
+
+			cpu {
+				onlinemask = <0xffff0000>;
+                                nae-rx-vc = <0>;
+                                nae-fb-vc = <3>;
+                                /* Setting this to '1' makes NAE driver to skip buf alloction and initial
+                                   sending of packet buffers*/
+                                nae-buf-alloc-by-app = <1>; 
+			};
+			uart {
+				id = <0>;
+			};
+			memory {
+				reg = <0xE0000000 0x60000000>;  // 1.5Gb @ 3584Mb
+			};
+			app-param{
+				onlinemask = <0xffff0000>;
+			};
+			nae{
+				owner = <0>;
+			};
 
                         fmn {
                                 node_0_vc_mask = <0x99999999 0x99999999 0x99999999 0x99999999>;
-- 
1.7.1

