From 20a7e6a680f89bf23929b1e582526b17e4df73fb Mon Sep 17 00:00:00 2001
From: Yonghong Song <ysong@netlogicmicro.com>
Date: Mon, 26 Sep 2011 23:18:10 -0700
Subject: Make linux-numa configure less aggressive assuming target memory configuration

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/boot/dts/linux-numa.dts b/arch/mips/boot/dts/linux-numa.dts
index 4723503..87574c2 100644
--- a/arch/mips/boot/dts/linux-numa.dts
+++ b/arch/mips/boot/dts/linux-numa.dts
@@ -41,8 +41,7 @@
 			//Make sure each node's memory is separated out 
 			memory {
 				reg = <0x01000000 0x0B000000	// 176M at 16M
-					0x20000000 0xa0000000  // 2586M at 512M
-				        0xe0000000 0x49000000>;
+					0x20000000 0xa0000000>; // 2586M at 512M
 			};
                         fmn {
                                 node_0_vc_mask = <0xffffffff 0xffffffff 0xffffffff 0xffffffff>;
@@ -56,8 +55,7 @@
 	};
 
 	chosen {
-		#bootargs = " xlp_noi2c console=ttyS0,115200 rdinit=/sbin/init  mem=4096m@4752m";
-		bootargs = " xlp_noi2c console=ttyS0,115200 rdinit=/sbin/init  mem=4g@4752m mem=4g@8848m";
+		bootargs = " xlp_noi2c console=ttyS0,115200 rdinit=/sbin/init mem=512m@3584m";
 	};
 
 	/* These binaries are downloaded at the resp physical memory locations
-- 
1.7.1

