From e2912bbcf8b400ecf8900472faf36c9c248da4e4 Mon Sep 17 00:00:00 2001
From: P. Sadik <psadik@netlogicmicro.com>
Date: Fri, 28 Oct 2011 22:36:00 +0530
Subject: Remove "rdinit=/sbin/init" from Kernel commandline. This is to use the /init script in the initramfs.

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/boot/dts/helinux.dts b/arch/mips/boot/dts/helinux.dts
index 30b2f53..e46080c 100644
--- a/arch/mips/boot/dts/helinux.dts
+++ b/arch/mips/boot/dts/helinux.dts
@@ -163,7 +163,7 @@
 	};
 
 	chosen {
-		bootargs = " xlp_noi2c console=ttyS0,115200 rdinit=/sbin/init ";
+		bootargs = " xlp_noi2c console=ttyS0,115200  ";
 	};
 
 	/* These binaries are downloaded at the resp physical memory locations
diff --git a/arch/mips/boot/dts/linux-loader.dts b/arch/mips/boot/dts/linux-loader.dts
index e6ff3ac..56c55ec 100644
--- a/arch/mips/boot/dts/linux-loader.dts
+++ b/arch/mips/boot/dts/linux-loader.dts
@@ -165,7 +165,7 @@
 	};
 
 	chosen {
-		bootargs = " xlp_noi2c console=ttyS0,115200 rdinit=/sbin/init";
+		bootargs = " xlp_noi2c console=ttyS0,115200 ";
 	};
 
 	/* These binaries are downloaded at the resp physical memory locations
diff --git a/arch/mips/boot/dts/linux-numa.dts b/arch/mips/boot/dts/linux-numa.dts
index 87574c2..f81e9a5 100644
--- a/arch/mips/boot/dts/linux-numa.dts
+++ b/arch/mips/boot/dts/linux-numa.dts
@@ -55,7 +55,7 @@
 	};
 
 	chosen {
-		bootargs = " xlp_noi2c console=ttyS0,115200 rdinit=/sbin/init mem=512m@3584m";
+		bootargs = " xlp_noi2c console=ttyS0,115200  mem=512m@3584m";
 	};
 
 	/* These binaries are downloaded at the resp physical memory locations
diff --git a/arch/mips/boot/dts/linux-perf.dts b/arch/mips/boot/dts/linux-perf.dts
index ef2993b..91dd8eb 100644
--- a/arch/mips/boot/dts/linux-perf.dts
+++ b/arch/mips/boot/dts/linux-perf.dts
@@ -52,7 +52,7 @@
 	};
 
 	chosen {
-		bootargs = " xlp_noi2c console=ttyS0,115200 rdinit=/sbin/init rhash_entries=1048576";
+		bootargs = " xlp_noi2c console=ttyS0,115200  rhash_entries=1048576";
 	};
 
 	/* These binaries are downloaded at the resp physical memory locations
diff --git a/arch/mips/boot/dts/linux.dts b/arch/mips/boot/dts/linux.dts
index e03fa54..457a158 100644
--- a/arch/mips/boot/dts/linux.dts
+++ b/arch/mips/boot/dts/linux.dts
@@ -54,7 +54,7 @@
 	};
 
 	chosen {
-		bootargs = " xlp_noi2c console=ttyS0,115200 rdinit=/sbin/init mem=512m@3584m ";
+		bootargs = " xlp_noi2c console=ttyS0,115200  mem=512m@3584m ";
 	};
 
 	/* These binaries are downloaded at the resp physical memory locations
diff --git a/arch/mips/boot/dts/multi-os-hybrid.dts b/arch/mips/boot/dts/multi-os-hybrid.dts
index 72df72e..271c0f8 100644
--- a/arch/mips/boot/dts/multi-os-hybrid.dts
+++ b/arch/mips/boot/dts/multi-os-hybrid.dts
@@ -189,7 +189,7 @@
 	};
 
 	chosen {
-		bootargs = " xlp_noi2c console=ttyS0,115200 rdinit=/sbin/init";
+		bootargs = " xlp_noi2c console=ttyS0,115200 ";
 	};
 
 	/* These binaries are downloaded at the resp physical memory locations
diff --git a/arch/mips/boot/dts/multi-os.dts b/arch/mips/boot/dts/multi-os.dts
index 35cd340..6031c3c 100644
--- a/arch/mips/boot/dts/multi-os.dts
+++ b/arch/mips/boot/dts/multi-os.dts
@@ -167,7 +167,7 @@
 	};
 
 	chosen {
-		bootargs = " xlp_noi2c console=ttyS0,115200 rdinit=/sbin/init";
+		bootargs = " xlp_noi2c console=ttyS0,115200 ";
 	};
 
 	/* These binaries are downloaded at the resp physical memory locations
diff --git a/arch/mips/boot/dts/netos-srtp.dts b/arch/mips/boot/dts/netos-srtp.dts
index 9f45794..18110f7 100644
--- a/arch/mips/boot/dts/netos-srtp.dts
+++ b/arch/mips/boot/dts/netos-srtp.dts
@@ -71,7 +71,7 @@
 	};
 
 	chosen {
-		bootargs = " xlp_noi2c console=ttyS0,115200 rdinit=/sbin/init";
+		bootargs = " xlp_noi2c console=ttyS0,115200 ";
 	};
 
 	soc {
diff --git a/arch/mips/boot/dts/netos.dts b/arch/mips/boot/dts/netos.dts
index 1419315..af5c4f3 100644
--- a/arch/mips/boot/dts/netos.dts
+++ b/arch/mips/boot/dts/netos.dts
@@ -73,7 +73,7 @@
 	};
 
 	chosen {
-		bootargs = " xlp_noi2c console=ttyS0,115200 rdinit=/sbin/init";
+		bootargs = " xlp_noi2c console=ttyS0,115200 ";
 	};
 
 	soc {
diff --git a/arch/mips/boot/dts/storm-helinux.dts b/arch/mips/boot/dts/storm-helinux.dts
index 8cdca44..45d6669 100644
--- a/arch/mips/boot/dts/storm-helinux.dts
+++ b/arch/mips/boot/dts/storm-helinux.dts
@@ -103,7 +103,7 @@
 	};
 
 	chosen {
-		bootargs = " xlp_noi2c console=ttyS0,115200 rdinit=/sbin/init ";
+		bootargs = " xlp_noi2c console=ttyS0,115200  ";
 	};
 
 	/* These binaries are downloaded at the resp physical memory locations
diff --git a/arch/mips/boot/dts/storm-linux-perf.dts b/arch/mips/boot/dts/storm-linux-perf.dts
index 3613ff2..c0c8477 100644
--- a/arch/mips/boot/dts/storm-linux-perf.dts
+++ b/arch/mips/boot/dts/storm-linux-perf.dts
@@ -52,7 +52,7 @@
 	};
 
 	chosen {
-		bootargs = " xlp_noi2c console=ttyS0,115200 rdinit=/sbin/init rhash_entries=1048576";
+		bootargs = " xlp_noi2c console=ttyS0,115200  rhash_entries=1048576";
 	};
 
 	/* These binaries are downloaded at the resp physical memory locations
diff --git a/arch/mips/boot/dts/storm-linux.dts b/arch/mips/boot/dts/storm-linux.dts
index ccf7101..2212ba2 100644
--- a/arch/mips/boot/dts/storm-linux.dts
+++ b/arch/mips/boot/dts/storm-linux.dts
@@ -52,7 +52,7 @@
 	};
 
 	chosen {
-		bootargs = " xlp_noi2c console=ttyS0,115200 rdinit=/sbin/init mem=512m@3584m ";
+		bootargs = " xlp_noi2c console=ttyS0,115200  mem=512m@3584m ";
 	};
 
 	/* These binaries are downloaded at the resp physical memory locations
diff --git a/arch/mips/boot/dts/storm-multi-os.dts b/arch/mips/boot/dts/storm-multi-os.dts
index 7192be0..717a804 100644
--- a/arch/mips/boot/dts/storm-multi-os.dts
+++ b/arch/mips/boot/dts/storm-multi-os.dts
@@ -166,7 +166,7 @@
 	};
 
 	chosen {
-		bootargs = " xlp_noi2c console=ttyS0,115200 rdinit=/sbin/init";
+		bootargs = " xlp_noi2c console=ttyS0,115200 ";
 	};
 
 	/* These binaries are downloaded at the resp physical memory locations
diff --git a/arch/mips/boot/dts/storm-netos.dts b/arch/mips/boot/dts/storm-netos.dts
index c0a5f95..8354196 100644
--- a/arch/mips/boot/dts/storm-netos.dts
+++ b/arch/mips/boot/dts/storm-netos.dts
@@ -71,7 +71,7 @@
 	};
 
 	chosen {
-		bootargs = " xlp_noi2c console=ttyS0,115200 rdinit=/sbin/init";
+		bootargs = " xlp_noi2c console=ttyS0,115200 ";
 	};
 
 	soc {
diff --git a/arch/mips/boot/dts/storm-tiny.dts b/arch/mips/boot/dts/storm-tiny.dts
index e9e6d26..27868b8 100644
--- a/arch/mips/boot/dts/storm-tiny.dts
+++ b/arch/mips/boot/dts/storm-tiny.dts
@@ -35,7 +35,7 @@
 	};
 
 	chosen {
-		bootargs = "xlp_noi2c console=ttyS0,115200 rdinit=/sbin/init";
+		bootargs = "xlp_noi2c console=ttyS0,115200 ";
 	};
 
 	soc {
diff --git a/arch/mips/boot/dts/tiny.dts b/arch/mips/boot/dts/tiny.dts
index fc5abad..75edc2c 100644
--- a/arch/mips/boot/dts/tiny.dts
+++ b/arch/mips/boot/dts/tiny.dts
@@ -35,7 +35,7 @@
 	};
 
 	chosen {
-		bootargs = "xlp_noi2c console=ttyS0,115200 rdinit=/sbin/init";
+		bootargs = "xlp_noi2c console=ttyS0,115200 ";
 	};
 
 	soc {
diff --git a/arch/mips/boot/dts/xlpev-linux.dts b/arch/mips/boot/dts/xlpev-linux.dts
index 9ee37ef..9252380 100644
--- a/arch/mips/boot/dts/xlpev-linux.dts
+++ b/arch/mips/boot/dts/xlpev-linux.dts
@@ -52,7 +52,7 @@
 	};
 
 	chosen {
-		bootargs = " xlp_noi2c console=ttyS0,115200 rdinit=/sbin/init mem=512m@3584m ";
+		bootargs = " xlp_noi2c console=ttyS0,115200  mem=512m@3584m ";
 	};
 
 	/* These binaries are downloaded at the resp physical memory locations
diff --git a/arch/mips/boot/dts/xlpev-tiny.dts b/arch/mips/boot/dts/xlpev-tiny.dts
index fc5abad..75edc2c 100644
--- a/arch/mips/boot/dts/xlpev-tiny.dts
+++ b/arch/mips/boot/dts/xlpev-tiny.dts
@@ -35,7 +35,7 @@
 	};
 
 	chosen {
-		bootargs = "xlp_noi2c console=ttyS0,115200 rdinit=/sbin/init";
+		bootargs = "xlp_noi2c console=ttyS0,115200 ";
 	};
 
 	soc {
diff --git a/arch/mips/boot/dts/xlppcie-linux.dts b/arch/mips/boot/dts/xlppcie-linux.dts
index aea883d..bfacb2f 100644
--- a/arch/mips/boot/dts/xlppcie-linux.dts
+++ b/arch/mips/boot/dts/xlppcie-linux.dts
@@ -52,7 +52,7 @@
 	};
 
 	chosen {
-		bootargs = " xlp_noi2c console=ttyS0,115200 rdinit=/sbin/init ";
+		bootargs = " xlp_noi2c console=ttyS0,115200  ";
 	};
 
 	/* These binaries are downloaded at the resp physical memory locations
diff --git a/arch/mips/boot/dts/xlppcie-tiny.dts b/arch/mips/boot/dts/xlppcie-tiny.dts
index fc5abad..75edc2c 100644
--- a/arch/mips/boot/dts/xlppcie-tiny.dts
+++ b/arch/mips/boot/dts/xlppcie-tiny.dts
@@ -35,7 +35,7 @@
 	};
 
 	chosen {
-		bootargs = "xlp_noi2c console=ttyS0,115200 rdinit=/sbin/init";
+		bootargs = "xlp_noi2c console=ttyS0,115200 ";
 	};
 
 	soc {
diff --git a/arch/mips/boot/dts/xlr.dts b/arch/mips/boot/dts/xlr.dts
index 60a1f03..cccfcce 100644
--- a/arch/mips/boot/dts/xlr.dts
+++ b/arch/mips/boot/dts/xlr.dts
@@ -47,7 +47,7 @@
 	};
 
 	chosen {
-		bootargs = "console=ttyS0,38400 rdinit=/sbin/init mem=240m@16m";
+		bootargs = "console=ttyS0,38400  mem=240m@16m";
 		bootenv = "board=arizona";
 	};
 
-- 
1.7.1

