From 744e84f01a6518abaa34d2509cd44af543eda327 Mon Sep 17 00:00:00 2001
From: Vikas Gupta <vikasg@netlogicmicro.com>
Date: Mon, 31 Oct 2011 16:24:57 +0530
Subject: 1) Using storm-nae.dts for netos/helinux. 2) Create a debug file that can be used to view the .dts files used for creating a particular dtb.

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/boot/dts/storm-nae.dts b/arch/mips/boot/dts/storm-nae.dts
index 48e9684..5291cf8 100644
--- a/arch/mips/boot/dts/storm-nae.dts
+++ b/arch/mips/boot/dts/storm-nae.dts
@@ -14,36 +14,27 @@
 	// (each entry is a pair of (vfbid , dest-vc)
 	// Legal range: (vfbid (0 - 126), dest-vc (0 - 4095))
 	vfbid-map = <
-                     0    0    1   1   2   2   3   3		
-/*
 		     // NAE -> CPU mapping 
 		     // Freeback VC = 0
-		      0   0  1   4  2   8  3  12
-		      4  16  5  20  6  24  7  28
-		      8  32  9  36 10  40 11  44
-	 	     12  48 13  52 14  56 15  60
-		     16  64 17  68 18  72 19  76
-		     20  80 21  84 22  88 23  92
-		     24  96 25 100 26 104 27 108
-		     28 112 29 116 30 120 31 124
+		      0   1  1   5  2   9  3  13
+		      4  17  5  21  6  25  7  29
+		      8  33  9  37 10  41 11  45
+	 	     12  49 13  53 14  57 15  61
 		     // NAE -> NAE mapping 
-		     32	1000 33 1001 34 1002 35 1003
-		     36	1004 37 1005 38 1006 39 1007
-		     40	1008 41 1009 42 1010 43 1011
-		     44	1012 45 1013 46 1014 47 1015
-		     48	1016 49 1017 50 1018 51 1019
+		     32	496 33 497 34 498 35 499
+		     36	500 37 501 38 502 39 503
 
 		     // NULL FBID Should map to cpu0 to detect NAE send message errors 
 		     127 0
 
-*/
+
 		>;
         // FLOW Table Generation: CRC16 polynomial
   	flow_crc_poly = <0xffff>;
 
 	ucore {
 		src@1 {
-			file = "default.c";
+			file = "storm-linux-default.c";
 			mask = <0xff>;
 		};
 	};
@@ -100,6 +91,7 @@
 	
 		//ext phy addresses	
                 ext_phy_addr = <5 6 7 4>;
+                //ext_phy_addr = <9 10 11 8>;
                 ext_phy_bus = <0 0 0 0>;
 
                 // Max available descriptors are 1024 (across all complexes).
@@ -108,5 +100,30 @@
                 free_desc_sizes = <2048 2048 2048 2048>;
                 ucore_mask = <0xff 0xff 0xff 0xff>;
         };
+        
+	complex@1 {
+                device_type = "nae-complex";
+                mode = "xaui";
+
+                loopback_mode = <0>;
+
+                // Max ingress fifo size 256 units (size of one unit is 64 byte)
+                iface_fifo_sizes = <52>;
+
+                // Max parser sequence fifo size 1024 packets
+                // (if 1588 Timestamp is not required, then max size increases to 2048)
+                num_channels = < 1>;
+                parser_sequence_fifo_sizes = <120>;
+                //parser_sequence_fifo_sizes = <62 >;
+
+                // Since rx_buffer is per context, we can have a few different ways
+                //  to populate the operand
+                rx_buffer_sizes = <944>;
 
+                // Max available descriptors are 1024 (across all complexes).
+                // Per port num_free_descriptors must be even number
+                num_free_descs = < 150 >;
+                free_desc_sizes = < 2048 >;
+                ucore_mask = <0xff>;
+        };
 };
-- 
1.7.1

