From b796b7647e718b7565c1dd55036c634d6c04556d Mon Sep 17 00:00:00 2001
From: Siva Pochiraju <sivap@netlogicmicro.com>
Date: Sun, 6 Nov 2011 23:46:03 -0800
Subject: reduced memory for dom0 in storm-helinux.dts for boards with max memory 4G

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/boot/dts/storm-helinux.dts b/arch/mips/boot/dts/storm-helinux.dts
index 45d6669..2f63fa4 100644
--- a/arch/mips/boot/dts/storm-helinux.dts
+++ b/arch/mips/boot/dts/storm-helinux.dts
@@ -39,7 +39,7 @@
 			};
 			memory {
 				reg = <0x01000000 0x0B000000	// 176M at 16M
-					0xE0000000 0xC0000000>;  // 3GB @ 3.5GB
+					0xE0000000 0x20000000>;  // 512M @ 3.5GB
 			};
                         fmn {
                                 node_0_vc_mask = <0x66666666 0x66666666 0x66666666 0x66666666>;
-- 
1.7.1

