From c37e422cfff984b8be359283c7f6ad425321b879 Mon Sep 17 00:00:00 2001
From: Siva Pochiraju <sivap@netlogicmicro.com>
Date: Wed, 23 Nov 2011 17:46:29 +0530
Subject: Changed NAE frequency for xlppcie-tiny-nae.dts

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/boot/dts/xlppcie-tiny-nae.dts b/arch/mips/boot/dts/xlppcie-tiny-nae.dts
index b96e193..c65a943 100644
--- a/arch/mips/boot/dts/xlppcie-tiny-nae.dts
+++ b/arch/mips/boot/dts/xlppcie-tiny-nae.dts
@@ -8,7 +8,7 @@
 	compatible = "NETL,XLP8XX_A0";
 	#address-cells = <1>;
 	#size-cells = <1>;
-	frequency = <500>;
+	frequency = <250>;
 
 	// VFBID MAP: Upto 127 entries
 	// (each entry is a pair of (vfbid , dest-vc)
-- 
1.7.1

