From 297213792eeb0b7017387ef6a9723a922c6b2a89 Mon Sep 17 00:00:00 2001
From: Mehul <vmehul@netlogicmicro.com>
Date: Mon, 28 Nov 2011 14:39:15 +0530
Subject: Added napi-vc-mask in all linux*.dts.

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/boot/dts/helinux.dts b/arch/mips/boot/dts/helinux.dts
index e46080c..1a8380e 100644
--- a/arch/mips/boot/dts/helinux.dts
+++ b/arch/mips/boot/dts/helinux.dts
@@ -32,6 +32,7 @@
                                 nae-fb-vc = <2>;
 				sae-rx-vc = <1>; /* sharing with nae */
 				sae-rx-sync-vc = <3>; /* should be exclusive */
+                                napi-vc-mask = <0x6>;
 
 			};
 			uart {
diff --git a/arch/mips/boot/dts/linux-loader.dts b/arch/mips/boot/dts/linux-loader.dts
index 56c55ec..f565d89 100644
--- a/arch/mips/boot/dts/linux-loader.dts
+++ b/arch/mips/boot/dts/linux-loader.dts
@@ -30,6 +30,7 @@
 				onlinemask = <0x0f>;
                                 nae-rx-vc = <0>;
                                 nae-fb-vc = <1>;
+				napi-vc-mask = <0x3>;
 			};
 			uart {
 				id = <0>;
diff --git a/arch/mips/boot/dts/linux-numa.dts b/arch/mips/boot/dts/linux-numa.dts
index f81e9a5..88d2825 100644
--- a/arch/mips/boot/dts/linux-numa.dts
+++ b/arch/mips/boot/dts/linux-numa.dts
@@ -32,6 +32,7 @@
 				onlinemask = <0xffffffff 0xffffffff 0xffffffff 0xffffffff>;
                                 nae-rx-vc = <0>;
                                 nae-fb-vc = <1>;
+                                napi-vc-mask = <0x3>;
 			};
 			uart {
 				id = <0>;
diff --git a/arch/mips/boot/dts/multi-os-hybrid.dts b/arch/mips/boot/dts/multi-os-hybrid.dts
index 271c0f8..74684de 100644
--- a/arch/mips/boot/dts/multi-os-hybrid.dts
+++ b/arch/mips/boot/dts/multi-os-hybrid.dts
@@ -30,6 +30,7 @@
 				onlinemask = <0x0f>;
                                 nae-rx-vc = <0>;
                                 nae-fb-vc = <1>;
+				napi-vc-mask = <0x3>;
 			};
 			uart {
 				id = <0>;
diff --git a/arch/mips/boot/dts/multi-os.dts b/arch/mips/boot/dts/multi-os.dts
index 6031c3c..f2078e2 100644
--- a/arch/mips/boot/dts/multi-os.dts
+++ b/arch/mips/boot/dts/multi-os.dts
@@ -30,6 +30,7 @@
 				onlinemask = <0x0f>;
                                 nae-rx-vc = <0>;
                                 nae-fb-vc = <1>;
+				napi-vc-mask = <0x3>;
 			};
 			uart {
 				id = <0>;
diff --git a/arch/mips/boot/dts/storm-helinux.dts b/arch/mips/boot/dts/storm-helinux.dts
index 2f63fa4..b4b8f46 100644
--- a/arch/mips/boot/dts/storm-helinux.dts
+++ b/arch/mips/boot/dts/storm-helinux.dts
@@ -30,6 +30,7 @@
 				onlinemask = <0xffff>;
                                 nae-rx-vc = <1>;
                                 nae-fb-vc = <2>;
+                                napi-vc-mask = <0x6>;
 
 			};
 			uart {
diff --git a/arch/mips/boot/dts/storm-multi-os.dts b/arch/mips/boot/dts/storm-multi-os.dts
index f701d58..7f82b41 100644
--- a/arch/mips/boot/dts/storm-multi-os.dts
+++ b/arch/mips/boot/dts/storm-multi-os.dts
@@ -30,6 +30,7 @@
 				onlinemask = <0x0f>;
                                 nae-rx-vc = <0>;
                                 nae-fb-vc = <1>;
+				napi-vc-mask = <0x3>;
 			};
 			uart {
 				id = <0>;
diff --git a/arch/mips/boot/dts/xlpev-linux.dts b/arch/mips/boot/dts/xlpev-linux.dts
index 9252380..84eb473 100644
--- a/arch/mips/boot/dts/xlpev-linux.dts
+++ b/arch/mips/boot/dts/xlpev-linux.dts
@@ -30,6 +30,7 @@
 				onlinemask = <0xffffffff>;
                                 nae-rx-vc = <0>;
                                 nae-fb-vc = <1>;
+				napi-vc-mask = <0x3>;
 			};
 			uart {
 				id = <0>;
diff --git a/arch/mips/boot/dts/xlppcie-linux.dts b/arch/mips/boot/dts/xlppcie-linux.dts
index bfacb2f..df96b7a 100644
--- a/arch/mips/boot/dts/xlppcie-linux.dts
+++ b/arch/mips/boot/dts/xlppcie-linux.dts
@@ -30,6 +30,8 @@
 				onlinemask = <0xffff>;
                                 nae-rx-vc = <0>;
                                 nae-fb-vc = <1>;
+				napi-vc-mask = <0x3>;
+
 			};
 			uart {
 				id = <0>;
diff --git a/arch/mips/boot/dts/xlppcie-multi-os.dts b/arch/mips/boot/dts/xlppcie-multi-os.dts
index f701d58..7f82b41 100644
--- a/arch/mips/boot/dts/xlppcie-multi-os.dts
+++ b/arch/mips/boot/dts/xlppcie-multi-os.dts
@@ -30,6 +30,7 @@
 				onlinemask = <0x0f>;
                                 nae-rx-vc = <0>;
                                 nae-fb-vc = <1>;
+				napi-vc-mask = <0x3>;
 			};
 			uart {
 				id = <0>;
-- 
1.7.1

