From 44000c5c031fa36a234d75a23e4cbc7a5f6bae60 Mon Sep 17 00:00:00 2001
From: Hareesh Ramachandran <hareesh@london.(none)>
Date: Mon, 12 Dec 2011 04:59:34 -0800
Subject: Updated the perf ucore and perf nae config changes

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/boot/dts/storm-linux-perf-nae.dts b/arch/mips/boot/dts/storm-linux-perf-nae.dts
index c16173f..25256fa 100644
--- a/arch/mips/boot/dts/storm-linux-perf-nae.dts
+++ b/arch/mips/boot/dts/storm-linux-perf-nae.dts
@@ -40,6 +40,23 @@
 		};
 	};
 
+	freein-fifo-config {
+		/* If shared is true, 
+			1. Ucore is going to use upto max queues(16 for XLP, 8 for storm) for buffer mgmt
+		   	2. Onnchip desc size per queue, same value will be configured for all the queues.
+		   If shared is false, 
+	 		1. Onchip desc size per queue will be configured using num_free_desc specified at the port level.
+		   If spill descs per queue is specified,  
+			1. spill will be enabled for all the queues(except mgmt) using the spill memory specified in the OS config file.
+		*/
+
+		freein-fifo-shared = <1>;
+		freein-fifo-onchip-num-descs = <56>;
+		freein-fifo-spill-num-descs = <1200>;
+	};
+
+
+
 	// Packet Ordering Engine (POE)
 	poe {	
 	
-- 
1.7.1

