From 3aa62ccc3ccb35f2a511c52f45b900bdb8cb6146 Mon Sep 17 00:00:00 2001
From: Yonghong Song <ysong@netlogicmicro.com>
Date: Mon, 19 Dec 2011 12:50:27 -0800
Subject: Change linux NUMA dts file default node to 2

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/boot/dts/linux-numa.dts b/arch/mips/boot/dts/linux-numa.dts
index 88d2825..6e48dd3 100644
--- a/arch/mips/boot/dts/linux-numa.dts
+++ b/arch/mips/boot/dts/linux-numa.dts
@@ -25,11 +25,11 @@
 
 			#address-cells = <1>;
 			#size-cells = <1>;
-			#nodes = <4>;
-			#cpumask-cells=<4>;
+			#nodes = <2>;
+			#cpumask-cells=<2>;
 
 			cpu {
-				onlinemask = <0xffffffff 0xffffffff 0xffffffff 0xffffffff>;
+				onlinemask = <0xffffffff 0xffffffff>;
                                 nae-rx-vc = <0>;
                                 nae-fb-vc = <1>;
                                 napi-vc-mask = <0x3>;
-- 
1.7.1

