From be84ced5abed6b28bccff41afc5337f3d5c31e31 Mon Sep 17 00:00:00 2001
From: divyams <divyams@netlogicmicro.com>
Date: Tue, 10 Jan 2012 17:57:55 +0530
Subject: Changes for poe_app performance improvement in bypass and enque mode

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/boot/dts/nae.dts b/arch/mips/boot/dts/nae.dts
index fad842f..4610034 100644
--- a/arch/mips/boot/dts/nae.dts
+++ b/arch/mips/boot/dts/nae.dts
@@ -62,6 +62,10 @@
 	poe {	
 	
 		// Supported modes are 'enque', 'serial', 'bypass'
+		// For enque mode support , change mode in 
+		// "sysconfig/nae/ucore/default.c" from 
+		// "FWD_DIST_VEC" to "FWD_ENQ_DIST_VEC"  
+		
 		mode = "bypass";
 
 		// Each vector is 512 bit with masb indicating vc 512 and lsb indicating vc 0
diff --git a/arch/mips/boot/dts/netos.dts b/arch/mips/boot/dts/netos.dts
index af5c4f3..3506e7a 100644
--- a/arch/mips/boot/dts/netos.dts
+++ b/arch/mips/boot/dts/netos.dts
@@ -61,6 +61,9 @@
 			memory {
 				reg = <0x20000000 0xa0000000>;  // 2560M @ 512M
 			};
+			app-param{
+                                onlinemask = <0x1>;
+                        };
 
 			/* In hyperexec-linux mode, vc_mask needs to be synchronized with linux fmn node! */
                         fmn {
-- 
1.7.1

