From 26abe311193612a3403dcf42c2d2fcc33bfae4bc Mon Sep 17 00:00:00 2001
From: Jayanthi A <jayanthia@netlogicmicro.com>
Date: Wed, 1 Feb 2012 16:59:49 +0530
Subject: Removed xlpev*.dts files

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/boot/dts/perf/nae.dts b/arch/mips/boot/dts/perf/nae.dts
deleted file mode 100644
index 1ef92d8..0000000
--- a/arch/mips/boot/dts/perf/nae.dts
+++ /dev/null
@@ -1,27 +0,0 @@
-                vfbid-map = <
-                          0  1    1 5     2 9     3 13
-                          4  17   5 21    6 25    7 29
-                          8  33   9 37    10 41   11 45
-                          12 49   13 53   14 57   15 61
-                          16 65   17 69   18 73   19 77
-                          20 81   21 85   22 89   23 93
-                          24 97   25 101  26 105  27 109
-                          28 113  29 117  30 121  31 125
-
-                             // NAE -> NAE mapping
-                          32 1000 33 1001 34 1002 35 1003
-                          36 1004 37 1005 38 1006 39 1007
-                          40 1008 41 1009 42 1010 43 1011
-                          44 1012 45 1013 46 1014 47 1015
-                          48 1016 49 1017 50 1018 51 1019
-
-                        // NULL FBID Should map to cpu0 to detect NAE send message errors
-                          127 0
-                >;
-                ucore {
-                        src@1 {
-                                file = "perf-ucore.c";
-                                mask = <0xffff>;
-                        };
-                };
-
diff --git a/arch/mips/boot/dts/xlp3xx-tiny.dts b/arch/mips/boot/dts/xlp3xx-tiny.dts
index 4afdf56..4eeee63 100644
--- a/arch/mips/boot/dts/xlp3xx-tiny.dts
+++ b/arch/mips/boot/dts/xlp3xx-tiny.dts
@@ -102,7 +102,7 @@
                         	// Max parser sequence fifo size 1024 packets
 	                        // (if 1588 Timestamp is not required, then max size increases to 2048)
         	                num-channels = <1 1 1 1>;
-                	        parser-sequence-fifo-size = <62 62 62 62>;
+                	        parser-sequence-fifo-size = <30 30 30 30>;
 
                         	// Since rx_buffer is per context, we can have a few different ways
 	                        //  to populate the operand
diff --git a/arch/mips/boot/dts/xlp4xx-helinux.dts b/arch/mips/boot/dts/xlp4xx-helinux.dts
index e775a5e..c4156a6 100644
--- a/arch/mips/boot/dts/xlp4xx-helinux.dts
+++ b/arch/mips/boot/dts/xlp4xx-helinux.dts
@@ -27,7 +27,7 @@
 			#size-cells = <1>;
 
 			cpu {
-				onlinemask = <0xffffffff>;
+				onlinemask = <0xffff>;
                                 nae-rx-vc = <1>;
                                 nae-fb-vc = <2>;
 				sae-rx-vc = <1>; /* sharing with nae */
@@ -41,8 +41,8 @@
 				sharedcfg = <0x1f000000>;
 			};
 			memory {
-				reg = <0x01000000 0x0B000000	// 176M at 16M
-					0xE0000000 0xC0000000>;  // 3GB @ 3.5GB
+                                reg = <0x01000000 0x0B000000    // 176M at 16M
+                                       0x20000000 0x20000000>;  // 512M at 512M
 			};
                         fmn {
                                 node_0_vc_mask = <0x66666666 0x66666666 0x66666666 0x66666666>;
@@ -59,7 +59,7 @@
 			os = "netos";
 
 			heap-size = <0x0 0x8000000>; // 128M
-			shared-mem = <0x0 0x40000000 0x0 0x10000000>; // 128M @ 1GB
+			shared-mem = <0x0 0x40000000 0x0 0x10000000>; // 256M @ 1GB
 
 			#address-cells = <1>;
 			#size-cells = <1>;
@@ -136,7 +136,7 @@
 			};
 
 			cpu {
-				onlinemask = <0xffff0000>;
+				onlinemask = <0xff>;
                                 nae-rx-vc = <0>;
                                 nae-fb-vc = <3>;
                                 /* Setting this to '1' makes NAE driver to skip buf alloction and initial
@@ -150,7 +150,7 @@
 				reg = <0xE0000000 0x60000000>;  // 1.5Gb @ 3584Mb
 			};
 			app-param{
-				onlinemask = <0xffff0000>;
+				onlinemask = <0xff>;
 			};
 			nae{
 				owner = <0>;
diff --git a/arch/mips/boot/dts/xlp4xx-netos.dts b/arch/mips/boot/dts/xlp4xx-netos.dts
index c03c173..2988a48 100644
--- a/arch/mips/boot/dts/xlp4xx-netos.dts
+++ b/arch/mips/boot/dts/xlp4xx-netos.dts
@@ -25,13 +25,13 @@
 			os = "netos";
 
 			heap-size = <0x0 0x2000000>; // 32M
-			shared-mem = <0x0 0xe0000000 0x0 0x10000000>; // 128M @ 3.5GB
+			shared-mem = <0x0 0x40000000 0x0 0x01000000>; //  1GB
 
 			#address-cells = <1>;
 			#size-cells = <1>;
 
 			app-bootcmd = "smpload";
-			app-loadaddr = "0xf0000000"; // 3.75GB
+			app-loadaddr = "0x4000000"; // 3.75GB
 
 			kernel {
 				/* Kernel Memory allocated from xen domain-heap */
@@ -59,7 +59,7 @@
 				id = <0>;
 			};
 			memory {
-				reg = <0x20000000 0xa0000000>;  // 2560M @ 512M
+				reg = <0x20000000 0x20000000>;  // 2560M @ 512M
 			};
 			app-param{
                                 onlinemask = <0xffff>;
diff --git a/arch/mips/boot/dts/xlp8xx-linux-loader.dts b/arch/mips/boot/dts/xlp8xx-linux-loader.dts
new file mode 100644
index 0000000..a98a611
--- /dev/null
+++ b/arch/mips/boot/dts/xlp8xx-linux-loader.dts
@@ -0,0 +1,202 @@
+/* XLP8XX Device Tree Source
+ *
+ */
+
+/dts-v1/;
+/ {
+	model = "MIPS,XLP8XX";
+	compatible = "NETL,XLP8XX_A0";
+
+	#address-cells = <1>;
+	#size-cells = <1>;
+
+	hypervisor {
+		hypervisor-name = "Xen";
+		alloc_dom0_memory = <0>;
+		bootargs = "ncores=8 dom0_loadaddr=0x72000000 dom0_size=0x1c000000 dom0_cpumask=0x0f -- ";
+		domain_heap = <0x30000000 0x20000000>; /* 0.75 GB -> 1.25 GB */
+	};
+	doms {
+		#address-cells = <1>;
+		#size-cells = <1>;
+		dom@0 {
+			device_type = "domain";
+			os = "linux";
+
+			#address-cells = <1>;
+			#size-cells = <1>;
+
+			cpu {
+				onlinemask = <0x0f>;
+                                nae-rx-vc = <0>;
+                                nae-fb-vc = <1>;
+				napi-vc-mask = <0x3>;
+			};
+			uart {
+				id = <0>;
+				owner = <1>;
+				sharedcfg = <0x1f000000>;
+			};
+			memory {
+				reg = <	0x01000000 0x0B000000	// 176M at 16M
+					0x20000000 0x30000000   // 768M at 512M
+					0x80000000 0x10000000>;	// 256M at 2G
+			};
+			linuxloader{
+				loadermask = <0xfffffff0>;
+				memory = <0x00000002 0x00000000
+					  0x00000000 0x80000000>; //2G @ 8G
+			};
+			pic {
+			};
+		};
+		dom@1 {
+			device_type = "domain";
+			os = "netos";
+
+			heap-size = <0x0 0x2000000>;                   // 32M
+			shared-mem = <0x0 0xe0000000 0x0 0x04000000>;  // 3584MB -> 3648 MB (64MB)
+
+			#address-cells = <1>;
+			#size-cells = <1>;
+
+			app-bootcmd = "ampload";
+			app-loadaddr = "0xf4000000"; // 3.75 GB + 64 MB
+
+			kernel {
+				/* Kernel Memory allocated from xen domain-heap */
+				size = <0x02000000>;  // 32M
+
+				/* Location in Dom0 ramdisk */
+				location = "/tmp/xen/netos.elf";
+
+				/* Domain destroy currently doesn't work with hyperexec */
+				dom_destroy_hack = <0>;
+			};
+
+			cpu {
+				onlinemask = <0xf0>;
+                                nae-rx-vc = <0>;
+                                nae-fb-vc = <1>;
+			};
+			uart {
+				id = <0>;
+				sharedcfg = <0x1f000000>;
+			};
+			memory {
+				reg = <0x50000000 0x10000000>;  // 1280M -> 1536M (256M)
+			};
+			nae{
+				owner = <0>;
+			};
+		};
+		dom@2 {
+			device_type = "domain";
+			os = "netos";
+
+			heap-size = <0x0 0x2000000>;                   // 32M
+			shared-mem = <0x0 0xe4000000 0x0 0x04000000>;  // 3648MB -> 3712 (64MB)
+
+			#address-cells = <1>;
+			#size-cells = <1>;
+
+			app-bootcmd = "smpload";
+			app-loadaddr = "0xf4000000"; // 3.75 GB + 64 MB
+
+			kernel {
+				/* Kernel Memory allocated from xen domain-heap */
+				size = <0x02000000>;  // 32M
+
+				/* Location in Dom0 ramdisk */
+				location = "/tmp/xen/netos.elf";
+
+				/* Domain destroy currently doesn't work with hyperexec */
+				dom_destroy_hack = <0>;
+			};
+
+			cpu {
+				onlinemask = <0xf00>;
+                                nae-rx-vc = <0>;
+                                nae-fb-vc = <1>;
+                                /* Setting this to '1' makes NAE driver to skip buf alloction and initial
+                                   sending of packet buffers*/
+                                nae-buf-alloc-by-app = <0>;
+			};
+			uart {
+				id = <0>;
+				sharedcfg = <0x1f000000>;
+			};
+			memory {
+				reg = <0x60000000 0x30000000>;  // 1536M -> 2304M (768M)
+			};
+			nae{
+				owner = <0>;
+			};
+		};
+		dom@3 {
+			device_type = "domain";
+			os = "netos";
+
+			heap-size = <0x0 0x2000000>;                   // 32M
+			shared-mem = <0x0 0xe8000000 0x0 0x04000000>;  // 3712MB -> 3776 MB (64MB)
+
+			#address-cells = <1>;
+			#size-cells = <1>;
+
+			app-bootcmd = "ampload";
+			app-loadaddr = "0xf4000000"; // 3.75 GB + 64 MB
+
+			kernel {
+				/* Kernel Memory allocated from xen domain-heap */
+				size = <0x02000000>;  // 32M
+
+				/* Location in Dom0 ramdisk */
+				location = "/tmp/xen/netos.elf";
+
+				/* Domain destroy currently doesn't work with hyperexec */
+				dom_destroy_hack = <0>;
+			};
+
+			cpu {
+				onlinemask = <0xfffff000>;
+                                nae-rx-vc = <0>;
+                                nae-fb-vc = <1>;
+			};
+			uart {
+				id = <0>;
+				sharedcfg = <0x1f000000>;
+			};
+			memory {
+				reg = <0x90000000 0x30000000>;  // 2304M -> 3072M (768MB)
+			};
+			nae{
+				owner = <0>;
+			};
+		};
+	};
+
+	chosen {
+		bootargs = " xlp_noi2c console=ttyS0,115200 ";
+	};
+
+	/* These binaries are downloaded at the resp physical memory locations
+	 * by their corresponding bootstrapping loaders. For example,
+	 * u-boot is loaded by x-load at 193M, sysconfig is loaded by u-boot
+	 * at 1M, hypervisor by u-boot hypervisor at 136M.
+	 * On simulator all these binaries are pre-loaded by using '-F' option
+	 * for convenience. On silicon, they will be loaded by their corresponding
+	 * temporary download locations the stage-1/stage-2 firmware
+	 */
+	firmware-download {
+		u-boot        = <0x0c100000 0x3f00000>;  /* 63M @ 193M */
+		sysconfig     = <0x00100000 0x00100000>; /* 1M @ 1M */
+		hypervisor    = <0x08800000 0x00800000>; /* 8M @ 136M */
+		dom0          = <0x09000000 0x03000000>; /* 48M @ 144M */
+		domU-ramdisk  = <0x60000000 0x60800000>; /* 8M @ 1536M */
+	};
+
+	soc {
+
+	};
+};
+
diff --git a/arch/mips/boot/dts/xlpev-helinux.dts b/arch/mips/boot/dts/xlpev-helinux.dts
deleted file mode 100644
index dbcf648..0000000
--- a/arch/mips/boot/dts/xlpev-helinux.dts
+++ /dev/null
@@ -1,233 +0,0 @@
-/* XLP8XX Device Tree Source
- *
- */
-
-/dts-v1/;
-/ {
-	model = "MIPS,XLP8XX";
-	compatible = "NETL,XLP8XX_A0";
-
-	#address-cells = <1>;
-	#size-cells = <1>;
-
-	hypervisor {
-		hypervisor-name = "Xen";
-		alloc_dom0_memory = <0>;
-		bootargs = "ncores=8 dom0_loadaddr=0x72000000 dom0_size=0x1c000000 dom0_cpumask=0xffffffff -- ";
-		domain_heap = <0x80000000 0x20000000>;
-	};
-	doms {
-		#address-cells = <1>;
-		#size-cells = <1>;
-		dom@0 {
-			device_type = "domain";
-			os = "linux";
-
-			#address-cells = <1>;
-			#size-cells = <1>;
-
-			cpu {
-				onlinemask = <0xffffffff>;
-                                nae-rx-vc = <1>;
-                                nae-fb-vc = <1>;
-				sae-rx-vc = <1>; /* sharing with nae */
-				sae-rx-sync-vc = <3>; /* should be exclusive */
-                                napi-vc-mask = <0x1>;
-
-			};
-			uart {
-				id = <0>;
-				owner = <1>;
-				sharedcfg = <0x1f000000>;
-			};
-			memory {
-				reg = <0x01000000 0x0B000000	// 176M at 16M
-					0xE0000000 0xC0000000>;  // 3GB @ 3.5GB
-			};
-                        fmn {
-                                node_0_vc_mask = <0x22222222 0x22222222 0x22222222 0x22222222>;
-                                node_1_vc_mask = <0x22222222 0x22222222 0x22222222 0x22222222>;
-                                node_2_vc_mask = <0x22222222 0x22222222 0x22222222 0x22222222>;
-                                node_3_vc_mask = <0x22222222 0x22222222 0x22222222 0x22222222>;
-                        };
-			pic {
-			};
-		};
-
-		dom@1 {
-			device_type = "domain";
-			os = "netos";
-
-			heap-size = <0x0 0x8000000>; // 128M
-			shared-mem = <0x0 0x40000000 0x0 0x10000000>; // 128M @ 1GB
-
-			#address-cells = <1>;
-			#size-cells = <1>;
-
-			app-bootcmd = "smpload";
-			app-loadaddr = "0xf0000000"; // 3.75GB
-
-			kernel {
-				/* Kernel Memory allocated from xen domain-heap */
-				size = <0x02000000>;  // 32M
-
-				/* Location in Dom0 ramdisk */
-				location = "/tmp/xen/netos.elf";
-
-				/* Domain destroy currently doesn't work with hyperexec */
-				dom_destroy_hack = <0>;
-			};
-
-			cpu {
-				onlinemask = <0xffffffff>;
-                                nae-rx-vc = <0>;
-                                nae-fb-vc = <3>;
-				sae-rx-vc = <0>; /* sharing with nae */
-				sae-rx-sync-vc = <0>;  /* sharing with nae */
-				/* helinux, as only 1 thread per cpu, it should poll for both nae and sae 
-				   responses. Also there is no other vc as linux is using other 2 vcs */
-                                /* Setting this to '1' makes NAE driver to skip buf alloction and initial
-                                   sending of packet buffers*/
-                                nae-buf-alloc-by-app = <1>; 
-			};
-			uart {
-				id = <0>;
-			};
-			memory {
-				reg = <0x60000000 0x60000000>;  // 1.5GB @ 1.5GB
-			};
-			app-param{
-				onlinemask = <0xffffffff>;
-			};
-			nae{
-				owner = <1>;
-			};
-
-                        fmn {
-                                node_0_vc_mask = <0x99999999 0x99999999 0x99999999 0x99999999>;
-                                node_1_vc_mask = <0x99999999 0x99999999 0x99999999 0x99999999>;
-                                node_2_vc_mask = <0x99999999 0x99999999 0x99999999 0x99999999>;
-                                node_3_vc_mask = <0x99999999 0x99999999 0x99999999 0x99999999>;
-                        };
-            kbp {
-                /* distribute total of 4096 buffer id across cpus. required only for cpus sending requests.
-                 * note : KBP compare/search instructions must use even-numbered buffers. i.e. (buffer_id & 0x1 == 0) is True
-                 */
-                buffers-cpuid-offset-quantity = <  0    0   64
-                                                   1   64   64
-                                                   2  128   64
-                                                   3  192   64
-                                                   4  256   64
-                                                   5  320   64
-                                                   6  384   64
-                                                   7  448   64
-                                                   8  512   64
-                                                   9  576   64
-                                                  10  640   64
-                                                  11  704   64
-                                                  12  768   64
-                                                  13  832   64
-                                                  14  896   64
-                                                  15  960   64
-                                                  16 1024   64
-                                                  17 1088   64
-                                                  18 1152   64
-                                                  19 1216   64
-                                                  20 1280   64
-                                                  21 1344   64
-                                                  22 1408   64
-                                                  23 1472   64
-                                                  24 1536   64
-                                                  25 1600   64
-                                                  26 1664   64
-                                                  27 1728   64
-                                                  28 1792   64
-                                                  29 1856   64
-                                                  30 1920   64
-                                                  31 1984   64>;
-                /* starting address for KBP Driver Data. two 32-bit unsigned integers forms one 64-bit memory address.
-                 * kbp driver requires 2 MBytes of dedicated memory space to store data */
-                memory-offsethigh-offsetlow = <0x00000001 0xA0000000>;
-            };
-
-		};
-
-		dom@2 {
-			device_type = "domain";
-			os = "netos";
-
-			heap-size = <0x0 0x8000000>; // 128M
-			shared-mem = <0x0 0x48000000 0x0 0x10000000>; // 128M @ 1152Mb
-
-			#address-cells = <1>;
-			#size-cells = <1>;
-
-			app-bootcmd = "smpload";
-			app-loadaddr = "0xf0000000"; // 3.75GB
-
-			kernel {
-				/* Kernel Memory allocated from xen domain-heap */
-				size = <0x02000000>;  // 32M
-
-				/* Location in Dom0 ramdisk */
-				location = "/tmp/xen/netos.elf";
-
-				/* Domain destroy currently doesn't work with hyperexec */
-				dom_destroy_hack = <0>;
-			};
-
-			cpu {
-				onlinemask = <0xffff0000>;
-                                nae-rx-vc = <0>;
-                                nae-fb-vc = <3>;
-                                /* Setting this to '1' makes NAE driver to skip buf alloction and initial
-                                   sending of packet buffers*/
-                                nae-buf-alloc-by-app = <1>; 
-			};
-			uart {
-				id = <0>;
-			};
-			memory {
-				reg = <0xE0000000 0x60000000>;  // 1.5Gb @ 3584Mb
-			};
-			app-param{
-				onlinemask = <0xffff0000>;
-			};
-			nae{
-				owner = <0>;
-			};
-
-                        fmn {
-                                node_0_vc_mask = <0x99999999 0x99999999 0x99999999 0x99999999>;
-                                node_1_vc_mask = <0x99999999 0x99999999 0x99999999 0x99999999>;
-                                node_2_vc_mask = <0x99999999 0x99999999 0x99999999 0x99999999>;
-                                node_3_vc_mask = <0x99999999 0x99999999 0x99999999 0x99999999>;
-                        };
-		};
-	};
-
-	chosen {
-		bootargs = " xlp_noi2c console=ttyS0,115200  ";
-	};
-
-	/* These binaries are downloaded at the resp physical memory locations
-	 * by their corresponding bootstrapping loaders. For example,
-	 * u-boot is loaded by x-load at 193M, sysconfig is loaded by u-boot
-	 * at 1M, hypervisor by u-boot hypervisor at 136M.
-	 * On simulator all these binaries are pre-loaded by using '-F' option
-	 * for convenience. On silicon, they will be loaded by their corresponding
-	 * temporary download locations the stage-1/stage-2 firmware
-	 */
-	firmware-download {
-		u-boot        = <0x0c100000 0x3f00000>;  /* 63M @ 193M */
-		sysconfig     = <0x00100000 0x00100000>; /* 1M @ 1M */
-		hypervisor    = <0x08800000 0x00800000>; /* 8M @ 136M */
-		dom0          = <0x09000000 0x03000000>; /* 48M @ 144M */
-		domU-ramdisk  = <0x60000000 0x60800000>; /* 8M @ 1536M */
-	};
-
-	soc {
-
-	};
-};
-
diff --git a/arch/mips/boot/dts/xlpev-linux-loader.dts b/arch/mips/boot/dts/xlpev-linux-loader.dts
deleted file mode 100644
index a98a611..0000000
--- a/arch/mips/boot/dts/xlpev-linux-loader.dts
+++ /dev/null
@@ -1,202 +0,0 @@
-/* XLP8XX Device Tree Source
- *
- */
-
-/dts-v1/;
-/ {
-	model = "MIPS,XLP8XX";
-	compatible = "NETL,XLP8XX_A0";
-
-	#address-cells = <1>;
-	#size-cells = <1>;
-
-	hypervisor {
-		hypervisor-name = "Xen";
-		alloc_dom0_memory = <0>;
-		bootargs = "ncores=8 dom0_loadaddr=0x72000000 dom0_size=0x1c000000 dom0_cpumask=0x0f -- ";
-		domain_heap = <0x30000000 0x20000000>; /* 0.75 GB -> 1.25 GB */
-	};
-	doms {
-		#address-cells = <1>;
-		#size-cells = <1>;
-		dom@0 {
-			device_type = "domain";
-			os = "linux";
-
-			#address-cells = <1>;
-			#size-cells = <1>;
-
-			cpu {
-				onlinemask = <0x0f>;
-                                nae-rx-vc = <0>;
-                                nae-fb-vc = <1>;
-				napi-vc-mask = <0x3>;
-			};
-			uart {
-				id = <0>;
-				owner = <1>;
-				sharedcfg = <0x1f000000>;
-			};
-			memory {
-				reg = <	0x01000000 0x0B000000	// 176M at 16M
-					0x20000000 0x30000000   // 768M at 512M
-					0x80000000 0x10000000>;	// 256M at 2G
-			};
-			linuxloader{
-				loadermask = <0xfffffff0>;
-				memory = <0x00000002 0x00000000
-					  0x00000000 0x80000000>; //2G @ 8G
-			};
-			pic {
-			};
-		};
-		dom@1 {
-			device_type = "domain";
-			os = "netos";
-
-			heap-size = <0x0 0x2000000>;                   // 32M
-			shared-mem = <0x0 0xe0000000 0x0 0x04000000>;  // 3584MB -> 3648 MB (64MB)
-
-			#address-cells = <1>;
-			#size-cells = <1>;
-
-			app-bootcmd = "ampload";
-			app-loadaddr = "0xf4000000"; // 3.75 GB + 64 MB
-
-			kernel {
-				/* Kernel Memory allocated from xen domain-heap */
-				size = <0x02000000>;  // 32M
-
-				/* Location in Dom0 ramdisk */
-				location = "/tmp/xen/netos.elf";
-
-				/* Domain destroy currently doesn't work with hyperexec */
-				dom_destroy_hack = <0>;
-			};
-
-			cpu {
-				onlinemask = <0xf0>;
-                                nae-rx-vc = <0>;
-                                nae-fb-vc = <1>;
-			};
-			uart {
-				id = <0>;
-				sharedcfg = <0x1f000000>;
-			};
-			memory {
-				reg = <0x50000000 0x10000000>;  // 1280M -> 1536M (256M)
-			};
-			nae{
-				owner = <0>;
-			};
-		};
-		dom@2 {
-			device_type = "domain";
-			os = "netos";
-
-			heap-size = <0x0 0x2000000>;                   // 32M
-			shared-mem = <0x0 0xe4000000 0x0 0x04000000>;  // 3648MB -> 3712 (64MB)
-
-			#address-cells = <1>;
-			#size-cells = <1>;
-
-			app-bootcmd = "smpload";
-			app-loadaddr = "0xf4000000"; // 3.75 GB + 64 MB
-
-			kernel {
-				/* Kernel Memory allocated from xen domain-heap */
-				size = <0x02000000>;  // 32M
-
-				/* Location in Dom0 ramdisk */
-				location = "/tmp/xen/netos.elf";
-
-				/* Domain destroy currently doesn't work with hyperexec */
-				dom_destroy_hack = <0>;
-			};
-
-			cpu {
-				onlinemask = <0xf00>;
-                                nae-rx-vc = <0>;
-                                nae-fb-vc = <1>;
-                                /* Setting this to '1' makes NAE driver to skip buf alloction and initial
-                                   sending of packet buffers*/
-                                nae-buf-alloc-by-app = <0>;
-			};
-			uart {
-				id = <0>;
-				sharedcfg = <0x1f000000>;
-			};
-			memory {
-				reg = <0x60000000 0x30000000>;  // 1536M -> 2304M (768M)
-			};
-			nae{
-				owner = <0>;
-			};
-		};
-		dom@3 {
-			device_type = "domain";
-			os = "netos";
-
-			heap-size = <0x0 0x2000000>;                   // 32M
-			shared-mem = <0x0 0xe8000000 0x0 0x04000000>;  // 3712MB -> 3776 MB (64MB)
-
-			#address-cells = <1>;
-			#size-cells = <1>;
-
-			app-bootcmd = "ampload";
-			app-loadaddr = "0xf4000000"; // 3.75 GB + 64 MB
-
-			kernel {
-				/* Kernel Memory allocated from xen domain-heap */
-				size = <0x02000000>;  // 32M
-
-				/* Location in Dom0 ramdisk */
-				location = "/tmp/xen/netos.elf";
-
-				/* Domain destroy currently doesn't work with hyperexec */
-				dom_destroy_hack = <0>;
-			};
-
-			cpu {
-				onlinemask = <0xfffff000>;
-                                nae-rx-vc = <0>;
-                                nae-fb-vc = <1>;
-			};
-			uart {
-				id = <0>;
-				sharedcfg = <0x1f000000>;
-			};
-			memory {
-				reg = <0x90000000 0x30000000>;  // 2304M -> 3072M (768MB)
-			};
-			nae{
-				owner = <0>;
-			};
-		};
-	};
-
-	chosen {
-		bootargs = " xlp_noi2c console=ttyS0,115200 ";
-	};
-
-	/* These binaries are downloaded at the resp physical memory locations
-	 * by their corresponding bootstrapping loaders. For example,
-	 * u-boot is loaded by x-load at 193M, sysconfig is loaded by u-boot
-	 * at 1M, hypervisor by u-boot hypervisor at 136M.
-	 * On simulator all these binaries are pre-loaded by using '-F' option
-	 * for convenience. On silicon, they will be loaded by their corresponding
-	 * temporary download locations the stage-1/stage-2 firmware
-	 */
-	firmware-download {
-		u-boot        = <0x0c100000 0x3f00000>;  /* 63M @ 193M */
-		sysconfig     = <0x00100000 0x00100000>; /* 1M @ 1M */
-		hypervisor    = <0x08800000 0x00800000>; /* 8M @ 136M */
-		dom0          = <0x09000000 0x03000000>; /* 48M @ 144M */
-		domU-ramdisk  = <0x60000000 0x60800000>; /* 8M @ 1536M */
-	};
-
-	soc {
-
-	};
-};
-
diff --git a/arch/mips/boot/dts/xlpev-linux-perf.dts b/arch/mips/boot/dts/xlpev-linux-perf.dts
deleted file mode 100644
index 013867c..0000000
--- a/arch/mips/boot/dts/xlpev-linux-perf.dts
+++ /dev/null
@@ -1,105 +0,0 @@
-/* XLP8XX Device Tree Source
- *
- */
-
-/dts-v1/;
-/ {
-	model = "MIPS,XLP8XX";
-	compatible = "NETL,XLP8XX_A0";
-
-	#address-cells = <1>;
-	#size-cells = <1>;
-
-	hypervisor {
-		hypervisor-name = "Xen";
-		alloc_dom0_memory = <0>;
-		bootargs = "ncores=8 dom0_loadaddr=0x72000000 dom0_size=0x1c000000 dom0_cpumask=0xffffffff -- ";
-		domain_heap = <0x80000000 0x20000000>;
-	};
-	doms {
-		#address-cells = <1>;
-		#size-cells = <1>;
-		dom@0 {
-			device_type = "domain";
-			os = "linux";
-
-			#address-cells = <1>;
-			#size-cells = <1>;
-
-			cpu {
-				onlinemask = <0xffffffff>;
-                                nae-rx-vc = <0>;
-                                nae-fb-vc = <1>;
-			};
-			uart {
-				id = <0>;
-				owner = <1>;
-				sharedcfg = <0x1f000000>;
-			};
-			memory {
-				reg = <0x01000000 0x0B000000	// 176M at 16M
-					0x20000000 0x60000000>;  // 2586M at 512M
-			};
-                        fmn {
-                                node_0_vc_mask = <0x00000000 0x00000000 0x00000000 0x00000000>;
-                                node_1_vc_mask = <0xffffffff 0xffffffff 0xffffffff 0xffffffff>;
-                                node_2_vc_mask = <0xffffffff 0xffffffff 0xffffffff 0xffffffff>;
-                                node_3_vc_mask = <0xffffffff 0xffffffff 0xffffffff 0xffffffff>;
-                        };
-			pic {
-			};
-		};
-	};
-
-	fmn-config {
-		/* credit and queue size configuration, 
-		   A single entry message consumes 12 byte, 2 entry 20 byte etc. 
-		   so queue size of 16K can store 16 * 1024 /12 = 1365 single 
-		   entry messages 
-		   Stations which are not specified in the list will be configured
-		   with the default values given below 
-		   Names are cpu, popq, pcie, gdx, regx, rsa, crypto, srio, cmp, poe, nae 
-		*/
-
-		default-credits = <50>;
-		default-queue-size = <16384>;
-		cpu {
-			queue-size = <32768>; /* queue size for all the cpu vcs */
-			nae = <100>;	/* credits from cpu to nae */
-		};
-		poe {
-			cpu = <500>;
-		};
-	};
-
-	extra-mem-config {
-		/* used only if spill enabled */
-		freein-fifo-spill-mem-range = <0x00000000 0x0fe00000 0x00000000 0x00100000>; // 1MB @ 254MB
-		fmn-spill-mem-range = <0x00000000 0x10000000 0x00000000 0x02000000>; //32MB @ 256MB
-	};
-
-	chosen {
-		bootargs = " xlp_noi2c console=ttyS0,115200  rhash_entries=1048576";
-	};
-
-	/* These binaries are downloaded at the resp physical memory locations
-	 * by their corresponding bootstrapping loaders. For example,
-	 * u-boot is loaded by x-load at 193M, sysconfig is loaded by u-boot
-	 * at 1M, hypervisor by u-boot hypervisor at 136M.
-	 * On simulator all these binaries are pre-loaded by using '-F' option
-	 * for convenience. On silicon, they will be loaded by their corresponding
-	 * temporary download locations the stage-1/stage-2 firmware
-	 */
-	firmware-download {
-		u-boot        = <0x0c100000 0x3f00000>;  /* 63M @ 193M */
-		sysconfig     = <0x00100000 0x00100000>; /* 1M @ 1M */
-		hypervisor    = <0x08800000 0x00800000>; /* 8M @ 136M */
-		dom0          = <0x09000000 0x03000000>; /* 48M @ 144M */
-		domU-ramdisk  = <0x60000000 0x60800000>; /* 8M @ 1536M */
-	};
-
-	soc {
-
-	};
-};
-
diff --git a/arch/mips/boot/dts/xlpev-linux.dts b/arch/mips/boot/dts/xlpev-linux.dts
deleted file mode 100644
index 6834fde..0000000
--- a/arch/mips/boot/dts/xlpev-linux.dts
+++ /dev/null
@@ -1,83 +0,0 @@
-/* XLP8XX Device Tree Source
- *
- */
-
-/dts-v1/;
-/ {
-	model = "MIPS,XLP8XX";
-	compatible = "NETL,XLP8XX_A0";
-
-	#address-cells = <1>;
-	#size-cells = <1>;
-
-	hypervisor {
-		hypervisor-name = "Xen";
-		alloc_dom0_memory = <0>;
-		bootargs = "ncores=8 dom0_loadaddr=0x72000000 dom0_size=0x1c000000 dom0_cpumask=0xffffffff -- ";
-		domain_heap = <0x80000000 0x20000000>;
-	};
-	doms {
-		#address-cells = <1>;
-		#size-cells = <1>;
-		dom@0 {
-			device_type = "domain";
-			os = "linux";
-
-			#address-cells = <1>;
-			#size-cells = <1>;
-
-			cpu {
-				onlinemask = <0xffffffff>;
-                                nae-rx-vc = <0>;
-                                nae-fb-vc = <1>;
-				napi-vc-mask = <0x3>;
-				sae-rx-vc = <0>; /* sharing with nae */
-				sae-rx-sync-vc  = <3>; /* should be exclusive */
-				ipsec-async-vc  = <0>;
-				ipsec-sync-vc   = <2>;
-			};
-			uart {
-				id = <0>;
-				owner = <1>;
-				sharedcfg = <0x1f000000>;
-			};
-			memory {
-				reg = <0x01000000 0x0B000000	// 176M at 16M
-					0x20000000 0xa0000000>;  // 2586M at 512M
-			};
-                        fmn {
-                                node_0_vc_mask = <0x33333333 0x33333333 0x33333333 0x33333333>;
-                                node_1_vc_mask = <0x33333333 0x33333333 0x33333333 0x33333333>;
-                                node_2_vc_mask = <0x33333333 0x33333333 0x33333333 0x33333333>;
-                                node_3_vc_mask = <0x33333333 0x33333333 0x33333333 0x33333333>;
-                        };
-			pic {
-			};
-		};
-	};
-
-	chosen {
-		bootargs = " xlp_noi2c console=ttyS0,115200  mem=512m@3584m ";
-	};
-
-	/* These binaries are downloaded at the resp physical memory locations
-	 * by their corresponding bootstrapping loaders. For example,
-	 * u-boot is loaded by x-load at 193M, sysconfig is loaded by u-boot
-	 * at 1M, hypervisor by u-boot hypervisor at 136M.
-	 * On simulator all these binaries are pre-loaded by using '-F' option
-	 * for convenience. On silicon, they will be loaded by their corresponding
-	 * temporary download locations the stage-1/stage-2 firmware
-	 */
-	firmware-download {
-		u-boot        = <0x0c100000 0x3f00000>;  /* 63M @ 193M */
-		sysconfig     = <0x00100000 0x00100000>; /* 1M @ 1M */
-		hypervisor    = <0x08800000 0x00800000>; /* 8M @ 136M */
-		dom0          = <0x09000000 0x03000000>; /* 48M @ 144M */
-		domU-ramdisk  = <0x60000000 0x60800000>; /* 8M @ 1536M */
-	};
-
-	soc {
-
-	};
-};
-
diff --git a/arch/mips/boot/dts/xlpev-netos.dts b/arch/mips/boot/dts/xlpev-netos.dts
deleted file mode 100644
index 1f0d8c6..0000000
--- a/arch/mips/boot/dts/xlpev-netos.dts
+++ /dev/null
@@ -1,125 +0,0 @@
-/* XLP8XX Device Tree Source
- *
- */
-
-/dts-v1/;
-/ {
-	model = "MIPS,XLP8XX";
-	compatible = "NETL,XLP8XX_A0";
-
-	#address-cells = <1>;
-	#size-cells = <1>;
-
-	hypervisor {
-		hypervisor-name = "Xen";
-		alloc_dom0_memory = <1>;
-		bootargs = "dom0_loadaddr=0x72000000 dom0_cpumask=0x0f -- ";
-		domain_heap = <0x50000000 0x20000000>;
-	};
-
-	doms {
-		#address-cells = <1>;
-		#size-cells = <1>;
-		dom@0 {
-			device_type = "domain";
-			os = "netos";
-
-			heap-size = <0x0 0x2000000>; // 32M
-			shared-mem = <0x0 0xe0000000 0x0 0x10000000>; // 128M @ 3.5GB
-
-			#address-cells = <1>;
-			#size-cells = <1>;
-
-			app-bootcmd = "smpload";
-			app-loadaddr = "0xf0000000"; // 3.75GB
-
-			kernel {
-				/* Kernel Memory allocated from xen domain-heap */
-				size = <0x02000000>;  // 32M
-
-				/* Location in Dom0 ramdisk */
-				location = "/tmp/xen/netos.elf";
-
-				/* Domain destroy currently doesn't work with hyperexec */
-				dom_destroy_hack = <0>;
-			};
-
-			cpu {
-				onlinemask = <0xffffffff>;
-                                nae-rx-vc = <0>;
-                                nae-fb-vc = <3>;
-				sae-rx-vc = <0>; /* sharing with nae */
-				sae-rx-sync-vc = <2>; /* should be exclusive */
-
-                                /* Setting this to '1' makes NAE driver to skip buf alloction and initial
-                                   sending of packet buffers*/
-                                nae-buf-alloc-by-app = <1>; 
-			};
-			uart {
-				id = <0>;
-			};
-			memory {
-				reg = <0x20000000 0xa0000000>;  // 2560M @ 512M
-			};
-			app-param{
-                                onlinemask = <0xffffffff>;
-                        };
-
-			/* In hyperexec-linux mode, vc_mask needs to be synchronized with linux fmn node! */
-                        fmn {
-                                node_0_vc_mask = <0x99999999 0x99999999 0x99999999 0x99999999>;
-                                node_1_vc_mask = <0x99999999 0x99999999 0x99999999 0x99999999>;
-                                node_2_vc_mask = <0x99999999 0x99999999 0x99999999 0x99999999>;
-                                node_3_vc_mask = <0x99999999 0x99999999 0x99999999 0x99999999>;
-                        };
-            kbp {
-                /* distribute total of 4096 buffer id across cpus. required only for cpus sending requests.
-                 * note : KBP compare/search instructions must use even-numbered buffers. i.e. (buffer_id & 0x1 == 0) is True
-                 */
-                buffers-cpuid-offset-quantity = <  0    0   64
-                                                   1   64   64
-                                                   2  128   64
-                                                   3  192   64
-                                                   4  256   64
-                                                   5  320   64
-                                                   6  384   64
-                                                   7  448   64
-                                                   8  512   64
-                                                   9  576   64
-                                                  10  640   64
-                                                  11  704   64
-                                                  12  768   64
-                                                  13  832   64
-                                                  14  896   64
-                                                  15  960   64
-                                                  16 1024   64
-                                                  17 1088   64
-                                                  18 1152   64
-                                                  19 1216   64
-                                                  20 1280   64
-                                                  21 1344   64
-                                                  22 1408   64
-                                                  23 1472   64
-                                                  24 1536   64
-                                                  25 1600   64
-                                                  26 1664   64
-                                                  27 1728   64
-                                                  28 1792   64
-                                                  29 1856   64
-                                                  30 1920   64
-                                                  31 1984   64>;
-                /* starting address for KBP Driver Data. two 32-bit unsigned integers forms one 64-bit memory address.
-                 * kbp driver requires 2 MBytes of dedicated memory space to store data */
-                memory-offsethigh-offsetlow = <0x00000001 0x00000000>;
-            };
-		};
-	};
-
-	chosen {
-		bootargs = " xlp_noi2c console=ttyS0,115200 ";
-	};
-
-	soc {
-
-	};
-};
diff --git a/arch/mips/boot/dts/xlpev-tiny.dts b/arch/mips/boot/dts/xlpev-tiny.dts
deleted file mode 100644
index 4f6f128..0000000
--- a/arch/mips/boot/dts/xlpev-tiny.dts
+++ /dev/null
@@ -1,123 +0,0 @@
-/* XLP8XX Device Tree Source 
- *
- */
-
-/dts-v1/;
-/ {
-	model = "MIPS,XLP8XX";
-	compatible = "NETL,XLP8XX_A0";
-	#address-cells = <1>;
-	#size-cells = <1>;
-	doms {
-		#address-cells = <1>;
-		#size-cells = <1>;
-		dom@0 {
-			device_type = "domain";
-			heap-size = <0x0 0x4000000>;
-			shared-mem = <0x0 0x80000000 0x0 0x20000000>;
-			#address-cells = <1>;
-			#size-cells = <1>;
-			cpu {
-				onlinemask = <0xf>;
-				nae-rx-vc = <0>;
-				nae-fb-vc = <3>;
-			};
-			uart {
-				id = <0>;
-			};
-			memory {
-				reg = <0x00100000 0x0FF00000	// 255M at 1M
-					   0x20000000 0x20000000>;  // 512M at 512M
-			};
-			pic {
-			};
-		};
-	};
-
-	chosen {
-		bootargs = "xlp_noi2c console=ttyS0,115200 ";
-	};
-
-	soc {
-	    nae@node-0 {
-		model = "MIPS,XLP8XX NAE CFG";
-		compatible = "NETL,XLP8XX_A0";
-		#address-cells = <1>;
-		#size-cells = <1>;
-		frequency = <250>;
-
-		/include/ "ucore.dts"
-
-	        // VFBID MAP: Upto 127 entries
-        	// (each entry is a pair of (vfbid , dest-vc)
-	        // Legal range: (vfbid (0 - 126), dest-vc (0 - 4095))
-        	vfbid-config {
-                	vfbid-map = <
-	                     0    0    1   1   2   2   3   3
-        	        >;
-	        };
-
-        	// Packet Ordering Engine (POE)
-	        poe {
-	
-        	        mode = "bypass";
-
-                	// Each vector is 512 bit with masb indicating vc 512 and lsb indicating vc 0
-	                // Format : 512-bit-vector is specified as 16 32-bit words
-        	        // Left most word has the vc range 511-479 right most word has vc range 31 - 0
-                	// Each word has the MSB select higer vc number and LSB select lower vc num
-	                distribution_vectors {
-        	                     dv0  = <0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0x1>;
-                                     dv1  = <0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0>;
-                                     dv2  = <0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0>;
-                                     dv3  = <0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0>;
-                                     dv4  = <0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0>;
-                                     dv5  = <0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0>;
-                                     dv6  = <0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0>;
-                                     dv7  = <0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0>;
-                                     dv8  = <0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0>;
-                                     dv9  = <0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0>;
-                                     dv10 = <0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0>;
-                                     dv11 = <0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0>;
-                                     dv12 = <0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0>;
-                                     dv13 = <0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0>;
-                                     dv14 = <0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0>;
-                                     dv15 = <0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0>;
-                	};
-        	};
-
-	        complex@4 {
-        	        device_type = "nae-complex";
-                	mode = "sgmii";
-
-	                sgmii {
-        	                loopback = <0 0>;
-	
-                	        mgmt-port = <1 0>;
-
-                        	// Max ingress fifo size 256 units (size of one unit is 64 byte)
-	                        iface-fifo-size = <13 13>;
-
-        	                ext-phy-addr = <16 17>;
-                	        ext-phy-bus = <0 0>;
-
-                        	// Max parser sequence fifo size 1024 packets
-	                        // (if 1588 Timestamp is not required, then max size increases to 2048)
-        	                num-channels = <1 1>;
-                	        parser-sequence-fifo-size = <62 62>;
-
-                        	// Since rx_buffer is per context, we can have a few different ways
-	                        //  to populate the operand
-        	                rx-buffer-size = <128 128>;
-
-
-                	        // Max available descriptors are 1024 (across all complexes).
-                        	// Per port num_free_descriptors must be even number
-	                        num-free-descs = <52 52>;
-        	                free-desc-size = <2048 2048>;
-                	        ucore-mask = <0xffff 0xffff>;
-    	            };
-        	};
-	    };
-	};
-};
-- 
1.7.1

