From b991c975568a311b8926f62e797637b079ad9064 Mon Sep 17 00:00:00 2001
From: Yonghong Song <ysong@netlogicmicro.com>
Date: Fri, 3 Feb 2012 09:08:10 -0800
Subject: Fix the incorrect value for domain cpu onlinemask

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/boot/dts/xlp8xx-multi-os.dts b/arch/mips/boot/dts/xlp8xx-multi-os.dts
index 8efd054..da637c6 100644
--- a/arch/mips/boot/dts/xlp8xx-multi-os.dts
+++ b/arch/mips/boot/dts/xlp8xx-multi-os.dts
@@ -27,7 +27,7 @@
             #size-cells = <1>;
 
             cpu {
-                onlinemask = <0xFFFF0000>;
+                onlinemask = <0xf>;
                 nae-rx-vc = <0>;
                 nae-fb-vc = <1>;
                 napi-vc-mask = <0x3>;
@@ -76,7 +76,7 @@
             };
 
             cpu {
-                onlinemask = <0xFFFFFFFF>;
+                onlinemask = <0xf0>;
                 nae-rx-vc = <0>;
                 nae-fb-vc = <1>;
             };
-- 
1.7.1

