From d153117db811d4d0bef3164790d13a0b0caf0eab Mon Sep 17 00:00:00 2001
From: Yonghong Song <ysong@netlogicmicro.com>
Date: Fri, 3 Feb 2012 12:49:42 -0800
Subject: Generate multi-os config files

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/boot/dts/xlp8xx-multi-os.dts b/arch/mips/boot/dts/xlp8xx-multi-os.dts
index da637c6..f893664 100644
--- a/arch/mips/boot/dts/xlp8xx-multi-os.dts
+++ b/arch/mips/boot/dts/xlp8xx-multi-os.dts
@@ -13,7 +13,7 @@
     hypervisor {
         hypervisor-name = "Xen";
         alloc_dom0_memory = <0>;
-        bootargs = "ncores=8 dom0_loadaddr=0x72000000 dom0_size=0x1c000000 dom0_cpumask=0xFFFFFFFF -- ";
+        bootargs = "ncores=8 dom0_loadaddr=0x72000000 dom0_size=0x1c000000 dom0_cpumask=0xf -- ";
         domain_heap = <0x30000000 0x20000000>; /* 0.75 GB -> 1.25 GB */
     };
     doms {
-- 
1.7.1

