From df87d7be47ddae162715670d38820b27fc03347e Mon Sep 17 00:00:00 2001
From: Yonghong Song <ysong@broadcom.com>
Date: Mon, 12 Mar 2012 13:45:26 -0700
Subject: Correct hypervisor:domain_heap values for hybrid multi-os dts files

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/boot/dts/xlp4xx-multi-os-hybrid.dts b/arch/mips/boot/dts/xlp4xx-multi-os-hybrid.dts
index 888e530..27dc1d4 100644
--- a/arch/mips/boot/dts/xlp4xx-multi-os-hybrid.dts
+++ b/arch/mips/boot/dts/xlp4xx-multi-os-hybrid.dts
@@ -14,7 +14,7 @@
 		hypervisor-name = "Xen";
 		alloc_dom0_memory = <0>;
 		bootargs = "ncores=8 dom0_loadaddr=0x72000000 dom0_size=0x1c000000 dom0_cpumask=0x0f -- ";
-		domain_heap = <0x30000000 0x20000000>; /* 0.75 GB -> 1.25 GB */
+		domain_heap = <0x4e800000 0x10000000>; /* 1256MB -> 1512MB */
 	};
 	doms {
 		#address-cells = <1>;
diff --git a/arch/mips/boot/dts/xlp8xx-multi-os-hybrid.dts b/arch/mips/boot/dts/xlp8xx-multi-os-hybrid.dts
index 888e530..be7b37a 100644
--- a/arch/mips/boot/dts/xlp8xx-multi-os-hybrid.dts
+++ b/arch/mips/boot/dts/xlp8xx-multi-os-hybrid.dts
@@ -14,7 +14,7 @@
 		hypervisor-name = "Xen";
 		alloc_dom0_memory = <0>;
 		bootargs = "ncores=8 dom0_loadaddr=0x72000000 dom0_size=0x1c000000 dom0_cpumask=0x0f -- ";
-		domain_heap = <0x30000000 0x20000000>; /* 0.75 GB -> 1.25 GB */
+		domain_heap = <0xf0000000 0xc0000000>; /* 3584MB -> 6656MB */
 	};
 	doms {
 		#address-cells = <1>;
-- 
1.7.1

