From 17fb70bab745cdecafa2eadddc98731796da6a84 Mon Sep 17 00:00:00 2001
From: Yonghong Song <ysong@broadcom.com>
Date: Thu, 15 Mar 2012 10:48:31 -0700
Subject: Increase linux memory size (in multi-os environment for xlp3xx/xlp4xx)

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/boot/dts/xlp3xx-multi-os.dts b/arch/mips/boot/dts/xlp3xx-multi-os.dts
index 5d72558..cd79663 100644
--- a/arch/mips/boot/dts/xlp3xx-multi-os.dts
+++ b/arch/mips/boot/dts/xlp3xx-multi-os.dts
@@ -38,7 +38,8 @@
 				sharedcfg = <0x1f000000>;
 			};
 			memory {
-				reg = <0x01000000 0x0f000000>;	// 240M at 16M
+				reg = < 0x01000000 0x0F000000    // 240M at 16M
+					0x20000000 0x20000000>;  // 512M at 512M
 			};
 			pic {
 			};
diff --git a/arch/mips/boot/dts/xlp4xx-multi-os.dts b/arch/mips/boot/dts/xlp4xx-multi-os.dts
index 5d72558..c1a3fce 100644
--- a/arch/mips/boot/dts/xlp4xx-multi-os.dts
+++ b/arch/mips/boot/dts/xlp4xx-multi-os.dts
@@ -38,7 +38,8 @@
 				sharedcfg = <0x1f000000>;
 			};
 			memory {
-				reg = <0x01000000 0x0f000000>;	// 240M at 16M
+				reg = < 0x01000000 0x0f000000	// 240M at 16M
+					0x20000000 0x20000000>; // 512M at 512M
 			};
 			pic {
 			};
-- 
1.7.1

