From a0595c95c89b818e13f87be5ea25b8bceec0c548 Mon Sep 17 00:00:00 2001
From: Hareesh R <hareeshr@broadcom.com>
Date: Mon, 26 Mar 2012 12:28:41 +0530
Subject: New dts file has been added for Hybrid-NAE with Helinux mode

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/boot/dts/xlp4xx-multi-os-hybrid.dts b/arch/mips/boot/dts/xlp4xx-multi-os-hybrid.dts
deleted file mode 100644
index 27dc1d4..0000000
--- a/arch/mips/boot/dts/xlp4xx-multi-os-hybrid.dts
+++ /dev/null
@@ -1,128 +0,0 @@
-/* XLP8XX Device Tree Source
- *
- */
-
-/dts-v1/;
-/ {
-	model = "MIPS,XLP8XX";
-	compatible = "NETL,XLP8XX_A0";
-
-	#address-cells = <1>;
-	#size-cells = <1>;
-
-	hypervisor {
-		hypervisor-name = "Xen";
-		alloc_dom0_memory = <0>;
-		bootargs = "ncores=8 dom0_loadaddr=0x72000000 dom0_size=0x1c000000 dom0_cpumask=0x0f -- ";
-		domain_heap = <0x4e800000 0x10000000>; /* 1256MB -> 1512MB */
-	};
-	doms {
-		#address-cells = <1>;
-		#size-cells = <1>;
-		dom@0 {
-			device_type = "domain";
-			os = "linux";
-
-			#address-cells = <1>;
-			#size-cells = <1>;
-
-			cpu {
-				onlinemask = <0x0f>;
-				nae-rx-vc = <0>;
-				nae-fb-vc = <1>;
-				napi-vc-mask = <0x3>;
-			};
-			uart {
-				id = <0>;
-				owner = <1>;
-				sharedcfg = <0x1f000000>;
-			};
-			memory {
-				reg = <	0x04000000 0x08000000    // 128M at 64M
-					0x20000000 0x20000000>;  // 512M at 512M
-
-			};
-			pic {
-			};
-		};
-		dom@1 {
-			device_type = "domain";
-			os = "netos";
-
-			heap-size = <0x0 0x2000000>;		   // 32M
-			shared-mem = <0x0 0xe0000000 0x0 0x04000000>;  // 3584MB -> 3648 MB (64MB)
-
-			#address-cells = <1>;
-			#size-cells = <1>;
-
-			app-bootcmd = "smpload";
-			app-loadaddr = "0xf4000000"; // 3.75 GB + 64 MB
-
-			kernel {
-				/* Kernel Memory allocated from xen domain-heap */
-				size = <0x02000000>;  // 32M
-
-				/* Location in Dom0 ramdisk */
-				location = "/tmp/xen/netos.elf";
-
-				/* Domain destroy currently doesn't work with hyperexec */
-				dom_destroy_hack = <0>;
-			};
-
-			cpu {
-				onlinemask = <0xf0>;
-				nae-rx-vc = <0>;
-				nae-fb-vc = <1>;
-			};
-			uart {
-				id = <0>;
-				sharedcfg = <0x1f000000>;
-			};
-			memory {
-				reg = <0x50000000 0x10000000>;  // 1280M -> 1536M (256M)
-			};
-			nae {
-				owner = <0>;
-			};
-
-			app-param{
-				linux-mem = <	0x00000000 0x04000000	// START ADDR <MSB LSB> 64MB
-						0x00000000 0x3C000000>;	// SIZE <MSB LSB> 960 MB
-				/* CP -> control plane
-				   FP - Fast path
-				*/
-				nae-rx-vc = <0>;//VC for FP and CP for nae rx
-				sae-rx-vc = <1>; //VC for FP and CP for sae rx
-				io-fb-vc = <2>;//IO freeeback VC for FP and CP
-				cp-fp-vc = <3>;//VC for receiving msgs from cp/fp
-				cp-to-fp-dst = <19>;//FP cpu 1 will process
-				fp-to-cp-dst = <3>;//CP cpu 0 will process
-			};
-		};
-	};
-
-	chosen {
-		bootargs = " xlp_noi2c console=ttyS0,115200 ";
-	};
-
-	/* These binaries are downloaded at the resp physical memory locations
-	 * by their corresponding bootstrapping loaders. For example,
-	 * u-boot is loaded by x-load at 193M, sysconfig is loaded by u-boot
-	 * at 1M, hypervisor by u-boot hypervisor at 136M.
-	 * On simulator all these binaries are pre-loaded by using '-F' option
-	 * for convenience. On silicon, they will be loaded by their corresponding
-	 * temporary download locations the stage-1/stage-2 firmware
-	 */
-	firmware-download {
-		u-boot	= <0x0c100000 0x3f00000>;  /* 63M @ 193M */
-		sysconfig     = <0x00100000 0x00100000>; /* 1M @ 1M */
-		hypervisor    = <0x08800000 0x00800000>; /* 8M @ 136M */
-		dom0	  = <0x09000000 0x03000000>; /* 48M @ 144M */
-		domU-ramdisk  = <0x60000000 0x60800000>; /* 8M @ 1536M */
-	};
-
-	soc {
-
-	};
-};
-
diff --git a/arch/mips/boot/dts/xlp8xx-multi-os-hybrid-helinux.dts b/arch/mips/boot/dts/xlp8xx-multi-os-hybrid-helinux.dts
new file mode 100644
index 0000000..7f2def7
--- /dev/null
+++ b/arch/mips/boot/dts/xlp8xx-multi-os-hybrid-helinux.dts
@@ -0,0 +1,187 @@
+/* XLP8XX Device Tree Source
+ *
+ */
+
+/dts-v1/;
+/ {
+	model = "MIPS,XLP8XX";
+	compatible = "NETL,XLP8XX_A0";
+
+	#address-cells = <1>;
+	#size-cells = <1>;
+
+	hypervisor {
+		hypervisor-name = "Xen";
+		alloc_dom0_memory = <0>;
+		bootargs = "ncores=8 dom0_loadaddr=0x72000000 dom0_size=0x1c000000 dom0_cpumask=0x0f -- ";
+		domain_heap = <0xf0000000 0xc0000000>; /* 3584MB -> 6656MB */
+	};
+	doms {
+		#address-cells = <1>;
+		#size-cells = <1>;
+		dom@0 {
+			device_type = "domain";
+			os = "linux";
+
+			#address-cells = <1>;
+			#size-cells = <1>;
+
+			cpu {
+				onlinemask = <0xffffffff>;
+				nae-rx-vc = <0>;
+				nae-fb-vc = <1>;
+				napi-vc-mask = <0x3>;
+			};
+			uart {
+				id = <0>;
+				owner = <1>;
+				sharedcfg = <0x1f000000>;
+			};
+			memory {
+				reg = <0x01000000 0x0B000000
+					0x20000000 0x20000000>;  // 512M at 512M
+
+			};
+			pic {
+			};
+			nae {
+				owner = <0 0 0 0>;
+				freein-fifo-mask = <0x30000 0 0 0>;
+				sh-domains = <1>;
+				/* node offsets and size */
+				vfbtbl-sw-offset = <0 0 0 0>;
+				vfbtbl-sw-nentries = <32 32 32 32>;
+				vfbtbl-hw-offset = <0 0 0 0>;
+				vfbtbl-hw-nentries = <0 0 0 0>;
+			};
+			fmn {
+				node_0_vc_mask = <0x33333333 0x33333333 0x33333333 0x33333333>;
+				node_1_vc_mask = <0x33333333 0x33333333 0x33333333 0x33333333>;
+				node_2_vc_mask = <0x33333333 0x33333333 0x33333333 0x33333333>;
+				node_3_vc_mask = <0x33333333 0x33333333 0x33333333 0x33333333>;
+			};
+
+
+		};
+		dom@1 {
+			device_type = "domain";
+			os = "netos";
+
+			heap-size = <0x0 0x2000000>;		   // 32M
+			shared-mem = <0x0 0xe0000000 0x0 0x04000000>;  // 3584MB -> 3648 MB (64MB)
+
+			#address-cells = <1>;
+			#size-cells = <1>;
+
+			app-bootcmd = "smpload";
+			app-loadaddr = "0xf4000000"; // 3.75 GB + 64 MB
+
+			kernel {
+				/* Kernel Memory allocated from xen domain-heap */
+				size = <0x02000000>;  // 32M
+
+				/* Location in Dom0 ramdisk */
+				location = "/tmp/xen/netos.elf";
+
+				/* Domain destroy currently doesn't work with hyperexec */
+				dom_destroy_hack = <0>;
+			};
+
+			cpu {
+				onlinemask = <0xf0>;
+				nae-rx-vc = <2>;
+				nae-fb-vc = <3>;
+			};
+			uart {
+				id = <0>;
+				sharedcfg = <0x1f000000>;
+			};
+			memory {
+				reg = <0x50000000 0x10000000>;  // 1280M -> 1536M (256M)
+			};
+			nae {
+				owner = <1 0 0 0>;
+				freein-fifo-mask = <0x1111 0 0 0>;
+				sh-domains = <2>;
+				/* node offsets and size */
+				vfbtbl-sw-offset = <0 0 0 0>;
+				vfbtbl-sw-nentries = <0 0 0 0>;
+				vfbtbl-hw-offset = <32 32 32 32>;
+				vfbtbl-hw-nentries = <20 20 20 20>;
+			};
+		};
+		dom@2 {
+			device_type = "domain";
+			os = "netos";
+
+			heap-size = <0x0 0x2000000>;		   // 32M
+			shared-mem = <0x0 0xe4000000 0x0 0x04000000>;  // 3584MB -> 3648 MB (64MB)
+
+			#address-cells = <1>;
+			#size-cells = <1>;
+
+			app-bootcmd = "smpload";
+			app-loadaddr = "0xf4000000"; // 3.75 GB + 64 MB
+
+			kernel {
+				/* Kernel Memory allocated from xen domain-heap */
+				size = <0x02000000>;  // 32M
+
+				/* Location in Dom0 ramdisk */
+				location = "/tmp/xen/netos.elf";
+
+				/* Domain destroy currently doesn't work with hyperexec */
+				dom_destroy_hack = <0>;
+			};
+
+			cpu {
+				onlinemask = <0xf00>;
+				nae-rx-vc = <2>;
+				nae-fb-vc = <3>;
+			};
+			uart {
+				id = <0>;
+				sharedcfg = <0x1f000000>;
+			};
+			memory {
+				reg = <0x50000000 0x10000000>;  // 1280M -> 1536M (256M)
+			};
+			nae {
+				owner = <0 0 0 0>;
+				freein-fifo-mask = <0x2222 0 0 0>;
+				sh-domains = <1>;
+				/* node offsets and size */
+				vfbtbl-sw-offset = <0 0 0 0>;
+				vfbtbl-sw-nentries = <0 0 0 0>;
+				vfbtbl-hw-offset = <32 32 32 32>;
+				vfbtbl-hw-nentries = <20 20 20 20>;
+			};
+		};
+
+	};
+
+	chosen {
+		bootargs = " xlp_noi2c console=ttyS0,115200 ";
+	};
+
+	/* These binaries are downloaded at the resp physical memory locations
+	 * by their corresponding bootstrapping loaders. For example,
+	 * u-boot is loaded by x-load at 193M, sysconfig is loaded by u-boot
+	 * at 1M, hypervisor by u-boot hypervisor at 136M.
+	 * On simulator all these binaries are pre-loaded by using '-F' option
+	 * for convenience. On silicon, they will be loaded by their corresponding
+	 * temporary download locations the stage-1/stage-2 firmware
+	 */
+	firmware-download {
+		u-boot	= <0x0c100000 0x3f00000>;  /* 63M @ 193M */
+		sysconfig     = <0x00100000 0x00100000>; /* 1M @ 1M */
+		hypervisor    = <0x08800000 0x00800000>; /* 8M @ 136M */
+		dom0	  = <0x09000000 0x03000000>; /* 48M @ 144M */
+		domU-ramdisk  = <0x60000000 0x60800000>; /* 8M @ 1536M */
+	};
+
+	soc {
+
+	};
+};
+
diff --git a/arch/mips/boot/dts/xlp8xx-multi-os-hybrid.dts b/arch/mips/boot/dts/xlp8xx-multi-os-hybrid.dts
index 4dab0d5..4d8681a 100644
--- a/arch/mips/boot/dts/xlp8xx-multi-os-hybrid.dts
+++ b/arch/mips/boot/dts/xlp8xx-multi-os-hybrid.dts
@@ -27,7 +27,7 @@
 			#size-cells = <1>;
 
 			cpu {
-				onlinemask = <0xffffffff>;
+				onlinemask = <0xf>;
 				nae-rx-vc = <0>;
 				nae-fb-vc = <1>;
 				napi-vc-mask = <0x3>;
@@ -38,15 +38,15 @@
 				sharedcfg = <0x1f000000>;
 			};
 			memory {
-				reg = <0x01000000 0x0B000000
+				reg = <0x02000000 0x0A000000
 					0x20000000 0x20000000>;  // 512M at 512M
 
 			};
 			pic {
 			};
 			nae {
-				owner = <0 0 0 0>;
-				freein-fifo-mask = <0x3 0 0 0>;
+				owner = <1 0 0 0>;
+				freein-fifo-mask = <0x3ffff 0 0 0>;
 				sh-domains = <1>;
 				/* node offsets and size */
 				vfbtbl-sw-offset = <0 0 0 0>;
@@ -55,10 +55,10 @@
 				vfbtbl-hw-nentries = <0 0 0 0>;
 			};
 			fmn {
-				node_0_vc_mask = <0x33333333 0x33333333 0x33333333 0x33333333>;
-				node_1_vc_mask = <0x33333333 0x33333333 0x33333333 0x33333333>;
-				node_2_vc_mask = <0x33333333 0x33333333 0x33333333 0x33333333>;
-				node_3_vc_mask = <0x33333333 0x33333333 0x33333333 0x33333333>;
+				node_0_vc_mask = <0x00000000 0x00000000 0x00000000 0x00000000>;
+				node_1_vc_mask = <0x00000000 0x00000000 0x00000000 0x00000000>;
+				node_2_vc_mask = <0x00000000 0x00000000 0x00000000 0x00000000>;
+				node_3_vc_mask = <0x00000000 0x00000000 0x00000000 0x00000000>;
 			};
 
 
@@ -88,58 +88,6 @@
 			};
 
 			cpu {
-				onlinemask = <0x0f>;
-				nae-rx-vc = <2>;
-				nae-fb-vc = <3>;
-			};
-			uart {
-				id = <0>;
-				sharedcfg = <0x1f000000>;
-			};
-			memory {
-				reg = <0x50000000 0x10000000>;  // 1280M -> 1536M (256M)
-			};
-			nae {
-				owner = <1 0 0 0>;
-				freein-fifo-mask = <0x1111 0 0 0>;
-				sh-domains = <2>;
-				/* node offsets and size */
-				vfbtbl-sw-offset = <0 0 0 0>;
-				vfbtbl-sw-nentries = <0 0 0 0>;
-				vfbtbl-hw-offset = <32 32 32 32>;
-				vfbtbl-hw-nentries = <20 20 20 20>;
-			};
-			app-param{
-				/* Only used in xen loading */
-				linux-mem = <	0x00000000 0x04000000	// START ADDR <MSB LSB> 64MB
-						0x00000000 0x3C000000>;	// SIZE <MSB LSB> 960 MB
-			};
-		};
-		dom@2 {
-			device_type = "domain";
-			os = "netos";
-
-			heap-size = <0x0 0x2000000>;		   // 32M
-			shared-mem = <0x0 0xe4000000 0x0 0x04000000>;  // 3584MB -> 3648 MB (64MB)
-
-			#address-cells = <1>;
-			#size-cells = <1>;
-
-			app-bootcmd = "smpload";
-			app-loadaddr = "0xf4000000"; // 3.75 GB + 64 MB
-
-			kernel {
-				/* Kernel Memory allocated from xen domain-heap */
-				size = <0x02000000>;  // 32M
-
-				/* Location in Dom0 ramdisk */
-				location = "/tmp/xen/netos.elf";
-
-				/* Domain destroy currently doesn't work with hyperexec */
-				dom_destroy_hack = <0>;
-			};
-
-			cpu {
 				onlinemask = <0xf0>;
 				nae-rx-vc = <2>;
 				nae-fb-vc = <3>;
@@ -153,8 +101,8 @@
 			};
 			nae {
 				owner = <0 0 0 0>;
-				freein-fifo-mask = <0x2222 0 0 0>;
-				sh-domains = <1>;
+				freein-fifo-mask = <0x1111 0 0 0>;
+				sh-domains = <2>;
 				/* node offsets and size */
 				vfbtbl-sw-offset = <0 0 0 0>;
 				vfbtbl-sw-nentries = <0 0 0 0>;
@@ -163,11 +111,10 @@
 			};
 			app-param{
 				/* Only used in xen loading */
-				linux-mem = <	0x00000000 0x04000000	// START ADDR <MSB LSB> 64MB
-						0x00000000 0x3C000000>;	// SIZE <MSB LSB> 960 MB
+				linux-mem = <	0x00000000 0x02000000	// START ADDR <MSB LSB> 16MB
+						0x00000000 0x3e000000>;	// SIZE <MSB LSB> 1008 MB
 			};
 		};
-
 	};
 
 	chosen {
-- 
1.7.1

