From afd8f3561f2849937d9f0d78574f16a295d52fea Mon Sep 17 00:00:00 2001
From: Hareesh R <hareeshr@broadcom.com>
Date: Thu, 29 Mar 2012 19:25:59 +0530
Subject: Moved the num-nodes variable from fmn.dts to master dts file Hybrid multios helinux dts file is valid only for multi node setup

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/boot/dts/hybrid-nae.dts b/arch/mips/boot/dts/hybrid-nae.dts
deleted file mode 100644
index c719930..0000000
--- a/arch/mips/boot/dts/hybrid-nae.dts
+++ /dev/null
@@ -1,692 +0,0 @@
-  nae@node-0 {
-	model = "MIPS,XLP8XX NAE CFG";
-	compatible = "NETL,XLP8XX_A0";
-	#address-cells = <1>;
-	#size-cells = <1>;
-	frequency = <500>;
-
-	/include/ "ucore.dts"
-
-        freein-fifo-config {
-                /* If shared is true,
-                        1. Ucore is going to use upto max queues(16 for XLP, 8 for storm) for buffer mgmt
-                        2. Onnchip desc size per queue, same value will be configured for all the queues.
-                   If shared is false,
-                        1. Onchip desc size per queue will be configured using num_free_desc specified at the port level.
-                   If spill descs per queue is specified,
-                        1. spill will be enabled for all the queues(except mgmt) using the spill memory specified in the OS config file.
-                */
-		freein-fifo-spill-mem-range = <0x00000000 0x0fe00000 0x00000000 0x00100000>; // 1MB @ 254MB
-                freein-fifo-shared = <1>;
-                freein-fifo-onchip-num-descs = <56>; 
-                freein-fifo-spill-num-descs = <1200>;
-        };
-
-	// VFBID MAP: Upto 127 entries
-	// (each entry is a pair of (vfbid , dest-vc)
-	// Legal range: (vfbid (0 - 126), dest-vc (0 - 4095))
-	vfbid-config {
-		shared = <1>;
-		vfbid-map = <
-		      // NULL FBID Should map to cpu0 to detect NAE send message errors
-		      127 0
-
-                >;
-	};
-
-	// Packet Ordering Engine (POE)
-	poe {	
-	
-		mode = "bypass";
-
-                distribution-enable = <1>;
-                // 16 bit mask
-                dist-drop-enable = <0>;
-                // 8 bit mask
-                class-drop-enable = <0>;
-                drop-timer = <0>;
-                dest-threshold = <0xa>;
-                dist-threshold = <0xa 0xa 0xa 0xa>;
-                statistics-enable = <0>;
-
-		// Each vector is 512 bit with masb indicating vc 512 and lsb indicating vc 0
-		// Format : 512-bit-vector is specified as 16 32-bit words
-		// Left most word has the vc range 511-479 right most word has vc range 31 - 0
-		// Each word has the MSB select higer vc number and LSB select lower vc num
-                distribution_vectors {
-                                     dv0  = <0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0x1>;
-                                     dv1  = <0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0>;
-                                     dv2  = <0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0>;
-                                     dv3  = <0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0>;
-                                     dv4  = <0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0>;
-                                     dv5  = <0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0>;
-                                     dv6  = <0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0>;
-                                     dv7  = <0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0>;
-                                     dv8  = <0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0>;
-                                     dv9  = <0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0>;
-                                     dv10 = <0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0>;
-                                     dv11 = <0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0>;
-                                     dv12 = <0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0>;
-                                     dv13 = <0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0>;
-                                     dv14 = <0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0>;
-                                     dv15 = <0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0>;
-                };
-	};
-
-	complex@0 {
-		device_type = "nae-complex";
-		mode = "xaui";
-		xaui {
-			loopback = <0>;
-			// Max ingress fifo size 256 units (size of one unit is 64 byte)
-			iface-fifo-size = <55>;
-
-			// Max parser sequence fifo size 1024 packets 
-			// (if 1588 Timestamp is not required, then max size increases to 2048)
-			parser-sequence-fifo-size = <225>;
-
-			// Since rx_buffer is per context, the buffer size given here will be divided by num-channels 
-			num-channels = <1>;
-			rx-buffer-size = <944>;
-
-			// Max available descriptors are 1024  (across all complexes).
-			// Per port num_free_descriptors must be even number
-			num-free-descs = <150>;
-			free-desc-size = <2048>;
-			// In xlp3xx number of ucores available is 8. so, ucore_mask for xlp3xx is (ucore-mask & 0xFF)
-			ucore-mask = <0xffff>;
-		};
-		sgmii {
-	                loopback = <0 0 0 0>;
-			mgmt-port = <0 0 0 0>;
-        	        // Max ingress fifo size 256 units (size of one unit is 64 byte)
-                	iface-fifo-size = <13 13 13 13>;
-
-	                ext-phy-addr = <4 7 6 5>;
-        	        ext-phy-bus = <1 1 1 1>;
-
-	                // Max parser sequence fifo size 1024 packets
-        	        // (if 1588 Timestamp is not required, then max size increases to 2048)
-                	num-channels = <1 1 1 1>;
-	                parser-sequence-fifo-size = <62 62 62 62>;
-
-        	        rx-buffer-size = <128 128 128 128>;
-
-                	// Max available descriptors are 1024 (across all complexes).
-	                // Per port num_free_descriptors must be even number
-        	        num-free-descs = <52 52 52 52>;
-                	free-desc-size = <2048 2048 2048 2048>;
-	                ucore-mask = <0xffff 0xffff 0xffff 0xffff>;
-		};
-                interlaken {
-                        //select daughter board : cortina or loopback
-                        db = "cortina";
-
-                        loopback = <0>;
-
-                        // Maximum number of lanes per interface = 8 (from 2 complexes)
-                        num-lanes = <4>;
-                        // lanerate 0 = 3.125Gbps , 1 = 6.25Gbps
-                        lane-rate = <0>;
-
-                        // Max ingress fifo size 256 units (size of one unit is 64 byte)
-                        iface-fifo-size = <110>;
-
-                        // Max parser sequence fifo size 1024 packets
-                        // (if 1588 Timestamp is not required, then max size increases to 2048)
-                        parser-sequence-fifo-size = <225>;
-
-			// Since rx_buffer is per context, the buffer size given here will be divided by num-channels
-                        num-channels = <4>;
-                        rx-buffer-size = <1840>;
-
-                        // Max available descriptors are 1024  (across all complexes).
-                        // Per port num_free_descriptors must be even number
-                        num-free-descs = <300>;
-                        free-desc-size = <2048>;
-                        ucore-mask = <0xffff>;
-                };
-
-	};
-	complex@1 {
-		device_type = "nae-complex";
-		mode = "xaui";
-		xaui {
-			loopback = <0>;
-			// Max ingress fifo size 256 units (size of one unit is 64 byte)
-			iface-fifo-size = <55>;
-
-			// Max parser sequence fifo size 1024 packets 
-			// (if 1588 Timestamp is not required, then max size increases to 2048)
-			parser-sequence-fifo-size = <225>;
-			
-			// Since rx_buffer is per context, the buffer size given here will be divided by num-channels
-			num-channels = <1>;
-			rx-buffer-size = <944>;
-	
-			// Max available descriptors are 1024  (across all complexes).
-			// Per port num_free_descriptors must be even number
-			num-free-descs = <150>;
-			free-desc-size = <2048>;
-			ucore-mask = <0xffff>;
-		};
-		sgmii {
-	                loopback = <0 0 0 0>;
-
-	                // Max ingress fifo size 256 units (size of one unit is 64 byte)
-        	        iface-fifo-size = <13 13 13 13>;
-
-                	ext-phy-addr = <8 11 10 9>;
-	                ext-phy-bus = <1 1 1 1>;
-
-        	        // Max parser sequence fifo size 1024 packets
-                	// (if 1588 Timestamp is not required, then max size increases to 2048)
-	                num-channels = <1 1 1 1>;
-        	        parser-sequence-fifo-size = <62 62 62 62>;
-
-			// Since rx_buffer is per context, the buffer size given here will be divided by num-channels
-        	        rx-buffer-size = <128 128 128 128>;
-
-
-                	// Max available descriptors are 1024 (across all complexes).
-	                // Per port num_free_descriptors must be even number
-        	        num-free-descs = <52 52 52 52>;
-                	free-desc-size = <2048 2048 2048 2048>;
-	                ucore-mask = <0xffff 0xffff 0xffff 0xffff>;
-		};
-
-	};
-	complex@2 {
-		device_type = "nae-complex";
-		mode = "xaui";
-
-                xaui {
-                        loopback = <0>;
-
-                        // Max ingress fifo size 256 units (size of one unit is 64 byte)
-                        iface-fifo-size = <55>;
-
-                        // Max parser sequence fifo size 1024 packets
-                        // (if 1588 Timestamp is not required, then max size increases to 2048)
-                        parser-sequence-fifo-size = <225>;
-			
-			// Since rx_buffer is per context, the buffer size given here will be divided by num-channels
-                        num-channels = <1>;
-                        rx-buffer-size = <944>;
-
-                        // Max available descriptors are 1024  (across all complexes).
-                        // Per port num_free_descriptors must be even number
-                        num-free-descs = <150>;
-                        free-desc-size = <2048>;
-                        ucore-mask = <0xffff>;
-                };
-
-		interlaken {
-	                //select daughter board : cortina or loopback
-        	        db = "cortina";
-
-                	loopback = <0>;
-
-	                // Maximum number of lanes per interface = 8 (from 2 complexes)
-        	        num-lanes = <4>;
-                	// lanerate 0 = 3.125Gbps , 1 = 6.25Gbps
-	                lane-rate = <0>;
-	
-        	        // Max ingress fifo size 256 units (size of one unit is 64 byte)
-                	iface-fifo-size = <110>;
-
-	                // Max parser sequence fifo size 1024 packets
-        	        // (if 1588 Timestamp is not required, then max size increases to 2048)
-                	parser-sequence-fifo-size = <225>;
-			
-			// Since rx_buffer is per context, the buffer size given here will be divided by num-channels
-                	num-channels = <4>;
-	                rx-buffer-size = <1840>;
-
-	                // Max available descriptors are 1024  (across all complexes).
-        	        // Per port num_free_descriptors must be even number
-                	num-free-descs = <300>;
-	                free-desc-size = <2048>;
-        	        ucore-mask = <0xffff>;
-		};
-		sgmii {
-        	        loopback = <0 0 0 0>;
-
-                	// Max ingress fifo size 256 units (size of one unit is 64 byte)
-	                iface-fifo-size = <13 13 13 13>;
-
-        	        ext-phy-addr = <0 3 2 1>;
-                	ext-phy-bus = <1 1 1 1>;
-
-	                // Max parser sequence fifo size 1024 packets
-        	        // (if 1588 Timestamp is not required, then max size increases to 2048)
-                	num-channels = <1 1 1 1>;
-	                parser-sequence-fifo-size = <62 62 62 62>;
-
-			// Since rx_buffer is per context, the buffer size given here will be divided by num-channels
-	                rx-buffer-size = <128 128 128 128>;
-
-
-        	        // Max available descriptors are 1024 (across all complexes).
-                	// Per port num_free_descriptors must be even number
-	                num-free-descs = <52 52 52 52>;
-        	        free-desc-size = <2048 2048 2048 2048>;
-                	ucore-mask = <0xffff 0xffff 0xffff 0xffff>;
-		};
-	};
-	complex@3 {
-		device_type = "nae-complex";
-		mode = "xaui";
-		xaui {
-			loopback = <0>;
-
-			// Max ingress fifo size 256 units (size of one unit is 64 byte)
-			iface-fifo-size = <55>;
-
-			// Max parser sequence fifo size 1024 packets 
-			// (if 1588 Timestamp is not required, then max size increases to 2048)
-			parser-sequence-fifo-size = <225>;
-	
-			// Since rx_buffer is per context, the buffer size given here will be divided by num-channels
-			num-channels = <1>;
-			rx-buffer-size = <944>;
-
-			// Max available descriptors are 1024  (across all complexes).
-			// Per port num_free_descriptors must be even number
-			num-free-descs = <150>;
-			free-desc-size = <2048>;
-			ucore-mask = <0xffff>;
-		};
-                sgmii {
-                        loopback = <0 0 0 0>;
-
-                        // Max ingress fifo size 256 units (size of one unit is 64 byte)
-                        iface-fifo-size = <13 13 13 13>;
-
-                        ext-phy-addr = <12 15 14 13>;
-                        ext-phy-bus = <1 1 1 1>;
-
-                        // Max parser sequence fifo size 1024 packets
-                        // (if 1588 Timestamp is not required, then max size increases to 2048)
-                        num-channels = <1 1 1 1>;
-                        parser-sequence-fifo-size = <62 62 62 62>;
-
-			// Since rx_buffer is per context, the buffer size given here will be divided by num-channels
-                        rx-buffer-size = <128 128 128 128>;
-
-
-                        // Max available descriptors are 1024 (across all complexes).
-                        // Per port num_free_descriptors must be even number
-                        num-free-descs = <52 52 52 52>;
-                        free-desc-size = <2048 2048 2048 2048>;
-                        ucore-mask = <0xffff 0xffff 0xffff 0xffff>;
-                };
-	};
-	complex@4 {
-		device_type = "nae-complex";
-		mode = "sgmii";
-
-		sgmii {
-			loopback = <0 0>;
-	
-			mgmt-port = <1 0>;
-
-			// Max ingress fifo size 256 units (size of one unit is 64 byte)
-			iface-fifo-size = <13 13>;
-                
-			ext-phy-addr = <16 17>;
-        	        ext-phy-bus = <0 0>;
-
-			// Max parser sequence fifo size 1024 packets 
-			// (if 1588 Timestamp is not required, then max size increases to 2048)
-			num-channels = <1 1>;
-			parser-sequence-fifo-size = <62 62>;
-
-			// Since rx_buffer is per context, the buffer size given here will be divided by num-channels
-			rx-buffer-size = <128 128>;
-				   
-
-			// Max available descriptors are 1024 (across all complexes).
-			// Per port num_free_descriptors must be even number
-			num-free-descs = <52 52>;
-			free-desc-size = <2048 2048>;
-			ucore-mask = <0xffff 0xffff>;
-		};
-	};
-};
-   nae@node-1 {
-	model = "MIPS,XLP8XX NAE CFG";
-	compatible = "NETL,XLP8XX_A0";
-	#address-cells = <1>;
-	#size-cells = <1>;
-	frequency = <500>;
-
-	/include/ "ucore.dts"
-
-        freein-fifo-config {
-                /* If shared is true,
-                        1. Ucore is going to use upto max queues(16 for XLP, 8 for storm) for buffer mgmt
-                        2. Onnchip desc size per queue, same value will be configured for all the queues.
-                   If shared is false,
-                        1. Onchip desc size per queue will be configured using num_free_desc specified at the port level.
-                   If spill descs per queue is specified,
-                        1. spill will be enabled for all the queues(except mgmt) using the spill memory specified in the OS config file.
-                */
-
-                freein-fifo-shared = <0>;
-                freein-fifo-onchip-num-descs = <0>;
-                freein-fifo-spill-num-descs = <0>;
-        };
-
-	// VFBID MAP: Upto 127 entries
-	// (each entry is a pair of (vfbid , dest-vc)
-	// Legal range: (vfbid (0 - 126), dest-vc (0 - 4095))
-	vfbid-config {
-		shared = <1>;
-		vfbid-map = <
-		      // NULL FBID Should map to cpu0 to detect NAE send message errors
-		      127 0
-
-                >;
-	};
-
-	// Packet Ordering Engine (POE)
-	poe {	
-	
-		mode = "bypass";
-
-		// Each vector is 512 bit with masb indicating vc 512 and lsb indicating vc 0
-		// Format : 512-bit-vector is specified as 16 32-bit words
-		// Left most word has the vc range 511-479 right most word has vc range 31 - 0
-		// Each word has the MSB select higer vc number and LSB select lower vc num
-		
-		distribution-enable = <1>;
-		// 16 bit mask
-		dist-drop-enable = <0>;	 
-		// 8 bit mask
-		class-drop-enable = <0>;
-		drop-timer = <0>;
-		dest-threshold = <0xa>;
-		dist-threshold = <0xa 0xa 0xa 0xa>;
-		statistics-enable = <0>;
- 
-                distribution_vectors {
-                                     dv0  = <0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0x1>;
-                                     dv1  = <0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0>;
-                                     dv2  = <0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0>;
-                                     dv3  = <0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0>;
-                                     dv4  = <0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0>;
-                                     dv5  = <0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0>;
-                                     dv6  = <0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0>;
-                                     dv7  = <0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0>;
-                                     dv8  = <0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0>;
-                                     dv9  = <0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0>;
-                                     dv10 = <0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0>;
-                                     dv11 = <0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0>;
-                                     dv12 = <0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0>;
-                                     dv13 = <0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0>;
-                                     dv14 = <0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0>;
-                                     dv15 = <0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0>;
-                };
-	};
-
-	complex@0 {
-		device_type = "nae-complex";
-		mode = "xaui";
-		xaui {
-			loopback = <0>;
-			// Max ingress fifo size 256 units (size of one unit is 64 byte)
-			iface-fifo-size = <55>;
-
-			// Max parser sequence fifo size 1024 packets 
-			// (if 1588 Timestamp is not required, then max size increases to 2048)
-			parser-sequence-fifo-size = <225>;
-
-			// Since rx_buffer is per context, the buffer size given here will be divided by num-channels
-			num-channels = <1>;
-			rx-buffer-size = <944>;
-
-			// Max available descriptors are 1024  (across all complexes).
-			// Per port num_free_descriptors must be even number
-			num-free-descs = <150>;
-			free-desc-size = <2048>;
-			ucore-mask = <0xffff>;
-		};
-		sgmii {
-	                loopback = <0 0 0 0>;
-	
-        	        // Max ingress fifo size 256 units (size of one unit is 64 byte)
-                	iface-fifo-size = <13 13 13 13>;
-
-	                ext-phy-addr = <4 7 6 5>;
-        	        ext-phy-bus = <1 1 1 1>;
-
-	                // Max parser sequence fifo size 1024 packets
-        	        // (if 1588 Timestamp is not required, then max size increases to 2048)
-                	num-channels = <1 1 1 1>;
-	                parser-sequence-fifo-size = <62 62 62 62>;
-
-			// Since rx_buffer is per context, the buffer size given here will be divided by num-channels
-        	        rx-buffer-size = <128 128 128 128>;
-
-
-                	// Max available descriptors are 1024 (across all complexes).
-	                // Per port num_free_descriptors must be even number
-        	        num-free-descs = <52 52 52 52>;
-                	free-desc-size = <2048 2048 2048 2048>;
-	                ucore-mask = <0xffff 0xffff 0xffff 0xffff>;
-		};
-                interlaken {
-                        //select daughter board : cortina or loopback
-                        db = "cortina";
-
-                        loopback = <0>;
-
-                        // Maximum number of lanes per interface = 8 (from 2 complexes)
-                        num-lanes = <4>;
-                        // lanerate 0 = 3.125Gbps , 1 = 6.25Gbps
-                        lane-rate = <0>;
-
-                        // Max ingress fifo size 256 units (size of one unit is 64 byte)
-                        iface-fifo-size = <110>;
-
-                        // Max parser sequence fifo size 1024 packets
-                        // (if 1588 Timestamp is not required, then max size increases to 2048)
-                        parser-sequence-fifo-size = <225>;
-
-                        // Since rx_buffer is per context, the buffer size given here will be divided by num-channels
-                        num-channels = <4>;
-                        rx-buffer-size = <1840>;
-
-                        // Max available descriptors are 1024  (across all complexes).
-                        // Per port num_free_descriptors must be even number
-                        num-free-descs = <300>;
-                        free-desc-size = <2048>;
-                        ucore-mask = <0xffff>;
-                };
-	};
-	complex@1 {
-		device_type = "nae-complex";
-		mode = "xaui";
-		xaui {
-			loopback = <0>;
-			// Max ingress fifo size 256 units (size of one unit is 64 byte)
-			iface-fifo-size = <55>;
-
-			// Max parser sequence fifo size 1024 packets 
-			// (if 1588 Timestamp is not required, then max size increases to 2048)
-			parser-sequence-fifo-size = <225>;
-
-			// Since rx_buffer is per context, the buffer size given here will be divided by num-channels
-			num-channels = <1>;
-			rx-buffer-size = <944>;
-	
-			// Max available descriptors are 1024  (across all complexes).
-			// Per port num_free_descriptors must be even number
-			num-free-descs = <150>;
-			free-desc-size = <2048>;
-			ucore-mask = <0xffff>;
-		};
-		sgmii {
-	                loopback = <0 0 0 0>;
-
-	                // Max ingress fifo size 256 units (size of one unit is 64 byte)
-        	        iface-fifo-size = <13 13 13 13>;
-
-                	ext-phy-addr = <8 11 10 9>;
-	                ext-phy-bus = <1 1 1 1>;
-
-        	        // Max parser sequence fifo size 1024 packets
-                	// (if 1588 Timestamp is not required, then max size increases to 2048)
-	                num-channels = <1 1 1 1>;
-        	        parser-sequence-fifo-size = <62 62 62 62>;
-
-			// Since rx_buffer is per context, the buffer size given here will be divided by num-channels
-        	        rx-buffer-size = <128 128 128 128>;
-
-
-                	// Max available descriptors are 1024 (across all complexes).
-	                // Per port num_free_descriptors must be even number
-        	        num-free-descs = <52 52 52 52>;
-                	free-desc-size = <2048 2048 2048 2048>;
-	                ucore-mask = <0xffff 0xffff 0xffff 0xffff>;
-		};
-
-	};
-	complex@2 {
-		device_type = "nae-complex";
-		mode = "xaui";
-
-                xaui {
-                        loopback = <0>;
-
-                        // Max ingress fifo size 256 units (size of one unit is 64 byte)
-                        iface-fifo-size = <55>;
-
-                        // Max parser sequence fifo size 1024 packets
-                        // (if 1588 Timestamp is not required, then max size increases to 2048)
-                        parser-sequence-fifo-size = <225>;
-
-                        num-channels = <1>;
-                        rx-buffer-size = <944>;
-
-                        // Max available descriptors are 1024  (across all complexes).
-                        // Per port num_free_descriptors must be even number
-                        num-free-descs = <150>;
-                        free-desc-size = <2048>;
-                        ucore-mask = <0xffff>;
-                };
-
-		interlaken {
-	                //select daughter board : cortina or loopback
-        	        db = "cortina";
-
-                	loopback = <0>;
-
-	                // Maximum number of lanes per interface = 8 (from 2 complexes)
-        	        num-lanes = <4>;
-                	// lanerate 0 = 3.125Gbps , 1 = 6.25Gbps
-	                lane-rate = <0>;
-	
-        	        // Max ingress fifo size 256 units (size of one unit is 64 byte)
-                	iface-fifo-size = <110>;
-
-	                // Max parser sequence fifo size 1024 packets
-        	        // (if 1588 Timestamp is not required, then max size increases to 2048)
-                	parser-sequence-fifo-size = <225>;
-
-			// Since rx_buffer is per context, the buffer size given here will be divided by num-channels
-                	num-channels = <4>;
-	                rx-buffer-size = <1840>;
-
-	                // Max available descriptors are 1024  (across all complexes).
-        	        // Per port num_free_descriptors must be even number
-                	num-free-descs = <300>;
-	                free-desc-size = <2048>;
-        	        ucore-mask = <0xffff>;
-		};
-		sgmii {
-        	        loopback = <0 0 0 0>;
-
-                	// Max ingress fifo size 256 units (size of one unit is 64 byte)
-	                iface-fifo-size = <13 13 13 13>;
-
-        	        ext-phy-addr = <0 3 2 1>;
-                	ext-phy-bus = <1 1 1 1>;
-
-	                // Max parser sequence fifo size 1024 packets
-        	        // (if 1588 Timestamp is not required, then max size increases to 2048)
-                	num-channels = <1 1 1 1>;
-	                parser-sequence-fifo-size = <62 62 62 62>;
-
-			// Since rx_buffer is per context, the buffer size given here will be divided by num-channels
-	                rx-buffer-size = <128 128 128 128>;
-
-
-        	        // Max available descriptors are 1024 (across all complexes).
-                	// Per port num_free_descriptors must be even number
-	                num-free-descs = <52 52 52 52>;
-        	        free-desc-size = <2048 2048 2048 2048>;
-                	ucore-mask = <0xffff 0xffff 0xffff 0xffff>;
-		};
-	};
-	complex@3 {
-		device_type = "nae-complex";
-		mode = "xaui";
-		xaui {
-			loopback = <0>;
-
-			// Max ingress fifo size 256 units (size of one unit is 64 byte)
-			iface-fifo-size = <55>;
-
-			// Max parser sequence fifo size 1024 packets 
-			// (if 1588 Timestamp is not required, then max size increases to 2048)
-			parser-sequence-fifo-size = <225>;
-
-			// Since rx_buffer is per context, the buffer size given here will be divided by num-channels
-			num-channels = <1>;
-			rx-buffer-size = <944>;
-
-			// Max available descriptors are 1024  (across all complexes).
-			// Per port num_free_descriptors must be even number
-			num-free-descs = <150>;
-			free-desc-size = <2048>;
-			ucore-mask = <0xffff>;
-		};
-	};
-	complex@4 {
-		device_type = "nae-complex";
-		mode = "sgmii";
-
-		sgmii {
-			loopback = <0 0>;
-			mgmt-port = <1 0>;
-			// Max ingress fifo size 256 units (size of one unit is 64 byte)
-			iface-fifo-size = <13 13>;
- 		               
-			ext-phy-addr = <16 17>;
-        	        ext-phy-bus = <0 0>;
-
-			// Max parser sequence fifo size 1024 packets 
-			// (if 1588 Timestamp is not required, then max size increases to 2048)
-			num-channels = <1 1>;
-			parser-sequence-fifo-size = <62 62>;
-	
-			// Since rx_buffer is per context, the buffer size given here will be divided by num-channels
-			rx-buffer-size = <128 128>;
-				   
-
-			// Max available descriptors are 1024 (across all complexes).
-			// Per port num_free_descriptors must be even number
-			num-free-descs = <52 52>;
-			free-desc-size = <2048 2048>;
-			ucore-mask = <0xffff 0xffff>;
-		};
-	};
-};
-nae@node-2 {
-
-};
-nae@node-3 {
-
-};
diff --git a/arch/mips/boot/dts/mnxlp8xx-linux-perf.dts b/arch/mips/boot/dts/mnxlp8xx-linux-perf.dts
new file mode 100644
index 0000000..dabc454
--- /dev/null
+++ b/arch/mips/boot/dts/mnxlp8xx-linux-perf.dts
@@ -0,0 +1,95 @@
+/* XLP8XX Device Tree Source
+ *
+ */
+
+/dts-v1/;
+/ {
+	model = "MIPS,XLP8XX";
+	compatible = "NETL,XLP8XX_A0";
+
+	#address-cells = <1>;
+	#size-cells = <1>;
+
+	hypervisor {
+		hypervisor-name = "Xen";
+		alloc_dom0_memory = <0>;
+		bootargs = "ncores=8 dom0_loadaddr=0x72000000 dom0_size=0x1c000000 dom0_cpumask=0xffffffff -- ";
+		domain_heap = <0x80000000 0x20000000>;
+	};
+	doms {
+		#address-cells = <1>;
+		#size-cells = <1>;
+		dom@0 {
+			device_type = "domain";
+			os = "linux";
+
+			#address-cells = <1>;
+			#size-cells = <1>;
+			#nodes = <2>;
+			#cpumask-cells=<2>;
+
+			cpu {
+				onlinemask = <0xffffffff 0xffffffff>;
+                                nae-rx-vc = <0>;
+                                nae-fb-vc = <1>;
+			};
+			uart {
+				id = <0>;
+				owner = <1>;
+				sharedcfg = <0x1f000000>;
+			};
+			//Make sure each node's memory is separated out 
+			memory {
+				reg = <0x01000000 0x0B000000	// 176M at 16M
+					0x20000000 0xa0000000>; // 2586M at 512M
+			};
+                        fmn {
+                                node_0_vc_mask = <0x00000000 0x00000000 0x00000000 0x00000000>;
+                                node_1_vc_mask = <0x00000000 0x00000000 0x00000000 0x00000000>;
+                                node_2_vc_mask = <0x00000000 0x00000000 0x00000000 0x00000000>;
+                                node_3_vc_mask = <0x00000000 0x00000000 0x00000000 0x00000000>;                       
+			};
+			pic {
+			};
+			nae {
+				owner = <1 1 0 0>;
+				freein-fifo-mask = <0xffff 0xffff 0 0>;
+				/* node offsets and size */
+				vfbtbl-sw-offset = <0 0 0 0>;
+				vfbtbl-sw-nentries = <32 32 32 32>;
+				/* hw entries 40 for 2 nodes , 20 fifos each */
+				vfbtbl-hw-offset = <32 32 32 32>;
+				vfbtbl-hw-nentries = <40 40 40 40>;
+			};
+
+		};
+	};
+
+	chosen {
+		bootargs = " xlp_noi2c console=ttyS0,115200  mem=512m@3584m";
+	};
+
+	/* These binaries are downloaded at the resp physical memory locations
+	 * by their corresponding bootstrapping loaders. For example,
+	 * u-boot is loaded by x-load at 193M, sysconfig is loaded by u-boot
+	 * at 1M, hypervisor by u-boot hypervisor at 136M.
+	 * On simulator all these binaries are pre-loaded by using '-F' option
+	 * for convenience. On silicon, they will be loaded by their corresponding
+	 * temporary download locations the stage-1/stage-2 firmware
+	 */
+	firmware-download {
+		u-boot        = <0x0c100000 0x3f00000>;  /* 63M @ 193M */
+		sysconfig     = <0x00100000 0x00100000>; /* 1M @ 1M */
+		hypervisor    = <0x08800000 0x00800000>; /* 8M @ 136M */
+		dom0          = <0x09000000 0x03000000>; /* 48M @ 144M */
+		domU-ramdisk  = <0x60000000 0x60800000>; /* 8M @ 1536M */
+	};
+
+	soc {
+		nodes {
+			num-nodes = <2>;
+		};
+
+	};
+};
+
diff --git a/arch/mips/boot/dts/mnxlp8xx-linux.dts b/arch/mips/boot/dts/mnxlp8xx-linux.dts
index 28e9281..84cf1eb 100644
--- a/arch/mips/boot/dts/mnxlp8xx-linux.dts
+++ b/arch/mips/boot/dts/mnxlp8xx-linux.dts
@@ -80,6 +80,9 @@
 	};
 
 	soc {
+		nodes {
+			num-nodes = <2>;
+		};
 
 	};
 };
diff --git a/arch/mips/boot/dts/mnxlp8xx-multi-os-hybrid-helinux.dts b/arch/mips/boot/dts/mnxlp8xx-multi-os-hybrid-helinux.dts
new file mode 100644
index 0000000..889d7e5
--- /dev/null
+++ b/arch/mips/boot/dts/mnxlp8xx-multi-os-hybrid-helinux.dts
@@ -0,0 +1,195 @@
+/* XLP8XX Device Tree Source
+ *
+ */
+
+/dts-v1/;
+/ {
+	model = "MIPS,XLP8XX";
+	compatible = "NETL,XLP8XX_A0";
+
+	#address-cells = <1>;
+	#size-cells = <1>;
+
+	hypervisor {
+		hypervisor-name = "Xen";
+		alloc_dom0_memory = <0>;
+		bootargs = "ncores=8 dom0_loadaddr=0x72000000 dom0_size=0x1c000000 dom0_cpumask=0x0f -- ";
+		domain_heap = <0xf0000000 0xc0000000>; /* 3584MB -> 6656MB */
+	};
+	doms {
+		#address-cells = <1>;
+		#size-cells = <1>;
+		dom@0 {
+			device_type = "domain";
+			os = "linux";
+
+			#address-cells = <1>;
+			#size-cells = <1>;
+			#nodes = <2>;
+			#cpumask-cells=<2>;
+
+			cpu {
+				onlinemask = <0xffffffff 0xffffffff>;
+				nae-rx-vc = <0>;
+				nae-fb-vc = <1>;
+			};
+			uart {
+				id = <0>;
+				owner = <1>;
+				sharedcfg = <0x1f000000>;
+			};
+			memory {
+				reg = <0x01000000 0x0B000000
+					0x20000000 0xa0000000>; // 2586M at 512M
+
+			};
+			pic {
+			};
+			nae {
+				owner = <1 1 0 0>;
+				freein-fifo-mask = <0x30000 0x30000 0 0>;
+				sh-domains = <1 2>;
+				/* node offsets and size */
+				vfbtbl-sw-offset = <0 0 0 0>;
+				vfbtbl-sw-nentries = <32 32 32 32>;
+				vfbtbl-hw-offset = <0 0 0 0>;
+				vfbtbl-hw-nentries = <0 0 0 0>;
+				dummy-pktdata-addr = <0x0 0x0ff00000>; // 255 MB
+			};
+			fmn {
+                                node_0_vc_mask = <0x00000000 0x00000000 0x00000000 0x00000000>;
+                                node_1_vc_mask = <0x00000000 0x00000000 0x00000000 0x00000000>;
+                                node_2_vc_mask = <0x00000000 0x00000000 0x00000000 0x00000000>;
+                                node_3_vc_mask = <0x00000000 0x00000000 0x00000000 0x00000000>;                       
+			};
+
+
+		};
+		dom@1 {
+			device_type = "domain";
+			os = "netos";
+
+			heap-size = <0x0 0x2000000>;		   // 32M
+			shared-mem = <0x1 0x00000000 0x0 0x04000000>;  // 64MB @ 4G
+
+			#address-cells = <1>;
+			#size-cells = <1>;
+
+			app-bootcmd = "smpload";
+			app-loadaddr = "0xf4000000"; // 3.75 GB + 64 MB
+
+			kernel {
+				/* Kernel Memory allocated from xen domain-heap */
+				size = <0x02000000>;  // 32M
+
+				/* Location in Dom0 ramdisk */
+				location = "/tmp/xen/netos.elf";
+
+				/* Domain destroy currently doesn't work with hyperexec */
+				dom_destroy_hack = <0>;
+			};
+
+			cpu {
+				onlinemask = <0x00000000 0xffffffff>;
+				nae-rx-vc = <2>;
+				nae-fb-vc = <2>;
+			};
+			uart {
+				id = <0>;
+				sharedcfg = <0x1f000000>;
+			};
+			memory {
+				reg = <0x50000000 0x10000000>;  // 1280M -> 1536M (256M)
+			};
+			nae {
+				owner = <0 0 0 0>;
+				freein-fifo-mask = <0xffff 0 0 0>;
+				sh-domains = <0>;
+				/* node offsets and size */
+				vfbtbl-sw-offset = <0 0 0 0>;
+				vfbtbl-sw-nentries = <0 0 0 0>;
+				vfbtbl-hw-offset = <32 32 32 32>;
+				vfbtbl-hw-nentries = <20 20 20 20>;
+
+				dummy-pktdata-addr = <0x0 0x0ff00000>; // 255 MB
+			};
+		};
+		dom@2 {
+			device_type = "domain";
+			os = "netos";
+
+			heap-size = <0x0 0x2000000>;		   // 32M
+			shared-mem = <0x2 0x29000000 0x0 0x04000000>;  // 64MB @ 8848MB
+
+			#address-cells = <1>;
+			#size-cells = <1>;
+
+			app-bootcmd = "smpload";
+			app-loadaddr = "0xf4000000"; // 3.75 GB + 64 MB
+
+			kernel {
+				/* Kernel Memory allocated from xen domain-heap */
+				size = <0x02000000>;  // 32M
+
+				/* Location in Dom0 ramdisk */
+				location = "/tmp/xen/netos.elf";
+
+				/* Domain destroy currently doesn't work with hyperexec */
+				dom_destroy_hack = <0>;
+			};
+
+			cpu {
+				onlinemask = <0xffffffff 0x00000000>;
+				nae-rx-vc = <2>;
+				nae-fb-vc = <2>;
+			};
+			uart {
+				id = <0>;
+				sharedcfg = <0x1f000000>;
+			};
+			memory {
+				reg = <0x50000000 0x10000000>;  // 1280M -> 1536M (256M)
+			};
+			nae {
+				owner = <0 0 0 0>;
+				freein-fifo-mask = <0x0 0xffff 0 0>;
+				sh-domains = <0>;
+				/* node offsets and size */
+				vfbtbl-sw-offset = <0 0 0 0>;
+				vfbtbl-sw-nentries = <0 0 0 0>;
+				vfbtbl-hw-offset = <32 32 32 32>;
+				vfbtbl-hw-nentries = <20 20 20 20>;
+
+				dummy-pktdata-addr = <0x0 0x0ff00000>; // 255 MB
+			};
+		};
+
+	};
+
+	chosen {
+		bootargs = " xlp_noi2c console=ttyS0,115200 mem=512m@3584m";
+	};
+
+	/* These binaries are downloaded at the resp physical memory locations
+	 * by their corresponding bootstrapping loaders. For example,
+	 * u-boot is loaded by x-load at 193M, sysconfig is loaded by u-boot
+	 * at 1M, hypervisor by u-boot hypervisor at 136M.
+	 * On simulator all these binaries are pre-loaded by using '-F' option
+	 * for convenience. On silicon, they will be loaded by their corresponding
+	 * temporary download locations the stage-1/stage-2 firmware
+	 */
+	firmware-download {
+		u-boot	= <0x0c100000 0x3f00000>;  /* 63M @ 193M */
+		sysconfig     = <0x00100000 0x00100000>; /* 1M @ 1M */
+		hypervisor    = <0x08800000 0x00800000>; /* 8M @ 136M */
+		dom0	  = <0x09000000 0x03000000>; /* 48M @ 144M */
+		domU-ramdisk  = <0x60000000 0x60800000>; /* 8M @ 1536M */
+	};
+
+	soc {
+		nodes {
+			num-nodes = <2>;
+		};
+	};
+};
+
diff --git a/arch/mips/boot/dts/mnxlpev-linux.dts b/arch/mips/boot/dts/mnxlpev-linux.dts
index 28e9281..84cf1eb 100644
--- a/arch/mips/boot/dts/mnxlpev-linux.dts
+++ b/arch/mips/boot/dts/mnxlpev-linux.dts
@@ -80,6 +80,9 @@
 	};
 
 	soc {
+		nodes {
+			num-nodes = <2>;
+		};
 
 	};
 };
diff --git a/arch/mips/boot/dts/netos-srtp.dts b/arch/mips/boot/dts/netos-srtp.dts
index 18110f7..3f3d454 100644
--- a/arch/mips/boot/dts/netos-srtp.dts
+++ b/arch/mips/boot/dts/netos-srtp.dts
@@ -75,6 +75,9 @@
 	};
 
 	soc {
+		nodes {
+			num-nodes = <1>;
+		};
 
 	};
 };
diff --git a/arch/mips/boot/dts/xlp3xx-helinux.dts b/arch/mips/boot/dts/xlp3xx-helinux.dts
index 74a3b70..d5860fd 100644
--- a/arch/mips/boot/dts/xlp3xx-helinux.dts
+++ b/arch/mips/boot/dts/xlp3xx-helinux.dts
@@ -126,6 +126,9 @@
 	};
 
 	soc {
+		nodes {
+			num-nodes = <1>;
+		};
 
 	};
 };
diff --git a/arch/mips/boot/dts/xlp3xx-linux-loader.dts b/arch/mips/boot/dts/xlp3xx-linux-loader.dts
index 4698d7a..3285b27 100644
--- a/arch/mips/boot/dts/xlp3xx-linux-loader.dts
+++ b/arch/mips/boot/dts/xlp3xx-linux-loader.dts
@@ -112,6 +112,9 @@
 	};
 
 	soc {
+		nodes {
+			num-nodes = <1>;
+		};
 
 	};
 };
diff --git a/arch/mips/boot/dts/xlp3xx-linux-perf.dts b/arch/mips/boot/dts/xlp3xx-linux-perf.dts
index c0c8477..8d150ec 100644
--- a/arch/mips/boot/dts/xlp3xx-linux-perf.dts
+++ b/arch/mips/boot/dts/xlp3xx-linux-perf.dts
@@ -41,13 +41,25 @@
 					0x20000000 0x60000000>;  // 2586M at 512M
 			};
                         fmn {
-								node_0_vc_mask = <0x00000000 0x00000000 0x00000000 0x00000000>;
+				node_0_vc_mask = <0x00000000 0x00000000 0x00000000 0x00000000>;
                                 node_1_vc_mask = <0xffffffff 0xffffffff 0xffffffff 0xffffffff>;
                                 node_2_vc_mask = <0xffffffff 0xffffffff 0xffffffff 0xffffffff>;
                                 node_3_vc_mask = <0xffffffff 0xffffffff 0xffffffff 0xffffffff>;
                         };
 			pic {
 			};
+			nae {
+				owner = <1 0 0 0>;
+				freein-fifo-mask = <0xff 0 0 0>;
+				/* node offsets and size */
+				vfbtbl-sw-offset = <0 0 0 0>;
+				vfbtbl-sw-nentries = <32 32 32 32>;
+				/* hw entries */
+				vfbtbl-hw-offset = <32 32 32 32>;
+				vfbtbl-hw-nentries = <20 20 20 20>;
+			};
+
+
 		};
 	};
 
@@ -72,6 +84,9 @@
 	};
 
 	soc {
+		nodes {
+			num-nodes = <1>;
+		};
 
 	};
 };
diff --git a/arch/mips/boot/dts/xlp3xx-linux.dts b/arch/mips/boot/dts/xlp3xx-linux.dts
index 8d1ae4e..e4c1c6d 100644
--- a/arch/mips/boot/dts/xlp3xx-linux.dts
+++ b/arch/mips/boot/dts/xlp3xx-linux.dts
@@ -78,6 +78,9 @@
 	};
 
 	soc {
+		nodes {
+			num-nodes = <1>;
+		};
 
 	};
 };
diff --git a/arch/mips/boot/dts/xlp3xx-multi-os.dts b/arch/mips/boot/dts/xlp3xx-multi-os.dts
index fc2dc11..b92fd87 100644
--- a/arch/mips/boot/dts/xlp3xx-multi-os.dts
+++ b/arch/mips/boot/dts/xlp3xx-multi-os.dts
@@ -188,6 +188,9 @@
 	};
 
 	soc {
+		nodes {
+			num-nodes = <1>;
+		};
 
 	};
 };
diff --git a/arch/mips/boot/dts/xlp3xx-netos.dts b/arch/mips/boot/dts/xlp3xx-netos.dts
index 5c57145..e90fe99 100644
--- a/arch/mips/boot/dts/xlp3xx-netos.dts
+++ b/arch/mips/boot/dts/xlp3xx-netos.dts
@@ -83,6 +83,9 @@
 	};
 
 	soc {
+		nodes {
+			num-nodes = <1>;
+		};
 
 	};
 };
diff --git a/arch/mips/boot/dts/xlp4xx-helinux.dts b/arch/mips/boot/dts/xlp4xx-helinux.dts
index c4156a6..87330f8 100644
--- a/arch/mips/boot/dts/xlp4xx-helinux.dts
+++ b/arch/mips/boot/dts/xlp4xx-helinux.dts
@@ -186,6 +186,9 @@
 	};
 
 	soc {
+		nodes {
+			num-nodes = <1>;
+		};
 
 	};
 };
diff --git a/arch/mips/boot/dts/xlp4xx-linux.dts b/arch/mips/boot/dts/xlp4xx-linux.dts
index f9d8f32..3807ac1 100644
--- a/arch/mips/boot/dts/xlp4xx-linux.dts
+++ b/arch/mips/boot/dts/xlp4xx-linux.dts
@@ -78,6 +78,9 @@
 	};
 
 	soc {
+		nodes {
+			num-nodes = <1>;
+		};
 
 	};
 };
diff --git a/arch/mips/boot/dts/xlp4xx-multi-os.dts b/arch/mips/boot/dts/xlp4xx-multi-os.dts
index c1a3fce..29fdd18 100644
--- a/arch/mips/boot/dts/xlp4xx-multi-os.dts
+++ b/arch/mips/boot/dts/xlp4xx-multi-os.dts
@@ -188,6 +188,9 @@
 	};
 
 	soc {
+		nodes {
+			num-nodes = <1>;
+		};
 
 	};
 };
diff --git a/arch/mips/boot/dts/xlp4xx-netos.dts b/arch/mips/boot/dts/xlp4xx-netos.dts
index 9260682..de0b44f 100644
--- a/arch/mips/boot/dts/xlp4xx-netos.dts
+++ b/arch/mips/boot/dts/xlp4xx-netos.dts
@@ -82,6 +82,9 @@
 	};
 
 	soc {
+		nodes {
+			num-nodes = <1>;
+		};
 
 	};
 };
diff --git a/arch/mips/boot/dts/xlp8xx-helinux.dts b/arch/mips/boot/dts/xlp8xx-helinux.dts
index 76807a5..0aea1dc 100644
--- a/arch/mips/boot/dts/xlp8xx-helinux.dts
+++ b/arch/mips/boot/dts/xlp8xx-helinux.dts
@@ -226,6 +226,9 @@
 	};
 
 	soc {
+		nodes {
+			num-nodes = <1>;
+		};
 
 	};
 };
diff --git a/arch/mips/boot/dts/xlp8xx-linux-loader.dts b/arch/mips/boot/dts/xlp8xx-linux-loader.dts
index a98a611..e910c54 100644
--- a/arch/mips/boot/dts/xlp8xx-linux-loader.dts
+++ b/arch/mips/boot/dts/xlp8xx-linux-loader.dts
@@ -196,6 +196,9 @@
 	};
 
 	soc {
+		nodes {
+			num-nodes = <1>;
+		};
 
 	};
 };
diff --git a/arch/mips/boot/dts/xlp8xx-linux-perf.dts b/arch/mips/boot/dts/xlp8xx-linux-perf.dts
index db17d00..3ff9f25 100644
--- a/arch/mips/boot/dts/xlp8xx-linux-perf.dts
+++ b/arch/mips/boot/dts/xlp8xx-linux-perf.dts
@@ -48,6 +48,17 @@
                         };
 			pic {
 			};
+			nae {
+				owner = <1 0 0 0>;
+				freein-fifo-mask = <0xffff 0 0 0>;
+				/* node offsets and size */
+				vfbtbl-sw-offset = <0 0 0 0>;
+				vfbtbl-sw-nentries = <32 32 32 32>;
+				/* hw entries */
+				vfbtbl-hw-offset = <32 32 32 32>;
+				vfbtbl-hw-nentries = <20 20 20 20>;
+			};
+
 		};
 	};
 	chosen {
@@ -71,6 +82,9 @@
 	};
 
 	soc {
+		nodes {
+			num-nodes = <1>;
+		};
 
 	};
 };
diff --git a/arch/mips/boot/dts/xlp8xx-linux.dts b/arch/mips/boot/dts/xlp8xx-linux.dts
index 6834fde..943ec3a 100644
--- a/arch/mips/boot/dts/xlp8xx-linux.dts
+++ b/arch/mips/boot/dts/xlp8xx-linux.dts
@@ -77,6 +77,9 @@
 	};
 
 	soc {
+		nodes {
+			num-nodes = <1>;
+		};
 
 	};
 };
diff --git a/arch/mips/boot/dts/xlp8xx-multi-os-hybrid-helinux.dts b/arch/mips/boot/dts/xlp8xx-multi-os-hybrid-helinux.dts
deleted file mode 100644
index 7f2def7..0000000
--- a/arch/mips/boot/dts/xlp8xx-multi-os-hybrid-helinux.dts
+++ /dev/null
@@ -1,187 +0,0 @@
-/* XLP8XX Device Tree Source
- *
- */
-
-/dts-v1/;
-/ {
-	model = "MIPS,XLP8XX";
-	compatible = "NETL,XLP8XX_A0";
-
-	#address-cells = <1>;
-	#size-cells = <1>;
-
-	hypervisor {
-		hypervisor-name = "Xen";
-		alloc_dom0_memory = <0>;
-		bootargs = "ncores=8 dom0_loadaddr=0x72000000 dom0_size=0x1c000000 dom0_cpumask=0x0f -- ";
-		domain_heap = <0xf0000000 0xc0000000>; /* 3584MB -> 6656MB */
-	};
-	doms {
-		#address-cells = <1>;
-		#size-cells = <1>;
-		dom@0 {
-			device_type = "domain";
-			os = "linux";
-
-			#address-cells = <1>;
-			#size-cells = <1>;
-
-			cpu {
-				onlinemask = <0xffffffff>;
-				nae-rx-vc = <0>;
-				nae-fb-vc = <1>;
-				napi-vc-mask = <0x3>;
-			};
-			uart {
-				id = <0>;
-				owner = <1>;
-				sharedcfg = <0x1f000000>;
-			};
-			memory {
-				reg = <0x01000000 0x0B000000
-					0x20000000 0x20000000>;  // 512M at 512M
-
-			};
-			pic {
-			};
-			nae {
-				owner = <0 0 0 0>;
-				freein-fifo-mask = <0x30000 0 0 0>;
-				sh-domains = <1>;
-				/* node offsets and size */
-				vfbtbl-sw-offset = <0 0 0 0>;
-				vfbtbl-sw-nentries = <32 32 32 32>;
-				vfbtbl-hw-offset = <0 0 0 0>;
-				vfbtbl-hw-nentries = <0 0 0 0>;
-			};
-			fmn {
-				node_0_vc_mask = <0x33333333 0x33333333 0x33333333 0x33333333>;
-				node_1_vc_mask = <0x33333333 0x33333333 0x33333333 0x33333333>;
-				node_2_vc_mask = <0x33333333 0x33333333 0x33333333 0x33333333>;
-				node_3_vc_mask = <0x33333333 0x33333333 0x33333333 0x33333333>;
-			};
-
-
-		};
-		dom@1 {
-			device_type = "domain";
-			os = "netos";
-
-			heap-size = <0x0 0x2000000>;		   // 32M
-			shared-mem = <0x0 0xe0000000 0x0 0x04000000>;  // 3584MB -> 3648 MB (64MB)
-
-			#address-cells = <1>;
-			#size-cells = <1>;
-
-			app-bootcmd = "smpload";
-			app-loadaddr = "0xf4000000"; // 3.75 GB + 64 MB
-
-			kernel {
-				/* Kernel Memory allocated from xen domain-heap */
-				size = <0x02000000>;  // 32M
-
-				/* Location in Dom0 ramdisk */
-				location = "/tmp/xen/netos.elf";
-
-				/* Domain destroy currently doesn't work with hyperexec */
-				dom_destroy_hack = <0>;
-			};
-
-			cpu {
-				onlinemask = <0xf0>;
-				nae-rx-vc = <2>;
-				nae-fb-vc = <3>;
-			};
-			uart {
-				id = <0>;
-				sharedcfg = <0x1f000000>;
-			};
-			memory {
-				reg = <0x50000000 0x10000000>;  // 1280M -> 1536M (256M)
-			};
-			nae {
-				owner = <1 0 0 0>;
-				freein-fifo-mask = <0x1111 0 0 0>;
-				sh-domains = <2>;
-				/* node offsets and size */
-				vfbtbl-sw-offset = <0 0 0 0>;
-				vfbtbl-sw-nentries = <0 0 0 0>;
-				vfbtbl-hw-offset = <32 32 32 32>;
-				vfbtbl-hw-nentries = <20 20 20 20>;
-			};
-		};
-		dom@2 {
-			device_type = "domain";
-			os = "netos";
-
-			heap-size = <0x0 0x2000000>;		   // 32M
-			shared-mem = <0x0 0xe4000000 0x0 0x04000000>;  // 3584MB -> 3648 MB (64MB)
-
-			#address-cells = <1>;
-			#size-cells = <1>;
-
-			app-bootcmd = "smpload";
-			app-loadaddr = "0xf4000000"; // 3.75 GB + 64 MB
-
-			kernel {
-				/* Kernel Memory allocated from xen domain-heap */
-				size = <0x02000000>;  // 32M
-
-				/* Location in Dom0 ramdisk */
-				location = "/tmp/xen/netos.elf";
-
-				/* Domain destroy currently doesn't work with hyperexec */
-				dom_destroy_hack = <0>;
-			};
-
-			cpu {
-				onlinemask = <0xf00>;
-				nae-rx-vc = <2>;
-				nae-fb-vc = <3>;
-			};
-			uart {
-				id = <0>;
-				sharedcfg = <0x1f000000>;
-			};
-			memory {
-				reg = <0x50000000 0x10000000>;  // 1280M -> 1536M (256M)
-			};
-			nae {
-				owner = <0 0 0 0>;
-				freein-fifo-mask = <0x2222 0 0 0>;
-				sh-domains = <1>;
-				/* node offsets and size */
-				vfbtbl-sw-offset = <0 0 0 0>;
-				vfbtbl-sw-nentries = <0 0 0 0>;
-				vfbtbl-hw-offset = <32 32 32 32>;
-				vfbtbl-hw-nentries = <20 20 20 20>;
-			};
-		};
-
-	};
-
-	chosen {
-		bootargs = " xlp_noi2c console=ttyS0,115200 ";
-	};
-
-	/* These binaries are downloaded at the resp physical memory locations
-	 * by their corresponding bootstrapping loaders. For example,
-	 * u-boot is loaded by x-load at 193M, sysconfig is loaded by u-boot
-	 * at 1M, hypervisor by u-boot hypervisor at 136M.
-	 * On simulator all these binaries are pre-loaded by using '-F' option
-	 * for convenience. On silicon, they will be loaded by their corresponding
-	 * temporary download locations the stage-1/stage-2 firmware
-	 */
-	firmware-download {
-		u-boot	= <0x0c100000 0x3f00000>;  /* 63M @ 193M */
-		sysconfig     = <0x00100000 0x00100000>; /* 1M @ 1M */
-		hypervisor    = <0x08800000 0x00800000>; /* 8M @ 136M */
-		dom0	  = <0x09000000 0x03000000>; /* 48M @ 144M */
-		domU-ramdisk  = <0x60000000 0x60800000>; /* 8M @ 1536M */
-	};
-
-	soc {
-
-	};
-};
-
diff --git a/arch/mips/boot/dts/xlp8xx-multi-os-hybrid.dts b/arch/mips/boot/dts/xlp8xx-multi-os-hybrid.dts
index 4d8681a..6b8e500 100644
--- a/arch/mips/boot/dts/xlp8xx-multi-os-hybrid.dts
+++ b/arch/mips/boot/dts/xlp8xx-multi-os-hybrid.dts
@@ -90,7 +90,7 @@
 			cpu {
 				onlinemask = <0xf0>;
 				nae-rx-vc = <2>;
-				nae-fb-vc = <3>;
+				nae-fb-vc = <2>;
 			};
 			uart {
 				id = <0>;
@@ -111,8 +111,8 @@
 			};
 			app-param{
 				/* Only used in xen loading */
-				linux-mem = <	0x00000000 0x02000000	// START ADDR <MSB LSB> 16MB
-						0x00000000 0x3e000000>;	// SIZE <MSB LSB> 1008 MB
+				linux-mem = <	0x00000000 0x02000000	// START ADDR <MSB LSB> 32MB
+						0x00000000 0x3e000000>;	// SIZE <MSB LSB> 992 MB
 			};
 		};
 	};
@@ -138,6 +138,9 @@
 	};
 
 	soc {
+		nodes {
+			num-nodes = <1>;
+		};
 
 	};
 };
diff --git a/arch/mips/boot/dts/xlp8xx-multi-os.dts b/arch/mips/boot/dts/xlp8xx-multi-os.dts
index c181dcd..2e13d7f 100644
--- a/arch/mips/boot/dts/xlp8xx-multi-os.dts
+++ b/arch/mips/boot/dts/xlp8xx-multi-os.dts
@@ -234,6 +234,9 @@
     };
 
     soc {
+		nodes {
+			num-nodes = <1>;
+		};
 
     };
 };
diff --git a/arch/mips/boot/dts/xlp8xx-netos.dts b/arch/mips/boot/dts/xlp8xx-netos.dts
index 8d7c531..2822c7b 100644
--- a/arch/mips/boot/dts/xlp8xx-netos.dts
+++ b/arch/mips/boot/dts/xlp8xx-netos.dts
@@ -122,6 +122,9 @@
 	};
 
 	soc {
+		nodes {
+			num-nodes = <1>;
+		};
 
 	};
 };
-- 
1.7.1

