From e75f5c03be3467068e5ef53c632b802c85e2a6e1 Mon Sep 17 00:00:00 2001
From: reshmic <reshmic@broadcom.com>
Date: Tue, 3 Apr 2012 17:44:36 +0530
Subject: Added sae-rx-vc sae-rx-sync-vc in xlp3xx-helinux.dts

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/boot/dts/xlp3xx-helinux.dts b/arch/mips/boot/dts/xlp3xx-helinux.dts
index d5860fd..f69a439 100644
--- a/arch/mips/boot/dts/xlp3xx-helinux.dts
+++ b/arch/mips/boot/dts/xlp3xx-helinux.dts
@@ -30,6 +30,8 @@
 				onlinemask = <0xffff>;
                                 nae-rx-vc = <1>;
                                 nae-fb-vc = <2>;
+                                sae-rx-vc = <1>; /*  */
+                                sae-rx-sync-vc  = <3>;
                                 napi-vc-mask = <0x6>;
 
 			};
@@ -82,6 +84,8 @@
                                 nae-fb-vc = <3>;
                                 /* Setting this to '1' makes NAE driver to skip buf alloction and initial
                                    sending of packet buffers*/
+                                sae-rx-vc = <0>; /* shared with nae; Has to be exclusive of linux dom nae vc */
+                                sae-rx-sync-vc  = <3>; /* shared with nae; Has to be exclusive of linux dom nae vc */
                                 nae-buf-alloc-by-app = <1>; 
 			};
 			uart {
-- 
1.7.1

