From e7d84ebd98eb31768638064352e7ad69c09e19bf Mon Sep 17 00:00:00 2001
From: P. Sadik <psadik@broadcom.com>
Date: Fri, 27 Apr 2012 14:06:44 +0530
Subject: Remove xlp_noi2c from Linux kernel command line.

This parameter obsolete.
[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/boot/dts/mnxlp8xx-hybrid-nae-helinux.dts b/arch/mips/boot/dts/mnxlp8xx-hybrid-nae-helinux.dts
index 889d7e5..2a90979 100644
--- a/arch/mips/boot/dts/mnxlp8xx-hybrid-nae-helinux.dts
+++ b/arch/mips/boot/dts/mnxlp8xx-hybrid-nae-helinux.dts
@@ -167,7 +167,7 @@
 	};
 
 	chosen {
-		bootargs = " xlp_noi2c console=ttyS0,115200 mem=512m@3584m";
+		bootargs = " console=ttyS0,115200 mem=512m@3584m";
 	};
 
 	/* These binaries are downloaded at the resp physical memory locations
diff --git a/arch/mips/boot/dts/mnxlp8xx-linux.dts b/arch/mips/boot/dts/mnxlp8xx-linux.dts
index b84f86e..242d4d8 100644
--- a/arch/mips/boot/dts/mnxlp8xx-linux.dts
+++ b/arch/mips/boot/dts/mnxlp8xx-linux.dts
@@ -72,7 +72,7 @@
 	};
 
 	chosen {
-		bootargs = " xlp_noi2c console=ttyS0,115200  mem=512m@3584m";
+		bootargs = " console=ttyS0,115200  mem=512m@3584m";
 	};
 
 	/* These binaries are downloaded at the resp physical memory locations
diff --git a/arch/mips/boot/dts/netos-srtp.dts b/arch/mips/boot/dts/netos-srtp.dts
index 3f3d454..0840c01 100644
--- a/arch/mips/boot/dts/netos-srtp.dts
+++ b/arch/mips/boot/dts/netos-srtp.dts
@@ -71,7 +71,7 @@
 	};
 
 	chosen {
-		bootargs = " xlp_noi2c console=ttyS0,115200 ";
+		bootargs = " console=ttyS0,115200 ";
 	};
 
 	soc {
diff --git a/arch/mips/boot/dts/xlp3xx-helinux.dts b/arch/mips/boot/dts/xlp3xx-helinux.dts
index f69a439..8e79584 100644
--- a/arch/mips/boot/dts/xlp3xx-helinux.dts
+++ b/arch/mips/boot/dts/xlp3xx-helinux.dts
@@ -110,7 +110,7 @@
 	};
 
 	chosen {
-		bootargs = " xlp_noi2c console=ttyS0,115200  ";
+		bootargs = " console=ttyS0,115200  ";
 	};
 
 	/* These binaries are downloaded at the resp physical memory locations
diff --git a/arch/mips/boot/dts/xlp3xx-linux-loader.dts b/arch/mips/boot/dts/xlp3xx-linux-loader.dts
index 3285b27..f66061a 100644
--- a/arch/mips/boot/dts/xlp3xx-linux-loader.dts
+++ b/arch/mips/boot/dts/xlp3xx-linux-loader.dts
@@ -92,7 +92,7 @@
 	}; 
 
 	chosen {
-		bootargs = " xlp_noi2c console=ttyS0,115200 ";
+		bootargs = " console=ttyS0,115200 ";
 	};
 
 	/* These binaries are downloaded at the resp physical memory locations
diff --git a/arch/mips/boot/dts/xlp3xx-linux.dts b/arch/mips/boot/dts/xlp3xx-linux.dts
index ff3a73a..bcb1838 100644
--- a/arch/mips/boot/dts/xlp3xx-linux.dts
+++ b/arch/mips/boot/dts/xlp3xx-linux.dts
@@ -70,7 +70,7 @@
 	};
 
 	chosen {
-		bootargs = " xlp_noi2c console=ttyS0,115200 mem=512m@3584m rhash_entries=1048576";
+		bootargs = " console=ttyS0,115200 mem=512m@3584m rhash_entries=1048576";
 	};
 
 	/* These binaries are downloaded at the resp physical memory locations
diff --git a/arch/mips/boot/dts/xlp3xx-multi-os.dts b/arch/mips/boot/dts/xlp3xx-multi-os.dts
index b92fd87..dec4357 100644
--- a/arch/mips/boot/dts/xlp3xx-multi-os.dts
+++ b/arch/mips/boot/dts/xlp3xx-multi-os.dts
@@ -168,7 +168,7 @@
 	};
 
 	chosen {
-		bootargs = " xlp_noi2c console=ttyS0,115200 ";
+		bootargs = " console=ttyS0,115200 ";
 	};
 
 	/* These binaries are downloaded at the resp physical memory locations
diff --git a/arch/mips/boot/dts/xlp3xx-netos.dts b/arch/mips/boot/dts/xlp3xx-netos.dts
index 2c2e0b9..65f4b2d 100644
--- a/arch/mips/boot/dts/xlp3xx-netos.dts
+++ b/arch/mips/boot/dts/xlp3xx-netos.dts
@@ -80,7 +80,7 @@
 	};
 
 	chosen {
-		bootargs = " xlp_noi2c console=ttyS0,115200 ";
+		bootargs = " console=ttyS0,115200 ";
 	};
 
 	soc {
diff --git a/arch/mips/boot/dts/xlp3xx-tiny.dts b/arch/mips/boot/dts/xlp3xx-tiny.dts
index 4eeee63..3cf00c4 100644
--- a/arch/mips/boot/dts/xlp3xx-tiny.dts
+++ b/arch/mips/boot/dts/xlp3xx-tiny.dts
@@ -1,4 +1,4 @@
-/* XLP8XX Device Tree Source 
+/* XLP8XX Device Tree Source
  *
  */
 
@@ -35,7 +35,7 @@
 	};
 
 	chosen {
-		bootargs = "xlp_noi2c console=ttyS0,115200 ";
+		bootargs = " console=ttyS0,115200 ";
 	};
 
 	soc {
@@ -56,10 +56,10 @@
 	                     0    0    1   1   2   2   3   3
         	        >;
 	        };
-	
+
         	// Packet Ordering Engine (POE)
 	        poe {
-	
+
         	        mode = "bypass";
 
                 	// Each vector is 512 bit with masb indicating vc 512 and lsb indicating vc 0
@@ -89,13 +89,13 @@
 	        complex@0 {
         	        device_type = "nae-complex";
                 	mode = "sgmii";
-	
+
         	        sgmii {
                 	        loopback = <0 0 0 0>;
 
                         	// Max ingress fifo size 256 units (size of one unit is 64 byte)
 	                        iface-fifo-size = <13 13 13 13>;
-	
+
         	                ext-phy-addr = <4 7 6 5>;
                 	        ext-phy-bus = <0 0 0 0>;
 
diff --git a/arch/mips/boot/dts/xlp4xx-helinux.dts b/arch/mips/boot/dts/xlp4xx-helinux.dts
index 87330f8..83f5ad1 100644
--- a/arch/mips/boot/dts/xlp4xx-helinux.dts
+++ b/arch/mips/boot/dts/xlp4xx-helinux.dts
@@ -166,7 +166,7 @@
 	};
 
 	chosen {
-		bootargs = " xlp_noi2c console=ttyS0,115200  ";
+		bootargs = " console=ttyS0,115200  ";
 	};
 
 	/* These binaries are downloaded at the resp physical memory locations
diff --git a/arch/mips/boot/dts/xlp4xx-linux.dts b/arch/mips/boot/dts/xlp4xx-linux.dts
index 89209bb..c67bb2a 100644
--- a/arch/mips/boot/dts/xlp4xx-linux.dts
+++ b/arch/mips/boot/dts/xlp4xx-linux.dts
@@ -69,7 +69,7 @@
 	};
 
 	chosen {
-		bootargs = " xlp_noi2c console=ttyS0,115200 rhash_entries=1048576";
+		bootargs = " console=ttyS0,115200 rhash_entries=1048576";
 	};
 
 	/* These binaries are downloaded at the resp physical memory locations
diff --git a/arch/mips/boot/dts/xlp4xx-multi-os.dts b/arch/mips/boot/dts/xlp4xx-multi-os.dts
index 29fdd18..0dc355e 100644
--- a/arch/mips/boot/dts/xlp4xx-multi-os.dts
+++ b/arch/mips/boot/dts/xlp4xx-multi-os.dts
@@ -168,7 +168,7 @@
 	};
 
 	chosen {
-		bootargs = " xlp_noi2c console=ttyS0,115200 ";
+		bootargs = " console=ttyS0,115200 ";
 	};
 
 	/* These binaries are downloaded at the resp physical memory locations
diff --git a/arch/mips/boot/dts/xlp4xx-netos.dts b/arch/mips/boot/dts/xlp4xx-netos.dts
index 35dcf62..801a854 100644
--- a/arch/mips/boot/dts/xlp4xx-netos.dts
+++ b/arch/mips/boot/dts/xlp4xx-netos.dts
@@ -79,7 +79,7 @@
 	};
 
 	chosen {
-		bootargs = " xlp_noi2c console=ttyS0,115200 ";
+		bootargs = " console=ttyS0,115200 ";
 	};
 
 	soc {
diff --git a/arch/mips/boot/dts/xlp4xx-tiny.dts b/arch/mips/boot/dts/xlp4xx-tiny.dts
index 4de22a7..aaca571 100644
--- a/arch/mips/boot/dts/xlp4xx-tiny.dts
+++ b/arch/mips/boot/dts/xlp4xx-tiny.dts
@@ -1,4 +1,4 @@
-/* XLP8XX Device Tree Source 
+/* XLP8XX Device Tree Source
  *
  */
 
@@ -35,7 +35,7 @@
 	};
 
 	chosen {
-		bootargs = "xlp_noi2c console=ttyS0,115200 ";
+		bootargs = " console=ttyS0,115200 ";
 	};
 
 	soc {
diff --git a/arch/mips/boot/dts/xlp8xx-helinux.dts b/arch/mips/boot/dts/xlp8xx-helinux.dts
index 0aea1dc..a0115d0 100644
--- a/arch/mips/boot/dts/xlp8xx-helinux.dts
+++ b/arch/mips/boot/dts/xlp8xx-helinux.dts
@@ -206,7 +206,7 @@
 	};
 
 	chosen {
-		bootargs = " xlp_noi2c console=ttyS0,115200  ";
+		bootargs = " console=ttyS0,115200  ";
 	};
 
 	/* These binaries are downloaded at the resp physical memory locations
diff --git a/arch/mips/boot/dts/xlp8xx-hybrid-nae-netos.dts b/arch/mips/boot/dts/xlp8xx-hybrid-nae-netos.dts
index 717ed9c..2a3ee2e 100644
--- a/arch/mips/boot/dts/xlp8xx-hybrid-nae-netos.dts
+++ b/arch/mips/boot/dts/xlp8xx-hybrid-nae-netos.dts
@@ -114,7 +114,7 @@
 	};
 
 	chosen {
-		bootargs = " xlp_noi2c console=ttyS0,115200 ";
+		bootargs = " console=ttyS0,115200 ";
 	};
 
 	/* These binaries are downloaded at the resp physical memory locations
diff --git a/arch/mips/boot/dts/xlp8xx-linux-loader.dts b/arch/mips/boot/dts/xlp8xx-linux-loader.dts
index e910c54..40985c9 100644
--- a/arch/mips/boot/dts/xlp8xx-linux-loader.dts
+++ b/arch/mips/boot/dts/xlp8xx-linux-loader.dts
@@ -176,7 +176,7 @@
 	};
 
 	chosen {
-		bootargs = " xlp_noi2c console=ttyS0,115200 ";
+		bootargs = " console=ttyS0,115200 ";
 	};
 
 	/* These binaries are downloaded at the resp physical memory locations
diff --git a/arch/mips/boot/dts/xlp8xx-linux.dts b/arch/mips/boot/dts/xlp8xx-linux.dts
index 8999d22..cc4bb4c 100644
--- a/arch/mips/boot/dts/xlp8xx-linux.dts
+++ b/arch/mips/boot/dts/xlp8xx-linux.dts
@@ -69,7 +69,7 @@
 		};
 	};
 	chosen {
-		bootargs = " xlp_noi2c console=ttyS0,115200 mem=512m@3584m rhash_entries=1048576";
+		bootargs = " console=ttyS0,115200 mem=512m@3584m rhash_entries=1048576";
 	};
 
 	/* These binaries are downloaded at the resp physical memory locations
diff --git a/arch/mips/boot/dts/xlp8xx-multi-os.dts b/arch/mips/boot/dts/xlp8xx-multi-os.dts
index 2e13d7f..8169c35 100644
--- a/arch/mips/boot/dts/xlp8xx-multi-os.dts
+++ b/arch/mips/boot/dts/xlp8xx-multi-os.dts
@@ -214,7 +214,7 @@
     };
 
     chosen {
-        bootargs = " xlp_noi2c console=ttyS0,115200 ";
+        bootargs = " console=ttyS0,115200 ";
     };
 
     /* These binaries are downloaded at the resp physical memory locations
diff --git a/arch/mips/boot/dts/xlp8xx-netos.dts b/arch/mips/boot/dts/xlp8xx-netos.dts
index 0616cbe..9cffd39 100644
--- a/arch/mips/boot/dts/xlp8xx-netos.dts
+++ b/arch/mips/boot/dts/xlp8xx-netos.dts
@@ -119,7 +119,7 @@
 	};
 
 	chosen {
-		bootargs = " xlp_noi2c console=ttyS0,115200 ";
+		bootargs = " console=ttyS0,115200 ";
 	};
 
 	soc {
diff --git a/arch/mips/boot/dts/xlp8xx-tiny.dts b/arch/mips/boot/dts/xlp8xx-tiny.dts
index 4de22a7..aaca571 100644
--- a/arch/mips/boot/dts/xlp8xx-tiny.dts
+++ b/arch/mips/boot/dts/xlp8xx-tiny.dts
@@ -1,4 +1,4 @@
-/* XLP8XX Device Tree Source 
+/* XLP8XX Device Tree Source
  *
  */
 
@@ -35,7 +35,7 @@
 	};
 
 	chosen {
-		bootargs = "xlp_noi2c console=ttyS0,115200 ";
+		bootargs = " console=ttyS0,115200 ";
 	};
 
 	soc {
-- 
1.7.1

