From 65c4f3e811d0875c806a9438ce70c01622ad7b69 Mon Sep 17 00:00:00 2001
From: ShengLiang Song <shengliang.song@broadcom.com>
Date: Mon, 2 Jul 2012 12:23:42 -0700
Subject: xlp2xx: initial support

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/boot/dts/xlp2xx-helinux.dts b/arch/mips/boot/dts/xlp2xx-helinux.dts
new file mode 100644
index 0000000..3a8eca1
--- /dev/null
+++ b/arch/mips/boot/dts/xlp2xx-helinux.dts
@@ -0,0 +1,139 @@
+/* XLP2XX Device Tree Source
+ *
+ */
+
+/dts-v1/;
+/ {
+	model = "MIPS,XLP2XX";
+	compatible = "NETL,XLP2XX_A0";
+
+	#address-cells = <1>;
+	#size-cells = <1>;
+
+	hypervisor {
+		hypervisor-name = "Xen";
+		alloc_dom0_memory = <0>;
+		bootargs = "ncores=4 dom0_loadaddr=0x72000000 dom0_size=0x1c000000 dom0_cpumask=0xffff -- ";
+		domain_heap = <0x80000000 0x20000000>;
+	};
+	doms {
+		#address-cells = <1>;
+		#size-cells = <1>;
+		dom@0 {
+			device_type = "domain";
+			os = "linux";
+
+			#address-cells = <1>;
+			#size-cells = <1>;
+
+			cpu {
+				onlinemask = <0xffff>;
+                                nae-rx-vc = <1>;
+                                nae-fb-vc = <2>;
+                                sae-rx-vc = <1>; /*  */
+                                sae-rx-sync-vc  = <3>;
+                                napi-vc-mask = <0x6>;
+
+			};
+			uart {
+				id = <0>;
+				owner = <1>;
+				sharedcfg = <0x1f000000>;
+			};
+			memory {
+				reg = <0x01000000 0x0B000000	// 176M at 16M
+					0xE0000000 0x20000000>;  // 512M @ 3.5GB
+			};
+                        fmn {
+                                node_0_vc_mask = <0x66666666 0x66666666 0x66666666 0x66666666>;
+                                node_1_vc_mask = <0x66666666 0x66666666 0x66666666 0x66666666>;
+                                node_2_vc_mask = <0x66666666 0x66666666 0x66666666 0x66666666>;
+                                node_3_vc_mask = <0x66666666 0x66666666 0x66666666 0x66666666>;
+                        };
+			pic {
+			};
+		};
+
+		dom@1 {
+			device_type = "domain";
+			os = "netos";
+
+			heap-size = <0x0 0x8000000>; // 128M
+			shared-mem = <0x0 0x40000000 0x0 0x10000000>; // 128M @ 1GB
+
+			#address-cells = <1>;
+			#size-cells = <1>;
+
+			app-bootcmd = "smpload";
+			app-loadaddr = "0xf0000000"; // 3.75GB
+
+			kernel {
+				/* Kernel Memory allocated from xen domain-heap */
+				size = <0x02000000>;  // 32M
+
+				/* Location in Dom0 ramdisk */
+				location = "/tmp/xen/netos.elf";
+
+				/* Domain destroy currently doesn't work with hyperexec */
+				dom_destroy_hack = <0>;
+			};
+
+			cpu {
+				onlinemask = <0xffff>;
+                                nae-rx-vc = <0>;
+                                nae-fb-vc = <3>;
+                                /* Setting this to '1' makes NAE driver to skip buf alloction and initial
+                                   sending of packet buffers*/
+                                sae-rx-vc = <0>; /* shared with nae; Has to be exclusive of linux dom nae vc */
+                                sae-rx-sync-vc  = <3>; /* shared with nae; Has to be exclusive of linux dom nae vc */
+                                nae-buf-alloc-by-app = <1>; 
+			};
+			uart {
+				id = <0>;
+			};
+			memory {
+				reg = <0x60000000 0x60000000>;  // 1.5GB @ 1.5GB
+			};
+			app-param{
+				onlinemask = <0xffff>;
+			};
+			nae{
+				owner = <1>;
+			};
+                        fmn {
+                                node_0_vc_mask = <0x99999999 0x99999999 0x99999999 0x99999999>;
+                                node_1_vc_mask = <0x99999999 0x99999999 0x99999999 0x99999999>;
+                                node_2_vc_mask = <0x99999999 0x99999999 0x99999999 0x99999999>;
+                                node_3_vc_mask = <0x99999999 0x99999999 0x99999999 0x99999999>;
+                        };
+		};
+	};
+
+	chosen {
+		bootargs = " console=ttyS0,115200  ";
+	};
+
+	/* These binaries are downloaded at the resp physical memory locations
+	 * by their corresponding bootstrapping loaders. For example,
+	 * u-boot is loaded by x-load at 193M, sysconfig is loaded by u-boot
+	 * at 1M, hypervisor by u-boot hypervisor at 136M.
+	 * On simulator all these binaries are pre-loaded by using '-F' option
+	 * for convenience. On silicon, they will be loaded by their corresponding
+	 * temporary download locations the stage-1/stage-2 firmware
+	 */
+	firmware-download {
+		u-boot        = <0x0c100000 0x3f00000>;  /* 63M @ 193M */
+		sysconfig     = <0x00100000 0x00100000>; /* 1M @ 1M */
+		hypervisor    = <0x08800000 0x00800000>; /* 8M @ 136M */
+		dom0          = <0x09000000 0x03000000>; /* 48M @ 144M */
+		domU-ramdisk  = <0x60000000 0x60800000>; /* 8M @ 1536M */
+	};
+
+	soc {
+		nodes {
+			num-nodes = <1>;
+		};
+
+	};
+};
+
diff --git a/arch/mips/boot/dts/xlp2xx-linux-loader.dts b/arch/mips/boot/dts/xlp2xx-linux-loader.dts
new file mode 100644
index 0000000..1087f1a
--- /dev/null
+++ b/arch/mips/boot/dts/xlp2xx-linux-loader.dts
@@ -0,0 +1,121 @@
+/* XLP2XX Device Tree Source
+ *
+ */
+
+/dts-v1/;
+/ {
+	model = "MIPS,XLP2XX";
+	compatible = "NETL,XLP2XX_A0";
+
+	#address-cells = <1>;
+	#size-cells = <1>;
+
+	hypervisor {
+		hypervisor-name = "Xen";
+		alloc_dom0_memory = <0>;
+		bootargs = "ncores=8 dom0_loadaddr=0x72000000 dom0_size=0x1c000000 dom0_cpumask=0x0f -- ";
+		domain_heap = <0x30000000 0x20000000>; /* 0.75 GB -> 1.25 GB */
+	};
+	doms {
+		#address-cells = <1>;
+		#size-cells = <1>;
+		dom@0 {
+			device_type = "domain";
+			os = "linux";
+
+			#address-cells = <1>;
+			#size-cells = <1>;
+
+			cpu {
+				onlinemask = <0x0f>;
+                                nae-rx-vc = <0>;
+                                nae-fb-vc = <1>;
+				napi-vc-mask = <0x3>;
+			};
+			uart {
+				id = <0>;
+				owner = <1>;
+				sharedcfg = <0x1f000000>;
+			};
+			memory {
+				reg = <	0x01000000 0x0B000000	// 176M at 16M
+					0x20000000 0x30000000>; // 768M at 512M
+			};
+			linuxloader{
+				loadermask = <0xfff0>;
+                                memory = <0x00000000 0x60000000
+                                          0x00000000 0x40000000>; //1024M at 1536M
+			};
+			pic {
+			};
+		};
+		dom@1 {
+			device_type = "domain";
+			os = "netos";
+
+			heap-size = <0x0 0x2000000>;                   // 32M
+                        shared-mem = <0x0 0xf0000000 0x0 0x04000000>;  // 3840MB -> 3904 MB (64MB)
+
+			#address-cells = <1>;
+			#size-cells = <1>;
+
+			app-bootcmd = "ampload";
+			app-loadaddr = "0xf4000000"; // 3904M + 64 MB
+
+			kernel {
+				/* Kernel Memory allocated from xen domain-heap */
+				size = <0x02000000>;  // 32M
+
+				/* Location in Dom0 ramdisk */
+				location = "/tmp/xen/netos.elf";
+
+				/* Domain destroy currently doesn't work with hyperexec */
+				dom_destroy_hack = <0>;
+			};
+
+			cpu {
+				onlinemask = <0xf0>;
+                                nae-rx-vc = <0>;
+                                nae-fb-vc = <1>;
+			};
+			uart {
+				id = <0>;
+				sharedcfg = <0x1f000000>;
+			};
+			memory {
+				reg = <0x50000000 0x10000000>;  // 1280M -> 1536M (256M)
+			};
+			nae{
+				owner = <0>;
+			};
+		};
+	}; 
+
+	chosen {
+		bootargs = " console=ttyS0,115200 ";
+	};
+
+	/* These binaries are downloaded at the resp physical memory locations
+	 * by their corresponding bootstrapping loaders. For example,
+	 * u-boot is loaded by x-load at 193M, sysconfig is loaded by u-boot
+	 * at 1M, hypervisor by u-boot hypervisor at 136M.
+	 * On simulator all these binaries are pre-loaded by using '-F' option
+	 * for convenience. On silicon, they will be loaded by their corresponding
+	 * temporary download locations the stage-1/stage-2 firmware
+	 */
+	firmware-download {
+		u-boot        = <0x0c100000 0x3f00000>;  /* 63M @ 193M */
+		sysconfig     = <0x00100000 0x00100000>; /* 1M @ 1M */
+		hypervisor    = <0x08800000 0x00800000>; /* 8M @ 136M */
+		dom0          = <0x09000000 0x03000000>; /* 48M @ 144M */
+		domU-ramdisk  = <0x60000000 0x60800000>; /* 8M @ 1536M */
+	};
+
+	soc {
+		nodes {
+			num-nodes = <1>;
+		};
+
+	};
+};
+
diff --git a/arch/mips/boot/dts/xlp2xx-linux.dts b/arch/mips/boot/dts/xlp2xx-linux.dts
new file mode 100644
index 0000000..26b0604
--- /dev/null
+++ b/arch/mips/boot/dts/xlp2xx-linux.dts
@@ -0,0 +1,117 @@
+/* XLP2XX Device Tree Source
+ *
+ */
+
+/dts-v1/;
+/ {
+	model = "MIPS,XLP2XX";
+	compatible = "NETL,XLP2XX_A0";
+
+	#address-cells = <1>;
+	#size-cells = <1>;
+
+	hypervisor {
+		hypervisor-name = "Xen";
+		alloc_dom0_memory = <0>;
+		bootargs = "ncores=4 dom0_loadaddr=0x72000000 dom0_size=0x1c000000 dom0_cpumask=0xffff -- ";
+		domain_heap = <0x80000000 0x20000000>;
+	};
+	doms {
+		#address-cells = <1>;
+		#size-cells = <1>;
+		dom@0 {
+			device_type = "domain";
+			os = "linux";
+
+			#address-cells = <1>;
+			#size-cells = <1>;
+
+			cpu {
+				onlinemask = <0xffff>;
+                                nae-rx-vc = <0>;
+                                nae-fb-vc = <1>;
+				napi-vc-mask = <0xb>;
+				sae-rx-vc = <0>; /*  */
+				sae-rx-sync-vc  = <3>; /* should be exclusive */
+				ipsec-async-vc  = <3>;
+				ipsec-sync-vc   = <2>; /* should be exclusive either  
+							  ipsec-sync-vc  or soc/srio@node-0  */
+
+			};
+			uart {
+				id = <0>;
+				owner = <1>;
+				sharedcfg = <0x1f000000>;
+			};
+			memory {
+				reg = <0x01000000 0x0B000000	// 176M at 16M
+					0x20000000 0xa0000000>;  // 2586M at 512M
+			};
+                        fmn {
+				/* make the node mask to 0xffffffff for SRIO */  
+                                node_0_vc_mask = <0xbbbbbbbb 0xbbbbbbbb 0xbbbbbbbb 0xbbbbbbbb>;
+                                node_1_vc_mask = <0xbbbbbbbb 0xbbbbbbbb 0xbbbbbbbb 0xbbbbbbbb>;
+                                node_2_vc_mask = <0xbbbbbbbb 0xbbbbbbbb 0xbbbbbbbb 0xbbbbbbbb>;
+                                node_3_vc_mask = <0xbbbbbbbb 0xbbbbbbbb 0xbbbbbbbb 0xbbbbbbbb>;
+                        };
+			pic {
+			};
+			nae {
+				owner = <1 0 0 0>;
+				freein-fifo-mask = <0xff 0 0 0>;
+				/* node offsets and size */
+				vfbtbl-sw-offset = <0 0 0 0>;
+				vfbtbl-sw-nentries = <32 32 32 32>;
+				/* hw entries */
+				vfbtbl-hw-offset = <32 32 32 32>;
+				vfbtbl-hw-nentries = <20 20 20 20>;
+			};
+
+
+		};
+	};
+
+	chosen {
+		bootargs = " console=ttyS0,115200 mem=512m@3584m rhash_entries=1048576";
+	};
+
+	/* These binaries are downloaded at the resp physical memory locations
+	 * by their corresponding bootstrapping loaders. For example,
+	 * u-boot is loaded by x-load at 193M, sysconfig is loaded by u-boot
+	 * at 1M, hypervisor by u-boot hypervisor at 136M.
+	 * On simulator all these binaries are pre-loaded by using '-F' option
+	 * for convenience. On silicon, they will be loaded by their corresponding
+	 * temporary download locations the stage-1/stage-2 firmware
+	 */
+	firmware-download {
+		u-boot        = <0x0c100000 0x3f00000>;  /* 63M @ 193M */
+		sysconfig     = <0x00100000 0x00100000>; /* 1M @ 1M */
+		hypervisor    = <0x08800000 0x00800000>; /* 8M @ 136M */
+		dom0          = <0x09000000 0x03000000>; /* 48M @ 144M */
+		domU-ramdisk  = <0x60000000 0x60800000>; /* 8M @ 1536M */
+	};
+
+	soc {
+		nodes {
+			num-nodes = <1>;
+		};
+		srio@node-0{
+                       /* type = "master" or "device". master is responsible for enumeration*/
+                       type = "master";
+                       /* SRIO_MODE_x1 = 1(not supported), SRIO_MODE_x4 = 2 */
+                       mode = <2>;
+                       /* BAUD_1250M = 1 (1.25G)
+                          BAUD_2500M = 2 (25.G)
+                          BAUD_3125M = 3 (3.125G)
+                          BAUD_5000M = 4 (5G)
+                          BAUD_6250M = 5 (6.25G) */
+                       baud-rate = <5>;
+                       rxvc = <2>;
+                       msgdstid = <2>;
+                       /* sys-size = 1 for 16bit device id, 0 for 8 bit device id */
+                       sys-size = <1>;
+               };
+
+	};
+};
+
diff --git a/arch/mips/boot/dts/xlp2xx-multi-os.dts b/arch/mips/boot/dts/xlp2xx-multi-os.dts
new file mode 100644
index 0000000..63ff66b
--- /dev/null
+++ b/arch/mips/boot/dts/xlp2xx-multi-os.dts
@@ -0,0 +1,197 @@
+/* XLP2XX Device Tree Source
+ *
+ */
+
+/dts-v1/;
+/ {
+	model = "MIPS,XLP2XX";
+	compatible = "NETL,XLP2XX_A0";
+
+	#address-cells = <1>;
+	#size-cells = <1>;
+
+	hypervisor {
+		hypervisor-name = "Xen";
+		alloc_dom0_memory = <0>;
+		bootargs = "ncores=4 dom0_loadaddr=0x72000000 dom0_size=0x1c000000 dom0_cpumask=0x0f -- ";
+		domain_heap = <0x4e800000 0x10000000>; /* 1256MB -> 1512MB */
+	};
+	doms {
+		#address-cells = <1>;
+		#size-cells = <1>;
+		dom@0 {
+			device_type = "domain";
+			os = "linux";
+
+			#address-cells = <1>;
+			#size-cells = <1>;
+
+			cpu {
+				onlinemask = <0x0f>;
+                                nae-rx-vc = <0>;
+                                nae-fb-vc = <1>;
+				napi-vc-mask = <0x3>;
+			};
+			uart {
+				id = <0>;
+				owner = <1>;
+				sharedcfg = <0x1f000000>;
+			};
+			memory {
+				reg = < 0x01000000 0x0F000000    // 240M at 16M
+					0x20000000 0x20000000>;  // 512M at 512M
+			};
+			pic {
+			};
+		};
+		dom@1 {
+			device_type = "domain";
+			os = "netos";
+
+			heap-size = <0x0 0x2000000>;                   // 32M
+			shared-mem = <0x0 0x62800000 0x0 0x02000000>;  // 1576MB -> 1608MB (32MB)
+
+			#address-cells = <1>;
+			#size-cells = <1>;
+
+			app-bootcmd = "ampload";
+			app-loadaddr = "0x82800000"; // 2088 MB
+
+			kernel {
+				/* Kernel Memory allocated from xen domain-heap */
+				size = <0x02000000>;  // 32M
+
+				/* Location in Dom0 ramdisk */
+				location = "/tmp/xen/netos.elf";
+
+				/* Domain destroy currently doesn't work with hyperexec */
+				dom_destroy_hack = <0>;
+			};
+
+			cpu {
+				onlinemask = <0xf0>;
+                                nae-rx-vc = <0>;
+                                nae-fb-vc = <1>;
+			};
+			uart {
+				id = <0>;
+				sharedcfg = <0x1f000000>;
+			};
+			memory {
+				reg = <0x5E800000 0x04000000>;  // 1512M -> 1576M (64M)
+			};
+			nae {
+				owner = <0>;
+			};
+		};
+		dom@2 {
+			device_type = "domain";
+			os = "netos";
+
+			heap-size = <0x0 0x2000000>;                   // 32M
+			shared-mem = <0x0 0x68800000 0x0 0x02000000>;  // 1672MB -> 1704MB (32MB)
+
+			#address-cells = <1>;
+			#size-cells = <1>;
+
+			app-bootcmd = "smpload";
+			app-loadaddr = "0x82800000"; // 2088 MB
+
+			kernel {
+				/* Kernel Memory allocated from xen domain-heap */
+				size = <0x02000000>;  // 32M
+
+				/* Location in Dom0 ramdisk */
+				location = "/tmp/xen/netos.elf";
+
+				/* Domain destroy currently doesn't work with hyperexec */
+				dom_destroy_hack = <0>;
+			};
+
+			cpu {
+				onlinemask = <0x0f00>;
+                                nae-rx-vc = <0>;
+                                nae-fb-vc = <1>;
+			};
+			uart {
+				id = <0>;
+				sharedcfg = <0x1f000000>;
+			};
+			memory {
+				reg = <0x64800000 0x04000000>;  // 1608M ->1672 M (64M) 
+			};
+			nae {
+				owner = <0>;
+			};
+		};
+	        dom@3 {
+                        device_type = "domain";
+                        os = "netos";
+
+                        heap-size = <0x0 0x2000000>;                   // 32M
+                        shared-mem = <0x0 0x7a800000 0x0 0x02000000>;  // 1960MB -> 1992MB (32MB)
+
+                        #address-cells = <1>;
+                        #size-cells = <1>;
+
+                        app-bootcmd = "ampload";
+                        app-loadaddr = "0x82800000"; // 2088Mb
+
+                        kernel {
+                                /* Kernel Memory allocated from xen domain-heap */
+                                size = <0x02000000>;  // 32M
+
+                                /* Location in Dom0 ramdisk */
+                                location = "/tmp/xen/netos.elf";
+
+                                /* Domain destroy currently doesn't work with hyperexec */
+                                dom_destroy_hack = <0>;
+                        };
+
+                        cpu {
+                                onlinemask = <0xf000>;
+                                nae-rx-vc = <0>;
+                                nae-fb-vc = <1>;
+                        };
+                        uart {
+                                id = <0>;
+                                sharedcfg = <0x1f000000>;
+                        };
+                        memory {
+                                reg = <0x6a800000 0x10000000>;  // 1704M -> 1960M (256MB)
+                        };
+                        nae {
+                                owner = <0>;
+                        };
+                };
+	
+	};
+
+	chosen {
+		bootargs = " console=ttyS0,115200 ";
+	};
+
+	/* These binaries are downloaded at the resp physical memory locations
+	 * by their corresponding bootstrapping loaders. For example,
+	 * u-boot is loaded by x-load at 193M, sysconfig is loaded by u-boot
+	 * at 1M, hypervisor by u-boot hypervisor at 136M.
+	 * On simulator all these binaries are pre-loaded by using '-F' option
+	 * for convenience. On silicon, they will be loaded by their corresponding
+	 * temporary download locations the stage-1/stage-2 firmware
+	 */
+	firmware-download {
+		u-boot        = <0x0c100000 0x3f00000>;  /* 63M @ 193M */
+		sysconfig     = <0x00100000 0x00100000>; /* 1M @ 1M */
+		hypervisor    = <0x08800000 0x00800000>; /* 8M @ 136M */
+		dom0          = <0x09000000 0x03000000>; /* 48M @ 144M */
+		domU-ramdisk  = <0x60000000 0x60800000>; /* 8M @ 1536M */
+	};
+
+	soc {
+		nodes {
+			num-nodes = <1>;
+		};
+
+	};
+};
+
diff --git a/arch/mips/boot/dts/xlp2xx-netos.dts b/arch/mips/boot/dts/xlp2xx-netos.dts
new file mode 100644
index 0000000..f5e3108
--- /dev/null
+++ b/arch/mips/boot/dts/xlp2xx-netos.dts
@@ -0,0 +1,92 @@
+/* XLP2XX Device Tree Source 
+ *
+ */
+
+/dts-v1/;
+/ {
+	model = "MIPS,XLP2XX";
+	compatible = "NETL,XLP2XX_A0";
+
+	#address-cells = <1>;
+	#size-cells = <1>;
+
+	hypervisor {
+		hypervisor-name = "Xen";
+		alloc_dom0_memory = <1>;
+		bootargs = "dom0_loadaddr=0x72000000 dom0_cpumask=0x0f -- ";
+		domain_heap = <0x50000000 0x20000000>;
+	};
+
+	doms {
+		#address-cells = <1>;
+		#size-cells = <1>;
+		dom@0 {
+			device_type = "domain";
+			os = "netos";
+
+			heap-size = <0x0 0x2000000>; // 32M
+			shared-mem = <0x0 0xe0000000 0x0 0x10000000>; // 128M @ 3.5GB
+
+			#address-cells = <1>;
+			#size-cells = <1>;
+
+			app-bootcmd = "smpload";
+			app-loadaddr = "0xf0000000"; // 3.75GB
+
+			kernel {
+				/* Kernel Memory allocated from xen domain-heap */
+				size = <0x02000000>;  // 32M
+
+				/* Location in Dom0 ramdisk */
+				location = "/tmp/xen/netos.elf";
+
+				/* Domain destroy currently doesn't work with hyperexec */
+				dom_destroy_hack = <0>;
+			};
+
+			cpu {
+				onlinemask = <0xffff>;
+                                nae-rx-vc = <0>;
+                                nae-fb-vc = <3>;
+				sae-rx-vc = <0>; /* sharing with nae */
+				sae-rx-sync-vc = <2>; /* should be exclusive */
+
+
+                                /* Setting this to '1' makes NAE driver to skip buf alloction and initial
+                                   sending of packet buffers*/
+                                nae-buf-alloc-by-app = <1>; 
+			};
+			uart {
+				id = <0>;
+			};
+			memory {
+				reg = <0x20000000 0xa0000000>;  // 2560M @ 512M
+			};
+			app-param{
+                                onlinemask = <0xffff>;
+				enable-nae-jumbo = <0>;
+                        };
+			nae{
+				owner = <1>;
+			};
+			/* In hyperexec-linux mode, vc_mask needs to be synchronized with linux fmn node! */
+                        fmn {
+                                node_0_vc_mask = <0x99999999 0x99999999 0x99999999 0x99999999>;
+                                node_1_vc_mask = <0x99999999 0x99999999 0x99999999 0x99999999>;
+                                node_2_vc_mask = <0x99999999 0x99999999 0x99999999 0x99999999>;
+                                node_3_vc_mask = <0x99999999 0x99999999 0x99999999 0x99999999>;
+                        };
+		};
+	};
+
+	chosen {
+		bootargs = " console=ttyS0,115200 ";
+	};
+
+	soc {
+		nodes {
+			num-nodes = <1>;
+		};
+
+	};
+};
diff --git a/arch/mips/boot/dts/xlp2xx-tiny.dts b/arch/mips/boot/dts/xlp2xx-tiny.dts
new file mode 100644
index 0000000..822b72c
--- /dev/null
+++ b/arch/mips/boot/dts/xlp2xx-tiny.dts
@@ -0,0 +1,141 @@
+/* XLP2XX Device Tree Source
+ *
+ */
+
+/dts-v1/;
+/ {
+	model = "MIPS,XLP2XX";
+	compatible = "NETL,XLP2XX_A0";
+	#address-cells = <1>;
+	#size-cells = <1>;
+	doms {
+		#address-cells = <1>;
+		#size-cells = <1>;
+		dom@0 {
+			device_type = "domain";
+			heap-size = <0x0 0x4000000>;
+			shared-mem = <0x0 0x80000000 0x0 0x20000000>;
+			#address-cells = <1>;
+			#size-cells = <1>;
+			cpu {
+				onlinemask = <0xf>;
+				nae-rx-vc = <0>;
+				nae-fb-vc = <3>;
+			};
+			uart {
+				id = <0>;
+			};
+			memory {
+				reg = <0x00100000 0x0FF00000	// 255M at 1M
+					   0x20000000 0x20000000>;  // 512M at 512M
+			};
+			pic {
+			};
+		};
+	};
+
+	chosen {
+		bootargs = " console=ttyS0,115200 ";
+	};
+
+	soc {
+	    nae@node-0 {
+        	model = "MIPS,XLP2XX NAE CFG";
+	        compatible = "NETL,XLP2XX_A0";
+        	#address-cells = <1>;
+	        #size-cells = <1>;
+        	frequency = <250>;
+
+	        /include/ "ucore.dts"
+
+	        // VFBID MAP: Upto 127 entries
+        	// (each entry is a pair of (vfbid , dest-vc)
+	        // Legal range: (vfbid (0 - 126), dest-vc (0 - 4095))
+        	vfbid-config {
+                	vfbid-map = <
+	                     0    0    1   1   2   2   3   3
+        	        >;
+	        };
+
+        	// Packet Ordering Engine (POE)
+	        poe {
+
+        	        mode = "bypass";
+
+                	// Each vector is 512 bit with masb indicating vc 512 and lsb indicating vc 0
+	                // Format : 512-bit-vector is specified as 16 32-bit words
+        	        // Left most word has the vc range 511-479 right most word has vc range 31 - 0
+                	// Each word has the MSB select higer vc number and LSB select lower vc num
+	                distribution_vectors {
+                                     dv0  = <0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0x1>;
+                                     dv1  = <0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0>;
+                                     dv2  = <0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0>;
+                                     dv3  = <0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0>;
+                                     dv4  = <0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0>;
+                                     dv5  = <0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0>;
+                                     dv6  = <0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0>;
+                                     dv7  = <0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0>;
+                                     dv8  = <0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0>;
+                                     dv9  = <0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0>;
+                                     dv10 = <0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0>;
+                                     dv11 = <0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0>;
+                                     dv12 = <0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0>;
+                                     dv13 = <0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0>;
+                                     dv14 = <0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0>;
+                                     dv15 = <0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0>;
+        	        };
+	        };
+
+	        complex@0 {
+        	        device_type = "nae-complex";
+                	mode = "sgmii";
+
+        	        sgmii {
+                	        loopback = <0 0 0 0>;
+
+                        	// Max ingress fifo size 256 units (size of one unit is 64 byte)
+	                        iface-fifo-size = <13 13 13 13>;
+
+        	                ext-phy-addr = <4 7 6 5>;
+                	        ext-phy-bus = <0 0 0 0>;
+
+                        	// Max parser sequence fifo size 1024 packets
+	                        // (if 1588 Timestamp is not required, then max size increases to 2048)
+        	                num-channels = <1 1 1 1>;
+                	        parser-sequence-fifo-size = <30 30 30 30>;
+
+                        	// Since rx_buffer is per context, we can have a few different ways
+	                        //  to populate the operand
+        	                rx-buffer-size = <128 128 128 128>;
+
+
+                	        // Max available descriptors are 1024 (across all complexes).
+                        	// Per port num_free_descriptors must be even number
+	                        num-free-descs = <52 52 52 52>;
+        	                free-desc-size = <2048 2048 2048 2048>;
+                	        ucore-mask = <0xff 0xff 0xff 0xff>;
+	                };
+	                xaui {
+        	        	loopback = <0>;
+                	        // Max ingress fifo size 256 units (size of one unit is 64 byte)
+                        	iface-fifo-size = <55>;
+
+	                       	// Max parser sequence fifo size 1024 packets
+        	               	// (if 1588 Timestamp is not required, then max size increases to 2048)
+                	       	parser-sequence-fifo-size = <225>;
+
+                        	// Since rx_buffer is per context, the buffer size given here will be divided by num-channels
+	                       	num-channels = <1>;
+        	               	rx-buffer-size = <944>;
+
+                	       	// Max available descriptors are 1024  (across all complexes).
+                        	// Per port num_free_descriptors must be even number
+	                        num-free-descs = <150>;
+        	                free-desc-size = <2048>;
+                	        // In xlp3xx number of ucores available is 8. so, ucore_mask for xlp3xx is (ucore-mask & 0xFF)
+                        	ucore-mask = <0xff>;
+	                };
+        	};
+	   };
+	};
+};
diff --git a/arch/mips/boot/dts/xlp3xx-linux-loader.dts b/arch/mips/boot/dts/xlp3xx-linux-loader.dts
index f66061a..10396e0 100644
--- a/arch/mips/boot/dts/xlp3xx-linux-loader.dts
+++ b/arch/mips/boot/dts/xlp3xx-linux-loader.dts
@@ -1,11 +1,11 @@
-/* XLP8XX Device Tree Source
+/* XLP3XX Device Tree Source
  *
  */
 
 /dts-v1/;
 / {
-	model = "MIPS,XLP8XX";
-	compatible = "NETL,XLP8XX_A0";
+	model = "MIPS,XLP3XX";
+	compatible = "NETL,XLP3XX_A0";
 
 	#address-cells = <1>;
 	#size-cells = <1>;
diff --git a/arch/mips/boot/dts/xlp3xx-linux.dts b/arch/mips/boot/dts/xlp3xx-linux.dts
index 3033cff..2fe30e6 100644
--- a/arch/mips/boot/dts/xlp3xx-linux.dts
+++ b/arch/mips/boot/dts/xlp3xx-linux.dts
@@ -1,4 +1,4 @@
-/* XLP8XX Device Tree Source
+/* XLP3XX Device Tree Source
  *
  */
 
diff --git a/arch/mips/boot/dts/xlp3xx-multi-os.dts b/arch/mips/boot/dts/xlp3xx-multi-os.dts
index dec4357..59ee946 100644
--- a/arch/mips/boot/dts/xlp3xx-multi-os.dts
+++ b/arch/mips/boot/dts/xlp3xx-multi-os.dts
@@ -1,11 +1,11 @@
-/* XLP8XX Device Tree Source
+/* XLP3XX Device Tree Source
  *
  */
 
 /dts-v1/;
 / {
-	model = "MIPS,XLP8XX";
-	compatible = "NETL,XLP8XX_A0";
+	model = "MIPS,XLP3XX";
+	compatible = "NETL,XLP3XX_A0";
 
 	#address-cells = <1>;
 	#size-cells = <1>;
diff --git a/arch/mips/boot/dts/xlp3xx-tiny.dts b/arch/mips/boot/dts/xlp3xx-tiny.dts
index 3cf00c4..5a652ed 100644
--- a/arch/mips/boot/dts/xlp3xx-tiny.dts
+++ b/arch/mips/boot/dts/xlp3xx-tiny.dts
@@ -1,4 +1,4 @@
-/* XLP8XX Device Tree Source
+/* XLP3XX Device Tree Source
  *
  */
 
@@ -40,8 +40,8 @@
 
 	soc {
 	    nae@node-0 {
-        	model = "MIPS,XLP8XX NAE CFG";
-	        compatible = "NETL,XLP8XX_A0";
+		model = "MIPS,XLP3XX NAE CFG";
+		compatible = "NETL,XLP3XX_A0";
         	#address-cells = <1>;
 	        #size-cells = <1>;
         	frequency = <250>;
-- 
1.7.1

