From b7bcdb3765fde51d048f8e25e5c8cffe11d020a9 Mon Sep 17 00:00:00 2001
From: ShengLiang Song <shengliang.song@broadcom.com>
Date: Wed, 2 May 2012 16:55:27 -0700
Subject: uboot: added complex-1 dts

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/boot/dts/xlp3xx-tiny.dts b/arch/mips/boot/dts/xlp3xx-tiny.dts
index 5a652ed..3f06215 100644
--- a/arch/mips/boot/dts/xlp3xx-tiny.dts
+++ b/arch/mips/boot/dts/xlp3xx-tiny.dts
@@ -98,7 +98,6 @@
 
         	                ext-phy-addr = <4 7 6 5>;
                 	        ext-phy-bus = <0 0 0 0>;
-
                         	// Max parser sequence fifo size 1024 packets
 	                        // (if 1588 Timestamp is not required, then max size increases to 2048)
         	                num-channels = <1 1 1 1>;
@@ -115,25 +114,33 @@
         	                free-desc-size = <2048 2048 2048 2048>;
                 	        ucore-mask = <0xff 0xff 0xff 0xff>;
 	                };
-	                xaui {
-        	        	loopback = <0>;
-                	        // Max ingress fifo size 256 units (size of one unit is 64 byte)
-                        	iface-fifo-size = <55>;
+        	};
+	        complex@1 {
+        	        device_type = "nae-complex";
+                	mode = "sgmii";
+        	        sgmii {
+                	        loopback = <0 0 0 0>;
+
+                        	// Max ingress fifo size 256 units (size of one unit is 64 byte)
+	                        iface-fifo-size = <13 13 13 13>;
+        	                ext-phy-addr = <8 11 10 9>;
+                	        ext-phy-bus = <0 0 0 0>;
+
+                        	// Max parser sequence fifo size 1024 packets
+	                        // (if 1588 Timestamp is not required, then max size increases to 2048)
+        	                num-channels = <1 1 1 1>;
+                	        parser-sequence-fifo-size = <30 30 30 30>;
 
-	                       	// Max parser sequence fifo size 1024 packets
-        	               	// (if 1588 Timestamp is not required, then max size increases to 2048)
-                	       	parser-sequence-fifo-size = <225>;
+                        	// Since rx_buffer is per context, we can have a few different ways
+	                        //  to populate the operand
+        	                rx-buffer-size = <128 128 128 128>;
 
-                        	// Since rx_buffer is per context, the buffer size given here will be divided by num-channels
-	                       	num-channels = <1>;
-        	               	rx-buffer-size = <944>;
 
-                	       	// Max available descriptors are 1024  (across all complexes).
+                	        // Max available descriptors are 1024 (across all complexes).
                         	// Per port num_free_descriptors must be even number
-	                        num-free-descs = <150>;
-        	                free-desc-size = <2048>;
-                	        // In xlp3xx number of ucores available is 8. so, ucore_mask for xlp3xx is (ucore-mask & 0xFF)
-                        	ucore-mask = <0xff>;
+	                        num-free-descs = <52 52 52 52>;
+        	                free-desc-size = <2048 2048 2048 2048>;
+                	        ucore-mask = <0xff 0xff 0xff 0xff>;
 	                };
         	};
 	   };
-- 
1.7.1

