From 69db1101b5b132169d09a9a35a397780561729fe Mon Sep 17 00:00:00 2001
From: Prasanna Mumbai <prasanna.mumbai@broadcom.com>
Date: Wed, 1 Aug 2012 16:13:11 +0530
Subject: xlp2xx: Updated the CPU masks and fmn vc masks

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/boot/dts/xlp2xx-helinux.dts b/arch/mips/boot/dts/xlp2xx-helinux.dts
index 00189b3..734ed1c 100644
--- a/arch/mips/boot/dts/xlp2xx-helinux.dts
+++ b/arch/mips/boot/dts/xlp2xx-helinux.dts
@@ -27,7 +27,7 @@
 			#size-cells = <1>;
 
 			cpu {
-				onlinemask = <0xffff>;
+				onlinemask = <0xff>;
                                 nae-rx-vc = <1>;
                                 nae-fb-vc = <2>;
                                 sae-rx-vc = <1>; /*  */
@@ -45,10 +45,7 @@
 					0xE0000000 0x20000000>;  // 512M @ 3.5GB
 			};
                         fmn {
-                                node_0_vc_mask = <0x66666666 0x66666666 0x66666666 0x66666666>;
-                                node_1_vc_mask = <0x66666666 0x66666666 0x66666666 0x66666666>;
-                                node_2_vc_mask = <0x66666666 0x66666666 0x66666666 0x66666666>;
-                                node_3_vc_mask = <0x66666666 0x66666666 0x66666666 0x66666666>;
+                                node_0_vc_mask = <0x66666666>;
                         };
 			pic {
 			};
@@ -79,7 +76,7 @@
 			};
 
 			cpu {
-				onlinemask = <0xffff>;
+				onlinemask = <0xff>;
                                 nae-rx-vc = <0>;
                                 nae-fb-vc = <3>;
                                 /* Setting this to '1' makes NAE driver to skip buf alloction and initial
@@ -95,7 +92,7 @@
 				reg = <0x60000000 0x60000000>;  // 1.5GB @ 1.5GB
 			};
 			app-param{
-				onlinemask = <0xffff>;
+				onlinemask = <0xff>;
 			};
 			nae{
 			};
@@ -105,10 +102,7 @@
 				rsa = <0x1>;
 			};
                         fmn {
-                                node_0_vc_mask = <0x99999999 0x99999999 0x99999999 0x99999999>;
-                                node_1_vc_mask = <0x99999999 0x99999999 0x99999999 0x99999999>;
-                                node_2_vc_mask = <0x99999999 0x99999999 0x99999999 0x99999999>;
-                                node_3_vc_mask = <0x99999999 0x99999999 0x99999999 0x99999999>;
+                                node_0_vc_mask = <0x99999999>;
                         };
 		};
 	};
diff --git a/arch/mips/boot/dts/xlp2xx-linux.dts b/arch/mips/boot/dts/xlp2xx-linux.dts
index 07a6314..ff2d4ac 100644
--- a/arch/mips/boot/dts/xlp2xx-linux.dts
+++ b/arch/mips/boot/dts/xlp2xx-linux.dts
@@ -27,7 +27,7 @@
 			#size-cells = <1>;
 
 			cpu {
-				onlinemask = <0xffff>;
+				onlinemask = <0xff>;
                                 nae-rx-vc = <0>;
                                 nae-fb-vc = <1>;
 				napi-vc-mask = <0xb>;
@@ -49,10 +49,7 @@
 			};
                         fmn {
 				/* make the node mask to 0xffffffff for SRIO */  
-                                node_0_vc_mask = <0xbbbbbbbb 0xbbbbbbbb 0xbbbbbbbb 0xbbbbbbbb>;
-                                node_1_vc_mask = <0xbbbbbbbb 0xbbbbbbbb 0xbbbbbbbb 0xbbbbbbbb>;
-                                node_2_vc_mask = <0xbbbbbbbb 0xbbbbbbbb 0xbbbbbbbb 0xbbbbbbbb>;
-                                node_3_vc_mask = <0xbbbbbbbb 0xbbbbbbbb 0xbbbbbbbb 0xbbbbbbbb>;
+                                node_0_vc_mask = <0xbbbbbbbb>;
                         };
 			pic {
 			};
diff --git a/arch/mips/boot/dts/xlp2xx-multi-os.dts b/arch/mips/boot/dts/xlp2xx-multi-os.dts
index a8c49eb..70c3d88 100644
--- a/arch/mips/boot/dts/xlp2xx-multi-os.dts
+++ b/arch/mips/boot/dts/xlp2xx-multi-os.dts
@@ -88,87 +88,6 @@
 				rsa = <0x0>;
 			};
 		};
-		dom@2 {
-			device_type = "domain";
-			os = "netos";
-
-			heap-size = <0x0 0x2000000>;                   // 32M
-			shared-mem = <0x0 0x68800000 0x0 0x02000000>;  // 1672MB -> 1704MB (32MB)
-
-			#address-cells = <1>;
-			#size-cells = <1>;
-
-			app-bootcmd = "smpload";
-			app-loadaddr = "0x82800000"; // 2088 MB
-
-			kernel {
-				/* Kernel Memory allocated from xen domain-heap */
-				size = <0x02000000>;  // 32M
-
-				/* Location in Dom0 ramdisk */
-				location = "/tmp/xen/netos.elf";
-
-				/* Domain destroy currently doesn't work with hyperexec */
-				dom_destroy_hack = <0>;
-			};
-
-			cpu {
-				onlinemask = <0x0f00>;
-                                nae-rx-vc = <0>;
-                                nae-fb-vc = <1>;
-			};
-			uart {
-				id = <0>;
-				sharedcfg = <0x1f000000>;
-			};
-			memory {
-				reg = <0x64800000 0x04000000>;  // 1608M ->1672 M (64M) 
-			};
-			nae {
-				owner = <0>;
-			};
-		};
-	        dom@3 {
-                        device_type = "domain";
-                        os = "netos";
-
-                        heap-size = <0x0 0x2000000>;                   // 32M
-                        shared-mem = <0x0 0x7a800000 0x0 0x02000000>;  // 1960MB -> 1992MB (32MB)
-
-                        #address-cells = <1>;
-                        #size-cells = <1>;
-
-                        app-bootcmd = "ampload";
-                        app-loadaddr = "0x82800000"; // 2088Mb
-
-                        kernel {
-                                /* Kernel Memory allocated from xen domain-heap */
-                                size = <0x02000000>;  // 32M
-
-                                /* Location in Dom0 ramdisk */
-                                location = "/tmp/xen/netos.elf";
-
-                                /* Domain destroy currently doesn't work with hyperexec */
-                                dom_destroy_hack = <0>;
-                        };
-
-                        cpu {
-                                onlinemask = <0xf000>;
-                                nae-rx-vc = <0>;
-                                nae-fb-vc = <1>;
-                        };
-                        uart {
-                                id = <0>;
-                                sharedcfg = <0x1f000000>;
-                        };
-                        memory {
-                                reg = <0x6a800000 0x10000000>;  // 1704M -> 1960M (256MB)
-                        };
-                        nae {
-                                owner = <0>;
-                        };
-                };
-	
 	};
 
 	chosen {
diff --git a/arch/mips/boot/dts/xlp2xx-netos.dts b/arch/mips/boot/dts/xlp2xx-netos.dts
index a7b8b59..0f87746 100644
--- a/arch/mips/boot/dts/xlp2xx-netos.dts
+++ b/arch/mips/boot/dts/xlp2xx-netos.dts
@@ -45,7 +45,7 @@
 			};
 
 			cpu {
-				onlinemask = <0xffff>;
+				onlinemask = <0xff>;
                                 nae-rx-vc = <0>;
                                 nae-fb-vc = <3>;
 				sae-rx-vc = <0>; /* sharing with nae */
@@ -63,7 +63,7 @@
 				reg = <0x20000000 0xa0000000>;  // 2560M @ 512M
 			};
 			app-param{
-                                onlinemask = <0xffff>;
+                                onlinemask = <0xff>;
 				enable-nae-jumbo = <0>;
                         };
 			nae{
@@ -75,10 +75,7 @@
 			};
 			/* In hyperexec-linux mode, vc_mask needs to be synchronized with linux fmn node! */
                         fmn {
-                                node_0_vc_mask = <0x99999999 0x99999999 0x99999999 0x99999999>;
-                                node_1_vc_mask = <0x99999999 0x99999999 0x99999999 0x99999999>;
-                                node_2_vc_mask = <0x99999999 0x99999999 0x99999999 0x99999999>;
-                                node_3_vc_mask = <0x99999999 0x99999999 0x99999999 0x99999999>;
+                                node_0_vc_mask = <0x99999999>;
                         };
 		};
 	};
-- 
1.7.1

