From 802d8f280bc24e8618a93a4dae426f7b24d7dd0b Mon Sep 17 00:00:00 2001
From: Prasanna Mumbai <prasanna.mumbai@broadcom.com>
Date: Thu, 2 Aug 2012 16:33:29 +0530
Subject: Change POE-SPILL-LENGTH to 64K for testing poe spill-fill

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/boot/dts/xlp4xx-netos.dts b/arch/mips/boot/dts/xlp4xx-netos.dts
index 98fbca3..f3e375e 100644
--- a/arch/mips/boot/dts/xlp4xx-netos.dts
+++ b/arch/mips/boot/dts/xlp4xx-netos.dts
@@ -31,7 +31,7 @@
 			#size-cells = <1>;
 
 			app-bootcmd = "smpload";
-			app-loadaddr = "0x4000000"; // 3.75GB
+			app-loadaddr = "0x2000000"; // 3.75GB
 
 			kernel {
 				/* Kernel Memory allocated from xen domain-heap */
-- 
1.7.1

