From a856234d11966ca190f57370ff19ee7d52daa9fa Mon Sep 17 00:00:00 2001
From: Prasanna Mumbai <prasanna.mumbai@broadcom.com>
Date: Tue, 7 Aug 2012 16:37:00 +0530
Subject: fmn: add FMN owner-config flag for DTS files.

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/boot/dts/mnxlp8xx-helinux-nae.dts b/arch/mips/boot/dts/mnxlp8xx-helinux-nae.dts
index 9ee85f0..67c930e 100644
--- a/arch/mips/boot/dts/mnxlp8xx-helinux-nae.dts
+++ b/arch/mips/boot/dts/mnxlp8xx-helinux-nae.dts
@@ -121,6 +121,7 @@
 				nae = <0x3>;   // currently for node0 and node1
 				sae = <0x1>;   // for node0
 				rsa = <0x1>;
+				fmn = <0x1>;
 			};
 			app-param{
 				onlinemask = <0xffffffff>;
diff --git a/arch/mips/boot/dts/mnxlp8xx-hybrid-nae-helinux.dts b/arch/mips/boot/dts/mnxlp8xx-hybrid-nae-helinux.dts
index 3914848..77ad6dc 100644
--- a/arch/mips/boot/dts/mnxlp8xx-hybrid-nae-helinux.dts
+++ b/arch/mips/boot/dts/mnxlp8xx-hybrid-nae-helinux.dts
@@ -60,6 +60,7 @@
 				nae = <0x3>;   // currently for node0 and node1
 				sae = <0x1>;   // for node0
 				rsa = <0x1>;
+				fmn = <0x1>;
 			};
 			fmn {
                                 node_0_vc_mask = <0x33333333 0x33333333 0x33333333 0x33333333>;
diff --git a/arch/mips/boot/dts/mnxlp8xx-linux.dts b/arch/mips/boot/dts/mnxlp8xx-linux.dts
index 72bd6a6..b4d973a 100644
--- a/arch/mips/boot/dts/mnxlp8xx-linux.dts
+++ b/arch/mips/boot/dts/mnxlp8xx-linux.dts
@@ -70,6 +70,7 @@
 				nae = <0x3>;   // currently for node0 and node1
 				sae = <0x1>;   // for node0
 				rsa = <0x1>;
+				fmn = <0x1>;
 			};
 
 		};
diff --git a/arch/mips/boot/dts/xlp2xx-helinux.dts b/arch/mips/boot/dts/xlp2xx-helinux.dts
index 734ed1c..bfb14b0 100644
--- a/arch/mips/boot/dts/xlp2xx-helinux.dts
+++ b/arch/mips/boot/dts/xlp2xx-helinux.dts
@@ -100,6 +100,7 @@
 				nae = <0x1>;   // currently for node0 and node1
 				sae = <0x1>;   // for node0
 				rsa = <0x1>;
+				fmn = <0x1>;
 			};
                         fmn {
                                 node_0_vc_mask = <0x99999999>;
diff --git a/arch/mips/boot/dts/xlp2xx-linux-loader.dts b/arch/mips/boot/dts/xlp2xx-linux-loader.dts
index 7ad4f8a..f125274 100644
--- a/arch/mips/boot/dts/xlp2xx-linux-loader.dts
+++ b/arch/mips/boot/dts/xlp2xx-linux-loader.dts
@@ -91,6 +91,7 @@
 				nae = <0x0>;
 				sae = <0x0>;
 				rsa = <0x0>;
+				fmn = <0x0>;
 			};
 		};
 	}; 
diff --git a/arch/mips/boot/dts/xlp2xx-linux.dts b/arch/mips/boot/dts/xlp2xx-linux.dts
index ff2d4ac..c5536cf 100644
--- a/arch/mips/boot/dts/xlp2xx-linux.dts
+++ b/arch/mips/boot/dts/xlp2xx-linux.dts
@@ -66,6 +66,7 @@
 				nae = <0x1>;   // currently for node 0
 				sae = <0x1>;
 				rsa = <0x1>;
+				fmn = <0x1>;
 			};
 
 
diff --git a/arch/mips/boot/dts/xlp2xx-multi-os.dts b/arch/mips/boot/dts/xlp2xx-multi-os.dts
index 70c3d88..15f1b70 100644
--- a/arch/mips/boot/dts/xlp2xx-multi-os.dts
+++ b/arch/mips/boot/dts/xlp2xx-multi-os.dts
@@ -86,6 +86,7 @@
 				nae = <0x0>;
 				sae = <0x0>;
 				rsa = <0x0>;
+				fmn = <0x0>;
 			};
 		};
 	};
diff --git a/arch/mips/boot/dts/xlp2xx-netos.dts b/arch/mips/boot/dts/xlp2xx-netos.dts
index 0f87746..a047605 100644
--- a/arch/mips/boot/dts/xlp2xx-netos.dts
+++ b/arch/mips/boot/dts/xlp2xx-netos.dts
@@ -72,6 +72,7 @@
 				nae = <0x1>;
 				sae = <0x1>;
 				rsa = <0x1>;
+				fmn = <0x1>;
 			};
 			/* In hyperexec-linux mode, vc_mask needs to be synchronized with linux fmn node! */
                         fmn {
diff --git a/arch/mips/boot/dts/xlp2xx-tiny.dts b/arch/mips/boot/dts/xlp2xx-tiny.dts
index 70a2df5..2d45084 100644
--- a/arch/mips/boot/dts/xlp2xx-tiny.dts
+++ b/arch/mips/boot/dts/xlp2xx-tiny.dts
@@ -33,6 +33,7 @@
 			};
 			owner-config {
 				nae = <0x1>;
+				fmn = <0x1>;
 			};
 		};
 	};
diff --git a/arch/mips/boot/dts/xlp3xx-helinux.dts b/arch/mips/boot/dts/xlp3xx-helinux.dts
index 66c1964..3d8661b 100644
--- a/arch/mips/boot/dts/xlp3xx-helinux.dts
+++ b/arch/mips/boot/dts/xlp3xx-helinux.dts
@@ -103,6 +103,7 @@
 				nae = <0x1>;   // currently for node 0
 				sae = <0x1>;
 				rsa = <0x1>;
+				fmn = <0x1>;
 			};
                         fmn {
                                 node_0_vc_mask = <0x99999999 0x99999999 0x99999999 0x99999999>;
diff --git a/arch/mips/boot/dts/xlp3xx-linux-loader.dts b/arch/mips/boot/dts/xlp3xx-linux-loader.dts
index c0fd27d..252b680 100644
--- a/arch/mips/boot/dts/xlp3xx-linux-loader.dts
+++ b/arch/mips/boot/dts/xlp3xx-linux-loader.dts
@@ -91,6 +91,7 @@
 				nae = <0x0>;   // currently for node 0
 				sae = <0x0>;
 				rsa = <0x0>;
+				fmn = <0x0>;
 			};
 		};
 	}; 
diff --git a/arch/mips/boot/dts/xlp3xx-linux.dts b/arch/mips/boot/dts/xlp3xx-linux.dts
index 42fb0f7..07fec56 100644
--- a/arch/mips/boot/dts/xlp3xx-linux.dts
+++ b/arch/mips/boot/dts/xlp3xx-linux.dts
@@ -69,6 +69,7 @@
 				nae = <0x1>;   // currently for node 0
 				sae = <0x1>;
 				rsa = <0x1>;
+				fmn = <0x1>;
 			};
 
 
diff --git a/arch/mips/boot/dts/xlp3xx-multi-os.dts b/arch/mips/boot/dts/xlp3xx-multi-os.dts
index bf06f72..9e72b7f 100644
--- a/arch/mips/boot/dts/xlp3xx-multi-os.dts
+++ b/arch/mips/boot/dts/xlp3xx-multi-os.dts
@@ -86,6 +86,7 @@
 				nae = <0x0>;
 				sae = <0x0>;
 				rsa = <0x0>;
+				fmn = <0x0>;
 			};
 		};
 		dom@2 {
@@ -130,6 +131,7 @@
 				nae = <0x0>;
 				sae = <0x0>;
 				rsa = <0x0>;
+				fmn = <0x0>;
 			};
 		};
 	        dom@3 {
@@ -174,6 +176,7 @@
 				nae = <0x0>;
 				sae = <0x0>;
 				rsa = <0x0>;
+				fmn = <0x0>;
 			};
                 };
 	
diff --git a/arch/mips/boot/dts/xlp3xx-netos.dts b/arch/mips/boot/dts/xlp3xx-netos.dts
index 3f63abd..c656d04 100644
--- a/arch/mips/boot/dts/xlp3xx-netos.dts
+++ b/arch/mips/boot/dts/xlp3xx-netos.dts
@@ -72,6 +72,7 @@
 				nae = <0x1>;   // currently for node 0
 				sae = <0x1>;
 				rsa = <0x1>;
+				fmn = <0x1>;
 			};
 			/* In hyperexec-linux mode, vc_mask needs to be synchronized with linux fmn node! */
                         fmn {
diff --git a/arch/mips/boot/dts/xlp3xx-tiny.dts b/arch/mips/boot/dts/xlp3xx-tiny.dts
index 6dce756..da63a96 100644
--- a/arch/mips/boot/dts/xlp3xx-tiny.dts
+++ b/arch/mips/boot/dts/xlp3xx-tiny.dts
@@ -33,6 +33,7 @@
 			};
 			owner-config {
 				nae = <0x1>;
+				fmn = <0x1>;
 			};
 		};
 	};
diff --git a/arch/mips/boot/dts/xlp4xx-helinux.dts b/arch/mips/boot/dts/xlp4xx-helinux.dts
index 6aeed58..c7bba46 100644
--- a/arch/mips/boot/dts/xlp4xx-helinux.dts
+++ b/arch/mips/boot/dts/xlp4xx-helinux.dts
@@ -105,6 +105,7 @@
 				nae = <0x1>;   // currently for node 0
 				sae = <0x1>;
 				rsa = <0x1>;
+				fmn = <0x1>;
 			};
 
                         fmn {
@@ -162,6 +163,7 @@
 				nae = <0x0>;
 				sae = <0x0>;
 				rsa = <0x0>;
+				fmn = <0x0>;
 			};
 
                         fmn {
diff --git a/arch/mips/boot/dts/xlp4xx-linux.dts b/arch/mips/boot/dts/xlp4xx-linux.dts
index 4d30653..059327e 100644
--- a/arch/mips/boot/dts/xlp4xx-linux.dts
+++ b/arch/mips/boot/dts/xlp4xx-linux.dts
@@ -68,6 +68,7 @@
 				nae = <0x1>;   // currently for node 0
 				sae = <0x1>;
 				rsa = <0x1>;
+				fmn = <0x1>;
 			};
 		};
 	};
diff --git a/arch/mips/boot/dts/xlp4xx-multi-os.dts b/arch/mips/boot/dts/xlp4xx-multi-os.dts
index 1a3aaac..e5c1037 100644
--- a/arch/mips/boot/dts/xlp4xx-multi-os.dts
+++ b/arch/mips/boot/dts/xlp4xx-multi-os.dts
@@ -86,6 +86,7 @@
 				nae = <0x0>;
 				sae = <0x0>;
 				rsa = <0x0>;
+				fmn = <0x0>;
 			};
 		};
 		dom@2 {
@@ -130,6 +131,7 @@
 				nae = <0x0>;
 				sae = <0x0>;
 				rsa = <0x0>;
+				fmn = <0x0>;
 			};
 		};
 	        dom@3 {
@@ -174,6 +176,7 @@
 				nae = <0x0>;
 				sae = <0x0>;
 				rsa = <0x0>;
+				fmn = <0x0>;
 			};
                 };
 	
diff --git a/arch/mips/boot/dts/xlp4xx-netos.dts b/arch/mips/boot/dts/xlp4xx-netos.dts
index f3e375e..01569df 100644
--- a/arch/mips/boot/dts/xlp4xx-netos.dts
+++ b/arch/mips/boot/dts/xlp4xx-netos.dts
@@ -71,6 +71,7 @@
 				nae = <0x1>;   // currently for node 0
 				sae = <0x1>;
 				rsa = <0x1>;
+				fmn = <0x1>;
 			};
 			/* In hyperexec-linux mode, vc_mask needs to be synchronized with linux fmn node! */
                         fmn {
diff --git a/arch/mips/boot/dts/xlp4xx-tiny.dts b/arch/mips/boot/dts/xlp4xx-tiny.dts
index ce62e44..b9ca9ac 100644
--- a/arch/mips/boot/dts/xlp4xx-tiny.dts
+++ b/arch/mips/boot/dts/xlp4xx-tiny.dts
@@ -33,6 +33,7 @@
 			};
 			owner-config {
 				nae = <0x1>;
+				fmn = <0x1>;
 			};
 		};
 	};
diff --git a/arch/mips/boot/dts/xlp8xx-helinux-nat.dts b/arch/mips/boot/dts/xlp8xx-helinux-nat.dts
index 944b1cd..a4972e4 100644
--- a/arch/mips/boot/dts/xlp8xx-helinux-nat.dts
+++ b/arch/mips/boot/dts/xlp8xx-helinux-nat.dts
@@ -65,6 +65,7 @@
 				nae = <0x1>;   // currently for node 0
 				sae = <0x1>;
 				rsa = <0x1>;
+				fmn = <0x1>;
 			};
 
 		};
diff --git a/arch/mips/boot/dts/xlp8xx-helinux.dts b/arch/mips/boot/dts/xlp8xx-helinux.dts
index b67a4e4..5f267a4 100644
--- a/arch/mips/boot/dts/xlp8xx-helinux.dts
+++ b/arch/mips/boot/dts/xlp8xx-helinux.dts
@@ -106,6 +106,7 @@
 				nae = <0x1>;   // currently for node 0
 				sae = <0x1>;
 				rsa = <0x1>;
+				fmn = <0x1>;
 			};
 
                         fmn {
diff --git a/arch/mips/boot/dts/xlp8xx-hybrid-nae-netos.dts b/arch/mips/boot/dts/xlp8xx-hybrid-nae-netos.dts
index ce8a484..2d24c5a 100644
--- a/arch/mips/boot/dts/xlp8xx-hybrid-nae-netos.dts
+++ b/arch/mips/boot/dts/xlp8xx-hybrid-nae-netos.dts
@@ -58,6 +58,7 @@
 				nae = <0x1>;   // currently for node 0
 				sae = <0x1>;
 				rsa = <0x1>;
+				fmn = <0x1>;
 			};
 			fmn {
 				node_0_vc_mask = <0x33333333 0x33333333 0x33333333 0x33333333>;
@@ -118,6 +119,7 @@
 				nae = <0x0>;
 				sae = <0x0>;
 				rsa = <0x0>;
+				fmn = <0x0>;
 			};
 
 			app-param {
diff --git a/arch/mips/boot/dts/xlp8xx-linux-loader.dts b/arch/mips/boot/dts/xlp8xx-linux-loader.dts
index 886da00..c6ae07d 100644
--- a/arch/mips/boot/dts/xlp8xx-linux-loader.dts
+++ b/arch/mips/boot/dts/xlp8xx-linux-loader.dts
@@ -92,6 +92,7 @@
 				nae = <0x0>;
 				sae = <0x0>;
 				rsa = <0x0>;
+				fmn = <0x0>;
 			};
 		};
 		dom@2 {
@@ -139,6 +140,7 @@
 				nae = <0x0>;
 				sae = <0x0>;
 				rsa = <0x0>;
+				fmn = <0x0>;
 			};
 		};
 		dom@3 {
@@ -183,6 +185,7 @@
 				nae = <0x0>;
 				sae = <0x0>;
 				rsa = <0x0>;
+				fmn = <0x0>;
 			};
 		};
 	};
diff --git a/arch/mips/boot/dts/xlp8xx-linux.dts b/arch/mips/boot/dts/xlp8xx-linux.dts
index 1e29d8e..4c54c5e 100644
--- a/arch/mips/boot/dts/xlp8xx-linux.dts
+++ b/arch/mips/boot/dts/xlp8xx-linux.dts
@@ -68,6 +68,7 @@
 				nae = <0x1>;   // currently for node 0
 				sae = <0x1>;
 				rsa = <0x1>;
+				fmn = <0x1>;
 			};
 		};
 	};
diff --git a/arch/mips/boot/dts/xlp8xx-multi-os.dts b/arch/mips/boot/dts/xlp8xx-multi-os.dts
index 3ce5802..4372c80 100644
--- a/arch/mips/boot/dts/xlp8xx-multi-os.dts
+++ b/arch/mips/boot/dts/xlp8xx-multi-os.dts
@@ -93,6 +93,7 @@
 	    	nae = <0x0>;
 		sae = <0x0>;
 		rsa = <0x0>;
+		fmn = <0x1>;
 	    };
             kbp {
                 /* distribute total of 4096 buffer id across cpus. required only for cpus sending requests.
@@ -177,6 +178,7 @@
 	    	nae = <0x0>;
 		sae = <0x0>;
 		rsa = <0x0>;
+		fmn = <0x0>;
 	    };
         };
         dom@3 {
@@ -218,7 +220,8 @@
             nae {
             };
 	    owner-config {
-	    nae = <0x0>;
+	    	nae = <0x0>;
+		fmn = <0x1>;
 	    };
         };
     };
diff --git a/arch/mips/boot/dts/xlp8xx-netos-nat.dts b/arch/mips/boot/dts/xlp8xx-netos-nat.dts
index c29b60a..ca47402 100644
--- a/arch/mips/boot/dts/xlp8xx-netos-nat.dts
+++ b/arch/mips/boot/dts/xlp8xx-netos-nat.dts
@@ -74,6 +74,7 @@
 				nae = <0x1>;   // currently for node 0
 				sae = <0x1>;
 				rsa = <0x1>;
+				fmn = <0x1>;
 			};
 		};
 	};
diff --git a/arch/mips/boot/dts/xlp8xx-netos.dts b/arch/mips/boot/dts/xlp8xx-netos.dts
index 7d20b3f..4dac487 100644
--- a/arch/mips/boot/dts/xlp8xx-netos.dts
+++ b/arch/mips/boot/dts/xlp8xx-netos.dts
@@ -72,6 +72,7 @@
 				nae = <0x1>;   // currently for node 0
 				sae = <0x1>;
 				rsa = <0x1>;
+				fmn = <0x1>;
 			};
 			/* In hyperexec-linux mode, vc_mask needs to be synchronized with linux fmn node! */
                         fmn {
diff --git a/arch/mips/boot/dts/xlp8xx-tiny.dts b/arch/mips/boot/dts/xlp8xx-tiny.dts
index 8996677..f4bc675 100644
--- a/arch/mips/boot/dts/xlp8xx-tiny.dts
+++ b/arch/mips/boot/dts/xlp8xx-tiny.dts
@@ -33,6 +33,7 @@
 			};
 			owner-config {
 				nae = <0x1>;
+				fmn = <0x1>;
 			};
 		};
 	};
-- 
1.7.1

