From 3b5a4189837c266f6fc6410b909ae371c6ea2009 Mon Sep 17 00:00:00 2001
From: Sreenidhi BR <sreenira@broadcom.com>
Date: Fri, 17 Aug 2012 11:23:12 +0530
Subject: nae: add app parameters <enable-nae-p2p> <enable-nae-hw-parser> to 2xx/3xx/4xx netos/helinux DTS files. These are disabled by default, they can be enabled when required.

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/boot/dts/xlp2xx-helinux.dts b/arch/mips/boot/dts/xlp2xx-helinux.dts
index bfb14b0..d9a7176 100644
--- a/arch/mips/boot/dts/xlp2xx-helinux.dts
+++ b/arch/mips/boot/dts/xlp2xx-helinux.dts
@@ -93,6 +93,9 @@
 			};
 			app-param{
 				onlinemask = <0xff>;
+				enable-nae-hw-parser = <0>;
+				enable-nae-jumbo = <0>;
+				enable-nae-p2p = <0>;
 			};
 			nae{
 			};
diff --git a/arch/mips/boot/dts/xlp2xx-netos.dts b/arch/mips/boot/dts/xlp2xx-netos.dts
index a047605..4326486 100644
--- a/arch/mips/boot/dts/xlp2xx-netos.dts
+++ b/arch/mips/boot/dts/xlp2xx-netos.dts
@@ -65,6 +65,8 @@
 			app-param{
                                 onlinemask = <0xff>;
 				enable-nae-jumbo = <0>;
+				enable-nae-hw-parser = <0>;
+				enable-nae-p2p = <0>;
                         };
 			nae{
 			};
diff --git a/arch/mips/boot/dts/xlp3xx-helinux.dts b/arch/mips/boot/dts/xlp3xx-helinux.dts
index 3d8661b..8b98dcc 100644
--- a/arch/mips/boot/dts/xlp3xx-helinux.dts
+++ b/arch/mips/boot/dts/xlp3xx-helinux.dts
@@ -96,6 +96,9 @@
 			};
 			app-param{
 				onlinemask = <0xffff>;
+				enable-nae-hw-parser = <0>;
+				enable-nae-jumbo = <0>;
+				enable-nae-p2p = <0>;
 			};
 			nae{
 			};
diff --git a/arch/mips/boot/dts/xlp3xx-netos.dts b/arch/mips/boot/dts/xlp3xx-netos.dts
index c656d04..96b66e7 100644
--- a/arch/mips/boot/dts/xlp3xx-netos.dts
+++ b/arch/mips/boot/dts/xlp3xx-netos.dts
@@ -65,6 +65,8 @@
 			app-param{
                                 onlinemask = <0xffff>;
 				enable-nae-jumbo = <0>;
+				enable-nae-hw-parser = <0>;
+				enable-nae-p2p = <0>;
                         };
 			nae{
 			};
diff --git a/arch/mips/boot/dts/xlp4xx-helinux.dts b/arch/mips/boot/dts/xlp4xx-helinux.dts
index c7bba46..dea5108 100644
--- a/arch/mips/boot/dts/xlp4xx-helinux.dts
+++ b/arch/mips/boot/dts/xlp4xx-helinux.dts
@@ -98,6 +98,9 @@
 			};
 			app-param{
 				onlinemask = <0xffff>;
+				enable-nae-hw-parser = <0>;
+				enable-nae-jumbo = <0>;
+				enable-nae-p2p = <0>;
 			};
 			nae{
 			};
diff --git a/arch/mips/boot/dts/xlp4xx-netos.dts b/arch/mips/boot/dts/xlp4xx-netos.dts
index 01569df..71ccc55 100644
--- a/arch/mips/boot/dts/xlp4xx-netos.dts
+++ b/arch/mips/boot/dts/xlp4xx-netos.dts
@@ -64,6 +64,8 @@
 			app-param{
                                 onlinemask = <0xffff>;
 				enable-nae-jumbo = <0>;
+				enable-nae-hw-parser = <0>;
+				enable-nae-p2p = <0>;
                         };
 			nae{
 			};
-- 
1.7.1

