From 6296095e1a0ac9f46e21e6cc5b6b7f74c7343fc0 Mon Sep 17 00:00:00 2001
From: Sreenidhi BR <sreenira@broadcom.com>
Date: Tue, 21 Aug 2012 18:56:07 +0530
Subject: sysconfig: updated FMN ownership for dom-0 in all applicable DTS files.

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/boot/dts/mnxlp8xx-helinux-nae.dts b/arch/mips/boot/dts/mnxlp8xx-helinux-nae.dts
index 67c930e..1b89582 100644
--- a/arch/mips/boot/dts/mnxlp8xx-helinux-nae.dts
+++ b/arch/mips/boot/dts/mnxlp8xx-helinux-nae.dts
@@ -78,6 +78,12 @@
 					segth  = <34>;	/*Maximum number of tx entries*/
 				};
 			};
+			owner-config {
+				nae = <0x0>;
+				sae = <0x1>;   // for node0
+				rsa = <0x1>;
+				fmn = <0xf>;
+			};
 		};
 		dom@1 {
 			device_type = "domain";
@@ -121,7 +127,7 @@
 				nae = <0x3>;   // currently for node0 and node1
 				sae = <0x1>;   // for node0
 				rsa = <0x1>;
-				fmn = <0x1>;
+				fmn = <0xf>;
 			};
 			app-param{
 				onlinemask = <0xffffffff>;
diff --git a/arch/mips/boot/dts/mnxlp8xx-helinux-nat.dts b/arch/mips/boot/dts/mnxlp8xx-helinux-nat.dts
index 2bbc6c8..10af712 100644
--- a/arch/mips/boot/dts/mnxlp8xx-helinux-nat.dts
+++ b/arch/mips/boot/dts/mnxlp8xx-helinux-nat.dts
@@ -53,7 +53,6 @@
 			pic {
 			};
 			nae {
-				owner = <1 1 0 0>;
 				sh-domains = <1 2>;
 				freein-fifo-mask = <0x30000 0x30000 0 0>;
 				/* node offsets and size */
@@ -64,6 +63,12 @@
 				vfbtbl-hw-nentries = <0 0 0 0>;
 				dummy-pktdata-addr = <0x0 0x0ff00000>; // 255 MB
 			};
+			owner-config {
+				nae = <0x3>;   // currently for node0 and node1
+				sae = <0x1>;   // for node0
+				rsa = <0x1>;
+				fmn = <0xf>;
+			};
 
 		};
 
@@ -106,7 +111,6 @@
 				onlinemask = <0xffff>;
 			};
 			nae{
-				owner = <0 0 0 0>;
 				sh-domains = <0>;
 				freein-fifo-mask = <0xffff 0 0 0>;
 				/* node offsets and size */
@@ -117,6 +121,12 @@
 				vfbtbl-hw-nentries = <20 20 20 20>;
 				dummy-pktdata-addr = <0x0 0x0ff00000>; // 255 MB
 			};
+			owner-config {
+				nae = <0x0>;   
+				sae = <0x0>;
+				rsa = <0x0>;
+				fmn = <0x0>;
+			};
 		};
 
 
@@ -156,7 +166,6 @@
 				reg = <0x00000000 0x00000000>;  // not used
 			};
 			nae {
-				owner = <0 0 0 0>;
 				freein-fifo-mask = <0x0 0xffff 0 0>;
 				sh-domains = <0>;
 				/* node offsets and size */
@@ -167,6 +176,12 @@
 
 				dummy-pktdata-addr = <0x0 0x0ff00000>; // 255 MB
 			};
+			owner-config {
+				nae = <0x0>;
+				sae = <0x0>;
+				rsa = <0x0>;
+				fmn = <0x0>;
+			};
 		};
 	};
 
diff --git a/arch/mips/boot/dts/mnxlp8xx-hybrid-nae-helinux.dts b/arch/mips/boot/dts/mnxlp8xx-hybrid-nae-helinux.dts
index 77ad6dc..49b5358 100644
--- a/arch/mips/boot/dts/mnxlp8xx-hybrid-nae-helinux.dts
+++ b/arch/mips/boot/dts/mnxlp8xx-hybrid-nae-helinux.dts
@@ -60,7 +60,7 @@
 				nae = <0x3>;   // currently for node0 and node1
 				sae = <0x1>;   // for node0
 				rsa = <0x1>;
-				fmn = <0x1>;
+				fmn = <0xf>;
 			};
 			fmn {
                                 node_0_vc_mask = <0x33333333 0x33333333 0x33333333 0x33333333>;
@@ -122,6 +122,7 @@
 				nae = <0x0>;
 				sae = <0x0>;
 				rsa = <0x0>;
+				fmn = <0x0>;
 			};
 		};
 		dom@2 {
@@ -175,6 +176,7 @@
 				nae = <0x0>;   // currently for node 0
 				sae = <0x0>;
 				rsa = <0x0>;
+				fmn = <0x0>;
 			};
 		};
 
diff --git a/arch/mips/boot/dts/mnxlp8xx-linux.dts b/arch/mips/boot/dts/mnxlp8xx-linux.dts
index b4d973a..6ec8002 100644
--- a/arch/mips/boot/dts/mnxlp8xx-linux.dts
+++ b/arch/mips/boot/dts/mnxlp8xx-linux.dts
@@ -70,7 +70,7 @@
 				nae = <0x3>;   // currently for node0 and node1
 				sae = <0x1>;   // for node0
 				rsa = <0x1>;
-				fmn = <0x1>;
+				fmn = <0xf>;
 			};
 
 		};
diff --git a/arch/mips/boot/dts/netos-srtp.dts b/arch/mips/boot/dts/netos-srtp.dts
index 0840c01..abb77d4 100644
--- a/arch/mips/boot/dts/netos-srtp.dts
+++ b/arch/mips/boot/dts/netos-srtp.dts
@@ -67,6 +67,12 @@
                                 node_2_vc_mask = <0x99999999 0x99999999 0x99999999 0x99999999>;
                                 node_3_vc_mask = <0x99999999 0x99999999 0x99999999 0x99999999>;
                         };
+			owner-config {
+				nae = <0x0>;
+				sae = <0x1>;
+				rsa = <0x1>;
+				fmn = <0x1>;
+			};
 		};
 	};
 
diff --git a/arch/mips/boot/dts/xlp2xx-helinux.dts b/arch/mips/boot/dts/xlp2xx-helinux.dts
index d9a7176..e0ee634 100644
--- a/arch/mips/boot/dts/xlp2xx-helinux.dts
+++ b/arch/mips/boot/dts/xlp2xx-helinux.dts
@@ -49,6 +49,12 @@
                         };
 			pic {
 			};
+			owner-config {
+				nae = <0x0>;
+				sae = <0x1>;
+				rsa = <0x1>;
+				fmn = <0x1>;
+			};
 		};
 
 		dom@1 {
diff --git a/arch/mips/boot/dts/xlp2xx-linux-loader.dts b/arch/mips/boot/dts/xlp2xx-linux-loader.dts
index f125274..8e03dfc 100644
--- a/arch/mips/boot/dts/xlp2xx-linux-loader.dts
+++ b/arch/mips/boot/dts/xlp2xx-linux-loader.dts
@@ -48,6 +48,12 @@
 			};
 			pic {
 			};
+			owner-config {
+				nae = <0x0>;
+				sae = <0x1>;
+				rsa = <0x1>;
+				fmn = <0x1>;
+			};
 		};
 		dom@1 {
 			device_type = "domain";
diff --git a/arch/mips/boot/dts/xlp2xx-multi-os.dts b/arch/mips/boot/dts/xlp2xx-multi-os.dts
index 15f1b70..a0b9e71 100644
--- a/arch/mips/boot/dts/xlp2xx-multi-os.dts
+++ b/arch/mips/boot/dts/xlp2xx-multi-os.dts
@@ -43,6 +43,12 @@
 			};
 			pic {
 			};
+			owner-config {
+				nae = <0x0>;
+				sae = <0x1>;
+				rsa = <0x1>;
+				fmn = <0x1>;
+			};
 		};
 		dom@1 {
 			device_type = "domain";
diff --git a/arch/mips/boot/dts/xlp3xx-helinux.dts b/arch/mips/boot/dts/xlp3xx-helinux.dts
index 8b98dcc..1256501 100644
--- a/arch/mips/boot/dts/xlp3xx-helinux.dts
+++ b/arch/mips/boot/dts/xlp3xx-helinux.dts
@@ -52,6 +52,12 @@
                         };
 			pic {
 			};
+			owner-config {
+				nae = <0x0>;
+				sae = <0x1>;
+				rsa = <0x1>;
+				fmn = <0x1>;
+			};
 		};
 
 		dom@1 {
diff --git a/arch/mips/boot/dts/xlp3xx-linux-loader.dts b/arch/mips/boot/dts/xlp3xx-linux-loader.dts
index 252b680..e260f23 100644
--- a/arch/mips/boot/dts/xlp3xx-linux-loader.dts
+++ b/arch/mips/boot/dts/xlp3xx-linux-loader.dts
@@ -48,6 +48,12 @@
 			};
 			pic {
 			};
+			owner-config {
+				nae = <0x0>;
+				sae = <0x1>;
+				rsa = <0x1>;
+				fmn = <0x1>;
+			};
 		};
 		dom@1 {
 			device_type = "domain";
diff --git a/arch/mips/boot/dts/xlp3xx-multi-os.dts b/arch/mips/boot/dts/xlp3xx-multi-os.dts
index 9e72b7f..3b38cfc 100644
--- a/arch/mips/boot/dts/xlp3xx-multi-os.dts
+++ b/arch/mips/boot/dts/xlp3xx-multi-os.dts
@@ -43,6 +43,12 @@
 			};
 			pic {
 			};
+			owner-config {
+				nae = <0x0>;
+				sae = <0x1>;
+				rsa = <0x1>;
+				fmn = <0x1>;
+			};
 		};
 		dom@1 {
 			device_type = "domain";
diff --git a/arch/mips/boot/dts/xlp4xx-helinux.dts b/arch/mips/boot/dts/xlp4xx-helinux.dts
index dea5108..b5a5f8d 100644
--- a/arch/mips/boot/dts/xlp4xx-helinux.dts
+++ b/arch/mips/boot/dts/xlp4xx-helinux.dts
@@ -52,6 +52,12 @@
                         };
 			pic {
 			};
+			owner-config {
+				nae = <0x0>;
+				sae = <0x1>;
+				rsa = <0x1>;
+				fmn = <0x1>;
+			};
 		};
 
 		dom@1 {
diff --git a/arch/mips/boot/dts/xlp4xx-multi-os.dts b/arch/mips/boot/dts/xlp4xx-multi-os.dts
index e5c1037..a4df8a0 100644
--- a/arch/mips/boot/dts/xlp4xx-multi-os.dts
+++ b/arch/mips/boot/dts/xlp4xx-multi-os.dts
@@ -43,6 +43,12 @@
 			};
 			pic {
 			};
+			owner-config {
+				nae = <0x0>;
+				sae = <0x1>;
+				rsa = <0x1>;
+				fmn = <0x1>;
+			};
 		};
 		dom@1 {
 			device_type = "domain";
diff --git a/arch/mips/boot/dts/xlp8xx-helinux-snort.dts b/arch/mips/boot/dts/xlp8xx-helinux-snort.dts
index 34da779..5e5e353 100644
--- a/arch/mips/boot/dts/xlp8xx-helinux-snort.dts
+++ b/arch/mips/boot/dts/xlp8xx-helinux-snort.dts
@@ -52,6 +52,12 @@
                         };
 			pic {
 			};
+			owner-config {
+				nae = <0x0>;
+				sae = <0x0>;
+				rsa = <0x0>;
+				fmn = <0x1>;
+			};
 		};
 
 		dom@1 {
@@ -100,7 +106,6 @@
 				onlinemask = <0xffffffff>;
 			};
 			nae{
-				owner = <1>;
 			};
 
                         fmn {
@@ -149,6 +154,12 @@
                  * kbp driver requires 2 MBytes of dedicated memory space to store data */
                 memory-offsethigh-offsetlow = <0x00000001 0x00000000>;
             };
+			owner-config {
+				nae = <0x1>;
+				sae = <0x0>;
+				rsa = <0x0>;
+				fmn = <0x0>;
+			};
 		};
 
 		dom@2 {
@@ -193,7 +204,6 @@
 				onlinemask = <0xffff0000>;
 			};
 			nae{
-				owner = <0>;
 			};
 
                         fmn {
diff --git a/arch/mips/boot/dts/xlp8xx-helinux.dts b/arch/mips/boot/dts/xlp8xx-helinux.dts
index c2eb702..c9fa889 100644
--- a/arch/mips/boot/dts/xlp8xx-helinux.dts
+++ b/arch/mips/boot/dts/xlp8xx-helinux.dts
@@ -52,6 +52,12 @@
                         };
 			pic {
 			};
+			owner-config {
+				nae = <0x0>;
+				sae = <0x1>;
+				rsa = <0x1>;
+				fmn = <0x1>;
+			};
 		};
 
 		dom@1 {
diff --git a/arch/mips/boot/dts/xlp8xx-linux-loader.dts b/arch/mips/boot/dts/xlp8xx-linux-loader.dts
index c6ae07d..31fe663 100644
--- a/arch/mips/boot/dts/xlp8xx-linux-loader.dts
+++ b/arch/mips/boot/dts/xlp8xx-linux-loader.dts
@@ -49,6 +49,12 @@
 			};
 			pic {
 			};
+			owner-config {
+				nae = <0x0>;
+				sae = <0x1>;
+				rsa = <0x1>;
+				fmn = <0x1>;
+			};
 		};
 		dom@1 {
 			device_type = "domain";
diff --git a/arch/mips/boot/dts/xlp8xx-multi-os.dts b/arch/mips/boot/dts/xlp8xx-multi-os.dts
index 4372c80..ce8960d 100644
--- a/arch/mips/boot/dts/xlp8xx-multi-os.dts
+++ b/arch/mips/boot/dts/xlp8xx-multi-os.dts
@@ -50,6 +50,12 @@
                 node_2_vc_mask = <0x33333333 0x33333333 0x33333333 0x33333333>;
                 node_3_vc_mask = <0x33333333 0x33333333 0x33333333 0x33333333>;
             };
+			owner-config {
+				nae = <0x0>;
+				sae = <0x1>;
+				rsa = <0x1>;
+				fmn = <0x1>;
+			};
         };
         dom@1 {
             device_type = "domain";
@@ -93,7 +99,7 @@
 	    	nae = <0x0>;
 		sae = <0x0>;
 		rsa = <0x0>;
-		fmn = <0x1>;
+		fmn = <0x0>;
 	    };
             kbp {
                 /* distribute total of 4096 buffer id across cpus. required only for cpus sending requests.
@@ -221,7 +227,7 @@
             };
 	    owner-config {
 	    	nae = <0x0>;
-		fmn = <0x1>;
+		fmn = <0x0>;
 	    };
         };
     };
diff --git a/arch/mips/boot/dts/xlp8xx-pcienet-netos.dts b/arch/mips/boot/dts/xlp8xx-pcienet-netos.dts
index 51f2c96..debbb53 100644
--- a/arch/mips/boot/dts/xlp8xx-pcienet-netos.dts
+++ b/arch/mips/boot/dts/xlp8xx-pcienet-netos.dts
@@ -112,6 +112,12 @@
                  * kbp driver requires 2 MBytes of dedicated memory space to store data */
                 memory-offsethigh-offsetlow = <0x00000001 0x00000000>;
             };
+			owner-config {
+				nae = <0x0>;
+				sae = <0x1>;
+				rsa = <0x1>;
+				fmn = <0x1>;
+			};
 		};
 	};
 
diff --git a/arch/mips/boot/dts/xlp8xx-tiny.dts b/arch/mips/boot/dts/xlp8xx-tiny.dts
index f4bc675..f667c8c 100644
--- a/arch/mips/boot/dts/xlp8xx-tiny.dts
+++ b/arch/mips/boot/dts/xlp8xx-tiny.dts
@@ -34,6 +34,8 @@
 			owner-config {
 				nae = <0x1>;
 				fmn = <0x1>;
+				sae = <0x1>;
+				rsa = <0x1>;
 			};
 		};
 	};
-- 
1.7.1

