From e0c1c1e9ee1388eff3dbf436c12eac25b11ed67d Mon Sep 17 00:00:00 2001
From: ShengLiang Song <shengliang.song@broadcom.com>
Date: Fri, 10 Aug 2012 10:05:01 -0700
Subject: xlp2xx: Updated xlp2xx ucore mask to 0xff

- Updated sysconfig for Linux and U-boot.
[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/boot/dts/xlp2xx-linux.dts b/arch/mips/boot/dts/xlp2xx-linux.dts
index c5536cf..03d72b8 100644
--- a/arch/mips/boot/dts/xlp2xx-linux.dts
+++ b/arch/mips/boot/dts/xlp2xx-linux.dts
@@ -49,7 +49,7 @@
 			};
                         fmn {
 				/* make the node mask to 0xffffffff for SRIO */  
-                                node_0_vc_mask = <0xbbbbbbbb>;
+                                node_0_vc_mask = <0xbbbbbbbb 0xbbbbbbbb 0xbbbbbbbb 0xbbbbbbbb>;
                         };
 			pic {
 			};
diff --git a/arch/mips/boot/dts/xlp2xx-multi-os.dts b/arch/mips/boot/dts/xlp2xx-multi-os.dts
index a0b9e71..b35f472 100644
--- a/arch/mips/boot/dts/xlp2xx-multi-os.dts
+++ b/arch/mips/boot/dts/xlp2xx-multi-os.dts
@@ -95,6 +95,94 @@
 				fmn = <0x0>;
 			};
 		};
+		dom@2 {
+			device_type = "domain";
+			os = "netos";
+
+			heap-size = <0x0 0x2000000>;                   // 32M
+			shared-mem = <0x0 0x68800000 0x0 0x02000000>;  // 1672MB -> 1704MB (32MB)
+
+			#address-cells = <1>;
+			#size-cells = <1>;
+
+			app-bootcmd = "smpload";
+			app-loadaddr = "0x82800000"; // 2088 MB
+
+			kernel {
+				/* Kernel Memory allocated from xen domain-heap */
+				size = <0x02000000>;  // 32M
+
+				/* Location in Dom0 ramdisk */
+				location = "/tmp/xen/netos.elf";
+
+				/* Domain destroy currently doesn't work with hyperexec */
+				dom_destroy_hack = <0>;
+			};
+
+			cpu {
+				onlinemask = <0x0f00>;
+                                nae-rx-vc = <0>;
+                                nae-fb-vc = <1>;
+			};
+			uart {
+				id = <0>;
+				sharedcfg = <0x1f000000>;
+			};
+			memory {
+				reg = <0x64800000 0x04000000>;  // 1608M ->1672 M (64M) 
+			};
+			nae {
+			};
+			owner-config {
+				nae = <0x0>;
+				sae = <0x0>;
+				rsa = <0x0>;
+			};
+		};
+	        dom@3 {
+                        device_type = "domain";
+                        os = "netos";
+
+                        heap-size = <0x0 0x2000000>;                   // 32M
+                        shared-mem = <0x0 0x7a800000 0x0 0x02000000>;  // 1960MB -> 1992MB (32MB)
+
+                        #address-cells = <1>;
+                        #size-cells = <1>;
+
+                        app-bootcmd = "ampload";
+                        app-loadaddr = "0x82800000"; // 2088Mb
+
+                        kernel {
+                                /* Kernel Memory allocated from xen domain-heap */
+                                size = <0x02000000>;  // 32M
+
+                                /* Location in Dom0 ramdisk */
+                                location = "/tmp/xen/netos.elf";
+
+                                /* Domain destroy currently doesn't work with hyperexec */
+                                dom_destroy_hack = <0>;
+                        };
+
+                        cpu {
+                                onlinemask = <0xf000>;
+                                nae-rx-vc = <0>;
+                                nae-fb-vc = <1>;
+                        };
+                        uart {
+                                id = <0>;
+                                sharedcfg = <0x1f000000>;
+                        };
+                        memory {
+                                reg = <0x6a800000 0x10000000>;  // 1704M -> 1960M (256MB)
+                        };
+                        nae {
+                        };
+			owner-config {
+				nae = <0x0>;
+				sae = <0x0>;
+				rsa = <0x0>;
+			};
+                };
 	};
 
 	chosen {
diff --git a/arch/mips/boot/dts/xlp2xx-tiny.dts b/arch/mips/boot/dts/xlp2xx-tiny.dts
index 2d45084..2af9ee8 100644
--- a/arch/mips/boot/dts/xlp2xx-tiny.dts
+++ b/arch/mips/boot/dts/xlp2xx-tiny.dts
@@ -119,27 +119,35 @@
         	                free-desc-size = <2048 2048 2048 2048>;
                 	        ucore-mask = <0xff 0xff 0xff 0xff>;
 	                };
-	                xaui {
-        	        	loopback = <0>;
-                	        // Max ingress fifo size 256 units (size of one unit is 64 byte)
-                        	iface-fifo-size = <55>;
+        	};
+		complex@1 {
+        	        device_type = "nae-complex";
+                	mode = "sgmii";
+        	        sgmii {
+                	        loopback = <0 0 0 0>;
+
+                        	// Max ingress fifo size 256 units (size of one unit is 64 byte)
+	                        iface-fifo-size = <13 13 13 13>;
+        	                ext-phy-addr = <8 11 10 9>;
+                	        ext-phy-bus = <0 0 0 0>;
+
+                        	// Max parser sequence fifo size 1024 packets
+	                        // (if 1588 Timestamp is not required, then max size increases to 2048)
+        	                num-channels = <1 1 1 1>;
+                	        parser-sequence-fifo-size = <30 30 30 30>;
 
-	                       	// Max parser sequence fifo size 1024 packets
-        	               	// (if 1588 Timestamp is not required, then max size increases to 2048)
-                	       	parser-sequence-fifo-size = <225>;
+                        	// Since rx_buffer is per context, we can have a few different ways
+	                        //  to populate the operand
+        	                rx-buffer-size = <128 128 128 128>;
 
-                        	// Since rx_buffer is per context, the buffer size given here will be divided by num-channels
-	                       	num-channels = <1>;
-        	               	rx-buffer-size = <944>;
 
-                	       	// Max available descriptors are 1024  (across all complexes).
+                	        // Max available descriptors are 1024 (across all complexes).
                         	// Per port num_free_descriptors must be even number
-	                        num-free-descs = <150>;
-        	                free-desc-size = <2048>;
-                	        // In xlp3xx number of ucores available is 8. so, ucore_mask for xlp3xx is (ucore-mask & 0xFF)
-                        	ucore-mask = <0xff>;
+	                        num-free-descs = <52 52 52 52>;
+        	                free-desc-size = <2048 2048 2048 2048>;
+                	        ucore-mask = <0xff 0xff 0xff 0xff>;
 	                };
-        	};
 	   };
 	};
+   };	
 };
-- 
1.7.1

