From a9a6642b69ee5a167c57512949d60f306dfc91b0 Mon Sep 17 00:00:00 2001
From: Prasanna Mumbai <prasanna.mumbai@broadcom.com>
Date: Wed, 10 Oct 2012 10:48:16 +0530
Subject: Frequency: add frequency config for the DTS files.

Signed-off-by: Prasanna Mumbai <prasanna.mumbai@broadcom.com>
[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/boot/dts/mnxlp8xx-helinux-nae.dts b/arch/mips/boot/dts/mnxlp8xx-helinux-nae.dts
index 1b89582..ef6706d 100644
--- a/arch/mips/boot/dts/mnxlp8xx-helinux-nae.dts
+++ b/arch/mips/boot/dts/mnxlp8xx-helinux-nae.dts
@@ -190,5 +190,12 @@
 			num-nodes = <2>;
 		};
 	};
+	frequency-config {
+		nae = <500>;
+		sae = <250>;
+		rsa = <250>;
+		dtre = <250>;
+		cde = <250>;
+	};
 };
 
diff --git a/arch/mips/boot/dts/mnxlp8xx-helinux-nat.dts b/arch/mips/boot/dts/mnxlp8xx-helinux-nat.dts
index 3f3c4ed..93b4242 100644
--- a/arch/mips/boot/dts/mnxlp8xx-helinux-nat.dts
+++ b/arch/mips/boot/dts/mnxlp8xx-helinux-nat.dts
@@ -224,5 +224,12 @@
 		};
 
 	};
+	frequency-config {
+		nae = <500>;
+		sae = <250>;
+		rsa = <250>;
+		dtre = <250>;
+		cde = <250>;
+	};
 };
 
diff --git a/arch/mips/boot/dts/mnxlp8xx-hybrid-nae-helinux.dts b/arch/mips/boot/dts/mnxlp8xx-hybrid-nae-helinux.dts
index caff28e..df1adc8 100644
--- a/arch/mips/boot/dts/mnxlp8xx-hybrid-nae-helinux.dts
+++ b/arch/mips/boot/dts/mnxlp8xx-hybrid-nae-helinux.dts
@@ -219,5 +219,12 @@
 			num-nodes = <2>;
 		};
 	};
+	frequency-config {
+		nae = <500>;
+		sae = <250>;
+		rsa = <250>;
+		dtre = <250>;
+		cde = <250>;
+	};
 };
 
diff --git a/arch/mips/boot/dts/mnxlp8xx-linux.dts b/arch/mips/boot/dts/mnxlp8xx-linux.dts
index e81ec9d..aa5b16e 100644
--- a/arch/mips/boot/dts/mnxlp8xx-linux.dts
+++ b/arch/mips/boot/dts/mnxlp8xx-linux.dts
@@ -101,5 +101,12 @@
 		};
 
 	};
+	frequency-config {
+		nae = <500>;
+		sae = <250>;
+		rsa = <250>;
+		dtre = <250>;
+		cde = <250>;
+	};
 };
 
diff --git a/arch/mips/boot/dts/xlp2xx-helinux.dts b/arch/mips/boot/dts/xlp2xx-helinux.dts
index e0ee634..2fe3aad 100644
--- a/arch/mips/boot/dts/xlp2xx-helinux.dts
+++ b/arch/mips/boot/dts/xlp2xx-helinux.dts
@@ -143,5 +143,12 @@
 		};
 
 	};
+	frequency-config {
+		nae = <250>;
+		sae = <500>;
+		rsa = <500>;
+		dtre = <250>;
+		cde = <250>;
+	};
 };
 
diff --git a/arch/mips/boot/dts/xlp2xx-linux-loader.dts b/arch/mips/boot/dts/xlp2xx-linux-loader.dts
index 8e03dfc..66a1f25 100644
--- a/arch/mips/boot/dts/xlp2xx-linux-loader.dts
+++ b/arch/mips/boot/dts/xlp2xx-linux-loader.dts
@@ -128,5 +128,12 @@
 		};
 
 	};
+	frequency-config {
+		nae = <250>;
+		sae = <500>;
+		rsa = <500>;
+		dtre = <250>;
+		cde = <250>;
+	};
 };
 
diff --git a/arch/mips/boot/dts/xlp2xx-linux.dts b/arch/mips/boot/dts/xlp2xx-linux.dts
index 03d72b8..8fb165d 100644
--- a/arch/mips/boot/dts/xlp2xx-linux.dts
+++ b/arch/mips/boot/dts/xlp2xx-linux.dts
@@ -115,5 +115,12 @@
                };
 
 	};
+	frequency-config {
+		nae = <250>;
+		sae = <500>;
+		rsa = <500>;
+		dtre = <250>;
+		cde = <250>;
+	};
 };
 
diff --git a/arch/mips/boot/dts/xlp2xx-multi-os.dts b/arch/mips/boot/dts/xlp2xx-multi-os.dts
index b35f472..9e0d361 100644
--- a/arch/mips/boot/dts/xlp2xx-multi-os.dts
+++ b/arch/mips/boot/dts/xlp2xx-multi-os.dts
@@ -211,5 +211,12 @@
 		};
 
 	};
+	frequency-config {
+		nae = <250>;
+		sae = <500>;
+		rsa = <500>;
+		dtre = <250>;
+		cde = <250>;
+	};
 };
 
diff --git a/arch/mips/boot/dts/xlp2xx-netos.dts b/arch/mips/boot/dts/xlp2xx-netos.dts
index 4326486..d310286 100644
--- a/arch/mips/boot/dts/xlp2xx-netos.dts
+++ b/arch/mips/boot/dts/xlp2xx-netos.dts
@@ -93,4 +93,11 @@
 		};
 
 	};
+	frequency-config {
+		nae = <250>;
+		sae = <500>;
+		rsa = <500>;
+		dtre = <250>;
+		cde = <250>;
+	};
 };
diff --git a/arch/mips/boot/dts/xlp2xx-tiny.dts b/arch/mips/boot/dts/xlp2xx-tiny.dts
index 2af9ee8..5ec1daf 100644
--- a/arch/mips/boot/dts/xlp2xx-tiny.dts
+++ b/arch/mips/boot/dts/xlp2xx-tiny.dts
@@ -150,4 +150,11 @@
 	   };
 	};
    };	
+	frequency-config {
+		nae = <250>;
+		sae = <250>;
+		rsa = <250>;
+		dtre = <250>;
+		cde = <250>;
+	};
 };
diff --git a/arch/mips/boot/dts/xlp3xx-helinux.dts b/arch/mips/boot/dts/xlp3xx-helinux.dts
index 1256501..7d03941 100644
--- a/arch/mips/boot/dts/xlp3xx-helinux.dts
+++ b/arch/mips/boot/dts/xlp3xx-helinux.dts
@@ -149,5 +149,12 @@
 		};
 
 	};
+	frequency-config {
+		nae = <250>;
+		sae = <250>;
+		rsa = <250>;
+		dtre = <250>;
+		cde = <250>;
+	};
 };
 
diff --git a/arch/mips/boot/dts/xlp3xx-linux-loader.dts b/arch/mips/boot/dts/xlp3xx-linux-loader.dts
index e260f23..ea9102e 100644
--- a/arch/mips/boot/dts/xlp3xx-linux-loader.dts
+++ b/arch/mips/boot/dts/xlp3xx-linux-loader.dts
@@ -128,5 +128,12 @@
 		};
 
 	};
+	frequency-config {
+		nae = <250>;
+		sae = <250>;
+		rsa = <250>;
+		dtre = <250>;
+		cde = <250>;
+	};
 };
 
diff --git a/arch/mips/boot/dts/xlp3xx-linux.dts b/arch/mips/boot/dts/xlp3xx-linux.dts
index 07fec56..c6a06f0 100644
--- a/arch/mips/boot/dts/xlp3xx-linux.dts
+++ b/arch/mips/boot/dts/xlp3xx-linux.dts
@@ -118,5 +118,12 @@
                };
 
 	};
+	frequency-config {
+		nae = <250>;
+		sae = <250>;
+		rsa = <250>;
+		dtre = <250>;
+		cde = <250>;
+	};
 };
 
diff --git a/arch/mips/boot/dts/xlp3xx-multi-os.dts b/arch/mips/boot/dts/xlp3xx-multi-os.dts
index 3b38cfc..6290ec3 100644
--- a/arch/mips/boot/dts/xlp3xx-multi-os.dts
+++ b/arch/mips/boot/dts/xlp3xx-multi-os.dts
@@ -214,5 +214,12 @@
 		};
 
 	};
+	frequency-config {
+		nae = <250>;
+		sae = <250>;
+		rsa = <250>;
+		dtre = <250>;
+		cde = <250>;
+	};
 };
 
diff --git a/arch/mips/boot/dts/xlp3xx-netos.dts b/arch/mips/boot/dts/xlp3xx-netos.dts
index 96b66e7..f7efff9 100644
--- a/arch/mips/boot/dts/xlp3xx-netos.dts
+++ b/arch/mips/boot/dts/xlp3xx-netos.dts
@@ -96,4 +96,11 @@
 		};
 
 	};
+	frequency-config {
+		nae = <250>;
+		sae = <250>;
+		rsa = <250>;
+		dtre = <250>;
+		cde = <250>;
+	};
 };
diff --git a/arch/mips/boot/dts/xlp3xx-tiny.dts b/arch/mips/boot/dts/xlp3xx-tiny.dts
index da63a96..b862b82 100644
--- a/arch/mips/boot/dts/xlp3xx-tiny.dts
+++ b/arch/mips/boot/dts/xlp3xx-tiny.dts
@@ -149,4 +149,11 @@
         	};
 	   };
 	};
+	frequency-config {
+		nae = <250>;
+		sae = <250>;
+		rsa = <250>;
+		dtre = <250>;
+		cde = <250>;
+	};
 };
diff --git a/arch/mips/boot/dts/xlp4xx-helinux.dts b/arch/mips/boot/dts/xlp4xx-helinux.dts
index b5a5f8d..5891eac 100644
--- a/arch/mips/boot/dts/xlp4xx-helinux.dts
+++ b/arch/mips/boot/dts/xlp4xx-helinux.dts
@@ -210,5 +210,12 @@
 		};
 
 	};
+	frequency-config {
+		nae = <500>;
+		sae = <250>;
+		rsa = <250>;
+		dtre = <250>;
+		cde = <250>;
+	};
 };
 
diff --git a/arch/mips/boot/dts/xlp4xx-linux.dts b/arch/mips/boot/dts/xlp4xx-linux.dts
index 21262f9..57f2deb 100644
--- a/arch/mips/boot/dts/xlp4xx-linux.dts
+++ b/arch/mips/boot/dts/xlp4xx-linux.dts
@@ -98,5 +98,12 @@
 		};
 
 	};
+	frequency-config {
+		nae = <500>;
+		sae = <250>;
+		rsa = <250>;
+		dtre = <250>;
+		cde = <250>;
+	};
 };
 
diff --git a/arch/mips/boot/dts/xlp4xx-multi-os.dts b/arch/mips/boot/dts/xlp4xx-multi-os.dts
index a4df8a0..bd43311 100644
--- a/arch/mips/boot/dts/xlp4xx-multi-os.dts
+++ b/arch/mips/boot/dts/xlp4xx-multi-os.dts
@@ -214,5 +214,12 @@
 		};
 
 	};
+	frequency-config {
+		nae = <500>;
+		sae = <250>;
+		rsa = <250>;
+		dtre = <250>;
+		cde = <250>;
+	};
 };
 
diff --git a/arch/mips/boot/dts/xlp4xx-netos.dts b/arch/mips/boot/dts/xlp4xx-netos.dts
index 6684036..acd57b9 100644
--- a/arch/mips/boot/dts/xlp4xx-netos.dts
+++ b/arch/mips/boot/dts/xlp4xx-netos.dts
@@ -95,4 +95,11 @@
 		};
 
 	};
+	frequency-config {
+		nae = <500>;
+		sae = <250>;
+		rsa = <250>;
+		dtre = <250>;
+		cde = <250>;
+	};
 };
diff --git a/arch/mips/boot/dts/xlp4xx-tiny.dts b/arch/mips/boot/dts/xlp4xx-tiny.dts
index f4bc675..061f487 100644
--- a/arch/mips/boot/dts/xlp4xx-tiny.dts
+++ b/arch/mips/boot/dts/xlp4xx-tiny.dts
@@ -123,4 +123,11 @@
                 };
             };
 	};
+	frequency-config {
+		nae = <500>;
+		sae = <250>;
+		rsa = <250>;
+		dtre = <250>;
+		cde = <250>;
+	};
 };
diff --git a/arch/mips/boot/dts/xlp8xx-helinux-nat.dts b/arch/mips/boot/dts/xlp8xx-helinux-nat.dts
index 369d2c7..0109762 100644
--- a/arch/mips/boot/dts/xlp8xx-helinux-nat.dts
+++ b/arch/mips/boot/dts/xlp8xx-helinux-nat.dts
@@ -160,5 +160,12 @@
 		};
 
 	};
+	frequency-config {
+		nae = <500>;
+		sae = <250>;
+		rsa = <250>;
+		dtre = <250>;
+		cde = <250>;
+	};
 };
 
diff --git a/arch/mips/boot/dts/xlp8xx-helinux-snort.dts b/arch/mips/boot/dts/xlp8xx-helinux-snort.dts
index 5e5e353..c67db6b 100644
--- a/arch/mips/boot/dts/xlp8xx-helinux-snort.dts
+++ b/arch/mips/boot/dts/xlp8xx-helinux-snort.dts
@@ -241,5 +241,12 @@
 		};
 
 	};
+	frequency-config {
+		nae = <500>;
+		sae = <250>;
+		rsa = <250>;
+		dtre = <250>;
+		cde = <250>;
+	};
 };
 
diff --git a/arch/mips/boot/dts/xlp8xx-helinux.dts b/arch/mips/boot/dts/xlp8xx-helinux.dts
index c9fa889..d706db4 100644
--- a/arch/mips/boot/dts/xlp8xx-helinux.dts
+++ b/arch/mips/boot/dts/xlp8xx-helinux.dts
@@ -249,5 +249,12 @@
 		};
 
 	};
+	frequency-config {
+		nae = <500>;
+		sae = <250>;
+		rsa = <250>;
+		dtre = <250>;
+		cde = <250>;
+	};
 };
 
diff --git a/arch/mips/boot/dts/xlp8xx-hybrid-nae-netos.dts b/arch/mips/boot/dts/xlp8xx-hybrid-nae-netos.dts
index 5bc4266..085f31a 100644
--- a/arch/mips/boot/dts/xlp8xx-hybrid-nae-netos.dts
+++ b/arch/mips/boot/dts/xlp8xx-hybrid-nae-netos.dts
@@ -158,5 +158,12 @@
 		};
 
 	};
+	frequency-config {
+		nae = <500>;
+		sae = <250>;
+		rsa = <250>;
+		dtre = <250>;
+		cde = <250>;
+	};
 };
 
diff --git a/arch/mips/boot/dts/xlp8xx-linux-loader.dts b/arch/mips/boot/dts/xlp8xx-linux-loader.dts
index 31fe663..6e88397 100644
--- a/arch/mips/boot/dts/xlp8xx-linux-loader.dts
+++ b/arch/mips/boot/dts/xlp8xx-linux-loader.dts
@@ -222,5 +222,12 @@
 		};
 
 	};
+	frequency-config {
+		nae = <500>;
+		sae = <250>;
+		rsa = <250>;
+		dtre = <250>;
+		cde = <250>;
+	};
 };
 
diff --git a/arch/mips/boot/dts/xlp8xx-linux.dts b/arch/mips/boot/dts/xlp8xx-linux.dts
index c8606cb..e30843e 100644
--- a/arch/mips/boot/dts/xlp8xx-linux.dts
+++ b/arch/mips/boot/dts/xlp8xx-linux.dts
@@ -97,5 +97,12 @@
 		};
 
 	};
+	frequency-config {
+		nae = <500>;
+		sae = <250>;
+		rsa = <250>;
+		dtre = <250>;
+		cde = <250>;
+	};
 };
 
diff --git a/arch/mips/boot/dts/xlp8xx-multi-os.dts b/arch/mips/boot/dts/xlp8xx-multi-os.dts
index ce8960d..7594830 100644
--- a/arch/mips/boot/dts/xlp8xx-multi-os.dts
+++ b/arch/mips/boot/dts/xlp8xx-multi-os.dts
@@ -258,5 +258,12 @@
 		};
 
     };
+    frequency-config {
+	nae = <500>;
+	sae = <250>;
+	rsa = <250>;
+	dtre = <250>;
+	cde = <250>;
+    };
 };
 
diff --git a/arch/mips/boot/dts/xlp8xx-netos-nat.dts b/arch/mips/boot/dts/xlp8xx-netos-nat.dts
index 68ed779..5e253fd 100644
--- a/arch/mips/boot/dts/xlp8xx-netos-nat.dts
+++ b/arch/mips/boot/dts/xlp8xx-netos-nat.dts
@@ -88,4 +88,11 @@
 		};
 
 	};
+	frequency-config {
+		nae = <500>;
+		sae = <250>;
+		rsa = <250>;
+		dtre = <250>;
+		cde = <250>;
+	};
 };
diff --git a/arch/mips/boot/dts/xlp8xx-netos.dts b/arch/mips/boot/dts/xlp8xx-netos.dts
index 5d46797..6fc14aa 100644
--- a/arch/mips/boot/dts/xlp8xx-netos.dts
+++ b/arch/mips/boot/dts/xlp8xx-netos.dts
@@ -135,4 +135,11 @@
 		};
 
 	};
+	frequency-config {
+		nae = <500>;
+		sae = <250>;
+		rsa = <250>;
+		dtre = <250>;
+		cde = <250>;
+	};
 };
diff --git a/arch/mips/boot/dts/xlp8xx-pcienet-netos.dts b/arch/mips/boot/dts/xlp8xx-pcienet-netos.dts
index debbb53..0dba903 100644
--- a/arch/mips/boot/dts/xlp8xx-pcienet-netos.dts
+++ b/arch/mips/boot/dts/xlp8xx-pcienet-netos.dts
@@ -128,4 +128,11 @@
 	soc {
 
 	};
+	frequency-config {
+		nae = <500>;
+		sae = <250>;
+		rsa = <250>;
+		dtre = <250>;
+		cde = <250>;
+	};
 };
diff --git a/arch/mips/boot/dts/xlp8xx-tiny.dts b/arch/mips/boot/dts/xlp8xx-tiny.dts
index f667c8c..b9ce648 100644
--- a/arch/mips/boot/dts/xlp8xx-tiny.dts
+++ b/arch/mips/boot/dts/xlp8xx-tiny.dts
@@ -125,4 +125,11 @@
                 };
             };
 	};
+	frequency-config {
+		nae = <250>;
+		sae = <250>;
+		rsa = <250>;
+		dtre = <250>;
+		cde = <250>;
+	};
 };
-- 
1.7.1

