From 074327d2e8e21069392de97260d28b1142ac897d Mon Sep 17 00:00:00 2001
From: Vikas Gupta <vikas.gupta@broadcom.com>
Date: Mon, 15 Oct 2012 20:14:14 +0530
Subject: XLP3XX-XMC: Support for XLP316L XMC config files.

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/boot/dts/xlp3xx-xmc-linux.dts b/arch/mips/boot/dts/xlp3xx-xmc-linux.dts
new file mode 100644
index 0000000..fe27d51
--- /dev/null
+++ b/arch/mips/boot/dts/xlp3xx-xmc-linux.dts
@@ -0,0 +1,122 @@
+/* XLP3XX Device Tree Source
+ *
+ */
+
+/dts-v1/;
+/ {
+	model = "MIPS,XLP3XX";
+	compatible = "NETL,XLP3XX_A0";
+
+	#address-cells = <1>;
+	#size-cells = <1>;
+
+	hypervisor {
+		hypervisor-name = "Xen";
+		alloc_dom0_memory = <0>;
+		bootargs = "ncores=4 dom0_loadaddr=0x72000000 dom0_size=0x1c000000 dom0_cpumask=0xffff -- ";
+		domain_heap = <0x80000000 0x20000000>;
+	};
+	doms {
+		#address-cells = <1>;
+		#size-cells = <1>;
+		dom@0 {
+			device_type = "domain";
+			os = "linux";
+
+			#address-cells = <1>;
+			#size-cells = <1>;
+
+			cpu {
+				onlinemask = <0xffff>;
+                                nae-rx-vc = <0>;
+                                nae-fb-vc = <1>;
+				napi-vc-mask = <0xb>;
+				sae-rx-vc = <0>; /*  */
+				sae-rx-sync-vc  = <3>; /* should be exclusive */
+				ipsec-async-vc  = <3>;
+				ipsec-sync-vc   = <2>; /* should be exclusive either  
+							  ipsec-sync-vc  or soc/srio@node-0  */
+
+			};
+			uart {
+				id = <0>;
+				owner = <1>;
+				sharedcfg = <0x1f000000>;
+			};
+			memory {
+				reg = <0x01000000 0x0B000000	// 176M at 16M
+					0x20000000 0x20000000>;  // 2586M at 512M
+			};
+                        fmn {
+				/* make the node mask to 0xffffffff for SRIO */  
+                                node_0_vc_mask = <0xbbbbbbbb 0xbbbbbbbb 0xbbbbbbbb 0xbbbbbbbb>;
+                                node_1_vc_mask = <0xbbbbbbbb 0xbbbbbbbb 0xbbbbbbbb 0xbbbbbbbb>;
+                                node_2_vc_mask = <0xbbbbbbbb 0xbbbbbbbb 0xbbbbbbbb 0xbbbbbbbb>;
+                                node_3_vc_mask = <0xbbbbbbbb 0xbbbbbbbb 0xbbbbbbbb 0xbbbbbbbb>;
+                        };
+			pic {
+			};
+			nae {
+				freein-fifo-mask = <0xff 0 0 0>;
+				/* node offsets and size */
+				vfbtbl-sw-offset = <0 0 0 0>;
+				vfbtbl-sw-nentries = <32 32 32 32>;
+				/* hw entries */
+				vfbtbl-hw-offset = <32 32 32 32>;
+				vfbtbl-hw-nentries = <20 20 20 20>;
+			};
+			owner-config {
+				nae = <0x1>;   // currently for node 0
+				sae = <0x1>;
+				rsa = <0x1>;
+				fmn = <0x1>;
+			};
+
+
+		};
+	};
+
+	chosen {
+		bootargs = " console=ttyS0,115200 mem=512m@3584m rhash_entries=1048576";
+	};
+
+	/* These binaries are downloaded at the resp physical memory locations
+	 * by their corresponding bootstrapping loaders. For example,
+	 * u-boot is loaded by x-load at 193M, sysconfig is loaded by u-boot
+	 * at 1M, hypervisor by u-boot hypervisor at 136M.
+	 * On simulator all these binaries are pre-loaded by using '-F' option
+	 * for convenience. On silicon, they will be loaded by their corresponding
+	 * temporary download locations the stage-1/stage-2 firmware
+	 */
+	firmware-download {
+		u-boot        = <0x0c100000 0x3f00000>;  /* 63M @ 193M */
+		sysconfig     = <0x00100000 0x00100000>; /* 1M @ 1M */
+		hypervisor    = <0x08800000 0x00800000>; /* 8M @ 136M */
+		dom0          = <0x09000000 0x03000000>; /* 48M @ 144M */
+		domU-ramdisk  = <0x60000000 0x60800000>; /* 8M @ 1536M */
+	};
+
+	soc {
+		nodes {
+			num-nodes = <1>;
+		};
+		srio@node-0{
+                       /* type = "master" or "device". master is responsible for enumeration*/
+                       type = "master";
+                       /* SRIO_MODE_x1 = 1(not supported), SRIO_MODE_x4 = 2 */
+                       mode = <2>;
+                       /* BAUD_1250M = 1 (1.25G)
+                          BAUD_2500M = 2 (25.G)
+                          BAUD_3125M = 3 (3.125G)
+                          BAUD_5000M = 4 (5G)
+                          BAUD_6250M = 5 (6.25G) */
+                       baud-rate = <5>;
+                       rxvc = <2>;
+                       msgdstid = <2>;
+                       /* sys-size = 1 for 16bit device id, 0 for 8 bit device id */
+                       sys-size = <1>;
+               };
+
+	};
+};
+
diff --git a/arch/mips/boot/dts/xlp3xx-xmc-tiny.dts b/arch/mips/boot/dts/xlp3xx-xmc-tiny.dts
new file mode 100644
index 0000000..d946391
--- /dev/null
+++ b/arch/mips/boot/dts/xlp3xx-xmc-tiny.dts
@@ -0,0 +1,125 @@
+/* XLP3XX Device Tree Source
+ *
+ */
+
+/dts-v1/;
+/ {
+	model = "MIPS,XLP3XX";
+	compatible = "NETL,XLP3XX_A0";
+	#address-cells = <1>;
+	#size-cells = <1>;
+	doms {
+		#address-cells = <1>;
+		#size-cells = <1>;
+		dom@0 {
+			device_type = "domain";
+			heap-size = <0x0 0x4000000>;
+			shared-mem = <0x0 0x80000000 0x0 0x20000000>;
+			#address-cells = <1>;
+			#size-cells = <1>;
+			cpu {
+				onlinemask = <0xf>;
+				nae-rx-vc = <0>;
+				nae-fb-vc = <3>;
+			};
+			uart {
+				id = <0>;
+			};
+			memory {
+				reg = <0x00100000 0x0FF00000	// 255M at 1M
+					   0x20000000 0x20000000>;  // 512M at 512M
+			};
+			pic {
+			};
+			owner-config {
+				nae = <0x1>;
+				fmn = <0x1>;
+			};
+		};
+	};
+
+	chosen {
+		bootargs = " console=ttyS0,115200 ";
+	};
+
+	soc {
+	    nae@node-0 {
+		model = "MIPS,XLP3XX NAE CFG";
+		compatible = "NETL,XLP3XX_A0";
+        	#address-cells = <1>;
+	        #size-cells = <1>;
+        	frequency = <250>;
+
+	        /include/ "ucore.dts"
+
+	        // VFBID MAP: Upto 127 entries
+        	// (each entry is a pair of (vfbid , dest-vc)
+	        // Legal range: (vfbid (0 - 126), dest-vc (0 - 4095))
+        	vfbid-config {
+                	vfbid-map = <
+	                     0    0    1   1   2   2   3   3
+        	        >;
+	        };
+
+        	// Packet Ordering Engine (POE)
+	        poe {
+
+        	        mode = "bypass";
+
+                	// Each vector is 512 bit with masb indicating vc 512 and lsb indicating vc 0
+	                // Format : 512-bit-vector is specified as 16 32-bit words
+        	        // Left most word has the vc range 511-479 right most word has vc range 31 - 0
+                	// Each word has the MSB select higer vc number and LSB select lower vc num
+	                distribution_vectors {
+                                     dv0  = <0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0x1>;
+                                     dv1  = <0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0>;
+                                     dv2  = <0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0>;
+                                     dv3  = <0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0>;
+                                     dv4  = <0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0>;
+                                     dv5  = <0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0>;
+                                     dv6  = <0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0>;
+                                     dv7  = <0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0>;
+                                     dv8  = <0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0>;
+                                     dv9  = <0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0>;
+                                     dv10 = <0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0>;
+                                     dv11 = <0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0>;
+                                     dv12 = <0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0>;
+                                     dv13 = <0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0>;
+                                     dv14 = <0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0>;
+                                     dv15 = <0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0>;
+        	        };
+	        };
+
+	        complex@0 {
+        	        device_type = "nae-complex";
+                	mode = "sgmii";
+			cpld = "no";
+
+        	        sgmii {
+                	        loopback = <0 0 0 0>;
+
+                        	// Max ingress fifo size 256 units (size of one unit is 64 byte)
+	                        iface-fifo-size = <13 13 13 13>;
+
+        	                ext-phy-addr = <4 7 2 1>;
+                	        ext-phy-bus = <0 0 0 0>;
+                        	// Max parser sequence fifo size 1024 packets
+	                        // (if 1588 Timestamp is not required, then max size increases to 2048)
+        	                num-channels = <1 1 1 1>;
+                	        parser-sequence-fifo-size = <30 30 30 30>;
+
+                        	// Since rx_buffer is per context, we can have a few different ways
+	                        //  to populate the operand
+        	                rx-buffer-size = <128 128 128 128>;
+
+
+                	        // Max available descriptors are 1024 (across all complexes).
+                        	// Per port num_free_descriptors must be even number
+	                        num-free-descs = <52 52 52 52>;
+        	                free-desc-size = <2048 2048 2048 2048>;
+                	        ucore-mask = <0xff 0xff 0xff 0xff>;
+	                };
+        	};
+	   };
+	};
+};
-- 
1.7.1

