From 7533098b8baa608dd5d55f15164d7c26d115787d Mon Sep 17 00:00:00 2001
From: Vikas Gupta <vikas.gupta@broadcom.com>
Date: Fri, 2 Nov 2012 17:19:44 +0530
Subject: xlp3xx-xmc: fix for phy addresses.

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/boot/dts/xlp3xx-xmc-tiny.dts b/arch/mips/boot/dts/xlp3xx-xmc-tiny.dts
index d946391..ff175f9 100644
--- a/arch/mips/boot/dts/xlp3xx-xmc-tiny.dts
+++ b/arch/mips/boot/dts/xlp3xx-xmc-tiny.dts
@@ -101,7 +101,7 @@
                         	// Max ingress fifo size 256 units (size of one unit is 64 byte)
 	                        iface-fifo-size = <13 13 13 13>;
 
-        	                ext-phy-addr = <4 7 2 1>;
+        	                ext-phy-addr = <1 7 6 5>;
                 	        ext-phy-bus = <0 0 0 0>;
                         	// Max parser sequence fifo size 1024 packets
 	                        // (if 1588 Timestamp is not required, then max size increases to 2048)
-- 
1.7.1

