From 333f2eb334806e7a0e30efbb504cbfec5b9ed020 Mon Sep 17 00:00:00 2001
From: Mehul Vora <mehulv@broadcom.com>
Date: Thu, 10 Jan 2013 09:27:27 +0530
Subject: xlp-pcie-net : Added dtb files for xlp pcie network driver.

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/boot/dts/xlp8xx-helinux-pcienet.dts b/arch/mips/boot/dts/xlp8xx-helinux-pcienet.dts
new file mode 100644
index 0000000..4617a4e
--- /dev/null
+++ b/arch/mips/boot/dts/xlp8xx-helinux-pcienet.dts
@@ -0,0 +1,122 @@
+/* XLP8XX Device Tree Source 
+ *
+ */
+
+/dts-v1/;
+/ {
+	model = "MIPS,XLP8XX";
+	compatible = "NETL,XLP8XX_A0";
+
+	#address-cells = <1>;
+	#size-cells = <1>;
+
+	hypervisor {
+		hypervisor-name = "Xen";
+		alloc_dom0_memory = <1>;
+		bootargs = "dom0_loadaddr=0x72000000 dom0_cpumask=0x0f -- ";
+		domain_heap = <0x50000000 0x20000000>;
+	};
+
+	doms {
+		#address-cells = <1>;
+		#size-cells = <1>;
+		dom@0 {
+			device_type = "domain";
+			os = "linux";
+
+			#address-cells = <1>;
+			#size-cells = <1>;
+
+			cpu {
+				onlinemask = <0xffff>;
+                napi-vc-mask = <0x0>;
+			};
+			uart {
+				id = <0>;
+				owner = <1>;
+				sharedcfg = <0x1f000000>;
+			};
+			memory {
+				reg = <0x01000000 0x0B000000	// 176M at 16M
+					0x20000000 0x26000000>;  //  608MB @ 512MB
+			};
+			fmn {
+           		node_0_vc_mask = <0x00000000 0x00000000 0x00000000 0x00000000>;
+           		node_1_vc_mask = <0x00000000 0x00000000 0x00000000 0x00000000>;
+           		node_2_vc_mask = <0x00000000 0x00000000 0x00000000 0x00000000>;
+           		node_3_vc_mask = <0x00000000 0x00000000 0x00000000 0x00000000>;
+			};
+			pic {
+			};
+			owner-config {
+				nae = <0x0>;   // currently for node 0
+				sae = <0x0>;
+				rsa = <0x0>;
+				fmn = <0x1>;
+			};
+		};
+
+		dom@1 {
+			device_type = "domain";
+			os = "netos";
+
+			heap-size = <0x0 0x2000000>; // 32M
+			shared-mem = <0x0 0x50000000 0x0 0x20000000>; // 512M @ 1.25GB
+
+			#address-cells = <1>;
+			#size-cells = <1>;
+
+			app-bootcmd = "smpload";
+			app-loadaddr = "0xf0000000"; // 3.75GB
+
+			kernel {
+				/* Kernel Memory allocated from xen domain-heap */
+				size = <0x02000000>;  // 32M
+
+				/* Location in Dom0 ramdisk */
+				location = "/tmp/xen/netos.elf";
+
+				/* Domain destroy currently doesn't work with hyperexec */
+				dom_destroy_hack = <0>;
+			};
+
+			cpu {
+				onlinemask = <0xffff>;
+                nae-rx-vc = <2>;
+                nae-fb-vc = <3>;
+			};
+			uart {
+				id = <0>;
+			};
+			memory {
+				reg = <0x60000000 0x60000000>;  // 1.5GB @ 1.5GB
+			};
+			app-param{
+				onlinemask = <0xffff>;
+			};
+			owner-config {
+				nae = <0x1>;
+				fmn = <0x0>;
+				sae = <0x0>;
+				rsa = <0x0>;
+			};
+		};
+	};
+
+	chosen {
+		bootargs = " xlp_noi2c console=ttyS0,115200 autoboot_pciedrv";
+	};
+
+	soc {
+		nodes {
+			num-nodes = <1>;
+		};
+	};
+	frequency-config {
+		nae = <500>;
+		sae = <500>;
+		rsa = <500>;
+		dtre = <250>;
+		cde = <250>;
+	};
+};
diff --git a/arch/mips/boot/dts/xlp8xx-pcienet-netos.dts b/arch/mips/boot/dts/xlp8xx-pcienet-netos.dts
index 109e418..1db15c3 100644
--- a/arch/mips/boot/dts/xlp8xx-pcienet-netos.dts
+++ b/arch/mips/boot/dts/xlp8xx-pcienet-netos.dts
@@ -25,7 +25,7 @@
 			os = "netos";
 
 			heap-size = <0x0 0x2000000>; // 32M
-			shared-mem = <0x0 0x40000000 0x0 0x8000000>; // 128M @ 1GB
+			shared-mem = <0x0 0x50000000 0x0 0x20000000>; // 512M @ 1.25GB
 
 			#address-cells = <1>;
 			#size-cells = <1>;
@@ -45,7 +45,7 @@
 			};
 
 			cpu {
-				onlinemask = <0xffffffff>;
+				onlinemask = <0xffff>;
                                 nae-rx-vc = <1>;
                                 nae-fb-vc = <3>;
 				sae-rx-vc = <0>; /* sharing with nae */
@@ -62,7 +62,7 @@
 				reg = <0x20000000 0x20000000>;  // 512M @ 512M
 			};
 			app-param{
-                                onlinemask = <0xffffffff>;
+                                onlinemask = <0xffff>;
                         };
 
 			/* In hyperexec-linux mode, vc_mask needs to be synchronized with linux fmn node! */
-- 
1.7.1

