From 94d54e4c35884428da1ba95206109ebeecad5030 Mon Sep 17 00:00:00 2001
From: Vikas Gupta <vikas.gupta@broadcom.com>
Date: Tue, 2 Apr 2013 10:41:03 -0700
Subject: dts: Updated the dts files to follow the standard Linux convention.

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/boot/dts/mn4xlp8xx-helinux-nat.dts b/arch/mips/boot/dts/mn4xlp8xx-helinux-nat.dts
index e2297eb..26465bd 100644
--- a/arch/mips/boot/dts/mn4xlp8xx-helinux-nat.dts
+++ b/arch/mips/boot/dts/mn4xlp8xx-helinux-nat.dts
@@ -39,10 +39,6 @@
 				owner = <1>;
 				sharedcfg = <0x1f000000>;
 			};
-			memory {
-				reg = <0x01000000 0x0B000000	// 176M at 16M
-					0x20000000 0x40000000>;  //  1GB @ 512MB
-			};
                         fmn {
                                 node_0_vc_mask = <0x33333333 0x33333333 0x33333333 0x33333333>;
                                 node_1_vc_mask = <0x33333333 0x33333333 0x33333333 0x33333333>;
@@ -198,7 +194,7 @@
 	};
 
 	chosen {
-		bootargs = " console=ttyS0,115200 mem=2048m@8848m";
+		bootargs = " console=ttyS0,115200 ";
 	};
 
 	/* These binaries are downloaded at the resp physical memory locations
diff --git a/arch/mips/boot/dts/mnxlp8xx-hybrid-nae-helinux.dts b/arch/mips/boot/dts/mnxlp8xx-hybrid-nae-helinux.dts
index dd80688..ab27c3f 100644
--- a/arch/mips/boot/dts/mnxlp8xx-hybrid-nae-helinux.dts
+++ b/arch/mips/boot/dts/mnxlp8xx-hybrid-nae-helinux.dts
@@ -46,13 +46,6 @@
 				owner = <1>;
 				sharedcfg = <0x1f000000>;
 			};
-			memory {
-				reg = <0x01000000 0x0B000000
-					0x20000000 0xa0000000>; // 2586M at 512M
-
-			};
-			pic {
-			};
 			nae {
 				freein-fifo-mask = <0x30000 0x30000 0 0>;
 				sh-domains = <1 2>;
@@ -202,7 +195,7 @@
 	};
 
 	chosen {
-		bootargs = " console=ttyS0,115200 mem=512m@3584m mem=2048m@10240m";
+		bootargs = " console=ttyS0,115200 ";
 	};
 
 	/* These binaries are downloaded at the resp physical memory locations
diff --git a/arch/mips/boot/dts/xlp8xx-helinux-snort.dts b/arch/mips/boot/dts/xlp8xx-helinux-snort.dts
index 2f14401..7c23aa9 100644
--- a/arch/mips/boot/dts/xlp8xx-helinux-snort.dts
+++ b/arch/mips/boot/dts/xlp8xx-helinux-snort.dts
@@ -4,12 +4,17 @@
 
 /dts-v1/;
 / {
-	model = "MIPS,XLP8XX";
-	compatible = "NETL,XLP8XX_A0";
-
-	#address-cells = <1>;
-	#size-cells = <1>;
-
+	model = "netlogic,XLP-EVP";
+        compatible = "netlogic,xlp";
+        #address-cells = <2>;
+        #size-cells = <2>;
+	/*memory range for linux. This will not appear in dom0*/
+        memory {
+                device_type = "memory";
+                reg =  <0 0x01000000 0 0x0B000000   
+                        0x0 0xe0000000 0x1 0xc5f00000>; // 7263M@3584M
+        };
+	
 	hypervisor {
 		hypervisor-name = "Xen";
 		alloc_dom0_memory = <0>;
@@ -40,18 +45,12 @@
 				owner = <1>;
 				sharedcfg = <0x1f000000>;
 			};
-			memory {
-				reg = <0x01000000 0x0B000000>;	// 176M at 16M
-//					0xE0000000 0xC0000000>;  // 3GB @ 3.5GB  5263M@3584M
-			};
                         fmn {
                                 node_0_vc_mask = <0x00000000 0x00000000 0x00000000 0x00000000>;
                                 node_1_vc_mask = <0x22222222 0x22222222 0x22222222 0x22222222>;
                                 node_2_vc_mask = <0x22222222 0x22222222 0x22222222 0x22222222>;
                                 node_3_vc_mask = <0x22222222 0x22222222 0x22222222 0x22222222>;
                         };
-			pic {
-			};
 			owner-config {
 				nae = <0x0>;
 				sae = <0x0>;
@@ -216,7 +215,7 @@
 	};
 
 	chosen {
-		bootargs = " console=ttyS0,115200  mem=7263M@3584M";
+		bootargs = " console=ttyS0,115200 ";
 	};
 
 	/* These binaries are downloaded at the resp physical memory locations
-- 
1.7.1

