From 2dc9cb453d8d9905ccd2e46ed03321a725aeee5f Mon Sep 17 00:00:00 2001
From: Sreenidhi B R <sreenira@broadcom.com>
Date: Thu, 30 May 2013 16:13:18 +0530
Subject: pktmem: add "pktmem" section to DTS file, with default values.

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/boot/dts/xlp8xx-helinux.dts b/arch/mips/boot/dts/xlp8xx-helinux.dts
index 08de7bb..b7c59a0 100644
--- a/arch/mips/boot/dts/xlp8xx-helinux.dts
+++ b/arch/mips/boot/dts/xlp8xx-helinux.dts
@@ -15,6 +15,15 @@
                         0 0xE0000000 0 0xC0000000>;
         };
 
+	/* memory range for HEv2 pkt mem. Only one contig memory per node supported */
+	pktmem {
+		device_type = "pktmem";
+		reg = <0 0x08000000 0 0x40000000 /* node0 */
+	               1 0x0 1 0x0               /* node1 */
+		       2 0x0 2 0x0               /* node2 */
+		       3 0x0 3 0x0>;             /* node3 */
+	};
+
 	hypervisor {
 		hypervisor-name = "Xen";
 		alloc_dom0_memory = <0>;
-- 
1.7.1

