From 36bbfea58462497562bc10876d8d408e5c1fd9a6 Mon Sep 17 00:00:00 2001
From: Sreenidhi B R <sreenira@broadcom.com>
Date: Fri, 7 Jun 2013 17:05:25 +0530
Subject: HEv2: pktmem section for xlp8xx-linux.dts

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/boot/dts/xlp8xx-linux.dts b/arch/mips/boot/dts/xlp8xx-linux.dts
index a970fbd..6c2fcd5 100644
--- a/arch/mips/boot/dts/xlp8xx-linux.dts
+++ b/arch/mips/boot/dts/xlp8xx-linux.dts
@@ -16,6 +16,14 @@
                         0 0xe0000000 0 0x20000000>;
         };
 
+	/* memory range for HEv2 pkt mem. Only one contig memory per node supported */
+	pktmem {
+		device_type = "pktmem";
+		reg = <0 0x08000000 0 0x40000000 /* node0 */
+	               1 0x0 1 0x0               /* node1 */
+		       2 0x0 2 0x0               /* node2 */
+		       3 0x0 3 0x0>;             /* node3 */
+	};
 
 	hypervisor {
 		hypervisor-name = "Xen";
-- 
1.7.1

