From 6079a4fb70ace96b3b5de30847a0efb4b3720c39 Mon Sep 17 00:00:00 2001
From: Vikas Gupta <vikas.gupta@broadcom.com>
Date: Tue, 4 Jun 2013 20:02:03 -0700
Subject: xlp9xx: changed the mdio bus address and complex mode. NOTE: CPU mask to be reverted soon.

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/boot/dts/xlp9xx-linux.dts b/arch/mips/boot/dts/xlp9xx-linux.dts
index d970f80..dbba6af 100644
--- a/arch/mips/boot/dts/xlp9xx-linux.dts
+++ b/arch/mips/boot/dts/xlp9xx-linux.dts
@@ -82,7 +82,7 @@
 		 * 20core: netlogic,cpumask=<0xffff 0xffffffff 0xffffffff>
 		 * 15 core, skipping core #5: netlogic,cpumask=<0xffffffff 0xff0fffff>
 		 */
-		netlogic,cpumask = <0xff>;
+		netlogic,cpumask = <0x01>;
 		bootargs = " console=ttyS0,115200  rhash_entries=1048576";
 	};
 
-- 
1.7.1

