From a886d039ab8b42385b4d74612582febabde7c892 Mon Sep 17 00:00:00 2001
From: Ganesan Ramalingam <ganesanr@broadcom.com>
Date: Fri, 14 Jun 2013 00:14:27 +0530
Subject: PNOR: Add support for PNOR on Griffin

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/boot/dts/xlp9xx-linux-tmpl.dts b/arch/mips/boot/dts/xlp9xx-linux-tmpl.dts
index a7261a9..ffc86fe 100644
--- a/arch/mips/boot/dts/xlp9xx-linux-tmpl.dts
+++ b/arch/mips/boot/dts/xlp9xx-linux-tmpl.dts
@@ -50,6 +50,7 @@
 				#size-cells = <1>;
 
 				reg = <1>;		/* Chip Select */
+
 				spi-max-frequency = <40000000>;
 
 				partition@0 {
@@ -75,6 +76,40 @@
 				};
 			};
 		};
+		nor_flash@1,0 {
+			compatible = "cfi-flash";
+			#address-cells = <1>;
+			#size-cells = <1>;
+			bank-width = <2>;
+			reg = <1 0 0x1000000>;
+
+			partition@0 {
+				label = "x-loader";
+				reg = <0x0 0x100000>; /* 1M */
+				read-only;
+			};
+
+			partition@100000 {
+				label = "u-boot";
+				reg = <0x100000 0x100000>; /* 1M */
+			};
+
+			partition@200000 {
+				label = "kernel";
+				reg = <0x200000 0x500000>; /* 5M */
+			};
+
+			partition@700000 {
+				label = "rootfs";
+				reg = <0x700000 0x800000>; /* 8M */
+			};
+
+			partition@f00000 {
+				label = "env";
+				reg = <0xf00000 0x100000>; /* 1M */
+				read-only;
+			};
+		};
 	};
 
 	chosen {
-- 
1.7.1

