From bc86d72cd170d2cced95eff41c8821d5d4d2ed16 Mon Sep 17 00:00:00 2001
From: kopal <kopal@broadcom.com>
Date: Thu, 27 Jun 2013 20:27:04 +0530
Subject: dts: updated dts to add eeprom support

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/boot/dts/xlp9xx-linux-tmpl.dts b/arch/mips/boot/dts/xlp9xx-linux-tmpl.dts
index 4d29271..cbf1f4a 100644
--- a/arch/mips/boot/dts/xlp9xx-linux-tmpl.dts
+++ b/arch/mips/boot/dts/xlp9xx-linux-tmpl.dts
@@ -57,6 +57,11 @@
 				compatible = "dallas,ds1374";
 				reg = <0x68>;
 			};
+			
+			eeprom@57 {
+                                compatible = "xlpge_eeprom";
+                                reg = <0x57>;
+                        };
 
 		};
 		i2c2: i2c@113500 {
-- 
1.7.1

