From 1afa49ed3e9140fcdff0f14524e45ad4a1a73f36 Mon Sep 17 00:00:00 2001
From: Virendra Pathak <vpathak@broadcom.com>
Date: Wed, 10 Jul 2013 09:50:17 +0530
Subject: xlp9xx: Updated helinux dts file for xlp9xx.

Updated:
-memory parameter
-onlinemask
-frequency configuration
[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/boot/dts/xlp9xx-helinux.dts b/arch/mips/boot/dts/xlp9xx-helinux.dts
index 9d76278..81ada94 100644
--- a/arch/mips/boot/dts/xlp9xx-helinux.dts
+++ b/arch/mips/boot/dts/xlp9xx-helinux.dts
@@ -7,8 +7,25 @@
 	model = "netlogic,XLP-GVP";
 	compatible = "netlogic,xlp";
 
-	#address-cells = <1>;
-	#size-cells = <1>;
+	#address-cells = <2>;
+	#size-cells = <2>;
+
+	memory {
+                device_type = "memory";
+                reg =  <0 0x00100000 0 0x0FF00000       // 255M at 1M
+                        0 0x20000000 0 0xa0000000>;     // 2560M at 512M
+        };
+
+        /* memory range for HEv2 pkt mem. Only one contig memory per node supported */
+        pktmem {
+                device_type = "pktmem";
+
+                /* format: node number, 64-bit addr, 64-bit size */
+                reg = <0 0x0 0xe0000000 0x0 0x20000000
+                       1 0x0 0x0 0x0 0x0
+                       2 0x0 0x0 0x0 0x0
+                       3 0x0 0x0 0x0 0x0>;
+        };
 
 	hypervisor {
 		hypervisor-name = "Xen";
@@ -27,7 +44,7 @@
 			#size-cells = <1>;
 
 			cpu {
-				onlinemask = <0x1>;
+				onlinemask = <0x1 0x0 0x0>;
                                 nae-rx-vc = <1>;
                                 nae-fb-vc = <1>;
 				sae-rx-vc = <1>; /* sharing with nae */
@@ -40,10 +57,6 @@
 				owner = <1>;
 				sharedcfg = <0x1f000000>;
 			};
-			memory {
-				reg = <0x01000000 0x0B000000	// 176M at 16M
-					0xE0000000 0xC0000000>;  // 3GB @ 3.5GB
-			};
                         fmn {
                                 node_0_vc_mask = <0x22222222 0x22222222 0x22222222 0x22222222>;
                                 node_1_vc_mask = <0x22222222 0x22222222 0x22222222 0x22222222>;
@@ -65,7 +78,6 @@
 			os = "netos";
 
 			heap-size = <0x0 0x8000000>; // 128M
-			shared-mem = <0x0 0x40000000 0x0 0x10000000>; // 128M @ 1GB
 
 			#address-cells = <1>;
 			#size-cells = <1>;
@@ -85,7 +97,7 @@
 			};
 
 			cpu {
-				onlinemask = <0x1>;
+				onlinemask = <0xffff>;
                                 nae-rx-vc = <0>;
                                 nae-fb-vc = <3>;
 				sae-rx-vc = <0>; /* sharing with nae */
@@ -99,16 +111,21 @@
 			uart {
 				id = <0>;
 			};
-			memory {
-				reg = <0x60000000 0x60000000>;  // 1.5GB @ 1.5GB
-			};
 			app-param{
-				onlinemask = <0x1>;
+				onlinemask = <0xffff>;
 				enable-nae-hw-parser = <0>;
 				enable-nae-jumbo = <0>;
 				enable-nae-p2p = <0>;
 			};
 			nae{
+				freein-fifo-mask = <0xffff 0 0 0>;
+                                /* node offsets and size */
+                                vfbtbl-sw-offset = <0 0 0 0>;
+                                vfbtbl-sw-nentries = <32 32 32 32>;
+                                /* hw entries */
+                                vfbtbl-hw-offset = <32 32 32 32>;
+                                vfbtbl-hw-nentries = <20 20 20 20>;
+                                dummy-pktdata-addr = <0x0 0x0ff00000>; // 255 MB
 			};
 			owner-config {
 				nae = <0x1>;   // currently for node 0
@@ -123,46 +140,6 @@
                                 node_2_vc_mask = <0x99999999 0x99999999 0x99999999 0x99999999>;
                                 node_3_vc_mask = <0x99999999 0x99999999 0x99999999 0x99999999>;
                         };
-            kbp {
-                /* distribute total of 4096 buffer id across cpus. required only for cpus sending requests.
-                 * note : KBP compare/search instructions must use even-numbered buffers. i.e. (buffer_id & 0x1 == 0) is True
-                 */
-                buffers-cpuid-offset-quantity = <  0    0   64
-                                                   1   64   64
-                                                   2  128   64
-                                                   3  192   64
-                                                   4  256   64
-                                                   5  320   64
-                                                   6  384   64
-                                                   7  448   64
-                                                   8  512   64
-                                                   9  576   64
-                                                  10  640   64
-                                                  11  704   64
-                                                  12  768   64
-                                                  13  832   64
-                                                  14  896   64
-                                                  15  960   64
-                                                  16 1024   64
-                                                  17 1088   64
-                                                  18 1152   64
-                                                  19 1216   64
-                                                  20 1280   64
-                                                  21 1344   64
-                                                  22 1408   64
-                                                  23 1472   64
-                                                  24 1536   64
-                                                  25 1600   64
-                                                  26 1664   64
-                                                  27 1728   64
-                                                  28 1792   64
-                                                  29 1856   64
-                                                  30 1920   64
-                                                  31 1984   64>;
-                /* starting address for KBP Driver Data. two 32-bit unsigned integers forms one 64-bit memory address.
-                 * kbp driver requires 2 MBytes of dedicated memory space to store data */
-                memory-offsethigh-offsetlow = <0x00000001 0x00000000>;
-            };
 		};
 
 		dom@2 {
@@ -170,7 +147,6 @@
 			os = "netos";
 
 			heap-size = <0x0 0x8000000>; // 128M
-			shared-mem = <0x0 0x48000000 0x0 0x10000000>; // 128M @ 1152Mb
 
 			#address-cells = <1>;
 			#size-cells = <1>;
@@ -190,7 +166,7 @@
 			};
 
 			cpu {
-				onlinemask = <0xffff0000>;
+				onlinemask = <0x000f>;
                                 nae-rx-vc = <0>;
                                 nae-fb-vc = <3>;
                                 /* Setting this to '1' makes NAE driver to skip buf alloction and initial
@@ -204,14 +180,15 @@
 				reg = <0xE0000000 0x60000000>;  // 1.5Gb @ 3584Mb
 			};
 			app-param{
-				onlinemask = <0xffff0000>;
+				onlinemask = <0x000f>;
 			};
 			nae{
 			};
 			owner-config {
 				nae = <0x0>;
-				sae = <0x0>;
-				rsa = <0x0>;
+				sae = <0x1>;
+				rsa = <0x1>;
+				fmn = <0x1>;
 			};
 
                         fmn {
@@ -224,6 +201,7 @@
 	};
 
 	chosen {
+		netlogic,cpumask = <0x01>;
 		bootargs = " console=ttyS0,115200  ";
 	};
 
@@ -252,7 +230,7 @@
 	frequency-config {
 		nae = <500>;
 		sae = <250>;
-		rsa = <250>;
+		rsa = <667>;
 		dtre = <250>;
 		cde = <250>;
 	};
-- 
1.7.1

