From 947eafa10d90fc630ae308823efecce1616d0f65 Mon Sep 17 00:00:00 2001
From: P. Sadik <psadik@broadcom.com>
Date: Tue, 16 Jul 2013 10:42:24 +0530
Subject: xlp9xx : kvm: Reset initrd parameters to zero.

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/boot/dts/xlp9xx-helinux-guest.dts b/arch/mips/boot/dts/xlp9xx-helinux-guest.dts
index 342cb16..998f7cf 100644
--- a/arch/mips/boot/dts/xlp9xx-helinux-guest.dts
+++ b/arch/mips/boot/dts/xlp9xx-helinux-guest.dts
@@ -68,8 +68,8 @@
 	chosen {
 		netlogic,cpumask = <0x1>;
 		bootargs = " console=ttyS0,115200  rhash_entries=1048576";
-		linux,initrd-start=<0x00c00000>;
-		linux,initrd-end=<0x02c00000>;
+		linux,initrd-start=<0x0>;
+		linux,initrd-end=<0x0>;
 	};
 
 	soc {
diff --git a/arch/mips/boot/dts/xlp9xx-linux-guest.dts b/arch/mips/boot/dts/xlp9xx-linux-guest.dts
index 7cd0b45..5bbc84a 100644
--- a/arch/mips/boot/dts/xlp9xx-linux-guest.dts
+++ b/arch/mips/boot/dts/xlp9xx-linux-guest.dts
@@ -30,8 +30,8 @@
 	chosen {
 		netlogic,cpumask = <0x1>;
 		bootargs = " console=ttyS0,115200  rhash_entries=1048576";
-		linux,initrd-start=<0x00c00000>;
-		linux,initrd-end=<0x02c00000>;
+		linux,initrd-start=<0x0>;
+		linux,initrd-end=<0x0>;
 	};
 
 	soc {
-- 
1.7.1

