From db239b9edda435923edfaed6b37015cf4e34b5de Mon Sep 17 00:00:00 2001
From: Sandip Matte <sandipm@broadcom.com>
Date: Tue, 23 Jul 2013 14:18:43 +0530
Subject: lzs: Add lzs vc

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/boot/dts/xlp9xx-helinux.dts b/arch/mips/boot/dts/xlp9xx-helinux.dts
index e9459f5..510fb8f 100644
--- a/arch/mips/boot/dts/xlp9xx-helinux.dts
+++ b/arch/mips/boot/dts/xlp9xx-helinux.dts
@@ -102,6 +102,7 @@
                                 nae-rx-vc = <0>;
                                 nae-fb-vc = <3>;
 				dtre-rx-vc = <2>;
+				lzs-vc = <2>;
 				sae-rx-vc = <0>; /* sharing with nae */
 				sae-rx-sync-vc = <0>;  /* sharing with nae */
 				/* helinux, as only 1 thread per cpu, it should poll for both nae and sae 
-- 
1.7.1

