From d8a70ad9d284cd02b8878d14f0da051cdde73dba Mon Sep 17 00:00:00 2001
From: Kamlakant Patel <kamlakant.patel@broadcom.com>
Date: Mon, 29 Jul 2013 17:44:22 +0530
Subject: dts: update DTS for SPI-MMC support

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/boot/dts/xlp8xx-linux-tmpl.dts b/arch/mips/boot/dts/xlp8xx-linux-tmpl.dts
index 00c0dcc..4d511fb 100644
--- a/arch/mips/boot/dts/xlp8xx-linux-tmpl.dts
+++ b/arch/mips/boot/dts/xlp8xx-linux-tmpl.dts
@@ -150,6 +150,14 @@
 			#size-cells = <0>;
 			reg = <0 0x3a000 0x100>;
 
+			spi_mmc@0 {
+				compatible = "mmc-spi-slot";
+				reg = <0>;      /* Chip Select */
+
+				voltage-ranges = <3300 3300>;
+				spi-max-frequency = <50000000>;
+			};
+
 			spi_nor@1 {
 				compatible = "spansion,s25sl12801";
 				#address-cells = <1>;
@@ -198,6 +206,14 @@
 					reg = <0x1c00000 0x0>;
 				};
 			};
+
+			spi_mmc@3 {
+				compatible = "mmc-spi-slot";
+				reg = <3>;      /* Chip Select */
+
+				voltage-ranges = <3300 3300>;
+				spi-max-frequency = <50000000>;
+			};
 		};
 	};
 
-- 
1.7.1

