From 6c4fac6a1d01ec6211c5d3a902f0e0f71d6f8fbe Mon Sep 17 00:00:00 2001
From: Prasanna Mumbai <prasanna.mumbai@broadcom.com>
Date: Tue, 3 Sep 2013 10:06:37 +0530
Subject: HEv2 - Poe app enable/disable param

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/boot/dts/xlp2xx-helinux.dts b/arch/mips/boot/dts/xlp2xx-helinux.dts
index 931be6b..24c1af3 100644
--- a/arch/mips/boot/dts/xlp2xx-helinux.dts
+++ b/arch/mips/boot/dts/xlp2xx-helinux.dts
@@ -112,6 +112,7 @@
 				enable-nae-hw-parser = <0>;
 				enable-nae-jumbo = <0>;
 				enable-nae-p2p = <0>;
+				enable-poe = <0>;
 			};
 			nae{
 			};
diff --git a/arch/mips/boot/dts/xlp3xx-helinux.dts b/arch/mips/boot/dts/xlp3xx-helinux.dts
index 19d8adc..be52418 100644
--- a/arch/mips/boot/dts/xlp3xx-helinux.dts
+++ b/arch/mips/boot/dts/xlp3xx-helinux.dts
@@ -115,6 +115,7 @@
 				enable-nae-hw-parser = <0>;
 				enable-nae-jumbo = <0>;
 				enable-nae-p2p = <0>;
+				enable-poe = <0>;
 			};
 			nae{
 			};
diff --git a/arch/mips/boot/dts/xlp8xx-helinux.dts b/arch/mips/boot/dts/xlp8xx-helinux.dts
index 44d33b8..4e7823b 100644
--- a/arch/mips/boot/dts/xlp8xx-helinux.dts
+++ b/arch/mips/boot/dts/xlp8xx-helinux.dts
@@ -120,6 +120,7 @@
 				enable-nae-hw-parser = <0>;
 				enable-nae-jumbo = <0>;
 				enable-nae-p2p = <0>;
+				enable-poe = <0>;
 			};
 			nae{
 			};
diff --git a/arch/mips/boot/dts/xlp9xx-helinux.dts b/arch/mips/boot/dts/xlp9xx-helinux.dts
index cbe51c1..91063ca 100644
--- a/arch/mips/boot/dts/xlp9xx-helinux.dts
+++ b/arch/mips/boot/dts/xlp9xx-helinux.dts
@@ -123,6 +123,7 @@
 				enable-nae-hw-parser = <0>;
 				enable-nae-jumbo = <0>;
 				enable-nae-p2p = <0>;
+				enable-poe = <0>;
 			};
 			nae{
 				freein-fifo-mask = <0xffff 0 0 0>;
-- 
1.7.1

