From b51c0858cf4625d3475e4bb4e5a4b36b89598af3 Mon Sep 17 00:00:00 2001
From: Hareesh R <hareeshr@broadcom.com>
Date: Fri, 6 Sep 2013 09:59:11 +0530
Subject: nae : netsoc nae shared mode changes

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/boot/dts/xlp9xx-helinux.dts b/arch/mips/boot/dts/xlp9xx-helinux.dts
index 5f0e3de..c47cd5a 100644
--- a/arch/mips/boot/dts/xlp9xx-helinux.dts
+++ b/arch/mips/boot/dts/xlp9xx-helinux.dts
@@ -42,7 +42,7 @@
 			#size-cells = <1>;
 
 			cpu {
-				onlinemask = <0x1 0x0 0x0>;
+				onlinemask = <0x1>;
                                 nae-rx-vc = <0>;
                                 nae-fb-vc = <1>;
 				sae-rx-vc = <0>; /* sharing with nae */
diff --git a/arch/mips/boot/dts/xlp9xx-linux.dts b/arch/mips/boot/dts/xlp9xx-linux.dts
index f4bc90d..416d235 100644
--- a/arch/mips/boot/dts/xlp9xx-linux.dts
+++ b/arch/mips/boot/dts/xlp9xx-linux.dts
@@ -45,7 +45,7 @@
 			#cpumask-cells=<3>;
 
 			cpu {
-				onlinemask = <0x1 0x0 0x0>; /* <0xffffffff 0xffffffff 0xffff> */
+				onlinemask = <0x1>; /* */
                                 nae-rx-vc = <0>;
                                 nae-fb-vc = <1>;
 				napi-vc-mask = <0xb>;
@@ -123,7 +123,7 @@
 			device_type = "domain";
 
 			cpu {   
-                                onlinemask = <0xffff 0x0 0x0>; 
+                                onlinemask = <0xffff>; 
                                 nae-rx-vc = <2>;
 			};
 
-- 
1.7.1

