From f3b41d7e9d647ee6088b194fc34ce0e418a33872 Mon Sep 17 00:00:00 2001
From: Virendra Pathak <vpathak@broadcom.com>
Date: Thu, 12 Sep 2013 10:53:59 +0530
Subject: HEv2 : added amp/smp mode option in app-param for HEv2 apps

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/boot/dts/xlp2xx-helinux.dts b/arch/mips/boot/dts/xlp2xx-helinux.dts
index cd7cc00..45abcb8 100644
--- a/arch/mips/boot/dts/xlp2xx-helinux.dts
+++ b/arch/mips/boot/dts/xlp2xx-helinux.dts
@@ -114,6 +114,7 @@
 				enable-nae-p2p = <0>;
 				enable-poe = <0>;
 				enable-microstruct = <0>;
+				launch-mode = "smp";	/* smp or amp */
 			};
 			nae{
 			};
diff --git a/arch/mips/boot/dts/xlp3xx-helinux.dts b/arch/mips/boot/dts/xlp3xx-helinux.dts
index d061fdb..f18742f 100644
--- a/arch/mips/boot/dts/xlp3xx-helinux.dts
+++ b/arch/mips/boot/dts/xlp3xx-helinux.dts
@@ -117,6 +117,7 @@
 				enable-nae-p2p = <0>;
 				enable-poe = <0>;
 				enable-microstruct = <0>;
+				launch-mode = "smp";	/* smp or amp */
 			};
 			nae{
 			};
diff --git a/arch/mips/boot/dts/xlp4xx-helinux.dts b/arch/mips/boot/dts/xlp4xx-helinux.dts
index 2d85846..b3affd1 100644
--- a/arch/mips/boot/dts/xlp4xx-helinux.dts
+++ b/arch/mips/boot/dts/xlp4xx-helinux.dts
@@ -107,6 +107,7 @@
 				enable-nae-jumbo = <0>;
 				enable-nae-p2p = <0>;
 				enable-microstruct = <0>;
+				launch-mode = "smp";	/* smp or amp */
 			};
 			nae{
 			};
diff --git a/arch/mips/boot/dts/xlp8xx-helinux.dts b/arch/mips/boot/dts/xlp8xx-helinux.dts
index 1c3a62f..36370ed 100644
--- a/arch/mips/boot/dts/xlp8xx-helinux.dts
+++ b/arch/mips/boot/dts/xlp8xx-helinux.dts
@@ -122,6 +122,7 @@
 				enable-nae-p2p = <0>;
 				enable-poe = <0>;
 				enable-microstruct = <0>;
+				launch-mode = "smp";	/* smp or amp */
 			};
 			nae{
 			};
diff --git a/arch/mips/boot/dts/xlp9xx-helinux.dts b/arch/mips/boot/dts/xlp9xx-helinux.dts
index c47cd5a..d958de5 100644
--- a/arch/mips/boot/dts/xlp9xx-helinux.dts
+++ b/arch/mips/boot/dts/xlp9xx-helinux.dts
@@ -125,6 +125,7 @@
 				enable-nae-p2p = <0>;
 				enable-poe = <0>;
 				enable-microstruct = <0>;
+				launch-mode = "smp";	/* smp or amp */
 			};
 			nae{
 				freein-fifo-mask = <0xffff 0 0 0>;
-- 
1.7.1

