From b3f4d3ea9d6af8cd705c59aaebcf73853916750d Mon Sep 17 00:00:00 2001
From: Sunil Kulkarni <sunilku@broadcom.com>
Date: Tue, 1 Oct 2013 02:09:34 -0700
Subject: sysconfig: Change default SOC frequency for RSA and SAE to 767 MHz in default linux 9XX dts.

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/boot/dts/xlp9xx-linux.dts b/arch/mips/boot/dts/xlp9xx-linux.dts
index 3633a87..2adfdd2 100644
--- a/arch/mips/boot/dts/xlp9xx-linux.dts
+++ b/arch/mips/boot/dts/xlp9xx-linux.dts
@@ -228,8 +228,8 @@
 
         frequency-config {
                 nae = <500>;
-                sae = <250>;
-                rsa = <667>;
+                sae = <767>;
+                rsa = <767>;
                 dtre = <250>;
                 cde = <250>;
         };
-- 
1.7.1

