From 54cb836f751404257d9bfe7477fc1863eff9fada Mon Sep 17 00:00:00 2001
From: Jayanthi A <jayanthi.annadurai@broadcom.com>
Date: Mon, 14 Oct 2013 15:57:43 -0700
Subject: Fixed online mask for dom1

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/boot/dts/xlp9xx-linux.dts b/arch/mips/boot/dts/xlp9xx-linux.dts
index 2adfdd2..d2b2eff 100644
--- a/arch/mips/boot/dts/xlp9xx-linux.dts
+++ b/arch/mips/boot/dts/xlp9xx-linux.dts
@@ -148,7 +148,7 @@
                                 };
 			};
 			app-param{
-                                onlinemask = <0xffff>;
+                                onlinemask = <0 0 0xffff>;
 			};
 		};
                 dom@2 { 
-- 
1.7.1

