From 3aa6d9c1514e774c4c67ccf7ce4dc39a5e05211e Mon Sep 17 00:00:00 2001
From: Jayanthi A <jayanthi.annadurai@broadcom.com>
Date: Wed, 30 Oct 2013 15:37:37 -0700
Subject: xlp9xx: added fmn dts file for nae loopback application

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/boot/dts/xlp9xx-helinux.dts b/arch/mips/boot/dts/xlp9xx-helinux.dts
index 46b8581..d16280c 100644
--- a/arch/mips/boot/dts/xlp9xx-helinux.dts
+++ b/arch/mips/boot/dts/xlp9xx-helinux.dts
@@ -102,7 +102,7 @@
 			};
 
 			cpu {
-				onlinemask = <0xffffffff>;
+				onlinemask = <0xffff 0xffffffff 0xffffffff>;
                                 nae-rx-vc = <2>;
                                 nae-fb-vc = <3>;
 				dtre-rx-vc = <2>;
@@ -127,7 +127,7 @@
 				   onlinemask = <0x0 0x0 0xff    # Node 1 cpus 7-0
 						 0x0 0x0 0xffff> # node 0 cpus 15-0
 				 */
-				onlinemask = <0xffffffff>;
+				onlinemask = <0xffff 0xffffffff 0xffffffff>;
 				enable-nae-hw-parser = <0>;
 				enable-nae-jumbo = <0>;
 				enable-nae-p2p = <0>;
-- 
1.7.1

