From 8a8960d3bf7e480a4f505dadc5a4af1053b7adf2 Mon Sep 17 00:00:00 2001
From: Kamlakant Patel <kamlakant.patel@broadcom.com>
Date: Sat, 9 Nov 2013 12:47:50 +0530
Subject: spi: add interrupts property for spi-controller

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/boot/dts/xlp2xx-linux-tmpl.dts b/arch/mips/boot/dts/xlp2xx-linux-tmpl.dts
index 8bcd895..59129d5 100644
--- a/arch/mips/boot/dts/xlp2xx-linux-tmpl.dts
+++ b/arch/mips/boot/dts/xlp2xx-linux-tmpl.dts
@@ -153,6 +153,9 @@
 			#size-cells = <0>;
 			reg = <0 0x3a000 0x100>;
 
+			interrupts = <34>;
+			interrupt-parent = <&pic>;
+
 			spi_nor@1 {
 				compatible = "spansion,s25sl12801";
 				#address-cells = <1>;
diff --git a/arch/mips/boot/dts/xlp3xx-linux-tmpl.dts b/arch/mips/boot/dts/xlp3xx-linux-tmpl.dts
index 76a72ed..d5759a3 100644
--- a/arch/mips/boot/dts/xlp3xx-linux-tmpl.dts
+++ b/arch/mips/boot/dts/xlp3xx-linux-tmpl.dts
@@ -154,6 +154,9 @@
 			#size-cells = <0>;
 			reg = <0 0x3a000 0x100>;
 
+			interrupts = <34>;
+			interrupt-parent = <&pic>;
+
 			spi_nor@1 {
 				compatible = "spansion,s25sl12801";
 				#address-cells = <1>;
diff --git a/arch/mips/boot/dts/xlp8xx-linux-tmpl.dts b/arch/mips/boot/dts/xlp8xx-linux-tmpl.dts
index d52f2ff..5732b10 100644
--- a/arch/mips/boot/dts/xlp8xx-linux-tmpl.dts
+++ b/arch/mips/boot/dts/xlp8xx-linux-tmpl.dts
@@ -166,6 +166,9 @@
 			#size-cells = <0>;
 			reg = <0 0x3a000 0x100>;
 
+			interrupts = <34>;
+			interrupt-parent = <&pic>;
+
 			spi_mmc@0 {
 				compatible = "mmc-spi-slot";
 				reg = <0>;      /* Chip Select */
diff --git a/arch/mips/boot/dts/xlp9xx-linux-tmpl.dts b/arch/mips/boot/dts/xlp9xx-linux-tmpl.dts
index 60bcd01..ead11ab 100644
--- a/arch/mips/boot/dts/xlp9xx-linux-tmpl.dts
+++ b/arch/mips/boot/dts/xlp9xx-linux-tmpl.dts
@@ -85,8 +85,11 @@
 			#size-cells = <0>;
 			reg = <0 0x13a000 0x100>;
 
+			interrupts = <34>;
+			interrupt-parent = <&pic>;
+
 			spi_nor@1 {
-				compatible = "spansion,s25sl12801";
+				compatible = "micron,n25q128";
 				#address-cells = <1>;
 				#size-cells = <1>;
 
-- 
1.7.1

