From 57cf492a6affe975463228ff6c881073304e1249 Mon Sep 17 00:00:00 2001
From: Kamlakant Patel <kamlakant.patel@broadcom.com>
Date: Fri, 29 Nov 2013 20:00:10 +0530
Subject: mtd: add nand flash controller support for Griffin

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/boot/dts/xlp9xx-linux-tmpl.dts b/arch/mips/boot/dts/xlp9xx-linux-tmpl.dts
index ead11ab..f2cbb1c 100644
--- a/arch/mips/boot/dts/xlp9xx-linux-tmpl.dts
+++ b/arch/mips/boot/dts/xlp9xx-linux-tmpl.dts
@@ -137,6 +137,31 @@
 				};
 			};
 		};
+
+		nand_flash@0,0 {
+			compatible = "brcm,xlp9xx-nand";
+			#address-cells = <1>;
+			#size-cells = <1>;
+			reg = <0 0x139000 0x1000>;
+
+			nand-ecc-mode = "hw";
+
+			partition@0 {
+				label = "firmware";
+				reg = <0x0 0x20000000>;
+			};
+
+			partition@1 {
+				label = "kernel";
+				reg = <0x20000000 0x20000000>;
+			};
+
+			partition@2 {
+				label = "user";
+				reg = <0x40000000 0x0>;
+			};
+		};
+
 		nor_flash@1,0 {
 			compatible = "cfi-flash";
 			#address-cells = <1>;
-- 
1.7.1

