From 8f31bdbdf4bb27966d53fc700a1eb9280e204946 Mon Sep 17 00:00:00 2001
From: Siva Krishna Gudivada <sivag@broadcom.com>
Date: Tue, 3 Dec 2013 16:04:19 +0530
Subject: DPI: Changed online mask to enable 64 cpus

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/boot/dts/xlp9xx-linux-dpi.dts b/arch/mips/boot/dts/xlp9xx-linux-dpi.dts
index 8b72989..4362dbc 100644
--- a/arch/mips/boot/dts/xlp9xx-linux-dpi.dts
+++ b/arch/mips/boot/dts/xlp9xx-linux-dpi.dts
@@ -197,7 +197,7 @@
 		 * 20core: netlogic,cpumask=<0xffff 0xffffffff 0xffffffff>
 		 * 15 core, skipping core #5: netlogic,cpumask=<0xffffffff 0xff0fffff>
 		 */
-		netlogic,cpumask = <0x01>;
+		netlogic,cpumask = <0xffffffff 0xffffffff>;
 		bootargs = " console=ttyS0,115200  rhash_entries=1048576";
 		/*linux,initrd-start=<0x00c00000>;
 		linux,initrd-end=<0x02c00000>;*/
-- 
1.7.1

