From e5ab023f6f0791634094581c7a2153148e56e0e8 Mon Sep 17 00:00:00 2001
From: Jayanthi A <jayanthi.annadurai@broadcom.com>
Date: Tue, 12 Nov 2013 14:47:34 -0800
Subject: xlp9xx:Selecting DV based on the context number. str-fwd is enabled by default for nae loopback application

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/boot/dts/xlp9xx-helinux.dts b/arch/mips/boot/dts/xlp9xx-helinux.dts
index 7a6a187..ce3efad 100644
--- a/arch/mips/boot/dts/xlp9xx-helinux.dts
+++ b/arch/mips/boot/dts/xlp9xx-helinux.dts
@@ -100,7 +100,7 @@
 			};
 
 			cpu {
-				onlinemask =<0xffff 0xffffffff 0xffffffff>; /* <0x1111 0x11111111 0x11111111>; */
+				onlinemask = <0xffff 0xffffffff 0xffffffff>; /* <0x1111 0x11111111 0x11111111>; */
                                 nae-rx-vc = <2>;
                                 nae-fb-vc = <3>;
 				dtre-rx-vc = <2>;
-- 
1.7.1

