From 1ff079036b369c0fea9b21db61f9dac0abadb45a Mon Sep 17 00:00:00 2001
From: Pavani Reddy <pavani@broadcom.com>
Date: Mon, 18 Nov 2013 14:12:43 +0530
Subject: xlp8xx-hybrid-nae-helinux.dts : add isolcpus

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/boot/dts/xlp8xx-hybrid-nae-helinux.dts b/arch/mips/boot/dts/xlp8xx-hybrid-nae-helinux.dts
index 103bb8e..635073a 100644
--- a/arch/mips/boot/dts/xlp8xx-hybrid-nae-helinux.dts
+++ b/arch/mips/boot/dts/xlp8xx-hybrid-nae-helinux.dts
@@ -203,7 +203,7 @@
 
 	chosen {
         netlogic,cpumask = <0xffffffff>;
-		bootargs = " console=ttyS0,115200 ";
+		bootargs = " console=ttyS0,115200 isolcpus=1-31";
 	};
 
 	/* These binaries are downloaded at the resp physical memory locations
-- 
1.7.1

