From 3b6ba48592e3287bb0f3f2974967f85846b89e0f Mon Sep 17 00:00:00 2001
From: Rahul Gupta <guptar@broadcom.com>
Date: Thu, 21 Nov 2013 11:47:53 +0530
Subject: 9xx brcm_nat_app: Fixed bug #1355 & #1234. Changed build-dtb.sh- common/soc-fmn-ver1.dts,used "onlinemask" new convention. "freein-fifo-replenish-desc-info" Credits increased to 64.

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/boot/dts/xlp9xx-helinux-nat.dts b/arch/mips/boot/dts/xlp9xx-helinux-nat.dts
index 0b47818..a8e8995 100644
--- a/arch/mips/boot/dts/xlp9xx-helinux-nat.dts
+++ b/arch/mips/boot/dts/xlp9xx-helinux-nat.dts
@@ -42,7 +42,7 @@ pktmem {
 			#size-cells = <1>;
 
 			cpu {
-				onlinemask = <0xffffffff>;
+				onlinemask = <0x0 0x0 0xffffffff>;
                 nae-rx-vc = <0>;
                 nae-fb-vc = <1>;
 			//	sae-rx-vc = <0>; /* sharing with nae */
@@ -115,7 +115,7 @@ pktmem {
 			};
 
 			cpu {
-				onlinemask = <0xffffffff>;
+				onlinemask = <0x0 0x0 0xffffffff>;
                                 nae-rx-vc = <2>;
                                 nae-fb-vc = <3>;
 				dtre-rx-vc = <2>;
@@ -132,7 +132,7 @@ pktmem {
 				id = <0>;
 			};
 			app-param{
-				onlinemask = <0xffffffff>;
+				onlinemask = <0x0 0x0 0xffffffff>;
 				enable-nae-hw-parser = <0>;
 				enable-nae-jumbo = <0>;
 				enable-nae-p2p = <0>;
@@ -153,7 +153,7 @@ pktmem {
 				/* < node paddr-H32bits paddr-L32bits size-H32bits size-L32bits */
 				freein-fifo-replenish-addr-info = <0x0 0x0 0x88000000 0x0 0x10000000>; /* 256M @2176M*/
 				/* < node fifomask descsize headroom ndescs > */
-				freein-fifo-replenish-desc-info = <0 0x00ff 2048 0 32>;
+				freein-fifo-replenish-desc-info = <0 0x00ff 2048 0 64>;
 			};
 			nae-1 {
 				sh-domains = <0>;
@@ -170,7 +170,7 @@ pktmem {
 				/* < node paddr-H32bits paddr-L32bits size-H32bits size-L32bits */
 				freein-fifo-replenish-addr-info = <0x0 0x0 0x98000000 0x0 0x10000000>; /* 256M @2432M*/
 				/* < node fifomask descsize headroom ndescs > */
-				freein-fifo-replenish-desc-info = <0 0x00ff 2048 0 32>;
+				freein-fifo-replenish-desc-info = <0 0x00ff 2048 0 64>;
 			};
 
 			owner-config {
-- 
1.7.1

