From 227f905f76f7f5146a4ac90545b8792f4bb3afd9 Mon Sep 17 00:00:00 2001
From: Pavani Reddy <pavani@broadcom.com>
Date: Thu, 21 Nov 2013 16:02:14 +0530
Subject: hybrid-nae: made soc/nodes to 1, added isolcpus ; build-dtb.sh : replaced nlm-fmn.dts with soc-fmn-ver1.dts

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/boot/dts/xlp9xx-hybrid-nae-helinux.dts b/arch/mips/boot/dts/xlp9xx-hybrid-nae-helinux.dts
index e22230d..6ab5ea7 100644
--- a/arch/mips/boot/dts/xlp9xx-hybrid-nae-helinux.dts
+++ b/arch/mips/boot/dts/xlp9xx-hybrid-nae-helinux.dts
@@ -231,7 +231,7 @@
 
 	chosen {
 		netlogic,cpumask = <0xffffffff>;
-		bootargs = " console=ttyS0,115200 ";
+		bootargs = " console=ttyS0,115200 isolcpus=1-31 " ;
 	};
 
 	/* These binaries are downloaded at the resp physical memory locations
@@ -252,7 +252,7 @@
 
 	soc {
 		nodes {
-			num-nodes = <2>;
+			num-nodes = <1>;
 		};
 	};
 	frequency-config {
-- 
1.7.1

