From 85edc7b9f6225803e58fb81135589d2f4d84ce99 Mon Sep 17 00:00:00 2001
From: Deepak Jangid <deepak.jangid@broadcom.com>
Date: Wed, 13 Nov 2013 02:19:19 -0800
Subject: snort: 8xx-snort dts edited onlinemask to 32 CPUs

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/boot/dts/xlp8xx-helinux-snort.dts b/arch/mips/boot/dts/xlp8xx-helinux-snort.dts
index cb76bf2..36900f2 100644
--- a/arch/mips/boot/dts/xlp8xx-helinux-snort.dts
+++ b/arch/mips/boot/dts/xlp8xx-helinux-snort.dts
@@ -43,7 +43,7 @@
 			#size-cells = <1>;
 
 			cpu {
-				onlinemask = <0x1>;
+				onlinemask = <0xffffffff>;
                                 nae-rx-vc = <1>;
                                 nae-fb-vc = <1>;
 				sae-rx-vc = <1>; /* sharing with nae */
@@ -93,7 +93,7 @@
 			};
 
 			cpu {
-				onlinemask = <0x1>;
+				onlinemask = <0xffffffff>;
 				nae-rx-vc = <0>;
 				nae-fb-vc = <3>;
 				sae-rx-vc = <0>; /* sharing with nae */
@@ -111,7 +111,7 @@
 				reg = <0x60000000 0x60000000>;  // 1.5GB @ 1.5GB
 			};
 			app-param {
-				onlinemask = <0x1>;
+				onlinemask = <0xffffffff>;
 			};
 			nae {
 			};
-- 
1.7.1

