From b75d516e357470bdd340fc47a0fbe106d07f9c5a Mon Sep 17 00:00:00 2001
From: Pavani Reddy <pavani@broadcom.com>
Date: Mon, 2 Dec 2013 14:36:33 +0530
Subject: xlp2xx-pcie-helinux-pcieperf.dts : add netlogic cpumask parameter

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/boot/dts/xlp2xx-pcie-helinux-pcieperf.dts b/arch/mips/boot/dts/xlp2xx-pcie-helinux-pcieperf.dts
index b2d6301..99d49ee 100644
--- a/arch/mips/boot/dts/xlp2xx-pcie-helinux-pcieperf.dts
+++ b/arch/mips/boot/dts/xlp2xx-pcie-helinux-pcieperf.dts
@@ -114,6 +114,7 @@
 	};
 
 	chosen {
+        netlogic,cpumask = <0xff>;
 		bootargs = " xlp_noi2c console=ttyS0,115200 ";
 		/*"autoboot_pciedrv";*/
 		/*bootargs = " xlp_noi2c console=ttyS0,115200 console=pcie_co,115200 autoboot_pciedrv console_over_pcie"; */
-- 
1.7.1

