From 563df538a14af57bee7c069dd59187b0bdbf3278 Mon Sep 17 00:00:00 2001
From: Anurag <anurag.gopinath@broadcom.com>
Date: Tue, 3 Dec 2013 23:32:49 +0530
Subject: XLP9xx : Correct the base address of UART1.

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/boot/dts/xlp9xx-linux-tmpl.dts b/arch/mips/boot/dts/xlp9xx-linux-tmpl.dts
index 30a3fc1..ee15f92 100644
--- a/arch/mips/boot/dts/xlp9xx-linux-tmpl.dts
+++ b/arch/mips/boot/dts/xlp9xx-linux-tmpl.dts
@@ -25,7 +25,7 @@
 		serial0: serial@30000 {
 			device_type = "serial";
 			compatible = "ns16550";
-			reg = <0 0x112100 0xa00>;
+			reg = <0 0x112100 0x20>;
 			reg-shift = <2>;
 			reg-io-width = <4>;
 			clock-frequency = <125000000>;
@@ -35,7 +35,7 @@
 		serial1: serial@31000 {
 			device_type = "serial";
 			compatible = "ns16550";
-			reg = <0 0x113100 0xa00>;
+			reg = <0 0x112120 0x20>;
 			reg-shift = <2>;
 			reg-io-width = <4>;
 			clock-frequency = <125000000>;
-- 
1.7.1

