From 1820d7a9b7c52796cd94e4968cce8f98ea286580 Mon Sep 17 00:00:00 2001
From: ajesh <ajesh@broadcom.com>
Date: Thu, 12 Dec 2013 15:41:37 +0530
Subject: MNXLP8xx: Added cpumask for multinode

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/boot/dts/mnxlp8xx-linux.dts b/arch/mips/boot/dts/mnxlp8xx-linux.dts
index 1ca8c79..afaac3e 100644
--- a/arch/mips/boot/dts/mnxlp8xx-linux.dts
+++ b/arch/mips/boot/dts/mnxlp8xx-linux.dts
@@ -92,6 +92,7 @@
 		bootargs = " console=ttyS0,115200 ";
 		/*linux,initrd-start=<0x00c00000>;
 		linux,initrd-end=<0x02c00000>;*/
+		netlogic,cpumask = <0xffffffff 0xffffffff>;
 	};
 
 	/* These binaries are downloaded at the resp physical memory locations
-- 
1.7.1

