From e6776342726595511adf8dd838741580a2cc9eb0 Mon Sep 17 00:00:00 2001
From: ajesh <ajesh@broadcom.com>
Date: Fri, 13 Dec 2013 16:56:37 +0530
Subject: mnxlp9xx: Changes for bringing up nae driver.

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/boot/dts/mnxlp9xx-linux.dts b/arch/mips/boot/dts/mnxlp9xx-linux.dts
index a8991a1..e26c3f8 100644
--- a/arch/mips/boot/dts/mnxlp9xx-linux.dts
+++ b/arch/mips/boot/dts/mnxlp9xx-linux.dts
@@ -19,6 +19,14 @@
 	 *               0 0xe0000000 0 0x20000000>;
 	 * };
 	 */
+		memory {
+                device_type = "memory";
+                reg =  <0 0x00000000 0 0x0B000000
+                        0 0x20000000 0 0xa0000000
+                        0 0xe0000000 0 0x20000000
+                        0x2 0x80000000 0 0x80000000>;
+        };
+
 
         /* memory range for HEv2 pkt mem. Only one contig memory per node supported */
         pktmem {
@@ -46,11 +54,11 @@
 
 			#address-cells = <1>;
 			#size-cells = <1>;
-			#nodes = <1>;
-			#cpumask-cells=<3>;
+			#nodes = <2>;
+			#cpumask-cells=<4>;
 
 			cpu {
-				onlinemask = <0x1>; /* */
+			onlinemask = <0x0 0x0 0xffffffff 0xffffffff>; /* */
                                 nae-rx-vc = <0>;
                                 nae-fb-vc = <1>;
 				napi-vc-mask = <0xb>;
@@ -77,6 +85,33 @@
                         };
 			pic {
 			};
+			ici-config{
+                enable = <1>;
+                total_credits = <673>;
+                node_0_link_mask = <0x2>;   //Bit mask of active links
+                node_1_link_mask = <0x2>;   //Bit mask of active links
+                node_2_link_mask = <0x0>;   //Bit mask of active links
+                node_3_link_mask = <0x0>;   //Bit mask of active links
+                gcu{
+                    own_credit = <75 12 75 75 75 75 75 12>; /*VC1 and VC7 always shows 0 transactions.*/
+                    shared_credit = <1 1 1 1 1 1 1 1>;
+                    txwght = <63 12 63 63 63 63 63 12>; /*Scheduling Priority, Higher Value = More Time Slot*/
+                    segth  = <34 34 34 34 34 34 34 34>; /*Maximum number of tx entries*/
+                };
+                fmn{
+                    own_credit = <75 12 12 75>; /*VC1 and VC2 always shows 0 transactions.*/
+                    shared_credit = <1 1 1 1>;
+                    txwght = <8 2 2 8 >;    /*Scheduling Priority, Higher Value = More Time Slot*/
+                    segth  = <34 34 34 34>; /*Maximum number of tx entries*/
+                };
+                pic{
+                    own_credit = <12>;
+                    shared_credit = <1>;
+                    txwght = <1>;   /*Scheduling Priority, Higher Value = More Time Slot*/
+                    segth  = <34>;  /*Maximum number of tx entries*/
+                };
+            };
+
 			nae-0 {
 				freein-fifo-mask = <0xffff 0 0 0>;
 				/* node offsets and size */
@@ -98,10 +133,10 @@
 				dummy-pktdata-addr = <0x0 0x0ff00000>; // 255 MB
 			};
 			owner-config {
-				nae = <0x1>;   // currently for node 0
+				nae = <0x3>;   // currently for node 0
 				sae = <0x1>;
 				rsa = <0x1>;
-				fmn = <0x1>;
+				fmn = <0x3>;
 			};
 			resource-allocation {
                                 net@0 {
@@ -204,6 +239,8 @@
 		 *			     0 0xffff 0xffffffff 0xffffffff>;
 		 * 15 core(core 5 disabled): netlogic,cpumask=<0xffffffff 0xff0fffff>;
 		 */
+		netlogic,cpumask = <0x0 0x0 0xffffffff 0xffffffff
+		                    0x0 0x0 0xffffffff 0xffffffff>;					
 		bootargs = " console=ttyS0,115200  rhash_entries=1048576";
 		/*linux,initrd-start=<0x00c00000>;
 		linux,initrd-end=<0x02c00000>;*/
@@ -227,7 +264,7 @@
 
 	soc {
 		nodes {
-			num-nodes = <1>;
+			num-nodes = <2>;
 		};
 
 	};
-- 
1.7.1

