From a1cccc5194f87082177c87f88ae7efc552d05664 Mon Sep 17 00:00:00 2001
From: Pavani Reddy <pavani@broadcom.com>
Date: Tue, 17 Dec 2013 16:07:56 +0530
Subject: mnxlp9xx : added naeperf dts files for mnxlp9xx

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/boot/dts/mnxlp9xx-helinux-nae.dts b/arch/mips/boot/dts/mnxlp9xx-helinux-nae.dts
index 0e59d32..dbd5638 100644
--- a/arch/mips/boot/dts/mnxlp9xx-helinux-nae.dts
+++ b/arch/mips/boot/dts/mnxlp9xx-helinux-nae.dts
@@ -145,6 +145,7 @@
 					      0xffff 0xffffffff 0xffffffff>;
 				enable_ici_pkt_traffic = <0>;
 				
+/*
 				//Enable this for mode #1 and #2 Local Node Memory - For 16G+16G
 				nae_0_pkt_pool_start = <0x140>;	//Multiple of 16MB
 				nae_0_pkt_pool_size = <0x1000000>;
@@ -152,7 +153,6 @@
 				nae_1_pkt_pool_size = <0x1000000>;
 
 				
-/*
 				//Enable this for mode #1 and #2 Local Node Memory - For 16G+16G
 				nae_0_pkt_pool_start = <0x140>;	//Multiple of 16MB
 				nae_0_pkt_pool_size = <0x1000000>;
@@ -160,13 +160,13 @@
 				nae_1_pkt_pool_size = <0x1000000>;
 */
 
-/*
 				//Enable this for mode #1 and #2 Local Node Memory - For 8G+8G
 				nae_0_pkt_pool_start = <0x140>;	//Multiple of 16MB
 				nae_0_pkt_pool_size = <0x1000000>;
 				nae_1_pkt_pool_start = <0x241>;	//Multiple of 16MB
 				nae_1_pkt_pool_size = <0x1000000>;
 
+/*
 				//Enable this for mode #3 Remote Node Memory  - For 8G+8G
 				nae_0_pkt_pool_start = <0x241>;	//Multiple of 16MB
 				nae_0_pkt_pool_size = <0x1000000>;
-- 
1.7.1

