From 6f53a34c95a6da003b19106fb6086f9ebcf9748d Mon Sep 17 00:00:00 2001
From: Saswat Dash <saswat.dash@broadcom.com>
Date: Thu, 19 Dec 2013 19:35:54 +0530
Subject: nae: Added NAE parameter nae-rx-cpu-mask

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/boot/dts/xlp2xx-linux.dts b/arch/mips/boot/dts/xlp2xx-linux.dts
index dd82aaa..d661b7f 100644
--- a/arch/mips/boot/dts/xlp2xx-linux.dts
+++ b/arch/mips/boot/dts/xlp2xx-linux.dts
@@ -54,6 +54,7 @@
                                 node_0_vc_mask = <0xbbbbbbbb 0xbbbbbbbb 0xbbbbbbbb 0xbbbbbbbb>;
                         };
 			nae {
+				nae-rx-cpu-mask = <>;
 				freein-fifo-mask = <0xff 0 0 0>;
 				/* node offsets and size */
 				vfbtbl-sw-offset = <0 0 0 0>;
diff --git a/arch/mips/boot/dts/xlp3xx-linux.dts b/arch/mips/boot/dts/xlp3xx-linux.dts
index 3eaf700..3b0cd7d 100644
--- a/arch/mips/boot/dts/xlp3xx-linux.dts
+++ b/arch/mips/boot/dts/xlp3xx-linux.dts
@@ -64,6 +64,12 @@
 
 			};
 			nae {
+				/* Example for different values of nae-rx-cpu-mask
+				 * for 16 CPUs: nae-rx-cpu-mask = <0xffff>;
+				 * for 8 CPUs: nae-rx-cpu-mask = <0xff>;
+				 * if left empty, it takes the value of onlinemask.
+				 */
+				nae-rx-cpu-mask = <>;
 				freein-fifo-mask = <0xff 0 0 0>;
 				/* node offsets and size */
 				vfbtbl-sw-offset = <0 0 0 0>;
diff --git a/arch/mips/boot/dts/xlp8xx-linux.dts b/arch/mips/boot/dts/xlp8xx-linux.dts
index 713197f..a036670 100644
--- a/arch/mips/boot/dts/xlp8xx-linux.dts
+++ b/arch/mips/boot/dts/xlp8xx-linux.dts
@@ -66,6 +66,11 @@
                                 node_3_vc_mask = <0xbbbbbbbb 0xbbbbbbbb 0xbbbbbbbb 0xbbbbbbbb>;
                         };
 			nae {
+				/* Example for different values of nae-rx-cpu-mask
+				 * for 32 CPUs: nae-rx-cpu-mask = <0xffffffff>;
+				 * if left empty, it takes the value of onlinemask.
+				 */
+				nae-rx-cpu-mask = <>;
 				freein-fifo-mask = <0x3ffff 0 0 0>;
 				/* node offsets and size */
 				vfbtbl-sw-offset = <0 0 0 0>;
diff --git a/arch/mips/boot/dts/xlp9xx-linux.dts b/arch/mips/boot/dts/xlp9xx-linux.dts
index b29e113..c1a6507 100644
--- a/arch/mips/boot/dts/xlp9xx-linux.dts
+++ b/arch/mips/boot/dts/xlp9xx-linux.dts
@@ -73,6 +73,12 @@
 			pic {
 			};
 			nae-0 {
+				/* Example for different values of nae-rx-cpu-mask
+				 * for 80 CPUs: nae-rx-cpu-mask = <0xffff 0xffffffff 0xffffffff>;
+				 * for 64 CPUs: nae-rx-cpu-mask = <0xffffffff 0xffffffff>;
+				 * for 32 CPUs: nae-rx-cpu-mask = <0xffffffff>;
+				 */
+				nae-rx-cpu-mask = <>;
 				freein-fifo-mask = <0xffff 0 0 0>;
 				/* node offsets and size */
 				vfbtbl-sw-offset = <0 0 0 0>;
@@ -83,6 +89,7 @@
 				dummy-pktdata-addr = <0x0 0x0ff00000>; // 255 MB
 			};
 			nae-1 {
+				nae-rx-cpu-mask = <>;
 				freein-fifo-mask = <0xffff 0 0 0>;
 				/* node offsets and size */
 				vfbtbl-sw-offset = <0 0 0 0>;
-- 
1.7.1

