From dfe8cea819946ef97bbf0ae17c1a5c76f4645688 Mon Sep 17 00:00:00 2001
From: Subhendu Sekhar Behera <sbehera@broadcom.com>
Date: Mon, 27 Jan 2014 18:14:29 +0530
Subject: defconfig & dts: add support for watchdog

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/boot/dts/xlp2xx-ep-linux-tmpl.dts b/arch/mips/boot/dts/xlp2xx-ep-linux-tmpl.dts
index 321bd7b..75f4adf 100644
--- a/arch/mips/boot/dts/xlp2xx-ep-linux-tmpl.dts
+++ b/arch/mips/boot/dts/xlp2xx-ep-linux-tmpl.dts
@@ -86,6 +86,10 @@
 			reg = <0 0x4000 0x200>;
 		};
 
+		watchdog@0 {
+			compatible = "netlogic,nlm-wdt";
+		};
+
 		nor_flash@1,0 {
 			compatible = "cfi-flash";
 			#address-cells = <1>;
diff --git a/arch/mips/boot/dts/xlp2xx-linux-tmpl.dts b/arch/mips/boot/dts/xlp2xx-linux-tmpl.dts
index 59129d5..9e97361 100644
--- a/arch/mips/boot/dts/xlp2xx-linux-tmpl.dts
+++ b/arch/mips/boot/dts/xlp2xx-linux-tmpl.dts
@@ -86,6 +86,10 @@
 			reg = <0 0x4000 0x200>;
 		};
 
+		watchdog@0 {
+			compatible = "netlogic,nlm-wdt";
+		};
+
 		nor_flash@1,0 {
 			compatible = "cfi-flash";
 			#address-cells = <1>;
diff --git a/arch/mips/boot/dts/xlp2xx-pcie-linux-tmpl.dts b/arch/mips/boot/dts/xlp2xx-pcie-linux-tmpl.dts
index 53c0eb2..10b54ef 100644
--- a/arch/mips/boot/dts/xlp2xx-pcie-linux-tmpl.dts
+++ b/arch/mips/boot/dts/xlp2xx-pcie-linux-tmpl.dts
@@ -86,6 +86,10 @@
 			reg = <0 0x4000 0x200>;
 		};
 
+		watchdog@0 {
+			compatible = "netlogic,nlm-wdt";
+		};
+
 		nand_flash@0,0 {
 			compatible = "netlogic,xlp-nand";
 			#address-cells = <1>;
diff --git a/arch/mips/boot/dts/xlp2xx-xmc-linux-tmpl.dts b/arch/mips/boot/dts/xlp2xx-xmc-linux-tmpl.dts
index 09feb9c..6c25d68 100644
--- a/arch/mips/boot/dts/xlp2xx-xmc-linux-tmpl.dts
+++ b/arch/mips/boot/dts/xlp2xx-xmc-linux-tmpl.dts
@@ -86,6 +86,10 @@
 			reg = <0 0x4000 0x200>;
 		};
 
+		watchdog@0 {
+			compatible = "netlogic,nlm-wdt";
+		};
+
 		nand_flash@0,0 {
 			compatible = "netlogic,xlp-nand";
 			#address-cells = <1>;
diff --git a/arch/mips/boot/dts/xlp3xx-linux-tmpl.dts b/arch/mips/boot/dts/xlp3xx-linux-tmpl.dts
index 17a2c1a..d6b6fa0 100644
--- a/arch/mips/boot/dts/xlp3xx-linux-tmpl.dts
+++ b/arch/mips/boot/dts/xlp3xx-linux-tmpl.dts
@@ -87,6 +87,10 @@
 			reg = <0 0x4000 0x200>;
 		};
 
+		watchdog@0 {
+			compatible = "netlogic,nlm-wdt";
+		};
+
 		nor_flash@1,0 {
 			compatible = "cfi-flash";
 			#address-cells = <1>;
diff --git a/arch/mips/boot/dts/xlp3xx-xmc-linux-tmpl.dts b/arch/mips/boot/dts/xlp3xx-xmc-linux-tmpl.dts
index 2128087..6efe2bd 100644
--- a/arch/mips/boot/dts/xlp3xx-xmc-linux-tmpl.dts
+++ b/arch/mips/boot/dts/xlp3xx-xmc-linux-tmpl.dts
@@ -88,6 +88,10 @@
 			reg = <0 0x4000 0x200>;
 		};
 
+		watchdog@0 {
+			compatible = "netlogic,nlm-wdt";
+		};
+
 		nand_flash@0,0 {
 			compatible = "netlogic,xlp-nand";
 			#address-cells = <1>;
diff --git a/arch/mips/boot/dts/xlp8xx-linux-mn-tmpl.dts b/arch/mips/boot/dts/xlp8xx-linux-mn-tmpl.dts
index 9151099..9e233cf 100644
--- a/arch/mips/boot/dts/xlp8xx-linux-mn-tmpl.dts
+++ b/arch/mips/boot/dts/xlp8xx-linux-mn-tmpl.dts
@@ -24,6 +24,11 @@
 			reg = <0 0x4000 0x200>;
 			interrupt-controller;
 		};
+
+		watchdog@0 {
+			compatible = "netlogic,nlm-wdt";
+		};
+
 		serial00: serial@30000 {
 			device_type = "serial";
 			compatible = "ns16550";
@@ -107,6 +112,7 @@
 			reg = <0 0x44000 0x200>;
 			interrupt-controller;
 		};
+
 		serial10: serial@70000 {
 			device_type = "serial";
 			compatible = "ns16550";
diff --git a/arch/mips/boot/dts/xlp8xx-linux-mn4-tmpl.dts b/arch/mips/boot/dts/xlp8xx-linux-mn4-tmpl.dts
index 3e70fe9..031a2b5 100644
--- a/arch/mips/boot/dts/xlp8xx-linux-mn4-tmpl.dts
+++ b/arch/mips/boot/dts/xlp8xx-linux-mn4-tmpl.dts
@@ -24,6 +24,11 @@
 			reg = <0 0x4000 0x200>;
 			interrupt-controller;
 		};
+
+		watchdog@0 {
+			compatible = "netlogic,nlm-wdt";
+		};
+
 		serial00: serial@30000 {
 			device_type = "serial";
 			compatible = "ns16550";
@@ -107,6 +112,7 @@
 			reg = <0 0x44000 0x200>;
 			interrupt-controller;
 		};
+
 		serial10: serial@70000 {
 			device_type = "serial";
 			compatible = "ns16550";
@@ -145,6 +151,7 @@
 			#interrupt-cells = <1>;
 			reg = <0 0x84000 0x200>;
 		};
+
 		pic3: pic@c4000 {
 			compatible = "netlogic,xlp-pic";
 			interrupt-controller;
diff --git a/arch/mips/boot/dts/xlp8xx-linux-tmpl.dts b/arch/mips/boot/dts/xlp8xx-linux-tmpl.dts
index 79e7992..59f121a 100644
--- a/arch/mips/boot/dts/xlp8xx-linux-tmpl.dts
+++ b/arch/mips/boot/dts/xlp8xx-linux-tmpl.dts
@@ -87,6 +87,10 @@
 			reg = <0 0x4000 0x200>;
 		};
 
+		watchdog@0 {
+			compatible = "netlogic,nlm-wdt";
+		};
+
 		nor_flash@1,0 {
 			compatible = "cfi-flash";
 			#address-cells = <1>;
diff --git a/arch/mips/boot/dts/xlp9xx-linux-mn-tmpl.dts b/arch/mips/boot/dts/xlp9xx-linux-mn-tmpl.dts
index 501463f..8c239f2 100644
--- a/arch/mips/boot/dts/xlp9xx-linux-mn-tmpl.dts
+++ b/arch/mips/boot/dts/xlp9xx-linux-mn-tmpl.dts
@@ -33,6 +33,10 @@
 			reg = <0 0x110000 0x200>;
 		};
 
+		watchdog@0 {
+			compatible = "netlogic,nlm-wdt";
+		};
+
 		i2c0: i2c@113100 {
 			compatible = "netlogic,xlp9xx-i2c";
 			#address-cells = <1>;
diff --git a/arch/mips/boot/dts/xlp9xx-linux-tmpl.dts b/arch/mips/boot/dts/xlp9xx-linux-tmpl.dts
index 94309d4..642fb64 100644
--- a/arch/mips/boot/dts/xlp9xx-linux-tmpl.dts
+++ b/arch/mips/boot/dts/xlp9xx-linux-tmpl.dts
@@ -49,6 +49,10 @@
 			reg = <0 0x110000 0x200>;
 		};
 
+		watchdog@0 {
+			compatible = "netlogic,nlm-wdt";
+		};
+
 		i2c0: i2c@113100 {
 			compatible = "netlogic,xlp9xx-i2c";
 			#address-cells = <1>;
-- 
1.7.1

