From 1c321bffe09ff3272008c697f88d6334fd23236f Mon Sep 17 00:00:00 2001
From: Virendra Pathak <vpathak@broadcom.com>
Date: Thu, 30 Jan 2014 09:25:26 +0530
Subject: HEv2: added timer service vc in app-param

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/boot/dts/xlp8xx-helinux.dts b/arch/mips/boot/dts/xlp8xx-helinux.dts
index f2f3805..ff5f64a 100644
--- a/arch/mips/boot/dts/xlp8xx-helinux.dts
+++ b/arch/mips/boot/dts/xlp8xx-helinux.dts
@@ -122,6 +122,7 @@
 				enable-microstruct = <0>;
 				enable-popq-mode = <0>; /* If set to 1, popq mode is enabled else disabled */
 				launch-mode = "smp";	/* smp or amp */
+				timer-service-vc = <2>;
 			};
 			nae{
 			};
diff --git a/arch/mips/boot/dts/xlp9xx-helinux.dts b/arch/mips/boot/dts/xlp9xx-helinux.dts
index 6a54de4..8082d72 100644
--- a/arch/mips/boot/dts/xlp9xx-helinux.dts
+++ b/arch/mips/boot/dts/xlp9xx-helinux.dts
@@ -134,6 +134,7 @@
 				enable-popq-mode = <0>; /* If set to 1, popq mode is enabled else disabled */
 				enable-popq-priority = <0>; /* If set to 1, popq priority mode is enabled else disabled */
 				launch-mode = "smp";	/* smp or amp */
+				timer-service-vc = <2>;
 			};
 			nae{
 				freein-fifo-mask = <0xffff 0 0 0>;
-- 
1.7.1

