From 57a06d4bd9a70f0444f96139cf7525fcf5ed16a9 Mon Sep 17 00:00:00 2001
From: Virendra Pathak <vpathak@broadcom.com>
Date: Thu, 13 Feb 2014 12:10:13 +0530
Subject: HEv2: updated app_param onlinemask per node to 4 word entry

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/boot/dts/mnxlp9xx-helinux-nae.dts b/arch/mips/boot/dts/mnxlp9xx-helinux-nae.dts
index dbd5638..1b67fc7 100644
--- a/arch/mips/boot/dts/mnxlp9xx-helinux-nae.dts
+++ b/arch/mips/boot/dts/mnxlp9xx-helinux-nae.dts
@@ -141,8 +141,16 @@
 				fmn = <0x0>;
 			};
 			app-param{
-				onlinemask = <0xffff 0xffffffff 0xffffffff
-					      0xffff 0xffffffff 0xffffffff>;
+				/* For multinode follow the below convention.
+				   1. provide node specific info on a separate line
+				   2. on each line provide the cpu mask as (MSB)79-64 , 63-32, 31-0 (LSB)
+
+				   example:
+				   onlinemask = <0x0 0x0 0x0 0xff    # Node 1 cpus 7-0
+						 0x0 0x0 0x0 0xffff> # node 0 cpus 15-0
+				 */
+				onlinemask = <0x0 0x0 0xffffffff 0xffffffff
+					      0x0 0x0 0xffffffff 0xffffffff>;
 				enable_ici_pkt_traffic = <0>;
 				
 /*
diff --git a/arch/mips/boot/dts/xlp9xx-helinux.dts b/arch/mips/boot/dts/xlp9xx-helinux.dts
index 8082d72..776fd2f 100644
--- a/arch/mips/boot/dts/xlp9xx-helinux.dts
+++ b/arch/mips/boot/dts/xlp9xx-helinux.dts
@@ -122,8 +122,8 @@
 				   2. on each line provide the cpu mask as (MSB)79-64 , 63-32, 31-0 (LSB)
 
 				   example:
-				   onlinemask = <0x0 0x0 0xff    # Node 1 cpus 7-0
-						 0x0 0x0 0xffff> # node 0 cpus 15-0
+				   onlinemask = <0x0 0x0 0x0 0xff    # Node 1 cpus 7-0
+						 0x0 0x0 0x0 0xffff> # node 0 cpus 15-0
 				 */
 				onlinemask = <0xffffffff>; /*<0x11111111>; */
 				enable-nae-hw-parser = <0>;
-- 
1.7.1

