From f59c5c23f8de247fdd2f78b70d73f97689a70fa5 Mon Sep 17 00:00:00 2001
From: Abhishek Joshi <abjoshi@broadcom.com>
Date: Mon, 24 Feb 2014 12:07:48 +0530
Subject: dts: added pktmem node for xlp4xx dts files

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/boot/dts/xlp4xx-helinux.dts b/arch/mips/boot/dts/xlp4xx-helinux.dts
index bb1c210..b81b4c9 100644
--- a/arch/mips/boot/dts/xlp4xx-helinux.dts
+++ b/arch/mips/boot/dts/xlp4xx-helinux.dts
@@ -15,6 +15,17 @@
 			0 0x20000000 0 0x20000000>;  // 512M at 512M
         };
 
+	/* memory range for HEv2 pkt mem. Only one contig memory per node supported */
+	pktmem {
+		device_type = "pktmem";
+
+		/* format: node number, 64-bit addr, 64-bit size */
+		reg = <0 0 0x80000000 0 0x08000000
+	               1 0x0 0x0 0x0 0x0
+		       2 0x0 0x0 0x0 0x0
+		       3 0x0 0x0 0x0 0x0>;
+	};
+
 	hypervisor {
 		hypervisor-name = "Xen";
 		alloc_dom0_memory = <0>;
diff --git a/arch/mips/boot/dts/xlp4xx-linux.dts b/arch/mips/boot/dts/xlp4xx-linux.dts
index b507f62..048d610 100644
--- a/arch/mips/boot/dts/xlp4xx-linux.dts
+++ b/arch/mips/boot/dts/xlp4xx-linux.dts
@@ -15,6 +15,17 @@
 			0 0x20000000 0 0x60000000>;  // 1536M at 512M
         };
 
+	/* memory range for HEv2 pkt mem. Only one contig memory per node supported */
+	pktmem {
+		device_type = "pktmem";
+
+		/* format: node number, 64-bit addr, 64-bit size */
+		reg = <0 0 0x80000000 0 0x08000000
+	               1 0x0 0x0 0x0 0x0
+		       2 0x0 0x0 0x0 0x0
+		       3 0x0 0x0 0x0 0x0>;
+	};
+
 	hypervisor {
 		hypervisor-name = "Xen";
 		alloc_dom0_memory = <0>;
-- 
1.7.1

