From 68141a4f8b143a97142d22ee3ad653658538c73c Mon Sep 17 00:00:00 2001
From: Siva Krishna Gudivada <sivag@broadcom.com>
Date: Wed, 26 Mar 2014 12:34:02 +0530
Subject: DPI: increased linux memory in dts file

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/boot/dts/xlp2xx-dpi-nae-mc.dts b/arch/mips/boot/dts/xlp2xx-dpi-nae-mc.dts
index 0b7229a..f3401d5 100644
--- a/arch/mips/boot/dts/xlp2xx-dpi-nae-mc.dts
+++ b/arch/mips/boot/dts/xlp2xx-dpi-nae-mc.dts
@@ -12,7 +12,7 @@
         memory {
                 device_type = "memory";
                 reg =  <0 0x01000000 0 0x0B000000   
-                        0 0xe0000000 0 0x20000000>;
+                        0 0x20000000 0 0x40000000>;
         };
 
 	hypervisor {
-- 
1.7.1

