From fa625914974a1230d6034f9e797426e29d54016e Mon Sep 17 00:00:00 2001
From: Pavani Reddy <pavani@broadcom.com>
Date: Fri, 28 Mar 2014 18:05:18 +0530
Subject: hybrid_nae_dts : changed the freein-fifo-mask for xlp9xx

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/boot/dts/xlp9xx-hybrid-nae-helinux.dts b/arch/mips/boot/dts/xlp9xx-hybrid-nae-helinux.dts
index bc4ecee..3c9e49b 100644
--- a/arch/mips/boot/dts/xlp9xx-hybrid-nae-helinux.dts
+++ b/arch/mips/boot/dts/xlp9xx-hybrid-nae-helinux.dts
@@ -56,7 +56,7 @@
 			};
 			nae-0 {
 				sh-domains = <1>;
-				freein-fifo-mask = <0xff00 0x0 0 0>;
+				freein-fifo-mask = <0x8000 0x0 0 0>;
 				/* node offsets and size */
 				vfbtbl-sw-offset = <0 0 0 0>;
 				vfbtbl-sw-nentries = <80 32 32 80>;
@@ -65,7 +65,7 @@
 			};
             nae-1 {
 				sh-domains = <1>;
-				freein-fifo-mask = <0xff00 0x0 0 0>;
+				freein-fifo-mask = <0x8000 0x0 0 0>;
 				/* node offsets and size */
 				vfbtbl-sw-offset = <0 0 0 0>;
 				vfbtbl-sw-nentries = <80 32 32 80>;
@@ -126,7 +126,7 @@
 				reg = <0x50000000 0x10000000>;  // 1280M -> 1536M (256M)
 			};
 			nae-0 {
-				freein-fifo-mask = <0x00ff 0 0 0>;
+				freein-fifo-mask = <0x7fff 0 0 0>;
 				sh-domains = <0>;
 				/* node offsets and size */
 				vfbtbl-sw-offset = <0 0 0 0>;
@@ -141,7 +141,7 @@
 				freein-fifo-replenish-desc-info = <0 0x00ff 2048 0 150>;
 			};
 			nae-1 {
-				freein-fifo-mask = <0x00ff 0 0 0>;
+				freein-fifo-mask = <0x7fff 0 0 0>;
 				sh-domains = <0>;
 				/* node offsets and size */
 				vfbtbl-sw-offset = <0 0 0 0>;
-- 
1.7.1

