From 0dd2c1e8c65ce627835a4f8e4ca06c8c138af14e Mon Sep 17 00:00:00 2001
From: Hareesh R <hareeshr@broadcom.com>
Date: Tue, 1 Apr 2014 17:10:01 +0530
Subject: crypto: VC specification in domain 1 for halx

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/boot/dts/xlp9xx-linux.dts b/arch/mips/boot/dts/xlp9xx-linux.dts
index 39b6a27..38c08d6 100644
--- a/arch/mips/boot/dts/xlp9xx-linux.dts
+++ b/arch/mips/boot/dts/xlp9xx-linux.dts
@@ -132,6 +132,8 @@
 			cpu {   
                                 onlinemask = <0xf>; 
                                 nae-rx-vc = <2>;
+				sae-rx-vc = <2>; 
+				sae-rx-sync-vc  = <3>; /* should be exclusive */
 			};
 
 			resource-allocation {
-- 
1.7.1

