From c3418e6cbea679a90a7e3c0860626522f641340b Mon Sep 17 00:00:00 2001
From: Pavani Reddy <pavani@broadcom.com>
Date: Mon, 7 Apr 2014 12:21:56 +0530
Subject: multinode dts : changed the memory region for mode 3 support

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/boot/dts/mnxlp8xx-helinux-nae.dts b/arch/mips/boot/dts/mnxlp8xx-helinux-nae.dts
index 46a4ffb..a5b1121 100644
--- a/arch/mips/boot/dts/mnxlp8xx-helinux-nae.dts
+++ b/arch/mips/boot/dts/mnxlp8xx-helinux-nae.dts
@@ -143,6 +143,7 @@
 				onlinemask = <0xffffffff 0xffffffff>;
 				enable_ici_pkt_traffic = <0>;
 				
+/*
 				//Enable this for mode #1 and #2 Local Node Memory - For 16G+16G
 				nae_0_pkt_pool_start = <0x140>;	//Multiple of 16MB
 				nae_0_pkt_pool_size = <0x1000000>;
@@ -150,21 +151,20 @@
 				nae_1_pkt_pool_size = <0x1000000>;
 
 				
-/*
-				//Enable this for mode #1 and #2 Local Node Memory - For 16G+16G
-				nae_0_pkt_pool_start = <0x140>;	//Multiple of 16MB
+				//Enable this for mode #3 Local Node Memory - For 16G+16G
+				nae_0_pkt_pool_start = <0x540>;	//Multiple of 16MB
 				nae_0_pkt_pool_size = <0x1000000>;
-				nae_1_pkt_pool_start = <0x241>;	//Multiple of 16MB
+				nae_1_pkt_pool_start = <0x140>;	//Multiple of 16MB
 				nae_1_pkt_pool_size = <0x1000000>;
 */
 
-/*
 				//Enable this for mode #1 and #2 Local Node Memory - For 8G+8G
 				nae_0_pkt_pool_start = <0x140>;	//Multiple of 16MB
 				nae_0_pkt_pool_size = <0x1000000>;
 				nae_1_pkt_pool_start = <0x241>;	//Multiple of 16MB
 				nae_1_pkt_pool_size = <0x1000000>;
 
+/*
 				//Enable this for mode #3 Remote Node Memory  - For 8G+8G
 				nae_0_pkt_pool_start = <0x241>;	//Multiple of 16MB
 				nae_0_pkt_pool_size = <0x1000000>;
diff --git a/arch/mips/boot/dts/mnxlp9xx-helinux-nae.dts b/arch/mips/boot/dts/mnxlp9xx-helinux-nae.dts
index 1b67fc7..8b4fd40 100644
--- a/arch/mips/boot/dts/mnxlp9xx-helinux-nae.dts
+++ b/arch/mips/boot/dts/mnxlp9xx-helinux-nae.dts
@@ -161,10 +161,10 @@
 				nae_1_pkt_pool_size = <0x1000000>;
 
 				
-				//Enable this for mode #1 and #2 Local Node Memory - For 16G+16G
-				nae_0_pkt_pool_start = <0x140>;	//Multiple of 16MB
+				//Enable this for mode #3 Local Node Memory - For 16G+16G
+				nae_0_pkt_pool_start = <0x540>;	//Multiple of 16MB
 				nae_0_pkt_pool_size = <0x1000000>;
-				nae_1_pkt_pool_start = <0x241>;	//Multiple of 16MB
+				nae_1_pkt_pool_start = <0x140>;	//Multiple of 16MB
 				nae_1_pkt_pool_size = <0x1000000>;
 */
 
-- 
1.7.1

