From 04ae59090aef9209cd02fc5841f5351327fa2eda Mon Sep 17 00:00:00 2001
From: Anurag <anurag.gopinath@broadcom.com>
Date: Mon, 7 Apr 2014 19:32:14 +0530
Subject: XLP9XX Multinode : Updates to make Slave node route packets

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/boot/dts/mnxlp9xx-linux.dts b/arch/mips/boot/dts/mnxlp9xx-linux.dts
index e26c3f8..3b062f0 100644
--- a/arch/mips/boot/dts/mnxlp9xx-linux.dts
+++ b/arch/mips/boot/dts/mnxlp9xx-linux.dts
@@ -113,7 +113,7 @@
             };
 
 			nae-0 {
-				freein-fifo-mask = <0xffff 0 0 0>;
+				freein-fifo-mask = <0xffff 0xffff 0 0>;
 				/* node offsets and size */
 				vfbtbl-sw-offset = <0 0 0 0>;
 				vfbtbl-sw-nentries = <80 32 32 80>;
@@ -123,7 +123,7 @@
 				dummy-pktdata-addr = <0x0 0x0ff00000>; // 255 MB
 			};
 			nae-1 {
-				freein-fifo-mask = <0xffff 0 0 0>;
+				freein-fifo-mask = <0xffff 0xffff 0 0>;
 				/* node offsets and size */
 				vfbtbl-sw-offset = <0 0 0 0>;
 				vfbtbl-sw-nentries = <80 80 80 80>;
-- 
1.7.1

