From 06d9ced7f0bf8690c7b3cf7c8d2e0cc34832e62a Mon Sep 17 00:00:00 2001
From: Saswat Kumar Dash <saswat.dash@broadcom.com>
Date: Mon, 5 May 2014 04:16:00 -0700
Subject: xlp5xx: modified vc mask for halx support

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/boot/dts/xlp5xx-linux.dts b/arch/mips/boot/dts/xlp5xx-linux.dts
index 5f75665..556b773 100644
--- a/arch/mips/boot/dts/xlp5xx-linux.dts
+++ b/arch/mips/boot/dts/xlp5xx-linux.dts
@@ -60,9 +60,9 @@
 				owner = <1>;
 				sharedcfg = <0x1f000000>;
 			};
-                        fmn {
-				/*node_0_vc_mask = <0xbbbbbbbb 0xbbbbbbbb 0xbbbbbbbb 0xbbbbbbbb >;*/
-				node_0_vc_mask = <0xffffffff 0xffffffff 0xffffffff 0xffffffff >;
+                        fmn {	/* make the node mask to 0xffffffff for SRIO */
+				node_0_vc_mask = <0xbbbbbbbb 0xbbbbbbbb 0xbbbbbbbb 0xbbbbbbbb >;
+				/*node_0_vc_mask = <0xffffffff 0xffffffff 0xffffffff 0xffffffff >;*/
                         };
 			pic {
 			};
-- 
1.7.1

