From abc01def540c11954371eb520f45bce58db74491 Mon Sep 17 00:00:00 2001
From: P. Sadik <psadik@broadcom.com>
Date: Fri, 16 May 2014 18:46:17 +0530
Subject: xlp5xx: Remove hardcoded MAC Address

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/boot/dts/xlp5xx-linux-tmpl.dts b/arch/mips/boot/dts/xlp5xx-linux-tmpl.dts
index b6fa45d..b7740a7 100644
--- a/arch/mips/boot/dts/xlp5xx-linux-tmpl.dts
+++ b/arch/mips/boot/dts/xlp5xx-linux-tmpl.dts
@@ -232,7 +232,7 @@
 		        interrupt-parent = <&gpio>;
 
 		        phy-mode = "mii";
-		        mac-address = [00 11 22 33 44 55];
+		        smsc,save-mac-address;
 		        smsc,irq-push-pull;
 		};
                 rio: xlp_rio@2c000 {
-- 
1.7.1

