From 9c766bf8e39c1507cc165088818056e8df839fef Mon Sep 17 00:00:00 2001
From: kopal <kopal@broadcom.com>
Date: Mon, 19 May 2014 17:49:23 +0530
Subject: xlp5xx: Added temperature sensor (max6653) to the correct I2C bus

Changed max6653 from i2c-2 to i2c-1.
[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/boot/dts/xlp5xx-linux-tmpl.dts b/arch/mips/boot/dts/xlp5xx-linux-tmpl.dts
index b7740a7..b7496bc 100644
--- a/arch/mips/boot/dts/xlp5xx-linux-tmpl.dts
+++ b/arch/mips/boot/dts/xlp5xx-linux-tmpl.dts
@@ -62,6 +62,11 @@
 			#size-cells = <0>;
 			reg = <0 0x113300 0x100>;
 			clock-frequency = <133000000>;
+			
+			dtt@2c {
+                                compatible = "max6653";
+                                reg = <0x2c>;
+			};
 		};
 		i2c2: i2c@113500 {
 			compatible = "netlogic,xlp9xx-i2c";
@@ -75,11 +80,6 @@
 				reg = <0x68>;
 			};
 
-			dtt@2c {
-                                compatible = "max6653";
-                                reg = <0x2c>;
-			};
-
 			eeprom@57 {
                                 compatible = "xlpge_eeprom";
                                 reg = <0x57>;
-- 
1.7.1

