From dcc4b9c9a46989aacda3d715aa62d5e7bb81c869 Mon Sep 17 00:00:00 2001
From: kopal <kopal@broadcom.com>
Date: Tue, 20 May 2014 09:43:13 +0530
Subject: xlp5xx: Added EEPROM to correct I2c bus.

changed EEPROM from i2c-2 to i2c-1
[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/boot/dts/xlp5xx-linux-tmpl.dts b/arch/mips/boot/dts/xlp5xx-linux-tmpl.dts
index b7496bc..524d87c 100644
--- a/arch/mips/boot/dts/xlp5xx-linux-tmpl.dts
+++ b/arch/mips/boot/dts/xlp5xx-linux-tmpl.dts
@@ -67,6 +67,11 @@
                                 compatible = "max6653";
                                 reg = <0x2c>;
 			};
+
+			eeprom@57 {
+                                compatible = "xlpge_eeprom";
+                                reg = <0x57>;
+                        };
 		};
 		i2c2: i2c@113500 {
 			compatible = "netlogic,xlp9xx-i2c";
@@ -79,11 +84,6 @@
 				compatible = "dallas,ds1374";
 				reg = <0x68>;
 			};
-
-			eeprom@57 {
-                                compatible = "xlpge_eeprom";
-                                reg = <0x57>;
-                        };
 		};
 		i2c3: i2c@113700 {
 			compatible = "netlogic,xlp9xx-i2c";
-- 
1.7.1

