From b318d4c0ddfa388381f7e6a13e3a50b3e4261da1 Mon Sep 17 00:00:00 2001
From: Saswat Kumar Dash <saswat.dash@broadcom.com>
Date: Thu, 22 May 2014 05:35:34 -0700
Subject: xlp5xx: changed onlinemask to 32cpus

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/boot/dts/xlp5xx-linux.dts b/arch/mips/boot/dts/xlp5xx-linux.dts
index d198bb8..c24b715 100644
--- a/arch/mips/boot/dts/xlp5xx-linux.dts
+++ b/arch/mips/boot/dts/xlp5xx-linux.dts
@@ -45,7 +45,7 @@
 			#cpumask-cells=<3>;
 
 			cpu {
-				onlinemask = <0x1>; /* */
+				onlinemask = <0xffffffff>; /* */
                                 nae-rx-vc = <0>;
                                 nae-fb-vc = <1>;
 				napi-vc-mask = <0xb>;
@@ -104,7 +104,7 @@
 			device_type = "domain";
 
 			cpu {
-                                onlinemask = <0x01>;
+                                onlinemask = <0xffffffff>;
                                 nae-rx-vc = <2>;
 			};
 
@@ -120,7 +120,7 @@
 				};
 			};
 			app-param{
-                                onlinemask = <0x01>;
+                                onlinemask = <0xffffffff>;
 			};
 		};
                 dom@2 {
-- 
1.7.1

