From 054bfdee41f59213334fc86b888c2d25d392574f Mon Sep 17 00:00:00 2001
From: Subhendu Sekhar Behera <sbehera@broadcom.com>
Date: Thu, 22 May 2014 19:16:02 +0530
Subject: XLP5XX: Enable watchdog driver

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/boot/dts/xlp5xx-linux-tmpl.dts b/arch/mips/boot/dts/xlp5xx-linux-tmpl.dts
index 524d87c..f77ce99 100644
--- a/arch/mips/boot/dts/xlp5xx-linux-tmpl.dts
+++ b/arch/mips/boot/dts/xlp5xx-linux-tmpl.dts
@@ -49,6 +49,10 @@
 			reg = <0 0x110000 0x200>;
 		};
 
+		watchdog@0 {
+			compatible = "netlogic,nlm-wdt";
+		};
+
 		i2c0: i2c@113100 {
 			compatible = "netlogic,xlp9xx-i2c";
 			#address-cells = <1>;
-- 
1.7.1

