From 77c75430f891a67c18df2eb62b455428ec891d3a Mon Sep 17 00:00:00 2001
From: Virendra Pathak <vpathak@broadcom.com>
Date: Tue, 22 Apr 2014 16:37:33 +0530
Subject: ale: added ale guest helinux dts file

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/boot/dts/xlp9xx-helinux-guest.dts b/arch/mips/boot/dts/xlp9xx-helinux-guest.dts
index 3644d87..495b90b 100644
--- a/arch/mips/boot/dts/xlp9xx-helinux-guest.dts
+++ b/arch/mips/boot/dts/xlp9xx-helinux-guest.dts
@@ -58,15 +58,16 @@
 
 			cpu {
 				dtre-rx-vc = <2>;
+				onlinemask = <0xffff>;
 			};
 			app-param {
-				onlinemask = <0x1>;
+				onlinemask = <0xffff>;
 			};
 		};
 	};
 
 	chosen {
-		netlogic,cpumask = <0x1>;
+		netlogic,cpumask = <0xffff>;
 		bootargs = " console=ttyS0,115200  rhash_entries=1048576";
 		linux,initrd-start=<0x0>;
 		linux,initrd-end=<0x0>;
-- 
1.7.1

