From 7d0c091f90cb8474a133530139c3eeab0bfaa441 Mon Sep 17 00:00:00 2001
From: Debayan Ghosh <debayan.ghosh@broadcom.com>
Date: Tue, 29 Apr 2014 17:47:14 +0530
Subject: soc : Removed num-nodes configuration from dts, runtime determination introduced

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/boot/dts/mn4xlp8xx-helinux-nae.dts b/arch/mips/boot/dts/mn4xlp8xx-helinux-nae.dts
index 0ec2e3d..583059c 100644
--- a/arch/mips/boot/dts/mn4xlp8xx-helinux-nae.dts
+++ b/arch/mips/boot/dts/mn4xlp8xx-helinux-nae.dts
@@ -185,10 +185,8 @@
 	};
 
 	soc {
-		nodes {
-			num-nodes = <4>;
-		};
 	};
+
 	frequency-config {
 		nae = <500>;
 		sae = <250>;
diff --git a/arch/mips/boot/dts/mn4xlp8xx-helinux-nat.dts b/arch/mips/boot/dts/mn4xlp8xx-helinux-nat.dts
index 7cc32b2..d76a328 100644
--- a/arch/mips/boot/dts/mn4xlp8xx-helinux-nat.dts
+++ b/arch/mips/boot/dts/mn4xlp8xx-helinux-nat.dts
@@ -211,11 +211,8 @@
 	};
 
 	soc {
-		nodes {
-			num-nodes = <4>;
-		};
-
 	};
+
 	frequency-config {
 		nae = <500>;
 		sae = <250>;
diff --git a/arch/mips/boot/dts/mn4xlp8xx-hybrid-nae-helinux.dts b/arch/mips/boot/dts/mn4xlp8xx-hybrid-nae-helinux.dts
index a05c148..c9b07f3 100644
--- a/arch/mips/boot/dts/mn4xlp8xx-hybrid-nae-helinux.dts
+++ b/arch/mips/boot/dts/mn4xlp8xx-hybrid-nae-helinux.dts
@@ -219,10 +219,8 @@
 	};
 
 	soc {
-		nodes {
-			num-nodes = <4>;
-		};
 	};
+
 	frequency-config {
 		nae = <500>;
 		sae = <250>;
diff --git a/arch/mips/boot/dts/mn4xlp8xx-linux.dts b/arch/mips/boot/dts/mn4xlp8xx-linux.dts
index 026c54c..0781b51 100644
--- a/arch/mips/boot/dts/mn4xlp8xx-linux.dts
+++ b/arch/mips/boot/dts/mn4xlp8xx-linux.dts
@@ -107,11 +107,8 @@
 	};
 
 	soc {
-		nodes {
-			num-nodes = <4>;
-		};
-
 	};
+
 	frequency-config {
 		nae = <500>;
 		sae = <250>;
diff --git a/arch/mips/boot/dts/mnxlp8xx-helinux-nae.dts b/arch/mips/boot/dts/mnxlp8xx-helinux-nae.dts
index a5b1121..3a0dd92 100644
--- a/arch/mips/boot/dts/mnxlp8xx-helinux-nae.dts
+++ b/arch/mips/boot/dts/mnxlp8xx-helinux-nae.dts
@@ -198,10 +198,8 @@
 	};
 
 	soc {
-		nodes {
-		num-nodes = <2>;
-		};
 	};
+	
 	frequency-config {
 		nae = <500>;
 		sae = <500>;
diff --git a/arch/mips/boot/dts/mnxlp8xx-helinux-nat-32GB.dts b/arch/mips/boot/dts/mnxlp8xx-helinux-nat-32GB.dts
index 3d09b7a..f70ed5d 100644
--- a/arch/mips/boot/dts/mnxlp8xx-helinux-nat-32GB.dts
+++ b/arch/mips/boot/dts/mnxlp8xx-helinux-nat-32GB.dts
@@ -216,11 +216,8 @@
 	};
 
 	soc {
-		nodes {
-			num-nodes = <2>;
-		};
-
 	};
+
 	frequency-config {
 		nae = <500>;
 		sae = <500>;
diff --git a/arch/mips/boot/dts/mnxlp8xx-helinux-nat.dts b/arch/mips/boot/dts/mnxlp8xx-helinux-nat.dts
index f456766..3c055d5 100644
--- a/arch/mips/boot/dts/mnxlp8xx-helinux-nat.dts
+++ b/arch/mips/boot/dts/mnxlp8xx-helinux-nat.dts
@@ -217,11 +217,8 @@
 	};
 
 	soc {
-		nodes {
-			num-nodes = <2>;
-		};
-
 	};
+	
 	frequency-config {
 		nae = <500>;
 		sae = <500>;
diff --git a/arch/mips/boot/dts/mnxlp8xx-hybrid-nae-helinux.dts b/arch/mips/boot/dts/mnxlp8xx-hybrid-nae-helinux.dts
index 55d2b4e..453e20c 100644
--- a/arch/mips/boot/dts/mnxlp8xx-hybrid-nae-helinux.dts
+++ b/arch/mips/boot/dts/mnxlp8xx-hybrid-nae-helinux.dts
@@ -211,10 +211,8 @@
 	};
 
 	soc {
-		nodes {
-			num-nodes = <2>;
-		};
 	};
+
 	frequency-config {
 		nae = <500>;
 		sae = <500>;
diff --git a/arch/mips/boot/dts/mnxlp8xx-linux.dts b/arch/mips/boot/dts/mnxlp8xx-linux.dts
index afaac3e..c39394c 100644
--- a/arch/mips/boot/dts/mnxlp8xx-linux.dts
+++ b/arch/mips/boot/dts/mnxlp8xx-linux.dts
@@ -112,11 +112,8 @@
 	};
 
 	soc {
-		nodes {
-			num-nodes = <2>;
-		};
-
 	};
+
 	frequency-config {
 		nae = <500>;
 		sae = <500>;
diff --git a/arch/mips/boot/dts/mnxlp9xx-helinux-nae.dts b/arch/mips/boot/dts/mnxlp9xx-helinux-nae.dts
index 8b4fd40..2864824 100644
--- a/arch/mips/boot/dts/mnxlp9xx-helinux-nae.dts
+++ b/arch/mips/boot/dts/mnxlp9xx-helinux-nae.dts
@@ -209,10 +209,8 @@
 	};
 
 	soc {
-		nodes {
-		num-nodes = <2>;
-		};
 	};
+
 	frequency-config {
 		nae = <500>;
 		sae = <500>;
diff --git a/arch/mips/boot/dts/mnxlp9xx-linux-dpi.dts b/arch/mips/boot/dts/mnxlp9xx-linux-dpi.dts
index 72c0bf9..aca679d 100644
--- a/arch/mips/boot/dts/mnxlp9xx-linux-dpi.dts
+++ b/arch/mips/boot/dts/mnxlp9xx-linux-dpi.dts
@@ -263,10 +263,6 @@
 	};
 
 	soc {
-		nodes {
-			num-nodes = <2>;
-		};
-
 	};
 };
 
diff --git a/arch/mips/boot/dts/mnxlp9xx-linux.dts b/arch/mips/boot/dts/mnxlp9xx-linux.dts
index 23326bc..b80151a 100644
--- a/arch/mips/boot/dts/mnxlp9xx-linux.dts
+++ b/arch/mips/boot/dts/mnxlp9xx-linux.dts
@@ -263,11 +263,8 @@
 	};
 
 	soc {
-		nodes {
-			num-nodes = <2>;
-		};
-
 	};
+
         frequency-config {
                 nae = <500>;
                 sae = <767>;
diff --git a/arch/mips/boot/dts/xlp8xx-helinux-nat.dts b/arch/mips/boot/dts/xlp8xx-helinux-nat.dts
index 82a19aa..fe055c4 100644
--- a/arch/mips/boot/dts/xlp8xx-helinux-nat.dts
+++ b/arch/mips/boot/dts/xlp8xx-helinux-nat.dts
@@ -161,11 +161,8 @@
 	};
 
 	soc {
-		nodes {
-			num-nodes = <1>;
-		};
-
 	};
+	
 	frequency-config {
 		nae = <500>;
 		sae = <500>;
diff --git a/arch/mips/boot/dts/xlp8xx-helinux-pcienet.dts b/arch/mips/boot/dts/xlp8xx-helinux-pcienet.dts
index 7611dba..30f24bb 100644
--- a/arch/mips/boot/dts/xlp8xx-helinux-pcienet.dts
+++ b/arch/mips/boot/dts/xlp8xx-helinux-pcienet.dts
@@ -118,10 +118,8 @@
 	};
 
 	soc {
-		nodes {
-			num-nodes = <1>;
-		};
 	};
+
 	frequency-config {
 		nae = <500>;
 		sae = <500>;
diff --git a/arch/mips/boot/dts/xlp8xx-helinux-pcieperf.dts b/arch/mips/boot/dts/xlp8xx-helinux-pcieperf.dts
index 8c63340..6960a6a 100644
--- a/arch/mips/boot/dts/xlp8xx-helinux-pcieperf.dts
+++ b/arch/mips/boot/dts/xlp8xx-helinux-pcieperf.dts
@@ -121,10 +121,8 @@
 	};
 
 	soc {
-		nodes {
-			num-nodes = <1>;
-		};
 	};
+
 	frequency-config {
 		nae = <250>;
 	};
diff --git a/arch/mips/boot/dts/xlp8xx-helinux-snort.dts b/arch/mips/boot/dts/xlp8xx-helinux-snort.dts
index 3a12d66..06ec562 100644
--- a/arch/mips/boot/dts/xlp8xx-helinux-snort.dts
+++ b/arch/mips/boot/dts/xlp8xx-helinux-snort.dts
@@ -243,11 +243,8 @@
 	};
 
 	soc {
-		nodes {
-			num-nodes = <1>;
-		};
-
 	};
+
 	frequency-config {
 		nae = <500>;
 		sae = <500>;
diff --git a/arch/mips/boot/dts/xlp8xx-helinux.dts b/arch/mips/boot/dts/xlp8xx-helinux.dts
index ff5f64a..ddcd8c4 100644
--- a/arch/mips/boot/dts/xlp8xx-helinux.dts
+++ b/arch/mips/boot/dts/xlp8xx-helinux.dts
@@ -262,11 +262,8 @@
 	};
 
 	soc {
-		nodes {
-			num-nodes = <1>;
-		};
-
 	};
+	
 	frequency-config {
 		nae = <500>;
 		sae = <500>;
diff --git a/arch/mips/boot/dts/xlp8xx-linux.dts b/arch/mips/boot/dts/xlp8xx-linux.dts
index a036670..8a27e1d 100644
--- a/arch/mips/boot/dts/xlp8xx-linux.dts
+++ b/arch/mips/boot/dts/xlp8xx-linux.dts
@@ -111,11 +111,8 @@
 	};
 
 	soc {
-		nodes {
-			num-nodes = <1>;
-		};
-
 	};
+
 	frequency-config {
 		nae = <500>;
 		sae = <500>;
diff --git a/arch/mips/boot/dts/xlp9xx-dpi-nae-mc.dts b/arch/mips/boot/dts/xlp9xx-dpi-nae-mc.dts
index a112ec4..55d31e7 100755
--- a/arch/mips/boot/dts/xlp9xx-dpi-nae-mc.dts
+++ b/arch/mips/boot/dts/xlp9xx-dpi-nae-mc.dts
@@ -131,10 +131,8 @@
 	};
 
 	soc {
-		nodes {
-			num-nodes = <1>;
-		};
 	};
+
 	frequency-config {
 		nae = <500>;
 		sae = <500>;
diff --git a/arch/mips/boot/dts/xlp9xx-helinux-guest.dts b/arch/mips/boot/dts/xlp9xx-helinux-guest.dts
index 495b90b..d9bb2fc 100644
--- a/arch/mips/boot/dts/xlp9xx-helinux-guest.dts
+++ b/arch/mips/boot/dts/xlp9xx-helinux-guest.dts
@@ -74,10 +74,6 @@
 	};
 
 	soc {
-		nodes {
-			num-nodes = <1>;
-		};
-
 	};
 };
 
diff --git a/arch/mips/boot/dts/xlp9xx-helinux-nat.dts b/arch/mips/boot/dts/xlp9xx-helinux-nat.dts
index 080721b..b964271 100644
--- a/arch/mips/boot/dts/xlp9xx-helinux-nat.dts
+++ b/arch/mips/boot/dts/xlp9xx-helinux-nat.dts
@@ -277,11 +277,8 @@ pktmem {
 	};
 
 	soc {
-		nodes {
-			num-nodes = <1>;
-		};
-
 	};
+
 	frequency-config {
 		nae = <500>;
 		sae = <250>;
diff --git a/arch/mips/boot/dts/xlp9xx-helinux-pcieperf.dts b/arch/mips/boot/dts/xlp9xx-helinux-pcieperf.dts
index c5e985d..3337af2 100644
--- a/arch/mips/boot/dts/xlp9xx-helinux-pcieperf.dts
+++ b/arch/mips/boot/dts/xlp9xx-helinux-pcieperf.dts
@@ -241,11 +241,8 @@
 	};
 
 	soc {
-		nodes {
-			num-nodes = <1>;
-		};
-
 	};
+
 	frequency-config {
 		nae = <500>;
 		sae = <250>;
diff --git a/arch/mips/boot/dts/xlp9xx-helinux-snort.dts b/arch/mips/boot/dts/xlp9xx-helinux-snort.dts
index cc89e29..84e4b94 100644
--- a/arch/mips/boot/dts/xlp9xx-helinux-snort.dts
+++ b/arch/mips/boot/dts/xlp9xx-helinux-snort.dts
@@ -308,11 +308,8 @@
 	};
 
 	soc {
-		nodes {
-			num-nodes = <1>;
-		};
-
 	};
+
 	frequency-config {
 		nae = <500>;
 		sae = <250>;
diff --git a/arch/mips/boot/dts/xlp9xx-helinux.dts b/arch/mips/boot/dts/xlp9xx-helinux.dts
index cb96f4e..1a2e00d 100644
--- a/arch/mips/boot/dts/xlp9xx-helinux.dts
+++ b/arch/mips/boot/dts/xlp9xx-helinux.dts
@@ -286,11 +286,8 @@
 	};
 
 	soc {
-		nodes {
-			num-nodes = <1>;
-		};
-
 	};
+
 	frequency-config {
 		nae = <500>;
 		sae = <250>;
diff --git a/arch/mips/boot/dts/xlp9xx-hybrid-nae-helinux.dts b/arch/mips/boot/dts/xlp9xx-hybrid-nae-helinux.dts
index 5dc67b7..d73b4dd 100644
--- a/arch/mips/boot/dts/xlp9xx-hybrid-nae-helinux.dts
+++ b/arch/mips/boot/dts/xlp9xx-hybrid-nae-helinux.dts
@@ -251,10 +251,8 @@
 	};
 
 	soc {
-		nodes {
-			num-nodes = <1>;
-		};
 	};
+
 	frequency-config {
 		nae = <500>;
 		sae = <500>;
diff --git a/arch/mips/boot/dts/xlp9xx-linux-dpi.dts b/arch/mips/boot/dts/xlp9xx-linux-dpi.dts
index 4362dbc..eb3ee50 100644
--- a/arch/mips/boot/dts/xlp9xx-linux-dpi.dts
+++ b/arch/mips/boot/dts/xlp9xx-linux-dpi.dts
@@ -220,10 +220,6 @@
 	};
 
 	soc {
-		nodes {
-			num-nodes = <1>;
-		};
-
 	};
 
         frequency-config {
diff --git a/arch/mips/boot/dts/xlp9xx-linux-guest.dts b/arch/mips/boot/dts/xlp9xx-linux-guest.dts
index 8e70f98..48fa5f8 100644
--- a/arch/mips/boot/dts/xlp9xx-linux-guest.dts
+++ b/arch/mips/boot/dts/xlp9xx-linux-guest.dts
@@ -89,10 +89,6 @@
 	};
 
 	soc {
-		nodes {
-			num-nodes = <1>;
-		};
-
 	};
 };
 
diff --git a/arch/mips/boot/dts/xlp9xx-linux.dts b/arch/mips/boot/dts/xlp9xx-linux.dts
index 38c08d6..8fd9577 100644
--- a/arch/mips/boot/dts/xlp9xx-linux.dts
+++ b/arch/mips/boot/dts/xlp9xx-linux.dts
@@ -229,10 +229,6 @@
 	};
 
 	soc {
-		nodes {
-			num-nodes = <1>;
-		};
-
 	};
 
         frequency-config {
-- 
1.7.1

