From 0610a6390c9a1d30fa2d9d00df8e8bf711610f06 Mon Sep 17 00:00:00 2001
From: Sreekanth Molagavalli <sreekanth.molagavalli@broadcom.com>
Date: Fri, 4 Jul 2014 16:02:29 +0530
Subject: xlp2xx_xmc: Change default cpu mask to 0xff.

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/boot/dts/xlp2xx-xmc-linux.dts b/arch/mips/boot/dts/xlp2xx-xmc-linux.dts
index efc7839..0ee7b9b 100644
--- a/arch/mips/boot/dts/xlp2xx-xmc-linux.dts
+++ b/arch/mips/boot/dts/xlp2xx-xmc-linux.dts
@@ -33,7 +33,7 @@
 			#size-cells = <1>;
 
 			cpu {
-				onlinemask = <0x1>;
+				onlinemask = <0xff>;
                                 nae-rx-vc = <0>;
                                 nae-fb-vc = <1>;
 				napi-vc-mask = <0xb>;
@@ -73,7 +73,7 @@
 	};
 
 	chosen {
-		netlogic,cpumask = <0x01>;
+		netlogic,cpumask = <0xff>;
 		bootargs = " console=ttyS0,115200  rhash_entries=1048576";
 		/*linux,initrd-start=<0x00c00000>;
 		linux,initrd-end=<0x02c00000>;*/
-- 
1.7.1

