From fe4a97b629a1de9e3bb9e82cbfdb40d2c00048aa Mon Sep 17 00:00:00 2001
From: rahul gupta <guptar@broadcom.com>
Date: Tue, 22 Jul 2014 20:02:22 +0530
Subject: dpdk: brcm_pcie_pmd added pcienet-dpdk-ucore

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/boot/dts/xlp2xx-pcie-helinux-pcienet.dts b/arch/mips/boot/dts/xlp2xx-pcie-helinux-pcienet.dts
index ead9dac..bff93ba 100644
--- a/arch/mips/boot/dts/xlp2xx-pcie-helinux-pcienet.dts
+++ b/arch/mips/boot/dts/xlp2xx-pcie-helinux-pcienet.dts
@@ -121,10 +121,15 @@
 
 	chosen {
         netlogic,cpumask = <0xff>;
-		/*"autoboot_pciedrv";*/
+		
+		/* For kernel host driver (enabled by default)-*/
+		/*"autoboot_pciedrv_2xx";*/
         /*bootargs = " xlp_noi2c console=ttyS0,115200 ";*/
-	    /*bootargs = " xlp_noi2c console=ttyS0,115200 autoboot_pciedrv console_over_pcie";*/
-	    bootargs = " xlp_noi2c console=ttyS0,115200 autoboot_pciedrv";
+	    /*bootargs = " xlp_noi2c console=ttyS0,115200 autoboot_pciedrv_2xx console_over_pcie";*/
+	    /*bootargs = " xlp_noi2c console=ttyS0,115200 autoboot_pciedrv_2xx";*/
+	    
+		/* For dpdk pmd driver enable- */
+		bootargs = " xlp_noi2c console=ttyS0,115200 autoboot_pciedrv_2xx_dpdk";
 	};
 
 	soc {
-- 
1.7.1

