From 4a10fa472a1e43b9a02bdf30c3b799918b0f79bf Mon Sep 17 00:00:00 2001
From: Krishnamurthy D V <krishnad@broadcom.com>
Date: Fri, 25 Jul 2014 03:55:39 -0700
Subject: pcieperf: Change memory range for pktmem

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/boot/dts/xlp5xx-pcie-helinux-pcieperf.dts b/arch/mips/boot/dts/xlp5xx-pcie-helinux-pcieperf.dts
index c74236d..410ac11 100644
--- a/arch/mips/boot/dts/xlp5xx-pcie-helinux-pcieperf.dts
+++ b/arch/mips/boot/dts/xlp5xx-pcie-helinux-pcieperf.dts
@@ -21,7 +21,7 @@
                 device_type = "pktmem";
 
                 /* format: node number, 64-bit addr, 64-bit size */
-                reg = <0 0x0 0xa0000000 0x0 0x80000000
+                reg = <0 0x0 0xe0000000 0x0 0x8000000
                        1 0x0 0x0 0x0 0x0
                        2 0x0 0x0 0x0 0x0
                        3 0x0 0x0 0x0 0x0>;
-- 
1.7.1

