From 2b015a70a0fbf32d9cba585cecf9f45df6e1da48 Mon Sep 17 00:00:00 2001
From: Siva Krishna Gudivada <sivag@broadcom.com>
Date: Thu, 14 Aug 2014 12:25:41 +0530
Subject: DPI: added pktmem section in dts files

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/boot/dts/xlp2xx-dpi-nae-mc.dts b/arch/mips/boot/dts/xlp2xx-dpi-nae-mc.dts
index 4093f75..3fc9ae1 100644
--- a/arch/mips/boot/dts/xlp2xx-dpi-nae-mc.dts
+++ b/arch/mips/boot/dts/xlp2xx-dpi-nae-mc.dts
@@ -15,6 +15,16 @@
                         0 0x20000000 0 0x40000000>;
         };
 
+        pktmem {
+           device_type = "pktmem";
+
+           /* format: node number, 64-bit addr, 64-bit size */
+           reg = <0 0 0x80000000 0 0x08000000
+                  1 0x0 0x0 0x0 0x0
+                  2 0x0 0x0 0x0 0x0
+                  3 0x0 0x0 0x0 0x0>;
+        };
+
 	hypervisor {
 		hypervisor-name = "Xen";
 		alloc_dom0_memory = <0>;
diff --git a/arch/mips/boot/dts/xlp3xx-dpi-nae-mc.dts b/arch/mips/boot/dts/xlp3xx-dpi-nae-mc.dts
index 80c2f9b..482315e 100644
--- a/arch/mips/boot/dts/xlp3xx-dpi-nae-mc.dts
+++ b/arch/mips/boot/dts/xlp3xx-dpi-nae-mc.dts
@@ -15,6 +15,17 @@
                 reg =  <0 0x00000000 0 0x0B000000   
                         0 0x20000000 0 0x40000000>;
         };
+
+        pktmem {
+                device_type = "pktmem";
+
+                /* format: node number, 64-bit addr, 64-bit size */
+                reg = <0 0 0x80000000 0 0x08000000
+                       1 0x0 0x0 0x0 0x0
+                       2 0x0 0x0 0x0 0x0
+                       3 0x0 0x0 0x0 0x0>;
+        };
+
 	hypervisor {
 		hypervisor-name = "Xen";
 		alloc_dom0_memory = <0>;
diff --git a/arch/mips/boot/dts/xlp5xx-dpi-nae-mc.dts b/arch/mips/boot/dts/xlp5xx-dpi-nae-mc.dts
index 0a5c97c..38c8830 100755
--- a/arch/mips/boot/dts/xlp5xx-dpi-nae-mc.dts
+++ b/arch/mips/boot/dts/xlp5xx-dpi-nae-mc.dts
@@ -1,10 +1,10 @@
-/* XLP9XX Device Tree Source
+/* XLP5XX Device Tree Source
  *
  */
 
 /dts-v1/;
 / {
-	model = "netlogic,XLP-GVP";
+	model = "netlogic,XLP-RVP";
         compatible = "netlogic,xlp";
         #address-cells = <2>;
         #size-cells = <2>;
-- 
1.7.1

