From 4599c7f2077cfd8d1a8f03c18909a236ad5fdd50 Mon Sep 17 00:00:00 2001
From: Prasanna Mumbai <prasanna.mumbai@broadcom.com>
Date: Thu, 21 Aug 2014 14:25:42 +0530
Subject: dtre: correct the dtre vc number.

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/boot/dts/xlp-config.dts b/arch/mips/boot/dts/xlp-config.dts
index 4a6f7b3..3b96f04 100644
--- a/arch/mips/boot/dts/xlp-config.dts
+++ b/arch/mips/boot/dts/xlp-config.dts
@@ -66,7 +66,7 @@
 				onlinemask = <0xffff>;
 				nae-rx-vc = <2>;
 				nae-fb-vc = <2>;
-				dtre-rx-vc = <3>;
+				dtre-rx-vc = <2>;
 				sae-rx-sync-vc = <2>;  /* sharing with nae */
 				sae-rx-vc = <2>; /* sharing with nae */
 				/* helinux, as only 1 thread per cpu, it should poll for both nae and sae
diff --git a/arch/mips/boot/dts/xlp9xx-config.dts b/arch/mips/boot/dts/xlp9xx-config.dts
index 4a302a5..76b5ae1 100644
--- a/arch/mips/boot/dts/xlp9xx-config.dts
+++ b/arch/mips/boot/dts/xlp9xx-config.dts
@@ -119,7 +119,7 @@
 				onlinemask = <0xffff>;
 				nae-rx-vc = <2>;
 				nae-fb-vc = <2>;
-				dtre-rx-vc = <3>;
+				dtre-rx-vc = <2>;
 				lzs-vc = <2>;
 				sae-rx-sync-vc = <2>;  /* sharing with nae */
 				sae-rx-vc = <2>; /* sharing with nae */
-- 
1.7.1

