From e8c02f18d2e4114994e36481a1530c8b8794a011 Mon Sep 17 00:00:00 2001
From: Virendra Pathak <vpathak@broadcom.com>
Date: Wed, 16 Jul 2014 12:26:11 +0530
Subject: DeviceManager: added device manager section in hybrid dts

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/boot/dts/xlp9xx-config.dts b/arch/mips/boot/dts/xlp9xx-config.dts
index 76b5ae1..9282584 100644
--- a/arch/mips/boot/dts/xlp9xx-config.dts
+++ b/arch/mips/boot/dts/xlp9xx-config.dts
@@ -159,6 +159,11 @@
 				freein-fifo-replenish-desc-info = <0 0x00ff 2048 0 150>;
 			};
 
+			device-manager {
+				port-mask = <0x3f>;
+				lifo-mask = <0x0 0xffffffff>;
+			};
+
 			resource-allocation {
 				net@0 {
 					port-mask = <0x3 0 0 0>;
diff --git a/arch/mips/boot/dts/xlp9xx-hybrid-nae-helinux.dts b/arch/mips/boot/dts/xlp9xx-hybrid-nae-helinux.dts
index e8db107..3aa2d8c 100644
--- a/arch/mips/boot/dts/xlp9xx-hybrid-nae-helinux.dts
+++ b/arch/mips/boot/dts/xlp9xx-hybrid-nae-helinux.dts
@@ -175,6 +175,12 @@
 				/* < node fifomask descsize headroom ndescs > */
 				freein-fifo-replenish-desc-info = <0 0x00ff 2048 0 750>;
 			};
+
+			device-manager {
+				port-mask = <0x3f>;
+				lifo-mask = <0x0 0xffffffff>;
+			};
+
 			resource-allocation {
 				net@0 {
 					port-mask = <0x3 0 0 0>;
-- 
1.7.1

