From c33cf696a19887c3ad3b21b0d440ea31b99e618f Mon Sep 17 00:00:00 2001
From: Subhendu Sekhar Behera <sbehera@broadcom.com>
Date: Thu, 25 Sep 2014 10:13:02 +0530
Subject: xlp_ep: Remove rtc configuration

Not needed in this board
[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/boot/dts/xlp2xx-ep-linux-kernel.dts b/arch/mips/boot/dts/xlp2xx-ep-linux-kernel.dts
index c10c558..8e9e50b 100644
--- a/arch/mips/boot/dts/xlp2xx-ep-linux-kernel.dts
+++ b/arch/mips/boot/dts/xlp2xx-ep-linux-kernel.dts
@@ -64,11 +64,6 @@
 			interrupt-parent = <&pic>;
 			interrupts = <31>;
 
-			rtc@68 {
-				compatible = "dallas,ds1374";
-				reg = <0x68>;
-			};
-
 			dtt@2c {
 				compatible = "maxim,max6639";
 				reg = <0x4c>;
-- 
1.7.1

