From f2574c3e5c92083dd4ee316de81e37db5c575121 Mon Sep 17 00:00:00 2001
From: Tanmay Jagdale <tanmayj@broadcom.com>
Date: Wed, 15 Oct 2014 17:16:46 +0530
Subject: dts: fixed the direct switch to raminit issue

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/boot/dts/xlp2xx-ep-linux-kernel.dts b/arch/mips/boot/dts/xlp2xx-ep-linux-kernel.dts
index c67566c..f64e2c9 100644
--- a/arch/mips/boot/dts/xlp2xx-ep-linux-kernel.dts
+++ b/arch/mips/boot/dts/xlp2xx-ep-linux-kernel.dts
@@ -248,7 +248,7 @@
 	};
 
 	chosen {
-		bootargs = "console=ttyS0,115200 rdinit=/sbin/init";
+		bootargs = "console=ttyS0,115200 rhash_entries=1048576";
 		netlogic,cpumask = <0xff>;
 		netlogic,reserve-mem = <256>;
 	};
diff --git a/arch/mips/boot/dts/xlp2xx-linux-kernel.dts b/arch/mips/boot/dts/xlp2xx-linux-kernel.dts
index b485ac0..007c635 100644
--- a/arch/mips/boot/dts/xlp2xx-linux-kernel.dts
+++ b/arch/mips/boot/dts/xlp2xx-linux-kernel.dts
@@ -242,7 +242,7 @@
 	};
 
 	chosen {
-		bootargs = "console=ttyS0,115200 rdinit=/sbin/init";
+		bootargs = "console=ttyS0,115200 rhash_entries=1048576";
 		netlogic,cpumask = <0xff>;
 		netlogic,reserve-mem = <256>;
 	};
diff --git a/arch/mips/boot/dts/xlp2xx-pcie-linux-kernel.dts b/arch/mips/boot/dts/xlp2xx-pcie-linux-kernel.dts
index d6563b8..ee08600 100644
--- a/arch/mips/boot/dts/xlp2xx-pcie-linux-kernel.dts
+++ b/arch/mips/boot/dts/xlp2xx-pcie-linux-kernel.dts
@@ -154,7 +154,7 @@
 	};
 
 	chosen {
-		bootargs = "console=ttyS0,115200 rdinit=/sbin/init";
+		bootargs = "console=ttyS0,115200 rhash_entries=1048576";
 		netlogic,cpumask = <0xff>;
 		netlogic,reserve-mem = <256>;
 	};
diff --git a/arch/mips/boot/dts/xlp2xx-xmc-linux-kernel.dts b/arch/mips/boot/dts/xlp2xx-xmc-linux-kernel.dts
index 861c6e1..4f4d483 100644
--- a/arch/mips/boot/dts/xlp2xx-xmc-linux-kernel.dts
+++ b/arch/mips/boot/dts/xlp2xx-xmc-linux-kernel.dts
@@ -137,7 +137,7 @@
 	};
 
 	chosen {
-		bootargs = "console=ttyS0,115200 rdinit=/sbin/init";
+		bootargs = "console=ttyS0,115200 rhash_entries=1048576";
 		netlogic,cpumask = <0xff>;
 		netlogic,reserve-mem = <128>;
 	};
diff --git a/arch/mips/boot/dts/xlp3xx-linux-kernel.dts b/arch/mips/boot/dts/xlp3xx-linux-kernel.dts
index 35f0cf5..7c165e2 100644
--- a/arch/mips/boot/dts/xlp3xx-linux-kernel.dts
+++ b/arch/mips/boot/dts/xlp3xx-linux-kernel.dts
@@ -247,7 +247,7 @@
 	};
 
 	chosen {
-		bootargs = "console=ttyS0,115200 rdinit=/sbin/init";
+		bootargs = "console=ttyS0,115200 rhash_entries=1048576";
 		netlogic,cpumask = <0xffff>;
 		netlogic,reserve-mem = <256>;
 	};
diff --git a/arch/mips/boot/dts/xlp3xx-xmc-linux-kernel.dts b/arch/mips/boot/dts/xlp3xx-xmc-linux-kernel.dts
index 3ebfc73..6d08112 100644
--- a/arch/mips/boot/dts/xlp3xx-xmc-linux-kernel.dts
+++ b/arch/mips/boot/dts/xlp3xx-xmc-linux-kernel.dts
@@ -143,7 +143,7 @@
 	};
 
 	chosen {
-		bootargs = "console=ttyS0,115200 rdinit=/sbin/init";
+		bootargs = "console=ttyS0,115200 rhash_entries=1048576";
 		netlogic,cpumask = <0xffff>;
 		netlogic,reserve-mem = <128>;
 	};
diff --git a/arch/mips/boot/dts/xlp5xx-linux-kernel.dts b/arch/mips/boot/dts/xlp5xx-linux-kernel.dts
index 02099c7..145ae42 100644
--- a/arch/mips/boot/dts/xlp5xx-linux-kernel.dts
+++ b/arch/mips/boot/dts/xlp5xx-linux-kernel.dts
@@ -245,7 +245,7 @@
 	};
 
 	chosen {
-		bootargs = "console=ttyS0,115200 rdinit=/sbin/init";
+		bootargs = "console=ttyS0,115200 rhash_entries=1048576";
 		netlogic,cpumask = <0xffffffff>;
 		netlogic,reserve-mem = <256>;
 		netlogic,frfifo-mem = <128>;
diff --git a/arch/mips/boot/dts/xlp5xx-pcie-linux-kernel.dts b/arch/mips/boot/dts/xlp5xx-pcie-linux-kernel.dts
index b78c1ab..f259378 100644
--- a/arch/mips/boot/dts/xlp5xx-pcie-linux-kernel.dts
+++ b/arch/mips/boot/dts/xlp5xx-pcie-linux-kernel.dts
@@ -198,7 +198,7 @@
 	};
 
 	chosen {
-		bootargs = "console=ttyS0,115200 rdinit=/sbin/init";
+		bootargs = "console=ttyS0,1152 rhash_entries=1048576";
 		netlogic,reserve-mem = <256>;
 	};
 };
diff --git a/arch/mips/boot/dts/xlp8xx-linux-kernel.dts b/arch/mips/boot/dts/xlp8xx-linux-kernel.dts
index 1f65bb1..ebdd682 100644
--- a/arch/mips/boot/dts/xlp8xx-linux-kernel.dts
+++ b/arch/mips/boot/dts/xlp8xx-linux-kernel.dts
@@ -229,7 +229,7 @@
 	};
 
 	chosen {
-		bootargs = "console=ttyS0,115200 rdinit=/sbin/init";
+		bootargs = "console=ttyS0,115200 rhash_entries=1048576";
 		netlogic,cpumask = <0xffffffff>;
 		netlogic,reserve-mem = <256>;
 		netlogic,frfifo-mem = <128>;
diff --git a/arch/mips/boot/dts/xlp8xx-linux-mn-kernel.dts b/arch/mips/boot/dts/xlp8xx-linux-mn-kernel.dts
index 62d9ca9..52ce32c 100644
--- a/arch/mips/boot/dts/xlp8xx-linux-mn-kernel.dts
+++ b/arch/mips/boot/dts/xlp8xx-linux-mn-kernel.dts
@@ -187,4 +187,11 @@
 			interrupt-controller;
 		};
 	};
+
+	chosen {
+		bootargs = "console=ttyS0,115200 rhash_entries=1048576";
+		netlogic.cpumask = <0xffffffff>;
+		netlogic,reserve-mem = <256>;
+		netlogic,frin-fifo-mem = <128>;
+	};
 };
diff --git a/arch/mips/boot/dts/xlp8xx-linux-mn4-kernel.dts b/arch/mips/boot/dts/xlp8xx-linux-mn4-kernel.dts
index d81de53..22a9d8c 100644
--- a/arch/mips/boot/dts/xlp8xx-linux-mn4-kernel.dts
+++ b/arch/mips/boot/dts/xlp8xx-linux-mn4-kernel.dts
@@ -211,4 +211,11 @@
 			reg = <0 0xea000 0x30>;
 		};
 	};
+
+	chosen {
+		bootargs = "console=ttyS0,115200 rhash_entries=1048576";
+		netlogic.cpumask = <0xffffffff>;
+		netlogic,reserve-mem = <256>;
+		netlogic,frin-fifo-mem = <128>;
+	};
 };
diff --git a/arch/mips/boot/dts/xlp9xx-linux-kernel.dts b/arch/mips/boot/dts/xlp9xx-linux-kernel.dts
index 8518d01..1d0d34d 100644
--- a/arch/mips/boot/dts/xlp9xx-linux-kernel.dts
+++ b/arch/mips/boot/dts/xlp9xx-linux-kernel.dts
@@ -235,7 +235,7 @@
 	};
 
 	chosen {
-		bootargs = "console=ttyS0,115200 rdinit=/sbin/init";
+		bootargs = "console=ttyS0,115200 rhash_entries=1048576";
 		netlogic,cpumask = <0xffffffff 0xffffffff>;
 		netlogic,reserve-mem = <256>;
 		netlogic,frfifo-mem = <256>;
diff --git a/arch/mips/boot/dts/xlp9xx-linux-mn-kernel.dts b/arch/mips/boot/dts/xlp9xx-linux-mn-kernel.dts
index 241e428..25a61e1 100644
--- a/arch/mips/boot/dts/xlp9xx-linux-mn-kernel.dts
+++ b/arch/mips/boot/dts/xlp9xx-linux-mn-kernel.dts
@@ -207,7 +207,7 @@
 	};
 
 	chosen {
-		bootargs = "console=ttyS0,115200 rdinit=/sbin/init";
+		bootargs = "console=ttyS0,115200 rhash_entries=1048576";
 		netlogic,cpumask = <0xffffffff 0xffffffff>;
 		netlogic,reserve-mem = <256 256>;
 	};
-- 
1.7.1

