From 2c1e282e51d70a8a06b9dd6d4cd9da1a5e803c40 Mon Sep 17 00:00:00 2001
From: Subhendu Sekhar Behera <sbehera@broadcom.com>
Date: Mon, 20 Oct 2014 11:12:23 +0530
Subject: i2c: add interrupt support

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/boot/dts/xlp5xx-linux-kernel.dts b/arch/mips/boot/dts/xlp5xx-linux-kernel.dts
index 10669f5..ff3ddc2 100644
--- a/arch/mips/boot/dts/xlp5xx-linux-kernel.dts
+++ b/arch/mips/boot/dts/xlp5xx-linux-kernel.dts
@@ -57,18 +57,22 @@
 		};
 
 		i2c0: i2c@113100 {
-			compatible = "netlogic,xlp9xx-i2c";
+			compatible = "netlogic,xlp-i2c";
 			#address-cells = <1>;
 			#size-cells = <0>;
 			reg = <0 0x113100 0x100>;
 			clock-frequency = <133000000>;
+			interrupts = <30>;
+			interrupt-parent = <&pic>;
 		};
 		i2c1: i2c@113300 {
-			compatible = "netlogic,xlp9xx-i2c";
+			compatible = "netlogic,xlp-i2c";
 			#address-cells = <1>;
 			#size-cells = <0>;
 			reg = <0 0x113300 0x100>;
 			clock-frequency = <133000000>;
+			interrupts = <31>;
+			interrupt-parent = <&pic>;
 			
 			dtt@2c {
                                 compatible = "max6653";
@@ -81,11 +85,13 @@
                         };
 		};
 		i2c2: i2c@113500 {
-			compatible = "netlogic,xlp9xx-i2c";
+			compatible = "netlogic,xlp-i2c";
 			#address-cells = <1>;
 			#size-cells = <0>;
 			reg = <0 0x113500 0x100>;
 			clock-frequency = <133000000>;
+			interrupts = <32>;
+			interrupt-parent = <&pic>;
 
 			rtc@68 {
 				compatible = "dallas,ds1374";
@@ -93,11 +99,13 @@
 			};
 		};
 		i2c3: i2c@113700 {
-			compatible = "netlogic,xlp9xx-i2c";
+			compatible = "netlogic,xlp-i2c";
 			#address-cells = <1>;
 			#size-cells = <0>;
 			reg = <0 0x113700 0x100>;
 			clock-frequency = <133000000>;
+			interrupts = <33>;
+			interrupt-parent = <&pic>;
 		};
 
 		spi: xlp_spi@13a000 {
diff --git a/arch/mips/boot/dts/xlp9xx-linux-kernel.dts b/arch/mips/boot/dts/xlp9xx-linux-kernel.dts
index d79c517..1bd9cc5 100644
--- a/arch/mips/boot/dts/xlp9xx-linux-kernel.dts
+++ b/arch/mips/boot/dts/xlp9xx-linux-kernel.dts
@@ -62,18 +62,22 @@
 		};
 
 		i2c0: i2c@113100 {
-			compatible = "netlogic,xlp9xx-i2c";
+			compatible = "netlogic,xlp-i2c";
 			#address-cells = <1>;
 			#size-cells = <0>;
 			reg = <0 0x113100 0x100>;
 			clock-frequency = <133000000>;
+			interrupts = <30>;
+			interrupt-parent = <&pic>;
 		};
 		i2c1: i2c@113300 {
-			compatible = "netlogic,xlp9xx-i2c";
+			compatible = "netlogic,xlp-i2c";
 			#address-cells = <1>;
 			#size-cells = <0>;
 			reg = <0 0x113300 0x100>;
 			clock-frequency = <133000000>;
+			interrupts = <31>;
+			interrupt-parent = <&pic>;
 
 			rtc@68 {
 				compatible = "dallas,ds1374";
@@ -92,18 +96,22 @@
 
 		};
 		i2c2: i2c@113500 {
-			compatible = "netlogic,xlp9xx-i2c";
+			compatible = "netlogic,xlp-i2c";
 			#address-cells = <1>;
 			#size-cells = <0>;
 			reg = <0 0x113500 0x100>;
 			clock-frequency = <133000000>;
+			interrupts = <32>;
+			interrupt-parent = <&pic>;
 		};
 		i2c3: i2c@113700 {
-			compatible = "netlogic,xlp9xx-i2c";
+			compatible = "netlogic,xlp-i2c";
 			#address-cells = <1>;
 			#size-cells = <0>;
 			reg = <0 0x113700 0x100>;
 			clock-frequency = <133000000>;
+			interrupts = <33>;
+			interrupt-parent = <&pic>;
 		};
 
 		spi: xlp_spi@13a000 {
-- 
1.7.1

