From a6e055588c88182b7c43ee9b11f9aa1219f7cace Mon Sep 17 00:00:00 2001
From: Tanmay Jagdale <tanmayj@broadcom.com>
Date: Fri, 28 Nov 2014 09:55:39 +0530
Subject: xlp9xx-mn: fixed the cpumask in multinode dts file

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/boot/dts/xlp9xx-linux-mn-kernel.dts b/arch/mips/boot/dts/xlp9xx-linux-mn-kernel.dts
index 37cbabf..5fa4591 100644
--- a/arch/mips/boot/dts/xlp9xx-linux-mn-kernel.dts
+++ b/arch/mips/boot/dts/xlp9xx-linux-mn-kernel.dts
@@ -217,7 +217,8 @@
 
 	chosen {
 		bootargs = "console=ttyS0,115200 rhash_entries=1048576";
-		netlogic,cpumask = <0xffffffff 0xffffffff>;
+		netlogic,cpumask = <0x0 0x0 0xffffffff 0xffffffff
+					0x0 0x0 0xffffffff 0xffffffff>;
 		netlogic,reserve-mem = <256 256>;
 	};
 };
-- 
1.7.1

