From 0128e31bbb9c3d086c1ef17b456492e72819e436 Mon Sep 17 00:00:00 2001
From: Subhendu Sekhar Behera <sbehera@broadcom.com>
Date: Mon, 24 Nov 2014 12:05:47 +0530
Subject: dts: random number generator support

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/boot/dts/xlp2xx-ep-linux-kernel.dts b/arch/mips/boot/dts/xlp2xx-ep-linux-kernel.dts
index 183f719..8d8b62d 100644
--- a/arch/mips/boot/dts/xlp2xx-ep-linux-kernel.dts
+++ b/arch/mips/boot/dts/xlp2xx-ep-linux-kernel.dts
@@ -100,7 +100,7 @@
 		};
 
 		rng@0 {
-			compatible = "netlogic,nlm-drng";
+			compatible = "netlogic,xlp-drng";
 			reg = <0 0x2a000 0x30>;
 		};
 
diff --git a/arch/mips/boot/dts/xlp2xx-linux-kernel.dts b/arch/mips/boot/dts/xlp2xx-linux-kernel.dts
index d1132bc..76182b4 100644
--- a/arch/mips/boot/dts/xlp2xx-linux-kernel.dts
+++ b/arch/mips/boot/dts/xlp2xx-linux-kernel.dts
@@ -94,7 +94,7 @@
 		};
 
 		rng@0 {
-			compatible = "netlogic,nlm-drng";
+			compatible = "netlogic,xlp-drng";
 			reg = <0 0x2a000 0x30>;
 		};
 
diff --git a/arch/mips/boot/dts/xlp2xx-pcie-linux-kernel.dts b/arch/mips/boot/dts/xlp2xx-pcie-linux-kernel.dts
index 4363005..fb13ff3 100644
--- a/arch/mips/boot/dts/xlp2xx-pcie-linux-kernel.dts
+++ b/arch/mips/boot/dts/xlp2xx-pcie-linux-kernel.dts
@@ -100,7 +100,7 @@
 		};
 
 		rng@0 {
-			compatible = "netlogic,nlm-drng";
+			compatible = "netlogic,xlp-drng";
 			reg = <0 0x2a000 0x30>;
 		};
 
diff --git a/arch/mips/boot/dts/xlp2xx-xmc-linux-kernel.dts b/arch/mips/boot/dts/xlp2xx-xmc-linux-kernel.dts
index 4f4d483..802f266 100644
--- a/arch/mips/boot/dts/xlp2xx-xmc-linux-kernel.dts
+++ b/arch/mips/boot/dts/xlp2xx-xmc-linux-kernel.dts
@@ -80,7 +80,7 @@
 		};
 
 		rng@0 {
-			compatible = "netlogic,nlm-drng";
+			compatible = "netlogic,xlp-drng";
 			reg = <0 0x2a000 0x30>;
 		};
 
diff --git a/arch/mips/boot/dts/xlp3xx-linux-kernel.dts b/arch/mips/boot/dts/xlp3xx-linux-kernel.dts
index 0a642b2..bfbe7a6 100644
--- a/arch/mips/boot/dts/xlp3xx-linux-kernel.dts
+++ b/arch/mips/boot/dts/xlp3xx-linux-kernel.dts
@@ -94,7 +94,7 @@
 		};
 
 		rng@0 {
-			compatible = "netlogic,nlm-drng";
+			compatible = "netlogic,xlp-drng";
 			reg = <0 0x2a000 0x30>;
 		};
 
diff --git a/arch/mips/boot/dts/xlp3xx-xmc-linux-kernel.dts b/arch/mips/boot/dts/xlp3xx-xmc-linux-kernel.dts
index bbac1ac..9230389 100644
--- a/arch/mips/boot/dts/xlp3xx-xmc-linux-kernel.dts
+++ b/arch/mips/boot/dts/xlp3xx-xmc-linux-kernel.dts
@@ -95,7 +95,7 @@
 		};
 
 		rng@0 {
-			compatible = "netlogic,nlm-drng";
+			compatible = "netlogic,xlp-drng";
 			reg = <0 0x2a000 0x30>;
 		};
 
diff --git a/arch/mips/boot/dts/xlp4xx-linux-kernel.dts b/arch/mips/boot/dts/xlp4xx-linux-kernel.dts
index cb35d1f..d3900c2 100644
--- a/arch/mips/boot/dts/xlp4xx-linux-kernel.dts
+++ b/arch/mips/boot/dts/xlp4xx-linux-kernel.dts
@@ -101,7 +101,7 @@
 		};
 
 		rng@0 {
-			compatible = "netlogic,nlm-drng";
+			compatible = "netlogic,xlp-drng";
 			reg = <0 0x2a000 0x30>;
 		};
 
diff --git a/arch/mips/boot/dts/xlp5xx-linux-kernel.dts b/arch/mips/boot/dts/xlp5xx-linux-kernel.dts
index ff3ddc2..fba978a 100644
--- a/arch/mips/boot/dts/xlp5xx-linux-kernel.dts
+++ b/arch/mips/boot/dts/xlp5xx-linux-kernel.dts
@@ -56,6 +56,11 @@
 			compatible = "netlogic,nlm-wdt";
 		};
 
+		rng@0 {
+			compatible = "netlogic,xlp-drng";
+			reg = <0 0x12a000 0x80>;
+		};
+
 		i2c0: i2c@113100 {
 			compatible = "netlogic,xlp-i2c";
 			#address-cells = <1>;
diff --git a/arch/mips/boot/dts/xlp5xx-pcie-linux-kernel.dts b/arch/mips/boot/dts/xlp5xx-pcie-linux-kernel.dts
index b4292d4..db489fe 100644
--- a/arch/mips/boot/dts/xlp5xx-pcie-linux-kernel.dts
+++ b/arch/mips/boot/dts/xlp5xx-pcie-linux-kernel.dts
@@ -62,6 +62,11 @@
 			compatible = "netlogic,nlm-wdt";
 		};
 
+		rng@0 {
+			compatible = "netlogic,xlp-drng";
+			reg = <0 0x12a000 0x80>;
+		};
+
 		i2c0: i2c@113100 {
 			compatible = "netlogic,xlp9xx-i2c";
 			#address-cells = <1>;
diff --git a/arch/mips/boot/dts/xlp8xx-linux-kernel.dts b/arch/mips/boot/dts/xlp8xx-linux-kernel.dts
index 51826dd..bbbfa9e 100644
--- a/arch/mips/boot/dts/xlp8xx-linux-kernel.dts
+++ b/arch/mips/boot/dts/xlp8xx-linux-kernel.dts
@@ -94,7 +94,7 @@
 		};
 
 		rng@0 {
-			compatible = "netlogic,nlm-drng";
+			compatible = "netlogic,xlp-drng";
 			reg = <0 0x2a000 0x30>;
 		};
 
diff --git a/arch/mips/boot/dts/xlp8xx-linux-mn-kernel.dts b/arch/mips/boot/dts/xlp8xx-linux-mn-kernel.dts
index 9a28ef9..555218b 100644
--- a/arch/mips/boot/dts/xlp8xx-linux-mn-kernel.dts
+++ b/arch/mips/boot/dts/xlp8xx-linux-mn-kernel.dts
@@ -39,7 +39,7 @@
 		};
 
 		rng@0 {
-			compatible = "netlogic,nlm-drng";
+			compatible = "netlogic,xlp-drng";
 			reg = <0 0x2a000 0x30>;
 		};
 
@@ -160,7 +160,7 @@
 		};
 
 		rng@1 {
-			compatible = "netlogic,nlm-drng";
+			compatible = "netlogic,xlp-drng";
 			reg = <0 0x6a000 0x30>;
 		};
 
diff --git a/arch/mips/boot/dts/xlp8xx-linux-mn4-kernel.dts b/arch/mips/boot/dts/xlp8xx-linux-mn4-kernel.dts
index 328e701..f58029c 100644
--- a/arch/mips/boot/dts/xlp8xx-linux-mn4-kernel.dts
+++ b/arch/mips/boot/dts/xlp8xx-linux-mn4-kernel.dts
@@ -39,7 +39,7 @@
 		};
 
 		rng@0 {
-			compatible = "netlogic,nlm-drng";
+			compatible = "netlogic,xlp-drng";
 			reg = <0 0x2a000 0x30>;
 		};
 
@@ -160,7 +160,7 @@
 		};
 
 		rng@1 {
-			compatible = "netlogic,nlm-drng";
+			compatible = "netlogic,xlp-drng";
 			reg = <0 0x6a000 0x30>;
 		};
 
diff --git a/arch/mips/boot/dts/xlp9xx-linux-kernel.dts b/arch/mips/boot/dts/xlp9xx-linux-kernel.dts
index 0016710..9a342e3 100644
--- a/arch/mips/boot/dts/xlp9xx-linux-kernel.dts
+++ b/arch/mips/boot/dts/xlp9xx-linux-kernel.dts
@@ -57,7 +57,7 @@
 		};
 
 		rng@0 {
-			compatible = "netlogic,nlm-drng";
+			compatible = "netlogic,xlp-drng";
 			reg = <0 0x12a000 0x80>;
 		};
 
diff --git a/arch/mips/boot/dts/xlp9xx-linux-mn-kernel.dts b/arch/mips/boot/dts/xlp9xx-linux-mn-kernel.dts
index 5fa4591..93add7f 100644
--- a/arch/mips/boot/dts/xlp9xx-linux-mn-kernel.dts
+++ b/arch/mips/boot/dts/xlp9xx-linux-mn-kernel.dts
@@ -47,7 +47,7 @@
 		};
 
 		rng@0 {
-			compatible = "netlogic,nlm-drng";
+			compatible = "netlogic,xlp-drng";
 			reg = <0 0x12a000 0x80>;
 		};
 
-- 
1.7.1

