From 44f59aa8ecdb275d07007be6ca3ccf44d5c69997 Mon Sep 17 00:00:00 2001
From: Anusha Nelluri <anushan@broadcom.com>
Date: Mon, 24 Nov 2014 17:59:47 +0530
Subject: dts: Added xlp9xx-guest2.dts file

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/boot/dts/xlp9xx-guest2.dts b/arch/mips/boot/dts/xlp9xx-guest2.dts
new file mode 100644
index 0000000..4380708
--- /dev/null
+++ b/arch/mips/boot/dts/xlp9xx-guest2.dts
@@ -0,0 +1,143 @@
+/*
+ * Device Tree Source
+ * For XLPII Guest
+ */
+
+/dts-v1/;
+/ {
+	model = "netlogic,XLP-GVP";
+	compatible = "netlogic,xlp";
+	#address-cells = <0x2>;
+	#size-cells = <0x2>;
+
+	memory {
+		device_type = "memory";
+		reg = <0x0 0x100000 0x0 0xff00000 0x0 0x20000000 0x0 0xa0000000>;
+	};
+
+	pktmem {
+		device_type = "pktmem";
+		reg = <0x0 0x2 0x20000000 0x0 0x10000000
+			0x1 0x0 0x0 0x0 0x0
+			0x2 0x0 0x0 0x0 0x0
+			0x3 0x0 0x0 0x0 0x0>;
+	};
+
+	doms {
+		#address-cells = <0x1>;
+		#size-cells = <0x1>;
+
+		dom@0 {
+			device_type = "domain";
+			os = "linux";
+
+			resource-allocation {
+				net@0 {
+					port-mask = <0x4 0x0 0x0 0x0>;
+					frinlifo-mask = <0xffff 0x0 0x0 0x0>;
+					frinlifo-shared-mode = <0 0 0 0>;
+					vfbid-sw-offset = <0x0 0x0 0x0 0x0>;
+					vfbid-sw-entries = <0x0 0x0 0x0 0x0>;
+					ucore-mask = <0x1 0x0 0x0 0x0>;
+					sram-off = <0x0 0x0 0x0 0x0>;
+				};
+
+				net@1 {
+					port-mask = <0x4 0 0 0>;
+					frinlifo-mask = <0xffff 0 0 0>;
+					frinlifo-shared-mode = <0 0 0 0>;
+					vfbid-sw-offset = <0 0 0 0>;
+					vfbid-sw-entries = <0 0 0 0>;
+					ucore-mask = <0x1 0 0 0>;
+					sram-off = <0 0 0 0>;
+				};
+			};
+
+			dtre {
+				dtre_init_only = <0x0>;
+			};
+		};
+
+		dom@1 {
+			device_type = "domain";
+			os = "linux";
+
+			cpu {
+				onlinemask = <0xffff0000>;
+				nae-rx-vc = <0x2>;
+				dtre-rx-vc = <0x2>;
+			};
+
+			app-param {
+				onlinemask = <0xffff0000>;
+			};
+
+			resource-allocation {
+				net@0 {
+					port-mask = <0x3 0x0 0x0 0x0>;
+					frinlifo-mask = <0xff 0x0 0x0 0x0>;
+					frinlifo-shared-mode = <0 0 0 0>;
+					vfbid-sw-offset = <0x0 0x0 0x0 0x0>;
+					vfbid-sw-entries = <0x0 0x0 0x0 0x0>;
+					ucore-mask = <0x2 0x0 0x0 0x0>;
+					sram-off = <0x0 0x0 0x0 0x0>;
+				};
+				net@1 {
+					port-mask = <0x3 0x0 0x0 0x0>;
+					frinlifo-mask = <0xff 0x0 0x0 0x0>;
+					frinlifo-shared-mode = <0 0 0 0>;
+					vfbid-sw-offset = <0x0 0x0 0x0 0x0>;
+					vfbid-sw-entries = <0x0 0x0 0x0 0x0>;
+					ucore-mask = <0x2 0x0 0x0 0x0>;
+					sram-off = <0x0 0x0 0x0 0x0>;
+				};
+			};
+		};
+	};
+
+	soc {
+		#address-cells = <0x2>;
+		#size-cells = <0x1>;
+		compatible = "simple-bus";
+		ranges = <0x0 0x0 0x0 0x18000000 0x4000000 0x1 0x0 0x0 0x16000000 0x2000000>;
+
+		serial@30000 {
+			device_type = "serial";
+			compatible = "ns16550";
+			reg = <0x0 0x112100 0x20>;
+			reg-shift = <0x2>;
+			reg-io-width = <0x4>;
+			clock-frequency = <0x7735940>;
+			interrupt-parent = <0x1>;
+			interrupts = <0x11>;
+		};
+
+		serial@31000 {
+			device_type = "serial";
+			compatible = "ns16550";
+			reg = <0x0 0x112120 0x20>;
+			reg-shift = <0x2>;
+			reg-io-width = <0x4>;
+			clock-frequency = <0x7735940>;
+			interrupt-parent = <0x1>;
+			interrupts = <0x12>;
+		};
+
+		pic@11000 {
+			interrupt-controller;
+			#address-cells = <0x0>;
+			#interrupt-cells = <0x1>;
+			reg = <0x0 0x110000 0x200>;
+			linux,phandle = <0x1>;
+			phandle = <0x1>;
+		};
+
+		/include/ "generic-soc-fmn-ver1.dtsi"
+		/include/ "generic-soc-net-ver1.dtsi"
+	};
+
+	chosen {
+		bootargs = "console=ttyS0,115200 rdinit=/sbin/init";
+		netlogic,cpumask = <0xffffffff>;
+	};
+};
-- 
1.7.1

