From 38237c3c42ad1710dc905aaea50974fcdbe036cf Mon Sep 17 00:00:00 2001
From: Tanmay Jagdale <tanmayj@broadcom.com>
Date: Thu, 4 Dec 2014 15:22:31 +0530
Subject: xlp-config; set freein-fifo mask for multi-nod

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/boot/dts/xlp-config.dts b/arch/mips/boot/dts/xlp-config.dts
index 7e5562a..ecb1a6b 100644
--- a/arch/mips/boot/dts/xlp-config.dts
+++ b/arch/mips/boot/dts/xlp-config.dts
@@ -33,7 +33,7 @@
 			};
 
 			nae {
-				freein-fifo-mask = <0x3ffff 0x0 0 0>;
+				freein-fifo-mask = <0x3ffff 0x3ffff 0 0>; /* Node-0 Node-1 Node-2 Node-3 */
 				/* node offsets and size */
 				vfbtbl-sw-offset = <0 0 0 0>;
 				vfbtbl-sw-nentries = <32 32 32 32>;
diff --git a/arch/mips/boot/dts/xlp9xx-config.dts b/arch/mips/boot/dts/xlp9xx-config.dts
index f02c898..4399aa4 100644
--- a/arch/mips/boot/dts/xlp9xx-config.dts
+++ b/arch/mips/boot/dts/xlp9xx-config.dts
@@ -39,7 +39,7 @@
 			};
 
 			nae-0 {
-				freein-fifo-mask = <0xffff 0x0 0 0>;
+				freein-fifo-mask = <0xffff 0xffff 0 0>; /* Node-0  Node-1 Node-2 Node-3 */
 				/* node offsets and size */
 				vfbtbl-sw-offset = <0 0 0 0>;
 				vfbtbl-sw-nentries = <80 32 32 80>;
@@ -48,7 +48,7 @@
 			};
 
 	    		nae-1 {
-				freein-fifo-mask = <0xffff 0x0 0 0>;
+				freein-fifo-mask = <0xffff 0xffff 0 0>; /* Node-0  Node-1 Node-2 Node-3 */
 				/* node offsets and size */
 				vfbtbl-sw-offset = <0 0 0 0>;
 				vfbtbl-sw-nentries = <80 80 80 80>;
-- 
1.7.1

