From 4dd42670c2bc3bc207ac00d45c4d6ced6d452a34 Mon Sep 17 00:00:00 2001
From: Saswat Kumar Dash <saswat.dash@broadcom.com>
Date: Sun, 7 Dec 2014 21:18:38 -0800
Subject: multinode: added netlogic,spill-mem under chosen/ section

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/boot/dts/generic-fmn.dtsi b/arch/mips/boot/dts/generic-fmn.dtsi
index 94baf8e..b7f282a 100644
--- a/arch/mips/boot/dts/generic-fmn.dtsi
+++ b/arch/mips/boot/dts/generic-fmn.dtsi
@@ -46,7 +46,7 @@ fmn@node-1 {
 	default-credits = <50>;
 	default-queue-size = <28672>;
 	fmn-spill-mem-base = <0x0 0x0>; /* 0 - dynamic allocation */
-	fmn-spill-mem-size = <0x00000000 0x02000000>;
+	fmn-spill-mem-size = <0x0 0x0>; /* set the size in netlogic,spill-mem under chosen/ section */
 
 	q-config {
 	};
diff --git a/arch/mips/boot/dts/generic-soc-fmn-ver1.dtsi b/arch/mips/boot/dts/generic-soc-fmn-ver1.dtsi
index 9ad5257..e73751c 100644
--- a/arch/mips/boot/dts/generic-soc-fmn-ver1.dtsi
+++ b/arch/mips/boot/dts/generic-soc-fmn-ver1.dtsi
@@ -114,7 +114,7 @@ fmn@node-1 {
 	default-credits = <16>;
 	default-queue-size = <28672>;
 	fmn-spill-mem-base = <0 0>; /*0 - dynamic allocation */
-	fmn-spill-mem-size = <0x0 0x02000000>;
+	fmn-spill-mem-size = <0x0 0x0>; /* set the size in netlogic,spill-mem under chosen/ section */
 	q-config {
 		/*
 		 * Supported station names :
diff --git a/arch/mips/boot/dts/xlp2xx-ep-linux-kernel.dts b/arch/mips/boot/dts/xlp2xx-ep-linux-kernel.dts
index 8d8b62d..afeddac 100644
--- a/arch/mips/boot/dts/xlp2xx-ep-linux-kernel.dts
+++ b/arch/mips/boot/dts/xlp2xx-ep-linux-kernel.dts
@@ -260,5 +260,6 @@
 		bootargs = "console=ttyS0,115200 rhash_entries=1048576";
 		netlogic,cpumask = <0xff>;
 		netlogic,reserve-mem = <256>;
+		netlogic,spill-mem = <32>; /* max spill size : 512 MB*/
 	};
 };
diff --git a/arch/mips/boot/dts/xlp2xx-xmc-linux-kernel.dts b/arch/mips/boot/dts/xlp2xx-xmc-linux-kernel.dts
index 802f266..355230f 100644
--- a/arch/mips/boot/dts/xlp2xx-xmc-linux-kernel.dts
+++ b/arch/mips/boot/dts/xlp2xx-xmc-linux-kernel.dts
@@ -140,5 +140,6 @@
 		bootargs = "console=ttyS0,115200 rhash_entries=1048576";
 		netlogic,cpumask = <0xff>;
 		netlogic,reserve-mem = <128>;
+		netlogic,spill-mem = <32>; /* max spill size : 512 MB*/
 	};
 };
diff --git a/arch/mips/boot/dts/xlp3xx-xmc-linux-kernel.dts b/arch/mips/boot/dts/xlp3xx-xmc-linux-kernel.dts
index 9230389..46e7fdf 100644
--- a/arch/mips/boot/dts/xlp3xx-xmc-linux-kernel.dts
+++ b/arch/mips/boot/dts/xlp3xx-xmc-linux-kernel.dts
@@ -155,5 +155,6 @@
 		bootargs = "console=ttyS0,115200 rhash_entries=1048576";
 		netlogic,cpumask = <0xffff>;
 		netlogic,reserve-mem = <128>;
+		netlogic,spill-mem = <32>; /* max spill size : 512 MB*/
 	};
 };
diff --git a/arch/mips/boot/dts/xlp5xx-pcie-linux-kernel.dts b/arch/mips/boot/dts/xlp5xx-pcie-linux-kernel.dts
index db489fe..48df268 100644
--- a/arch/mips/boot/dts/xlp5xx-pcie-linux-kernel.dts
+++ b/arch/mips/boot/dts/xlp5xx-pcie-linux-kernel.dts
@@ -214,5 +214,6 @@
 	chosen {
 		bootargs = "console=ttyS0,1152 rhash_entries=1048576";
 		netlogic,reserve-mem = <256>;
+		netlogic,spill-mem = <32>; /* max spill size : 512 MB*/
 	};
 };
diff --git a/arch/mips/boot/dts/xlp8xx-linux-mn-kernel.dts b/arch/mips/boot/dts/xlp8xx-linux-mn-kernel.dts
index 555218b..8dcb944 100644
--- a/arch/mips/boot/dts/xlp8xx-linux-mn-kernel.dts
+++ b/arch/mips/boot/dts/xlp8xx-linux-mn-kernel.dts
@@ -202,5 +202,6 @@
 		netlogic.cpumask = <0xffffffff>;
 		netlogic,reserve-mem = <256>;
 		netlogic,frin-fifo-mem = <128>;
+		netlogic,spill-mem = <32>; /* max spill size : 512 MB*/
 	};
 };
diff --git a/arch/mips/boot/dts/xlp8xx-linux-mn4-kernel.dts b/arch/mips/boot/dts/xlp8xx-linux-mn4-kernel.dts
index f58029c..d251354 100644
--- a/arch/mips/boot/dts/xlp8xx-linux-mn4-kernel.dts
+++ b/arch/mips/boot/dts/xlp8xx-linux-mn4-kernel.dts
@@ -226,5 +226,6 @@
 		netlogic.cpumask = <0xffffffff>;
 		netlogic,reserve-mem = <256>;
 		netlogic,frin-fifo-mem = <128>;
+		netlogic,spill-mem = <32>; /* max spill size : 512 MB*/
 	};
 };
diff --git a/arch/mips/boot/dts/xlp9xx-linux-mn-kernel.dts b/arch/mips/boot/dts/xlp9xx-linux-mn-kernel.dts
index 93add7f..6c7d5b0 100644
--- a/arch/mips/boot/dts/xlp9xx-linux-mn-kernel.dts
+++ b/arch/mips/boot/dts/xlp9xx-linux-mn-kernel.dts
@@ -220,5 +220,6 @@
 		netlogic,cpumask = <0x0 0x0 0xffffffff 0xffffffff
 					0x0 0x0 0xffffffff 0xffffffff>;
 		netlogic,reserve-mem = <256 256>;
+		netlogic,spill-mem = <32>; /* max spill size : 512 MB*/
 	};
 };
-- 
1.7.1

