From 76b9faa9702ca917bab9430e4cfa3321eeb1e14d Mon Sep 17 00:00:00 2001
From: Saswat Kumar Dash <saswat.dash@broadcom.com>
Date: Tue, 9 Dec 2014 22:59:31 -0800
Subject: guest: increased guest memory to 3GB

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/boot/dts/xlp9xx-guest.dts b/arch/mips/boot/dts/xlp9xx-guest.dts
index 3c51fe4..10ac35b 100644
--- a/arch/mips/boot/dts/xlp9xx-guest.dts
+++ b/arch/mips/boot/dts/xlp9xx-guest.dts
@@ -12,7 +12,7 @@
 
 	memory {
 		device_type = "memory";
-		reg = <0x0 0x100000 0x0 0xff00000 0x0 0x20000000 0x0 0xa0000000>;
+		reg = <0x0 0x100000 0x0 0xff00000 0x0 0x20000000 0x0 0xa0000000 0x0 0xe0000000 0x0 0x20000000>;
 	};
 
 	pktmem {
-- 
1.7.1

