From 7391a92f359d76ca4c647d81de7891078ae5feae Mon Sep 17 00:00:00 2001
From: Hareesh R <hareeshr@broadcom.com>
Date: Wed, 10 Dec 2014 18:41:36 +0530
Subject: Openssl : Fix for response missing from sae engine

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/boot/dts/generic-soc-fmn-ver1.dtsi b/arch/mips/boot/dts/generic-soc-fmn-ver1.dtsi
index e73751c..148a6e7 100644
--- a/arch/mips/boot/dts/generic-soc-fmn-ver1.dtsi
+++ b/arch/mips/boot/dts/generic-soc-fmn-ver1.dtsi
@@ -32,8 +32,8 @@ fmn@node-0 {
 		};
 
 		crypto {
-			onchip-entries = <256>;
-			num-sae-queues = <1>;
+			onchip-entries = <64>;
+			/* num-sae-queues = <1>; */
 			queue-size = <49152>;
 			credits-from {
 				stn@node-0 {
-- 
1.7.1

