From 45b95d8493d7bc1d23860e29fe4c09dc38a79711 Mon Sep 17 00:00:00 2001
From: Kamlakant Patel <kamlakant.patel@broadcom.com>
Date: Thu, 11 Dec 2014 15:46:44 +0530
Subject: multinode: enable node 1 spi/flash devices

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/boot/dts/xlp8xx-linux-mn-kernel.dts b/arch/mips/boot/dts/xlp8xx-linux-mn-kernel.dts
index 69b1f0e..c115972 100644
--- a/arch/mips/boot/dts/xlp8xx-linux-mn-kernel.dts
+++ b/arch/mips/boot/dts/xlp8xx-linux-mn-kernel.dts
@@ -177,7 +177,7 @@
 			};
 		};
 
-		spi: xlp_spi@3a000 {
+		spi0: xlp_spi@3a000 {
 			compatible = "netlogic,xlp-spi";
 			#address-cells = <1>;
 			#size-cells = <0>;
@@ -282,16 +282,96 @@
 			interrupt-parent = <&pic1>;
 			interrupts = <18>;
 		};
-		gpio1: xlp_gpio@74000 {
-			compatible = "netlogic,xlp-gpio";
-			reg = <0 0x74000 0x1000>;
-			#gpio-cells = <2>;
-			gpio-controller;
 
-			#interrupt-cells = <1>;
+		nand_flash1@79000 {
+			compatible = "netlogic,xlp-nand";
+			#address-cells = <1>;
+			#size-cells = <0>;
+			reg = <0 0x79000 0x1000>;
+
+			interrupts = <37>;
 			interrupt-parent = <&pic1>;
-			interrupts = <39>;
-			interrupt-controller;
+			nand-ecc-mode = "hw";
+
+			nand_cs1 {
+				#address-cells = <1>;
+				#size-cells = <1>;
+				reg = <1>;
+
+				partition@0 {
+					label = "firmware";
+					reg = <0x0 0x20000000>;
+				};
+
+				partition@1 {
+					label = "kernel";
+					reg = <0x20000000 0x20000000>;
+				};
+
+				partition@2 {
+					label = "user";
+					reg = <0x40000000 0x0>;
+				};
+			};
+		};
+
+		spi1: xlp_spi@7a000 {
+			compatible = "netlogic,xlp-spi";
+			#address-cells = <1>;
+			#size-cells = <0>;
+			reg = <0 0x7a000 0x100>;
+
+			interrupts = <34>;
+			interrupt-parent = <&pic1>;
+
+			spi_nor@1 {
+				compatible = "spansion,s25sl12801";
+				#address-cells = <1>;
+				#size-cells = <1>;
+
+				reg = <1>;		/* Chip Select */
+				spi-max-frequency = <40000000>;
+
+				partition@0 {
+					label = "x-loader";
+					reg = <0x0 0x200000>; /* 2M */
+					read-only;
+				};
+
+				partition@200000 {
+					label = "kernel";
+					reg = <0x200000 0x500000>; /* 5M */
+				};
+
+				partition@700000 {
+					label = "rootfs";
+					reg = <0x700000 0x800000>; /* 8M */
+				};
+
+				partition@f00000 {
+					label = "env";
+					reg = <0xf00000 0x100000>; /* 1M */
+					read-only;
+				};
+			};
+
+			spi_nand@2 {
+				compatible = "spinand,mt29f";
+				#address-cells = <1>;
+				#size-cells = <1>;
+				reg = <2>;	/* SPI NAND CS */
+				spi-max-frequency = <50000000>;
+
+				partition@0 {
+					label = "firmware";
+					reg = <0x0 0x1c00000>;
+				};
+
+				partition@1 {
+					label = "kernel";
+					reg = <0x1c00000 0x0>;
+				};
+			};
 		};
 	};
 
diff --git a/arch/mips/boot/dts/xlp9xx-linux-mn-kernel.dts b/arch/mips/boot/dts/xlp9xx-linux-mn-kernel.dts
index 81415ea..bde05ac 100644
--- a/arch/mips/boot/dts/xlp9xx-linux-mn-kernel.dts
+++ b/arch/mips/boot/dts/xlp9xx-linux-mn-kernel.dts
@@ -105,7 +105,7 @@
 			interrupt-parent = <&pic0>;
 		};
 
-		spi: xlp_spi@13a000 {
+		spi0: xlp_spi@13a000 {
 			compatible = "netlogic,xlp-spi";
 			#address-cells = <1>;
 			#size-cells = <0>;
@@ -228,7 +228,7 @@
 			};
 		};
 
-		gpio: xlp_gpio@114000 {
+		gpio0: xlp_gpio@114000 {
 			compatible = "netlogic,xlp-gpio";
 			reg = <0 0x114100 0x1000>;
 			#gpio-cells = <2>;
@@ -240,6 +240,7 @@
 			interrupt-controller;
 		};
 
+		/* Node 1 devices */
 		pic1: pic1@610000 {
 			compatible = "netlogic,xlp-pic";
 			interrupt-controller;
@@ -301,6 +302,94 @@
 			interrupt-parent = <&pic1>;
 		};
 
+		nand_flash1@639000 {
+			compatible = "brcm,xlp9xx-nand";
+			#address-cells = <1>;
+			#size-cells = <0>;
+			reg = <0 0x639000 0x1000>;
+
+			nand-ecc-mode = "hw";
+
+			nand_cs1 {
+				#address-cells = <1>;
+				#size-cells = <1>;
+				reg = <1>;
+
+				partition@0 {
+					label = "firmware";
+					reg = <0x0 0x20000000>;
+				};
+
+				partition@1 {
+					label = "kernel";
+					reg = <0x20000000 0x20000000>;
+				};
+
+				partition@2 {
+					label = "user";
+					reg = <0x40000000 0x0>;
+				};
+			};
+		};
+
+		spi1: xlp_spi@63a000 {
+			compatible = "netlogic,xlp-spi";
+			#address-cells = <1>;
+			#size-cells = <0>;
+			reg = <0 0x63a000 0x100>;
+
+			interrupts = <34>;
+			interrupt-parent = <&pic1>;
+
+			spi_nor@1 {
+				compatible = "spansion,s25sl12801";
+				#address-cells = <1>;
+				#size-cells = <1>;
+
+				reg = <1>;		/* Chip Select */
+
+				spi-max-frequency = <40000000>;
+
+				partition@0 {
+					label = "x-loader";
+					reg = <0x0 0x200000>; /* 2M */
+					read-only;
+				};
+
+				partition@200000 {
+					label = "kernel";
+					reg = <0x200000 0x500000>; /* 5M */
+				};
+
+				partition@700000 {
+					label = "rootfs";
+					reg = <0x700000 0x800000>; /* 8M */
+				};
+
+				partition@f00000 {
+					label = "env";
+					reg = <0xf00000 0x100000>; /* 1M */
+					read-only;
+				};
+			};
+			spi_nand@2 {
+				compatible = "spinand,mt29f";
+				#address-cells = <1>;
+				#size-cells = <1>;
+				reg = <2>;	/* SPI NAND CS */
+				spi-max-frequency = <50000000>;
+
+				partition@0 {
+					label = "firmware";
+					reg = <0x0 0x1c00000>;
+				};
+
+				partition@1 {
+					label = "kernel";
+					reg = <0x1c00000 0x0>;
+				};
+			};
+		};
 	};
 
 	chosen {
-- 
1.7.1

