From 76050d33aee5a2cc58cfa6a354d2c20a35a2909e Mon Sep 17 00:00:00 2001
From: Tanmay Jagdale <tanmayj@broadcom.com>
Date: Wed, 17 Dec 2014 11:22:33 +0530
Subject: xlp5xx-pcie: fix console speed

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/boot/dts/xlp5xx-pcie-linux-kernel.dts b/arch/mips/boot/dts/xlp5xx-pcie-linux-kernel.dts
index 568f7cd..029736c 100644
--- a/arch/mips/boot/dts/xlp5xx-pcie-linux-kernel.dts
+++ b/arch/mips/boot/dts/xlp5xx-pcie-linux-kernel.dts
@@ -213,7 +213,7 @@
 	};
 
 	chosen {
-		bootargs = "console=ttyS0,1152 rhash_entries=1048576";
+		bootargs = "console=ttyS0,115200 rhash_entries=1048576";
 		netlogic,reserve-mem = <256>;
 		netlogic,spill-mem = <32>; /* max spill size : 512 MB*/
 	};
-- 
1.7.1

