From 8a78003a0766f2dd17dacdc8ef21e9ed825d0043 Mon Sep 17 00:00:00 2001
From: Saswat Kumar Dash <saswat.dash@broadcom.com>
Date: Tue, 16 Dec 2014 22:35:33 -0800
Subject: legacy: added netlogic,fmn-spill under chosen/ section

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/boot/dts/xlp2xx-pcie-linux-kernel.dts b/arch/mips/boot/dts/xlp2xx-pcie-linux-kernel.dts
index 36be059..065e2e3 100644
--- a/arch/mips/boot/dts/xlp2xx-pcie-linux-kernel.dts
+++ b/arch/mips/boot/dts/xlp2xx-pcie-linux-kernel.dts
@@ -167,5 +167,6 @@
 		bootargs = "console=ttyS0,115200 rhash_entries=1048576";
 		netlogic,cpumask = <0xff>;
 		netlogic,reserve-mem = <256>;
+		netlogic,spill-mem = <32>;
 	};
 };
-- 
1.7.1

