From 0a6be74f10aa64310014e685133ba8de3bef5f06 Mon Sep 17 00:00:00 2001
From: Zi Shen Lim <zlim@netlogicmicro.com>
Date: Mon, 12 Jul 2010 20:22:39 -0700
Subject: Remove DTC warnings

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/dts/xlp.dts b/dts/xlp.dts
index 704a4b0..e20e90f 100644
--- a/dts/xlp.dts
+++ b/dts/xlp.dts
@@ -8,15 +8,15 @@
 	compatible = "NETL,XLP8XX_A0";
 	#address-cells = <1>;
 	#size-cells = <1>;
-
 	doms {
 		#address-cells = <1>;
 		#size-cells = <1>;
-
 		dom@0 {
 			device_type = "domain";
 			heap-size = <0x0 0x4000000>;
 			shared-mem = <0x0 0x80000000 0x0 0x20000000>;
+			#address-cells = <1>;
+			#size-cells = <1>;
 			cpu {
 				onlinemask = <0xff>;
 			};
-- 
1.7.1

