From 9b286959dd153ba8fd605b2bbec67b76a8e9e8a3 Mon Sep 17 00:00:00 2001
From: Siva Pochiraju <sivap@netlogicmicro.com>
Date: Fri, 16 Jul 2010 16:27:53 -0700
Subject: Fixed compiler warning

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/libfdt/fdt.c b/libfdt/fdt.c
index 2acaec5..bbbda5c 100644
--- a/libfdt/fdt.c
+++ b/libfdt/fdt.c
@@ -79,7 +79,7 @@ const void *fdt_offset_ptr(const void *fdt, int offset, int len)
 	const char *p;
 
 	if (fdt_version(fdt) >= 0x11)
-		if (((offset + len) < offset)
+		if ((((unsigned int) offset + (unsigned int) len) < (unsigned int) offset)
 		    || ((offset + len) > fdt_size_dt_struct(fdt)))
 			return NULL;
 
-- 
1.7.1

