From c14d5b5b49b86b80065d692c4d4c96ed151e5c30 Mon Sep 17 00:00:00 2001
From: Prasad Boddupalli <pboddupalli@netlogicmicro.com>
Date: Wed, 21 Jul 2010 11:08:31 -0700
Subject: changes to include Xen header files depending on the compilation macro

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/libfdt/libfdt_env.h b/libfdt/libfdt_env.h
index 449bf60..26c31fb 100644
--- a/libfdt/libfdt_env.h
+++ b/libfdt/libfdt_env.h
@@ -1,9 +1,14 @@
 #ifndef _LIBFDT_ENV_H
 #define _LIBFDT_ENV_H
 
+#ifdef __XEN__
+#include <xen/types.h>
+#include <xen/string.h>
+#else
 #include <stddef.h>
 #include <stdint.h>
 #include <string.h>
+#endif
 
 #define _B(n)	((unsigned long long)((uint8_t *)&x)[n])
 static inline uint32_t fdt32_to_cpu(uint32_t x)
-- 
1.7.1

