From 7f64d016514f9602ea96df06d5b59df0f80ad97e Mon Sep 17 00:00:00 2001
From: Siva Pochiraju <sivap@netlogicmicro.com>
Date: Sat, 7 Aug 2010 13:57:13 -0700
Subject: Added rx_vc and fb_vc for cpu

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/dts/xlp.dts b/dts/xlp.dts
index e20e90f..1dae99c 100644
--- a/dts/xlp.dts
+++ b/dts/xlp.dts
@@ -19,6 +19,8 @@
 			#size-cells = <1>;
 			cpu {
 				onlinemask = <0xff>;
+				nae-rx-vc = <0>;
+				nae-fb-vc = <3>;
 			};
 			uart {
 				id = <0>;
-- 
1.7.1

