From b2c60f3a8ff0e0cbe58dc0990fc97687a980fb1a Mon Sep 17 00:00:00 2001
From: Krishnamurthy D V <kmurthy@netlogicmicro.com>
Date: Thu, 5 Aug 2010 19:31:16 +0530
Subject: FDT config file for XLR

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/dts/xlr.dts b/dts/xlr.dts
new file mode 100644
index 0000000..8c21202
--- /dev/null
+++ b/dts/xlr.dts
@@ -0,0 +1,318 @@
+/* XLR,XLS Device Tree Source
+ *
+ */
+
+/dts-v1/;
+/ {
+	model = "MIPS,XLR(S)";
+	compatible = "NETL,XLR";
+	#address-cells = <1>;
+	#size-cells = <1>;
+	doms {
+		#address-cells = <1>;
+		#size-cells = <1>;
+		dom@0 {
+			device_type = "domain";
+			heap-size = <0x0 0x4000000>;
+			shared-mem = <0x0 0x80000000 0x0 0x20000000>;
+			#address-cells = <1>;
+			#size-cells = <1>;
+			cpu {
+				onlinemask = <0xffffffff>;
+			};
+			uart {
+				id = <0>;
+			};
+			memory {
+				reg = <0x00100000 0x0FF00000	// 255M at 1M
+					   0x20000000 0x20000000>;  // 512M at 512M
+			};
+			pic {
+			};
+		};
+		dom@1 {
+			device_type = "domain";
+			cpu {
+				onlinemask = <0xffff0000>;
+			};
+			uart {
+				id = <1>;
+			};
+			memory {
+			};
+		};
+		dom@2 {
+			device_type = "domain";
+			cpu {
+			};
+			uart {
+			};
+			memory {
+			};
+		};
+		dom@3 {
+			device_type = "domain";
+			cpu {
+			};
+			uart {
+			};
+			memory {
+			};
+		};
+		dom@4 {
+			device_type = "domain";
+			cpu {
+			};
+			uart {
+			};
+			memory {
+			};
+		};
+		dom@5 {
+			device_type = "domain";
+			cpu {
+			};
+			uart {
+			};
+			memory {
+			};
+		};
+		dom@6 {
+			device_type = "domain";
+			cpu {
+			};
+			uart {
+			};
+			memory {
+			};
+		};
+		dom@7 {
+			device_type = "domain";
+			cpu {
+			};
+			uart {
+			};
+			memory {
+			};
+		};
+		dom@8 {
+			device_type = "domain";
+			cpu {
+			};
+			uart {
+			};
+			memory {
+			};
+		};
+		dom@9 {
+			device_type = "domain";
+			cpu {
+			};
+			uart {
+			};
+			memory {
+			};
+		};
+		dom@10 {
+			device_type = "domain";
+			cpu {
+			};
+			uart {
+			};
+			memory {
+			};
+		};
+		dom@11 {
+			device_type = "domain";
+			cpu {
+			};
+			uart {
+			};
+			memory {
+			};
+		};
+		dom@12 {
+			device_type = "domain";
+			cpu {
+			};
+			uart {
+			};
+			memory {
+			};
+		};
+		dom@13 {
+			device_type = "domain";
+			cpu {
+			};
+			uart {
+			};
+			memory {
+			};
+		};
+		dom@14 {
+			device_type = "domain";
+			cpu {
+			};
+			uart {
+			};
+			memory {
+			};
+		};
+		dom@15 {
+			device_type = "domain";
+			cpu {
+			};
+			uart {
+			};
+			memory {
+			};
+		};
+		dom@16 {
+			device_type = "domain";
+			cpu {
+			};
+			uart {
+			};
+			memory {
+			};
+		};
+		dom@17 {
+			device_type = "domain";
+			cpu {
+			};
+			uart {
+			};
+			memory {
+			};
+		};
+		dom@18 {
+			device_type = "domain";
+			cpu {
+			};
+			uart {
+			};
+			memory {
+			};
+		};
+		dom@19 {
+			device_type = "domain";
+			cpu {
+			};
+			uart {
+			};
+			memory {
+			};
+		};
+		dom@20 {
+			device_type = "domain";
+			cpu {
+			};
+			uart {
+			};
+			memory {
+			};
+		};
+		dom@21 {
+			device_type = "domain";
+			cpu {
+			};
+			uart {
+			};
+			memory {
+			};
+		};
+		dom@23 {
+			device_type = "domain";
+			cpu {
+			};
+			uart {
+			};
+			memory {
+			};
+		};
+		dom@24 {
+			device_type = "domain";
+			cpu {
+			};
+			uart {
+			};
+			memory {
+			};
+		};
+		dom@25 {
+			device_type = "domain";
+			cpu {
+			};
+			uart {
+			};
+			memory {
+			};
+		};
+		dom@26 {
+			device_type = "domain";
+			cpu {
+			};
+			uart {
+			};
+			memory {
+			};
+		};
+		dom@27 {
+			device_type = "domain";
+			cpu {
+			};
+			uart {
+			};
+			memory {
+			};
+		};
+		dom@28 {
+			device_type = "domain";
+			cpu {
+			};
+			uart {
+			};
+			memory {
+			};
+		};
+		dom@29 {
+			device_type = "domain";
+			cpu {
+			};
+			uart {
+			};
+			memory {
+			};
+		};
+		dom@30 {
+			device_type = "domain";
+			cpu {
+			};
+			uart {
+			};
+			memory {
+			};
+		};
+		dom@31 {
+			device_type = "domain";
+			cpu {
+			};
+			uart {
+			};
+			memory {
+			};
+		};
+	};
+
+	chosen {
+		bootargs = "console=ttyS0,115200 rdinit=/sbin/init";
+	};
+
+	system {
+		boardmajor = <2>;
+		boardminor = <1>;
+		macaddress = [ 00 0f 30 00 09 6a ];
+		cpufrequency = <1190000000>;
+		globalshmem = <0x800ff000>;
+	};
+};
-- 
1.7.1

