From 192333515766e4f19a033454e09c656f76b2e91e Mon Sep 17 00:00:00 2001
From: Krishnamurthy D V <kmurthy@netlogicmicro.com>
Date: Thu, 2 Sep 2010 19:21:27 +0530
Subject: Adding fields required for hyperexec XLR apps

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/dts/xlr.dts b/dts/xlr.dts
index 8c21202..34a227a 100644
--- a/dts/xlr.dts
+++ b/dts/xlr.dts
@@ -306,9 +306,11 @@
 
 	chosen {
 		bootargs = "console=ttyS0,115200 rdinit=/sbin/init";
+		bootenv = "board=arizona"
 	};
 
 	system {
+		iobase = <0xbef00000>;
 		boardmajor = <2>;
 		boardminor = <1>;
 		macaddress = [ 00 0f 30 00 09 6a ];
-- 
1.7.1

