From 6a0b861990007772f26caddd8a373d5541e2bdb6 Mon Sep 17 00:00:00 2001
From: Venu Vadapalli <vvadapalli@netlogicmicro.com>
Date: Wed, 4 Aug 2010 00:54:40 -0700
Subject: Added memory chunks, labels and sharedcfg nodes for devices

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/dts/sysconfig.dts b/dts/sysconfig.dts
index 29b4fc5..d54336c 100644
--- a/dts/sysconfig.dts
+++ b/dts/sysconfig.dts
@@ -1,4 +1,4 @@
-/* XLP8XX Device Tree Source 
+/* XLP8XX Device Tree Source
  *
  */
 
@@ -11,15 +11,23 @@
 	#size-cells = <1>;
 
 	hypervisor {
+
 		hypervisor-name = "Xen";
+
 		cpu_possible_map = <0xff>;
+
 		is_dom0_hvm = <1>;
+
 		bootargs = "dom0_loadaddr=0x09000000 dom0_cpumask=0x0f -- ";
 
 		uart {
 			/* Shared ports, not owner */
 			port@0 {};
 		};
+
+		memory {
+			map = <&mem_hypervisor>;
+		};
 	};
 
 	doms {
@@ -29,6 +37,8 @@
 			device_type = "domain";
 			os = "linux";
 
+			bootargs = " xlp_noi2c console=ttyS0,115200 rdinit=/sbin/init";
+
 			#address-cells = <1>;
 			#size-cells = <1>;
 
@@ -43,8 +53,7 @@
 			};
 
 			memory {
-				reg = <0x04000000 0x0C000000	// 192M at 64M
-					0x20000000 0x20000000>;  // 512M at 512M
+				map = <&mem_dom0_lowmem &mem_dom0_highmem>;
 			};
 
 			pic { };
@@ -67,7 +76,7 @@
 			os = "netos";
 
 			heap-size = <0x0 0x4000000>;
-			shared-mem = <0x0 0x80000000 0x0 0x20000000>;
+			shared-mem = <&mem_shared>;
 
 			#address-cells = <1>;
 			#size-cells = <1>;
@@ -81,7 +90,7 @@
 				port@1 {};
 			};
 			memory {
-				reg = <0x30000000 0x10000000>;  // 256M @ 768M
+				map = <&mem_dom1>;
 			};
 			nae@mode0 {
 				/* owner, could be shared - read sharedcfg in sysconfig */
@@ -98,24 +107,54 @@
 		};
 	};
 
-	chosen {
-		bootargs = " xlp_noi2c console=ttyS0,115200 rdinit=/sbin/init";
-	};
-
 	sysconfig {
 
 		uart {
 			port@0 {
 				baud-rate = <115200>;
-				sharedcfg {
-					memory = <0x1F000000 0x00400000>;  // 4M @ 496M
-				};
+				sharedcfg { memory = <&mem_uart0>; };
 			};
 			port@1 {
 				baud-rate = <115200>;
-				sharedcfg {
-					memory = <0x1F400000 0x00400000>;  // 4M @ 500M
-				};
+				sharedcfg { memory = <&mem_uart1>; };
+			};
+		};
+
+		memory {
+			#address-cells = <1>;
+			#size-cells = <1>;
+
+			mem_expvec: chunk@0 {
+				range = <0x00000000 0x00100000>;	// 1M @ 0M
+				sharedcfg {};
+			};
+			mem_hypervisor: chunk@1 {
+				range = <0x00100000 0x03f000000>;	// 63M @ 1M
+				sharedcfg {};
+			};
+			mem_dom0_lowmem: chunk@2 {
+				range = <0x04000000 0x0C000000>;	// 192M at 64M
+				sharedcfg {};
+			};
+			mem_uart0: chunk@3 {
+				range = <0x1F000000 0x00400000>;  	// 4M @ 496M
+				sharedcfg {};
+			};
+			mem_uart1: chunk@4 {
+				range = <0x1F400000 0x00400000>;  	// 4M @ 500M
+				sharedcfg {};
+			};
+			mem_dom0_highmem: chunk@5 {
+				range = <0x20000000 0x20000000>;  	// 512M at 512M
+				sharedcfg {};
+			};
+			mem_dom1: chunk@6 {
+				range = <0x30000000 0x10000000>;  	// 256M @ 768M
+				sharedcfg {};
+			};
+			mem_shared: chunk@7 {
+				shared-mem = <0x0 0x80000000 0x0 0x20000000>; // 512M @ 2G
+				sharedcfg {};
 			};
 		};
 
@@ -127,7 +166,6 @@
                         start-port-id = <0>;
                         num-port-regs = <4>;
 			port@0 {
-				sharedcfg { };
 
 	                        regs = <
         	                        0x35 0xd
@@ -135,9 +173,10 @@
                         	        0x8a 0xd
                                 	0x8a 0xd
 	                        >;
+
+				sharedcfg { };
 			};
                         port@1 {
-				sharedcfg { };
 
 				regs = <
 	                                0x35 0x35
@@ -145,9 +184,10 @@
                 	                0x8a 0x35
                         	        0x8a 0x1f
 	                        >;
+
+				sharedcfg { };
 			};
                         port@2 {
-				sharedcfg { };
 
 				regs = <
 	                                0x35 0x8a
@@ -155,9 +195,10 @@
                 	                0x8a 0x8
                         	        0x8a 0x8
 	                        >;
+
+				sharedcfg { };
 			};
                         port@3 {
-				sharedcfg { };
 
 				regs = <
 	                                0x35 0x14
@@ -165,6 +206,8 @@
                 	                0x8a 0x14
                         	        0x8a 0x14
 	                        >;
+
+				sharedcfg { };
 			};
 		};
 		nae@mode1 {
@@ -175,7 +218,6 @@
                         start-port-id = <0>;
                         num-port-regs = <2>;
 			port@0 {
-				sharedcfg { };
 
 	                        regs = <
         	                        0x35 0xd
@@ -183,9 +225,11 @@
                         	        0x8a 0xd
                                 	0x8a 0xd
 	                        >;
+
+				sharedcfg { };
+
 			};
                         port@1 {
-				sharedcfg { };
 
 				regs = <
 	                                0x35 0x35
@@ -193,6 +237,8 @@
                 	                0x8a 0x35
                         	        0x8a 0x1f
 	                        >;
+
+				sharedcfg { };
 			};
 		};
 
-- 
1.7.1

