From 059f8cc160528c19b3e850d7e1fee3acbd8c3f20 Mon Sep 17 00:00:00 2001
From: Venu Vadapalli <vvadapalli@netlogicmicro.com>
Date: Wed, 4 Aug 2010 01:26:04 -0700
Subject: Removed doms node, defined cpu sets

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/dts/sysconfig.dts b/dts/sysconfig.dts
index d54336c..d430f3e 100644
--- a/dts/sysconfig.dts
+++ b/dts/sysconfig.dts
@@ -30,84 +30,90 @@
 		};
 	};
 
-	doms {
+	dom@0 {
+		os = "linux";
+
+		bootargs = " xlp_noi2c console=ttyS0,115200 rdinit=/sbin/init";
+
 		#address-cells = <1>;
 		#size-cells = <1>;
-		dom@0 {
-			device_type = "domain";
-			os = "linux";
 
-			bootargs = " xlp_noi2c console=ttyS0,115200 rdinit=/sbin/init";
+		cpu {
+			set = <&cpu_dom0>;
+		};
 
-			#address-cells = <1>;
-			#size-cells = <1>;
+		uart {
+			/* owner, could be shared - read sharedcfg in sysconfig */
+			port@0 { owner; };
+			port@1 { owner; };
+		};
 
-			cpu {
-				onlinemask = <0xf>;
-			};
+		memory {
+			map = <&mem_dom0_lowmem &mem_dom0_highmem>;
+		};
 
-			uart {
-				/* owner, could be shared - read sharedcfg in sysconfig */
-				port@0 { owner; };
-				port@1 { owner; };
-			};
+		pic { };
 
-			memory {
-				map = <&mem_dom0_lowmem &mem_dom0_highmem>;
-			};
+		nae@mode0 {
+			/* Shared ports, not owner */
+			port@0 {};
+			port@1 {};
+			port@2 {};
+			port@3 {};
+		};
+		nae@mode1 {
+			/* owner, could be shared - read sharedcfg in sysconfig */
+			port@0 { owner;	};
+			port@1 { owner;	};
+		};
+	};
 
-			pic { };
+	dom@1 {
+		os = "netos";
 
-			nae@mode0 {
-				/* Shared ports, not owner */
-				port@0 {};
-				port@1 {};
-				port@2 {};
-				port@3 {};
-			};
-			nae@mode1 {
-				/* owner, could be shared - read sharedcfg in sysconfig */
-				port@0 { owner;	};
-				port@1 { owner;	};
-			};
+		heap-size = <0x0 0x4000000>;
+		shared-mem = <&mem_shared>;
+
+		#address-cells = <1>;
+		#size-cells = <1>;
+
+		cpu {
+			set = <&cpu_dom1>;
+		};
+		uart {
+			/* Shared ports, not owner */
+			port@0 {};
+			port@1 {};
+		};
+		memory {
+			map = <&mem_dom1>;
+		};
+		nae@mode0 {
+			/* owner, could be shared - read sharedcfg in sysconfig */
+			port@0 { owner;	};
+			port@1 { owner;	};
+			port@2 { owner;	};
+			port@3 { owner;	};
+		};
+		nae@mode1 {
+			/* Shared ports, not owner */
+			port@0 {};
+			port@1 {};
 		};
-		dom@1 {
-			device_type = "domain";
-			os = "netos";
+	};
 
-			heap-size = <0x0 0x4000000>;
-			shared-mem = <&mem_shared>;
+	sysconfig {
 
-			#address-cells = <1>;
-			#size-cells = <1>;
+		cpu {
+			#mask-cells = <1>;
 
-			cpu {
-				onlinemask = <0x10>;
-			};
-			uart {
-				/* Shared ports, not owner */
-				port@0 {};
-				port@1 {};
+			cpu_dom0: set@0 {
+				mask = <0x0f>;
 			};
-			memory {
-				map = <&mem_dom1>;
-			};
-			nae@mode0 {
-				/* owner, could be shared - read sharedcfg in sysconfig */
-				port@0 { owner;	};
-				port@1 { owner;	};
-				port@2 { owner;	};
-				port@3 { owner;	};
-			};
-			nae@mode1 {
-				/* Shared ports, not owner */
-				port@0 {};
-				port@1 {};
+			cpu_dom1: set@1 {
+				mask = <0xf0>;
 			};
 		};
-	};
-
-	sysconfig {
 
 		uart {
 			port@0 {
@@ -153,6 +159,9 @@
 				sharedcfg {};
 			};
 			mem_shared: chunk@7 {
+				#address-cells = <2>;
+				#size-cells = <2>;
+
 				shared-mem = <0x0 0x80000000 0x0 0x20000000>; // 512M @ 2G
 				sharedcfg {};
 			};
-- 
1.7.1

