From 71a5a7f1dcfb9616a762342b7bad32fe7b47d7a7 Mon Sep 17 00:00:00 2001
From: Venu Vadapalli <vvadapalli@netlogicmicro.com>
Date: Wed, 4 Aug 2010 17:14:13 -0700
Subject: Added new sample configuration files

	netos.dts: standalone netos on xen
	sysconfig-xen.dts: linux+netos on xen
	sysconfig.dts: Added more nodes
[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/dts/netos.dts b/dts/netos.dts
new file mode 100644
index 0000000..12173d0
--- /dev/null
+++ b/dts/netos.dts
@@ -0,0 +1,48 @@
+/* XLP8XX Device Tree Source 
+ *
+ */
+
+/dts-v1/;
+/ {
+	model = "MIPS,XLP8XX";
+	compatible = "NETL,XLP8XX_A0";
+
+	#address-cells = <1>;
+	#size-cells = <1>;
+
+	hypervisor {
+		hypervisor-name = "Xen";
+		cpu_possible_map = <0xff>;
+		is_dom0_hvm = <0>;
+		bootargs = "dom0_loadaddr=0x09000000 dom0_cpumask=0x01 -- ";
+	};
+
+	doms {
+		#address-cells = <1>;
+		#size-cells = <1>;
+		dom@0 {
+			device_type = "domain";
+			os = "netos";
+
+			heap-size = <0x0 0x4000000>; // 64M
+			shared-mem = <0x0 0x80000000 0x0 0x20000000>; //512M @ 2GB
+
+			#address-cells = <1>;
+			#size-cells = <1>;
+
+			cpu {
+				onlinemask = <0x01>;
+			};
+			uart {
+				id = <0>;
+			};
+			memory {
+				reg = <0x30000000 0x10000000>;  // 256M @ 768M
+			};
+		};
+	};
+
+	chosen {
+		bootargs = " xlp_noi2c console=ttyS0,115200 rdinit=/sbin/init";
+	};
+};
diff --git a/dts/sysconfig-xen.dts b/dts/sysconfig-xen.dts
new file mode 100644
index 0000000..25a0810
--- /dev/null
+++ b/dts/sysconfig-xen.dts
@@ -0,0 +1,67 @@
+/* XLP8XX Device Tree Source
+ *
+ */
+
+/dts-v1/;
+/ {
+	model = "MIPS,XLP8XX";
+	compatible = "NETL,XLP8XX_A0";
+
+	#address-cells = <1>;
+	#size-cells = <1>;
+
+	hypervisor {
+		hypervisor-name = "Xen";
+		cpu_possible_map = <0xff>;
+		is_dom0_hvm = <1>;
+		bootargs = "dom0_loadaddr=0x09000000 dom0_cpumask=0x01 -- ";
+	};
+	doms {
+		#address-cells = <1>;
+		#size-cells = <1>;
+		dom@0 {
+			device_type = "domain";
+			os = "linux";
+
+			#address-cells = <1>;
+			#size-cells = <1>;
+
+			cpu {
+				onlinemask = <0x01>;
+			};
+			uart {
+				id = <0>;
+			};
+			memory {
+				reg = <0x04000000 0x0C000000	// 192M at 64M
+					0x20000000 0x20000000>;  // 512M at 512M
+			};
+			pic {
+			};
+		};
+		dom@1 {
+			device_type = "domain";
+			os = "netos";
+
+			heap-size = <0x0 0x4000000>;
+			shared-mem = <0x0 0x80000000 0x0 0x20000000>;
+
+			#address-cells = <1>;
+			#size-cells = <1>;
+
+			cpu {
+				onlinemask = <0x10>;
+			};
+			uart {
+				id = <1>;
+			};
+			memory {
+				reg = <0x30000000 0x10000000>;  // 256M @ 768M
+			};
+		};
+	};
+
+	chosen {
+		bootargs = " xlp_noi2c console=ttyS0,115200 rdinit=/sbin/init";
+	};
+};
diff --git a/dts/sysconfig.dts b/dts/sysconfig.dts
index d430f3e..a2226a5 100644
--- a/dts/sysconfig.dts
+++ b/dts/sysconfig.dts
@@ -7,9 +7,6 @@
 	model = "MIPS,XLP8XX";
 	compatible = "NETL,XLP8XX_A0";
 
-	#address-cells = <1>;
-	#size-cells = <1>;
-
 	hypervisor {
 
 		hypervisor-name = "Xen";
@@ -35,9 +32,6 @@
 
 		bootargs = " xlp_noi2c console=ttyS0,115200 rdinit=/sbin/init";
 
-		#address-cells = <1>;
-		#size-cells = <1>;
-
 		cpu {
 			set = <&cpu_dom0>;
 		};
@@ -102,7 +96,7 @@
 		};
 	};
 
-	sysconfig {
+	soc {
 
 		cpu {
 			#mask-cells = <1>;
@@ -134,31 +128,35 @@
 				range = <0x00000000 0x00100000>;	// 1M @ 0M
 				sharedcfg {};
 			};
-			mem_hypervisor: chunk@1 {
-				range = <0x00100000 0x03f000000>;	// 63M @ 1M
+			mem_syscfgdb: chunk@1 {
+				range = <0x00100000 0x00100000>;	// 1M @ 1M
 				sharedcfg {};
 			};
-			mem_dom0_lowmem: chunk@2 {
+			mem_hypervisor: chunk@2 {
+				range = <0x00100000 0x03E000000>;	// 63M @ 2M
+				sharedcfg {};
+			};
+			mem_dom0_lowmem: chunk@3 {
 				range = <0x04000000 0x0C000000>;	// 192M at 64M
 				sharedcfg {};
 			};
-			mem_uart0: chunk@3 {
+			mem_uart0: chunk@4 {
 				range = <0x1F000000 0x00400000>;  	// 4M @ 496M
 				sharedcfg {};
 			};
-			mem_uart1: chunk@4 {
+			mem_uart1: chunk@5 {
 				range = <0x1F400000 0x00400000>;  	// 4M @ 500M
 				sharedcfg {};
 			};
-			mem_dom0_highmem: chunk@5 {
+			mem_dom0_highmem: chunk@6 {
 				range = <0x20000000 0x20000000>;  	// 512M at 512M
 				sharedcfg {};
 			};
-			mem_dom1: chunk@6 {
+			mem_dom1: chunk@7 {
 				range = <0x30000000 0x10000000>;  	// 256M @ 768M
 				sharedcfg {};
 			};
-			mem_shared: chunk@7 {
+			mem_shared: chunk@8 {
 				#address-cells = <2>;
 				#size-cells = <2>;
 
@@ -252,4 +250,8 @@
 		};
 
 	};
+
+	board {
+	};
+
 };
-- 
1.7.1

