From 795defd3aa00f15ed8d4f521b296aee0c079ef17 Mon Sep 17 00:00:00 2001
From: Venu Vadapalli <vvadapalli@netlogicmicro.com>
Date: Tue, 17 Aug 2010 15:09:45 -0700
Subject: Parameters to use 3 domains

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/dts/sysconfig-xen.dts b/dts/sysconfig-xen.dts
index f0aebc4..dd4ba8b 100644
--- a/dts/sysconfig-xen.dts
+++ b/dts/sysconfig-xen.dts
@@ -14,7 +14,8 @@
 		hypervisor-name = "Xen";
 		cpu_possible_map = <0xff>;
 		is_dom0_hvm = <1>;
-		bootargs = "ncores=1 dom0_loadaddr=0x09000000 dom0_cpumask=0x01 -- ";
+		bootargs = "ncores=1 dom0_loadaddr=0x09000000 dom0_size=0x1c000000 dom0_cpumask=0x01 -- ";
+		domain_heap = <0x80000000 0x20000000>;
 	};
 	doms {
 		#address-cells = <1>;
@@ -31,10 +32,12 @@
 			};
 			uart {
 				id = <0>;
+				owner = <1>;
+				sharedcfg = <0x1f000000>;
 			};
 			memory {
 				reg = <0x04000000 0x0C000000	// 192M at 64M
-					0x20000000 0x10000000>;  // 512M at 512M
+					0x20000000 0x10000000>;  // 256M at 512M
 			};
 			pic {
 			};
@@ -44,7 +47,7 @@
 			os = "netos";
 
 			heap-size = <0x0 0x4000000>;                   // 64M
-			shared-mem = <0x0 0x80000000 0x0 0x20000000>;  // 512M @ 2GB
+			shared-mem = <0x0 0x40000000 0x0 0x08000000>;  // 128M @ 1GB
 
 			#address-cells = <1>;
 			#size-cells = <1>;
@@ -60,7 +63,31 @@
 				sharedcfg = <0x1f000000>;
 			};
 			memory {
-				reg = <0x30000000 0x10000000>;  // 256M @ 768M
+				reg = <0x30000000 0x08000000>;  // 256M @ 768M
+			};
+		};
+		dom@2 {
+			device_type = "domain";
+			os = "netos";
+
+			heap-size = <0x0 0x4000000>;                   // 64M
+			shared-mem = <0x0 0x48000000 0x0 0x08000000>;  // 128M @ 1.128GB
+
+			#address-cells = <1>;
+			#size-cells = <1>;
+
+			app-bootcmd = "ampload";
+			app-loadaddr = "0x47000000";
+
+			cpu {
+				onlinemask = <0x04>;
+			};
+			uart {
+				id = <0>;
+				sharedcfg = <0x1f000000>;
+			};
+			memory {
+				reg = <0x38000000 0x08000000>;  // 128M @ 896M
 			};
 		};
 	};
-- 
1.7.1

