From a0e47b959b7607730ca43a2889c4d54107753edc Mon Sep 17 00:00:00 2001
From: Venu Vadapalli <vvadapalli@netlogicmicro.com>
Date: Tue, 17 Aug 2010 18:32:24 -0700
Subject: Use 1.5GB (unused area) for preloading hyperexec app

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/dts/sysconfig-xen.dts b/dts/sysconfig-xen.dts
index dd4ba8b..9ab8507 100644
--- a/dts/sysconfig-xen.dts
+++ b/dts/sysconfig-xen.dts
@@ -53,7 +53,7 @@
 			#size-cells = <1>;
 
 			app-bootcmd = "ampload";
-			app-loadaddr = "0x47000000";
+			app-loadaddr = "0x60000000";
 
 			cpu {
 				onlinemask = <0x08>;
@@ -77,7 +77,7 @@
 			#size-cells = <1>;
 
 			app-bootcmd = "ampload";
-			app-loadaddr = "0x47000000";
+			app-loadaddr = "0x60000000";
 
 			cpu {
 				onlinemask = <0x04>;
@@ -95,4 +95,21 @@
 	chosen {
 		bootargs = " xlp_noi2c console=ttyS0,115200 rdinit=/sbin/init";
 	};
+
+	/* These binaries are downloaded at the resp physical memory locations
+	 * by their corresponding bootstrapping loaders. For example,
+	 * u-boot is loaded by x-load at 193M, sysconfig is loaded by u-boot
+	 * at 1M, hypervisor by u-boot hypervisor at 136M.
+	 * On simulator all these binaries are pre-loaded by using '-F' option
+	 * for convenience. On silicon, they will be loaded by their corresponding
+	 * temporary download locations the stage-1/stage-2 firmware
+	 */
+	firmware-download {
+		u-boot        = <0x0c100000 0x3f00000>;  /* 63M @ 193M */
+		sysconfig     = <0x00100000 0x00100000>; /* 1M @ 1M */
+		hypervisor    = <0x08800000 0x00800000>; /* 8M @ 136M */
+		dom0          = <0x09000000 0x03000000>; /* 48M @ 144M */
+		domU-ramdisk  = <0x60000000 0x60800000>; /* 8M @ 1536M */
+	};
 };
+
-- 
1.7.1

