From 8865c3832146459d0b397e522b22d13866358761 Mon Sep 17 00:00:00 2001
From: Prasad Boddupalli <pboddupalli@netlogicmicro.com>
Date: Thu, 19 Aug 2010 12:56:30 -0700
Subject: fix comments

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/dts/sysconfig-xen.dts b/dts/sysconfig-xen.dts
index 9ab8507..a2b9cf0 100644
--- a/dts/sysconfig-xen.dts
+++ b/dts/sysconfig-xen.dts
@@ -63,7 +63,7 @@
 				sharedcfg = <0x1f000000>;
 			};
 			memory {
-				reg = <0x30000000 0x08000000>;  // 256M @ 768M
+				reg = <0x30000000 0x08000000>;  // 128M @ 768M
 			};
 		};
 		dom@2 {
-- 
1.7.1

