From 92026313fde0e124d850c21dca0431908e796a3e Mon Sep 17 00:00:00 2001
From: Venu Vadapalli <vvadapalli@netlogicmicro.com>
Date: Thu, 2 Sep 2010 15:07:32 -0700
Subject: Auto generated xen config files and new kernel node for NetOS doms

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/dts/sysconfig-xen-dom-1.cfg b/dts/sysconfig-xen-dom-1.cfg
index c458b30..284f3ff 100644
--- a/dts/sysconfig-xen-dom-1.cfg
+++ b/dts/sysconfig-xen-dom-1.cfg
@@ -2,24 +2,11 @@
 
 kernel = "/tmp/xen/netos.elf"
 
-# Initial memory allocation (in megabytes) for the new domain.
-#
-# WARNING: Creating a domain with insufficient memory may cause out of
-#          memory errors. The domain needs enough memory to boot kernel
-#          and modules. Allocating less than 32MBs is not recommended.
 memory = 32
 
-# A name for your domain. All domains must have different names.
-name = "NetOS@1"
+name = "netos@1"
 
-# List of which CPUS this domain is allowed to use, default Xen picks
-#cpus = ""         # leave to Xen to pick
-#cpus = "0"        # all vcpus run on CPU0
-#cpus = "0-3,5,^1" # all vcpus run on cpus 0,2,3,5
-#cpus = ["3", "2"] # VCPU0 runs on CPU1, VCPU1 runs on CPU2
-cpus = ["128", "64"] # VCPU0 runs on CPU1
+cpus = ["32", "16"]
 
-# Number of Virtual CPUS to use, default is 1
 vcpus = 2
 
-
diff --git a/dts/sysconfig-xen-dom-2.cfg b/dts/sysconfig-xen-dom-2.cfg
index bcc687f..a4b1a72 100644
--- a/dts/sysconfig-xen-dom-2.cfg
+++ b/dts/sysconfig-xen-dom-2.cfg
@@ -2,23 +2,11 @@
 
 kernel = "/tmp/xen/netos.elf"
 
-# Initial memory allocation (in megabytes) for the new domain.
-#
-# WARNING: Creating a domain with insufficient memory may cause out of
-#          memory errors. The domain needs enough memory to boot kernel
-#          and modules. Allocating less than 32MBs is not recommended.
 memory = 32
 
-# A name for your domain. All domains must have different names.
-name = "NetOS@2"
+name = "netos@2"
 
-# List of which CPUS this domain is allowed to use, default Xen picks
-#cpus = ""         # leave to Xen to pick
-#cpus = "0"        # all vcpus run on CPU0
-#cpus = "0-3,5,^1" # all vcpus run on cpus 0,2,3,5
-#cpus = ["3", "2"] # VCPU0 runs on CPU1, VCPU1 runs on CPU2
-cpus = ["32", "16"] # VCPU0 runs on CPU1
+cpus = ["128", "64"]
 
-# Number of Virtual CPUS to use, default is 1
 vcpus = 2
 
diff --git a/dts/sysconfig-xen.dts b/dts/sysconfig-xen.dts
index bd098d8..f7f7ea0 100644
--- a/dts/sysconfig-xen.dts
+++ b/dts/sysconfig-xen.dts
@@ -55,8 +55,16 @@
 			app-bootcmd = "ampload";
 			app-loadaddr = "0x60000000";
 
+			kernel {
+				/* Kernel Memory allocated from xen domain-heap */
+				size = <0x02000000>;  // 32M
+
+				/* Location in Dom0 ramdisk */
+				location = "/tmp/xen/netos.elf";
+			};
+
 			cpu {
-				onlinemask = <0xC0>;
+				onlinemask = <0x30>;
 			};
 			uart {
 				id = <0>;
@@ -79,8 +87,16 @@
 			app-bootcmd = "smpload";
 			app-loadaddr = "0x60000000";
 
+			kernel {
+				/* Kernel Memory allocated from xen domain-heap */
+				size = <0x02000000>;  // 32M
+
+				/* Location in Dom0 ramdisk */
+				location = "/tmp/xen/netos.elf";
+			};
+
 			cpu {
-				onlinemask = <0x30>;
+				onlinemask = <0xC0>;
 			};
 			uart {
 				id = <0>;
-- 
1.7.1

