From 78d685948ae3819a2c86ed66f940ff0aaddd76fe Mon Sep 17 00:00:00 2001
From: Venu Vadapalli <vvadapalli@netlogicmicro.com>
Date: Thu, 2 Sep 2010 16:00:21 -0700
Subject: Deprecated xen dts/cfg files as they are now available in sysconfig repository

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/dts/sysconfig-xen-dom-1.cfg b/dts/sysconfig-xen-dom-1.cfg
deleted file mode 100644
index 284f3ff..0000000
--- a/dts/sysconfig-xen-dom-1.cfg
+++ /dev/null
@@ -1,12 +0,0 @@
-#  -*- mode: python; -*-
-
-kernel = "/tmp/xen/netos.elf"
-
-memory = 32
-
-name = "netos@1"
-
-cpus = ["32", "16"]
-
-vcpus = 2
-
diff --git a/dts/sysconfig-xen-dom-2.cfg b/dts/sysconfig-xen-dom-2.cfg
deleted file mode 100644
index a4b1a72..0000000
--- a/dts/sysconfig-xen-dom-2.cfg
+++ /dev/null
@@ -1,12 +0,0 @@
-#  -*- mode: python; -*-
-
-kernel = "/tmp/xen/netos.elf"
-
-memory = 32
-
-name = "netos@2"
-
-cpus = ["128", "64"]
-
-vcpus = 2
-
diff --git a/dts/sysconfig-xen.dts b/dts/sysconfig-xen.dts
deleted file mode 100644
index f7f7ea0..0000000
--- a/dts/sysconfig-xen.dts
+++ /dev/null
@@ -1,131 +0,0 @@
-/* XLP8XX Device Tree Source
- *
- */
-
-/dts-v1/;
-/ {
-	model = "MIPS,XLP8XX";
-	compatible = "NETL,XLP8XX_A0";
-
-	#address-cells = <1>;
-	#size-cells = <1>;
-
-	hypervisor {
-		hypervisor-name = "Xen";
-		cpu_possible_map = <0xff>;
-		is_dom0_hvm = <1>;
-		bootargs = "ncores=2 dom0_loadaddr=0x09000000 dom0_size=0x1c000000 dom0_cpumask=0x0f -- ";
-		domain_heap = <0x80000000 0x20000000>;
-	};
-	doms {
-		#address-cells = <1>;
-		#size-cells = <1>;
-		dom@0 {
-			device_type = "domain";
-			os = "linux";
-
-			#address-cells = <1>;
-			#size-cells = <1>;
-
-			cpu {
-				onlinemask = <0x0f>;
-			};
-			uart {
-				id = <0>;
-				owner = <1>;
-				sharedcfg = <0x1f000000>;
-			};
-			memory {
-				reg = <0x04000000 0x0C000000	// 192M at 64M
-					0x20000000 0x10000000>;  // 256M at 512M
-			};
-			pic {
-			};
-		};
-		dom@1 {
-			device_type = "domain";
-			os = "netos";
-
-			heap-size = <0x0 0x2000000>;                   // 32M
-			shared-mem = <0x0 0x40000000 0x0 0x08000000>;  // 128M @ 1GB
-
-			#address-cells = <1>;
-			#size-cells = <1>;
-
-			app-bootcmd = "ampload";
-			app-loadaddr = "0x60000000";
-
-			kernel {
-				/* Kernel Memory allocated from xen domain-heap */
-				size = <0x02000000>;  // 32M
-
-				/* Location in Dom0 ramdisk */
-				location = "/tmp/xen/netos.elf";
-			};
-
-			cpu {
-				onlinemask = <0x30>;
-			};
-			uart {
-				id = <0>;
-				sharedcfg = <0x1f000000>;
-			};
-			memory {
-				reg = <0x30000000 0x08000000>;  // 128M @ 768M
-			};
-		};
-		dom@2 {
-			device_type = "domain";
-			os = "netos";
-
-			heap-size = <0x0 0x2000000>;                   // 32M
-			shared-mem = <0x0 0x48000000 0x0 0x08000000>;  // 128M @ 1.128GB
-
-			#address-cells = <1>;
-			#size-cells = <1>;
-
-			app-bootcmd = "smpload";
-			app-loadaddr = "0x60000000";
-
-			kernel {
-				/* Kernel Memory allocated from xen domain-heap */
-				size = <0x02000000>;  // 32M
-
-				/* Location in Dom0 ramdisk */
-				location = "/tmp/xen/netos.elf";
-			};
-
-			cpu {
-				onlinemask = <0xC0>;
-			};
-			uart {
-				id = <0>;
-				sharedcfg = <0x1f000000>;
-			};
-			memory {
-				reg = <0x38000000 0x08000000>;  // 128M @ 896M
-			};
-		};
-	};
-
-	chosen {
-		bootargs = " xlp_noi2c console=ttyS0,115200 rdinit=/sbin/init";
-	};
-
-	/* These binaries are downloaded at the resp physical memory locations
-	 * by their corresponding bootstrapping loaders. For example,
-	 * u-boot is loaded by x-load at 193M, sysconfig is loaded by u-boot
-	 * at 1M, hypervisor by u-boot hypervisor at 136M.
-	 * On simulator all these binaries are pre-loaded by using '-F' option
-	 * for convenience. On silicon, they will be loaded by their corresponding
-	 * temporary download locations the stage-1/stage-2 firmware
-	 */
-	firmware-download {
-		u-boot        = <0x0c100000 0x3f00000>;  /* 63M @ 193M */
-		sysconfig     = <0x00100000 0x00100000>; /* 1M @ 1M */
-		hypervisor    = <0x08800000 0x00800000>; /* 8M @ 136M */
-		dom0          = <0x09000000 0x03000000>; /* 48M @ 144M */
-		domU-ramdisk  = <0x60000000 0x60800000>; /* 8M @ 1536M */
-	};
-};
-
-- 
1.7.1

