From 9e763c24d064beb561ef9929b5c6a5d20b6116a1 Mon Sep 17 00:00:00 2001
From: Vikas Gupta <vikasg@lnx3-234.(none)>
Date: Thu, 16 Jun 2011 21:26:36 -0700
Subject: Build fix for u-boot

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/libfdt/contrib/fdt_helper.c b/libfdt/contrib/fdt_helper.c
index 21a5756..de109ea 100644
--- a/libfdt/contrib/fdt_helper.c
+++ b/libfdt/contrib/fdt_helper.c
@@ -23,8 +23,9 @@
  THE POSSIBILITY OF SUCH DAMAGE.
 *******************************#NETL_2#**********************************/
 #if defined(__KERNEL__)
+#if !defined(NLM_HAL_UBOOT)
 #include <linux/kernel.h>
-
+#endif
 #else
 #include <stdio.h>
 #include <sys/types.h>
@@ -45,6 +46,11 @@ static int fdt_helper_print  = 1;
 	if (fdt_helper_print) \
 		printf(x);    \
 } while (0)
+#elif defined(NLM_HAL_UBOOT)
+#define fdtprint(x...) do {   \
+	if (fdt_helper_print) \
+		printf(x);    \
+} while (0)
 #else
 #define fdtprint(x...) do {   \
 	if (fdt_helper_print) \
-- 
1.7.1

