From 5154a8b543d369b2bfa29e00fa261f07e8ce29ac Mon Sep 17 00:00:00 2001
From: Hareesh R <hareesh@netlogicmicro.com>
Date: Wed, 21 Dec 2011 12:25:42 +0530
Subject: crypto finish function added

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/crypto/cryptoapi.c b/crypto/cryptoapi.c
index dc2a674..56471c3 100644
--- a/crypto/cryptoapi.c
+++ b/crypto/cryptoapi.c
@@ -139,6 +139,21 @@ int nlm_crypto_lib_init(void)
 }
 
 /**
+* @brief The crypto lib finish fuction releases the resources allocted  
+*
+* @return 
+*  - On sucess,  returns NLM_CRYPTO_OK. 
+*  - On failure, returns NLM_CRYPTO_ERROR.
+*	  
+* @ingroup crypto
+*/
+
+int nlm_crypto_lib_finish(void)
+{
+	return crypto_app_lib_finish();
+}
+
+/**
 * @brief The crypto open sync session create a context memory for the given sync session. 
 * The context memory is allocated from the shared mem so that it can be accessed 
 * from both kernel and userspace. In netos mode, as the application affinity is always set, 
diff --git a/crypto/include/cryptodev.h b/crypto/include/cryptodev.h
index 466f3c1..d706ff2 100644
--- a/crypto/include/cryptodev.h
+++ b/crypto/include/cryptodev.h
@@ -174,6 +174,7 @@ struct nlm_crypto_init_params
 };
 
 extern int crypto_app_lib_init(struct nlm_crypto_init_params *);
+extern int crypto_app_lib_finish(void);
 
 extern int contig_memory_init(void *shmaddr, unsigned long long paddr, unsigned long long size);
 /* alignment */
diff --git a/crypto/include/nlmcrypto.h b/crypto/include/nlmcrypto.h
index 916c479..4e75fa6 100644
--- a/crypto/include/nlmcrypto.h
+++ b/crypto/include/nlmcrypto.h
@@ -641,6 +641,7 @@ static inline int my_cpu_id(void)
 extern unsigned long xlp_rsa_base;
 extern unsigned long xlp_sae_base;
 int nlm_crypto_lib_init(void);
+int nlm_crypto_lib_finish(void);
 nlm_crypto_ctx_t *nlm_crypto_open_sync_session(int sync_mode, int cpu, void *arg);
 nlm_crypto_ctx_t *nlm_crypto_open_async_session(int max_outstanding_reqs, 
 		int (*callback)(nlm_crypto_ctx_t *ctxt, void *ctrl, void *param, void *arg), void *arg);
-- 
1.7.1

