From cef986a3cda7f904b89fef97869ac03319f345df Mon Sep 17 00:00:00 2001
From: ShengLiang Song <shengliang.song@broadcom.com>
Date: Wed, 4 Apr 2012 20:17:09 -0700
Subject: Removed fdt parse msg

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/libfdt/contrib/fdt_helper.c b/libfdt/contrib/fdt_helper.c
index de109ea..44c17ef 100644
--- a/libfdt/contrib/fdt_helper.c
+++ b/libfdt/contrib/fdt_helper.c
@@ -123,15 +123,19 @@ int copy_fdt_prop(void *fdt, const char *path, const char *prop,
 
 	nodeoffset = fdt_path_offset(fdt, path);
 	if (nodeoffset < 0) {
+#if 0
 		fdtprint("%s: Failed to parse path %s\n",
 		         fdt_strerror(nodeoffset), path);
+#endif
 		return nodeoffset;
 	}
 
 	pval = fdt_getprop(fdt, nodeoffset, prop, &plen);
 	if (pval == NULL) {
+#if 0
 		fdtprint("%s: Failed to parse property %s\n",
 		         fdt_strerror(plen), prop);
+#endif
 		return plen;
 	}
 
-- 
1.7.1

