From d12f1f909b00511cedb58d2f73d258b56714749f Mon Sep 17 00:00:00 2001
From: Wei Zhang <wezhang@broadcom.com>
Date: Fri, 20 Apr 2012 10:03:39 -0700
Subject: uboot clean up

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/libfdt/contrib/fdt_helper.c b/libfdt/contrib/fdt_helper.c
index 44c17ef..5a273eb 100644
--- a/libfdt/contrib/fdt_helper.c
+++ b/libfdt/contrib/fdt_helper.c
@@ -41,6 +41,7 @@
 
 static int fdt_helper_print  = 1;
 
+extern int printf(const char *format, ...);
 #if !defined(__KERNEL__)
 #define fdtprint(x...) do {   \
 	if (fdt_helper_print) \
-- 
1.7.1

