From 814a40fe0373a3069103e7ff0ba2d67f5a00d534 Mon Sep 17 00:00:00 2001
From: Jayachandran C <jchandra@broadcom.com>
Date: Wed, 17 Apr 2013 02:26:04 +0530
Subject: libfdt: support for compiling with linux kernel

This allows us to replace the fdt code that was in poart of the SDK kernel
under linux/arch/mips/netlogic/boot with this library
[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/libfdt/libfdt_env.h b/libfdt/libfdt_env.h
index cf2b98c..3785453 100644
--- a/libfdt/libfdt_env.h
+++ b/libfdt/libfdt_env.h
@@ -19,6 +19,9 @@
 #include <stddef.h>
 #include <types.h>
 #include <string.h>
+#elif defined(NLM_HAL_LINUX_KERNEL)
+#include <linux/types.h>
+#include <linux/string.h>
 #else
 #include <stddef.h>
 #include <stdint.h>
-- 
1.7.1

