From 7c2cbcd5de22f087ae673ec028de213c603dafed Mon Sep 17 00:00:00 2001
From: P. Sadik <psadik@broadcom.com>
Date: Fri, 26 Apr 2013 15:11:26 +0530
Subject: linux-3.7: Updates for compiling libraries against linux-3.7

Use Linux's standard Kbuild mechanism to build libraries as loadable modules.
[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/libfdt/Kbuild b/libfdt/Kbuild
new file mode 100644
index 0000000..44c65f3
--- /dev/null
+++ b/libfdt/Kbuild
@@ -0,0 +1,2 @@
+obj-m += fdt_interface.o
+fdt_interface-objs := fdt.o fdt_ro.o fdt_wip.o fdt_sw.o fdt_rw.o fdt_strerror.o contrib/fdt_export.o contrib/fdt_helper.o
diff --git a/libfdt/contrib/fdt_export.c b/libfdt/contrib/fdt_export.c
new file mode 100644
index 0000000..a1140b4
--- /dev/null
+++ b/libfdt/contrib/fdt_export.c
@@ -0,0 +1,20 @@
+#include <linux/types.h>
+#include <linux/kernel.h>
+#include <linux/module.h>
+
+#include <libfdt.h>
+#include "fdt_helper.h"
+
+EXPORT_SYMBOL(fdt_check_header);
+EXPORT_SYMBOL(fdt_path_offset);
+EXPORT_SYMBOL(fdt_getprop);
+EXPORT_SYMBOL(fdt_strerror);
+EXPORT_SYMBOL(fdt_get_path);
+EXPORT_SYMBOL(fdt_del_node);
+EXPORT_SYMBOL(fdt_add_subnode);
+EXPORT_SYMBOL(fdt_pack);
+EXPORT_SYMBOL(fdt_node_offset_by_prop_value);
+EXPORT_SYMBOL(fdt_node_offset_by_compatible);
+EXPORT_SYMBOL(fdt_setprop);
+EXPORT_SYMBOL(fdt_parent_offset);
+EXPORT_SYMBOL(copy_fdt_prop);
-- 
1.7.1

