From d9a685b3297c0a640ed09f8926278004046dd880 Mon Sep 17 00:00:00 2001
From: Yonghong Song <ysong@broadcom.com>
Date: Wed, 1 May 2013 10:36:51 -0700
Subject: fdtlib: Add a Makefile to build libfdtlib.a

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/libfdt/Makefile b/libfdt/Makefile
new file mode 100644
index 0000000..7014338
--- /dev/null
+++ b/libfdt/Makefile
@@ -0,0 +1,21 @@
+
+FDTLIB_DIR = $(SDK_ROOT)/libraries/fdt/libfdt
+FDTLIB_OBJS = fdt.o fdt_ro.o fdt_wip.o fdt_sw.o fdt_rw.o fdt_strerror.o contrib/fdt_export.o contrib/fdt_helper.o
+FDTLIB = libfdtlib.a
+
+include $(SDK_ROOT)/linux-userspace/Makefile.mk
+
+CFLAGS += -DNLM_HAL_LINUX_USER -I./ -I./contrib/
+
+$(FDTLIB): $(FDTLIB_OBJS)
+	echo "Build fdtlib library"
+	$(AR) -rcsP $(FDTLIB) $(FDTLIB_OBJS)
+
+all: $(FDTLIB)
+
+clean:
+	$(RM) *.a $(FDTLIB_OBJS)
+
+install:
+
+.PHONY: all clean install $(FDTLIB)
-- 
1.7.1

