From e59eb1d2f4a12e166bcb5b8d2d076543f73f686d Mon Sep 17 00:00:00 2001
From: Anurag <anurag.gopinath@broadcom.com>
Date: Thu, 16 May 2013 17:39:16 +0530
Subject: CDE : Adding support for XLP9XX and moving to the new HAL structure

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/netlogic/lib/syslib/include/nlm_hal_cde.h b/arch/mips/netlogic/lib/syslib/include/nlm_hal_cde.h
index de7a569..7248a44 100644
--- a/arch/mips/netlogic/lib/syslib/include/nlm_hal_cde.h
+++ b/arch/mips/netlogic/lib/syslib/include/nlm_hal_cde.h
@@ -89,10 +89,30 @@ static __inline__ uint64_t nlm_hal_cde_make_data_desc(int eof, int type, int sod
                 );
   return src_desc;
 }
+static __inline__ int nlm_hal_9xxcde_make_cmp_msg(uint64_t *msg0, uint64_t *msg1, int readp, int op, int rtn_bkt,
+				   int length, uint64_t src_addr)
+{
+  int stid = XLP_STNID_CMP;
+  *msg0 = ( ((
+                ((uint64_t) readp <<63) |
+                ((uint64_t) op << 62)  | 
+                ((uint64_t) length << 40) |
+                ((uint64_t) src_addr & 0xffffffffffULL))
+                &0xc003ffffffffffff) );
+  *msg1 = ( ((
+                ((uint64_t) rtn_bkt << 50) |
+                ((uint64_t) 0x3ffffffffffff))
+                &0x7fffffffffffffff));
+  return stid;
+}
+
 
 /* Return error when parameter are invalid */
-static __inline__ int nlm_hal_cde_send_request(uint32_t dst_vc, uint64_t entry0)
+static __inline__ int nlm_hal_cde_send_request(uint32_t dst_vc, uint64_t entry0, uint64_t entry1)
 {
+    if (cpu_is_xlp9xx())
+	return nlm_hal_send_msg2(dst_vc, 0, entry0, entry1);
+    else
 	return nlm_hal_send_msg1(dst_vc, 0, entry0);
 
 }
-- 
1.7.1

