From 4276e897373f4158d768de26a5f8e4e36a5fa538 Mon Sep 17 00:00:00 2001
From: Vikas Gupta <vikas.gupta@broadcom.com>
Date: Wed, 22 May 2013 03:10:18 -0700
Subject: netlib: remove redundant defines

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/netlogic/lib/netlib/include/netsoc_common.h b/arch/mips/netlogic/lib/netlib/include/netsoc_common.h
index b172fb1..16c6fcc 100644
--- a/arch/mips/netlogic/lib/netlib/include/netsoc_common.h
+++ b/arch/mips/netlogic/lib/netlib/include/netsoc_common.h
@@ -69,9 +69,6 @@
 #define DEFAULT_ILK_CAL_SLOTS           26
 
 
-#define GMAC    1
-#define XGMAC   2
-
 
 typedef struct {
         uint32_t iftype[5];
-- 
1.7.1

