From c6e9448aa0ce01c220cf4037cb43e4d59f80a2b5 Mon Sep 17 00:00:00 2001
From: Alok Agrawat <alok@broadcom.com>
Date: Wed, 22 May 2013 05:35:58 -0700
Subject: netlib: Enable debug print for non-Linux code as well.

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/netlogic/lib/netlib/include/netsoc_common.h b/arch/mips/netlogic/lib/netlib/include/netsoc_common.h
index 16c6fcc..77f4513 100644
--- a/arch/mips/netlogic/lib/netlib/include/netsoc_common.h
+++ b/arch/mips/netlogic/lib/netlib/include/netsoc_common.h
@@ -69,10 +69,9 @@
 #define DEFAULT_ILK_CAL_SLOTS           26
 
 
-
 typedef struct {
         uint32_t iftype[5];
-        union { 
+        union {
                 struct {
                         uint32_t phymode[4];
                         uint32_t phy_bus[4];
@@ -117,13 +116,13 @@ struct netsoc_info{
 		uint32_t max_ms_fifo;
 		uint32_t max_tx_fifo;
 		uint32_t max_pktlen_fifo;
-        
-                uint32_t max_stg2_offset; 
+
+                uint32_t max_stg2_offset;
                 uint32_t max_eh_offset;
                 uint32_t max_frout_offset;
                 uint32_t max_ms_offset;
                 uint32_t max_tx_offset;
-                uint32_t max_pktlen_offset; 
+                uint32_t max_pktlen_offset;
 	}egress;
 	struct {
 		uint32_t max_l2_entries;
@@ -190,7 +189,7 @@ static inline uint32_t __netsoc_get_max_nae(void)
         if (is_nlm_xlp9xx()) {
                 return XLP9XX_MAX_NAE_PERNODE;
         }
-        else {  
+        else {
                 return XLP_MAX_NAE_PERNODE;
         }
 }
@@ -200,7 +199,7 @@ static inline uint32_t __netsoc_get_max_poe(void)
         if (is_nlm_xlp9xx()) {
                 return XLP9XX_MAX_POE_PERNODE;
         }
-        else {  
+        else {
                 return XLP_MAX_POE_PERNODE;
         }
 }
@@ -211,7 +210,7 @@ static inline uint32_t __netsoc_get_max_ale(void)
         if (is_nlm_xlp9xx()) {
                 return MAX_ALE_PERNODE;
         }
-        else {  
+        else {
                 return 0;
         }
 }
@@ -270,10 +269,10 @@ static inline uint32_t get_free_desc_size(uint32_t intf_type)
 static inline uint32_t get_intf_fifo_size(uint32_t intf_type)
 {
         switch(intf_type) {
-        case SGMII_IF: 
+        case SGMII_IF:
                 return (current_netsoc->ingress.max_if_fifo / current_netsoc->max_ports);
-        case XAUI_IF:  
-        case RXAUI_IF: 
+        case XAUI_IF:
+        case RXAUI_IF:
                 return (current_netsoc->ingress.max_if_fifo / current_netsoc->max_ports) * 4;
         case INTERLAKEN_IF:
                 return (current_netsoc->ingress.max_if_fifo / current_netsoc->max_ports) * 8;	//FIXME J
@@ -285,17 +284,17 @@ static inline uint32_t get_intf_fifo_size(uint32_t intf_type)
 static inline uint32_t get_iface_credit(uint32_t intf_type)
 {
 	uint32_t size = get_intf_fifo_size(intf_type);
-	return (size >> 2);	
+	return (size >> 2);
 }
 
 static inline uint32_t get_prsr_seq_fifo_size(uint32_t intf_type)
 {
         switch(intf_type) {
-        case SGMII_IF:  
+        case SGMII_IF:
                 return DEFAULT_SGMII_PARSERSEQFIFO_SZ;
         case XAUI_IF:
-        case RXAUI_IF:  
-                return DEFAULT_XAUI_PARSERSEQFIFO_SZ;   
+        case RXAUI_IF:
+                return DEFAULT_XAUI_PARSERSEQFIFO_SZ;
         case INTERLAKEN_IF:
                 return DEFAULT_ILK_PARSERSEQFIFO_SZ;
         default:
@@ -306,11 +305,11 @@ static inline uint32_t get_prsr_seq_fifo_size(uint32_t intf_type)
 static inline uint32_t get_rx_buf_size(uint32_t intf_type)
 {
         switch(intf_type) {
-        case SGMII_IF:     
+        case SGMII_IF:
                 return DEFAULT_SGMII_RXFIFO_SZ;
         case XAUI_IF:
-        case RXAUI_IF:  
-                return DEFAULT_XAUI_RXFIFO_SZ;   
+        case RXAUI_IF:
+                return DEFAULT_XAUI_RXFIFO_SZ;
         case INTERLAKEN_IF:
                 return DEFAULT_ILK_RXFIFO_SZ;
         default:
@@ -346,7 +345,7 @@ static inline void update_default_config(int intf_type, uint32_t *config, uint32
                         *config++ = val;
                 }
         }
-        else {  
+        else {
                 *config = val;
         }
 }
diff --git a/arch/mips/netlogic/lib/netlib/include/netsoc_libiface.h b/arch/mips/netlogic/lib/netlib/include/netsoc_libiface.h
index 45dd30a..f789f0f 100644
--- a/arch/mips/netlogic/lib/netlib/include/netsoc_libiface.h
+++ b/arch/mips/netlogic/lib/netlib/include/netsoc_libiface.h
@@ -186,7 +186,8 @@ static inline void netsoc_api_writel(uint64_t base, uint32_t index, uint32_t val
                 printk(fmt, ##args); \
         } \
 }
-#elif defined(NLM_HAL_LINUX_USER)
+//#elif //defined(NLM_HAL_LINUX_USER)
+#else
 #define netsoc_api_print(dbglvl, fmt, args...)  \
 { \
         if (dbglvl & netsoc_api_dbglvl) { \
-- 
1.7.1

