From 5da2a45f7195b57b58ad822deebf196672e0ec98 Mon Sep 17 00:00:00 2001
From: Yonghong Song <ysong@broadcom.com>
Date: Thu, 23 May 2013 10:08:03 -0700
Subject: syslib: remove simulator workaround for various PCIM base addresses

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/netlogic/lib/syslib/src/nlm_hal.c b/arch/mips/netlogic/lib/syslib/src/nlm_hal.c
index fd90a7e..619559b 100644
--- a/arch/mips/netlogic/lib/syslib/src/nlm_hal.c
+++ b/arch/mips/netlogic/lib/syslib/src/nlm_hal.c
@@ -401,6 +401,7 @@ __inline__ void nlm_hal_init(void)
 
         	/* PCI enumeration of supported devices*/
 
+#if 0
 		/* FIXME: this is a workaround for simulator, once uboot is ready for simulator,
 		 * the below should be removed.
 		 */
@@ -414,7 +415,9 @@ __inline__ void nlm_hal_init(void)
 			xlp_mac_base[0] = 0x0;
 			xlp_nae_base[0] = 0x0;
 			xlp_nae_base[0] = 0x0;
-		} else {
+		} else 
+#endif
+		{
 			xlp_fmn_base[node] = mask & nlm_hal_read_32bit_reg((0x18000000 + XLP_CFG_BASE(node, XLP_FMN)), PCI_MEM_BAR_0);
 
 			xlp_mac_base[node] = mask & nlm_hal_read_32bit_reg((0x18000000 + XLP_CFG_BASE(node, XLP_NAE)), PCI_MEM_BAR_0); //0x18018000
-- 
1.7.1

