From a69f82a525c2ddcd75fb5d6cf53f2a1d4e07147a Mon Sep 17 00:00:00 2001
From: Virendra Pathak <vpathak@broadcom.com>
Date: Wed, 5 Jun 2013 13:10:04 +0530
Subject: compilation: fix linux userspace compilation

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/netlogic/lib/syslib/src/nlm_hal.c b/arch/mips/netlogic/lib/syslib/src/nlm_hal.c
index 3027a51..256e8e6 100644
--- a/arch/mips/netlogic/lib/syslib/src/nlm_hal.c
+++ b/arch/mips/netlogic/lib/syslib/src/nlm_hal.c
@@ -1231,7 +1231,7 @@ EXPORT_SYMBOL(nlm_hal_is_xlp_a0);
 EXPORT_SYMBOL(nlm_hal_is_xlp_le);
 EXPORT_SYMBOL(nlm_hal_get_dev_base);
 EXPORT_SYMBOL(nlm_hal_set_sae_freq);
-EXPORT_SYMBOL(nlm_hal_get_chip_feature);
+EXPORT_SYMBOL(nlm_hal_get_sae_chip_feature);
 EXPORT_SYMBOL(nlm_hal_set_rsa_freq);
 EXPORT_SYMBOL(nlm_hal_set_dtre_freq);
 EXPORT_SYMBOL(nlm_hal_set_cde_freq);
-- 
1.7.1

