From d78c855efe01973bb304812dfef04b65d033dd70 Mon Sep 17 00:00:00 2001
From: Jayanthi A <jayanthi.annadurai@broadcom.com>
Date: Sat, 1 Jun 2013 03:20:19 -0700
Subject: netlib: Fixed POE DV register offset

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/netlogic/lib/netlib/include/netsoc_poe.h b/arch/mips/netlogic/lib/netlib/include/netsoc_poe.h
index 2a97f1c..9152314 100644
--- a/arch/mips/netlogic/lib/netlib/include/netsoc_poe.h
+++ b/arch/mips/netlogic/lib/netlib/include/netsoc_poe.h
@@ -105,7 +105,7 @@ typedef enum {
         FMNRX_MODE_MAX,
 }class_map_t;
 
-#define NETV1_POE_DIST_VEC0_CELL0_BASE  0x108
+#define NETV1_POE_DIST_VEC0_CELL0_BASE  0x110
 #define NETV1_DV_REG_BASE(vec) 	(NETV1_POE_DIST_VEC0_CELL0_BASE + (16 * vec) - 1)
 
 static inline void __netsoc_v1_config_core_thrsel(poe_t *poe, cpu_mask_t *cpumap, uint32_t vc_sel_groupid)
-- 
1.7.1

