From 2e6e39182e9ca639a10492f68a18f34dd20cf753 Mon Sep 17 00:00:00 2001
From: Alok Agrawat <alok@broadcom.com>
Date: Mon, 3 Jun 2013 18:58:20 -0700
Subject: nae: fixed comiplation issues.

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/netlogic/lib/fmnlib/nlm_hal_fmn_config.c b/arch/mips/netlogic/lib/fmnlib/nlm_hal_fmn_config.c
index 9e88e8b..662388c 100644
--- a/arch/mips/netlogic/lib/fmnlib/nlm_hal_fmn_config.c
+++ b/arch/mips/netlogic/lib/fmnlib/nlm_hal_fmn_config.c
@@ -989,7 +989,6 @@ int nlm_hal_setup_outq(int node, int max_nodes)
 				val = (1ULL << 63) | (qid << 6) | (qid << 0);
 
 #if !defined(NLM_HAL_UBOOT) && !defined(NLM_HAL_NETLBOOT)
-#error "compile"
 				/* spills, maximum spill per VC: 64 * 4KB = 256KB,
 				 * In our configuration, each VC: 4KB
 				 */
diff --git a/arch/mips/netlogic/lib/netlib/external/ext_phy_bcm.c b/arch/mips/netlogic/lib/netlib/external/ext_phy_bcm.c
index 6756034..78aa030 100644
--- a/arch/mips/netlogic/lib/netlib/external/ext_phy_bcm.c
+++ b/arch/mips/netlogic/lib/netlib/external/ext_phy_bcm.c
@@ -192,7 +192,7 @@ static void bcm_start_an(net_port_t *netport)
                 __netsoc_mdio_wr(nae, NLM_HAL_INT_MDIO, 0, int_inf, 0x00, status|(1<<9)); // Restart XLP AN
         }while(count<100);
 
-        netsoc_api_print(NETSOC_APIDBG_DEFAULT,"Autonegotiation is NOT OK for PHY-SGMII inf=0x%x int_inf =0x%x\n", phyaddr, int_inf);
+        //netsoc_api_print(NETSOC_APIDBG_DEFAULT,"Autonegotiation is NOT OK for PHY-SGMII inf=0x%x int_inf =0x%x\n", phyaddr, int_inf);
         return;
 }
 
diff --git a/arch/mips/netlogic/lib/netlib/src/netsoc_nae.c b/arch/mips/netlogic/lib/netlib/src/netsoc_nae.c
index f168d1e..ce773df 100644
--- a/arch/mips/netlogic/lib/netlib/src/netsoc_nae.c
+++ b/arch/mips/netlogic/lib/netlib/src/netsoc_nae.c
@@ -2360,7 +2360,7 @@ void __netsoc_init_egress(nae_t *nae)
         uint32_t tx_cfg =  netsoc_read_nae_reg(nae_base, TX_CONFIG);
 
         if ((is_nlm_xlp3xx()) || is_nlm_xlp2xx() || is_nlm_xlp9xx() || (is_nlm_xlp8xx_bx())) {
-                netsoc_write_nae_reg(nae_base, TX_CONFIG, tx_cfg | NAE_TX_ENABLE | NAE_TX_ACE  | NAE_TX_COMPATIBLE | (1 << 3) | );
+                netsoc_write_nae_reg(nae_base, TX_CONFIG, tx_cfg | NAE_TX_ENABLE | NAE_TX_ACE  | NAE_TX_COMPATIBLE | (1 << 3) );
         }
         else {
                 netsoc_write_nae_reg(nae_base, TX_CONFIG, tx_cfg | NAE_TX_ENABLE | NAE_TX_ACE);
-- 
1.7.1

