From 427327860c99c9382088d2f3223bd68f000ae8d3 Mon Sep 17 00:00:00 2001
From: Sreenidhi B R <sreenira@broadcom.com>
Date: Tue, 11 Jun 2013 16:56:32 +0530
Subject: xlp9xx: add cores 15-19 for fmn

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/netlogic/lib/fmnlib/nlm_hal_fmn_config.c b/arch/mips/netlogic/lib/fmnlib/nlm_hal_fmn_config.c
index 662388c..ece89b1 100644
--- a/arch/mips/netlogic/lib/fmnlib/nlm_hal_fmn_config.c
+++ b/arch/mips/netlogic/lib/fmnlib/nlm_hal_fmn_config.c
@@ -80,6 +80,10 @@ nlm_fmn_config_t xlp3xx_fmn_config[] = {
 [XLP_MSG_HANDLE_CPU13] = {XLP_3XX_INVALID_STATION, 0},
 [XLP_MSG_HANDLE_CPU14] = {XLP_3XX_INVALID_STATION, 0},
 [XLP_MSG_HANDLE_CPU15] = {XLP_3XX_INVALID_STATION, 0},
+[XLP_MSG_HANDLE_CPU16] = {XLP_3XX_INVALID_STATION, 0},
+[XLP_MSG_HANDLE_CPU17] = {XLP_3XX_INVALID_STATION, 0},
+[XLP_MSG_HANDLE_CPU18] = {XLP_3XX_INVALID_STATION, 0},
+[XLP_MSG_HANDLE_CPU19] = {XLP_3XX_INVALID_STATION, 0},
 
 [XLP_MSG_HANDLE_PCIE0] = {XLP_PCIE0_VC_BASE, XLP_PCIE0_VC_LIMIT},
 [XLP_MSG_HANDLE_PCIE1] = {XLP_PCIE1_VC_BASE, XLP_PCIE1_VC_LIMIT},
@@ -131,6 +135,10 @@ nlm_fmn_config_t xlp2xx_fmn_config[] = {
 [XLP_MSG_HANDLE_CPU13] = {XLP_2XX_INVALID_STATION, 0},
 [XLP_MSG_HANDLE_CPU14] = {XLP_2XX_INVALID_STATION, 0},
 [XLP_MSG_HANDLE_CPU15] = {XLP_2XX_INVALID_STATION, 0},
+[XLP_MSG_HANDLE_CPU16] = {XLP_2XX_INVALID_STATION, 0},
+[XLP_MSG_HANDLE_CPU17] = {XLP_2XX_INVALID_STATION, 0},
+[XLP_MSG_HANDLE_CPU18] = {XLP_2XX_INVALID_STATION, 0},
+[XLP_MSG_HANDLE_CPU19] = {XLP_2XX_INVALID_STATION, 0},
 
 [XLP_MSG_HANDLE_PCIE0] = {XLP_PCIE0_VC_BASE, XLP_PCIE0_VC_LIMIT},
 [XLP_MSG_HANDLE_PCIE1] = {XLP_PCIE1_VC_BASE, XLP_PCIE1_VC_LIMIT},
@@ -180,6 +188,10 @@ nlm_fmn_config_t xlp9xx_fmn_config[] = {
 [XLP_MSG_HANDLE_CPU13] = {XLP_STNID_CPU13, XLP_CPU13_VC_LIMIT},
 [XLP_MSG_HANDLE_CPU14] = {XLP_STNID_CPU14, XLP_CPU14_VC_LIMIT},
 [XLP_MSG_HANDLE_CPU15] = {XLP_STNID_CPU15, XLP_CPU15_VC_LIMIT},
+[XLP_MSG_HANDLE_CPU16] = {XLP_STNID_CPU16, XLP_CPU16_VC_LIMIT},
+[XLP_MSG_HANDLE_CPU17] = {XLP_STNID_CPU17, XLP_CPU17_VC_LIMIT},
+[XLP_MSG_HANDLE_CPU18] = {XLP_STNID_CPU18, XLP_CPU18_VC_LIMIT},
+[XLP_MSG_HANDLE_CPU19] = {XLP_STNID_CPU19, XLP_CPU19_VC_LIMIT},
 
 [XLP_MSG_HANDLE_PCIE0]   = {XLP_9XX_PCIE0_VC_BASE, XLP_9XX_PCIE0_VC_LIMIT},
 [XLP_MSG_HANDLE_PCIE1]   = {XLP_9XX_PCIE1_VC_BASE, XLP_9XX_PCIE1_VC_LIMIT},
@@ -734,6 +746,7 @@ static void nlm_hal_write_fmn_credit(int node, int max_nodes)
 
 		fmn_config = &xlp9xx_fmn_config[0];
 
+		fmn_modify_qsize_credit_config(node, XLP_MSG_BLK_CPU, 20, XLP_CPU0_VC_BASE, XLP_CPU19_VC_LIMIT);
 		fmn_modify_qsize_credit_config(node, XLP_MSG_BLK_GDX, 1, XLP_9XX_GDX0_VC_BASE, XLP_9XX_GDX0_VC_LIMIT);
 		fmn_modify_qsize_credit_config(node, XLP_MSG_BLK_REGX, 1, XLP_9XX_RGX_VC_BASE, XLP_9XX_RGX_VC_LIMIT);
 		fmn_modify_qsize_credit_config(node, XLP_MSG_BLK_CMP, 1, XLP_9XX_CMP_VC_BASE, XLP_9XX_CMP_VC_LIMIT);
diff --git a/arch/mips/netlogic/lib/syslib/include/nlm_hal_xlp_dev.h b/arch/mips/netlogic/lib/syslib/include/nlm_hal_xlp_dev.h
index 0bc6615..474e13c 100644
--- a/arch/mips/netlogic/lib/syslib/include/nlm_hal_xlp_dev.h
+++ b/arch/mips/netlogic/lib/syslib/include/nlm_hal_xlp_dev.h
@@ -338,6 +338,10 @@ extern int is_nlm_xlp(unsigned int chipid, unsigned int rev,  unsigned int ext);
 #define XLP_STNID_CPU13         0xD0
 #define XLP_STNID_CPU14         0xE0
 #define XLP_STNID_CPU15         0xF0
+#define XLP_STNID_CPU16         0x100
+#define XLP_STNID_CPU17         0x110
+#define XLP_STNID_CPU18         0x120
+#define XLP_STNID_CPU19         0x130
 
 #define XLP_CPU8_VC_LIMIT       0x8F
 #define XLP_CPU9_VC_LIMIT       0x9F
@@ -347,6 +351,10 @@ extern int is_nlm_xlp(unsigned int chipid, unsigned int rev,  unsigned int ext);
 #define XLP_CPU13_VC_LIMIT      0xDF
 #define XLP_CPU14_VC_LIMIT      0xEF
 #define XLP_CPU15_VC_LIMIT      0xFF
+#define XLP_CPU16_VC_LIMIT      0x10F
+#define XLP_CPU17_VC_LIMIT      0x11F
+#define XLP_CPU18_VC_LIMIT      0x12F
+#define XLP_CPU19_VC_LIMIT      0x13F
 
 #define XLP_9XX_PCIE0_VC_BASE	0x180
 #define XLP_9XX_PCIE0_VC_LIMIT	0x181
@@ -555,6 +563,10 @@ enum XLP_MSG_HANDLES  {
         XLP_MSG_HANDLE_CPU13,
         XLP_MSG_HANDLE_CPU14,
         XLP_MSG_HANDLE_CPU15,
+        XLP_MSG_HANDLE_CPU16,
+        XLP_MSG_HANDLE_CPU17,
+        XLP_MSG_HANDLE_CPU18,
+        XLP_MSG_HANDLE_CPU19,
 
 	/* io */
         XLP_MSG_HANDLE_PCIE0,
-- 
1.7.1

