From 7eba772bbec0a5ba73b9a4793328cccd41d8d612 Mon Sep 17 00:00:00 2001
From: Vikas Gupta <vikas.gupta@broadcom.com>
Date: Thu, 13 Jun 2013 20:26:18 -0700
Subject: xlp3xx: Fix for PMA workaroud calling wrong function.

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/netlogic/lib/netlib/include/netsoc_nae.h b/arch/mips/netlogic/lib/netlib/include/netsoc_nae.h
index a70c10a..9e47fd3 100644
--- a/arch/mips/netlogic/lib/netlib/include/netsoc_nae.h
+++ b/arch/mips/netlogic/lib/netlib/include/netsoc_nae.h
@@ -502,7 +502,7 @@ static inline uint8_t __netsoc_read_pma_reg(nae_t *nae, int block, int lane_ctrl
         netsoc_write_mac_reg(mac_base, lane_ctrl, regval
                         | (serdes_reg << PHY_LANE_CTRL_ADDR_POS));
 
-        while (((serdes_val = netsoc_read_nae_reg(mac_base, lane_ctrl)) & PHY_LANE_CTRL_CMD_PENDING));
+        while (((serdes_val = netsoc_read_mac_reg(mac_base, lane_ctrl)) & PHY_LANE_CTRL_CMD_PENDING));
 
         return (serdes_val & 0xFF);
 }
diff --git a/arch/mips/netlogic/lib/netlib/src/netsoc_pcs.c b/arch/mips/netlogic/lib/netlib/src/netsoc_pcs.c
index bd61591..41e4d92 100644
--- a/arch/mips/netlogic/lib/netlib/src/netsoc_pcs.c
+++ b/arch/mips/netlogic/lib/netlib/src/netsoc_pcs.c
@@ -108,9 +108,9 @@ static inline void nlm_hal_PMFF_ALL_workaround(nae_t *nae, int block, int lane_c
         volatile int val = 0;
 
         if (is_nlm_xlp3xx_ax() || is_nlm_xlp3xx_b0()) {
-                val = __netsoc_indirect_pma_register_read(nae, block, lane_ctrl, SER_GEN1_PWR_DOWN);
+                val =__netsoc_read_pma_reg(nae, block, lane_ctrl, SER_GEN1_PWR_DOWN);
                 val = val | SERDES_PMFF_ALL_SET;
-                __netsoc_indirect_pma_register_wite(nae, block, lane_ctrl, SER_GEN1_PWR_DOWN, val);
+                __netsoc_write_pma_reg(nae, block, lane_ctrl, SER_GEN1_PWR_DOWN, val);
         }
 }
 
-- 
1.7.1

