From 1aaee8dd51800dedebfe83cdd424e2c01e9a8f0b Mon Sep 17 00:00:00 2001
From: K B S Naveen KUNDULA <kundula@lc-blr-153.ban.broadcom.com>
Date: Fri, 21 Jun 2013 12:32:16 +0530
Subject: Cleanup: Fixed the description of parameter for txpll reset function

   This fixes JIRA issue XLPSDK-1526
[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/netlogic/lib/netlib/src/nlm_hal_nae.c b/arch/mips/netlogic/lib/netlib/src/nlm_hal_nae.c
index acacb86..192ff37 100644
--- a/arch/mips/netlogic/lib/netlib/src/nlm_hal_nae.c
+++ b/arch/mips/netlogic/lib/netlib/src/nlm_hal_nae.c
@@ -921,7 +921,7 @@ void xlp2xx_nae_lane_reset_txpll(int node, int block, int lane_ctrl, int mode)
 *
 * @param [in] node Node number
 * @param [in] block NAE Register Memory Map Block
-* @param [in] lane_ctrl Lane to reset (0-3)
+* @param [in] lane_ctrl Lane to reset (4-7). 4 for lane 0, 5 for lane 1, 6 for lane 2, 7 for lane 3 
 * @param [in] mode 0-unused, 1-XAUI/HS-SGMII, 2-SGMII, 3-INTERLAKEN, 6-PHYMODE_RXAUI
 *
 * @return
-- 
1.7.1

