From 02a519ced042d02c8d506916650a52563c72ac7a Mon Sep 17 00:00:00 2001
From: Yonghong Song <ysong@broadcom.com>
Date: Mon, 1 Jul 2013 11:17:46 -0700
Subject: syslib: dtre: fix the BUS issue for xlp9xx dtre initialization

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/netlogic/lib/syslib/src/nlm_hal.c b/arch/mips/netlogic/lib/syslib/src/nlm_hal.c
index e429237..6a66016 100644
--- a/arch/mips/netlogic/lib/syslib/src/nlm_hal.c
+++ b/arch/mips/netlogic/lib/syslib/src/nlm_hal.c
@@ -893,8 +893,8 @@ void nlm_hal_set_dtre_freq(int node, int freq)
    */
 static void nlm_hal_dtr_9xx_init(void *fdt)
 {
-	uint64_t base0 = nlm_hal_get_dev_base (XLP_DTR_NODE, XLP_DTR_BUS, XLP_DTR_DEVICE, XLP_DTR_FUNC);
-	uint64_t base1 = nlm_hal_get_dev_base (XLP_DTR_NODE, XLP_DTR_BUS, XLP_DTR_DEVICE, 6);
+	uint64_t base0 = nlm_hal_get_dev_base (XLP_DTR_NODE, 1, XLP_DTR_DEVICE, XLP_DTR_FUNC);
+	uint64_t base1 = nlm_hal_get_dev_base (XLP_DTR_NODE, 1, XLP_DTR_DEVICE, 6);
 	int frequency = 0, node = 0;
 
 	/* xlp9xx has two DTRE engines.  We need to initialize both */
-- 
1.7.1

