From 15508d53238dcb25d883bf8c8b6512d7d6fe51c0 Mon Sep 17 00:00:00 2001
From: Vikas Gupta <vikas.gupta@broadcom.com>
Date: Mon, 17 Jun 2013 16:29:41 -0700
Subject: nae: fix for wrong check for intf type.

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/netlogic/lib/netlib/include/netsoc_common.h b/arch/mips/netlogic/lib/netlib/include/netsoc_common.h
index e291830..414c4a0 100644
--- a/arch/mips/netlogic/lib/netlib/include/netsoc_common.h
+++ b/arch/mips/netlogic/lib/netlib/include/netsoc_common.h
@@ -354,7 +354,7 @@ static inline uint32_t get_max_lanes(uint32_t intf_type)
 	}
 	else if(INTERLAKEN_IF==intf_type){
 		return 8;	
-	}else if((SGMII_IF==intf_type) && (XFI_IF==intf_type)){
+	}else if((SGMII_IF==intf_type) || (XFI_IF==intf_type)){
 		return 1;
 	}else
 		return 0;	
diff --git a/arch/mips/netlogic/lib/netlib/src/netsoc_nae.c b/arch/mips/netlogic/lib/netlib/src/netsoc_nae.c
index f3b6006..1fabc0e 100644
--- a/arch/mips/netlogic/lib/netlib/src/netsoc_nae.c
+++ b/arch/mips/netlogic/lib/netlib/src/netsoc_nae.c
@@ -1277,6 +1277,7 @@ void __netsoc_config_ingress_fifo(nae_t *nae)
 			break;
 		   case SGMII_IF:
 			fifo_xoff_thresh = 6;
+			rx_buf_size = nae->ports[port].rx_buf_size;
 			break;
 		   case INTERLAKEN_IF:
 			fifo_xoff_thresh = 12;
@@ -1292,7 +1293,8 @@ void __netsoc_config_ingress_fifo(nae_t *nae)
                                         rx_buf_base, nae->ports[port].rx_buf_size);
 
 	//	if(nae->ports[port].vlan_pri_en && rx_buf_size)
-			__netsoc_config_context_xoff_thr(nae, thrgrp,  (rx_buf_size << 2) / nae->ports[port].num_channels);
+			if(rx_buf_size)
+				__netsoc_config_context_xoff_thr(nae, thrgrp,  (rx_buf_size << 2) / nae->ports[port].num_channels);
 	//	else
 	//		thrgrp = 0;
 
-- 
1.7.1

