From 9c570bd5fc90d91f0367f19b197307b51b1cf1fb Mon Sep 17 00:00:00 2001
From: Alok Agrawat <alok@broadcom.com>
Date: Thu, 20 Jun 2013 16:39:40 -0700
Subject: Fixed netsoc_fdt compilation error.

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/netlogic/lib/netlib/src/netsoc_fdt.c b/arch/mips/netlogic/lib/netlib/src/netsoc_fdt.c
index 831f751..73625d6 100644
--- a/arch/mips/netlogic/lib/netlib/src/netsoc_fdt.c
+++ b/arch/mips/netlogic/lib/netlib/src/netsoc_fdt.c
@@ -1138,7 +1138,7 @@ static int fdt_parse_port_config(void *fdt, nae_t *nae_cfg)
                         nae_cfg->xgmii_speed[block]=1; /*1G SGMII*/
                 }
                 //printk("\n#@#@@@###################@@@@@@@@@@@@@@@@@@@@@@@@@@@@@ nae_cfg->msec_port_enable = %x block = %d cmplx_cfg.msec_port_enable = %x\n",
-                nae_cfg->msec_port_enable, block, cmplx_cfg.msec_port_enable);
+                //nae_cfg->msec_port_enable, block, cmplx_cfg.msec_port_enable);
 #ifdef MACSEC_DEBUG
                 netsoc_api_print(NETSOC_APIDBG_FDT," nae_cfg->msec_port_enable = %x block = %d cmplx_cfg.msec_port_enable = %x\n",
                 nae_cfg->msec_port_enable, block, cmplx_cfg.msec_port_enable);
-- 
1.7.1

