From 9689fb4e59fbf99c8d3d83ed0c3e674ea287699c Mon Sep 17 00:00:00 2001
From: Alok Agrawat <alok@broadcom.com>
Date: Fri, 21 Jun 2013 17:40:38 -0700
Subject: Disabled MACSec.

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/netlogic/lib/netlib/src/netsoc_fdt.c b/arch/mips/netlogic/lib/netlib/src/netsoc_fdt.c
index 73625d6..1c8f997 100644
--- a/arch/mips/netlogic/lib/netlib/src/netsoc_fdt.c
+++ b/arch/mips/netlogic/lib/netlib/src/netsoc_fdt.c
@@ -184,6 +184,8 @@ static int fdt_get_interface_type(void *fdt, char *nae_port_str, int block, int
                 }
                 else if (!strcmp(port_type_str, "interlaken")) {
                         intf_type = DC_ILK;
+prinkt("Interlaker selected \n");
+
                 }
                 else {
                         intf_type = UNKNOWN_IF;
@@ -973,7 +975,7 @@ static void extract_complex_params(void *fdt, int intf_type, char *nae_port_str,
 	netsoc_api_print(NETSOC_APIDBG_FDT,"Extracted complex configuration\n");
 }
 
-#define MACSEC_DEBUG 1
+//#define MACSEC_DEBUG 1
 
 static int fdt_parse_port_config(void *fdt, nae_t *nae_cfg)
 {
@@ -1159,8 +1161,8 @@ static int fdt_parse_port_config(void *fdt, nae_t *nae_cfg)
                         nae_port->ext_phy_bus = cmplx_cfg.ext_phy_bus[offset];
                         nae_port->loopback = cmplx_cfg.loopback[offset];
                         nae_port->rxaui_mode = cmplx_cfg.rxaui_submode[offset];;
-                        //nae_port->msec_enable = cmplx_cfg.msec_port_enable[offset];;
-                        nae_port->msec_enable = 1;
+                        nae_port->msec_enable = cmplx_cfg.msec_port_enable[offset];;
+                        //nae_port->msec_enable = 1;
 
                         if (is_nlm_xlp9xx()|| is_nlm_xlp3xx() || is_nlm_xlp2xx()) {
                                 nae_port->ext_phy_bus = 0;
-- 
1.7.1

