From 1ea150f5db48da7bc08d074ca0a2cdf860ac6c08 Mon Sep 17 00:00:00 2001
From: Alok Agrawat <alok@broadcom.com>
Date: Fri, 21 Jun 2013 18:16:44 -0700
Subject: disbaled print

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/netlogic/lib/netlib/src/netsoc_fdt.c b/arch/mips/netlogic/lib/netlib/src/netsoc_fdt.c
index 1c8f997..87220a5 100644
--- a/arch/mips/netlogic/lib/netlib/src/netsoc_fdt.c
+++ b/arch/mips/netlogic/lib/netlib/src/netsoc_fdt.c
@@ -184,7 +184,7 @@ static int fdt_get_interface_type(void *fdt, char *nae_port_str, int block, int
                 }
                 else if (!strcmp(port_type_str, "interlaken")) {
                         intf_type = DC_ILK;
-prinkt("Interlaker selected \n");
+//prinkt("Interlaker selected \n");
 
                 }
                 else {
-- 
1.7.1

