From d591c4d0faec1c04640a12953db3f4ffa1fb251b Mon Sep 17 00:00:00 2001
From: Alok Agrawat <alok@broadcom.com>
Date: Mon, 24 Jun 2013 17:26:39 -0700
Subject: MACSEC: SCI mask set correctly for xlp9XX.

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/netlogic/lib/netlib/include/netsoc_dev.h b/arch/mips/netlogic/lib/netlib/include/netsoc_dev.h
index 0cdcdc5..0fa47fc 100644
--- a/arch/mips/netlogic/lib/netlib/include/netsoc_dev.h
+++ b/arch/mips/netlogic/lib/netlib/include/netsoc_dev.h
@@ -302,8 +302,8 @@ RX_MSEC_SEC_TAG2		= 0x409,
 
 #define RX_MSEC_SCI_MASK_LO(i) (0x40b + i *2)
 #define RX_MSEC_SCI_MASK_HI(i) (0x40c + i *2)
-#define RX_MSEC_SCI_MASK_LO_9xx(i) (0x40c + i *2)
-#define RX_MSEC_SCI_MASK_HI_9xx(i) (0x40d + i *2)
+#define RX_MSEC_SCI_MASK_LO_9XX(i) (0x40c + i *2)
+#define RX_MSEC_SCI_MASK_HI_9XX(i) (0x40d + i *2)
 
 /*
  *  POE
diff --git a/arch/mips/netlogic/lib/netlib/src/netsoc_nae.c b/arch/mips/netlogic/lib/netlib/src/netsoc_nae.c
index 04e8ecd..7cab413 100644
--- a/arch/mips/netlogic/lib/netlib/src/netsoc_nae.c
+++ b/arch/mips/netlogic/lib/netlib/src/netsoc_nae.c
@@ -3194,8 +3194,8 @@ void __netsoc_msec_set_rx_cam(nae_t *nae, int index, int port_num, uint64_t sci,
 	netsoc_write_nae_reg(nae_base, RX_MSEC_MEM_CTRL_REG, cntrl_reg_val);
  	if(is_nlm_xlp9xx())
  	{
- 		netsoc_write_nae_reg(nae_base, RX_MSEC_SCI_MASK_LO(port_num), sci_mask & 0xffffffff);
- 		netsoc_write_nae_reg(nae_base, RX_MSEC_SCI_MASK_HI(port_num), (sci_mask >> 32) & 0xffffffff);
+ 		netsoc_write_nae_reg(nae_base, RX_MSEC_SCI_MASK_LO_9XX(port_num), sci_mask & 0xffffffff);
+ 		netsoc_write_nae_reg(nae_base, RX_MSEC_SCI_MASK_HI_9XX(port_num), (sci_mask >> 32) & 0xffffffff);
  	}
  	else
  	{
-- 
1.7.1

