From 002fe21fa75c859f2111eb7a67621931bc0eeb80 Mon Sep 17 00:00:00 2001
From: ctseng <ctseng@broadcom.com>
Date: Tue, 25 Jun 2013 14:37:17 -0700
Subject: prevent header file from refining macro which has been defined in other header file(s).

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/netlogic/lib/syslib/include/nlm_hal.h b/arch/mips/netlogic/lib/syslib/include/nlm_hal.h
index 88bc0a7..2ee58f8 100644
--- a/arch/mips/netlogic/lib/syslib/include/nlm_hal.h
+++ b/arch/mips/netlogic/lib/syslib/include/nlm_hal.h
@@ -203,7 +203,9 @@ static inline uint32_t get_xlp3xx_epid(void)
         return epid;
 }
 
+#ifndef DEFAULT_CPU_IO_BASE
 #define DEFAULT_CPU_IO_BASE 0xffffffffb8000000ULL
+#endif
 #define HDR_OFFSET 0x100
 #define BDF2OFFSET(bus, dev, func)  (((bus) << 20) | ((dev) << 15) | ((func) << 12))
 #define XLP_XKPHYS_UNCACHED       0x9000000000000000ULL
-- 
1.7.1

