From 133aef19d28cdbc5be5b673078560665e4bd8cae Mon Sep 17 00:00:00 2001
From: ctseng <ctseng@broadcom.com>
Date: Tue, 25 Jun 2013 17:34:46 -0700
Subject: removed compilation warnings (variable used but not initialized).

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/netlogic/lib/netlib/include/netsoc_libiface.h b/arch/mips/netlogic/lib/netlib/include/netsoc_libiface.h
index a2033a0..b76aca1 100644
--- a/arch/mips/netlogic/lib/netlib/include/netsoc_libiface.h
+++ b/arch/mips/netlogic/lib/netlib/include/netsoc_libiface.h
@@ -180,7 +180,7 @@ static inline void netsoc_api_delay(uint32_t x)
 
 static inline uint32_t netsoc_api_readl(uint64_t base, uint32_t index)
 {
-	uint32_t ret;
+	uint32_t ret = 0;
 
 	if (netsoc_lib_readl != NULL) {
 		ret = netsoc_lib_readl(base, index);
-- 
1.7.1

