From 294b15629091fc7475f53531d23d8a971ff91011 Mon Sep 17 00:00:00 2001
From: Virendra Pathak <vpathak@broadcom.com>
Date: Mon, 15 Jul 2013 17:52:15 +0530
Subject: poe: workaround in poe for enabling brcm_nae app

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/netlogic/lib/netlib/include/netsoc_common.h b/arch/mips/netlogic/lib/netlib/include/netsoc_common.h
index 12e9f54..75b9dc1 100644
--- a/arch/mips/netlogic/lib/netlib/include/netsoc_common.h
+++ b/arch/mips/netlogic/lib/netlib/include/netsoc_common.h
@@ -65,12 +65,14 @@
 
 
 #define DEFAULT_SGMII_CAL_SLOTS         3
-//#define DEFAULT_XAUI_CAL_SLOTS          13
-#define DEFAULT_XAUI_CAL_SLOTS          8
+#define DEFAULT_XAUI_CAL_SLOTS          13
+//#define DEFAULT_XAUI_CAL_SLOTS          8
 #define DEFAULT_ILK_CAL_SLOTS           26
 #define DEFAULT_XFI_CAL_SLOTS          7
 #define DEFAULT_XLAUI_CAL_SLOTS        13
 
+#define DEFAULT_XLP9XX_XAUI_PARSERSEQFIFO_SZ   480
+
 
 
 
@@ -316,6 +318,8 @@ static inline uint32_t get_prsr_seq_fifo_size(uint32_t intf_type)
                 return DEFAULT_SGMII_PARSERSEQFIFO_SZ;
         case XAUI_IF:
         case XLAUI_IF:
+		if (is_nlm_xlp9xx())
+			return DEFAULT_XLP9XX_XAUI_PARSERSEQFIFO_SZ;
                 return DEFAULT_XAUI_PARSERSEQFIFO_SZ;
         case RXAUI_IF:
 		if(is_nlm_xlp9xx())
diff --git a/arch/mips/netlogic/lib/netlib/include/netsoc_poe.h b/arch/mips/netlogic/lib/netlib/include/netsoc_poe.h
index 9152314..88c729e 100644
--- a/arch/mips/netlogic/lib/netlib/include/netsoc_poe.h
+++ b/arch/mips/netlogic/lib/netlib/include/netsoc_poe.h
@@ -163,7 +163,7 @@ static inline void __netsoc_v1_config_poedin_spill(poe_t *poe, uint32_t class_nu
 	uint64_t poebase = poe->pcie_base;
 	uint32_t  spillbase_lo, spillbase_hi;
 
-        spillbase >>= 6;
+        //spillbase >>= 6;
 	spillbase_lo = spillbase & 0xFFFFFFFF;
 	spillbase_hi = (spillbase >> 32) & 0xFF;
 
@@ -171,7 +171,8 @@ static inline void __netsoc_v1_config_poedin_spill(poe_t *poe, uint32_t class_nu
 	netsoc_write_poe_pcie_reg(poebase, (POE_DIN_SPILL_BASELO_C0 + (class_num * 2)), spillbase_lo);
         netsoc_write_poe_pcie_reg(poebase, (POE_DIN_SPILL_BASEHI_C0 + (class_num * 2)), spillbase_hi);
 
-	netsoc_write_poe_pcie_reg(poebase, POE_DIN_SPILL_MAXLINE_C0 + class_num, (spillsz >> 6));
+	netsoc_write_poe_pcie_reg(poebase, POE_DIN_SPILL_MAXLINE_C0 + class_num, 0x2);
+	//netsoc_write_poe_pcie_reg(poebase, POE_DIN_SPILL_MAXLINE_C0 + class_num, (spillsz >> 6));
 	
 } 
 
diff --git a/arch/mips/netlogic/lib/netlib/src/netsoc_poe.c b/arch/mips/netlogic/lib/netlib/src/netsoc_poe.c
index cceb10a..2d9adf6 100644
--- a/arch/mips/netlogic/lib/netlib/src/netsoc_poe.c
+++ b/arch/mips/netlogic/lib/netlib/src/netsoc_poe.c
@@ -378,7 +378,7 @@ int __netsoc_init_poe(poe_t *poe)
         	__netsoc_init_poe_enq_deq_spill(poe, poe->poe_address_map, poe->poe_address_len);
 
 	if (is_nlm_xlp9xx()) {
-		__netsoc_init_poe_din_spill(poe, poe->din_spill_base, poe->din_spill_size, 0x8); 
+		__netsoc_init_poe_din_spill(poe, poe->din_spill_base, poe->din_spill_size, 0x28); 
 	}
 
         if (poe->dist_en == 1) {
-- 
1.7.1

