From 332e3a5d786d7f673262aa74d58023d621347430 Mon Sep 17 00:00:00 2001
From: Wei Zhang <wezhang@broadcom.com>
Date: Mon, 15 Jul 2013 09:42:07 -0700
Subject: 9xx : fix compile fail. xlp9xx_get_node_id need direct cop0 access

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/netlogic/lib/syslib/include/nlm_hal.h b/arch/mips/netlogic/lib/syslib/include/nlm_hal.h
index be8e7f4..194e8e2 100644
--- a/arch/mips/netlogic/lib/syslib/include/nlm_hal.h
+++ b/arch/mips/netlogic/lib/syslib/include/nlm_hal.h
@@ -212,7 +212,7 @@ static inline uint32_t get_xlp3xx_epid(void)
 
 #define BRIDGE BDF2OFFSET(0, 0, 0)
 
-#ifdef NLM_HAL_LINUX_KERNEL
+#if defined(NLM_HAL_LINUX_KERNEL) || defined(NLM_HAL_UBOOT)
 inline static u32 xlp9xx_get_node_id(void)
 {
         u32 node = 0;
-- 
1.7.1

