From c26fea76c2c2386082d21f6fc1dd79da682c8d38 Mon Sep 17 00:00:00 2001
From: Jayanthi A <jayanthi.annadurai@broadcom.com>
Date: Tue, 23 Jul 2013 09:55:05 -0700
Subject: NETSOC: poe destination threshold configured

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/netlogic/lib/netlib/src/netsoc_poe.c b/arch/mips/netlogic/lib/netlib/src/netsoc_poe.c
index 643d3b3..5673bdc 100644
--- a/arch/mips/netlogic/lib/netlib/src/netsoc_poe.c
+++ b/arch/mips/netlogic/lib/netlib/src/netsoc_poe.c
@@ -399,6 +399,7 @@ int __netsoc_init_poe(poe_t *poe)
 			xlp_config_dist_table(poe);
 		}
         }
+	netsoc_write_poe_pcie_reg(poe->pcie_base, POE_DEST_THRESHOLD, poe->dest_threshold);
 
         if (poe->dist_drop_enable) {
                 netsoc_api_print(NETSOC_APIDBG_DEFAULT, "enabling poe drop based on dist vec\n");
-- 
1.7.1

