From a620a5f756deac4c3343ce040a853c5f72ac83b5 Mon Sep 17 00:00:00 2001
From: Sreenidhi B R <sreenira@broadcom.com>
Date: Fri, 2 Aug 2013 10:43:57 +0530
Subject: netlib: compilation fix.

exporting symbol needed by halx.
[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/netlogic/lib/netlib/include/export_sym.h b/arch/mips/netlogic/lib/netlib/include/export_sym.h
index acfe777..fc656a6 100644
--- a/arch/mips/netlogic/lib/netlib/include/export_sym.h
+++ b/arch/mips/netlogic/lib/netlib/include/export_sym.h
@@ -62,6 +62,7 @@ EXPORT_SYMBOL(netsoc_disable_nae_hwparser);
 EXPORT_SYMBOL(netsoc_enable_nae_hwparser);
 EXPORT_SYMBOL(netsoc_config_nae_l4parser);
 EXPORT_SYMBOL(netsoc_config_nae_l3parser);
+EXPORT_SYMBOL(netsoc_config_nae_l2parser_perport);
 EXPORT_SYMBOL(netsoc_lib_writel);
 EXPORT_SYMBOL(netsoc_lib_readl);
 
diff --git a/arch/mips/netlogic/lib/netlib/include/netsoc_libiface.h b/arch/mips/netlogic/lib/netlib/include/netsoc_libiface.h
index 12c5f8f..2f528a7 100644
--- a/arch/mips/netlogic/lib/netlib/include/netsoc_libiface.h
+++ b/arch/mips/netlogic/lib/netlib/include/netsoc_libiface.h
@@ -516,6 +516,7 @@ extern int netsoc_mdio_read(net_port_t *netport, uint32_t type, uint32_t phyaddr
 extern int netsoc_enable_nae_hwparser(nae_t *nae);
 extern int netsoc_disable_nae_hwparser(nae_t *nae);
 extern int netsoc_config_nae_l2parser(nae_t *nae, l2_parser_config_t *l2);
+extern int netsoc_config_nae_l2parser_perport(nae_t *nae, l2_parser_config_t *l2, uint32_t portnum);
 extern int netsoc_config_nae_l3parser(nae_t *nae, l3_parser_config_t *l3_cfg, int entry);
 extern int netsoc_config_nae_l4parser(nae_t *nae, l4_parser_config_t *l4_cfg, int entry);
 extern int netsoc_build_crc32_table(nae_t *nae, uint32_t *crc32_tbl);
-- 
1.7.1

