From b7f4fc6686ceb48c9d2fbe4fe8c9905e4fe091fb Mon Sep 17 00:00:00 2001
From: Jayanthi A <jayanthi.annadurai@broadcom.com>
Date: Fri, 26 Jul 2013 10:08:22 -0700
Subject: netlib: added API to receive 3 entry message from netsoc

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/netlogic/lib/netlib/include/export_sym.h b/arch/mips/netlogic/lib/netlib/include/export_sym.h
index fc656a6..06ebeda 100644
--- a/arch/mips/netlogic/lib/netlib/include/export_sym.h
+++ b/arch/mips/netlogic/lib/netlib/include/export_sym.h
@@ -56,7 +56,7 @@ EXPORT_SYMBOL(netsoc_get_frame_size);
 EXPORT_SYMBOL(netsoc_get_flowid);
 EXPORT_SYMBOL(netsoc_build_crc32_table);
 EXPORT_SYMBOL(netsoc_crc16);
-
+EXPORT_SYMBOL(netsoc_config_nae_l2parser_perport);
 EXPORT_SYMBOL(netsoc_config_nae_l2parser);
 EXPORT_SYMBOL(netsoc_disable_nae_hwparser);
 EXPORT_SYMBOL(netsoc_enable_nae_hwparser);
diff --git a/arch/mips/netlogic/lib/netlib/include/netsoc_libiface.h b/arch/mips/netlogic/lib/netlib/include/netsoc_libiface.h
index 2f528a7..0e2a71e 100644
--- a/arch/mips/netlogic/lib/netlib/include/netsoc_libiface.h
+++ b/arch/mips/netlogic/lib/netlib/include/netsoc_libiface.h
@@ -515,6 +515,7 @@ extern int netsoc_mdio_write(net_port_t *netport, uint32_t type, uint32_t phyadd
 extern int netsoc_mdio_read(net_port_t *netport, uint32_t type, uint32_t phyaddr, uint32_t regidx);
 extern int netsoc_enable_nae_hwparser(nae_t *nae);
 extern int netsoc_disable_nae_hwparser(nae_t *nae);
+extern int netsoc_config_nae_l2parser_perport(nae_t *nae, l2_parser_config_t *l2, uint32_t port);
 extern int netsoc_config_nae_l2parser(nae_t *nae, l2_parser_config_t *l2);
 extern int netsoc_config_nae_l2parser_perport(nae_t *nae, l2_parser_config_t *l2, uint32_t portnum);
 extern int netsoc_config_nae_l3parser(nae_t *nae, l3_parser_config_t *l3_cfg, int entry);
diff --git a/arch/mips/netlogic/lib/netlib/include/netsoc_msgiface.h b/arch/mips/netlogic/lib/netlib/include/netsoc_msgiface.h
index 5dded0d..fc0818e 100644
--- a/arch/mips/netlogic/lib/netlib/include/netsoc_msgiface.h
+++ b/arch/mips/netlogic/lib/netlib/include/netsoc_msgiface.h
@@ -86,6 +86,37 @@ static __inline__ int netsoc_fmn_recv_2(uint32_t rxvc, uint32_t *node, uint32_t
         return size;
 }
 
+static __inline__ int netsoc_fmn_recv_3(uint32_t rxvc, uint32_t *node, uint32_t *nae_id, uint64_t *msg0, uint64_t *msg1, uint64_t *msg2)
+{
+        uint32_t src_id, size, code;
+        int rx_status = 0;
+
+
+        if(xlp_message_receive_3(rxvc,
+                              &src_id,
+                              &size,
+                              &code,
+                              msg0,
+                              msg1,
+			      msg2
+                              )) {
+
+                rx_status = xlp_read_rx_status();
+
+                if ((rx_status >> 28) & (1 << rxvc)) {
+                }
+                return 0;
+
+        }
+#ifdef NETSOC_MSG_DEBUG 
+        printf("src_id %d size %d code %d msg0 0x%llx msg1 0x%llx msg2 0x%llx\n",
+                         src_id, size, code, msg0, msg1, msg2);
+#endif  
+        *node = GET_NODE(src_id);
+        *nae_id = GET_BLOCK_ID(src_id);
+        return size;
+}
+
 /*
         To send a freein buffer to NAE free in descriptor queue
         destvc: FMN NAE-RX VC number
diff --git a/arch/mips/netlogic/lib/netlib/include/netsoc_parser.h b/arch/mips/netlogic/lib/netlib/include/netsoc_parser.h
index fe27352..7966e6a 100644
--- a/arch/mips/netlogic/lib/netlib/include/netsoc_parser.h
+++ b/arch/mips/netlogic/lib/netlib/include/netsoc_parser.h
@@ -38,8 +38,8 @@
 #define XLP_HWPAR_L3HDR_LEN     16  //L3 header extract length
 #define XLP_HWPAR_MAX_L3HDR     3   //number of areas of L3 header can be extracted
 
-#define XLP_HWPAR_L4HDR_LEN     2   //L4 header extract length
-#define XLP_HWPAR_MAX_L4HDR     16   //number of areas of L4 header can be extracted 
+#define XLP_HWPAR_L4HDR_LEN     16   //L4 header extract length
+#define XLP_HWPAR_MAX_L4HDR     2   //number of areas of L4 header can be extracted 
 
 
 
-- 
1.7.1

