From 26cf4f1c241e928bf874ea57e0107682f9bd042b Mon Sep 17 00:00:00 2001
From: Vikas Gupta <vikas.gupta@broadcom.com>
Date: Wed, 7 Aug 2013 15:10:57 +0530
Subject: xlp9xx: Disable comma bypass for SGMII.

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/netlogic/lib/netlib/src/netsoc_pcs.c b/arch/mips/netlogic/lib/netlib/src/netsoc_pcs.c
index 0059c98..2aea9c7 100644
--- a/arch/mips/netlogic/lib/netlib/src/netsoc_pcs.c
+++ b/arch/mips/netlogic/lib/netlib/src/netsoc_pcs.c
@@ -351,7 +351,7 @@ static void xlp9xx_lane_reset(nae_t *nae, int block, int lane_ctrl, int mode)
 	}
 	
 	if(mode==PHYMODE_SGMII || mode==PHYMODE_HS_SGMII){
-        	val = (1<<20) | (1<<29) | (1<<31);
+        	val = (1<<20) | (1<<29);
 	}else{
         	val = (1<<20) | (1<<29) | (1<<31);
 	}
-- 
1.7.1

