From 492d1846ed9c6e2f357a3044ebd39727136bde77 Mon Sep 17 00:00:00 2001
From: Jayanthi A <jayanthi.annadurai@broadcom.com>
Date: Thu, 8 Aug 2013 13:50:40 -0700
Subject: netlib: Removed module_init and module_exit

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/netlogic/lib/netlib/include/netsoc_msgiface.h b/arch/mips/netlogic/lib/netlib/include/netsoc_msgiface.h
index 0aaa4f1..fc0818e 100644
--- a/arch/mips/netlogic/lib/netlib/include/netsoc_msgiface.h
+++ b/arch/mips/netlogic/lib/netlib/include/netsoc_msgiface.h
@@ -117,37 +117,6 @@ static __inline__ int netsoc_fmn_recv_3(uint32_t rxvc, uint32_t *node, uint32_t
         return size;
 }
 
-static __inline__ int netsoc_fmn_recv_3(uint32_t rxvc, uint32_t *node, uint32_t *nae_id, uint64_t *msg0, uint64_t *msg1, uint64_t *msg2)
-{
-        uint32_t src_id, size, code;
-        int rx_status = 0;
-
-
-        if(xlp_message_receive_3(rxvc,
-                              &src_id,
-                              &size,
-                              &code,
-                              msg0,
-                              msg1,
-			      msg2
-                              )) {
-
-                rx_status = xlp_read_rx_status();
-
-                if ((rx_status >> 28) & (1 << rxvc)) {
-                }
-                return 0;
-
-        }
-#ifdef NETSOC_MSG_DEBUG 
-        printf("src_id %d size %d code %d msg0 0x%llx msg1 0x%llx msg2 0x%llx\n",
-                         src_id, size, code, msg0, msg1, msg2);
-#endif  
-        *node = GET_NODE(src_id);
-        *nae_id = GET_BLOCK_ID(src_id);
-        return size;
-}
-
 /*
         To send a freein buffer to NAE free in descriptor queue
         destvc: FMN NAE-RX VC number
diff --git a/arch/mips/netlogic/lib/netlib/src/netsoc_api.c b/arch/mips/netlogic/lib/netlib/src/netsoc_api.c
index 204a635..000b471 100644
--- a/arch/mips/netlogic/lib/netlib/src/netsoc_api.c
+++ b/arch/mips/netlogic/lib/netlib/src/netsoc_api.c
@@ -1703,22 +1703,4 @@ void brcm_netsoc_lib_init(struct netsoc_lib_param *libparam)
 
 #ifdef NLM_HAL_LINUX_KERNEL
 #include "export_sym.h"
-
-extern void *fdt;
-static int __init brcm_hal_netsoc_init(void)
-{
-	return 0;
-}
-
-static void __init brcm_hal_netsoc_exit(void)
-{
-
-}
-module_init(brcm_hal_netsoc_init);
-module_exit(brcm_hal_netsoc_exit);
-
-MODULE_LICENSE("Proprietary");
-MODULE_AUTHOR("Broadcom");
-MODULE_DESCRIPTION("Broadcom Hal interface");
-
 #endif
-- 
1.7.1

