From 16cfa7380c8074890de45760a0c553ba710e865f Mon Sep 17 00:00:00 2001
From: Xiaoyi Liu <xiaoyi@broadcom.com>
Date: Thu, 22 Aug 2013 02:10:29 -0400
Subject: saesoc_lib: in case of designer freeback, the regular freeback should not be included in calculation of pkt desc len.

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/netlogic/lib/seclib/saesoc_lib.h b/arch/mips/netlogic/lib/seclib/saesoc_lib.h
index 546c3a6..bc77cfe 100644
--- a/arch/mips/netlogic/lib/seclib/saesoc_lib.h
+++ b/arch/mips/netlogic/lib/seclib/saesoc_lib.h
@@ -1853,10 +1853,13 @@ static inline int saesoc_process_request(
 			return -CRYPTOSOC_EINVAL;
 		}
 
-		fmnentry2 = mparam != NULL ? mparam->async_rsp_arg : 0ULL;
-		ddesc->descs[nddescs] = fmnentry2;
-		rsp_msg_ptr = &ddesc->descs[nddescs];
-		nddescs++;
+                /* Designer freeback is only enabled in asyn mode */
+                if (dpriv->ndesfb_descs == 0) {
+		    fmnentry2 = mparam != NULL ? mparam->async_rsp_arg : 0ULL;
+		    ddesc->descs[nddescs] = fmnentry2;
+		    rsp_msg_ptr = &ddesc->descs[nddescs];
+	            nddescs++;
+                }
 
 		fmnentry0 = saesoc_gen2_form_fmn_entry0(rsp_vc, dpriv->ndesfb_descs, 
 			dpriv->ndesfb_descs == 0 ? 0:1, nsdescs, 
-- 
1.7.1

