From aec3ed659a6dedad2e1db56ec09e6e42e673fce2 Mon Sep 17 00:00:00 2001
From: Sreenidhi B R <sreenira@broadcom.com>
Date: Wed, 11 Sep 2013 04:09:28 -0700
Subject: debug: Order of debug print variables fixed in fmnlib.

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/netlogic/lib/fmnlib/nlm_hal_fmn_config.c b/arch/mips/netlogic/lib/fmnlib/nlm_hal_fmn_config.c
index 300c90a..a48e1dc 100644
--- a/arch/mips/netlogic/lib/fmnlib/nlm_hal_fmn_config.c
+++ b/arch/mips/netlogic/lib/fmnlib/nlm_hal_fmn_config.c
@@ -1042,8 +1042,8 @@ int nlm_hal_setup_outq(int node, int max_nodes)
 #ifdef FMN_DEBUG
 				nlm_print("qid: %d, q_ram_base: %d, q_ram_start_page: %d, q_ram_end_page: %d.\n",
 						qid, (q_ram_base & 0x1f),
-						(q_ram_end_page & 0x3f),
-						(q_ram_start_page & 0x3f));
+						(q_ram_start_page & 0x3f),
+						(q_ram_end_page & 0x3f));
 #endif
 
 				q_ram_start_page = q_ram_end_page +1;
-- 
1.7.1

