From ca8ce577db4cbe01da71187662cb5c79daa60cb3 Mon Sep 17 00:00:00 2001
From: Hareesh R <hareeshr@broadcom.com>
Date: Wed, 18 Sep 2013 01:28:31 -0700
Subject: netlib: Moved linux shinfo to a common file

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/netlogic/lib/netlib/include/nlm_nae.h b/arch/mips/netlogic/lib/netlib/include/nlm_nae.h
index db1eca9..818459e 100644
--- a/arch/mips/netlogic/lib/netlib/include/nlm_nae.h
+++ b/arch/mips/netlogic/lib/netlib/include/nlm_nae.h
@@ -69,6 +69,8 @@
 
 #define NLM_NAE_LNX_SHINFO_FL_VALID     0x01
 #define NLM_NAE_LNX_SHINFO_FL_JUMBO_EN  0x02
+#define NLM_NAE_LNX_SHINFO_FL_VER_1     0x04
+#define NLM_NAE_LNX_SHINFO_FL_ALE_BYPASS     0x08
 
 #define NLM_NAE_LNX_SHINFO_FWD_INFO_LCPU_2_PCPU_OFF 0
 #define NLM_NAE_LNX_SHINFO_FWD_INFO_LCPU_2_PCPU_MASK 0xff
@@ -92,6 +94,23 @@ struct nlm_nae_shinfo {
 	unsigned int *fwd_info;
 };
 
+/*
+ * This strucutre has been used to communicate the nae 
+ * shared info between ucore and the application/OS
+ */
+struct nlm_nae_linux_shinfo {
+	unsigned int flags;
+	int rxvc;
+	int domid;
+	int mode;
+	int node;
+	int naeid;
+	int num_cpus;
+	/* variable sizes depending on the num cpus */
+	unsigned int fwd_info[0]; 
+};
+
+
 enum higig_mode{
 	NO_HIGIG,
 	HIGIG,
-- 
1.7.1

