From 60627daabb9b023c06c901760a4c6d02b5caeb7e Mon Sep 17 00:00:00 2001
From: Virendra Pathak <vpathak@broadcom.com>
Date: Thu, 26 Sep 2013 11:29:26 +0530
Subject: syslib: Added APIs for extrcating cpu info for all chip types

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/netlogic/lib/syslib/include/brcm_xlp_cpu.h b/arch/mips/netlogic/lib/syslib/include/brcm_xlp_cpu.h
new file mode 100644
index 0000000..674d1f2
--- /dev/null
+++ b/arch/mips/netlogic/lib/syslib/include/brcm_xlp_cpu.h
@@ -0,0 +1,61 @@
+
+/*-
+ * Copyright (c) 2003-2012 Broadcom Corporation
+ * All Rights Reserved
+ *
+ * This software is available to you under a choice of one of two
+ * licenses.  You may choose to be licensed under the terms of the GNU
+ * General Public License (GPL) Version 2, available from the file
+ * http://www.gnu.org/licenses/gpl-2.0.txt
+ * or the Broadcom license below:
+
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in
+ *    the documentation and/or other materials provided with the
+ *    distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY BROADCOM ``AS IS'' AND ANY EXPRESS OR
+ * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL BROADCOM OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR
+ * BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+ * WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE
+ * OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN
+ * IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ * #BRCM_4# */
+
+#ifndef _BRCM_XLP_CPU_H
+#define _BRCM_XLP_CPU_H
+
+#define DEFAULT_NODE_NUM	1
+#define NLM_MAX_NODES		4
+#define MAX_VC_PERTHREAD	4
+
+#ifndef NLM_NCPUS_PER_NODE 
+#define NLM_NCPUS_PER_NODE	32
+#endif
+
+#define XLP9XX_MAX_NODES		8
+#define XLP9XX_MAX_CPU_PER_NODE   	80
+
+#define XLP3XX_MAX_CPU_PER_NODE 	16
+#define XLP2XX_MAX_CPU_PER_NODE 	8
+
+#ifndef __ASSEMBLY__
+
+inline int brcm_get_max_cpu_num(void);
+inline int brcm_get_max_node_num(void);
+
+#endif /* __ASSEMBLY__ */
+
+#endif /* #ifndef _BRCM_XLP_CPU_H */
diff --git a/arch/mips/netlogic/lib/syslib/include/nlm_hal_xlp_dev.h b/arch/mips/netlogic/lib/syslib/include/nlm_hal_xlp_dev.h
index 854799e..3f16499 100644
--- a/arch/mips/netlogic/lib/syslib/include/nlm_hal_xlp_dev.h
+++ b/arch/mips/netlogic/lib/syslib/include/nlm_hal_xlp_dev.h
@@ -38,15 +38,9 @@
 #ifndef NLM_HAL_XLP_DEV_H
 #define NLM_HAL_XLP_DEV_H
 
-#define NLM_MAX_NODES           4
-
-#define MAX_VC_PERTHREAD        4
+#include <brcm_xlp_cpu.h>
 
 #define XLP_CACHELINE_SIZE	64
-
-#ifndef NLM_NCPUS_PER_NODE
-#define NLM_NCPUS_PER_NODE	32
-#endif
 /*
  * This File has all the XLP Device specific Defines
  */
diff --git a/arch/mips/netlogic/lib/syslib/src/brcm_xlp_cpu.c b/arch/mips/netlogic/lib/syslib/src/brcm_xlp_cpu.c
new file mode 100644
index 0000000..c023ad2
--- /dev/null
+++ b/arch/mips/netlogic/lib/syslib/src/brcm_xlp_cpu.c
@@ -0,0 +1,92 @@
+
+/*-
+ * Copyright (c) 2003-2012 Broadcom Corporation
+ * All Rights Reserved
+ *
+ * This software is available to you under a choice of one of two
+ * licenses.  You may choose to be licensed under the terms of the GNU
+ * General Public License (GPL) Version 2, available from the file
+ * http://www.gnu.org/licenses/gpl-2.0.txt
+ * or the Broadcom license below:
+
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in
+ *    the documentation and/or other materials provided with the
+ *    distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY BROADCOM ``AS IS'' AND ANY EXPRESS OR
+ * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL BROADCOM OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR
+ * BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+ * WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE
+ * OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN
+ * IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ * #BRCM_4# */
+
+#include <brcm_xlp_cpu.h>
+#include <nlm_hal_xlp_dev.h>
+
+/**
+ * @brief brcm_get_max_cpu_num - return the total number of cpus on a chip
+ *
+ * @param [in]	- none
+ *
+ * @return
+ * 	  - total number of cpus on a chip
+ *
+ * @see brcm_get_relative_cpu_num
+ **/
+
+inline int brcm_get_max_cpu_num(void)
+{
+	int max_cpu = 0;
+
+	if (is_nlm_xlp9xx()) {
+		max_cpu = XLP9XX_MAX_CPU_PER_NODE;
+	} else if (is_nlm_xlp8xx()) {
+		max_cpu = NLM_NCPUS_PER_NODE;
+	} else if (is_nlm_xlp3xx()) {
+		max_cpu = XLP3XX_MAX_CPU_PER_NODE;
+	} else if (is_nlm_xlp2xx()) {
+		max_cpu = XLP2XX_MAX_CPU_PER_NODE;
+	} else {
+		max_cpu = NLM_NCPUS_PER_NODE;
+	}
+	return max_cpu;
+}
+
+/**
+ * @brief brcm_get_max_node_num - return the total number of nodes
+ *
+ * @param [in]	- none
+ *
+ * @return
+ * 	  - total number of nodes on a chip
+ *
+ * @see brcm_get_node_num
+ **/
+
+inline int brcm_get_max_node_num(void)
+{
+	int max_node = 0;
+	
+	if (is_nlm_xlp9xx()) {
+		max_node = XLP9XX_MAX_NODES;
+	} else if (is_nlm_xlp8xx()) {
+		max_node = NLM_MAX_NODES;
+	} else {
+		max_node = DEFAULT_NODE_NUM;
+	}
+	return max_node;
+}
-- 
1.7.1

