From 0a6002fabc3cac894d243d0d3254e59b1c31c847 Mon Sep 17 00:00:00 2001
From: Jayanthi A <jayanthi.annadurai@broadcom.com>
Date: Fri, 4 Oct 2013 21:58:51 -0700
Subject: syslib: Fixed xlp9xx system block base address

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/netlogic/lib/syslib/include/nlm_hal_xlp_dev.h b/arch/mips/netlogic/lib/syslib/include/nlm_hal_xlp_dev.h
index 3f16499..9015c18 100644
--- a/arch/mips/netlogic/lib/syslib/include/nlm_hal_xlp_dev.h
+++ b/arch/mips/netlogic/lib/syslib/include/nlm_hal_xlp_dev.h
@@ -497,6 +497,7 @@ extern int is_nlm_xlp(unsigned int chipid, unsigned int rev,  unsigned int ext);
 #define XLP_SYS_DEVICE			0x6
 #define XLP_SYS_FUNC			0x5
 
+#define XLP9XX_SYS_BUS			0x1
 #define XLP9XX_SYS_DEVICE		0x6
 #define XLP9XX_SYS_FUNC			0x0
 
diff --git a/arch/mips/netlogic/lib/syslib/src/nlm_hal.c b/arch/mips/netlogic/lib/syslib/src/nlm_hal.c
index e9e5bb6..43cfbc5 100644
--- a/arch/mips/netlogic/lib/syslib/src/nlm_hal.c
+++ b/arch/mips/netlogic/lib/syslib/src/nlm_hal.c
@@ -408,7 +408,7 @@ __inline__ void nlm_hal_init(void)
 
 			xlp_poe_base_pcim[node] = (uint64_t)xlp9xx_cpu_mem_base(-1, XLP9XX_PCIE_DEV_BASE(XLP_POE));
 			xlp_poe_base_pcie[node] = (xlp_io_base | XLP9XX_PCIE_DEV_BASE(XLP_POE)) & 0x1fffffff; /* For now . Will be fixed soon.*/
-			xlp_sys_base[node] = (xlp_io_base | XLP9XX_PCIE_DEV_BASE( XLP_SYS)) & 0x1fffffff; /*For now . Will be fixed soon.*/
+			xlp_sys_base[node] = (xlp_io_base | (XLP9XX_SYS_BUS << 20) | XLP9XX_PCIE_DEV_BASE(XLP9XX_SYS)) & 0x1fffffff; /*For now . Will be fixed soon.*/
 
 			xlp_sae_base[node] = (xlp_io_base | XLP9XX_SAE_RSA_BUS_NUM | XLP9XX_PCIE_DEV_BASE(XLP_SAE)) & 0x1fffffff; /*For now . Will be fixed soon.*/
 			xlp_rsa_base[node] = (xlp_io_base | XLP9XX_SAE_RSA_BUS_NUM | XLP9XX_PCIE_DEV_BASE(XLP_RSA)) & 0x1fffffff; /* For now . Will be fixed soon.*/
-- 
1.7.1

