From 300560bbea39eb1b55fcd98ab3d56dd5a2856534 Mon Sep 17 00:00:00 2001
From: Jayanthi A <jayanthi.annadurai@broadcom.com>
Date: Wed, 9 Oct 2013 06:36:54 -0700
Subject: fmnlib: Fixed xlp9xx credit configuration

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/netlogic/lib/fmnlib/nlm_hal_fmn_config.c b/arch/mips/netlogic/lib/fmnlib/nlm_hal_fmn_config.c
index 06fdb08..68cd3ae 100644
--- a/arch/mips/netlogic/lib/fmnlib/nlm_hal_fmn_config.c
+++ b/arch/mips/netlogic/lib/fmnlib/nlm_hal_fmn_config.c
@@ -752,6 +752,7 @@ static int  fmn_9xx_is_vc_valid(int node, unsigned int base_vc)
 	return 0;
 }
 
+
 static void nlm_hal_write_fmn_credit(int node, int max_nodes) 
 {
 	int src, qid, hndl = 0, dst_node;
@@ -775,6 +776,7 @@ static void nlm_hal_write_fmn_credit(int node, int max_nodes)
 		fmn_modify_qsize_credit_config(node, XLP_MSG_BLK_RSA_ECC, 1, XLP_9XX_RSA_VC_BASE, XLP_9XX_RSA_VC_LIMIT);
 		fmn_modify_qsize_credit_config(node, XLP_MSG_BLK_CRYPTO, 1, XLP_9XX_SEC_VC_BASE, XLP_9XX_SEC_VC_LIMIT);
 
+
 		/* go through each valid station */
 		for(hndl = 0; hndl < XLP_MSG_HANDLE_MAX; hndl++) {
 			int base_vc = fmn_config->base_vc;
@@ -793,7 +795,7 @@ static void nlm_hal_write_fmn_credit(int node, int max_nodes)
 
 						for (; qid <= xlp9xx_fmn_config[station].vc_limit; qid++) {
 							/* FIXME: define/config fmn_cfg_value properly for XLP9XX */
-#if 0
+#if 1
 							credits = fmn_cfg_value[qid] - 1;
 #else
 							credits = 50;
-- 
1.7.1

