From 0ec674839b99e8c85276a4d960dd7fd181294c5b Mon Sep 17 00:00:00 2001
From: Jayanthi A <jayanthi.annadurai@broadcom.com>
Date: Wed, 9 Oct 2013 06:39:16 -0700
Subject: netlib: Enabled xlp9xx rxbuffer drop mode

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/netlogic/lib/netlib/src/netsoc_nae.c b/arch/mips/netlogic/lib/netlib/src/netsoc_nae.c
index c574c3f..0bab60f 100644
--- a/arch/mips/netlogic/lib/netlib/src/netsoc_nae.c
+++ b/arch/mips/netlogic/lib/netlib/src/netsoc_nae.c
@@ -1236,6 +1236,8 @@ void __netsoc_config_rxbuffer(nae_t *nae, int context_base, int num_channels, ui
 				context_base + offset, base, size, thrgrp);
 		val = 0x80000000 | ((base << 2) & 0x3fff); /* base */
                 val |= (((size << 2)  & 0x3fff) << 16); /* size */
+		if (is_nlm_xlp9xx()) 
+			val |= (1<<15);
                 netsoc_write_nae_reg(nae_base, RX_BUFFER_BASE_DEPTH_REG, val);
 
 		// Read back
-- 
1.7.1

