From dc1c46a0f9e3d0efc33376edfa411d38be8ec4fe Mon Sep 17 00:00:00 2001
From: Prasanna Mumbai <prasanna.mumbai@broadcom.com>
Date: Thu, 10 Oct 2013 16:48:51 +0530
Subject: FMN: use node id for multinode case

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/netlogic/lib/fmnlib/nlm_hal_fmn_config.c b/arch/mips/netlogic/lib/fmnlib/nlm_hal_fmn_config.c
index 68cd3ae..52aab90 100644
--- a/arch/mips/netlogic/lib/fmnlib/nlm_hal_fmn_config.c
+++ b/arch/mips/netlogic/lib/fmnlib/nlm_hal_fmn_config.c
@@ -800,7 +800,7 @@ static void nlm_hal_write_fmn_credit(int node, int max_nodes)
 #else
 							credits = 50;
 #endif
-							nlh_write_cfg_reg64(xlp_fmn_base[0] + (1 << 14), 
+							nlh_write_cfg_reg64(xlp_fmn_base[node] + (1 << 14),
 								(0ULL << 42) /* type, credit value + 1 */
 								| (credits << 26) /* credit count */
 								| ((qid | (dst_node << 10)) << 13) /* destination */
-- 
1.7.1

