From a6d8434dbe9ac56005525c453d5af69117ef7fa0 Mon Sep 17 00:00:00 2001
From: Jayanthi A <jayanthi.annadurai@broadcom.com>
Date: Thu, 10 Oct 2013 19:46:36 -0700
Subject: netlib: Fixed poe reset. Disabled BIU timeout timer

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/netlogic/lib/netlib/src/netsoc_poe.c b/arch/mips/netlogic/lib/netlib/src/netsoc_poe.c
index 7fb3ff2..fb1dd5f 100644
--- a/arch/mips/netlogic/lib/netlib/src/netsoc_poe.c
+++ b/arch/mips/netlogic/lib/netlib/src/netsoc_poe.c
@@ -368,6 +368,9 @@ int __netsoc_init_poe(poe_t *poe)
         uint64_t poe_pcie_base = poe->pcie_base;
         uint32_t index =0,thrsel_mask = 0;
  
+	if (poe->nae->owned == 0)
+		return 0;
+
 	netsoc_api_print(NETSOC_APIDBG_DEFAULT, "%s\n",__func__);
 #ifndef XLP_SIM
 	__netsoc_reset_poe(poe);
@@ -424,7 +427,7 @@ int __netsoc_init_poe(poe_t *poe)
 	if (is_nlm_xlp9xx()) {
 	        netsoc_write_poe_pcie_reg(poe_pcie_base, POE_TX_TIMER, 0x2);
 		netsoc_write_poe_pcie_reg(poe_pcie_base, 0x184, 0x300);
-		netsoc_write_poe_pcie_reg(poe_pcie_base, 0x183, 0x11);
+		netsoc_write_poe_pcie_reg(poe_pcie_base, 0x183, 0x10);
 	}
 	else {
 		netsoc_write_poe_pcie_reg(poe_pcie_base, POE_TX_TIMER, 0x3);
-- 
1.7.1

