From c10c208256bb0046bcc703ad5685d6bb6a7264e6 Mon Sep 17 00:00:00 2001
From: Jayanthi A <jayanthi.annadurai@broadcom.com>
Date: Fri, 11 Oct 2013 20:47:22 -0700
Subject: netlib: Fixed xaui mode phy configuration

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/netlogic/lib/netlib/src/netsoc_pcs.c b/arch/mips/netlogic/lib/netlib/src/netsoc_pcs.c
index 467c0f9..b373cc0 100644
--- a/arch/mips/netlogic/lib/netlib/src/netsoc_pcs.c
+++ b/arch/mips/netlogic/lib/netlib/src/netsoc_pcs.c
@@ -350,7 +350,7 @@ static void xlp9xx_lane_reset(nae_t *nae, int block, int lane_ctrl, int mode)
 		return; /*Only one lane in complex 2 */
 	}
 	
-	if(mode==PHYMODE_SGMII || mode==PHYMODE_HS_SGMII){
+	if (mode==PHYMODE_SGMII) { // FIXME PHYMODE_HS_SGMII support is yet to be added; || mode==PHYMODE_HS_SGMII){
         	val = (1<<20) | (1<<29);
 	}else{
         	val = (1<<20) | (1<<29) | (1<<31);
-- 
1.7.1

