From 2cc68388a532eabdae5948bf0574e721b95df6ff Mon Sep 17 00:00:00 2001
From: Vikas Gupta <vikas.gupta@broadcom.com>
Date: Wed, 16 Oct 2013 01:57:05 -0700
Subject: xlp9xx: extend the divisor value for GIG MDIO.

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/netlogic/lib/netlib/src/netsoc_nae.c b/arch/mips/netlogic/lib/netlib/src/netsoc_nae.c
index a99df36..5f9de06 100644
--- a/arch/mips/netlogic/lib/netlib/src/netsoc_nae.c
+++ b/arch/mips/netlogic/lib/netlib/src/netsoc_nae.c
@@ -93,7 +93,7 @@ static uint32_t nae_get_EXT_G_MDIO_DIV(void)
 	return (
 	       is_nlm_xlp3xx_ax() ||
 	       is_nlm_xlp8xx_ax() ||
-	       is_nlm_xlp8xx_b0() ) ?  EXT_G_MDIO_DIV : EXT_G_MDIO_DIV_WITH_HW_DIV64_11;
+	       is_nlm_xlp8xx_b0() || is_nlm_xlp9xx()) ?  EXT_G_MDIO_DIV : EXT_G_MDIO_DIV_WITH_HW_DIV64_11;
 }
 
 /* INT_MDIO_CTRL, block7, 0x799
-- 
1.7.1

