From 5c64fed27563e993d4af1045492326a8b18115ce Mon Sep 17 00:00:00 2001
From: Virendra Pathak <vpathak@broadcom.com>
Date: Tue, 22 Oct 2013 11:34:52 +0530
Subject: fmn: added support for configuring popq priority class

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/netlogic/lib/fmnlib/nlm_hal_fmn.h b/arch/mips/netlogic/lib/fmnlib/nlm_hal_fmn.h
index f8987cd..ea7dd49 100644
--- a/arch/mips/netlogic/lib/fmnlib/nlm_hal_fmn.h
+++ b/arch/mips/netlogic/lib/fmnlib/nlm_hal_fmn.h
@@ -54,6 +54,9 @@
 #define FMN_MAX_Q_SIZE (256ULL * 1024)
 #define FMN_Q_PAGE_SIZE (4ULL * 1024)
 
+#define POPQ_CLASS_CONFIG_BASE  	0x4068
+#define POPQ_CLASS_CONFIG_OFFSET	8
+
 enum FMN_MSG_BLKS {
 	XLP_MSG_BLK_CPU = 0, 
 	XLP_MSG_BLK_POPQ, 
@@ -293,6 +296,7 @@ extern void nlm_hal_set_fmn_interrupt(int irq);
 
 extern void nlm_hal_disable_vc_intr(int node, int vc);
 extern void nlm_hal_enable_vc_intr(int node, int vc);
+int brcm_hal_config_popq_class(int node, int popq_class, uint8_t width, uint16_t base_id);
 #endif /* #ifndef _NLH_FMN_H */
 
 	
diff --git a/arch/mips/netlogic/lib/fmnlib/nlm_hal_fmn_config.c b/arch/mips/netlogic/lib/fmnlib/nlm_hal_fmn_config.c
index 52aab90..c818dcf 100644
--- a/arch/mips/netlogic/lib/fmnlib/nlm_hal_fmn_config.c
+++ b/arch/mips/netlogic/lib/fmnlib/nlm_hal_fmn_config.c
@@ -615,6 +615,42 @@ void nlm_hal_disable_vc_intr(int node, int vc)
 	}
 }
 
+/**
+ * @brief brcm_hal_config_popq_class - configure popq priority class
+ *
+ * @param [in]	node-		node number
+ * 		popq_class-	popq priority class number
+ * 		width-		width of Priority class Pop output queue
+ * 		base_id-	Base queue Id for Priority class output queue
+ *
+ * @return
+ * 	  - 0		success
+ * 	  - <0 	error
+ *
+ * @note -  nlm_hal_init should be called before calling this function.
+ *
+ */
+
+int brcm_hal_config_popq_class(int node, int popq_class, uint8_t width, uint16_t base_id)
+{
+	uint64_t val = 0;
+	int offset = 0;
+	
+	if (is_nlm_xlp9xx()) {
+		offset = popq_class * POPQ_CLASS_CONFIG_OFFSET;
+		val |= ((1ULL << 63) | ((0x7f & (~width)) << 16) | (base_id & 0x1fff));
+		nlh_write_cfg_reg64(xlp_fmn_base[node] + (POPQ_CLASS_CONFIG_BASE + offset), val);
+#if FMN_DEBUG
+		nlm_print("%s \n", __func__);
+		val = nlh_read_cfg_reg64(xlp_fmn_base[node] + (POPQ_CLASS_CONFIG_BASE + offset));
+		nlm_print("val: 0x%llx \n", val);
+#endif
+		return 0;
+	} else {
+		return -1;
+	}
+}
+
 /*********************************************************************
  * nlm_hal_set_fmn_interrupt
  *
-- 
1.7.1

