From 18a1949a247dcc0c0ce1e6de73cd4daace152f9e Mon Sep 17 00:00:00 2001
From: ctseng <ctseng@broadcom.com>
Date: Thu, 24 Oct 2013 12:55:21 -0700
Subject: syslib: remove compilation warning from HAL.

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/netlogic/lib/fmnlib/nlm_hal_fmn.h b/arch/mips/netlogic/lib/fmnlib/nlm_hal_fmn.h
index ea7dd49..4f6ce14 100644
--- a/arch/mips/netlogic/lib/fmnlib/nlm_hal_fmn.h
+++ b/arch/mips/netlogic/lib/fmnlib/nlm_hal_fmn.h
@@ -267,7 +267,7 @@ static __inline__ int nlm_hal_enable_rxqdepth(int node, int qid)
         if (is_nlm_xlp9xx()) {
                 //*(uint64_t *)(xlp_fmn_base[node] + (1 << 13) + (vc << 3)) = 0x0;
                 val = nlh_read_cfg_reg64(xlp_fmn_base[node] + (qid << 3));
-                val |= (1<<51);
+                val |= (1UL<<51);
                 nlh_write_cfg_reg64(xlp_fmn_base[node] + (qid << 3), val);
                 return 0;
         }
@@ -282,7 +282,7 @@ static __inline__ int nlm_hal_disable_rxqdepth(int node, int qid)
         if (is_nlm_xlp9xx()) {
                 //*(uint64_t *)(xlp_fmn_base[node] + (1 << 13) + (vc << 3)) = 0x0;
                 val = nlh_read_cfg_reg64(xlp_fmn_base[node] + (qid << 3));
-                val &= ~(1<<51);
+                val &= ~(1UL<<51);
                 nlh_write_cfg_reg64(xlp_fmn_base[node] + (qid << 3), val);
                 return 0;
         }
diff --git a/arch/mips/netlogic/lib/syslib/include/xlp9xx_sys.h b/arch/mips/netlogic/lib/syslib/include/xlp9xx_sys.h
index 70e12c8..8659b19 100644
--- a/arch/mips/netlogic/lib/syslib/include/xlp9xx_sys.h
+++ b/arch/mips/netlogic/lib/syslib/include/xlp9xx_sys.h
@@ -186,7 +186,7 @@ static inline int xlp9xx_set_nae_frequency(int node, uint64_t freq)
     uint32_t val, dev_id = XLP9XX_CLKDEV_NET;
     uint64_t sys_base = XLP9XX_SYS_CLKDEV_BASE(node);
 
-    nlm_print("NAE frequency %lld MHz\n",get_nae_frequency(node) / mhz);
+    nlm_print("NAE frequency %lu MHz\n", get_nae_frequency(node) / mhz);
 		
     if (freq == get_nae_frequency(node))
 	return 0;	
@@ -242,7 +242,7 @@ static inline int xlp9xx_set_nae_frequency(int node, uint64_t freq)
       }
     }
 
-    nlm_print("NAE frequency changed to %lld MHz\n",get_nae_frequency(node) / mhz);
+    nlm_print("NAE frequency changed to %lu MHz\n", get_nae_frequency(node) / mhz);
     return 0;
 }
 
@@ -293,7 +293,7 @@ static inline int set_crypto_block_frequency(int node, int dev_id, uint64_t freq
     else
         block_name = "RSA";
 
-    nlm_print("%s frequency %lld MHz\n", block_name,  
+    nlm_print("%s frequency %lu MHz\n", block_name,  
               get_crypto_block_frequency(node, dev_id) / mhz);
 
     if (freq == get_crypto_block_frequency(node, dev_id))
@@ -341,7 +341,7 @@ static inline int set_crypto_block_frequency(int node, int dev_id, uint64_t freq
       }
     }
 
-    nlm_print("%s frequency changed to %lld MHz\n",block_name, 
+    nlm_print("%s frequency changed to %lu MHz\n",block_name, 
               get_crypto_block_frequency(node, dev_id) / mhz);
     return 0;
 }
-- 
1.7.1

