From 5ec4625bfdabcf8d9cd9eb8a1c49cbeb937fdabb Mon Sep 17 00:00:00 2001
From: Kevin James <kevinj@broadcom.com>
Date: Fri, 25 Oct 2013 04:24:43 -0700
Subject: netlib:Fixed Rx & Tx enable during nae soft reset

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/netlogic/lib/netlib/src/nlm_hal_nae.c b/arch/mips/netlogic/lib/netlib/src/nlm_hal_nae.c
index ff61680..8a84ce8 100644
--- a/arch/mips/netlogic/lib/netlib/src/nlm_hal_nae.c
+++ b/arch/mips/netlogic/lib/netlib/src/nlm_hal_nae.c
@@ -5538,9 +5538,7 @@ int nlm_hal_open_if(int node, int intf_type, int block)
 
 			// Reset GMAC 
 			mac_cfg1 = read_gmac_reg(node, block, MAC_CONF1);
-		        write_gmac_reg(node, block , MAC_CONF1, mac_cfg1 | INF_SOFTRESET(1) |
-					           INF_RX_ENABLE(1) |
-						   INF_TX_ENABLE(1));
+		        write_gmac_reg(node, block , MAC_CONF1, mac_cfg1 | INF_SOFTRESET(1));
 
 			// Default 1G
 		        write_gmac_reg(node, block , MAC_CONF2,  INF_PREMBL_LEN(0x7) |
-- 
1.7.1

