From 7186df0a31e6a682a2ef949712093fc68805eaa3 Mon Sep 17 00:00:00 2001
From: Jayanthi A <jayanthi.annadurai@broadcom.com>
Date: Wed, 23 Oct 2013 19:08:32 -0700
Subject: netlib: Fixed number of nodes

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/netlogic/lib/netlib/src/netsoc_config.c b/arch/mips/netlogic/lib/netlib/src/netsoc_config.c
index 4a9210c..3c7fb8f 100644
--- a/arch/mips/netlogic/lib/netlib/src/netsoc_config.c
+++ b/arch/mips/netlogic/lib/netlib/src/netsoc_config.c
@@ -333,7 +333,7 @@ static void dump_context_config(void)
 
 static int get_max_nodes(void)
 {
-	return 1; //nlm_node_cfg.num_nodes;
+	return nlm_node_cfg.num_nodes;
 }
 
 int __netsoc_restart_ucore_using_fdt(nae_t *nae, void *fdt)
-- 
1.7.1

