From de18422b76e39e531a7a8bbfe79ecb4e07cb87a9 Mon Sep 17 00:00:00 2001
From: Jayanthi A <jayanthi.annadurai@broadcom.com>
Date: Thu, 24 Oct 2013 18:10:46 -0700
Subject: netlib: egress tx fifo max offset is 2046

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/netlogic/lib/netlib/src/netsoc_nae.c b/arch/mips/netlogic/lib/netlib/src/netsoc_nae.c
index 8ea693a..47adc3f 100644
--- a/arch/mips/netlogic/lib/netlib/src/netsoc_nae.c
+++ b/arch/mips/netlogic/lib/netlib/src/netsoc_nae.c
@@ -679,8 +679,8 @@ void __netsoc_config_egress_fifo_carvings(nae_t *nae, int start_ctxt, int num_ct
 	for(i=start_ctxt;i<limit;i++){
 		size   = current_netsoc->egress.max_tx_fifo / nae->num_contexts;
 
-		if(size > current_netsoc->egress.max_tx_offset)
-			offset = current_netsoc->egress.max_tx_offset - 1;
+		if(size > (current_netsoc->egress.max_tx_offset - 2))
+			offset = current_netsoc->egress.max_tx_offset - 2;
 		else
 			offset = size -1;
 		netsoc_write_nae_reg(nae_base, TX_PKT_PMEM_CMD1, offset);
@@ -691,6 +691,8 @@ void __netsoc_config_egress_fifo_carvings(nae_t *nae, int start_ctxt, int num_ct
 		netsoc_api_print(NETSOC_APIDBG_CONFIG, "program_nae_pkt_pmem: ctxt_num:%d start:%d size:%d\n", i, start, offset);
 		netsoc_write_nae_reg(nae_base, TX_PKT_PMEM_CMD0, data);
 		start+= offset + 1;
+		if (start & 0x1)
+			start++;
 	}
 	nae->pktfifo_base = start;
 
-- 
1.7.1

