From be3ff8a2104860c634d22bbf26f0a3eb97e3d58d Mon Sep 17 00:00:00 2001
From: Vikas Gupta <vikas.gupta@broadcom.com>
Date: Wed, 30 Oct 2013 10:38:16 -0700
Subject: syslib: fix for base address.

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/netlogic/lib/syslib/include/nlm_hal_xlp_dev.h b/arch/mips/netlogic/lib/syslib/include/nlm_hal_xlp_dev.h
index da42e13..bd4bd66 100644
--- a/arch/mips/netlogic/lib/syslib/include/nlm_hal_xlp_dev.h
+++ b/arch/mips/netlogic/lib/syslib/include/nlm_hal_xlp_dev.h
@@ -503,7 +503,7 @@ extern int is_nlm_xlp(unsigned int chipid, unsigned int rev,  unsigned int ext);
 
 #define XLP9XX_SYS_FUNC_CLKDEV  2
 
-#define XLP9XX_SYS_CLKDEV_BASE(node)   (KSEG1 +0x18000000 + (XLP9XX_SYS_BUS << 20) + (((node * XLP_MAXDEV_PERNODE) + XLP9XX_SYS_DEVICE) << 15) + (XLP9XX_SYS_FUNC_CLKDEV << 12))
+#define XLP9XX_SYS_CLKDEV_BASE(node)   (0x18000000 + (XLP9XX_SYS_BUS << 20) + (((node * XLP_MAXDEV_PERNODE) + XLP9XX_SYS_DEVICE) << 15) + (XLP9XX_SYS_FUNC_CLKDEV << 12))
 
 #define XLP_MAXDEV_PERNODE		8
 
-- 
1.7.1

