From 1ff25f5227ad102700276423c74437850f71f1c9 Mon Sep 17 00:00:00 2001
From: Virendra Pathak <vpathak@broadcom.com>
Date: Tue, 5 Nov 2013 19:52:55 +0530
Subject: netlib: removed the duplicate definition of node config structure

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/netlogic/lib/netlib/src/nlm_hal_phy.c b/arch/mips/netlogic/lib/netlib/src/nlm_hal_phy.c
index b6e5cac..5c7de3d 100644
--- a/arch/mips/netlogic/lib/netlib/src/nlm_hal_phy.c
+++ b/arch/mips/netlogic/lib/netlib/src/nlm_hal_phy.c
@@ -38,7 +38,6 @@
 #include "fdt_helper.h"
 
 
-struct nlm_node_config  nlm_node_cfg;
 
 static int reg_num_phys;
 void sgmii_scan_phys(int node);
-- 
1.7.1

