From ad12a5cad76e4d2473df1026f5fd634165ec66c4 Mon Sep 17 00:00:00 2001
From: Subhendu Sekhar Behera <sbehera@broadcom.com>
Date: Mon, 4 Nov 2013 10:55:13 +0530
Subject: Fix: use NLM_HAL_DO_DIV macro for 64-bit division.

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/netlogic/lib/syslib/include/xlp9xx_sys.h b/arch/mips/netlogic/lib/syslib/include/xlp9xx_sys.h
index 8659b19..8fc4459 100644
--- a/arch/mips/netlogic/lib/syslib/include/xlp9xx_sys.h
+++ b/arch/mips/netlogic/lib/syslib/include/xlp9xx_sys.h
@@ -182,14 +182,22 @@ static inline int xlp9xx_set_nae_frequency(int node, uint64_t freq)
 {
     int sel, div, dev_shift, i;
     const uint64_t mhz = 1000000;
-    int dev_freq = freq / mhz;
+    int dev_freq;
+    uint64_t nae_freq;
     uint32_t val, dev_id = XLP9XX_CLKDEV_NET;
     uint64_t sys_base = XLP9XX_SYS_CLKDEV_BASE(node);
 
-    nlm_print("NAE frequency %lu MHz\n", get_nae_frequency(node) / mhz);
+    nae_freq = get_nae_frequency(node);
+
+    NLM_HAL_DO_DIV(nae_freq, mhz);
+
+    nlm_print("NAE frequency %lu MHz\n", nae_freq);
 		
     if (freq == get_nae_frequency(node))
-	return 0;	
+	return 0;
+
+    NLM_HAL_DO_DIV(freq, mhz);
+    dev_freq = freq;
     switch (dev_freq) {
         case 133: sel = 0; div = 0; break;
         case 66:  sel = 0; div = 1; break;
@@ -242,7 +250,7 @@ static inline int xlp9xx_set_nae_frequency(int node, uint64_t freq)
       }
     }
 
-    nlm_print("NAE frequency changed to %lu MHz\n", get_nae_frequency(node) / mhz);
+    nlm_print("NAE frequency changed to %lu MHz\n", nae_freq);
     return 0;
 }
 
@@ -283,18 +291,24 @@ static inline int set_crypto_block_frequency(int node, int dev_id, uint64_t freq
 {
     int sel, div, dev_shift, i;
     const uint64_t mhz = 1000000;
-    int dev_freq = freq / mhz;
+    int dev_freq;
+    uint64_t block_freq;
     uint32_t val;
     uint64_t sys_base = XLP9XX_SYS_CLKDEV_BASE(node);
     char* block_name;
 
+    NLM_HAL_DO_DIV(freq, mhz);
+    dev_freq = freq;
+
     if (dev_id == XLP9XX_CLKDEV_SEC)
         block_name = "SAE";
     else
         block_name = "RSA";
 
-    nlm_print("%s frequency %lu MHz\n", block_name,  
-              get_crypto_block_frequency(node, dev_id) / mhz);
+    block_freq = get_crypto_block_frequency(node, dev_id);
+    NLM_HAL_DO_DIV(block_freq, mhz);
+
+    nlm_print("%s frequency %lu MHz\n", block_name, block_freq);
 
     if (freq == get_crypto_block_frequency(node, dev_id))
         return 0;
@@ -342,7 +356,7 @@ static inline int set_crypto_block_frequency(int node, int dev_id, uint64_t freq
     }
 
     nlm_print("%s frequency changed to %lu MHz\n",block_name, 
-              get_crypto_block_frequency(node, dev_id) / mhz);
+              block_freq);
     return 0;
 }
 
-- 
1.7.1

