From 96a4036fd965a59fffac5dbf4d5bcd10ed6e23b0 Mon Sep 17 00:00:00 2001
From: Vikas Gupta <vikas.gupta@broadcom.com>
Date: Thu, 14 Nov 2013 00:53:32 -0800
Subject: nae: fix gmac-mdio read return value.

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/netlogic/lib/netlib/src/netsoc_nae.c b/arch/mips/netlogic/lib/netlib/src/netsoc_nae.c
index 1389f86..9469c5c 100644
--- a/arch/mips/netlogic/lib/netlib/src/netsoc_nae.c
+++ b/arch/mips/netlogic/lib/netlib/src/netsoc_nae.c
@@ -188,7 +188,7 @@ int __netsoc_int_gmac_mdio_read(nae_t *nae, uint32_t bus, uint32_t phyaddr, uint
 
 	/* Read the data back
 	 */
-	return netsoc_read_mac_reg(mac_base, INT_MDIO_RD_STAT + bus * 4);
+	return (netsoc_read_mac_reg(mac_base, INT_MDIO_RD_STAT + bus * 4) & 0xffff);
 }
 
 /* Internal MDIO WRITE Routines
-- 
1.7.1

