From acfb26cd1ef84c6497db42088eac181e75f29062 Mon Sep 17 00:00:00 2001
From: Jayanthi A <jayanthi.annadurai@broadcom.com>
Date: Fri, 15 Nov 2013 13:51:03 -0800
Subject: netlib: Fixed POE fbp base address

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/netlogic/lib/netlib/src/netsoc_poe.c b/arch/mips/netlogic/lib/netlib/src/netsoc_poe.c
index a1f6b01..486829b 100644
--- a/arch/mips/netlogic/lib/netlib/src/netsoc_poe.c
+++ b/arch/mips/netlogic/lib/netlib/src/netsoc_poe.c
@@ -89,7 +89,7 @@ int __netsoc_init_poe_ext_storage(poe_t *poe, uint64_t fbp_base_phys,
         /* Free Buffer Pool config */
         nlm_print (" POE Free Buffer Pool config ...\n");
 
-        if (is_nlm_xlp3xx() || is_nlm_xlp2xx()) {
+        if (is_nlm_xlp3xx() || is_nlm_xlp2xx() || is_nlm_xlp9xx()) {
                 a = (uint64_t)XLP3XX_EXT_FBP_START_ADDR;
                 num_messages = XLP3XX_MAX_POE_EXT_MSG_STORAGE;
                 fbp_sp_init = XLP3XX_POE_FBP_SP_INIT;
-- 
1.7.1

