From d62f77eff68ce7c317e6bb910fdbe149213aff1c Mon Sep 17 00:00:00 2001
From: Pavani Reddy <pavani@broadcom.com>
Date: Fri, 15 Nov 2013 18:25:34 +0530
Subject: netlib: Adding an API for setting parser sequencer threshold

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/netlogic/lib/netlib/src/netsoc_nae.c b/arch/mips/netlogic/lib/netlib/src/netsoc_nae.c
index ec3b0d4..e3c1bcd 100644
--- a/arch/mips/netlogic/lib/netlib/src/netsoc_nae.c
+++ b/arch/mips/netlogic/lib/netlib/src/netsoc_nae.c
@@ -1560,6 +1560,15 @@ void __netsoc_config_nae_global(nae_t *nae)
 	__netsoc_config_flow_base(nae); // must be after one of the above functions.
 }
 
+void  __netsoc_setparser_seq_thold(nae_t *nae , int thrs)
+{
+        uint32_t value;
+        value = netsoc_read_nae_reg(nae->nae_base , RX_FREE_LIFO_TH);
+        value = value & 0xffffff00;
+        value |= (thrs & 0xff);
+        netsoc_write_nae_reg(nae->nae_base, RX_FREE_LIFO_TH, value);
+
+}
 
 void __netsoc_set_frequency(nae_t *nae, uint32_t frequency)
 {
-- 
1.7.1

