From 74a449207df27c8582670bff47f7d64c32ef90d4 Mon Sep 17 00:00:00 2001
From: Hareesh R <hareeshr@broadcom.com>
Date: Tue, 19 Nov 2013 10:29:26 +0530
Subject: 2xx: Added b0 macro

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/netlogic/lib/syslib/include/nlm_hal_xlp_dev.h b/arch/mips/netlogic/lib/syslib/include/nlm_hal_xlp_dev.h
index a455884..de5eae3 100644
--- a/arch/mips/netlogic/lib/syslib/include/nlm_hal_xlp_dev.h
+++ b/arch/mips/netlogic/lib/syslib/include/nlm_hal_xlp_dev.h
@@ -197,6 +197,7 @@ extern int is_nlm_xlp(unsigned int chipid, unsigned int rev,  unsigned int ext);
 #define is_nlm_xlp3xx_lite()  (is_nlm_xlp3xx() && (!is_nlm_xlp3xx_B(XLP_REVISION_ANY)) )
 
 #define is_nlm_xlp2xx()	    is_nlm_xlp(0x2000, XLP_REVISION_ANY,  0)
+#define is_nlm_xlp2xx_b0()	    is_nlm_xlp(0x2000, XLP_REVISION_B0,  0)
 #define is_nlm_xlp9xx()	    is_nlm_xlp(0x9000, XLP_REVISION_ANY,  0)
 
 #endif /*__XLP_CHIPID_MACROS__ */
-- 
1.7.1

