From a148a76a00444ebefb0dc72c31bfecda610f33c2 Mon Sep 17 00:00:00 2001
From: Hareesh R <hareeshr@broadcom.com>
Date: Thu, 28 Nov 2013 15:56:25 +0530
Subject: seclib: Retry the request when it fails

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/netlogic/lib/seclib/cryptosoc_lib.c b/arch/mips/netlogic/lib/seclib/cryptosoc_lib.c
index c43d766..b67554e 100644
--- a/arch/mips/netlogic/lib/seclib/cryptosoc_lib.c
+++ b/arch/mips/netlogic/lib/seclib/cryptosoc_lib.c
@@ -787,11 +787,15 @@ static int cryptosoc_process_sync_shrd_request(int gen_type2,
 
 	_uint64_t msg2;
 	int req_id, rv;
+	int nretry = retry_count;
+	retry_count &= 0xfffffff;
+try_again:
 
 	cryptosoc_lock(&cpriv->lock);
 	if(cpriv->req_tail == 0) {
 		cryptosoc_unlock(&cpriv->lock);
-		return -CRYPTOSOC_EAGAIN;
+		rv = -CRYPTOSOC_EAGAIN;
+		goto get_out;
 	}
 	cpriv->req_tail--;
 	req_id = cpriv->free_req_index[cpriv->req_tail];
@@ -836,6 +840,11 @@ shrd_exit:
 	cpriv->free_req_index[cpriv->req_tail] = req_id;
 	cpriv->req_tail++;
 	cryptosoc_unlock(&cpriv->lock);
+
+get_out:
+	if((rv == -CRYPTOSOC_EAGAIN) && (nretry == -1 || retry_count > 0))
+		goto try_again;
+
 	return rv;
 
 }
-- 
1.7.1

