From cde9b5d758813b5964813d4f0a1bf57f00fb8050 Mon Sep 17 00:00:00 2001
From: Saswat Dash <saswat.dash@broadcom.com>
Date: Thu, 28 Nov 2013 17:06:14 +0530
Subject: fmnlib: Added credit config for PCIE blocks

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/netlogic/lib/fmnlib/nlm_hal_fmn.h b/arch/mips/netlogic/lib/fmnlib/nlm_hal_fmn.h
index 1d21ff8..b632733 100644
--- a/arch/mips/netlogic/lib/fmnlib/nlm_hal_fmn.h
+++ b/arch/mips/netlogic/lib/fmnlib/nlm_hal_fmn.h
@@ -76,6 +76,7 @@ enum FMN_MSG_BLKS {
 	XLP_MSG_BLK_POE,
 	XLP_MSG_BLK_NAE,
 	XLP_MSG_BLK_REGX,
+	XLP_MSG_BLK_KBP,
 	XLP_MSG_BLK_SRIO,
 	XLP_MSG_BLK_LZS,
 	XLP_MSG_BLK_MAX = XLP_MSG_BLK_LZS + 1,
diff --git a/arch/mips/netlogic/lib/fmnlib/nlm_hal_fmn_config.c b/arch/mips/netlogic/lib/fmnlib/nlm_hal_fmn_config.c
index 43311af..2d036ef 100644
--- a/arch/mips/netlogic/lib/fmnlib/nlm_hal_fmn_config.c
+++ b/arch/mips/netlogic/lib/fmnlib/nlm_hal_fmn_config.c
@@ -224,7 +224,7 @@ nlm_fmn_config_t xlp9xx_fmn_config[] = {
 
 extern struct nlm_node_config nlm_node_cfg;
 
-/* #define FMN_DEBUG 1  */
+/* #define FMN_DEBUG 1 */ 
 static int max_msg_blks = XLP_MSG_BLK_MAX;
 
 static unsigned int fmn_cfg_value[XLP_NET_VC_LIMIT + 1];
@@ -252,6 +252,7 @@ static struct fmn_qsize_credit_config fmn_qsize_credit_cfg[XLP9XX_MSG_BLK_MAX] =
 	[XLP_MSG_BLK_REGX] =    { "regx",   XLP_3XX_REGEX_VC_BASE, XLP_3XX_REGEX_VC_LIMIT, 1, 1 },
 	[XLP_MSG_BLK_SRIO] =    { "srio",   XLP_3XX_SRIO_VC_BASE,  XLP_3XX_SRIO_VC_LIMIT,  1, 1 },
 	[XLP_MSG_BLK_LZS] =    { "lzs",   XLP_9XX_LZS_VC_BASE,  XLP_9XX_LZS_VC_LIMIT,  1, 1 },
+	[XLP_MSG_BLK_KBP] =	{"kbp",	XLP_9XX_KBP_VC_BASE,	XLP_9XX_KBP_VC_LIMIT,	1, 1},
         [XLP_MSG_BLK_POE_1] =  { "poe1",    XLP_9XX_POE1_VC_BASE,   XLP_9XX_POE1_VC_LIMIT, 1, 1 },
         [XLP_MSG_BLK_NAE_1] =  { "nae1",    XLP_9XX_NET1_VC_BASE,   XLP_9XX_NET1_VC_LIMIT, 1, 1 },
 	[XLP_MSG_BLK_ALE] =    { "ale",     XLP_9XX_ALE_VC_BASE,   XLP_9XX_ALE_VC_LIMIT, 1, 1 },
@@ -610,12 +611,15 @@ static void fmn_update_credit(int node, int b_stid, int dst_node)
 	for(s_stn = 0; s_stn < max_msg_blks; s_stn++) {
 		if(!fmn_q_config[s_stn].valid)
 			continue;
+		
 		if(b_stid >=  fmn_q_config[s_stn].b_stid && 
 				b_stid <= fmn_q_config[s_stn].e_stid) {
 			credits = fmn_q_config[s_stn].credits[dst_node];
 			break;
 		}
+	
 	}
+
 	if(credits == NULL) {
 		nlm_print("ERROR in Credit config: Station id not found, configuring default credit\n");
 #ifdef FMN_DEBUG
@@ -965,7 +969,6 @@ static void nlm_hal_write_fmn_credit(int node, int max_nodes)
 			if (fmn_9xx_is_vc_valid(node, base_vc) )
 			{
 				int station;
-
 				/* configure credits from this base station to all other appropriate VC's */
 				for(dst_node = 0 ;dst_node < max_nodes; dst_node++) {
 					fmn_update_credit(node, base_vc, dst_node);
@@ -997,6 +1000,7 @@ static void nlm_hal_write_fmn_credit(int node, int max_nodes)
 			}
 			fmn_config++;
 		}
+
 	} else if (is_nlm_xlp3xx() || is_nlm_xlp2xx()) {
 		nlm_print(" XLP3XX/XLP2XX FMN configuration \n");
 		if(is_nlm_xlp316()){
@@ -1686,10 +1690,15 @@ void nlm_hal_fmn_init(void *fdt, int node)
 
                fmn_modify_qsize_credit_config(node, XLP_MSG_BLK_NAE, 1, XLP_9XX_NET0_VC_BASE, XLP_9XX_NET0_VC_LIMIT);
                fmn_modify_qsize_credit_config(node, XLP_MSG_BLK_POE, 1, XLP_9XX_POE0_VC_BASE, XLP_9XX_POE0_VC_LIMIT);
+
+		fmn_modify_qsize_credit_config(node, XLP_MSG_BLK_PCIE0, 1, XLP_9XX_PCIE0_VC_BASE, XLP_9XX_PCIE0_VC_LIMIT);
+		fmn_modify_qsize_credit_config(node, XLP_MSG_BLK_PCIE1, 1, XLP_9XX_PCIE1_VC_BASE, XLP_9XX_PCIE1_VC_LIMIT);
+		fmn_modify_qsize_credit_config(node, XLP_MSG_BLK_PCIE2, 1, XLP_9XX_PCIE2_VC_BASE, XLP_9XX_PCIE2_VC_LIMIT);
+		fmn_modify_qsize_credit_config(node, XLP_MSG_BLK_PCIE3, 1, XLP_9XX_PCIE3_VC_BASE, XLP_9XX_PCIE3_VC_LIMIT);
         }
 
 	update_fmn_config();
-
+	
 	if (max_nodes < 0) {
 		nlm_print("FMN Configuration failed\n");
 		while(1);
-- 
1.7.1

