From a7895769bcf00fd7fcde2b4c4ccae725009b13ba Mon Sep 17 00:00:00 2001
From: Vikas Gupta <vikas.gupta@broadcom.com>
Date: Fri, 29 Nov 2013 22:43:35 +0530
Subject: xlp9xx: changed the xlgmac threshold value.

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/netlogic/lib/netlib/src/netsoc_nae.c b/arch/mips/netlogic/lib/netlib/src/netsoc_nae.c
index 3b68813..2982f7a 100644
--- a/arch/mips/netlogic/lib/netlib/src/netsoc_nae.c
+++ b/arch/mips/netlogic/lib/netlib/src/netsoc_nae.c
@@ -2323,7 +2323,7 @@ static void __netsoc_config_xlgmac(net_port_t *netport)
 
 	__netsoc_xlgmac_softreset(mac_base, 0x3);
 	__netsoc_xlgmac_softreset(mac_base, 0x0);
-	__netsoc_xlgmac_threshold(mac_base, 0x1f);
+	__netsoc_xlgmac_threshold(mac_base, 0xf);
 	__netsoc_set_xlgmac_enable(netport);
 	__netsoc_set_xlgmac_pause_config(netport);
 
-- 
1.7.1

