From 961e244af176a62359fac103cdeae23d76d8ae03 Mon Sep 17 00:00:00 2001
From: Vikas Gupta <vikas.gupta@broadcom.com>
Date: Mon, 2 Dec 2013 23:21:31 +0530
Subject: xlp2xx: fix number of ucores and mask.

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/netlogic/lib/netlib/include/xlp2xx_netsoc.h b/arch/mips/netlogic/lib/netlib/include/xlp2xx_netsoc.h
index d2b86aa..2e34989 100644
--- a/arch/mips/netlogic/lib/netlib/include/xlp2xx_netsoc.h
+++ b/arch/mips/netlogic/lib/netlib/include/xlp2xx_netsoc.h
@@ -34,10 +34,10 @@
 
 #define XLP2XX_MAX_PORTS                8
 #define XLP2XX_MAX_FREEIN_LIFOS         8
-#define XLP2XX_MAX_UCORE                2
+#define XLP2XX_MAX_UCORE                8
 #define XLP2XX_MAX_CONTEXTS             64
 #define XLP2XX_MAX_NAE_COMPLEX          2
-#define XLP2XX_UCORE_MASK               0x3
+#define XLP2XX_UCORE_MASK               0xFF
 
 // INGRESS 
 
-- 
1.7.1

