From 96f19e9ed47200d98965ebb650a744643de66566 Mon Sep 17 00:00:00 2001
From: Pavani Reddy <pavani@broadcom.com>
Date: Fri, 29 Nov 2013 18:41:45 +0530
Subject: fmnlib: changed fmnlib for multinode support

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/netlogic/lib/fmnlib/nlm_hal_fmn_config.c b/arch/mips/netlogic/lib/fmnlib/nlm_hal_fmn_config.c
index 2d036ef..cb1f18f 100644
--- a/arch/mips/netlogic/lib/fmnlib/nlm_hal_fmn_config.c
+++ b/arch/mips/netlogic/lib/fmnlib/nlm_hal_fmn_config.c
@@ -628,8 +628,6 @@ static void fmn_update_credit(int node, int b_stid, int dst_node)
 		return;
 	}
 
-	nlm_print("b_stid %d, dst_node %d.\n", b_stid, dst_node);
-
 	for(d_stn = 0; d_stn < max_msg_blks; d_stn++) {
 #ifdef FMN_DEBUG
 		nlm_print("update credit s_stn %d d_stn %d credits %d\n",s_stn, d_stn, credits[d_stn]); 
@@ -1519,8 +1517,9 @@ int parse_fdt_fmn_config(void *fdt)
 	uint64_t spill_base = 0ULL;
 #endif
 	uint32_t qsize, credits, src_node, s_stn, d_stn;
+	static int init_done = 0;
 
-
+	
 	sprintf(fmn_cfg_str,"/soc/nodes");
 	nodeoffset = fdt_path_offset(fdt, fmn_cfg_str);
         if(nodeoffset >= 0) {
@@ -1529,9 +1528,11 @@ int parse_fdt_fmn_config(void *fdt)
 			max_nodes = fdt32_to_cpu(*(unsigned int *)pval);
 		}
 	}
-
 	nlm_print("Number of nodes %d \n",max_nodes);
 
+	if(init_done)
+		return max_nodes;
+
 	nlm_node_cfg.num_nodes = max_nodes;
 	
 	for(node = 0; node < max_nodes; node++) {
@@ -1622,6 +1623,7 @@ int parse_fdt_fmn_config(void *fdt)
 
 	parse_queue_config(fdt, max_nodes);
 
+	init_done = 1;
 	return max_nodes;
 }
 
@@ -1713,6 +1715,7 @@ void nlm_hal_fmn_init(void *fdt, int node)
 	 */
 	if(nlm_hal_setup_outq(fdt, node, max_nodes) < 0)
 		while(1);
+
 	nlm_hal_write_fmn_credit(node, max_nodes);
 }
 
-- 
1.7.1

