From 91799c9177b4951dbdf11e4d07a10049285afb5f Mon Sep 17 00:00:00 2001
From: Vikas Gupta <vikas.gupta@broadcom.com>
Date: Fri, 6 Dec 2013 23:07:40 +0530
Subject: xlp2xx: fix for passing phymode argument in pma2p0 configuration.

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/netlogic/lib/netlib/src/netsoc_pcs.c b/arch/mips/netlogic/lib/netlib/src/netsoc_pcs.c
index 32b4f7f..8977d2b 100644
--- a/arch/mips/netlogic/lib/netlib/src/netsoc_pcs.c
+++ b/arch/mips/netlogic/lib/netlib/src/netsoc_pcs.c
@@ -1133,7 +1133,7 @@ static void xlp2xx_xaui_pcs_init(nae_t *nae, uint32_t xaui_cplx_mask, uint32_t p
                         nlm_hal_config_pma2p0_mem_xaui_125();
                 }
 		
-		xlp2xx_config_block_pma2p0(nae, block, nae->serdes_init_cnt, xaui_speed, 0);
+		xlp2xx_config_block_pma2p0(nae, block, nae->serdes_init_cnt, xaui_speed, phymode);
 		nae->serdes_init_cnt++;	
 
                 for( lane_ctrl = PHY_LANE_0_CTRL; lane_ctrl <= PHY_LANE_3_CTRL; lane_ctrl++){
-- 
1.7.1

