From b2e9a40f9f4156d618700c1a14ce7b966ae67f85 Mon Sep 17 00:00:00 2001
From: ajesh <ajesh@broadcom.com>
Date: Tue, 10 Dec 2013 15:24:35 +0530
Subject: XAUI: Correct bus number for phy configuration

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/netlogic/lib/netlib/src/netsoc_nae.c b/arch/mips/netlogic/lib/netlib/src/netsoc_nae.c
index 2982f7a..b4e45f0 100644
--- a/arch/mips/netlogic/lib/netlib/src/netsoc_nae.c
+++ b/arch/mips/netlogic/lib/netlib/src/netsoc_nae.c
@@ -3873,6 +3873,7 @@ static void __netsoc_nlp1042c2_init(nae_t* nae)
 		if((nae->ports[port].iftype == RXAUI_IF) || (nae->ports[port].iftype == XAUI_IF)){
 			phyaddr = nae->ports[port].ext_phy_addr;
 			phybus = nae->ports[port].ext_phy_bus;
+			bus = phybus;
 			submode = nae->ports[port].rxaui_mode;
 			netsoc_api_print(NETSOC_APIDBG_CONFIG, "Initializing external NLP1042 PHY phyaddr= 0x%x bus = 0x%x submode=0x%x\n",  phyaddr, phybus, submode);
 			if(submode == NLP1042_PHY_10G_NONE){
-- 
1.7.1

