From 1cc2105093949dfa9261b2f4f675de9c5db363d3 Mon Sep 17 00:00:00 2001
From: Alok Agrawat <alok@broadcom.com>
Date: Tue, 10 Dec 2013 23:09:20 -0800
Subject: interlaken: fixed nae-app for XLP8xx.

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/netlogic/lib/netlib/external/cortina_cs34x7/nlm_cortina_cs34x7.c b/arch/mips/netlogic/lib/netlib/external/cortina_cs34x7/nlm_cortina_cs34x7.c
index a344331..07f4a26 100644
--- a/arch/mips/netlogic/lib/netlib/external/cortina_cs34x7/nlm_cortina_cs34x7.c
+++ b/arch/mips/netlogic/lib/netlib/external/cortina_cs34x7/nlm_cortina_cs34x7.c
@@ -213,6 +213,9 @@ int cortina_get_mac_stats(int node, int hwport, int port, void *devstats)
         int cs;
 	unsigned long long data;
 
+	if (!is_ilk_card_onslot(hwport/4)) 
+		return -1;
+
 	if (hwport == XLP_ILK_PORT_0)
                 cs = XLP_ILK_PORT0_CS;
         else if (hwport == XLP_ILK_PORT_1)
@@ -370,16 +373,17 @@ int nlm_hal_init_cs34x7(int hwport,int num_lanes, int lane_rate, int nae_id)
 		nlm_print("Maximum lanes supported on EVP2 is 4\n");
 		return -1;
 	}
+	if(nae_id == 1)
+		hwport += 8;
 
-#if 0
-	if (!is_ilk_card_onslot(hwport/4)) {
-		nlm_print("No interlaken card on slot %d\n",hwport/4);
-		return -1;
+	if(!is_nlm_xlp9xx())	
+	{	
+		if (!is_ilk_card_onslot(hwport/4)) {
+			nlm_print("No interlaken card on slot %d\n",hwport/4);
+			return -1;
+		}
 	}
-#endif
 
-	if(nae_id == 1)
-		hwport += 8;
 	switch(hwport) {
 		case XLP_ILK_PORT_0:
 			cortina_cs34x7_init(XLP_ILK_PORT0_CS, (XLP_ILK_MAX_LANES - num_lanes), lane_rate);
-- 
1.7.1

