From 3569dd4df0d3e8992e1f60d596c98570134dfc3c Mon Sep 17 00:00:00 2001
From: Manoj Vellala <manojv@broadcom.com>
Date: Mon, 16 Dec 2013 02:51:13 -0800
Subject: seclib: Hash algorithm CBC_MAC doesnot support HMAC.

	set hmac = 0 for HASH_MODE_CBC
[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/netlogic/lib/seclib/saesoc_lib.h b/arch/mips/netlogic/lib/seclib/saesoc_lib.h
index 7f71b96..4b43544 100644
--- a/arch/mips/netlogic/lib/seclib/saesoc_lib.h
+++ b/arch/mips/netlogic/lib/seclib/saesoc_lib.h
@@ -666,8 +666,9 @@ static inline int saesoc_new_session(
 			int hmac = initp->hash.key_len?1:0;
 			int mode ;
 			if ( (initp->hash.mode == SAESOC_HASH_MODE_XCBC) ||
-			  	(initp->hash.mode == SAESOC_HASH_MODE_GCM) || 
-				(initp->hash.mode == SAESOC_HASH_MODE_CCM))
+			  	(initp->hash.mode == SAESOC_HASH_MODE_GCM) ||
+                                (initp->hash.mode == SAESOC_HASH_MODE_CCM) ||
+				(initp->hash.mode == SAESOC_HASH_MODE_CBC))
 			       hmac = 0;	
 			spriv->hash_en = 1;
 			spriv->addl_hash_len = initp->hash.addl_hash_len;
-- 
1.7.1

