From 782bd8025ddac19365d3af0a1a09a05cc3b76188 Mon Sep 17 00:00:00 2001
From: Saswat Dash <saswat.dash@broadcom.com>
Date: Mon, 3 Feb 2014 18:05:41 +0530
Subject: netlib: corrected typo

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/netlogic/lib/netlib/src/netsoc_fdt.c b/arch/mips/netlogic/lib/netlib/src/netsoc_fdt.c
index 78db869..84ee86d 100644
--- a/arch/mips/netlogic/lib/netlib/src/netsoc_fdt.c
+++ b/arch/mips/netlogic/lib/netlib/src/netsoc_fdt.c
@@ -480,7 +480,7 @@ start_for_dom:
 						cpu_mask[i] = fdt32_to_cpu(pval[max]);
 					right_shift_int_array(cpu_mask, node * num_cpus_per_node, num_cpus_per_node);
 				}
-		sprintf(dom_node_str, "/doms/dom@%d/cpu, domid");
+		sprintf(dom_node_str, "/doms/dom@%d/cpu", domid);
 		nodeoffset = fdt_path_offset(fdt, dom_node_str);
 		if (nodeoffset >= 0) {
              		pval = (unsigned int *)fdt_getprop(fdt, nodeoffset, "nae-rx-vc", &plen);
-- 
1.7.1

