From bf0a940890ee476f0cda734aed84981b51b25fd6 Mon Sep 17 00:00:00 2001
From: Xiaoyi Liu <xiaoyi@broadcom.com>
Date: Tue, 11 Feb 2014 17:16:33 -0500
Subject: fmnlib: outputq should only be configued for the valid stations for a given platform.

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/netlogic/lib/fmnlib/nlm_hal_fmn_config.c b/arch/mips/netlogic/lib/fmnlib/nlm_hal_fmn_config.c
index 634b776..6da5a63 100644
--- a/arch/mips/netlogic/lib/fmnlib/nlm_hal_fmn_config.c
+++ b/arch/mips/netlogic/lib/fmnlib/nlm_hal_fmn_config.c
@@ -1196,6 +1196,9 @@ int nlm_hal_setup_outq(void *fdt, int node, int max_nodes)
 		nlm_print(" XLP9XX/XLP7XX FMN output queue configuration \n");
 
 		for (station = 0; station < max_msg_blks; station++) {
+
+			if (!fmn_q_config[station].valid) continue;
+
 			qid = fmn_q_config[station].b_stid;
 			if ( fmn_9xx_is_vc_valid(node, qid) == 0 ) continue;
 
-- 
1.7.1

