From 06e70c0b170c4fb7b8400d81e9b660b375d4b29a Mon Sep 17 00:00:00 2001
From: Virendra Pathak <vpathak@broadcom.com>
Date: Wed, 19 Feb 2014 21:14:00 +0530
Subject: syslib: updated function number of DTRE1 for xlp9xx

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/netlogic/lib/syslib/include/nlm_hal_xlp_dev.h b/arch/mips/netlogic/lib/syslib/include/nlm_hal_xlp_dev.h
index 9302e09..21f8882 100644
--- a/arch/mips/netlogic/lib/syslib/include/nlm_hal_xlp_dev.h
+++ b/arch/mips/netlogic/lib/syslib/include/nlm_hal_xlp_dev.h
@@ -458,6 +458,7 @@ extern int is_nlm_xlp(unsigned int chipid, unsigned int rev,  unsigned int ext);
 #define XLP_DTR_BUS                     0x0
 #define XLP_DTR_DEVICE                  0x5
 #define XLP_DTR_FUNC                    0x0
+#define XLP_DTR_FUNC_1                  0x6  /* for xlp9xx */
 #define XLP_DTR_MASTER_CONTROL_REG      0x40
 #define XLP_DTR_CHANNEL_CONTROL_REG_0   0x48
 #define XLP_DTR_CHANNEL_CONTROL_REG_1   0x49
-- 
1.7.1

