From f58a16cd7f6c1e5b0c2a0df9d2cf13f5885f6a6f Mon Sep 17 00:00:00 2001
From: Alok Agrawat <alok@broadcom.com>
Date: Wed, 26 Feb 2014 03:10:43 -0800
Subject: jumbo nae: enabled jumbo frame size for interlaken.

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/netlogic/lib/netlib/src/netsoc_nae.c b/arch/mips/netlogic/lib/netlib/src/netsoc_nae.c
index 5c6608d..425d219 100644
--- a/arch/mips/netlogic/lib/netlib/src/netsoc_nae.c
+++ b/arch/mips/netlogic/lib/netlib/src/netsoc_nae.c
@@ -3117,6 +3117,13 @@ int __netsoc_config_frame_size(net_port_t *netport, uint32_t size)
 		case XLAUI_IF:
 			return NETSOC_API_SUCCESS;  /*default its setting 16K*/
 			break;
+                case INTERLAKEN_IF:
+                        if(nae->nae_id)
+                                return cortina_set_max_framesize(netport->hw_port_id+8, (netport->hw_port_id/MAX_PORTS_PERBLOCK), size);
+                        else
+                                return cortina_set_max_framesize(netport->hw_port_id, (netport->hw_port_id/MAX_PORTS_PERBLOCK), size);
+                        break;
+
 		default:
 			return -NETSOC_API_PARAM_INVALID;
 	}
-- 
1.7.1

