From 4e42c5d272975aa7414025f87d09583d48e0793e Mon Sep 17 00:00:00 2001
From: Alok Agrawat <alok@broadcom.com>
Date: Wed, 26 Feb 2014 03:47:43 -0800
Subject: ilk: enabled interlaken card detection from cpld for xlp9xx.

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/netlogic/lib/netlib/external/cortina_cs34x7/nlm_cortina_cs34x7.c b/arch/mips/netlogic/lib/netlib/external/cortina_cs34x7/nlm_cortina_cs34x7.c
index 07f4a26..94a265f 100644
--- a/arch/mips/netlogic/lib/netlib/external/cortina_cs34x7/nlm_cortina_cs34x7.c
+++ b/arch/mips/netlogic/lib/netlib/external/cortina_cs34x7/nlm_cortina_cs34x7.c
@@ -339,7 +339,6 @@ int cortina_set_max_framesize(int hwport, int port, uint16_t framesize)
 		cs = XLP_ILK_PORT1_CS;
 	else
 		return -1;
-
 	if (port > (CORTINA_MAX_PORTS - 1))
 		return -1;
 
@@ -376,12 +375,9 @@ int nlm_hal_init_cs34x7(int hwport,int num_lanes, int lane_rate, int nae_id)
 	if(nae_id == 1)
 		hwport += 8;
 
-	if(!is_nlm_xlp9xx())	
-	{	
-		if (!is_ilk_card_onslot(hwport/4)) {
-			nlm_print("No interlaken card on slot %d\n",hwport/4);
-			return -1;
-		}
+	if (!is_ilk_card_onslot(hwport/4)) {
+		nlm_print("No interlaken card on slot %d\n",hwport/4);
+		return -1;
 	}
 
 	switch(hwport) {
-- 
1.7.1

