From 4c0c44f6b1f4e8ff2bec548cdbf98c66db44ffd5 Mon Sep 17 00:00:00 2001
From: Virendra Pathak <vpathak@broadcom.com>
Date: Sat, 15 Mar 2014 13:28:44 +0530
Subject: cleanup: Removed compiler warnings for halx

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/netlogic/lib/syslib/include/xlp9xx_sys.h b/arch/mips/netlogic/lib/syslib/include/xlp9xx_sys.h
index c633de2..6e769bf 100644
--- a/arch/mips/netlogic/lib/syslib/include/xlp9xx_sys.h
+++ b/arch/mips/netlogic/lib/syslib/include/xlp9xx_sys.h
@@ -254,7 +254,7 @@ static inline int xlp9xx_set_nae_frequency(int node, uint64_t freq)
 
     NLM_HAL_DO_DIV(nae_freq, mhz);
 
-    nlm_print("NAE frequency %lu MHz node :%d\n", nae_freq,node);
+    nlm_print("NAE frequency %llu MHz node :%d\n", (unsigned long long) nae_freq, node);
 		
     if (freq == get_nae_frequency(node))
 	return 0;
@@ -313,7 +313,7 @@ static inline int xlp9xx_set_nae_frequency(int node, uint64_t freq)
       }
     }
 
-    nlm_print("NAE frequency changed to %lu MHz\n", nae_freq);
+    nlm_print("NAE frequency changed to %llu MHz\n", (unsigned long long)nae_freq);
     return 0;
 }
 
@@ -371,7 +371,7 @@ static inline int set_crypto_block_frequency(int node, int dev_id, uint64_t freq
     block_freq = get_crypto_block_frequency(node, dev_id);
     NLM_HAL_DO_DIV(block_freq, mhz);
 
-    nlm_print("%s frequency %lu MHz\n", block_name, block_freq);
+    nlm_print("%s frequency %llu MHz\n", block_name, (unsigned long long)block_freq);
 
     if (freq == get_crypto_block_frequency(node, dev_id))
         return 0;
@@ -418,8 +418,8 @@ static inline int set_crypto_block_frequency(int node, int dev_id, uint64_t freq
       }
     }
 
-    nlm_print("%s frequency changed to %lu MHz\n",block_name, 
-              block_freq);
+    nlm_print("%s frequency changed to %llu MHz\n",block_name, 
+              (unsigned long long)block_freq);
     return 0;
 }
 
-- 
1.7.1

