From 065024d9c0a559782bd9240f2dec4369baeac9e3 Mon Sep 17 00:00:00 2001
From: Debayan Ghosh <debayan.ghosh@broadcom.com>
Date: Mon, 24 Mar 2014 23:09:30 +0530
Subject: cache_lock: added register definitions for l3 cache lock

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/netlogic/lib/syslib/include/nlm_hal_xlp_dev.h b/arch/mips/netlogic/lib/syslib/include/nlm_hal_xlp_dev.h
index 21f8882..400cec1 100644
--- a/arch/mips/netlogic/lib/syslib/include/nlm_hal_xlp_dev.h
+++ b/arch/mips/netlogic/lib/syslib/include/nlm_hal_xlp_dev.h
@@ -428,6 +428,14 @@ extern int is_nlm_xlp(unsigned int chipid, unsigned int rev,  unsigned int ext);
 #define XLP9XX_SAE_RSA_BUS_NUM		(1<<20)
 #define XLP9XX_PCIE_DEV_BASE(SOC)		(((SOC##_DEVICE) << 15) | (SOC##_FUNC << 12))
 
+/* l3 cpu cache registers */
+#define XLP9XX_L3_LINE_LCK0_REG_ID  0x110
+#define XLP9XX_L3_LINE_LCK1_REG_ID  0x111
+#define XLP9XX_L3_STATUS_REG_ID     0x117
+#define XLP9XX_CMU_CFG_TARGET_ID_REG_ID  0x103
+#define XLP9XX_LINE_FLUSH_REG0_ID      0xE9
+#define XLP9XX_LINE_FLUSH_REG1_ID      0xEA
+#define XLP9XX_L3_LINE_FLUSH_STATUS_REG_ID  0xEB
 
 
 /*Sw Code */
-- 
1.7.1

