From 2064a43440bc6df9cf6ece4ee2ed78d10476681a Mon Sep 17 00:00:00 2001
From: Alok Agrawat <alok@broadcom.com>
Date: Fri, 21 Mar 2014 05:04:36 -0700
Subject: interlaken: fixed a typo which caused interlaken to fail.

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/netlogic/lib/netlib/src/netsoc_pcs.c b/arch/mips/netlogic/lib/netlib/src/netsoc_pcs.c
index e277462..6ccfb1a 100644
--- a/arch/mips/netlogic/lib/netlib/src/netsoc_pcs.c
+++ b/arch/mips/netlogic/lib/netlib/src/netsoc_pcs.c
@@ -727,7 +727,7 @@ void xlp9xx_nae_write_pma2p0_mem(uint64_t mac_base, int block, int xgmii_speed,
 			if (kr_en)
 				__netsoc_write_pma2p0_reg(mac_base, lane_no, 0x1c, 0x50);
 			else
-				__netsoc_write_pma2p0_reg(mac_base, lane_no, 0x1c, 0xc0);
+				__netsoc_write_pma2p0_reg(mac_base, lane_no, 0x1c, 0x00);
 				
 			__netsoc_write_pma2p0_reg(mac_base, lane_no, 0x1b, 0x70);
              }
-- 
1.7.1

