From 8280c545779d1452dc0e6be88546bf1a71adb40f Mon Sep 17 00:00:00 2001
From: ajesh <ajesh@broadcom.com>
Date: Thu, 3 Apr 2014 19:38:03 +0530
Subject: nae: scan nae-1 mdio bus for RXAUI.

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/netlogic/lib/netlib/src/netsoc_nae.c b/arch/mips/netlogic/lib/netlib/src/netsoc_nae.c
index 45c1c92..e0fb246 100644
--- a/arch/mips/netlogic/lib/netlib/src/netsoc_nae.c
+++ b/arch/mips/netlogic/lib/netlib/src/netsoc_nae.c
@@ -3889,7 +3889,16 @@ static void __netsoc_nlp1042c2_init(nae_t* nae)
 	uint32_t cplx_mask_10g_phy = nae->xaui_complex_map | nae->rxaui_complex_map;
 	int phyaddr=0, phybus=0, i; /* use these variables carefully as they are used in headers included below*/
 	int max_ports= current_netsoc->max_ports;
-
+	nae_t* ext_nae;
+	nae_t* temp_nae;
+	if(nae->nae_id == 0)
+	{
+		ext_nae = nae->sibling;
+	}
+	else
+	{
+		ext_nae = nae;
+	}
 	netsoc_api_print(NETSOC_APIDBG_CONFIG,"Initializing external NLP1042 PHY with cmplex mask = 0x%x\n", cplx_mask_10g_phy);
 
 	for(port=0; port<max_ports; port++){
@@ -3906,9 +3915,13 @@ static void __netsoc_nlp1042c2_init(nae_t* nae)
 				netsoc_api_print(NETSOC_APIDBG_CONFIG, "NLP1042C2 XAUI p:%d 0xc205:%x \n", port, data);
 				#include "NETSOC_NLP1042C2_XAUI.h"
 			}else{
-				data = NETSOC_NLM_C45_READ(nae, phybus, phyaddr, 1, 0xc205);
+				//data = NETSOC_NLM_C45_READ(nae, phybus, phyaddr, 1, 0xc205);
+				data = NETSOC_NLM_C45_READ(ext_nae, phybus, phyaddr, 1, 0xc205);
+				temp_nae = nae;
+				nae = ext_nae;
 				netsoc_api_print(NETSOC_APIDBG_CONFIG, "NLP1042C2 RXAUI p:%d 0xc205:%x \n", port, data);
 				#include "NETSOC_NLP1042C2_RXAUI_Dune.h"
+				nae = temp_nae;
 			}
 		}
 	}
-- 
1.7.1

