From 71a370093a2013c00a3d6b9d82fc01be146e09d4 Mon Sep 17 00:00:00 2001
From: Vikas Gupta <vikas.gupta@broadcom.com>
Date: Wed, 9 Apr 2014 11:55:46 +0530
Subject: nae: fix for accessing the neighbour NAE while configuring for NLP1042 PHY.

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/netlogic/lib/netlib/src/netsoc_fdt.c b/arch/mips/netlogic/lib/netlib/src/netsoc_fdt.c
index a9f3916..afbdd21 100644
--- a/arch/mips/netlogic/lib/netlib/src/netsoc_fdt.c
+++ b/arch/mips/netlogic/lib/netlib/src/netsoc_fdt.c
@@ -875,6 +875,10 @@ static void extract_complex_params(void *fdt, int intf_type, char *nae_port_str,
 			for(i=0; i<sizeof(cmplx->ext_phy_bus)/sizeof(uint32_t); i++)
 				cmplx->ext_phy_bus[i] = 0;
 		}
+        	if (GET_PORT_PROP("ext-phy-sbl",&cmplx->ext_phy_sbl, sizeof(cmplx->ext_phy_sbl)) < 0) {
+			for(i=0; i<sizeof(cmplx->ext_phy_sbl)/sizeof(uint32_t); i++)
+				cmplx->ext_phy_sbl[i] = 0;
+		}
 	}
 
 	if (intf_type == SGMII_IF) {
diff --git a/arch/mips/netlogic/lib/netlib/src/netsoc_nae.c b/arch/mips/netlogic/lib/netlib/src/netsoc_nae.c
index e0fb246..8328ba2 100644
--- a/arch/mips/netlogic/lib/netlib/src/netsoc_nae.c
+++ b/arch/mips/netlogic/lib/netlib/src/netsoc_nae.c
@@ -3880,33 +3880,33 @@ int __netsoc_init_ucore(nae_t *nae)
 	return NETSOC_API_SUCCESS;
 }
 
-static void __netsoc_nlp1042c2_init(nae_t* nae)
+static void __netsoc_nlp1042c2_init(nae_t* curr_nae)
 {
 	int data;
 	int bus = 0, port = 0;
 	int dev_addr = 1;
 	int  submode;
-	uint32_t cplx_mask_10g_phy = nae->xaui_complex_map | nae->rxaui_complex_map;
+	uint32_t cplx_mask_10g_phy = curr_nae->xaui_complex_map | curr_nae->rxaui_complex_map;
 	int phyaddr=0, phybus=0, i; /* use these variables carefully as they are used in headers included below*/
 	int max_ports= current_netsoc->max_ports;
-	nae_t* ext_nae;
-	nae_t* temp_nae;
-	if(nae->nae_id == 0)
-	{
-		ext_nae = nae->sibling;
-	}
-	else
-	{
-		ext_nae = nae;
-	}
 	netsoc_api_print(NETSOC_APIDBG_CONFIG,"Initializing external NLP1042 PHY with cmplex mask = 0x%x\n", cplx_mask_10g_phy);
 
 	for(port=0; port<max_ports; port++){
+		nae_t *nae = curr_nae;
+		net_port_t *netport = &nae->ports[port];
+
 		if((nae->ports[port].iftype == RXAUI_IF) || (nae->ports[port].iftype == XAUI_IF)){
-			phyaddr = nae->ports[port].ext_phy_addr;
-			phybus = nae->ports[port].ext_phy_bus;
+			phyaddr = netport->ext_phy_addr;
+			phybus = netport->ext_phy_bus;
 			bus = phybus;
-			submode = nae->ports[port].rxaui_mode;
+			submode = netport->rxaui_mode;
+			if(netport->ext_phy_sbl){
+				nae = nae->sibling;
+				if(!nae){
+					netsoc_api_print(NETSOC_APIDBG_ERROR, "Neighbour NAE is not accessible \n");
+					return;
+				}
+			}
 			netsoc_api_print(NETSOC_APIDBG_CONFIG, "Initializing external NLP1042 PHY phyaddr= 0x%x bus = 0x%x submode=0x%x\n",  phyaddr, phybus, submode);
 			if(submode == NLP1042_PHY_10G_NONE){
 				continue;
@@ -3915,13 +3915,9 @@ static void __netsoc_nlp1042c2_init(nae_t* nae)
 				netsoc_api_print(NETSOC_APIDBG_CONFIG, "NLP1042C2 XAUI p:%d 0xc205:%x \n", port, data);
 				#include "NETSOC_NLP1042C2_XAUI.h"
 			}else{
-				//data = NETSOC_NLM_C45_READ(nae, phybus, phyaddr, 1, 0xc205);
-				data = NETSOC_NLM_C45_READ(ext_nae, phybus, phyaddr, 1, 0xc205);
-				temp_nae = nae;
-				nae = ext_nae;
+				data = NETSOC_NLM_C45_READ(nae, phybus, phyaddr, 1, 0xc205);
 				netsoc_api_print(NETSOC_APIDBG_CONFIG, "NLP1042C2 RXAUI p:%d 0xc205:%x \n", port, data);
 				#include "NETSOC_NLP1042C2_RXAUI_Dune.h"
-				nae = temp_nae;
 			}
 		}
 	}
-- 
1.7.1

