From 2e4a45594b60917b7dd1be641a2df296d977e5aa Mon Sep 17 00:00:00 2001
From: Vikas Gupta <vikas.gupta@broadcom.com>
Date: Thu, 6 Feb 2014 02:35:12 -0800
Subject: xlp5xx: added is_nlm_xlp5xx api.

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/netlogic/lib/syslib/include/nlm_hal_xlp_dev.h b/arch/mips/netlogic/lib/syslib/include/nlm_hal_xlp_dev.h
index f4bea70..a793682 100644
--- a/arch/mips/netlogic/lib/syslib/include/nlm_hal_xlp_dev.h
+++ b/arch/mips/netlogic/lib/syslib/include/nlm_hal_xlp_dev.h
@@ -226,6 +226,7 @@ extern int is_nlm_xlp(unsigned int chipid, unsigned int rev,  unsigned int ext);
 #define is_nlm_xlp532() is_nlm_xlp532_rev(XLP_REVISION_ANY)
 #define is_nlm_xlp524() is_nlm_xlp524_rev(XLP_REVISION_ANY)
 #define is_nlm_xlp516() is_nlm_xlp516_rev(XLP_REVISION_ANY)
+#define is_nlm_xlp5xx()	     is_nlm_xlp5xx_rev(XLP_REVISION_ANY)
 
 #endif /*__XLP_CHIPID_MACROS__ */
 
-- 
1.7.1

