From 04ca872d75adb36c3dd7a0046a68682fb74d8f17 Mon Sep 17 00:00:00 2001
From: Vikas Gupta <vikas.gupta@broadcom.com>
Date: Thu, 6 Feb 2014 23:02:56 -0800
Subject: xlp5xx: Added missing macro is_nlm_xlp5xx_rev

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/netlogic/lib/syslib/include/nlm_hal_xlp_dev.h b/arch/mips/netlogic/lib/syslib/include/nlm_hal_xlp_dev.h
index a793682..e253218 100644
--- a/arch/mips/netlogic/lib/syslib/include/nlm_hal_xlp_dev.h
+++ b/arch/mips/netlogic/lib/syslib/include/nlm_hal_xlp_dev.h
@@ -219,6 +219,7 @@ extern int is_nlm_xlp(unsigned int chipid, unsigned int rev,  unsigned int ext);
 #define is_nlm_xlp524_L(rev)      ( is_nlm_xlp(0x5064, rev, CPU_EXTPID_XLP_5XX_L))
 #define is_nlm_xlp516_L(rev)      ( is_nlm_xlp(0x5044, rev, CPU_EXTPID_XLP_5XX_L))
 
+#define is_nlm_xlp5xx_rev(rev) (is_nlm_xlp5xx_B(rev) || is_nlm_xlp5xx_L(rev))
 #define is_nlm_xlp532_rev(rev) (is_nlm_xlp532_B(rev) || is_nlm_xlp532_L(rev))
 #define is_nlm_xlp524_rev(rev) (is_nlm_xlp524_B(rev) || is_nlm_xlp524_L(rev))
 #define is_nlm_xlp516_rev(rev) (is_nlm_xlp516_B(rev) || is_nlm_xlp516_L(rev))
-- 
1.7.1

