From 720c35c3fd67bf828d903039060579ae4da3ead3 Mon Sep 17 00:00:00 2001
From: Virendra Pathak <vpathak@broadcom.com>
Date: Sat, 22 Feb 2014 13:20:56 +0530
Subject: 5xx: Fixed invalidation of CPU stations during fmn init

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/netlogic/lib/fmnlib/nlm_hal_fmn_config.c b/arch/mips/netlogic/lib/fmnlib/nlm_hal_fmn_config.c
index 709ed4a..027e75d 100644
--- a/arch/mips/netlogic/lib/fmnlib/nlm_hal_fmn_config.c
+++ b/arch/mips/netlogic/lib/fmnlib/nlm_hal_fmn_config.c
@@ -1813,20 +1813,23 @@ void nlm_hal_fmn_init(void *fdt, int node)
 
 		fmn_modify_qsize_credit_config(node, XLP_MSG_BLK_POPQ, 0, XLP_5XX_POPQ_VC_BASE, XLP_5XX_POPQ_VC_LIMIT);
 
-		if (is_nlm_xlp516()) {
-			for (hndl = XLP_MSG_HANDLE_CPU4; hndl < XLP_MSG_HANDLE_CPU8; hndl++) {
-				xlp5xx_fmn_config[hndl].base_vc = XLP_5XX_INVALID_STATION;
-				xlp5xx_fmn_config[hndl].vc_limit =0;
-			}
-			fmn_modify_qsize_credit_config(node, XLP_MSG_BLK_CPU, 4, XLP_CPU0_VC_BASE, XLP_CPU3_VC_LIMIT);
+		if (is_nlm_xlp532()) {
+			fmn_modify_qsize_credit_config(node, XLP_MSG_BLK_CPU, 8, XLP_CPU0_VC_BASE, XLP_CPU7_VC_LIMIT);
 		} else if (is_nlm_xlp524()) {
 			for (hndl = XLP_MSG_HANDLE_CPU6; hndl < XLP_MSG_HANDLE_CPU8; hndl++) {
 				xlp5xx_fmn_config[hndl].base_vc = XLP_5XX_INVALID_STATION;
 				xlp5xx_fmn_config[hndl].vc_limit =0;
 			}
 			fmn_modify_qsize_credit_config(node, XLP_MSG_BLK_CPU, 6, XLP_CPU0_VC_BASE, XLP_CPU5_VC_LIMIT);
+		} else if (is_nlm_xlp516()) {
+			for (hndl = XLP_MSG_HANDLE_CPU4; hndl < XLP_MSG_HANDLE_CPU8; hndl++) {
+				xlp5xx_fmn_config[hndl].base_vc = XLP_5XX_INVALID_STATION;
+				xlp5xx_fmn_config[hndl].vc_limit =0;
+			}
+			fmn_modify_qsize_credit_config(node, XLP_MSG_BLK_CPU, 4, XLP_CPU0_VC_BASE, XLP_CPU3_VC_LIMIT);
 		} else {
-			fmn_modify_qsize_credit_config(node, XLP_MSG_BLK_CPU, 8, XLP_CPU0_VC_BASE, XLP_CPU7_VC_LIMIT);
+			nlm_print("Error: Invalid xlp5xx chip type\n");
+			while(1);
 		}
 
 		fmn_modify_qsize_credit_config(node, XLP_MSG_BLK_GDX, 1, XLP_5XX_GDX0_VC_BASE, XLP_5XX_GDX0_VC_LIMIT);
-- 
1.7.1

