From 1f6942b73f8deb4b224368b9768a3ddf394de1ab Mon Sep 17 00:00:00 2001
From: johuang <john.huang@broadcom.com>
Date: Fri, 7 Mar 2014 14:26:03 -0800
Subject: 5XX: Fix is_xlp5xx()

     Add XLP_REVISION_ANY.
[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/netlogic/lib/syslib/src/nlm_hal_cpu_info.c b/arch/mips/netlogic/lib/syslib/src/nlm_hal_cpu_info.c
index 6443e0c..46c5a11 100644
--- a/arch/mips/netlogic/lib/syslib/src/nlm_hal_cpu_info.c
+++ b/arch/mips/netlogic/lib/syslib/src/nlm_hal_cpu_info.c
@@ -343,7 +343,10 @@ static inline int is_xlp5xx(uint8_t num_cores, uint8_t num_threads, uint32_t rev
 		epid = (uint8_t) ((cfg3 >>27) & 0x7);
 		ncores   = (8 - bitcount(efuse_cfg6() & 0xFF));
 		nthreads =  4 - ((cfg3 >> 24) & 3);
-		
+
+		if (rev == XLP_REVISION_ANY)
+			return 1;
+
 		if (exttype == epid)
 		{
 			if ((num_cores * num_threads) == CPU_NUM_ANY) return 1;
-- 
1.7.1

