From fe07a9dae600f2cc4bcbe047d8459618d032ca38 Mon Sep 17 00:00:00 2001
From: Anurag <anurag.gopinath@broadcom.com>
Date: Tue, 11 Mar 2014 06:52:56 +0530
Subject: XLP5XX : Add SRIO support

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/netlogic/lib/fmnlib/nlm_hal_fmn_config.c b/arch/mips/netlogic/lib/fmnlib/nlm_hal_fmn_config.c
index 027e75d..5a1a255 100644
--- a/arch/mips/netlogic/lib/fmnlib/nlm_hal_fmn_config.c
+++ b/arch/mips/netlogic/lib/fmnlib/nlm_hal_fmn_config.c
@@ -258,7 +258,7 @@ nlm_fmn_config_t xlp5xx_fmn_config[] = {
 [XLP_MSG_HANDLE_REGX]    = {XLP_5XX_RGX_VC_BASE, XLP_5XX_RGX_VC_LIMIT},
 [XLP_MSG_HANDLE_RSA_ECC] = {XLP_5XX_RSA_VC_BASE, XLP_5XX_RSA_VC_LIMIT},
 [XLP_MSG_HANDLE_CRYPTO]  = {XLP_5XX_SEC_VC_BASE, XLP_5XX_SEC_VC_LIMIT},
-[XLP_MSG_HANDLE_SRIO]    = {XLP_5XX_INVALID_STATION, 0},
+[XLP_MSG_HANDLE_SRIO] = {XLP_5XX_SRIO_VC_BASE, XLP_5XX_SRIO_VC_LIMIT},
 [XLP_MSG_HANDLE_CMP]    = {XLP_5XX_CMP_VC_BASE, XLP_5XX_CMP_VC_LIMIT},
 [XLP_MSG_HANDLE_LZS]    = {XLP_5XX_LZS_VC_BASE, XLP_5XX_LZS_VC_LIMIT},
 [XLP_MSG_HANDLE_KBP]    = {XLP_5XX_KBP_VC_BASE, XLP_5XX_KBP_VC_LIMIT}, /*FIXME*/
@@ -302,7 +302,7 @@ static struct fmn_qsize_credit_config fmn_qsize_credit_cfg[XLP9XX_MSG_BLK_MAX] =
 	[XLP_MSG_BLK_POE] =     { "poe",    XLP_POE_VC_BASE,       XLP_POE_VC_LIMIT,       1, 1 },
 	[XLP_MSG_BLK_NAE] =     { "nae",    XLP_NET_VC_BASE,       XLP_NET_VC_LIMIT,       1, 1 },
 	[XLP_MSG_BLK_REGX] =    { "regx",   XLP_3XX_REGEX_VC_BASE, XLP_3XX_REGEX_VC_LIMIT, 1, 1 },
-	[XLP_MSG_BLK_SRIO] =    { "srio",   XLP_3XX_SRIO_VC_BASE,  XLP_3XX_SRIO_VC_LIMIT,  1, 1 },
+	[XLP_MSG_BLK_SRIO] =    { "srio",   XLP_5XX_SRIO_VC_BASE,  XLP_5XX_SRIO_VC_LIMIT,  1, 1 },
 	[XLP_MSG_BLK_LZS] =    { "lzs",   XLP_9XX_LZS_VC_BASE,  XLP_9XX_LZS_VC_LIMIT,  1, 1 },
         [XLP_MSG_BLK_POE_1] =  { "poe1",    XLP_9XX_POE1_VC_BASE,   XLP_9XX_POE1_VC_LIMIT, 1, 1 },
         [XLP_MSG_BLK_NAE_1] =  { "nae1",    XLP_9XX_NET1_VC_BASE,   XLP_9XX_NET1_VC_LIMIT, 1, 1 },
@@ -1804,7 +1804,6 @@ void nlm_hal_fmn_init(void *fdt, int node)
 		fmn_modify_qsize_credit_config(node, XLP_MSG_BLK_PCIE3, 1, XLP_9XX_PCIE3_VC_BASE, XLP_9XX_PCIE3_VC_LIMIT);
 	} else if (is_nlm_xlp5xx()) {
 		
-		fmn_invalidate_blocks(node, XLP_MSG_BLK_SRIO);
 		fmn_invalidate_blocks(node, XLP_MSG_BLK_GDX1);
 		fmn_invalidate_blocks(node, XLP_MSG_BLK_NAE_1);
 		fmn_invalidate_blocks(node, XLP_MSG_BLK_POE_1);
@@ -1837,6 +1836,7 @@ void nlm_hal_fmn_init(void *fdt, int node)
 		fmn_modify_qsize_credit_config(node, XLP_MSG_BLK_CMP, 1, XLP_5XX_CMP_VC_BASE, XLP_5XX_CMP_VC_LIMIT);
 		fmn_modify_qsize_credit_config(node, XLP_MSG_BLK_RSA_ECC, 1, XLP_5XX_RSA_VC_BASE, XLP_5XX_RSA_VC_LIMIT);
 		fmn_modify_qsize_credit_config(node, XLP_MSG_BLK_CRYPTO, 1, XLP_5XX_SEC_VC_BASE, XLP_5XX_SEC_VC_LIMIT);
+		fmn_modify_qsize_credit_config(node, XLP_MSG_BLK_SRIO, 1, XLP_5XX_SRIO_VC_BASE, XLP_5XX_SRIO_VC_LIMIT);
 
 		fmn_modify_qsize_credit_config(node, XLP_MSG_BLK_NAE, 1, XLP_5XX_NET0_VC_BASE, XLP_5XX_NET0_VC_LIMIT);
 		fmn_modify_qsize_credit_config(node, XLP_MSG_BLK_POE, 1, XLP_5XX_POE0_VC_BASE, XLP_5XX_POE0_VC_LIMIT);
diff --git a/arch/mips/netlogic/lib/syslib/include/nlm_hal_xlp_dev.h b/arch/mips/netlogic/lib/syslib/include/nlm_hal_xlp_dev.h
index fbbbc9c..c716742 100644
--- a/arch/mips/netlogic/lib/syslib/include/nlm_hal_xlp_dev.h
+++ b/arch/mips/netlogic/lib/syslib/include/nlm_hal_xlp_dev.h
@@ -478,8 +478,11 @@ extern int is_nlm_xlp(unsigned int chipid, unsigned int rev,  unsigned int ext);
 #define XLP_5XX_RSA_VC_LIMIT	0x1C3
 #define XLP_5XX_RGX_VC_BASE	0x1CC
 #define XLP_5XX_RGX_VC_LIMIT	0x1CF
-#define XLP_5XX_KBP_VC_BASE	0x1DC 
-#define XLP_5XX_KBP_VC_LIMIT	0x1DF 
+#define XLP_5XX_KBP_VC_BASE	0x1E6 
+#define XLP_5XX_KBP_VC_LIMIT	0x1EA  //Changed AGN
+
+#define XLP_5XX_SRIO_VC_BASE    0x1DC 
+#define XLP_5XX_SRIO_VC_LIMIT   0x1E5
 
 #define XLP_5XX_POE0_VC_BASE	0x200
 #define XLP_5XX_POE0_VC_LIMIT	0x207
@@ -608,6 +611,7 @@ extern int is_nlm_xlp(unsigned int chipid, unsigned int rev,  unsigned int ext);
 
 #define XLP_NA_REG_BLOCK_SIZE       0x2000 /* 8KB */
 #define PCI_MEM_BAR_0               0x0004
+#define PCI_MEM_BAR_1               0x0005
 #define MAX_NUMBER_OF_SERDES_LANE   0x8
 
 
diff --git a/arch/mips/netlogic/lib/syslib/include/nlm_srio_xlp.h b/arch/mips/netlogic/lib/syslib/include/nlm_srio_xlp.h
index 881a15f..4ee86fd 100644
--- a/arch/mips/netlogic/lib/syslib/include/nlm_srio_xlp.h
+++ b/arch/mips/netlogic/lib/syslib/include/nlm_srio_xlp.h
@@ -1,4 +1,3 @@
-
 /*-
  * Copyright (c) 2003-2012 Broadcom Corporation
  * All Rights Reserved
@@ -104,12 +103,14 @@
 /* SRIO VC */
 #define XLP_SRIO_TXVC_BASE	280
 #define XLP_SRIO_ADDRQ_VC	288
+#define XLP_SRIO_5XX_ADDRQ_VC   0x1E0 //480	
 
 #define NLM_SRIO_BUS_NUM 	0
 #define NLM_SRIO_DEV_NUM 	5
 #define NLM_SRIO_FUN_NUM 	4 
 
-#define NLM_SRIO_CFG_BASE	 ( 0x18000000 | (NLM_SRIO_DEV_NUM << 15) | (NLM_SRIO_FUN_NUM << 12))   
+#define NLM_3XX_SRIO_CFG_BASE       ( 0x18000000 | (NLM_SRIO_DEV_NUM << 15) | (NLM_SRIO_FUN_NUM << 12))
+#define NLM_SRIO_CFG_BASE	 ( 0x18000000 | (0x1<<20) | (NLM_SRIO_DEV_NUM << 15) | (NLM_SRIO_FUN_NUM << 12)) 
 /* SriComplex Configuration Register */
 
 #define SRICOMCONFIG    	0x40
@@ -247,10 +248,15 @@
 #define DATAMSGEN1               0x63
 #define DATAMSGEN2               0x64
 #define DATAMSGEN3               0x65
+#define DATAMSGEN4               0x66
+#define DATAMSGEN5               0x67
+#define DATAMSGEN6               0x68
+#define DATAMSGEN7               0x69
 
 /* Destination ID for Data Message Distribution Table */
 /* Total of 256 entries */
 #define DATAMSG_DESTID		0x66
+#define DATAMSG_DESTID_5XX		0x6A
 
 /* Switch CAM Table */
 /* Total of 32 entries */
@@ -269,6 +275,7 @@
 /* Direct IO Operation Table */
 /* Totll of 32 entries */
 #define DIRECTIO_CMD		0x187
+#define DIRECTIO_CMD_5XX		0x18B
    #define DIRIO_CMD_DESTID	0
    #define DIRIO_CMD_PORT	16
    #define DIRIO_CMD_TT		18
@@ -280,24 +287,31 @@
 /* Diret IO Response Table */
 /* Total of 32 entires */
 #define DIRECTIO_RESP   	0x1a7
+#define DIRECTIO_RESP_5XX   	0x1AB
 
 /* Diret IO Ctrl */
 #define DIRECTIO_CTRL		0x1c7
+#define DIRECTIO_CTRL_5XX	0x1CB
 
 /* Byte Swap Register */
 #define BYTE_SWAP 		0x1ca
+#define BYTE_SWAP_5XX 		0x1ce
 
 /* Alloc Base Register */
 #define ALLOC_BASE		0x1cb
+#define ALLOC_BASE_5XX		0x1cf
 
 /* Alloc Limit Register */
 #define ALLOC_LIMIT		0x1cc
+#define ALLOC_LIMIT_5XX		0x1d0
 
 /* Read Exclusive Base Register */
 #define RDEXC_BASE		0x1cd
+#define RDEXC_BASE_5XX		0x1d1
 
 /* Read Exclusive Limit Register */
 #define RDEXC_LIMIT		0x1ce
+#define RDEXC_LIMIT_5XX		0x1d2
 
 /* Sri Error Source 0 Register */
 #define SRIERR_SRC0		0x1cf
@@ -319,20 +333,21 @@
 
 /* FMB Egress SECC Info Register */
 #define SRI_SECCINF        	0x1d3
+#define SRI_SECCINF_5XX        	0x1d7
 
 /* FMB Egress MECC Info Register */
 #define SRI_MECCINF		0x1d4
+#define SRI_MECCINF_5XX		0x1d8
 
 /* Sri Response Timeout Info 0 Register */
 #define SRI_RSPTMINF0		0x1d5
-
 #define SRI_RSPTMINF1           0x1d6
 #define SRI_RSPTMINF2           0x1d7
-
 #define SRI_RSPTMINF3           0x1d8
 
 
 /*RIO Specific registers */
+#define RIO_PE_FEAT             0x10
 #define RIO_SRC_OP		0x18
 #define RIO_PORTMAINT_HDR	0x100
 #define RIO_LNKTOUT_CSR		0x120
@@ -349,23 +364,30 @@
 #define RIO_PORT2CTRL_CSR       0x19c               
 #define RIO_PORT3CTRL_CSR       0x1bc               
 
+#define RIO_PLM_SP0_IMP_SPEC_CTL 0x10080
+#define RIO_PLM_SP1_IMP_SPEC_CTL 0x10100
+#define RIO_PLM_SP2_IMP_SPEC_CTL 0x10180
+#define RIO_PLM_SP3_IMP_SPEC_CTL 0x10200
 /*/ SBB */
 
 #define DIRIO_RESP_MASK		0xF
 /* dirio 40 bit phys addr */
 #define DIRIO_PHYS_ADDR		0
 #define DIRIO_PHYS_THRD		32
-#define DIRIO_PHYS_OP		37
+
+#define DIRIO_PHYS_OP_3XX       37
+#define DIRIO_PHYS_OP		36
 #define DIRIO_PHYS_SRIO		39
 
 /* dirio opcode */
-#define DIRIO_OP_NREAD		0
-#define DIRIO_OP_MREAD		3
+#define DIRIO_OP_MREAD_3XX		3
+#define DIRIO_OP_NREAD          0
+#define DIRIO_OP_MREAD          1
 
-#define DIRIO_OP_NWRITE		0
-#define DIRIO_OP_NWRITE_R	1
-#define DIRIO_OP_MWRITE		2
-#define DIRIO_OP_PWRITE		3
+#define DIRIO_OP_NWRITE         0
+#define DIRIO_OP_NWRITE_R       1
+#define DIRIO_OP_MWRITE         2
+#define DIRIO_OP_PWRITE         3
 
 
 #define SRIO_DIRIO_CMD(dst, port, tt, crc, crf, prio, hop) \
@@ -373,7 +395,6 @@
 	 (crf << DIRIO_CMD_CRF) | (crc << DIRIO_CMD_CRC) | \
 	 (tt << DIRIO_CMD_TT) | (port << DIRIO_CMD_PORT) | dst)
 
-	
 #define SRIO_DIRIO_40BITPHYS(addr, cpu, opcode) \
 	(uint64_t)((1ULL << DIRIO_PHYS_SRIO) | ((uint64_t)opcode << DIRIO_PHYS_OP) | ((uint64_t)cpu << DIRIO_PHYS_THRD) | addr)
 
@@ -449,11 +470,10 @@
 #define SRICOMCONFIG_RXDMSG_TIMER_EN	0x00000010
 #define SRICOMCONFIG_TXDMSG_TIMER_EN	0x00000040
 
-#define DOORBELL_DESTID		0x1F0
-
 #define SRIO_DATAMSG_SIZE	0x1D9
-
+#define SRIO_DATAMSG_SIZE_5XX	0x1DD
 #define SRIO_SUPPRESS_CTRL      0x1E8
+#define SRIO_SUPPRESS_CTRL_5XX      0x1FE
 #define TYPE2_SUPRS_EN          0x0001
 #define TYPE5_NORESP_SUPRS_EN   0x0002
 #define TYPE5_RESP_SUPRS_EN     0x0004
@@ -462,10 +482,16 @@
 #define TYPE10_SUPRS_EN         0x0020
 #define TYPE11_SUPRS_EN         0x0040
 
+#define DOORBELL_DESTID		0x1F0
 #define RXEDMESG_TIMEOUT	0x210		
 #define TXEDMESG_TIMEOUT	0x212
+#define FLUSHADDRQ              0x278
 
-#define FLUSHADDRQ             0x278
+#define DOORBELL_ENABLE_5XX     0x203 //New reg , Check AGN
+#define DOORBELL_DESTID_5XX		0x204
+#define RXEDMESG_TIMEOUT_5XX	0x284		
+#define TXEDMESG_TIMEOUT_5XX	0x286
+#define FLUSHADDRQ_5XX              0x2e9
 
 #define MAX_PACKET_LEN			4096
 #define MAX_P2PDMSG_PACKET_LEN		1024
@@ -482,6 +508,9 @@
 #define SRIO_INJECT_CRC		(0x2000000000000000ULL)
 #define SRIO_SEND_FREEBACK	(0x0400000000000000ULL)
 #define SRIO_SEND_RESP		(0x0200000000000000ULL)
+#define XLP5XX_SRIO_SEND_RESP_ALL       (0x0200000000000000ULL)
+#define XLP5XX_SRIO_SEND_RESP_ERROR     (0x0000000000000000ULL)
+#define XLP5XX_SRIO_SEND_RESP_SUPPRESS  (0x0400000000000000ULL)
 #define SRIO_NODMA_RESP		(0x0000000080000000ULL)
 
 #define	SRIO_P2P_POS		63	/* P2P or P2D desc */
@@ -547,7 +576,6 @@
 #define SRIO_STREAM_SIZE_MASK	0xffffULL
 #define SRIO_STREAM_ID_MASK	0xffffULL
 
-
 #define SRIO_DMSG_SIZE_POS	52
 #define SRIO_DMSG_LEN_POS	48
 
-- 
1.7.1

