From 770ce787710409994ed492cdceb4372a605ec89b Mon Sep 17 00:00:00 2001
From: Sandip Matte <sandipm@broadcom.com>
Date: Mon, 10 Mar 2014 18:17:49 +0530
Subject: Add LZS station information in the fmn config structure

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/netlogic/lib/fmnlib/nlm_hal_fmn_config.c b/arch/mips/netlogic/lib/fmnlib/nlm_hal_fmn_config.c
index 5a1a255..79bc85f 100644
--- a/arch/mips/netlogic/lib/fmnlib/nlm_hal_fmn_config.c
+++ b/arch/mips/netlogic/lib/fmnlib/nlm_hal_fmn_config.c
@@ -1834,6 +1834,7 @@ void nlm_hal_fmn_init(void *fdt, int node)
 		fmn_modify_qsize_credit_config(node, XLP_MSG_BLK_GDX, 1, XLP_5XX_GDX0_VC_BASE, XLP_5XX_GDX0_VC_LIMIT);
 		fmn_modify_qsize_credit_config(node, XLP_MSG_BLK_REGX, 1, XLP_5XX_RGX_VC_BASE, XLP_5XX_RGX_VC_LIMIT);
 		fmn_modify_qsize_credit_config(node, XLP_MSG_BLK_CMP, 1, XLP_5XX_CMP_VC_BASE, XLP_5XX_CMP_VC_LIMIT);
+		fmn_modify_qsize_credit_config(node, XLP_MSG_BLK_LZS, 1, XLP_5XX_LZS_VC_BASE, XLP_5XX_LZS_VC_LIMIT);
 		fmn_modify_qsize_credit_config(node, XLP_MSG_BLK_RSA_ECC, 1, XLP_5XX_RSA_VC_BASE, XLP_5XX_RSA_VC_LIMIT);
 		fmn_modify_qsize_credit_config(node, XLP_MSG_BLK_CRYPTO, 1, XLP_5XX_SEC_VC_BASE, XLP_5XX_SEC_VC_LIMIT);
 		fmn_modify_qsize_credit_config(node, XLP_MSG_BLK_SRIO, 1, XLP_5XX_SRIO_VC_BASE, XLP_5XX_SRIO_VC_LIMIT);
-- 
1.7.1

