From 863ecc1717098b852edfd52c14136f8b1f9fd768 Mon Sep 17 00:00:00 2001
From: Saswat Kumar Dash <saswat.dash@broadcom.com>
Date: Sun, 4 May 2014 22:41:33 -0700
Subject: xlp5xx: modified check for netsoc lib param validation

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/netlogic/lib/netlib/include/netsoc_nae.h b/arch/mips/netlogic/lib/netlib/include/netsoc_nae.h
index 4ec3bdf..d56f1af 100644
--- a/arch/mips/netlogic/lib/netlib/include/netsoc_nae.h
+++ b/arch/mips/netlogic/lib/netlib/include/netsoc_nae.h
@@ -1412,7 +1412,7 @@ static inline int  __netsoc_validate_lib_param (struct netsoc_lib_param *param)
 {
 	 if(param == NULL) return 0;
 
-        if(!(is_nlm_xlp9xx()) && (param->kr_en)){
+        if(!(IS_NLM_XLP9XX_FAMILY) && (param->kr_en)){
                         return -1;
         }else{
                 return 1;
-- 
1.7.1

