From 43feb817661c1cc29a5d8685e00ebe1af770a7d0 Mon Sep 17 00:00:00 2001
From: Prasanna Mumbai <prasanna.mumbai@broadcom.com>
Date: Thu, 22 May 2014 08:18:23 -0700
Subject: 5xx: change the check is_nlm_9xx to IS_NLM_9XX_FAMILY

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/netlogic/lib/syslib/src/brcm_xlp_cpu.c b/arch/mips/netlogic/lib/syslib/src/brcm_xlp_cpu.c
index 0d9b249..466ce4a 100644
--- a/arch/mips/netlogic/lib/syslib/src/brcm_xlp_cpu.c
+++ b/arch/mips/netlogic/lib/syslib/src/brcm_xlp_cpu.c
@@ -108,7 +108,7 @@ inline int brcm_get_max_node_num(void)
 
 inline int brcm_get_popq_vc_base(void)
 {
-	if (is_nlm_xlp9xx()) {
+	if (IS_NLM_XLP9XX_FAMILY) {
 		return XLP_9XX_POPQ_VC_BASE;
 	} else {
 		return XLP_POPQ_VC_BASE;
@@ -128,7 +128,7 @@ inline int brcm_get_popq_vc_base(void)
 
 inline int brcm_get_poe_vc_base(int poe_blk)
 {
-	if(is_nlm_xlp9xx()) {
+	if(IS_NLM_XLP9XX_FAMILY) {
 		if (poe_blk == 0) {
 			return XLP_9XX_POE0_VC_BASE;
 		} else {
@@ -171,7 +171,7 @@ inline uint64_t brcm_get_poe_pcie_base(int node, int poe_block)
 	}
 	bus = (is_nlm_xlp9xx()) ? 1 : 0 ;
 	if (poe_block == 0) {
-		if (is_nlm_xlp9xx()) {
+		if (IS_NLM_XLP9XX_FAMILY) {
 			poe_pcie_base = xlp9xx_poe0_base[node];
 		} else {
 			poe_pcie_base = nlm_hal_get_dev_base(node, bus, XLP_POE_DEVICE, XLP_POE_FUNC);
-- 
1.7.1

