From ea5af04b75f13ecb747c85047e79b9dcc480c16e Mon Sep 17 00:00:00 2001
From: johuang <john.huang@broadcom.com>
Date: Fri, 23 May 2014 15:40:49 -0700
Subject: xlp5xx: fixed the efuse_config register base address

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/netlogic/lib/syslib/include/nlm_hal_xlp_dev.h b/arch/mips/netlogic/lib/syslib/include/nlm_hal_xlp_dev.h
index 7f4d212..46bfeb9 100644
--- a/arch/mips/netlogic/lib/syslib/include/nlm_hal_xlp_dev.h
+++ b/arch/mips/netlogic/lib/syslib/include/nlm_hal_xlp_dev.h
@@ -512,6 +512,11 @@ extern int is_nlm_xlp(unsigned int chipid, unsigned int rev,  unsigned int ext);
 #define XLP_5XX_NET1_RX_VC_BASE      712
 #define XLP_5XX_NET1_RX_VC_LIMIT     727
 
+#define XLP5XX_EFUSE_BUS			0x1
+#define XLP5XX_EFUSE_DEVICE			0x6
+#define XLP5XX_EFUSE_FUNC			0x1
+
+#define XLP5XX_PCIE_DEV_BASE(SOC)		(((SOC##_DEVICE) << 15) | (SOC##_FUNC << 12))
 
 #define XLP9XX_SAE_RSA_BUS_NUM		(1<<20)
 #define XLP9XX_PCIE_DEV_BASE(SOC)		(((SOC##_DEVICE) << 15) | (SOC##_FUNC << 12))
diff --git a/arch/mips/netlogic/lib/syslib/src/nlm_hal_cpu_info.c b/arch/mips/netlogic/lib/syslib/src/nlm_hal_cpu_info.c
index 46c5a11..ca0190f 100644
--- a/arch/mips/netlogic/lib/syslib/src/nlm_hal_cpu_info.c
+++ b/arch/mips/netlogic/lib/syslib/src/nlm_hal_cpu_info.c
@@ -332,16 +332,19 @@ static inline int xlp3xx_get_num_of_cores(uint32_t core_mask, uint32_t epid)
 
 static inline int is_xlp5xx(uint8_t num_cores, uint8_t num_threads, uint32_t rev, uint32_t exttype)
 {
-	uint32_t pid, cfg3;
+	uint32_t pid;
 	uint8_t epid;
 	int ncores, nthreads;
 	
 	pid=get_proc_id();
 	if( pid == CHIP_PROCESSOR_ID_XLP_5XX )
 	{
-		cfg3 = efuse_cfg3();
+		volatile uint32_t *fuse_mmio = xlp9xx_cpu_io_mmio(-1, XLP5XX_PCIE_DEV_BASE(XLP5XX_EFUSE));
+		volatile uint32_t cfg3 = fuse_mmio[0xc3];
+		volatile uint32_t cfg6 = fuse_mmio[0xc6];
+
 		epid = (uint8_t) ((cfg3 >>27) & 0x7);
-		ncores   = (8 - bitcount(efuse_cfg6() & 0xFF));
+		ncores   = (8 - bitcount(cfg6 & 0xFF));
 		nthreads =  4 - ((cfg3 >> 24) & 3);
 
 		if (rev == XLP_REVISION_ANY)
-- 
1.7.1

