From cf28a84726d13a1d5fb825ea08ff3a3e55261358 Mon Sep 17 00:00:00 2001
From: Abhishek Joshi <abjoshi@broadcom.com>
Date: Tue, 15 Apr 2014 18:25:53 +0530
Subject: fmn: avoid allocation of on-chip memory to disabled output queues

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/netlogic/lib/fmnlib/nlm_hal_fmn_config.c b/arch/mips/netlogic/lib/fmnlib/nlm_hal_fmn_config.c
index 79bc85f..c91ffc2 100644
--- a/arch/mips/netlogic/lib/fmnlib/nlm_hal_fmn_config.c
+++ b/arch/mips/netlogic/lib/fmnlib/nlm_hal_fmn_config.c
@@ -1335,11 +1335,13 @@ int nlm_hal_setup_outq(void *fdt, int node, int max_nodes)
 					q_enable = 0ULL;
 					fmn_cfg_value[qid] = 0;
 					oq_num32pages = 2;
+					continue;
 				}
 				else if ((station == XLP_MSG_BLK_NAE_1) && (qid >= (XLP_9XX_NET1_TX_VC_BASE + num_valid_queues)) && (qid < XLP_9XX_NET1_RX_VC_BASE) ) {
 					fmn_cfg_value[qid] = 0;
 					q_enable = 0ULL;
 					oq_num32pages = 2;
+					continue;
 				}
                                 else if ((station == XLP_MSG_BLK_CRYPTO) && ((qid >= (XLP_9XX_SEC_VC_BASE + num_valid_queues)) && (qid <= XLP_9XX_SEC_VC_LIMIT))) {
 #ifdef FMN_DEBUG
@@ -1348,6 +1350,7 @@ int nlm_hal_setup_outq(void *fdt, int node, int max_nodes)
                                         fmn_cfg_value[qid] = 0;
                                         q_enable = 0ULL;
                                         oq_num32pages = 2;
+					continue;
                                 }
 
 				/* enable output queue and setup on-chip output queue */
-- 
1.7.1

