From 6788d8a57475c66c28fea55e5d6e2bb0de770cd6 Mon Sep 17 00:00:00 2001
From: Kevin James <kevinj@broadcom.com>
Date: Thu, 6 Mar 2014 05:34:47 -0800
Subject: nae: removed 32k alignment for NAE freein spill

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/netlogic/lib/netlib/src/netsoc_nae.c b/arch/mips/netlogic/lib/netlib/src/netsoc_nae.c
index ccb015f..ab20c18 100644
--- a/arch/mips/netlogic/lib/netlib/src/netsoc_nae.c
+++ b/arch/mips/netlogic/lib/netlib/src/netsoc_nae.c
@@ -4087,8 +4087,6 @@ static int __netsoc_config_freein_fifo(nae_t *nae)
                                 netsoc_api_print(NETSOC_APIDBG_CONFIG, "Freein fifo cfg %d spl_claddrhi %x spillsize %d cachelines\n",
 								 i, ((reg >> 30) & 0x3), (reg & 0x3fffffff));
                                 spill_addr += spillsz;
-                                /* align to 32k, already >> 6 is done above, so align to 1k */
-                                spill_addr = (spill_addr + 1023) & (~1023);
                         } else {
                                 netsoc_write_nae_reg(nae->nae_base, FREE_SPILL0_MEM_CFG, 0);
                                 netsoc_write_nae_reg(nae->nae_base, FREE_SPILL1_MEM_CFG, 0);
-- 
1.7.1

