From 705f99a443688ba06c459ae833748ecd7c4a02df Mon Sep 17 00:00:00 2001
From: Abhishek Joshi <abjoshi@broadcom.com>
Date: Wed, 30 Apr 2014 19:41:41 +0530
Subject: fmn: corrected popq priority class configuration function

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/netlogic/lib/fmnlib/nlm_hal_fmn_config.c b/arch/mips/netlogic/lib/fmnlib/nlm_hal_fmn_config.c
index c91ffc2..58675cc 100644
--- a/arch/mips/netlogic/lib/fmnlib/nlm_hal_fmn_config.c
+++ b/arch/mips/netlogic/lib/fmnlib/nlm_hal_fmn_config.c
@@ -828,8 +828,11 @@ int brcm_hal_config_popq_class(int node, int popq_class, uint8_t width, uint16_t
 {
 	uint64_t val = 0;
 	int offset = 0;
+	unsigned long fmn_base_local[NLM_MAX_NODES] = {0};
+	unsigned long long mask = ~0xf;
 
 	if (is_nlm_xlp9xx()) {
+		fmn_base_local[node] = mask & nlh_read_cfg_reg32(xlp9xx_fmn_base[node] + 0x10);
 		if ((popq_class < 0) || (popq_class > POPQ_MAX_CLASS_NUM)) {
 			nlm_print("ERROR: popq class number out of range. \
 					popq_class: %d, popq class range: 0 - %d\n", \
@@ -838,10 +841,10 @@ int brcm_hal_config_popq_class(int node, int popq_class, uint8_t width, uint16_t
 		}
 		offset = popq_class * POPQ_CLASS_CONFIG_OFFSET;
 		val |= ((1ULL << 63) | ((0x7f & (~width)) << 16) | (base_id & 0x1fff));
-		nlh_write_cfg_reg64(xlp_fmn_base[node] + (POPQ_CLASS_CONFIG_BASE + offset), val);
+		nlh_write_cfg_reg64(fmn_base_local[node] + (POPQ_CLASS_CONFIG_BASE + offset), val);
 #if FMN_DEBUG
 		nlm_print("%s \n", __func__);
-		val = nlh_read_cfg_reg64(xlp_fmn_base[node] + (POPQ_CLASS_CONFIG_BASE + offset));
+		val = nlh_read_cfg_reg64(fmn_base_local[node] + (POPQ_CLASS_CONFIG_BASE + offset));
 		nlm_print("val: 0x%llx \n", val);
 #endif
 		return 0;
-- 
1.7.1

