From 05f7194e19dbfe62e9446cc24752a9a4bf25e4ee Mon Sep 17 00:00:00 2001
From: Debayan Ghosh <debayan.ghosh@broadcom.com>
Date: Fri, 2 May 2014 22:33:13 +0530
Subject: lzs: Implemented common compression api(s) for lzs engine

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/netlogic/lib/syslib/include/nlm_hal.h b/arch/mips/netlogic/lib/syslib/include/nlm_hal.h
index d18c048..e12fd6b 100644
--- a/arch/mips/netlogic/lib/syslib/include/nlm_hal.h
+++ b/arch/mips/netlogic/lib/syslib/include/nlm_hal.h
@@ -65,6 +65,7 @@ enum freq_config {
         NLM_SAE,
         NLM_DTRE,
         NLM_CDE,
+        NLM_LZS
 };
 
 struct nlm_netl_proc_info{
diff --git a/arch/mips/netlogic/lib/syslib/src/nlm_hal.c b/arch/mips/netlogic/lib/syslib/src/nlm_hal.c
index 20e3304..f5192ae 100644
--- a/arch/mips/netlogic/lib/syslib/src/nlm_hal.c
+++ b/arch/mips/netlogic/lib/syslib/src/nlm_hal.c
@@ -596,7 +596,17 @@ int nlm_hal_get_fdt_freq(void *fdt, int type)
 		else
 			ret = freq;
 #ifdef FREQ_DEBUG
-		nlm_print("sae frequency is %d\n", ret);
+		nlm_print("cde frequency is %d\n", ret);
+#endif
+		break;
+	case NLM_LZS:
+		if(copy_fdt_prop(fdt, path_str, "lzs", PROP_CELL, &freq, sizeof(uint32_t)) < 0)
+			nlm_print("Unable to find the frequency in the FDT file for type:%d, \
+					using the default value\n", type);
+		else
+			ret = freq;
+#ifdef FREQ_DEBUG
+		nlm_print("cde frequency is %d\n", ret);
 #endif
 		break;
 	default:
-- 
1.7.1

