From 76b63e9c9f704292830f05d3abed8fef06cd69e5 Mon Sep 17 00:00:00 2001
From: Vikas Gupta <vikas.gupta@broadcom.com>
Date: Sun, 11 May 2014 23:20:36 -0700
Subject: nae: Increased the divisor value for XGIG MDIO to support NLP PHY.

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/netlogic/lib/netlib/src/netsoc_nae.c b/arch/mips/netlogic/lib/netlib/src/netsoc_nae.c
index 5a6a908..b6d1603 100644
--- a/arch/mips/netlogic/lib/netlib/src/netsoc_nae.c
+++ b/arch/mips/netlogic/lib/netlib/src/netsoc_nae.c
@@ -116,7 +116,7 @@ static uint32_t nae_get_INT_MDIO_DIV(void)
 
 static uint32_t nae_get_EXT_XG_MDIO_DIV(void)
 {
-	return ((0x7F << EXT_XG_MDIO_CTRL_XDIV_POS) | (2 << EXT_XG_MDIO_CTRL_MCDIV_POS));
+	return ((0x7F << EXT_XG_MDIO_CTRL_XDIV_POS) | (3 << EXT_XG_MDIO_CTRL_MCDIV_POS));
 }
 
 /*
-- 
1.7.1

