From 2a609297ba2297b1587f0d849e71d4c92aca55b3 Mon Sep 17 00:00:00 2001
From: Vikas Gupta <vikas.gupta@broadcom.com>
Date: Thu, 5 Jun 2014 01:21:29 -0700
Subject: syslib: build fix.

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/netlogic/lib/syslib/include/xlp9xx_sys.h b/arch/mips/netlogic/lib/syslib/include/xlp9xx_sys.h
index 74c4bfc..090b027 100644
--- a/arch/mips/netlogic/lib/syslib/include/xlp9xx_sys.h
+++ b/arch/mips/netlogic/lib/syslib/include/xlp9xx_sys.h
@@ -158,6 +158,8 @@ static inline void xlp9xx_set_clkdev(int node, int dev_id, int sel, int div)
                         break;
                 }
         }
+}
+
 static void inline xlp9xx_setup_devpll(uint64_t sys_clk_base, int dev_pll, int freq)
 {
         volatile uint32_t pll_change, mul;
-- 
1.7.1

