From bdfdfc290968e57de3c99611a7060b231299c2a6 Mon Sep 17 00:00:00 2001
From: Abhishek Joshi <abjoshi@broadcom.com>
Date: Fri, 30 May 2014 10:14:15 +0530
Subject: fmn: correction for configuration of kbp station on 5xx

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/netlogic/lib/fmnlib/nlm_hal_fmn_config.c b/arch/mips/netlogic/lib/fmnlib/nlm_hal_fmn_config.c
index 2a2a3c8..b15ebf6 100644
--- a/arch/mips/netlogic/lib/fmnlib/nlm_hal_fmn_config.c
+++ b/arch/mips/netlogic/lib/fmnlib/nlm_hal_fmn_config.c
@@ -1020,10 +1020,10 @@ static void nlm_hal_write_fmn_credit(int node, int max_nodes)
 	if (!is_nlm_xlp9xx()) {
 		fmn_invalidate_blocks(node, XLP_MSG_BLK_POE_1);
 		fmn_invalidate_blocks(node, XLP_MSG_BLK_NAE_1);
-		fmn_invalidate_blocks(node, XLP_MSG_BLK_KBP);
 
 		if (!is_nlm_xlp5xx()) {
 			fmn_invalidate_blocks(node, XLP_MSG_BLK_ALE);
+			fmn_invalidate_blocks(node, XLP_MSG_BLK_KBP);
 		}
 	}
 
@@ -1853,6 +1853,7 @@ void nlm_hal_fmn_init(void *fdt, int node)
 		fmn_modify_qsize_credit_config(node, XLP_MSG_BLK_PCIE1, 1, XLP_5XX_PCIE1_VC_BASE, XLP_5XX_PCIE1_VC_LIMIT);
 		fmn_modify_qsize_credit_config(node, XLP_MSG_BLK_PCIE2, 1, XLP_5XX_PCIE2_VC_BASE, XLP_5XX_PCIE2_VC_LIMIT);
 		fmn_modify_qsize_credit_config(node, XLP_MSG_BLK_PCIE3, 1, XLP_5XX_PCIE3_VC_BASE, XLP_5XX_PCIE3_VC_LIMIT);
+		fmn_modify_qsize_credit_config(node, XLP_MSG_BLK_KBP, 1, XLP_5XX_KBP_VC_BASE, XLP_5XX_KBP_VC_LIMIT);
 	}
 
 	update_fmn_config();
-- 
1.7.1

