From 82cb93f8afb24e2f624baa21474f7f854bda8ab5 Mon Sep 17 00:00:00 2001
From: Vikas Gupta <vikas.gupta@broadcom.com>
Date: Mon, 9 Jun 2014 12:56:37 +0530
Subject: syslib: build fix.

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/netlogic/lib/syslib/src/nlm_hal.c b/arch/mips/netlogic/lib/syslib/src/nlm_hal.c
index 271a0a6..2beb442 100644
--- a/arch/mips/netlogic/lib/syslib/src/nlm_hal.c
+++ b/arch/mips/netlogic/lib/syslib/src/nlm_hal.c
@@ -338,34 +338,6 @@ __inline__ int nlm_hal_is_xlp_le(void)
 	return little_endian;
 }
 
-static void inline xlp9xx_setup_devpll(uint64_t sys_clk_base, int dev_pll, int freq)
-{
-	volatile uint32_t pll_change, mul, mul0;
-	switch (freq) {
-		case 400: mul = 0 ; break; 
-		case 500: mul = 3 ; break; 
-		case 667: mul = 8 ; break; 
-		case 700: mul = 0x9 ; break; 
-		case 767: mul = 0xb ; break; 
-		default: 
-			  mul=0; break;
-	}
-	nlm_print("dev _pll = %d frq= %d mul = %d\n", dev_pll, freq, mul);
-	
-	mul0 = syssoc_read_reg(sys_clk_base, DEV_PLLCTRL_1 + (4*dev_pll));
-	nlm_print("DEV PLL[%d] present freq  =%d\n", dev_pll, 400 + 33*mul0);
-
-	syssoc_write_reg(sys_clk_base, DEV_PLLCTRL_1 + (4*dev_pll), mul);
-	syssoc_write_reg(sys_clk_base, XLP9XX_SYS_PLLCHG_CTRL, (1<<(2+dev_pll)));
-	pll_change = (1<<(2+dev_pll));
-	while (pll_change != 0) {
-		pll_change = syssoc_read_reg(sys_clk_base, XLP9XX_SYS_PLLCHG_CTRL);
-		//nlm_print("CLK_REG_SYSPLLCHGCTRL_ADDR = 0x%x\n", pll_change);
-	}
-	mul = syssoc_read_reg(sys_clk_base, DEV_PLLCTRL_1 + (4*dev_pll));
-	nlm_print("DEV PLL[%d] freq changed to =%d\n", dev_pll, 400 + 33*mul);
-
-}
 
 /*
  * @brief nlm_hal_fdt_device_clock_init function is used to read the frequency
-- 
1.7.1

