From d93bb6c26b30286fe5f16e8573c399e7e08d7639 Mon Sep 17 00:00:00 2001
From: Subhendu Sekhar Behera <sbehera@broadcom.com>
Date: Wed, 4 Jun 2014 14:02:27 +0530
Subject: XLP9xx: Change soc frequency return value.

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/netlogic/lib/syslib/include/nlm_hal.h b/arch/mips/netlogic/lib/syslib/include/nlm_hal.h
index e12fd6b..a79e8da 100644
--- a/arch/mips/netlogic/lib/syslib/include/nlm_hal.h
+++ b/arch/mips/netlogic/lib/syslib/include/nlm_hal.h
@@ -160,7 +160,7 @@ extern void nlm_hal_get_rsa_vc_nums(int *vcbase, int *vclimit);
 extern void nlm_hal_set_rsa_data_mode(int node, int bigend_fmt);
 
 extern int xlp9xx_set_nae_frequency(int node, uint64_t freq);
-extern int xlp9xx_set_soc_frequency(int node, int dev_id, uint64_t freq);
+extern uint64_t xlp9xx_set_soc_frequency(int node, int dev_id, uint64_t freq);
 extern void xlp9xx_device_clock_init(int node, void *fdt);	
 extern uint64_t xlp9xx_get_dev_freq(int node, int dev_id);
 
diff --git a/arch/mips/netlogic/lib/syslib/src/nlm_evp_cpld.c b/arch/mips/netlogic/lib/syslib/src/nlm_evp_cpld.c
index db6ffd7..c086313 100644
--- a/arch/mips/netlogic/lib/syslib/src/nlm_evp_cpld.c
+++ b/arch/mips/netlogic/lib/syslib/src/nlm_evp_cpld.c
@@ -248,6 +248,7 @@ void set_gbu_frequency(int node, int frequency)
 	else if(is_nlm_xlp9xx())
 	{
 		set_freq = xlp9xx_set_soc_frequency(node, XLP9XX_CLKDEV_GBU, frequency * mhz);
+		NLM_HAL_DO_DIV(set_freq,mhz);
 	}
 	else {
 		set_freq = nlm_hal_set_soc_freq(node, DFS_DEVICE_NOR, frequency * mhz);
diff --git a/arch/mips/netlogic/lib/syslib/src/nlm_hal.c b/arch/mips/netlogic/lib/syslib/src/nlm_hal.c
index 2beb442..1c45a4c 100644
--- a/arch/mips/netlogic/lib/syslib/src/nlm_hal.c
+++ b/arch/mips/netlogic/lib/syslib/src/nlm_hal.c
@@ -385,7 +385,7 @@ void xlp9xx_device_clock_init(int node, void *fdt)
 	//	xlp9xx_setup_devpll(sys_clk_base, sel, dev_clock[sel]);
 }
 
-int xlp9xx_set_soc_frequency(int node, int dev_id, uint64_t freq)
+uint64_t xlp9xx_set_soc_frequency(int node, int dev_id, uint64_t freq)
 {
 
 	switch(dev_id) {
@@ -406,7 +406,7 @@ int xlp9xx_set_soc_frequency(int node, int dev_id, uint64_t freq)
 			nlm_print("%s error\n",__func__);
 			break;
 	}
-        return 0;
+        return xlp9xx_get_dev_freq(node, dev_id);
 }
 
 
-- 
1.7.1

