From 143fab4182893636c3edfc33ee1a9d7fa05aeeb7 Mon Sep 17 00:00:00 2001
From: Kevin Joseph James <kevinj@broadcom.com>
Date: Fri, 2 May 2014 01:17:12 -0700
Subject: Fix: Redefinition of hard_smp_processor_id() in linux-userspace

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/netlogic/lib/syslib/include/nlm_xlp.h b/arch/mips/netlogic/lib/syslib/include/nlm_xlp.h
index fd4edfa..f57c506 100644
--- a/arch/mips/netlogic/lib/syslib/include/nlm_xlp.h
+++ b/arch/mips/netlogic/lib/syslib/include/nlm_xlp.h
@@ -1,7 +1,9 @@
 #ifndef _HAL_XLP_
 #define _HAL_XLP_
 
-#ifndef _ASM_NLM_MIPS_EXTS_H
+#ifdef NLM_HAL_LINUX_KERNEL
+#include <asm/netlogic/mips-extns.h>
+#else
 static inline int hard_smp_processor_id(void)
 {
 	int id;
-- 
1.7.1

