From 02531136aa7f05c9c4c5be1a8b123aedac7817f5 Mon Sep 17 00:00:00 2001
From: Sreekanth Molagavalli <sreekanth.molagavalli@broadcom.com>
Date: Tue, 24 Jun 2014 12:35:55 +0530
Subject: general: Print only default debug messages.

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/netlogic/lib/netlib/src/netsoc_api.c b/arch/mips/netlogic/lib/netlib/src/netsoc_api.c
index 86b8183..cd79a47 100644
--- a/arch/mips/netlogic/lib/netlib/src/netsoc_api.c
+++ b/arch/mips/netlogic/lib/netlib/src/netsoc_api.c
@@ -47,7 +47,7 @@ void *(*netsoc_lib_contig_malloc)(uint32_t align, uint32_t size);
 void (*netsoc_lib_contig_free)(void *buf);
 uint64_t (*netsoc_lib_virttophys)(void *);
 void *(*netsoc_lib_phystovirt)(uint64_t);
-unsigned int netsoc_api_dbglvl = 0xff; // NETSOC_APIDBG_DEFAULT | NETSOC_APIDBG_GLOBAL;
+unsigned int netsoc_api_dbglvl = NETSOC_APIDBG_DEFAULT;
 uint64_t (*netsoc_lib_readq)(uint64_t base, uint32_t index);
 void (*netsoc_lib_writeq)(uint64_t base, uint32_t index, uint64_t val);
 
-- 
1.7.1

