From e8a2aa37f8f8040d2fff1c180e77cf9cf540bc28 Mon Sep 17 00:00:00 2001
From: Vikas Gupta <vikas.gupta@broadcom.com>
Date: Wed, 25 Jun 2014 15:43:29 +0530
Subject: xlp9xx/5xx: Support for adjusting interface max frame length for xlaui

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/netlogic/lib/netlib/src/netsoc_nae.c b/arch/mips/netlogic/lib/netlib/src/netsoc_nae.c
index 0608294..254b8bd 100644
--- a/arch/mips/netlogic/lib/netlib/src/netsoc_nae.c
+++ b/arch/mips/netlogic/lib/netlib/src/netsoc_nae.c
@@ -3094,6 +3094,7 @@ int __netsoc_config_frame_size(net_port_t *netport, uint32_t size)
 {
         nae_t *nae = netport->nae;
         uint64_t mac_base;
+	volatile uint32_t val;
 
         switch(netport->iftype) {
                 case SGMII_IF:
@@ -3107,7 +3108,10 @@ int __netsoc_config_frame_size(net_port_t *netport, uint32_t size)
 			netsoc_write_mac_reg(mac_base, XAUI_MAX_FRAME_LEN , ((size/4) << 16) | size);
 			break;
 		case XLAUI_IF:
-			return NETSOC_API_SUCCESS;  /*default its setting 16K*/
+			val  = __netsoc_read_xlgmac_reg(netport, XLAUI_FRM_LENGTH);
+			val &= 0xffffc000;
+			val |= size; // 0xff for max_frame
+			__netsoc_write_xlgmac_reg(netport, XLAUI_FRM_LENGTH, val);
 			break;
                 case INTERLAKEN_IF:
                         if(nae->nae_id)
-- 
1.7.1

