From 33662f8b137584826bd7fe6a4ead26e1ff109a6f Mon Sep 17 00:00:00 2001
From: kopal <kopal@broadcom.com>
Date: Wed, 25 Jun 2014 15:54:25 +0530
Subject: netlib: Added higig support for xlp2xx-b0

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/netlogic/lib/netlib/src/netsoc_nae.c b/arch/mips/netlogic/lib/netlib/src/netsoc_nae.c
index 254b8bd..700f5a7 100644
--- a/arch/mips/netlogic/lib/netlib/src/netsoc_nae.c
+++ b/arch/mips/netlogic/lib/netlib/src/netsoc_nae.c
@@ -2205,7 +2205,7 @@ static void __netsoc_config_xaui(net_port_t *netport)
         }
         netsoc_write_mac_reg(mac_base, NETIOR_XGMAC_CTRL1, val | (1<<13));
 
-	if (IS_NLM_XLP9XX_FAMILY) {
+	if (IS_NLM_XLP9XX_FAMILY || (is_nlm_xlp2xx_b0() && higig_type)) {
                 /*
                  * Configuring the OFF/ON timer
                  * 31:16  - In PFC mode is used as the Xoff value
-- 
1.7.1

