From 399de9ce5c12e28520ace6c7f7998124aa1781ef Mon Sep 17 00:00:00 2001
From: Anusha Nelluri <anushan@broadcom.com>
Date: Fri, 11 Jul 2014 13:08:12 +0530
Subject: xlp5xx: Added configuration of popq class support for xlp5xx.

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/netlogic/lib/fmnlib/nlm_hal_fmn_config.c b/arch/mips/netlogic/lib/fmnlib/nlm_hal_fmn_config.c
index 94a7a44..71af5a2 100644
--- a/arch/mips/netlogic/lib/fmnlib/nlm_hal_fmn_config.c
+++ b/arch/mips/netlogic/lib/fmnlib/nlm_hal_fmn_config.c
@@ -814,7 +814,7 @@ void nlm_hal_disable_vc_intr(int node, int vc)
  * @param [in]	node-		node number
  * 		popq_class-	popq priority class number
  * 		width-		width of Priority class Pop output queue
- * 		base_id-	Base queue Id for Priority class output queue
+ * 		base_id-	Relative Base queue Id for Priority class output queue
  *
  * @return
  * 	  - 0		success
@@ -831,7 +831,9 @@ int brcm_hal_config_popq_class(int node, int popq_class, uint8_t width, uint16_t
 	unsigned long fmn_base_local[NLM_MAX_NODES] = {0};
 	unsigned long long mask = ~0xf;
 
-	if (is_nlm_xlp9xx()) {
+	if (IS_NLM_XLP9XX_FAMILY) {
+		//Relative base queue id is received as an argument. POPQ Base address for 9xx/5xx will be returned by brcm_get_popq_vc_base.
+		base_id = brcm_get_popq_vc_base() + base_id;
 		fmn_base_local[node] = mask & nlh_read_cfg_reg32(xlp9xx_fmn_base[node] + 0x10);
 		if ((popq_class < 0) || (popq_class > POPQ_MAX_CLASS_NUM)) {
 			nlm_print("ERROR: popq class number out of range. \
diff --git a/arch/mips/netlogic/lib/syslib/src/brcm_xlp_cpu.c b/arch/mips/netlogic/lib/syslib/src/brcm_xlp_cpu.c
index 466ce4a..f17a46b 100644
--- a/arch/mips/netlogic/lib/syslib/src/brcm_xlp_cpu.c
+++ b/arch/mips/netlogic/lib/syslib/src/brcm_xlp_cpu.c
@@ -108,8 +108,10 @@ inline int brcm_get_max_node_num(void)
 
 inline int brcm_get_popq_vc_base(void)
 {
-	if (IS_NLM_XLP9XX_FAMILY) {
+	if (is_nlm_xlp9xx()) {
 		return XLP_9XX_POPQ_VC_BASE;
+	} else if (is_nlm_xlp5xx()) {
+		return XLP_5XX_POPQ_VC_BASE;
 	} else {
 		return XLP_POPQ_VC_BASE;
 	}
-- 
1.7.1

