From 9dd74b4c9346cd5f3a8b3ea99dd9359cfa192b3b Mon Sep 17 00:00:00 2001
From: Anusha Nelluri <anushan@broadcom.com>
Date: Tue, 15 Jul 2014 10:55:01 +0530
Subject: fmnlib: moving brcm_get_popq_vc_base API from syslib to fmnlib

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/netlogic/lib/fmnlib/nlm_hal_fmn.h b/arch/mips/netlogic/lib/fmnlib/nlm_hal_fmn.h
index e62e9bd..50bf7be 100644
--- a/arch/mips/netlogic/lib/fmnlib/nlm_hal_fmn.h
+++ b/arch/mips/netlogic/lib/fmnlib/nlm_hal_fmn.h
@@ -307,6 +307,7 @@ extern void nlm_hal_fmn_init(void *fdt, int node);
 extern void nlm_hal_set_fmn_interrupt(int irq);
 extern void nlm_hal_disable_vc_intr(int node, int vc);
 extern void nlm_hal_enable_vc_intr(int node, int vc);
+extern inline int brcm_get_popq_vc_base(void);
 int brcm_hal_config_popq_class(int node, int popq_class, uint8_t width, uint16_t base_id);
 #endif /* #ifndef _NLH_FMN_H */
 
diff --git a/arch/mips/netlogic/lib/fmnlib/nlm_hal_fmn_config.c b/arch/mips/netlogic/lib/fmnlib/nlm_hal_fmn_config.c
index 71af5a2..4e4e4f3 100644
--- a/arch/mips/netlogic/lib/fmnlib/nlm_hal_fmn_config.c
+++ b/arch/mips/netlogic/lib/fmnlib/nlm_hal_fmn_config.c
@@ -807,6 +807,27 @@ void nlm_hal_disable_vc_intr(int node, int vc)
 		nlm_hal_write_outq_config(node, vc, val);
 	}
 }
+/**
+* @brief brcm_get_popq_vc_base- return the popq vc base
+*
+* @param [in]  - none
+*
+* @return
+*    - POPQ vc base id
+*
+* @see brcm_get_poe_vc_base 
+**/
+
+inline int brcm_get_popq_vc_base(void)
+{
+    if (is_nlm_xlp9xx()) {
+       return XLP_9XX_POPQ_VC_BASE;
+    } else if (is_nlm_xlp5xx()) {
+       return XLP_5XX_POPQ_VC_BASE;
+    } else {
+       return XLP_POPQ_VC_BASE;
+    }
+}
 
 /**
  * @brief brcm_hal_config_popq_class - configure popq priority class
diff --git a/arch/mips/netlogic/lib/syslib/include/brcm_xlp_cpu.h b/arch/mips/netlogic/lib/syslib/include/brcm_xlp_cpu.h
index 8bf00f9..774f5f3 100644
--- a/arch/mips/netlogic/lib/syslib/include/brcm_xlp_cpu.h
+++ b/arch/mips/netlogic/lib/syslib/include/brcm_xlp_cpu.h
@@ -59,7 +59,6 @@
 
 inline int brcm_get_max_cpu_num(void);
 inline int brcm_get_max_node_num(void);
-inline int brcm_get_popq_vc_base(void);
 inline int brcm_get_poe_vc_base(int poe_blk);
 inline uint64_t brcm_get_poe_pcie_base(int node, int poe_block);
 
diff --git a/arch/mips/netlogic/lib/syslib/src/brcm_xlp_cpu.c b/arch/mips/netlogic/lib/syslib/src/brcm_xlp_cpu.c
index f17a46b..ee291f8 100644
--- a/arch/mips/netlogic/lib/syslib/src/brcm_xlp_cpu.c
+++ b/arch/mips/netlogic/lib/syslib/src/brcm_xlp_cpu.c
@@ -96,28 +96,6 @@ inline int brcm_get_max_node_num(void)
 }
 
 /**
- * @brief brcm_get_popq_vc_base- return the popq vc base
- *
- * @param [in]	- none
- *
- * @return
- * 	  - POPQ vc base id
- *
- * @see brcm_get_poe_vc_base 
- **/
-
-inline int brcm_get_popq_vc_base(void)
-{
-	if (is_nlm_xlp9xx()) {
-		return XLP_9XX_POPQ_VC_BASE;
-	} else if (is_nlm_xlp5xx()) {
-		return XLP_5XX_POPQ_VC_BASE;
-	} else {
-		return XLP_POPQ_VC_BASE;
-	}
-}
-
-/**
  * @brief brcm_get_poe_vc_base- return the pop vc base
  *
  * @param poe_blk - poe block number
-- 
1.7.1

