From 59b988e6e76e1d144715270f6367e9cc931af63a Mon Sep 17 00:00:00 2001
From: Subhendu Sekhar Behera <sbehera@broadcom.com>
Date: Fri, 1 Aug 2014 15:22:29 +0530
Subject: XLP9XX: add ICS function & device number

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/netlogic/lib/syslib/include/nlm_hal_xlp_dev.h b/arch/mips/netlogic/lib/syslib/include/nlm_hal_xlp_dev.h
index b7d31f7..da2f09c 100644
--- a/arch/mips/netlogic/lib/syslib/include/nlm_hal_xlp_dev.h
+++ b/arch/mips/netlogic/lib/syslib/include/nlm_hal_xlp_dev.h
@@ -451,7 +451,10 @@ extern int is_nlm_xlp(unsigned int chipid, unsigned int rev,  unsigned int ext);
 #define XLP9XX_POE1_FUNC                4 
 
 #define XLP9XX_ALE_DEVICE               3
-#define XLP9XX_ALE_FUNC                 5 
+#define XLP9XX_ALE_FUNC                 5
+
+#define XLP9XX_ICS_DEVICE               2
+#define XLP9XX_ICS_FUNC                 1
 
 #define XLP9XX_SYS_BUS			0x1
 #define XLP9XX_SYS_DEVICE		0x6
diff --git a/arch/mips/netlogic/lib/syslib/include/nlm_xlp.h b/arch/mips/netlogic/lib/syslib/include/nlm_xlp.h
index f57c506..2aa62c8 100644
--- a/arch/mips/netlogic/lib/syslib/include/nlm_xlp.h
+++ b/arch/mips/netlogic/lib/syslib/include/nlm_xlp.h
@@ -20,6 +20,7 @@ extern cpumask_t phys_cpu_present_map;
 #define NLM_MAX_THREADS_PER_CPU 4
 #define NLM_MAX_VC_PER_THREAD   4
 
+#define NLM_MAX_CORE_PER_NODE_9XX	32
 #define NLM_MAX_CPU_NODE_9XX		8
 #define NUM_CPUMASK_WORDS_9XX		2
 #define NLM_MAX_CPU_PER_NODE_9XX	80
-- 
1.7.1

