From cfcd0998c1255884194259fb30f8851b042f2aa8 Mon Sep 17 00:00:00 2001
From: kopal <kopal@broadcom.com>
Date: Wed, 24 Sep 2014 14:56:14 +0530
Subject: nae: changed hardcoded values to macros for 9xx_B0.

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/netlogic/lib/netlib/include/netsoc_dev.h b/arch/mips/netlogic/lib/netlib/include/netsoc_dev.h
index 7897cf9..7c51d1e 100644
--- a/arch/mips/netlogic/lib/netlib/include/netsoc_dev.h
+++ b/arch/mips/netlogic/lib/netlib/include/netsoc_dev.h
@@ -574,11 +574,19 @@ enum {
 /*PRM: PMA2P0  CONFIG REGISTERS*/
 #define PMA2P0_CMD 			0x9
 #define PMA2P0_CTL0			0xA
+	#define PMA2P0_CTL_SYNTH_PWR_DOWN       (1<<8)
 	#define PMA2P0_CTL_POR			(1<<9)
-	#define XLP5XX_PMA2P0_CTL_POR		(1<<31)
 	#define PMA2P0_CTL_SYNTH_RST		(1<<10)
+	#define PMA2P0_CTL_REG_RESET            (1<<16)
+	#define PMA2P0_CTL_SYNTH_PWR_DOWN_B0    (1<<20)
+	#define PMA2P0_CTL_SYNTH_RST_B0      	(1<<24)
+	#define PMA2P0_CTL_POR_B0	        (1<<28)
+	#define XLP5XX_PMA2P0_CTL_POR		(1<<31)
 	#define PMA2P0_CTL_RTHR			(0xf<<12)	
-#define PMA2P0_CTL1			0xB
+	#define PMA2P0_CTL_RX_PWR_DOWN          0xf0
+	#define PMA2P0_CTL_TX_PWR_DOWN          0xf
+
+	#define PMA2P0_CTL1			0xB
 
 	#define PMA2P0_CTL_MASK_DR	(0x7)
 	#define PMA2P0_CTL_RXAUI_10G_DR	(0x3)
diff --git a/arch/mips/netlogic/lib/netlib/src/netsoc_pcs.c b/arch/mips/netlogic/lib/netlib/src/netsoc_pcs.c
index f592b2d..137101b 100644
--- a/arch/mips/netlogic/lib/netlib/src/netsoc_pcs.c
+++ b/arch/mips/netlogic/lib/netlib/src/netsoc_pcs.c
@@ -783,8 +783,7 @@ void xlp9xx_nae_write_pma2p0_mem(uint64_t mac_base, int block, int xgmii_speed,
 		if(!is_nlm_xlp9xx_b0())
                 	pma2p0_config &= ~(PMA2P0_CTL_POR | PMA2P0_CTL_SYNTH_RST | PMA2P0_CTL_RTHR);
                	else
-                       pma2p0_config &= 0xF0FF03FF;
-
+                       	pma2p0_config &= ~(PMA2P0_CTL_RTHR | PMA2P0_CTL_SYNTH_RST_B0 | PMA2P0_CTL_POR_B0);
 	}
 	netsoc_write_mac_reg(mac_base, PMA2P0_CTL0, pma2p0_config);
 
@@ -862,7 +861,7 @@ static void xlp9xx_config_pma2p0_serdes(nae_t *nae, uint32_t block, uint32_t pma
                        pma2p0_config |= PMA2P0_CTL_POR | PMA2P0_CTL_SYNTH_RST | PMA2P0_CTL_RTHR | (1<<16);
                	} else {
                        pma2p0_config &= 0x00FF0FFF;
-                       pma2p0_config |= ~0x00FF0FFF;
+                       pma2p0_config |= (PMA2P0_CTL_RTHR | PMA2P0_CTL_SYNTH_RST_B0 | PMA2P0_CTL_POR_B0 | PMA2P0_CTL_REG_RESET);
                	}
 
 	}
@@ -882,7 +881,7 @@ static void xlp9xx_config_pma2p0_serdes(nae_t *nae, uint32_t block, uint32_t pma
                        pma2p0_config &= ~(PMA2P0_CTL_POR | (1<<16));
                        pma2p0_config &= 0xFFFFFE00;
                } else 
-                       pma2p0_config &= 0x0F0FFF00;
+                       pma2p0_config &= ~(PMA2P0_CTL_POR_B0 | PMA2P0_CTL_REG_RESET | PMA2P0_CTL_SYNTH_PWR_DOWN_B0 | PMA2P0_CTL_RX_PWR_DOWN | PMA2P0_CTL_TX_PWR_DOWN);
 
 	}
 	netsoc_write_mac_reg(mac_base, PMA2P0_CTL0, pma2p0_config);
-- 
1.7.1

