From 5dcb92d8919e989e6a8b46e509d7acf5fa929c46 Mon Sep 17 00:00:00 2001
From: Vikas Gupta <vikas.gupta@broadcom.com>
Date: Thu, 25 Sep 2014 10:14:56 +0530
Subject: xlp9xx: Support for B0 version for XLAUI

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/netlogic/lib/netlib/include/netsoc_nae.h b/arch/mips/netlogic/lib/netlib/include/netsoc_nae.h
index 1531cab..bf7fa06 100644
--- a/arch/mips/netlogic/lib/netlib/include/netsoc_nae.h
+++ b/arch/mips/netlogic/lib/netlib/include/netsoc_nae.h
@@ -978,6 +978,8 @@ static inline uint32_t __netsoc_read_xlgmac_reg(net_port_t *netport, uint32_t ad
 
 	// Write the command
 	netsoc_write_mac_reg(mac_base , XLGMACPCS_CMD, cmd);
+	if(is_nlm_xlp9xx_b0())
+		netsoc_write_mac_reg(mac_base , XLGMACPCS_CMD, cmd);
 
 	// Wait for Busy bit to clear
 	val = (1<<31);
diff --git a/arch/mips/netlogic/lib/netlib/src/netsoc_nae.c b/arch/mips/netlogic/lib/netlib/src/netsoc_nae.c
index 8639e59..3e2f4d8 100644
--- a/arch/mips/netlogic/lib/netlib/src/netsoc_nae.c
+++ b/arch/mips/netlogic/lib/netlib/src/netsoc_nae.c
@@ -2041,13 +2041,15 @@ void __netsoc_set_xlgmac_enable(net_port_t *netport)
 	val = __netsoc_read_xlgmac_reg(netport, XLAUI_TX_FIFO_SECTIONS);
 	val &= 0xffff0000;
 	val |= 0x9;
-	val = 9; // Workaround for XLGMAC read bug
+	if(!is_nlm_xlp9xx_b0())
+		val = 9; // Workaround for XLGMAC read bug
         __netsoc_write_xlgmac_reg(netport, XLAUI_TX_FIFO_SECTIONS, val);
 
 	val = __netsoc_read_xlgmac_reg(netport, XLAUI_RX_FIFO_SECTIONS);
 	val &= 0xffff0000;
 	val |= 0x4;
-	val = 0x4; // Workaround for XLGMAC read bug
+	if(!is_nlm_xlp9xx_b0())
+		val = 0x4; // Workaround for XLGMAC read bug
         __netsoc_write_xlgmac_reg(netport, XLAUI_RX_FIFO_SECTIONS, val);
 
 	val = 0x220d3; // PFC disable
-- 
1.7.1

