From 1e1dfb5c00ff2cadbcad970a93075cd5365d3a3b Mon Sep 17 00:00:00 2001
From: Pranav Bhalerao <pranavb@broadcom.com>
Date: Fri, 3 Oct 2014 18:23:00 +0530
Subject: seclib: Changes to remove Camellia encryption from XLP9XX_b0 & XLP5XX.

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/netlogic/lib/seclib/saesoc_defs.h b/arch/mips/netlogic/lib/seclib/saesoc_defs.h
index 5907840..790a807 100644
--- a/arch/mips/netlogic/lib/seclib/saesoc_defs.h
+++ b/arch/mips/netlogic/lib/seclib/saesoc_defs.h
@@ -116,6 +116,7 @@
 #define SAESOC_FF_3DES_KEY_SWAP_SUPPORT 0x2000
 #define SAESOC_FF_HASH_SHA3_SUPPORT 0x4000
 #define SAESOC_FF_CIPHER_CBC_TERM_SUPPORT 0x8000
+#define SAESOC_FF_NO_CAMELLIA_ENC_SUPPORT 0x10000
 /*!defines saesoc_feature_flags > */
 
 /* for generation 1 like eagle, storm, the below 1 is a must */
-- 
1.7.1

