From 38ed57a9795b24b80eb22d7fbe378df07b316fa4 Mon Sep 17 00:00:00 2001
From: Vikas Gupta <vikas.gupta@broadcom.com>
Date: Tue, 7 Oct 2014 14:51:39 +0530
Subject: xlp9xx B0: Fix for pma2p0 power on reset.

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/netlogic/lib/netlib/src/netsoc_pcs.c b/arch/mips/netlogic/lib/netlib/src/netsoc_pcs.c
index 137101b..bffa497 100644
--- a/arch/mips/netlogic/lib/netlib/src/netsoc_pcs.c
+++ b/arch/mips/netlogic/lib/netlib/src/netsoc_pcs.c
@@ -843,7 +843,11 @@ static void xlp9xx_config_pma2p0_serdes(nae_t *nae, uint32_t block, uint32_t pma
 
 		}else{
 			pma2p0_config = netsoc_read_mac_reg(netsoc_get_macreg_base_for_phy(nae->mac_base, 2), PMA2P0_CTL0);
-			pma2p0_config &= ~PMA2P0_CTL_POR;
+			if(is_nlm_xlp9xx_b0())
+				pma2p0_config &= ~PMA2P0_CTL_POR_B0;
+			else 
+				pma2p0_config &= ~PMA2P0_CTL_POR;
+				
 			netsoc_write_mac_reg(netsoc_get_macreg_base_for_phy(nae->mac_base, 2), PMA2P0_CTL0, pma2p0_config);
 		}
 	}
-- 
1.7.1

