From 1f9edb137f04f62c6e41a486521bdc3312a4678b Mon Sep 17 00:00:00 2001
From: Vikas Gupta <vikas.gupta@broadcom.com>
Date: Fri, 17 Oct 2014 23:39:26 +0530
Subject: xlp9xx/5xx: Fix for rx_buf_size calcualtion for XFI.

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/netlogic/lib/netlib/include/netsoc_common.h b/arch/mips/netlogic/lib/netlib/include/netsoc_common.h
index efeb6c6..8601be8 100644
--- a/arch/mips/netlogic/lib/netlib/include/netsoc_common.h
+++ b/arch/mips/netlogic/lib/netlib/include/netsoc_common.h
@@ -394,7 +394,9 @@ static inline uint32_t get_rx_buf_size(int block, uint32_t intf_type)
 		}else
 			return DEFAULT_SGMII_RXFIFO_SZ;
 	case XFI_IF:
-                return DEFAULT_XAUI_RXFIFO_SZ/4;
+ 		if (IS_NLM_XLP9XX_FAMILY)
+	 			return DEFAULT_XLP9XX_XAUI_RXFIFO_SZ/4;
+                	return DEFAULT_XAUI_RXFIFO_SZ/4;
         case XAUI_IF:
         case XLAUI_IF:
  		if (IS_NLM_XLP9XX_FAMILY)
-- 
1.7.1

