From cddb402d3de60d143af7e62eab6e19885e4900f1 Mon Sep 17 00:00:00 2001
From: Vikas Gupta <vikas.gupta@broadcom.com>
Date: Tue, 21 Oct 2014 08:50:49 +0530
Subject: xlp9xx b0 : Workaround for XFI on xlp9xx B0

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/netlogic/lib/netlib/src/netsoc_pcs.c b/arch/mips/netlogic/lib/netlib/src/netsoc_pcs.c
index bffa497..e5b9626 100644
--- a/arch/mips/netlogic/lib/netlib/src/netsoc_pcs.c
+++ b/arch/mips/netlogic/lib/netlib/src/netsoc_pcs.c
@@ -768,6 +768,23 @@ void xlp9xx_nae_write_pma2p0_mem(uint64_t mac_base, int block, int xgmii_speed,
              }
 	}
 
+      if(phy_mode == PHYMODE_XFI && is_nlm_xlp9xx_b0() ){
+               for (lane_no=0; lane_no<max_lanes; lane_no=lane_no+lwidth) {
+                       pma2p0_cmd = __netsoc_read_pma2p0_reg(mac_base,  lane_no, 7);
+                       if (lane_no == 0) {
+                               pma2p0_cmd |= (1<<2);
+                       } else {
+                               pma2p0_cmd &= ~(1<<2);
+                       }
+                       __netsoc_write_pma2p0_reg(mac_base, lane_no, 7, pma2p0_cmd);
+                       pma2p0_cmd = __netsoc_read_pma2p0_reg(mac_base, PMA2P0_COMMON_LANE0, 13);
+                       pma2p0_cmd |= (1<<1) | (1<<0);
+                       __netsoc_write_pma2p0_reg(mac_base, PMA2P0_COMMON_LANE0, 13, pma2p0_cmd);
+               }
+
+
+       }               
+
 
 	for(delay=0; delay<10000; delay++);
 
-- 
1.7.1

