From 071102237407b812f5a52c9e78b7c07c322fdcdc Mon Sep 17 00:00:00 2001
From: Saswat Kumar Dash <saswat.dash@broadcom.com>
Date: Tue, 28 Oct 2014 05:57:40 -0700
Subject: ale: workaround for xlp9xx AX

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/netlogic/lib/alelib/ale_api.h b/arch/mips/netlogic/lib/alelib/ale_api.h
index 26d129c..801bf4a 100644
--- a/arch/mips/netlogic/lib/alelib/ale_api.h
+++ b/arch/mips/netlogic/lib/alelib/ale_api.h
@@ -209,11 +209,11 @@ ale_direct_search_req(ale_t* ale,              //ale pointer
 
     ale_print(ALE_DBG_INFO, "%s: Prefix: %x %x %x %x\n",
               __func__, str_ptr[0], str_ptr[1], str_ptr[2], str_ptr[3]);
-    if (is_nlm_xlp9xx()) {
+    if (is_nlm_xlp9xx_ax()) {
     	desc[i++] = (uint64_t)reverse_32(str_ptr[1]) << 32 | (uint64_t)reverse_32(str_ptr[0]); 
     	if(str_len > 64)
         	desc[i++] = (uint64_t)reverse_32(str_ptr[3]) << 32 | (uint64_t)reverse_32(str_ptr[2]); 
-     }else if (is_nlm_xlp5xx()) {
+     } else {
 	desc[i++] = (uint64_t)(str_ptr[0]) << 32 | (uint64_t)(str_ptr[1]);
 	if (str_len > 64)
 		desc[i++] = (uint64_t)(str_ptr[2]) << 32 | (uint64_t)(str_ptr[3]);
@@ -281,7 +281,7 @@ ale_memory_search_req(ale_t* ale,             //ale pointer
        ale_print(ALE_DBG_INFO, "%s: Prefix[%d]: %x %x %x %x\n",
                   __func__, i, str_ptr[0], str_ptr[1], str_ptr[2], str_ptr[3]);
 
-	if (is_nlm_xlp9xx()) {
+	if (is_nlm_xlp9xx_ax()) {
         	*req_ptr++ = (uint64_t)reverse_32(str_ptr[1]) << 32 | (uint64_t)reverse_32(str_ptr[0]);  
         	desc_cnt++;
         	if(str_len > 64){
@@ -290,7 +290,7 @@ ale_memory_search_req(ale_t* ale,             //ale pointer
         }  
         req_cnt++;
         prefix++;
-    }else if (is_nlm_xlp5xx()) {
+    } else {
 	*req_ptr++ = (uint64_t)(str_ptr[0]) << 32 | (uint64_t)(str_ptr[1]);
 	desc_cnt++;
 	if (str_len > 64) {
-- 
1.7.1

