From c4fde1ff09e13cb0270213cafdb907f19fcdbe99 Mon Sep 17 00:00:00 2001
From: kopal <kopal@broadcom.com>
Date: Mon, 4 Aug 2014 09:55:00 +0530
Subject: XLP9XX:XLAUI: Do not allow jumbo frames when MTU is less than 1500.

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/netlogic/lib/netlib/src/netsoc_nae.c b/arch/mips/netlogic/lib/netlib/src/netsoc_nae.c
index ab12f6a..968b4ab 100644
--- a/arch/mips/netlogic/lib/netlib/src/netsoc_nae.c
+++ b/arch/mips/netlogic/lib/netlib/src/netsoc_nae.c
@@ -2058,12 +2058,6 @@ void __netsoc_set_xlgmac_enable(net_port_t *netport)
 	}
         __netsoc_write_xlgmac_reg(netport, XLAUI_COMMNAD_CONFIG, val);
 
-
-	val  = __netsoc_read_xlgmac_reg(netport, XLAUI_FRM_LENGTH);
-	val &= 0xffffc000;
-	val |= 0x3fe0; // 0xff for max_frame
-	__netsoc_write_xlgmac_reg(netport, XLAUI_FRM_LENGTH, val);
-
 }
 
 
-- 
1.7.1

