From 9117260b58969549dc148af21e5e6efd0bc6cd3f Mon Sep 17 00:00:00 2001
From: Kevin Joseph James <kevin.james@broadcom.com>
Date: Wed, 21 May 2014 03:15:14 -0700
Subject: netlib: Delay SGMII auto-negotiation until the ports are opened

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/netlogic/lib/netlib/include/netsoc_haliface.h b/arch/mips/netlogic/lib/netlib/include/netsoc_haliface.h
index d9d3d17..fded3f4 100644
--- a/arch/mips/netlogic/lib/netlib/include/netsoc_haliface.h
+++ b/arch/mips/netlogic/lib/netlib/include/netsoc_haliface.h
@@ -92,6 +92,7 @@ extern uint32_t __netsoc_bkpl_wait_for_an_complete(nae_t *nae, uint32_t block, u
 extern int  __netsoc_bkpl_wait_for_ld_lt_done(nae_t *nae, uint32_t block, uint32_t lane_no);
 void __netsoc_init_kr_module(void);
 
+extern void __netsoc_config_sgmii(net_port_t *netport);
 extern void __netsoc_sgmii_pcs_init(nae_t *nae, uint32_t sgmii_cplx_mask);
 extern void __netsoc_xaui_pcs_init(nae_t *nae, uint32_t xaui_cplx_mask, uint32_t phymode);
 extern void __netsoc_xfi_pcs_init(nae_t *nae, uint32_t xfi_cplx_mask, uint32_t phymode);
diff --git a/arch/mips/netlogic/lib/netlib/src/netsoc_api.c b/arch/mips/netlogic/lib/netlib/src/netsoc_api.c
index 973c0f3..8f016f3 100644
--- a/arch/mips/netlogic/lib/netlib/src/netsoc_api.c
+++ b/arch/mips/netlogic/lib/netlib/src/netsoc_api.c
@@ -326,6 +326,10 @@ net_port_t *netsoc_open_port(nae_t *nae, uint32_t port)
 		return NULL;
 
 	netport = &nae->ports[port];
+
+	if (netport->iftype == SGMII_IF)
+		__netsoc_config_sgmii(netport);
+
 	__netsoc_mac_enable(netport);
 	return netport;
 }
diff --git a/arch/mips/netlogic/lib/netlib/src/netsoc_nae.c b/arch/mips/netlogic/lib/netlib/src/netsoc_nae.c
index 2bb151b..f152a0c 100644
--- a/arch/mips/netlogic/lib/netlib/src/netsoc_nae.c
+++ b/arch/mips/netlogic/lib/netlib/src/netsoc_nae.c
@@ -2088,7 +2088,7 @@ void __netsoc_set_xlgmac_enable(net_port_t *netport)
 }
 
 
-static void __netsoc_config_sgmii(net_port_t *netport)
+void __netsoc_config_sgmii(net_port_t *netport)
 {
         unsigned int mac_cfg1 = 0, mac_cfg2 = 0;
         unsigned int netwk_inf = 0, val;
@@ -2940,8 +2940,8 @@ int __netsoc_config_port(net_port_t *netport)
 			netsoc_api_print(NETSOC_APIDBG_PORT, "init ext phy:%d\n", netport->hw_port_id);
 			__netsoc_init_ext_phy(netport);
 #endif
-			// Configure speed and mode
-			__netsoc_config_sgmii(netport);
+			// Configure speed and mode..Moving this to netsoc_open_port()
+			//__netsoc_config_sgmii(netport);
 			break;
 		default:
 			netsoc_api_print(NETSOC_APIDBG_ERROR, "Unknown interface type\n");
-- 
1.7.1

