From 0364fc41ae962f5e5d6c78031394e530d3a2addf Mon Sep 17 00:00:00 2001
From: Vikas Gupta <vikas.gupta@broadcom.com>
Date: Wed, 27 Aug 2014 23:22:53 +0530
Subject: xlp9xx: added flow control support for XLAUI.

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/netlogic/lib/netlib/src/netsoc_nae.c b/arch/mips/netlogic/lib/netlib/src/netsoc_nae.c
index 8c38564..307a7a0 100644
--- a/arch/mips/netlogic/lib/netlib/src/netsoc_nae.c
+++ b/arch/mips/netlogic/lib/netlib/src/netsoc_nae.c
@@ -3028,6 +3028,13 @@ int __netsoc_flow_control_status(nae_t *nae, uint32_t port)
 				return 1;
 			break;
 
+		case XLAUI_IF:
+        		mac_base = netsoc_get_macreg_base_for_xlgmac(nae->mac_base, netport->hw_port_id);
+        		val = netsoc_read_mac_reg(mac_base, XLGMACPCS_CTRL);
+			if(val & 0x5)
+				return 1;
+			return 0;
+			break;
 		case XFI_IF:
 		case XAUI_IF:
                 case RXAUI_IF:
@@ -3206,6 +3213,17 @@ int __netsoc_config_flow_control(net_port_t *netport, uint32_t enable)
 
 			netsoc_api_print(NETSOC_APIDBG_PORT,"%s: GMAC enable %d NETIOR_GMAC_CTRL1 0x%x\n",__func__, enable, netsoc_read_mac_reg(mac_base, NETIOR_GMAC_CTRL1));
 			break;
+		case XLAUI_IF:
+        		mac_base = netsoc_get_macreg_base_for_xlgmac(nae->mac_base, netport->hw_port_id);
+			if(enable){
+				val = netsoc_read_mac_reg(mac_base, XLGMACPCS_CTRL);
+        			netsoc_write_mac_reg(mac_base, XLGMACPCS_CTRL, val | 0x5);
+			}else{
+				val = netsoc_read_mac_reg(mac_base, XLGMACPCS_CTRL);
+				val &= ~(0x5);
+        			netsoc_write_mac_reg(mac_base, XLGMACPCS_CTRL, val);
+			}
+			break;
 		case XAUI_IF:
 		case XFI_IF:
 		case RXAUI_IF:
-- 
1.7.1

