From fa6073fd1245b1b27ef436848672e978d2ffe14c Mon Sep 17 00:00:00 2001
From: Vikas Gupta <vikas.gupta@broadcom.com>
Date: Fri, 12 Sep 2014 12:07:36 +0530
Subject: netlib: removed redefinition of macro.

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/netlogic/lib/netlib/include/netsoc_dev.h b/arch/mips/netlogic/lib/netlib/include/netsoc_dev.h
index e79ccd1..ede061d 100644
--- a/arch/mips/netlogic/lib/netlib/include/netsoc_dev.h
+++ b/arch/mips/netlogic/lib/netlib/include/netsoc_dev.h
@@ -979,7 +979,6 @@ enum {
 #define XGMAC_AMIIM_CONFIG               0x12
 #define XGMAC_AMIIM_LINKFAIL             0x13
 #define XGMAC_AMIIM_INDICATOR            0x14
-#define MAC_FILTER_CONFIG                0x5C
 #define HASH_TABLE_VECTOR_BASE           0x60
 
 #define NETIOR_XGMAC_CTRL2		0x7E
-- 
1.7.1

