From 92f84fb52f8007f50fbc2d90324638ec4760d300 Mon Sep 17 00:00:00 2001
From: Saswat Kumar Dash <saswat.dash@broadcom.com>
Date: Fri, 3 Oct 2014 03:18:41 -0700
Subject: syslib: fixed CDE/LZS SoC frequency setting

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/netlogic/lib/syslib/src/nlm_hal.c b/arch/mips/netlogic/lib/syslib/src/nlm_hal.c
index 0de4625..1956298 100644
--- a/arch/mips/netlogic/lib/syslib/src/nlm_hal.c
+++ b/arch/mips/netlogic/lib/syslib/src/nlm_hal.c
@@ -402,6 +402,12 @@ uint64_t xlp9xx_set_soc_frequency(int node, int dev_id, uint64_t freq)
 		case XLP9XX_CLKDEV_GDX:
 			xlp9xx_set_dtre_frequency(node, freq);
 			break;
+		case XLP9XX_CLKDEV_CMP:
+			xlp9xx_set_cmp_frequency(node, freq);
+			break;
+		case XLP9XX_CLKDEV_LZS:
+			xlp9xx_set_lzs_frequency(node, freq);
+			break;
 		default:
 			nlm_print("%s error\n",__func__);
 			break;
@@ -1093,7 +1099,7 @@ void nlm_hal_set_cde_freq(int node, int freq)
 	if(is_nlm_xlp2xx()) {
 		nlm_hal_xlp2xx_set_clkdev_frq(node, XLP2XX_CLKDEVICE_CMP, freq * mhz);
 	}
-	else if (is_nlm_xlp9xx()) {
+	else if (IS_NLM_XLP9XX_FAMILY) {
 		xlp9xx_set_soc_frequency(node, XLP9XX_CLKDEV_CMP, freq * mhz);
 	}
 	else {
-- 
1.7.1

