From f6cc39720b80909b02fe6513116beb8511f2a0f1 Mon Sep 17 00:00:00 2001
From: kopal <kopal@broadcom.com>
Date: Thu, 16 Oct 2014 15:52:03 +0530
Subject: netlib: fixed prototype for netsoc_get_link_status().

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/netlogic/lib/netlib/src/netsoc_api.c b/arch/mips/netlogic/lib/netlib/src/netsoc_api.c
index e825581..17513a5 100644
--- a/arch/mips/netlogic/lib/netlib/src/netsoc_api.c
+++ b/arch/mips/netlogic/lib/netlib/src/netsoc_api.c
@@ -74,7 +74,7 @@ register_broadcom_phy,
 
 int netsoc_get_phy_status(net_port_t *port, uint32_t *duplex, uint32_t *speed)
 {
-	return __netsoc_get_link_status (port, duplex, speed);
+	return __netsoc_get_link_status (port, speed, duplex);
 }
 
 
-- 
1.7.1

