From c6a66409e00730500b93357d54889db349d75025 Mon Sep 17 00:00:00 2001
From: Vikas Gupta <vikas.gupta@broadcom.com>
Date: Fri, 14 Nov 2014 21:56:06 +0530
Subject: nae mdio: reduce the divisor value as it suits most of the boards.

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/netlogic/lib/netlib/src/netsoc_nae.c b/arch/mips/netlogic/lib/netlib/src/netsoc_nae.c
index b09a8f8..2c0c0bd 100644
--- a/arch/mips/netlogic/lib/netlib/src/netsoc_nae.c
+++ b/arch/mips/netlogic/lib/netlib/src/netsoc_nae.c
@@ -36,7 +36,7 @@
 #include "netsoc_libiface.h"
 #include "ext_phy.h"
 
-int cl45_div = 3;
+int cl45_div = 2;
 #ifdef NLM_HAL_LINUX_KERNEL
 #include <linux/module.h>
 module_param(cl45_div, int, 0);
-- 
1.7.1

