From e13ce456e52bfc75587fe6dab8e51f354f2185e2 Mon Sep 17 00:00:00 2001
From: Vikas Gupta <vikas.gupta@broadcom.com>
Date: Mon, 17 Nov 2014 14:56:32 +0530
Subject: nae: fix for build break and warnings.

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/netlogic/lib/netlib/include/netsoc_haliface.h b/arch/mips/netlogic/lib/netlib/include/netsoc_haliface.h
index 2df0809..26f5f59 100644
--- a/arch/mips/netlogic/lib/netlib/include/netsoc_haliface.h
+++ b/arch/mips/netlogic/lib/netlib/include/netsoc_haliface.h
@@ -100,7 +100,8 @@ extern void __netsoc_xaui_pcs_init(nae_t *nae, uint32_t xaui_cplx_mask, uint32_t
 extern void __netsoc_xfi_pcs_init(nae_t *nae, uint32_t xfi_cplx_mask, uint32_t phymode);
 extern void __netsoc_xlaui_pcs_init(nae_t *nae, uint32_t xlaui_cplx_mask, uint32_t phymode);
 extern int __netsoc_interlaken_pcs_init(nae_t *nae, uint32_t ilk_complex_map);
-extern void __netsoc_wait_for_pcs_sync(nae);
+extern void __netsoc_wait_for_pcs_sync(nae_t *nae);
+extern void __netsoc_check_for_remote_fault(nae_t *nae);
 extern int register_marvell_phy(void);
 extern int register_broadcom_phy(void);
 extern int register_external_phy(ext_phy_t *ext_phy);
@@ -118,7 +119,7 @@ extern int __netsoc_alloc_poe_enq_storage(poe_t *poe);
 extern void __netsoc_enable_poe_stats(poe_t *poe);
 extern void __netsoc_print_poe_stats(poe_t *poe);
 extern void __netsoc_kr_init_an(nae_t *nae, int block);
-extern  __netsoc_port_link_status(net_port_t* netport);
+extern  int __netsoc_port_link_status(net_port_t* netport);
 
 extern unsigned char pma2p0_mem_sgmii_4page [256];
 extern unsigned char pma2p0_mem_sgmii [256];
diff --git a/arch/mips/netlogic/lib/netlib/include/nlm_hal_nae.h b/arch/mips/netlogic/lib/netlib/include/nlm_hal_nae.h
index b758f94..6092fdf 100644
--- a/arch/mips/netlogic/lib/netlib/include/nlm_hal_nae.h
+++ b/arch/mips/netlogic/lib/netlib/include/nlm_hal_nae.h
@@ -420,6 +420,7 @@ static inline int get_sgmii_speed_config(unsigned int speed)
 	if(speed==SPEED_10M) return 0;
 	if(speed==SPEED_100M) return 0x1;
 	if(speed==SPEED_1000M) return 0x2;
+	return 0;
 }	
 
 /* MAC_CONF1 */
-- 
1.7.1

