From 2742c94b26251f0a31df823e54b7d8a94f6d3818 Mon Sep 17 00:00:00 2001
From: Shashank Shekhar <shekhars@broadcom.com>
Date: Wed, 30 Apr 2014 04:07:46 -0700
Subject: NAE: Corrected parser sequencer FIFO size limitation

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/netlogic/lib/netlib/src/netsoc_fdt.c b/arch/mips/netlogic/lib/netlib/src/netsoc_fdt.c
index ebc312a..ae323dd 100644
--- a/arch/mips/netlogic/lib/netlib/src/netsoc_fdt.c
+++ b/arch/mips/netlogic/lib/netlib/src/netsoc_fdt.c
@@ -891,12 +891,6 @@ static void extract_complex_params(void *fdt, int intf_type, char *nae_port_str,
 	}
 
 	if (intf_type == SGMII_IF) {
-                if (is_nlm_xlp3xx()) {
-                     for (i = 0 ; i < 4; i++) {
-                        if (cmplx->prsr_seq_fifo_size[i] > XLP3XX_SGMII_PARSERSEQ_FIFO_MAX)
-                                cmplx->prsr_seq_fifo_size[i] = XLP3XX_SGMII_PARSERSEQ_FIFO_MAX;
-                     }
-                }
                 GET_PORT_PROP("ext-phy-addr", &cmplx->ext_phy_addr, sizeof(cmplx->ext_phy_addr));
                 GET_PORT_PROP("ext-phy-bus", &cmplx->ext_phy_bus, sizeof(cmplx->ext_phy_bus));
                 GET_PORT_PROP("ext-phy-sbl", &cmplx->ext_phy_sbl, sizeof(cmplx->ext_phy_sbl));
-- 
1.7.1

