From 4539360c6bf4d35da3fea8496d0aa2b919d22bb4 Mon Sep 17 00:00:00 2001
From: Vikas Gupta <vikas.gupta@broadcom.com>
Date: Wed, 19 Nov 2014 14:36:40 +0530
Subject: nae: fix typo in configuring threshold low for spill in Free-in-Lifo config.

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/netlogic/lib/netlib/src/netsoc_nae.c b/arch/mips/netlogic/lib/netlib/src/netsoc_nae.c
index 2c0c0bd..ca3513c 100644
--- a/arch/mips/netlogic/lib/netlib/src/netsoc_nae.c
+++ b/arch/mips/netlogic/lib/netlib/src/netsoc_nae.c
@@ -4208,7 +4208,7 @@ static int __netsoc_config_freein_fifo(nae_t *nae)
         if(spillsz) {
 		if (IS_NLM_XLP9XX_FAMILY) {
 			th_hi = minsz / 4;
-			th_hi = minsz / 2;
+			th_lo = minsz / 2;
 		}
 		else {
 	                th_hi = 6; /* Defualt value */
-- 
1.7.1

