From 1dc85ea6d586df328c39c0414739cd91389f6502 Mon Sep 17 00:00:00 2001
From: kopal <kopal@broadcom.com>
Date: Thu, 27 Nov 2014 14:36:48 +0530
Subject: nae: fix for incorrect mapping of SGMII speed in Marvel Phy.

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/netlogic/lib/netlib/external/ext_phy_mvl.c b/arch/mips/netlogic/lib/netlib/external/ext_phy_mvl.c
index 7ca8b0c..73047ae 100644
--- a/arch/mips/netlogic/lib/netlib/external/ext_phy_mvl.c
+++ b/arch/mips/netlogic/lib/netlib/external/ext_phy_mvl.c
@@ -74,6 +74,25 @@ static int mvl_get_phy_status(net_port_t *netport, uint32_t *speed, uint32_t *du
         *speed = (extstatus >> 14) & 0x3;
         *duplex =  (extstatus >> 13) & 0x1;
 	prev_link_stat=netport->link_stat;
+
+	switch(*speed) {
+                case 0x0:
+                        *speed = SPEED_10M;
+                break;
+
+                case 0x1:
+                        *speed = SPEED_100M;
+                break;
+
+                case 0x2:
+                        *speed = SPEED_1000M;
+                break;
+
+                default:
+                        netsoc_api_print(NETSOC_APIDBG_DEFAULT,"Unknown operating speed, extstatus = 0x%04x\n", extstatus);
+                break;
+        }
+
         if (extstatus & 0x0400) {
 		netport->link_stat=1;
                 if(netport->link_stat!=prev_link_stat)
-- 
1.7.1

