From db6ac4cfcaa15bd1826287216d4e5d4cd55e4d2b Mon Sep 17 00:00:00 2001
From: Vikas Gupta <vikas.gupta@broadcom.com>
Date: Wed, 3 Dec 2014 19:16:03 +0530
Subject: xlp2xx nae: fix for flow control.

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/netlogic/lib/netlib/src/netsoc_nae.c b/arch/mips/netlogic/lib/netlib/src/netsoc_nae.c
index fca6526..45882f8 100644
--- a/arch/mips/netlogic/lib/netlib/src/netsoc_nae.c
+++ b/arch/mips/netlogic/lib/netlib/src/netsoc_nae.c
@@ -2277,7 +2277,7 @@ static void __netsoc_config_xaui(net_port_t *netport)
         }
         netsoc_write_mac_reg(mac_base, NETIOR_XGMAC_CTRL1, val | (1<<13));
 
-	if (IS_NLM_XLP9XX_FAMILY || (is_nlm_xlp2xx_b0() && higig_type)) {
+	if (IS_NLM_XLP9XX_FAMILY || (is_nlm_xlp2xx_b0())) {
                 /*
                  * Configuring the OFF/ON timer
                  * 31:16  - In PFC mode is used as the Xoff value
-- 
1.7.1

