From 150a1ecfa9affcf19fa94d261be6d5901a932424 Mon Sep 17 00:00:00 2001
From: Abhishek Joshi <abjoshi@broadcom.com>
Date: Wed, 3 Dec 2014 08:54:49 -0800
Subject: syslib: restore max nodes to four

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/netlogic/lib/syslib/src/nlm_hal.c b/arch/mips/netlogic/lib/syslib/src/nlm_hal.c
index 32cf64c..56bb5e3 100644
--- a/arch/mips/netlogic/lib/syslib/src/nlm_hal.c
+++ b/arch/mips/netlogic/lib/syslib/src/nlm_hal.c
@@ -162,10 +162,7 @@ __inline__ void nlm_hal_enumerate_pci(void)
 	uint32_t active_nodes=0;
 	int node = 0;
 	int max_nodes = 0;
-	if(IS_NLM_XLP9XX_FAMILY)
-		max_nodes = XLP9XX_MAX_NODES;
-	else
-		max_nodes = NLM_MAX_NODES;
+	max_nodes = NLM_MAX_NODES;
 	for(node = 0 ; node < max_nodes ; node++)
 	{
 		if(nlm_hal_is_node_online(node))
@@ -186,7 +183,8 @@ __inline__ void nlm_hal_enumerate_pci(void)
 */
 int nlm_hal_is_node_online(int node)
 {
-	int ret, base;
+	int ret = 0;
+	uint64_t base = 0;
 	if(IS_NLM_XLP9XX_FAMILY) {
 		/* the system function value is same as the node value */
 		base = nlm_hal_get_dev_base(0, 0, 0, node);
@@ -638,10 +636,7 @@ __inline__ void nlm_hal_init(void)
 	enable_KX(flags);
 #endif
 	nlm_hal_enumerate_pci();
-	if(IS_NLM_XLP9XX_FAMILY)
-		max_nodes = XLP9XX_MAX_NODES;
-	else
-		max_nodes = NLM_MAX_NODES;
+	max_nodes = NLM_MAX_NODES;
 	xlp_io_base = KSEG1 + 0x18000000;
 	nlm_node_cfg.num_nodes = nlm_hal_get_num_nodes();
 	nlm_print("Active Nodes: %d\n",nlm_node_cfg.num_nodes);
-- 
1.7.1

