From a7350b4de9529b49cf2d942e8c6da96523605a72 Mon Sep 17 00:00:00 2001
From: PUNYA BHEEMESH <bheemesh@broadcom.com>
Date: Fri, 5 Dec 2014 14:59:14 +0530
Subject: netlib: default enable MAC_FILTERING in libraries

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/netlogic/lib/netlib/include/netsoc_nae.h b/arch/mips/netlogic/lib/netlib/include/netsoc_nae.h
index b6b276e..02e4ee7 100644
--- a/arch/mips/netlogic/lib/netlib/include/netsoc_nae.h
+++ b/arch/mips/netlogic/lib/netlib/include/netsoc_nae.h
@@ -44,7 +44,7 @@
 #define NAE_RX_CLASS_VALID		1
 #define MIN_VALID_CLASS_INDX		0
 #define MAX_VALID_CLASS_INDX		1
-//#define MAC_FILTERING
+#define MAC_FILTERING
 
 extern struct netsoc_info *current_netsoc;
 
diff --git a/arch/mips/netlogic/lib/netlib/src/nlm_hal_nae.c b/arch/mips/netlogic/lib/netlib/src/nlm_hal_nae.c
index a46b607..d18f723 100644
--- a/arch/mips/netlogic/lib/netlib/src/nlm_hal_nae.c
+++ b/arch/mips/netlogic/lib/netlib/src/nlm_hal_nae.c
@@ -1523,10 +1523,6 @@ static int local_load_ucore(int node, int ucore_mask, unsigned int *opcodes, int
 	ucore_cfg = ucore_cfg | (VAL_UCORE_RESET(ucore_mask));
         nlm_hal_write_nae_reg(node, RX_UCORE_CFG, ucore_cfg);
 
-	/* Enable CAM 384 * 64 entries */
-	nlm_hal_write_nae_reg(node, RX_UCORE_CFG, ucore_cfg);
-
-		/* write zero to reset bits to pull them out of reset */
 	ucore_cfg = ucore_cfg & (~VAL_UCORE_RESET(ucore_mask)) & ~(1 << 24);
         nlm_hal_write_nae_reg(node, RX_UCORE_CFG, ucore_cfg);
 
-- 
1.7.1

