From e46519c72caa41936b165281ab2c9eca7c12b2a7 Mon Sep 17 00:00:00 2001
From: Saswat Kumar Dash <saswat.dash@broadcom.com>
Date: Sun, 7 Dec 2014 21:49:23 -0800
Subject: multinode: fmn spill memory allocation on multinode

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/netlogic/lib/fmnlib/nlm_hal_fmn_config.c b/arch/mips/netlogic/lib/fmnlib/nlm_hal_fmn_config.c
index 06ad9b0..f147010 100644
--- a/arch/mips/netlogic/lib/fmnlib/nlm_hal_fmn_config.c
+++ b/arch/mips/netlogic/lib/fmnlib/nlm_hal_fmn_config.c
@@ -275,7 +275,7 @@ nlm_fmn_config_t xlp5xx_fmn_config[] = {
 };
 
 extern struct nlm_node_config nlm_node_cfg;
-extern unsigned long spill_mem_address;
+extern unsigned long spill_mem_address[NLM_MAX_NODES];
 extern unsigned long spill_mem_size;
 
 //#define FMN_DEBUG 1
@@ -1718,11 +1718,11 @@ int parse_fdt_fmn_config(void *fdt)
 #ifdef NLM_HAL_LINUX_KERNEL
 		if (nlm_node_cfg.fmn_cfg[node]->fmn_spill_size == 0ULL) {
 			nlm_node_cfg.fmn_cfg[node]->fmn_spill_size = spill_mem_size;
-			spill_base = spill_mem_address;
+			spill_base = spill_mem_address[node];
 		}
 		if ((pval == NULL) || (spill_base == 0ULL)){
 			if (spill_base == 0ULL) {
-				nlm_print("Node %d FMN spill_mem alloc failed \n", node);
+				nlm_print("Node %d FMN spill_mem alloc failed\n", node);
 				nlm_node_cfg.fmn_cfg[node]->fmn_spill_size = 0;
 			}
 		}
-- 
1.7.1

