From e2ad1060d90c21ab4fb190cfdac113e69350c520 Mon Sep 17 00:00:00 2001
From: Pavani Reddy <pavani@broadcom.com>
Date: Wed, 10 Dec 2014 11:31:33 +0530
Subject: Multinode : Change the start cpu of slave node to 128

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/netlogic/lib/netlib/src/netsoc_fdt.c b/arch/mips/netlogic/lib/netlib/src/netsoc_fdt.c
index 5ddf8d8..da7fd45 100644
--- a/arch/mips/netlogic/lib/netlib/src/netsoc_fdt.c
+++ b/arch/mips/netlogic/lib/netlib/src/netsoc_fdt.c
@@ -487,8 +487,11 @@ start_for_dom:
 					max = (plen / sizeof(unsigned int)) - 1;
 					for(i = 0; i < plen / sizeof(unsigned int); i++, max--)
 						cpu_mask[i] = fdt32_to_cpu(pval[max]);
-					right_shift_int_array(cpu_mask, node * num_cpus_per_node, num_cpus_per_node);
-				}
+					if(is_nlm_xlp9xx())
+						right_shift_int_array(cpu_mask, node * num_cpus_per_node * 2, num_cpus_per_node * 2);/* for xlp9xx slave cpu starts from 128 */
+					else
+						right_shift_int_array(cpu_mask, node * num_cpus_per_node, num_cpus_per_node);
+		 }
 		sprintf(dom_node_str, "/doms/dom@%d/cpu", domid);
 		nodeoffset = fdt_path_offset(fdt, dom_node_str);
 		if (nodeoffset >= 0) {
-- 
1.7.1

