From 4b336051f88b5eb9651b9186e43d0d033823a956 Mon Sep 17 00:00:00 2001
From: PUNYA BHEEMESH <bheemesh@broadcom.com>
Date: Wed, 10 Dec 2014 11:43:08 +0530
Subject: fmnlib: Fix for EPSW-2716

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/netlogic/lib/fmnlib/nlm_hal_fmn_config.c b/arch/mips/netlogic/lib/fmnlib/nlm_hal_fmn_config.c
index f147010..86c703a 100644
--- a/arch/mips/netlogic/lib/fmnlib/nlm_hal_fmn_config.c
+++ b/arch/mips/netlogic/lib/fmnlib/nlm_hal_fmn_config.c
@@ -761,7 +761,7 @@ static void fmn_validate_credit(int node, int max_nodes)
                                         fmn_q_config[d_stn].q_name, reqd_size, qsize);
 #endif
 
-		if(reqd_size >= qsize)
+		if(reqd_size > qsize)
 			nlm_print("WARN ... Credits overflow.. dst stn %s, reqd %d cfgd %d\n",
 					fmn_q_config[d_stn].q_name, reqd_size, qsize);
 	}
-- 
1.7.1

