From 92949f017ac0a9bc3830768431de6293d5cdf4e2 Mon Sep 17 00:00:00 2001
From: Vikas Gupta <vikas.gupta@broadcom.com>
Date: Wed, 10 Dec 2014 14:32:27 +0530
Subject: nae rxaui: Updated the NLP PHY configuration.

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/netlogic/lib/netlib/external/NETSOC_NLP1042C2_RXAUI_Dune.h b/arch/mips/netlogic/lib/netlib/external/NETSOC_NLP1042C2_RXAUI_Dune.h
index 39f05cd..6bb8fee 100644
--- a/arch/mips/netlogic/lib/netlib/external/NETSOC_NLP1042C2_RXAUI_Dune.h
+++ b/arch/mips/netlogic/lib/netlib/external/NETSOC_NLP1042C2_RXAUI_Dune.h
@@ -25,9 +25,9 @@ NETSOC_NLM_C45_WRITE(nae, bus, phyaddr, dev_addr, 0xc240, 0x3800); // 13:11= 0x7
 NETSOC_NLM_C45_WRITE(nae, bus, phyaddr, dev_addr, 0xc242, 0x0100); // 15:13= 0x0
 NETSOC_NLM_C45_WRITE(nae, bus, phyaddr, dev_addr, 0xc246, 0x0180); //  0:0 = 0x0
 NETSOC_NLM_C45_WRITE(nae, bus, phyaddr, dev_addr, 0xc246, 0x0182); //  1:1 = 0x1
-NETSOC_NLM_C45_WRITE(nae, bus, phyaddr, dev_addr, 0xc243, 0x07e1); //  4:0 = 0x1
-NETSOC_NLM_C45_WRITE(nae, bus, phyaddr, dev_addr, 0xc243, 0x07e1); // 10:5 = 0x3f
-NETSOC_NLM_C45_WRITE(nae, bus, phyaddr, dev_addr, 0xc243, 0x17e1); // 15:11= 0x2
+NETSOC_NLM_C45_WRITE(nae, bus, phyaddr, dev_addr, 0xc243, 0x07e0); //  4:0 = 0x1
+NETSOC_NLM_C45_WRITE(nae, bus, phyaddr, dev_addr, 0xc243, 0x09e0); // 10:5 = 0x3f
+NETSOC_NLM_C45_WRITE(nae, bus, phyaddr, dev_addr, 0xc243, 0x19e0); // 15:11= 0x2
 NETSOC_NLM_C45_WRITE(nae, bus, phyaddr, dev_addr, 0xcb0b, 0x0001); //  1:0 = 0x1
 NETSOC_NLM_C45_WRITE(nae, bus, phyaddr, dev_addr, 0xcb0b, 0x0011); //  5:4 = 0x1
 NETSOC_NLM_C45_WRITE(nae, bus, phyaddr, dev_addr, 0xcb0b, 0x0311); //  9:8 = 0x3
@@ -66,10 +66,10 @@ NETSOC_NLM_C45_WRITE(nae, bus, phyaddr, dev_addr, 0xcb1b, 0x1000); // 15:12= 0x1
 NETSOC_NLM_C45_WRITE(nae, bus, phyaddr, dev_addr, 0xca51, 0x0001); //  0:0 = 0x1
 NETSOC_NLM_C45_WRITE(nae, bus, phyaddr, dev_addr, 0xca51, 0x0011); //  7:1 = 0x8
 NETSOC_NLM_C45_WRITE(nae, bus, phyaddr, dev_addr, 0xc2c3, 0x000f); //  3:0 = 0xf
-NETSOC_NLM_C45_WRITE(nae, bus, phyaddr, dev_addr, 0xc2ca, 0x43c0); // 15:11= 0x8
-NETSOC_NLM_C45_WRITE(nae, bus, phyaddr, dev_addr, 0xc2d0, 0x43c0); // 15:11= 0x8
-NETSOC_NLM_C45_WRITE(nae, bus, phyaddr, dev_addr, 0xc2d6, 0x43c0); // 15:11= 0x8
-NETSOC_NLM_C45_WRITE(nae, bus, phyaddr, dev_addr, 0xc2dc, 0x43c0); // 15:11= 0x8
+NETSOC_NLM_C45_WRITE(nae, bus, phyaddr, dev_addr, 0xc2ca, 0x4700); // 15:11= 0x8
+NETSOC_NLM_C45_WRITE(nae, bus, phyaddr, dev_addr, 0xc2d0, 0x4700); // 15:11= 0x8
+NETSOC_NLM_C45_WRITE(nae, bus, phyaddr, dev_addr, 0xc2d6, 0x4700); // 15:11= 0x8
+NETSOC_NLM_C45_WRITE(nae, bus, phyaddr, dev_addr, 0xc2dc, 0x4700); // 15:11= 0x8
 NETSOC_NLM_C45_WRITE(nae, bus, phyaddr, dev_addr, 0xc2f1, 0x5f15); //  5:1 = 0xa
 NETSOC_NLM_C45_WRITE(nae, bus, phyaddr, dev_addr, 0xc2f1, 0x5e15); //  8:6 = 0x0
 NETSOC_NLM_C45_WRITE(nae, bus, phyaddr, dev_addr, 0xc2f5, 0x5f15); //  5:1 = 0xa
-- 
1.7.1

