From 6136aa62fc186acb5f351b084b1a070d9524cbc7 Mon Sep 17 00:00:00 2001
From: Anshu Kumari <anshuk@broadcom.com>
Date: Fri, 12 Dec 2014 16:03:38 +0530
Subject: Libraries : bit7 of XAUI Configuration Register set to 1

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/netlogic/lib/netlib/include/netsoc_dev.h b/arch/mips/netlogic/lib/netlib/include/netsoc_dev.h
index b681ce4..bbbcd89 100644
--- a/arch/mips/netlogic/lib/netlib/include/netsoc_dev.h
+++ b/arch/mips/netlogic/lib/netlib/include/netsoc_dev.h
@@ -862,6 +862,7 @@ enum {
     #define XAUI_CONFIG_TFEN            0x40000000
     #define XAUI_CONFIG_RCTLEN          0x20000000
     #define XAUI_CONFIG_RFEN            0x10000000
+    #define XAUI_CONFIG_RCTLSHRTP	0X00001000
     #define XAUI_CONFIG_DRPLT64         0x00000020
     #define XAUI_CONFIG_LENCHK          0x00000008
     #define XAUI_CONFIG_GENFCS          0x00000004
diff --git a/arch/mips/netlogic/lib/netlib/src/netsoc_nae.c b/arch/mips/netlogic/lib/netlib/src/netsoc_nae.c
index 3ad39eb..8f243c8 100644
--- a/arch/mips/netlogic/lib/netlib/src/netsoc_nae.c
+++ b/arch/mips/netlogic/lib/netlib/src/netsoc_nae.c
@@ -2226,7 +2226,8 @@ static void __netsoc_config_xaui(net_port_t *netport)
         netsoc_write_mac_reg(mac_base, XAUI_CONFIG_0, 0x0);
 
         /* Enable tx/rx frame */
-        val = 0xA10000A8;
+	//val = 0xA00001028;
+	val = (0 | XAUI_CONFIG_TCTLEN | XAUI_CONFIG_RCTLEN | XAUI_CONFIG_RCTLSHRTP | XAUI_CONFIG_DRPLT64 | XAUI_CONFIG_LENCHK);
         val |= (XAUI_CONFIG_LENCHK | XAUI_CONFIG_GENFCS | XAUI_CONFIG_PAD_64);
         netsoc_write_mac_reg(mac_base, XAUI_CONFIG_1, val );
         /* write max frame len*/
-- 
1.7.1

