From f33fbea4ee7ecdbc90287bf6801dec9646e65a43 Mon Sep 17 00:00:00 2001
From: Vikas Gupta <vikas.gupta@broadcom.com>
Date: Fri, 12 Dec 2014 17:09:29 +0530
Subject: xlp5xx nae: put workaround for A0 to trigger xoff from RX buffer before parser xoff gets triggered.

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/netlogic/lib/netlib/src/netsoc_nae.c b/arch/mips/netlogic/lib/netlib/src/netsoc_nae.c
index 8f243c8..a85cbe6 100644
--- a/arch/mips/netlogic/lib/netlib/src/netsoc_nae.c
+++ b/arch/mips/netlogic/lib/netlib/src/netsoc_nae.c
@@ -1263,6 +1263,17 @@ void __netsoc_config_context_xoff_thr(nae_t *nae, net_port_t *netport, int thgrp
 		maxbuf=maxbuf<<2;
 		reaction_len = inflight_len + mtu_len;	
 		xoff = ((maxbuf * NAE_RX_THR_BYTE_UNIT) - reaction_len - inflight_len - mtu_len) / NAE_RX_THR_BYTE_UNIT;
+		if(is_nlm_xlp5xx()){ /*h-w limitaiton on xlp5xx A0*/
+			if((netport->block_id==0)||(netport->block_id==1)){
+				if (netport->iftype == XLAUI_IF || netport->iftype == XAUI_IF ){
+					xoff = 652*4;
+				}else if(netport->iftype == RXAUI_IF){
+					xoff = 652*2;
+				}else{
+					xoff = 652;
+				}
+			}
+		}
 		xon = xoff - NAE_THR_SEPARATION;
 		if (xoff < 2*NAE_THR_SEPARATION) {
 			val = maxbuf / 16 ;
@@ -1395,6 +1406,18 @@ void __netsoc_config_parserfifo(nae_t *nae, int hw_port_id, uint32_t start, uint
         {
 		int line_delay = nae->mtu_config/smallest_pkt_len;
 		xoff = size - 2*(inflight_len/smallest_pkt_len) - line_delay; /*assuming other side is also having similar inflight delay*/
+		if(is_nlm_xlp5xx()){ /*h-w limitaiton on xlp5xx A0*/
+			if(hw_port_id != 8){ /*leave management port aside*/
+				if (iftype == XLAUI_IF || iftype == XAUI_IF ){
+					xoff = 250*4;
+				}else if (iftype == RXAUI_IF) {
+					xoff = 250*2;
+				}else{
+					xoff = 250;
+				}	
+			}
+		}
+
 	        xon = xoff - NAE_THR_SEPARATION;
 		if (xoff < 2*NAE_THR_SEPARATION) {
                         xoff = size - 30;
-- 
1.7.1

