From 0f8826e726754cf86553b613d876fe36e6ffe4f8 Mon Sep 17 00:00:00 2001
From: Vikas Gupta <vikas.gupta@broadcom.com>
Date: Sat, 13 Dec 2014 17:33:50 +0530
Subject: nae mac: added mac enable/disable support for xlaui

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/netlogic/lib/netlib/src/netsoc_nae.c b/arch/mips/netlogic/lib/netlib/src/netsoc_nae.c
index c7e87d8..0b459b5 100644
--- a/arch/mips/netlogic/lib/netlib/src/netsoc_nae.c
+++ b/arch/mips/netlogic/lib/netlib/src/netsoc_nae.c
@@ -2126,7 +2126,7 @@ void __netsoc_set_xlgmac_enable(net_port_t *netport)
 		val = 0x4; // Workaround for XLGMAC read bug
         __netsoc_write_xlgmac_reg(netport, XLAUI_RX_FIFO_SECTIONS, val);
 
-	val = 0x220d3; // PFC disable
+	val = 0x220d0; // PFC disable
         if (netport->vlan_pri_en) {
 		val |= (1<<19); /*PFC enable*/
 	}
@@ -2361,7 +2361,6 @@ static void __netsoc_config_xaui(net_port_t *netport)
 	val = netsoc_read_mac_reg(mac_base, NETIOR_XGMAC_CTRL3);
 	val &= 0xfffffc00;
 	val |= netport->context_base;
-
         netsoc_write_mac_reg(mac_base, NETIOR_XGMAC_CTRL3, val);
 	netsoc_api_print(NETSOC_APIDBG_PORT, "%s XAUI Config Complete block %d \n", __func__, netport->hw_port_id >> 2); return;
 }
@@ -3010,7 +3009,7 @@ int __netsoc_config_port(net_port_t *netport)
 */
 void __netsoc_mac_disable(net_port_t *netport)
 {
-        unsigned int mac_cfg1 = 0, xaui_cfg = 0;
+        unsigned int mac_cfg1 = 0, port_cfg = 0;
         unsigned int netwk_inf = 0;
 	nae_t *nae = netport->nae;
 	uint64_t mac_base;
@@ -3033,11 +3032,16 @@ void __netsoc_mac_disable(net_port_t *netport)
 					mac_base = netsoc_get_macreg_base_for_xgmac2(nae->mac_base, netport->hw_port_id);
 				}
 			}
-			xaui_cfg=netsoc_read_mac_reg(mac_base, XAUI_CONFIG_1);
-			netsoc_write_mac_reg(mac_base, XAUI_CONFIG_1, xaui_cfg &
+			port_cfg=netsoc_read_mac_reg(mac_base, XAUI_CONFIG_1);
+			netsoc_write_mac_reg(mac_base, XAUI_CONFIG_1, port_cfg &
                                                         (~(XAUI_CONFIG_TFEN | XAUI_CONFIG_RFEN)));
 			netsoc_api_print(NETSOC_APIDBG_PORT,"%s: XGMAC enable XAUI_CONFIG_1 0x%x\n",__func__, netsoc_read_mac_reg(mac_base, XAUI_CONFIG_1));
 			break;
+		case XLAUI_IF:
+        			mac_base = netsoc_get_macreg_base_for_xlgmac(nae->mac_base, netport->hw_port_id);
+				port_cfg = 0x220d0; // PFC disable
+        			__netsoc_write_xlgmac_reg(netport, XLAUI_COMMNAD_CONFIG, port_cfg);
+			break;
 		case INTERLAKEN_IF:
 			break;
 	}
@@ -3182,7 +3186,7 @@ void __netsoc_setup_hwaddr(net_port_t *netport, unsigned char *dev_addr)
 */
 void __netsoc_mac_enable(net_port_t *netport)
 {
-        unsigned int mac_cfg1 = 0, xaui_cfg = 0;
+        unsigned int mac_cfg1 = 0, port_cfg = 0;
         unsigned int netwk_inf = 0;
 	nae_t *nae = netport->nae;
 	uint64_t mac_base;
@@ -3205,14 +3209,19 @@ void __netsoc_mac_enable(net_port_t *netport)
 			//mac_base = netsoc_get_macreg_base_for_xgmac0(nae->mac_base, netport->hw_port_id);
 			mac_base = __netsoc_get_xgmac_base(netport);
 
-			netsoc_api_print(NETSOC_APIDBG_PORT,"%s: XGMAC enabling.. XAUI_CONFIG_1 0x%x\n",__func__, xaui_cfg);
-			xaui_cfg=netsoc_read_mac_reg(mac_base, XAUI_CONFIG_1);
-	                netsoc_write_mac_reg(mac_base, XAUI_CONFIG_1, xaui_cfg |
+			netsoc_api_print(NETSOC_APIDBG_PORT,"%s: XGMAC enabling.. XAUI_CONFIG_1 0x%x\n",__func__, port_cfg);
+			port_cfg=netsoc_read_mac_reg(mac_base, XAUI_CONFIG_1);
+	                netsoc_write_mac_reg(mac_base, XAUI_CONFIG_1, port_cfg |
 							XAUI_CONFIG_TFEN | XAUI_CONFIG_RFEN);
-			xaui_cfg=netsoc_read_mac_reg(mac_base, XAUI_CONFIG_1);
-			netsoc_api_print(NETSOC_APIDBG_PORT,"%s: XGMAC enabled XAUI_CONFIG_1 0x%x\n",__func__, xaui_cfg);
+			port_cfg=netsoc_read_mac_reg(mac_base, XAUI_CONFIG_1);
+			netsoc_api_print(NETSOC_APIDBG_PORT,"%s: XGMAC enabled XAUI_CONFIG_1 0x%x\n",__func__, port_cfg);
+			break;
+		case XLAUI_IF:
+			mac_base = netsoc_get_macreg_base_for_xlgmac(nae->mac_base, netport->hw_port_id);
+			port_cfg = 0x220d3; // PFC disable
+			__netsoc_write_xlgmac_reg(netport, XLAUI_COMMNAD_CONFIG, port_cfg);
+			break;
 
-                        break;
                 case INTERLAKEN_IF:
                         break;
         }
-- 
1.7.1

