From b29349090eeb3d44f2f3b7fdba71a218c4193175 Mon Sep 17 00:00:00 2001
From: kopal <kopal@broadcom.com>
Date: Mon, 15 Dec 2014 11:32:39 +0530
Subject: nae: added EXPORT SYMBOL for promiscuous mode APIs.

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/netlogic/lib/netlib/include/export_sym.h b/arch/mips/netlogic/lib/netlib/include/export_sym.h
index 50c767c..a1f6133 100644
--- a/arch/mips/netlogic/lib/netlib/include/export_sym.h
+++ b/arch/mips/netlogic/lib/netlib/include/export_sym.h
@@ -8,6 +8,8 @@ EXPORT_SYMBOL(cntx2netport);
 
 EXPORT_SYMBOL(netsoc_mac_disable);
 EXPORT_SYMBOL(netsoc_mac_enable);
+EXPORT_SYMBOL(netsoc_promisc_enable);
+EXPORT_SYMBOL(netsoc_promisc_disable);
 EXPORT_SYMBOL(netsoc_setup_hwaddr);
 EXPORT_SYMBOL(netsoc_get_phy_status);
 EXPORT_SYMBOL(netsoc_start_autoneg);
-- 
1.7.1

