From ad60030f6c5b764f91126889a8b911cfc69a72f1 Mon Sep 17 00:00:00 2001
From: Rahul Jain <rajain@netlogicmicro.com>
Date: Thu, 7 Jul 2011 16:28:37 +0530
Subject: ?lock_kernel issue while building linux_kernel_no_preempt

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/drivers/misc/netlogic/linux-loader/phnx_loader.c b/drivers/misc/netlogic/linux-loader/phnx_loader.c
index e697c7c..494e83d 100644
--- a/drivers/misc/netlogic/linux-loader/phnx_loader.c
+++ b/drivers/misc/netlogic/linux-loader/phnx_loader.c
@@ -38,6 +38,7 @@
 #include <asm/io.h>
 #include <asm/system.h>
 #include <linux/rwsem.h>
+#include <linux/smp_lock.h>
 
 #define Message(fmt, args...)      //printk("[%s]-[%d] "fmt"\n",__FUNCTION__,__LINE__,##args)
 #define dbg_msg(fmt, args...)      //printk("[%s]-[%d] "fmt"\n",__FUNCTION__,__LINE__,##args)
-- 
1.7.1

