From b1d2f6e2cb896b8e1069eee61133e4da78bbb36f Mon Sep 17 00:00:00 2001
From: Vikas Gupta <vikasg@netlogicmicro.com>
Date: Tue, 27 Sep 2011 19:31:09 +0530
Subject: Jumbo nae: Fix frame length settings for XAUI AND SGMII.

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/drivers/misc/netlogic/nae_jumbo/xlp_nae_jumbo.c b/drivers/misc/netlogic/nae_jumbo/xlp_nae_jumbo.c
index 342fa24..f0f7c9b 100755
--- a/drivers/misc/netlogic/nae_jumbo/xlp_nae_jumbo.c
+++ b/drivers/misc/netlogic/nae_jumbo/xlp_nae_jumbo.c
@@ -741,13 +741,9 @@ static void nlm_xlp_nae_init(void)
 			case SGMII_IF:
 				priv->index = nae_cfg.ports[i].hw_port_id & 0x3;
 				priv->phy.addr = nae_cfg.ports[i].hw_port_id;
-				write_gmac_reg(nae_cfg.ports[i].hw_port_id , MAX_FRM, 0x2300);
-				printk("MAX frame length set =0x%x for SGMII intf %d\n", read_gmac_reg(nae_cfg.ports[i].hw_port_id, MAX_FRM), nae_cfg.ports[i].hw_port_id);
 				break;
 			case XAUI_IF:
 				priv->index = XGMAC;
-				nlm_hal_write_mac_reg(nae_cfg.ports[i].hw_port_id, XGMAC, XAUI_MAX_FRAME_LEN,  0x8c02300);
-				printk("MAX frame length for set =0x%x for XAUI intf %d \n", nlm_hal_read_mac_reg(nae_cfg.ports[i].hw_port_id, XGMAC, XAUI_MAX_FRAME_LEN),nae_cfg.ports[i].hw_port_id);
 				break;	
 			case INTERLAKEN_IF:
 				priv->index = INTERLAKEN;
@@ -1264,6 +1260,14 @@ static void xlp_mac_setup_hwaddr(struct dev_data *priv)
 
         nlm_hal_write_mac_reg(priv->block, priv->index, MAC_FILTER_CONFIG, (1 << MAC_FILTER_BCAST_EN_POS) |
 						 (1 << MAC_FILTER_MCAST_EN_POS) | (1 << MAC_FILTER_ADDR0_VALID_POS) );
+	if(priv->type == XAUI_IF){
+		nlm_hal_write_mac_reg(priv->block, priv->index, XAUI_MAX_FRAME_LEN, 0x9002400);
+		printk("MAX frame length  set=0x%x for XAUI intf %d \n", nlm_hal_read_mac_reg(priv->block, priv->index, XAUI_MAX_FRAME_LEN), priv->block);
+	}
+	if(priv->type == SGMII_IF){			
+		nlm_hal_write_mac_reg(priv->block, priv->index, MAX_FRM, 0x2400);
+		printk("MAX frame length set=0x%x for SGMII intf %d\n", nlm_hal_read_mac_reg(priv->block, priv->index, MAX_FRM), priv->index);
+	}
 
 }
 
@@ -1293,7 +1297,6 @@ static int nlm_xlp_nae_change_mtu(struct net_device *dev, int new_mtu)
 	struct dev_data *priv = netdev_priv(dev);
 	unsigned long flags;
 	printk("nlm_xlp_nae_change_mtu called \n");
-	printk("XAUI RX STATS =0x%x\n", nlm_hal_read_mac_reg(0, XGMAC, 0x28));
 
 	if ((new_mtu > DEFAULT_JUMBO_MTU) || (new_mtu < MIN_ETH_FRAME_SIZE)) {
 		return -EINVAL;
-- 
1.7.1

