From 202d87ada565cb486a4acb142c660cef4f3457d1 Mon Sep 17 00:00:00 2001
From: Vikas Gupta <vikasg@netlogicmicro.com>
Date: Thu, 29 Sep 2011 16:53:45 +0530
Subject: Fix for LE mode. NAE expect P2D descriptors written in memory to be in BE.

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/drivers/misc/netlogic/nae_jumbo/xlp_nae_jumbo.c b/drivers/misc/netlogic/nae_jumbo/xlp_nae_jumbo.c
index f0f7c9b..1096f4e 100755
--- a/drivers/misc/netlogic/nae_jumbo/xlp_nae_jumbo.c
+++ b/drivers/misc/netlogic/nae_jumbo/xlp_nae_jumbo.c
@@ -933,7 +933,7 @@ static inline int create_p2p_desc(uint64_t paddr, uint64_t len, uint64_t *p2pmsg
 	int plen;
 	do {
 		plen = len >= MAX_PACKET_SZ_PER_MSG ? (MAX_PACKET_SZ_PER_MSG - 64): len;
-		p2pmsg[idx] = nae_tx_desc(P2D_NEOP, 0, NULL_VFBID, plen, paddr);
+		p2pmsg[idx] = cpu_to_be64(nae_tx_desc(P2D_NEOP, 0, NULL_VFBID, plen, paddr));
 		len -= plen;
 		paddr += plen;
 		idx++;
@@ -944,7 +944,7 @@ static inline int create_p2p_desc(uint64_t paddr, uint64_t len, uint64_t *p2pmsg
 
 static inline void create_last_p2p_desc(uint64_t *p2pmsg, struct sk_buff *skb, int idx)
 {
-	p2pmsg[idx - 1] |= ((uint64_t)P2D_EOP << 62);
+	p2pmsg[idx - 1] |= cpu_to_be64(((uint64_t)P2D_EOP << 62));
 	p2pmsg[P2P_SKB_OFF] = (uint64_t)skb;
 }
 
-- 
1.7.1

