From 5e84d3b1edcb381d958cfff381d582dda64759cb Mon Sep 17 00:00:00 2001
From: Rahul Jain <rajain@netlogicmicro.com>
Date: Sat, 10 Dec 2011 21:43:09 +0530
Subject: Vuart fixes for LinuxLoader

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/drivers/misc/netlogic/linux-loader/phnx_loader.h b/drivers/misc/netlogic/linux-loader/phnx_loader.h
index 429884a..2fa01c3 100644
--- a/drivers/misc/netlogic/linux-loader/phnx_loader.h
+++ b/drivers/misc/netlogic/linux-loader/phnx_loader.h
@@ -32,8 +32,8 @@
 #define PHNX_LOADER_INFO_MAGIC 			0x600ddeed
 #define XLR_MAP_SLAVE_DEVICE 			0x1
 #define XLR_MAP_UNCACHED 			0x1
-#define XLR_MAP_CACHED 				0x2
-#define PHNX_APP_LOADER_CHRDEV_NAME 		"xlr_app_loader"
+#define XLR_MAP_CACHED				0x2
+#define PHNX_APP_LOADER_CHRDEV_NAME 		"linux_loader"
 #define MAX_NUM_LOADER_MEM_BLK 			4
 #define SHARED_MEM_ADDR            		(3 << 20)
 #define XLP_APP_LOADER_MAJOR       		(245)
-- 
1.7.1

