From cbc2af292f6ff0d41d78a26e6b78b318d759a0fb Mon Sep 17 00:00:00 2001
From: Jayanthi A <jayanthia@netlogicmicro.com>
Date: Mon, 30 Jan 2012 10:08:16 +0530
Subject: Removed naecfg_hack

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/drivers/misc/netlogic/nae-perf/net_common.h b/drivers/misc/netlogic/nae-perf/net_common.h
index d67f2f5..9cecbcf 100644
--- a/drivers/misc/netlogic/nae-perf/net_common.h
+++ b/drivers/misc/netlogic/nae-perf/net_common.h
@@ -65,11 +65,6 @@ static __inline__ void press_key_to_continue(void) {
 #define press_key_to_continue()
 #endif
 
-enum NAE_REG_CMD {
-	CMD_READ = 0,
-	CMD_WRITE
-};
-
 #define NAE_RX_ENABLE 0x1
 #define NAE_TX_ENABLE 0x1
 
-- 
1.7.1

