From acea9ad2acb470b0895e5ad567e857737584d62e Mon Sep 17 00:00:00 2001
From: Rahul Jain <rajain@netlogicmicro.com>
Date: Sat, 11 Feb 2012 18:59:59 +0530
Subject: nae-perf:Fix for ABI 32 and preempt linux kernel build

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/drivers/misc/netlogic/nae-perf/xlp_nae.c b/drivers/misc/netlogic/nae-perf/xlp_nae.c
index 582a03a..8da70d4 100755
--- a/drivers/misc/netlogic/nae-perf/xlp_nae.c
+++ b/drivers/misc/netlogic/nae-perf/xlp_nae.c
@@ -732,7 +732,9 @@ static int mac_refill_frin_one_buffer(struct net_device *dev, int cpu, int len)
 	priv = netdev_priv(ndev);
 	skb->dev = ndev;
 
+#ifdef CONFIG_NLM_NET_OPTS
 	skb->queue_id = freein_fifo;
+#endif
 
 	mac_put_skb_back_ptr(skb);
 
@@ -912,7 +914,9 @@ static int xlp_poll_lower(int budget, int cpu)
 		if (err) {
 			vaddr = (uint64_t)bus_to_virt(addr);
 			skb = mac_get_skb_back_ptr(vaddr);
+#ifdef CONFIG_NLM_NET_OPTS
 			mac_refill_frin_skb(addr, skb->queue_id);
+#endif
 			STATS_INC(priv->stats.rx_errors);
 			STATS_INC(priv->stats.rx_dropped);
 			err_replenish_count[LAST_RCVD_INDEX(cpu)]++;
@@ -948,7 +952,9 @@ static int xlp_poll_lower(int budget, int cpu)
 		len = len  - ETH_FCS_LEN;
 
 		skb = mac_get_skb_back_ptr(vaddr);
+#ifdef CONFIG_NLM_NET_OPTS
 		//printk ("cpu %d Freein_fifo %d len %d skb->len %d \n", cpu, skb->queue_id, len, skb->len);
+#endif
 #ifdef ENABLE_SANITY_CHECKS
 		if (!skb) {
 			STATS_INC(priv->stats.rx_dropped);
@@ -1136,7 +1142,9 @@ static void nlm_replenish_per_cpu_buffer(int qindex, int bufcnt)
 			printk("[%s] alloc skb failed\n",__FUNCTION__);
 			break;
 		}
+#ifdef CONFIG_NLM_NET_OPTS
 		skb->queue_id = vc_index;
+#endif
 		/* Store skb in back_ptr */
 		mac_put_skb_back_ptr(skb);
 		code = 0;
@@ -1727,8 +1735,10 @@ static int nlm_xlp_nae_start_xmit(struct sk_buff *skb, struct net_device *dev)
 	{
 		last_rcvd_skb[LAST_RCVD_INDEX(cpu)] = NULL;
 		/* Do h/w replenishment and 4 CPUs share a FIFO */
+#ifdef CONFIG_NLM_NET_OPTS
 		msg0 = nae_tx_desc(P2D_NEOP, 0, skb->queue_id - frin_queue_base + 32,
 				0, last_rcvd_skb_phys[LAST_RCVD_INDEX(cpu)]);
+#endif
 		hw_repl = 1;
 		//printk ("%s skb->len %d \n", __func__, skb->len);
 
-- 
1.7.1

