From 3ca1dab789f1e469c65da91e88637981be8791c5 Mon Sep 17 00:00:00 2001
From: Hareesh R <hareeshr@broadcom.com>
Date: Thu, 15 Mar 2012 23:52:42 +0530
Subject: Prototype change for ucore shared data write

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/drivers/misc/netlogic/nae-perf/xlp_nae.c b/drivers/misc/netlogic/nae-perf/xlp_nae.c
index cb0d0d5..c017b10 100755
--- a/drivers/misc/netlogic/nae-perf/xlp_nae.c
+++ b/drivers/misc/netlogic/nae-perf/xlp_nae.c
@@ -1105,7 +1105,7 @@ static void nlm_update_ucore_shared_memory(int node)
 	else
 		data[34] = 8;
 
-	nlm_hal_write_ucore_shared_mem(data, sizeof(data)/sizeof(uint32_t));
+	nlm_hal_write_ucore_shared_mem(0, data, sizeof(data)/sizeof(uint32_t));
 }
 
 static void nlm_replenish_per_cpu_buffer(int qindex, int bufcnt)
-- 
1.7.1

