From 7cccf106496822199787315a82907267a907be29 Mon Sep 17 00:00:00 2001
From: Yonghong Song <ysong@broadcom.com>
Date: Fri, 14 Sep 2012 15:37:53 -0700
Subject: errata: add ehb after sync/sc

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/drivers/misc/netlogic/linux-loader/xlr_lib_launch.c b/drivers/misc/netlogic/linux-loader/xlr_lib_launch.c
index d2a5092..fa18008 100644
--- a/drivers/misc/netlogic/linux-loader/xlr_lib_launch.c
+++ b/drivers/misc/netlogic/linux-loader/xlr_lib_launch.c
@@ -78,7 +78,12 @@ static void (*cb_prelaunch_init_kseg)(unsigned long);
 static __inline__ void lib_sync(void)
 {
 	__asm__ volatile(
+		".set push\n"
+		".set noreorder\n"
+		".set mips64r2\n"
 		"sync\n"
+		"ehb\n"
+		".set pop\n"
 	);
 
 }
-- 
1.7.1

