From 74b1d54e70e2d04b222e2bbb87a6efdc8c8ce63d Mon Sep 17 00:00:00 2001
From: Alok Agrawat <alok@broadcom.com>
Date: Thu, 4 Oct 2012 05:32:53 -0700
Subject: cde: Fixed rmmod crash.

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/drivers/misc/netlogic/nlm_cde/nlm_cde.c b/drivers/misc/netlogic/nlm_cde/nlm_cde.c
index 640e3d0..0857042 100644
--- a/drivers/misc/netlogic/nlm_cde/nlm_cde.c
+++ b/drivers/misc/netlogic/nlm_cde/nlm_cde.c
@@ -654,7 +654,6 @@ int xlp_cde_open(struct inode *inode, struct file *filp)
 void nlm_cde_cleanup(cmp_data_t *cmp_data)
 {
   int i;
-
   if (cmp_data->src)  
 	free_pages ((ulong)cmp_data->src, get_order (MAX_INPUT_BUFFER_SIZE));
   if (cmp_data->target)  
@@ -665,24 +664,25 @@ void nlm_cde_cleanup(cmp_data_t *cmp_data)
 	  kfree(cmp_data->scratch);
   if (cmp_data->scratch_inf)  
 	  kfree(cmp_data->scratch_inf);
-  if (cmp_data)
-    kfree(cmp_data);
+
   free_pages((ulong)cde_mem_start_1, get_order(cde_mem_size_1));
   free_pages((ulong)cde_mem_start_2, get_order(cde_mem_size_2));
   free_pages((ulong)cde_mem_start_3, get_order(cde_mem_size_3));
   free_pages((ulong)cde_mem_start_4, get_order(cde_mem_size_4));
   free_pages((ulong)cde_mem_start_5, get_order(cde_mem_size_5));
 
-    cmp_data->src = NULL;
-    cmp_data->target = NULL;
-    cmp_data->src_desc = NULL; 
-    cmp_data->scratch = NULL;
-    cmp_data->scratch_inf = NULL;
-    cmp_data = NULL;
+  cmp_data->src = NULL;
+  cmp_data->target = NULL;
+  cmp_data->src_desc = NULL; 
+  cmp_data->scratch = NULL;
+  cmp_data->scratch_inf = NULL;
+
+  if (cmp_data)
+    kfree(cmp_data);
 
   for (i = 0; i < NUM_FREE_DESCRIPTORS; i++) {
     if (page_array[i].data_array)	{
-	free_cache_aligned_mem((void*)page_array_tmp_data_array[i]);
+	free_cache_aligned_mem((void*)page_array[i].data_array);
 	page_array[i].data_array = NULL;
     }	
   }
-- 
1.7.1

