From ae9f23b5a88cf822343fb4af517f6272231f345c Mon Sep 17 00:00:00 2001
From: Alok Agrawat <alok@broadcom.com>
Date: Thu, 4 Oct 2012 04:19:57 -0700
Subject: cde: Fixed the test app.

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/drivers/misc/netlogic/nlm_cde/cdetest.c b/drivers/misc/netlogic/nlm_cde/cdetest.c
index 064fd13..27e5b9b 100644
--- a/drivers/misc/netlogic/nlm_cde/cdetest.c
+++ b/drivers/misc/netlogic/nlm_cde/cdetest.c
@@ -29,6 +29,7 @@ THE POSSIBILITY OF SUCH DAMAGE.
 #include <stdio.h>
 #include <string.h>
 #include <fcntl.h>
+#include <sys/mman.h>
 
 static int sleep_usec = 100;
 static int LEN = 1024 * 16;
@@ -104,6 +105,7 @@ int main(int argc, char *argv[])
         char buf[LEN];
 	char buf1[LEN];
         char buf2[LEN];
+	unsigned char *cde_mmap_mem = NULL;
 
 	int tfd = open(argv[1], O_RDONLY);
         int vfd = open("out.txt",O_WRONLY);
@@ -130,6 +132,11 @@ int main(int argc, char *argv[])
 	}
 //        test();
   //      return 0;
+#if 1
+       cde_mmap_mem = mmap (0, 2048, PROT_READ | PROT_WRITE, MAP_SHARED, dfd, 0);
+       if(cde_mmap_mem)
+               strcpy(cde_mmap_mem, "HelloWorld\n");
+#endif
 
         cde.compression_type = 1;
         cde.sof = 1;
-- 
1.7.1

