From 6d18215340e6c851eaf16c1d4186ce8dd333f772 Mon Sep 17 00:00:00 2001
From: Anurag <anurag.gopinath@broadcom.com>
Date: Thu, 16 May 2013 17:36:51 +0530
Subject: CDE : Adding support for XLP9XX and moving to the new HAL structure

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/drivers/misc/netlogic/nlm_cde/nlm_cde.c b/drivers/misc/netlogic/nlm_cde/nlm_cde.c
index bcd6039..2319df8 100644
--- a/drivers/misc/netlogic/nlm_cde/nlm_cde.c
+++ b/drivers/misc/netlogic/nlm_cde/nlm_cde.c
@@ -41,12 +41,16 @@
 #include <asm/uaccess.h>	/* for get_user, etc. */
 #include <linux/init.h>		/* for __init, module_{init,exit} */
 #include <linux/poll.h>		/* for POLLIN, etc. */
-#include <asm/netlogic/msgring.h>
+#include <nlm_msgring.h>
+#include <nlm_hal.h>
 #include <asm/mutex.h>
 
-#include <asm/netlogic/hal/nlm_hal_fmn.h>
-#include <asm/netlogic/hal/nlm_hal_macros.h>
-#include <asm/netlogic/hal/nlm_hal_cde.h>
+#include <nlm_xlp.h>
+#include <nlm_hal_fmn.h>
+#include <nlm_hal_macros.h>
+#include <asm/netlogic/xlp-hal/xlp.h>
+#include <nlm_hal_cde.h>
+#include <nlm_hal_fmn_dp.h>
 #include "nlm_cde.h"
 #include "nlm_cde_api.h"
 
@@ -204,7 +208,7 @@ static int send_message(int stid, struct msgrng_msg *msg)
 	printk("send_message = stid = %x msg = %llx\n",stid, msg->msg0);
 #endif
   msgrng_access_enable(mflags);
-  nlm_hal_cde_send_request(stid, msg->msg0);
+  nlm_hal_cde_send_request(stid, msg->msg0, msg->msg1);
   msgrng_access_disable(mflags);
 
   return ret;
@@ -333,6 +337,10 @@ int create_inf_message(cmp_data_t *cmp_data, int type, int start_of_file, int en
       cur_blk++;
     }
 
+    if (cpu_is_xlp9xx())
+    stid = nlm_hal_9xxcde_make_cmp_msg((uint64_t*)&cmp_msg[i].msg0, (uint64_t*)&cmp_msg[i].msg1, 0, cmp_data->op, rtn_bkt,
+			cur_desc, (uint64_t)virt_to_phys(cmp_data->src_desc + desc_idx));
+    else
     stid = nlm_hal_cde_make_cmp_msg((uint64_t*)&cmp_msg[i].msg0, 0, cmp_data->op, rtn_bkt,
 			cur_desc, (uint64_t)virt_to_phys(cmp_data->src_desc + desc_idx));
 #ifdef CDE_DEBUG
@@ -437,6 +445,10 @@ int create_message(cmp_data_t *cmp_data, int type, int start_of_file, int end_of
       cur_blk++;
     }
 
+    if (cpu_is_xlp9xx())
+    stid = nlm_hal_9xxcde_make_cmp_msg((uint64_t*)&cmp_msg[i].msg0,(uint64_t*)&cmp_msg[i].msg1, 0, cmp_data->op, rtn_bkt,
+			cur_desc, (uint64_t)virt_to_phys(cmp_data->src_desc + desc_idx));
+    else
     stid = nlm_hal_cde_make_cmp_msg((uint64_t*)&cmp_msg[i].msg0, 0, cmp_data->op, rtn_bkt,
 			cur_desc, (uint64_t)virt_to_phys(cmp_data->src_desc + desc_idx));
 #ifdef CDE_DEBUG
@@ -499,7 +511,7 @@ static int nlm_hal_cde_read_cmp_msg(char *buffer, uint64_t payload)
   uint64_t i,j,num_bytes;
   int offset = 0;
   uint64_t *desc, dest_addr;
-  int num_desc = (payload >> 40) & 0x3fff;
+  int num_desc = (payload >> 40) & 0x1fff;
   char * tmp_ptr;
   for (i = 0; i < num_desc; i++) {
     desc = phys_to_virt(payload & 0xffffffffffUll) + i*8; //64 byte descriptors //dliao: why i*8??
@@ -510,7 +522,7 @@ static int nlm_hal_cde_read_cmp_msg(char *buffer, uint64_t payload)
     tmp_ptr = (char *) phys_to_virt(dest_addr & 0xffffffffffUll);
 
     for (j = 0; j < num_bytes; j++) {
-      buffer[offset+j] = tmp_ptr[j];
+      buffer[offset+j] = tmp_ptr[j]; //AGN: make it Memcpy, and need to check perf difference
     }
     offset = offset + num_bytes;
   }
@@ -545,6 +557,9 @@ nlm_xlp_cde_msgring_handler(uint32_t vc, uint32_t src_id,
 #endif
 
   offset = cmp_data->target_size;
+  if (cpu_is_xlp9xx())
+  cmp_data->target_size += nlm_hal_cde_read_cmp_msg((char *) cmp_data->target + offset, msg->msg0);
+  else
   cmp_data->target_size += nlm_hal_cde_read_cmp_msg((char *) cmp_data->target + offset, msg->msg1);
 
 #ifdef CDE_DEBUG
@@ -970,6 +985,8 @@ static int __init xlp_cde_init(void)
     printk("xlp_cde_init done\n\n");
     if(is_nlm_xlp2xx())
   	num_cde_engine = NUM_CDE_ENGINE_XLP2XX;
+    else if (cpu_is_xlp9xx())
+  	num_cde_engine = NUM_CDE_ENGINE_XLP9XX;
     else
   	num_cde_engine = NUM_CDE_ENGINE;
 {
diff --git a/drivers/misc/netlogic/nlm_cde/nlm_cde.h b/drivers/misc/netlogic/nlm_cde/nlm_cde.h
index 3cbe33d..11264e8 100644
--- a/drivers/misc/netlogic/nlm_cde/nlm_cde.h
+++ b/drivers/misc/netlogic/nlm_cde/nlm_cde.h
@@ -34,6 +34,7 @@
 #define NLM_CDE_VC_BASE 297
 #define NUM_CDE_ENGINE 4
 #define NUM_CDE_ENGINE_XLP2XX 	1
+#define NUM_CDE_ENGINE_XLP9XX 	3
 
 #define CDE_INFLATE 			1
 #define CDE_DEFLATE 			2
-- 
1.7.1

