From 922db4bbf60e884abe58680721a0f855e3d66c72 Mon Sep 17 00:00:00 2001
From: Lakshman Garlapati <kumarga@broadcom.com>
Date: Wed, 22 May 2013 18:57:28 +0530
Subject: DPI: added XLP9XX support

Modified Makefile to support XLP9XX.
[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/drivers/misc/netlogic/netl7driver/netl7driver.c b/drivers/misc/netlogic/netl7driver/netl7driver.c
index a798d9f..063a8fb 100755
--- a/drivers/misc/netlogic/netl7driver/netl7driver.c
+++ b/drivers/misc/netlogic/netl7driver/netl7driver.c
@@ -49,6 +49,10 @@
 
 #define DRV_NAME "NetL7_V2.4"
 
+#ifndef VM_RESERVED
+#define VM_RESERVED (VM_DONTEXPAND | VM_DONTDUMP)
+#endif
+
 /* Uncomment if MP support is required */
 //#define MP_SUPPORT 1
 
@@ -584,7 +588,7 @@ nlm_device_mmap (struct file *file_p, struct vm_area_struct *vm_p)
   
   /*NLM_INFO (":mmap The file reference count is %ld\n", file_p->f_count.counter);*/
 
-  vm_p->vm_flags |= VM_DONTEXPAND | VM_DONTDUMP;
+  vm_p->vm_flags |= VM_RESERVED;
   vm_p->vm_flags |= VM_IO;
   cacheable = 1;
   
-- 
1.7.1

