From fd4805b6d98faeb507cf4b3e954f2190e6fc8b7e Mon Sep 17 00:00:00 2001
From: Sreenidhi B R <sreenira@broadcom.com>
Date: Tue, 11 Jun 2013 16:55:28 +0530
Subject: xlp9xx: add vc map structure for xlp9xx.

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/drivers/misc/netlogic/soc_interface/on_chip.c b/drivers/misc/netlogic/soc_interface/on_chip.c
index ec5bc94..652caeb 100644
--- a/drivers/misc/netlogic/soc_interface/on_chip.c
+++ b/drivers/misc/netlogic/soc_interface/on_chip.c
@@ -268,6 +268,60 @@ static uint16_t xlp2xx_vc_to_handle_map[MAX_VC] = {
 
 };
 
+/* TODO: popq handlers are to be added.
+   */
+
+static uint16_t xlp9xx_vc_to_handle_map[MAX_VC] = {
+	[0 ... 15] = XLP_MSG_HANDLE_CPU0,
+	[16 ... 31] = XLP_MSG_HANDLE_CPU1,
+	[32 ... 47] = XLP_MSG_HANDLE_CPU2,
+	[48 ... 63] = XLP_MSG_HANDLE_CPU3,
+	[64 ... 79] = XLP_MSG_HANDLE_CPU4,
+	[80 ... 95] = XLP_MSG_HANDLE_CPU5,
+	[96 ... 111] = XLP_MSG_HANDLE_CPU6,
+	[112 ... 127] = XLP_MSG_HANDLE_CPU7,
+	[128 ... 143] = XLP_MSG_HANDLE_CPU8,
+	[144 ... 159] = XLP_MSG_HANDLE_CPU9,
+	[160 ... 175] = XLP_MSG_HANDLE_CPU10,
+	[176 ... 191] = XLP_MSG_HANDLE_CPU11,
+	[192 ... 207] = XLP_MSG_HANDLE_CPU12,
+	[208 ... 223] = XLP_MSG_HANDLE_CPU13,
+	[224 ... 239] = XLP_MSG_HANDLE_CPU14,
+	[240 ... 255] = XLP_MSG_HANDLE_CPU15,
+	[256 ... 271] = XLP_MSG_HANDLE_CPU16,
+	[272 ... 287] = XLP_MSG_HANDLE_CPU17,
+	[288 ... 303] = XLP_MSG_HANDLE_CPU18,
+	[304 ... 319] = XLP_MSG_HANDLE_CPU19,
+	[320 ... 383] = XLP_MSG_HANDLE_INVALID,
+	[384 ... 385] = XLP_MSG_HANDLE_PCIE0,
+	[386 ... 387] = XLP_MSG_HANDLE_PCIE1,
+	[388 ... 389] = XLP_MSG_HANDLE_PCIE2,
+	[390 ... 391] = XLP_MSG_HANDLE_PCIE3,
+	[392 ... 393] = XLP_MSG_HANDLE_GDX_0,
+	[394 ... 395] = XLP_MSG_HANDLE_GDX_1,
+	[396 ... 399] = XLP_MSG_HANDLE_INVALID,
+	[400 ... 406] = XLP_MSG_HANDLE_CMP,
+	[407 ... 411] = XLP_MSG_HANDLE_INVALID,
+	[412 ... 418] = XLP_MSG_HANDLE_LZS,
+	[419 ... 423] = XLP_MSG_HANDLE_INVALID,
+	[424 ... 431] = XLP_MSG_HANDLE_CRYPTO,
+	[432 ... 447] = XLP_MSG_HANDLE_INVALID,
+	[448 ... 451] = XLP_MSG_HANDLE_RSA_ECC,
+	[452 ... 459] = XLP_MSG_HANDLE_INVALID,
+	[460 ... 463] = XLP_MSG_HANDLE_REGX,
+	[464 ... 475] = XLP_MSG_HANDLE_INVALID,
+	[476 ... 479] = XLP_MSG_HANDLE_KBP,
+	[480 ... 511] = XLP_MSG_HANDLE_INVALID,
+	[512 ... 519] = XLP_MSG_HANDLE_POE_0,
+	[520 ... 599] = XLP_MSG_HANDLE_NAE_0,
+	[600 ... 639] = XLP_MSG_HANDLE_INVALID,
+	[640 ... 647] = XLP_MSG_HANDLE_POE_1,
+	[648 ... 727] = XLP_MSG_HANDLE_NAE_1,
+	[728 ... 729] = XLP_MSG_HANDLE_INVALID,
+	[730 ... 731] = XLP_MSG_HANDLE_ALE,
+	[732 ... 4095]= XLP_MSG_HANDLE_INVALID
+};
+
 static uint16_t *vc_to_handle = vc_to_handle_map;
 
 /******************************************************************************************
@@ -1042,6 +1096,8 @@ static int __init on_chip_init(void)
 		vc_to_handle = xlp2xx_vc_to_handle_map;
 	else if (is_nlm_xlp3xx())
 		vc_to_handle = xlp3xx_vc_to_handle_map;
+	else if (is_nlm_xlp9xx())
+		vc_to_handle = xlp9xx_vc_to_handle_map;
 	else
 		vc_to_handle = vc_to_handle_map;
 
-- 
1.7.1

