From 444fb8b862475724f40d459e35ef939418f03fe9 Mon Sep 17 00:00:00 2001
From: Yonghong Song <ysong@broadcom.com>
Date: Tue, 25 Jun 2013 13:41:53 -0700
Subject: HEv2: kmod: add ioctl hook for packet memory driver for compatibility mode

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/drivers/misc/netlogic/pkt_pool_mem/pkt_pool_mem.c b/drivers/misc/netlogic/pkt_pool_mem/pkt_pool_mem.c
index 4440d1a..e5efbf3 100644
--- a/drivers/misc/netlogic/pkt_pool_mem/pkt_pool_mem.c
+++ b/drivers/misc/netlogic/pkt_pool_mem/pkt_pool_mem.c
@@ -211,6 +211,9 @@ static long pktmem_ioctl (struct file *fptr, unsigned int cmd, unsigned long dp)
 static const struct file_operations pktmem_ops = {
 	.owner  = THIS_MODULE,
 	.unlocked_ioctl = pktmem_ioctl,
+#ifdef CONFIG_COMPAT
+	.compat_ioctl = pktmem_ioctl,
+#endif
 };
 
 static inline int get_fdt_prop(void *fdt, const char* node_path,
-- 
1.7.1

