From 82c89918946c26ad39eb5a54fb5c8e974f414a81 Mon Sep 17 00:00:00 2001
From: Prasanna Mumbai <prasanna.mumbai@broadcom.com>
Date: Wed, 3 Jul 2013 15:19:22 +0530
Subject: HEv2 : Update pktpoolmem for page alignment & some improvements

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/drivers/misc/netlogic/pkt_pool_mem/pkt_pool_mem.c b/drivers/misc/netlogic/pkt_pool_mem/pkt_pool_mem.c
index d4a41e4..98c0d06 100644
--- a/drivers/misc/netlogic/pkt_pool_mem/pkt_pool_mem.c
+++ b/drivers/misc/netlogic/pkt_pool_mem/pkt_pool_mem.c
@@ -56,7 +56,7 @@ static int print_memory(brcm_devmem_ioctl_t * param)
 
 static int alloc_memory (brcm_devmem_ioctl_t * param)
 {
-	int alloc;
+	int alloc, align_len;
 	memory_pool_t *mem, *newmem;
 
 	/* take care of shared memory request first */
@@ -64,6 +64,11 @@ static int alloc_memory (brcm_devmem_ioctl_t * param)
 	alloc = 0;
 	if (param->shr_id != NULL)
 	{
+		if (param->size < 0x100000)
+		{
+			printk ("Packet allocation request should be > 1MB.\n");
+			return -1;
+		}
 		while (mem != NULL)
 		{
 			if (strcmp(param->shr_id,mem->shr_id) == 0)
@@ -76,11 +81,20 @@ static int alloc_memory (brcm_devmem_ioctl_t * param)
 		}
 	}
 
+
 	if (alloc)
 		return 0;
 
 	mem = &mem_pool_data[param->node_id];
 	alloc = 0;
+
+	/*
+	 * Align the packet allocation to 64K page size.
+	 */
+	if((align_len = param->size % 65536) != 0)
+		param->size += (65536 - align_len) ;
+
+
 	while (mem != NULL)
 	{
 		if (mem->in_use == 0)
-- 
1.7.1

