From 95b47c2eaa628a388afb4c02160cec9e59edb543 Mon Sep 17 00:00:00 2001
From: Prasanna Mumbai <prasanna.mumbai@broadcom.com>
Date: Fri, 5 Jul 2013 14:18:24 +0530
Subject: HEv2: Create a header file to be shared between userspace & kernel.

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/drivers/misc/netlogic/pkt_pool_mem/pkt_pool_kern_iface.h b/drivers/misc/netlogic/pkt_pool_mem/pkt_pool_kern_iface.h
new file mode 100644
index 0000000..395ff01
--- /dev/null
+++ b/drivers/misc/netlogic/pkt_pool_mem/pkt_pool_kern_iface.h
@@ -0,0 +1,17 @@
+#ifndef __PKT_POOL_KERN_IFACE_H
+#define __PKT_POOL_KERN_IFACE_H
+
+typedef struct brcm_devmem_ioctl {
+	unsigned int node_id;
+	char shr_id[64];
+	unsigned long long phys_addr;
+	unsigned long long size;
+} brcm_devmem_ioctl_t;
+
+enum pktmem_ops {
+	ALLOC_POOL = 0,
+	FREE_POOL,
+	PRINT_POOL,
+};
+
+#endif
diff --git a/drivers/misc/netlogic/pkt_pool_mem/pkt_pool_mem.c b/drivers/misc/netlogic/pkt_pool_mem/pkt_pool_mem.c
index 98c0d06..2b50884 100644
--- a/drivers/misc/netlogic/pkt_pool_mem/pkt_pool_mem.c
+++ b/drivers/misc/netlogic/pkt_pool_mem/pkt_pool_mem.c
@@ -6,6 +6,7 @@
 #include <linux/fs.h>
 #include <linux/mm.h>
 #include "libfdt.h"
+#include "pkt_pool_kern_iface.h"
 
 
 #define MAX_NODES 	8
@@ -23,19 +24,6 @@ typedef struct memory_pool
 
 static memory_pool_t mem_pool_data[MAX_NODES];
 
-typedef struct brcm_devmem_ioctl {
-	unsigned int node_id;
-	char shr_id[64];
-	unsigned long long phys_addr;
-	unsigned long long size;
-} brcm_devmem_ioctl_t;
-
-enum pktmem_ops {
-	ALLOC_POOL = 0,
-	FREE_POOL,
-	PRINT_POOL,
-};
-
 #undef PKTMEM_DEBUG
 
 static int print_memory(brcm_devmem_ioctl_t * param)
-- 
1.7.1

