From d5b93d27991d70c6e04aebcfb1a9c7176584a84f Mon Sep 17 00:00:00 2001
From: Vikas Gupta <vikas.gupta@broadcom.com>
Date: Mon, 17 Jun 2013 16:32:23 -0700
Subject: fmn: registering handler according to outputq number.

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/drivers/misc/netlogic/soc_interface/on_chip.c b/drivers/misc/netlogic/soc_interface/on_chip.c
index b963504..4df329e 100644
--- a/drivers/misc/netlogic/soc_interface/on_chip.c
+++ b/drivers/misc/netlogic/soc_interface/on_chip.c
@@ -131,7 +131,7 @@ static uint16_t vc_to_handle_map[MAX_VC] = {
 	[281 ... 296] = XLP_MSG_HANDLE_CRYPTO,
 	[297 ... 304] = XLP_MSG_HANDLE_CMP,
 	[305 ... 383] = XLP_MSG_HANDLE_INVALID,
-	[384 ... 391] = XLP_MSG_HANDLE_NAE_0,
+	[384 ... 391] = XLP_MSG_HANDLE_POE_0,
 	[392 ... 475] = XLP_MSG_HANDLE_INVALID,
 	[476 ... 1019] = XLP_MSG_HANDLE_NAE_0,
 	[1020 ... 1023] = XLP_MSG_HANDLE_INVALID,
@@ -155,7 +155,7 @@ static uint16_t vc_to_handle_map[MAX_VC] = {
 	[1305 ... 1320] = XLP_MSG_HANDLE_CRYPTO,
 	[1321 ... 1328] = XLP_MSG_HANDLE_CMP,
 	[1329 ... 1407] = XLP_MSG_HANDLE_INVALID,
-	[1408 ... 1415] = XLP_MSG_HANDLE_NAE_0,
+	[1408 ... 1415] = XLP_MSG_HANDLE_POE_0,
 	[1416 ... 1499] = XLP_MSG_HANDLE_INVALID,
 	[1500 ... 2043] = XLP_MSG_HANDLE_NAE_0,
 	[2044 ... 2047] = XLP_MSG_HANDLE_INVALID,
@@ -179,7 +179,7 @@ static uint16_t vc_to_handle_map[MAX_VC] = {
 	[2329 ... 2344] = XLP_MSG_HANDLE_CRYPTO,
 	[2345 ... 2352] = XLP_MSG_HANDLE_CMP,
 	[2353 ... 2431] = XLP_MSG_HANDLE_INVALID,
-	[2432 ... 2439] = XLP_MSG_HANDLE_NAE_0,
+	[2432 ... 2439] = XLP_MSG_HANDLE_POE_0,
 	[2440 ... 2523] = XLP_MSG_HANDLE_INVALID,
 	[2524 ... 3067] = XLP_MSG_HANDLE_NAE_0,
 	[3068 ... 3071] = XLP_MSG_HANDLE_INVALID,
@@ -204,7 +204,7 @@ static uint16_t vc_to_handle_map[MAX_VC] = {
 	[3353 ... 3368] = XLP_MSG_HANDLE_CRYPTO,
 	[3369 ... 3376] = XLP_MSG_HANDLE_CMP,
 	[3377 ... 3455] = XLP_MSG_HANDLE_INVALID,
-	[3456 ... 3463] = XLP_MSG_HANDLE_NAE_0,
+	[3456 ... 3463] = XLP_MSG_HANDLE_POE_0,
 	[3464 ... 3547] = XLP_MSG_HANDLE_INVALID,
 	[3548 ... 4091] = XLP_MSG_HANDLE_NAE_0,
 	[4092 ... 4095] = XLP_MSG_HANDLE_INVALID,
@@ -237,7 +237,7 @@ static uint16_t xlp3xx_vc_to_handle_map[MAX_VC] = {
 	[276 ... 279] = XLP_MSG_HANDLE_CRYPTO,
 	[280 ... 288] = XLP_MSG_HANDLE_SRIO,
 	[289 ... 383] = XLP_MSG_HANDLE_INVALID,
-	[384 ... 391] = XLP_MSG_HANDLE_NAE_0,
+	[384 ... 391] = XLP_MSG_HANDLE_POE_0,
 	[392 ... 431] = XLP_MSG_HANDLE_INVALID,
 	[432 ... 511] = XLP_MSG_HANDLE_NAE_0,
 	[512 ... 4095]= XLP_MSG_HANDLE_INVALID
@@ -261,7 +261,7 @@ static uint16_t xlp2xx_vc_to_handle_map[MAX_VC] = {
 	[273 ... 275] = XLP_MSG_HANDLE_INVALID,
 	[276 ... 276] = XLP_MSG_HANDLE_CRYPTO,
 	[277 ... 383] = XLP_MSG_HANDLE_INVALID,
-	[384 ... 391] = XLP_MSG_HANDLE_NAE_0,
+	[384 ... 391] = XLP_MSG_HANDLE_POE_0,
 	[392 ... 431] = XLP_MSG_HANDLE_INVALID,
 	[432 ... 503] = XLP_MSG_HANDLE_NAE_0,
 	[504 ... 4095]= XLP_MSG_HANDLE_INVALID
-- 
1.7.1

