From 93854557f81c894401e420bff172fc48b6534c3e Mon Sep 17 00:00:00 2001
From: Alok Agrawat <alok@broadcom.com>
Date: Thu, 20 Jun 2013 15:25:38 -0700
Subject: CDE: Add mmap call incdetest.

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/drivers/misc/netlogic/nlm_cde/cdetest.c b/drivers/misc/netlogic/nlm_cde/cdetest.c
index f4fb307..7fdf917 100644
--- a/drivers/misc/netlogic/nlm_cde/cdetest.c
+++ b/drivers/misc/netlogic/nlm_cde/cdetest.c
@@ -134,8 +134,8 @@ int main(int argc, char *argv[])
 //        test();
   //      return 0;
 printf("\\\\\\\\\\\\\\\\\\\\\\\\\\\\\n CDE TEST APPLICATION >>>>>>>>>>>>>>>>>\n");
-#if 0
-	size = ((1024 + 65) * 1024 ) *2;
+#if 1
+	size = ((1024 + 65) * 1024 ) *4;
 printf("\n size = %d\n", size);
        cde_mmap_mem = mmap (0, size, PROT_READ | PROT_WRITE, MAP_SHARED, dfd, 0);
        if(cde_mmap_mem)
-- 
1.7.1

