From a29b2f4ff26921437e3a603f17aa1651c5edd154 Mon Sep 17 00:00:00 2001
From: P. Sadik <psadik@broadcom.com>
Date: Wed, 10 Jul 2013 17:19:55 +0530
Subject: xlp9xx: add vc map structure for xlp9xx.

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/drivers/misc/netlogic/soc_interface/on_chip.c b/drivers/misc/netlogic/soc_interface/on_chip.c
index 94dcf09..4df329e 100644
--- a/drivers/misc/netlogic/soc_interface/on_chip.c
+++ b/drivers/misc/netlogic/soc_interface/on_chip.c
@@ -104,41 +104,6 @@ static int xlp_napi_vc_count = 0;
 static int xlp_fmn_init_done = 0;
 extern unsigned int xlp_napi_vc_mask;
 
-/* This structure will work for CDE. It will work for NAE also but it is not correfcft.
-   Need to fill all valid station numbers*/
-static uint16_t xlp9xx_vc_to_handle_map[MAX_VC] = {
-	[0 ... 15] = XLP_MSG_HANDLE_CPU0,
-	[16 ... 31] = XLP_MSG_HANDLE_CPU1,
-	[32 ... 47] = XLP_MSG_HANDLE_CPU2,
-	[48 ... 63] = XLP_MSG_HANDLE_CPU3,
-	[64 ... 79] = XLP_MSG_HANDLE_CPU4,
-	[80 ... 95] = XLP_MSG_HANDLE_CPU5,
-	[96 ... 111] = XLP_MSG_HANDLE_CPU6,
-	[112 ... 127] = XLP_MSG_HANDLE_CPU7,
-	[128 ... 143] = XLP_MSG_HANDLE_CPU0,
-	[144 ... 159] = XLP_MSG_HANDLE_CPU1,
-	[160 ... 175] = XLP_MSG_HANDLE_CPU2,
-	[176 ... 191] = XLP_MSG_HANDLE_CPU3,
-	[192 ... 207] = XLP_MSG_HANDLE_CPU4,
-	[208 ... 223] = XLP_MSG_HANDLE_CPU5,
-	[224 ... 239] = XLP_MSG_HANDLE_CPU6,
-	[240 ... 255] = XLP_MSG_HANDLE_CPU7,
-	[256 ... 257] = XLP_MSG_HANDLE_PCIE0,
-	[258 ... 259] = XLP_MSG_HANDLE_PCIE1,
-	[260 ... 261] = XLP_MSG_HANDLE_PCIE2,
-	[262 ... 263] = XLP_MSG_HANDLE_PCIE3,
-	[264 ... 267] = XLP_MSG_HANDLE_DTRE,
-	[268 ... 271] = XLP_MSG_HANDLE_GDX,
-	[272 ... 280] = XLP_MSG_HANDLE_RSA_ECC,
-	[281 ... 296] = XLP_MSG_HANDLE_CRYPTO,
-	[400 ... 406] = XLP_MSG_HANDLE_CMP,
-	[305 ... 383] = XLP_MSG_HANDLE_INVALID,
-	[384 ... 391] = XLP_MSG_HANDLE_NAE_0,
-//	[392 ... 475] = XLP_MSG_HANDLE_INVALID,
-	[476 ... 1019] = XLP_MSG_HANDLE_NAE_0,
-	[1020 ... 1023] = XLP_MSG_HANDLE_INVALID,
-};
-
 static uint16_t vc_to_handle_map[MAX_VC] = {
 	[0 ... 15] = XLP_MSG_HANDLE_CPU0,
 	[16 ... 31] = XLP_MSG_HANDLE_CPU1,
@@ -647,6 +612,7 @@ int register_xlp_msgring_handler(int major,
 		       __FUNCTION__, __LINE__, major, XLP_MAX_TX_STNS, action);
 		return ret;
 	}
+
 	/* Check if the message station is valid, if not return error */
 	spin_lock_irqsave(&msgrng_lock, flags);
 
@@ -1099,7 +1065,7 @@ static int __init on_chip_init(void)
 		vc_to_handle = xlp2xx_vc_to_handle_map;
 	else if (is_nlm_xlp3xx())
 		vc_to_handle = xlp3xx_vc_to_handle_map;
-	else if(is_nlm_xlp9xx())
+	else if (is_nlm_xlp9xx())
 		vc_to_handle = xlp9xx_vc_to_handle_map;
 	else
 		vc_to_handle = vc_to_handle_map;
-- 
1.7.1

