From 3c31d825b6635ea3beb09715d4263f897cf75b2b Mon Sep 17 00:00:00 2001
From: Lakshman Garlapati <kumarga@broadcom.com>
Date: Tue, 9 Jul 2013 04:42:10 -0700
Subject: DPI : Changes to bringup XLP9XX

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/drivers/misc/netlogic/netl7driver/netl7driver.c b/drivers/misc/netlogic/netl7driver/netl7driver.c
index 063a8fb..47727c0 100755
--- a/drivers/misc/netlogic/netl7driver/netl7driver.c
+++ b/drivers/misc/netlogic/netl7driver/netl7driver.c
@@ -73,8 +73,10 @@ module_param (verbose, int, S_IRUGO);
 #define NLM_VERB(f, args...) do { if (verbose) printk (KERN_INFO DRV_NAME " "f, ##args); } while (0)
 
 #define NLM_PCI_VENDOR_ID (0x184e)
+#define NLM_XLP9XX_PCI_VENDOR_ID (0x14e4)
 #define NLM_PCI_DEV_ID (0x8)
-#define NLM_STORM_PCI_DEV_ID (0x1019)
+#define NLM_XLP3XX_PCI_DEV_ID (0x1019)
+#define NLM_XLP9XX_PCI_DEV_ID (0x9019)
 
 enum nlm_device_type
 {
@@ -181,7 +183,8 @@ nlm_pci_probe (struct pci_dev *nlm_dev, const struct pci_device_id *nlm_id)
       break;
     case NLM_PCI_DEV_ID + 4:
     case NLM_PCI_DEV_ID + 5:
-    case NLM_STORM_PCI_DEV_ID:
+    case NLM_XLP3XX_PCI_DEV_ID:
+    case NLM_XLP9XX_PCI_DEV_ID:
       type = NLM_DEVICE_MARS3;
       break;
     default:
@@ -256,7 +259,8 @@ static const struct __devinitdata pci_device_id netl7_pci_ids[] =
   {PCI_DEVICE (NLM_PCI_VENDOR_ID, NLM_PCI_DEV_ID + 3)},
   {PCI_DEVICE (NLM_PCI_VENDOR_ID, NLM_PCI_DEV_ID + 4)},
   {PCI_DEVICE (NLM_PCI_VENDOR_ID, NLM_PCI_DEV_ID + 5)},
-  {PCI_DEVICE (NLM_PCI_VENDOR_ID, NLM_STORM_PCI_DEV_ID)},
+  {PCI_DEVICE (NLM_PCI_VENDOR_ID, NLM_XLP3XX_PCI_DEV_ID)},
+  {PCI_DEVICE (NLM_XLP9XX_PCI_VENDOR_ID, NLM_XLP9XX_PCI_DEV_ID)},
   {0,},
 };
 
-- 
1.7.1

