From 722c822161ccc56c2cbc9bfded236331477cdbc0 Mon Sep 17 00:00:00 2001
From: Sreenidhi B R <sreenira@broadcom.com>
Date: Wed, 18 Sep 2013 00:46:42 -0700
Subject: pktmem: add usage of copy_from_user and copy_to_user macros for data protection

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/drivers/misc/netlogic/pkt_pool_mem/pkt_pool_mem.c b/drivers/misc/netlogic/pkt_pool_mem/pkt_pool_mem.c
index 4502e80..bd257a3 100644
--- a/drivers/misc/netlogic/pkt_pool_mem/pkt_pool_mem.c
+++ b/drivers/misc/netlogic/pkt_pool_mem/pkt_pool_mem.c
@@ -173,15 +173,17 @@ static int free_memory (brcm_devmem_ioctl_t * param)
 static long pktmem_ioctl (struct file *fptr, unsigned int cmd, unsigned long dp)
 {
 	int rc;
-	brcm_devmem_ioctl_t *dptr;
+	brcm_devmem_ioctl_t *dptr, data;
 
 #ifdef PKTMEM_DEBUG
 	printk("pktmem: icoctl enter, cmd %d.\n", cmd);
 #endif
 	spin_lock(&ioctl_lock);
-	dptr = (brcm_devmem_ioctl_t *)dp;
+	dptr = &data;
 	rc = 0;
 
+	copy_from_user(dptr, dp, sizeof(brcm_devmem_ioctl_t));
+
 #ifdef PKTMEM_DEBUG
 	printk("pktmem: memory map before\n");
 	print_memory(dptr);
@@ -208,6 +210,9 @@ static long pktmem_ioctl (struct file *fptr, unsigned int cmd, unsigned long dp)
 	print_memory(dptr);
 	printk("pktmem: ioctl returns rc %d.\n", rc);
 #endif
+
+	copy_to_user(dp, dptr, sizeof(brcm_devmem_ioctl_t));
+
 	spin_unlock(&ioctl_lock);
 	return rc;
 }
-- 
1.7.1

