From b4eccef5ca2a8eafe621d6daf24fd2908e7411e5 Mon Sep 17 00:00:00 2001
From: Abhishek Joshi <abjoshi@broadcom.com>
Date: Sun, 27 Oct 2013 19:11:55 +0530
Subject: hev2: Fixed app failure for 64 bit LE mode

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/drivers/misc/netlogic/pkt_pool_mem/pkt_pool_mem.c b/drivers/misc/netlogic/pkt_pool_mem/pkt_pool_mem.c
index bd257a3..c1d88a0 100644
--- a/drivers/misc/netlogic/pkt_pool_mem/pkt_pool_mem.c
+++ b/drivers/misc/netlogic/pkt_pool_mem/pkt_pool_mem.c
@@ -292,8 +292,13 @@ static int __init brcm_xlp_ppm_init (void)
 			mem_pool_data[node].next = NULL;
 			memset(mem_pool_data[node].shr_id, '\0', 64);
 			mem_pool_data[node].in_use = 0;
+#ifdef __MIPSEL__
+			mem_pool_data[node].start = ((((unsigned long long)pktregs[i+2])<<32) | (pktregs[i+1]));
+			mem_pool_data[node].size = ((((unsigned long long)pktregs[i+4])<<32) | (pktregs[i+3]));
+#else
 			mem_pool_data[node].start = ((((unsigned long long)pktregs[i+1])<<32) | (pktregs[i+2]));
 			mem_pool_data[node].size = ((((unsigned long long)pktregs[i+3])<<32) | (pktregs[i+4]));
+#endif
 		}
 	}
 	else
-- 
1.7.1

