From ebd4ebab790897a5c2463ac0185649ed2b1ac0ad Mon Sep 17 00:00:00 2001
From: Hareesh R <hareeshr@broadcom.com>
Date: Tue, 29 Oct 2013 13:48:23 +0530
Subject: openssl:Fix for memory allocation

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/drivers/misc/netlogic/nlmcrypto/cryptodrv.c b/drivers/misc/netlogic/nlmcrypto/cryptodrv.c
index 431753e..faca178 100644
--- a/drivers/misc/netlogic/nlmcrypto/cryptodrv.c
+++ b/drivers/misc/netlogic/nlmcrypto/cryptodrv.c
@@ -444,10 +444,11 @@ static long crypto_driver_ioctl(struct file *filp, unsigned int cmd,
 		case NLM_CRYPTO_GET_CHIP_FEATURE:
 		{
 			uint32_t *ptr = (uint32_t *) arg;
-			int nlm_crypto_chip_features;
+			int chip_type = 1;
 
-			nlm_crypto_chip_features = nlm_hal_get_sae_chip_feature();
-			copy_to_user((ptr), &nlm_crypto_chip_features,sizeof(nlm_crypto_chip_features));
+			if(is_nlm_xlp9xx())
+				chip_type = 2;
+			copy_to_user((ptr), &chip_type,sizeof(chip_type));
 
 			break;
 		}
-- 
1.7.1

