From e23084bb3fda2eafc508cae1a0842d93f364ebab Mon Sep 17 00:00:00 2001
From: Abhishek Joshi <abjoshi@broadcom.com>
Date: Tue, 29 Oct 2013 17:57:37 +0530
Subject: hev2: pktmem support for LE mode

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/drivers/misc/netlogic/pkt_pool_mem/pkt_pool_mem.c b/drivers/misc/netlogic/pkt_pool_mem/pkt_pool_mem.c
index c1d88a0..42a8cf6 100644
--- a/drivers/misc/netlogic/pkt_pool_mem/pkt_pool_mem.c
+++ b/drivers/misc/netlogic/pkt_pool_mem/pkt_pool_mem.c
@@ -14,6 +14,13 @@
 #define PKTMEM_MAJOR 	123
 static DEFINE_SPINLOCK(ioctl_lock);
 
+#ifdef __MIPSEL__
+#define swap32(x) ((uint32_t)(                                                           \
+				(((uint32_t)(x) & (uint32_t)0x000000ffUL) << 24) | \
+				(((uint32_t)(x) & (uint32_t)0x0000ff00UL) <<  8) | \
+				(((uint32_t)(x) & (uint32_t)0x00ff0000UL) >>  8) | \
+				(((uint32_t)(x) & (uint32_t)0xff000000UL) >> 24)))
+#endif
 
 typedef struct memory_pool 
 {
@@ -272,6 +279,9 @@ static int __init brcm_xlp_ppm_init (void)
 	int i, node;
 	void *fdt;
 	uint32_t pktregs[MAX_NODES * 5];
+#ifdef __MIPSEL__
+	int j;
+#endif
 
 	i = register_chrdev(PKTMEM_MAJOR, "brcm_pktmem", &pktmem_ops);
 	if (i < 0)
@@ -293,12 +303,11 @@ static int __init brcm_xlp_ppm_init (void)
 			memset(mem_pool_data[node].shr_id, '\0', 64);
 			mem_pool_data[node].in_use = 0;
 #ifdef __MIPSEL__
-			mem_pool_data[node].start = ((((unsigned long long)pktregs[i+2])<<32) | (pktregs[i+1]));
-			mem_pool_data[node].size = ((((unsigned long long)pktregs[i+4])<<32) | (pktregs[i+3]));
-#else
+			for (j=0; j < 5; j++)
+				pktregs[i+j]=swap32(pktregs[i+j]);
+#endif
 			mem_pool_data[node].start = ((((unsigned long long)pktregs[i+1])<<32) | (pktregs[i+2]));
 			mem_pool_data[node].size = ((((unsigned long long)pktregs[i+3])<<32) | (pktregs[i+4]));
-#endif
 		}
 	}
 	else
-- 
1.7.1

