From 2850102c22376b566d805b4e1619858718461458 Mon Sep 17 00:00:00 2001
From: Anurag <anurag.gopinath@broadcom.com>
Date: Tue, 26 Nov 2013 01:08:39 +0530
Subject: CDE : Add support for ioctl from 64 bit applications

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/drivers/misc/netlogic/nlm_cde/nlm_cde.c b/drivers/misc/netlogic/nlm_cde/nlm_cde.c
index 2710a9b..0fafb58 100644
--- a/drivers/misc/netlogic/nlm_cde/nlm_cde.c
+++ b/drivers/misc/netlogic/nlm_cde/nlm_cde.c
@@ -173,6 +173,7 @@ static const struct file_operations xlp_cde_fops = {
 	.read = xlp_cde_read,
 	.write = xlp_cde_write,
 	.compat_ioctl = xlp_cde_ioctl,
+	.unlocked_ioctl= xlp_cde_ioctl,
 	.mmap = xlp_cde_mmap,
 	.release = xlp_cde_release,
 };
-- 
1.7.1

