From fdfb9d27500e7610b18cf0d6e7c3725cd6390520 Mon Sep 17 00:00:00 2001
From: Pavani Reddy <pavani@broadcom.com>
Date: Fri, 29 Nov 2013 19:12:06 +0530
Subject: soc_interface: Fixed multinode credit initialization.

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/drivers/misc/netlogic/soc_interface/on_chip.c b/drivers/misc/netlogic/soc_interface/on_chip.c
index 8f23eb9..4606002 100644
--- a/drivers/misc/netlogic/soc_interface/on_chip.c
+++ b/drivers/misc/netlogic/soc_interface/on_chip.c
@@ -1442,18 +1442,19 @@ static int __init nlm_fmn_init(void)
         int node, rv;
 	int node_exist[8] = {1, 0, 0, 0, 0, 0, 0, 0};
 	int max_nodes = NLM_MAX_NODES;
+	uint32_t fmn_owner_mask = 0;
 
 	if(is_nlm_xlp9xx())
 		max_nodes = NLM_MAX_CPU_NODE_9XX;
 
         /* Better to do node level fmn initialization */
-        if(get_dom_fmn_node_ownership(fdt, 0)) {
-                for(node = 0; node < max_nodes; node++) {
-                        if(!node_exist[node])
+	fmn_owner_mask = get_dom_fmn_node_ownership(fdt, 0);
+        if(fmn_owner_mask) {
+                for(node = 0; node < max_nodes; node++) 
+		{
+			if(!(fmn_owner_mask & (1<<node)))
                                 continue;
-                        rv = work_on_cpu(node_exist[node] >> 16, fmn_config, (void *)(long)node);
-                        if(rv < 0)
-                                panic("Fmn init failed\n");
+			fmn_config((void *)(long)node);
                 }
         } else
                 printk("Skipping FMN initialization due to lack of ownership\n");
-- 
1.7.1

