From d5ca9e6fb4986d0ed57dfb6e3f3badbd3927f5f3 Mon Sep 17 00:00:00 2001
From: Deepak Jangid <deepak.jangid@broadcom.com>
Date: Tue, 17 Dec 2013 02:57:01 -0800
Subject: soc_interface: Don't acknowledge MSG RING interrupt when interrupt handler returns 1. This is needed for drivers that have their own NAPI handlers.

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/drivers/misc/netlogic/soc_interface/on_chip.c b/drivers/misc/netlogic/soc_interface/on_chip.c
index 4606002..e86e7aa 100644
--- a/drivers/misc/netlogic/soc_interface/on_chip.c
+++ b/drivers/misc/netlogic/soc_interface/on_chip.c
@@ -839,7 +839,11 @@ void nlm_xlp_msgring_int_handler(unsigned int irq, struct pt_regs *regs)
 			for(vc = 0; vc < 4; vc++) {
 				if(!(vcmask & (1<<vc)))
 					continue;
-				xlp_intr_vc_handler(vc);
+				/*Don't acknowledge MSG RING interrupt when 
+				  intr_vc_handler returns 1*/
+				if(xlp_intr_vc_handler(vc) == 1){
+					pop_vc_mask = pop_vc_mask & ~(1<<vc);
+				}
 			}
 		}
 
-- 
1.7.1

