From e537dba4e67ae901481652a936df483c47f7d94d Mon Sep 17 00:00:00 2001
From: Virendra Pathak <vpathak@broadcom.com>
Date: Thu, 19 Dec 2013 07:52:28 +0530
Subject: halx_netio: updated halx_netio module

-  w.r.t directory structure changes in halx
[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/drivers/misc/netlogic/halx_netio/Makefile b/drivers/misc/netlogic/halx_netio/Makefile
index 9828a1e..1573a5f 100644
--- a/drivers/misc/netlogic/halx_netio/Makefile
+++ b/drivers/misc/netlogic/halx_netio/Makefile
@@ -4,7 +4,7 @@ HAL_NET_DIR = $(SDK_ROOT)/libraries/netlib
 
 ccflags-y += -DCONFIG_XLP_FMN_SUPPORT -DNLM_HAL_LINUX_KERNEL \
      -I$(SDK_ROOT)/libraries/fdt/libfdt -I$(HALX_DIR)/include -I$(HAL_NET_DIR)/include \
-     -I$(HALX_DIR)/halx-brcm/include -I$(HALX_DIR)/halx-brcm/arch/xlp
+     -I$(HALX_DIR)/src/include -I$(HALX_DIR)/src/arch/xlp
 
 KBUILD_EXTRA_SYMBOLS:= $(SDK_ROOT)/libraries/Module.symvers
 
-- 
1.7.1

