From c30f90343ca5c96b0f0a2cc6b71cceb2ed321dde Mon Sep 17 00:00:00 2001
From: Debayan Ghosh <debayan.ghosh@broadcom.com>
Date: Thu, 20 Mar 2014 23:58:44 +0530
Subject: HEv2: Added l3 cacheline locking support and demo app

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/drivers/misc/netlogic/pkt_pool_mem/pkt_pool_kern_iface.h b/drivers/misc/netlogic/pkt_pool_mem/pkt_pool_kern_iface.h
index 395ff01..5b869a4 100644
--- a/drivers/misc/netlogic/pkt_pool_mem/pkt_pool_kern_iface.h
+++ b/drivers/misc/netlogic/pkt_pool_mem/pkt_pool_kern_iface.h
@@ -12,6 +12,7 @@ enum pktmem_ops {
 	ALLOC_POOL = 0,
 	FREE_POOL,
 	PRINT_POOL,
+	CACHE_FLUSH
 };
 
 #endif
diff --git a/drivers/misc/netlogic/pkt_pool_mem/pkt_pool_mem.c b/drivers/misc/netlogic/pkt_pool_mem/pkt_pool_mem.c
index 42a8cf6..f138797 100644
--- a/drivers/misc/netlogic/pkt_pool_mem/pkt_pool_mem.c
+++ b/drivers/misc/netlogic/pkt_pool_mem/pkt_pool_mem.c
@@ -5,6 +5,8 @@
 #include <linux/slab.h>
 #include <linux/fs.h>
 #include <linux/mm.h>
+#include <linux/delay.h>
+#include <asm/cacheops.h>
 #include "libfdt.h"
 #include "pkt_pool_kern_iface.h"
 
@@ -13,6 +15,16 @@
 #define MB(x) 		(x*1024*1024)
 #define PKTMEM_MAJOR 	123
 static DEFINE_SPINLOCK(ioctl_lock);
+static DEFINE_SPINLOCK(cache_lock);
+
+#define cache_op(op,addr)                                               \
+        __asm__ __volatile__(                                           \
+        "       .set    push                                    \n"     \
+        "       .set    noreorder                               \n"     \
+        "       cache   %0, %1                                  \n"     \
+        "       .set    pop                                     \n"     \
+        :                                                               \
+        : "i" (op), "R" (*(unsigned char *)(addr)))
 
 #ifdef __MIPSEL__
 #define swap32(x) ((uint32_t)(                                                           \
@@ -177,14 +189,29 @@ static int free_memory (brcm_devmem_ioctl_t * param)
 		return -1;
 }
 
+
 static long pktmem_ioctl (struct file *fptr, unsigned int cmd, unsigned long dp)
 {
+	
 	int rc;
 	brcm_devmem_ioctl_t *dptr, data;
-
+	
 #ifdef PKTMEM_DEBUG
 	printk("pktmem: icoctl enter, cmd %d.\n", cmd);
 #endif
+	/* Handling special cmd for cache flush operation */
+	if(cmd == CACHE_FLUSH) 
+	{
+		spin_lock(&cache_lock);
+		uint64_t paddr = dp;
+		cache_op(Hit_Writeback_Inv_D,paddr);
+		cache_op(Hit_Writeback_Inv_D,paddr+32);
+		udelay(30);
+		cache_op(Hit_Writeback_Inv_SD,paddr);	
+		spin_unlock(&cache_lock);
+		return 0;
+	}
+
 	spin_lock(&ioctl_lock);
 	dptr = &data;
 	rc = 0;
-- 
1.7.1

