From 0ea08aa92f0473017a447d74672171a1fa6cca4b Mon Sep 17 00:00:00 2001
From: Virendra Pathak <vpathak@broadcom.com>
Date: Wed, 26 Mar 2014 18:54:16 +0530
Subject: build: Fix for halx_netio compilation due to change in halx include path change

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/drivers/misc/netlogic/halx_netio/Makefile b/drivers/misc/netlogic/halx_netio/Makefile
index a64d7c3..6b24789 100644
--- a/drivers/misc/netlogic/halx_netio/Makefile
+++ b/drivers/misc/netlogic/halx_netio/Makefile
@@ -4,7 +4,7 @@ HAL_NET_DIR = $(SDK_ROOT)/libraries/netlib
 HAL_ROOT = $(SDK_ROOT)/libraries
 
 ccflags-y += -DCONFIG_XLP_FMN_SUPPORT -DNLM_HAL_LINUX_KERNEL \
-     -I$(HAL_ROOT)/fdt/libfdt -I$(HAL_ROOT)/syslib/include/ -I$(HAL_ROOT)/alelib/ -I$(HAL_ROOT)/fmnlib -I$(HALX_DIR)/include -I$(HAL_NET_DIR)/include \
+     -I$(HAL_ROOT)/fdt/libfdt -I$(HAL_ROOT)/syslib/include/ -I$(HAL_ROOT)/alelib/ -I$(HAL_ROOT)/fmnlib -I$(HALX_DIR)/usr/include -I$(HAL_NET_DIR)/include \
      -I$(HALX_DIR)/src/include -I$(HALX_DIR)/src/arch/xlp
 
 KBUILD_EXTRA_SYMBOLS:= $(SDK_ROOT)/libraries/Module.symvers
-- 
1.7.1

