From 9bcc218e82b3e1088e3ce0fcbf11ecdbdd556270 Mon Sep 17 00:00:00 2001
From: Abhishek Joshi <abjoshi@broadcom.com>
Date: Tue, 11 Feb 2014 22:53:12 +0530
Subject: dtre: added xlp5xx support

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/drivers/misc/netlogic/dtre/nlm_adma.c b/drivers/misc/netlogic/dtre/nlm_adma.c
index 8ca2167..79a0c5d 100644
--- a/drivers/misc/netlogic/dtre/nlm_adma.c
+++ b/drivers/misc/netlogic/dtre/nlm_adma.c
@@ -1488,9 +1488,15 @@ static int __devinit nlm_adma_probe(struct platform_device *pdev)
 			printk("Error: NLM-ADMA unable to register for msgring handler\n");
 			return -1;
 		}
-	}
-	else
-	{
+	} else if (is_nlm_xlp5xx()) {
+		nlm_dtre_min_vc = DTRE_XLP5XX_MIN_VC;
+
+		if (register_xlp_msgring_handler(XLP_MSG_HANDLE_GDX_0, nlm_dtre_msgring_handler, NULL))
+		{
+			printk("Error: NLM-ADMA unable to register for msgring handler\n");
+			return -1;
+		}
+	} else {
 		nlm_dtre_min_vc = DTRE_MIN_VC;
 
 		if (register_xlp_msgring_handler(XLP_MSG_HANDLE_DTRE, nlm_dtre_msgring_handler, NULL)){
@@ -1501,8 +1507,9 @@ static int __devinit nlm_adma_probe(struct platform_device *pdev)
 
 	if (is_nlm_xlp2xx()) {
 		num_channels = DTRE_2XX_NUM_CHANNELS;
-	}
-	else {
+	} else if (is_nlm_xlp5xx()) {
+		num_channels = DTRE_5XX_NUM_CHANNELS;
+	} else {
 		num_channels = DTRE_NUM_CHANNELS;
 	}
 
@@ -1564,13 +1571,20 @@ static int __devinit nlm_adma_probe(struct platform_device *pdev)
 			} else {
 				is_raid = 0;
 			}
-		}
-		else if (is_nlm_xlp2xx()) {
+		} else if (is_nlm_xlp5xx()) {
+			/* xlp5xx:
+			 * channel 0 : RAID
+			 * channel 1: DMA
+			 */
+			if (loop == 0) {
+				is_raid = 1;
+			} else {
+				is_raid = 0;
+			}
+		} else if (is_nlm_xlp2xx()) {
 			/* for xlp2xx, there are TWO channels, both for DMA */
 			is_raid = 0;
-		}
-		else
-		{
+		} else {
 			/* set RAID capabilities for chan 0 and 
 			   MEM capabilities for chan 1-3 */
 			if (loop == 0) {
@@ -1649,9 +1663,19 @@ static int __devinit nlm_adma_probe(struct platform_device *pdev)
 
 		/* set channel 0 is raid - bit 4 */
 		nlm_hal_write_32bit_reg (base, 0x40, value | (0x10));
-	}
-	else if (!is_nlm_xlp2xx())
-	{
+	} else if (is_nlm_xlp5xx()) {
+		/* xlp5xx: xlp5xx have only one complex
+		 */
+
+		/* for DTRE, B/D/F = 1/5/0 , since 5xx is closely related to xlp9xx,
+		   it is assumed that the bus number is 1, as is the case in xlp9xx
+		*/
+		base = nlm_hal_get_dev_base (0 /*node*/, 1 /*B*/, 5 /*D*/, 0 /*F*/);
+		value = nlm_hal_read_32bit_reg (base, 0x40);
+
+		/* set channel 0 is raid - bit 4 */
+		nlm_hal_write_32bit_reg (base, 0x40, value | (0x10));
+	} else if (!is_nlm_xlp2xx()) {
 		/* for DTRE, B/D/F = 0/5/0 */
 		base = nlm_hal_get_dev_base (0 /*node*/, 0 /*B*/, 5 /*D*/, 0 /*F*/);
 		value = nlm_hal_read_32bit_reg (base, 0x40);
diff --git a/drivers/misc/netlogic/dtre/nlm_adma.h b/drivers/misc/netlogic/dtre/nlm_adma.h
index 77e73e2..e039fc3 100644
--- a/drivers/misc/netlogic/dtre/nlm_adma.h
+++ b/drivers/misc/netlogic/dtre/nlm_adma.h
@@ -56,8 +56,11 @@
 #define DTRE_MAX_VC		267
 #define DTRE_NUM_CHANNELS	4
 #define DTRE_2XX_NUM_CHANNELS	2
+#define DTRE_5XX_NUM_CHANNELS	2
 #define DTRE_9XX_MIN_VC		392
 #define DTRE_9XX_MAX_VC		395
+#define DTRE_XLP5XX_MIN_VC	392
+#define DTRE_XLP5XX_MAX_VC	393
 
 
 
-- 
1.7.1

