From dc51f94ba80cbe7bf1ee62d85d88fd7f739a6f92 Mon Sep 17 00:00:00 2001
From: Anurag <anurag.gopinath@broadcom.com>
Date: Tue, 11 Mar 2014 08:11:43 +0530
Subject: XLP5XX : Add SRIO support

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/drivers/misc/netlogic/soc_interface/on_chip.c b/drivers/misc/netlogic/soc_interface/on_chip.c
index 3c3f51b..19933d3 100644
--- a/drivers/misc/netlogic/soc_interface/on_chip.c
+++ b/drivers/misc/netlogic/soc_interface/on_chip.c
@@ -705,8 +705,8 @@ static uint16_t xlp5xx_vc_to_handle_map[MAX_VC_5XX] = {
 	[452 ... 459] = XLP_MSG_HANDLE_INVALID,
 	[460 ... 463] = XLP_MSG_HANDLE_REGX,
 	[464 ... 475] = XLP_MSG_HANDLE_INVALID,
-	[476 ... 479] = XLP_MSG_HANDLE_KBP,
-	[480 ... 511] = XLP_MSG_HANDLE_INVALID,
+	[476 ... 485] = XLP_MSG_HANDLE_SRIO,
+	[486 ... 511] = XLP_MSG_HANDLE_INVALID,
 	[512 ... 519] = XLP_MSG_HANDLE_POE_0,
 	[520 ... 599] = XLP_MSG_HANDLE_NAE_0,
 	[600 ... 729] = XLP_MSG_HANDLE_INVALID,
-- 
1.7.1

