From 91f4d74f856c96ce16aac81ddf9623a80db0a96a Mon Sep 17 00:00:00 2001
From: Anurag <anurag.gopinath@broadcom.com>
Date: Wed, 30 Apr 2014 19:24:23 +0530
Subject: Fix : Compile issue from XLP5XX merge

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/drivers/misc/netlogic/nlm_cde/nlm_cde.c b/drivers/misc/netlogic/nlm_cde/nlm_cde.c
index 5e2541c..866a00b 100644
--- a/drivers/misc/netlogic/nlm_cde/nlm_cde.c
+++ b/drivers/misc/netlogic/nlm_cde/nlm_cde.c
@@ -1165,9 +1165,11 @@ static int __init xlp_cde_init(void)
 	if(is_nlm_xlp2xx()) {
 		is_2xx = 1;
 		num_cde_engine = NUM_CDE_ENGINE_XLP2XX;
-	}else if (is_nlm_xlp9xx()) {
+	}
+        else if (is_nlm_xlp9xx()) {
 		is_9xx = 1;
 		num_cde_engine = NUM_CDE_ENGINE_XLP9XX;
+        }
 	else if (is_nlm_xlp5xx())
 		num_cde_engine = NUM_CDE_ENGINE_XLP5XX;
 	else if (cpu_is_xlp9xx())
-- 
1.7.1

