From 2ec76f86006c16fabf0943b684e06f6be5409f3e Mon Sep 17 00:00:00 2001
From: Tanmay Jagdale <tanmayj@broadcom.com>
Date: Fri, 1 Aug 2014 16:41:49 +0530
Subject: fdtconf: fixed ucore code location in dtb

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/drivers/misc/netlogic/fdtconf/fdtconf.c b/drivers/misc/netlogic/fdtconf/fdtconf.c
index c76c3c5..326b6f5 100644
--- a/drivers/misc/netlogic/fdtconf/fdtconf.c
+++ b/drivers/misc/netlogic/fdtconf/fdtconf.c
@@ -51,7 +51,7 @@
 /* Path to ucore code in device tree */
 #define	UCORE_0		"/soc/net@node-0/nae-0/ucore/src@1"
 #define	UCORE_1		"/soc/net@node-0/nae-1/ucore/src@1"
-#define	UCORE		"/soc/nae@node-0/nae/ucore/src@1"
+#define	UCORE		"/soc/nae@node-0/ucore/src@1"
 
 void *fdt;
 EXPORT_SYMBOL(fdt);
-- 
1.7.1

