From 5f0f9274be0e33ea34854f6292f22be9e0d8a635 Mon Sep 17 00:00:00 2001
From: Pranav Bhalerao <pranavb@broadcom.com>
Date: Wed, 1 Oct 2014 16:18:08 +0530
Subject: crypto: XLP9XX B0 Payload offset changes

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/drivers/misc/netlogic/nlmcrypto/cryptodrv.c b/drivers/misc/netlogic/nlmcrypto/cryptodrv.c
index 83e911f..e21dcd4 100644
--- a/drivers/misc/netlogic/nlmcrypto/cryptodrv.c
+++ b/drivers/misc/netlogic/nlmcrypto/cryptodrv.c
@@ -461,7 +461,9 @@ static long crypto_driver_ioctl(struct file *filp, unsigned int cmd,
 				chip_type = 2;
 			else if(is_nlm_xlp2xx())
 				chip_type = 5; //both 0 and 2nd bit set
-
+                          
+                        if (is_nlm_xlp9xx_bx())
+                                chip_type = 10; /* both 1st & 3rd bit set */
 			copy_to_user((ptr), &chip_type,sizeof(chip_type));
 
 			break;
-- 
1.7.1

