From 3fa284cbebe7b71944daf2f32a5fc81e02b4bfbc Mon Sep 17 00:00:00 2001
From: Kancherla Tejaswi <ktejaswi@broadcom.com>
Date: Mon, 24 Nov 2014 15:49:46 +0530
Subject: cleanup: removed hard coded values for node config

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/drivers/misc/netlogic/soc_interface/on_chip.c b/drivers/misc/netlogic/soc_interface/on_chip.c
index c7647c4..fab918f 100644
--- a/drivers/misc/netlogic/soc_interface/on_chip.c
+++ b/drivers/misc/netlogic/soc_interface/on_chip.c
@@ -1625,7 +1625,7 @@ static int __init on_chip_init(void)
 	} else {
 		parse_fmn_config();
 	}
-
+	nlm_hal_init();
 	parse_fdt_sae_vc_config();
 
 	/* Set netlogic_io_base to the run time value */
@@ -1635,8 +1635,6 @@ static int __init on_chip_init(void)
 	msgring_registered.value = 0;
 #endif
 
-	nlm_hal_init();
-
 	for (i = 0; i < NR_CPUS; i++)
 		for (j = 0; j < NLM_MAX_COUNTERS; j++)
 			atomic_set(&nlm_common_counters[i][j], 0);
-- 
1.7.1

