From d43d33785c321b58773fc86b833c408be8925cdc Mon Sep 17 00:00:00 2001
From: Hareesh R <hareeshr@broadcom.com>
Date: Wed, 26 Nov 2014 19:15:07 +0530
Subject: cmem : set_pet_at replaced with set_pte

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/drivers/misc/netlogic/cmem/cmem.c b/drivers/misc/netlogic/cmem/cmem.c
index 02a6f80..e9943e5 100644
--- a/drivers/misc/netlogic/cmem/cmem.c
+++ b/drivers/misc/netlogic/cmem/cmem.c
@@ -266,7 +266,8 @@ static int remap_pte_fn(pte_t *pte, struct page *pmd_page,
 	pval = pte_mkspecial(pfn_pte(pfn, prot));
 	
 	Message("%s addr %lx phys %llx\n", __FUNCTION__, addr, pfn << PAGE_SHIFT);
-	set_pte_at(current->mm, addr, pte, pval);
+	//set_pte_at(current->mm, addr, pte, pval);
+	set_pte(pte, pval);
 	args[0]++;
         return 0;
 }
@@ -275,7 +276,8 @@ static int unmap_pte_fn(pte_t *pte, struct page *pmd_page,
                         unsigned long addr, void *data)
 {
 	Message("%s addr %lx\n", __FUNCTION__, addr);
-	set_pte_at(current->mm, addr, pte, __pte(0));
+	//set_pte_at(current->mm, addr, pte, __pte(0));
+	set_pte(pte, __pte(0));
 	return 0;
 }
 
-- 
1.7.1

