From a1dfae88d5f041c4f89d8e1853ea2793edd0a415 Mon Sep 17 00:00:00 2001
From: henry shao <hshao@netlogicmicro.com>
Date: Wed, 30 Jun 2010 16:28:33 -0700
Subject: NAE: fix warning message

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/drivers/net/ethernet/broadcom/nae/init_nae.c b/drivers/net/ethernet/broadcom/nae/init_nae.c
index 18cb5d3..2d0de0a 100644
--- a/drivers/net/ethernet/broadcom/nae/init_nae.c
+++ b/drivers/net/ethernet/broadcom/nae/init_nae.c
@@ -51,7 +51,7 @@ int init_gmac(unsigned int  inf)
 
     // set softreset
     write_gmac_reg( inf , INF_MAC_CONFIG1, INF_SOFTRESET(1)|
-	                                       /*  INF_LOOP_BACK(1)|*/
+	                                        /* INF_LOOP_BACK(1)|*/
                                                  INF_RX_ENABLE(1)|
                                                  INF_TX_ENABLE(1) );
 
@@ -174,7 +174,7 @@ int init_ucore(uint32_t ucore_mask, int if_num)
      * */
 
 
-init_ingress(void)
+void init_ingress(void)
 {
     unsigned int rx_cfg = 0;
 
@@ -196,7 +196,7 @@ init_ingress(void)
                                      NAE_RX_ENABLE|
                                      NAE_MAX_MESSAGE_SIZE(0x3)|
                                      NAE_RX_STATUS_MASK(0x3f)|
-                                     NAE_FRINDESCCLSIZE(4)
+                                     NAE_FRINDESCCLSIZE(24)
                                     );
 
 }
@@ -213,7 +213,7 @@ init_ingress(void)
   * */
 
 
-init_egress(void)
+void init_egress(void)
 {
     uint32_t tx_cfg =  nlm_hal_read_nae_reg(TX_CONFIG);
 
-- 
1.7.1

