From 7dbc218f1e7829de493aa8d3adb0cb56a562c573 Mon Sep 17 00:00:00 2001
From: henry shao <hshao@netlogicmicro.com>
Date: Wed, 30 Jun 2010 16:29:40 -0700
Subject: NAE: fix function prototype does not match problem

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/drivers/net/ethernet/broadcom/nae/net_common.h b/drivers/net/ethernet/broadcom/nae/net_common.h
index 5706de7..d502b3b 100644
--- a/drivers/net/ethernet/broadcom/nae/net_common.h
+++ b/drivers/net/ethernet/broadcom/nae/net_common.h
@@ -29,8 +29,8 @@ typedef struct fmn_credit_struct {
 extern int init_gmac(unsigned int inf);
 extern int init_tx_if_credit( /*uint32_t*/__u32 credit_val, unsigned int if_bmask);
 extern int init_ucore(uint32_t ucore_mask, int if_num);
-extern int init_ingress(void);
-extern int init_egress(void);
+extern void init_ingress(void);
+extern void init_egress(void);
 extern int fmn_init(const fmn_credit_type *credit);
 extern void *xlp_init_buffer( size_t size,
                    size_t pbase ,
-- 
1.7.1

