From e72b867e565d83435812e57c2aaf4819e84c98ea Mon Sep 17 00:00:00 2001
From: henry shao <hshao@netlogicmicro.com>
Date: Wed, 7 Jul 2010 18:15:25 -0700
Subject: NAE: clean up for mac_refill_frin_desc() function parameter and some extra cleanup

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/drivers/net/ethernet/broadcom/nae/init_fmn.c b/drivers/net/ethernet/broadcom/nae/init_fmn.c
index a1025c2..4a2361f 100644
--- a/drivers/net/ethernet/broadcom/nae/init_fmn.c
+++ b/drivers/net/ethernet/broadcom/nae/init_fmn.c
@@ -34,11 +34,6 @@ int fmn_init(const fmn_credit_type *credit)
 
          }
 
-/*
-        log_dbg("config sqid %d dqid %d with credit %d\n", credit[i].s_qid,
-                                                              credit[i].d_qid,
-                                                              credit[i].credit);
-*/
         if(credit[i].flag & SET_UP_QUEUE )
 	   nlm_hal_write_outq_config(credit[i].d_qid, q_cfg);
 
@@ -61,20 +56,15 @@ void *init_nae_free_pool(int num_queue, unsigned char *pktmem , int num_bytes, i
     int q = 0, i = 0;
     int  code = 0;
 
-//	msgrng_access_enable();
     for( q = 0 ; q < num_queue ; q++)
     {
 
-//             log_dbg("%s queue num %d %d\n", __FUNCTION__, nae_rx_qid, q);
 	   for( i= 0; i < num_desc ; i ++ )
         {
 
-//             msg0.entry[0] = VIRT_TO_PHYS((unsigned long)buf &0xffffffffc0ULL);
              msg0.entry[0] = (unsigned long long)buf &0xffffffffc0ULL;
-//             log_dbg("%s free desc %llx\n", __FUNCTION__, msg0.entry[0]);
              msg_print(1, 0, nae_rx_qid, &msg0);
 
-//	xlp_message_send(nae_rx_qid, 1, code, &msg0);
 #if 1
              if (nlm_hal_send_msg1( nae_rx_qid,
                                     code,
@@ -86,10 +76,8 @@ void *init_nae_free_pool(int num_queue, unsigned char *pktmem , int num_bytes, i
 #endif
                 buf += num_bytes;
         }
-//        getchar();
         nae_rx_qid ++;
     }
-//	msgrng_access_disable();
 
 return (void *)buf;
 }
diff --git a/drivers/net/ethernet/broadcom/nae/xlp_nae.c b/drivers/net/ethernet/broadcom/nae/xlp_nae.c
index c304d98..4862f9d 100755
--- a/drivers/net/ethernet/broadcom/nae/xlp_nae.c
+++ b/drivers/net/ethernet/broadcom/nae/xlp_nae.c
@@ -51,7 +51,6 @@
 #include <asm/uaccess.h>
 
 #include <asm/netlogic/msgring.h>
-#include <asm/netlogic/xlr_mac.h>
 
 #include <asm/netlogic/hal/nlm_hal_fmn.h>
 #include <asm/netlogic/hal/nlm_hal_nae.h>
@@ -160,12 +159,12 @@ static struct net_device_stats *nlm_xlp_mac_get_stats(struct net_device *dev);
 static void  nlm_xlp_napi_poll(struct napi_struct *napi, int* budget);
 
 static struct net_device *dev_mac[MAX_GMAC_PORT];
-struct net_device *dev_mac_type[MAX_NET_TYPES][MAX_GMAC_PORT];
+struct net_device *dev_mac_type[MAX_XLP_NET_TYPES][MAX_GMAC_PORT];
 
 
 extern struct proc_dir_entry *nlm_root_proc;
 static struct tasklet_struct mac_refill_task[MAX_GMAC_PORT];
-static void mac_refill_frin_desc(struct net_device *dev);
+static void mac_refill_frin_desc(unsigned long dev);
 
 extern void nlm_xlp_mac_set_enable(struct dev_data *priv, int flag);
 
@@ -277,14 +276,19 @@ static inline void nlm_xlp_free_skb(struct xlp_msg *msg)
  * @dev -  this is per device based function
  *
  **********************************************************************/
-static void mac_refill_frin_desc(struct net_device *dev)
+static void mac_refill_frin_desc(unsigned long dev)
 {
-	struct dev_data* priv = netdev_priv(dev);
-        int ret = 0, mflags = 0, i, code,limit;
+	struct dev_data* priv;
+	struct net_device *ndev;
+        int ret, mflags, i, code,limit;
         struct xlp_msg msg;
 	struct sk_buff * skb;
 	uint64_t *idx_ptr;
 
+	ndev = (struct net_device *) dev;
+	priv = netdev_priv(ndev);
+	ret = 0;
+
 	limit = priv->inited ? MIN_FRIN_DESC_THRESHD*4: MIN_FRIN_DESC_THRESHD*4*10;
 
 	for(i=1; i <= limit; i++)
@@ -296,7 +300,7 @@ static void mac_refill_frin_desc(struct net_device *dev)
 			return;
 		}
 
-		skb->dev = dev;
+		skb->dev = ndev;
 
 		/* Send the free Rx desc to the MAC */
 		mac_put_skb_back_ptr(skb);
@@ -419,7 +423,6 @@ static void nlm_xlp_nae_init(void)
 	spin_lock_init(&nlm_xlp_nae_lock);
 
 	//initial base address for nae, gmac, fmn ...
-	//nlm_hal_init();
 
 	spin_lock_irq(&nlm_xlp_nae_lock);
 
@@ -1024,7 +1027,7 @@ static void nlm_xlp_nae_msgring_handler(uint32_t vc, uint32_t src_id,
 					uint64_t msg0, uint64_t msg1,
 					uint64_t msg2, uint64_t msg3, void* data)
 {
-        struct net_device *ndev;
+        struct net_device *pdev;
         struct dev_data *priv;
 	unsigned int len, port = 0, src, cpu, context;
         unsigned char* buf;
@@ -1044,17 +1047,17 @@ static void nlm_xlp_nae_msgring_handler(uint32_t vc, uint32_t src_id,
 
                 if(addr && (len==0))
                 {
-			ndev = (struct net_device*)dev_mac[port];
-			if(!ndev)
+			pdev = (struct net_device*)dev_mac[port];
+			if(!pdev)
 				return;
-			priv = netdev_priv(ndev);
+			priv = netdev_priv(pdev);
 
                         skb = (struct sk_buff *)bus_to_virt(addr);
                         if(skb)
                         {
                                 priv->stats.rx_packets++;
                                 if(priv->stats.rx_packets % MIN_FRIN_DESC_THRESHD == 15){
-                                       netif_tx_wake_all_queues(ndev);
+                                       netif_tx_wake_all_queues(pdev);
                                 }
                                 dev_kfree_skb_any(skb);
                         }
@@ -1067,7 +1070,7 @@ static void nlm_xlp_nae_msgring_handler(uint32_t vc, uint32_t src_id,
                 else if(addr == 0)
                 {
                         //case tx queue is stopped
-                        netif_tx_wake_all_queues(ndev);
+                        netif_tx_wake_all_queues(pdev);
                 }
 	}
 	else if(vc == 0)
@@ -1084,10 +1087,10 @@ static void nlm_xlp_nae_msgring_handler(uint32_t vc, uint32_t src_id,
 		{
 			port = 0;
 		}
-		ndev = (struct net_device*)dev_mac[port];
-		if(!ndev)
+		pdev = (struct net_device*)dev_mac[port];
+		if(!pdev)
 			return;
-		priv = netdev_priv(ndev);
+		priv = netdev_priv(pdev);
 
 
 		if(!len || addr == 0)
diff --git a/drivers/net/ethernet/broadcom/nae/xlp_nae.h b/drivers/net/ethernet/broadcom/nae/xlp_nae.h
index b4c18e5..f85af9a 100644
--- a/drivers/net/ethernet/broadcom/nae/xlp_nae.h
+++ b/drivers/net/ethernet/broadcom/nae/xlp_nae.h
@@ -1,7 +1,15 @@
 #ifndef _XLP_NAE_H
 #define _XLP_NAE_H
 
-#define MAX_CPUS	8
+#define MAX_CPUS		8
+#define MAC_MAX_FRAME_SIZE      1600
+#define MAC_SKB_BACK_PTR_SIZE   SMP_CACHE_BYTES
+
+
+#define MAC_PREPAD		0
+#define BYTE_OFFSET             2
+#define NLM_RX_BUF_SIZE (MAC_MAX_FRAME_SIZE+BYTE_OFFSET+MAC_PREPAD+MAC_SKB_BACK_PTR_SIZE+SMP_CACHE_BYTES)
+#define MAC_CRC_LEN             4
 
 struct cpu_stat {
         unsigned long tx_packets;
@@ -10,6 +18,9 @@ struct cpu_stat {
         unsigned long interrupts;
 };
 
+
+typedef enum xlp_net_types { TYPE_XLP_GMAC = 0, TYPE_XLP_XGMAC, TYPE_XLP_XAUI, TYPE_XLP_INTERLAKEN, MAX_XLP_NET_TYPES }xlp_interface_t;
+
 typedef enum { xlp_mac_speed_10, xlp_mac_speed_100,
                xlp_mac_speed_1000, xlp_mac_speed_rsvd
 } xlp_mac_speed_t;
-- 
1.7.1

