From 8eed35798ec09612ce3ee22405f6dd6df34d1037 Mon Sep 17 00:00:00 2001
From: henry shao <hshao@netlogicmicro.com>
Date: Wed, 1 Sep 2010 17:06:33 -0700
Subject: NAE: clean up debug code on xlp_nae

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/drivers/net/ethernet/broadcom/nae/xlp_nae.c b/drivers/net/ethernet/broadcom/nae/xlp_nae.c
index cf55523..0a1bae2 100755
--- a/drivers/net/ethernet/broadcom/nae/xlp_nae.c
+++ b/drivers/net/ethernet/broadcom/nae/xlp_nae.c
@@ -280,7 +280,6 @@ static void mac_refill_frin_desc(unsigned long dev)
 
 	/* TODO ALERT: priv->num_desc access needs to atomic */
 	limit = MAX_NUM_DESC - priv->num_desc;
-	printk("[%s]: sending %d desc\n", __func__, limit);
 
 	for(i=1; i <= limit; i++)
 	{
@@ -332,14 +331,12 @@ retry_send:
                                 printk("TX message Q fulll\n");
 			}
 
-			printk("[%s]: retrying...\n", __func__);
 			goto retry_send;
                 }
 
 		msgrng_access_disable(mflags);
 		priv->num_desc += 1;
 	}
-	printk("[%s]: priv->num_desc = %d\n", __func__, priv->num_desc);
 
         return;
 }
-- 
1.7.1

