From dde127d03afbb2cac87c78d6b8bded975aca196e Mon Sep 17 00:00:00 2001
From: Venu Vadapalli <vvadapalli@netlogicmicro.com>
Date: Sun, 5 Sep 2010 12:47:24 -0700
Subject: NAE: Fix NAE txc handling again to reintroduce parsing of context field

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/drivers/net/ethernet/broadcom/nae/xlp_nae.c b/drivers/net/ethernet/broadcom/nae/xlp_nae.c
index ad9794e..4be35d0 100755
--- a/drivers/net/ethernet/broadcom/nae/xlp_nae.c
+++ b/drivers/net/ethernet/broadcom/nae/xlp_nae.c
@@ -896,20 +896,25 @@ static void nlm_xlp_nae_msgring_handler(uint32_t vc, uint32_t src_id,
 
 	if( vc == 1 && size == 1)
 	{
-		/* Transmit Complete */
-		addr = msg0 & 0xfffffffff0ULL;
-		port = msg0 & 0x0f;
-		len = (msg0 >> 40) & 0x3fff;
+		/* Process Transmit Complete */
+		addr = msg0 & 0xffffffffffULL;
+
+		/* context is XLP_SGMII_RCV_CONTEXT_NUM + three bit vlan type
+		 * or vlan priority
+		 */
+		context = (msg0 >> 40) & 0x3fff;
+		port = context / XLP_SGMII_RCV_CONTEXT_NUM;
 
 		/* update dev and port to be accurate */
-		if(port < 0 || port >= MAX_GMAC_PORT)
-		{
-			port = 0;
+		if(port >= MAX_GMAC_PORT) {
+			printk("[%s]: [txc] Bad port %d (context=%d)\n",
+			       __func__, port, context);
 		}
 
 		pdev = (struct net_device*)dev_mac[port];
 		if(!pdev) {
-			printk("[%s]: [txc] wrong port=%d? pdev = NULL!\n", __func__, port);
+			printk("[%s]: [txc] wrong port=%d? pdev = NULL!\n",
+			       __func__, port);
 			return;
 		}
 		priv = netdev_priv(pdev);
-- 
1.7.1

