From cea44ac8023d819b735158aff8c6272aa7e20447 Mon Sep 17 00:00:00 2001
From: Venu Vadapalli <vvadapalli@netlogicmicro.com>
Date: Tue, 7 Sep 2010 17:42:30 -0700
Subject: Use new HAL naecfg_hack

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/drivers/net/ethernet/broadcom/nae/xlp_nae.c b/drivers/net/ethernet/broadcom/nae/xlp_nae.c
index 4be35d0..7078d5e 100755
--- a/drivers/net/ethernet/broadcom/nae/xlp_nae.c
+++ b/drivers/net/ethernet/broadcom/nae/xlp_nae.c
@@ -102,9 +102,12 @@ module_param(debug, int, 0);
 extern int hwemul;
 module_param(hwemul, int, 0);
 
-extern int  xlp_sgmii_ports;
+extern int xlp_sgmii_ports;
 module_param(xlp_sgmii_ports, int, 0);
 
+extern int naecfg_hack;
+module_param(naecfg_hack, int, 0);
+
 unsigned char eth_hw_addr[18][6] = {
 	{0x00,0x01,0x02,0x03,0x04,0x05},
 	{0x00,0x01,0x02,0x03,0x04,0x06},
@@ -436,7 +439,8 @@ static void nlm_xlp_nae_init(void)
 	uint32_t val;
 
 	printk("======= Module Parameters =========\n");
-	printk("debug = %d, hwemul=%d, xlp_sgmii_ports=%d\n", debug, hwemul, xlp_sgmii_ports);
+	printk("debug = %d, hwemul=%d, xlp_sgmii_ports=%d naecfg_hack=%d\n",
+	       debug, hwemul, xlp_sgmii_ports, naecfg_hack);
 
 	if (hwemul) {
 		unsigned long mflags = 0;
-- 
1.7.1

