From 47dc05b7c6a8f904152190bff87eb49966af990c Mon Sep 17 00:00:00 2001
From: Venu Vadapalli <vvadapalli@netlogicmicro.com>
Date: Sat, 11 Sep 2010 12:46:44 -0700
Subject: Use new soc node in FDT file, check FDT header for sanity

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/drivers/net/ethernet/broadcom/nae/init_nae.c b/drivers/net/ethernet/broadcom/nae/init_nae.c
index 6737a13..dc1a614 100644
--- a/drivers/net/ethernet/broadcom/nae/init_nae.c
+++ b/drivers/net/ethernet/broadcom/nae/init_nae.c
@@ -19,7 +19,7 @@ static void parse_ucore_config(void)
 
 	uc_mask = num_opcodes = ucore_cfg = 0;
 
-	node = finddevice("/chosen/nae-cfg/ucore/src@1");
+	node = finddevice("/soc/nae-cfg/ucore/src@1");
 	if(!node) {
 		printk("[%s] Unable to parse ucore configuration!\n", __func__);
 		return;
@@ -91,7 +91,7 @@ static void parse_fdt_nae_config(void)
 	/* Parse Nae Config */
 	start_port = num_nae_regs = num_intf_regs = 0;
 
-	node = finddevice("/chosen/nae-cfg");
+	node = finddevice("/soc/nae-cfg");
 	if(!node) {
 		printk("[%s] Unable to parse nae configuration!\n", __func__);
 		return;
@@ -125,7 +125,7 @@ static void parse_fdt_nae_config(void)
 	{
 		uint32_t *nae_regs = 0, *intf_regs = 0;
 
-		sprintf(domstr, "/chosen/nae-cfg/port@%d", i);
+		sprintf(domstr, "/soc/nae-cfg/port@%d", i);
 
 		subnode = finddevice(domstr);
 		if(!subnode) continue;
@@ -177,13 +177,17 @@ static void parse_fdt_nae_config(void)
 
 void initialize_nae(void)
 {
+	if (check_header()) {
+		printk("Sanity check on FDT blob failed! Aborting\n");
+		return;
+	}
 	printk("Configuring ucore...\n");
 	parse_ucore_config();
 
 	printk("Configuring NAE...\n");
 	parse_fdt_nae_config();
 
-	printk("Configuring NAE...\n");
+	printk("Configuring CPU-NAE...\n");
 	parse_fdt_cpu_config();
 
 	printk("NAE configuration done!\n");
-- 
1.7.1

