From 907bcfd5b61ffe7e4b1a94a25388ffc980ac972c Mon Sep 17 00:00:00 2001
From: Venu Vadapalli <vvadapalli@netlogicmicro.com>
Date: Sat, 2 Oct 2010 16:14:58 -0700
Subject: Added gitignore files

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/drivers/net/ethernet/broadcom/nae/.gitignore b/drivers/net/ethernet/broadcom/nae/.gitignore
index 5027e4e..2d6f993 100644
--- a/drivers/net/ethernet/broadcom/nae/.gitignore
+++ b/drivers/net/ethernet/broadcom/nae/.gitignore
@@ -4,4 +4,4 @@ TAGS
 modules.order
 *.ko
 *.mod.c
-
+.tmp_versions
-- 
1.7.1

