From b77002b490c42b7e2a6e892fc748be286f7e63b1 Mon Sep 17 00:00:00 2001
From: Venu Vadapalli <vvadapalli@netlogicmicro.com>
Date: Wed, 29 Sep 2010 14:56:12 -0700
Subject: Initial set of changes for bringup

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/drivers/net/ethernet/broadcom/nae/init_nae.c b/drivers/net/ethernet/broadcom/nae/init_nae.c
index dc1a614..236dd65 100644
--- a/drivers/net/ethernet/broadcom/nae/init_nae.c
+++ b/drivers/net/ethernet/broadcom/nae/init_nae.c
@@ -3,6 +3,7 @@
 #include <linux/mm.h>
 
 #include <asm/netlogic/hal/nlm_hal_nae.h>
+#include <asm/netlogic/hal/nlm_hal_xlp_dev.h>
 #include <ops.h>
 
 #include "net_common.h"
@@ -87,6 +88,7 @@ static void parse_fdt_nae_config(void)
 	char port_type_str[MAX_PROP_LEN];
 	int size = 0;
 	uint32_t start_port, num_nae_regs, num_intf_regs;
+	int num_ports = 0;
 
 	/* Parse Nae Config */
 	start_port = num_nae_regs = num_intf_regs = 0;
@@ -105,7 +107,7 @@ static void parse_fdt_nae_config(void)
 	if(getprop(node, "start-port-id", &start_port, sizeof(uint32_t)) < 0)
 		printk("fdt missing start-port-id\n");
 
-	if(getprop(node, "num-ports", &netif_ports, sizeof(uint32_t)) < 0)
+	if(getprop(node, "num-ports", &num_ports, sizeof(uint32_t)) < 0)
 		printk("fdt missing num-ports\n");
 
 	/*******************************************************************************/
@@ -116,20 +118,55 @@ static void parse_fdt_nae_config(void)
 	if(getprop(node, "num-intf-regs", &num_intf_regs, sizeof(uint32_t)) < 0)
 		printk("fdt missing num-if-regs\n");
 
-	printk("num-ports = %d, start-port-id = %d\n", netif_ports, start_port);
+	printk("num-ports = %d, start-port-id = %d\n", num_ports, start_port);
 
 	if (debug)
 		printk("num-nae-regs = %d, num-intf-regs = %d\n", num_nae_regs, num_intf_regs);
 
-	for(i = 0; i < netif_ports; i++)
+	//nlm_hal_sgmii_pcs_init();
+
+	for(i = 0; i < num_ports; i++)
 	{
 		uint32_t *nae_regs = 0, *intf_regs = 0;
+		int mgmt_intf = 0;
+		int hw_port_id = 0;
+		int txq_range[2];
+		int rxq;
 
 		sprintf(domstr, "/soc/nae-cfg/port@%d", i);
 
 		subnode = finddevice(domstr);
 		if(!subnode) continue;
 
+		if (getprop(subnode, "mgmt", &mgmt_intf, sizeof(uint32_t)) < 0)
+			printk("fdt missing mgmt\n");
+		else {
+			if (!mgmt_intf) {
+				printk("Skipping non-mgmt port[%d]\n", i);
+				continue;
+			}
+		}
+
+		netif_ports++;
+
+		if (getprop(subnode, "tx-que-range", txq_range, sizeof(uint32_t) * 2) < 0)
+			printk("fdt missing tx-que-range\n");
+		else {
+			printk("tx-que-range[%d %d]\n", txq_range[0], txq_range[1]);
+		}
+
+		if (getprop(subnode, "rx-que", &rxq, sizeof(uint32_t)) < 0)
+			printk("fdt missing rx-que\n");
+		else {
+			printk("rx-que[%d]\n", rxq);
+		}
+
+		if (getprop(subnode, "hw-port-id", &hw_port_id, sizeof(uint32_t)) < 0)
+			printk("fdt missing hw-port-id\n");
+		else {
+			printk("hw-port-id[%d]\n", hw_port_id);
+		}
+
 		getprop(subnode, "type", port_type_str, MAX_PROP_LEN);
 		if (!strcmp(port_type_str, "SGMII_IF")) port_type = SGMII_IF;
 		else if (!strcmp(port_type_str, "XAUI_IF")) port_type = XAUI_IF;
@@ -155,18 +192,18 @@ static void parse_fdt_nae_config(void)
 		}
 		getprop(subnode, "intf-regs", intf_regs, size);
 
-		printk("Configuring per-port interface registers for port@%d\n", i);
+		//printk("Configuring per-port interface registers for port@%d\n", i);
 		/* Configure per port interface registers */
-		nlm_hal_init_if_regs(port_type, i, intf_regs, num_intf_regs);
+		//nlm_hal_init_if_regs(port_type, hw_port_id, intf_regs, num_intf_regs);
 
-		printk("Configuring per-port NAE registers for port@%d\n", i);
+		//printk("Configuring per-port NAE registers for port@%d\n", i);
 		/* Configure per port NAE registers */
-		nlm_hal_init_nae_regs(port_type, nae_regs, num_nae_regs);
+		//nlm_hal_init_nae_regs(port_type, nae_regs, num_nae_regs);
 
 		kfree(nae_regs);
 		kfree(intf_regs);
 
-		if (nlm_hal_open_if(port_type, i) < 0) {
+		if (nlm_hal_open_if(port_type, hw_port_id) < 0) {
 			printk("[%s] Unable to open port %d\n", __func__, i);
 			continue;
 		}
diff --git a/drivers/net/ethernet/broadcom/nae/xlp_nae.c b/drivers/net/ethernet/broadcom/nae/xlp_nae.c
index 014fdae..a8d133c 100755
--- a/drivers/net/ethernet/broadcom/nae/xlp_nae.c
+++ b/drivers/net/ethernet/broadcom/nae/xlp_nae.c
@@ -80,8 +80,9 @@
 #define ETH_MTU_SIZE		 	1536
 #define MIN_ETH_FRAME_SIZE		64
 
-#define  DUMP_PKT(x, y)	if (debug == 2)  {	\
+#define  DUMP_PKT(str, x, y)	if (debug == 2)  {	\
 	int i;      				\
+        printk(" %s \n", str);                  \
         for(i = 0; i < y; i++)			\
         {					\
                 printk("%02x ", x[i]);		\
@@ -445,13 +446,14 @@ static void nlm_xlp_nae_init(void)
 	printk("======= Module Parameters =========\n");
 	printk("debug = %d, hwemul=%d, naecfg_hack=%d\n",
 	       debug, hwemul, naecfg_hack);
-
+#if 0
 	if (hwemul) {
 		unsigned long mflags = 0;
 		msgrng_access_enable(mflags);
 		nlm_hal_fmn_init(0xE0000000ULL, 0x1000000, 50);
 		msgrng_access_disable(mflags);
 	}
+#endif
 
 	initialize_nae();
 
@@ -478,8 +480,8 @@ static void nlm_xlp_nae_init(void)
 		priv->inited	= 0;
 		priv->block 	=(i & 0xff) >> 2;
 		priv->index 	= i & 0x3;
-		priv->nae_tx_qid 	= XLP_NET_TX_VC_BASE+i;
-		priv->nae_rx_qid 	= XLP_NET_RX_VC_BASE+i;
+		priv->nae_tx_qid 	= 492 + i;
+		priv->nae_rx_qid 	= 1016 + i;
 		register_netdev(dev);
 
 		dev_mac_type[priv->type][priv->port] = dev;
@@ -684,7 +686,7 @@ static int nlm_xlp_nae_start_xmit(struct sk_buff *skb, struct net_device *dev)
 
 	msg.entry[2] = msg.entry[3] = 0;
 
-	DUMP_PKT(skb->data, skb->len);
+	DUMP_PKT(__func__, skb->data, skb->len);
 
 	__sync();
 
@@ -885,7 +887,7 @@ static void nlm_xlp_nae_msgring_handler(uint32_t vc, uint32_t src_id,
 		 * or vlan priority
 		 */
 		context = (msg0 >> 40) & 0x3fff;
-		port = context / XLP_SGMII_RCV_CONTEXT_NUM;
+		port = 0;//context / XLP_SGMII_RCV_CONTEXT_NUM;
 
 		/* update dev and port to be accurate */
 		if(port >= MAX_GMAC_PORT) {
@@ -905,7 +907,7 @@ static void nlm_xlp_nae_msgring_handler(uint32_t vc, uint32_t src_id,
 		if(skb)
 		{
 			if (debug) {
-				printk("[%s]: addr=%llx, skb=%p, context=%d, port=%d\n",
+				printk("[%s][TXC] addr=%llx, skb=%p, context=%d, port=%d\n",
 				       __func__, addr, skb, context, port);
 			}
 			dev_kfree_skb_any(skb);
@@ -924,7 +926,7 @@ static void nlm_xlp_nae_msgring_handler(uint32_t vc, uint32_t src_id,
 		context = msg1 >> 54;
 
 		/*context is XLP_SGMII_RCV_CONTEXT_NUM + three bit vlan type or vlan priority*/
-		port = context / XLP_SGMII_RCV_CONTEXT_NUM;
+		port = 0; //context / XLP_SGMII_RCV_CONTEXT_NUM;
 
 		/* update dev and port to be accurate */
 		if(port < 0 || port >= MAX_GMAC_PORT)
@@ -942,11 +944,11 @@ static void nlm_xlp_nae_msgring_handler(uint32_t vc, uint32_t src_id,
 		buf = (unsigned char *)vaddr;
 
 		if (debug) {
-			printk("[%s]: addr=%llx, len=%d, context=%d, port=%d, vaddr=%llx, buf=%p\n",
+			printk("[%s][RX] addr=%llx, len=%d, context=%d, port=%d, vaddr=%llx, buf=%p\n",
 			       __func__, addr, len, context, port, vaddr, buf);
 		}
 
-		DUMP_PKT(buf , len);
+		DUMP_PKT("RX Packet: ", buf , len);
 
 		len = len  - MAC_CRC_LEN;
 
@@ -978,8 +980,6 @@ static void nlm_xlp_nae_msgring_handler(uint32_t vc, uint32_t src_id,
 			{
 				tasklet_schedule(&mac_refill_task[port]);
 			}
-
-			DUMP_PKT(skb->data, skb->len);
 		}
 		else if(!skb)
 		{
-- 
1.7.1

