From ffd185e25a298b88528fee49e203f4da2a680325 Mon Sep 17 00:00:00 2001
From: Venu Vadapalli <vvadapalli@netlogicmicro.com>
Date: Sat, 2 Oct 2010 16:10:59 -0700
Subject: Added gitignore files

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/drivers/net/ethernet/broadcom/nae/.gitignore b/drivers/net/ethernet/broadcom/nae/.gitignore
index 2d6f993..d7384ab 100644
--- a/drivers/net/ethernet/broadcom/nae/.gitignore
+++ b/drivers/net/ethernet/broadcom/nae/.gitignore
@@ -5,3 +5,4 @@ modules.order
 *.ko
 *.mod.c
 .tmp_versions
+
-- 
1.7.1

