From 00625eb71c2d683a17efeaace59dc9937c431235 Mon Sep 17 00:00:00 2001
From: Venu Vadapalli <vvadapalli@netlogicmicro.com>
Date: Wed, 20 Oct 2010 19:55:30 -0700
Subject: NAE: Increment frin_to_be_sent when dropping rx packets

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/drivers/net/ethernet/broadcom/nae/xlp_nae.c b/drivers/net/ethernet/broadcom/nae/xlp_nae.c
index ac2958c..f64c6a3 100755
--- a/drivers/net/ethernet/broadcom/nae/xlp_nae.c
+++ b/drivers/net/ethernet/broadcom/nae/xlp_nae.c
@@ -887,7 +887,7 @@ static void nlm_xlp_nae_msgring_handler(uint32_t vc, uint32_t src_id,
 			priv->stats.rx_dropped++;
 
 			dev_kfree_skb_any(skb);
-			return;
+			goto out;
 		}
 
 		skb_put(skb, len);
@@ -903,6 +903,7 @@ static void nlm_xlp_nae_msgring_handler(uint32_t vc, uint32_t src_id,
 		priv->stats.rx_packets++;
 		priv->cpu_stats[cpu].rx_packets++;
 
+	out:
 		if (atomic64_inc_return(&priv->frin_to_be_sent) > frin_desc_thres);
 		{
 			tasklet_schedule(&mac_refill_task[port]);
-- 
1.7.1

