From 57af5163877fa31de0d018a8f1269076c7cdc4ae Mon Sep 17 00:00:00 2001
From: Zi Shen Lim <zlim@netlogicmicro.com>
Date: Thu, 9 Dec 2010 13:21:34 -0800
Subject: NAE: skip non-existant ports when removing module

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/drivers/net/ethernet/broadcom/nae/xlp_nae.c b/drivers/net/ethernet/broadcom/nae/xlp_nae.c
index d12d091..d807ae1 100755
--- a/drivers/net/ethernet/broadcom/nae/xlp_nae.c
+++ b/drivers/net/ethernet/broadcom/nae/xlp_nae.c
@@ -1041,6 +1041,9 @@ static void nlm_xlp_nae_remove(void)
 	for (i = 0; i < MAX_GMAC_PORT; i++)
 	{
 		dev = dev_mac[i];
+
+		if (dev == 0) continue;
+
 		priv = netdev_priv(dev);
 		netif_napi_del(&priv->napi);
 		unregister_netdev(dev);
-- 
1.7.1

