From 19425dc0e349a2d4bedd6b48d70252548c8a96ac Mon Sep 17 00:00:00 2001
From: Mehul <vmehul@netlogicmicro.com>
Date: Wed, 23 Mar 2011 19:39:54 +0530
Subject: Added hybrid nae application and char driver to map shared memory. Added msgrng_access_enable/disable during nae init.

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/drivers/net/ethernet/broadcom/nae/init_nae.c b/drivers/net/ethernet/broadcom/nae/init_nae.c
index 16635ef..c18f1c0 100644
--- a/drivers/net/ethernet/broadcom/nae/init_nae.c
+++ b/drivers/net/ethernet/broadcom/nae/init_nae.c
@@ -39,13 +39,16 @@ static void config_fmn(void)
 int initialize_nae(uint32_t cm0, uint32_t cm1, uint32_t cm2, uint32_t cm3)
 {
 	int dom_id = 0;
+	unsigned long mflags;
 
 	config_fmn();
 
+	msgrng_access_enable(mflags);
 	nlm_hal_init_nae(fdt, dom_id);
 
 	printk("Overriding HAL POE configuration based on current active cpumask\n");
 	nlm_hal_init_poe_distvec(0, cm0, cm1, cm2, cm3, (1 << nae_cfg.rx_vc));
 
+	/* msgrng_access_disable(mflags); */
 	return 0;
 }
-- 
1.7.1

