From 154366201c6166931ce1bf259eacce9a2e1cdc8d Mon Sep 17 00:00:00 2001
From: Mehul <vmehul@netlogicmicro.com>
Date: Thu, 24 Mar 2011 15:13:50 +0530
Subject: Uncommented msgrng_access_disable.

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/drivers/net/ethernet/broadcom/nae/init_nae.c b/drivers/net/ethernet/broadcom/nae/init_nae.c
index c18f1c0..42940f3 100644
--- a/drivers/net/ethernet/broadcom/nae/init_nae.c
+++ b/drivers/net/ethernet/broadcom/nae/init_nae.c
@@ -49,6 +49,6 @@ int initialize_nae(uint32_t cm0, uint32_t cm1, uint32_t cm2, uint32_t cm3)
 	printk("Overriding HAL POE configuration based on current active cpumask\n");
 	nlm_hal_init_poe_distvec(0, cm0, cm1, cm2, cm3, (1 << nae_cfg.rx_vc));
 
-	/* msgrng_access_disable(mflags); */
+	msgrng_access_disable(mflags);
 	return 0;
 }
-- 
1.7.1

