From af4d31f2b8745f0009212fb9fa954676ab92be14 Mon Sep 17 00:00:00 2001
From: jayanthi <jayanthia@rmicorp.com>
Date: Mon, 16 May 2011 18:29:44 +0530
Subject: NAE driver: Fix typo

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/drivers/net/ethernet/broadcom/nae/xlp_nae.c b/drivers/net/ethernet/broadcom/nae/xlp_nae.c
index e14ae73..8f8ab1d 100755
--- a/drivers/net/ethernet/broadcom/nae/xlp_nae.c
+++ b/drivers/net/ethernet/broadcom/nae/xlp_nae.c
@@ -972,7 +972,7 @@ static void nlm_xlp_nae_msgring_handler(uint32_t vc, uint32_t src_id,
 		priv->cpu_stats[cpu].rx_packets++;
 
 	out:
-		if (atomic64_inc_return(&priv->frin_to_be_sent) > frin_desc_thres);
+		if (atomic64_inc_return(&priv->frin_to_be_sent) > frin_desc_thres)
 		{
 			tasklet_schedule(&mac_refill_task[port]);
 			//mac_refill_frin_desc((unsigned long) skb->dev) ;
-- 
1.7.1

