From 365fda9fc6b6fc05dbf5e86dd8e38ba8552d67d6 Mon Sep 17 00:00:00 2001
From: Krishnamurthy D V <kmurthy@netlogicmicro.com>
Date: Tue, 21 Jun 2011 18:31:08 +0530
Subject: using msgring bkup timer instead of ipi intr

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/drivers/net/ethernet/broadcom/nae/xlp_nae.c b/drivers/net/ethernet/broadcom/nae/xlp_nae.c
index 8f8ab1d..5bb0ead 100755
--- a/drivers/net/ethernet/broadcom/nae/xlp_nae.c
+++ b/drivers/net/ethernet/broadcom/nae/xlp_nae.c
@@ -536,7 +536,7 @@ static int  nlm_xlp_nae_open (struct net_device *dev)
 	priv->phy_oldlinkstat = -1;
 
 	netif_tx_start_all_queues(dev);
-	add_timer(&priv->link_timer);
+//	add_timer(&priv->link_timer);
 
 /*	napi_enable(&priv->napi);*/
 
-- 
1.7.1

