From 16b1995a3c559ce16a8e2bfa7d7f9697be81c894 Mon Sep 17 00:00:00 2001
From: Jayanthi Annadurai <jayanthia@netlogicmicro.com>
Date: Fri, 17 Jun 2011 14:10:25 -0700
Subject: Added SRIO support

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/drivers/net/ethernet/broadcom/nae/xlp_nae.c b/drivers/net/ethernet/broadcom/nae/xlp_nae.c
index 5bb0ead..c6fc49f 100755
--- a/drivers/net/ethernet/broadcom/nae/xlp_nae.c
+++ b/drivers/net/ethernet/broadcom/nae/xlp_nae.c
@@ -1115,7 +1115,7 @@ static void nlm_xlp_nae_remove(void)
 		if (dev == 0) continue;
 
 		priv = netdev_priv(dev);
-		netif_napi_del(&priv->napi);
+//		netif_napi_del(&priv->napi);
 		unregister_netdev(dev);
 		free_netdev(dev);
 	}
-- 
1.7.1

