From 2042cd05cf192509e751a24b04fb96bf444db5c3 Mon Sep 17 00:00:00 2001
From: Om Narasimhan <onarasimhan@netlogicmicro.com>
Date: Tue, 2 Aug 2011 19:47:55 -0700
Subject: changes for new IRT-MAP subsystem.

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/drivers/net/ethernet/broadcom/nae/init_nae.c b/drivers/net/ethernet/broadcom/nae/init_nae.c
index 42940f3..302251f 100644
--- a/drivers/net/ethernet/broadcom/nae/init_nae.c
+++ b/drivers/net/ethernet/broadcom/nae/init_nae.c
@@ -26,7 +26,7 @@ static void config_fmn(void)
 	/* Configure FMN again but only cpu credits */
 	msgrng_access_enable(mflags);
 
-	nlm_xlp_msgring_int_handler(IRQ_MSGRING, NULL);
+	nlm_xlp_msgring_int_handler(XLP_IRQ_MSGRING, NULL);
 
 	/* Configure credits to non-n0c0 cores */
 	nlm_hal_fmn_init(0x10000000, 0x02000000, 50);
diff --git a/drivers/net/ethernet/broadcom/nae/xlp_nae.c b/drivers/net/ethernet/broadcom/nae/xlp_nae.c
index c6fc49f..17d8a08 100755
--- a/drivers/net/ethernet/broadcom/nae/xlp_nae.c
+++ b/drivers/net/ethernet/broadcom/nae/xlp_nae.c
@@ -57,7 +57,7 @@
 
 #include <asm/netlogic/hal/nlm_hal_fmn.h>
 #include <asm/netlogic/hal/nlm_hal_nae.h>
-#include <asm/netlogic/hal/nlm_hal_pic.h>
+#include <asm/netlogic/xlp_hal_pic.h>
 
 #include "net_common.h"
 #include "xlp_nae.h"
@@ -1076,14 +1076,13 @@ static void nlm_xlp_mac_timer(unsigned long data)
 
 		/* Send IRQ_MSGRING vector in an IPI to all cpus but the current one */
 		if (cpumask_lo)
-			nlm_hal_write_pic_reg(mmio, PIC_IPI_CTL, (IRQ_MSGRING << 20) | cpumask_lo );
+			nlm_hal_write_pic_reg(mmio, PIC_IPI_CTL, (XLP_IRQ_MSGRING << 20) | cpumask_lo );
 
 		if (cpumask_hi)
-			nlm_hal_write_pic_reg(mmio, PIC_IPI_CTL, (IRQ_MSGRING << 20) | (1 << 16)
-					      | (cpumask_hi));
+			nlm_hal_write_pic_reg(mmio, PIC_IPI_CTL, (XLP_IRQ_MSGRING << 20) | (1 << 16) | (cpumask_hi));
 
 		/* Run IPI handler on this cpu too */
-		nlm_xlp_msgring_int_handler(IRQ_MSGRING, NULL);
+		nlm_xlp_msgring_int_handler(XLP_IRQ_MSGRING, NULL);
 	}
 
         priv->link_timer.expires = jiffies + next_tick;
-- 
1.7.1

