From f26f51151822fde14445daf4492fb576f23c5bd7 Mon Sep 17 00:00:00 2001
From: jayanthi <jayanthia@rmicorp.com>
Date: Fri, 1 Jul 2011 11:41:11 +0530
Subject: Fix: Free in desc threshold, rmmod nae crash

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/drivers/net/ethernet/broadcom/nae/xlp_nae.c b/drivers/net/ethernet/broadcom/nae/xlp_nae.c
index 17d8a08..6c597db 100755
--- a/drivers/net/ethernet/broadcom/nae/xlp_nae.c
+++ b/drivers/net/ethernet/broadcom/nae/xlp_nae.c
@@ -445,6 +445,7 @@ static void nlm_xlp_nae_init(void)
 		dev->dev_addr = eth_hw_addr[i];
 		priv->port	= i;
 
+		priv->frin_desc_thres = nae_cfg.ports[i].num_free_desc / 3;
 		atomic64_set(&priv->frin_to_be_sent, nae_cfg.ports[i].num_free_desc);
 		atomic64_set(&priv->num_replenishes, 0);
 		atomic64_set(&priv->total_frin_sent, 0);
@@ -972,7 +973,7 @@ static void nlm_xlp_nae_msgring_handler(uint32_t vc, uint32_t src_id,
 		priv->cpu_stats[cpu].rx_packets++;
 
 	out:
-		if (atomic64_inc_return(&priv->frin_to_be_sent) > frin_desc_thres)
+		if (atomic64_inc_return(&priv->frin_to_be_sent) > priv->frin_desc_thres)
 		{
 			tasklet_schedule(&mac_refill_task[port]);
 			//mac_refill_frin_desc((unsigned long) skb->dev) ;
@@ -1061,6 +1062,7 @@ static void nlm_xlp_mac_timer(unsigned long data)
         int next_tick = HZ / 1000; /* 1ms */
 
 	/* printk("[%s] A0 Workaround, forcing FMN int handling \n",__func__); */
+#ifdef XLP_MSGRNG_TIMER
 	if (priv->inited)
 	{
 		uint32_t cpumask = cpumask_to_uint32(&cpu_present_map); /* doesn't handle non-n0 nodes */
@@ -1084,7 +1086,7 @@ static void nlm_xlp_mac_timer(unsigned long data)
 		/* Run IPI handler on this cpu too */
 		nlm_xlp_msgring_int_handler(XLP_IRQ_MSGRING, NULL);
 	}
-
+#endif
         priv->link_timer.expires = jiffies + next_tick;
         add_timer(&priv->link_timer);
 }
@@ -1114,7 +1116,7 @@ static void nlm_xlp_nae_remove(void)
 		if (dev == 0) continue;
 
 		priv = netdev_priv(dev);
-//		netif_napi_del(&priv->napi);
+		//netif_napi_del(&priv->napi);
 		unregister_netdev(dev);
 		free_netdev(dev);
 	}
diff --git a/drivers/net/ethernet/broadcom/nae/xlp_nae.h b/drivers/net/ethernet/broadcom/nae/xlp_nae.h
index 3b3d480..485f2e1 100644
--- a/drivers/net/ethernet/broadcom/nae/xlp_nae.h
+++ b/drivers/net/ethernet/broadcom/nae/xlp_nae.h
@@ -64,6 +64,7 @@ struct dev_data
         unsigned short type;
         struct sk_buff* skb;
         int phy_oldlinkstat;
+	uint32_t  frin_desc_thres;
         atomic64_t frin_to_be_sent;
 	atomic64_t num_replenishes;
 	atomic64_t total_frin_sent;
-- 
1.7.1

