From b0eca46f2b73027fa2d508f87976c757d00bc636 Mon Sep 17 00:00:00 2001
From: Yonghong Song <ysong@netlogicmicro.com>
Date: Thu, 20 Oct 2011 09:34:21 -0700
Subject: Fix some variable declaration to avoid conflict when kernel inline nae driver is used

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/drivers/net/ethernet/broadcom/nae/init_nae.c b/drivers/net/ethernet/broadcom/nae/init_nae.c
index 302251f..167e0a6 100644
--- a/drivers/net/ethernet/broadcom/nae/init_nae.c
+++ b/drivers/net/ethernet/broadcom/nae/init_nae.c
@@ -13,8 +13,6 @@
 
 #include "net_common.h"
 
-extern int rely_on_firmware_config;
-
 static void config_fmn(void)
 {
 	unsigned long mflags = 0;
diff --git a/drivers/net/ethernet/broadcom/nae/net_common.h b/drivers/net/ethernet/broadcom/nae/net_common.h
index 0d0e8e8..d67f2f5 100644
--- a/drivers/net/ethernet/broadcom/nae/net_common.h
+++ b/drivers/net/ethernet/broadcom/nae/net_common.h
@@ -160,7 +160,6 @@ static __inline__ void tx_print(uint64_t msg0) {
 	       (msg0) & 0xffffffffffULL);
 }
 
-extern int debug;
 extern void *fdt;
 
 struct nae_port {
diff --git a/drivers/net/ethernet/broadcom/nae/xlp_nae.c b/drivers/net/ethernet/broadcom/nae/xlp_nae.c
index 6c597db..dd89021 100755
--- a/drivers/net/ethernet/broadcom/nae/xlp_nae.c
+++ b/drivers/net/ethernet/broadcom/nae/xlp_nae.c
@@ -107,7 +107,7 @@
 	printk("\n"); }
 
 /* Module Parameters */
-int debug = 0;
+static int debug = 0;
 module_param(debug, int, 0);
 
 static int drop_uboot_pkt = 1;
@@ -117,7 +117,7 @@ static unsigned long stats_uboot_pkts;
 extern int naecfg_hack;
 module_param(naecfg_hack, int, 0);
 
-int rely_on_firmware_config = 1;
+extern int rely_on_firmware_config;
 module_param(rely_on_firmware_config, int, 0);
 
 /***************************************************************
-- 
1.7.1

