From a7ebc9c75eefacc44e791c13495c68be13dca5a6 Mon Sep 17 00:00:00 2001
From: Jonathan Pui <jpui@netlogicmicro.com>
Date: Wed, 5 Oct 2011 18:43:11 -0700
Subject: Changed drivers to use static variables.

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/drivers/net/ethernet/broadcom/nae/xlp_nae.c b/drivers/net/ethernet/broadcom/nae/xlp_nae.c
index dd89021..faab2cf 100755
--- a/drivers/net/ethernet/broadcom/nae/xlp_nae.c
+++ b/drivers/net/ethernet/broadcom/nae/xlp_nae.c
@@ -117,14 +117,14 @@ static unsigned long stats_uboot_pkts;
 extern int naecfg_hack;
 module_param(naecfg_hack, int, 0);
 
-extern int rely_on_firmware_config;
+static int rely_on_firmware_config = 1;
 module_param(rely_on_firmware_config, int, 0);
 
 /***************************************************************
  *
  * Below parameters are set during FDT file parsing
  */
-int frin_desc_thres = 24;
+static int frin_desc_thres = 24;
 module_param(frin_desc_thres, int, 0);
 
 static uint32_t nae_rx_vc = 0;
-- 
1.7.1

