From b125e5a1411d8470e57c409e24434595c4090b79 Mon Sep 17 00:00:00 2001
From: Vikas Gupta <vikasg@netlogicmicro.com>
Date: Mon, 7 Nov 2011 12:10:09 +0530
Subject: Commented debug messages.

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/drivers/net/ethernet/broadcom/nae/xlp_nae.c b/drivers/net/ethernet/broadcom/nae/xlp_nae.c
index ce13b29..d24d3ef 100755
--- a/drivers/net/ethernet/broadcom/nae/xlp_nae.c
+++ b/drivers/net/ethernet/broadcom/nae/xlp_nae.c
@@ -651,7 +651,7 @@ static int mac_fill_rxfr(struct net_device *dev)
 #error "Jumbo support cannot be enabled with CONFIG_PHOENIX_HW_BUFFER_MGMT"
 #endif /* CONFIG_PHOENIX_HW_BUFFER_MGMT */
 
-	printk(" mac_fill_rxfr--------\n");
+	//printk(" mac_fill_rxfr--------\n");
 	//if (!priv->init_frin_desc) return ret;
 	//priv->init_frin_desc = 0;	
 
@@ -673,7 +673,7 @@ static int mac_fill_rxfr(struct net_device *dev)
 
 		pa = page_to_phys(pg);
 		va = page_address(pg);
-		printk("Virt addr alloc_page=0x%lx\n", (unsigned long)va);
+		//printk("Virt addr alloc_page=0x%lx\n", (unsigned long)va);
 		j = 0;
 		while(nr_buffs) {
 			rx_cookie = (struct jumbo_rx_cookie *)va;
-- 
1.7.1

