From 6d2b80808520889e79b2b1c64d2facca899b5a27 Mon Sep 17 00:00:00 2001
From: Vikas Gupta <vikasg@netlogicmicro.com>
Date: Tue, 8 Nov 2011 20:39:37 +0530
Subject: Fix FCS error check for P2P based packet

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/drivers/net/ethernet/broadcom/nae/xlp_nae.c b/drivers/net/ethernet/broadcom/nae/xlp_nae.c
index d24d3ef..846617c 100755
--- a/drivers/net/ethernet/broadcom/nae/xlp_nae.c
+++ b/drivers/net/ethernet/broadcom/nae/xlp_nae.c
@@ -1440,7 +1440,7 @@ static void nlm_xlp_nae_msgring_handler(uint32_t vc, uint32_t src_id,
 				p2d = be64_to_cpu(*(uint64_t*)vaddr);
 				p2d_addr[idx] = p2d & 0xffffffffe0ULL;
 				p2d_len[idx] = (p2d >> 40) & 0x3fff;
-				bad_pkt = (p2d >> 4) & 0x1;
+				bad_pkt |= (p2d >> 4) & 0x1;
 				len += (p2d >> 40) & 0x3fff;
 				//printk("P2D is at = 0x%lx physbuff= 0x%lx len= 0x%x\n", vaddr, p2d_addr[idx],  p2d_len[idx] );
 			}
-- 
1.7.1

