From 575b3650dd853b1975f20858f00af7d60e833d44 Mon Sep 17 00:00:00 2001
From: Jayanthi A <jayanthia@netlogicmicro.com>
Date: Wed, 8 Feb 2012 15:40:49 +0530
Subject: Fixed vfbid and dv for non-contiguous onlinemask

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/drivers/net/ethernet/broadcom/nae/xlp_nae.c b/drivers/net/ethernet/broadcom/nae/xlp_nae.c
index 5fe3d7a..15f9ae3 100755
--- a/drivers/net/ethernet/broadcom/nae/xlp_nae.c
+++ b/drivers/net/ethernet/broadcom/nae/xlp_nae.c
@@ -54,6 +54,7 @@
 #include <asm/uaccess.h>
 #include <asm/netlogic/msgring.h>
 #include <asm/netlogic/cpumask.h>
+#include <asm/netlogic/xlp.h>
 
 #include <asm/netlogic/hal/nlm_hal_fmn.h>
 #include <asm/netlogic/hal/nlm_hal_nae.h>
@@ -728,7 +729,7 @@ static int nlm_initialize_vfbid(int node, int fbvc)
 		vfbid_tbl[i] = 0;
 
 	for (cpu = 0; cpu < NR_CPUS ; cpu++) {
-		if(!cpu_isset(cpu, cpu_present_map))
+		if(!cpu_isset(cpu, phys_cpu_present_map))
                         continue;
 		dst_node = cpu / 32;
 		vfbid_tbl[cpu] = (dst_node << 10) | (((cpu % 32) * 4) + fbvc);
@@ -750,12 +751,16 @@ static void nlm_xlp_nae_init(void)
 	struct proc_dir_entry *entry;
 	nlm_nae_config_ptr nae_cfg;
 	uint32_t cpu_mask[NLM_MAX_NODES];
+	char buf[CPUMASK_BUF];
 	
 	for(i=0; i < NLM_MAX_NODES; i++)
 		cpu_mask[i] = 0;
+        
+	cpumask_scnprintf(buf, CPUMASK_BUF, &phys_cpu_present_map);
+        printk("phys_cpu_present_map %s\n",buf);
 
 	for (cpu = 0, node = 0; cpu < NR_CPUS ; cpu++) {
-                if(!cpu_isset(cpu, cpu_present_map))
+                if(!cpu_isset(cpu, phys_cpu_present_map))
                         continue;
 		node = cpu / 32;
 		cpu_mask[node] |= (1 << (cpu % 32));
-- 
1.7.1

