From 26f5fb09452575f04c582d3a35571de8735dda74 Mon Sep 17 00:00:00 2001
From: Vikas Gupta <vikasg@netlogicmicro.com>
Date: Wed, 8 Feb 2012 17:43:41 +0530
Subject: FIxed 32 Bit Compilation failure

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/drivers/net/ethernet/broadcom/nae/xlp_nae.c b/drivers/net/ethernet/broadcom/nae/xlp_nae.c
index 3548017..07e2d7c 100755
--- a/drivers/net/ethernet/broadcom/nae/xlp_nae.c
+++ b/drivers/net/ethernet/broadcom/nae/xlp_nae.c
@@ -812,9 +812,9 @@ static void nlm_xlp_nae_init(void)
 			priv->port	= i;
 
 			priv->frin_desc_thres = nae_cfg->ports[i].num_free_desc / 3;
-			atomic64_set(&priv->frin_to_be_sent, nae_cfg->ports[i].num_free_desc);
-			atomic64_set(&priv->num_replenishes, 0);
-			atomic64_set(&priv->total_frin_sent, 0);
+			atomic_set(&priv->frin_to_be_sent, nae_cfg->ports[i].num_free_desc);
+			atomic_set(&priv->num_replenishes, 0);
+			atomic_set(&priv->total_frin_sent, 0);
 
 			priv->inited = 0;
 			priv->node = node;
-- 
1.7.1

