From 5cc496a3528459c8be623bc59785d5c0f15fc1c3 Mon Sep 17 00:00:00 2001
From: Hareesh R <hareeshr@broadcom.com>
Date: Wed, 25 Apr 2012 18:20:36 +0530
Subject: napi should work based on bkup timer in a2/a1 board

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/drivers/net/ethernet/broadcom/nae/xlp_nae.c b/drivers/net/ethernet/broadcom/nae/xlp_nae.c
index 8d292b0..11472d3 100755
--- a/drivers/net/ethernet/broadcom/nae/xlp_nae.c
+++ b/drivers/net/ethernet/broadcom/nae/xlp_nae.c
@@ -1044,11 +1044,14 @@ static void nlm_xlp_nae_init(void)
 	printk("num_descs_per_normalq=%d num_descs_per_jumboq=%d perf_mode=%s enable_lro=%d enable_jumbo=%d\n",
 	       num_descs_per_normalq, num_descs_per_jumboq, mode_str[perf_mode], enable_lro, enable_jumbo);
 
+#if 0
+	/* commenting out the below code, as it should use the bkup timer */
 	/* msgring intr may not work in [8421]xxAx parts, disabling the napi */
 	if(is_nlm_xlp8xx_ax() && enable_napi) {
 		printk("NAPI cannot be enabled on XLP Ax parts\n");
 		enable_napi = 0;
 	}
+#endif
 	
 	for(i=0; i<NR_CPUS; i++)
 		nlm_mode[CPU_INDEX(i)] = perf_mode;
-- 
1.7.1

