From 8c000a37c4a81688334824d39cbb93fbff6801f2 Mon Sep 17 00:00:00 2001
From: Alok Agrawat <alok@broadcom.com>
Date: Wed, 26 Sep 2012 03:44:28 -0700
Subject: nae: Remove compilation warnings

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/drivers/net/ethernet/broadcom/nae/xlp_hw.c b/drivers/net/ethernet/broadcom/nae/xlp_hw.c
index 996897d..946d6a4 100755
--- a/drivers/net/ethernet/broadcom/nae/xlp_hw.c
+++ b/drivers/net/ethernet/broadcom/nae/xlp_hw.c
@@ -206,7 +206,7 @@ static int xlp_config_msec_tx(struct net_device *dev, struct ethtool_cmd *cmd)
 	int ret = 0;
 	if(is_nlm_xlp2xx())
 	{
-		struct dev_data *priv = netdev_priv(dev);
+	//	struct dev_data *priv = netdev_priv(dev);
 		printk("%s node = %d port_enable = %x preamble_len = %x packet_num = %d pn_thrshld = %d\n", __FUNCTION__, cmd->node, cmd->port_enable, cmd->preamble_len, cmd->packet_num, cmd->win_size_thrshld);
 		nlm_hal_msec_tx_config(cmd->node, cmd->port_enable, cmd->preamble_len, cmd->packet_num, cmd->win_size_thrshld);
 	}
@@ -218,7 +218,7 @@ static int xlp_config_msec_tx_mem(struct net_device *dev, struct ethtool_cmd *cm
 	int ret = 0;
 	if(is_nlm_xlp2xx())
 	{
-		struct dev_data *priv = netdev_priv(dev);
+	//	struct dev_data *priv = netdev_priv(dev);
 		printk("%s node = %d context = %x tci = %x sci = %llx key = %llx\n", __FUNCTION__, cmd->node, cmd->port, cmd->tci, cmd->sci, *(uint64_t*)cmd->key);
 		nlm_hal_msec_tx_mem_config(cmd->node, cmd->port/*context*/, cmd->tci, cmd->sci, cmd->key);
 	}
@@ -230,7 +230,7 @@ static int xlp_config_msec_rx(struct net_device *dev, struct ethtool_cmd *cmd)
 	int ret = 0;
 	if(is_nlm_xlp2xx())
 	{
-		struct dev_data *priv = netdev_priv(dev);
+//		struct dev_data *priv = netdev_priv(dev);
 		printk("%s node = %d port_enable = %x preamble_len = %x packet_num = %d pn_thrshld = %d\n", __FUNCTION__, cmd->node, cmd->port_enable, cmd->preamble_len, cmd->packet_num, cmd->win_size_thrshld);
 		nlm_hal_msec_rx_config(cmd->node, cmd->port_enable, cmd->preamble_len, cmd->packet_num, cmd->win_size_thrshld);
 	}
@@ -242,7 +242,7 @@ static int xlp_config_msec_rx_mem(struct net_device *dev, struct ethtool_cmd *cm
 	int ret = 0;
 	if(is_nlm_xlp2xx())
 	{
-		struct dev_data *priv = netdev_priv(dev);
+//		struct dev_data *priv = netdev_priv(dev);
 		printk("%s node = %d, port = %d, index = %d, sci = %llx,  sci_mask = %llx key = %llx\n",__FUNCTION__, cmd->node, cmd->port, cmd->index, cmd->sci, cmd->sci_mask, *(uint64_t*)cmd->key); 
 		nlm_hal_msec_rx_mem_config(cmd->node, cmd->port, cmd->index, cmd->sci, cmd->key, cmd->sci_mask);
 	}
-- 
1.7.1

