From c8b612ec9662cdb482826af813e2bb8db39a8816 Mon Sep 17 00:00:00 2001
From: Alok Agrawat <alok@broadcom.com>
Date: Tue, 23 Oct 2012 06:50:30 -0700
Subject: XMC: Added timer support for autonegotiation

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/drivers/net/ethernet/broadcom/nae/xlp_nae.c b/drivers/net/ethernet/broadcom/nae/xlp_nae.c
index 16b47c1..b7b8b15 100755
--- a/drivers/net/ethernet/broadcom/nae/xlp_nae.c
+++ b/drivers/net/ethernet/broadcom/nae/xlp_nae.c
@@ -51,6 +51,7 @@
 #include <linux/net_tstamp.h>
 #include <linux/clocksource.h>
 #include <linux/timecompare.h>
+#include <linux/timer.h>
 
 #include <net/ip.h>
 
@@ -2665,6 +2666,27 @@ static void nlm_xlp_nae_remove(void)
 
 }
 
+struct timer_list phy_int_timer;
+static void phy_st_timer_handler(unsigned long data)
+{
+        struct timer_list *timer = &phy_int_timer;
+
+	nlm_hal_restart_an(0, 0);
+
+        timer->expires = jiffies + (HZ * 2);
+        add_timer(timer);
+}
+void init_phy_state_timer(void *data)
+{
+        struct timer_list *timer = &phy_int_timer;
+
+        init_timer(timer);
+        timer->expires = jiffies + 10;
+        timer->data = 0;
+        timer->function = phy_st_timer_handler;
+        add_timer(timer);
+}
+
 static struct pci_driver soc_driver = {
 	.name             = XLP_SOC_MAC_DRIVER,
 	.id_table         = soc_pci_table,
@@ -2676,6 +2698,7 @@ static int __init nlm_xlp_mac_init(void)
 {
 
 	nlm_xlp_nae_init();
+	init_phy_state_timer(NULL);
 
 	if(enable_napi)
 		nlm_xlp_enable_napi();
-- 
1.7.1

