From e8afb54a797abf4b44310f8f6c294fae1bd8da5b Mon Sep 17 00:00:00 2001
From: Vikas Gupta <vikas.gupta@broadcom.com>
Date: Fri, 4 Jan 2013 03:25:02 -0800
Subject: nae: Fix compilation error if NLM_NET_OPTS option enabled.

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/drivers/net/ethernet/broadcom/nae/xlpge.h b/drivers/net/ethernet/broadcom/nae/xlpge.h
index cc16373..55e3cae 100644
--- a/drivers/net/ethernet/broadcom/nae/xlpge.h
+++ b/drivers/net/ethernet/broadcom/nae/xlpge.h
@@ -493,5 +493,15 @@ extern uint64_t nlm_mode[];
 extern int nlm_prepad_len;
 extern int load_balance_timer_run;
 
+#ifdef CONFIG_NLM_NET_OPTS
+extern struct sk_buff *last_rcvd_skb[NR_CPUS * 8] ____cacheline_aligned;
+extern uint32_t last_rcvd_len[NR_CPUS * 8] ____cacheline_aligned;
+extern uint32_t last_rcvd_node[NR_CPUS * 8] ____cacheline_aligned;
+extern uint64_t last_rcvd_skb_phys[NR_CPUS * 8] ____cacheline_aligned;
+extern int get_hw_frfifo_queue_id(int rxnode, nlm_nae_config_ptr nae_cfg,
+                                  int cpu, uint32_t truesize, int hw_port_id);
+#endif
+
+
 #endif /* __ASSEMBLY__ */
 #endif
diff --git a/drivers/net/ethernet/broadcom/nae/xlpge_nae.c b/drivers/net/ethernet/broadcom/nae/xlpge_nae.c
index 25d079f..c8177cf 100644
--- a/drivers/net/ethernet/broadcom/nae/xlpge_nae.c
+++ b/drivers/net/ethernet/broadcom/nae/xlpge_nae.c
@@ -86,6 +86,14 @@ struct file_operations nlm_load_balance_proc_fops = {
 	.release = seq_release,
 };
 #endif
+
+#ifdef CONFIG_NLM_NET_OPTS
+struct sk_buff *last_rcvd_skb[NR_CPUS * 8] ____cacheline_aligned;
+uint32_t last_rcvd_len[NR_CPUS * 8] ____cacheline_aligned;
+uint32_t last_rcvd_node[NR_CPUS * 8] ____cacheline_aligned;
+uint64_t last_rcvd_skb_phys[NR_CPUS * 8] ____cacheline_aligned;
+#endif
+
 static unsigned short nlm_select_queue(struct net_device *dev,
 				       struct sk_buff *skb)
 {
@@ -826,7 +834,7 @@ static void nlm_enable_l3_l4_parser(int node)
 
 #ifdef CONFIG_NLM_NET_OPTS
 /* Get the hardware replenishment queue id */
-static int get_hw_frfifo_queue_id(int rxnode, nlm_nae_config_ptr nae_cfg,
+int get_hw_frfifo_queue_id(int rxnode, nlm_nae_config_ptr nae_cfg,
 				  int cpu, uint32_t truesize, int hw_port_id)
 {
 	/*
-- 
1.7.1

