From 26574c0bb6c548a0c4615b987160f51c3eea58d7 Mon Sep 17 00:00:00 2001
From: Sreenidhi B R <sreenira@broadcom.com>
Date: Wed, 9 Jan 2013 12:21:25 +0530
Subject: nae: reprogram rx buffer xon/xoff threshold, when mtu size changes (jumbo frame support)

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/drivers/net/ethernet/broadcom/nae/xlpge_nae.c b/drivers/net/ethernet/broadcom/nae/xlpge_nae.c
index c8177cf..0b14f64 100644
--- a/drivers/net/ethernet/broadcom/nae/xlpge_nae.c
+++ b/drivers/net/ethernet/broadcom/nae/xlpge_nae.c
@@ -373,6 +373,10 @@ static int initialize_nae_per_node(int node, uint32_t *phys_cpu_map, int mode,
 	 * for jumbo skbs */
 	if (*jumbo_enabled) {
 		int mine = 1;
+
+		/* update rx xon/xoff thresholds for jumbo */
+		nlm_hal_set_context_xon_xoff_threshold(node, ETH_JUMBO_DATA_LEN);
+
 		for (i = 0; i < nae_cfg->frin_total_queue; i++) {
 			if ((1 << i) & nae_cfg->freein_fifo_dom_mask) {
 				if (mine) {
-- 
1.7.1

