From 5976866f42867ca3f186d8d79795999a87b2b4e6 Mon Sep 17 00:00:00 2001
From: Vinod Mopuri <vmopuri@lc-sj1-5012.sj.broadcom.com>
Date: Wed, 16 Jan 2013 02:04:54 -0800
Subject: nae: Fixed compilation errors Added definitions that was missed in 3.0 nae driver migration.

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/drivers/net/ethernet/broadcom/nae/xlpge_rx.c b/drivers/net/ethernet/broadcom/nae/xlpge_rx.c
index a233426..74eced3 100644
--- a/drivers/net/ethernet/broadcom/nae/xlpge_rx.c
+++ b/drivers/net/ethernet/broadcom/nae/xlpge_rx.c
@@ -51,6 +51,14 @@ static uint32_t nlm_pcpu_mask = 0;
 static uint32_t *ucore_shared_data = NULL;
 #endif
 
+#ifdef CONFIG_NLM_NET_OPTS
+extern struct sk_buff *last_rcvd_skb[NR_CPUS * 8] ____cacheline_aligned;
+extern uint32_t last_rcvd_len[NR_CPUS * 8] ____cacheline_aligned;
+extern uint32_t last_rcvd_node[NR_CPUS * 8] ____cacheline_aligned;
+extern uint64_t last_rcvd_skb_phys[NR_CPUS * 8] ____cacheline_aligned;
+struct dev_data *last_rcvd_priv[NR_CPUS * 8] ____cacheline_aligned;
+#endif
+
 uint64_t receive_count[NR_CPUS * 8] __cacheline_aligned;
 
 inline void process_tx_complete(int cpu, uint32_t src_id, uint64_t msg0)
diff --git a/drivers/net/ethernet/broadcom/nae/xlpge_tx.c b/drivers/net/ethernet/broadcom/nae/xlpge_tx.c
index 55a74aa..4ccb95d 100644
--- a/drivers/net/ethernet/broadcom/nae/xlpge_tx.c
+++ b/drivers/net/ethernet/broadcom/nae/xlpge_tx.c
@@ -41,6 +41,10 @@
 
 #include "xlpge.h"
 
+#ifdef CONFIG_NLM_NET_OPTS
+extern struct dev_data *last_rcvd_priv[NR_CPUS * 8] ____cacheline_aligned;
+#endif
+
 uint64_t fast_replenish_count[NR_CPUS * 8] __cacheline_aligned;
 uint64_t slow_replenish_count[NR_CPUS * 8] __cacheline_aligned;
 uint64_t err_replenish_count[NR_CPUS * 8] __cacheline_aligned;
-- 
1.7.1

