From 80ff7bdf379e0adcef30fdd2f699a72ba7a4f892 Mon Sep 17 00:00:00 2001
From: Vikas Gupta <vikas.gupta@broadcom.com>
Date: Fri, 24 May 2013 16:57:44 +0530
Subject: nae: compilation fix for changed enum name in netlib.

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/drivers/net/ethernet/broadcom/nae/xlpge_nae.c b/drivers/net/ethernet/broadcom/nae/xlpge_nae.c
index 0ed578c..54a53c0 100644
--- a/drivers/net/ethernet/broadcom/nae/xlpge_nae.c
+++ b/drivers/net/ethernet/broadcom/nae/xlpge_nae.c
@@ -1446,7 +1446,7 @@ static int nlm_per_port_nae_init(nae_t* nae_cfg, int port, int maxnae)
 			(nae_cfg->ports[port].hw_port_id / 4),
 			XGMAC, XAUI_MAX_FRAME_LEN , 0x01800600);
 	#endif /*No frame length should be done here.*/
-		priv->index = XGMAC;
+		priv->index = XGMAC0;
 		break;
 	case INTERLAKEN_IF:
 		priv->index = INTERLAKEN;
diff --git a/drivers/net/ethernet/broadcom/nae/xlpge_rx.c b/drivers/net/ethernet/broadcom/nae/xlpge_rx.c
index f0061d1..e395b07 100644
--- a/drivers/net/ethernet/broadcom/nae/xlpge_rx.c
+++ b/drivers/net/ethernet/broadcom/nae/xlpge_rx.c
@@ -669,7 +669,7 @@ static inline void process_rx_packets(int cpu, unsigned int src_id,
 	skb_reserve(skb, nlm_prepad_len);
 #endif
 
-	if (priv->index == XGMAC)
+	if (priv->index == XGMAC0)
 		msec_port = (priv->port | 0xf) << (4 * priv->block);
 	else
 		msec_port = 1 << port;
diff --git a/drivers/net/ethernet/broadcom/nae/xlpge_tso.c b/drivers/net/ethernet/broadcom/nae/xlpge_tso.c
index b000879..8885373 100644
--- a/drivers/net/ethernet/broadcom/nae/xlpge_tso.c
+++ b/drivers/net/ethernet/broadcom/nae/xlpge_tso.c
@@ -103,7 +103,7 @@ inline int tso_xmit_skb(struct sk_buff *skb, struct net_device *dev)
 		nae_cfg->sectag_offset[priv->port],
 		nae_cfg->sectag_len[priv->port], nae_cfg->icv_len[priv->port]);
 #endif
-	if(priv->index == XGMAC)
+	if(priv->index == XGMAC0)
 		msec_port = (priv->port | 0xf) << (4 * priv->block);
 	else
 		msec_port = 1 << priv->port;
-- 
1.7.1

