From 538fae881c7f445004d4053e6aa72944417f4aec Mon Sep 17 00:00:00 2001
From: Vikas Gupta <vikas.gupta@broadcom.com>
Date: Tue, 4 Jun 2013 19:35:27 -0700
Subject: xlp9xx: changes according to number of descriptors at RX path.

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/drivers/net/ethernet/broadcom/nae/xlpge_rx.c b/drivers/net/ethernet/broadcom/nae/xlpge_rx.c
index e395b07..0ce7ee5 100644
--- a/drivers/net/ethernet/broadcom/nae/xlpge_rx.c
+++ b/drivers/net/ethernet/broadcom/nae/xlpge_rx.c
@@ -70,7 +70,7 @@ inline void process_tx_complete(int cpu, uint32_t src_id, uint64_t msg0)
         uint64_t addr;
         uint32_t context, port, node;
 
-        Message("%s cpu %d src_id %d\n", __FUNCTION__, cpu, src_id);
+        printk("%s cpu %d src_id %d\n", __FUNCTION__, cpu, src_id);
 
         /* Process Transmit Complete, addr is the skb pointer */
         addr = msg0 & 0xffffffffffULL;
@@ -588,7 +588,7 @@ static inline void nlm_update_flow_stats(unsigned int *prepad,
 #endif
 
 static inline void process_rx_packets(int cpu, unsigned int src_id, 
-		unsigned long long msg0, unsigned long long msg1)
+		unsigned long long msg0, unsigned long long msg1, unsigned long long msg2)
 {
 	uint64_t addr;
 	uint32_t len, context, truesize;
@@ -600,6 +600,9 @@ static inline void process_rx_packets(int cpu, unsigned int src_id,
 	nae_t* nae_cfg;
 	uint32_t msec_port;
 
+	if(is_nlm_xlp9xx()){
+		msg1 = msg2;	
+	}
 	err = (msg1 >> 4) & 0x1;
 
 	/* Rx packet */
@@ -613,7 +616,7 @@ static inline void process_rx_packets(int cpu, unsigned int src_id,
 		
 	node = (src_id >> 10) & 0x3;
 
-	Message("%s in cpu %d src_id %d len %d context %d node %d err %d\n",
+	printk("%s in cpu %d src_id %d len %d context %d node %d err %d\n",
 		__func__, cpu, src_id, len, context, node, err);
 
 
@@ -804,16 +807,16 @@ void xlp_poll_upper(int cpu)
 static int xlp_poll_lower(int budget, int cpu)
 {
 	int status;
-	uint64_t msg0 = 0, msg1 = 0;
+	uint64_t msg0 = 0, msg1 = 0, msg2=0;
 	int no_rx_pkt_rcvd = 0;
 	uint32_t src_id = 0, size = 0, code;
 	unsigned long __attribute__ ((unused)) mflags;
 
 	while (budget--) {
 		msgrng_access_enable(mflags);
-		status = xlp_message_receive_2(nae_rx_vc, &src_id, &size,
-				&code, &msg0, &msg1);
-		Message("xlp_message_receive_2 src_id = 0x%x size=0x%x \n", src_id, size);
+		status = xlp_message_receive_3(nae_rx_vc, &src_id, &size,
+				&code, &msg0, &msg1, &msg2);
+		printk("xlp_message_receive_3 src_id = 0x%x size=0x%x \n", src_id, size);
 		msgrng_access_disable(mflags);
 
 		if (status) {
@@ -824,13 +827,13 @@ static int xlp_poll_lower(int budget, int cpu)
 
 		no_rx_pkt_rcvd++;
 #ifdef ENABLE_SANITY_CHECKS
-		if ((size != 2) && (size != 1)) {
+		if ((size < 1) && (size > 3)) {
 			printk("Unexpected single entry packet\n");
 			continue;
 		}
 #endif
-		if (size == 2)
-			process_rx_packets(cpu, src_id, msg0, msg1);
+		if (size >= 2)
+			process_rx_packets(cpu, src_id, msg0, msg1, msg2);
 		else if (size == 1)
 			process_tx_complete(cpu, src_id, msg0);
 		else {
@@ -862,7 +865,7 @@ static void nlm_xlp_nae_msgring_handler(uint32_t vc, uint32_t src_id,
 	int cpu = hard_smp_processor_id();
 
 	if (vc == nae_rx_vc && size == 2)
-		 process_rx_packets(cpu, src_id, msg0, msg1);
+		 process_rx_packets(cpu, src_id, msg0, msg1, msg2);
 	else if (vc == nae_fb_vc && size == 1)
 		process_tx_complete(cpu, src_id, msg0);
 	else {
diff --git a/drivers/net/ethernet/broadcom/nae/xlpge_tx.c b/drivers/net/ethernet/broadcom/nae/xlpge_tx.c
index de93ba1..b7810c8 100644
--- a/drivers/net/ethernet/broadcom/nae/xlpge_tx.c
+++ b/drivers/net/ethernet/broadcom/nae/xlpge_tx.c
@@ -108,7 +108,7 @@ int nlm_xlp_nae_start_xmit(struct sk_buff *skb, struct net_device *dev)
 		qid = nae_cfg->vfbtbl_sw_offset + (cpu % NLM_NCPUS_PER_NODE);
 		msg0 = nae_tx_desc(DESC_TYPE_P2DNEOP, qid, 0, virt_to_bus(skb));
 
-		Message("Tx, tx complete to cpu, cpu %d len %d qid %d\n",
+		printk("Tx, tx complete to cpu, cpu %d len %d qid %d\n",
 			cpu, skb->len, qid);
 	}
 	
-- 
1.7.1

