From a2ad77610d7cb704ef2e95aef3cab261102709e2 Mon Sep 17 00:00:00 2001
From: kopal <kopal@broadcom.com>
Date: Wed, 26 Jun 2013 15:36:49 +0530
Subject: nae: fixed ethtool write functionality

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/drivers/net/ethernet/broadcom/nae/xlpge_eeprom.c b/drivers/net/ethernet/broadcom/nae/xlpge_eeprom.c
index 5ae298b..f849871 100644
--- a/drivers/net/ethernet/broadcom/nae/xlpge_eeprom.c
+++ b/drivers/net/ethernet/broadcom/nae/xlpge_eeprom.c
@@ -52,7 +52,6 @@ int eeprom_write (int devaddr, int addr, int alen, u8 * wr_buf, int len)
        	if(client==NULL){
 		return -1;
 	}
-	memset(wr_buf, '0', sizeof(wr_buf)); 
 	i2c_smbus_write_i2c_block_data(client, addr, len, wr_buf);	
 	return 0;
 }
-- 
1.7.1

