From 2444cd1fa947ba52382e870874eb681baf6f749a Mon Sep 17 00:00:00 2001
From: Vikas Gupta <vikas.gupta@broadcom.com>
Date: Mon, 17 Jun 2013 18:29:33 -0700
Subject: nae: statistics should be read from netlib API.

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/drivers/net/ethernet/broadcom/nae/xlpge_ethtool.c b/drivers/net/ethernet/broadcom/nae/xlpge_ethtool.c
index 5163750..c996ad8 100644
--- a/drivers/net/ethernet/broadcom/nae/xlpge_ethtool.c
+++ b/drivers/net/ethernet/broadcom/nae/xlpge_ethtool.c
@@ -301,6 +301,13 @@ static void xlp_get_strings (struct net_device *dev, u32 stringset, u8 *buf)
 
 void xlp_get_mac_stats(struct net_device *dev, struct net_device_stats *stats)
 {
+
+	struct dev_data *priv = netdev_priv(dev);
+	net_port_t* netport = priv->nae_port;
+	netsoc_get_port_stats( netport ,(dev_stat_t*)stats);
+#if 0
+
+
 	struct dev_data *priv = netdev_priv(dev);
 #ifdef CONFIG_64BIT
 	uint64_t val;
@@ -363,6 +370,7 @@ void xlp_get_mac_stats(struct net_device *dev, struct net_device_stats *stats)
 	stats->tx_carrier_errors = STATS_RD(TX_DROP_FRAME_COUNTER);
 	stats->tx_fifo_errors = STATS_RD(TX_DROP_FRAME_COUNTER);
 	*/
+#endif
 	return;
 }
 
-- 
1.7.1

