From 81871af65146d42d4a502348c6b13121fb9a6c5c Mon Sep 17 00:00:00 2001
From: Vikas Gupta <vikas.gupta@broadcom.com>
Date: Tue, 18 Jun 2013 19:00:50 -0700
Subject: nae: mac address array should be read according to nae-id.

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/drivers/net/ethernet/broadcom/nae/xlpge_nae.c b/drivers/net/ethernet/broadcom/nae/xlpge_nae.c
index 23f31c5..d35f971 100644
--- a/drivers/net/ethernet/broadcom/nae/xlpge_nae.c
+++ b/drivers/net/ethernet/broadcom/nae/xlpge_nae.c
@@ -1425,7 +1425,7 @@ static int nlm_per_port_nae_init(nae_t* nae_cfg, int port, int maxnae)
 	/* set ethtool_ops which is inside xlp_ethtool.c file*/
 	xlp_set_ethtool_ops(dev);
 
-	dev->dev_addr 	= eth_hw_addr[node][port];
+	dev->dev_addr 	= eth_hw_addr[node][(nae_cfg->nae_id)*MAX_GMAC_PORT + port];
 	priv->port	= port;
 	priv->hw_port_id = nae_cfg->ports[port].hw_port_id;
 
diff --git a/drivers/net/ethernet/broadcom/nae/xlpge_rx.c b/drivers/net/ethernet/broadcom/nae/xlpge_rx.c
index e15565b..076fabe 100644
--- a/drivers/net/ethernet/broadcom/nae/xlpge_rx.c
+++ b/drivers/net/ethernet/broadcom/nae/xlpge_rx.c
@@ -804,7 +804,6 @@ void xlp_poll_upper(int cpu)
 			__sync();
 
 			process_tx_complete(cpu, src_id, msg0);
-
 	}
 }
 
-- 
1.7.1

