From 602e61d2dfb190c7385b0ccf107989854dc6e2a6 Mon Sep 17 00:00:00 2001
From: Vikas Gupta <vikas.gupta@broadcom.com>
Date: Wed, 19 Jun 2013 17:18:16 -0700
Subject: nae: context to port mismatch fixed.

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/drivers/net/ethernet/broadcom/nae/xlpge_rx.c b/drivers/net/ethernet/broadcom/nae/xlpge_rx.c
index 076fabe..08c60c8 100644
--- a/drivers/net/ethernet/broadcom/nae/xlpge_rx.c
+++ b/drivers/net/ethernet/broadcom/nae/xlpge_rx.c
@@ -70,7 +70,7 @@ inline void process_tx_complete(int cpu, uint32_t src_id, uint64_t msg0)
         uint64_t *p2pfbdesc;
 #endif
         uint64_t addr;
-        uint32_t context, port, node;
+        uint32_t context, node;
 
         Message("%s cpu %d src_id %d\n", __FUNCTION__, cpu, src_id);
 
@@ -84,7 +84,7 @@ inline void process_tx_complete(int cpu, uint32_t src_id, uint64_t msg0)
         	context = (msg0 >> 40) & 0x3ff;
 	}
         node = (src_id >> 10) & 0x3;
-        port = *(cntx2port[node] + context);
+        //port = *(cntx2port[node] + context);
 
 
 #ifdef TSO_ENABLED
@@ -618,11 +618,12 @@ static inline void process_rx_packets(int cpu, unsigned int src_id,
 		
 	node = (src_id >> 10) & 0x3;
 
-	Message("%s in cpu %d src_id %d len %d context %d node %d err %d\n",
-		__func__, cpu, src_id, len, context, node, err);
 
-
-	port = *(cntx2port[node] + context);
+	vaddr = (uint64_t)(unsigned long)bus_to_virt(addr);
+	nae_cfg = (uint64_t*)mac_get_nae_back_ptr(vaddr);
+	port = nae_cfg->cntx2port[context]; 
+	Message("%s in cpu %d src_id %d len %d context %d node %d err %d port=%d\n",
+		__func__, cpu, src_id, len, context, node, err, port);
 
 #ifdef ENABLE_SANITY_CHECKS
 	if (port >= MAX_GMAC_PORT) {
@@ -632,12 +633,11 @@ static inline void process_rx_packets(int cpu, unsigned int src_id,
 		return;
 	}
 #endif
-	vaddr = (uint64_t)(unsigned long)bus_to_virt(addr);
-	nae_cfg = (uint64_t*)mac_get_nae_back_ptr(vaddr);
+
 
 #ifdef ENABLE_SANITY_CHECKS
 	if(nae_cfg->nae_id!=0 && nae_cfg->nae_id !=1)
-		printk("Ooops.....NAE ID is wrong\n");
+		printk("Nae ID is  wrong\n");
 #endif
 	pdev = per_cpu_netdev[node][cpu][nae_cfg->nae_id][port];
 	 
-- 
1.7.1

