From 65f1b388d97fa2ea17ff9e4010f234fd360c1153 Mon Sep 17 00:00:00 2001
From: Vikas Gupta <vikas.gupta@broadcom.com>
Date: Wed, 19 Jun 2013 22:30:39 -0700
Subject: xlp9xx: fix for context field in descriptors

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/drivers/net/ethernet/broadcom/nae/xlpge_rx.c b/drivers/net/ethernet/broadcom/nae/xlpge_rx.c
index 08c60c8..abac48f 100644
--- a/drivers/net/ethernet/broadcom/nae/xlpge_rx.c
+++ b/drivers/net/ethernet/broadcom/nae/xlpge_rx.c
@@ -77,7 +77,7 @@ inline void process_tx_complete(int cpu, uint32_t src_id, uint64_t msg0)
         /* Process Transmit Complete, addr is the skb pointer */
         addr = msg0 & 0xffffffffffULL;
 
-	if((is_nlm_xlp3xx() || is_nlm_xlp2xx())){ 
+	if((is_nlm_xlp3xx() || is_nlm_xlp2xx() ||is_nlm_xlp9xx())){ 
         	context = (msg0 >> 40) & 0x3f;
 	}
 	else{
@@ -610,7 +610,7 @@ static inline void process_rx_packets(int cpu, unsigned int src_id,
 	/* Rx packet */
 	addr	= msg1 & 0xffffffffc0ULL;
 	len	= (msg1 >> 40) & 0x3fff;
-	if((is_nlm_xlp3xx() || is_nlm_xlp2xx())){ 
+	if((is_nlm_xlp3xx() || is_nlm_xlp2xx() ||is_nlm_xlp9xx())){ 
 		context = (msg1 >> 54) & 0x3f;
 	}else{
 		context = (msg1 >> 54) & 0x3ff;
-- 
1.7.1

