From 60516bfea0acb25ee63a229275812254ade2bcdf Mon Sep 17 00:00:00 2001
From: Ashok Kumar <ashoks@broadcom.com>
Date: Wed, 17 Jul 2013 06:17:15 +0530
Subject: nae: configure poe distr vector if it is enabled.

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/drivers/net/ethernet/broadcom/nae/xlpge_nae.c b/drivers/net/ethernet/broadcom/nae/xlpge_nae.c
index d35f971..077ab20 100644
--- a/drivers/net/ethernet/broadcom/nae/xlpge_nae.c
+++ b/drivers/net/ethernet/broadcom/nae/xlpge_nae.c
@@ -476,6 +476,18 @@ static int initialize_nae_per_node(nae_t * nae_cfg, uint32_t *phys_cpu_map, int
 
 	init_dummy_entries_for_port_fifos(nae_cfg);
 
+	if (nae_cfg->poe->dist_en)
+	{
+		cpu_mask_t cpumap = {0};
+
+		/* FIXME: for multi-node */
+		for (i = 0;i < NLM_MAX_NODES;i++)
+			cpumap.mask[i] = phys_cpu_map[i];
+
+		netsoc_config_poe_distvec(nae_cfg->poe, nae_cfg->shinfo[0].domid, &cpumap,
+								  nae_cfg->shinfo[0].rxvc);
+	}
+
 #if 0
 	if (is_nlm_xlp2xx()) {
 		nlm_hal_msec_tx_default_config(node,
-- 
1.7.1

