From db4c11218dc83296e4170b3c4a7530dcb46522e0 Mon Sep 17 00:00:00 2001
From: kopal <kopal@broadcom.com>
Date: Tue, 6 Aug 2013 16:42:13 +0530
Subject: nae: Remove debug prints.

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/drivers/net/ethernet/broadcom/nae/xlpge_rx.c b/drivers/net/ethernet/broadcom/nae/xlpge_rx.c
index 4a899b8..224c122 100644
--- a/drivers/net/ethernet/broadcom/nae/xlpge_rx.c
+++ b/drivers/net/ethernet/broadcom/nae/xlpge_rx.c
@@ -822,7 +822,6 @@ static int xlp_poll_lower(int budget, int cpu)
 		msgrng_access_enable(mflags);
 		status = xlp_message_receive_3(nae_rx_vc, &src_id, &size,
 				&code, &msg0, &msg1, &msg2);
-		printk("xlp_message_receive_3 src_id = 0x%x size=0x%x \n", src_id, size);
 		msgrng_access_disable(mflags);
 
 		if (status) {
-- 
1.7.1

