From 77d6637af068aff111f85c8de616a05aa54a927b Mon Sep 17 00:00:00 2001
From: Hareesh R <hareeshr@broadcom.com>
Date: Wed, 20 Nov 2013 15:46:28 +0530
Subject: seclib:3des key swap changes

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/drivers/net/ethernet/broadcom/nae/xlpge_tso.h b/drivers/net/ethernet/broadcom/nae/xlpge_tso.h
index 681476f..8d1ce04 100644
--- a/drivers/net/ethernet/broadcom/nae/xlpge_tso.h
+++ b/drivers/net/ethernet/broadcom/nae/xlpge_tso.h
@@ -329,7 +329,7 @@ retry_send:
 		ret = xlp_message_send_1(priv->nae_tx_qid, 0, msg);
 	msgrng_access_disable(mflags);
 	if(ret)	{
-		printk("Transmit failed\n");
+	//	printk("Transmit failed\n");
 		xlp_poll_upper(cpu);
 		retry_cnt++;
 		if(retry_cnt >= 128) {
-- 
1.7.1

