From 87d850fedd7bd4bbad5886feeb9314a9bc5d72d9 Mon Sep 17 00:00:00 2001
From: kopal <kopal@broadcom.com>
Date: Mon, 24 Feb 2014 11:20:05 +0530
Subject: NAE: Added support to change MTU size for RXAUI.

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/drivers/net/ethernet/broadcom/nae/xlpge_nae.c b/drivers/net/ethernet/broadcom/nae/xlpge_nae.c
index 1e2247f..3dccca3 100644
--- a/drivers/net/ethernet/broadcom/nae/xlpge_nae.c
+++ b/drivers/net/ethernet/broadcom/nae/xlpge_nae.c
@@ -1233,7 +1233,8 @@ static int nlm_xlp_nae_change_mtu(struct net_device *dev, int new_mtu)
 			//nlm_xlp_mac_set_enable(priv, 0); /* Disable MAC TX/RX */
 	}
 
-	if (priv->type==SGMII_IF || priv->type==XAUI_IF || priv->type== XLAUI_IF || priv->type== XFI_IF)
+	if (priv->type==SGMII_IF || priv->type==XAUI_IF || priv->type== XLAUI_IF || priv->type== XFI_IF
+										|| priv->type==RXAUI_IF)
 		netsoc_set_framesize(nae_port, local_mtu);
 	else if (priv->type==INTERLAKEN_IF){
 		/*TODO: Add IL frame set in brcm_netsoc */
-- 
1.7.1

