From faff4c44fd97ad839687f58c018212fe4b6ef149 Mon Sep 17 00:00:00 2001
From: Vikas Gupta <vikas.gupta@broadcom.com>
Date: Mon, 24 Feb 2014 12:56:16 +0530
Subject: xlp9xx: Option added for taking argumnets from app command line.

	* option added for KR4.
[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/drivers/net/ethernet/broadcom/nae/xlpge_nae.c b/drivers/net/ethernet/broadcom/nae/xlpge_nae.c
index 3dccca3..d43abf5 100644
--- a/drivers/net/ethernet/broadcom/nae/xlpge_nae.c
+++ b/drivers/net/ethernet/broadcom/nae/xlpge_nae.c
@@ -613,7 +613,7 @@ int initialize_nae(uint32_t *phys_cpu_map, int mode, int *jumbo_enabled)
 
 	brcm_netsoc_lib_init(&mod_api);
 
-	if (init_netsoc(fdt, dom_id) < 0) {
+	if (init_netsoc(fdt, dom_id, &mod_api) < 0) {
                 printk("NETSOC initialization failed \n");
 		return -1;
 	}
-- 
1.7.1

