From 9433a8a61593e87ab1ffd044d50ef29cd5f0f851 Mon Sep 17 00:00:00 2001
From: P. Sadik <psadik@broadcom.com>
Date: Tue, 4 Mar 2014 13:21:23 +0530
Subject: nae: set default value of mod_api struct to 0.

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/drivers/net/ethernet/broadcom/nae/xlpge_nae.c b/drivers/net/ethernet/broadcom/nae/xlpge_nae.c
index d43abf5..8e158c0 100644
--- a/drivers/net/ethernet/broadcom/nae/xlpge_nae.c
+++ b/drivers/net/ethernet/broadcom/nae/xlpge_nae.c
@@ -601,6 +601,8 @@ int initialize_nae(uint32_t *phys_cpu_map, int mode, int *jumbo_enabled)
 	nae_t* nae;
 	struct netsoc_lib_param mod_api;
 
+	memset(&mod_api, 0, sizeof (struct netsoc_lib_param));
+
 	msgrng_access_enable(mflags);
 	mod_api.contig_alloc = allocate_contig_phys;
 	mod_api.malloc = NULL;
@@ -705,7 +707,7 @@ int replenish_freein_fifos(void)
 	int max_descs_pqueue, num_descs, max_nae_units;
 	unsigned int blk_cmplx_map, cmplx;
 	unsigned int ndescs_nq, ndescs_jq;
-	
+
 	/*get max nae*/
 	max_nae_units = get_num_nae_pernode();
 	for (node = 0; node < NLM_MAX_NODES; node++) {
@@ -809,7 +811,7 @@ static void phy_st_timer_handler(unsigned long data)
 					if((arr[i]>-1)&&(arr[i]<tmp))
 						netsoc_start_autoneg(&nae_cfg->ports[i]);
 					else
-						continue;	
+						continue;
 					}
 		}
 	}
@@ -1383,7 +1385,7 @@ static int nlm_per_port_nae_init(nae_t* nae_cfg, int port, int maxnae)
 	int cpu;
 	int node = nae_cfg->node;
 	static int port_index = 0;
-	
+
 	if (!nae_cfg->ports[port].valid)
 		return -1;
 
@@ -1416,7 +1418,7 @@ static int nlm_per_port_nae_init(nae_t* nae_cfg, int port, int maxnae)
 	priv->type	= nae_cfg->ports[port].iftype;
 	priv->mgmt_port	= nae_cfg->ports[port].mgmt;
 	priv->port_index = port_index++;
-	
+
 	switch(nae_cfg->ports[port].iftype) {
 	case SGMII_IF:
 		priv->index = nae_cfg->ports[port].hw_port_id & 0x3;
-- 
1.7.1

