From 0236ef9469994ee96711f36b38ecb9a897a83e52 Mon Sep 17 00:00:00 2001
From: Alok Agrawat <alok@broadcom.com>
Date: Thu, 3 Apr 2014 03:11:40 -0700
Subject: nae: Enable Jumbo support in Interlaken driver.

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/drivers/net/ethernet/broadcom/nae/xlpge_nae.c b/drivers/net/ethernet/broadcom/nae/xlpge_nae.c
index f284aca..6ff6b0c 100644
--- a/drivers/net/ethernet/broadcom/nae/xlpge_nae.c
+++ b/drivers/net/ethernet/broadcom/nae/xlpge_nae.c
@@ -1261,13 +1261,8 @@ static int nlm_xlp_nae_change_mtu(struct net_device *dev, int new_mtu)
 	}
 
 	if (priv->type==SGMII_IF || priv->type==XAUI_IF || priv->type== XLAUI_IF || priv->type== XFI_IF
-										|| priv->type==RXAUI_IF)
+										|| priv->type==RXAUI_IF || priv->type==INTERLAKEN_IF)
 		netsoc_set_framesize(nae_port, local_mtu);
-	else if (priv->type==INTERLAKEN_IF){
-		/*TODO: Add IL frame set in brcm_netsoc */
-		//nlm_hal_set_ilk_framesize(priv->node, priv->block,
-		//	priv->phy.addr, local_mtu);
-	}
 	else {
 		spin_unlock_irqrestore(&priv->lock, flags);
 		return -1;
-- 
1.7.1

