From 13d679c37ea4beaeb43b0038854f621a146dce44 Mon Sep 17 00:00:00 2001
From: Anusha Nelluri <anushan@broadcom.com>
Date: Tue, 15 Jul 2014 17:38:46 +0530
Subject: halx_brcm: initializing nae_family_9xx to 1 for 9xx/5xx in hybrid Mode

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/drivers/net/ethernet/broadcom/nae/xlpge_init.h b/drivers/net/ethernet/broadcom/nae/xlpge_init.h
index 29241d8..efc6d6d 100644
--- a/drivers/net/ethernet/broadcom/nae/xlpge_init.h
+++ b/drivers/net/ethernet/broadcom/nae/xlpge_init.h
@@ -37,6 +37,7 @@
 
 extern int perf_mode;
 extern int num_cpus_per_node;
+extern int nae_family_9xx; 
 
 extern int get_num_cpus_per_node(void);
 extern int replenish_freein_fifos(void);
diff --git a/drivers/net/ethernet/broadcom/nae/xlpge_tso.h b/drivers/net/ethernet/broadcom/nae/xlpge_tso.h
index 7c5e1f7..369f30f 100644
--- a/drivers/net/ethernet/broadcom/nae/xlpge_tso.h
+++ b/drivers/net/ethernet/broadcom/nae/xlpge_tso.h
@@ -44,7 +44,6 @@
 //#define MACSEC_DEBUG	1
 extern struct p2p_desc_mem p2p_desc_mem[NR_CPUS];
 extern int force_tso;
-extern int nae_family_9xx;
  
 
 static __inline__ uint64_t nae_tso_desc0(
-- 
1.7.1

