From 910d839afaf2ebff51dedb23052d7a5b715d0a30 Mon Sep 17 00:00:00 2001
From: Pavani Reddy <pavani@broadcom.com>
Date: Fri, 1 Aug 2014 14:13:09 +0530
Subject: nae : Change the size of phys_cpu_map to fix multi-node issue nae on 9xx

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/drivers/net/ethernet/broadcom/nae/xlpge_nae.c b/drivers/net/ethernet/broadcom/nae/xlpge_nae.c
index 7617816..7d07a1e 100644
--- a/drivers/net/ethernet/broadcom/nae/xlpge_nae.c
+++ b/drivers/net/ethernet/broadcom/nae/xlpge_nae.c
@@ -97,7 +97,7 @@ static uint32_t lnx_frfifo_jumbo_mask[NLM_MAX_NODES][MAX_NAE_PERNODE];
 int enable_jumbo = 0;
 module_param(enable_jumbo, int, 0);
 struct p2p_desc_mem p2p_desc_mem[NR_CPUS] __cacheline_aligned;
-static unsigned int phys_cpu_map[NLM_MAX_NODES];
+static unsigned int phys_cpu_map[NR_CPUS/sizeof(int)];
 struct timer_list phy_int_timer;
 struct task_struct *kthread_an;
 #ifdef CONFIG_NLM_ENABLE_LOAD_BALANCING
@@ -1592,8 +1592,8 @@ void nlm_xlp_nae_init(void)
 	for (i = 0; i < NR_CPUS; i++) {
 	        if(!cpu_isset(i, phys_cpu_present_map))
                         continue;
-		phys_cpu_map[i / NLM_NCPUS_PER_NODE] |=
-			(1 << (i % NLM_NCPUS_PER_NODE));
+		phys_cpu_map[i / (NR_CPUS/sizeof(int))] |=
+			(1 << (i % (NR_CPUS/sizeof(int))));
 	}
 
 	num_cpus_per_node = get_num_cpus_per_node();
-- 
1.7.1

