From 23b8face6a6fa38fff87887a4beb117af0a0c2b9 Mon Sep 17 00:00:00 2001
From: Pavani Reddy <pavani@broadcom.com>
Date: Tue, 21 Oct 2014 16:39:07 +0530
Subject: Nae : Enable flow control by default

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/drivers/net/ethernet/broadcom/nae/xlpge_nae.c b/drivers/net/ethernet/broadcom/nae/xlpge_nae.c
index c713e08..af3cc24 100644
--- a/drivers/net/ethernet/broadcom/nae/xlpge_nae.c
+++ b/drivers/net/ethernet/broadcom/nae/xlpge_nae.c
@@ -1109,7 +1109,7 @@ static int  nlm_xlp_nae_open (struct net_device *dev)
 		else
 			priv->flow_ctrl=0;
 		//nlm_xlp_mac_set_enable(priv, 1);
-		netsoc_disable_flow_control(priv->nae_port);
+		//netsoc_disable_flow_control(priv->nae_port);
 
 	return ret;
 }
-- 
1.7.1

