From 2021d915d60f35cb74faae1ccc8d703a33c4a565 Mon Sep 17 00:00:00 2001
From: kopal <kopal@broadcom.com>
Date: Thu, 3 Jul 2014 23:54:59 +0530
Subject: Ethtool: report correct link status

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/drivers/net/ethernet/broadcom/nae/xlpge_ethtool.c b/drivers/net/ethernet/broadcom/nae/xlpge_ethtool.c
index 3ed091b..c425dcf 100644
--- a/drivers/net/ethernet/broadcom/nae/xlpge_ethtool.c
+++ b/drivers/net/ethernet/broadcom/nae/xlpge_ethtool.c
@@ -291,17 +291,15 @@ static int xlp_nway_reset(struct net_device *dev)
 static u32 xlp_get_link(struct net_device *dev)
 {
 	struct dev_data *priv = netdev_priv(dev);
-	struct nlm_hal_mii_info mii_info;
+        net_port_t* nae_port = priv->nae_port;
 	ulong flags;
-	if (priv->type != SGMII_IF)
-		return -EIO;
-
+	uint32_t duplex, speed;
 	spin_lock_irqsave(&priv->lock, flags);
-	//nlm_hal_status_ext_phy( priv->node, priv->phy.addr, &mii_info);
 
-	spin_unlock_irqrestore(&priv->lock, flags);
+	netsoc_get_phy_status(nae_port, &duplex, &speed);
 
-	if(mii_info.link_stat==1)
+	spin_unlock_irqrestore(&priv->lock, flags);
+	if(nae_port->link_stat==1)
                 return 1;
 
 	return 0;
-- 
1.7.1

