From f96efeab87778bdac16d6c8909aec74cb4d2319e Mon Sep 17 00:00:00 2001
From: Vikas Gupta <vikas.gupta@broadcom.com>
Date: Sat, 13 Dec 2014 15:37:33 +0530
Subject: nae : disable mac when port is closed.

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/drivers/net/ethernet/broadcom/nae/xlpge_nae.c b/drivers/net/ethernet/broadcom/nae/xlpge_nae.c
index c7b2aaa..0078e1d 100644
--- a/drivers/net/ethernet/broadcom/nae/xlpge_nae.c
+++ b/drivers/net/ethernet/broadcom/nae/xlpge_nae.c
@@ -1138,7 +1138,7 @@ static int  nlm_xlp_nae_stop (struct net_device *dev)
 	spin_lock_irq(&priv->lock);
 
 	if (nae_cfg->owned){
-		//nlm_xlp_mac_set_enable(priv, 0);
+		netsoc_close_port(priv->nae_port);	
 	}
 	priv->inited = 0;
 	netif_tx_stop_all_queues(dev);
-- 
1.7.1

