From 7a6ec392839a148647ff073284ba3eb02e5cce73 Mon Sep 17 00:00:00 2001
From: Alok Agrawat <aagrawat@netlogicmicro.com>
Date: Fri, 7 Oct 2011 15:57:02 +0530
Subject: Crypto: Fixed the crash observed while pumping packets at line rate.

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/drivers/crypto/sae/nlm_crypto.c b/drivers/crypto/sae/nlm_crypto.c
index fa7391f..a637464 100644
--- a/drivers/crypto/sae/nlm_crypto.c
+++ b/drivers/crypto/sae/nlm_crypto.c
@@ -76,7 +76,7 @@ static int xlp_sae_release(struct inode *, struct file *);
 static int nlm_fb_vc;
 static uint64_t nlm_tx_id, nlm_err_msg;
 struct semaphore sem;
-volatile int msg_hdl = 0;
+volatile int msg_hdl[32];
 
 extern struct proc_dir_entry *nlm_root_proc;
 //static const char *versionstr = "XLP2.0";
@@ -912,9 +912,9 @@ crypto_cipher_auth_op(struct crypto_session *session, struct crypto_param *cop)
     if(session->aip)
       update_auth_stats(session->aip->auth_alg, session->aip->auth_mode, cop->auth_len);
     nlm_fb_vc = crypto_get_fb_vc();
-    nlm_tx_id = (uint64_t) cop;
 
-    msg_hdl = 0;
+    nlm_tx_id = hard_smp_processor_id();
+    msg_hdl[nlm_tx_id] = 0;
     nlm_sae_make_desc((void *) session->cntrl_desc, (void *) session->pkt_desc, &cprm);
     ret = nlm_hal_crypto_send_request(crypto_get_vc(), nlm_fb_vc,
 	    (void *) session->cntrl_desc,
@@ -923,7 +923,7 @@ crypto_cipher_auth_op(struct crypto_session *session, struct crypto_param *cop)
 	    virt_to_phys(session->pkt_desc),
 	    &cprm, nlm_tx_id);
 
-    while(msg_hdl == 0);/*Wait till message ring handler is invoked by freeback message*/
+    while(msg_hdl[nlm_tx_id] == 0);/*Wait till message ring handler is invoked by freeback message*/
 
     ret = crypto_fb_msg_print_err(nlm_err_msg);
 
@@ -1008,12 +1008,11 @@ nlm_xlp_sae_msgring_handler(uint32_t vc, uint32_t src_id,
 	uint64_t msg0, uint64_t msg1,
 	uint64_t msg2, uint64_t msg3, void *data)
 {
+    uint64_t tx_id;
     if (size == 2) {
-	if (nlm_tx_id ==
-		nlm_hal_crypto_process_response(vc, code, src_id, msg0,
-		    msg1, &nlm_err_msg)) {
-	    msg_hdl = 1;
-	}
+	tx_id = nlm_hal_crypto_process_response(vc, code, src_id, msg0,
+		    msg1, &nlm_err_msg);
+	    msg_hdl[tx_id] = 1;
     }
 }
 
-- 
1.7.1

