From 74d0665d81564b0d87b6a0cb5926ac2357a06ab4 Mon Sep 17 00:00:00 2001
From: Alok Agrawat <aagrawat@netlogicmicro.com>
Date: Thu, 13 Oct 2011 11:51:26 +0530
Subject: modified linux crypto to accomodate changes done for bug 16851

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/drivers/crypto/sae/nlm_crypto.c b/drivers/crypto/sae/nlm_crypto.c
index a637464..e582256 100644
--- a/drivers/crypto/sae/nlm_crypto.c
+++ b/drivers/crypto/sae/nlm_crypto.c
@@ -830,10 +830,10 @@ crypto_cipher_auth_op(struct crypto_session *session, struct crypto_param *cop)
 
     i = 0;
     while (i < cop->nr_frags) {
-	cprm.src_phy[i].buf =
-	    (unsigned char *) virt_to_phys(cop->src[i].buf);
-	cprm.dst_phy[i].buf =
-	    (unsigned char *) virt_to_phys(cop->dst[i].buf);
+	cprm.src_phy[i].buf_phy =
+	    (uint64_t) virt_to_phys(cop->src[i].buf);
+	cprm.dst_phy[i].buf_phy =
+	    (uint64_t) virt_to_phys(cop->dst[i].buf);
 	cprm.src_phy[i].iov_len = cop->src[i].iov_len;
 	cprm.dst_phy[i].iov_len = cop->dst[i].iov_len;
 
@@ -874,7 +874,7 @@ crypto_cipher_auth_op(struct crypto_session *session, struct crypto_param *cop)
 
     cprm.cipher_key_len = cop->cipher_key_len;
     cprm.auth_key_len = cop->auth_key_len;
-    cprm.hash_dst_address = (unsigned char*)virt_to_phys(cop->hash_dst_address);
+    cprm.hash_dst_address = (uint64_t)virt_to_phys(cop->hash_dst_address);
     cprm.nr_frags = cop->nr_frags;
 
     cprm.designer_freeback_id = cop->designer_freeback_id;
diff --git a/drivers/crypto/sae/nlm_crypto_api.h b/drivers/crypto/sae/nlm_crypto_api.h
index 5dab998..8ceb735 100644
--- a/drivers/crypto/sae/nlm_crypto_api.h
+++ b/drivers/crypto/sae/nlm_crypto_api.h
@@ -83,6 +83,7 @@ THE POSSIBILITY OF SUCH DAMAGE.
 
 struct crypto_iovec {
     unsigned char *buf;
+    uint64_t buf_phy;
     uint32_t iov_len;
 };
 
-- 
1.7.1

