From f93b1543942f18c04501f387e21497c00c67d667 Mon Sep 17 00:00:00 2001
From: reshmic <reshmic@broadcom.com>
Date: Sat, 29 Sep 2012 18:37:28 +0530
Subject: kmod/sae : donot swap the 3des key in case of 2xx chip

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/drivers/crypto/sae/nlm_crypto.c b/drivers/crypto/sae/nlm_crypto.c
index 87f2d6e..fc1bb4b 100644
--- a/drivers/crypto/sae/nlm_crypto.c
+++ b/drivers/crypto/sae/nlm_crypto.c
@@ -29,6 +29,7 @@ THE POSSIBILITY OF SUCH DAMAGE.
 
 #include <asm/netlogic/hal/nlm_hal_fmn.h>
 #include <asm/netlogic/hal/nlm_hal_macros.h>
+#include <asm/netlogic/hal/nlm_hal_xlp_dev.h>
 #include <linux/crypto.h>
 #include "nlm_async.h"
 
@@ -71,6 +72,7 @@ static int xlp_sae_release(struct inode *, struct file *);
 struct nlm_crypto_stat crypto_stat[MAX_CPU];
 int crypto_vc_base;
 int crypto_vc_limit;
+int chip_feature = 0 ;
 
 
 
@@ -704,6 +706,14 @@ nlm_crypto_init(void)
     return ret;
 }
 
+static void  init_sae()
+{
+	if( is_nlm_xlp2xx())
+		chip_feature = (ZUC | DES3_KEY_SWAP); 
+	else
+		chip_feature = 0x0;
+}
+
     static int __init
 xlp_sae_init(void)
 {
@@ -722,6 +732,7 @@ xlp_sae_init(void)
     	return -1;
     }
     nlm_crypto_init();
+    init_sae();
     if(ipsec_async_vc != -1){
     	xlp_crypt_alg_init();
     	xlp_aead_alg_init();
diff --git a/drivers/crypto/sae/nlm_enc.c b/drivers/crypto/sae/nlm_enc.c
index 48aa677..4e032d5 100755
--- a/drivers/crypto/sae/nlm_enc.c
+++ b/drivers/crypto/sae/nlm_enc.c
@@ -109,11 +109,14 @@ xlp_setkey_des3(struct crypto_ablkcipher *tfm, const u8 * in_key, unsigned int l
 {
 	struct nlm_enc_ctx * nlm_ctx = nlm_crypto_ablkcipher_ctx_2(tfm); 
 	uint64_t key[3] ;
-	
-	memcpy(key,&in_key[16],8);
-        memcpy(&key[1],&in_key[8],8);
-        memcpy(&key[2],&in_key[0],8);
-
+	if( CHIP_SUPPORTS(DES3_KEY_SWAP) ) {
+		memcpy((uint8_t *)key,in_key,len);
+	}
+	else {
+		memcpy(key,&in_key[16],8);
+		memcpy(&key[1],&in_key[8],8);
+		memcpy(&key[2],&in_key[0],8);
+	}
 
 	nlm_crypto_fill_pkt_ctrl(&nlm_ctx->ctrl,0,0,0,cipher_alg,cipher_mode,0,(unsigned char*)&key[0],len,0,0);
 	crypto_ablkcipher_crt(tfm)->ivsize = cipher_mode_iv_len[cipher_alg][ cipher_mode];
-- 
1.7.1

