From adb594f3602bda212d7ff024bb8a6502f6bc09bf Mon Sep 17 00:00:00 2001
From: reshmic <reshmic@broadcom.com>
Date: Wed, 26 Dec 2012 14:16:59 +0530
Subject: crypto: Modified brcm headers

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/drivers/crypto/sae/nlm_aead.c b/drivers/crypto/sae/nlm_aead.c
index a5a29ea..67d481a 100755
--- a/drivers/crypto/sae/nlm_aead.c
+++ b/drivers/crypto/sae/nlm_aead.c
@@ -5,11 +5,13 @@
  * entities identified below. The licensing terms, warranty terms and other
  * terms specified in the header of the original work apply to this derived work
  *
- * This software is available to you under the terms of the GNU General Public
+ * #BRCM_1# */
+
+/* This software is available to you under the terms of the GNU General Public
  * License (GPL) Version 2, available from the file 
  * http://www.gnu.org/licenses/gpl-2.0.txt
- *
- * #BRCM_1# */
+*/
+
 #include <linux/rtnetlink.h>
 #include <crypto/algapi.h>
 #include <crypto/aes.h>
diff --git a/drivers/crypto/sae/nlm_async.h b/drivers/crypto/sae/nlm_async.h
index ca81f7f..6b9cb26 100644
--- a/drivers/crypto/sae/nlm_async.h
+++ b/drivers/crypto/sae/nlm_async.h
@@ -5,11 +5,13 @@
  * entities identified below. The licensing terms, warranty terms and other
  * terms specified in the header of the original work apply to this derived work
  *
- * This software is available to you under the terms of the GNU General Public
+ * #BRCM_1# */
+
+/* This software is available to you under the terms of the GNU General Public
  * License (GPL) Version 2, available from the file 
  * http://www.gnu.org/licenses/gpl-2.0.txt
- *
- * #BRCM_1# */
+*/
+
 #ifndef __NLM_ASYNC_H
 #define __NLM_ASYNC_H
 
diff --git a/drivers/crypto/sae/nlm_auth.c b/drivers/crypto/sae/nlm_auth.c
index cb6a36c..51f2f88 100644
--- a/drivers/crypto/sae/nlm_auth.c
+++ b/drivers/crypto/sae/nlm_auth.c
@@ -5,11 +5,13 @@
  * entities identified below. The licensing terms, warranty terms and other
  * terms specified in the header of the original work apply to this derived work
  *
- * This software is available to you under the terms of the GNU General Public
+ * #BRCM_1# */
+
+/* This software is available to you under the terms of the GNU General Public
  * License (GPL) Version 2, available from the file 
  * http://www.gnu.org/licenses/gpl-2.0.txt
- *
- * #BRCM_1# */
+*/
+
 #include <crypto/sha.h>
 #include <nlm_hal_fmn.h>
 #include <crypto/aes.h>
diff --git a/drivers/crypto/sae/nlm_crypto.c b/drivers/crypto/sae/nlm_crypto.c
index eb83b47..b61e70e 100644
--- a/drivers/crypto/sae/nlm_crypto.c
+++ b/drivers/crypto/sae/nlm_crypto.c
@@ -5,11 +5,12 @@
  * entities identified below. The licensing terms, warranty terms and other
  * terms specified in the header of the original work apply to this derived work
  *
+ * #BRCM_1# */
+/*
  * This software is available to you under the terms of the GNU General Public
  * License (GPL) Version 2, available from the file 
  * http://www.gnu.org/licenses/gpl-2.0.txt
- *
- * #BRCM_1# */
+ */
 #include <asm/netlogic/xlp.h>
 #include <asm/netlogic/msgring.h>
 #include <linux/proc_fs.h>
diff --git a/drivers/crypto/sae/nlm_enc.c b/drivers/crypto/sae/nlm_enc.c
index 1c217c0..7510859 100755
--- a/drivers/crypto/sae/nlm_enc.c
+++ b/drivers/crypto/sae/nlm_enc.c
@@ -5,11 +5,13 @@
  * entities identified below. The licensing terms, warranty terms and other
  * terms specified in the header of the original work apply to this derived work
  *
- * This software is available to you under the terms of the GNU General Public
+ * #BRCM_1# */
+
+/* This software is available to you under the terms of the GNU General Public
  * License (GPL) Version 2, available from the file 
  * http://www.gnu.org/licenses/gpl-2.0.txt
- *
- * #BRCM_1# */
+*/
+
 #include <linux/crypto.h>
 #include <crypto/aes.h>
 #include <crypto/des.h>
diff --git a/drivers/crypto/sae/nlmcrypto_ifc.h b/drivers/crypto/sae/nlmcrypto_ifc.h
index c23e4e3..1a17cc1 100644
--- a/drivers/crypto/sae/nlmcrypto_ifc.h
+++ b/drivers/crypto/sae/nlmcrypto_ifc.h
@@ -5,11 +5,13 @@
  * entities identified below. The licensing terms, warranty terms and other
  * terms specified in the header of the original work apply to this derived work
  *
- * This software is available to you under the terms of the GNU General Public
+ * #BRCM_1# */
+
+/* This software is available to you under the terms of the GNU General Public
  * License (GPL) Version 2, available from the file 
  * http://www.gnu.org/licenses/gpl-2.0.txt
- *
- * #BRCM_1# */
+*/
+
 #ifndef _NLM_CRYPTO_IFC_H
 #define _NLM_CRYPTO_IFC_H
 extern void *linuxu_shvaddr;
-- 
1.7.1

