From 526f856c710df4c05d99abbea47605bd9151a20f Mon Sep 17 00:00:00 2001
From: reshmic <reshmic@broadcom.com>
Date: Mon, 1 Apr 2013 19:57:17 +0530
Subject: crypto: Renamed nlm_hal_get_chip_feature to nlm_hal_get_sae_chip_feature

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/drivers/crypto/sae/nlm_crypto.c b/drivers/crypto/sae/nlm_crypto.c
index be6374d..f66777b 100644
--- a/drivers/crypto/sae/nlm_crypto.c
+++ b/drivers/crypto/sae/nlm_crypto.c
@@ -713,7 +713,7 @@ xlp_sae_init(void)
     }
     nlm_crypto_init();
     init_sae();
-    nlm_crypto_chip_features = nlm_hal_get_chip_feature();
+    nlm_crypto_chip_features = nlm_hal_get_sae_chip_feature();
     if(ipsec_async_vc != -1){
     	xlp_crypt_alg_init();
     	xlp_aead_alg_init();
-- 
1.7.1

