From a6a275712ac93eae117bc34b00a5a0a5c4ada912 Mon Sep 17 00:00:00 2001
From: reshmic <reshmic@broadcom.com>
Date: Thu, 13 Jun 2013 16:43:13 +0530
Subject: sae : crypto_get_soc_vc_numbers to get vc base and limit

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/drivers/crypto/sae/nlm_crypto.c b/drivers/crypto/sae/nlm_crypto.c
index ea0c711..2001ab8 100644
--- a/drivers/crypto/sae/nlm_crypto.c
+++ b/drivers/crypto/sae/nlm_crypto.c
@@ -43,8 +43,8 @@ int cryptosoc_lib_priv_finish(void){
 }
 int crypto_get_soc_vc_numbers(enum cryptosoc_soc_type soc, int *base_vc, int *lmt_vc)
 {
-	*base_vc = 276;
-	*lmt_vc = 279;
+	if ( soc == CRYPTOSOC_SAE )
+		nlm_hal_get_crypto_vc_nums(base_vc,lmt_vc);
 	return 0;
 }
 
@@ -369,7 +369,7 @@ nlm_crypto_init(void)
     }
 
 #endif
-   nlm_hal_get_crypto_vc_nums(&crypto_vc_base, &crypto_vc_limit); 
+   crypto_get_soc_vc_numbers(CRYPTOSOC_SAE,&crypto_vc_base, &crypto_vc_limit); 
    printf("crypto vc base %d\n",crypto_vc_base);
 
     if (register_xlp_msgring_handler
@@ -393,7 +393,7 @@ static void  init_sae(void)
 	for_each_node(node) {
 
 		nlm_hal_set_sae_freq(node, freq);
-		//nlm_hal_set_sae_engine_sel(node);
+		nlm_hal_set_sae_engine_sel(node);
 	}
 
 }
-- 
1.7.1

