From b1a5c60888f1c1ec3f57afd6c6bc8fcb45c39c80 Mon Sep 17 00:00:00 2001
From: Sandip Matte <sandipm@broadcom.com>
Date: Tue, 20 May 2014 12:08:24 +0530
Subject: crypto: Support linux crypto driver on 5xx chip.

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/drivers/crypto/sae/nlm_crypto.c b/drivers/crypto/sae/nlm_crypto.c
index f9e3a74..60c15bb 100644
--- a/drivers/crypto/sae/nlm_crypto.c
+++ b/drivers/crypto/sae/nlm_crypto.c
@@ -38,7 +38,7 @@ extern int cryptosoc_ptype_gen;
 
 int cryptosoc_lib_priv_init(struct cryptosoc_lib_params *params)
 {
-        if (is_nlm_xlp9xx())
+	if (is_nlm_xlp9xx() || is_nlm_xlp5xx())
                 cryptosoc_ptype_gen = CRYPTOSOC_PTYPE_IS_GEN2;
         else
                 cryptosoc_ptype_gen = CRYPTOSOC_PTYPE_IS_GEN1;
@@ -435,7 +435,7 @@ xlp_sae_init(void)
 
     printk(KERN_ERR ",\n XLP SAE/Crypto Initialization \n");
     printk("ipsec_sync_vc %d ipsec_async_vc %d\n",ipsec_sync_vc,ipsec_async_vc);
-    if (is_nlm_xlp9xx())
+    if (is_nlm_xlp9xx() || is_nlm_xlp5xx())
         cryptosoc_ptype_gen = CRYPTOSOC_PTYPE_IS_GEN2;
     else
         cryptosoc_ptype_gen = CRYPTOSOC_PTYPE_IS_GEN1;
-- 
1.7.1

