From 1f59e61a37b26a203f54511831e601bb8406b5e8 Mon Sep 17 00:00:00 2001
From: Pranav Bhalerao <pranavb@broadcom.com>
Date: Mon, 8 Dec 2014 12:35:39 +0530
Subject: kmod/crypto:In multinode setup init_sae is trying to initialize all available node irrespective of its status whether its online or offline. System is initialized with total 64 nodes. SAE must be initialized only for online nodes.

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/drivers/crypto/sae/nlm_crypto.c b/drivers/crypto/sae/nlm_crypto.c
index 94855db..c39920e 100644
--- a/drivers/crypto/sae/nlm_crypto.c
+++ b/drivers/crypto/sae/nlm_crypto.c
@@ -405,8 +405,7 @@ static void  init_sae(void)
 
 	freq = nlm_hal_get_fdt_freq(fdt, NLM_SAE);
 
-	for_each_node(node) {
-
+	for_each_online_node(node) {
 		nlm_hal_set_sae_freq(node, freq);
 		nlm_hal_set_sae_engine_sel(node);
 	}
-- 
1.7.1

