From f7e71bfe1d60dce65b37a8640be7166c6891eda2 Mon Sep 17 00:00:00 2001
From: Hareesh R <hareesh@netlogicmicro.com>
Date: Fri, 12 Aug 2011 15:02:33 +0530
Subject: crypto fdctxt cleanup support

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/netlogic/lib/seclib/nlmcrypto_ifc.h b/arch/mips/netlogic/lib/seclib/nlmcrypto_ifc.h
index 7621a47..876bdcc 100644
--- a/arch/mips/netlogic/lib/seclib/nlmcrypto_ifc.h
+++ b/arch/mips/netlogic/lib/seclib/nlmcrypto_ifc.h
@@ -57,15 +57,6 @@ static inline void *crypto_phys_to_virt(unsigned long long paddr)
 static inline int nlm_crypto_app_lib_init(int *sae_rx_vc, int *sae_rx_sync_vc, void **pcpu_sync_lock_ptr)
 {
 	extern int nlm_crypto_app_linuxu_init(int *sae_rx_vc, int *sae_rx_sync_vc, void **pcpu_sync_lock_ptr);
-#ifdef NLM_CRYPTO_HELINUX
-	extern void fdt_get_sae_vc_map(int *asyncvc, int *syncvc);
-	unsigned long long virtstart, physstart;
-	if (contig_pinfo(&virtstart, &physstart) < 0)
-		return -1;
-	linuxu_shvaddr = (void *)virtstart;
-	linuxu_shoff = physstart;
-	fdt_get_sae_vc_map(sae_rx_vc, sae_rx_sync_vc);
-#endif
 	return nlm_crypto_app_linuxu_init(sae_rx_vc, sae_rx_sync_vc, pcpu_sync_lock_ptr);
 }
 
-- 
1.7.1

