From 722cd1823675b29db4c5be7269c7d36246034b33 Mon Sep 17 00:00:00 2001
From: Jiang Lu <lu.jiang@windriver.com>
Date: Thu, 7 Aug 2014 12:36:35 +0800
Subject: netlogic:enable SELECT_MEMORY_MODEL for xlp

Signed-off-by: Jiang Lu <lu.jiang@windriver.com>
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/netlogic/Kconfig b/arch/mips/netlogic/Kconfig
index 34288ad..5880e79 100644
--- a/arch/mips/netlogic/Kconfig
+++ b/arch/mips/netlogic/Kconfig
@@ -1,9 +1,13 @@
 if NLM_XLP_BOARD || NLM_XLR_BOARD
 
+config ARCH_SELECT_MEMORY_MODEL
+	bool
+
 if NLM_XLP_BOARD
 config DT_XLP_EVP
 	bool "Built-in device tree for XLP EVP boards"
 	default y
+	select ARCH_SELECT_MEMORY_MODEL
 	help
 	  Add an FDT blob for XLP EVP boards into the kernel.
 	  This DTB will be used if the firmware does not pass in a DTB
-- 
1.7.1

