From 3d6fd43506d51f51b099c46680458e38be9c26ae Mon Sep 17 00:00:00 2001
From: Quanyang Wang <quanyang.wang@windriver.com>
Date: Fri, 3 Apr 2015 12:25:14 +0800
Subject: bcm-xlp: fix Makefile typo

Fix a typo in Makefile

Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/boot/Makefile b/arch/mips/boot/Makefile
index 784a719..dc0fdac 100644
--- a/arch/mips/boot/Makefile
+++ b/arch/mips/boot/Makefile
@@ -42,7 +42,7 @@ $(obj)/vmlinux.srec: $(VMLINUX) FORCE
 	$(call if_changed,srec)
 
 # Rule to build device tree blobs
-DTS_FLAGS ?= -p 1024
+DTC_FLAGS ?= -p 1024
 dtstree		:= $(srctree)/$(src)/dts
 
 $(obj)/%.dtb: $(src)/dts/%.dts FORCE
-- 
1.7.1

