From 663a5459ab1e270e21b3aa30b53df5ff4db6d6a7 Mon Sep 17 00:00:00 2001
From: Mark Asselstine <mark.asselstine@windriver.com>
Date: Wed, 16 Jun 2010 08:35:45 -0700
Subject: [PATCH] grsec: quiet preemption check warning in gr_log_resource()

A call to preempt_enable_no_resched() in gr_log_resource() was flagged by the
additional checks brought in with the preempt_irq commit
[sched: Debug missed preemption checks]

resulting in

------------[ cut here ]------------
WARNING: at /home/masselst/git/wrlinux-x/builds/cav_octeon_cn56xx/build/linux/kernel/sched.c:3652 preempt_enable_no_resched+0x60/0x70()
BUG: init:1 task might have lost a preemption check!
Modules linked in:
Call Trace:
[<ffffffff8110c7e8>] dump_stack+0x8/0x34
[<ffffffff81150668>] warn_slowpath_common+0x70/0xb0
[<ffffffff811506fc>] warn_slowpath_fmt+0x34/0x40
[<ffffffff81144ad0>] preempt_enable_no_resched+0x60/0x70
[<ffffffff81341c78>] gr_learn_resource+0x48/0x250
[<ffffffff811df530>] may_expand_vm+0x38/0x58
[<ffffffff811e1e2c>] mmap_region+0x20c/0x608
[<ffffffff8124d954>] elf_map+0x15c/0x1b0
[<ffffffff8124e1a4>] load_elf_binary+0x7fc/0x18c0
[<ffffffff812044dc>] search_binary_handler+0x334/0x490
[<ffffffff81205e7c>] do_execve+0x3f4/0x508
[<ffffffff811209bc>] sys_execve+0x5c/0x88
[<ffffffff81102bc4>] handle_sys64+0x44/0x60

---[ end trace 96ebf50545eab5a7 ]---

since the original code seemed to want to ensure there was no schedule() made at
this point in time the call has been converted to __preempt_enable_no_resched()
to quiet the kernel WARNING.

Signed-off-by: Mark Asselstine <mark.asselstine@windriver.com>
---
 grsecurity/gracl_res.c |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/grsecurity/gracl_res.c b/grsecurity/gracl_res.c
index dfa3049..10b4880 100644
--- a/grsecurity/gracl_res.c
+++ b/grsecurity/gracl_res.c
@@ -56,7 +56,7 @@ gr_log_resource(const struct task_struct *task,
 		      (!gt && wanted >= task->signal->rlim[res].rlim_cur)) &&
 		     task->signal->rlim[res].rlim_cur != RLIM_INFINITY))
 		gr_log_res_ulong2_str(GR_DONT_AUDIT, GR_RESOURCE_MSG, task, wanted, restab_log[res], task->signal->rlim[res].rlim_cur);
-	preempt_enable_no_resched();
+	__preempt_enable_no_resched();
 
 	return;
 out_rcu_unlock:
-- 
1.6.5.2

