From 5b2151faf7ca8717895bd880dd87668f2228c838 Mon Sep 17 00:00:00 2001
From: Bruce Ashfield <bruce.ashfield@windriver.com>
Date: Sun, 30 May 2010 04:22:25 -0400
Subject: [PATCH] grsec: lttng: replace user_mode_vm with user_mode

Signed-off-by: Bruce Ashfield <bruce.ashfield@windriver.com>
---
 arch/x86/kernel/immediate.c |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/arch/x86/kernel/immediate.c b/arch/x86/kernel/immediate.c
index 5d77ac3..ef58735 100644
--- a/arch/x86/kernel/immediate.c
+++ b/arch/x86/kernel/immediate.c
@@ -156,7 +156,7 @@ static int imv_notifier(struct notifier_block *nb,
 	enum die_val die_val = (enum die_val) val;
 	struct die_args *args = data;
 
-	if (!args->regs || user_mode_vm(args->regs))
+	if (!args->regs || user_mode(args->regs))
 		return NOTIFY_DONE;
 
 	if (die_val == DIE_INT3) {
-- 
1.6.5.2

