From 77b849dbb60452cc24e9dc3ad254070696a4851d Mon Sep 17 00:00:00 2001
From: Phil Elwell <phil@raspberrypi.org>
Date: Fri, 7 Oct 2016 12:37:09 +0100
Subject: [PATCH 104/203] scripts/mkknlimg: Change string for DDT detection

The old "of_overlay_apply" string does not appear in 4.8 kernel builds.
"of_cfs_init" is both present and a more accurate indication of support
for dynamically loading overlays (although the trailer is now of little
practical use and the firmware assumes DT support and will look for
both .dtbo and -overlay.dtb overlays).

Signed-off-by: Phil Elwell <phil@raspberrypi.org>
[Xulin: Original patch taken from
https://github.com/raspberrypi/linux.git branch rpi-4.8.y]
Signed-off-by: Xulin Sun <xulin.sun@windriver.com>
---
 scripts/mkknlimg |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/scripts/mkknlimg b/scripts/mkknlimg
index 78c5845..a21f7e3 100755
--- a/scripts/mkknlimg
+++ b/scripts/mkknlimg
@@ -77,7 +77,7 @@ my $wanted_strings =
 	'brcm,bcm2709' => FLAG_PI | FLAG_DTOK | FLAG_270X,
 	'brcm,bcm2835' => FLAG_PI | FLAG_DTOK | FLAG_283X,
 	'brcm,bcm2836' => FLAG_PI | FLAG_DTOK | FLAG_283X,
-	'of_overlay_apply' => FLAG_DTOK | FLAG_DDTK,
+	'of_cfs_init' => FLAG_DTOK | FLAG_DDTK,
 };
 
 my $res = try_extract($kernel_file, $tmpfile1);
-- 
1.7.5.4

