From c58d6da2856f7a45a2a06f24f47f2c5b63b866b0 Mon Sep 17 00:00:00 2001
From: Yang Shi <yang.shi@windriver.com>
Date: Tue, 13 Jan 2015 12:48:33 -0800
Subject: [PATCH 354/524] grsec: changes to fs_configfs from
 grsecurity-3.0-3.14.28-201501120819

Signed-off-by: Yang Shi <yang.shi@windriver.com>
---
 fs/configfs/dir.c | 10 ++++++++--
 1 file changed, 8 insertions(+), 2 deletions(-)

diff --git a/fs/configfs/dir.c b/fs/configfs/dir.c
index e081acb..911df21 100644
--- a/fs/configfs/dir.c
+++ b/fs/configfs/dir.c
@@ -1548,7 +1548,8 @@ static int configfs_readdir(struct file *file, struct dir_context *ctx)
 	}
 	for (p = q->next; p != &parent_sd->s_children; p = p->next) {
 		struct configfs_dirent *next;
-		const char *name;
+		const unsigned char * name;
+		char d_name[sizeof(next->s_dentry->d_iname)];
 		int len;
 		struct inode *inode = NULL;
 
@@ -1557,7 +1558,12 @@ static int configfs_readdir(struct file *file, struct dir_context *ctx)
 			continue;
 
 		name = configfs_get_name(next);
-		len = strlen(name);
+		if (next->s_dentry && name == next->s_dentry->d_iname) {
+			len =  next->s_dentry->d_name.len;
+			memcpy(d_name, name, len);
+			name = d_name;
+		} else
+			len = strlen(name);
 
 		/*
 		 * We'll have a dentry and an inode for
-- 
2.0.2

