From 20fec8a19a01b4d8730870c16bbbf7e126f4ee1d Mon Sep 17 00:00:00 2001
From: Yang Shi <yang.shi@windriver.com>
Date: Tue, 13 Jan 2015 12:49:45 -0800
Subject: [PATCH 515/524] grsec: changes to sound_soc from
 grsecurity-3.0-3.14.28-201501120819

Signed-off-by: Yang Shi <yang.shi@windriver.com>
---
 sound/soc/soc-core.c | 6 ++++--
 1 file changed, 4 insertions(+), 2 deletions(-)

diff --git a/sound/soc/soc-core.c b/sound/soc/soc-core.c
index fe1df50..09d3be6 100644
--- a/sound/soc/soc-core.c
+++ b/sound/soc/soc-core.c
@@ -2254,8 +2254,10 @@ int snd_soc_set_ac97_ops_of_reset(struct snd_ac97_bus_ops *ops,
 	if (ret)
 		return ret;
 
-	ops->warm_reset = snd_soc_ac97_warm_reset;
-	ops->reset = snd_soc_ac97_reset;
+	pax_open_kernel();
+	*(void **)&ops->warm_reset = snd_soc_ac97_warm_reset;
+	*(void **)&ops->reset = snd_soc_ac97_reset;
+	pax_close_kernel();
 
 	snd_ac97_rst_cfg = cfg;
 	return 0;
-- 
2.0.2

