From 836ffc3874f4fba837c9912a7e8e82f75453eb07 Mon Sep 17 00:00:00 2001
From: Yang Shi <yang.shi@windriver.com>
Date: Tue, 13 Jan 2015 12:48:55 -0800
Subject: [PATCH 405/524] grsec: changes to include_net_9p from
 grsecurity-3.0-3.14.28-201501120819

Signed-off-by: Yang Shi <yang.shi@windriver.com>
---
 include/net/9p/transport.h | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/include/net/9p/transport.h b/include/net/9p/transport.h
index 9a36d92..0aafe2a 100644
--- a/include/net/9p/transport.h
+++ b/include/net/9p/transport.h
@@ -60,7 +60,7 @@ struct p9_trans_module {
 	int (*cancel) (struct p9_client *, struct p9_req_t *req);
 	int (*zc_request)(struct p9_client *, struct p9_req_t *,
 			  char *, char *, int , int, int, int);
-};
+} __do_const;
 
 void v9fs_register_trans(struct p9_trans_module *m);
 void v9fs_unregister_trans(struct p9_trans_module *m);
-- 
2.0.2

