From 6e43cbb219df8c1d2434ba5292275beb35928461 Mon Sep 17 00:00:00 2001
From: Yang Shi <yang.shi@windriver.com>
Date: Tue, 13 Jan 2015 12:49:37 -0800
Subject: [PATCH 497/524] grsec: changes to security_selinux_include from
 grsecurity-3.0-3.14.28-201501120819

Signed-off-by: Yang Shi <yang.shi@windriver.com>
---
 security/selinux/include/xfrm.h | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/security/selinux/include/xfrm.h b/security/selinux/include/xfrm.h
index 9f05847..7933395 100644
--- a/security/selinux/include/xfrm.h
+++ b/security/selinux/include/xfrm.h
@@ -46,7 +46,7 @@ static inline void selinux_xfrm_notify_policyload(void)
 {
 	struct net *net;
 
-	atomic_inc(&flow_cache_genid);
+	atomic_inc_unchecked(&flow_cache_genid);
 	rtnl_lock();
 	for_each_net(net)
 		rt_genid_bump_all(net);
-- 
2.0.2

