From 188e7360150f7aa943aab1771613165c12fdaabc Mon Sep 17 00:00:00 2001
From: Bruce Ashfield <bruce.ashfield@windriver.com>
Date: Sun, 30 May 2010 04:32:51 -0400
Subject: [PATCH] grsec: immediate: adjust to grsec module structures

Signed-off-by: Bruce Ashfield <bruce.ashfield@windriver.com>
---
 kernel/module.c |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/kernel/module.c b/kernel/module.c
index b8603fe..83a12dd 100644
--- a/kernel/module.c
+++ b/kernel/module.c
@@ -2721,7 +2721,7 @@ SYSCALL_DEFINE3(init_module, void __user *, umod,
 	module_put(mod);
 #ifdef USE_IMMEDIATE
 	imv_unref(mod->immediate, mod->immediate + mod->num_immediate,
-		mod->module_init, mod->init_size);
+		mod->module_init_rw, mod->init_size_rw);
 #endif
 	trim_init_extable(mod);
 #ifdef CONFIG_KALLSYMS
-- 
1.6.5.2

