From 2fc702a0f1f0bbb9f4029957f901368dfd6bb2e9 Mon Sep 17 00:00:00 2001
From: Joe MacDonald <joe.macdonald@windriver.com>
Date: Tue, 17 Jul 2012 10:41:11 -0400
Subject: [PATCH 2/6] pax: Disable gcc plugins

The Wind River toolchain doesn't support gcc plugins, so we update the
default behavior to not build the pax plugins and still allow them to be
turned on by passing ENABLE_PAX_PLUGINS and ENABLE_PAX_CONSTIFY_PLUGIN
respectively.

Signed-off-by: Joe MacDonald <joe.macdonald@windriver.com>
---
 Makefile |    4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/Makefile b/Makefile
index ba3f29a..75ce9d2 100644
--- a/Makefile
+++ b/Makefile
@@ -565,10 +565,10 @@ else
 KBUILD_CFLAGS	+= -O2
 endif
 
-ifndef DISABLE_PAX_PLUGINS
+ifdef ENABLE_PAX_PLUGINS
 PLUGINCC := $(shell $(CONFIG_SHELL) $(srctree)/scripts/gcc-plugin.sh "$(HOSTCC)" "$(HOSTCXX)" "$(CC)")
 ifneq ($(PLUGINCC),)
-ifndef DISABLE_PAX_CONSTIFY_PLUGIN
+ifdef ENABLE_PAX_CONSTIFY_PLUGIN
 ifndef CONFIG_UML
 CONSTIFY_PLUGIN_CFLAGS := -fplugin=$(objtree)/tools/gcc/constify_plugin.so -DCONSTIFY_PLUGIN
 endif
-- 
1.7.9.7

