From 918833c4949c05374e0f989654d0b9c385857b7f Mon Sep 17 00:00:00 2001
From: Paul Gortmaker <paul.gortmaker@windriver.com>
Date: Wed, 15 Jan 2014 19:01:38 -0500
Subject: [PATCH 1/3] pax: disable the plugin checks for now

Need to determine if supporting plugins in a cross compile
environment is worthwhile.  In the interim, the host checks
are counterproductive in forcing a failure and suggesting
that plugins get installed.

Signed-off-by: Paul Gortmaker <paul.gortmaker@windriver.com>
---
 Makefile | 1 +
 1 file changed, 1 insertion(+)

diff --git a/Makefile b/Makefile
index 280465e..c41adfd 100644
--- a/Makefile
+++ b/Makefile
@@ -577,6 +577,7 @@ else
 KBUILD_CFLAGS	+= -O2
 endif
 
+DISABLE_PAX_PLUGINS := yes_I_know_what_Im_doing
 ifndef DISABLE_PAX_PLUGINS
 ifeq ($(call cc-ifversion, -ge, 0408, y), y)
 PLUGINCC := $(shell $(CONFIG_SHELL) $(srctree)/scripts/gcc-plugin.sh "$(HOSTCXX)" "$(HOSTCXX)" "$(CC)")
-- 
1.8.5.2.233.g932f7e4

