From 61fd41848126cbae678679794bdac26a014fe5a0 Mon Sep 17 00:00:00 2001
From: Shinya Kuribayashi <shinya.kuribayashi@necel.com>
Date: Tue, 28 Apr 2009 14:17:47 +0900
Subject: [PATCH] net: nec_candy: Use print_mac()

Signed-off-by: Shinya Kuribayashi <shinya.kuribayashi@necel.com>
---
 drivers/net/nec_candy.c |    8 +++-----
 1 files changed, 3 insertions(+), 5 deletions(-)

diff --git a/drivers/net/nec_candy.c b/drivers/net/nec_candy.c
index c19f6d4..1543f77 100644
--- a/drivers/net/nec_candy.c
+++ b/drivers/net/nec_candy.c
@@ -1763,6 +1763,7 @@ static int nec_candy_probe(struct device *dev)
 	struct candy_private *pp;
 	struct nec_candy_platform_data *pdata = dev->platform_data;
 	struct platform_device *pdev = to_platform_device(dev);
+	DECLARE_MAC_BUF(mac);
 
 	printk(KERN_INFO "%s\n", version);
 
@@ -1933,12 +1934,9 @@ static int nec_candy_probe(struct device *dev)
 		goto err6;
 
 	printk(KERN_INFO DRV_NAME ": Probe candy chip at "
-	       "0x%08x, irq %d, Phy ID 0x%08x, "
-	       "MAC Addr %02x:%02x:%02x:%02x:%02x:%02x\n",
+	       "0x%08x, irq %d, Phy ID 0x%08x, MAC Addr %s\n",
 	       res->start, ndev->irq, pp->mii_if.phy_id,
-	       ndev->dev_addr[0], ndev->dev_addr[1],
-	       ndev->dev_addr[2], ndev->dev_addr[3],
-	       ndev->dev_addr[4], ndev->dev_addr[5]);
+	       print_mac(mac, ndev->dev_addr));
 
 	return 0;
 
-- 
1.6.0.4

