From 00a50778318fdb2714caa0cc8c732ea78130b23e Mon Sep 17 00:00:00 2001
From: Bruce Ashfield <bruce.ashfield@windriver.com>
Date: Wed, 2 Jun 2010 00:27:39 -0400
Subject: [PATCH 05/12] revoke: use proper fync op

Signed-off-by: Bruce Ashfield <bruce.ashfield@windriver.com>
---
 fs/revoke.c |    3 ++-
 1 files changed, 2 insertions(+), 1 deletions(-)

diff --git a/fs/revoke.c b/fs/revoke.c
index 9101f5e..ac666d2 100644
--- a/fs/revoke.c
+++ b/fs/revoke.c
@@ -17,6 +17,7 @@
 #include <linux/mount.h>
 #include <linux/namei.h>
 #include <linux/fdtable.h>
+#include <linux/slab.h>
 
 static void revoke_aliases(struct inode *inode)
 {
@@ -446,6 +447,6 @@ asmlinkage long sys_revokeat(int dfd, const char __user *filename)
 
 int generic_file_revoke(struct file *file)
 {
-	return do_fsync(file, 1);
+	return vfs_fsync(file, file->f_path.dentry, 1);
 }
 EXPORT_SYMBOL(generic_file_revoke);
-- 
1.6.5.2

