From 70178e3ede1c81dc0750ff191db801bfe2d1d11f Mon Sep 17 00:00:00 2001
From: Jonas Rendel <Jonas.Rendel@windriver.com>
Date: Wed, 26 Nov 2008 09:36:00 +0100
Subject: [PATCH] enable cpu isolation on mips architecture

---
 arch/mips/Kconfig |    1 +
 1 files changed, 1 insertions(+), 0 deletions(-)

diff --git a/arch/mips/Kconfig b/arch/mips/Kconfig
index 1e06d23..dc791fa 100644
--- a/arch/mips/Kconfig
+++ b/arch/mips/Kconfig
@@ -1812,6 +1812,7 @@ config HZ
 	default 1024 if HZ_1024
 
 source "kernel/Kconfig.preempt"
+source "kernel/Kconfig.cpuisol"
 
 config MIPS_INSANE_LARGE
 	bool "Support for large 64-bit configurations"
-- 
1.5.6

