From ae4a64a2541d054f67b01c966b6e5b638cfcab57 Mon Sep 17 00:00:00 2001
From: Jonas Rendel <Jonas.Rendel@windriver.com>
Date: Wed, 26 Nov 2008 09:36:00 +0100
Subject: [PATCH 8/8] enable cpu isolation on mips architecture

---
 arch/mips/Kconfig |    1 +
 1 files changed, 1 insertions(+), 0 deletions(-)

diff --git a/arch/mips/Kconfig b/arch/mips/Kconfig
index aa35f05..53bcc69 100644
--- a/arch/mips/Kconfig
+++ b/arch/mips/Kconfig
@@ -2036,6 +2036,7 @@ config HZ
 	default 1024 if HZ_1024
 
 source "kernel/Kconfig.preempt"
+source "kernel/Kconfig.cpuisol"
 
 config MIPS_INSANE_LARGE
 	bool "Support for large 64-bit configurations"
-- 
1.7.0.4

