From fc24575a3c902c6fabb071f80eb91c9ea43e57eb Mon Sep 17 00:00:00 2001
From: Jonas Rendel <Jonas.Rendel@windriver.com>
Date: Wed, 26 Nov 2008 09:36:00 +0100
Subject: [PATCH 8/8] cpuisol: enable cpu isolation on mips architecture

---
 arch/mips/Kconfig |    1 +
 1 files changed, 1 insertions(+), 0 deletions(-)

diff --git a/arch/mips/Kconfig b/arch/mips/Kconfig
index 838a9e5..a95d6b7 100644
--- a/arch/mips/Kconfig
+++ b/arch/mips/Kconfig
@@ -2044,6 +2044,7 @@ config HZ
 	default 1024 if HZ_1024
 
 source "kernel/Kconfig.preempt"
+source "kernel/Kconfig.cpuisol"
 
 config MIPS_INSANE_LARGE
 	bool "Support for large 64-bit configurations"
-- 
1.6.5.2

