From fec9d7c93f2270ea521b9503a8acc73a3f32a783 Mon Sep 17 00:00:00 2001
From: Yonghong Song <ysong@broadcom.com>
Date: Thu, 9 May 2013 10:45:56 -0700
Subject: [PATCH] kvm: fix guest multi-core booting

Commit 9c76100eb1a50f6ce0e74a73e6164afeece11c30 from Broadcom SDK 3.0.2

Signed-off-by: Yonghong Song <ysong@broadcom.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/netlogic/kvm/kvm_sysmgt.c b/arch/mips/netlogic/kvm/kvm_sysmgt.c
index 1fcbd57..015f5b5 100644
--- a/arch/mips/netlogic/kvm/kvm_sysmgt.c
+++ b/arch/mips/netlogic/kvm/kvm_sysmgt.c
@@ -68,15 +68,21 @@ void kvm_handle_pcie_sysmgt(struct pt_regs *regs, unsigned long write,
 			return;
 		}
 	} else if (rindex == 0x42) {
-		/* cpu reset register */
+		/* cpu reset register*/
 		if (write) {
-			int i, v;
+			int i, nval, oval;
 
-			arch->sysmgt.regs[rindex] = regs->regs[reg_num];
-			v = regs->regs[reg_num] & 0xff;
-			for (i = 1; i < 8; i++) {
-				unsigned int core_enable = !(v & (1 << i));
-				if (core_enable) {
+			oval = arch->sysmgt.regs[rindex];
+			nval = regs->regs[reg_num];
+			arch->sysmgt.regs[rindex] = nval;
+
+			for (i = 1; i < (KVM_MAX_VCPUS >> 2); i++) {
+				/* if it is already out of reset, skip */
+				if (!(oval & (1 << i)))
+					continue;
+
+				/* not out of reset yet, should it? */
+				if (!(nval & (1 << i))) {
 					unsigned int val = i << 2, badinstr, epc_badinstr;
 					val |= (KVM_EXIT_ENABLE_CORE << 24);
 
-- 
1.9.1

