From 0bc31b2d51f4fa94fbb58ca882fcfbf57f502377 Mon Sep 17 00:00:00 2001
From: Yonghong Song <ysong@broadcom.com>
Date: Wed, 31 Jul 2013 14:20:42 -0700
Subject: [PATCH] kvm: uart: still indicate tx interrupt when rx buffer is NUL

Commit 755f246833ffbaa834b40f5b84348820fdef0699 from Broadcom SDK 3.0.2

Signed-off-by: Yonghong Song <ysong@broadcom.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/netlogic/kvm/kvm_uart.c b/arch/mips/netlogic/kvm/kvm_uart.c
index 9c06b3a..a67e32b 100644
--- a/arch/mips/netlogic/kvm/kvm_uart.c
+++ b/arch/mips/netlogic/kvm/kvm_uart.c
@@ -149,9 +149,10 @@ void kvm_handle_pcie_uart(struct pt_regs *regs, unsigned long write,
 					arch->uart.lsr &= 0xfe;
 
 					/* clear "receive data ready" and interrupt */
-					if ((arch->uart.iir & 0xe) == 4) {
+					if ((arch->uart.ier & 0x2) == 0) {
 						arch->uart.iir = 0x1;
-					}
+					} else
+						arch->uart.iir = 0x2;
 				}
 			}
 		}
@@ -171,16 +172,16 @@ void kvm_handle_pcie_uart(struct pt_regs *regs, unsigned long write,
 		} else {
 			if (arch->uart.lcr & (1 << 7))
 				regs->regs[reg_num] = arch->uart.dlb2;
-			else
+			else {
 				regs->regs[reg_num] = arch->uart.ier;
+			}
 		}
 	} else if (rindex == 0x42) {
 		/* uart interrupt source /fifo control */
 		if (write)
 			; /* do nothing */
 		else {
-			unsigned int tmp = arch->uart.iir;
-			regs->regs[reg_num] = tmp;
+			regs->regs[reg_num] = arch->uart.iir;
 		}
 	} else if (rindex == 0x43) {
 		/* line control */
-- 
1.9.1

