From ef5335ce3c856066da1e27fa5195ad033cff3f67 Mon Sep 17 00:00:00 2001
From: Subhendu Sekhar Behera <sbehera@broadcom.com>
Date: Mon, 13 Jan 2014 20:19:52 +0530
Subject: [PATCH] kvm: support for outta tree buildling

Commit e37846a58ed7910c7703ca4010ee7ecddbfb6253 from Broadcom SDK 3.0.2

Signed-off-by: Subhendu Sekhar Behera <sbehera@broadcom.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/kvm/Makefile b/arch/mips/kvm/Makefile
index 59cd048..7432246 100644
--- a/arch/mips/kvm/Makefile
+++ b/arch/mips/kvm/Makefile
@@ -1,6 +1,8 @@
 # Makefile for kernel virtual machines on xlp
+obj-$(CONFIG_KVM) := kvm_xlp.o
 
 common-objs = $(addprefix ../../../virt/kvm/, kvm_main.o)
 
-obj-$(CONFIG_KVM) += $(common-objs) kvm.o xlp.o context_switch.o
+kvm_xlp-y += $(common-objs) kvm.o xlp.o context_switch.o
+
 obj-$(CONFIG_IOMMU_API) += kvm_iommu.o
-- 
1.9.1

