From 4bee20ebce7ca0a90aadd37e508b6a51786abe80 Mon Sep 17 00:00:00 2001
From: Yonghong Song <ysong@broadcom.com>
Date: Tue, 18 Jun 2013 14:33:15 -0700
Subject: [PATCH] kvm: handle virtualization traps due to accessing L2 Cache
 Config register

Commit 88e7a3c824b520f79eadfa380799506e25c79b0d from Broadcom SDK 3.0.2

Signed-off-by: Yonghong Song <ysong@broadcom.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/netlogic/kvm/kvm_traps.c b/arch/mips/netlogic/kvm/kvm_traps.c
index bbe4bf6..c1fccd8 100644
--- a/arch/mips/netlogic/kvm/kvm_traps.c
+++ b/arch/mips/netlogic/kvm/kvm_traps.c
@@ -145,6 +145,9 @@ static void process_psi(struct pt_regs *regs)
 			} else if (cr == 0x400) {
 				/* mmu setup */
 				regs->regs[rt] = 0x0;
+			} else if (cr == 0x800) {
+				/* scu config */
+				regs->regs[rt] = 0x0;
 			} else
 				printk("=== unhandled mfcr cr %x\n", cr);
 
@@ -176,6 +179,9 @@ static void process_psi(struct pt_regs *regs)
 			} else if (cr == 0x700) {
 				/* schedule defeature */
 				;
+			} else if (cr == 0x800) {
+				/* scu config */
+				;
 			} else if (cr == 0xa00) {
 				/* map thread mode: this is the way to wake up
 				 * other threads in the same core.
-- 
1.9.1

