From d57a1dc52d384d18deea23ba2fc13fbc07ddc883 Mon Sep 17 00:00:00 2001
From: Yonghong Song <ysong@broadcom.com>
Date: Fri, 26 Jul 2013 09:29:21 -0700
Subject: [PATCH] kvm: fix a bug in changing user page table pointer

Commit 1be3479e2c92cd88b930c1bcb81c18b6ffa8c126 from Broadcom SDK 3.0.2

Signed-off-by: Yonghong Song <ysong@broadcom.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/include/asm/stackframe.h b/arch/mips/include/asm/stackframe.h
index 83606c4..78e257f 100644
--- a/arch/mips/include/asm/stackframe.h
+++ b/arch/mips/include/asm/stackframe.h
@@ -17,6 +17,7 @@
 #include <asm/asmmacro.h>
 #include <asm/mipsregs.h>
 #include <asm/asm-offsets.h>
+#include <asm/thread_info.h>
 
 /*
  * For SMTC kernel, global IE should be left set, and interrupts
@@ -174,8 +175,8 @@
 		LONG_S  v0, PT_OSSCRATCH7(sp)
 
 		/* restore HVA page table */
-		mfc0    k0, $4, 0
-		srl     k0, k0, 23
+		dmfc0    k0, ASM_SMP_CPUID_REG
+		dsrl    k0, k0, SMP_CPUID_PTRSHIFT
 		PTR_LA  k1, pgd_current
 		daddu   v1, k1, k0
 		LONG_L  k1, VCPU_ARCH_HVA_PGD(v0)
@@ -201,8 +202,8 @@
 		dmtc0   v0, $22, 7
 
 		/* restore GPA page table */
-		mfc0    k0, $4, 0
-		srl     k0, k0, 23
+		dmfc0   k0, ASM_SMP_CPUID_REG
+		dsrl    k0, k0, SMP_CPUID_PTRSHIFT
 		PTR_LA  k1, pgd_current
 		daddu   v1, k1, k0
 		LONG_L  k1, 0(v1)
diff --git a/arch/mips/kvm/context_switch.S b/arch/mips/kvm/context_switch.S
index 5939a38..1b4958b 100644
--- a/arch/mips/kvm/context_switch.S
+++ b/arch/mips/kvm/context_switch.S
@@ -234,8 +234,8 @@ FEXPORT(__kvm_vcpu_run_guest)
 	mtc0	t0, $10, 4
 
         /* Switch to gpa page table for refill and hardware page table */
-        mfc0    t0, $4, 0
-        srl     t0, t0, 23
+        dmfc0    t0, ASM_SMP_CPUID_REG
+        dsrl    t0, t0, SMP_CPUID_PTRSHIFT
         PTR_LA  t1, pgd_current
         daddu   t2, t1, t0
 
-- 
1.9.1

