From 14c08569a4e111b26b2042c9c373566908d4e526 Mon Sep 17 00:00:00 2001
From: Yonghong Song <ysong@broadcom.com>
Date: Thu, 15 Aug 2013 15:41:49 -0700
Subject: [PATCH] kvm: bug fix for MMIO guest exit

Commit b99f09e4889f2c643be98a88bc67545895817e64 from Broadcom SDK 3.0.2

Signed-off-by: Yonghong Song <ysong@broadcom.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/kvm/kvm.c b/arch/mips/kvm/kvm.c
index 84d1bb4..3b710fd 100644
--- a/arch/mips/kvm/kvm.c
+++ b/arch/mips/kvm/kvm.c
@@ -444,7 +444,14 @@ int kvm_arch_vcpu_ioctl_run(struct kvm_vcpu *vcpu, struct kvm_run *run)
 		run->exit_reason = ret >> 24;
 		run->mmio.phys_addr = vcpu->arch.mmio.addr;
 		run->mmio.len = vcpu->arch.mmio.len;
-		memcpy(run->mmio.data, &vcpu->arch.mmio.val, 8);
+
+		if (run->mmio.len == 1)
+			run->mmio.data[0] = (uint8_t)vcpu->arch.mmio.val;
+		else if (run->mmio.len == 4)
+			*(uint32_t *)run->mmio.data = (uint32_t)vcpu->arch.mmio.val;
+		else if (run->mmio.len == 8)
+			*(uint64_t *)run->mmio.data = (uint64_t)vcpu->arch.mmio.val;
+
 		run->mmio.is_write = vcpu->arch.mmio.is_write;
 		if (!run->mmio.is_write)
 			vcpu->mmio_needed = 1;
-- 
1.9.1

