From 9009b36134ddb104b51b1971cc7ac733684da8e7 Mon Sep 17 00:00:00 2001
From: Yonghong Song <ysong@broadcom.com>
Date: Wed, 8 May 2013 14:29:38 -0700
Subject: [PATCH] kvm: avoid using smp_processor_id() in places where process
 may be preempted

Commit 6c0a7ff69dabd2bdfe3cfcd2c5420bfc2e4730a4 from Broadcom SDK 3.0.2

Signed-off-by: Yonghong Song <ysong@broadcom.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/kvm/kvm.c b/arch/mips/kvm/kvm.c
index a1a25d4..f366afd 100644
--- a/arch/mips/kvm/kvm.c
+++ b/arch/mips/kvm/kvm.c
@@ -108,7 +108,7 @@ static void kvm_sync_gpa_map(struct kvm *kvm)
 			int i;
 			unsigned long *ptr;
 
-			pgdp = (pgd_t *)pgd_current[smp_processor_id()] + __pgd_offset(address);
+			pgdp = current->mm->pgd + __pgd_offset(address);
 			pudp = pud_offset(pgdp, address);
 			pmdp = pmd_offset(pudp, address);
 			ptep = pte_offset(pmdp, address);
@@ -217,7 +217,7 @@ static void kvm_set_gpa_pa_map(struct kvm *kvm, struct kvm_userspace_memory_regi
 
 	for (address = hva_base; address < (hva_base + msize);
 	     address += PAGE_SIZE, gpa += PAGE_SIZE) {
-		pgdp = (pgd_t *)pgd_current[smp_processor_id()] + __pgd_offset(address);
+		pgdp = current->mm->pgd + __pgd_offset(address);
 		pudp = pud_offset(pgdp, address);
 		pmdp = pmd_offset(pudp, address);
 		ptep = pte_offset(pmdp, address);
diff --git a/arch/mips/mm/fault.c b/arch/mips/mm/fault.c
index 379aa4c..59d104f 100644
--- a/arch/mips/mm/fault.c
+++ b/arch/mips/mm/fault.c
@@ -50,7 +50,7 @@ static void nlm_kvm_populate_page(struct pt_regs *regs, unsigned long address,
         unsigned long	*ptr;
         int		i;
 
-	pgdp = (pgd_t *)pgd_current[smp_processor_id()] + __pgd_offset(address);
+	pgdp = current->mm->pgd + __pgd_offset(address);
 	pudp = pud_offset(pgdp, address);
 	pmdp = pmd_offset(pudp, address);
 	ptep = pte_offset(pmdp, address);
-- 
1.9.1

