From 617aba4944c44b2d13b1726006032d92899cf0b9 Mon Sep 17 00:00:00 2001
From: Yonghong Song <ysong@broadcom.com>
Date: Tue, 30 Jul 2013 09:39:24 -0700
Subject: [PATCH] kvm: fix a memory leak during injecting char to guest uart

Commit 859f1dac6054a73f908cf72e783cef256bdee804 from Broadcom SDK 3.0.2

Signed-off-by: Yonghong Song <ysong@broadcom.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/kvm/kvm.c b/arch/mips/kvm/kvm.c
index 2d7294b3..a05b03d 100644
--- a/arch/mips/kvm/kvm.c
+++ b/arch/mips/kvm/kvm.c
@@ -756,6 +756,7 @@ long kvm_arch_vm_ioctl(struct file *filp, unsigned int ioctl, unsigned long arg)
 		for (i = 0; i < input.size; i++) {
 			kvm_uart_insert_char(kvm, buf[i]);
 		}
+		vfree(buf);
 		break;
 	}
 	case KVM_MIPS_EXIT_REQUEST:
-- 
1.9.1

