From db3ae419020d240068e953d6616ca22d3232ebb3 Mon Sep 17 00:00:00 2001
From: Yonghong Song <ysong@broadcom.com>
Date: Wed, 9 Oct 2013 15:17:58 -0700
Subject: [PATCH] kvm: fix memory free bug

Commit 70934e76d5edf6b93adbcb39353fde95e6d9b5f5 from Broadcom SDK 3.0.2

o The bug caused a double free for kmalloc'd memory

Signed-off-by: Yonghong Song <ysong@broadcom.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/kvm/kvm.c b/arch/mips/kvm/kvm.c
index 2859215..cb46737 100644
--- a/arch/mips/kvm/kvm.c
+++ b/arch/mips/kvm/kvm.c
@@ -722,7 +722,6 @@ void kvm_arch_destroy_vm(struct kvm *kvm)
 {
 	xlp_kvm_destroy_vm(kvm);
 	kvm_free_vcpus(kvm);
-	kvm_free_physmem(kvm);
 }
 
 void kvm_arch_free_memslot(struct kvm_memory_slot *free,
-- 
1.9.1

