From eaf1ce180e910eee280331b9ba8ca3396cf41fa8 Mon Sep 17 00:00:00 2001
From: Ashok Kumar <ashoks@broadcom.com>
Date: Fri, 14 Feb 2014 12:42:28 +0530
Subject: [PATCH] kvm: cleanup and fix for previous commit break

Commit b22557c46e98c77cefbc9d699b53c6e2e4a33bb9 from Broadcom SDK 3.0.2

Signed-off-by: Ashok Kumar <ashoks@broadcom.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/kvm/kvm.c b/arch/mips/kvm/kvm.c
index f3b95b8..e46edc2 100644
--- a/arch/mips/kvm/kvm.c
+++ b/arch/mips/kvm/kvm.c
@@ -467,9 +467,10 @@ int kvm_arch_vcpu_ioctl_run(struct kvm_vcpu *vcpu, struct kvm_run *run)
 		run->exit_reason = ret >> 24;
 		break;
 	case KVM_EXIT_QUIT_KVM:
+#if 0
 		printk("KVM_EXIT_QUIT_KVM Running guest (vcpu %p, kvm %p, Entry PC: 0x%llx, GuestId 0x%llx, Guest vcpu id: 0x%llx) ...\n",
                 vcpu, vcpu->kvm, r->root_epc, r->root_guestctl1 & 0xff, r->guest_ebase & 0x3ff);
-
+#endif
 		/* quit kvm */
 		run->exit_reason = ret >> 24;
 		/* also clear guestctl1 */
@@ -521,6 +522,7 @@ int kvm_arch_vcpu_ioctl_run(struct kvm_vcpu *vcpu, struct kvm_run *run)
 		} else {
 			vcpu->arch.pio_needed = 1;
 		}
+		break;
 
 	case KVM_EXIT_HC_HALX_NETSOC:
 		run->mips_hypcall.args[0] = vcpu->arch.mips_hypcall.args[0];
-- 
1.9.1

