From 6ce6618edc6b7c4a95e44b0da0b6246e12335e74 Mon Sep 17 00:00:00 2001
From: Yonghong Song <ysong@broadcom.com>
Date: Tue, 9 Jul 2013 09:44:00 -0700
Subject: [PATCH] kvm: Add a hypercall interface to get the
 guest_cpuid->root_cpuid mapping

Commit 3e311f1e567879e15dd5b871e220c15b9a22efaf from Broadcom SDK 3.0.2

Signed-off-by: Yonghong Song <ysong@broadcom.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/include/asm/netlogic/kvm_para.h b/arch/mips/include/asm/netlogic/kvm_para.h
index a1865f4..7e0f9c8 100644
--- a/arch/mips/include/asm/netlogic/kvm_para.h
+++ b/arch/mips/include/asm/netlogic/kvm_para.h
@@ -37,4 +37,26 @@
 
 extern int is_nlm_guest_os;
 
+#define KVM_HC_GET_CPUID	0x1  /* p1r1 */
+
+static inline int do_hypcall_p1r1(int num, uint64_t param0, uint64_t *ret0)
+{
+	int	 ret;
+	uint64_t result;
+
+	__asm__ __volatile__(
+		"move $2, %2\n"
+		"move $3, %3\n"
+		"hypcall\n"
+		"move %0, $2\n"
+		"move %1, $3\n"
+		: "=r"(ret), "=r"(result)
+		: "r"(num), "r"(param0)
+		: "$2","$3", "memory"
+	);
+
+	*ret0 = result;
+	return ret;
+}
+
 #endif
diff --git a/arch/mips/netlogic/kvm/kvm_traps.c b/arch/mips/netlogic/kvm/kvm_traps.c
index 87867ef..5d84ba1 100644
--- a/arch/mips/netlogic/kvm/kvm_traps.c
+++ b/arch/mips/netlogic/kvm/kvm_traps.c
@@ -39,6 +39,7 @@
 #include <asm/uaccess.h>
 #include <asm/kvm.h>
 #include <asm/kvm_host.h>
+#include <asm/netlogic/kvm_para.h>
 #include <asm/netlogic/kvm_xlp.h>
 #include <asm/netlogic/mips-extns.h>
 
@@ -344,12 +345,41 @@ static void process_hc(struct pt_regs *regs)
 {
 	unsigned int badinstr, epc_badinstr;
 	unsigned long epc;
+	int hc_num;
 
 	epc = regs->cp0_epc;
 	kvm_get_badinstr(regs, &badinstr, &epc_badinstr);
 
-	printk("Hypercall exception not implemented:\n");
-	printk("\tepc 0x%lx, badinstr: 0x%x\n", epc, badinstr);
+	hc_num = regs->regs[2];
+	switch (hc_num) {
+	case KVM_HC_GET_CPUID:
+	{
+		int i, guest_cpuid = regs->regs[3], host_cpuid = -1;
+		struct kvm *kvm;
+		struct kvm_vcpu *vcpu;
+
+		kvm = ((struct kvm_vcpu *)regs->cp0_osscratch7)->kvm;
+		kvm_for_each_vcpu(i, vcpu, kvm) {
+			if (vcpu->vcpu_id == guest_cpuid) {
+				host_cpuid = vcpu->arch.host_vcpuid;
+				break;
+			}
+		}
+
+		/* result */
+		if (host_cpuid == (-1)) {
+			regs->regs[2] = -1; /* fail */
+		} else {
+			regs->regs[2] = 0; /* succeed */
+			regs->regs[3] = host_cpuid;
+		}
+		break;
+	}
+	default:
+		printk("Hypercall exception (number %d) not implemented:\n", hc_num);
+		printk("\tepc 0x%lx, badinstr: 0x%x\n", epc, badinstr);
+		break;
+	}
 
 	compute_guest_return_epc(regs, epc_badinstr);
 }
diff --git a/arch/mips/netlogic/xlp/setup.c b/arch/mips/netlogic/xlp/setup.c
index 312322f..b08f479 100644
--- a/arch/mips/netlogic/xlp/setup.c
+++ b/arch/mips/netlogic/xlp/setup.c
@@ -33,6 +33,7 @@
  */
 
 #include <linux/kernel.h>
+#include <linux/export.h>
 #include <linux/of_fdt.h>
 #include <linux/module.h>
 
-- 
1.9.1

