From 73e19327e57a7e023f6e84c26470355c1f406a11 Mon Sep 17 00:00:00 2001
From: Yonghong Song <ysong@broadcom.com>
Date: Mon, 6 May 2013 08:50:15 -0700
Subject: [PATCH] kvm: save/restore guest CP0 state during context switch

Commit 6cd6150b2f810fcc1696a79b9cde80bc5e63ad6f from Broadcom SDK 3.0.2

Signed-off-by: Yonghong Song <ysong@broadcom.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/kernel/r4k_switch.S b/arch/mips/kernel/r4k_switch.S
index 4381dc3..ff5b4ab 100644
--- a/arch/mips/kernel/r4k_switch.S
+++ b/arch/mips/kernel/r4k_switch.S
@@ -68,6 +68,10 @@
 						# clobbers t1
 1:
 
+#ifdef CONFIG_KVM
+	cpu_save_vm_state a0
+#endif
+
 	/*
 	 * The order of restoring the registers takes care of the race
 	 * updating $28, $29 and kernelsp without disabling ints.
@@ -101,6 +105,9 @@
 	and	a2, a3
 	or	a2, t1
 	mtc0	a2, CP0_STATUS
+#ifdef CONFIG_KVM
+	cpu_restore_vm_state a1
+#endif
 #ifdef CONFIG_MIPS_MT_SMTC
 	_ehb
 	andi	t0, t0, VPECONTROL_TE
diff --git a/arch/mips/kvm/context_switch.S b/arch/mips/kvm/context_switch.S
index 62bb40f..07ffd53 100644
--- a/arch/mips/kvm/context_switch.S
+++ b/arch/mips/kvm/context_switch.S
@@ -170,6 +170,7 @@
 	.endm
 
 FEXPORT(__kvm_vcpu_run_guest)
+	.set	noat
 
 	/* essentially, it is a context switch, we must
 	 * save all relevant registers, and then load
-- 
1.9.1

