From 10e646920b139ba0394f09edfd5358fa9985c56a Mon Sep 17 00:00:00 2001
From: Yonghong Song <ysong@broadcom.com>
Date: Thu, 30 May 2013 09:16:37 -0700
Subject: [PATCH] kvm: add vhost to Kconfig (when virtualization is enabled)

Commit 5561953a3d0b22ec3acd8553a4c2fa34475cca33 from Broadcom SDK 3.0.2

o Also comment out a couple of kernel debug messages

Signed-off-by: Yonghong Song <ysong@broadcom.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/kvm/Kconfig b/arch/mips/kvm/Kconfig
index ff63681..6d38d4d 100644
--- a/arch/mips/kvm/Kconfig
+++ b/arch/mips/kvm/Kconfig
@@ -21,4 +21,6 @@ config KVM
 	select PREEMPT_NOTIFIERS
 	select ANON_INODES
 
+source drivers/vhost/Kconfig
+
 endif
diff --git a/arch/mips/netlogic/kvm/kvm_sysmgt.c b/arch/mips/netlogic/kvm/kvm_sysmgt.c
index 015f5b5..c3f1016 100644
--- a/arch/mips/netlogic/kvm/kvm_sysmgt.c
+++ b/arch/mips/netlogic/kvm/kvm_sysmgt.c
@@ -86,7 +86,9 @@ void kvm_handle_pcie_sysmgt(struct pt_regs *regs, unsigned long write,
 					unsigned int val = i << 2, badinstr, epc_badinstr;
 					val |= (KVM_EXIT_ENABLE_CORE << 24);
 
+#if 0
 					printk("Enable core %d ...\n", i);
+#endif
 					kvm_get_badinstr(regs, &badinstr, &epc_badinstr);
 					compute_guest_return_epc(regs, epc_badinstr);
 					kvm_save_guest_context(regs, kvm_get_vcpu_guest_regs(regs));	
diff --git a/arch/mips/netlogic/kvm/kvm_traps.c b/arch/mips/netlogic/kvm/kvm_traps.c
index 53d0dfa..bbe4bf6 100644
--- a/arch/mips/netlogic/kvm/kvm_traps.c
+++ b/arch/mips/netlogic/kvm/kvm_traps.c
@@ -196,8 +196,11 @@ static void process_psi(struct pt_regs *regs)
 						val |= (3 << 16);
 					val |= (KVM_EXIT_SPAWN_THREADS << 24);
 
+#if 0
 					printk("Tentatively Leaving the guest (Spawning %d new threads)...\n",
 						(regs->regs[rt] == 2) ? 1 : 3);
+#endif
+
 					compute_guest_return_epc(regs, epc_badinstr);
 					kvm_save_guest_context(regs, kvm_get_vcpu_guest_regs(regs));
 					__kvm_vcpu_leave_guest((struct kvm_vcpu *)regs->cp0_osscratch7, val);
-- 
1.9.1

