From 56d6765e1628385252e0f52b978be10842a6f38b Mon Sep 17 00:00:00 2001
From: Nam Ninh <nam.ninh@windriver.com>
Date: Tue, 25 Nov 2014 17:59:24 -0500
Subject: [PATCH] kvm: enable MSI support for XLP

Enable MSI for XLP because the hardware supports it.

Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/Kconfig b/arch/mips/Kconfig
index 5d20397..54d953b 100644
--- a/arch/mips/Kconfig
+++ b/arch/mips/Kconfig
@@ -831,6 +831,7 @@ config NLM_XLP_BOARD
 	select CEVT_R4K
 	select CSRC_R4K
 	select IRQ_CPU
+	select ARCH_SUPPORTS_MSI
 	select ZONE_DMA32 if 64BIT
 	select SYNC_R4K
 	select SYS_HAS_EARLY_PRINTK
diff --git a/arch/mips/include/asm/netlogic/xlp-hal/pic.h b/arch/mips/include/asm/netlogic/xlp-hal/pic.h
index 03faaac..2761297 100644
--- a/arch/mips/include/asm/netlogic/xlp-hal/pic.h
+++ b/arch/mips/include/asm/netlogic/xlp-hal/pic.h
@@ -222,6 +222,8 @@
 #define PIC_IRT_PCIE_LINK_INDEX(num)	((num) + PIC_IRT_PCIE_LINK_0_INDEX)
 
 #define PIC_XLP9XX_IRT_PCIE_LINK_0_INDEX	191
+#define PIC_XLP9XX_IRT_PCIE_LINK_INDEX(num) \
+	((num) + PIC_XLP9XX_IRT_PCIE_LINK_0_INDEX)
 
 #define PIC_CLOCK_TIMER			7
 
-- 
1.9.1

