From 22aa5b1a3ad56deb7da2cf3233f6af09163303bb Mon Sep 17 00:00:00 2001
From: Abhishek Joshi <abjoshi@broadcom.com>
Date: Mon, 7 Apr 2014 21:33:55 +0530
Subject: [PATCH] kvm: removed unnecessary prints under debug

Commit 5c995c1b0021909a57ba60d82463a2dc88dd0642 from Broadcom SDK 3.0.2

Signed-off-by: Abhishek Joshi <abjoshi@broadcom.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/include/asm/netlogic/kvm_xlp.h b/arch/mips/include/asm/netlogic/kvm_xlp.h
index 1c206bf..c8f782a 100644
--- a/arch/mips/include/asm/netlogic/kvm_xlp.h
+++ b/arch/mips/include/asm/netlogic/kvm_xlp.h
@@ -39,6 +39,7 @@
 
 /* Maximum # of guest ids supported by hardware */
 #define KVM_MAX_NUM_GID		256
+/*#define KVM_HYPERCALL_DEBUG*/
 
 static inline struct kvm_arch * kvm_get_arch(struct pt_regs *regs)
 {
diff --git a/arch/mips/kvm/kvm.c b/arch/mips/kvm/kvm.c
index e46edc2..f8f3cd9 100644
--- a/arch/mips/kvm/kvm.c
+++ b/arch/mips/kvm/kvm.c
@@ -528,8 +528,10 @@ int kvm_arch_vcpu_ioctl_run(struct kvm_vcpu *vcpu, struct kvm_run *run)
 		run->mips_hypcall.args[0] = vcpu->arch.mips_hypcall.args[0];
 		run->mips_hypcall.cmd = vcpu->arch.mips_hypcall.cmd;
 		run->exit_reason = ret >> 24;
+#ifdef KVM_HYPERCALL_DEBUG
 		printk("KVM_EXIT_HC_HALX_NETSOC. run->mips_hypcall.args[0] = 0x%lx\n", (long unsigned int)run->mips_hypcall.args[0]);
 		printk("KVM_EXIT_HC_HALX_NETSOC. run->exit_reason = 0x%lx, run->mips_hypcall.cmd = %d\n", (long unsigned int)run->exit_reason, (int)run->mips_hypcall.cmd);
+#endif
 		break;
 #if 0
 	case KVM_EXIT_IRQ_WINDOW_OPEN:
diff --git a/arch/mips/netlogic/kvm/kvm_traps.c b/arch/mips/netlogic/kvm/kvm_traps.c
index f8ab256..fd01b5c 100644
--- a/arch/mips/netlogic/kvm/kvm_traps.c
+++ b/arch/mips/netlogic/kvm/kvm_traps.c
@@ -115,7 +115,9 @@ static int gpa_to_hva (struct kvm_vcpu *vcpu, unsigned long gpa_address, unsigne
 			break;
 		}
 	}
+#ifdef KVM_HYPERCALL_DEBUG
 	printk("hva address 0x%lx \n", address);
+#endif
 	*hva_address = address;
 	return 0;
 }
@@ -503,7 +505,9 @@ static void process_hc(struct pt_regs *regs)
 		int guest_cpuid = 0;
 
 		kvm = ((struct kvm_vcpu *)regs->cp0_osscratch7)->kvm;
+#ifdef KVM_HYPERCALL_DEBUG
 		printk("Hypercall exception (number %d) netsoc: 0x%lx\n",hc_num, regs->regs[3]);
+#endif
 
 		kvm_for_each_vcpu(i, vcpu, kvm) {
 			if (vcpu->vcpu_id == guest_cpuid) {
@@ -527,10 +531,10 @@ static void process_hc(struct pt_regs *regs)
 		gpa_addr = regs->regs[3];
 		sub_cmd = regs->regs[4];
 		gpa_to_hva((struct kvm_vcpu *)regs->cp0_osscratch7, gpa_addr, &hva_addr);
-
+#ifdef KVM_HYPERCALL_DEBUG
 		printk("Tentatively Leaving the guest ...gpa 0x%lx hva 0x%lx\n", gpa_addr, hva_addr);
 		printk("KVM_HC_HALX_NETSOC: sub_cmd = %d\n",  sub_cmd);
-
+#endif
 		vcpu->arch.mips_hypcall.args[0] = hva_addr;
 		vcpu->arch.mips_hypcall.cmd = sub_cmd;
 
@@ -573,7 +577,9 @@ void process_virt_exception(struct pt_regs *regs)
 		process_fc(regs);
 		break;
 	case 0x2:
+#ifdef KVM_HYPERCALL_DEBUG
 		printk("Hypercall\n");
+#endif
 		process_hc(regs);
 		break;
 	case 0x3:
-- 
1.9.1

