From 066c87afbf46854ac5e722e1bfbea6ca594aa17f Mon Sep 17 00:00:00 2001
From: Yang Shi <yang.shi@windriver.com>
Date: Thu, 1 May 2014 11:58:45 -0700
Subject: [PATCH] grsec: Check if host GCC support -Wno-empty-body

grsec adds -Wno-empty-body compiler option to Makefile, but it is not supported
on older version GCC, so it could cause kernel configure failure on CentOS 5.x.
Although cc-option is used to test if compiler supports this option, but
cc-option just can test options for cross compiler rather than host compiler.

So, added cc-option-host to test options for host compiler, then use it to
pass -Wno-empty-body conditionally.

Signed-off-by: Yang Shi <yang.shi@windriver.com>
---
 Makefile               | 12 ++++++------
 scripts/Kbuild.include |  5 +++++
 2 files changed, 11 insertions(+), 6 deletions(-)

diff --git a/Makefile b/Makefile
index e9f40a7..6f65e20 100644
--- a/Makefile
+++ b/Makefile
@@ -239,11 +239,15 @@ CONFIG_SHELL := $(shell if [ -x "$$BASH" ]; then echo $$BASH; \
 	  else if [ -x /bin/bash ]; then echo /bin/bash; \
 	  else echo sh; fi ; fi)
 
+# We need some generic definitions (do not try to remake the file).
+$(srctree)/scripts/Kbuild.include: ;
+include $(srctree)/scripts/Kbuild.include
+
 HOSTCC       = gcc
 HOSTCXX      = g++
-HOSTCFLAGS   = -Wall -W -Wmissing-prototypes -Wstrict-prototypes -Wno-unused-parameter -Wno-missing-field-initializers -O2 -fomit-frame-pointer -fno-delete-null-pointer-checks
-HOSTCFLAGS  += $(call cc-option, -Wno-empty-body)
+HOSTCFLAGS   := -Wall -W -Wmissing-prototypes -Wstrict-prototypes -Wno-unused-parameter -Wno-missing-field-initializers -O2 -fomit-frame-pointer -fno-delete-null-pointer-checks
 HOSTCXXFLAGS = -O2 -Wall -W -fno-delete-null-pointer-checks
+HOSTCFLAGS   += $(call cc-option-host, -Wno-empty-body)
 
 # Decide whether to build built-in, modular, or both.
 # Normally, just do built-in.
@@ -319,10 +323,6 @@ export quiet Q KBUILD_VERBOSE
 # Look for make include files relative to root of kernel src
 MAKEFLAGS += --include-dir=$(srctree)
 
-# We need some generic definitions (do not try to remake the file).
-$(srctree)/scripts/Kbuild.include: ;
-include $(srctree)/scripts/Kbuild.include
-
 # Make variables (CC, etc...)
 
 AS		= $(CROSS_COMPILE)as
diff --git a/scripts/Kbuild.include b/scripts/Kbuild.include
index 547e15d..a8dd807 100644
--- a/scripts/Kbuild.include
+++ b/scripts/Kbuild.include
@@ -117,6 +117,11 @@ cc-option = $(call try-run,\
 cc-option-yn = $(call try-run,\
 	$(CC) $(KBUILD_CPPFLAGS) $(KBUILD_CFLAGS) $(1) -c -x c /dev/null -o "$$TMP",y,n)
 
+# cc-option-host
+# Usage: cflags-y += $(call cc-option-host,-march=winchip-c6,-march=i586)
+cc-option-host = $(call try-run,\
+	$(HOSTCC) $(HOSTCXXFLAGS) $(HOSTCFLAGS) $(1) -c -x c /dev/null -o "$$TMP",$(1),$(2))
+
 # cc-option-align
 # Prefix align with either -falign or -malign
 cc-option-align = $(subst -functions=0,,\
-- 
1.9.2

