From 7fa43e88a8e1c19dc4582643416baeb8bf2673c3 Mon Sep 17 00:00:00 2001
From: "Roy.Li" <rongqing.li@windriver.com>
Date: Mon, 4 Mar 2013 15:09:26 +0800
Subject: [PATCH 14/16] grsecurity: call cputime_to_secs to convert cputime to
 seconds

When MICROSTATE_ACCT is enabled, The type of cputime_t is changed from
unsigned long to u64, which will lead to a 64-bit integer division in
32bit platform, to make it work, we must use special macro to do it.

        ./asm-generic/cputime_msa.h:typedef u64 cputime_t;
        ./asm-generic/cputime.h:typedef unsigned long __nocast cputime_t;

In fact, task->utime can not be divided by HZ, since once MICROSTATE_ACCT is
enabled, the task->{utime, stime} does not mean jiffies, but microseconds, the
macro of cputime_to_secs should always be used to convert from cputtime to
seconds.

cputime_to_secs has many protypes, when enable MICROSTATE_ACCT, The one
from cputime_msa.h will be used to do conversion.

        ./asm-generic/cputime_msa.h:#define cputime_to_secs(msa)
                ((unsigned long) cputime_div(msa, NSEC_PER_SEC))
        ./asm-generic/cputime.h:#define cputime_to_secs(jif)
                (cputime_to_jiffies(jif) / HZ)

Signed-off-by: Roy.Li <rongqing.li@windriver.com>
---
 grsecurity/gracl.c |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/grsecurity/gracl.c b/grsecurity/gracl.c
index 1de6aab..27f30d9 100644
--- a/grsecurity/gracl.c
+++ b/grsecurity/gracl.c
@@ -3994,7 +3994,7 @@ gr_acl_handle_psacct(struct task_struct *task, const long code)
 	runtime -= wmin * 60;
 	wsec = runtime;
 
-	cputime = (task->utime + task->stime) / HZ;
+	cputime = cputime_to_secs(task->utime + task->stime);
 	cday = cputime / (3600 * 24);
 	cputime -= cday * (3600 * 24);
 	chr = cputime / 3600;
-- 
1.7.9.5

