From 6a1f6fd0947a634a8def05c8547bd7efd24a5636 Mon Sep 17 00:00:00 2001
From: Yang Shi <yang.shi@windriver.com>
Date: Mon, 12 Jan 2015 13:04:50 -0800
Subject: [PATCH] pax: disable the plugin checks for now

Need to determine if supporting plugins in a cross compile
environment is worthwhile.  In the interim, the host checks
are counterproductive in forcing a failure and suggesting
that plugins get installed.

Signed-off-by: Paul Gortmaker <paul.gortmaker@windriver.com>
[Adjust context on 3.14 kernel]
Signed-off-by: Yang Shi <yang.shi@windriver.com>
---
 Makefile | 1 +
 1 file changed, 1 insertion(+)

diff --git a/Makefile b/Makefile
index 97dc0c5..3ad6980 100644
--- a/Makefile
+++ b/Makefile
@@ -589,6 +589,7 @@ endif
 # Tell gcc to never replace conditional load with a non-conditional one
 KBUILD_CFLAGS	+= $(call cc-option,--param=allow-store-data-races=0)
 
+DISABLE_PAX_PLUGINS := yes_I_know_what_Im_doing
 ifndef DISABLE_PAX_PLUGINS
 ifeq ($(call cc-ifversion, -ge, 0408, y), y)
 PLUGINCC := $(shell $(CONFIG_SHELL) $(srctree)/scripts/gcc-plugin.sh "$(HOSTCXX)" "$(HOSTCXX)" "$(CC)")
-- 
2.0.2

