From cfb28809b645b89a1beb5ecdd247ef09fbe8e191 Mon Sep 17 00:00:00 2001
From: Thomas Gleixner <tglx@linutronix.de>
Date: Tue, 20 Dec 2011 19:06:57 +0100
Subject: [PATCH] kvm: Allow to query the msi host irq

Signed-off-by: Thomas Gleixner <tglx@linutronix.de>
Signed-off-by: Michael Barabanov <michael.barabanov@windriver.com>
[PG: relocate to arch/x86 for 4.1 kernel]
Signed-off-by: Paul Gortmaker <paul.gortmaker@windriver.com>

diff --git a/arch/x86/kvm/assigned-dev.c b/arch/x86/kvm/assigned-dev.c
index ef86fb060d82..2d2d26017369 100644
--- a/arch/x86/kvm/assigned-dev.c
+++ b/arch/x86/kvm/assigned-dev.c
@@ -919,12 +919,18 @@ static int kvm_vm_ioctl_query_irq(struct kvm *kvm, u32 assigned_dev_id,
 				      assigned_dev_id);
 	if (!dev)
 		goto out;
-	if (entry >= dev->entries_nr)
-		goto out;
+
 	if (dev->host_irq_disabled)
 		goto out;
 
-	r = dev->host_msix_entries[entry].vector;
+	if (entry >= dev->entries_nr) {
+		if (entry)
+			goto out;
+		/* MSI case */
+		r = dev->host_irq;
+	} else {
+		r = dev->host_msix_entries[entry].vector;
+	}
 out:
 	mutex_unlock(&kvm->lock);
 	return r;
-- 
2.5.0

