From 6137d43d09d0042588159b85be630abd40640878 Mon Sep 17 00:00:00 2001
From: Paul Gortmaker <paul.gortmaker@windriver.com>
Date: Thu, 21 Nov 2013 13:43:42 -0500
Subject: [PATCH] kvm: avoid 64 bit math in vmx.c tracing on 32 bit builds

To fix:

 ERROR: "__udivdi3" [arch/x86/kvm/kvm-intel.ko] undefined!

Signed-off-by: Paul Gortmaker <paul.gortmaker@windriver.com>

diff --git a/arch/x86/kvm/vmx.c b/arch/x86/kvm/vmx.c
index d938be0..0b8f263 100644
--- a/arch/x86/kvm/vmx.c
+++ b/arch/x86/kvm/vmx.c
@@ -4703,7 +4703,7 @@ static int handle_io(struct kvm_vcpu *vcpu)
 		}
 		else {
 			guest_time += kvm_register_read(vcpu, VCPU_REGS_RAX);
-			host_guest_delta = (host_time - guest_time) / 1000;
+			host_guest_delta = div_s64(host_time - guest_time, 1000);
 
 			trace_printk("Guest time: %llu\n", guest_time);
 			trace_printk("Host time: %llu\n", host_time);
-- 
1.8.5.rc3

