From f9d2168483354a3ea8d3f54d8bba6c9982a01afc Mon Sep 17 00:00:00 2001
From: Paul Gortmaker <paul.gortmaker@windriver.com>
Date: Tue, 5 Feb 2013 14:32:32 -0500
Subject: [PATCH] kvm-rt: swait_head_has_waiters is now swaitqueue_active

Signed-off-by: Paul Gortmaker <paul.gortmaker@windriver.com>
Signed-off-by: Michael Barabanov <michael.barabanov@windriver.com>

diff --git a/virt/kvm/kvm_main.c b/virt/kvm/kvm_main.c
index 710ea15..983d203 100644
--- a/virt/kvm/kvm_main.c
+++ b/virt/kvm/kvm_main.c
@@ -1597,7 +1597,7 @@ void kvm_vcpu_on_spin(struct kvm_vcpu *me)
 				break;
 			if (vcpu == me)
 				continue;
-			if (swait_head_has_waiters(&vcpu->wq))
+			if (swaitqueue_active(&vcpu->wq))
 				continue;
 			if (kvm_vcpu_yield_to(vcpu)) {
 				kvm->last_boosted_vcpu = i;
-- 
1.7.9.5

