From 8d48538af006ac1b1bbe2ff4f4554e5d7adf427c Mon Sep 17 00:00:00 2001
From: Thomas Gleixner <tglx@linutronix.de>
Date: Tue, 20 Dec 2011 19:06:57 +0100
Subject: [PATCH] kvm: Allow to query the msi host irq

Signed-off-by: Thomas Gleixner <tglx@linutronix.de>
Signed-off-by: Michael Barabanov <michael.barabanov@windriver.com>

diff --git a/virt/kvm/assigned-dev.c b/virt/kvm/assigned-dev.c
index e7d84f7..c4cb792 100644
--- a/virt/kvm/assigned-dev.c
+++ b/virt/kvm/assigned-dev.c
@@ -861,12 +861,18 @@ static int kvm_vm_ioctl_query_irq(struct kvm *kvm, u32 assigned_dev_id,
 				      assigned_dev_id);
 	if (!dev)
 		goto out;
-	if (entry >= dev->entries_nr)
-		goto out;
+
 	if (dev->host_irq_disabled)
 		goto out;
 
-	r = dev->host_msix_entries[entry].vector;
+	if (entry >= dev->entries_nr) {
+		if (entry)
+			goto out;
+		/* MSI case */
+		r = dev->host_irq;
+	} else {
+		r = dev->host_msix_entries[entry].vector;
+	}
 out:
 	mutex_unlock(&kvm->lock);
 	return r;
-- 
1.7.9.5

