From b5bf2986e9f7be0975ffa3302b4c759171f91118 Mon Sep 17 00:00:00 2001
From: ltian <le.tian@windriver.com>
Date: Fri, 12 Nov 2010 15:56:12 +0800
Subject: [PATCH 049/132] Cavium: Export cvmx_bootmem_alloc_named_range

Source: Internal development

cvmx_bootmem_alloc_named_range is called by octeon-ethernet
driver, export the symbol so that octeon-ethernet driver can be
loaded as a kernel module.

Signed-off-by: ltian <le.tian@windriver.com>
---
 simple_exec/cvmx-bootmem.c |    2 ++
 1 files changed, 2 insertions(+), 0 deletions(-)

diff --git a/simple_exec/cvmx-bootmem.c b/simple_exec/cvmx-bootmem.c
index bb373fb..2b6d94a 100644
--- a/simple_exec/cvmx-bootmem.c
+++ b/simple_exec/cvmx-bootmem.c
@@ -469,6 +469,8 @@ void *cvmx_bootmem_alloc_named_range(uint64_t size, uint64_t min_addr, uint64_t
         return NULL;
 
 }
+EXPORT_SYMBOL(cvmx_bootmem_alloc_named_range);
+
 void *cvmx_bootmem_alloc_named_address(uint64_t size, uint64_t address, const char *name)
 {
     return(cvmx_bootmem_alloc_named_range(size, address, address + size, 0, name));
-- 
1.6.5.2

