From b3a25ca258c6082e057b9f99bbc75913bf302074 Mon Sep 17 00:00:00 2001
From: ltian <le.tian@windriver.com>
Date: Thu, 11 Nov 2010 16:25:37 +0800
Subject: [PATCH 021/132] Cavium: Implement exception table swap

Source: SDK 2.0.0-366

No exception table swap function is specified by default. Provide a
swap function to exchange two entries.

Signed-off-by: ltian <le.tian@windriver.com>
---
 lib/extable.c |   13 ++++++++++++-
 1 files changed, 12 insertions(+), 1 deletions(-)

diff --git a/lib/extable.c b/lib/extable.c
index 4cac81e..6d5144e 100644
--- a/lib/extable.c
+++ b/lib/extable.c
@@ -33,11 +33,22 @@ static int cmp_ex(const void *a, const void *b)
 	return 0;
 }
 
+static void swap_ex(void *a, void *b, int size)
+{
+	struct exception_table_entry t;
+	struct exception_table_entry *ea = a;
+	struct exception_table_entry *eb = b;
+
+	t = *ea;
+	*ea = *eb;
+	*eb = t;
+}
+
 void sort_extable(struct exception_table_entry *start,
 		  struct exception_table_entry *finish)
 {
 	sort(start, finish - start, sizeof(struct exception_table_entry),
-	     cmp_ex, NULL);
+	     cmp_ex, swap_ex);
 }
 
 #ifdef CONFIG_MODULES
-- 
1.6.5.2

