From 68a0fb606027bc58b0f041abc8b40aba12f2eb5d Mon Sep 17 00:00:00 2001
From: ltian <le.tian@windriver.com>
Date: Thu, 11 Nov 2010 10:28:53 +0800
Subject: [PATCH 011/132] Cavium: SDK2.0 EDAC driver support files for Cavium Octeon

These files update EDAC driver support.
The files in their current state were provided by
Ralf Baechle <ralf.baechle@windriver.com>.

Integrated-by: ltian <le.tian@windriver.com>
---
 drivers/edac/octeon_edac-pc.c |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/drivers/edac/octeon_edac-pc.c b/drivers/edac/octeon_edac-pc.c
index e219ab7..bfb8828 100644
--- a/drivers/edac/octeon_edac-pc.c
+++ b/drivers/edac/octeon_edac-pc.c
@@ -76,7 +76,7 @@ static int unregister_co_cache_error_notifier(struct notifier_block *nb)
 
 static inline int co_cache_error_call_notifiers(unsigned long val)
 {
-        return raw_notifier_call_chain(&co_cache_error_chain, val, NULL);
+	return raw_notifier_call_chain(&co_cache_error_chain, val, NULL);
 }
 
 static int __devinit co_cache_error_probe(struct platform_device *pdev)
-- 
1.6.5.2

