From 7b2ecc315b62659f0ec5084ac7285a8c1016f2d6 Mon Sep 17 00:00:00 2001
From: Yang Shi <yang.shi@windriver.com>
Date: Thu, 22 Apr 2010 18:27:08 +0800
Subject: [PATCH] lttng: MIPS: Use 64 bit counter for trace clock on Octeon CPUs

Cavium Octeon CPUs have a 64-bit cycle counter that is synchronized
when the CPUs are brought on-line. So for this case we don't need
any fancy stuff.

Signed-off-by: Yang Shi <yang.shi@windriver.com>
---
 include/asm-mips/trace-clock.h |   31 +++++++++++++++++++++++++++++++
 1 files changed, 31 insertions(+), 0 deletions(-)

diff --git a/include/asm-mips/trace-clock.h b/include/asm-mips/trace-clock.h
index 4c96cb5..8d8f01b 100644
--- a/include/asm-mips/trace-clock.h
+++ b/include/asm-mips/trace-clock.h
@@ -12,6 +12,36 @@
 
 extern u64 trace_clock_read_synthetic_tsc(void);
 
+#ifdef CONFIG_CPU_CAVIUM_OCTEON
+
+extern uint64_t octeon_get_clock_rate(void);
+
+static inline u32 trace_clock_read32(void)
+{
+	return (u32)read_c0_cvmcount();
+}
+
+
+static inline u64 trace_clock_read64(void)
+{
+	return read_c0_cvmcount();
+}
+
+static inline void trace_clock_add_timestamp(unsigned long ticks)
+{ }
+
+static inline unsigned int trace_clock_frequency(void)
+{
+	return (unsigned int)octeon_get_clock_rate(); 
+}
+
+static inline u32 trace_clock_freq_scale(void)
+{
+	return 1;
+}
+
+#else
+
 /*
  * MIPS get_cycles only returns a 32 bits TSC (see timex.h). The assumption
  * there is that the reschedule is done every 8 seconds or so. Given that
@@ -40,5 +70,6 @@ static inline u32 trace_clock_freq_scale(void)
 {
 	return 1;
 }
+#endif
 
 #endif /* _ASM_MIPS_TRACE_CLOCK_H */
-- 
1.6.0.4

