From 209dd0c48a8cf491aeacddbc3fa509e41ea87c47 Mon Sep 17 00:00:00 2001
From: ltian <le.tian@windriver.com>
Date: Mon, 15 Nov 2010 14:12:55 +0800
Subject: [PATCH 078/132] RapidIO: Octeon: Fix shutdown code

Signed-off-by: Chad Reese <kreese@caviumnetworks.com>
Signed-off-by: ltian <le.tian@windriver.com>
---
 drivers/rapidio/devices/cn6xxx.c |    5 +----
 1 files changed, 1 insertions(+), 4 deletions(-)

diff --git a/drivers/rapidio/devices/cn6xxx.c b/drivers/rapidio/devices/cn6xxx.c
index ab08d72..82b6d82 100644
--- a/drivers/rapidio/devices/cn6xxx.c
+++ b/drivers/rapidio/devices/cn6xxx.c
@@ -188,7 +188,7 @@ static int cn6xxx_probe(struct rio_dev *dev, const struct rio_device_id *id)
 	if (rio_request_inb_dbell(dev->net->hport, dev, 0, 1,
 		cn6xxx_doorbell)) {
 		dev_err(&dev->dev, "Register for incomming doorbells failed\n");
-		/* return -1; */
+		return -1;
 	}
 
 	for (index = 0; index < 16; index++) {
@@ -213,11 +213,8 @@ static int cn6xxx_probe(struct rio_dev *dev, const struct rio_device_id *id)
 
 static void cn6xxx_remove(struct rio_dev *dev)
 {
-
 	dev_info(&dev->dev, "Removed\n");
 	rio_release_inb_dbell(dev->net->hport, 0, 1);
-	rio_init_dbell_res(&res, 0, 1);
-	rio_release_outb_dbell(dev, &res);
 }
 
 static int cn6xxx_suspend(struct rio_dev *dev, u32 state)
-- 
1.6.5.2

