Subject: rt: irq tracer fix
From: Thomas Gleixner <tglx@linutronix.de>
Date: Fri, 09 Jan 2009 16:16:27 +0100

Signed-off-by: Thomas Gleixner <tglx@linutronix.de>
---
 kernel/irq/handle.c |    3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

Index: linux-2.6.26.8-rt13/kernel/irq/handle.c
===================================================================
--- linux-2.6.26.8-rt13.orig/kernel/irq/handle.c	2009-01-09 17:38:24.000000000 -0500
+++ linux-2.6.26.8-rt13/kernel/irq/handle.c	2009-01-09 17:39:40.000000000 -0500
@@ -164,7 +164,8 @@ irqreturn_t handle_IRQ_event(unsigned in
 	} while (action);
 
 	if (status & IRQF_SAMPLE_RANDOM) {
-		local_irq_enable();
+		if (hardirq_count() && (action->flags & IRQF_DISABLED))
+			local_irq_enable();
 		add_interrupt_randomness(irq);
 	}
 	local_irq_disable();
