From 1f566af1bded919a54b398b6e1e797cdf5607e4f Mon Sep 17 00:00:00 2001
From: Shinya Kuribayashi <shinya.kuribayashi@necel.com>
Date: Mon, 7 Sep 2009 09:46:38 +0900
Subject: [PATCH] MIPS: SS1000: Enable CLO / CLZ instructions

Register scan instructions have been added to the NEC VR5500A processor.
SS1000 processor supports them as well.

CLO  - Count Leading Ones
CLZ  - Count Leading Zeros
DCLO - Count Leading Ones in Doubleword
DCLZ - Count Leading Zeros in Doubleword

Signed-off-by: Shinya Kuribayashi <shinya.kuribayashi@necel.com>
---
 include/asm-mips/bitops.h |    6 ++++--
 1 files changed, 4 insertions(+), 2 deletions(-)

diff --git a/include/asm-mips/bitops.h b/include/asm-mips/bitops.h
index 49df8c4..664d80f 100644
--- a/include/asm-mips/bitops.h
+++ b/include/asm-mips/bitops.h
@@ -558,7 +558,8 @@ static inline void __clear_bit_unlock(unsigned long nr, volatile unsigned long *
 	__clear_bit(nr, addr);
 }
 
-#if defined(CONFIG_CPU_MIPS32) || defined(CONFIG_CPU_MIPS64)
+#if defined(CONFIG_CPU_MIPS32) || defined(CONFIG_CPU_MIPS64) || \
+    defined(CONFIG_CPU_SS1000)
 
 /*
  * Return the bit position (0..63) of the most significant 1 bit in a word
@@ -619,7 +620,8 @@ static inline int fls(int word)
 	return 32 - word;
 }
 
-#if defined(CONFIG_64BIT) && defined(CONFIG_CPU_MIPS64)
+#if defined(CONFIG_64BIT) && \
+   (defined(CONFIG_CPU_MIPS64) || defined(CONFIG_CPU_SS1000))
 static inline int fls64(__u64 word)
 {
 	__asm__("dclz %0, %1" : "=r" (word) : "r" (word));
-- 
1.6.3.3

