From 04fee9ccad8ea4e6c8b242944c7668e3fe839eb8 Mon Sep 17 00:00:00 2001
From: fupan li <fupan.li@windriver.com>
Date: Thu, 20 Mar 2014 09:28:08 +0800
Subject: [PATCH] fix the breaks of 'make headers_install' building rule.

The commit c3335092c31a5eb2f489eda2acabd89aedf447c0 : add DRM/I915 support for Haswell
had move the drm userspace api related header files into include/uapi directory, so the
drm entry in the include/Kbuild files should be removed.

Signed-off-by: fupan li <fupan.li@windriver.com>
---
 Makefile       |    2 ++
 include/Kbuild |    1 -
 2 files changed, 2 insertions(+), 1 deletions(-)

diff --git a/Makefile b/Makefile
index 4971274..0e18b55 100644
--- a/Makefile
+++ b/Makefile
@@ -1074,6 +1074,7 @@ headers_install: __headers
 	$(if $(wildcard $(srctree)/arch/$(hdr-arch)/include/asm/Kbuild),, \
 	$(error Headers not exportable for the $(SRCARCH) architecture))
 	$(Q)$(MAKE) $(hdr-inst)=include
+	$(Q)$(MAKE) $(hdr-inst)=include/uapi
 	$(Q)$(MAKE) $(hdr-inst)=arch/$(hdr-arch)/include/asm $(hdr-dst)
 
 PHONY += headers_check_all
@@ -1083,6 +1084,7 @@ headers_check_all: headers_install_all
 PHONY += headers_check
 headers_check: headers_install
 	$(Q)$(MAKE) $(hdr-inst)=include HDRCHECK=1
+	$(Q)$(MAKE) $(hdr-inst)=include/uapi HDRCHECK=1
 	$(Q)$(MAKE) $(hdr-inst)=arch/$(hdr-arch)/include/asm $(hdr-dst) HDRCHECK=1
 
 # ---------------------------------------------------------------------------
diff --git a/include/Kbuild b/include/Kbuild
index 8d226bf..ed03bef 100644
--- a/include/Kbuild
+++ b/include/Kbuild
@@ -7,6 +7,5 @@ header-y += sound/
 header-y += mtd/
 header-y += rdma/
 header-y += video/
-header-y += drm/
 header-y += xen/
 header-y += scsi/
-- 
1.7.5.4

