From 49871c1ef616b94b1672a1fa5c06bdc95dc9c451 Mon Sep 17 00:00:00 2001
From: auto commit <unknown@unknown>
Date: Wed, 24 Aug 2011 12:53:39 -0700
Subject: [PATCH 92/97] Cavium: Add include file to simple exec

Source: Local development

Addresses build failure due to missing kmalloc defines.

Signed-off-by: Phil Staub <Phil.Staub@windriver.com>
---
 simple_exec/cvmx-helper-ilk.c |    1 +
 1 files changed, 1 insertions(+), 0 deletions(-)

diff --git a/simple_exec/cvmx-helper-ilk.c b/simple_exec/cvmx-helper-ilk.c
index ed5b196..14a5954 100644
--- a/simple_exec/cvmx-helper-ilk.c
+++ b/simple_exec/cvmx-helper-ilk.c
@@ -48,6 +48,7 @@
 
 #ifdef CVMX_BUILD_FOR_LINUX_KERNEL
 #include <linux/module.h>
+#include <linux/slab.h>
 
 #include <asm/octeon/cvmx.h>
 #include <asm/octeon/cvmx-config.h>
-- 
1.7.0

