From 37e41bf6fe604da052baf7b2ca6db3f6f87d4e76 Mon Sep 17 00:00:00 2001
From: Yang Wei <Wei.Yang@windriver.com>
Date: Tue, 26 Aug 2014 18:33:04 +0800
Subject: [PATCH] lib:scatterlist: Do not enable warning if interrupt is
 forcely threaded

For CGL kernel, the device interrupt is forcely threaded, it also means
that interrupt handler always runs with irq enabled context, we should
skip the warning, just like the handle of preempt-rt kernel.

Signed-off-by: Yang Wei <Wei.Yang@windriver.com>
---
 lib/scatterlist.c |    3 ++-
 1 files changed, 2 insertions(+), 1 deletions(-)

diff --git a/lib/scatterlist.c b/lib/scatterlist.c
index 8c2f278..9e8af54 100644
--- a/lib/scatterlist.c
+++ b/lib/scatterlist.c
@@ -11,6 +11,7 @@
 #include <linux/scatterlist.h>
 #include <linux/highmem.h>
 #include <linux/kmemleak.h>
+#include <linux/interrupt.h>
 
 /**
  * sg_next - return the next scatterlist entry in a list
@@ -424,7 +425,7 @@ void sg_miter_stop(struct sg_mapping_iter *miter)
 			flush_kernel_dcache_page(miter->page);
 
 		if (miter->__flags & SG_MITER_ATOMIC) {
-			WARN_ON(!irqs_disabled());
+			WARN_ON(!force_irqthreads && !irqs_disabled());
 			kunmap_atomic(miter->addr);
 		} else
 			kunmap(miter->page);
-- 
1.7.5.4

