From cea2866c8ec9b1c017a2e5a5b9388748add0fb6d Mon Sep 17 00:00:00 2001
From: Yang Wei <Wei.Yang@windriver.com>
Date: Tue, 26 Aug 2014 14:18:08 +0800
Subject: [PATCH] mmc/octeon: Delete a unused +

Signed-off-by: Yang Wei <Wei.Yang@windriver.com>
---
 drivers/mmc/host/octeon_mmc.c |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/drivers/mmc/host/octeon_mmc.c b/drivers/mmc/host/octeon_mmc.c
index 50616d7..38a480c 100644
--- a/drivers/mmc/host/octeon_mmc.c
+++ b/drivers/mmc/host/octeon_mmc.c
@@ -699,7 +699,7 @@ static void octeon_mmc_request(struct mmc_host *mmc, struct mmc_request *mrq)
 		octeon_mmc_dbg("command has data\n");
 		if (cmd->data->flags & MMC_DATA_READ) {
 			sg_miter_start(&host->smi, mrq->data->sg,
-+				       mrq->data->sg_len, SG_MITER_ATOMIC | SG_MITER_TO_SG);
+				       mrq->data->sg_len, SG_MITER_ATOMIC | SG_MITER_TO_SG);
 	} else {
 			struct sg_mapping_iter *smi = &host->smi;
 			unsigned int data_len = mrq->data->blksz * mrq->data->blocks;
-- 
1.7.5.4

