From b640dc7b9281fd60cb8a9e2155e8e7a7b82d2e3d Mon Sep 17 00:00:00 2001
From: Paul Butler <paul.butler@windriver.com>
Date: Thu, 7 Mar 2013 19:25:26 -0600
Subject: [PATCH 10/76] arch/arm/mach-axxia/Makefile: adding i2c

Signed-off-by: Paul Butler <paul.butler@windriver.com>
---
 arch/arm/mach-axxia/Makefile |    1 +
 1 files changed, 1 insertions(+), 0 deletions(-)

diff --git a/arch/arm/mach-axxia/Makefile b/arch/arm/mach-axxia/Makefile
index 6e157dd..c124b4b 100644
--- a/arch/arm/mach-axxia/Makefile
+++ b/arch/arm/mach-axxia/Makefile
@@ -6,5 +6,6 @@ obj-y					+= clock.o
 obj-y                                   += io.o
 obj-y					+= timers.o
 obj-y					+= pci.o
+obj-$(CONFIG_I2C)			+= i2c.o
 obj-$(CONFIG_SMP)			+= platsmp.o headsmp.o
 obj-$(CONFIG_HOTPLUG_CPU)		+= hotplug.o
-- 
1.7.0.5

