From 8fbf16cd79fcd38869895c35e9baf52855f41ff3 Mon Sep 17 00:00:00 2001
From: Paul Butler <paul.butler@windriver.com>
Date: Thu, 7 Mar 2013 19:26:02 -0600
Subject: [PATCH 11/76] arch/arm/tools/mach-types: adding axxia in the mach-types

Signed-off-by: Paul Butler <paul.butler@windriver.com>
---
 arch/arm/tools/mach-types |    1 +
 1 files changed, 1 insertions(+), 0 deletions(-)

diff --git a/arch/arm/tools/mach-types b/arch/arm/tools/mach-types
index 6ed550c..863c920 100644
--- a/arch/arm/tools/mach-types
+++ b/arch/arm/tools/mach-types
@@ -1209,3 +1209,4 @@ familybox		MACH_FAMILYBOX		FAMILYBOX		4170
 ensemble_mx35		MACH_ENSEMBLE_MX35	ENSEMBLE_MX35		4171
 sc_sps_1		MACH_SC_SPS_1		SC_SPS_1		4172
 spear1310_reva_evb	MACH_SPEAR1310_REVA_EVB	SPEAR1310_REVA_EVB	7777
+axxia			MACH_AXXIA		AXXIA			4173
-- 
1.7.0.5

