From 1b697b5566dd0ac2f7f681b6b7916bd0f36824b5 Mon Sep 17 00:00:00 2001
From: Michael Bringmann <michael.bringmann@lsi.com>
Date: Thu, 4 Apr 2013 19:44:54 -0500
Subject: [PATCH 30/76] mach-axxia/i2c: remove unnecessary code

Signed-off-by: Michael Bringmann <michael.bringmann@lsi.com>
---
 arch/arm/mach-axxia/i2c.h |   14 ++------------
 1 files changed, 2 insertions(+), 12 deletions(-)

diff --git a/arch/arm/mach-axxia/i2c.h b/arch/arm/mach-axxia/i2c.h
index cc88104..8e31379 100644
--- a/arch/arm/mach-axxia/i2c.h
+++ b/arch/arm/mach-axxia/i2c.h
@@ -23,23 +23,13 @@
 
 
 /*
- * Minimum number of I2C busses to expect for an AXXIA platform.
- * Runtime checks will determine the actual number of busses that
- * are present depending upon chip type / chip version / etc.
+ * Default bus id to expect for an AXXIA platform.
  */
 #define ARCH_AXXIA_MAX_I2C_BUSSES       1
 #define ARCH_AXXIA_MAX_I2C_BUS_NR       2
 
-/*
- * Maximum byte size of I2C bus name string including null terminator
- */
-#define MAX_AXXIA_I2C_HWMOD_NAME_LEN    16
-
-
-#define AXXIA_I2C_SIZE                  0x0000003f
-#define AXXIA1_I2C_BASE                 0x00000000
-
 
 extern int axxia_register_i2c_busses(void);
 
+
 #endif /* __ASM__ARCH_AXXIA_I2C_H */
-- 
1.7.0.5

