From 7e17c6168c34db13529cf64a878aa689d0018acb Mon Sep 17 00:00:00 2001
From: Michael Bringmann <michael.bringmann@lsi.com>
Date: Thu, 4 Apr 2013 20:00:50 -0500
Subject: [PATCH 36/76] drivers/i2c/ai2c: Fix compile warning about unneeded label

Signed-off-by: Michael Bringmann <michael.bringmann@lsi.com>
---
 drivers/i2c/busses/ai2c/ai2c_mod.c |    1 -
 1 files changed, 0 insertions(+), 1 deletions(-)

diff --git a/drivers/i2c/busses/ai2c/ai2c_mod.c b/drivers/i2c/busses/ai2c/ai2c_mod.c
index 2882668..5dd856a 100644
--- a/drivers/i2c/busses/ai2c/ai2c_mod.c
+++ b/drivers/i2c/busses/ai2c/ai2c_mod.c
@@ -167,7 +167,6 @@ static int ai2c_master_xfer(
 		}
 	}
 
-ai2c_return:
 	AI2C_LOG(AI2C_MSG_EXIT, ">>>Exit ai2c_master_xfer %d\n", err);
 	return err;
 }
-- 
1.7.0.5

