From 0495dfd9b3c5db0aa1c3d3810322e7b2ead5d86d Mon Sep 17 00:00:00 2001
From: Paul Butler <paul.butler@windriver.com>
Date: Thu, 7 Mar 2013 15:45:46 -0600
Subject: [PATCH 72/76] drivers/usb/host/ehci-ci13612.c: replaced bugzilla comments

Signed-off-by: Paul Butler <paul.butler@windriver.com>
---
 drivers/usb/host/ehci-ci13612.c |    7 +++----
 1 files changed, 3 insertions(+), 4 deletions(-)

diff --git a/drivers/usb/host/ehci-ci13612.c b/drivers/usb/host/ehci-ci13612.c
index 5d3a6bc..151376c 100644
--- a/drivers/usb/host/ehci-ci13612.c
+++ b/drivers/usb/host/ehci-ci13612.c
@@ -27,13 +27,12 @@
 #include "ehci-ci13612.h"
 
 
-/* Patch the code to fix the bugs in Bugzilla */
 static void ci13612_usb_setup(struct usb_hcd *hcd)
 {
 	int USB_TXFIFOTHRES, VUSB_HS_TX_BURST;
 
-	/* fix Bugzilla #31874 */
-	/* fix Bugzilla #32212 */
+	/* Fix a HW erratum where the USB core may overrun its transmit FIFO. */
+	/* Fix a HW erratum where the USB core may incorrectly fill its transmit FIFO. */
 	VUSB_HS_TX_BURST = inl(USB_HWTXBUF) & 0x0f;
 	USB_TXFIFOTHRES = (inl(USB_TXFILLTUNING) & 0x3f0000) >> 16;
 
@@ -94,7 +93,7 @@ static int ehci_run_fix(struct usb_hcd *hcd)
 	unsigned burst_size;
 	int retval;
 
-	/* fix Bugzilla 33669 */
+	/* Fix a HW erratum during the USB reset process. */
 	port_status = ehci_readl(ehci, &ehci->regs->port_status[0]);
 	printk(KERN_INFO "ehci_run: port_status = 0x%x\n", port_status);
 	if (port_status & 0x100) {
-- 
1.7.0.5

