From 8981c153ea2dae8a85cab6279c85931f92f91a3e Mon Sep 17 00:00:00 2001
From: Ionut Neicu <ineicu@windriver.com>
Date: Thu, 12 Nov 2015 12:53:46 +0200
Subject: [PATCH 4/6] Add support for TI TCA9546A I2C switch

Signed-off-by: Ionut Neicu <ineicu@windriver.com>
---
 drivers/i2c/muxes/i2c-mux-pca954x.c |    6 ++++++
 1 files changed, 6 insertions(+), 0 deletions(-)

diff --git a/drivers/i2c/muxes/i2c-mux-pca954x.c b/drivers/i2c/muxes/i2c-mux-pca954x.c
index 550bd36..d8ea747 100644
--- a/drivers/i2c/muxes/i2c-mux-pca954x.c
+++ b/drivers/i2c/muxes/i2c-mux-pca954x.c
@@ -55,6 +55,7 @@ enum pca_type {
 	pca_9546,
 	pca_9547,
 	pca_9548,
+	tca_9546,
 };
 
 struct pca954x {
@@ -102,6 +103,10 @@ static const struct chip_desc chips[] = {
 		.nchans = 8,
 		.muxtype = pca954x_isswi,
 	},
+	[tca_9546] = {
+		.nchans = 4,
+		.muxtype = pca954x_isswi,
+	},
 };
 
 static const struct i2c_device_id pca954x_id[] = {
@@ -113,6 +118,7 @@ static const struct i2c_device_id pca954x_id[] = {
 	{ "pca9546", pca_9545 },
 	{ "pca9547", pca_9547 },
 	{ "pca9548", pca_9548 },
+	{ "tca9546", tca_9546 },
 	{ }
 };
 MODULE_DEVICE_TABLE(i2c, pca954x_id);
-- 
1.7.5.4

