From a196c917223310f77c4d67ed501594aa5653fbf7 Mon Sep 17 00:00:00 2001
From: Lans Zhang <jia.zhang@windriver.com>
Date: Thu, 25 Jul 2013 17:30:02 +0800
Subject: [PATCH 1181/1187] clanton: fix dependency.

Signed-off-by: Lans Zhang <jia.zhang@windriver.com>
---
 drivers/platform/x86/Kconfig |    6 +++---
 1 files changed, 3 insertions(+), 3 deletions(-)

diff --git a/drivers/platform/x86/Kconfig b/drivers/platform/x86/Kconfig
index 1c5ce3c..822fbe3 100644
--- a/drivers/platform/x86/Kconfig
+++ b/drivers/platform/x86/Kconfig
@@ -700,7 +700,7 @@ config INTEL_CLN_ECC_REFRESH_PERIOD
 	  hours.
 
 config INTEL_CLN_ECC_SCRUB
-	bool "Hardware ECC Scrub - /proc interface for Intel Clanton platform"
+	tristate "Hardware ECC Scrub - /proc interface for Intel Clanton platform"
 	depends on INTEL_CLN_SOC
 	help
 	   Say Y here to enable support for accessing the hardware memory
@@ -735,8 +735,8 @@ config INTEL_CLN_ECC_SCRUB_S3_CONFIG
 	  hardware memory ECC Scrubber.
 
 config INTEL_CLN_THERMAL
-	bool "Thermal driver for Intel Clanton platform"
-	depends on INTEL_CLN_SOC
+	tristate "Thermal driver for Intel Clanton platform"
+	depends on INTEL_CLN_SOC && THERMAL
 	help
 	  Say Y here to enable Clanton's Thermal driver plus the MSI's
 	  that can be hooked from the thermal sub-system
-- 
1.7.5.4

