From 1527e81184bb1e042433c2de47f8eef6e398e553 Mon Sep 17 00:00:00 2001
From: Lans Zhang <jia.zhang@windriver.com>
Date: Wed, 7 Jan 2015 15:05:20 +0800
Subject: [PATCH 30/31] Enable COMMON_CLK for Quark.

Signed-off-by: Lans Zhang <jia.zhang@windriver.com>
---
 arch/x86/Kconfig |    1 +
 1 files changed, 1 insertions(+), 0 deletions(-)

diff --git a/arch/x86/Kconfig b/arch/x86/Kconfig
index 500574a..7fe299b 100644
--- a/arch/x86/Kconfig
+++ b/arch/x86/Kconfig
@@ -437,6 +437,7 @@ config INTEL_QUARK_X1000_SOC
 	depends on M586TSC
 	select ARCH_REQUIRE_GPIOLIB
 	select I2C
+	select COMMON_CLK
 	---help---
 	  Quark X1000 SOC support . This option enables probing for various
 	  PCI-IDs of several on-chip devices provided by the X1000
-- 
1.7.5.4

