From fdd2e71b923bcaf0bd36fc3850afae567d48c4ed Mon Sep 17 00:00:00 2001
From: chunguang yang <chunguang.yang@windriver.com>
Date: Fri, 23 Jan 2015 14:33:19 +0800
Subject: [PATCH] Fix license conflict issue.

When built with CONFIG_PROVE_RCU enabled, cryptodev
module will have build error like this, due to it
uses GPL-only symbol.

FATAL: modpost: GPL-incompatible module cryptodev.ko uses GPL-only
symbol 'rcu_lock_map'
---
 crypto/ocf/Kconfig |    1 +
 1 files changed, 1 insertions(+), 0 deletions(-)

diff --git a/crypto/ocf/Kconfig b/crypto/ocf/Kconfig
index 6642997..b22f76c 100644
--- a/crypto/ocf/Kconfig
+++ b/crypto/ocf/Kconfig
@@ -2,6 +2,7 @@ menu "OCF Configuration"
 
 config OCF_OCF
 	tristate "OCF (Open Cryptograhic Framework)"
+	depends on !CONFIG_PROVE_RCU
 	help
 	  A linux port of the OpenBSD/FreeBSD crypto framework.
 
-- 
1.7.5.4

