From c0cf39b7ec2c21c480ddecc1f154bd0a422d622b Mon Sep 17 00:00:00 2001
From: Ionut Neicu <ineicu@windriver.com>
Date: Thu, 12 Nov 2015 13:46:37 +0200
Subject: [PATCH 6/6] fix compile error when CONFIG_PCAL_6416A is active

Signed-off-by: Ionut Neicu <ineicu@windriver.com>
---
 drivers/gpio/gpio-pcal6416a.c |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/drivers/gpio/gpio-pcal6416a.c b/drivers/gpio/gpio-pcal6416a.c
index 8d4e21d..d7dea1a 100644
--- a/drivers/gpio/gpio-pcal6416a.c
+++ b/drivers/gpio/gpio-pcal6416a.c
@@ -450,7 +450,7 @@ pcal6416a_get_alt_pdata(struct i2c_client *client)
 }
 #endif
 
-static int __devinit pcal6416a_probe(struct i2c_client *client,
+static int pcal6416a_probe(struct i2c_client *client,
 				   const struct i2c_device_id *id)
 {
 	struct pcal6416a_platform_data *pdata;
-- 
1.7.5.4

