From aca7af773db8d269e593a17783bf8880c88754c2 Mon Sep 17 00:00:00 2001
From: Lans Zhang <jia.zhang@windriver.com>
Date: Mon, 11 Nov 2013 13:25:20 +0800
Subject: [PATCH 12/12] Fix gip test dependency.

Signed-off-by: Lans Zhang <jia.zhang@windriver.com>
---
 drivers/mfd/Kconfig |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/drivers/mfd/Kconfig b/drivers/mfd/Kconfig
index 05d1209..4fd02b5 100644
--- a/drivers/mfd/Kconfig
+++ b/drivers/mfd/Kconfig
@@ -763,7 +763,7 @@ config INTEL_CLN_GIP
 config INTEL_CLN_GIP_TEST
 	tristate "Intel Clanton GIP support for Integration Testing"
 	depends on INTEL_CLN_GIP
-	depends on INTEL_QUARK_X1000_SOC_FPGAEMU || INTEL_QUARK_X1000_SOC_SVP
+	depends on INTEL_QUARK_X1000_SOC
 	select I2C_CHARDEV
 	select GPIO_SYSFS
 	select SPI
-- 
1.7.5.4

