From 92e8a0ba296f2ca753b1d139ce4940f37a2188b9 Mon Sep 17 00:00:00 2001
From: Peter Meerwald <pmeerw@pmeerw.net>
Date: Mon, 18 Jun 2012 20:33:03 +0200
Subject: [PATCH 0749/1187] iio: remove extra ; after function definition

commit 19ea4752fc1f45c778ae66443c1023aced14ba05	upstream

Signed-off-by: Peter Meerwald <pmeerw@pmeerw.net>
Acked-by: Jonathan Cameron <jic23@kernel.org>
Signed-off-by: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
Signed-off-by: Lans Zhang <jia.zhang@windriver.com>
---
 include/linux/iio/buffer.h |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/include/linux/iio/buffer.h b/include/linux/iio/buffer.h
index fb0fe46..ad4fb1a 100644
--- a/include/linux/iio/buffer.h
+++ b/include/linux/iio/buffer.h
@@ -184,7 +184,7 @@ static inline int iio_buffer_register(struct iio_dev *indio_dev,
 }
 
 static inline void iio_buffer_unregister(struct iio_dev *indio_dev)
-{};
+{}
 
 #endif /* CONFIG_IIO_BUFFER */
 
-- 
1.7.5.4

