From 21ea851ac54e5ace9cbf88a2cfe35b404baafb2b Mon Sep 17 00:00:00 2001
From: Pengyu Ma <pengyu.ma@windriver.com>
Date: Tue, 21 Apr 2015 16:46:51 +0800
Subject: [PATCH 31/32] iio/accel: update the new quark kconfig name in
 dependence

Signed-off-by: Pengyu Ma <pengyu.ma@windriver.com>
---
 drivers/iio/accel/Kconfig |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/drivers/iio/accel/Kconfig b/drivers/iio/accel/Kconfig
index 1bff06c..25bc842 100644
--- a/drivers/iio/accel/Kconfig
+++ b/drivers/iio/accel/Kconfig
@@ -30,7 +30,7 @@ config HID_SENSOR_ACCEL_3D
 
 config IIO_LIS331DLH_INTEL_QRK
         tristate "STMicroelectronics LIS331DLH accelerometer i2c driver for Intel Quark platform"
-        depends on INTEL_QUARK_X1000_SOC
+        depends on X86_INTEL_QUARK
 	depends on I2C && SYSFS
 	select IIO_ST_SENSORS_CORE
         help
-- 
1.7.5.4

