From a0fe409c3daf5219c08ebdacdde370a2ce9fe7fa Mon Sep 17 00:00:00 2001
From: Pengyu Ma <pengyu.ma@windriver.com>
Date: Sun, 6 Sep 2015 03:47:18 -0400
Subject: [PATCH] intel-quark: esram: use linux header

replace asm-generic/uaccess.h by linux/uaccess.h

Signed-off-by: Pengyu Ma <pengyu.ma@windriver.com>

diff --git a/arch/x86/platform/intel-quark/intel_qrk_esram.c b/arch/x86/platform/intel-quark/intel_qrk_esram.c
index c905ee773e24..ca130b7ce3dd 100644
--- a/arch/x86/platform/intel-quark/intel_qrk_esram.c
+++ b/arch/x86/platform/intel-quark/intel_qrk_esram.c
@@ -38,7 +38,7 @@
 #include <asm/io.h>
 #include <asm/pgtable.h>
 #include <asm/special_insns.h>
-#include <asm-generic/uaccess.h>
+#include <linux/uaccess.h>
 #include <asm/iosf_mbi.h>
 #include <linux/io.h>
 #include <linux/delay.h>
-- 
2.1.0

