From 6be9b236b0c6421d4385b20d3c620d14a0db136c Mon Sep 17 00:00:00 2001
From: Lans Zhang <jia.zhang@windriver.com>
Date: Thu, 23 Jan 2014 12:59:53 +0800
Subject: [PATCH 4/8] intel-quark: fix thermal driver dependency.

Signed-off-by: Lans Zhang <jia.zhang@windriver.com>
---
 drivers/platform/x86/quark/Kconfig |    4 ++--
 1 files changed, 2 insertions(+), 2 deletions(-)

diff --git a/drivers/platform/x86/quark/Kconfig b/drivers/platform/x86/quark/Kconfig
index dffa0fc..2565830 100644
--- a/drivers/platform/x86/quark/Kconfig
+++ b/drivers/platform/x86/quark/Kconfig
@@ -27,8 +27,8 @@ config INTEL_CLN_ECC_REFRESH_PERIOD
 	  hours.
 
 config INTEL_CLN_THERMAL
-	bool "Thermal driver for Intel Clanton platform"
-	depends on INTEL_QUARK_X1000_SOC
+	tristate "Thermal driver for Intel Clanton platform"
+	depends on INTEL_QUARK_X1000_SOC && THERMAL
 	help
 	  Say Y here to enable Clanton's Thermal driver plus the MSI's
 	  that can be hooked from the thermal sub-system
-- 
1.7.5.4

