From 947bcc467e095d5c3628ce6c8f97e170dc679daa Mon Sep 17 00:00:00 2001
From: Lans Zhang <jia.zhang@windriver.com>
Date: Thu, 6 Mar 2014 12:28:31 +0800
Subject: [PATCH 2/7] intel-quark: fix thermal driver dependency.

Signed-off-by: Lans Zhang <jia.zhang@windriver.com>
---
 drivers/platform/x86/quark/Kconfig |    4 ++--
 1 files changed, 2 insertions(+), 2 deletions(-)

diff --git a/drivers/platform/x86/quark/Kconfig b/drivers/platform/x86/quark/Kconfig
index 0e2c722..c389434 100644
--- a/drivers/platform/x86/quark/Kconfig
+++ b/drivers/platform/x86/quark/Kconfig
@@ -27,8 +27,8 @@ config INTEL_QRK_ECC_REFRESH_PERIOD
 	  hours.
 
 config INTEL_QRK_THERMAL
-	bool "Thermal driver for Intel Quark platform"
-	depends on INTEL_QUARK_X1000_SOC
+	tristate "Thermal driver for Intel Quark platform"
+	depends on INTEL_QUARK_X1000_SOC && THERMAL
 	help
 	  Say Y here to enable Quark's Thermal driver plus the MSI's
 	  that can be hooked from the thermal sub-system
-- 
1.7.5.4

