From c9449cf64dc1f419ecb3059d9c0aedce45916481 Mon Sep 17 00:00:00 2001
From: Lans Zhang <jia.zhang@windriver.com>
Date: Mon, 27 Jan 2014 13:03:14 +0800
Subject: [PATCH 2/2] intel-quark: fix 8250 serial driver build error.

Signed-off-by: Lans Zhang <jia.zhang@windriver.com>
---
 drivers/tty/serial/8250/8250_pci.c |    4 ++--
 1 files changed, 2 insertions(+), 2 deletions(-)

diff --git a/drivers/tty/serial/8250/8250_pci.c b/drivers/tty/serial/8250/8250_pci.c
index 03b4bb4..58cd75d 100644
--- a/drivers/tty/serial/8250/8250_pci.c
+++ b/drivers/tty/serial/8250/8250_pci.c
@@ -167,12 +167,12 @@ afavlab_setup(struct serial_private *priv, const struct pciserial_board *board,
 static int
 pci_intel_cln_setup(struct serial_private *priv,
 		const struct pciserial_board *board,
-		struct uart_8250_port *port, int idx)
+		struct uart_port *port, int idx)
 {
 	unsigned int bar, offset = board->first_offset;
 	bar = FL_GET_BASE(board->flags);
 
-	return setup_port(priv, &port->port, bar, offset, board->reg_shift);
+	return setup_port(priv, port, bar, offset, board->reg_shift);
 }
 
 /*
-- 
1.7.5.4

