From 6a093e37c153f469d82aa3ec6e23067569ace89c Mon Sep 17 00:00:00 2001
From: Lans Zhang <jia.zhang@windriver.com>
Date: Sat, 9 Nov 2013 15:03:57 +0800
Subject: [PATCH 05/12] intel-quark: fix driver dependency.

Signed-off-by: Lans Zhang <jia.zhang@windriver.com>
---
 drivers/platform/x86/quark/Kconfig |    6 +++---
 1 files changed, 3 insertions(+), 3 deletions(-)

diff --git a/drivers/platform/x86/quark/Kconfig b/drivers/platform/x86/quark/Kconfig
index 5d47629..04bfca1d 100644
--- a/drivers/platform/x86/quark/Kconfig
+++ b/drivers/platform/x86/quark/Kconfig
@@ -27,7 +27,7 @@ config INTEL_CLN_ECC_REFRESH_PERIOD
 	  hours.
 
 config INTEL_CLN_ECC_SCRUB
-	bool "Hardware ECC Scrub - /proc interface for Intel Clanton platform"
+	tristate "Hardware ECC Scrub - /proc interface for Intel Clanton platform"
 	depends on INTEL_QUARK_X1000_SOC
 	help
 	   Say Y here to enable support for accessing the hardware memory
@@ -62,8 +62,8 @@ config INTEL_CLN_ECC_SCRUB_S3_CONFIG
 	  hardware memory ECC Scrubber.
 
 config INTEL_CLN_THERMAL
-	bool "Thermal driver for Intel Clanton platform"
-	depends on INTEL_QUARK_X1000_SOC
+	tristate "Thermal driver for Intel Clanton platform"
+	depends on INTEL_QUARK_X1000_SOC && THERMAL
 	help
 	  Say Y here to enable Clanton's Thermal driver plus the MSI's
 	  that can be hooked from the thermal sub-system
-- 
1.7.5.4

