From 31993b63dd1697ad91c41b78617fc1453ffacc8a Mon Sep 17 00:00:00 2001
From: Lans Zhang <jia.zhang@windriver.com>
Date: Thu, 23 Jan 2014 13:42:34 +0800
Subject: [PATCH 8/8] intel-quark: workaround to silence the interrupts from
 gpe0E.

Signed-off-by: Lans Zhang <jia.zhang@windriver.com>
---
 drivers/gpio/gpio-sch.c |   11 +++++++++++
 1 files changed, 11 insertions(+), 0 deletions(-)

diff --git a/drivers/gpio/gpio-sch.c b/drivers/gpio/gpio-sch.c
index 807d35a..4482f7e 100644
--- a/drivers/gpio/gpio-sch.c
+++ b/drivers/gpio/gpio-sch.c
@@ -34,10 +34,12 @@ static DEFINE_SPINLOCK(gpio_lock);
 #define CGEN	(0x00)
 #define CGIO	(0x04)
 #define CGLV	(0x08)
+#define CGPE	(0x14)
 
 #define RGEN	(0x20)
 #define RGIO	(0x24)
 #define RGLV	(0x28)
+#define RGPE	(0x34)
 
 static unsigned short gpio_ba;
 
@@ -254,6 +256,15 @@ static int __devinit sch_gpio_probe(struct platform_device *pdev)
 	sch_gpio_core.dev = &pdev->dev;
 	sch_gpio_resume.dev = &pdev->dev;
 
+	pr_info("Initial Core-Well GPE %x Resume-Well GPE %x\n",
+		inb(gpio_ba + CGPE), inb(gpio_ba + RGPE));
+
+	outb(0, gpio_ba + CGPE);
+	outb(0, gpio_ba + RGPE);
+
+	pr_info("New Core-Well GPE %x Resume-Well GPE %x\n",
+		inb(gpio_ba + CGPE), inb(gpio_ba + RGPE));
+
 	err = gpiochip_add(&sch_gpio_core);
 	if (err < 0)
 		goto err_sch_gpio_core;
-- 
1.7.5.4

