From f0fad33e696f72bc40805ca49fd768f7ed2bef90 Mon Sep 17 00:00:00 2001
From: Lans Zhang <jia.zhang@windriver.com>
Date: Sat, 9 Nov 2013 15:01:20 +0800
Subject: [PATCH 03/12] intel-quark: fix build warning for 8250_pci.c.

Signed-off-by: Lans Zhang <jia.zhang@windriver.com>
---
 drivers/tty/serial/8250/8250_pci.c |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/drivers/tty/serial/8250/8250_pci.c b/drivers/tty/serial/8250/8250_pci.c
index 518a797..03b4bb4 100644
--- a/drivers/tty/serial/8250/8250_pci.c
+++ b/drivers/tty/serial/8250/8250_pci.c
@@ -172,7 +172,7 @@ pci_intel_cln_setup(struct serial_private *priv,
 	unsigned int bar, offset = board->first_offset;
 	bar = FL_GET_BASE(board->flags);
 
-	return setup_port(priv, port, bar, offset, board->reg_shift);
+	return setup_port(priv, &port->port, bar, offset, board->reg_shift);
 }
 
 /*
-- 
1.7.5.4

