From 238e96fb704d9346aaa259e39f14186541addd32 Mon Sep 17 00:00:00 2001
From: Lans Zhang <jia.zhang@windriver.com>
Date: Wed, 2 Oct 2013 14:14:39 +0800
Subject: [PATCH 1178/1187] Quark: add mtd support.

Signed-off-by: Lans Zhang <jia.zhang@windriver.com>
---
 drivers/mtd/devices/Kconfig |    5 +++++
 1 files changed, 5 insertions(+), 0 deletions(-)

diff --git a/drivers/mtd/devices/Kconfig b/drivers/mtd/devices/Kconfig
index 4cdb2af..137dbe9 100644
--- a/drivers/mtd/devices/Kconfig
+++ b/drivers/mtd/devices/Kconfig
@@ -151,6 +151,11 @@ config MTD_MTDRAM
 	  provide storage.  You probably want to say 'N' unless you're
 	  testing stuff.
 
+config MTD_MTD_CLN_ROM
+	bool "Simple R/O drive for SPI flash in Clanton x86 legacy block"
+	help
+	  Driver to enable reading directly from legacy block SPI /sketch part
+
 config MTDRAM_TOTAL_SIZE
 	int "MTDRAM device size in KiB"
 	depends on MTD_MTDRAM
-- 
1.7.5.4

