From 58ec020221f94adda16e5a3c47814264ca9ecae4 Mon Sep 17 00:00:00 2001
From: Jonathan Cameron <jic23@kernel.org>
Date: Sat, 6 Oct 2012 14:38:11 +0100
Subject: [PATCH 0985/1187] staging:iio:adc:max1363 white space cleanup

commit 0d9fa2ce939279ac7aab34b50b90196f8208af42	upstream

Signed-off-by: Jonathan Cameron <jic23@kernel.org>
Signed-off-by: Lans Zhang <jia.zhang@windriver.com>
---
 drivers/staging/iio/adc/max1363.c |    6 +-----
 1 files changed, 1 insertions(+), 5 deletions(-)

diff --git a/drivers/staging/iio/adc/max1363.c b/drivers/staging/iio/adc/max1363.c
index 3217753..2f8d433 100644
--- a/drivers/staging/iio/adc/max1363.c
+++ b/drivers/staging/iio/adc/max1363.c
@@ -765,9 +765,9 @@ static int max1363_read_event_config(struct iio_dev *indio_dev,
 				     u64 event_code)
 {
 	struct max1363_state *st = iio_priv(indio_dev);
-
 	int val;
 	int number = IIO_EVENT_CODE_EXTRACT_CHAN(event_code);
+
 	mutex_lock(&indio_dev->mlock);
 	if (IIO_EVENT_CODE_EXTRACT_DIR(event_code) == IIO_EV_DIR_FALLING)
 		val = (1 << number) & st->mask_low;
@@ -969,8 +969,6 @@ static struct attribute_group max1363_event_attribute_group = {
 	.name = "events",
 };
 
-#define MAX1363_EVENT_FUNCS						\
-
 static int max1363_update_scan_mode(struct iio_dev *indio_dev,
 				    const unsigned long *scan_mask)
 {
@@ -1388,8 +1386,6 @@ static const struct max1363_chip_info max1363_chip_info_tbl[] = {
 	},
 };
 
-
-
 static int max1363_initial_setup(struct max1363_state *st)
 {
 	st->setupbyte = MAX1363_SETUP_AIN3_IS_AIN3_REF_IS_VDD
-- 
1.7.5.4

