From 3a74c4fba7e315c550fdbc6ca91a66bb50b1b912 Mon Sep 17 00:00:00 2001
From: Yue Tao <yue.tao@windriver.com>
Date: Mon, 19 Aug 2013 18:27:22 -0700
Subject: [PATCH] checkpoint: Mon Aug 19 18:27:22 PDT 2013

---
 wrs/cfg/branch_status |   11 +++++++++++
 1 files changed, 11 insertions(+), 0 deletions(-)

diff --git a/wrs/cfg/branch_status b/wrs/cfg/branch_status
index 1c2ed71..4b70a33 100644
--- a/wrs/cfg/branch_status
+++ b/wrs/cfg/branch_status
@@ -964,3 +964,14 @@ ecgl:8b5fc82ab6a1b7b2525767d802f67aaa2452ff81
 cgl:cf7eacfce87f2a6eda244d6a96b9c9a957076c31
 cav_octeon-cgl:13e688338159a979695736f97ff34807d35bdbe6
 raza_xlr_atx-cgl:8ab0677caaa91855c60d36359d6a6bb542447675
+small:6ac2c3bccbc71ed31d46b1758a45aa8d2e32c8b2
+fsl_imx-small:7a6cc9916ca1426b887f975390ecf3fae5847093
+standard:5f45d9355b94f9fa4edc31f80aa09eb7d5c5f849
+fsl_imx-standard:ce1eb6b4f4acf33bbec443d18a2e378b299540ab
+cav_octeon-standard:c61f70d67dc4c86c96e5f96e37e55ede972dd803
+raza_xlr_atx-standard:dae7f51f8968ca8c8af2e52b385e97015ee32592
+preempt_rt:e7b904ab2dcf05d80489ed63a7672ab88e846551
+ecgl:fd9a266af790db8fa543ecb51091449bf04e035a
+cgl:fb4d05bb007fb4b89a5660bd862a26ea65df3cce
+cav_octeon-cgl:0a5593e1b706bf8ad703dec2725d148643b4ccdf
+raza_xlr_atx-cgl:9de01d31c5a6c4c5ea2a1d5cb54af23c5d36afc5
-- 
1.7.0

