From c1aaa01c454e8c3ef3c22f216f9677707bc72a03 Mon Sep 17 00:00:00 2001
From: Keith Busch <keith.busch@intel.com>
Date: Tue, 25 Jun 2013 15:15:23 -0600
Subject: [PATCH 32/72] NVMe: Update nvme_id_power_state with latest spec

commit 685585c25e8269cc355b59d1cd6fc65b8c5c4878 upstream

Signed-off-by: Keith Busch <keith.busch@intel.com>
Signed-off-by: Matthew Wilcox <matthew.r.wilcox@intel.com>
Signed-off-by: fupan li <fupan.li@windriver.com>
---
 include/uapi/linux/nvme.h |    8 +++++++-
 1 files changed, 7 insertions(+), 1 deletions(-)

diff --git a/include/uapi/linux/nvme.h b/include/uapi/linux/nvme.h
index 31ed566..989c04e 100644
--- a/include/uapi/linux/nvme.h
+++ b/include/uapi/linux/nvme.h
@@ -23,7 +23,8 @@
 
 struct nvme_id_power_state {
 	__le16			max_power;	/* centiwatts */
-	__u16			rsvd2;
+	__u8			rsvd2;
+	__u8			flags;
 	__le32			entry_lat;	/* microseconds */
 	__le32			exit_lat;	/* microseconds */
 	__u8			read_tput;
@@ -33,6 +34,11 @@ struct nvme_id_power_state {
 	__u8			rsvd16[16];
 };
 
+enum {
+	NVME_PS_FLAGS_MAX_POWER_SCALE	= 1 << 0,
+	NVME_PS_FLAGS_NON_OP_STATE	= 1 << 1,
+};
+
 struct nvme_id_ctrl {
 	__le16			vid;
 	__le16			ssvid;
-- 
1.7.5.4

