From 9268f3b02f7489ef7cde391a5f02b4620fb51b7e Mon Sep 17 00:00:00 2001
From: Xufeng Zhang <xufeng.zhang@windriver.com>
Date: Thu, 7 Jul 2011 16:44:09 +0800
Subject: [PATCH 32/47] powerpc/fsl: 85xx: document cache sram bindings

Add binding documentation for cache sram for QorIQ P1/P2 series platforms.

Extracted from vendor drop QorIQ-NONDPAA-SDK-V1-20110429_ltib.iso
linux-2.6.35-qoriq-powerpc-fsl-85xx-document-cache-sram-bi.patch.

Signed-off-by: Harninder Rai <harninder.rai@freescale.com>
Signed-off-by: Vivek Mahajan <vivek.mahajan@freescale.com>
Integrated-by: Xufeng Zhang <xufeng.zhang@windriver.com>
---
 .../powerpc/dts-bindings/fsl/85xx_cache_sram.txt   |   17 +++++++++++++++++
 1 files changed, 17 insertions(+), 0 deletions(-)
 create mode 100644 Documentation/powerpc/dts-bindings/fsl/85xx_cache_sram.txt

diff --git a/Documentation/powerpc/dts-bindings/fsl/85xx_cache_sram.txt b/Documentation/powerpc/dts-bindings/fsl/85xx_cache_sram.txt
new file mode 100644
index 0000000..d7ab3d9
--- /dev/null
+++ b/Documentation/powerpc/dts-bindings/fsl/85xx_cache_sram.txt
@@ -0,0 +1,17 @@
+Freescale's mpc85xx and some QorIQ platforms provide an option of
+configuring a part or full cache as SRAM. This cache SRAM representation
+in the device tree should be done as under:-
+
+Required properties:
+
+ - compatible: "fsl,l2sram"
+ - reg : offset of the cache-sram
+
+
+Example:
+
+      l2sram@fff00000 {
+              device_type = "memory";
+              compatible = "fsl,l2sram";
+              reg = <0x0 0xfff00000 0 0x10000>;
+      };
-- 
1.7.0.4

