From ca85b779a7d3c658586b4f1488070477e2634fb4 Mon Sep 17 00:00:00 2001
From: Xufeng Zhang <xufeng.zhang@windriver.com>
Date: Mon, 1 Aug 2011 11:05:54 +0800
Subject: [PATCH 22/47] talitos: fix kernel Call Trace when get processor id

talitos call smp_processor_id() to get processor id in some preemptible
contexts, this will lead to following kernel call trace:
BUG: using smp_processor_id() in preemptible [00000000] code: ping/1027
caller is crypto_edesc_alloc+0x24/0xb4
Call Trace:
[eead9a60] [c0007bac] show_stack+0x44/0x160 (unreliable)
[eead9a90] [c037174c] debug_smp_processor_id+0xdc/0xec
[eead9ab0] [c04d57a8] crypto_edesc_alloc+0x24/0xb4
[eead9ad0] [c04d59a0] talitos_edesc_alloc+0x168/0x278
[eead9b00] [c04d67b0] aead_givencrypt+0x3c/0xd0
[eead9b20] [f94b8390] esp_output+0x294/0x308 [esp4]
[eead9b80] [c05a2e48] xfrm_output_resume+0x220/0x3bc
[eead9bc0] [c059702c] xfrm4_output_finish+0x38/0x78
[eead9bd0] [c0554e2c] ip_local_out+0x34/0x48
[eead9be0] [c05550c0] ip_push_pending_frames+0x280/0x3b0
[eead9c10] [c0575350] raw_sendmsg+0x478/0x7bc
[eead9cb0] [c058054c] inet_sendmsg+0x38/0x7c
[eead9cd0] [c050e874] sock_sendmsg+0xc0/0x160
[eead9db0] [c050ebe0] sys_sendmsg+0x1e4/0x2a0
[eead9f00] [c051182c] sys_socketcall+0x1a0/0x320
[eead9f40] [c00118f4] ret_from_syscall+0x0/0x4
---
 drivers/crypto/talitos.c |    6 ++++++
 1 files changed, 6 insertions(+), 0 deletions(-)

diff --git a/drivers/crypto/talitos.c b/drivers/crypto/talitos.c
index 67a43e0..2bd9b6e 100644
--- a/drivers/crypto/talitos.c
+++ b/drivers/crypto/talitos.c
@@ -206,7 +206,9 @@ struct talitos_edesc *crypto_edesc_alloc(int len, int flags,
 {
 	int check;
 	struct talitos_edesc *ret;
+	preempt_disable();
 	u32 smp_processor_id = smp_processor_id();
+	preempt_enable();
 	check = in_softirq();
 
 	if (!check)
@@ -231,7 +233,9 @@ void crypto_edesc_free(struct talitos_edesc *edesc,
 			struct talitos_private *priv)
 {
 	int check;
+	preempt_disable();
 	u32 smp_processor_id = smp_processor_id();
+	preempt_enable();
 	check = in_softirq();
 
 	if (!check)
@@ -431,7 +435,9 @@ static int talitos_submit(struct device *dev, struct talitos_desc *desc,
 	struct talitos_private *priv = dev_get_drvdata(dev);
 	struct talitos_request *request;
 	u8 ch;
+	preempt_disable();
 	int grp_id = get_grp_id(priv);
+	preempt_enable();
 
 	u8 head, last_chan, total_chan;
 	if (priv->core_num_chan[grp_id] > 0) {
-- 
1.7.0.4

