From e58134d12a952864e70c50001b913e76d3b6b33c Mon Sep 17 00:00:00 2001
From: Xufeng Zhang <xufeng.zhang@windriver.com>
Date: Tue, 28 Jun 2011 14:50:29 +0800
Subject: [PATCH 19/47] talitos: document the channel remapping

Add the binding documentation for the channels remapping to allow
sec to map the selected channels to an alternate 4K region.

Extracted from vendor drop QorIQ-NONDPAA-SDK-V1-20110429_ltib.iso
linux-2.6.35-qoriq-talitos-document-the-channel-remapping.patch.

Signed-off-by: Sandeep Malik <Sandeep.Malik@freescale.com>
Integrated-by: Xufeng Zhang <xufeng.zhang@windriver.com>
---
 Documentation/powerpc/dts-bindings/fsl/sec.txt |   11 +++++++++++
 1 files changed, 11 insertions(+), 0 deletions(-)

diff --git a/Documentation/powerpc/dts-bindings/fsl/sec.txt b/Documentation/powerpc/dts-bindings/fsl/sec.txt
index 2b6f2d4..b0119f3 100644
--- a/Documentation/powerpc/dts-bindings/fsl/sec.txt
+++ b/Documentation/powerpc/dts-bindings/fsl/sec.txt
@@ -48,6 +48,16 @@ remaining bits are reserved for future SEC EUs.
 
   ..and so on and so forth.
 
+-fsl,channel-remap : The bitmask representing which descriptors are
+ remapped to alternate 4K memory region. It's a single 32-bit cell. The
+ channel remapping information is passed using the 4 bit channel map and
+ follows the following encoding mechanism:
+
+	bit 0 	= set if channel 0 is remapped to alternate 4K region.
+	bit 1 	= set if channel 1 is remapped to alternate 4K region.
+	bit 2 	= set if channel 2 is remapped to alternate 4K region.
+	bit 3 	= set if channel 3 is remapped to alternate 4K region.
+
 Optional properties:
 
 - interrupt-parent : the phandle for the interrupt controller that
@@ -65,4 +75,5 @@ Example:
 		fsl,channel-fifo-len = <24>;
 		fsl,exec-units-mask = <0xfe>;
 		fsl,descriptor-types-mask = <0x12b0ebf>;
+		fsl,channel-remap = <0x3>;
 	};
-- 
1.7.0.4

