From 08f1d8e453b7fb1b188681c7b6aecac75e5f4595 Mon Sep 17 00:00:00 2001
From: Zumeng Chen <zumeng.chen@windriver.com>
Date: Sat, 21 Aug 2010 07:59:05 -0700
Subject: [PATCH] tun: avoid redundant hash search with invalid name param

In many cases, __dev_get_by_name will return NULL when
ifr->ifr_name is invalid char array after copy_from_user,
which is possible when run tunctl without -t tap0 as
parameter. But as not often situation, the following
oops will be popped at hp_proliant_bl460c:

BUG: unable to handle kernel NULL pointer dereference at (null)
IP: [<ffffffff814f9fda>] __dev_get_by_name+0x6a/0xc0
Pid: 750, comm: tunctl Not tainted 2.6.34.1-WR4.0.0.0_standard
RIP: 0010:[<ffffffff814f9fda>]  [<ffffffff814f9fda>] __dev_get_by_name+0x6a/0xc0
RSP: 0018:ffff88011f143d78  EFLAGS: 00010246
RAX: 0000000000000000 RBX: ffff88011fa334d0 RCX: 0000000000000000
RDX: 0000000000000000 RSI: 0000000000000000 RDI: ffff88011f143dc8
RBP: ffff88011f143d98 R08: 0000000000000000 R09: 0000000000000003
R10: 0000000000000000 R11: 0000000000000202 R12: 00007fffdea25630
R13: ffff88011f143dc8 R14: ffff88011f6bed80 R15: 0000000000000028
FS:  00007f38f93ae700(0000) GS:ffff880001c00000(0000) knlGS:0000000000000000
CS:  0010 DS: 0000 ES: 0000 CR0: 000000008005003b
CR2: 0000000000000000 CR3: 000000011faef000 CR4: 00000000000006f0
DR0: 0000000000000000 DR1: 0000000000000000 DR2: 0000000000000000
DR3: 0000000000000000 DR6: 00000000ffff0ff0 DR7: 0000000000000400
Process tunctl (pid: 750, threadinfo ffff88011f142000, task ffff88011f0c39f0)
Stack:
 00000000400454ca 00000000400454ca 00007fffdea25630 0000000000000000
<0> ffff88011f143e48 ffffffffa00b2db0 0000000000000000 0000000000008002
<0> ffff88011fa334d0 ffff88011f143dc8 0000000000000000 0000000000000000
Call Trace:
 [<ffffffffa00b2db0>] __tun_chr_ioctl+0x1f0/0xba0 [tun]
 [<ffffffffa00b37ae>] tun_chr_ioctl+0xe/0x10 [tun]
 [<ffffffff81150b78>] vfs_ioctl+0x38/0xd0
 [<ffffffff81150d20>] do_vfs_ioctl+0x80/0x560
 [<ffffffff8115a0a2>] ? alloc_fd+0xf2/0x140
 [<ffffffff815d55b6>] ? _raw_spin_lock+0x16/0x40
 [<ffffffff81151289>] sys_ioctl+0x89/0x100
 [<ffffffff810360eb>] system_call_done+0x0/0x5
Obviously, there is NULL pointer dereference in hash search when
ifr.ifr_name is not set.

At the same time, this patch will bypass the redundant hash search
to give the default tap interface name to user.

Signed-Off-By: Zumeng Chen <zumeng.chen@windriver.com>
---
 drivers/net/tun.c |    6 ++++--
 1 files changed, 4 insertions(+), 2 deletions(-)

diff --git a/drivers/net/tun.c b/drivers/net/tun.c
index 4326520..8de6fcf 100644
--- a/drivers/net/tun.c
+++ b/drivers/net/tun.c
@@ -968,10 +968,12 @@ static int tun_set_iff(struct net *net, struct file *file, struct ifreq *ifr)
 {
 	struct sock *sk;
 	struct tun_struct *tun;
-	struct net_device *dev;
+	struct net_device *dev = NULL;
 	int err;
 
-	dev = __dev_get_by_name(net, ifr->ifr_name);
+	if(strlen(&(ifr->ifr_name[0])))
+		dev = __dev_get_by_name(net, ifr->ifr_name);
+
 	if (dev) {
 		const struct cred *cred = current_cred();
 
-- 
1.6.5.2

