From 55c1a216dd6e7a86b1f656e74a973431a8ae3817 Mon Sep 17 00:00:00 2001
From: Yue Tao <Yue.Tao@windriver.com>
Date: Wed, 31 Jul 2013 02:42:17 -0700
Subject: [PATCH] Disable parallel building on firmware

Signed-off-by: Yue Tao <Yue.Tao@windriver.com>
---
 firmware/Makefile |    1 +
 1 files changed, 1 insertions(+), 0 deletions(-)

diff --git a/firmware/Makefile b/firmware/Makefile
index f83a650..d916ffd 100644
--- a/firmware/Makefile
+++ b/firmware/Makefile
@@ -215,6 +215,7 @@ $(patsubst %,$(obj)/%.gen.S, $(fw-external-y)): %: $(wordsize_deps) \
 	$(call cmd,fwbin,$(fwabs)/$(patsubst $(obj)/%.gen.S,%,$@))
 
 # The .o files depend on the binaries directly; the .S files don't.
+.NOTPARALLEL: $(patsubst %,$(obj)/%.gen.o, $(fw-shipped-y))
 $(patsubst %,$(obj)/%.gen.o, $(fw-shipped-y)): %.gen.o: %
 $(patsubst %,$(obj)/%.gen.o, $(fw-external-y)): $(obj)/%.gen.o: $(fwdir)/%
 
-- 
1.7.5.4

