From 6e6685e9cba3f097f7d0fe5c03ca7055c6e40570 Mon Sep 17 00:00:00 2001
From: Zhong Hongbo <hongbo.zhong@windriver.com>
Date: Tue, 28 Jan 2014 13:24:34 +0800
Subject: [PATCH 1063/1072] Disable parallel building on firmware

Signed-off-by: Zhong Hongbo <hongbo.zhong@windriver.com>
---
 firmware/Makefile |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/firmware/Makefile b/firmware/Makefile
index 945d89c..51b4e5f 100644
--- a/firmware/Makefile
+++ b/firmware/Makefile
@@ -225,7 +225,7 @@ $(patsubst %,$(obj)/%.gen.o, $(fw-external-y)): $(obj)/%.gen.o: $(fwdir)/%
 $(obj)/%: $(obj)/%.ihex | $(objtree)/$(obj)/$$(dir %)
 	$(call cmd,ihex)
 
-.NOTPARALLEL: $(obj)/%
+.NOTPARALLEL: $(patsubst %,$(obj)/%.gen.o, $(fw-shipped-y))
 
 # Don't depend on ihex2fw if we're installing and it already exists.
 # Putting it after | in the dependencies doesn't seem sufficient when
-- 
1.7.5.4

