From a0e2a176ba691d251ef81be024d07583a930013d Mon Sep 17 00:00:00 2001
From: Axel Lin <axel.lin@ingics.com>
Date: Thu, 15 Aug 2013 14:06:37 +0800
Subject: [PATCH 1018/1072] spi: bitbang: Remove unused tmp variable

commit 2e29db400c4898ff72e2ea87d262921df00f1c87 upstream

Signed-off-by: Axel Lin <axel.lin@ingics.com>
Signed-off-by: Mark Brown <broonie@linaro.org>
Signed-off-by: Hongbo Zhong <hongbo.zhong@windriver.com>
---
 drivers/spi/spi-bitbang.c |    2 --
 1 files changed, 0 insertions(+), 2 deletions(-)

diff --git a/drivers/spi/spi-bitbang.c b/drivers/spi/spi-bitbang.c
index dd2e5d7..e3946e4 100644
--- a/drivers/spi/spi-bitbang.c
+++ b/drivers/spi/spi-bitbang.c
@@ -276,7 +276,6 @@ static int spi_bitbang_transfer_one(struct spi_master *master,
 	struct spi_bitbang 	*bitbang;
 	unsigned		nsecs;
 	struct spi_transfer	*t = NULL;
-	unsigned		tmp;
 	unsigned		cs_change;
 	int			status;
 	int			do_setup = -1;
@@ -290,7 +289,6 @@ static int spi_bitbang_transfer_one(struct spi_master *master,
 	 */
 	nsecs = 100;
 
-	tmp = 0;
 	cs_change = 1;
 	status = 0;
 
-- 
1.7.5.4

