From db39495ca58ec1353b8efb71ae05dbdaca6c95cb Mon Sep 17 00:00:00 2001
From: Li Jun <jun.li@freescale.com>
Date: Mon, 1 Jun 2015 10:34:42 +0800
Subject: [PATCH 1463/1594] MLK-11005-6 Documentation: usb: chipidea: add usb
 OTG ADP test

commit 8de6b09949cf2c6df24230f67fb9fa29795482da from
git://git.freescale.com/imx/linux-2.6-imx.git

Add how to enable USB OTG ADP feature and test sequence.

Signed-off-by: Li Jun <jun.li@freescale.com>
---
 Documentation/usb/chipidea.txt |   13 +++++++++++++
 1 files changed, 13 insertions(+), 0 deletions(-)

diff --git a/Documentation/usb/chipidea.txt b/Documentation/usb/chipidea.txt
index 44ddaeb..faf16c5 100644
--- a/Documentation/usb/chipidea.txt
+++ b/Documentation/usb/chipidea.txt
@@ -88,6 +88,19 @@ end device by:
 	echo 1 or 0 > /sys/bus/platform/devices/ci_hdrc.0/inputs/hnp_enable
 If this input is 0, only OTG role switch by ID pin detection can be supported.
 
+1.6 Enable and test USB OTG ADP
+----------------------
+Enable:
+1) Enable OTG FSM as 1.1
+2) Add "adp-support" properity in dts for your hardware
+
+Test sequence:
+1) Connect 2 OTG devices, enable HNP by enable hnp_enable at 2 boards
+	echo 1 > /sys/bus/platform/devices/ci_hdrc.0/inputs/hnp_enable
+2) After A-device ends session(remove B-device and wait 10s, or a_bus_drop)
+3) Remove B-device, B-device will show "OTG device detached".
+4) Insert B-device, A-device will enumerate B-device.
+
 2. How to enable USB as system wakeup source
 -----------------------------------
 Below is the example for how to enable USB as system wakeup source
-- 
1.7.5.4

