From 68fbdc55a6e4728415c8a15f78604c61f101bdf0 Mon Sep 17 00:00:00 2001
From: Li Zhou <li.zhou@windriver.com>
Date: Tue, 13 Mar 2018 02:11:02 +0000
Subject: [PATCH] dtc: add an exception for check_pci_device_reg

In Documentation/devicetree/bindings/pci/xilinx-nwl-pcie.txt, there
is an example for xilinux special dt node:
nwl_pcie: pcie@fd0e0000 {
......
         pcie_intc: legacy-interrupt-controller {
                interrupt-controller;
                #address-cells = <0>;
                #interrupt-cells = <1>;
        };
};
For node legacy-interrupt-controller, there isn't any PCI reg property.
While in check_pci_device_reg function to check the device tree, this
node is checked too and a warning is reported as:
arch/arm64/boot/dts/xilinx/zynqmp-zcu102-rev1.0.dtb:
Warning (pci_device_reg): Node /amba/pcie@fd0e0000/legacy-interrupt-controller missing PCI reg property

In fact this check shouldn't be done for legacy-interrupt-controller
node. So add exception for such case: when there isn't any @xxxxx behind
the node name, no need to check the reg property.

Fixes: 33c3985226d3 checks: Add bus checks for PCI buses

Signed-off-by: Li Zhou <li.zhou@windriver.com>
---
 scripts/dtc/checks.c |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/scripts/dtc/checks.c b/scripts/dtc/checks.c
index 5adfc8f..f9484f2 100644
--- a/scripts/dtc/checks.c
+++ b/scripts/dtc/checks.c
@@ -773,7 +773,7 @@ static void check_pci_device_reg(struct check *c, struct dt_info *dti, struct no
 	unsigned int dev, func, reg;
 	cell_t *cells;
 
-	if (!node->parent || (node->parent->bus != &pci_bus))
+	if (!node->parent || (node->parent->bus != &pci_bus) || (*unitname == '\0'))
 		return;
 
 	prop = get_property(node, "reg");
-- 
1.7.5.4

