From 57b79543bb0053b4dc1268ee4edebe5b880b8823 Mon Sep 17 00:00:00 2001
From: Mathieu Desnoyers <mathieu.desnoyers@polymtl.ca>
Date: Fri, 3 Sep 2010 10:17:45 -0400
Subject: [PATCH 002/757] Support architectures without dump tables

Signed-off-by: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
---
 drivers/staging/lttng/ltt-tracer.h | 10 ++++++++++
 1 file changed, 10 insertions(+)

diff --git a/drivers/staging/lttng/ltt-tracer.h b/drivers/staging/lttng/ltt-tracer.h
index 9564c3f..8d979b5 100644
--- a/drivers/staging/lttng/ltt-tracer.h
+++ b/drivers/staging/lttng/ltt-tracer.h
@@ -644,8 +644,18 @@ void ltt_statedump_unregister_kprobes_dump(void (*callback)(void *call_data));
 
 extern void ltt_dump_softirq_vec(void *call_data);
 
+#ifdef CONFIG_HAVE_LTT_DUMP_TABLES
 extern void ltt_dump_sys_call_table(void *call_data);
 extern void ltt_dump_idt_table(void *call_data);
+#else
+static inline void ltt_dump_sys_call_table(void *call_data)
+{
+}
+
+static inline void ltt_dump_idt_table(void *call_data)
+{
+}
+#endif
 
 /* Relay IOCTL */
 
-- 
2.0.2

