From 6af3145d11af93d6d1e45876976c1814b5342ef8 Mon Sep 17 00:00:00 2001
From: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
Date: Mon, 18 Apr 2011 12:27:35 -0400
Subject: [PATCH 055/933] Set FMODE_PREAD for stream_file

Signed-off-by: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
---
 drivers/staging/lttng/ltt-debugfs-abi.c | 6 ++++++
 1 file changed, 6 insertions(+)

diff --git a/drivers/staging/lttng/ltt-debugfs-abi.c b/drivers/staging/lttng/ltt-debugfs-abi.c
index 0ac8861..f5ead59 100644
--- a/drivers/staging/lttng/ltt-debugfs-abi.c
+++ b/drivers/staging/lttng/ltt-debugfs-abi.c
@@ -240,6 +240,12 @@ int lttng_abi_open_stream(struct file *channel_file)
 		ret = PTR_ERR(stream_file);
 		goto file_error;
 	}
+	/*
+	 * OPEN_FMODE, called within anon_inode_getfile/alloc_file, don't honor
+	 * FMODE_LSEEK, FMODE_PREAD nor FMODE_PWRITE. We need to read from this
+	 * file descriptor, so we set FMODE_PREAD here.
+	 */
+	stream_file->f_mode = FMODE_PREAD;
 	fd_install(stream_fd, stream_file);
 	/*
 	 * The stream holds a reference to the channel within the generic ring
-- 
2.0.2

