From 3c9fc78814ecf1eeeead948322b6f55c1ab07a02 Mon Sep 17 00:00:00 2001
From: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
Date: Tue, 10 May 2011 23:21:03 -0400
Subject: [PATCH 061/933] Cleanup makefile

Signed-off-by: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
---
 drivers/staging/lttng/Makefile | 12 ------------
 1 file changed, 12 deletions(-)

diff --git a/drivers/staging/lttng/Makefile b/drivers/staging/lttng/Makefile
index ac00f9c..cb839bf 100644
--- a/drivers/staging/lttng/Makefile
+++ b/drivers/staging/lttng/Makefile
@@ -15,18 +15,6 @@ ltt-relay-objs :=  ltt-events.o ltt-event-header.o ltt-debugfs-abi.o \
 
 obj-m += probes/
 
-#ltt-type-serializer.o
-#ltt-serialize.o
-#obj-m += ltt-marker-control.o
-#obj-m += ltt-trace-control.o
-#ltt-ascii.o
-
-#obj-m += ltt-statedump.o
-#obj-m += ltt-userspace-event.o
-#obj-m += ltt-filter.o
-#obj-m += ltt-kprobes.o
-#obj-m += probes/
-
 endif
 
 else
-- 
2.0.2

