From b5deede6c830616fe57948f006cf7e9dee0394a7 Mon Sep 17 00:00:00 2001
From: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
Date: Sat, 14 May 2011 10:01:22 -0400
Subject: [PATCH 076/757] Add sequence and array

Signed-off-by: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
---
 drivers/staging/lttng/ltt-events.c | 43 ++++++++++++++++++++++++++++++++++++++
 1 file changed, 43 insertions(+)

diff --git a/drivers/staging/lttng/ltt-events.c b/drivers/staging/lttng/ltt-events.c
index 29ea9ba..cc189b3 100644
--- a/drivers/staging/lttng/ltt-events.c
+++ b/drivers/staging/lttng/ltt-events.c
@@ -363,9 +363,52 @@ int _ltt_fields_metadata_statedump(struct ltt_session *session,
 				field->name);
 			break;
 		case atype_array:
+		{
+			const struct lttng_basic_type *elem_type;
+
+			elem_type = &field->type.u.array.elem_type;
+			ret = lttng_metadata_printf(session,
+				"		integer { size = %u; align = %u; signed = %u;%s } %s[%u];\n",
+				elem_type->u.basic.integer.size,
+				elem_type->u.basic.integer.alignment,
+				elem_type->u.basic.integer.signedness,
+#ifdef __BIG_ENDIAN
+				elem_type->u.basic.integer.reverse_byte_order ? " byte_order = le;" : "",
+#else
+				elem_type->u.basic.integer.reverse_byte_order ? " byte_order = be;" : "",
+#endif
+				field->name, field->type.u.array.length);
 			break;
+		}
 		case atype_sequence:
+		{
+			const struct lttng_basic_type *elem_type;
+			const struct lttng_basic_type *length_type;
+
+			elem_type = &field->type.u.sequence.elem_type;
+			length_type = &field->type.u.sequence.length_type;
+			ret = lttng_metadata_printf(session,
+				"		integer { size = %u; align = %u; signed = %u;%s } %s[ integer { size = %u; align = %u; signed = %u;%s } ];\n",
+				elem_type->u.basic.integer.size,
+				elem_type->u.basic.integer.alignment,
+				elem_type->u.basic.integer.signedness,
+#ifdef __BIG_ENDIAN
+				elem_type->u.basic.integer.reverse_byte_order ? " byte_order = le;" : "",
+#else
+				elem_type->u.basic.integer.reverse_byte_order ? " byte_order = be;" : "",
+#endif
+				field->name,
+					length_type->u.basic.integer.size,
+					length_type->u.basic.integer.alignment,
+					length_type->u.basic.integer.signedness,
+#ifdef __BIG_ENDIAN
+					length_type->u.basic.integer.reverse_byte_order ? " byte_order = le;" : "",
+#else
+					length_type->u.basic.integer.reverse_byte_order ? " byte_order = be;" : ""
+#endif
+				);
 			break;
+		}
 
 		case atype_string:
 			ret = lttng_metadata_printf(session,
-- 
2.0.2

