From 22dffbdfd183e99c1470c6d3ed70623e2d4dbee7 Mon Sep 17 00:00:00 2001
From: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
Date: Wed, 18 May 2011 18:50:26 -0400
Subject: [PATCH 083/757] Fix bitfield.h include directory

Signed-off-by: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
---
 drivers/staging/lttng/lib/bitfield.h | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/drivers/staging/lttng/lib/bitfield.h b/drivers/staging/lttng/lib/bitfield.h
index a509e35..861e6dc 100644
--- a/drivers/staging/lttng/lib/bitfield.h
+++ b/drivers/staging/lttng/lib/bitfield.h
@@ -19,7 +19,7 @@
  * all copies or substantial portions of the Software.
  */
 
-#include "ltt-endian.h"
+#include "../ltt-endian.h"
 
 #ifndef CHAR_BIT
 #define CHAR_BIT 8
-- 
2.0.2

