From ee89eb73f47c4ab8822e75efefcfc59e8800c659 Mon Sep 17 00:00:00 2001
From: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
Date: Thu, 19 May 2011 00:17:21 -0400
Subject: [PATCH 092/757] LTTng ftrace support module should be unloadable

Signed-off-by: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
---
 drivers/staging/lttng/probes/lttng-ftrace.c | 13 +++++++++++++
 1 file changed, 13 insertions(+)

diff --git a/drivers/staging/lttng/probes/lttng-ftrace.c b/drivers/staging/lttng/probes/lttng-ftrace.c
index 3ec000a..f219037 100644
--- a/drivers/staging/lttng/probes/lttng-ftrace.c
+++ b/drivers/staging/lttng/probes/lttng-ftrace.c
@@ -7,6 +7,12 @@
  * Dual LGPL v2.1/GPL v2 license.
  */
 
+/*
+ * Ftrace function tracer does not seem to provide synchronization between probe
+ * teardown and callback execution. Therefore, we make this module permanently
+ * loaded (unloadable).
+ */
+
 #include <linux/module.h>
 #include <linux/ftrace.h>
 #include <linux/slab.h>
@@ -135,6 +141,13 @@ void lttng_ftrace_unregister(struct ltt_event *event)
 }
 EXPORT_SYMBOL_GPL(lttng_ftrace_unregister);
 
+/* This module is permanent. */
+int lttng_ftrace_init(void)
+{
+	return 0;
+}
+module_init(lttng_ftrace_init)
+
 MODULE_LICENSE("GPL and additional rights");
 MODULE_AUTHOR("Mathieu Desnoyers");
 MODULE_DESCRIPTION("Linux Trace Toolkit Ftrace Support");
-- 
2.0.2

