From f298196d1eefff1384e25dcef39e400434ec1852 Mon Sep 17 00:00:00 2001
From: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
Date: Sat, 21 May 2011 15:13:32 -0400
Subject: [PATCH 104/933] LTTng metadata fix

Signed-off-by: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
---
 drivers/staging/lttng/ltt-events.c | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/drivers/staging/lttng/ltt-events.c b/drivers/staging/lttng/ltt-events.c
index 6d9eafa..5c4ebb1 100644
--- a/drivers/staging/lttng/ltt-events.c
+++ b/drivers/staging/lttng/ltt-events.c
@@ -680,9 +680,10 @@ int _ltt_session_metadata_statedump(struct ltt_session *session)
 		"		uint32_t magic;\n"
 		"		uint8_t  uuid[16];\n"
 		"		uint32_t stream_id;\n"
-		"	};\n",
+		"	};\n"
 		"};\n\n",
 		ltt_alignof(uint8_t) * CHAR_BIT,
+		ltt_alignof(uint16_t) * CHAR_BIT,
 		ltt_alignof(uint32_t) * CHAR_BIT,
 		ltt_alignof(uint64_t) * CHAR_BIT,
 		CTF_VERSION_MAJOR,
-- 
2.0.2

