From 8e64548c2eae07158de0b6125091ad5246873ae2 Mon Sep 17 00:00:00 2001
From: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
Date: Sat, 21 May 2011 15:19:57 -0400
Subject: [PATCH 105/757] update metadata formatting

Signed-off-by: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
---
 drivers/staging/lttng/ltt-events.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/drivers/staging/lttng/ltt-events.c b/drivers/staging/lttng/ltt-events.c
index 5c4ebb1..98cdf01 100644
--- a/drivers/staging/lttng/ltt-events.c
+++ b/drivers/staging/lttng/ltt-events.c
@@ -589,7 +589,7 @@ int _ltt_stream_packet_context_declare(struct ltt_session *session)
 		"	uint32_t content_size;\n"
 		"	uint32_t packet_size;\n"
 		"	uint32_t cpu_id;\n"
-		"};\n"
+		"};\n\n"
 		);
 }
 
-- 
2.0.2

