From 8a554bbb796920229edbf6682b3de1f107efcd91 Mon Sep 17 00:00:00 2001
From: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
Date: Sat, 21 May 2011 18:02:01 -0400
Subject: [PATCH 107/757] Fix metadata

Signed-off-by: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
---
 drivers/staging/lttng/ltt-events.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/drivers/staging/lttng/ltt-events.c b/drivers/staging/lttng/ltt-events.c
index 28d17fc..8a6f8e9 100644
--- a/drivers/staging/lttng/ltt-events.c
+++ b/drivers/staging/lttng/ltt-events.c
@@ -522,7 +522,7 @@ int _ltt_event_metadata_statedump(struct ltt_session *session,
 		"	name = %s;\n"
 		"	id = %u;\n"
 		"	stream_id = %u;\n"
-		"	event.fields := struct {\n",
+		"	fields := struct {\n",
 		event->desc->name,
 		event->id,
 		event->chan->id);
@@ -715,7 +715,7 @@ skip_session:
 	}
 
 	list_for_each_entry(event, &session->events, list) {
-		ret = _ltt_event_metadata_statedump(session, chan, event);
+		ret = _ltt_event_metadata_statedump(session, event->chan, event);
 		if (ret)
 			goto end;
 	}
-- 
2.0.2

