From 9b63f5ca52e24bcc4511739c15d9c2369db36e35 Mon Sep 17 00:00:00 2001
From: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
Date: Sun, 22 May 2011 10:04:47 -0400
Subject: [PATCH 110/757] Comment buffer creation behavior wrt packet headers

Signed-off-by: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
---
 drivers/staging/lttng/ltt-events.c | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/drivers/staging/lttng/ltt-events.c b/drivers/staging/lttng/ltt-events.c
index 46da159..17d8709 100644
--- a/drivers/staging/lttng/ltt-events.c
+++ b/drivers/staging/lttng/ltt-events.c
@@ -170,6 +170,11 @@ struct ltt_channel *ltt_channel_create(struct ltt_session *session,
 	chan->session = session;
 	init_waitqueue_head(&chan->notify_wait);
 	chan->id = session->free_chan_id++;
+	/*
+	 * Note: the channel creation op already writes into the packet
+	 * headers. Therefore the "chan" information used as input
+	 * should be already accessible.
+	 */
 	chan->chan = transport->ops.channel_create("[lttng]", chan, buf_addr,
 			subbuf_size, num_subbuf, switch_timer_interval,
 			read_timer_interval);
-- 
2.0.2

