From e4c31a1ea538239b08cecb76ad30559f6fda5723 Mon Sep 17 00:00:00 2001
From: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
Date: Mon, 23 May 2011 22:25:25 -0400
Subject: [PATCH 123/757] Add function tracer ABI

Signed-off-by: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
---
 drivers/staging/lttng/ltt-debugfs-abi.h | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/drivers/staging/lttng/ltt-debugfs-abi.h b/drivers/staging/lttng/ltt-debugfs-abi.h
index d7d6892..12592c8 100644
--- a/drivers/staging/lttng/ltt-debugfs-abi.h
+++ b/drivers/staging/lttng/ltt-debugfs-abi.h
@@ -79,5 +79,7 @@ struct lttng_kernel_tracer_version {
 	_IOW(0xF6, 0x47, struct lttng_kernel_kprobe)
 #define LTTNG_KERNEL_TRACER_VERSION	\
 	_IOR(0xF6, 0x48, struct lttng_kernel_tracer_version)
+#define LTTNG_KERNEL_FUNCTION_TRACER	\
+	_IOW(0xF6, 0x49, struct lttng_kernel_function_tracer)
 
 #endif /* _LTT_DEBUGFS_ABI_H */
-- 
2.0.2

