From 56e8b5526d80c171fce309781616dc38c535bd50 Mon Sep 17 00:00:00 2001
From: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
Date: Tue, 24 May 2011 13:22:35 -0400
Subject: [PATCH 136/757] Add missing mutex to event list

Signed-off-by: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
---
 drivers/staging/lttng/ltt-probes.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/drivers/staging/lttng/ltt-probes.c b/drivers/staging/lttng/ltt-probes.c
index 249fc33..ab4b349 100644
--- a/drivers/staging/lttng/ltt-probes.c
+++ b/drivers/staging/lttng/ltt-probes.c
@@ -90,6 +90,7 @@ void *tp_list_start(struct seq_file *m, loff_t *pos)
 	struct lttng_probe_desc *probe_desc;
 	int iter = 0, i;
 
+	mutex_lock(&probe_mutex);
 	list_for_each_entry(probe_desc, &probe_list, head) {
 		for (i = 0; i < probe_desc->nr_events; i++) {
 			if (iter++ >= *pos)
@@ -120,6 +121,7 @@ void *tp_list_next(struct seq_file *m, void *p, loff_t *ppos)
 static
 void tp_list_stop(struct seq_file *m, void *p)
 {
+	mutex_unlock(&probe_mutex);
 }
 
 static
-- 
2.0.2

