From 9cbc9689336ae1f74377d60ca8c796347ddd696b Mon Sep 17 00:00:00 2001
From: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
Date: Wed, 25 May 2011 21:43:23 -0400
Subject: [PATCH 150/757] Fix field ctx allocation

Signed-off-by: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
---
 drivers/staging/lttng/ltt-context.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/drivers/staging/lttng/ltt-context.c b/drivers/staging/lttng/ltt-context.c
index ff1ba2e..961d80e 100644
--- a/drivers/staging/lttng/ltt-context.c
+++ b/drivers/staging/lttng/ltt-context.c
@@ -28,7 +28,7 @@ struct lttng_ctx_field *lttng_append_context(struct lttng_ctx **ctx_p)
 	if (ctx->nr_fields + 1 > ctx->allocated_fields) {
 		struct lttng_ctx_field *new_fields;
 
-		ctx->allocated_fields = min_t(size_t, 1, 2 * ctx->allocated_fields);
+		ctx->allocated_fields = max_t(size_t, 1, 2 * ctx->allocated_fields);
 		new_fields = kzalloc(ctx->allocated_fields * sizeof(struct lttng_ctx_field), GFP_KERNEL);
 		if (!new_fields)
 			return NULL;
-- 
2.0.2

