From 6a05cc7c15a07a7e9dd6531982c963bc03b08461 Mon Sep 17 00:00:00 2001
From: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
Date: Wed, 25 May 2011 23:21:36 -0400
Subject: [PATCH 152/933] Fix context append incorrect copy length

Signed-off-by: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
---
 drivers/staging/lttng/ltt-context.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/drivers/staging/lttng/ltt-context.c b/drivers/staging/lttng/ltt-context.c
index 961d80e..e0cd073 100644
--- a/drivers/staging/lttng/ltt-context.c
+++ b/drivers/staging/lttng/ltt-context.c
@@ -33,7 +33,7 @@ struct lttng_ctx_field *lttng_append_context(struct lttng_ctx **ctx_p)
 		if (!new_fields)
 			return NULL;
 		if (ctx->fields)
-			memcpy(new_fields, ctx->fields, ctx->nr_fields);
+			memcpy(new_fields, ctx->fields, sizeof(*ctx->fields) * ctx->nr_fields);
 		kfree(ctx->fields);
 		ctx->fields = new_fields;
 	}
-- 
2.0.2

