From 5a1888183e31804b3031ced82645ae094b6cdf64 Mon Sep 17 00:00:00 2001
From: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
Date: Wed, 1 Jun 2011 13:20:43 -0400
Subject: [PATCH 153/933] Don't export lttng metadata event in event list

Signed-off-by: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
---
 drivers/staging/lttng/ltt-probes.c | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/drivers/staging/lttng/ltt-probes.c b/drivers/staging/lttng/ltt-probes.c
index ab4b349..94beb05 100644
--- a/drivers/staging/lttng/ltt-probes.c
+++ b/drivers/staging/lttng/ltt-probes.c
@@ -129,6 +129,11 @@ int tp_list_show(struct seq_file *m, void *p)
 {
 	const struct lttng_event_desc *probe_desc = p;
 
+	/*
+	 * Don't export lttng internal events (metadata).
+	 */
+	if (!strncmp(probe_desc->name, "lttng_", sizeof("lttng_") - 1))
+		return 0;
 	seq_printf(m,	"event { name = %s; };\n",
 		   probe_desc->name);
 	return 0;
-- 
2.0.2

