From daff9cc773cf4c6ba90c491e84fa0db7e417e48f Mon Sep 17 00:00:00 2001
From: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
Date: Wed, 1 Jun 2011 16:03:14 -0400
Subject: [PATCH 158/757] Add missing () to metadata for sequence

Signed-off-by: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
---
 drivers/staging/lttng/ltt-events.c | 11 +++++------
 1 file changed, 5 insertions(+), 6 deletions(-)

diff --git a/drivers/staging/lttng/ltt-events.c b/drivers/staging/lttng/ltt-events.c
index f98ef30..a4bc154 100644
--- a/drivers/staging/lttng/ltt-events.c
+++ b/drivers/staging/lttng/ltt-events.c
@@ -476,15 +476,14 @@ int _ltt_field_statedump(struct ltt_session *session,
 			length_type->u.basic.integer.signedness,
 			(length_type->u.basic.integer.encoding == lttng_encode_none)
 				? "none"
-				: (length_type->u.basic.integer.encoding == lttng_encode_UTF8)
+				: ((length_type->u.basic.integer.encoding == lttng_encode_UTF8)
 					? "UTF8"
-					: "ASCII",
+					: "ASCII"),
 			length_type->u.basic.integer.base,
 #ifdef __BIG_ENDIAN
 			length_type->u.basic.integer.reverse_byte_order ? " byte_order = le;" : "",
 #else
-			length_type->u.basic.integer.reverse_byte_order
-? " byte_order = be;" : "",
+			length_type->u.basic.integer.reverse_byte_order ? " byte_order = be;" : "",
 #endif
 			field->name,
 			elem_type->u.basic.integer.size,
@@ -492,9 +491,9 @@ int _ltt_field_statedump(struct ltt_session *session,
 			elem_type->u.basic.integer.signedness,
 			(elem_type->u.basic.integer.encoding == lttng_encode_none)
 				? "none"
-				: (elem_type->u.basic.integer.encoding == lttng_encode_UTF8)
+				: ((elem_type->u.basic.integer.encoding == lttng_encode_UTF8)
 					? "UTF8"
-					: "ASCII",
+					: "ASCII"),
 			elem_type->u.basic.integer.base,
 #ifdef __BIG_ENDIAN
 			elem_type->u.basic.integer.reverse_byte_order ? " byte_order = le;" : "",
-- 
2.0.2

