From d8bbabe4af38bd79d243f9ffc25a36d5433e3545 Mon Sep 17 00:00:00 2001
From: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
Date: Wed, 1 Jun 2011 16:09:45 -0400
Subject: [PATCH 159/933] Case alignment to uint for metadata print

Signed-off-by: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
---
 drivers/staging/lttng/ltt-events.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/drivers/staging/lttng/ltt-events.c b/drivers/staging/lttng/ltt-events.c
index a4bc154..2693dc2 100644
--- a/drivers/staging/lttng/ltt-events.c
+++ b/drivers/staging/lttng/ltt-events.c
@@ -472,7 +472,7 @@ int _ltt_field_statedump(struct ltt_session *session,
 			"		integer { size = %u; align = %u; signed = %u; encoding = %s; base = %u;%s } __%s_length;\n",
 			"		integer { size = %u; align = %u; signed = %u; encoding = %s; base = %u;%s } %s[ __%s_length ];\n",
 			length_type->u.basic.integer.size,
-			length_type->u.basic.integer.alignment,
+			(unsigned int) length_type->u.basic.integer.alignment,
 			length_type->u.basic.integer.signedness,
 			(length_type->u.basic.integer.encoding == lttng_encode_none)
 				? "none"
@@ -487,7 +487,7 @@ int _ltt_field_statedump(struct ltt_session *session,
 #endif
 			field->name,
 			elem_type->u.basic.integer.size,
-			elem_type->u.basic.integer.alignment,
+			(unsigned int) elem_type->u.basic.integer.alignment,
 			elem_type->u.basic.integer.signedness,
 			(elem_type->u.basic.integer.encoding == lttng_encode_none)
 				? "none"
-- 
2.0.2

