From 1df914a67a69735613815f3ac5b2e242e6a022eb Mon Sep 17 00:00:00 2001
From: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
Date: Wed, 1 Jun 2011 16:15:45 -0400
Subject: [PATCH 160/933] Split vsnprintf with large number of args in two

Signed-off-by: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
---
 drivers/staging/lttng/ltt-events.c | 11 +++++++----
 1 file changed, 7 insertions(+), 4 deletions(-)

diff --git a/drivers/staging/lttng/ltt-events.c b/drivers/staging/lttng/ltt-events.c
index 2693dc2..1b53d2c 100644
--- a/drivers/staging/lttng/ltt-events.c
+++ b/drivers/staging/lttng/ltt-events.c
@@ -470,7 +470,6 @@ int _ltt_field_statedump(struct ltt_session *session,
 		length_type = &field->type.u.sequence.length_type;
 		ret = lttng_metadata_printf(session,
 			"		integer { size = %u; align = %u; signed = %u; encoding = %s; base = %u;%s } __%s_length;\n",
-			"		integer { size = %u; align = %u; signed = %u; encoding = %s; base = %u;%s } %s[ __%s_length ];\n",
 			length_type->u.basic.integer.size,
 			(unsigned int) length_type->u.basic.integer.alignment,
 			length_type->u.basic.integer.signedness,
@@ -485,7 +484,12 @@ int _ltt_field_statedump(struct ltt_session *session,
 #else
 			length_type->u.basic.integer.reverse_byte_order ? " byte_order = be;" : "",
 #endif
-			field->name,
+			field->name);
+		if (ret)
+			return ret;
+
+		ret = lttng_metadata_printf(session,
+			"		integer { size = %u; align = %u; signed = %u; encoding = %s; base = %u;%s } %s[ __%s_length ];\n",
 			elem_type->u.basic.integer.size,
 			(unsigned int) elem_type->u.basic.integer.alignment,
 			elem_type->u.basic.integer.signedness,
@@ -501,8 +505,7 @@ int _ltt_field_statedump(struct ltt_session *session,
 			elem_type->u.basic.integer.reverse_byte_order ? " byte_order = be;" : "",
 #endif
 			field->name,
-			field->name
-			);
+			field->name);
 		break;
 	}
 
-- 
2.0.2

