From 52229413bb1cd68cb28e5a3d253eff1a35528b7b Mon Sep 17 00:00:00 2001
From: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
Date: Wed, 8 Jun 2011 16:10:49 -0400
Subject: [PATCH 171/757] Metadata: should write reserve_len bytes, not "len"

Signed-off-by: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
---
 drivers/staging/lttng/ltt-events.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/drivers/staging/lttng/ltt-events.c b/drivers/staging/lttng/ltt-events.c
index ffcc61e..e3c252d 100644
--- a/drivers/staging/lttng/ltt-events.c
+++ b/drivers/staging/lttng/ltt-events.c
@@ -399,7 +399,7 @@ int lttng_metadata_printf(struct ltt_session *session,
 				ret = waitret;
 			goto end;
 		}
-		chan->ops->event_write(&ctx, &str[pos], len);
+		chan->ops->event_write(&ctx, &str[pos], reserve_len);
 		chan->ops->event_commit(&ctx);
 	}
 end:
-- 
2.0.2

