From 9727611200d6a33ebc1b1746feb75e14a40e49a9 Mon Sep 17 00:00:00 2001
From: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
Date: Fri, 10 Jun 2011 11:20:26 -0400
Subject: [PATCH 179/933] Add TODO for ftrace

Signed-off-by: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
---
 drivers/staging/lttng/probes/lttng-ftrace.c | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/drivers/staging/lttng/probes/lttng-ftrace.c b/drivers/staging/lttng/probes/lttng-ftrace.c
index 476d6bc..b8bd344 100644
--- a/drivers/staging/lttng/probes/lttng-ftrace.c
+++ b/drivers/staging/lttng/probes/lttng-ftrace.c
@@ -11,6 +11,10 @@
  * Ftrace function tracer does not seem to provide synchronization between probe
  * teardown and callback execution. Therefore, we make this module permanently
  * loaded (unloadable).
+ *
+ * TODO: Move to register_ftrace_function() (which is exported for
+ * modules) for Linux >= 3.0. It is faster (only enables the selected
+ * functions), and will stay there.
  */
 
 #include <linux/module.h>
-- 
2.0.2

