From c28be3b0125757e3fe86c68eeb0b8ce73b52b42f Mon Sep 17 00:00:00 2001
From: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
Date: Wed, 22 Jun 2011 15:56:47 -0400
Subject: [PATCH 184/757] Don't allow enable/disable on metadata channel/events

Signed-off-by: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
---
 drivers/staging/lttng/ltt-events.c | 8 ++++++++
 1 file changed, 8 insertions(+)

diff --git a/drivers/staging/lttng/ltt-events.c b/drivers/staging/lttng/ltt-events.c
index a94e8f9..42d486c 100644
--- a/drivers/staging/lttng/ltt-events.c
+++ b/drivers/staging/lttng/ltt-events.c
@@ -139,6 +139,8 @@ int ltt_channel_enable(struct ltt_channel *channel)
 {
 	int old;
 
+	if (channel == channel->session->metadata)
+		return -EPERM;
 	old = xchg(&channel->enabled, 1);
 	if (old)
 		return -EEXIST;
@@ -149,6 +151,8 @@ int ltt_channel_disable(struct ltt_channel *channel)
 {
 	int old;
 
+	if (channel == channel->session->metadata)
+		return -EPERM;
 	old = xchg(&channel->enabled, 0);
 	if (!old)
 		return -EEXIST;
@@ -159,6 +163,8 @@ int ltt_event_enable(struct ltt_event *event)
 {
 	int old;
 
+	if (event->chan == event->chan->session->metadata)
+		return -EPERM;
 	old = xchg(&event->enabled, 1);
 	if (old)
 		return -EEXIST;
@@ -169,6 +175,8 @@ int ltt_event_disable(struct ltt_event *event)
 {
 	int old;
 
+	if (event->chan == event->chan->session->metadata)
+		return -EPERM;
 	old = xchg(&event->enabled, 0);
 	if (!old)
 		return -EEXIST;
-- 
2.0.2

