From 43ba70e8f2b6493647d61b2d4ba780bc182ce1b6 Mon Sep 17 00:00:00 2001
From: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
Date: Thu, 23 Jun 2011 21:29:54 -0400
Subject: [PATCH 190/933] Fix typo in uuid wrapper

Signed-off-by: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
---
 drivers/staging/lttng/wrapper/uuid.h | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/drivers/staging/lttng/wrapper/uuid.h b/drivers/staging/lttng/wrapper/uuid.h
index 569ccab..b85024f 100644
--- a/drivers/staging/lttng/wrapper/uuid.h
+++ b/drivers/staging/lttng/wrapper/uuid.h
@@ -22,7 +22,7 @@ typedef struct {
 static inline
 void uuid_le_gen(uuid_le *u)
 {
-	genarate_random_uuid(u);
+	generate_random_uuid(u);
 }
 
 #endif
-- 
2.0.2

