From ef15d6afbcbfd6c094744daeade6fa596fcaa4cb Mon Sep 17 00:00:00 2001
From: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
Date: Thu, 23 Jun 2011 22:13:37 -0400
Subject: [PATCH 193/757] Add check for old tracepoint kernel

Signed-off-by: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
---
 drivers/staging/lttng/ltt-events.h | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/drivers/staging/lttng/ltt-events.h b/drivers/staging/lttng/ltt-events.h
index 32c8735..4d4eea9 100644
--- a/drivers/staging/lttng/ltt-events.h
+++ b/drivers/staging/lttng/ltt-events.h
@@ -359,4 +359,8 @@ void lttng_ftrace_destroy_private(struct ltt_event *event)
 
 extern const struct file_operations lttng_tracepoint_list_fops;
 
+#if (LINUX_VERSION_CODE >= KERNEL_VERSION(2,6,35))
+#define TRACEPOINT_HAS_DATA_ARG
+#endif
+
 #endif /* _LTT_EVENTS_H */
-- 
2.0.2

