From 3f78e5b34fc5d0fc6e3fa3e72600a5c8944f64fa Mon Sep 17 00:00:00 2001
From: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
Date: Thu, 23 Jun 2011 22:50:25 -0400
Subject: [PATCH 197/933] Update spinlock wrapper

Signed-off-by: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
---
 drivers/staging/lttng/wrapper/spinlock.h | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/drivers/staging/lttng/wrapper/spinlock.h b/drivers/staging/lttng/wrapper/spinlock.h
index af00ef2..e38d389 100644
--- a/drivers/staging/lttng/wrapper/spinlock.h
+++ b/drivers/staging/lttng/wrapper/spinlock.h
@@ -13,10 +13,10 @@
 
 #define raw_spin_lock_init(lock)				\
 	do {							\
-		*(lock) = __RAW_SPIN_LOCK_UNLOCKED(lock);	\
+		*(lock) = __RAW_SPIN_LOCK_UNLOCKED;		\
 	} while (0)
 
-#define raw_spin_is_locked(lock)	spin_is_locked(lock)
+#define raw_spin_is_locked(lock)	__raw_spin_is_locked(lock)
 
 
 #endif
-- 
2.0.2

