From 780d9f7ec674db76994624fd613da8d82b94e2de Mon Sep 17 00:00:00 2001
From: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
Date: Thu, 23 Jun 2011 22:50:56 -0400
Subject: [PATCH 198/757] Remove unused variable

Signed-off-by: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
---
 drivers/staging/lttng/lttng-context-vtid.c | 1 -
 1 file changed, 1 deletion(-)

diff --git a/drivers/staging/lttng/lttng-context-vtid.c b/drivers/staging/lttng/lttng-context-vtid.c
index f81fff6..522d2bd 100644
--- a/drivers/staging/lttng/lttng-context-vtid.c
+++ b/drivers/staging/lttng/lttng-context-vtid.c
@@ -40,7 +40,6 @@ void vtid_record(struct lttng_ctx_field *field,
 int lttng_add_vtid_to_ctx(struct lttng_ctx **ctx)
 {
 	struct lttng_ctx_field *field;
-	int ret;
 
 	field = lttng_append_context(ctx);
 	if (!field)
-- 
2.0.2

