From 8a3e0eba6b7236f5b5fe869847a6b5da43400d8e Mon Sep 17 00:00:00 2001
From: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
Date: Thu, 23 Jun 2011 23:03:30 -0400
Subject: [PATCH 199/933] update spinlock wrapper

Signed-off-by: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
---
 drivers/staging/lttng/wrapper/spinlock.h | 9 ++++++---
 1 file changed, 6 insertions(+), 3 deletions(-)

diff --git a/drivers/staging/lttng/wrapper/spinlock.h b/drivers/staging/lttng/wrapper/spinlock.h
index e38d389..8b1ad99 100644
--- a/drivers/staging/lttng/wrapper/spinlock.h
+++ b/drivers/staging/lttng/wrapper/spinlock.h
@@ -11,9 +11,12 @@
 
 #if (LINUX_VERSION_CODE < KERNEL_VERSION(2,6,33))
 
-#define raw_spin_lock_init(lock)				\
-	do {							\
-		*(lock) = __RAW_SPIN_LOCK_UNLOCKED;		\
+#include <linux/string.h>
+
+#define raw_spin_lock_init(lock)					\
+	do {								\
+		raw_spinlock_t __lock = __RAW_SPIN_LOCK_UNLOCKED;	\
+		memcpy(lock, &__lock, sizeof(lock));			\
 	} while (0)
 
 #define raw_spin_is_locked(lock)	__raw_spin_is_locked(lock)
-- 
2.0.2

