From 6149d9496fd5680d1304f429ecf6ab21c89b6854 Mon Sep 17 00:00:00 2001
From: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
Date: Thu, 23 Jun 2011 23:04:34 -0400
Subject: [PATCH 200/757] Add PIPE_DEF_BUFFERS to splice wrapper

Signed-off-by: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
---
 drivers/staging/lttng/wrapper/splice.h | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/drivers/staging/lttng/wrapper/splice.h b/drivers/staging/lttng/wrapper/splice.h
index d419847..f75309a 100644
--- a/drivers/staging/lttng/wrapper/splice.h
+++ b/drivers/staging/lttng/wrapper/splice.h
@@ -16,4 +16,8 @@
 ssize_t wrapper_splice_to_pipe(struct pipe_inode_info *pipe,
 			       struct splice_pipe_desc *spd);
 
+#ifndef PIPE_DEF_BUFFERS
+#define PIPE_DEF_BUFFERS 16
+#endif
+
 #endif /* _LTT_WRAPPER_SPLICE_H */
-- 
2.0.2

