From 89e5e7149556c6df8aaefdedab3f8094ecc8443d Mon Sep 17 00:00:00 2001
From: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
Date: Thu, 14 Jul 2011 14:13:06 -0400
Subject: [PATCH 211/933] Fix context duplicate name check

Signed-off-by: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
---
 drivers/staging/lttng/ltt-context.c | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/drivers/staging/lttng/ltt-context.c b/drivers/staging/lttng/ltt-context.c
index 6c3890c..e9a45f3 100644
--- a/drivers/staging/lttng/ltt-context.c
+++ b/drivers/staging/lttng/ltt-context.c
@@ -21,6 +21,9 @@ int lttng_find_context(struct lttng_ctx *ctx, const char *name)
 	unsigned int i;
 
 	for (i = 0; i < ctx->nr_fields; i++) {
+		/* Skip allocated (but non-initialized) contexts */
+		if (!ctx->fields[i].event_field.name)
+			continue;
 		if (!strcmp(ctx->fields[i].event_field.name, name))
 			return 1;
 	}
-- 
2.0.2

