From dbaf85c1e17d05cdf10eadd816b7c5a2d0c2a60b Mon Sep 17 00:00:00 2001
From: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
Date: Tue, 19 Jul 2011 21:51:57 -0400
Subject: [PATCH 220/757] update readme

Signed-off-by: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
---
 drivers/staging/lttng/README | 10 ++++++++--
 1 file changed, 8 insertions(+), 2 deletions(-)

diff --git a/drivers/staging/lttng/README b/drivers/staging/lttng/README
index 48ac657..c6766afd 100644
--- a/drivers/staging/lttng/README
+++ b/drivers/staging/lttng/README
@@ -28,8 +28,14 @@ against, use:
 KERNELDIR=path_to_kernel_dir make
 KERNELDIR=path_to_kernel_dir make install
 
-Use lttng-tools to control the tracer. LTTng tools should automatically load the
-kernel modules when needed.
+Use lttng-tools (git://git.lttng.org/lttng-tools.git) to control the tracer.
+LTTng tools should automatically load the kernel modules when needed.
+
+Use Babeltrace (git://git.efficios.com/babeltrace.git) to print traces as a
+human-readable text log.
+
+Please note that the LTTng-UST 2.0 (user-space tracing counterpart of LTTng 2.0)
+is still in active development and not released yet.
 
 So far, it has been tested on vanilla kernels 2.6.38 and 2.6.39 (on x86 at the
 moment). It should work fine with newer kernels and other architectures, but
-- 
2.0.2

