From 325bbb2ec6c9814939a20dcfecdeaa792bcbf575 Mon Sep 17 00:00:00 2001
From: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
Date: Wed, 20 Jul 2011 10:02:23 -0400
Subject: [PATCH 221/933] Fix Readme

Signed-off-by: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
---
 drivers/staging/lttng/README | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/drivers/staging/lttng/README b/drivers/staging/lttng/README
index c6766afd..6c92e31 100644
--- a/drivers/staging/lttng/README
+++ b/drivers/staging/lttng/README
@@ -20,13 +20,13 @@ To build and install, you will need to have your kernel headers available (or
 access to your full kernel source tree), and use:
 
 make
-make install
+make modules_install
 
 If you need to specify the target directory to the kernel you want to build
 against, use:
 
 KERNELDIR=path_to_kernel_dir make
-KERNELDIR=path_to_kernel_dir make install
+KERNELDIR=path_to_kernel_dir make modules_install
 
 Use lttng-tools (git://git.lttng.org/lttng-tools.git) to control the tracer.
 LTTng tools should automatically load the kernel modules when needed.
-- 
2.0.2

