From fb40ca7918101b62ef91aaaabea2a2b0d11500c6 Mon Sep 17 00:00:00 2001
From: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
Date: Wed, 20 Jul 2011 11:00:11 -0400
Subject: [PATCH 222/933] Update README

Signed-off-by: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
---
 drivers/staging/lttng/README | 6 +++---
 1 file changed, 3 insertions(+), 3 deletions(-)

diff --git a/drivers/staging/lttng/README b/drivers/staging/lttng/README
index 6c92e31..745e563 100644
--- a/drivers/staging/lttng/README
+++ b/drivers/staging/lttng/README
@@ -11,9 +11,9 @@ need for additional patches. Other features:
 - Function tracer, perf counters and kprobes support,
 - Integrated interface for both kernel and userspace tracing,
 - Have the ability to attach "context" information to events in the
-  trace (e.g. any perf counter, pid, ppid, tid, comm name, etc). So
-  basically, all the perf "required fields" like "preempt count" and
-  "bkl count" are all optional, specified on a per-tracing-session basis
+  trace (e.g. any perf counter, pid, ppid, tid, comm name, etc).
+  basically, all the extra information fields to be collected with
+  events are optional, specified on a per-tracing-session basis
   (except for timestamp and event id, which are mandatory).
 
 To build and install, you will need to have your kernel headers available (or
-- 
2.0.2

