From 97cb68a67f0da0ba60b69f2b2d8146b250edc1df Mon Sep 17 00:00:00 2001
From: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
Date: Wed, 20 Jul 2011 11:03:52 -0400
Subject: [PATCH 223/757] update readme

Signed-off-by: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
---
 drivers/staging/lttng/README | 11 ++++++-----
 1 file changed, 6 insertions(+), 5 deletions(-)

diff --git a/drivers/staging/lttng/README b/drivers/staging/lttng/README
index 745e563..fb0561d 100644
--- a/drivers/staging/lttng/README
+++ b/drivers/staging/lttng/README
@@ -8,13 +8,14 @@ need for additional patches. Other features:
 
 - Produces CTF (Common Trace Format) natively,
   (http://www.efficios.com/ctf)
-- Function tracer, perf counters and kprobes support,
+- Tracepoints, Function tracer, CPU Performance Monitoring Unit (PMU)
+  counters and kprobes support,
 - Integrated interface for both kernel and userspace tracing,
 - Have the ability to attach "context" information to events in the
-  trace (e.g. any perf counter, pid, ppid, tid, comm name, etc).
-  basically, all the extra information fields to be collected with
-  events are optional, specified on a per-tracing-session basis
-  (except for timestamp and event id, which are mandatory).
+  trace (e.g. any PMU counter, pid, ppid, tid, comm name, etc).
+  All the extra information fields to be collected with events are
+  optional, specified on a per-tracing-session basis (except for
+  timestamp and event id, which are mandatory).
 
 To build and install, you will need to have your kernel headers available (or
 access to your full kernel source tree), and use:
-- 
2.0.2

