From a59cad2d09023012241d1932aec72fa148ff988e Mon Sep 17 00:00:00 2001
From: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
Date: Thu, 21 Jul 2011 14:54:59 -0400
Subject: [PATCH 230/933] Build fix: ifdef KVM build

Signed-off-by: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
---
 drivers/staging/lttng/probes/Makefile | 5 ++++-
 1 file changed, 4 insertions(+), 1 deletion(-)

diff --git a/drivers/staging/lttng/probes/Makefile b/drivers/staging/lttng/probes/Makefile
index f4642d5..1b1e51f 100644
--- a/drivers/staging/lttng/probes/Makefile
+++ b/drivers/staging/lttng/probes/Makefile
@@ -11,10 +11,13 @@ obj-m += lttng-types.o
 obj-m += lttng-probe-lttng.o
 
 obj-m += lttng-probe-sched.o
-obj-m += lttng-probe-kvm.o
 obj-m += lttng-probe-irq.o
 obj-m += lttng-probe-syscalls.o
 
+ifneq ($(CONFIG_KVM),)
+obj-m += lttng-probe-kvm.o
+endif
+
 ifneq ($(CONFIG_BLOCK),)
 ifneq ($(CONFIG_EVENT_TRACING),)	# need blk_cmd_buf_len
 obj-m += lttng-probe-block.o
-- 
2.0.2

