From 40774859acd8091994e05705d6670dfbafcc7371 Mon Sep 17 00:00:00 2001
From: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
Date: Sat, 23 Jul 2011 12:37:54 -0400
Subject: [PATCH 235/757] context: document and check that only last context is
 removed

Signed-off-by: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
---
 drivers/staging/lttng/ltt-context.c | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/drivers/staging/lttng/ltt-context.c b/drivers/staging/lttng/ltt-context.c
index e9a45f3..154e996 100644
--- a/drivers/staging/lttng/ltt-context.c
+++ b/drivers/staging/lttng/ltt-context.c
@@ -60,6 +60,9 @@ struct lttng_ctx_field *lttng_append_context(struct lttng_ctx **ctx_p)
 }
 EXPORT_SYMBOL_GPL(lttng_append_context);
 
+/*
+ * Remove last context field.
+ */
 void lttng_remove_context_field(struct lttng_ctx **ctx_p,
 				struct lttng_ctx_field *field)
 {
@@ -67,6 +70,7 @@ void lttng_remove_context_field(struct lttng_ctx **ctx_p,
 
 	ctx = *ctx_p;
 	ctx->nr_fields--;
+	WARN_ON_ONCE(&ctx->fields[ctx->nr_fields] != field);
 	memset(&ctx->fields[ctx->nr_fields], 0, sizeof(struct lttng_ctx_field));
 }
 EXPORT_SYMBOL_GPL(lttng_remove_context_field);
-- 
2.0.2

