From 46b920c8f1db87417382885a99ae59436a983ecb Mon Sep 17 00:00:00 2001
From: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
Date: Thu, 28 Jul 2011 15:16:17 -0400
Subject: [PATCH 241/933] Update ftrace wrapper

Signed-off-by: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
---
 drivers/staging/lttng/wrapper/ftrace.h | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/drivers/staging/lttng/wrapper/ftrace.h b/drivers/staging/lttng/wrapper/ftrace.h
index 9cf50d2..9c18cc5 100644
--- a/drivers/staging/lttng/wrapper/ftrace.h
+++ b/drivers/staging/lttng/wrapper/ftrace.h
@@ -55,14 +55,14 @@ static inline
 int wrapper_register_ftrace_function_probe(char *glob,
 		struct ftrace_probe_ops *ops, void *data)
 {
-	return unregister_ftrace_function_probe();
+	return register_ftrace_function_probe(glob, ops, data);
 }
 
 static inline
 void wrapper_unregister_ftrace_function_probe(char *glob,
 		struct ftrace_probe_ops *ops, void *data)
 {
-	return unregister_ftrace_function_probe();
+	return unregister_ftrace_function_probe(glob, ops, data);
 }
 #endif
 
-- 
2.0.2

