From 140af9a67c7ca855d5a915596bc71dbcc1bbf3d2 Mon Sep 17 00:00:00 2001
From: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
Date: Sat, 6 Aug 2011 02:30:42 -0400
Subject: [PATCH 250/757] Fix calibrate: add noinline

Signed-off-by: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
---
 drivers/staging/lttng/lttng-calibrate.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/drivers/staging/lttng/lttng-calibrate.c b/drivers/staging/lttng/lttng-calibrate.c
index 421d027..07e3c5b 100644
--- a/drivers/staging/lttng/lttng-calibrate.c
+++ b/drivers/staging/lttng/lttng-calibrate.c
@@ -11,6 +11,7 @@
 #include "ltt-debugfs-abi.h"
 #include "ltt-events.h"
 
+noinline
 void lttng_calibrate_kretprobe(void)
 {
 	asm volatile ("");
-- 
2.0.2

