From 578113b6e78135880484914d9e9ab3d150b18eb2 Mon Sep 17 00:00:00 2001
From: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
Date: Sat, 17 Sep 2011 09:52:46 -0400
Subject: [PATCH 281/757] Remove unused variable

Signed-off-by: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
---
 drivers/staging/lttng/lttng-syscalls.c | 2 --
 1 file changed, 2 deletions(-)

diff --git a/drivers/staging/lttng/lttng-syscalls.c b/drivers/staging/lttng/lttng-syscalls.c
index 48eaf16..13ca726 100644
--- a/drivers/staging/lttng/lttng-syscalls.c
+++ b/drivers/staging/lttng/lttng-syscalls.c
@@ -67,8 +67,6 @@ static struct trace_syscall_entry sc_table[] = {
 #include "instrumentation/syscalls/headers/syscalls_pointers.h"
 };
 
-static int sc_table_filled;
-
 #undef CREATE_SYSCALL_TABLE
 
 static void syscall_entry_unknown(struct ltt_channel *chan,
-- 
2.0.2

