From a03d712a708c8a4392db243d23874967f2f353af Mon Sep 17 00:00:00 2001
From: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
Date: Sat, 17 Sep 2011 10:00:56 -0400
Subject: [PATCH 282/757] Fix syscall exit tracing

Signed-off-by: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
---
 drivers/staging/lttng/lttng-syscalls.c | 6 ++++--
 1 file changed, 4 insertions(+), 2 deletions(-)

diff --git a/drivers/staging/lttng/lttng-syscalls.c b/drivers/staging/lttng/lttng-syscalls.c
index 13ca726..682a387 100644
--- a/drivers/staging/lttng/lttng-syscalls.c
+++ b/drivers/staging/lttng/lttng-syscalls.c
@@ -257,7 +257,8 @@ int lttng_syscalls_register(struct ltt_channel *chan, void *filter)
 	 * conflict with sys_exit syscall entry.
 	 */
 	ret = tracepoint_probe_register("sys_exit",
-			(void *) __event_probe__exit_syscall, chan);
+			(void *) __event_probe__exit_syscall,
+			chan->sc_unknown);
 	if (ret) {
 		WARN_ON_ONCE(tracepoint_probe_unregister("sys_enter",
 			(void *) syscall_entry_probe, chan));
@@ -275,7 +276,8 @@ int lttng_syscalls_unregister(struct ltt_channel *chan)
 	if (!chan->sc_table)
 		return 0;
 	ret = tracepoint_probe_unregister("sys_exit",
-			(void *) __event_probe__exit_syscall, chan);
+			(void *) __event_probe__exit_syscall,
+			chan->sc_unknown);
 	if (ret)
 		return ret;
 	ret = tracepoint_probe_unregister("sys_enter",
-- 
2.0.2

