From 4c6f2882bc59bf1deab9612d57a676a090c0e55d Mon Sep 17 00:00:00 2001
From: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
Date: Sat, 17 Sep 2011 11:46:54 -0400
Subject: [PATCH 288/757] Add __field_hex

Signed-off-by: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
---
 drivers/staging/lttng/probes/lttng-events.h | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/drivers/staging/lttng/probes/lttng-events.h b/drivers/staging/lttng/probes/lttng-events.h
index f86bd39..1d2def4 100644
--- a/drivers/staging/lttng/probes/lttng-events.h
+++ b/drivers/staging/lttng/probes/lttng-events.h
@@ -111,6 +111,10 @@ void trace_##_name(void *__data);
 #define __field_ext(_type, _item, _filter_type)			\
 	__field(_type, _item)
 
+#undef __field_hex
+#define __field_hex(_type, _item)				\
+	__field_full(_type, _item, __BYTE_ORDER, 16)
+
 #undef __field_network
 #define __field_network(_type, _item)				\
 	__field_full(_type, _item, __BIG_ENDIAN, 10)
-- 
2.0.2

