From e3db5dd428db0955946ecc0944176273baa13251 Mon Sep 17 00:00:00 2001
From: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
Date: Tue, 27 Sep 2011 08:17:43 -0400
Subject: [PATCH 303/933] Add sys_open filename (x86_64)

Signed-off-by: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
---
 .../headers/x86-64-syscalls-3.0.4_pointers_override.h         | 11 +++++++++++
 1 file changed, 11 insertions(+)

diff --git a/drivers/staging/lttng/instrumentation/syscalls/headers/x86-64-syscalls-3.0.4_pointers_override.h b/drivers/staging/lttng/instrumentation/syscalls/headers/x86-64-syscalls-3.0.4_pointers_override.h
index 0cdb32a..ccbca10 100644
--- a/drivers/staging/lttng/instrumentation/syscalls/headers/x86-64-syscalls-3.0.4_pointers_override.h
+++ b/drivers/staging/lttng/instrumentation/syscalls/headers/x86-64-syscalls-3.0.4_pointers_override.h
@@ -1,5 +1,16 @@
 #ifndef CREATE_SYSCALL_TABLE
 
+#define OVERRIDE_sys_open
+TRACE_EVENT(sys_open,
+	TP_PROTO(const char * filename, int flags, int mode),
+	TP_ARGS(filename, flags, mode),
+	TP_STRUCT__entry(__string(filename, filename) __field(int, flags) __field(int, mode)),
+	TP_fast_assign(tp_copy_string_from_user(filename, filename) tp_assign(flags, flags) tp_assign(mode, mode)),
+	TP_printk()
+)
+
 #else	/* CREATE_SYSCALL_TABLE */
 
+TRACE_SYSCALL_TABLE(sys_open, sys_open, 2, 3)
+
 #endif /* CREATE_SYSCALL_TABLE */
-- 
2.0.2

