From 5f4512ea4b280c1400efe9727d8c4f78db6ff975 Mon Sep 17 00:00:00 2001
From: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
Date: Tue, 27 Sep 2011 13:56:30 -0400
Subject: [PATCH 314/933] Fix syscall table

Signed-off-by: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
---
 drivers/staging/lttng/lttng-syscalls.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/drivers/staging/lttng/lttng-syscalls.c b/drivers/staging/lttng/lttng-syscalls.c
index aef7d4a..52c94f1 100644
--- a/drivers/staging/lttng/lttng-syscalls.c
+++ b/drivers/staging/lttng/lttng-syscalls.c
@@ -73,7 +73,7 @@ static const struct trace_syscall_entry sc_table[] = {
 //extern const struct trace_syscall_entry compat_sc_table[];
 //extern const size_t compat_sc_table_len;
 //temp disable
-static const struct trace_syscall_entry compat_sc_table[];
+static const struct trace_syscall_entry compat_sc_table[] = { };
 static const size_t compat_sc_table_len;
 
 static void syscall_entry_unknown(struct ltt_event *event,
-- 
2.0.2

