From 0615696ecf3fe9fde28d05767d4f8bc1c7ab11c1 Mon Sep 17 00:00:00 2001
From: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
Date: Fri, 30 Sep 2011 11:49:31 -0400
Subject: [PATCH 327/757] x86-32: override sys_mmap syscall instrumentation

Signed-off-by: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
---
 .../syscalls/headers/x86-32-syscalls-3.1.0-rc6_integers_override.h | 7 +++++++
 1 file changed, 7 insertions(+)

diff --git a/drivers/staging/lttng/instrumentation/syscalls/headers/x86-32-syscalls-3.1.0-rc6_integers_override.h b/drivers/staging/lttng/instrumentation/syscalls/headers/x86-32-syscalls-3.1.0-rc6_integers_override.h
index 8f685a8..02a83da 100644
--- a/drivers/staging/lttng/instrumentation/syscalls/headers/x86-32-syscalls-3.1.0-rc6_integers_override.h
+++ b/drivers/staging/lttng/instrumentation/syscalls/headers/x86-32-syscalls-3.1.0-rc6_integers_override.h
@@ -29,3 +29,10 @@
 #define OVERRIDE_TABLE_sys_setresgid16
 
 #endif
+
+#ifdef CREATE_SYSCALL_TABLE
+
+#define OVERRIDE_TABLE_sys_mmap
+TRACE_SYSCALL_TABLE(sys_mmap, sys_mmap, 90, 6)
+
+#endif
-- 
2.0.2

