From 9c14e15409edf4aa3f083658d940a6568ae6eece Mon Sep 17 00:00:00 2001
From: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
Date: Sun, 22 Jan 2012 15:52:05 -0500
Subject: [PATCH 350/933] Update TODO

Signed-off-by: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
---
 drivers/staging/lttng/TODO | 9 +--------
 1 file changed, 1 insertion(+), 8 deletions(-)

diff --git a/drivers/staging/lttng/TODO b/drivers/staging/lttng/TODO
index d0d380a..18d99d5 100644
--- a/drivers/staging/lttng/TODO
+++ b/drivers/staging/lttng/TODO
@@ -10,14 +10,7 @@ TODO:
 
 A) Cleanup/Testing
 
-	1) Remove debugfs "lttng" file (keep only procfs "lttng" file).
-	   The rationale for this is that this file is needed for
-	   user-level tracing support (LTTng-UST 2.0) intended to be
-	   used on production system, and therefore should be present as
-	   part of a "usually mounted" filesystem rather than a debug
-	   filesystem.
-
-	2) Test lib ring buffer snapshot feature.
+	1) Test lib ring buffer snapshot feature.
 	   When working on the lttngtop project, Julien Desfossez
 	   reported that he needed to push the consumer position
 	   forward explicitely with lib_ring_buffer_put_next_subbuf.
-- 
2.0.2

