From 6b768bd771dae7182579effa8281d1d4bfcd5a11 Mon Sep 17 00:00:00 2001
From: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
Date: Sun, 22 Jan 2012 15:53:11 -0500
Subject: [PATCH 351/933] Update TODO

Signed-off-by: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
---
 drivers/staging/lttng/TODO | 1 -
 1 file changed, 1 deletion(-)

diff --git a/drivers/staging/lttng/TODO b/drivers/staging/lttng/TODO
index 18d99d5..4420074 100644
--- a/drivers/staging/lttng/TODO
+++ b/drivers/staging/lttng/TODO
@@ -57,7 +57,6 @@ B) Features
 
 	3) Integrate the "statedump" module from LTTng 0.x into LTTng
 	   2.0.
-	     * Dependency: addition of "dynamic enumerations" type to CTF.
 	   See: http://git.lttng.org/?p=lttng-modules.git;a=shortlog;h=refs/heads/v0.19-stable
 	        ltt-statedump.c
 
-- 
2.0.2

