From 0c353796ab361bc5d47320c16f093cc76cd20c20 Mon Sep 17 00:00:00 2001
From: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
Date: Mon, 30 Jan 2012 11:27:58 -0500
Subject: [PATCH 361/933] Add missing double-quotes to clock uuid

Signed-off-by: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
---
 drivers/staging/lttng/lttng-events.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/drivers/staging/lttng/lttng-events.c b/drivers/staging/lttng/lttng-events.c
index c25d444..674494b 100644
--- a/drivers/staging/lttng/lttng-events.c
+++ b/drivers/staging/lttng/lttng-events.c
@@ -948,7 +948,7 @@ int _lttng_session_metadata_statedump(struct lttng_session *session)
 
 	if (!trace_clock_uuid(clock_uuid_s)) {
 		ret = lttng_metadata_printf(session,
-			"	uuid = %s;\n",
+			"	uuid = \"%s\";\n",
 			clock_uuid_s
 			);
 		if (ret)
-- 
2.0.2

