From b33e0153b49f1719899ac30da2d6d65d8efd9372 Mon Sep 17 00:00:00 2001
From: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
Date: Thu, 2 Feb 2012 11:55:18 -0500
Subject: [PATCH 362/933] Update version to 1.9.9

Signed-off-by: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
---
 drivers/staging/lttng/lttng-tracer.h | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/drivers/staging/lttng/lttng-tracer.h b/drivers/staging/lttng/lttng-tracer.h
index 571a715..99c8f08 100644
--- a/drivers/staging/lttng/lttng-tracer.h
+++ b/drivers/staging/lttng/lttng-tracer.h
@@ -25,9 +25,9 @@
 #include "lttng-tracer-core.h"
 #include "lttng-events.h"
 
-#define LTTNG_VERSION 0
+#define LTTNG_VERSION 1
 #define LTTNG_PATCHLEVEL 9
-#define LTTNG_SUBLEVEL 1
+#define LTTNG_SUBLEVEL 9
 
 #ifndef CHAR_BIT
 #define CHAR_BIT 8
-- 
2.0.2

