From e808408f03e0f61eeb59d9f6077c8b1472acfb8a Mon Sep 17 00:00:00 2001
From: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
Date: Thu, 2 Feb 2012 12:21:53 -0500
Subject: [PATCH 365/933] Remove unused defines

Signed-off-by: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
---
 drivers/staging/lttng/lttng-abi.h | 3 ---
 1 file changed, 3 deletions(-)

diff --git a/drivers/staging/lttng/lttng-abi.h b/drivers/staging/lttng/lttng-abi.h
index aaeceb6..fecf0ad 100644
--- a/drivers/staging/lttng/lttng-abi.h
+++ b/drivers/staging/lttng/lttng-abi.h
@@ -46,7 +46,6 @@ struct lttng_kernel_channel {
 	char padding[LTTNG_KERNEL_CHANNEL_PADDING];
 };
 
-#define LTTNG_KERNEL_KRETPROBE_PADDING	LTTNG_KERNEL_SYM_NAME_LEN + 32
 struct lttng_kernel_kretprobe {
 	uint64_t addr;
 
@@ -57,7 +56,6 @@ struct lttng_kernel_kretprobe {
 /*
  * Either addr is used, or symbol_name and offset.
  */
-#define LTTNG_KERNEL_KPROBE_PADDING	LTTNG_KERNEL_SYM_NAME_LEN + 32
 struct lttng_kernel_kprobe {
 	uint64_t addr;
 
@@ -65,7 +63,6 @@ struct lttng_kernel_kprobe {
 	char symbol_name[LTTNG_KERNEL_SYM_NAME_LEN];
 };
 
-#define LTTNG_KERNEL_FUNCTION_TRACER_PADDING	LTTNG_KERNEL_SYM_NAME_LEN + 32
 struct lttng_kernel_function_tracer {
 	char symbol_name[LTTNG_KERNEL_SYM_NAME_LEN];
 };
-- 
2.0.2

