From 4bb4b351df1a1d5267569bf2c02d34a4491b919c Mon Sep 17 00:00:00 2001
From: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
Date: Mon, 6 Feb 2012 23:23:38 -0500
Subject: [PATCH 368/933] Allow open /proc/lttng for read & write

Signed-off-by: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
---
 drivers/staging/lttng/lttng-abi.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/drivers/staging/lttng/lttng-abi.c b/drivers/staging/lttng/lttng-abi.c
index c449895..98f03c3 100644
--- a/drivers/staging/lttng/lttng-abi.c
+++ b/drivers/staging/lttng/lttng-abi.c
@@ -749,7 +749,7 @@ int __init lttng_abi_init(void)
 	int ret = 0;
 
 	wrapper_vmalloc_sync_all();
-	lttng_proc_dentry = proc_create_data("lttng", S_IWUSR, NULL,
+	lttng_proc_dentry = proc_create_data("lttng", S_IRUSR | S_IWUSR, NULL,
 					&lttng_fops, NULL);
 	
 	if (!lttng_proc_dentry) {
-- 
2.0.2

