From 07cd3ad4557cf5e844c19b5300e64311510ce128 Mon Sep 17 00:00:00 2001
From: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
Date: Wed, 8 Feb 2012 09:37:11 -0500
Subject: [PATCH 370/757] Update README

Signed-off-by: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
---
 drivers/staging/lttng/README | 9 +++++----
 1 file changed, 5 insertions(+), 4 deletions(-)

diff --git a/drivers/staging/lttng/README b/drivers/staging/lttng/README
index f4e3368..1fd49b6 100644
--- a/drivers/staging/lttng/README
+++ b/drivers/staging/lttng/README
@@ -1,7 +1,7 @@
 LTTng 2.0 modules
 
 Mathieu Desnoyers
-July 19, 2011
+February 8, 2012
 
 LTTng 2.0 kernel modules build against a vanilla or distribution kernel, without
 need for additional patches. Other features:
@@ -36,9 +36,6 @@ the kernel modules when needed. Use Babeltrace to print traces as a
 human-readable text log. These tools are available at the following URL:
 http://lttng.org/lttng2.0
 
-Please note that the LTTng-UST 2.0 (user-space tracing counterpart of LTTng 2.0)
-is still in active development and not released yet.
-
 So far, it has been tested on vanilla Linux kernels 2.6.38, 2.6.39 and 3.0 (on
 x86 32/64-bit, and powerpc 32-bit at the moment, build tested on ARM). It should
 work fine with newer kernels and other architectures, but expect build issues
@@ -49,6 +46,10 @@ note that lttng-modules 2.0 can build on a Linux kernel patched with the LTTng
 0.x patchset, but the lttng-modules 2.0 replace the lttng-modules 0.x, so both
 tracers cannot be installed at the same time for a given kernel version.
 
+LTTng-modules depends on having kallsyms enabled in the kernel it is
+built against. Ideally, if you want to have system call tracing, the
+"Trace Syscalls" feature should be enabled too. 
+
 * Note about Perf PMU counters support
 
 Each PMU counter has its zero value set when it is attached to a context with
-- 
2.0.2

