From 58ecb975572f3a58528015fe2d0c05d5329504dc Mon Sep 17 00:00:00 2001
From: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
Date: Thu, 9 Feb 2012 19:36:51 -0500
Subject: [PATCH 371/933] Update Changelog to 2.0-pre13

Signed-off-by: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
---
 drivers/staging/lttng/ChangeLog | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/drivers/staging/lttng/ChangeLog b/drivers/staging/lttng/ChangeLog
index 4f4ce0d..90b0e88 100644
--- a/drivers/staging/lttng/ChangeLog
+++ b/drivers/staging/lttng/ChangeLog
@@ -1,3 +1,8 @@
+2012-02-09 LTTng modules 2.0-pre13
+	* Update README
+	* environment: write sysname, release, version, domain to metadata
+	* Allow open /proc/lttng for read & write
+
 2012-02-02 LTTng modules 2.0-pre12
 	* Add x86 32/64 execve syscall instrumentation override
 	* Remove unused defines
-- 
2.0.2

