From cc4f4e6d95cd6bb30a09ccb0b23d52c01cba0310 Mon Sep 17 00:00:00 2001
From: Juha Niskanen <juniskane@gmail.com>
Date: Thu, 16 Feb 2012 11:55:42 -0500
Subject: [PATCH 382/757] statedump: Use old macro name for kernel 2.6.38

Signed-off-by: Juha Niskanen <juniskane@gmail.com>
Signed-off-by: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
---
 drivers/staging/lttng/lttng-statedump-impl.c | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/drivers/staging/lttng/lttng-statedump-impl.c b/drivers/staging/lttng/lttng-statedump-impl.c
index cfcf4fe..eccf1de 100644
--- a/drivers/staging/lttng/lttng-statedump-impl.c
+++ b/drivers/staging/lttng/lttng-statedump-impl.c
@@ -214,6 +214,11 @@ int lttng_enumerate_vm_maps(struct lttng_session *session)
 }
 
 #ifdef CONFIG_GENERIC_HARDIRQS
+
+#if (LINUX_VERSION_CODE < KERNEL_VERSION(2,6,39))
+#define irq_desc_get_chip(desc) get_irq_desc_chip(desc)
+#endif
+
 static
 void lttng_list_interrupts(struct lttng_session *session)
 {
-- 
2.0.2

