From 828983f3ed63c37ebd13bca599e54d30423e5f44 Mon Sep 17 00:00:00 2001
From: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
Date: Mon, 20 Feb 2012 15:32:01 -0500
Subject: [PATCH 384/757] Update Version to 2.0.0-rc1

Signed-off-by: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
---
 drivers/staging/lttng/ChangeLog | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/drivers/staging/lttng/ChangeLog b/drivers/staging/lttng/ChangeLog
index 9956098..ed89478 100644
--- a/drivers/staging/lttng/ChangeLog
+++ b/drivers/staging/lttng/ChangeLog
@@ -1,3 +1,7 @@
+2012-02-20 LTTng modules 2.0.0-rc1
+	* Standardize version across toolchain
+	* statedump: Use old macro name for kernel 2.6.38
+
 2012-02-16 LTTng modules 2.0-pre15
 	* Add timer instrumentation
 	* fix: need to undef mainline define
-- 
2.0.2

