From 20d43ab5164570f791d4c2cda9e1577ca950ce05 Mon Sep 17 00:00:00 2001
From: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
Date: Tue, 13 Mar 2012 16:11:37 -0400
Subject: [PATCH 392/757] Fix clock offset 32-bit multiplication overflow

Signed-off-by: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
---
 drivers/staging/lttng/lttng-events.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/drivers/staging/lttng/lttng-events.c b/drivers/staging/lttng/lttng-events.c
index ea6c116..2fedbea 100644
--- a/drivers/staging/lttng/lttng-events.c
+++ b/drivers/staging/lttng/lttng-events.c
@@ -887,7 +887,7 @@ uint64_t measure_clock_offset(void)
 	local_irq_restore(flags);
 
 	offset = (monotonic[0] + monotonic[1]) >> 1;
-	realtime = rts.tv_sec * NSEC_PER_SEC;
+	realtime = (uint64_t) rts.tv_sec * NSEC_PER_SEC;
 	realtime += rts.tv_nsec;
 	offset = realtime - offset;
 	return offset;
-- 
2.0.2

