From 6ac154ec332c6029f4649fd59ea24b6cc25ae2a0 Mon Sep 17 00:00:00 2001
From: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
Date: Fri, 16 Mar 2012 15:44:38 -0400
Subject: [PATCH 393/757] Update version to 2.0.0-rc3

Signed-off-by: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
---
 drivers/staging/lttng/ChangeLog | 6 ++++++
 1 file changed, 6 insertions(+)

diff --git a/drivers/staging/lttng/ChangeLog b/drivers/staging/lttng/ChangeLog
index 501da3e..aa8f4f0 100644
--- a/drivers/staging/lttng/ChangeLog
+++ b/drivers/staging/lttng/ChangeLog
@@ -1,3 +1,9 @@
+2012-03-16 LTTng modules 2.0.0-rc3
+	* Fix clock offset 32-bit multiplication overflow
+	* Fix : wrong assign of fd in state dump
+	* License cleanup, ifdef namespace cleanup
+	* Fix: ensure power of 2 check handles 64-bit size_t entirely
+
 2012-03-02 LTTng modules 2.0.0-rc2
 	* Fix: dmesg printout should not print metadata warnings
 	* Fix: use transport name as channel name
-- 
2.0.2

