From c81fef9dfe5445c88fa19acf5eea7c1c40922987 Mon Sep 17 00:00:00 2001
From: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
Date: Tue, 10 Apr 2012 20:36:17 -0400
Subject: [PATCH 401/933] README: Document that CONFIG_MODULES is required

Signed-off-by: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
---
 drivers/staging/lttng/README | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/drivers/staging/lttng/README b/drivers/staging/lttng/README
index c3568f8..89b4a7a 100644
--- a/drivers/staging/lttng/README
+++ b/drivers/staging/lttng/README
@@ -51,6 +51,8 @@ kernel version.
 
 * Kernel config options required
 
+CONFIG_MODULES: required
+  * Kernel modules support.
 CONFIG_KALLSYMS: required
   * See wrapper/ files. This is necessary until the few required missing
     symbols are exported to GPL modules from mainline.
-- 
2.0.2

