From 46c1c6dea9224e1cd697c9cf9f42b07451053d65 Mon Sep 17 00:00:00 2001
From: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
Date: Mon, 14 May 2012 11:12:00 -0400
Subject: [PATCH 409/933] Update instrumentation/events README file

Signed-off-by: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
---
 .../staging/lttng/instrumentation/events/README    | 22 +++++++++++++++++++++-
 1 file changed, 21 insertions(+), 1 deletion(-)

diff --git a/drivers/staging/lttng/instrumentation/events/README b/drivers/staging/lttng/instrumentation/events/README
index dad2cbb..84c3e89 100644
--- a/drivers/staging/lttng/instrumentation/events/README
+++ b/drivers/staging/lttng/instrumentation/events/README
@@ -1,7 +1,27 @@
-The workflow for updating patches from newer kernel:
+* Workflow for updating patches from newer kernel:
 
 Diff mainline/ and lttng-module/ directories.
 
 Pull the new headers from mainline kernel to mainline/.
 Copy them into lttng-modules.
 Apply diff. Fix conflicts.
+
+
+* Workflow to add new Tracepoint instrumentation to newer kernel,
+  and add support for it into LTTng:
+
+a) instrument the kernel with new trace events headers. If you want that
+   instrumentation distributed, you will have to push those changes into
+   the upstream Linux kernel first,
+b) copy those headers into lttng mainline/ directory,
+c) look at a diff from other headers between mainline/ and 
+   lttng/, and use that as a recipe to create a new lttng/
+   header from the mainline/ header,
+d) create a new file in probes/ for the new trace event header you added,
+e) add it to probes/Makefile,
+f) build, make modules_install,
+g) don't forget to load that new module too.
+
+Currently, LTTng policy is to only accept headers derived from trace
+event headers accepted into the Linux kernel upstream for tracepoints
+related to upstream kernel instrumentation.
-- 
2.0.2

