From 370e21f8e7f1e0f332c140a0347787bb93c96a88 Mon Sep 17 00:00:00 2001
From: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
Date: Thu, 5 Jul 2012 14:43:53 -0400
Subject: [PATCH 415/933] Cleanup: remove trailing whitespace in README

Signed-off-by: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
---
 drivers/staging/lttng/README | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/drivers/staging/lttng/README b/drivers/staging/lttng/README
index 89b4a7a..30f1338 100644
--- a/drivers/staging/lttng/README
+++ b/drivers/staging/lttng/README
@@ -79,7 +79,7 @@ CONFIG_PERF_EVENTS:
                lttng add-context -t perf:*
 CONFIG_EVENT_TRACING:
             needed to allow block layer tracing
-CONFIG_KPROBES: 
+CONFIG_KPROBES:
             Dynamic probe.
                lttng enable-event -k --probe ...
 CONFIG_KRETPROBES:
-- 
2.0.2

