From 11cf799452902c9e49da960ca2e175583b6d3869 Mon Sep 17 00:00:00 2001
From: Ralf Baechle <ralf@linux-mips.org>
Date: Mon, 9 Jul 2012 12:48:11 -0400
Subject: [PATCH 416/757] cleanup: fix typo in syscall instrumentation header

Signed-off-by: Ralf Baechle <ralf@linux-mips.org>
Signed-off-by: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
---
 .../syscalls/headers/x86-64-syscalls-3.0.4_integers_override.h          | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/drivers/staging/lttng/instrumentation/syscalls/headers/x86-64-syscalls-3.0.4_integers_override.h b/drivers/staging/lttng/instrumentation/syscalls/headers/x86-64-syscalls-3.0.4_integers_override.h
index 3d400f7..5f971e7 100644
--- a/drivers/staging/lttng/instrumentation/syscalls/headers/x86-64-syscalls-3.0.4_integers_override.h
+++ b/drivers/staging/lttng/instrumentation/syscalls/headers/x86-64-syscalls-3.0.4_integers_override.h
@@ -1,3 +1,3 @@
 /*
- * this is a place-holder for x86_64 interger syscall definition override.
+ * this is a place-holder for x86_64 integer syscall definition override.
  */
-- 
2.0.2

