From 5d9998f48fde3816cfc6a198c6885c5fe57b6754 Mon Sep 17 00:00:00 2001
From: Paul Woegerer <paul_woegerer@mentor.com>
Date: Wed, 14 Nov 2012 11:03:42 -0500
Subject: [PATCH 446/757] Add ifdefs to net probe to support Linux 2.6.39

Signed-off-by: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
---
 .../lttng/instrumentation/events/lttng-module/net.h | 21 +++++++++++++++++++++
 1 file changed, 21 insertions(+)

diff --git a/drivers/staging/lttng/instrumentation/events/lttng-module/net.h b/drivers/staging/lttng/instrumentation/events/lttng-module/net.h
index c25b0d9..589011c 100644
--- a/drivers/staging/lttng/instrumentation/events/lttng-module/net.h
+++ b/drivers/staging/lttng/instrumentation/events/lttng-module/net.h
@@ -8,29 +8,50 @@
 #include <linux/netdevice.h>
 #include <linux/ip.h>
 #include <linux/tracepoint.h>
+#include <linux/version.h>
 
 TRACE_EVENT(net_dev_xmit,
 
+#if (LINUX_VERSION_CODE >= KERNEL_VERSION(3,0,0))
 	TP_PROTO(struct sk_buff *skb,
 		 int rc,
 		 struct net_device *dev,
 		 unsigned int skb_len),
 
 	TP_ARGS(skb, rc, dev, skb_len),
+#else
+	TP_PROTO(struct sk_buff *skb,
+		 int rc),
+	
+	TP_ARGS(skb, rc),
+#endif
 
 	TP_STRUCT__entry(
 		__field(	void *,		skbaddr		)
 		__field(	unsigned int,	len		)
 		__field(	int,		rc		)
+#if (LINUX_VERSION_CODE >= KERNEL_VERSION(3,0,0))
 		__string(	name,		dev->name	)
+#else
+		__string(	name,		skb->dev->name	)
+#endif
 	),
 
+#if (LINUX_VERSION_CODE >= KERNEL_VERSION(3,0,0))
 	TP_fast_assign(
 		tp_assign(skbaddr, skb);
 		tp_assign(len, skb_len);
 		tp_assign(rc, rc);
 		tp_strcpy(name, dev->name);
 	),
+#else
+	TP_fast_assign(
+		tp_assign(skbaddr, skb);
+		tp_assign(len, skb->len);
+		tp_assign(rc, rc);
+		tp_strcpy(name, skb->dev->name);
+	),
+#endif
 
 	TP_printk("dev=%s skbaddr=%p len=%u rc=%d",
 		__get_str(name), __entry->skbaddr, __entry->len, __entry->rc)
-- 
2.0.2

