From f5bd9f4504e6db2e9ac0c5869b4c54918482cb19 Mon Sep 17 00:00:00 2001
From: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
Date: Thu, 9 May 2013 08:55:52 -0400
Subject: [PATCH 495/933] Version 2.2.0-rc2

Signed-off-by: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
---
 drivers/staging/lttng/ChangeLog | 11 +++++++++++
 1 file changed, 11 insertions(+)

diff --git a/drivers/staging/lttng/ChangeLog b/drivers/staging/lttng/ChangeLog
index dcf5e90..d0eb9c1 100644
--- a/drivers/staging/lttng/ChangeLog
+++ b/drivers/staging/lttng/ChangeLog
@@ -1,3 +1,14 @@
+2013-05-09 LTTng modules 2.2.0-rc2
+	* Remove useless lttng_metadata probe
+	* Fix: warning kvm_trace_exit_reason redefined for 3.9 kernel
+	* Fix: 3.9.0 block layer instrumentation build
+	* Revert "Stop build if CONFIG_TRACEPOINTS is not set"
+	* Stop build if CONFIG_TRACEPOINTS is not set
+	* Add comments to endifs.
+	* Fix: remove ARM set_tls system call override
+	* Cleanup: arm sys_mmap whitespace
+	* Clean up using global_dirty_limit wrapper for writeback probe
+
 2013-03-29 LTTng modules 2.2.0-rc1
 	* Update README
 	* Fix: vmscan instrumentation build
-- 
2.0.2

