From 50f644fb1233549a72c1d725705c8cac14bc026d Mon Sep 17 00:00:00 2001
From: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
Date: Thu, 8 Aug 2013 09:15:44 -0400
Subject: [PATCH 530/933] README: lttng-modules incompatible with lttng 0.x
 patchset

Signed-off-by: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
---
 drivers/staging/lttng/README | 6 +-----
 1 file changed, 1 insertion(+), 5 deletions(-)

diff --git a/drivers/staging/lttng/README b/drivers/staging/lttng/README
index b7d8b9f..9aa5b93 100644
--- a/drivers/staging/lttng/README
+++ b/drivers/staging/lttng/README
@@ -44,11 +44,7 @@ lttng-modules tree). It should work fine with newer kernels and other
 architectures, but expect build issues with kernels older than 2.6.36.
 The clock source currently used is the standard gettimeofday (slower,
 less scalable and less precise than the LTTng 0.x clocks).  Support for
-LTTng 0.x clocks will be added back soon into LTTng 2.0.  Please note
-that lttng-modules 2.0 can build on a Linux kernel patched with the
-LTTng 0.x patchset, but the lttng-modules 2.0 replace the lttng-modules
-0.x, so both tracers cannot be installed at the same time for a given
-kernel version.
+LTTng 0.x clocks will be added back soon into LTTng 2.0.
 
 
 * Kernel config options required
-- 
2.0.2

