From bd7607385fffbc93418bfdc0b04745ab8c3c9867 Mon Sep 17 00:00:00 2001
From: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
Date: Fri, 30 Aug 2013 14:41:31 -0400
Subject: [PATCH 532/933] Version 2.3.0-rc2

Signed-off-by: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
---
 drivers/staging/lttng/ChangeLog | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/drivers/staging/lttng/ChangeLog b/drivers/staging/lttng/ChangeLog
index b56214e..70ba791 100644
--- a/drivers/staging/lttng/ChangeLog
+++ b/drivers/staging/lttng/ChangeLog
@@ -1,3 +1,7 @@
+2013-08-30 LTTng modules 2.3.0-rc2
+	* Fix: metadata lttng channel refcount
+	* README: lttng-modules incompatible with lttng 0.x patchset
+
 2013-07-17 LTTng modules 2.3.0-rc1
 	* Fix: kvm x86 probes side-effect
 	* Fix: allow get/put subbuf in loop for metadata stream
-- 
2.0.2

