From ffc4da04ef76ce4b0db1cde32b8468ee657ae214 Mon Sep 17 00:00:00 2001
From: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
Date: Tue, 3 Sep 2013 00:16:58 -0400
Subject: [PATCH 534/757] Fix: ext3: 3.11 Linux support

Signed-off-by: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
---
 .../instrumentation/events/lttng-module/ext3.h     | 34 ++++++++++++++++++++++
 1 file changed, 34 insertions(+)

diff --git a/drivers/staging/lttng/instrumentation/events/lttng-module/ext3.h b/drivers/staging/lttng/instrumentation/events/lttng-module/ext3.h
index 32917fb..286a1df 100644
--- a/drivers/staging/lttng/instrumentation/events/lttng-module/ext3.h
+++ b/drivers/staging/lttng/instrumentation/events/lttng-module/ext3.h
@@ -295,6 +295,38 @@ DEFINE_EVENT(ext3__page_op, ext3_releasepage,
 	TP_ARGS(page)
 )
 
+#if (LINUX_VERSION_CODE >= KERNEL_VERSION(3,11,0))
+
+TRACE_EVENT(ext3_invalidatepage,
+	TP_PROTO(struct page *page, unsigned int offset, unsigned int length),
+
+	TP_ARGS(page, offset, length),
+
+	TP_STRUCT__entry(
+		__field(	pgoff_t, index			)
+		__field(	unsigned int, offset		)
+		__field(	unsigned int, length		)
+		__field(	ino_t,	ino			)
+		__field(	dev_t,	dev			)
+
+	),
+
+	TP_fast_assign(
+		tp_assign(index, page->index)
+		tp_assign(offset, offset)
+		tp_assign(length, length)
+		tp_assign(ino, page->mapping->host->i_ino)
+		tp_assign(dev, page->mapping->host->i_sb->s_dev)
+	),
+
+	TP_printk("dev %d,%d ino %lu page_index %lu offset %u length %u",
+		  MAJOR(__entry->dev), MINOR(__entry->dev),
+		  (unsigned long) __entry->ino,
+		  __entry->index, __entry->offset, __entry->length)
+)
+
+#else
+
 TRACE_EVENT(ext3_invalidatepage,
 	TP_PROTO(struct page *page, unsigned long offset),
 
@@ -321,6 +353,8 @@ TRACE_EVENT(ext3_invalidatepage,
 		  __entry->index, __entry->offset)
 )
 
+#endif
+
 TRACE_EVENT(ext3_discard_blocks,
 	TP_PROTO(struct super_block *sb, unsigned long blk,
 			unsigned long count),
-- 
2.0.2

