From 7ac902155abe1e706e770b73884cef6fbbd6939f Mon Sep 17 00:00:00 2001
From: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
Date: Sun, 24 Nov 2013 03:42:22 -0500
Subject: [PATCH 559/933] gcc blacklist: only apply when building with gcc

Signed-off-by: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
---
 drivers/staging/lttng/wrapper/compiler.h | 9 ++++++---
 1 file changed, 6 insertions(+), 3 deletions(-)

diff --git a/drivers/staging/lttng/wrapper/compiler.h b/drivers/staging/lttng/wrapper/compiler.h
index bd69fad..0c01632 100644
--- a/drivers/staging/lttng/wrapper/compiler.h
+++ b/drivers/staging/lttng/wrapper/compiler.h
@@ -27,12 +27,15 @@
  * Don't allow compiling with buggy compiler.
  */
 
+#ifdef GCC_VERSION
+
 /*
  * http://gcc.gnu.org/bugzilla/show_bug.cgi?id=58854
  */
-#ifdef __ARMEL__
-# if GCC_VERSION >= 40800 && GCC_VERSION <= 40802
-#  error Your gcc version produces clobbered frame accesses
+# ifdef __ARMEL__
+#  if GCC_VERSION >= 40800 && GCC_VERSION <= 40802
+#   error Your gcc version produces clobbered frame accesses
+#  endif
 # endif
 #endif
 
-- 
2.0.2

