From b3e6c9aa68dc25965888472a30057143c971aa97 Mon Sep 17 00:00:00 2001
From: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
Date: Tue, 10 Dec 2013 14:14:49 -0500
Subject: [PATCH 564/933] Version 2.4.0-rc2

Signed-off-by: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
---
 drivers/staging/lttng/ChangeLog | 8 ++++++++
 1 file changed, 8 insertions(+)

diff --git a/drivers/staging/lttng/ChangeLog b/drivers/staging/lttng/ChangeLog
index 9c656cf..8f12661 100644
--- a/drivers/staging/lttng/ChangeLog
+++ b/drivers/staging/lttng/ChangeLog
@@ -1,3 +1,11 @@
+2013-12-10 LTTng modules 2.4.0-rc2
+	* Fix: client_packet_header() uses wrong packet
+	* Cleanup: fix typo in ring buffer backend comment
+	* Fix: handle unknown event type parameter
+	* Fix: eliminate timestamp overlap between packets
+	* gcc blacklist: only apply when building with gcc
+	* Blacklist ARM gcc 4.8.0, 4.8.1, 4.8.2
+
 2013-11-15 LTTng modules 2.4.0-rc1
 	* Update ext4 instrumentation to Linux 3.12
 	* Update RCU instrumentation to Linux 3.12
-- 
2.0.2

