From 77e1705363f77ef4a11359914e77de109698544f Mon Sep 17 00:00:00 2001
From: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
Date: Fri, 14 Feb 2014 11:46:08 -0500
Subject: [PATCH 578/757] Version 2.4.0-rc4

Signed-off-by: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
---
 drivers/staging/lttng/ChangeLog | 6 ++++++
 1 file changed, 6 insertions(+)

diff --git a/drivers/staging/lttng/ChangeLog b/drivers/staging/lttng/ChangeLog
index 8963a98..f899ca5 100644
--- a/drivers/staging/lttng/ChangeLog
+++ b/drivers/staging/lttng/ChangeLog
@@ -1,3 +1,9 @@
+2014-02-14 LTTng modules 2.4.0-rc4
+	* Fix: use after free in ring buffer clients
+	* Fix: Add statedump exit so module is not permanent
+	* Cleanup: move callback init to static initialization
+	* Fix: OOT lttng-statedump tracepoints not visible with signed kernels
+
 2014-01-29 LTTng modules 2.4.0-rc3
 	* Fix: tracepoint event name mapping at unregister
 	* Fix: tracepoint name remapping
-- 
2.0.2

