From fb77d3bffecbf957abce5202537a3691cba00bed Mon Sep 17 00:00:00 2001
From: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
Date: Thu, 10 Apr 2014 15:46:14 -0400
Subject: [PATCH 595/757] Fix: add missing module version in v4l2 probe

Signed-off-by: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
---
 drivers/staging/lttng/probes/lttng-probe-v4l2.c | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/drivers/staging/lttng/probes/lttng-probe-v4l2.c b/drivers/staging/lttng/probes/lttng-probe-v4l2.c
index 709ccb2..3b1985c 100644
--- a/drivers/staging/lttng/probes/lttng-probe-v4l2.c
+++ b/drivers/staging/lttng/probes/lttng-probe-v4l2.c
@@ -23,6 +23,7 @@
 
 #include <linux/module.h>
 #include <linux/videodev2.h>
+#include "../lttng-tracer.h"
 
 /*
  * Create the tracepoint static inlines from the kernel to validate that our
@@ -42,3 +43,7 @@
 MODULE_LICENSE("GPL and additional rights");
 MODULE_AUTHOR("Wade Farnsworth <wade_farnsworth@mentor.com>");
 MODULE_DESCRIPTION("LTTng v4l2 probes");
+MODULE_VERSION(__stringify(LTTNG_MODULES_MAJOR_VERSION) "."
+	__stringify(LTTNG_MODULES_MINOR_VERSION) "."
+	__stringify(LTTNG_MODULES_PATCHLEVEL_VERSION)
+	LTTNG_MODULES_EXTRAVERSION);
-- 
2.0.2

