From 146f2ed50d715fb81d43a787a1515a8140d9834a Mon Sep 17 00:00:00 2001
From: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
Date: Tue, 9 Sep 2014 10:10:44 -0400
Subject: [PATCH 625/933] Add cscope to gitignore

Signed-off-by: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
---
 drivers/staging/lttng/.gitignore | 1 +
 1 file changed, 1 insertion(+)

diff --git a/drivers/staging/lttng/.gitignore b/drivers/staging/lttng/.gitignore
index 7a1e643..6deb422 100644
--- a/drivers/staging/lttng/.gitignore
+++ b/drivers/staging/lttng/.gitignore
@@ -30,6 +30,7 @@ modules.builtin
 *.lzma
 *.lzo
 *.gcno
+cscope.*
 
 #
 # Top-level generic files
-- 
2.0.2

