From 9a4359fd33647dbe8e9005f5c28abc675397ace0 Mon Sep 17 00:00:00 2001
From: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
Date: Thu, 11 Sep 2014 12:19:16 -0400
Subject: [PATCH 648/757] Fix: tracepoint list anonymous file name

Signed-off-by: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
---
 drivers/staging/lttng/lttng-abi.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/drivers/staging/lttng/lttng-abi.c b/drivers/staging/lttng/lttng-abi.c
index 0291811..6dc59ae 100644
--- a/drivers/staging/lttng/lttng-abi.c
+++ b/drivers/staging/lttng/lttng-abi.c
@@ -118,7 +118,7 @@ int lttng_abi_tracepoint_list(void)
 		goto fd_error;
 	}
 
-	tracepoint_list_file = anon_inode_getfile("[lttng_session]",
+	tracepoint_list_file = anon_inode_getfile("[lttng_tracepoint_list]",
 					  &lttng_tracepoint_list_fops,
 					  NULL, O_RDWR);
 	if (IS_ERR(tracepoint_list_file)) {
-- 
2.0.2

