From 4a4f6338974e419e3eaf20f039b2c70be230427b Mon Sep 17 00:00:00 2001
From: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
Date: Fri, 12 Sep 2014 15:25:38 -0400
Subject: [PATCH 652/933] syscall listing: show syscall ID

Signed-off-by: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
---
 drivers/staging/lttng/lttng-syscalls.c | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/drivers/staging/lttng/lttng-syscalls.c b/drivers/staging/lttng/lttng-syscalls.c
index 25b8431..0a8074f 100644
--- a/drivers/staging/lttng/lttng-syscalls.c
+++ b/drivers/staging/lttng/lttng-syscalls.c
@@ -1127,7 +1127,8 @@ int syscall_list_show(struct seq_file *m, void *p)
 		WARN_ON_ONCE(!(entry >= compat_sc_table
 			&& entry < compat_sc_table + ARRAY_SIZE(compat_sc_table)));
 	}
-	seq_printf(m,	"syscall { name = %s; bitness = %u; };\n",
+	seq_printf(m,	"syscall { id = %u; name = %s; bitness = %u; };\n",
+		entry - table,
 		entry->desc->name,
 		bitness);
 	return 0;
-- 
2.0.2

