From a80fd9fd594e4d60571fab2c2975e6ac5a1a8f9d Mon Sep 17 00:00:00 2001
From: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
Date: Fri, 12 Sep 2014 15:30:29 -0400
Subject: [PATCH 653/757] Cleanup: lttng-abi.h coding style

Signed-off-by: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
---
 drivers/staging/lttng/lttng-abi.h | 18 +++++++++---------
 1 file changed, 9 insertions(+), 9 deletions(-)

diff --git a/drivers/staging/lttng/lttng-abi.h b/drivers/staging/lttng/lttng-abi.h
index c59b8c6..7a09fa2 100644
--- a/drivers/staging/lttng/lttng-abi.h
+++ b/drivers/staging/lttng/lttng-abi.h
@@ -56,14 +56,14 @@ struct lttng_kernel_channel {
 	enum lttng_kernel_output output;	/* splice, mmap */
 	int overwrite;				/* 1: overwrite, 0: discard */
 	char padding[LTTNG_KERNEL_CHANNEL_PADDING];
-}__attribute__((packed));
+} __attribute__((packed));
 
 struct lttng_kernel_kretprobe {
 	uint64_t addr;
 
 	uint64_t offset;
 	char symbol_name[LTTNG_KERNEL_SYM_NAME_LEN];
-}__attribute__((packed));
+} __attribute__((packed));
 
 /*
  * Either addr is used, or symbol_name and offset.
@@ -73,11 +73,11 @@ struct lttng_kernel_kprobe {
 
 	uint64_t offset;
 	char symbol_name[LTTNG_KERNEL_SYM_NAME_LEN];
-}__attribute__((packed));
+} __attribute__((packed));
 
 struct lttng_kernel_function_tracer {
 	char symbol_name[LTTNG_KERNEL_SYM_NAME_LEN];
-}__attribute__((packed));
+} __attribute__((packed));
 
 struct lttng_kernel_syscall {
 	char enable;
@@ -101,13 +101,13 @@ struct lttng_kernel_event {
 		struct lttng_kernel_syscall syscall;
 		char padding[LTTNG_KERNEL_EVENT_PADDING2];
 	} u;
-}__attribute__((packed));
+} __attribute__((packed));
 
 struct lttng_kernel_tracer_version {
 	uint32_t major;
 	uint32_t minor;
 	uint32_t patchlevel;
-}__attribute__((packed));
+} __attribute__((packed));
 
 enum lttng_kernel_calibrate_type {
 	LTTNG_KERNEL_CALIBRATE_KRETPROBE,
@@ -115,7 +115,7 @@ enum lttng_kernel_calibrate_type {
 
 struct lttng_kernel_calibrate {
 	enum lttng_kernel_calibrate_type type;	/* type (input) */
-}__attribute__((packed));
+} __attribute__((packed));
 
 enum lttng_kernel_context_type {
 	LTTNG_KERNEL_CONTEXT_PID		= 0,
@@ -135,7 +135,7 @@ struct lttng_kernel_perf_counter_ctx {
 	uint32_t type;
 	uint64_t config;
 	char name[LTTNG_KERNEL_SYM_NAME_LEN];
-}__attribute__((packed));
+} __attribute__((packed));
 
 #define LTTNG_KERNEL_CONTEXT_PADDING1	16
 #define LTTNG_KERNEL_CONTEXT_PADDING2	LTTNG_KERNEL_SYM_NAME_LEN + 32
@@ -147,7 +147,7 @@ struct lttng_kernel_context {
 		struct lttng_kernel_perf_counter_ctx perf_counter;
 		char padding[LTTNG_KERNEL_CONTEXT_PADDING2];
 	} u;
-}__attribute__((packed));
+} __attribute__((packed));
 
 /* LTTng file descriptor ioctl */
 #define LTTNG_KERNEL_SESSION			_IO(0xF6, 0x45)
-- 
2.0.2

