From abc65f4676978535934953e26103097d7cc9c50e Mon Sep 17 00:00:00 2001
From: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
Date: Tue, 23 Sep 2014 19:48:01 -0400
Subject: [PATCH 657/757] Fix: syscall filtering: NULL pointer deref

Signed-off-by: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
---
 drivers/staging/lttng/lttng-syscalls.c | 8 ++++----
 1 file changed, 4 insertions(+), 4 deletions(-)

diff --git a/drivers/staging/lttng/lttng-syscalls.c b/drivers/staging/lttng/lttng-syscalls.c
index cba00160..1489934 100644
--- a/drivers/staging/lttng/lttng-syscalls.c
+++ b/drivers/staging/lttng/lttng-syscalls.c
@@ -1060,18 +1060,18 @@ int lttng_syscall_filter_disable(struct lttng_channel *chan,
 		goto error;
 	}
 	if (syscall_nr >= 0) {
-		if (!test_bit(syscall_nr, chan->sc_filter->sc)) {
+		if (!test_bit(syscall_nr, filter->sc)) {
 			ret = -EEXIST;
 			goto error;
 		}
-		bitmap_clear(chan->sc_filter->sc, syscall_nr, 1);
+		bitmap_clear(filter->sc, syscall_nr, 1);
 	}
 	if (compat_syscall_nr >= 0) {
-		if (!test_bit(compat_syscall_nr, chan->sc_filter->sc_compat)) {
+		if (!test_bit(compat_syscall_nr, filter->sc_compat)) {
 			ret = -EEXIST;
 			goto error;
 		}
-		bitmap_clear(chan->sc_filter->sc_compat, compat_syscall_nr, 1);
+		bitmap_clear(filter->sc_compat, compat_syscall_nr, 1);
 	}
 apply_filter:
 	if (!chan->sc_filter)
-- 
2.0.2

