From 1d67880f296554c1a4e568c3c562ea96e9f14873 Mon Sep 17 00:00:00 2001
From: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
Date: Tue, 23 Sep 2014 21:02:51 -0400
Subject: [PATCH 658/933] Fix: syscall: fail disable all if all already
 disabled

Signed-off-by: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
---
 drivers/staging/lttng/lttng-syscalls.c | 8 ++++++++
 1 file changed, 8 insertions(+)

diff --git a/drivers/staging/lttng/lttng-syscalls.c b/drivers/staging/lttng/lttng-syscalls.c
index 1489934..00584e3 100644
--- a/drivers/staging/lttng/lttng-syscalls.c
+++ b/drivers/staging/lttng/lttng-syscalls.c
@@ -1048,6 +1048,14 @@ int lttng_syscall_filter_disable(struct lttng_channel *chan,
 	}
 
 	if (!name) {
+		/* Fail if all syscalls are already disabled. */
+		if (bitmap_empty(filter->sc, NR_syscalls)
+			&& bitmap_empty(filter->sc_compat,
+				NR_compat_syscalls)) {
+			ret = -EEXIST;
+			goto error;
+		}
+
 		/* Disable all system calls */
 		bitmap_clear(filter->sc, 0, NR_syscalls);
 		bitmap_clear(filter->sc_compat, 0, NR_compat_syscalls);
-- 
2.0.2

