From eb178e5d4b8c4c8346241ed1d0879e3297ec97d5 Mon Sep 17 00:00:00 2001
From: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
Date: Wed, 24 Sep 2014 12:17:03 -0400
Subject: [PATCH 659/933] Fix: don't allow disabling syscalls when none are
 enabled

Signed-off-by: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
---
 drivers/staging/lttng/lttng-syscalls.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/drivers/staging/lttng/lttng-syscalls.c b/drivers/staging/lttng/lttng-syscalls.c
index 00584e3..5cac721 100644
--- a/drivers/staging/lttng/lttng-syscalls.c
+++ b/drivers/staging/lttng/lttng-syscalls.c
@@ -1036,6 +1036,8 @@ int lttng_syscall_filter_disable(struct lttng_channel *chan,
 	WARN_ON_ONCE(!chan->sc_table);
 
 	if (!chan->sc_filter) {
+		if (!chan->syscall_all)
+			return -EEXIST;
 		filter = kzalloc(sizeof(struct lttng_syscall_filter),
 				GFP_KERNEL);
 		if (!filter)
-- 
2.0.2

