From c417bbd585790e096679d69ac47d525af9577d68 Mon Sep 17 00:00:00 2001
From: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
Date: Wed, 24 Sep 2014 16:08:34 -0400
Subject: [PATCH 660/933] Fix: syscall listing: handle "enable all syscall"

Signed-off-by: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
---
 drivers/staging/lttng/lttng-syscalls.c | 6 +++---
 1 file changed, 3 insertions(+), 3 deletions(-)

diff --git a/drivers/staging/lttng/lttng-syscalls.c b/drivers/staging/lttng/lttng-syscalls.c
index 5cac721..846946a 100644
--- a/drivers/staging/lttng/lttng-syscalls.c
+++ b/drivers/staging/lttng/lttng-syscalls.c
@@ -1230,12 +1230,12 @@ long lttng_channel_syscall_mask(struct lttng_channel *channel,
 
 	for (bit = 0; bit < ARRAY_SIZE(sc_table); bit++) {
 		bt_bitfield_write_be(tmp_mask, char, bit, 1,
-			test_bit(bit, filter->sc));
+			filter ? test_bit(bit, filter->sc) : 1);
 	}
 	for (; bit < sc_tables_len; bit++) {
 		bt_bitfield_write_be(tmp_mask, char, bit, 1,
-			test_bit(bit - ARRAY_SIZE(sc_table),
-				filter->sc_compat));
+			filter ? test_bit(bit - ARRAY_SIZE(sc_table),
+				filter->sc_compat) : 1);
 	}
 	if (copy_to_user(usyscall_mask->mask, tmp_mask, bitmask_len))
 		ret = -EFAULT;
-- 
2.0.2

