From 1854e40647534d0404e3cae8f2c7e196a64ad353 Mon Sep 17 00:00:00 2001
From: Julien Desfossez <jdesfossez@efficios.com>
Date: Wed, 1 Oct 2014 11:05:08 -0400
Subject: [PATCH 667/933] Introduce macros to check Ubuntu kernel version

Signed-off-by: Julien Desfossez <jdesfossez@efficios.com>
Signed-off-by: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
---
 drivers/staging/lttng/lttng-kernel-version.h | 15 +++++++++++++++
 1 file changed, 15 insertions(+)

diff --git a/drivers/staging/lttng/lttng-kernel-version.h b/drivers/staging/lttng/lttng-kernel-version.h
index 0eb56b3..8fad8b0 100644
--- a/drivers/staging/lttng/lttng-kernel-version.h
+++ b/drivers/staging/lttng/lttng-kernel-version.h
@@ -24,6 +24,7 @@
  */
 
 #include <linux/version.h>
+#include <linux/vermagic.h>
 
 /*
  * This macro checks if the kernel version is between the two specified
@@ -33,4 +34,18 @@
 	(LINUX_VERSION_CODE >= KERNEL_VERSION(a_low, b_low, c_low) && \
 	 LINUX_VERSION_CODE < KERNEL_VERSION(a_high, b_high, c_high))
 
+#define LTTNG_UBUNTU_KERNEL_VERSION(a, b, c, d) \
+	(((a) << 24) + ((b) << 16) + (c << 8) + (d))
+
+#define LTTNG_UBUNTU_VERSION_CODE \
+	((LINUX_VERSION_CODE << 8) + UTS_UBUNTU_RELEASE_ABI)
+
+#define LTTNG_UBUNTU_KERNEL_RANGE(a_low, b_low, c_low, d_low, \
+		a_high, b_high, c_high, d_high) \
+	(defined(UTS_UBUNTU_RELEASE_ABI) && \
+		LTTNG_UBUNTU_VERSION_CODE >= \
+		LTTNG_UBUNTU_KERNEL_VERSION(a_low, b_low, c_low, d_low) && \
+		LTTNG_UBUNTU_VERSION_CODE < \
+		LTTNG_UBUNTU_KERNEL_VERSION(a_high, b_high, c_high, d_high))
+
 #endif /* _LTTNG_KERNEL_VERSION_H */
-- 
2.0.2

