From ec90cc0f3a2422edd1270979797609f6b886524d Mon Sep 17 00:00:00 2001
From: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
Date: Wed, 1 Oct 2014 12:05:41 -0400
Subject: [PATCH 671/933] Check for stale version.h files

Signed-off-by: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
---
 drivers/staging/lttng/Makefile | 14 ++++++++++++++
 1 file changed, 14 insertions(+)

diff --git a/drivers/staging/lttng/Makefile b/drivers/staging/lttng/Makefile
index 606b5ec..aa4835d 100644
--- a/drivers/staging/lttng/Makefile
+++ b/drivers/staging/lttng/Makefile
@@ -5,6 +5,20 @@
 ifneq ($(KERNELRELEASE),)
 ifneq ($(CONFIG_TRACEPOINTS),)
 
+lttng_check_linux_version = $(shell pwd)/include/linux/version.h
+lttng_check_generated_linux_version = $(shell pwd)/include/generated/uapi/linux/version.h
+
+#
+# Check for stale version.h, which can be a leftover from an old Linux
+# kernel tree moved to a newer kernel version, only pruned by make
+# distclean.
+#
+ifneq ($(wildcard $(lttng_check_linux_version)),)
+ifneq ($(wildcard $(lttng_check_generated_linux_version)),)
+$(error Duplicate version.h files found in $(lttng_check_linux_version) and $(lttng_check_generated_linux_version). Consider running make distclean on your kernel, or removing the stale $(lttng_check_linux_version) file)
+endif
+endif
+
 obj-m += lttng-ring-buffer-client-discard.o
 obj-m += lttng-ring-buffer-client-overwrite.o
 obj-m += lttng-ring-buffer-metadata-client.o
-- 
2.0.2

