From 7c9388a69c0547fb858dd42b864958c369803a4e Mon Sep 17 00:00:00 2001
From: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
Date: Thu, 2 Oct 2014 15:54:53 -0400
Subject: [PATCH 673/933] Cleanup: remove unused trace_clock_read32()

Signed-off-by: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
---
 drivers/staging/lttng/wrapper/trace-clock.h | 5 -----
 1 file changed, 5 deletions(-)

diff --git a/drivers/staging/lttng/wrapper/trace-clock.h b/drivers/staging/lttng/wrapper/trace-clock.h
index b145b82..5705d0c 100644
--- a/drivers/staging/lttng/wrapper/trace-clock.h
+++ b/drivers/staging/lttng/wrapper/trace-clock.h
@@ -140,11 +140,6 @@ static inline u64 trace_clock_monotonic_wrapper(void)
 }
 #endif /* #else #if (LINUX_VERSION_CODE >= KERNEL_VERSION(3,17,0)) */
 
-static inline u32 trace_clock_read32(void)
-{
-	return (u32) trace_clock_monotonic_wrapper();
-}
-
 static inline u64 trace_clock_read64(void)
 {
 	return (u64) trace_clock_monotonic_wrapper();
-- 
2.0.2

