From aaeec692c3e060ecb4e02b142936c6d31571d94a Mon Sep 17 00:00:00 2001
From: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
Date: Mon, 20 Oct 2014 18:21:22 -0400
Subject: [PATCH 680/933] Fix: syscall list ioctl number conflict

Signed-off-by: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
---
 drivers/staging/lttng/lttng-abi.h | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/drivers/staging/lttng/lttng-abi.h b/drivers/staging/lttng/lttng-abi.h
index 322989a..8f5c093 100644
--- a/drivers/staging/lttng/lttng-abi.h
+++ b/drivers/staging/lttng/lttng-abi.h
@@ -162,7 +162,7 @@ struct lttng_kernel_context {
 #define LTTNG_KERNEL_WAIT_QUIESCENT		_IO(0xF6, 0x48)
 #define LTTNG_KERNEL_CALIBRATE			\
 	_IOWR(0xF6, 0x49, struct lttng_kernel_calibrate)
-#define LTTNG_KERNEL_SYSCALL_LIST		_IO(0xF6, 0x49)
+#define LTTNG_KERNEL_SYSCALL_LIST		_IO(0xF6, 0x4A)
 
 /* Session FD ioctl */
 #define LTTNG_KERNEL_METADATA			\
-- 
2.0.2

