From 8880ae75f16bf4601b63dbf8de4e6ac5fa4e26b6 Mon Sep 17 00:00:00 2001
From: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
Date: Thu, 23 Apr 2015 17:51:50 -0400
Subject: [PATCH 716/757] Fix: atomic.h wrapper for kernels < 2.6.37

Signed-off-by: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
---
 drivers/staging/lttng/wrapper/atomic.h | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/drivers/staging/lttng/wrapper/atomic.h b/drivers/staging/lttng/wrapper/atomic.h
index 6a3cd07..f3a3dab 100644
--- a/drivers/staging/lttng/wrapper/atomic.h
+++ b/drivers/staging/lttng/wrapper/atomic.h
@@ -24,7 +24,11 @@
  */
 
 #include <linux/version.h>
+#if (LINUX_VERSION_CODE >= KERNEL_VERSION(2,6,37))
 #include <linux/atomic.h>
+#else
+#include <asm/atomic.h>
+#endif
 
 #if (LINUX_VERSION_CODE >= KERNEL_VERSION(3,16,0))
 static inline void lttng_smp_mb__before_atomic(void)
-- 
2.0.2

