From 68cef7e0bdb5e43d7e1a757cbd003638e503b462 Mon Sep 17 00:00:00 2001
From: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
Date: Thu, 16 Apr 2015 17:14:53 -0400
Subject: [PATCH 729/933] Be more restrictive in provider name check

Signed-off-by: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
---
 drivers/staging/lttng/lttng-probes.c | 8 ++++++++
 1 file changed, 8 insertions(+)

diff --git a/drivers/staging/lttng/lttng-probes.c b/drivers/staging/lttng/lttng-probes.c
index b5a7fc8..ee75e33 100644
--- a/drivers/staging/lttng/lttng-probes.c
+++ b/drivers/staging/lttng/lttng-probes.c
@@ -60,6 +60,14 @@ int check_event_provider(struct lttng_probe_desc *desc)
 				desc->provider,
 				provider_name_len))
 			return 0;	/* provider mismatch */
+		/*
+		 * The event needs to contain at least provider name + _ +
+		 * one or more letter.
+		 */
+		if (strlen(desc->event_desc[i]->name) <= provider_name_len + 1)
+			return 0;	/* provider mismatch */
+		if (desc->event_desc[i]->name[provider_name_len] != '_')
+			return 0;	/* provider mismatch */
 	}
 	return 1;
 }
-- 
2.0.2

