From 824f6e68416b6f2da5f3b7aa9e21e220644ba2d9 Mon Sep 17 00:00:00 2001
From: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
Date: Fri, 17 Apr 2015 14:49:38 -0400
Subject: [PATCH 736/933] Fix: kernels >= 2.6.33 put utsrelease.h in generated/

Signed-off-by: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
---
 drivers/staging/lttng/lttng-kernel-version.h | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/drivers/staging/lttng/lttng-kernel-version.h b/drivers/staging/lttng/lttng-kernel-version.h
index e8353fd..ec8cee0 100644
--- a/drivers/staging/lttng/lttng-kernel-version.h
+++ b/drivers/staging/lttng/lttng-kernel-version.h
@@ -24,7 +24,12 @@
  */
 
 #include <linux/version.h>
+
+#if (LINUX_VERSION_CODE >= KERNEL_VERSION(2,6,33))
+#include <generated/utsrelease.h>
+#else /* #if (LINUX_VERSION_CODE >= KERNEL_VERSION(2,6,33)) */
 #include <linux/utsrelease.h>
+#endif /* #else #if (LINUX_VERSION_CODE >= KERNEL_VERSION(2,6,33)) */
 
 /*
  * This macro checks if the kernel version is between the two specified
-- 
2.0.2

