From e3585903b9062a9e7198bb1d27e18c830634af8b Mon Sep 17 00:00:00 2001
From: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
Date: Thu, 23 Apr 2015 16:11:15 -0400
Subject: [PATCH 740/933] Fix: double-unlock of session mutex

Signed-off-by: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
---
 drivers/staging/lttng/lttng-events.c | 1 -
 1 file changed, 1 deletion(-)

diff --git a/drivers/staging/lttng/lttng-events.c b/drivers/staging/lttng/lttng-events.c
index 053f125..1ae894b 100644
--- a/drivers/staging/lttng/lttng-events.c
+++ b/drivers/staging/lttng/lttng-events.c
@@ -643,7 +643,6 @@ struct lttng_event *_lttng_event_create(struct lttng_channel *chan,
 	}
 	hlist_add_head(&event->hlist, head);
 	list_add(&event->list, &chan->session->events);
-	mutex_unlock(&sessions_mutex);
 	return event;
 
 statedump_error:
-- 
2.0.2

