From 520d1e9d5bc2394a0e9382f07160451bb99f2355 Mon Sep 17 00:00:00 2001
From: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
Date: Thu, 23 Apr 2015 17:42:06 -0400
Subject: [PATCH 746/933] Fix: add missing types.h include for older kernels

Signed-off-by: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
---
 drivers/staging/lttng/lttng-filter-validator.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/drivers/staging/lttng/lttng-filter-validator.c b/drivers/staging/lttng/lttng-filter-validator.c
index fb5d2d6..a996945 100644
--- a/drivers/staging/lttng/lttng-filter-validator.c
+++ b/drivers/staging/lttng/lttng-filter-validator.c
@@ -20,6 +20,7 @@
  * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
  */
 
+#include <linux/types.h>
 #include <linux/jhash.h>
 #include <linux/slab.h>
 
-- 
2.0.2

