From 68b61bc9b69aec0efbf4a57dbc4b6e39ee56c6a6 Mon Sep 17 00:00:00 2001
From: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
Date: Thu, 16 Jul 2015 13:02:23 -0400
Subject: [PATCH 749/757] Fix: ext4 instrumentation for 4.2 kernels

Signed-off-by: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
---
 .../instrumentation/events/lttng-module/ext4.h     | 30 ++++++++++++++++++++++
 1 file changed, 30 insertions(+)

diff --git a/drivers/staging/lttng/instrumentation/events/lttng-module/ext4.h b/drivers/staging/lttng/instrumentation/events/lttng-module/ext4.h
index 6169fef..1c2865f 100644
--- a/drivers/staging/lttng/instrumentation/events/lttng-module/ext4.h
+++ b/drivers/staging/lttng/instrumentation/events/lttng-module/ext4.h
@@ -1511,6 +1511,35 @@ LTTNG_TRACEPOINT_EVENT(ext4_da_update_reserve_space,
 #endif
 )
 
+#if (LINUX_VERSION_CODE >= KERNEL_VERSION(4,2,0))
+LTTNG_TRACEPOINT_EVENT(ext4_da_reserve_space,
+	TP_PROTO(struct inode *inode),
+
+	TP_ARGS(inode),
+
+	TP_STRUCT__entry(
+		__field(	dev_t,	dev			)
+		__field(	ino_t,	ino			)
+		__field(	__u64,	i_blocks		)
+		__field(	int,	reserved_data_blocks	)
+		__field(	int,	reserved_meta_blocks	)
+		__field(	TP_MODE_T, mode			)
+	),
+
+	TP_fast_assign(
+		tp_assign(dev, inode->i_sb->s_dev)
+		tp_assign(ino, inode->i_ino)
+		tp_assign(i_blocks, inode->i_blocks)
+		tp_assign(reserved_data_blocks,
+				EXT4_I(inode)->i_reserved_data_blocks)
+		tp_assign(reserved_meta_blocks,
+				EXT4_I(inode)->i_reserved_meta_blocks)
+		tp_assign(mode, inode->i_mode)
+	),
+
+	TP_printk()
+)
+#else /* #if (LINUX_VERSION_CODE >= KERNEL_VERSION(4,2,0)) */
 LTTNG_TRACEPOINT_EVENT(ext4_da_reserve_space,
 	TP_PROTO(struct inode *inode, int md_needed),
 
@@ -1546,6 +1575,7 @@ LTTNG_TRACEPOINT_EVENT(ext4_da_reserve_space,
 		  __entry->md_needed, __entry->reserved_data_blocks,
 		  __entry->reserved_meta_blocks)
 )
+#endif /* #else #if (LINUX_VERSION_CODE >= KERNEL_VERSION(4,2,0)) */
 
 LTTNG_TRACEPOINT_EVENT(ext4_da_release_space,
 	TP_PROTO(struct inode *inode, int freed_blocks),
-- 
2.0.2

