From c2172e0d741c863fc6962b24adf582b6eeef2ab8 Mon Sep 17 00:00:00 2001
From: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
Date: Fri, 4 Sep 2015 01:12:53 -0400
Subject: [PATCH 756/757] Fix: use after free in channel release

Signed-off-by: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
---
 drivers/staging/lttng/lttng-abi.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/drivers/staging/lttng/lttng-abi.c b/drivers/staging/lttng/lttng-abi.c
index 6993a46..bdc6c76 100644
--- a/drivers/staging/lttng/lttng-abi.c
+++ b/drivers/staging/lttng/lttng-abi.c
@@ -1234,8 +1234,8 @@ int lttng_metadata_channel_release(struct inode *inode, struct file *file)
 	struct lttng_channel *channel = file->private_data;
 
 	if (channel) {
-		lttng_metadata_channel_destroy(channel);
 		fput(channel->session->file);
+		lttng_metadata_channel_destroy(channel);
 	}
 
 	return 0;
-- 
2.0.2

