From 3f2faa4abf0aefbe0d5c48f9fe09571c8151aadf Mon Sep 17 00:00:00 2001
From: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
Date: Mon, 27 Apr 2015 17:32:54 -0400
Subject: [PATCH 757/933] Fix: x86 mmutrace event extra field for kernels >=
 3.11

Signed-off-by: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
---
 .../events/lttng-module/arch/x86/kvm/mmutrace.h    | 23 ++++++++++++++++++++++
 1 file changed, 23 insertions(+)

diff --git a/drivers/staging/lttng/instrumentation/events/lttng-module/arch/x86/kvm/mmutrace.h b/drivers/staging/lttng/instrumentation/events/lttng-module/arch/x86/kvm/mmutrace.h
index a71b2a1..e0234ff 100644
--- a/drivers/staging/lttng/instrumentation/events/lttng-module/arch/x86/kvm/mmutrace.h
+++ b/drivers/staging/lttng/instrumentation/events/lttng-module/arch/x86/kvm/mmutrace.h
@@ -3,6 +3,7 @@
 
 #include "../../../../../../probes/lttng-tracepoint-event.h"
 #include <linux/ftrace_event.h>
+#include <linux/version.h>
 
 #undef TRACE_SYSTEM
 #define TRACE_SYSTEM kvm_mmu
@@ -117,6 +118,26 @@ LTTNG_TRACEPOINT_EVENT_INSTANCE(kvm_mmu_page_class, kvm_mmu_prepare_zap_page,
 	TP_ARGS(sp)
 )
 
+#if (LINUX_VERSION_CODE >= KERNEL_VERSION(3,11,0))
+
+LTTNG_TRACEPOINT_EVENT_MAP(
+	mark_mmio_spte,
+
+	kvm_mmu_mark_mmio_spte,
+
+	TP_PROTO(u64 *sptep, gfn_t gfn, unsigned access, unsigned int gen),
+	TP_ARGS(sptep, gfn, access, gen),
+
+	TP_FIELDS(
+		ctf_integer(void *, sptep, sptep)
+		ctf_integer(gfn_t, gfn, gfn)
+		ctf_integer(unsigned, access, access)
+		ctf_integer(unsigned int, gen, gen)
+	)
+)
+
+#else /* #if (LINUX_VERSION_CODE >= KERNEL_VERSION(3,11,0)) */
+
 LTTNG_TRACEPOINT_EVENT_MAP(
 	mark_mmio_spte,
 
@@ -132,6 +153,8 @@ LTTNG_TRACEPOINT_EVENT_MAP(
 	)
 )
 
+#endif /* #else #if (LINUX_VERSION_CODE >= KERNEL_VERSION(3,11,0)) */
+
 LTTNG_TRACEPOINT_EVENT_MAP(
 	handle_mmio_page_fault,
 
-- 
2.0.2

