From abb362bbf76328363be6f4eb7bce49c0d8343410 Mon Sep 17 00:00:00 2001
From: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
Date: Thu, 11 Jun 2015 17:38:46 +0200
Subject: [PATCH 776/933] Fix: context printk mismatch on error

Signed-off-by: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
---
 drivers/staging/lttng/lttng-context.c | 20 ++++++++++----------
 1 file changed, 10 insertions(+), 10 deletions(-)

diff --git a/drivers/staging/lttng/lttng-context.c b/drivers/staging/lttng/lttng-context.c
index c2ce68d..d8ce977 100644
--- a/drivers/staging/lttng/lttng-context.c
+++ b/drivers/staging/lttng/lttng-context.c
@@ -226,43 +226,43 @@ int lttng_context_init(void)
 
 	ret = lttng_add_hostname_to_ctx(&lttng_static_ctx);
 	if (ret) {
-		printk(KERN_WARNING "Cannot add context lttng_add_procname_to_ctx");
+		printk(KERN_WARNING "Cannot add context lttng_add_hostname_to_ctx");
 	}
 	ret = lttng_add_nice_to_ctx(&lttng_static_ctx);
 	if (ret) {
-		printk(KERN_WARNING "Cannot add context lttng_add_procname_to_ctx");
+		printk(KERN_WARNING "Cannot add context lttng_add_nice_to_ctx");
 	}
 	ret = lttng_add_pid_to_ctx(&lttng_static_ctx);
 	if (ret) {
-		printk(KERN_WARNING "Cannot add context lttng_add_pthread_id_to_ctx");
+		printk(KERN_WARNING "Cannot add context lttng_add_pid_to_ctx");
 	}
 	ret = lttng_add_ppid_to_ctx(&lttng_static_ctx);
 	if (ret) {
-		printk(KERN_WARNING "Cannot add context lttng_add_procname_to_ctx");
+		printk(KERN_WARNING "Cannot add context lttng_add_ppid_to_ctx");
 	}
 	ret = lttng_add_prio_to_ctx(&lttng_static_ctx);
 	if (ret) {
-		printk(KERN_WARNING "Cannot add context lttng_add_vpid_to_ctx");
+		printk(KERN_WARNING "Cannot add context lttng_add_prio_to_ctx");
 	}
 	ret = lttng_add_procname_to_ctx(&lttng_static_ctx);
 	if (ret) {
-		printk(KERN_WARNING "Cannot add context lttng_add_vtid_to_ctx");
+		printk(KERN_WARNING "Cannot add context lttng_add_procname_to_ctx");
 	}
 	ret = lttng_add_tid_to_ctx(&lttng_static_ctx);
 	if (ret) {
-		printk(KERN_WARNING "Cannot add context lttng_add_procname_to_ctx");
+		printk(KERN_WARNING "Cannot add context lttng_add_tid_to_ctx");
 	}
 	ret = lttng_add_vppid_to_ctx(&lttng_static_ctx);
 	if (ret) {
-		printk(KERN_WARNING "Cannot add context lttng_add_procname_to_ctx");
+		printk(KERN_WARNING "Cannot add context lttng_add_vppid_to_ctx");
 	}
 	ret = lttng_add_vtid_to_ctx(&lttng_static_ctx);
 	if (ret) {
-		printk(KERN_WARNING "Cannot add context lttng_add_procname_to_ctx");
+		printk(KERN_WARNING "Cannot add context lttng_add_vtid_to_ctx");
 	}
 	ret = lttng_add_vpid_to_ctx(&lttng_static_ctx);
 	if (ret) {
-		printk(KERN_WARNING "Cannot add context lttng_add_procname_to_ctx");
+		printk(KERN_WARNING "Cannot add context lttng_add_vpid_to_ctx");
 	}
 	/* TODO: perf counters for filtering */
 	return 0;
-- 
2.0.2

