From 7b1449b62474a8c77650fbc9c33b62cb335919a6 Mon Sep 17 00:00:00 2001
From: Julien Desfossez <jdesfossez@efficios.com>
Date: Wed, 25 Nov 2015 12:54:03 -0500
Subject: [PATCH 824/933] Bump version number for development branch

Signed-off-by: Julien Desfossez <jdesfossez@efficios.com>
Signed-off-by: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
---
 drivers/staging/lttng/lttng-tracer.h | 9 ++++-----
 1 file changed, 4 insertions(+), 5 deletions(-)

diff --git a/drivers/staging/lttng/lttng-tracer.h b/drivers/staging/lttng/lttng-tracer.h
index 020097f..a1972c5 100644
--- a/drivers/staging/lttng/lttng-tracer.h
+++ b/drivers/staging/lttng/lttng-tracer.h
@@ -40,13 +40,12 @@
 #include "lttng-events.h"
 
 #define LTTNG_MODULES_MAJOR_VERSION 2
-#define LTTNG_MODULES_MINOR_VERSION 7
+#define LTTNG_MODULES_MINOR_VERSION 8
 #define LTTNG_MODULES_PATCHLEVEL_VERSION 0
-#define LTTNG_MODULES_EXTRAVERSION "-rc1"
+#define LTTNG_MODULES_EXTRAVERSION "-pre0"
 
-#define LTTNG_VERSION_NAME		"Herbe à Détourne"
-#define LTTNG_VERSION_DESCRIPTION	\
-	"Brewed with unrestrained amounts of Citra hop, the Herbe à Détourne is a fantastic New World Tripel brewed by \"Dieu du Ciel!\". Aromas of mango, cantaloupe melon and passion fruit, combined with a controlled bitter finish, unite in making this smooth golden-orange beer stand apart."
+#define LTTNG_VERSION_NAME		"TBD"
+#define LTTNG_VERSION_DESCRIPTION	"TBD"
 
 #ifndef CHAR_BIT
 #define CHAR_BIT 8
-- 
2.0.2

