From 6c607dc0ef4d83fdebd3909ae08d452e92c7e9a1 Mon Sep 17 00:00:00 2001
From: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
Date: Wed, 3 Feb 2016 19:17:35 -0500
Subject: [PATCH 842/933] Expect filter context identifiers starting with $ctx.

Signed-off-by: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
---
 drivers/staging/lttng/lttng-context.c | 8 +++++++-
 1 file changed, 7 insertions(+), 1 deletion(-)

diff --git a/drivers/staging/lttng/lttng-context.c b/drivers/staging/lttng/lttng-context.c
index 9bf4d80..1121f6e 100644
--- a/drivers/staging/lttng/lttng-context.c
+++ b/drivers/staging/lttng/lttng-context.c
@@ -56,14 +56,20 @@ EXPORT_SYMBOL_GPL(lttng_find_context);
 int lttng_get_context_index(struct lttng_ctx *ctx, const char *name)
 {
 	unsigned int i;
+	const char *subname;
 
 	if (!ctx)
 		return -1;
+	if (strncmp(name, "$ctx.", strlen("$ctx.")) == 0) {
+		subname = name + strlen("$ctx.");
+	} else {
+		subname = name;
+	}
 	for (i = 0; i < ctx->nr_fields; i++) {
 		/* Skip allocated (but non-initialized) contexts */
 		if (!ctx->fields[i].event_field.name)
 			continue;
-		if (!strcmp(ctx->fields[i].event_field.name, name))
+		if (!strcmp(ctx->fields[i].event_field.name, subname))
 			return i;
 	}
 	return -1;
-- 
2.0.2

