From 9e90ba9205cb37b9581153cab39cdb4c9793784b Mon Sep 17 00:00:00 2001
From: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
Date: Mon, 7 Mar 2016 18:24:25 -0500
Subject: [PATCH 857/933] PowerPC-64 ABIv1: add build error if KALLSYMS_ALL is
 missing

Signed-off-by: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
---
 drivers/staging/lttng/wrapper/kallsyms.h | 10 ++++++++++
 1 file changed, 10 insertions(+)

diff --git a/drivers/staging/lttng/wrapper/kallsyms.h b/drivers/staging/lttng/wrapper/kallsyms.h
index ad9e1f2..8200018 100644
--- a/drivers/staging/lttng/wrapper/kallsyms.h
+++ b/drivers/staging/lttng/wrapper/kallsyms.h
@@ -28,6 +28,16 @@
 
 #include <linux/kallsyms.h>
 
+/*
+ * PowerPC ABIv1 needs KALLSYMS_ALL to get the function descriptor,
+ * which is needed to perform the function call.
+ */
+#if defined(CONFIG_PPC64) && (!defined(_CALL_ELF) || _CALL_ELF < 2)
+# ifndef CONFIG_KALLSYMS_ALL
+#  error "LTTng-modules requires CONFIG_KALLSYMS_ALL on PowerPC ABIv1"
+# endif
+#endif
+
 static inline
 unsigned long kallsyms_lookup_funcptr(const char *name)
 {
-- 
2.0.2

