From 2012b4e3cb0d393467c0356b21c1c2c8ec456699 Mon Sep 17 00:00:00 2001
From: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
Date: Tue, 15 Mar 2016 11:03:08 -0400
Subject: [PATCH 868/933] Update gitignore

Signed-off-by: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
---
 drivers/staging/lttng/.gitignore | 4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

diff --git a/drivers/staging/lttng/.gitignore b/drivers/staging/lttng/.gitignore
index 6deb422..f22e828 100644
--- a/drivers/staging/lttng/.gitignore
+++ b/drivers/staging/lttng/.gitignore
@@ -42,7 +42,9 @@ cscope.*
 /vmlinuz
 /System.map
 /Module.markers
-/Module.symvers
+
+# ignore across tree
+Module.symvers
 
 #
 # git files that we don't want to ignore even it they are dot-files
-- 
2.0.2

