From a307d7706798e7c9c545d3f269bc55219969836f Mon Sep 17 00:00:00 2001
From: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
Date: Tue, 15 Mar 2016 12:24:12 -0400
Subject: [PATCH 877/933] probes/Kbuild: remove upper level header path

Signed-off-by: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
---
 drivers/staging/lttng/probes/Kbuild | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/drivers/staging/lttng/probes/Kbuild b/drivers/staging/lttng/probes/Kbuild
index 1c6b9d0..265b50a 100644
--- a/drivers/staging/lttng/probes/Kbuild
+++ b/drivers/staging/lttng/probes/Kbuild
@@ -2,7 +2,7 @@ MAKEFILEDIR = $(shell dirname $(lastword $(MAKEFILE_LIST)))
 
 include $(MAKEFILEDIR)/../Makefile.ABI.workarounds
 
-ccflags-y += -I$(PWD) -I$(PWD)/..
+ccflags-y += -I$(PWD)
 
 obj-$(CONFIG_LTTNG) += lttng-probe-sched.o
 obj-$(CONFIG_LTTNG) += lttng-probe-irq.o
-- 
2.0.2

