From 9553ebb7905a22b2f8a51b8207ea7bb484f07060 Mon Sep 17 00:00:00 2001
From: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
Date: Fri, 18 Mar 2016 17:32:21 -0400
Subject: [PATCH 879/933] Bump minor tracer ABI version

Signed-off-by: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
---
 drivers/staging/lttng/lttng-abi.h | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/drivers/staging/lttng/lttng-abi.h b/drivers/staging/lttng/lttng-abi.h
index e53ff27..e23393f 100644
--- a/drivers/staging/lttng/lttng-abi.h
+++ b/drivers/staging/lttng/lttng-abi.h
@@ -30,7 +30,7 @@
  * should be increased when an incompatible ABI change is done.
  */
 #define LTTNG_MODULES_ABI_MAJOR_VERSION		2
-#define LTTNG_MODULES_ABI_MINOR_VERSION		0
+#define LTTNG_MODULES_ABI_MINOR_VERSION		1
 
 #define LTTNG_KERNEL_SYM_NAME_LEN	256
 
-- 
2.0.2

