From ec5975c4fe26135b054ad19be13281da614f53e1 Mon Sep 17 00:00:00 2001
From: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
Date: Thu, 25 Aug 2016 17:39:47 -0400
Subject: [PATCH 922/933] Fix: timer wrapper: support kernels >= 4.8

Signed-off-by: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
---
 drivers/staging/lttng/wrapper/timer.h | 7 ++++---
 1 file changed, 4 insertions(+), 3 deletions(-)

diff --git a/drivers/staging/lttng/wrapper/timer.h b/drivers/staging/lttng/wrapper/timer.h
index 441a017..c1c0c95 100644
--- a/drivers/staging/lttng/wrapper/timer.h
+++ b/drivers/staging/lttng/wrapper/timer.h
@@ -28,7 +28,8 @@
 #include <lttng-kernel-version.h>
 
 
-#if (LTTNG_RT_VERSION_CODE >= LTTNG_RT_KERNEL_VERSION(4,6,4,8))
+#if (LTTNG_RT_VERSION_CODE >= LTTNG_RT_KERNEL_VERSION(4,6,4,8) \
+	|| LINUX_VERSION_CODE >= KERNEL_VERSION(4,8,0))
 
 #define lttng_init_timer_pinned(timer)					\
 	init_timer_pinned(timer)
@@ -39,7 +40,7 @@ static inline int lttng_mod_timer_pinned(struct timer_list *timer,
 	return mod_timer(timer, expires);
 }
 
-#else /* #if (LTTNG_RT_VERSION_CODE >= LTTNG_RT_KERNEL_VERSION(4,6,4,8)) */
+#else
 
 #define lttng_init_timer_pinned(timer)					\
 	init_timer(timer)
@@ -50,6 +51,6 @@ static inline int lttng_mod_timer_pinned(struct timer_list *timer,
 	return mod_timer_pinned(timer, expires);
 }
 
-#endif /* #else #if (LTTNG_RT_VERSION_CODE >= LTTNG_RT_KERNEL_VERSION(4,6,4,8)) */
+#endif
 
 #endif /* _LTTNG_WRAPPER_TIMER_H */
-- 
2.0.2

