From 863201d03e81fca4cbf9e3a414450a8314783efa Mon Sep 17 00:00:00 2001
From: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
Date: Tue, 30 Aug 2016 17:02:09 -0400
Subject: [PATCH 924/933] Fix: napi instrumentation for 4.8 kernels

Signed-off-by: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
---
 .../lttng/instrumentation/events/lttng-module/napi.h | 20 ++++++++++++++++++++
 1 file changed, 20 insertions(+)

diff --git a/drivers/staging/lttng/instrumentation/events/lttng-module/napi.h b/drivers/staging/lttng/instrumentation/events/lttng-module/napi.h
index 1126dd4..7b5ecfc 100644
--- a/drivers/staging/lttng/instrumentation/events/lttng-module/napi.h
+++ b/drivers/staging/lttng/instrumentation/events/lttng-module/napi.h
@@ -10,6 +10,24 @@
 
 #define NO_DEV "(no_device)"
 
+#if (LINUX_VERSION_CODE >= KERNEL_VERSION(4,8,0))
+
+LTTNG_TRACEPOINT_EVENT(napi_poll,
+
+	TP_PROTO(struct napi_struct *napi, int work, int budget),
+
+	TP_ARGS(napi, work, budget),
+
+	TP_FIELDS(
+		ctf_integer_hex(struct napi_struct *, napi, napi)
+		ctf_string(dev_name, napi->dev ? napi->dev->name : NO_DEV)
+		ctf_integer(int, work, work)
+		ctf_integer(int, budget, budget)
+	)
+)
+
+#else
+
 LTTNG_TRACEPOINT_EVENT(napi_poll,
 
 	TP_PROTO(struct napi_struct *napi),
@@ -22,6 +40,8 @@ LTTNG_TRACEPOINT_EVENT(napi_poll,
 	)
 )
 
+#endif
+
 #undef NO_DEV
 
 #endif /* LTTNG_TRACE_NAPI_H */
-- 
2.0.2

