From 658bda5af784ffcb6ade6302ff0af8be871bc334 Mon Sep 17 00:00:00 2001
From: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
Date: Thu, 23 Jun 2011 17:06:11 -0400
Subject: [PATCH 188/625] Add wait quiescent

Signed-off-by: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
Signed-off-by: Jianchuan Wang <jianchuan.wang@windriver.com>
---
 drivers/staging/lttng2/ltt-debugfs-abi.c | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/drivers/staging/lttng2/ltt-debugfs-abi.c b/drivers/staging/lttng2/ltt-debugfs-abi.c
index c6d9f60..6327d09 100644
--- a/drivers/staging/lttng2/ltt-debugfs-abi.c
+++ b/drivers/staging/lttng2/ltt-debugfs-abi.c
@@ -212,6 +212,9 @@ long lttng_ioctl(struct file *file, unsigned int cmd, unsigned long arg)
 				(struct lttng_kernel_tracer_version __user *) arg);
 	case LTTNG_KERNEL_TRACEPOINT_LIST:
 		return lttng_abi_tracepoint_list();
+	case LTTNG_KERNEL_WAIT_QUIESCENT:
+		synchronize_trace();
+		return 0;
 	default:
 		return -ENOIOCTLCMD;
 	}
-- 
2.0.2

