From 7a79127b9152f060fd8ae1da8b13d3fe8d28e43f Mon Sep 17 00:00:00 2001
From: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
Date: Fri, 29 Jul 2011 22:21:07 -0400
Subject: [PATCH 245/625] Add info about kretprobes

Signed-off-by: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
Signed-off-by: Jianchuan Wang <jianchuan.wang@windriver.com>
---
 drivers/staging/lttng2/README | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/drivers/staging/lttng2/README b/drivers/staging/lttng2/README
index 1a7cea1..3bd0d15 100644
--- a/drivers/staging/lttng2/README
+++ b/drivers/staging/lttng2/README
@@ -9,7 +9,7 @@ need for additional patches. Other features:
 - Produces CTF (Common Trace Format) natively,
   (http://www.efficios.com/ctf)
 - Tracepoints, Function tracer, CPU Performance Monitoring Unit (PMU)
-  counters and kprobes support,
+  counters, kprobes, and kretprobes support,
 - Integrated interface for both kernel and userspace tracing,
 - Have the ability to attach "context" information to events in the
   trace (e.g. any PMU counter, pid, ppid, tid, comm name, etc).
-- 
2.0.2

