From 7996cbb2b8907e060bf6d0cb5a63d757ba2bbbae Mon Sep 17 00:00:00 2001
From: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
Date: Thu, 21 Jul 2011 15:00:26 -0400
Subject: [PATCH 231/625] Build fix: check if have syscall tracepoint config

Signed-off-by: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
Signed-off-by: Jianchuan Wang <jianchuan.wang@windriver.com>
---
 drivers/staging/lttng2/probes/Makefile | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/drivers/staging/lttng2/probes/Makefile b/drivers/staging/lttng2/probes/Makefile
index 1b1e51f..8e6994a 100644
--- a/drivers/staging/lttng2/probes/Makefile
+++ b/drivers/staging/lttng2/probes/Makefile
@@ -12,7 +12,10 @@ obj-m += lttng-probe-lttng.o
 
 obj-m += lttng-probe-sched.o
 obj-m += lttng-probe-irq.o
+
+ifneq ($(CONFIG_HAVE_SYSCALL_TRACEPOINTS),)
 obj-m += lttng-probe-syscalls.o
+endif
 
 ifneq ($(CONFIG_KVM),)
 obj-m += lttng-probe-kvm.o
-- 
2.0.2

