From 09aa67e228cd68cd07baa31f4515110279eac774 Mon Sep 17 00:00:00 2001
From: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
Date: Tue, 16 Aug 2011 20:24:52 -0400
Subject: [PATCH 255/540] ChangeLog for v2.0-pre8

Signed-off-by: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
---
 drivers/staging/lttng2/ChangeLog | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/drivers/staging/lttng2/ChangeLog b/drivers/staging/lttng2/ChangeLog
index 26c197a..e37dad7 100644
--- a/drivers/staging/lttng2/ChangeLog
+++ b/drivers/staging/lttng2/ChangeLog
@@ -1,3 +1,6 @@
+2011-08-16 LTTng modules 2.0-pre8
+	* Fix metadata buffer wait/wakeup
+
 2011-08-12 LTTng modules 2.0-pre7
 	* Add missing module refcount to lttng_tracepoint_list_fops
 	* Add license info about lib/ringbuffer/ring_buffer_mmap.c being GPLv2
-- 
1.8.4.93.g57e4c17

