From f496ccc29b7feee83701156fa290d8bf2d6faed8 Mon Sep 17 00:00:00 2001
From: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
Date: Mon, 23 May 2011 23:50:38 -0400
Subject: [PATCH 128/540] Fix ftrace.h wrapper

Signed-off-by: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
---
 drivers/staging/lttng2/wrapper/ftrace.h | 5 ++---
 1 file changed, 2 insertions(+), 3 deletions(-)

diff --git a/drivers/staging/lttng2/wrapper/ftrace.h b/drivers/staging/lttng2/wrapper/ftrace.h
index da7c394..9cf50d2 100644
--- a/drivers/staging/lttng2/wrapper/ftrace.h
+++ b/drivers/staging/lttng2/wrapper/ftrace.h
@@ -53,15 +53,14 @@ void wrapper_unregister_ftrace_function_probe(char *glob,
 
 static inline
 int wrapper_register_ftrace_function_probe(char *glob,
-		struct ftrace_probe_ops *ops, void *data);
-
+		struct ftrace_probe_ops *ops, void *data)
 {
 	return unregister_ftrace_function_probe();
 }
 
 static inline
 void wrapper_unregister_ftrace_function_probe(char *glob,
-		struct ftrace_probe_ops *ops, void *data);
+		struct ftrace_probe_ops *ops, void *data)
 {
 	return unregister_ftrace_function_probe();
 }
-- 
1.8.4.93.g57e4c17

