From b52c9b0df3f10e4a401a32330375c83b0125301f Mon Sep 17 00:00:00 2001
From: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
Date: Wed, 18 May 2011 18:50:26 -0400
Subject: [PATCH 083/540] Fix bitfield.h include directory

Signed-off-by: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
---
 drivers/staging/lttng2/lib/bitfield.h | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/drivers/staging/lttng2/lib/bitfield.h b/drivers/staging/lttng2/lib/bitfield.h
index a509e35..861e6dc 100644
--- a/drivers/staging/lttng2/lib/bitfield.h
+++ b/drivers/staging/lttng2/lib/bitfield.h
@@ -19,7 +19,7 @@
  * all copies or substantial portions of the Software.
  */
 
-#include "ltt-endian.h"
+#include "../ltt-endian.h"
 
 #ifndef CHAR_BIT
 #define CHAR_BIT 8
-- 
1.8.4.93.g57e4c17

