From 885416d23b7e3e1678d82eb6c02cf70c4e01bbcf Mon Sep 17 00:00:00 2001
From: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
Date: Mon, 30 Jan 2012 09:19:23 -0500
Subject: [PATCH 359/540] Fix comment

Signed-off-by: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
---
 drivers/staging/lttng2/wrapper/splice.h | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/drivers/staging/lttng2/wrapper/splice.h b/drivers/staging/lttng2/wrapper/splice.h
index ee08c6c..ea3bfac 100644
--- a/drivers/staging/lttng2/wrapper/splice.h
+++ b/drivers/staging/lttng2/wrapper/splice.h
@@ -4,7 +4,7 @@
 /*
  * Copyright (C) 2011 Mathieu Desnoyers (mathieu.desnoyers@efficios.com)
  *
- * wrapper around vmalloc_sync_all. Using KALLSYMS to get its address when
+ * wrapper around splice_to_pipe. Using KALLSYMS to get its address when
  * available, else we need to have a kernel that exports this function to GPL
  * modules.
  *
-- 
1.8.4.93.g57e4c17

