From 7082f134a58ca3629a590b7e57e0c60405051c00 Mon Sep 17 00:00:00 2001
From: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
Date: Tue, 10 May 2011 23:12:46 -0400
Subject: [PATCH 059/540] Fix const warning

Signed-off-by: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
---
 drivers/staging/lttng2/ltt-events.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/drivers/staging/lttng2/ltt-events.c b/drivers/staging/lttng2/ltt-events.c
index 184f2da..93638cd 100644
--- a/drivers/staging/lttng2/ltt-events.c
+++ b/drivers/staging/lttng2/ltt-events.c
@@ -96,7 +96,7 @@ end:
 	return ret;
 }
 
-static struct ltt_transport *ltt_transport_find(char *name)
+static struct ltt_transport *ltt_transport_find(const char *name)
 {
 	struct ltt_transport *transport;
 
-- 
1.8.4.93.g57e4c17

