From 92064e0138b3849b0ebe196564008df254c58ef2 Mon Sep 17 00:00:00 2001
From: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
Date: Tue, 10 Apr 2012 20:36:17 -0400
Subject: [PATCH 401/540] README: Document that CONFIG_MODULES is required

Signed-off-by: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
---
 drivers/staging/lttng2/README | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/drivers/staging/lttng2/README b/drivers/staging/lttng2/README
index c3568f8..89b4a7a 100644
--- a/drivers/staging/lttng2/README
+++ b/drivers/staging/lttng2/README
@@ -51,6 +51,8 @@ kernel version.
 
 * Kernel config options required
 
+CONFIG_MODULES: required
+  * Kernel modules support.
 CONFIG_KALLSYMS: required
   * See wrapper/ files. This is necessary until the few required missing
     symbols are exported to GPL modules from mainline.
-- 
1.8.4.93.g57e4c17

