From 97b4a817777af498cb2827b6c091e480d2481084 Mon Sep 17 00:00:00 2001
From: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
Date: Thu, 16 Feb 2012 11:39:57 -0500
Subject: [PATCH 381/540] Update ChangeLog to 2.0-pre15

Signed-off-by: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
---
 drivers/staging/lttng2/ChangeLog | 6 ++++++
 1 file changed, 6 insertions(+)

diff --git a/drivers/staging/lttng2/ChangeLog b/drivers/staging/lttng2/ChangeLog
index 463aa1c..9956098 100644
--- a/drivers/staging/lttng2/ChangeLog
+++ b/drivers/staging/lttng2/ChangeLog
@@ -1,3 +1,9 @@
+2012-02-16 LTTng modules 2.0-pre15
+	* Add timer instrumentation
+	* fix: need to undef mainline define
+	* fix: Include signal.h instead of irq.h for prototype match check
+	* Add signal instrumentation
+
 2012-02-16 LTTng modules 2.0-pre14
 	* syscall tracing: sys_getcpu
 	* Add sys_clone x86 instrumentation
-- 
1.8.4.93.g57e4c17

