From d77100442e982e6418a115b6004b092014dad1f8 Mon Sep 17 00:00:00 2001
From: Bruce Ashfield <bruce.ashfield@windriver.com>
Date: Mon, 2 Apr 2012 11:36:07 -0400
Subject: [PATCH] unionfs: v3.4 build fixups

Signed-off-by: Bruce Ashfield <bruce.ashfield@windriver.com>
---
 fs/unionfs/union.h |   22 ++++++++++++++++++++--
 1 files changed, 20 insertions(+), 2 deletions(-)

diff --git a/fs/unionfs/union.h b/fs/unionfs/union.h
index 8e7fcfb..2d80523 100644
--- a/fs/unionfs/union.h
+++ b/fs/unionfs/union.h
@@ -48,8 +48,6 @@
 #include <linux/splice.h>
 #include <linux/sched.h>
 
-#include <asm/system.h>
-
 #include <linux/union_fs.h>
 
 /* the file system name */
@@ -678,4 +676,24 @@ extern void __show_inode_counts(const struct inode *inode,
 
 #endif /* not CONFIG_UNION_FS_DEBUG */
 
+static inline void __FD_SET(unsigned long __fd, __kernel_fd_set *__fdsetp)
+{
+       __set_bit(__fd, __fdsetp->fds_bits);
+}
+
+static inline void __FD_CLR(unsigned long __fd, __kernel_fd_set *__fdsetp)
+{
+       __clear_bit(__fd, __fdsetp->fds_bits);
+}
+
+static inline int __FD_ISSET(unsigned long __fd, const __kernel_fd_set *__fdsetp)
+{
+       return test_bit(__fd, __fdsetp->fds_bits);
+}
+
+static inline void __FD_ZERO(__kernel_fd_set *__fdsetp)
+{
+       memset(__fdsetp->fds_bits, 0, sizeof __fdsetp->fds_bits);
+}
+
 #endif	/* not _UNION_H_ */
-- 
1.7.5.4

