From 2fe432357baadf7254562bb12573fec1305191f7 Mon Sep 17 00:00:00 2001
From: Amy Fong <amy.fong@windriver.com>
Date: Tue, 19 Jul 2011 19:02:00 -0400
Subject: [PATCH 02/17] sysfs: Remove double free sysfs_get_sb

commit ba514a57f5c38d9d79ea15e75059e07f49238726 upstream.

Signed-off-by: Eric W. Biederman <ebiederm@aristanetworks.com>
Signed-off-by: Greg Kroah-Hartman <gregkh@suse.de>
---
 fs/sysfs/mount.c |    1 -
 1 files changed, 0 insertions(+), 1 deletions(-)

diff --git a/fs/sysfs/mount.c b/fs/sysfs/mount.c
index db0b1f2..50e4fb6 100644
--- a/fs/sysfs/mount.c
+++ b/fs/sysfs/mount.c
@@ -104,7 +104,6 @@ static int sysfs_get_sb(struct file_system_type *fs_type,
 	if (IS_ERR(sb) || sb->s_fs_info != info)
 		kfree(info);
 	if (IS_ERR(sb)) {
-		kfree(info);
 		error = PTR_ERR(sb);
 		goto out;
 	}
-- 
1.7.0.4

