From 63b315b3bca843b2d0241b9a9c7d13fff226ab2e Mon Sep 17 00:00:00 2001
From: Liu Changhui <changhui.liu@windriver.com>
Date: Fri, 29 Jan 2010 13:27:49 +0800
Subject: [PATCH] Platform: Add bootinfo fixup for rmi xls

RMI bootloader need boot_mem_map type parameter is 32 bit even though kernel
is 64 bit.

source: from RMI SDK1.7

Signed-off-by: Liu Changhui <changhui.liu@windriver.com>
---
 include/asm-mips/bootinfo.h |   14 ++++++++++++++
 1 files changed, 14 insertions(+), 0 deletions(-)

diff --git a/include/asm-mips/bootinfo.h b/include/asm-mips/bootinfo.h
index 5d410bb..a9fa8e2 100644
--- a/include/asm-mips/bootinfo.h
+++ b/include/asm-mips/bootinfo.h
@@ -78,7 +78,15 @@ struct boot_mem_map {
 	struct boot_mem_map_entry {
 		phys_t addr;	/* start of memory segment */
 		phys_t size;	/* size of memory segment */
+/*
+ * kernel get boot_mem_map from bootloader, so data structure
+ * need to consistent with bootloader.
+ */
+#ifdef CONFIG_RMI_PHOENIX
+		uint32_t type;		/* type of memory segment */
+#else
 		long type;		/* type of memory segment */
+#endif
 	} map[BOOT_MEM_MAP_MAX];
 };
 
@@ -109,4 +117,10 @@ extern void plat_kexec_setup(void);
  */
 extern void plat_mem_setup(void);
 
+#define MAX_EXCLUDE 16
+struct boot_mem_map_exclude_region {
+	uint64_t start;
+	uint64_t end;
+};
+
 #endif /* _ASM_BOOTINFO_H */
-- 
1.6.0.4

