From a99afaed34bad9497edbb181eb424414f50d09c6 Mon Sep 17 00:00:00 2001
From: ywang5 <yalei.wang@windriver.com>
Date: Fri, 19 Apr 2013 12:45:09 +0800
Subject: [PATCH 3/3] bsp: Add tpm and mmc device for adi-alberta board

For supporting the adi-alberta board, customize the phyid, the
mmc init process and the i2c tpm device SLB9645. Codes are ported
from the kernel provided by ADI.

signed-off-by: Yalei Wang <yalei.wang@windriver.com>
---
 arch/arm/mach-omap2/board-am335xevm.c |   21 ++++++++++++---------
 arch/arm/mach-omap2/devices.c         |    4 ++--
 2 files changed, 14 insertions(+), 11 deletions(-)

diff --git a/arch/arm/mach-omap2/board-am335xevm.c b/arch/arm/mach-omap2/board-am335xevm.c
index e941130..1499845 100644
--- a/arch/arm/mach-omap2/board-am335xevm.c
+++ b/arch/arm/mach-omap2/board-am335xevm.c
@@ -1567,13 +1567,7 @@ static void lis331dlh_init(int evm_id, int profile)
 
 static struct i2c_board_info am335x_i2c1_boardinfo[] = {
 	{
-		I2C_BOARD_INFO("tlv320aic3x", 0x1b),
-	},
-	{
-		I2C_BOARD_INFO("tsl2550", 0x39),
-	},
-	{
-		I2C_BOARD_INFO("tmp275", 0x48),
+		I2C_BOARD_INFO("slb9635tt", 0x20),
 	},
 };
 
@@ -2380,11 +2374,12 @@ static struct omap_musb_board_data musb_board_data = {
 	 * mode[4:7] = USB1PORT's mode
 	 * AM335X beta EVM has USB0 in OTG mode and USB1 in host mode.
 	 */
-	.mode           = (MUSB_HOST << 4) | MUSB_OTG,
+	.mode           = (MUSB_HOST << 4) | MUSB_HOST,
 	.power		= 500,
 	.instances	= 1,
 };
 
+#if 0 /* fbs */
 static int cpld_reg_probe(struct i2c_client *client,
 	    const struct i2c_device_id *id)
 {
@@ -2416,16 +2411,20 @@ static void evm_init_cpld(void)
 {
 	i2c_add_driver(&cpld_reg_driver);
 }
+#endif /* fbs */
 
 static void __init am335x_evm_i2c_init(void)
 {
 	/* Initially assume General Purpose EVM Config */
 	am335x_evm_id = GEN_PURP_EVM;
 
-	evm_init_cpld();
+/*	evm_init_cpld(); fbs*/
 
 	omap_register_i2c_bus(1, 100, am335x_i2c0_boardinfo,
 				ARRAY_SIZE(am335x_i2c0_boardinfo));
+
+	omap_register_i2c_bus(2, 100, am335x_i2c1_boardinfo,
+				ARRAY_SIZE(am335x_i2c1_boardinfo));
 }
 
 static struct resource am335x_rtc_resources[] = {
@@ -2583,6 +2582,10 @@ static void __init am335x_evm_init(void)
 	am335x_rtc_init();
 	clkout2_enable();
 	am335x_evm_i2c_init();
+
+	mmc0_init (0,0);
+	am33xx_cpsw_init(AM33XX_CPSW_MODE_RGMII, NULL, NULL);
+
 	omap_sdrc_init(NULL, NULL);
 	usb_musb_init(&musb_board_data);
 	omap_board_config = am335x_evm_config;
diff --git a/arch/arm/mach-omap2/devices.c b/arch/arm/mach-omap2/devices.c
index fe95391..a605603 100644
--- a/arch/arm/mach-omap2/devices.c
+++ b/arch/arm/mach-omap2/devices.c
@@ -1485,13 +1485,13 @@ static struct cpsw_slave_data am33xx_cpsw_slaves[] = {
 	{
 		.slave_reg_ofs  = 0x208,
 		.sliver_reg_ofs = 0xd80,
-		.phy_id		= "davinci_mdio-0:00",
+		.phy_id		= "davinci_mdio-0:02",
 		.dual_emac_reserved_vlan = CPSW_PORT_VLAN_SLAVE_0,
 	},
 	{
 		.slave_reg_ofs  = 0x308,
 		.sliver_reg_ofs = 0xdc0,
-		.phy_id		= "davinci_mdio-0:01",
+		.phy_id		= "davinci_mdio-0:03",
 		.dual_emac_reserved_vlan = CPSW_PORT_VLAN_SLAVE_1,
 	},
 };
-- 
1.7.5.4

