From 28d4cc51b36636a8792d32c8e2bfd2b3d0d88d5e Mon Sep 17 00:00:00 2001
From: Allan Stephens <allan.stephens@windriver.com>
Date: Wed, 11 May 2011 11:37:16 -0400
Subject: [PATCH 6/6] tipc: Update version number to TIPC 1.7.7

Signed-off-by: Allan Stephens <allan.stephens@windriver.com>
---
 net/tipc/tipc_core.h |    4 ++--
 1 files changed, 2 insertions(+), 2 deletions(-)

diff --git a/net/tipc/tipc_core.h b/net/tipc/tipc_core.h
index fa08239..4a9b227 100644
--- a/net/tipc/tipc_core.h
+++ b/net/tipc/tipc_core.h
@@ -2,7 +2,7 @@
  * net/tipc/tipc_core.h: Include file for TIPC global declarations
  *
  * Copyright (c) 2005-2006, Ericsson AB
- * Copyright (c) 2005-2008, 2010, Wind River Systems
+ * Copyright (c) 2005-2008, 2010-2011, Wind River Systems
  * All rights reserved.
  *
  * Redistribution and use in source and binary forms, with or without
@@ -61,7 +61,7 @@
 #include <linux/vmalloc.h>
 
 
-#define TIPC_MOD_VER "1.7.7-rc4"
+#define TIPC_MOD_VER "1.7.7"
 
 /*
  * Spinlock wrappers (lets TIPC common files run unchanged on other OS's)
-- 
1.7.0.4

