From 24eb09f8f07533cea013aac161df526cf1aee7bc Mon Sep 17 00:00:00 2001
From: Yue Tao <yue.tao@windriver.com>
Date: Sun, 3 Mar 2013 22:15:07 -0800
Subject: [PATCH] Update tipc version from TIPC 1.7.6 to 1.7.7 (continue)

---
 net/tipc/tipc_socket.c |    5 +++++
 1 files changed, 5 insertions(+), 0 deletions(-)

diff --git a/net/tipc/tipc_socket.c b/net/tipc/tipc_socket.c
index 7c37f0b..442fc9f 100644
--- a/net/tipc/tipc_socket.c
+++ b/net/tipc/tipc_socket.c
@@ -1341,10 +1341,15 @@ static u32 dispatch(struct tipc_port *tport, struct sk_buff *buf)
 	if (!sock_owned_by_user(sk)) {
 		res = filter_rcv(sk, buf);
 	} else {
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(2, 6, 34)
 		if (sk_add_backlog(sk, buf))
 			res = TIPC_ERR_OVERLOAD;
 		else
 			res = TIPC_OK;
+#else
+		sk_add_backlog(sk, buf);
+		res = TIPC_OK;
+#endif
 	}
 	bh_unlock_sock(sk);
 
-- 
1.7.0

