From 83a9cb884051b7b8fd0747b54c99da345f638037 Mon Sep 17 00:00:00 2001
From: Jack Tan <jack.tan@windriver.com>
Date: Fri, 19 Dec 2008 17:19:06 +0800
Subject: [PATCH] kgdbos support

KGDBoS support.

Signed-off-by: Jack Tan <jack.tan@windriver.com>
---
 drivers/serial/8250_kgdb.c |   11 +++++++++++
 1 files changed, 11 insertions(+), 0 deletions(-)

diff --git a/drivers/serial/8250_kgdb.c b/drivers/serial/8250_kgdb.c
index 84d8d33..f11fe89 100644
--- a/drivers/serial/8250_kgdb.c
+++ b/drivers/serial/8250_kgdb.c
@@ -74,12 +74,20 @@ int __weak kgdb8250_early_debug_ready(void)
 
 static inline unsigned int kgdb8250_ioread(u8 mask)
 {
+#ifdef CONFIG_RMI_PHOENIX
+	return __raw_readl(kgdb8250_addr + (mask << kgdb8250_port.regshift));
+#else
 	return ioread8(kgdb8250_addr + (mask << kgdb8250_port.regshift));
+#endif
 }
 
 static inline void kgdb8250_iowrite(u8 val, u8 mask)
 {
+#ifdef CONFIG_RMI_PHOENIX
+	return __raw_writel(val, kgdb8250_addr + (mask << kgdb8250_port.regshift));
+#else
 	iowrite8(val, kgdb8250_addr + (mask << kgdb8250_port.regshift));
+#endif
 }
 
 /*
@@ -330,6 +338,9 @@ static int kgdb8250_early_init(void)
 {
 	/* Internal driver setup. */
 	switch (kgdb8250_port.iotype) {
+#ifdef CONFIG_RMI_PHOENIX
+	case UPIO_AU:
+#endif
 	case UPIO_MEM:
 		kgdb8250_needs_request_mem_region = 0;
 		if (kgdb8250_port.mapbase)
-- 
1.6.0.3

