From 9cab12e3a4c3f9950c29e3a6e2d4f9916d72841b Mon Sep 17 00:00:00 2001
From: Paul Gortmaker <paul.gortmaker@windriver.com>
Date: Sun, 21 Jun 2009 17:28:43 -0400
Subject: [PATCH 39/49] rmi: properly initialize SMP boot spinlock

Signed-off-by: Paul Gortmaker <paul.gortmaker@windriver.com>
---
 arch/mips/rmi/phoenix/smp.c |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/arch/mips/rmi/phoenix/smp.c b/arch/mips/rmi/phoenix/smp.c
index 87c4f7d..1b639dd 100644
--- a/arch/mips/rmi/phoenix/smp.c
+++ b/arch/mips/rmi/phoenix/smp.c
@@ -277,7 +277,7 @@ struct plat_smp_ops phoenix_smp_ops = {
 };
 
 
-static spinlock_t smp_boot_lock;
+static spinlock_t smp_boot_lock = SPIN_LOCK_UNLOCKED;
 extern unsigned long phnx_ebase;
 extern void prom_reconfigure_thr_resources(void);
 extern void ptr_smp_boot(unsigned long, unsigned long, unsigned long);
-- 
1.6.3.2

