From 5a86240f7893e9f0a97adb7ac0204e75b1087c06 Mon Sep 17 00:00:00 2001
From: Bruce Ashfield <bruce.ashfield@windriver.com>
Date: Fri, 6 Nov 2009 01:10:09 -0500
Subject: [PATCH] Raza bsp delivers ghost breakpoints to ptrace clients

CQID: WIND00184370

1/1 [
Author: Zhang Xiao
Email: xiao.zhang@windriver.com
Subject: Ptrace fixup for raza_xlr_atx BSP
Date: Wed, 4 Nov 2009 13:37:44 +0800

Add a cache flush in ptrace to avoid cache inconsistency.

Signed-off-by: Zhang Xiao <xiao.zhang@windriver.com>
]

Signed-off-by: Bruce Ashfield <bruce.ashfield@windriver.com>
---
 Ptrace-fixup-for-raza_xlr_atx-BSP.patch |   37 +++++++++++++++++++++++++++++++
 1 files changed, 37 insertions(+), 0 deletions(-)
 create mode 100644 Ptrace-fixup-for-raza_xlr_atx-BSP.patch

diff --git a/Ptrace-fixup-for-raza_xlr_atx-BSP.patch b/Ptrace-fixup-for-raza_xlr_atx-BSP.patch
new file mode 100644
index 0000000..35f13fd
--- /dev/null
+++ b/Ptrace-fixup-for-raza_xlr_atx-BSP.patch
@@ -0,0 +1,37 @@
+From 1c5d9f5908f681faf9a20701706ac3bed6bb07aa Mon Sep 17 00:00:00 2001
+From: Zhang Xiao <xiao.zhang@windriver.com>
+Date: Wed, 4 Nov 2009 13:37:44 +0800
+Subject: [PATCH] Ptrace fixup for raza_xlr_atx BSP
+
+Add a cache flush in ptrace to avoid cache inconsistency.
+
+Signed-off-by: Zhang Xiao <xiao.zhang@windriver.com>
+---
+ arch/mips/kernel/ptrace32.c |    4 ++++
+ 1 files changed, 4 insertions(+), 0 deletions(-)
+
+diff --git a/arch/mips/kernel/ptrace32.c b/arch/mips/kernel/ptrace32.c
+index 76818be..306fcf8 100644
+--- a/arch/mips/kernel/ptrace32.c
++++ b/arch/mips/kernel/ptrace32.c
+@@ -35,6 +35,7 @@
+ #include <asm/system.h>
+ #include <asm/uaccess.h>
+ #include <asm/bootinfo.h>
++#include <asm/cacheflush.h>
+ 
+ int ptrace_getregs(struct task_struct *child, __s64 __user *data);
+ int ptrace_setregs(struct task_struct *child, __s64 __user *data);
+@@ -241,6 +242,9 @@ asmlinkage int sys32_ptrace(int request, int pid, int addr, int data)
+ 	case PTRACE_POKETEXT: /* write the word at location addr. */
+ 	case PTRACE_POKEDATA:
+ 		ret = 0;
++#ifdef CONFIG_RMI_PHOENIX
++		__flush_cache_all();
++#endif
+ 		if (access_process_vm(child, addr, &data, sizeof(data), 1)
+ 		    == sizeof(data))
+ 			break;
+-- 
+1.6.5.2
+
-- 
1.6.5.2

