From 76a4aa1dfe3543249fac7bd8e3be5e9cc25a1150 Mon Sep 17 00:00:00 2001
From: Wei Yang <Wei.Yang@windriver.com>
Date: Tue, 22 Jan 2013 15:54:12 +0800
Subject: [PATCH] Do not do the full mismatch check on PPC.EMB.apuinfo

The root cause of issue is PPC.EMB.apuinfo is marked as PROGBITS by"-m64"
option. the section should be a note section. the attribute should not be marked.
Since nothing is going to use the section, so skip the check on PPC.EMB.apuinfo

Allow .PPC.EMB.apuinfo line sections in modules to
prevent modpost warnings:
WARNING: net/l2tp/l2tp_ppp.o (.PPC.EMB.apuinfo): unexpected non-allocatable section.
Did you forget to use "ax"/"aw"  in a .S file?
Note that for example <linux/init.h> contains
section definitions for use in .S files

Signed-off-by: Wei Yang <Wei.Yang@windriver.com>
---
 scripts/mod/modpost.c |    1 +
 1 file changed, 1 insertion(+)

diff --git a/scripts/mod/modpost.c b/scripts/mod/modpost.c
index 00acd10..8d4d375 100644
--- a/scripts/mod/modpost.c
+++ b/scripts/mod/modpost.c
@@ -820,6 +820,7 @@ static const char *section_white_list[] =
 	".debug*",
 	".zdebug*",		/* Compressed debug sections. */
 	".GCC-command-line",	/* mn10300 */
+	".PPC.EMB.apuinfo",
 	".mdebug*",        /* alpha, score, mips etc. */
 	".pdr",            /* alpha, score, mips etc. */
 	".stab*",
-- 
1.7.9.7

