From 18223e5e7ea32f31712debf8d6e3129cd74633d5 Mon Sep 17 00:00:00 2001
From: yhe <yongli.he@windriver.com>
Date: Thu, 15 Nov 2012 13:02:06 +0800
Subject: [PATCH 3/3] dpaa/KEXEC: hook up FM shutdown

Signed-off-by: Yongli He <yongli.he@windriver.com>
---
 .../dpa/NetCommSw/src/wrapper/lnxwrp_fm.c          |    4 ++++
 1 file changed, 4 insertions(+)

diff --git a/drivers/net/ethernet/freescale/dpa/NetCommSw/src/wrapper/lnxwrp_fm.c b/drivers/net/ethernet/freescale/dpa/NetCommSw/src/wrapper/lnxwrp_fm.c
index 39593df..883dae1 100644
--- a/drivers/net/ethernet/freescale/dpa/NetCommSw/src/wrapper/lnxwrp_fm.c
+++ b/drivers/net/ethernet/freescale/dpa/NetCommSw/src/wrapper/lnxwrp_fm.c
@@ -1013,7 +1013,11 @@ static struct platform_driver fm_driver = {
         .owner          = THIS_MODULE,
     },
     .probe          = fm_probe,
+#if defined(CONFIG_KEXEC)
+    .shutdown       = fm_remove,
+#endif
     .remove         = __devexit_p(fm_remove)
+
 };
 
 t_Handle LNXWRP_FM_Init(void)
-- 
1.7.9.7

