From c90fae33d9cccc65febde23839c5b245ef35818d Mon Sep 17 00:00:00 2001
From: Alison Wang <b18965@freescale.com>
Date: Mon, 13 Jun 2016 14:58:02 +0800
Subject: [PATCH 112/388] Armv8: aarch32: Add ITS for LS1012A AArch32 support

ITS files of LS1012A QDS/RDB/FRDM boards are added for AArch32 support.

Signed-off-by: Alison Wang <alison.wang@nxp.com>
[Original patch taken from SDK-V2.0-1703]
Signed-off-by: Yanjiang Jin <yanjiang.jin@windriver.com>
---
 kernel-ls1012a-frdm-aarch32.its | 53 +++++++++++++++++++++++++++++++++++++++++
 kernel-ls1012a-qds-aarch32.its  | 53 +++++++++++++++++++++++++++++++++++++++++
 kernel-ls1012a-rdb-aarch32.its  | 53 +++++++++++++++++++++++++++++++++++++++++
 3 files changed, 159 insertions(+)
 create mode 100644 kernel-ls1012a-frdm-aarch32.its
 create mode 100644 kernel-ls1012a-qds-aarch32.its
 create mode 100644 kernel-ls1012a-rdb-aarch32.its

diff --git a/kernel-ls1012a-frdm-aarch32.its b/kernel-ls1012a-frdm-aarch32.its
new file mode 100644
index 0000000..7b6c497
--- /dev/null
+++ b/kernel-ls1012a-frdm-aarch32.its
@@ -0,0 +1,53 @@
+/*
+ * Copyright (C) 2016, Freescale Semiconductor
+ *
+ * This file is licensed under the terms of the GNU General Public
+ * License version 2. This program is licensed "as is" without any
+ * warranty of any kind, whether express or implied.
+ */
+
+/dts-v1/;
+
+/ {
+	description = "Image file for the LS1012A Linux Kernel";
+	#address-cells = <1>;
+
+	images {
+		kernel@1 {
+			description = "ARM32 Linux kernel";
+			data = /incbin/("./arch/arm/boot/zImage");
+			type = "kernel";
+			arch = "arm";
+			os = "linux";
+			compression = "none";
+			load = <0x80008000>;
+			entry = <0x80008000>;
+		};
+		fdt@1 {
+			description = "Flattened Device Tree blob";
+			data = /incbin/("./fsl-ls1012a-frdm.dtb");
+			type = "flat_dt";
+			arch = "arm";
+			compression = "none";
+			load = <0x90000000>;
+		};
+		ramdisk@1 {
+			description = "LS1 Ramdisk";
+                        data = /incbin/("./fsl-image-core-ls1012ardb-32b.ext2.gz");
+			type = "ramdisk";
+			arch = "arm";
+			os = "linux";
+			compression = "none";
+		};
+	};
+
+	configurations {
+		default = "config@1";
+		config@1 {
+			description = "Boot Linux kernel";
+			kernel = "kernel@1";
+			fdt = "fdt@1";
+			ramdisk = "ramdisk@1";
+		};
+	};
+};
diff --git a/kernel-ls1012a-qds-aarch32.its b/kernel-ls1012a-qds-aarch32.its
new file mode 100644
index 0000000..40262d7
--- /dev/null
+++ b/kernel-ls1012a-qds-aarch32.its
@@ -0,0 +1,53 @@
+/*
+ * Copyright (C) 2016, Freescale Semiconductor
+ *
+ * This file is licensed under the terms of the GNU General Public
+ * License version 2. This program is licensed "as is" without any
+ * warranty of any kind, whether express or implied.
+ */
+
+/dts-v1/;
+
+/ {
+	description = "Image file for the LS1012A Linux Kernel";
+	#address-cells = <1>;
+
+	images {
+		kernel@1 {
+			description = "ARM32 Linux kernel";
+			data = /incbin/("./arch/arm/boot/zImage");
+			type = "kernel";
+			arch = "arm";
+			os = "linux";
+			compression = "none";
+			load = <0x80008000>;
+			entry = <0x80008000>;
+		};
+		fdt@1 {
+			description = "Flattened Device Tree blob";
+			data = /incbin/("./fsl-ls1012a-qds.dtb");
+			type = "flat_dt";
+			arch = "arm";
+			compression = "none";
+			load = <0x90000000>;
+		};
+		ramdisk@1 {
+			description = "LS1 Ramdisk";
+                        data = /incbin/("./fsl-image-core-ls1012ardb-32b.ext2.gz");
+			type = "ramdisk";
+			arch = "arm";
+			os = "linux";
+			compression = "none";
+		};
+	};
+
+	configurations {
+		default = "config@1";
+		config@1 {
+			description = "Boot Linux kernel";
+			kernel = "kernel@1";
+			fdt = "fdt@1";
+			ramdisk = "ramdisk@1";
+		};
+	};
+};
diff --git a/kernel-ls1012a-rdb-aarch32.its b/kernel-ls1012a-rdb-aarch32.its
new file mode 100644
index 0000000..4e847ab
--- /dev/null
+++ b/kernel-ls1012a-rdb-aarch32.its
@@ -0,0 +1,53 @@
+/*
+ * Copyright (C) 2016, Freescale Semiconductor
+ *
+ * This file is licensed under the terms of the GNU General Public
+ * License version 2. This program is licensed "as is" without any
+ * warranty of any kind, whether express or implied.
+ */
+
+/dts-v1/;
+
+/ {
+	description = "Image file for the LS1012A Linux Kernel";
+	#address-cells = <1>;
+
+	images {
+		kernel@1 {
+			description = "ARM32 Linux kernel";
+			data = /incbin/("./arch/arm/boot/zImage");
+			type = "kernel";
+			arch = "arm";
+			os = "linux";
+			compression = "none";
+			load = <0x80008000>;
+			entry = <0x80008000>;
+		};
+		fdt@1 {
+			description = "Flattened Device Tree blob";
+			data = /incbin/("./fsl-ls1012a-rdb.dtb");
+			type = "flat_dt";
+			arch = "arm";
+			compression = "none";
+			load = <0x90000000>;
+		};
+		ramdisk@1 {
+			description = "LS1 Ramdisk";
+                        data = /incbin/("./fsl-image-core-ls1012ardb-32b.ext2.gz");
+			type = "ramdisk";
+			arch = "arm";
+			os = "linux";
+			compression = "none";
+		};
+	};
+
+	configurations {
+		default = "config@1";
+		config@1 {
+			description = "Boot Linux kernel";
+			kernel = "kernel@1";
+			fdt = "fdt@1";
+			ramdisk = "ramdisk@1";
+		};
+	};
+};
-- 
2.9.3

