From 8f3f159f5492429601fcd6116a50c75a37080328 Mon Sep 17 00:00:00 2001
From: Zumeng Chen <zumeng.chen@windriver.com>
Date: Fri, 2 Jul 2010 18:25:53 +0800
Subject: [PATCH] kmemcheck: fix not exist Kconfig dependency

FUNCTION_TRACER does not exist in 3.0.3, so remove it.
And RODATA should not be checked by kmemcheck, so add
it into here.

Signed-off-by: Zumeng Chen <zumeng.chen@windriver.com>
---
 lib/Kconfig.kmemcheck |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/lib/Kconfig.kmemcheck b/lib/Kconfig.kmemcheck
index 603c81b..787bd26 100644
--- a/lib/Kconfig.kmemcheck
+++ b/lib/Kconfig.kmemcheck
@@ -7,7 +7,7 @@ menuconfig KMEMCHECK
 	depends on !X86_USE_3DNOW
 	depends on SLUB || SLAB
 	depends on !CC_OPTIMIZE_FOR_SIZE
-	depends on !FUNCTION_TRACER
+	depends on !FTRACE && !FTRACE_STARTUP_TEST && !DEBUG_RODATA
 	select FRAME_POINTER
 	select STACKTRACE
 	default n
-- 
1.7.0.4

