From c3a5420779a0d41e658581e8864f81fa19b399d3 Mon Sep 17 00:00:00 2001
From: Paul Gortmaker <paul.gortmaker@windriver.com>
Date: Wed, 8 Oct 2008 14:24:06 -0400
Subject: [PATCH] [PATCH] patch move-native-irq.patch

From 4e7c9b885df2689bef4e3522aa30bcc7349864a4 Mon Sep 17 00:00:00 2001
Subject: [PATCH] patch move-native-irq.patch
---
 kernel/irq/migration.c |   14 ++++++++++++--
 1 files changed, 12 insertions(+), 2 deletions(-)

diff --git a/kernel/irq/migration.c b/kernel/irq/migration.c
index 77b7acc..e61da29 100644
--- a/kernel/irq/migration.c
+++ b/kernel/irq/migration.c
@@ -61,6 +61,7 @@ void move_masked_irq(int irq)
 void move_native_irq(int irq)
 {
 	struct irq_desc *desc = irq_desc + irq;
+	int mask = 1;
 
 	if (likely(!(desc->status & IRQ_MOVE_PENDING)))
 		return;
@@ -68,8 +69,17 @@ void move_native_irq(int irq)
 	if (unlikely(desc->status & IRQ_DISABLED))
 		return;
 
-	desc->chip->mask(irq);
+	/*
+	 * If the irq is already in progress, it should be masked.
+	 * If we unmask it, we might cause an interrupt storm on RT.
+	 */
+	if (unlikely(desc->status & IRQ_INPROGRESS))
+		mask = 0;
+
+	if (mask)
+		desc->chip->mask(irq);
 	move_masked_irq(irq);
-	desc->chip->unmask(irq);
+	if (mask)
+		desc->chip->unmask(irq);
 }
 
-- 
1.5.5.1

