From 39395d50749091007410acd1a199d27122abf6a9 Mon Sep 17 00:00:00 2001
From: Paul Gortmaker <paul.gortmaker@windriver.com>
Date: Wed, 8 Oct 2008 14:40:13 -0400
Subject: [PATCH] [PATCH] patch preempt-irqs-arm.patch

From cfdf2ea57964e6624a651b638adbe7f9521bb7bf Mon Sep 17 00:00:00 2001
Subject: [PATCH] patch preempt-irqs-arm.patch
---
 arch/arm/common/time-acorn.c |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/arch/arm/common/time-acorn.c b/arch/arm/common/time-acorn.c
index df0983a..7b9ef19 100644
--- a/arch/arm/common/time-acorn.c
+++ b/arch/arm/common/time-acorn.c
@@ -75,7 +75,7 @@ ioc_timer_interrupt(int irq, void *dev_id)
 
 static struct irqaction ioc_timer_irq = {
 	.name		= "timer",
-	.flags		= IRQF_DISABLED,
+	.flags		= IRQF_DISABLED | IRQF_NODELAY,
 	.handler	= ioc_timer_interrupt
 };
 
-- 
1.5.5.1

