From 7a5a62ac2408cc58d199dd95a2b18a219f09186f Mon Sep 17 00:00:00 2001
From: Paul Gortmaker <paul.gortmaker@windriver.com>
Date: Wed, 8 Oct 2008 14:43:04 -0400
Subject: [PATCH] [PATCH] patch preempt-irqs-i386-idle-poll-loop-fix.patch

From 6953d02de17ca3450470e22a570178a1005f2e95 Mon Sep 17 00:00:00 2001
Subject: [PATCH] patch preempt-irqs-i386-idle-poll-loop-fix.patch
---
 arch/x86/kernel/process.c |    4 +++-
 1 files changed, 3 insertions(+), 1 deletions(-)

diff --git a/arch/x86/kernel/process.c b/arch/x86/kernel/process.c
index 876e918..36e133e 100644
--- a/arch/x86/kernel/process.c
+++ b/arch/x86/kernel/process.c
@@ -185,7 +185,9 @@ static void mwait_idle(void)
 static void poll_idle(void)
 {
 	local_irq_enable();
-	cpu_relax();
+	do {
+		cpu_relax();
+	} while (!need_resched());
 }
 
 /*
-- 
1.5.5.1

