From 0bdc599b69cbc24c12a767e66ff2bf4191391549 Mon Sep 17 00:00:00 2001
From: Paul Gortmaker <paul.gortmaker@windriver.com>
Date: Wed, 8 Oct 2008 14:43:05 -0400
Subject: [PATCH] [PATCH] patch schedule-tail-balance-disable-irqs.patch

From 9b12cb650497be2bff2adc0a42b6abee716b8db4 Mon Sep 17 00:00:00 2001
Subject: [PATCH] patch schedule-tail-balance-disable-irqs.patch
---
 kernel/sched_rt.c |    4 ++--
 1 files changed, 2 insertions(+), 2 deletions(-)

diff --git a/kernel/sched_rt.c b/kernel/sched_rt.c
index 1113157..3688bb3 100644
--- a/kernel/sched_rt.c
+++ b/kernel/sched_rt.c
@@ -1224,9 +1224,9 @@ static void post_schedule_rt(struct rq *rq)
 	 * first via finish_lock_switch and then reaquire it here.
 	 */
 	if (unlikely(rq->rt.overloaded)) {
-		spin_lock_irq(&rq->lock);
+		spin_lock(&rq->lock);
 		push_rt_tasks(rq);
-		spin_unlock_irq(&rq->lock);
+		spin_unlock(&rq->lock);
 	}
 }
 
-- 
1.5.5.1

