From 3657a43528d282f04b37ed39bcd6f920e31830c1 Mon Sep 17 00:00:00 2001
From: Paul Gortmaker <paul.gortmaker@windriver.com>
Date: Wed, 8 Oct 2008 14:50:27 -0400
Subject: [PATCH] [PATCH] patch preempt-realtime-loopback.patch

From 2e30777c277d0951fcfd24a3e784698a4969cbea Mon Sep 17 00:00:00 2001
Subject: [PATCH] patch preempt-realtime-loopback.patch
---
 drivers/net/loopback.c |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/drivers/net/loopback.c b/drivers/net/loopback.c
index dfe1184..c570619 100644
--- a/drivers/net/loopback.c
+++ b/drivers/net/loopback.c
@@ -93,7 +93,7 @@ static int loopback_xmit(struct sk_buff *skb, struct net_device *dev)
 	lb_stats->packets++;
 	put_cpu();
 
-	netif_rx(skb);
+	netif_rx_ni(skb);
 
 	return 0;
 }
-- 
1.5.5.1

