From 6cab4abb34b6a4163ee3a4f95fd5244a3c630d9d Mon Sep 17 00:00:00 2001
From: Yong Zhang <yong.zhang@windriver.com>
Date: Fri, 24 May 2013 08:49:44 +0800
Subject: [PATCH] MIPS: XLP: PCI: initialize msi-x vector register

Signed-off-by: Yong Zhang <yong.zhang@windriver.com>
---
 arch/mips/pci/pci-xlp.c |    3 +++
 1 files changed, 3 insertions(+), 0 deletions(-)

diff --git a/arch/mips/pci/pci-xlp.c b/arch/mips/pci/pci-xlp.c
index 4f79ee9..5be7968 100644
--- a/arch/mips/pci/pci-xlp.c
+++ b/arch/mips/pci/pci-xlp.c
@@ -403,6 +403,9 @@ static void xlp_msi_controller_init(u8 node, int fn)
 		(virt_to_phys(xlp_msix_addr_start(node, fn)) >> 8));
 	nlh_pci_w32r(node, fn, 0x250,
 		(virt_to_phys(xlp_msix_addr_start(node, fn) + XLP_MSIX_ADDR_SIZE) >> 8));
+	msi = nlh_pci_r32r(node, fn, 0x2c);
+	msi &= ~(0x7ff << 16);
+	nlh_pci_w32r(node, fn, 0x2c, msi | (XLP_MSIX_PER_SLOT - 1) << 16);
 }
 
 /*
-- 
1.7.0

