From a2803ec3a86c48535ee88aadb9072ff270d283d9 Mon Sep 17 00:00:00 2001
From: Yanjiang Jin <yanjiang.jin@windriver.com>
Date: Thu, 6 Sep 2012 15:56:01 +0800
Subject: [PATCH] nlm_xlp_64_be: use phys_cpu_present_map to init NAE

Use phys_cpu_present_map to initialize NAE.

The problem:
The SDK and WRL both enable all available CPUs by default,
but if user modifies the online CPU map in the DTS, for instance,

from
onlinemask = <0xffffffff>
to
onlinemask = <0x11111111>

the network performance goes bad(ping packets with huge latency).

~ # ping 192.168.1.2
PING 192.168.1.2 (192.168.1.2): 56 data bytes
64 bytes from 192.168.1.2: seq=3 ttl=128 time=0.833 ms
.....
---
 drivers/net/nae/xlp_nae.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/drivers/net/nae/xlp_nae.c b/drivers/net/nae/xlp_nae.c
index 7b01911..0b3c833 100644
--- a/drivers/net/nae/xlp_nae.c
+++ b/drivers/net/nae/xlp_nae.c
@@ -914,7 +914,7 @@ static void nlm_xlp_nae_init(void)
 	pr_info("debug = %d, frin_desc_thres = %d naecfg_hack = %d drop_uboot_pkt = %d\n",
 	debug, frin_desc_thres, naecfg_hack, drop_uboot_pkt);
 
-	if (initialize_nae(cpumask_to_uint32(&cpu_online_map), 0, 0, 0))
+	if (initialize_nae(cpumask_to_uint32(&phys_cpu_present_map), 0, 0, 0))
 		return;
 	p2p_desc_mem_init();
 
-- 
1.7.11

