From 8e63bf21e18b3901c49973169d8d4da6a102b743 Mon Sep 17 00:00:00 2001
From: Yong Zhang <yong.zhang@windriver.com>
Date: Mon, 1 Dec 2014 14:53:08 +0800
Subject: [PATCH] Revert "nlm_xlp_64_be: update delay/udelay/mdelay"

This reverts commit d99c721f3b18[nlm_xlp_64_be:
update delay/udelay/mdelay].
On XLP, the clock frequence can be 66MHz or 133MHz on different
boards, restricting it to 66MHz is wrong. So restore back
to use PIC for timer again to make the timer counting correct.
---
 arch/mips/lib/delay.c |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/arch/mips/lib/delay.c b/arch/mips/lib/delay.c
index e61970f..27ceba0 100644
--- a/arch/mips/lib/delay.c
+++ b/arch/mips/lib/delay.c
@@ -23,7 +23,7 @@
 #include <asm/compiler.h>
 #include <asm/war.h>
 
-#if defined CONFIG_NLM_XLP && defined CONFIG_XLP_REPLACE_R4K_TIMER
+#if defined CONFIG_NLM_XLP /* For XLP with PIC timer */
 #include <asm/netlogic/xlp_irq.h>
 #include <linux/clockchips.h>
 extern int read_current_timer(unsigned long *cyc);
-- 
1.7.0

