From b94c8a1f0ec002f7759f9fcf0d91e54557052f9f Mon Sep 17 00:00:00 2001
From: Yongli He <yongli.he@windriver.com>
Date: Wed, 3 Sep 2008 11:37:58 +0800
Subject: [PATCH] symlink increase nested limit

Signed-off-by: Yongli he   <yongli.he@windriver.com>
Signed-off-by: MacDonald, Joe <Joe.MacDonald@windriver.com>
---
 fs/Kconfig            |    8 ++++++++
 include/linux/namei.h |    3 ++-
 2 files changed, 10 insertions(+), 1 deletions(-)

diff --git a/fs/Kconfig b/fs/Kconfig
index 7724d55..025fc42 100644
--- a/fs/Kconfig
+++ b/fs/Kconfig
@@ -499,6 +499,14 @@ config OCFS2_DEBUG_FS
 
 endif # BLOCK
 
+config MAX_NESTED_LINKS
+	int "Symblic link nest limit"
+	default 5
+	help
+	  Specifies the nesting limit on symbolic links.  You may increase
+	  this value, but setting it too high could lead to kernel stack
+	  overflows.
+
 config DNOTIFY
 	bool "Dnotify support"
 	default y
diff --git a/include/linux/namei.h b/include/linux/namei.h
index 68f8c32..4ab2873 100644
--- a/include/linux/namei.h
+++ b/include/linux/namei.h
@@ -13,7 +13,8 @@ struct open_intent {
 	struct file *file;
 };
 
-enum { MAX_NESTED_LINKS = 8 };
+enum { MAX_NESTED_LINKS = CONFIG_MAX_NESTED_LINKS };
+
 
 struct nameidata {
 	struct path	path;
-- 
1.5.5.1

