From 24b6d0b2b3dfdc9e03bd9d1f9656b399841c74fa Mon Sep 17 00:00:00 2001
From: qingtao.cao@windriver.com <qingtao.cao@windriver.com>
Date: Tue, 4 Nov 2008 19:39:19 +0800
Subject: [PATCH] EDAC: E7501 MCH DVNP fixup

Setup the e7501 MCH DVNP(Device Not Present) register to make all
Function 1 devices visible to software, otherwise e7501 will appear
to have only Function 1 (D0F0, D2F0 etc).

Note, this fixup won't make D0F1 visible because PCI_HEADER_TYPE of
D0F0 has been read already, so we have to manually setup its
dev->multifunction.
---
 arch/x86/pci/fixup.c |   31 +++++++++++++++++++++++++++++++
 1 files changed, 31 insertions(+), 0 deletions(-)

diff --git a/arch/x86/pci/fixup.c b/arch/x86/pci/fixup.c
index 4bdaa59..8b1d2c8 100644
--- a/arch/x86/pci/fixup.c
+++ b/arch/x86/pci/fixup.c
@@ -511,3 +511,34 @@ DECLARE_PCI_FIXUP_HEADER(PCI_VENDOR_ID_AMD, 0x1201, fam10h_pci_cfg_space_size);
 DECLARE_PCI_FIXUP_HEADER(PCI_VENDOR_ID_AMD, 0x1202, fam10h_pci_cfg_space_size);
 DECLARE_PCI_FIXUP_HEADER(PCI_VENDOR_ID_AMD, 0x1203, fam10h_pci_cfg_space_size);
 DECLARE_PCI_FIXUP_HEADER(PCI_VENDOR_ID_AMD, 0x1204, fam10h_pci_cfg_space_size);
+
+static void __devinit e7501_MCH_DVNP_setup(struct pci_dev *dev)
+{
+	u16 val;
+
+	/*
+ 	 * setup the e7501 MCH DVNP(Device Not Present) register to make all
+	 * Function 1 devices visible to software, otherwise e7501 will appear
+	 * to have only Function 1 (D0F0, D2F0 etc)
+	 */
+	pci_read_config_word(dev, 0xE0, &val);
+	printk(KERN_INFO "E7501_host_controller_DVNP is 0x%x\n", val);
+
+	pci_write_config_word(dev, 0xE0, (val & 0xFFE2));
+	pci_read_config_word(dev, 0xE0, &val);
+	printk(KERN_INFO "E7501_host_controller_DVNP is 0x%x\n", val);
+	
+	/*
+ 	 * However, since when this fixup takes effect, PCI_HEAD_TYEP of 
+ 	 * MCH itself has already been read, so we have to change its
+ 	 * dev->multifunction to be greater than zero so that 
+ 	 * pci_scan_slot() would be willing to search Function 1 device
+ 	 * on it.
+ 	 *
+ 	 * On a side note, since Device 2/3/4 have not been scanned, so
+ 	 * setup DVNP alone would be enough to make their Function 1
+ 	 * device visible to software.
+ 	 */ 
+	dev->multifunction = 0x1;
+}
+DECLARE_PCI_FIXUP_HEADER(PCI_VENDOR_ID_INTEL, PCI_DEVICE_ID_INTEL_E7501_MCH, e7501_MCH_DVNP_setup);
-- 
1.6.0.90.g436ed

