From 1353e7e213a02d45f91efbde10a9b9dd95587d39 Mon Sep 17 00:00:00 2001
From: Benjamin Walsh <benjamin.walsh@windriver.com>
Date: Fri, 7 Nov 2008 14:21:09 -0500
Subject: [PATCH] EDAC: add common driver for Maple

Provide a module to export a common device index counter for all modules
on one Maple target.
---
 drivers/edac/Kconfig      |    6 ++++++
 drivers/edac/Makefile     |    1 +
 drivers/edac/maple_edac.c |   45 +++++++++++++++++++++++++++++++++++++++++++++
 drivers/edac/maple_edac.h |   25 +++++++++++++++++++++++++
 4 files changed, 77 insertions(+), 0 deletions(-)
 create mode 100644 drivers/edac/maple_edac.c
 create mode 100644 drivers/edac/maple_edac.h

diff --git a/drivers/edac/Kconfig b/drivers/edac/Kconfig
index c999684..451e24e 100644
--- a/drivers/edac/Kconfig
+++ b/drivers/edac/Kconfig
@@ -172,4 +172,10 @@ config EDAC_CELL
 	  Cell Broadband Engine internal memory controller
 	  on platform without a hypervisor
 
+config EDAC_MAPLE
+	tristate "Maple board"
+	depends on EDAC_MM_EDAC && PPC_MAPLE
+	help
+	  Global variables for Maple board
+
 endif # EDAC
diff --git a/drivers/edac/Makefile b/drivers/edac/Makefile
index 2ae3582..374f9df 100644
--- a/drivers/edac/Makefile
+++ b/drivers/edac/Makefile
@@ -34,4 +34,5 @@ obj-$(CONFIG_EDAC_MPC85XX)		+= mpc85xx_edac.o
 obj-$(CONFIG_EDAC_MV64X60)		+= mv64x60_edac.o
 obj-$(CONFIG_EDAC_CELL)			+= cell_edac.o
 obj-$(CONFIG_EDAC_TEST_DEVICE)		+= test_device_edac.o
+obj-$(CONFIG_EDAC_MAPLE)		+= maple_edac.o
 
diff --git a/drivers/edac/maple_edac.c b/drivers/edac/maple_edac.c
new file mode 100644
index 0000000..04bf138
--- /dev/null
+++ b/drivers/edac/maple_edac.c
@@ -0,0 +1,45 @@
+/*
+ * Global variables for Maple board
+ *
+ * Author: Benjamin Walsh <benjamin.walsh@windriver.com>
+ *         Cao Qingtao <qingtao.cao@windriver.com>
+ *         Hu Yongqi <yongqi.hu@windriver.com>
+ * 2008 (c) Wind River Systems, Inc.
+ *
+ * This file is licensed under the terms of the GNU General Public
+ * License version 2. This program is licensed "as is" without any
+ * warranty of any kind, whether express or implied.
+ *
+ */
+
+#include <linux/module.h>
+#include <linux/init.h>
+
+#define MAPLE_EDAC_REVISION	" Ver: 1.0.0 " __DATE__
+#define MAPLE_EDAC_MOD_STR	"maple_edac"
+
+int maple_edac_mc_idx;
+EXPORT_SYMBOL_GPL(maple_edac_mc_idx);
+
+int maple_edac_dev_idx;
+EXPORT_SYMBOL_GPL(maple_edac_dev_idx);
+
+static int __init maple_edac_init(void)
+{
+	maple_edac_mc_idx = 0;
+	maple_edac_dev_idx = 0;
+
+	return 0;
+}
+
+static void __exit maple_edac_exit(void)
+{
+}
+
+module_init(maple_edac_init);
+module_exit(maple_edac_exit);
+
+MODULE_LICENSE("GPL");
+MODULE_AUTHOR("Wind River (http://windriver.com) Benjamin Walsh\n");
+MODULE_DESCRIPTION("Maple EDAC module");
+
diff --git a/drivers/edac/maple_edac.h b/drivers/edac/maple_edac.h
new file mode 100644
index 0000000..54a6056
--- /dev/null
+++ b/drivers/edac/maple_edac.h
@@ -0,0 +1,25 @@
+/*
+ * Global variables for Maple board
+ *
+ * Author: Benjamin Walsh <benjamin.walsh@windriver.com>
+ *         Cao Qingtao <qingtao.cao@windriver.com>
+ *         Hu Yongqi <yongqi.hu@windriver.com>
+ * 2008 (c) Wind River Systems, Inc.
+ *
+ * This file is licensed under the terms of the GNU General Public
+ * License version 2. This program is licensed "as is" without any
+ * warranty of any kind, whether express or implied.
+ *
+ */
+
+#ifndef _MAPLE_EDAC_H_
+#define _MAPLE_EDAC_H_
+
+/*
+ * All EDAC modules for Maple target should share one same copy of
+ * EDAC device index.
+ */
+extern int maple_edac_mc_idx;
+extern int maple_edac_dev_idx;
+
+#endif /* _MAPLE_EDAC_H_*/
-- 
1.5.5.1

