From a674dd706c8b9d6d985a281379861a95ba530fcf Mon Sep 17 00:00:00 2001
From: Peter Swain <pswain@cavium.com>
Date: Wed, 30 Jul 2014 18:05:50 -0700
Subject: [PATCH 733/974] raid6/test: fix makerule syntax

The makerule syntax must be /bin/sh-compatible.
Replaced some csh-ish '>&/dev/null' with '>/dev/null 2>&1',
which repairs the
  /bin/sh: 1: Syntax error: Bad fd number
errors which came from a
  make -C lib/raid6/test

Even though '>&' and '&>' are accepted by bash & friends in interactive
mode, this behavior disappears when invoked with argv[0] "sh".
It's always wrong in makerules
[Original patch taken from Cavium SDK 3.1.2-568]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
---
 lib/raid6/test/Makefile | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/lib/raid6/test/Makefile b/lib/raid6/test/Makefile
index 087332d..7dc9027 100644
--- a/lib/raid6/test/Makefile
+++ b/lib/raid6/test/Makefile
@@ -25,11 +25,11 @@ endif
 ifeq ($(IS_X86),yes)
         OBJS   += mmx.o sse1.o sse2.o avx2.o recov_ssse3.o recov_avx2.o
         CFLAGS += $(shell echo "vpbroadcastb %xmm0, %ymm1" |	\
-                    gcc -c -x assembler - >&/dev/null &&	\
+                    gcc -c -x assembler - >/dev/null 2>&1 &&	\
                     rm ./-.o && echo -DCONFIG_AS_AVX2=1)
 else
         HAS_ALTIVEC := $(shell echo -e '\#include <altivec.h>\nvector int a;' |\
-                         gcc -c -x c - >&/dev/null && \
+                         gcc -c -x c - >/dev/null 2>&1 && \
                          rm ./-.o && echo yes)
         ifeq ($(HAS_ALTIVEC),yes)
                 OBJS += altivec1.o altivec2.o altivec4.o altivec8.o
-- 
2.6.2

