From 7be4beaa63a35e15e2c80938cc45aee62670cf49 Mon Sep 17 00:00:00 2001
From: Nicolas Ferre <nicolas.ferre@atmel.com>
Date: Thu, 15 Mar 2012 13:49:21 +0100
Subject: ARM: at91/Kconfig: add AT91SAM9x5 family to AT91_EARLY_DBGU0 entry

commit 514982adc7a5439140e09facf492e58fc8a7fcc3 upstream.

Signed-off-by: Nicolas Ferre <nicolas.ferre@atmel.com>
---
 arch/arm/mach-at91/Kconfig | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/arch/arm/mach-at91/Kconfig b/arch/arm/mach-at91/Kconfig
index d2922a1..e0b8b10 100644
--- a/arch/arm/mach-at91/Kconfig
+++ b/arch/arm/mach-at91/Kconfig
@@ -529,7 +529,7 @@ choice
 	prompt "Select a UART for early kernel messages"
 
 config AT91_EARLY_DBGU0
-	bool "DBGU on rm9200, 9260/9g20, 9261/9g10 and 9rl"
+	bool "DBGU on rm9200, 9260/9g20, 9261/9g10, 9rl and 9x5"
 	depends on HAVE_AT91_DBGU0
 
 config AT91_EARLY_DBGU1
-- 
1.8.0.197.g5a90748

