From e2086addb7f23c3aa497a3b9e36559efd6059e62 Mon Sep 17 00:00:00 2001
From: Nicolas Ferre <nicolas.ferre@atmel.com>
Date: Tue, 26 Jun 2012 11:07:32 +0200
Subject: phy/micrel: Use proper phy in gmac

Signed-off-by: Nicolas Ferre <nicolas.ferre@atmel.com>
---
 drivers/net/phy/micrel.c | 8 ++++++++
 1 file changed, 8 insertions(+)

diff --git a/drivers/net/phy/micrel.c b/drivers/net/phy/micrel.c
index 590f902..cc7f75b 100644
--- a/drivers/net/phy/micrel.c
+++ b/drivers/net/phy/micrel.c
@@ -191,6 +191,7 @@ static int __init ksphy_init(void)
 {
 	int ret;
 
+#if 0
 	ret = phy_driver_register(&ks8001_driver);
 	if (ret)
 		goto err1;
@@ -208,9 +209,15 @@ static int __init ksphy_init(void)
 	ret = phy_driver_register(&ks8051_driver);
 	if (ret)
 		goto err5;
+#endif
+
+	ret = phy_driver_register(&ksz9021_driver);
+	if (ret)
+		goto err1;
 
 	return 0;
 
+#if 0
 err5:
 	phy_driver_unregister(&ks8041_driver);
 err4:
@@ -219,6 +226,7 @@ err3:
 	phy_driver_unregister(&ksz9021_driver);
 err2:
 	phy_driver_unregister(&ks8001_driver);
+#endif
 err1:
 	return ret;
 }
-- 
1.8.0.197.g5a90748

