From af15c74766bcffe30022f7c27e477d2fce4b8471 Mon Sep 17 00:00:00 2001
From: Ludovic Desroches <ludovic.desroches@atmel.com>
Date: Tue, 10 Jul 2012 12:03:54 +0200
Subject: phy/micrel: we need to register ks8051 phy for emac

Signed-off-by: Ludovic Desroches <ludovic.desroches@atmel.com>
---
 drivers/net/phy/micrel.c | 6 ++++--
 1 file changed, 4 insertions(+), 2 deletions(-)

diff --git a/drivers/net/phy/micrel.c b/drivers/net/phy/micrel.c
index cc7f75b..2d80e01 100644
--- a/drivers/net/phy/micrel.c
+++ b/drivers/net/phy/micrel.c
@@ -206,10 +206,10 @@ static int __init ksphy_init(void)
 	ret = phy_driver_register(&ks8041_driver);
 	if (ret)
 		goto err4;
+#endif
 	ret = phy_driver_register(&ks8051_driver);
 	if (ret)
-		goto err5;
-#endif
+		goto err2;
 
 	ret = phy_driver_register(&ksz9021_driver);
 	if (ret)
@@ -227,6 +227,8 @@ err3:
 err2:
 	phy_driver_unregister(&ks8001_driver);
 #endif
+err2:
+	phy_driver_unregister(&ks8051_driver);
 err1:
 	return ret;
 }
-- 
1.8.0.197.g5a90748

