From 3d6072e790e82ecbfbb3dfc84e22ba325db16f92 Mon Sep 17 00:00:00 2001
From: Nicolas Ferre <nicolas.ferre@atmel.com>
Date: Mon, 24 Sep 2012 12:27:10 +0200
Subject: AT91: board-dt add mci pinmux for 9x5

Signed-off-by: Nicolas Ferre <nicolas.ferre@atmel.com>
---
 arch/arm/mach-at91/board-dt.c | 24 ++++++++++++++++++++++++
 1 file changed, 24 insertions(+)

diff --git a/arch/arm/mach-at91/board-dt.c b/arch/arm/mach-at91/board-dt.c
index 9355d77..15830cb 100644
--- a/arch/arm/mach-at91/board-dt.c
+++ b/arch/arm/mach-at91/board-dt.c
@@ -50,6 +50,30 @@ static void __init at91_dt_device_init(void)
 		printk("AT91: i2c pin mux done\n");
 		at91_set_gpio_input(AT91_PIN_PA7, 1);
 		printk("AT91: qt1070 pin mux done\n");
+
+		at91_set_gpio_input(AT91_PIN_PD14, 1);
+		at91_set_deglitch(AT91_PIN_PD14, 1);
+		at91_set_gpio_input(AT91_PIN_PD15, 1);
+		at91_set_deglitch(AT91_PIN_PD15, 1);
+		/* CLK */
+		at91_set_A_periph(AT91_PIN_PA17, 0);
+		/* CMD */
+		at91_set_A_periph(AT91_PIN_PA16, 1);
+		/* DAT0, DAT1..DAT3 */
+		at91_set_A_periph(AT91_PIN_PA15, 1);
+		at91_set_A_periph(AT91_PIN_PA18, 1);
+		at91_set_A_periph(AT91_PIN_PA19, 1);
+		at91_set_A_periph(AT91_PIN_PA20, 1);
+		/* CLK */
+		at91_set_B_periph(AT91_PIN_PA13, 0);
+		/* CMD */
+		at91_set_B_periph(AT91_PIN_PA12, 1);
+		/* DAT0, DAT1..DAT3 */
+		at91_set_B_periph(AT91_PIN_PA11, 1);
+		at91_set_B_periph(AT91_PIN_PA2, 1);
+		at91_set_B_periph(AT91_PIN_PA3, 1);
+		at91_set_B_periph(AT91_PIN_PA4, 1);
+		printk("AT91: mci0/1 pin mux done\n");
 	}
 }
 
-- 
1.8.0.197.g5a90748

