From 1a4758d4c8790c591d2be79382e878526f6dfcb6 Mon Sep 17 00:00:00 2001
From: Josh Wu <josh.wu@atmel.com>
Date: Mon, 24 Sep 2012 14:50:45 +0800
Subject: mtd: atmel_nand: add 4k page nand flash support for PMECC.

---
 drivers/mtd/nand/atmel_nand.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/drivers/mtd/nand/atmel_nand.c b/drivers/mtd/nand/atmel_nand.c
index a92f603..e881c15 100644
--- a/drivers/mtd/nand/atmel_nand.c
+++ b/drivers/mtd/nand/atmel_nand.c
@@ -952,6 +952,7 @@ static int __init atmel_pmecc_nand_init_params(struct platform_device *pdev,
 	/* set ECC page size and oob layout */
 	switch (mtd->writesize) {
 	case 2048:
+	case 4096:
 		host->pmecc_degree = PMECC_GF_DIMENSION_13;
 		host->pmecc_cw_len = (1 << host->pmecc_degree) - 1;
 		host->pmecc_sector_number = mtd->writesize / sector_size;
@@ -977,7 +978,6 @@ static int __init atmel_pmecc_nand_init_params(struct platform_device *pdev,
 		break;
 	case 512:
 	case 1024:
-	case 4096:
 		/* TODO */
 		dev_warn(host->dev,
 			"Unsupported page size for PMECC, use Software ECC\n");
-- 
1.8.0.197.g5a90748

