From f89227d6aec0e90ca9b2969e806703e63ca2d324 Mon Sep 17 00:00:00 2001
From: Ludovic Desroches <ludovic.desroches@atmel.com>
Date: Tue, 16 Oct 2012 17:07:17 +0200
Subject: media: atmel_isi: add pinctrl

Signed-off-by: Ludovic Desroches <ludovic.desroches@atmel.com>
---
 drivers/media/video/atmel-isi.c | 8 ++++++++
 1 file changed, 8 insertions(+)

diff --git a/drivers/media/video/atmel-isi.c b/drivers/media/video/atmel-isi.c
index 7a44df4..ef80695 100644
--- a/drivers/media/video/atmel-isi.c
+++ b/drivers/media/video/atmel-isi.c
@@ -20,6 +20,7 @@
 #include <linux/kernel.h>
 #include <linux/module.h>
 #include <linux/platform_device.h>
+#include <linux/pinctrl/consumer.h>
 #include <linux/slab.h>
 
 #include <media/atmel-isi.h>
@@ -949,6 +950,7 @@ static int __devinit atmel_isi_probe(struct platform_device *pdev)
 	struct device *dev = &pdev->dev;
 	struct soc_camera_host *soc_host;
 	struct isi_platform_data *pdata;
+	struct pinctrl *pinctrl;
 
 	pdata = dev->platform_data;
 	if (!pdata || !pdata->data_width_flags || !pdata->mck_hz) {
@@ -961,6 +963,12 @@ static int __devinit atmel_isi_probe(struct platform_device *pdev)
 	if (!regs)
 		return -ENXIO;
 
+	pinctrl = devm_pinctrl_get_select_default(&pdev->dev);
+	if (IS_ERR(pinctrl)) {
+		dev_err(&pdev->dev, "Failed to request pinctrl\n");
+		return PTR_ERR(pinctrl);
+	}
+
 	pclk = clk_get(&pdev->dev, "isi_clk");
 	if (IS_ERR(pclk))
 		return PTR_ERR(pclk);
-- 
1.8.0.197.g5a90748

