From 09d9306960be4117e00c086bc90bb510e581fdc9 Mon Sep 17 00:00:00 2001
From: Jean-Christophe PLAGNIOL-VILLARD <plagnioj@jcrosoft.com>
Date: Fri, 19 Oct 2012 15:13:13 +0800
Subject: arm: at91sam9x5ek: drop temporary pinmux

Signed-off-by: Jean-Christophe PLAGNIOL-VILLARD <plagnioj@jcrosoft.com>
---
 arch/arm/mach-at91/board-dt.c | 41 -----------------------------------------
 1 file changed, 41 deletions(-)

diff --git a/arch/arm/mach-at91/board-dt.c b/arch/arm/mach-at91/board-dt.c
index 95798f5..16b6b3c 100644
--- a/arch/arm/mach-at91/board-dt.c
+++ b/arch/arm/mach-at91/board-dt.c
@@ -79,45 +79,6 @@ static struct atmel_lcdfb_info __initdata ek_lcdc_data = {
 	.lcd_wiring_mode		= ATMEL_LCDC_WIRING_RGB,
 };
 
-void __init at91sam9x5_pinmux_lcd(void)
-{
-	at91_set_A_periph(AT91_PIN_PC26, 0);    /* LCDPWM */
-
-	at91_set_A_periph(AT91_PIN_PC27, 0);    /* LCDVSYNC */
-	at91_set_A_periph(AT91_PIN_PC28, 0);    /* LCDHSYNC */
-
-	at91_set_A_periph(AT91_PIN_PC24, 0);    /* LCDDISP */
-	at91_set_A_periph(AT91_PIN_PC29, 0);    /* LCDDEN */
-	at91_set_A_periph(AT91_PIN_PC30, 0);    /* LCDPCK */
-
-	at91_set_A_periph(AT91_PIN_PC0, 0);     /* LCDD0 */
-	at91_set_A_periph(AT91_PIN_PC1, 0);     /* LCDD1 */
-	at91_set_A_periph(AT91_PIN_PC2, 0);     /* LCDD2 */
-	at91_set_A_periph(AT91_PIN_PC3, 0);     /* LCDD3 */
-	at91_set_A_periph(AT91_PIN_PC4, 0);     /* LCDD4 */
-	at91_set_A_periph(AT91_PIN_PC5, 0);     /* LCDD5 */
-	at91_set_A_periph(AT91_PIN_PC6, 0);     /* LCDD6 */
-	at91_set_A_periph(AT91_PIN_PC7, 0);     /* LCDD7 */
-	at91_set_A_periph(AT91_PIN_PC8, 0);     /* LCDD8 */
-	at91_set_A_periph(AT91_PIN_PC9, 0);     /* LCDD9 */
-	at91_set_A_periph(AT91_PIN_PC10, 0);    /* LCDD10 */
-	at91_set_A_periph(AT91_PIN_PC11, 0);    /* LCDD11 */
-	at91_set_A_periph(AT91_PIN_PC12, 0);    /* LCDD12 */
-	at91_set_A_periph(AT91_PIN_PC13, 0);    /* LCDD13 */
-	at91_set_A_periph(AT91_PIN_PC14, 0);    /* LCDD14 */
-	at91_set_A_periph(AT91_PIN_PC15, 0);    /* LCDD15 */
-	at91_set_A_periph(AT91_PIN_PC16, 0);    /* LCDD16 */
-	at91_set_A_periph(AT91_PIN_PC17, 0);    /* LCDD17 */
-	at91_set_A_periph(AT91_PIN_PC18, 0);    /* LCDD18 */
-	at91_set_A_periph(AT91_PIN_PC19, 0);    /* LCDD19 */
-	at91_set_A_periph(AT91_PIN_PC20, 0);    /* LCDD20 */
-	at91_set_A_periph(AT91_PIN_PC21, 0);    /* LCDD21 */
-	at91_set_A_periph(AT91_PIN_PC22, 0);    /* LCDD22 */
-	at91_set_A_periph(AT91_PIN_PC23, 0);    /* LCDD23 */
-
-	printk("AT91: lcd pin mux done\n");
-}
-
 struct of_dev_auxdata at91_auxdata_lookup[] __initdata = {
 	OF_DEV_AUXDATA("atmel,at91sam9x5-lcd", 0xf8038000, "atmel_hlcdfb_base", &ek_lcdc_data),
 	OF_DEV_AUXDATA("atmel,at91sam9x5-lcd", 0xf8038100, "atmel_hlcdfb_ovl", &ek_lcdc_data),
@@ -141,8 +102,6 @@ static void __init at91_dt_device_init(void)
 	if (of_machine_is_compatible("atmel,at91sam9x5")) {
 		at91_set_gpio_input(AT91_PIN_PA7, 1);
 		printk("AT91: qt1070 pin mux done\n");
-
-		at91sam9x5_pinmux_lcd();
 	}
 
 	of_platform_populate(NULL, of_default_bus_match_table, at91_auxdata_lookup, NULL);
-- 
1.8.0.197.g5a90748

