From 87437fe8c93900352b99d6f883e730e7577f2a22 Mon Sep 17 00:00:00 2001
From: Jean-Christophe PLAGNIOL-VILLARD <plagnioj@jcrosoft.com>
Date: Tue, 23 Oct 2012 13:47:30 +0800
Subject: pinctrl: at19: fix typo on PULL_UP

Signed-off-by: Jean-Christophe PLAGNIOL-VILLARD <plagnioj@jcrosoft.com>
---
 drivers/pinctrl/pinctrl-at91.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/drivers/pinctrl/pinctrl-at91.c b/drivers/pinctrl/pinctrl-at91.c
index 605fffd..1a2404b 100644
--- a/drivers/pinctrl/pinctrl-at91.c
+++ b/drivers/pinctrl/pinctrl-at91.c
@@ -58,7 +58,7 @@ static struct at91_gpio_chip *gpio_chips[MAX_GPIO_BANKS];
 
 static int gpio_banks;
 
-#define PULL_UP		(0 << 1)
+#define PULL_UP		(1 << 0)
 #define MULTI_DRIVE	(1 << 1)
 #define DEGLITCH	(1 << 2)
 #define PULL_DOWN	(1 << 3)
-- 
1.8.0.197.g5a90748

