From c0c66e44e482f311e2a550bea04cb4c7e1bfb732 Mon Sep 17 00:00:00 2001
From: Zumeng Chen <zumeng.chen@windriver.com>
Date: Thu, 13 May 2010 15:34:18 +0800
Subject: [PATCH 116/119] fix kernel crash due to null pointer

Signed-off-by: Zumeng Chen <zumeng.chen@windriver.com>
---
 kernel/vbus/devices/venet/tap.c |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/kernel/vbus/devices/venet/tap.c b/kernel/vbus/devices/venet/tap.c
index b40c315..16cdb74 100644
--- a/kernel/vbus/devices/venet/tap.c
+++ b/kernel/vbus/devices/venet/tap.c
@@ -287,6 +287,6 @@ static void __exit venettap_cleanup(void)
 	vbus_devclass_unregister(&venettap_devclass);
 }
 
-module_init(venettap_init);
+late_initcall(venettap_init);
 module_exit(venettap_cleanup);
 
-- 
1.6.5.2

