From ffe810048ccb7fdac83cf34fce2d0d34181043e5 Mon Sep 17 00:00:00 2001
From: Gregory Haskins <ghaskins@novell.com>
Date: Tue, 17 Nov 2009 15:48:49 -0500
Subject: [PATCH 084/119] vbus: fix xioevent leak in kvm-connector

Signed-off-by: Gregory Haskins <ghaskins@novell.com>
---
 kernel/vbus/connectors/kvm.c |    4 ++++
 1 files changed, 4 insertions(+), 0 deletions(-)

diff --git a/kernel/vbus/connectors/kvm.c b/kernel/vbus/connectors/kvm.c
index d006137..d4ff89b 100644
--- a/kernel/vbus/connectors/kvm.c
+++ b/kernel/vbus/connectors/kvm.c
@@ -1389,6 +1389,10 @@ vbus_kvm_chardev_release(struct inode *inode, struct file *filp)
 		_eventq_detach(&vkvm->eventq.queues[i]);
 
 	_fastcall_channels_release(vkvm);
+	if (vkvm->fastcalls.ioevent)
+		kvm_xioevent_deassign(vkvm->fastcalls.ioevent);
+	if (vkvm->shmsignals)
+		kvm_xioevent_deassign(vkvm->shmsignals);
 	mmput(vkvm->mm);
 	vbus_memctx_put(vkvm->ctx);
 	vbus_client_put(vkvm->client);
-- 
1.6.5.2

