From 25bbb94faa8dc4972c4b3c29b1481396a3079cd8 Mon Sep 17 00:00:00 2001
From: Gregory Haskins <ghaskins@novell.com>
Date: Wed, 26 Aug 2009 20:54:27 -0400
Subject: [PATCH 033/119] venettap: make vsg-cache allocation static

Signed-off-by: Gregory Haskins <ghaskins@novell.com>
---
 kernel/vbus/devices/venet-tap.c |   18 +++++++++++-------
 1 files changed, 11 insertions(+), 7 deletions(-)

diff --git a/kernel/vbus/devices/venet-tap.c b/kernel/vbus/devices/venet-tap.c
index 1faaafc..b96fa1f 100644
--- a/kernel/vbus/devices/venet-tap.c
+++ b/kernel/vbus/devices/venet-tap.c
@@ -82,6 +82,11 @@ enum {
 	TX_IOQ_CONGESTED,
 };
 
+#define VSG_DESC_SIZE(count) (sizeof(struct venet_sg) + \
+			      sizeof(struct venet_iov) * ((count) - 1))
+
+#define MAX_VSG_DESC_SIZE VSG_DESC_SIZE(MAX_SKB_FRAGS)
+
 struct venettap;
 
 struct venettap_rx_ops {
@@ -119,6 +124,7 @@ struct venettap {
 		struct venettap_rx_ops      *rx_ops;
 		wait_queue_head_t            rx_empty;
 		struct {
+			char                 buf[MAX_VSG_DESC_SIZE];
 			struct venet_sg     *desc;
 			size_t               len;
 			int                  enabled:1;
@@ -329,13 +335,11 @@ venettap_sg_decode(struct venettap *priv, void *ptr, int len)
 	 * SG is enabled, so we need to pull in the venet_sg
 	 * header before we can interpret the rest of the
 	 * packet
-	 *
-	 * FIXME: Make sure this is not too big
 	 */
-	if (unlikely(len > priv->vbus.sg.len)) {
-		kfree(priv->vbus.sg.desc);
-		priv->vbus.sg.desc = kzalloc(len, GFP_KERNEL);
-	}
+	if (unlikely(len > MAX_VSG_DESC_SIZE))
+		return -1;
+
+	priv->vbus.sg.len = len;
 
 	vsg = priv->vbus.sg.desc;
 	ctx = priv->vbus.ctx;
@@ -1084,6 +1088,7 @@ venettap_negcap_sg(struct venettap *priv, u32 requested)
 
 	if (ret & VENET_CAP_SG) {
 		priv->vbus.sg.enabled = true;
+		priv->vbus.sg.desc = (struct venet_sg *)priv->vbus.sg.buf;
 		priv->vbus.rx_ops = &venettap_sg_rx_ops;
 	}
 
@@ -1307,7 +1312,6 @@ venettap_vlink_release(struct vbus_connection *conn)
 
 	priv->vbus.sg.enabled = false;
 	priv->vbus.rx_ops = &venettap_flat_rx_ops;
-	kfree(priv->vbus.sg.desc);
 	priv->vbus.sg.desc = NULL;
 	priv->vbus.sg.len = 0;
 }
-- 
1.6.5.2

