From 666595ddc096b4903df03169c28f6700c640e354 Mon Sep 17 00:00:00 2001
From: Gregory Haskins <ghaskins@novell.com>
Date: Fri, 18 Sep 2009 20:03:36 -0400
Subject: [PATCH 042/119] venet: fix kbuild error

Signed-off-by: Gregory Haskins <ghaskins@novell.com>
---
 kernel/vbus/devices/venet/Kconfig |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/kernel/vbus/devices/venet/Kconfig b/kernel/vbus/devices/venet/Kconfig
index 81a4db5..4f89afb 100644
--- a/kernel/vbus/devices/venet/Kconfig
+++ b/kernel/vbus/devices/venet/Kconfig
@@ -3,7 +3,7 @@
 #
 
 config VBUS_VENETDEV
-       bool
+       tristate
        default n
 
 config VBUS_VENETTAP
-- 
1.6.5.2

