From 3fc2224d82e5b90054dc6ee38395e566a2fedf35 Mon Sep 17 00:00:00 2001
From: Gregory Haskins <ghaskins@novell.com>
Date: Wed, 26 Aug 2009 20:33:57 -0400
Subject: [PATCH 029/119] venettap: use NET_IP_ALIGN for reserved space in skb

Signed-off-by: Gregory Haskins <ghaskins@novell.com>
---
 kernel/vbus/devices/venet-tap.c |    4 ++--
 1 files changed, 2 insertions(+), 2 deletions(-)

diff --git a/kernel/vbus/devices/venet-tap.c b/kernel/vbus/devices/venet-tap.c
index 8a8fd3a..48bb9dc 100644
--- a/kernel/vbus/devices/venet-tap.c
+++ b/kernel/vbus/devices/venet-tap.c
@@ -555,7 +555,7 @@ venettap_rx(struct venettap *priv)
 			goto next;
 		}
 
-		skb = dev_alloc_skb(len+2);
+		skb = dev_alloc_skb(len + NET_IP_ALIGN);
 		if (unlikely(!skb)) {
 			printk(KERN_INFO "VENETTAP: skb alloc failed:"	\
 			       " memory squeeze.\n");
@@ -565,7 +565,7 @@ venettap_rx(struct venettap *priv)
 		}
 
 		/* align IP on 16B boundary */
-		skb_reserve(skb, 2);
+		skb_reserve(skb, NET_IP_ALIGN);
 
 		ret = rx_ops->import(priv, skb, (void *)iter.desc->ptr, len);
 		if (unlikely(ret < 0)) {
-- 
1.6.5.2

