From 3db8e28aaf97be2f87241c987198b35e41fc24f5 Mon Sep 17 00:00:00 2001
From: Shuo Liu <shuo.liu@windriver.com>
Date: Tue, 13 Oct 2009 13:03:18 +0800
Subject: [PATCH 2/4] Modification of the start of the fixmap

Fixed FIXADDR_TOP to make kmap_coherent work for BCM5621X.

Signed-off-by: Shuo Liu <shuo.liu@windriver.com>
---
 include/asm-mips/fixmap.h |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/include/asm-mips/fixmap.h b/include/asm-mips/fixmap.h
index 9cc8522..3f1d57b 100644
--- a/include/asm-mips/fixmap.h
+++ b/include/asm-mips/fixmap.h
@@ -67,7 +67,7 @@ enum fixed_addresses {
  * the start of the fixmap, and leave one page empty
  * at the top of mem..
  */
-#if defined(CONFIG_CPU_TX39XX) || defined(CONFIG_CPU_TX49XX)
+#if defined(CONFIG_CPU_TX39XX) || defined(CONFIG_CPU_TX49XX) || defined(CONFIG_BCM5621X)
 #define FIXADDR_TOP	((unsigned long)(long)(int)(0xff000000 - 0x20000))
 #else
 #define FIXADDR_TOP	((unsigned long)(long)(int)0xfffe0000)
-- 
1.6.5.2

