From 594259a969f87d14c6ba0ef10756041c6c4033c5 Mon Sep 17 00:00:00 2001
From: Xulin Sun <xulin.sun@windriver.com>
Date: Fri, 14 Aug 2015 16:50:53 +0800
Subject: [PATCH 324/451] temp: pci/designware: comment out MSI related code

Signed-off-by: Xulin Sun<xulin.sun@windriver.com>
---
 drivers/pci/host/pcie-designware.c |    9 +++++----
 1 files changed, 5 insertions(+), 4 deletions(-)

diff --git a/drivers/pci/host/pcie-designware.c b/drivers/pci/host/pcie-designware.c
index 91afdd2..1953853 100644
--- a/drivers/pci/host/pcie-designware.c
+++ b/drivers/pci/host/pcie-designware.c
@@ -152,7 +152,7 @@ static int dw_pcie_wr_own_conf(struct pcie_port *pp, int where, int size,
 	return ret;
 }
 
-#ifndef CONFIG_ARCH_FSL_LS2085A
+#if !defined(CONFIG_ARCH_FSL_LS2085A) && !defined(CONFIG_ARCH_FSL_LS1043A)
 static struct irq_chip dw_msi_irq_chip = {
 	.name = "PCI-MSI",
 	.irq_enable = unmask_msi_irq,
@@ -317,6 +317,7 @@ static void dw_msi_teardown_irq(struct msi_controller *chip, unsigned int irq)
 
 	clear_irq_range(pp, irq, 1, data->hwirq);
 }
+#endif
 
 static struct msi_controller dw_pcie_msi_chip = {
 	.setup_irq = NULL,
@@ -331,7 +332,7 @@ int dw_pcie_link_up(struct pcie_port *pp)
 		return 0;
 }
 
-#ifndef CONFIG_ARCH_FSL_LS2085A
+#if !defined(CONFIG_ARCH_FSL_LS2085A) && !defined(CONFIG_ARCH_FSL_LS1043A)
 static int dw_pcie_msi_map(struct irq_domain *domain, unsigned int irq,
 			irq_hw_number_t hwirq)
 {
@@ -531,7 +532,7 @@ int __init dw_pcie_host_init(struct pcie_port *pp)
 		dev_err(pp->dev, "Failed to parse the number of lanes\n");
 		return -EINVAL;
 	}
-#ifndef CONFIG_ARCH_FSL_LS2085A
+#if !defined(CONFIG_ARCH_FSL_LS2085A) && !defined(CONFIG_ARCH_FSL_LS1043A)
 	if (IS_ENABLED(CONFIG_PCI_MSI)) {
 		if (!pp->ops->msi_host_init) {
 			pp->irq_domain = irq_domain_add_linear(pp->dev->of_node,
@@ -577,7 +578,7 @@ int __init dw_pcie_host_init(struct pcie_port *pp)
 	dw_pcie_wr_own_conf(pp, PCIE_LINK_WIDTH_SPEED_CONTROL, 4, val);
 
 #ifdef CONFIG_PCI_MSI
-#ifndef CONFIG_ARCH_FSL_LS2085A
+#if !defined(CONFIG_ARCH_FSL_LS2085A) && !defined(CONFIG_ARCH_FSL_LS1043A)
 	dw_pcie_msi_chip.dev = pp->dev;
 	dw_pci.msi_ctrl = &dw_pcie_msi_chip;
 #endif
-- 
1.7.5.4

