From 006773b03f25c769255b9ffa6f2c69e0fabb99c8 Mon Sep 17 00:00:00 2001
From: Jack Tan <jack.tan@windriver.com>
Date: Fri, 14 Feb 2014 17:26:08 +0800
Subject: [PATCH] bcm-xlp2: compile error fixup

Based on SDK 3.0 (2013-10-29)

Signed-off-by: Jack Tan <jack.tan@windriver.com>

diff --git a/arch/mips/include/asm/mipsregs.h b/arch/mips/include/asm/mipsregs.h
index 3d0074e10595..41a3a7701499 100644
--- a/arch/mips/include/asm/mipsregs.h
+++ b/arch/mips/include/asm/mipsregs.h
@@ -13,6 +13,7 @@
 #ifndef _ASM_MIPSREGS_H
 #define _ASM_MIPSREGS_H
 
+#include <linux/types.h>
 #include <linux/linkage.h>
 #include <linux/types.h>
 #include <asm/hazards.h>
-- 
2.0.1

