From 44d3893dd44462a23c7a25573fa91435844065d0 Mon Sep 17 00:00:00 2001
From: Kamlakant Patel <kamlakant.patel@broadcom.com>
Date: Wed, 4 Dec 2013 10:04:02 +0530
Subject: [PATCH 007/160] sd/mmc: Fix mmc detection on xlp2xx B0 boards

Signed-off-by: Kamlakant Patel <kamlakant.patel@broadcom.com>
Signed-off-by: Jack Tan <jack.tan@windriver.com>
Signed-off-by: Jack Tan <jiankemeng@gmail.com>
---
 drivers/mmc/host/sdhci-xlp.c |   17 ++++++-----------
 1 file changed, 6 insertions(+), 11 deletions(-)

diff --git a/drivers/mmc/host/sdhci-xlp.c b/drivers/mmc/host/sdhci-xlp.c
index 966714a..fbf9bdb 100644
--- a/drivers/mmc/host/sdhci-xlp.c
+++ b/drivers/mmc/host/sdhci-xlp.c
@@ -45,7 +45,6 @@
 
 #define XLP_SLOT_SIZE		0x100
 #define XLP_NUM_SD_SLOT		2
-#define XLP_MMC_CAP_REG_LOW	0x204
 
 struct sdhci_xlp_chip {
 	struct pci_dev		*pdev;
@@ -68,17 +67,13 @@ static const struct pci_device_id xlp_pci_ids[] = {
 	},
 };
 
-unsigned int xlp_get_max_clock(struct sdhci_host *sd_host)
+static unsigned int xlp_get_max_clock(struct sdhci_host *sd_host)
 {
-	if (!cpu_is_xlp9xx())
-		return 0;
-#if 0
-	u32 val;
-
-	val = readl(sd_host->ioaddr, XLP_MMC_CAP_REG_LOW);
-	return ((val >> 7) & 0xff) << 20;
-#endif
-	return 133 << 20;  /* 133MHz */
+	/*
+	 * SDHCI_CAPABILITIES[15:8] register gives 0x00 instead of 0x85(133)
+	 * on xlp2xx B0 and xlp9xx chips. Forcibly returning 133MHz.
+	 */
+	return 133000000;
 }
 static struct sdhci_ops xlp_sdhci_ops = {
 	.get_max_clock = xlp_get_max_clock,
-- 
1.7.9.5

