From dd987447a94164e306ce416d9935f7cd8fb8cf4f Mon Sep 17 00:00:00 2001
From: Vikas Gupta <vikas.gupta@broadcom.com>
Date: Fri, 22 Nov 2013 02:22:27 -0800
Subject: [PATCH 073/160] nae: added XLAUI interface type for setting MTU

Signed-off-by: Jack Tan <jack.tan@windriver.com>
Signed-off-by: Jack Tan <jiankemeng@gmail.com>
---
 drivers/netlogic/nae/xlpge_nae.c |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/drivers/netlogic/nae/xlpge_nae.c b/drivers/netlogic/nae/xlpge_nae.c
index 2da4ea37..122d675 100644
--- a/drivers/netlogic/nae/xlpge_nae.c
+++ b/drivers/netlogic/nae/xlpge_nae.c
@@ -1218,7 +1218,7 @@ static int nlm_xlp_nae_change_mtu(struct net_device *dev, int new_mtu)
 			//nlm_xlp_mac_set_enable(priv, 0); /* Disable MAC TX/RX */
 	}
 
-	if (priv->type==SGMII_IF || priv->type==XAUI_IF)
+	if (priv->type==SGMII_IF || priv->type==XAUI_IF || priv->type== XLAUI_IF)
 		netsoc_set_framesize(nae_port, local_mtu);
 	else if (priv->type==INTERLAKEN_IF){
 		/*TODO: Add IL frame set in brcm_netsoc */
-- 
1.7.9.5

