From 266b826e72bd3eb138270a474942bf7023524dd2 Mon Sep 17 00:00:00 2001
From: Subhendu Sekhar Behera <sbehera@broadcom.com>
Date: Mon, 10 Mar 2014 11:11:39 +0530
Subject: [PATCH 095/163] Fix: Compilation Error

Signed-off-by: Jack Tan <jack.tan@windriver.com>
---
 arch/mips/pci/pci-xlp.c |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/arch/mips/pci/pci-xlp.c b/arch/mips/pci/pci-xlp.c
index a28cfb8c..8548534 100644
--- a/arch/mips/pci/pci-xlp.c
+++ b/arch/mips/pci/pci-xlp.c
@@ -331,7 +331,7 @@ int xlp_fixup_numa_node(struct device *dev)
 
 		/* fix the node numbers in top level devices */
 		if (pdev->bus->number == 0) {
-		}if (cpu_is_xlp9xx())
+			if (cpu_is_xlp9xx())
 				node = PCI_FUNC(pdev->devfn);
 			else
 				node = PCI_SLOT(pdev->devfn) / 8;
-- 
1.7.9.5

