From 1aec54f96e7a14994c291de35d8d490ddab0ee27 Mon Sep 17 00:00:00 2001
From: Kamlakant Patel <kamlakant.patel@broadcom.com>
Date: Mon, 2 Dec 2013 17:21:05 +0530
Subject: [PATCH] sd/mmc: fix simultaneous transfer issue on both mmc slots

Signed-off-by: Kamlakant Patel <kamlakant.patel@broadcom.com>
Signed-off-by: Jack Tan <jack.tan@windriver.com>
Signed-off-by: Jack Tan <jiankemeng@gmail.com>

diff --git a/drivers/mmc/host/sdhci.c b/drivers/mmc/host/sdhci.c
index 652a6cbdfa99..c8e732fbee5a 100644
--- a/drivers/mmc/host/sdhci.c
+++ b/drivers/mmc/host/sdhci.c
@@ -1305,6 +1305,16 @@ static int sdhci_set_power(struct sdhci_host *host, unsigned short power)
  *                                                                           *
 \*****************************************************************************/
 
+/**
+ * On Netlogic XLP platform, while doing transfers on both the slots
+ * simultaneously, hardware generates CRC error. This is a software workaround
+ * to avoid the CRC error.
+ */
+
+#ifdef CONFIG_CPU_XLP
+static DEFINE_MUTEX(sdhci_cmd_inprogress);
+#endif
+
 static void sdhci_request(struct mmc_host *mmc, struct mmc_request *mrq)
 {
 	struct sdhci_host *host;
@@ -1314,6 +1324,10 @@ static void sdhci_request(struct mmc_host *mmc, struct mmc_request *mrq)
 
 	host = mmc_priv(mmc);
 
+#ifdef CONFIG_CPU_XLP
+	mutex_lock(&sdhci_cmd_inprogress);
+#endif
+
 	sdhci_runtime_pm_get(host);
 
 	present = mmc_gpio_get_cd(host->mmc);
@@ -2150,6 +2164,10 @@ static void sdhci_tasklet_finish(unsigned long param)
 	mmiowb();
 	raw_spin_unlock_irqrestore(&host->lock, flags);
 
+#ifdef CONFIG_CPU_XLP
+	mutex_unlock(&sdhci_cmd_inprogress);
+#endif
+
 	mmc_request_done(host->mmc, mrq);
 	sdhci_runtime_pm_put(host);
 }
-- 
2.1.0

