From adaf71df40926654e6f4bbc7f707e4e07c6d0a91 Mon Sep 17 00:00:00 2001
From: Kamlakant Patel <kamlakant.patel@broadcom.com>
Date: Mon, 2 Dec 2013 17:21:05 +0530
Subject: [PATCH 006/160] sd/mmc: fix simultaneous transfer issue on both mmc
 slots

Signed-off-by: Kamlakant Patel <kamlakant.patel@broadcom.com>
Signed-off-by: Jack Tan <jack.tan@windriver.com>
Signed-off-by: Jack Tan <jiankemeng@gmail.com>
---
 drivers/mmc/host/sdhci.c |   18 ++++++++++++++++++
 1 file changed, 18 insertions(+)

diff --git a/drivers/mmc/host/sdhci.c b/drivers/mmc/host/sdhci.c
index 01fa2b1..21f1abf 100644
--- a/drivers/mmc/host/sdhci.c
+++ b/drivers/mmc/host/sdhci.c
@@ -1322,6 +1322,16 @@ static int sdhci_set_power(struct sdhci_host *host, unsigned short power)
  *                                                                           *
 \*****************************************************************************/
 
+/**
+ * On Netlogic XLP platform, while doing transfers on both the slots
+ * simultaneously, hardware generates CRC error. This is a software workaround
+ * to avoid the CRC error.
+ */
+
+#ifdef CONFIG_CPU_XLP
+static DEFINE_MUTEX(sdhci_cmd_inprogress);
+#endif
+
 static void sdhci_request(struct mmc_host *mmc, struct mmc_request *mrq)
 {
 	struct sdhci_host *host;
@@ -1331,6 +1341,10 @@ static void sdhci_request(struct mmc_host *mmc, struct mmc_request *mrq)
 
 	host = mmc_priv(mmc);
 
+#ifdef CONFIG_CPU_XLP
+	mutex_lock(&sdhci_cmd_inprogress);
+#endif
+
 	sdhci_runtime_pm_get(host);
 
 	raw_spin_lock_irqsave(&host->lock, flags);
@@ -2168,6 +2182,10 @@ static void sdhci_tasklet_finish(unsigned long param)
 	mmiowb();
 	raw_spin_unlock_irqrestore(&host->lock, flags);
 
+#ifdef CONFIG_CPU_XLP
+	mutex_unlock(&sdhci_cmd_inprogress);
+#endif
+
 	mmc_request_done(host->mmc, mrq);
 	sdhci_runtime_pm_put(host);
 }
-- 
1.7.9.5

