From c5708be993b0715f9884a7a227e87a8ccd29325e Mon Sep 17 00:00:00 2001
From: Paul Gortmaker <paul.gortmaker@windriver.com>
Date: Fri, 19 Jun 2009 10:29:07 -0400
Subject: [PATCH] rmi: delete unused clock defines

At one point there was a plan to use a PIC timer for a clock
interrupt, but it was scrapped.  These are leftovers from
that plan.

Signed-off-by: Paul Gortmaker <paul.gortmaker@windriver.com>
---
 include/asm-mips/rmi/pic.h |    5 -----
 1 files changed, 0 insertions(+), 5 deletions(-)

diff --git a/include/asm-mips/rmi/pic.h b/include/asm-mips/rmi/pic.h
index 5d22b42..3646727 100644
--- a/include/asm-mips/rmi/pic.h
+++ b/include/asm-mips/rmi/pic.h
@@ -44,7 +44,6 @@
 #define PIC_IRT_TIMER_5_INDEX      6
 #define PIC_IRT_TIMER_6_INDEX      7
 #define PIC_IRT_TIMER_7_INDEX      8
-#define PIC_IRT_CLOCK_INDEX        PIC_IRT_TIMER_7_INDEX
 #define PIC_IRT_UART_0_INDEX       9
 #define PIC_IRT_UART_1_INDEX       10
 #define PIC_IRT_I2C_0_INDEX       11
@@ -90,7 +89,6 @@
 #define PIC_SYS_TIMER_MAXVAL_1_BASE 0x110
 #define PIC_SYS_TIMER_0_BASE 0x120
 #define PIC_SYS_TIMER_1_BASE 0x130
-#define PIC_CLOCK_TIMER 7
 
 #define PIC_CTRL    0x00
 #define PIC_IPI     0x04
@@ -124,8 +122,6 @@
 #define PIC_IRT_1_TIMER_6     (PIC_IRT_1_BASE   + PIC_IRT_TIMER_6_INDEX)
 #define PIC_IRT_0_TIMER_7     (PIC_IRT_0_BASE   + PIC_IRT_TIMER_7_INDEX)
 #define PIC_IRT_1_TIMER_7     (PIC_IRT_1_BASE   + PIC_IRT_TIMER_7_INDEX)
-#define PIC_IRT_0_CLOCK       (PIC_IRT_0_TIMER_7)
-#define PIC_IRT_1_CLOCK       (PIC_IRT_1_TIMER_7)
 #define PIC_IRT_0_UART_0 (PIC_IRT_0_BASE + PIC_IRT_UART_0_INDEX)
 #define PIC_IRT_1_UART_0 (PIC_IRT_1_BASE + PIC_IRT_UART_0_INDEX)
 #define PIC_IRT_0_UART_1 (PIC_IRT_0_BASE + PIC_IRT_UART_1_INDEX)
@@ -159,7 +155,6 @@
 #define PIC_TIMER_5_IRQ (PIC_IRQ_BASE + PIC_IRT_TIMER_5_INDEX)
 #define PIC_TIMER_6_IRQ (PIC_IRQ_BASE + PIC_IRT_TIMER_6_INDEX)
 #define PIC_TIMER_7_IRQ (PIC_IRQ_BASE + PIC_IRT_TIMER_7_INDEX)
-#define PIC_CLOCK_IRQ   (PIC_TIMER_7_IRQ)
 #define PIC_UART_0_IRQ  (PIC_IRQ_BASE + PIC_IRT_UART_0_INDEX)
 #define PIC_UART_1_IRQ  (PIC_IRQ_BASE + PIC_IRT_UART_1_INDEX)
 #define PIC_I2C_0_IRQ   (PIC_IRQ_BASE + PIC_IRT_I2C_0_INDEX)
-- 
1.6.0.4

