From a42184652f2c0e4919e7c9e21616f58ff3de40bd Mon Sep 17 00:00:00 2001
From: Bruce Ashfield <bruce.ashfield@windriver.com>
Date: Wed, 15 Jun 2011 16:12:17 -0400
Subject: [PATCH] qemuppc: irq disable fixups

Signed-off-by: Bruce Ashfield <bruce.ashfield@windriver.com>
---
 kernel/irq/chip.c |    6 ++----
 1 files changed, 2 insertions(+), 4 deletions(-)

diff --git a/kernel/irq/chip.c b/kernel/irq/chip.c
index a813b57..72a7e2f 100644
--- a/kernel/irq/chip.c
+++ b/kernel/irq/chip.c
@@ -142,10 +142,8 @@ EXPORT_SYMBOL_GPL(irq_get_irq_data);
 static void irq_state_clr_disabled(struct irq_desc *desc)
 {
 #ifdef CONFIG_PPC_QEMU
-	struct irq_desc *desc = irq_data_to_desc(irq);
-
-	if (!(desc->status & IRQ_DELAYED_DISABLE))
-		desc->chip->mask(irq);
+	if (!(desc->status_use_accessors & IRQ_DELAYED_DISABLE))
+		desc->irq_data.chip->irq_mask(&desc->irq_data);
 #else
 	irqd_clear(&desc->irq_data, IRQD_IRQ_DISABLED);
 #endif /* CONFIG_PPC_QEMU */
-- 
1.7.0.4

