From 29b56507193cd6ac0da09249d075290ee37c927e Mon Sep 17 00:00:00 2001
From: Nikos Mavrogiannopoulos <nmav@gnutls.org>
Date: Sat, 28 Nov 2009 19:35:33 +0200
Subject: [PATCH 003/291] cryptodev: enable

---
 drivers/staging/crypto/cryptodev/example2.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/drivers/staging/crypto/cryptodev/example2.c b/drivers/staging/crypto/cryptodev/example2.c
index ddd35b58d39f..e25dc5ba0acc 100644
--- a/drivers/staging/crypto/cryptodev/example2.c
+++ b/drivers/staging/crypto/cryptodev/example2.c
@@ -108,7 +108,7 @@ test_crypto(int cfd)
 	} else {
 		fprintf(stderr, "HMAC test 1: passed\n");
 	}
-return 0;
+
 	sess.cipher = CRYPTO_AES_CBC;
 	sess.mac = CRYPTO_SHA1_HMAC;
 	sess.keylen = KEY_SIZE;
-- 
2.1.0

