From 1facb10176279240f87f26910235da1298ed84ac Mon Sep 17 00:00:00 2001
From: Nikos Mavrogiannopoulos <nmav@gnutls.org>
Date: Fri, 19 Feb 2010 17:18:56 +0100
Subject: [PATCH 014/291] cryptodev: Corrected

---
 drivers/staging/crypto/cryptodev/example-cipher.c | 8 ++++++++
 1 file changed, 8 insertions(+)

diff --git a/drivers/staging/crypto/cryptodev/example-cipher.c b/drivers/staging/crypto/cryptodev/example-cipher.c
index c6aa55d9f4cc..500d773aebbf 100644
--- a/drivers/staging/crypto/cryptodev/example-cipher.c
+++ b/drivers/staging/crypto/cryptodev/example-cipher.c
@@ -54,10 +54,18 @@ test_crypto(int cfd)
 		perror("ioctl(CIOCCRYPT)");
 		return 1;
 	}
+
+	if (ioctl(cfd, CIOCGSESSION, &sess)) {
+		perror("ioctl(CIOCGSESSION)");
+		return 1;
+	}
 	
 	/* Decrypt data.encrypted to data.decrypted */
+	cryp.ses = sess.ses;
+	cryp.len = sizeof(plaintext);
 	cryp.src = ciphertext;
 	cryp.dst = ciphertext;
+	cryp.iv = iv;
 	cryp.op = COP_DECRYPT;
 	if (ioctl(cfd, CIOCCRYPT, &cryp)) {
 		perror("ioctl(CIOCCRYPT)");
-- 
2.1.0

