From fa5bc644ef08bf15f02cfbb71137639cdf084e94 Mon Sep 17 00:00:00 2001
From: Nikos Mavrogiannopoulos <nmav@gnutls.org>
Date: Fri, 19 Feb 2010 23:56:20 +0100
Subject: [PATCH 020/291] cryptodev: Added

---
 drivers/staging/crypto/cryptodev/README | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/drivers/staging/crypto/cryptodev/README b/drivers/staging/crypto/cryptodev/README
index 49dfeba81dd9..5554a5efb6a1 100644
--- a/drivers/staging/crypto/cryptodev/README
+++ b/drivers/staging/crypto/cryptodev/README
@@ -2,3 +2,8 @@ This is a /dev/crypto device driver, equivalent to those in OpenBSD or
 FreeBSD. The main idea is to access of existing ciphers in kernel space 
 from userspace, thus enabling re-use of a hardware implementation of a
 cipher.
+
+It was initially written for linux 2.6.8 by Michal Ludvig. Compatibility
+fixes for *BSD cryptodev as well as Was ported to 2.6.27 blkcipher API 
+by Nikos Mavrogiannopoulos. Initial blkcipher async API porting by
+Michael Weiser.
-- 
2.1.0

