From ae138f5ebc0e0052003a8e8a56fc7a17051e34a7 Mon Sep 17 00:00:00 2001
From: Nikos Mavrogiannopoulos <nmav@gnutls.org>
Date: Sat, 20 Feb 2010 10:49:48 +0100
Subject: [PATCH 026/291] cryptodev: Added

---
 drivers/staging/crypto/cryptodev/Makefile | 9 +++++++++
 1 file changed, 9 insertions(+)

diff --git a/drivers/staging/crypto/cryptodev/Makefile b/drivers/staging/crypto/cryptodev/Makefile
index 2093acca563f..5a8eb2b85c5f 100644
--- a/drivers/staging/crypto/cryptodev/Makefile
+++ b/drivers/staging/crypto/cryptodev/Makefile
@@ -14,3 +14,12 @@ install:
 
 clean:
 	make -C $(KERNEL_DIR) SUBDIRS=`pwd` clean
+	rm -f $(hostprogs)
+
+hostprogs := example-cipher example-hmac
+example-cipher-objs := example-cipher.o
+example-hmac-objs := example-hmac.o
+
+check: $(hostprogs)
+	./example-cipher
+	./example-hmac
-- 
2.1.0

