From 7bee1f73cc294f96808bbafab0ef6b5e602b1b96 Mon Sep 17 00:00:00 2001
From: Nikos Mavrogiannopoulos <nmav@gnutls.org>
Date: Wed, 17 Mar 2010 18:04:57 +0100
Subject: [PATCH 036/291] cryptodev: Added

---
 drivers/staging/crypto/cryptodev/README | 4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

diff --git a/drivers/staging/crypto/cryptodev/README b/drivers/staging/crypto/cryptodev/README
index 5554a5efb6a1..0cc8fb6e0436 100644
--- a/drivers/staging/crypto/cryptodev/README
+++ b/drivers/staging/crypto/cryptodev/README
@@ -4,6 +4,8 @@ from userspace, thus enabling re-use of a hardware implementation of a
 cipher.
 
 It was initially written for linux 2.6.8 by Michal Ludvig. Compatibility
-fixes for *BSD cryptodev as well as Was ported to 2.6.27 blkcipher API 
+fixes for *BSD cryptodev as well as porting to 2.6.27 blkcipher API 
 by Nikos Mavrogiannopoulos. Initial blkcipher async API porting by
 Michael Weiser.
+
+Maintained by Nikos Mavrogiannopoulos (nmav [at] gnutls [dot] org)
-- 
2.1.0

