From 904fb3830fe2ad80974eefbb59f9dbaf87568266 Mon Sep 17 00:00:00 2001
From: Nikos Mavrogiannopoulos <nmav@gnutls.org>
Date: Fri, 18 Jun 2010 23:48:33 +0200
Subject: [PATCH 053/291] cryptodev: Made

---
 drivers/staging/crypto/cryptodev/NEWS | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/drivers/staging/crypto/cryptodev/NEWS b/drivers/staging/crypto/cryptodev/NEWS
index 87686e21bd0f..53e73d94fa5f 100644
--- a/drivers/staging/crypto/cryptodev/NEWS
+++ b/drivers/staging/crypto/cryptodev/NEWS
@@ -1,4 +1,4 @@
-Version 0.2 (unreleased)
+Version 0.2 (released 2010-06-18)
 
 * Added compat_ioctl() to allow working on systems where userspace is 32bits
 and kernel is operating in 64bit mode (Phil Sutter)
-- 
2.1.0

