From da6544bf3e41f90937661ea681228e9e1c4b6d0e Mon Sep 17 00:00:00 2001
From: Phil Sutter <phil.sutter@viprinet.com>
Date: Tue, 22 Jun 2010 10:51:49 +0200
Subject: [PATCH 059/291] cryptodev: allow

---
 drivers/staging/crypto/cryptodev/cryptodev_main.c | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/drivers/staging/crypto/cryptodev/cryptodev_main.c b/drivers/staging/crypto/cryptodev/cryptodev_main.c
index 6328624146e9..790dac3da61d 100644
--- a/drivers/staging/crypto/cryptodev/cryptodev_main.c
+++ b/drivers/staging/crypto/cryptodev/cryptodev_main.c
@@ -129,6 +129,9 @@ crypto_create_session(struct fcrypt *fcr, struct session_op *sop)
 		case CRYPTO_CAMELLIA_CBC:
 			alg_name = "cbc(camelia)";
 			break;
+		case CRYPTO_NULL:
+			alg_name = "ecb(cipher_null)";
+			break;
 		default:
 			dprintk(1,KERN_DEBUG,"%s: bad cipher: %d\n", __func__, sop->cipher);
 			return -EINVAL;
-- 
2.1.0

