From f97400fef699581e916d7bd1c0753520344a7019 Mon Sep 17 00:00:00 2001
From: Nikos Mavrogiannopoulos <nmav@gnutls.org>
Date: Wed, 23 Jun 2010 00:19:40 +0200
Subject: [PATCH 063/291] cryptodev: ignore

---
 drivers/staging/crypto/cryptodev/.gitignore | 1 +
 1 file changed, 1 insertion(+)

diff --git a/drivers/staging/crypto/cryptodev/.gitignore b/drivers/staging/crypto/cryptodev/.gitignore
index 740ebd5eb3fa..06eaf99189a7 100644
--- a/drivers/staging/crypto/cryptodev/.gitignore
+++ b/drivers/staging/crypto/cryptodev/.gitignore
@@ -9,5 +9,6 @@ Module.symvers
 modules.order
 examples/cipher
 examples/hmac
+examples/speed
 releases
 scripts
-- 
2.1.0

