From 5e72de7cf50f58375635938043b9ce760c39ec6a Mon Sep 17 00:00:00 2001
From: Nikos Mavrogiannopoulos <nmav@gnutls.org>
Date: Wed, 23 Jun 2010 08:41:30 +0200
Subject: [PATCH 064/291] cryptodev: added

---
 drivers/staging/crypto/cryptodev/README | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/drivers/staging/crypto/cryptodev/README b/drivers/staging/crypto/cryptodev/README
index 0cc8fb6e0436..9faff875b3d0 100644
--- a/drivers/staging/crypto/cryptodev/README
+++ b/drivers/staging/crypto/cryptodev/README
@@ -6,6 +6,6 @@ cipher.
 It was initially written for linux 2.6.8 by Michal Ludvig. Compatibility
 fixes for *BSD cryptodev as well as porting to 2.6.27 blkcipher API 
 by Nikos Mavrogiannopoulos. Initial blkcipher async API porting by
-Michael Weiser.
+Michael Weiser. Zero copy support was implemented by Phil Sutter.
 
 Maintained by Nikos Mavrogiannopoulos (nmav [at] gnutls [dot] org)
-- 
2.1.0

