From bbebb7bf531da4204cc3ca2b6bb3ce4754494bbb Mon Sep 17 00:00:00 2001
From: Nikos Mavrogiannopoulos <nmav@gnutls.org>
Date: Sat, 3 Jul 2010 15:30:57 +0200
Subject: [PATCH 068/291] cryptodev: removed

---
 drivers/staging/crypto/cryptodev/cryptodev.h | 1 -
 1 file changed, 1 deletion(-)

diff --git a/drivers/staging/crypto/cryptodev/cryptodev.h b/drivers/staging/crypto/cryptodev/cryptodev.h
index 613fe8524a67..7c254314830a 100644
--- a/drivers/staging/crypto/cryptodev/cryptodev.h
+++ b/drivers/staging/crypto/cryptodev/cryptodev.h
@@ -98,7 +98,6 @@ struct session_op {
 /* struct crypt_op flags */
 
 #define COP_FLAG_NONE		(0 << 0) /* totally no flag */
-#define COP_FLAG_ZCOPY		(1 << 0) /* operate directly on the userspace buffer */
 
 /* Stuff for bignum arithmetic and public key
  * cryptography - not supported yet by linux
-- 
2.1.0

