From ebaa6852e5381f88cead424478f3eea053dbe971 Mon Sep 17 00:00:00 2001
From: Nikos Mavrogiannopoulos <nmav@gnutls.org>
Date: Sun, 4 Jul 2010 10:47:51 +0200
Subject: [PATCH 070/291] cryptodev: More

---
 drivers/staging/crypto/cryptodev/.gitignore | 1 +
 1 file changed, 1 insertion(+)

diff --git a/drivers/staging/crypto/cryptodev/.gitignore b/drivers/staging/crypto/cryptodev/.gitignore
index 06eaf99189a7..8b6dce92f018 100644
--- a/drivers/staging/crypto/cryptodev/.gitignore
+++ b/drivers/staging/crypto/cryptodev/.gitignore
@@ -12,3 +12,4 @@ examples/hmac
 examples/speed
 releases
 scripts
+version.h
-- 
2.1.0

