From 5c8dff72e629c6e2dbae020eaf1e0d91feeca2e4 Mon Sep 17 00:00:00 2001
From: Phil Sutter <phil@nwl.cc>
Date: Mon, 12 Jul 2010 23:58:45 +0200
Subject: [PATCH 081/291] cryptodev: zero-copy:

---
 drivers/staging/crypto/cryptodev/cryptodev_main.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/drivers/staging/crypto/cryptodev/cryptodev_main.c b/drivers/staging/crypto/cryptodev/cryptodev_main.c
index 8b148a7c14ae..4f41bc51feca 100644
--- a/drivers/staging/crypto/cryptodev/cryptodev_main.c
+++ b/drivers/staging/crypto/cryptodev/cryptodev_main.c
@@ -532,10 +532,12 @@ static int get_userbuf(struct csession *ses,
 {
 	int src_pagecount, dst_pagecount = 0, pagecount, write_src = 1;
 
+	BUG_ON(!cop->src);
 	src_pagecount = PAGECOUNT(cop->src, cop->len);
 	if (!ses->cdata.init) {		/* hashing only */
 		write_src = 0;
 	} else if (cop->src != cop->dst) {	/* non-in-situ transformation */
+		BUG_ON(!cop->dst);
 		dst_pagecount = PAGECOUNT(cop->dst, cop->len);
 		write_src = 0;
 	}
-- 
2.1.0

