From 82095957acdd8aba166347be323f4c63d87903dc Mon Sep 17 00:00:00 2001
From: Nikos Mavrogiannopoulos <nmav@gnutls.org>
Date: Sat, 2 Oct 2010 14:19:32 +0200
Subject: [PATCH 090/291] cryptodev: avoid

---
 drivers/staging/crypto/cryptodev/Makefile | 6 ++++--
 1 file changed, 4 insertions(+), 2 deletions(-)

diff --git a/drivers/staging/crypto/cryptodev/Makefile b/drivers/staging/crypto/cryptodev/Makefile
index ff28fc440efe..e89f39416a43 100644
--- a/drivers/staging/crypto/cryptodev/Makefile
+++ b/drivers/staging/crypto/cryptodev/Makefile
@@ -5,10 +5,12 @@ cryptodev-objs = cryptodev_main.o cryptodev_cipher.o
 
 obj-m += cryptodev.o
 
-build:
-	@echo "#define VERSION \"$(VERSION)\"" > version.h
+build: version.h
 	make -C $(KERNEL_DIR) SUBDIRS=`pwd` modules
 
+version.h: Makefile
+	@echo "#define VERSION \"$(VERSION)\"" > version.h
+
 install:
 	make -C $(KERNEL_DIR) SUBDIRS=`pwd` modules_install
 	@echo "Installing cryptodev.h in /usr/include/crypto ..."
-- 
2.1.0

