From be6cfe45b29dad7ec2cd3be495fd51eaaf5239fd Mon Sep 17 00:00:00 2001
From: Nikos Mavrogiannopoulos <nmav@gnutls.org>
Date: Sat, 2 Oct 2010 14:33:20 +0200
Subject: [PATCH 091/291] cryptodev: DISABLE_ZCOPY

---
 drivers/staging/crypto/cryptodev/cryptodev_main.c | 8 --------
 1 file changed, 8 deletions(-)

diff --git a/drivers/staging/crypto/cryptodev/cryptodev_main.c b/drivers/staging/crypto/cryptodev/cryptodev_main.c
index a27da96fbd49..a912cbc63a74 100644
--- a/drivers/staging/crypto/cryptodev/cryptodev_main.c
+++ b/drivers/staging/crypto/cryptodev/cryptodev_main.c
@@ -507,8 +507,6 @@ __crypto_run_std(struct csession *ses_ptr, struct crypt_op *cop)
 	return ret;
 }
 
-#ifndef DISABLE_ZCOPY
-
 void release_user_pages(struct page **pg, int pagecount)
 {
 	while (pagecount--) {
@@ -638,8 +636,6 @@ __crypto_run_zc(struct csession *ses_ptr, struct crypt_op *cop)
 	return ret;
 }
 
-#endif /* DISABLE_ZCOPY */
-
 static int crypto_run(struct fcrypt *fcr, struct crypt_op *cop)
 {
 	struct csession *ses_ptr;
@@ -699,11 +695,7 @@ static int crypto_run(struct fcrypt *fcr, struct crypt_op *cop)
 	}
 
 	if (!(cop->flags & COP_FLAG_UPDATE) || cop->len != 0) {
-#ifdef DISABLE_ZCOPY
-		ret = __crypto_run_std(ses_ptr, cop);
-#else /* normal */
 		ret = __crypto_run_zc(ses_ptr, cop);
-#endif
 		if (unlikely(ret))
 			goto out_unlock;
 	}
-- 
2.1.0

