From 76cd1983c78fcb7ad065d90ba1830791d1acd4ae Mon Sep 17 00:00:00 2001
From: Nikos Mavrogiannopoulos <nmav@gnutls.org>
Date: Sat, 2 Oct 2010 14:37:53 +0200
Subject: [PATCH 092/291] cryptodev: Added

---
 drivers/staging/crypto/cryptodev/cryptodev.h | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/drivers/staging/crypto/cryptodev/cryptodev.h b/drivers/staging/crypto/cryptodev/cryptodev.h
index e2a8484df7cf..3620325f45cd 100644
--- a/drivers/staging/crypto/cryptodev/cryptodev.h
+++ b/drivers/staging/crypto/cryptodev/cryptodev.h
@@ -159,4 +159,8 @@ enum cryptodev_crk_op_t {
 #define CIOCKEY         _IOWR('c', 105, struct crypt_kop)
 #define CIOCASYMFEAT    _IOR('c', 106, __u32)
 
+/* to indicate that CRIOGET is not required in linux
+ */
+#define CRIOGET_NOT_NEEDED 1
+
 #endif /* L_CRYPTODEV_H */
-- 
2.1.0

