From 0c3b32812563cfcfcae083534d8130a10a94356f Mon Sep 17 00:00:00 2001
From: Nikos Mavrogiannopoulos <nmav@gnutls.org>
Date: Fri, 8 Oct 2010 00:27:34 +0200
Subject: [PATCH 096/291] cryptodev: corrected

---
 drivers/staging/crypto/cryptodev/NEWS | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/drivers/staging/crypto/cryptodev/NEWS b/drivers/staging/crypto/cryptodev/NEWS
index 298fefe74d31..c83a1dee829e 100644
--- a/drivers/staging/crypto/cryptodev/NEWS
+++ b/drivers/staging/crypto/cryptodev/NEWS
@@ -1,4 +1,4 @@
-Version 0.6 (released 2010-10-08)
+Version 0.7 (released 2010-10-08)
 
 * Added COP_FLAG_FINAL to make multi-update
 more efficient.
-- 
2.1.0

