From 0083824ba3b5206be8ff70293eb7da4a0dd87090 Mon Sep 17 00:00:00 2001
From: Nikos Mavrogiannopoulos <nmav@gnutls.org>
Date: Sun, 24 Oct 2010 13:41:26 +0200
Subject: [PATCH 105/291] cryptodev: removed

---
 drivers/staging/crypto/cryptodev/README | 5 -----
 1 file changed, 5 deletions(-)

diff --git a/drivers/staging/crypto/cryptodev/README b/drivers/staging/crypto/cryptodev/README
index 163780297011..b6191f7b4216 100644
--- a/drivers/staging/crypto/cryptodev/README
+++ b/drivers/staging/crypto/cryptodev/README
@@ -3,10 +3,5 @@ FreeBSD. The main idea is to access of existing ciphers in kernel space
 from userspace, thus enabling re-use of a hardware implementation of a
 cipher.
 
-Since version 0.4 /dev/crypto uses zero copy for operations between user-space
-and kernel space. We are not aware of any accelerators that this might cause
-a problem. If it does define DISABLE_ZCOPY in cryptodev_int.h recompile
-and let us know.
-
 For questions and suggestions please use the mailing lists at:
 http://home.gna.org/cryptodev-linux/lists.html
-- 
2.1.0

