From 9d1156486f2f0c54eaf6dd14b8359df3610cecbb Mon Sep 17 00:00:00 2001
From: Nikos Mavrogiannopoulos <nmav@gnutls.org>
Date: Sun, 24 Oct 2010 13:51:36 +0200
Subject: [PATCH 108/291] cryptodev: increased

---
 drivers/staging/crypto/cryptodev/Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/drivers/staging/crypto/cryptodev/Makefile b/drivers/staging/crypto/cryptodev/Makefile
index 721915b70215..d5bb6496ca8e 100644
--- a/drivers/staging/crypto/cryptodev/Makefile
+++ b/drivers/staging/crypto/cryptodev/Makefile
@@ -1,5 +1,5 @@
 KERNEL_DIR = /lib/modules/$(shell uname -r)/build
-VERSION = 0.7
+VERSION = 0.8
 
 cryptodev-objs = cryptodev_main.o cryptodev_cipher.o
 
-- 
2.1.0

