From 5e1858e0c94d382aa77f913576b2b530a27c2e64 Mon Sep 17 00:00:00 2001
From: Nikos Mavrogiannopoulos <nmav@gnutls.org>
Date: Sat, 6 Nov 2010 12:25:07 +0100
Subject: [PATCH 115/291] cryptodev: released

---
 drivers/staging/crypto/cryptodev/NEWS | 9 +++++++++
 1 file changed, 9 insertions(+)

diff --git a/drivers/staging/crypto/cryptodev/NEWS b/drivers/staging/crypto/cryptodev/NEWS
index c83a1dee829e..9f0783054115 100644
--- a/drivers/staging/crypto/cryptodev/NEWS
+++ b/drivers/staging/crypto/cryptodev/NEWS
@@ -1,3 +1,12 @@
+Version 0.8 (released 2010-11-06)
+
+* Made cryptodev aware of alignment constraints.
+
+* Added support for CRYPTO_AES_ECB.
+
+* Added asynchronous operation support using
+  CIOCASYNCCRYPT, CIOCASYNCFETCH ioctls and poll().
+
 Version 0.7 (released 2010-10-08)
 
 * Added COP_FLAG_FINAL to make multi-update
-- 
2.1.0

