From f4eeefb8239268608616e3315d9ff9bf6c228112 Mon Sep 17 00:00:00 2001
From: Phil Sutter <phil.sutter@viprinet.com>
Date: Tue, 18 Jan 2011 18:33:29 +0100
Subject: [PATCH 128/291] cryptodev: enable

Signed-off-by: Nikos Mavrogiannopoulos <nmav@gnutls.org>
---
 drivers/staging/crypto/cryptodev/cryptodev_main.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/drivers/staging/crypto/cryptodev/cryptodev_main.c b/drivers/staging/crypto/cryptodev/cryptodev_main.c
index cb81107c1195..0f93010a6ba1 100644
--- a/drivers/staging/crypto/cryptodev/cryptodev_main.c
+++ b/drivers/staging/crypto/cryptodev/cryptodev_main.c
@@ -1267,6 +1267,7 @@ cryptodev_compat_ioctl(struct file *file, unsigned int cmd, unsigned long arg_)
 	case CIOCASYMFEAT:
 	case CRIOGET:
 	case CIOCFSESSION:
+	case CIOCGSESSINFO:
 		return cryptodev_ioctl(file, cmd, arg_);
 
 	case COMPAT_CIOCGSESSION:
-- 
2.1.0

