From 7a226c89257a936321270958a3d49b40a37c0eeb Mon Sep 17 00:00:00 2001
From: Nikos Mavrogiannopoulos <nmav@gnutls.org>
Date: Sun, 23 Jan 2011 10:39:29 +0100
Subject: [PATCH 131/291] cryptodev: updated

---
 drivers/staging/crypto/cryptodev/NEWS | 16 ++++++++++++++++
 1 file changed, 16 insertions(+)

diff --git a/drivers/staging/crypto/cryptodev/NEWS b/drivers/staging/crypto/cryptodev/NEWS
index 9f0783054115..b90554ff5bd1 100644
--- a/drivers/staging/crypto/cryptodev/NEWS
+++ b/drivers/staging/crypto/cryptodev/NEWS
@@ -1,3 +1,19 @@
+Version 0.9 (unreleased)
+
+* Added additional test tools:
+  - sha_speed does performance testing of SHA1 and SHA256
+  - hashcrypt_speed additionally encrypts with AES128 and AES256
+
+* Allow updating the IV in userspace via the COP_FLAG_WRITE_IV
+flag.
+
+* Export the alignmask in an OCF compatible way.
+
+* Fix for kernel crash on passing incorrect session ID.
+
+* Added CIOCGSESSINFO to export additional information
+for each session.
+
 Version 0.8 (released 2010-11-06)
 
 * Made cryptodev aware of alignment constraints.
-- 
2.1.0

