From 015c11befcff4f99a479ea1fe96780615da763f3 Mon Sep 17 00:00:00 2001
From: Nikos Mavrogiannopoulos <nmav@gnutls.org>
Date: Tue, 12 Apr 2011 20:36:34 +0200
Subject: [PATCH 140/291] cryptodev: updated

---
 drivers/staging/crypto/cryptodev/NEWS | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/drivers/staging/crypto/cryptodev/NEWS b/drivers/staging/crypto/cryptodev/NEWS
index 2d78c23bd799..a97fa5ae43f1 100644
--- a/drivers/staging/crypto/cryptodev/NEWS
+++ b/drivers/staging/crypto/cryptodev/NEWS
@@ -1,3 +1,8 @@
+Version 1.0 (released 201-04-12)
+
+* Several fixes in the included examples. Based on
+patches by Vladimir Zapolskiy.
+
 Version 0.9 (released 2011-02-11)
 
 * Added additional test tools:
-- 
2.1.0

