From bcc2e05d02bfc0d734d8b40c972368121b0fae6c Mon Sep 17 00:00:00 2001
From: Nikos Mavrogiannopoulos <nmav@gnutls.org>
Date: Sun, 31 Jul 2011 03:04:14 +0200
Subject: [PATCH 141/291] cryptodev: Define

---
 drivers/staging/crypto/cryptodev/crypto/cryptodev.h | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/drivers/staging/crypto/cryptodev/crypto/cryptodev.h b/drivers/staging/crypto/cryptodev/crypto/cryptodev.h
index db4697c81669..f062f78a84cf 100644
--- a/drivers/staging/crypto/cryptodev/crypto/cryptodev.h
+++ b/drivers/staging/crypto/cryptodev/crypto/cryptodev.h
@@ -71,6 +71,8 @@ enum cryptodev_crypto_op_t {
 /* maximum length of verbose alg names (depends on CRYPTO_MAX_ALG_NAME) */
 #define CRYPTODEV_MAX_ALG_NAME		64
 
+#define HASH_MAX_LEN 64
+
 /* input of CIOCGSESSION */
 struct session_op {
 	/* Specify either cipher or mac
-- 
2.1.0

