From 3346fcb88b2c3d3e5e23c85e86b46b2a22eda0aa Mon Sep 17 00:00:00 2001
From: Nikos Mavrogiannopoulos <nmav@gnutls.org>
Date: Sat, 3 Dec 2011 19:22:04 +0100
Subject: [PATCH 155/291] cryptodev: documented

---
 drivers/staging/crypto/cryptodev/crypto/cryptodev.h | 16 ++++++++++++++++
 1 file changed, 16 insertions(+)

diff --git a/drivers/staging/crypto/cryptodev/crypto/cryptodev.h b/drivers/staging/crypto/cryptodev/crypto/cryptodev.h
index 5f46e4721d89..0491fbb93b52 100644
--- a/drivers/staging/crypto/cryptodev/crypto/cryptodev.h
+++ b/drivers/staging/crypto/cryptodev/crypto/cryptodev.h
@@ -144,6 +144,22 @@ struct crypt_auth_op {
 	__u32   iv_len;
 };
 
+/* In plain AEAD mode the following are required:
+ *  flags   : 0
+ *  iv      : the initialization vector (12 bytes)
+ *  auth_len: the length of the data to be authenticated
+ *  auth_src: the data to be authenticated
+ *  len     : length of data to be encrypted
+ *  src     : the data to be encrypted
+ *  dst     : space to hold encrypted data. It must have
+ *            at least a size of len + tag_size.
+ *  tag_size: the size of the desired authentication tag or zero to use
+ *            the maximum tag output.
+ *
+ * Note tag isn't being used because the Linux AEAD interface
+ * copies the tag just after data.
+ */
+
 /* In TLS mode the following are required:
  *  flags   : COP_FLAG_AEAD_TLS_TYPE
  *  iv      : the initialization vector
-- 
2.1.0

