From aa4d28f229911570737ca21dc1478657fbd26ad5 Mon Sep 17 00:00:00 2001
From: Nikos Mavrogiannopoulos <nmav@gnutls.org>
Date: Sun, 19 Feb 2012 12:32:33 +0100
Subject: [PATCH 173/291] cryptodev: bumped

---
 drivers/staging/crypto/cryptodev/Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/drivers/staging/crypto/cryptodev/Makefile b/drivers/staging/crypto/cryptodev/Makefile
index e0c005acb86b..b2548a4b879f 100644
--- a/drivers/staging/crypto/cryptodev/Makefile
+++ b/drivers/staging/crypto/cryptodev/Makefile
@@ -1,6 +1,6 @@
 KBUILD_CFLAGS += -I$(src)
 KERNEL_DIR = /lib/modules/$(shell uname -r)/build
-VERSION = 1.0
+VERSION = 1.1
 
 cryptodev-objs = ioctl.o main.o cryptlib.o authenc.o zc.o util.o
 
-- 
2.1.0

