From c207936dd029342494a9113077629e5e7c997516 Mon Sep 17 00:00:00 2001
From: Nikos Mavrogiannopoulos <nmav@gnutls.org>
Date: Mon, 20 Feb 2012 09:00:14 +0100
Subject: [PATCH 174/291] cryptodev: updated

---
 drivers/staging/crypto/cryptodev/NEWS | 10 ++++++++++
 1 file changed, 10 insertions(+)

diff --git a/drivers/staging/crypto/cryptodev/NEWS b/drivers/staging/crypto/cryptodev/NEWS
index 6384098f8f5f..483e68c0e108 100644
--- a/drivers/staging/crypto/cryptodev/NEWS
+++ b/drivers/staging/crypto/cryptodev/NEWS
@@ -4,6 +4,16 @@ Version 1.1 (unreleased)
 
 * Defined HASH_MAX_LEN is cryptodev.h
 
+* CIOCGSESSINFO ioctl() sets the SIOP_FLAG_KERNEL_DRIVER_ONLY 
+flag if the driver is only available through kernel
+driver (and is not just software cipher).
+
+* Added new encryption ioctl, CIOCAUTHCRYPT, which
+combines authentication and encryption. Operates
+in AEAD, TLS and SRTP modes (the API might change
+in later versions).
+
+
 Version 1.0 (released 201-04-12)
 
 * Several fixes in the included examples. Based on
-- 
2.1.0

