From 0c49c910b75d3619d50a4f4cf15d0bdb443c7364 Mon Sep 17 00:00:00 2001
From: Nikos Mavrogiannopoulos <nmav@gnutls.org>
Date: Mon, 20 Feb 2012 09:03:31 +0100
Subject: [PATCH 175/291] cryptodev: updated

---
 drivers/staging/crypto/cryptodev/Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/drivers/staging/crypto/cryptodev/Makefile b/drivers/staging/crypto/cryptodev/Makefile
index b2548a4b879f..1da4633cd77d 100644
--- a/drivers/staging/crypto/cryptodev/Makefile
+++ b/drivers/staging/crypto/cryptodev/Makefile
@@ -33,7 +33,7 @@ dist: clean
 	@echo Packing
 	@rm -f *.tar.gz
 	@mkdir $(TMPDIR)/$(FILEBASE)
-	@cp -ar crypto extras examples Makefile *.c *.h README NEWS \
+	@cp -ar crypto extras tests examples Makefile *.c *.h README NEWS \
 		AUTHORS COPYING $(TMPDIR)/$(FILEBASE)
 	@rm -rf $(TMPDIR)/$(FILEBASE)/.git* $(TMPDIR)/$(FILEBASE)/releases $(TMPDIR)/$(FILEBASE)/scripts
 	@tar -C /tmp -czf ./$(OUTPUT) $(FILEBASE)
-- 
2.1.0

