From e487b84b910a755110d0ffe34c072973634579c5 Mon Sep 17 00:00:00 2001
From: Nikos Mavrogiannopoulos <nmav@gnutls.org>
Date: Mon, 20 Feb 2012 09:05:55 +0100
Subject: [PATCH 176/291] cryptodev: updated

---
 drivers/staging/crypto/cryptodev/NEWS | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/drivers/staging/crypto/cryptodev/NEWS b/drivers/staging/crypto/cryptodev/NEWS
index 483e68c0e108..af35dd17d224 100644
--- a/drivers/staging/crypto/cryptodev/NEWS
+++ b/drivers/staging/crypto/cryptodev/NEWS
@@ -1,4 +1,4 @@
-Version 1.1 (unreleased)
+Version 1.1 (released 2012-02-20)
 
 * Fixed alignment issue in speed.c
 
@@ -14,7 +14,7 @@ in AEAD, TLS and SRTP modes (the API might change
 in later versions).
 
 
-Version 1.0 (released 201-04-12)
+Version 1.0 (released 2011-04-12)
 
 * Several fixes in the included examples. Based on
 patches by Vladimir Zapolskiy.
-- 
2.1.0

