From 6ad5af2c5fcc527dd100adc31c3e4ea3696ca98d Mon Sep 17 00:00:00 2001
From: Nikos Mavrogiannopoulos <nmav@gnutls.org>
Date: Mon, 20 Feb 2012 19:16:47 +0100
Subject: [PATCH 177/291] cryptodev: If

---
 drivers/staging/crypto/cryptodev/ioctl.c | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/drivers/staging/crypto/cryptodev/ioctl.c b/drivers/staging/crypto/cryptodev/ioctl.c
index 4330bceee7e6..b152500a7399 100644
--- a/drivers/staging/crypto/cryptodev/ioctl.c
+++ b/drivers/staging/crypto/cryptodev/ioctl.c
@@ -717,6 +717,8 @@ static int get_session_info(struct fcrypt *fcr, struct session_info_op *siop)
 #ifdef CRYPTO_ALG_KERN_DRIVER_ONLY
 		if (tfm->__crt_alg->cra_flags & CRYPTO_ALG_KERN_DRIVER_ONLY)
 			siop->flags |= SIOP_FLAG_KERNEL_DRIVER_ONLY;
+#else
+		siop->flags |= SIOP_FLAG_KERNEL_DRIVER_ONLY; /* set it by default */
 #endif
 	}
 	if (ses_ptr->hdata.init) {
@@ -725,6 +727,8 @@ static int get_session_info(struct fcrypt *fcr, struct session_info_op *siop)
 #ifdef CRYPTO_ALG_KERN_DRIVER_ONLY
 		if (tfm->__crt_alg->cra_flags & CRYPTO_ALG_KERN_DRIVER_ONLY)
 			siop->flags |= SIOP_FLAG_KERNEL_DRIVER_ONLY;
+#else
+		siop->flags |= SIOP_FLAG_KERNEL_DRIVER_ONLY; /* set it by default */
 #endif
 	}
 
-- 
2.1.0

