From 096ece5947c964fa37fef11d047e6f482dba45ee Mon Sep 17 00:00:00 2001
From: Nikos Mavrogiannopoulos <nmav@gnutls.org>
Date: Mon, 20 Feb 2012 19:42:12 +0100
Subject: [PATCH 178/291] cryptodev: updated

---
 drivers/staging/crypto/cryptodev/NEWS | 7 +++++++
 1 file changed, 7 insertions(+)

diff --git a/drivers/staging/crypto/cryptodev/NEWS b/drivers/staging/crypto/cryptodev/NEWS
index af35dd17d224..9d94d5d8adb7 100644
--- a/drivers/staging/crypto/cryptodev/NEWS
+++ b/drivers/staging/crypto/cryptodev/NEWS
@@ -1,3 +1,10 @@
+Version 1.2 (unreleased)
+
+* In kernels that do not distinguish between hw 
+accelerated ciphers or not set the SIOP_FLAG_KERNEL_DRIVER_ONLY
+flag unconditionally.
+
+
 Version 1.1 (released 2012-02-20)
 
 * Fixed alignment issue in speed.c
-- 
2.1.0

