From 3dad9605c9f6cb95ef80c627205759ac5c04bd19 Mon Sep 17 00:00:00 2001
From: Nikos Mavrogiannopoulos <nmav@gnutls.org>
Date: Fri, 24 Feb 2012 16:53:17 +0100
Subject: [PATCH 186/291] cryptodev: more

---
 drivers/staging/crypto/cryptodev/.gitignore | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/drivers/staging/crypto/cryptodev/.gitignore b/drivers/staging/crypto/cryptodev/.gitignore
index 894394ec7980..2728950eb757 100644
--- a/drivers/staging/crypto/cryptodev/.gitignore
+++ b/drivers/staging/crypto/cryptodev/.gitignore
@@ -22,3 +22,5 @@ tests/cipher-aead
 tests/fullspeed
 examples/aes
 lib/benchmark
+tests/cipher-aead-srtp
+tests/cipher-gcm
-- 
2.1.0

