From 7d48175751af33cb7b15e9725aceb38483b54835 Mon Sep 17 00:00:00 2001
From: Nikos Mavrogiannopoulos <nmav@gnutls.org>
Date: Fri, 24 Feb 2012 23:00:25 +0100
Subject: [PATCH 189/291] cryptodev: corrected

---
 drivers/staging/crypto/cryptodev/NEWS | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/drivers/staging/crypto/cryptodev/NEWS b/drivers/staging/crypto/cryptodev/NEWS
index d96d8aa4cb6b..b938c626e637 100644
--- a/drivers/staging/crypto/cryptodev/NEWS
+++ b/drivers/staging/crypto/cryptodev/NEWS
@@ -16,7 +16,7 @@ Version 1.1 (released 2012-02-20)
 
 * Fixed alignment issue in speed.c
 
-* Defined HASH_MAX_LEN is cryptodev.h
+* Defined HASH_MAX_LEN in cryptodev.h
 
 * CIOCGSESSINFO ioctl() sets the SIOP_FLAG_KERNEL_DRIVER_ONLY 
 flag if the driver is only available through kernel
-- 
2.1.0

