From 705e7c81046548dedfeca1e682dea183bde51e79 Mon Sep 17 00:00:00 2001
From: Nikos Mavrogiannopoulos <nmav@gnutls.org>
Date: Sun, 26 Feb 2012 21:22:09 +0100
Subject: [PATCH 194/291] cryptodev: documented

---
 drivers/staging/crypto/cryptodev/NEWS | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/drivers/staging/crypto/cryptodev/NEWS b/drivers/staging/crypto/cryptodev/NEWS
index d574d025e1b8..612dcb3828da 100644
--- a/drivers/staging/crypto/cryptodev/NEWS
+++ b/drivers/staging/crypto/cryptodev/NEWS
@@ -3,6 +3,9 @@ Version 1.3 (unreleased)
 * Return EBADMSG instead of ECANCELED on tag verification
 failure in authenc modes.
 
+* COP_FLAG_RESET can be combined with COP_FLAG_UPDATE for
+efficiency.
+
 
 Version 1.2 (released 2012-02-24)
 
-- 
2.1.0

