From 3a452e43f0ac54c6d799b03327657a8792ce6327 Mon Sep 17 00:00:00 2001
From: Nikos Mavrogiannopoulos <nmav@gnutls.org>
Date: Wed, 29 Feb 2012 23:28:19 +0100
Subject: [PATCH 200/291] cryptodev: documented

---
 drivers/staging/crypto/cryptodev/README | 14 ++++++++++++++
 1 file changed, 14 insertions(+)

diff --git a/drivers/staging/crypto/cryptodev/README b/drivers/staging/crypto/cryptodev/README
index b6191f7b4216..4ec3b8be46aa 100644
--- a/drivers/staging/crypto/cryptodev/README
+++ b/drivers/staging/crypto/cryptodev/README
@@ -5,3 +5,17 @@ cipher.
 
 For questions and suggestions please use the mailing lists at:
 http://home.gna.org/cryptodev-linux/lists.html
+
+
+=== How to combine with cryptographic libraries ===
+
+* GnuTLS: 
+
+GnuTLS needs to be compiled with --enable-cryptodev in order to take
+advantage of /dev/crypto. GnuTLS 3.0.14 or later is recommended.
+
+* OpenSSL:
+
+The current releases of openssl support /dev/crypto by applying the
+patches available in the extras subdirectory. In order to compile
+use the -DHAVE_CRYPTODEV -DUSE_CRYPTODEV_DIGESTS flags.
-- 
2.1.0

