From 642d7981ffc80a16bee5c1865b29dfa284a43e70 Mon Sep 17 00:00:00 2001
From: Jaren Johnston <jaren.johnston>
Date: Fri, 15 Jun 2012 08:44:43 +0200
Subject: [PATCH 213/291] cryptodev: Fix

Signed-off-by: Nikos Mavrogiannopoulos <nikos@esat.kuleuven.be>
---
 drivers/staging/crypto/cryptodev/zc.h | 9 +++++----
 1 file changed, 5 insertions(+), 4 deletions(-)

diff --git a/drivers/staging/crypto/cryptodev/zc.h b/drivers/staging/crypto/cryptodev/zc.h
index 757767733f86..a98e60d71a35 100644
--- a/drivers/staging/crypto/cryptodev/zc.h
+++ b/drivers/staging/crypto/cryptodev/zc.h
@@ -16,10 +16,11 @@ int get_userbuf(struct csession *ses,
                 struct scatterlist **src_sg, 
                 struct scatterlist **dst_sg);
 
-/* last page - first page + 1 */
-#define PAGECOUNT(buf, buflen) \
-	((((unsigned long)(buf + buflen - 1) & PAGE_MASK) >> PAGE_SHIFT) - \
-	(((unsigned long) buf               & PAGE_MASK) >> PAGE_SHIFT) + 1)
+/* buflen ? (last page - first page + 1) : 0 */
+#define PAGECOUNT(buf, buflen) ((buflen) \
+	? ((((unsigned long)(buf + buflen - 1)) >> PAGE_SHIFT) - \
+	   (((unsigned long)(buf             )) >> PAGE_SHIFT) + 1) \
+	: 0)
 
 #define DEFAULT_PREALLOC_PAGES 32
 
-- 
2.1.0

