From 1528afd8a97d5881fe96bbe637146d7b66c5c72d Mon Sep 17 00:00:00 2001
From: Nikos Mavrogiannopoulos <nmav@gnutls.org>
Date: Fri, 15 Jun 2012 15:51:26 +0200
Subject: [PATCH 215/291] cryptodev: documented

---
 drivers/staging/crypto/cryptodev/NEWS | 6 ++++++
 1 file changed, 6 insertions(+)

diff --git a/drivers/staging/crypto/cryptodev/NEWS b/drivers/staging/crypto/cryptodev/NEWS
index 64f02581f06a..3469b7ad9e20 100644
--- a/drivers/staging/crypto/cryptodev/NEWS
+++ b/drivers/staging/crypto/cryptodev/NEWS
@@ -1,3 +1,9 @@
+Version 1.5 (unreleased)
+
+* Allow empty plaintext and authenticated data in AEAD 
+ciphers. Patch by Jaren Johnston.
+
+
 Version 1.4 (released 2012-03-15)
 
 * Correctly report hw accelerated ciphers.
-- 
2.1.0

