From d7dd448dd1b6031bec9d9ba9e936487ecb1dcc53 Mon Sep 17 00:00:00 2001
From: Phil Sutter <phil.sutter@viprinet.com>
Date: Mon, 16 Jul 2012 15:10:13 +0200
Subject: [PATCH 222/291] cryptodev: cryptlib:

Signed-off-by: Phil Sutter <phil.sutter@viprinet.com>
---
 drivers/staging/crypto/cryptodev/cryptlib.h | 6 ++----
 1 file changed, 2 insertions(+), 4 deletions(-)

diff --git a/drivers/staging/crypto/cryptodev/cryptlib.h b/drivers/staging/crypto/cryptodev/cryptlib.h
index 729679cd266d..1745d0fe9888 100644
--- a/drivers/staging/crypto/cryptodev/cryptlib.h
+++ b/drivers/staging/crypto/cryptodev/cryptlib.h
@@ -36,10 +36,8 @@ ssize_t cryptodev_cipher_encrypt(struct cipher_data *cdata,
 inline static void cryptodev_cipher_auth(struct cipher_data *cdata,
 					 struct scatterlist *sg1, size_t len)
 {
-	if( len )
-		aead_request_set_assoc(cdata->async.arequest, sg1, len);
-	else /* for some reason we _have_ to call that */
-		aead_request_set_assoc(cdata->async.arequest, NULL, 0);
+	/* for some reason we _have_ to call that even for zero length sgs */
+	aead_request_set_assoc(cdata->async.arequest, len ? sg1 : NULL, len);
 }
 
 inline static void cryptodev_cipher_set_tag_size(struct cipher_data *cdata, int size)
-- 
2.1.0

