From 3a8d605f8dd77fac2c1273767353adb158459176 Mon Sep 17 00:00:00 2001
From: Nikos Mavrogiannopoulos <nmav@gnutls.org>
Date: Thu, 29 Nov 2012 11:38:08 +0100
Subject: [PATCH 227/291] cryptodev: install

---
 drivers/staging/crypto/cryptodev/Makefile | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/drivers/staging/crypto/cryptodev/Makefile b/drivers/staging/crypto/cryptodev/Makefile
index b36d68c5de45..702e231a78e6 100644
--- a/drivers/staging/crypto/cryptodev/Makefile
+++ b/drivers/staging/crypto/cryptodev/Makefile
@@ -13,6 +13,8 @@ build: version.h
 version.h: Makefile
 	@echo "#define VERSION \"$(VERSION)\"" > version.h
 
+install: modules_install
+
 modules_install:
 	make -C $(KERNEL_DIR) SUBDIRS=`pwd` modules_install
 	@echo "Installing cryptodev.h in $(PREFIX)/usr/include/crypto ..."
-- 
2.1.0

