From 7329042dd159ba2e8f6c2abb5fe3b8c7a5be11e0 Mon Sep 17 00:00:00 2001
From: Nikos Mavrogiannopoulos <nmav@gnutls.org>
Date: Tue, 15 Jan 2013 16:19:48 +0100
Subject: [PATCH 231/291] cryptodev: updated

---
 drivers/staging/crypto/cryptodev/NEWS | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/drivers/staging/crypto/cryptodev/NEWS b/drivers/staging/crypto/cryptodev/NEWS
index eaf78548daee..200b04e71c3a 100644
--- a/drivers/staging/crypto/cryptodev/NEWS
+++ b/drivers/staging/crypto/cryptodev/NEWS
@@ -3,7 +3,8 @@ Version 1.6 (unreleased)
 * Added modules_install target in Makefile
 
 * The included eng_cryptodev.c was modified to compile
-in OpenSSL 0.9.8.
+in OpenSSL 0.9.8, as well as, support the partial update
+in digests.
 
 
 Version 1.5 (released 2012-08-04)
-- 
2.1.0

