From f2ca686697de1d9d8a9da07768004e90f74ba64e Mon Sep 17 00:00:00 2001
From: Nikos Mavrogiannopoulos <nmav@gnutls.org>
Date: Fri, 25 Jan 2013 09:33:15 +0100
Subject: [PATCH 234/291] cryptodev: doc

---
 drivers/staging/crypto/cryptodev/NEWS | 4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

diff --git a/drivers/staging/crypto/cryptodev/NEWS b/drivers/staging/crypto/cryptodev/NEWS
index 200b04e71c3a..96ce681f9f1e 100644
--- a/drivers/staging/crypto/cryptodev/NEWS
+++ b/drivers/staging/crypto/cryptodev/NEWS
@@ -4,7 +4,9 @@ Version 1.6 (unreleased)
 
 * The included eng_cryptodev.c was modified to compile
 in OpenSSL 0.9.8, as well as, support the partial update
-in digests.
+in digests. Based on patch by Glenn Schmottlach.
+
+* Added SHA224. Patch by Yashpal Dutta.
 
 
 Version 1.5 (released 2012-08-04)
-- 
2.1.0

