From 0193887e0592e411b477130067fbf2dfef4424bf Mon Sep 17 00:00:00 2001
From: Nikos Mavrogiannopoulos <nikos@esat.kuleuven.be>
Date: Thu, 31 Jan 2013 10:14:00 +0100
Subject: [PATCH 235/291] cryptodev: updated

---
 drivers/staging/crypto/cryptodev/README | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/drivers/staging/crypto/cryptodev/README b/drivers/staging/crypto/cryptodev/README
index 2c18a8773097..1176e7029fbe 100644
--- a/drivers/staging/crypto/cryptodev/README
+++ b/drivers/staging/crypto/cryptodev/README
@@ -4,7 +4,7 @@ from userspace, thus enabling re-use of a hardware implementation of a
 cipher.
 
 For questions and suggestions please use the mailing lists at:
-http://home.gna.org/cryptodev-linux/lists.html
+http://cryptodev-linux.org/lists.html
 
 
 === How to combine with cryptographic libraries ===
-- 
2.1.0

