From 8e071898d75a6c3fa1b9745d3cdf1974294c62b1 Mon Sep 17 00:00:00 2001
From: Nikos Mavrogiannopoulos <nmav@gnutls.org>
Date: Sat, 2 Mar 2013 13:07:59 +0100
Subject: [PATCH 239/291] cryptodev: updated

---
 drivers/staging/crypto/cryptodev/NEWS | 4 ----
 1 file changed, 4 deletions(-)

diff --git a/drivers/staging/crypto/cryptodev/NEWS b/drivers/staging/crypto/cryptodev/NEWS
index 78962915a4fa..b04065f91667 100644
--- a/drivers/staging/crypto/cryptodev/NEWS
+++ b/drivers/staging/crypto/cryptodev/NEWS
@@ -5,10 +5,6 @@ Version 1.6 (unreleased)
 * Asynchronous operations will not be scheduled if zero
 copy is disabled.
 
-* The included eng_cryptodev.c was modified to compile
-in OpenSSL 0.9.8, as well as, support the partial update
-in digests. Based on patch by Glenn Schmottlach.
-
 * Added SHA224. Patch by Yashpal Dutta.
 
 
-- 
2.1.0

