From f6ebaae262f8ecf817cda7b3ee986d61a58ed64a Mon Sep 17 00:00:00 2001
From: Nikos Mavrogiannopoulos <nmav@gnutls.org>
Date: Tue, 23 Apr 2013 15:29:26 +0300
Subject: [PATCH 245/291] cryptodev: added

---
 drivers/staging/crypto/cryptodev/INSTALL | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/drivers/staging/crypto/cryptodev/INSTALL b/drivers/staging/crypto/cryptodev/INSTALL
index 7cc89eb78939..899309d30e8d 100644
--- a/drivers/staging/crypto/cryptodev/INSTALL
+++ b/drivers/staging/crypto/cryptodev/INSTALL
@@ -11,6 +11,8 @@ After that you should set your system to load the kernel module on system
 load. In most systems this can be done as:
 # echo cryptodev >>/etc/modules
 
+or in systemd-enabled systems:
+# echo "cryptodev" > /etc/modules-load.d/cryptodev.conf
 
 === Testing installation ===
 
-- 
2.1.0

