From d42091bc90272f316b5953f3dc509219290c24a0 Mon Sep 17 00:00:00 2001
From: Nikos Mavrogiannopoulos <nmav@gnutls.org>
Date: Tue, 7 May 2013 02:22:44 +0300
Subject: [PATCH 247/291] cryptodev: made

---
 drivers/staging/crypto/cryptodev/INSTALL | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/drivers/staging/crypto/cryptodev/INSTALL b/drivers/staging/crypto/cryptodev/INSTALL
index 1a93f9fd3aa1..18587510e4d6 100644
--- a/drivers/staging/crypto/cryptodev/INSTALL
+++ b/drivers/staging/crypto/cryptodev/INSTALL
@@ -9,7 +9,7 @@ and the latter installs the header files and the kernel module.
 
 After that you should set your system to load the kernel module on system
 load. In most systems this can be done as:
-# echo cryptodev >>/etc/modules
+# echo "cryptodev" >>/etc/modules
 
 or in systemd-enabled systems:
 # echo "cryptodev" > /etc/modules-load.d/cryptodev.conf
-- 
2.1.0

