From b35b1470716fc621a426e96f1c56425cf3268493 Mon Sep 17 00:00:00 2001
From: Cristian Stoica <cristian.stoica@freescale.com>
Date: Thu, 31 Oct 2013 11:56:58 +0200
Subject: [PATCH 254/291] cryptodev: fix

Signed-off-by: Cristian Stoica <cristian.stoica@freescale.com>
---
 drivers/staging/crypto/cryptodev/authenc.c | 3 +--
 1 file changed, 1 insertion(+), 2 deletions(-)

diff --git a/drivers/staging/crypto/cryptodev/authenc.c b/drivers/staging/crypto/cryptodev/authenc.c
index 8bff6776168c..74f9c9bbab47 100644
--- a/drivers/staging/crypto/cryptodev/authenc.c
+++ b/drivers/staging/crypto/cryptodev/authenc.c
@@ -655,8 +655,7 @@ __crypto_auth_run_zc(struct csession *ses_ptr, struct kernel_crypt_auth_op *kcao
 			int dst_len;
 
 			if (unlikely(ses_ptr->cdata.init == 0 ||
-					ses_ptr->cdata.stream == 0 ||
-					ses_ptr->cdata.aead == 0))
+					(ses_ptr->cdata.stream == 0 && ses_ptr->cdata.aead == 0)))
 			{
 				dprintk(0, KERN_ERR, "Only stream and AEAD ciphers are allowed for authenc\n");
 				ret = -EINVAL;
-- 
2.1.0

