From ede03794cabe049b69b8f9cedc4a537392f0de83 Mon Sep 17 00:00:00 2001
From: Cristian Stoica <cristian.stoica@freescale.com>
Date: Wed, 27 Nov 2013 18:44:16 +0200
Subject: [PATCH 257/291] cryptodev: fix

Signed-off-by: Cristian Stoica <cristian.stoica@freescale.com>
---
 drivers/staging/crypto/cryptodev/cryptlib.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/drivers/staging/crypto/cryptodev/cryptlib.c b/drivers/staging/crypto/cryptodev/cryptlib.c
index 44404b567f2e..54d5d410f09d 100644
--- a/drivers/staging/crypto/cryptodev/cryptlib.c
+++ b/drivers/staging/crypto/cryptodev/cryptlib.c
@@ -155,7 +155,7 @@ error:
 	} else {
 		if (out->async.arequest)
 			aead_request_free(out->async.arequest);
-		if (out->async.s)
+		if (out->async.as)
 			crypto_free_aead(out->async.as);
 	}
 	kfree(out->async.result);
-- 
2.1.0

