From 223f16255378fe5ada433d9ae6df9889c01b4682 Mon Sep 17 00:00:00 2001
From: Phil Sutter <phil@nwl.cc>
Date: Tue, 28 Jan 2014 04:23:06 +0100
Subject: [PATCH 264/291] cryptodev: Makefile:

Signed-off-by: Phil Sutter <phil@nwl.cc>
---
 drivers/staging/crypto/cryptodev/Makefile | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/drivers/staging/crypto/cryptodev/Makefile b/drivers/staging/crypto/cryptodev/Makefile
index 1e52c9f34861..302391121308 100644
--- a/drivers/staging/crypto/cryptodev/Makefile
+++ b/drivers/staging/crypto/cryptodev/Makefile
@@ -34,6 +34,9 @@ clean:
 check:
 	CFLAGS=$(CRYPTODEV_CFLAGS) KERNEL_DIR=$(KERNEL_DIR) make -C tests check
 
+checkpatch:
+	$(KERNEL_DIR)/scripts/checkpatch.pl --file *.c *.h
+
 FILEBASE = cryptodev-linux-$(VERSION)
 TMPDIR ?= /tmp
 OUTPUT = $(FILEBASE).tar.gz
-- 
2.1.0

