From e09c2b178a794b53f7087dc4535be8a2357d446e Mon Sep 17 00:00:00 2001
From: Phil Sutter <phil@nwl.cc>
Date: Wed, 29 Jan 2014 03:28:32 +0100
Subject: [PATCH 266/291] cryptodev: checkpatch

Signed-off-by: Phil Sutter <phil@nwl.cc>
---
 drivers/staging/crypto/cryptodev/authenc.c | 12 ++++++------
 1 file changed, 6 insertions(+), 6 deletions(-)

diff --git a/drivers/staging/crypto/cryptodev/authenc.c b/drivers/staging/crypto/cryptodev/authenc.c
index 9edcf07f221f..ee31de88e1f4 100644
--- a/drivers/staging/crypto/cryptodev/authenc.c
+++ b/drivers/staging/crypto/cryptodev/authenc.c
@@ -276,17 +276,17 @@ int kcaop_to_user(struct kernel_crypt_auth_op *kcaop,
 	return 0;
 }
 
-static void copy_tls_hash( struct scatterlist *dst_sg, int len, void *hash, int hash_len)
+static void copy_tls_hash(struct scatterlist *dst_sg, int len, void *hash, int hash_len)
 {
 	scatterwalk_map_and_copy(hash, dst_sg, len, hash_len, 1);
 }
 
-static void read_tls_hash( struct scatterlist *dst_sg, int len, void *hash, int hash_len)
+static void read_tls_hash(struct scatterlist *dst_sg, int len, void *hash, int hash_len)
 {
 	scatterwalk_map_and_copy(hash, dst_sg, len-hash_len, hash_len, 0);
 }
 
-static int pad_record( struct scatterlist *dst_sg, int len, int block_size)
+static int pad_record(struct scatterlist *dst_sg, int len, int block_size)
 {
 	uint8_t pad[block_size];
 	int pad_size = block_size - (len % block_size);
@@ -298,7 +298,7 @@ static int pad_record( struct scatterlist *dst_sg, int len, int block_size)
 	return pad_size;
 }
 
-static int verify_tls_record_pad( struct scatterlist *dst_sg, int len, int block_size)
+static int verify_tls_record_pad(struct scatterlist *dst_sg, int len, int block_size)
 {
 	uint8_t pad[256]; /* the maximum allowed */
 	uint8_t pad_size;
@@ -363,7 +363,7 @@ tls_auth_n_crypt(struct csession *ses_ptr, struct kernel_crypt_auth_op *kcaop,
 				return ret;
 			}
 
-			copy_tls_hash( dst_sg, len, hash_output, caop->tag_len);
+			copy_tls_hash(dst_sg, len, hash_output, caop->tag_len);
 			len += caop->tag_len;
 		}
 
@@ -407,7 +407,7 @@ tls_auth_n_crypt(struct csession *ses_ptr, struct kernel_crypt_auth_op *kcaop,
 				return -EINVAL;
 			}
 
-			read_tls_hash( dst_sg, len, vhash, caop->tag_len);
+			read_tls_hash(dst_sg, len, vhash, caop->tag_len);
 			len -= caop->tag_len;
 
 			if (auth_len > 0) {
-- 
2.1.0

