From 5a57a8242fa2cd4276048a1ca47635fb601e0f9d Mon Sep 17 00:00:00 2001
From: Phil Sutter <phil@nwl.cc>
Date: Wed, 29 Jan 2014 03:41:56 +0100
Subject: [PATCH 268/291] cryptodev: checkpatch

Signed-off-by: Phil Sutter <phil@nwl.cc>
---
 drivers/staging/crypto/cryptodev/authenc.c | 8 ++++----
 1 file changed, 4 insertions(+), 4 deletions(-)

diff --git a/drivers/staging/crypto/cryptodev/authenc.c b/drivers/staging/crypto/cryptodev/authenc.c
index 2c51199856f4..46e8d65918ee 100644
--- a/drivers/staging/crypto/cryptodev/authenc.c
+++ b/drivers/staging/crypto/cryptodev/authenc.c
@@ -595,8 +595,8 @@ __crypto_auth_run_zc(struct csession *ses_ptr, struct kernel_crypt_auth_op *kcao
 
 	if (caop->flags & COP_FLAG_AEAD_SRTP_TYPE) {
 		if (unlikely(ses_ptr->cdata.init != 0 &&
-			(ses_ptr->cdata.stream == 0 || ses_ptr->cdata.aead != 0)))
-		{
+		             (ses_ptr->cdata.stream == 0 ||
+			      ses_ptr->cdata.aead != 0))) {
 			dprintk(0, KERN_ERR, "Only stream modes are allowed in SRTP mode (but not AEAD)\n");
 			return -EINVAL;
 		}
@@ -655,8 +655,8 @@ __crypto_auth_run_zc(struct csession *ses_ptr, struct kernel_crypt_auth_op *kcao
 			int dst_len;
 
 			if (unlikely(ses_ptr->cdata.init == 0 ||
-					(ses_ptr->cdata.stream == 0 && ses_ptr->cdata.aead == 0)))
-			{
+			             (ses_ptr->cdata.stream == 0 &&
+				      ses_ptr->cdata.aead == 0))) {
 				dprintk(0, KERN_ERR, "Only stream and AEAD ciphers are allowed for authenc\n");
 				ret = -EINVAL;
 				goto free_auth_buf;
-- 
2.1.0

