From 450d8bf2ff6eb20c4b338ebfbec5df4c9d16754f Mon Sep 17 00:00:00 2001
From: Phil Sutter <phil@nwl.cc>
Date: Wed, 29 Jan 2014 03:44:03 +0100
Subject: [PATCH 269/291] cryptodev: checkpatch

Signed-off-by: Phil Sutter <phil@nwl.cc>
---
 drivers/staging/crypto/cryptodev/authenc.c | 6 ++++--
 1 file changed, 4 insertions(+), 2 deletions(-)

diff --git a/drivers/staging/crypto/cryptodev/authenc.c b/drivers/staging/crypto/cryptodev/authenc.c
index 46e8d65918ee..80a88c1af7ed 100644
--- a/drivers/staging/crypto/cryptodev/authenc.c
+++ b/drivers/staging/crypto/cryptodev/authenc.c
@@ -662,8 +662,10 @@ __crypto_auth_run_zc(struct csession *ses_ptr, struct kernel_crypt_auth_op *kcao
 				goto free_auth_buf;
 			}
 
-			if (caop->op == COP_ENCRYPT) dst_len = caop->len + cryptodev_cipher_get_tag_size(&ses_ptr->cdata);
-			else dst_len = caop->len;
+			if (caop->op == COP_ENCRYPT)
+				dst_len = caop->len + cryptodev_cipher_get_tag_size(&ses_ptr->cdata);
+			else
+				dst_len = caop->len;
 
 			ret = get_userbuf(ses_ptr, caop->src, caop->len, caop->dst, dst_len,
 					  kcaop->task, kcaop->mm, &src_sg, &dst_sg);
-- 
2.1.0

