From 11bd2c346ff4993e26b3a468a174b4759a5b4ea1 Mon Sep 17 00:00:00 2001
From: Phil Sutter <phil@nwl.cc>
Date: Wed, 29 Jan 2014 04:19:45 +0100
Subject: [PATCH 274/291] cryptodev: checkpatch

Signed-off-by: Phil Sutter <phil@nwl.cc>
---
 drivers/staging/crypto/cryptodev/zc.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/drivers/staging/crypto/cryptodev/zc.c b/drivers/staging/crypto/cryptodev/zc.c
index 6ca27ecd88df..3b986e7a3863 100644
--- a/drivers/staging/crypto/cryptodev/zc.c
+++ b/drivers/staging/crypto/cryptodev/zc.c
@@ -173,8 +173,8 @@ int get_userbuf(struct csession *ses,
 		return 0;
 	}
 
-	*src_sg = NULL; // default to no input
-	*dst_sg = NULL; // default to ignore output
+	*src_sg = NULL; /* default to no input */
+	*dst_sg = NULL; /* default to ignore output */
 
 	if (likely(src)) {
 		rc = __get_userbuf(src, src_len, 0, ses->readonly_pages,
-- 
2.1.0

