From a6938430bee232d4706a1d127c07b22b91ab4e88 Mon Sep 17 00:00:00 2001
From: Phil Sutter <phil@nwl.cc>
Date: Fri, 31 Jan 2014 02:33:05 +0100
Subject: [PATCH 279/291] cryptodev: allow

Signed-off-by: Phil Sutter <phil@nwl.cc>
---
 drivers/staging/crypto/cryptodev/Makefile | 10 +++++++++-
 1 file changed, 9 insertions(+), 1 deletion(-)

diff --git a/drivers/staging/crypto/cryptodev/Makefile b/drivers/staging/crypto/cryptodev/Makefile
index 302391121308..489c5f310000 100644
--- a/drivers/staging/crypto/cryptodev/Makefile
+++ b/drivers/staging/crypto/cryptodev/Makefile
@@ -34,8 +34,16 @@ clean:
 check:
 	CFLAGS=$(CRYPTODEV_CFLAGS) KERNEL_DIR=$(KERNEL_DIR) make -C tests check
 
+CPOPTS =
+ifneq (${SHOW_TYPES},)
+CPOPTS += --show-types
+endif
+ifneq (${IGNORE_TYPES},)
+CPOPTS += --ignore ${IGNORE_TYPES}
+endif
+
 checkpatch:
-	$(KERNEL_DIR)/scripts/checkpatch.pl --file *.c *.h
+	$(KERNEL_DIR)/scripts/checkpatch.pl ${CPOPTS} --file *.c *.h
 
 FILEBASE = cryptodev-linux-$(VERSION)
 TMPDIR ?= /tmp
-- 
2.1.0

