From 51fe9c7f7acd61d5f6824a325bb47233072c0c78 Mon Sep 17 00:00:00 2001
From: Phil Sutter <phil.sutter@viprinet.com>
Date: Mon, 24 Jan 2011 17:44:30 +0100
Subject: [PATCH 280/291] cryptodev: allow

Signed-off-by: Phil Sutter <phil@nwl.cc>
---
 drivers/staging/crypto/cryptodev/Makefile | 10 +++++++++-
 1 file changed, 9 insertions(+), 1 deletion(-)

diff --git a/drivers/staging/crypto/cryptodev/Makefile b/drivers/staging/crypto/cryptodev/Makefile
index 489c5f310000..c657c70366f4 100644
--- a/drivers/staging/crypto/cryptodev/Makefile
+++ b/drivers/staging/crypto/cryptodev/Makefile
@@ -13,8 +13,16 @@ cryptodev-objs = ioctl.o main.o cryptlib.o authenc.o zc.o util.o
 
 obj-m += cryptodev.o
 
+KERNEL_MAKE_OPTS := -C ${KERNEL_DIR} SUBDIRS=`pwd`
+ifneq (${ARCH},)
+KERNEL_MAKE_OPTS += ARCH=${ARCH}
+endif
+ifneq (${CROSS_COMPILE},)
+KERNEL_MAKE_OPTS += CROSS_COMPILE=${CROSS_COMPILE}
+endif
+
 build: version.h
-	make -C $(KERNEL_DIR) SUBDIRS=`pwd` modules
+	make ${KERNEL_MAKE_OPTS} modules
 
 version.h: Makefile
 	@echo "#define VERSION \"$(VERSION)\"" > version.h
-- 
2.1.0

