From b9551f570ecc1bc5d0c0990f05b72aa76f3e28ba Mon Sep 17 00:00:00 2001
From: Nikos Mavrogiannopoulos <nmav@redhat.com>
Date: Wed, 6 Aug 2014 12:00:21 +0200
Subject: [PATCH 291/291] cryptodev: updated

---
 drivers/staging/crypto/cryptodev/NEWS | 4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

diff --git a/drivers/staging/crypto/cryptodev/NEWS b/drivers/staging/crypto/cryptodev/NEWS
index 89176add454e..a0a7d3843c16 100644
--- a/drivers/staging/crypto/cryptodev/NEWS
+++ b/drivers/staging/crypto/cryptodev/NEWS
@@ -1,9 +1,11 @@
 Version 1.7 (unreleased)
 
-* Several bugfixes by Cristian Stoica.
+* Added support for composite AEAD keys by Cristian Stoica.
 
 * Added support for sysctl to modify verbosity by Nikolaos Tsakalakis.
 
+* Several bugfixes by Cristian Stoica.
+
 * When a driver requires aligned data but unaligned are provided, then
   zero copy is disabled to prevent driver failing to encrypt.
 
-- 
2.1.0

