From 911442c3c1b8dfcc6facad675040b69f211ab8a7 Mon Sep 17 00:00:00 2001
From: Nikos Mavrogiannopoulos <nmav@gnutls.org>
Date: Fri, 15 Jun 2012 15:51:26 +0200
Subject: [PATCH 215/291] documented updates

---
 drivers/staging/crypto/cryptodev/NEWS | 6 ++++++
 1 file changed, 6 insertions(+)

diff --git a/drivers/staging/crypto/cryptodev/NEWS b/drivers/staging/crypto/cryptodev/NEWS
index 64f0258..3469b7a 100644
--- a/drivers/staging/crypto/cryptodev/NEWS
+++ b/drivers/staging/crypto/cryptodev/NEWS
@@ -1,3 +1,9 @@
+Version 1.5 (unreleased)
+
+* Allow empty plaintext and authenticated data in AEAD 
+ciphers. Patch by Jaren Johnston.
+
+
 Version 1.4 (released 2012-03-15)
 
 * Correctly report hw accelerated ciphers.
-- 
2.0.2

