From ff8f17e1a56a0e31fab8ea65be7547a8ef6007a3 Mon Sep 17 00:00:00 2001
From: Cristian Stoica <cristian.stoica@freescale.com>
Date: Wed, 18 Sep 2013 14:09:02 +0300
Subject: [PATCH 251/292] fix incorrect debug message about hmac key length

Signed-off-by: Cristian Stoica <cristian.stoica@freescale.com>
---
 drivers/staging/crypto/cryptodev/ioctl.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/drivers/staging/crypto/cryptodev/ioctl.c b/drivers/staging/crypto/cryptodev/ioctl.c
index f84b79c81ed6..9cabb668bfaa 100644
--- a/drivers/staging/crypto/cryptodev/ioctl.c
+++ b/drivers/staging/crypto/cryptodev/ioctl.c
@@ -253,7 +253,7 @@ crypto_create_session(struct fcrypt *fcr, struct session_op *sop)
 		if (unlikely(sop->mackeylen > CRYPTO_HMAC_MAX_KEY_LEN)) {
 			dprintk(1, KERN_DEBUG,
 				"Setting key failed for %s-%zu.\n",
-				alg_name, (size_t)sop->mackeylen*8);
+				hash_name, (size_t)sop->mackeylen*8);
 			ret = -EINVAL;
 			goto error_hash;
 		}
-- 
1.8.1.2

