From 4cb51895bc5eeca176391c8f9b53eef7e039251a Mon Sep 17 00:00:00 2001
From: Nikos Mavrogiannopoulos <nmav@gnutls.org>
Date: Fri, 4 Dec 2009 01:27:22 +0200
Subject: [PATCH 006/292] License is GPL

---
 drivers/staging/crypto/cryptodev/cryptodev.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/drivers/staging/crypto/cryptodev/cryptodev.c b/drivers/staging/crypto/cryptodev/cryptodev.c
index f09a38d5aadb..53808043bb10 100644
--- a/drivers/staging/crypto/cryptodev/cryptodev.c
+++ b/drivers/staging/crypto/cryptodev/cryptodev.c
@@ -33,7 +33,7 @@
 
 MODULE_AUTHOR("Michal Ludvig <mludvig@logix.net.nz>");
 MODULE_DESCRIPTION("CryptoDev driver");
-MODULE_LICENSE("Dual BSD/GPL");
+MODULE_LICENSE("GPL");
 
 /* ====== Compile-time config ====== */
 
-- 
1.8.1.2

