From 308b903a8bcd43ee1ec6589384dffa568b63ce38 Mon Sep 17 00:00:00 2001
From: Nikos Mavrogiannopoulos <nmav@gnutls.org>
Date: Wed, 20 Mar 2013 21:50:41 +0100
Subject: [PATCH 241/292] released 1.6

---
 drivers/staging/crypto/cryptodev/NEWS | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/drivers/staging/crypto/cryptodev/NEWS b/drivers/staging/crypto/cryptodev/NEWS
index 9706613fa7b8..bd1b666ac902 100644
--- a/drivers/staging/crypto/cryptodev/NEWS
+++ b/drivers/staging/crypto/cryptodev/NEWS
@@ -1,4 +1,4 @@
-Version 1.6 (unreleased)
+Version 1.6 (released 2013-03-20)
 
 * Added modules_install target in Makefile
 
-- 
1.8.1.2

