From 9f2fa2fcedfe8861e2360e8013b66355ef3ef061 Mon Sep 17 00:00:00 2001
From: Nikos Mavrogiannopoulos <nmav@gnutls.org>
Date: Sun, 28 Aug 2011 13:59:42 +0200
Subject: [PATCH 144/291] updated

---
 drivers/staging/crypto/cryptodev/NEWS | 6 ++++++
 1 file changed, 6 insertions(+)

diff --git a/drivers/staging/crypto/cryptodev/NEWS b/drivers/staging/crypto/cryptodev/NEWS
index a97fa5a..6384098 100644
--- a/drivers/staging/crypto/cryptodev/NEWS
+++ b/drivers/staging/crypto/cryptodev/NEWS
@@ -1,3 +1,9 @@
+Version 1.1 (unreleased)
+
+* Fixed alignment issue in speed.c
+
+* Defined HASH_MAX_LEN is cryptodev.h
+
 Version 1.0 (released 201-04-12)
 
 * Several fixes in the included examples. Based on
-- 
2.0.2

