From 4e0969f7c89b084b7f5399a022300c5bbbe04e85 Mon Sep 17 00:00:00 2001
From: zumeng chen <zchen@pek-cc-pb02l.(none)>
Date: Thu, 19 Jun 2008 16:17:59 +0800
Subject: [PATCH] Ltt init kills bcm91x80 boot

Because HAVE_LTT_CLOCK is selected by "CONFIG MIPS"
so ltt will use include/asm-mips/ltt.h instead of
include/linux/ltt.h; and mips_hpt_frequency will be
involved by the function of ltt_frequency, it should
be initialized in time.c by sub-arch of mips.

But mips_hpt_frequency is not initialized by sibyte
in the file of arch/mips/sibyte/bcm1480/time.c, so
system will raise a instruction exception because of
dividing something by zero in ltt_init_synthetic_tsc.

This patch addes initialization of mips_hpt_frequency
into bcm1480/time.c, the boot process looks pass.
---
 arch/mips/sibyte/bcm1480/time.c |   12 ++++++++++++
 1 files changed, 12 insertions(+), 0 deletions(-)

diff --git a/arch/mips/sibyte/bcm1480/time.c b/arch/mips/sibyte/bcm1480/time.c
index 1680a68..5439018 100644
--- a/arch/mips/sibyte/bcm1480/time.c
+++ b/arch/mips/sibyte/bcm1480/time.c
@@ -17,6 +17,13 @@
  */
 #include <linux/init.h>
 
+#ifdef CONFIG_SIMULATION
+#define BCM1480_HPT_VALUE       50000
+#else
+#define BCM1480_HPT_VALUE       1000000
+#endif
+
+extern unsigned int mips_hpt_frequency;
 extern void sb1480_clockevent_init(void);
 extern void sb1480_clocksource_init(void);
 
@@ -24,4 +31,9 @@ void __init plat_time_init(void)
 {
 	sb1480_clocksource_init();
 	sb1480_clockevent_init();
+
+	/* To initialize mips_hpt_frequency used by ltt_frequency
+	 * otherwise there is div/zero when boot. 
+	 */
+	mips_hpt_frequency = BCM1480_HPT_VALUE;
 }
-- 
1.5.5.1

