From cd961e6eec5198777735e1eaad14db4babfcda44 Mon Sep 17 00:00:00 2001
From: Bruce Ashfield <bruce.ashfield@windriver.com>
Date: Wed, 9 Jul 2008 15:24:25 -0400
Subject: [PATCH] temporary build fixes for lttng

The manual uprev of lttng from 2.6.25 -> 2.6.26 missed a few
interfaces, making some of lttng unusable. This temporarily
fixes the build errors until the full package is officially
uprev'd.

 commit cc216c5d429892872f70f76975e243aef7ad9db1 required
 an update to ltt-tracer.c

    and

 ltt-statedump.c required the inclusion of fdtable.h

Signed-off-by: Bruce Ashfield <bruce.ashfield@windriver.com>
---
 ltt/ltt-statedump.c |    1 +
 1 files changed, 1 insertions(+), 0 deletions(-)

diff --git a/ltt/ltt-statedump.c b/ltt/ltt-statedump.c
index 0a57b77..0069834 100644
--- a/ltt/ltt-statedump.c
+++ b/ltt/ltt-statedump.c
@@ -26,6 +26,7 @@
 #include <linux/inetdevice.h>
 #include <linux/sched.h>
 #include <linux/mm.h>
+#include <linux/fdtable.h>
 #include <linux/marker.h>
 #include <linux/fdtable.h>
 
-- 
1.5.5.1

