From 7fb1b13d9e104762083873b299e6f673120f3c68 Mon Sep 17 00:00:00 2001
From: Mathieu Desnoyers <mathieu.desnoyers@polymtl.ca>
Date: Fri, 3 Oct 2008 13:43:00 -0400
Subject: [PATCH] [PATCH] check_compact_channel

From 81a4da3cdf77333ab0f25b1527c0e464bb0f0c74 Mon Sep 17 00:00:00 2001
Subject: [PATCH] check_compact_channel
---
 ltt/ltt-marker-control.c |    9 +++++++--
 1 files changed, 7 insertions(+), 2 deletions(-)

diff --git a/ltt/ltt-marker-control.c b/ltt/ltt-marker-control.c
index bcefddc..8868b83 100644
--- a/ltt/ltt-marker-control.c
+++ b/ltt/ltt-marker-control.c
@@ -332,9 +332,14 @@ int ltt_marker_connect(const char *mname, const char *pname,
 	/*
 	 * ID has priority over channel in case of conflict.
 	 */
-	if (id == MARKER_ID_COMPACT)
+	if (id == MARKER_ID_COMPACT) {
+#ifdef CONFIG_LTT_HEARTBEAT
 		pdata->channel = get_channel_index_from_name("compact");
-	else if (pdata->channel == get_channel_index_from_name("compact"))
+#else
+		ret = -EINVAL;
+		goto end;
+#endif
+	} else if (pdata->channel == get_channel_index_from_name("compact"))
 		pdata->channel = get_channel_index_from_name(DEFAULT_CHANNEL);
 	ret = marker_probe_register(mname, NULL, probe->probe_func, pdata);
 	if (ret)
-- 
1.5.5.1

