From a8fccc7110832398e7ceffb1a04403e37166adc6 Mon Sep 17 00:00:00 2001
From: Mathieu Desnoyers <mathieu.desnoyers@polymtl.ca>
Date: Tue, 12 Aug 2008 13:05:40 -0400
Subject: [PATCH] [PATCH] immediate-values-optimized-var-size-check

From b8c066d5d598c2afcc609d02fc69e45532940a35 Mon Sep 17 00:00:00 2001
Subject: [PATCH] immediate-values-optimized-var-size-check
---
 include/asm-x86/immediate.h |   14 ++++++++++----
 1 files changed, 10 insertions(+), 4 deletions(-)

diff --git a/include/asm-x86/immediate.h b/include/asm-x86/immediate.h
index 86de7bf..bc1251d 100644
--- a/include/asm-x86/immediate.h
+++ b/include/asm-x86/immediate.h
@@ -130,14 +130,20 @@ struct __imv {
 	({								\
 		uint32_t value;						\
 		BUILD_BUG_ON(sizeof(__typeof__(name##__imv)) > 1);	\
-		asm (".section __imv,\"aw\",@progbits\n\t"		\
-			_ASM_PTR "%c1, (3f)-4\n\t"			\
-			".byte 0, 5\n\t"				\
+		asm (".section __discard,\"\",@progbits\n\t"		\
+			"1:\n\t"					\
+			"mov $0,%0\n\t"					\
+			"2:\n\t"					\
+			".previous\n\t"					\
+			".section __imv,\"aw\",@progbits\n\t"		\
+			_ASM_PTR "%c1, (3f)-%c2\n\t"			\
+			".byte 0, (2b)-(1b)\n\t"			\
 			".previous\n\t"					\
 			"mov $0,%0\n\t"					\
 			"3:\n\t"					\
 			: "=a" (value)					\
-			: "i" (&name##__imv));				\
+			: "i" (&name##__imv),				\
+			  "i" (sizeof(value)));				\
 		value;							\
 	})
 
-- 
1.5.5.1

