From 4c84039ec4d993505dffd14d5a2420ecc709d944 Mon Sep 17 00:00:00 2001
From: Mathieu Desnoyers <mathieu.desnoyers@polymtl.ca>
Date: Thu, 30 Oct 2008 23:26:56 -0400
Subject: [PATCH] lttng-tracer-fix-leak

LTTng tracer fix leak

ltt_traces ltt_traces.head is saved to tmp, and is empty after the following
code:
    INIT_LIST_HEAD(&ltt_traces.head);
So remove the tracers will take no effect.

This patch fix to remove tracers used tmp.

Signed-off-by: Wei Yongjun <yjwei@cn.fujitsu.com>
Acked-by: Mathieu Desnoyers <mathieu.desnoyers@polymtl.ca>
---
 ltt/ltt-tracer.c |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/ltt/ltt-tracer.c b/ltt/ltt-tracer.c
index 98d4db1..68bfa35 100644
--- a/ltt/ltt-tracer.c
+++ b/ltt/ltt-tracer.c
@@ -793,7 +793,7 @@ static void __exit ltt_exit(void)
 	synchronize_sched();
 	/* Safe iteration is now permitted. It does not have to be RCU-safe
 	 * because no readers are left. */
-	list_for_each_safe(pos, n, &ltt_traces.head) {
+	list_for_each_safe(pos, n, &tmp) {
 		trace = container_of(pos, struct ltt_trace_struct, list);
 		/* _ltt_trace_destroy does a synchronize_sched() */
 		_ltt_trace_destroy(trace);
-- 
1.5.5.1

