From 3a15779663b8011360ce2a1ad0d8d50afa0bd888 Mon Sep 17 00:00:00 2001
From: Mathieu Desnoyers <mathieu.desnoyers@polymtl.ca>
Date: Fri, 3 Oct 2008 13:48:46 -0400
Subject: [PATCH] [PATCH] markers-use-imv-jump-liveliness

From 92cb06cc9e2e5e5b79853f055addd4ae16e7adc6 Mon Sep 17 00:00:00 2001
Subject: [PATCH] markers-use-imv-jump-liveliness
---
 include/linux/marker.h |    5 ++++-
 1 files changed, 4 insertions(+), 1 deletions(-)

diff --git a/include/linux/marker.h b/include/linux/marker.h
index 954871a..88761d5 100644
--- a/include/linux/marker.h
+++ b/include/linux/marker.h
@@ -76,10 +76,13 @@ struct marker {
 		{ __mark_empty_function, NULL}, NULL, NULL, NULL };	\
 		__mark_check_format(format, ## args);			\
 		if (!generic) {						\
-			if (unlikely(imv_cond(__mark_##name.state)))	\
+			if (unlikely(imv_cond(__mark_##name.state))) {	\
+				imv_cond_end();				\
 				(*__mark_##name.call)			\
 					(&__mark_##name, call_private,	\
 					## args);			\
+			} else						\
+				imv_cond_end();				\
 		} else {						\
 			if (unlikely(_imv_read(__mark_##name.state)))	\
 				(*__mark_##name.call)			\
-- 
1.5.5.1

