From 6e7e76a4d587ca266fb9fef45f65f3c831b38882 Mon Sep 17 00:00:00 2001
From: Mathieu Desnoyers <mathieu.desnoyers@polymtl.ca>
Date: Tue, 12 Aug 2008 10:33:00 -0400
Subject: [PATCH] ltt-relay-cleanup-code

LTT relay cleanup code

Remove extra tabs in ltt-relay.c.

Signed-off-by: Mathieu Desnoyers <mathieu.desnoyers@polymtl.ca>
---
 ltt/ltt-relay.c |   24 ++++++++++++------------
 1 files changed, 12 insertions(+), 12 deletions(-)

diff --git a/ltt/ltt-relay.c b/ltt/ltt-relay.c
index 4b3f80b..62eeeee 100644
--- a/ltt/ltt-relay.c
+++ b/ltt/ltt-relay.c
@@ -903,18 +903,18 @@ static inline void ltt_reserve_switch_old_subbuf(
 		struct rchan_buf *buf,
 		struct ltt_reserve_switch_offsets *offsets, u64 *tsc)
 {
-		ltt_channel->buffer_end(buf, *tsc, offsets->old,
-			SUBBUF_INDEX((offsets->old-1), buf->chan));
-		/* Must write buffer end before incrementing commit count */
-		smp_wmb();
-		offsets->commit_count =
-			local_add_return(buf->chan->subbuf_size
-				- (SUBBUF_OFFSET(offsets->old-1, buf->chan)+1),
-				&ltt_buf->commit_count[SUBBUF_INDEX(
-						offsets->old-1, buf->chan)]);
-		if (SUBBUF_OFFSET(offsets->commit_count, buf->chan) == 0)
-			ltt_deliver(buf, SUBBUF_INDEX((offsets->old-1),
-						buf->chan), NULL);
+	ltt_channel->buffer_end(buf, *tsc, offsets->old,
+		SUBBUF_INDEX((offsets->old-1), buf->chan));
+	/* Must write buffer end before incrementing commit count */
+	smp_wmb();
+	offsets->commit_count =
+		local_add_return(buf->chan->subbuf_size
+			- (SUBBUF_OFFSET(offsets->old-1, buf->chan)+1),
+			&ltt_buf->commit_count[SUBBUF_INDEX(
+					offsets->old-1, buf->chan)]);
+	if (SUBBUF_OFFSET(offsets->commit_count, buf->chan) == 0)
+		ltt_deliver(buf, SUBBUF_INDEX((offsets->old-1),
+					buf->chan), NULL);
 }
 
 /*
-- 
1.5.5.1

