From b94132fb2641eb77ecec7339838b4b026a7e3856 Mon Sep 17 00:00:00 2001
From: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
Date: Wed, 18 May 2011 18:42:20 -0400
Subject: [PATCH 139/248] tracepoints-improve-documentation-sample

After reading the tracepoints.txt documentation, I tried to
follow the instructions to run the sample, but just assigning
CONFIG_SAMPLE_TRACEPOINTS=m wasn't enough to generate the tracepoints
samples. I also need to set CONFIG_SAMPLES=y, otherwise nothing
happens.

So, this silly patch explicit says that CONFIG_SAMPLES should be set to
compile those files.

Signed-off-by: Breno Leitao <leitao@linux.vnet.ibm.com>
Signed-off-by: Mathieu Desnoyers <mathieu.desnoyers@polymtl.ca>
---
 Documentation/trace/tracepoints.txt |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/Documentation/trace/tracepoints.txt b/Documentation/trace/tracepoints.txt
index c0e1cee..d380250 100644
--- a/Documentation/trace/tracepoints.txt
+++ b/Documentation/trace/tracepoints.txt
@@ -106,7 +106,7 @@ used to export the defined tracepoints.
 See the example provided in samples/tracepoints
 
 Compile them with your kernel.  They are built during 'make' (not
-'make modules') when CONFIG_SAMPLE_TRACEPOINTS=m.
+'make modules') when CONFIG_SAMPLE=y and CONFIG_SAMPLE_TRACEPOINTS=m.
 
 Run, as root :
 modprobe tracepoint-sample (insmod order is not important)
-- 
1.7.0.4

