From fae8d985b0ec32e890727aa3565bbf664d7bbd6d Mon Sep 17 00:00:00 2001
From: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
Date: Thu, 13 May 2010 19:26:29 -0400
Subject: [PATCH 141/391] lttng-instrumentation/lttng-instrumentation-scheduler-arch

LTTng instrumentation scheduler arch

Scheduler instrumentation tracepoints probes, arch-specific
(sched_kthread_create).

Signed-off-by: Mathieu Desnoyers <mathieu.desnoyers@polymtl.ca>
---
 include/trace/sched.h |   11 +++++++++++
 1 files changed, 11 insertions(+), 0 deletions(-)
 create mode 100644 include/trace/sched.h

diff --git a/include/trace/sched.h b/include/trace/sched.h
new file mode 100644
index 0000000..a4b0307
--- /dev/null
+++ b/include/trace/sched.h
@@ -0,0 +1,11 @@
+#ifndef _LTTNG_TRACE_SCHED_H
+#define _LTTNG_TRACE_SCHED_H
+
+#include <linux/sched.h>
+#include <linux/tracepoint.h>
+
+DECLARE_TRACE(sched_kthread_create,
+	TP_PROTO(void *fn, int pid),
+		TP_ARGS(fn, pid));
+
+#endif /* _LTTNG_TRACE_SCHED_H */
-- 
1.6.5.2

