From 2eadf9d498e478db8c649895e1644d489ce396cf Mon Sep 17 00:00:00 2001
From: Mathieu Desnoyers <mathieu.desnoyers@polymtl.ca>
Date: Tue, 12 Aug 2008 13:06:05 -0400
Subject: [PATCH] [PATCH] restore-x86-nr-syscalls

From b9888c12bd697148e2d9f312dc0d6280c58249b2 Mon Sep 17 00:00:00 2001
Subject: [PATCH] restore-x86-nr-syscalls
---
 include/asm-x86/unistd_32.h |    2 ++
 1 files changed, 2 insertions(+), 0 deletions(-)

diff --git a/include/asm-x86/unistd_32.h b/include/asm-x86/unistd_32.h
index d739467..80864ab 100644
--- a/include/asm-x86/unistd_32.h
+++ b/include/asm-x86/unistd_32.h
@@ -339,6 +339,8 @@
 #define __NR_pipe2		331
 #define __NR_inotify_init1	332
 
+#define NR_syscalls		333
+
 #ifdef __KERNEL__
 
 #define __ARCH_WANT_IPC_PARSE_VERSION
-- 
1.5.5.1

