From 01c96c1c23f1335aeb5e2f89252602d60abfe5fc Mon Sep 17 00:00:00 2001
From: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
Date: Thu, 13 May 2010 19:26:38 -0400
Subject: [PATCH 153/391] lttng-statedump/lttng-statedump-kprobes

---
 ltt/ltt-statedump.c |    1 +
 1 files changed, 1 insertions(+), 0 deletions(-)

diff --git a/ltt/ltt-statedump.c b/ltt/ltt-statedump.c
index 2402d4a..1dfc0fb 100644
--- a/ltt/ltt-statedump.c
+++ b/ltt/ltt-statedump.c
@@ -339,6 +339,7 @@ static int do_ltt_statedump(struct ltt_probe_private_data *call_data)
 	list_interrupts(call_data);
 	ltt_enumerate_network_ip_interface(call_data);
 	ltt_dump_swap_files(call_data);
+	ltt_dump_kprobes_table(call_data);
 
 	/*
 	 * Fire off a work queue on each CPU. Their sole purpose in life
-- 
1.6.5.2

