From e0433379481ae1ef5d11f6da901f93eaf04abd8d Mon Sep 17 00:00:00 2001
From: Mathieu Desnoyers <mathieu.desnoyers@polymtl.ca>
Date: Wed, 8 Oct 2008 23:52:59 -0400
Subject: [PATCH] [PATCH] restore-x86-nr-syscalls

From faff6b5fbed70fc078606783e1d4ccbc618432c9 Mon Sep 17 00:00:00 2001
Subject: [PATCH] restore-x86-nr-syscalls
---
 include/asm-x86/unistd_32.h |    2 ++
 1 files changed, 2 insertions(+), 0 deletions(-)

diff --git a/include/asm-x86/unistd_32.h b/include/asm-x86/unistd_32.h
index d739467..80864ab 100644
--- a/include/asm-x86/unistd_32.h
+++ b/include/asm-x86/unistd_32.h
@@ -339,6 +339,8 @@
 #define __NR_pipe2		331
 #define __NR_inotify_init1	332
 
+#define NR_syscalls		333
+
 #ifdef __KERNEL__
 
 #define __ARCH_WANT_IPC_PARSE_VERSION
-- 
1.5.5.1

