From 4d28cbe16465e1ea28d22c87a977d825723fddca Mon Sep 17 00:00:00 2001
From: Mathieu Desnoyers <mathieu.desnoyers@polymtl.ca>
Date: Thu, 30 Oct 2008 23:27:25 -0400
Subject: [PATCH] lttng-instrumentation-scheduler-arch-tracepoints-probes

LTTng instrumentation sched arch tracepoint probe

sched_kthread_create tracepoints probe

Signed-off-by: Mathieu Desnoyers <mathieu.desnoyers@polymtl.ca>
---
 ltt/probes/kernel-trace.c |    7 +++++++
 1 files changed, 7 insertions(+), 0 deletions(-)

diff --git a/ltt/probes/kernel-trace.c b/ltt/probes/kernel-trace.c
index 27d2b8f..b0e0f16 100644
--- a/ltt/probes/kernel-trace.c
+++ b/ltt/probes/kernel-trace.c
@@ -160,6 +160,13 @@ void probe_sched_process_fork(struct task_struct *parent,
 		parent->pid, child->pid, child->tgid);
 }
 
+void probe_sched_kthread_create(void *fn, int pid)
+{
+	trace_mark_tp(kernel_kthread_create, sched_kthread_create,
+		probe_sched_kthread_create,
+		"fn %p pid %d", fn, pid);
+}
+
 void probe_timer_itimer_expired(struct signal_struct *sig)
 {
 	trace_mark_tp(kernel_timer_itimer_expired, timer_itimer_expired,
-- 
1.5.5.1

