From 49bef06669717db57015bde4442e5fdc40a1bd86 Mon Sep 17 00:00:00 2001
From: Mathieu Desnoyers <mathieu.desnoyers@polymtl.ca>
Date: Thu, 30 Oct 2008 23:27:29 -0400
Subject: [PATCH] lttng-statedump-kconfig-menus

LTTng statedump Kconfig mens

Add LTTng statedump support kconfig menu.

Signed-off-by: Mathieu Desnoyers <mathieu.desnoyers@polymtl.ca>
---
 ltt/Kconfig |   11 +++++++++++
 1 files changed, 11 insertions(+), 0 deletions(-)

diff --git a/ltt/Kconfig b/ltt/Kconfig
index 6cf9a7b..a17f3b5 100644
--- a/ltt/Kconfig
+++ b/ltt/Kconfig
@@ -147,4 +147,15 @@ config LTT_NETLINK_CONTROL
 	  If you enable this option, the Linux Trace Toolkit Netlink Controller
 	  will be either built in the kernel or as module.
 
+config LTT_STATEDUMP
+	tristate "Linux Trace Toolkit State Dump"
+	depends on LTT_TRACER
+	default m
+	help
+	  If you enable this option, the Linux Trace Toolkit State Dump will
+	  be either built in the kernel or as module.
+
+	  This module saves the state of the running kernel at trace start
+	  into the trace buffers along with the ongoing tracing information.
+
 endmenu
-- 
1.5.5.1

