From e867c14f4812e2d164b83f9242901aa172f7f0a5 Mon Sep 17 00:00:00 2001
From: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
Date: Wed, 18 May 2011 18:42:44 -0400
Subject: [PATCH 196/248] markers-modularize-lttng

---
 kernel/marker.c |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/kernel/marker.c b/kernel/marker.c
index 56b51b7..d22bbc6 100644
--- a/kernel/marker.c
+++ b/kernel/marker.c
@@ -26,7 +26,7 @@
 #include <linux/err.h>
 #include <linux/slab.h>
 #include <linux/immediate.h>
-#include <linux/ltt-tracer.h>
+#include <linux/ltt-channels.h>
 
 extern struct marker __start___markers[];
 extern struct marker __stop___markers[];
-- 
1.7.0.4

