From 5b3dcf66bf5ee54841d3a99c7b3d1598c64eba0b Mon Sep 17 00:00:00 2001
From: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
Date: Thu, 13 May 2010 19:27:02 -0400
Subject: [PATCH 199/391] memcpy-relay-fix

Signed-off-by: Mathieu Desnoyers <mathieu.desnoyers@polymtl.ca>
---
 include/linux/ltt-relay.h |   16 ++++++++++------
 1 files changed, 10 insertions(+), 6 deletions(-)

diff --git a/include/linux/ltt-relay.h b/include/linux/ltt-relay.h
index 52e8968..3781cbf 100644
--- a/include/linux/ltt-relay.h
+++ b/include/linux/ltt-relay.h
@@ -270,13 +270,17 @@ static inline void ltt_relay_do_copy(void *dest, const void *src, size_t len)
 		*(u64 *)dest = *(const u64 *)src;
 		break;
 	default:
-		/*
-		 * What we really want here is an inline memcpy, but we don't
-		 * have constants, so gcc generally uses a function call.
-		 */
-		for (; len > 0; len--)
-			*(u8 *)dest++ = *(const u8 *)src++;
+		goto memcpy_fallback;
 	}
+	return;
+
+memcpy_fallback:
+	/*
+	 * What we really want here is an inline memcpy, but we don't
+	 * have constants, so gcc generally uses a function call.
+	 */
+	for (; len > 0; len--)
+		*(u8 *)dest++ = *(const u8 *)src++;
 }
 #endif
 
-- 
1.6.5.2

