From bc16b9ae741b5cc7cb061160a7f74974bcb1bd08 Mon Sep 17 00:00:00 2001
From: Bruce Ashfield <bruce.ashfield@windriver.com>
Date: Wed, 12 Oct 2011 16:35:45 -0400
Subject: [PATCH 216/248] lttng: update alignment macros

Signed-off-by: Bruce Ashfield <bruce.ashfield@windriver.com>
---
 include/linux/align.h |    4 ++--
 1 files changed, 2 insertions(+), 2 deletions(-)

diff --git a/include/linux/align.h b/include/linux/align.h
index 6da9346..07b8a8c 100644
--- a/include/linux/align.h
+++ b/include/linux/align.h
@@ -35,7 +35,7 @@
  */
 #define offset_align(align_drift, alignment)				       \
 	({								       \
-		BUILD_AND_RUN_BUG_ON((alignment) == 0			       \
+		BUILD_RUNTIME_BUG_ON((alignment) == 0			       \
 				   || ((alignment) & ((alignment) - 1)));      \
 		(((alignment) - (align_drift)) & ((alignment) - 1));	       \
 	})
@@ -50,7 +50,7 @@
  */
 #define offset_align_floor(align_drift, alignment)			       \
 	({								       \
-		BUILD_AND_RUN_BUG_ON((alignment) == 0			       \
+		BUILD_RUNTIME_BUG_ON((alignment) == 0			       \
 				   || ((alignment) & ((alignment) - 1)));      \
 		(((align_drift) - (alignment)) & ((alignment) - 1);	       \
 	})
-- 
1.7.0.4

