From d4fab39035c2b587ad5ab8ccde8524dd408d3830 Mon Sep 17 00:00:00 2001
From: Mathieu Desnoyers <mathieu.desnoyers@polymtl.ca>
Date: Fri, 3 Sep 2010 10:17:45 -0400
Subject: [PATCH 218/248] Support architectures without dump tables

Signed-off-by: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
---
 ltt/ltt-tracer.h |   10 ++++++++++
 1 files changed, 10 insertions(+), 0 deletions(-)

diff --git a/ltt/ltt-tracer.h b/ltt/ltt-tracer.h
index 9564c3f..8d979b5 100644
--- a/ltt/ltt-tracer.h
+++ b/ltt/ltt-tracer.h
@@ -644,8 +644,18 @@ void ltt_statedump_unregister_kprobes_dump(void (*callback)(void *call_data));
 
 extern void ltt_dump_softirq_vec(void *call_data);
 
+#ifdef CONFIG_HAVE_LTT_DUMP_TABLES
 extern void ltt_dump_sys_call_table(void *call_data);
 extern void ltt_dump_idt_table(void *call_data);
+#else
+static inline void ltt_dump_sys_call_table(void *call_data)
+{
+}
+
+static inline void ltt_dump_idt_table(void *call_data)
+{
+}
+#endif
 
 /* Relay IOCTL */
 
-- 
1.7.0.4

