From 684515a3334d09d3933e963d48927a68a17395ce Mon Sep 17 00:00:00 2001
From: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
Date: Thu, 13 May 2010 19:27:14 -0400
Subject: [PATCH 220/391] ltt-relay-lockless-api-update

ltt relay lockless api update

update to new reader API from
ltt-relay-alloc-use-array-and-vmalloc_sync_all.patch

Signed-off-by: Mathieu Desnoyers <mathieu.desnoyers@polymtl.ca>
---
 ltt/ltt-relay-lockless.c |    4 ++--
 1 files changed, 2 insertions(+), 2 deletions(-)

diff --git a/ltt/ltt-relay-lockless.c b/ltt/ltt-relay-lockless.c
index 24393c2..89f850c 100644
--- a/ltt/ltt-relay-lockless.c
+++ b/ltt/ltt-relay-lockless.c
@@ -624,7 +624,7 @@ static int subbuf_splice_actor(struct file *in,
 
 	for (; spd.nr_pages < nr_pages; spd.nr_pages++) {
 		unsigned int this_len;
-		struct buf_page *page;
+		struct page *page;
 
 		if (!len)
 			break;
@@ -633,7 +633,7 @@ static int subbuf_splice_actor(struct file *in,
 
 		this_len = PAGE_SIZE - poff;
 		page = ltt_relay_read_get_page(buf, roffset);
-		spd.pages[spd.nr_pages] = page->page;
+		spd.pages[spd.nr_pages] = page;
 		spd.partial[spd.nr_pages].offset = poff;
 		spd.partial[spd.nr_pages].len = this_len;
 
-- 
1.6.5.2

