From 7212582d3ff758bac79ee03ef2348c02ab9a3975 Mon Sep 17 00:00:00 2001
From: Mathieu Desnoyers <mathieu.desnoyers@polymtl.ca>
Date: Sun, 24 Oct 2010 20:29:00 -0400
Subject: [PATCH 221/248] VFS: update to new unlocked_ioctl()

Signed-off-by: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
---
 ltt/ltt-relay-vfs.c |   15 ++++-----------
 1 files changed, 4 insertions(+), 11 deletions(-)

diff --git a/ltt/ltt-relay-vfs.c b/ltt/ltt-relay-vfs.c
index defbe2d..5abef77 100644
--- a/ltt/ltt-relay-vfs.c
+++ b/ltt/ltt-relay-vfs.c
@@ -109,7 +109,6 @@ static unsigned int ltt_poll(struct file *filp, poll_table *wait)
 /**
  *	ltt_ioctl - control on the debugfs file
  *
- *	@inode: the inode
  *	@filp: the file
  *	@cmd: the command
  *	@arg: command arg
@@ -129,9 +128,9 @@ static unsigned int ltt_poll(struct file *filp, poll_table *wait)
  *		returns the maximum size for sub-buffers.
  */
 static
-int ltt_ioctl(struct inode *inode, struct file *filp, unsigned int cmd,
-	      unsigned long arg)
+long ltt_ioctl(struct file *filp, unsigned int cmd, unsigned long arg)
 {
+	struct inode *inode = filp->f_dentry->d_inode;
 	struct ltt_chanbuf *buf = inode->i_private;
 	u32 __user *argp = (u32 __user *)arg;
 
@@ -185,13 +184,7 @@ int ltt_ioctl(struct inode *inode, struct file *filp, unsigned int cmd,
 static
 long ltt_compat_ioctl(struct file *file, unsigned int cmd, unsigned long arg)
 {
-	long ret = -ENOIOCTLCMD;
-
-	lock_kernel();
-	ret = ltt_ioctl(file->f_dentry->d_inode, file, cmd, arg);
-	unlock_kernel();
-
-	return ret;
+	return ltt_ioctl(file, cmd, arg);
 }
 #endif
 
@@ -200,7 +193,7 @@ static const struct file_operations ltt_file_operations = {
 	.release = ltt_release,
 	.poll = ltt_poll,
 	.splice_read = ltt_relay_file_splice_read,
-	.ioctl = ltt_ioctl,
+	.unlocked_ioctl = ltt_ioctl,
 	.llseek = ltt_relay_no_llseek,
 #ifdef CONFIG_COMPAT
 	.compat_ioctl = ltt_compat_ioctl,
-- 
1.7.0.4

