From 47a9f9e49b08a942a2ee39fd1f1f396205a57203 Mon Sep 17 00:00:00 2001
From: Mathieu Desnoyers <mathieu.desnoyers@polymtl.ca>
Date: Sun, 24 Oct 2010 20:29:39 -0400
Subject: [PATCH 222/248] probes/fs-trace.c: const char fs_exec

Signed-off-by: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
---
 ltt/probes/fs-trace.c |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/ltt/probes/fs-trace.c b/ltt/probes/fs-trace.c
index bca2827..ccfc282 100644
--- a/ltt/probes/fs-trace.c
+++ b/ltt/probes/fs-trace.c
@@ -25,7 +25,7 @@ void probe_fs_buffer_wait_end(void *_data, struct buffer_head *bh)
 		probe_fs_buffer_wait_end, "bh %p", bh);
 }
 
-void probe_fs_exec(void *_data, char *filename)
+void probe_fs_exec(void *_data, const char *filename)
 {
 	trace_mark_tp(fs, exec, fs_exec, probe_fs_exec, "filename %s",
 		filename);
-- 
1.7.0.4

