From 393d155b2e939140feb64ff4ff4165b8c5ba68f3 Mon Sep 17 00:00:00 2001
From: Mathieu Desnoyers <mathieu.desnoyers@polymtl.ca>
Date: Sun, 24 Oct 2010 20:35:25 -0400
Subject: [PATCH 223/248] block-trace: update to 2.6.36

Signed-off-by: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
---
 ltt/probes/block-trace.c |   30 +++++++++++++++---------------
 1 files changed, 15 insertions(+), 15 deletions(-)

diff --git a/ltt/probes/block-trace.c b/ltt/probes/block-trace.c
index 51ae2cd..fcb87d5 100644
--- a/ltt/probes/block-trace.c
+++ b/ltt/probes/block-trace.c
@@ -19,10 +19,10 @@ void probe_block_rq_abort(void *data, struct request_queue *q, struct request *r
 {
 	int rw = rq->cmd_flags & 0x03;
 
-	if (blk_discard_rq(rq))
-		rw |= (1 << BIO_RW_DISCARD);
+	if (rq->cmd_flags & REQ_DISCARD)
+		rw |= REQ_DISCARD;
 
-	if (blk_pc_request(rq)) {
+	if (rq->cmd_type == REQ_TYPE_BLOCK_PC) {
 		trace_mark_tp(block, rq_abort_pc, block_rq_abort,
 			probe_block_rq_abort,
 			"data_len %u rw %d errors %d",
@@ -46,10 +46,10 @@ void probe_block_rq_insert(void *data, struct request_queue *q, struct request *
 {
 	int rw = rq->cmd_flags & 0x03;
 
-	if (blk_discard_rq(rq))
-		rw |= (1 << BIO_RW_DISCARD);
+	if (rq->cmd_flags & REQ_DISCARD)
+		rw |= REQ_DISCARD;
 
-	if (blk_pc_request(rq)) {
+	if (rq->cmd_type == REQ_TYPE_BLOCK_PC) {
 		trace_mark_tp(block, rq_insert_pc, block_rq_insert,
 			probe_block_rq_insert,
 			"data_len %u rw %d errors %d",
@@ -73,10 +73,10 @@ void probe_block_rq_issue(void *data, struct request_queue *q, struct request *r
 {
 	int rw = rq->cmd_flags & 0x03;
 
-	if (blk_discard_rq(rq))
-		rw |= (1 << BIO_RW_DISCARD);
+	if (rq->cmd_flags & REQ_DISCARD)
+		rw |= REQ_DISCARD;
 
-	if (blk_pc_request(rq)) {
+	if (rq->cmd_type == REQ_TYPE_BLOCK_PC) {
 		trace_mark_tp(block, rq_issue_pc, block_rq_issue,
 			probe_block_rq_issue,
 			"data_len %u rw %d errors %d",
@@ -100,10 +100,10 @@ void probe_block_rq_requeue(void *data, struct request_queue *q, struct request
 {
 	int rw = rq->cmd_flags & 0x03;
 
-	if (blk_discard_rq(rq))
-		rw |= (1 << BIO_RW_DISCARD);
+	if (rq->cmd_flags & REQ_DISCARD)
+		rw |= REQ_DISCARD;
 
-	if (blk_pc_request(rq)) {
+	if (rq->cmd_type == REQ_TYPE_BLOCK_PC) {
 		trace_mark_tp(block, rq_requeue_pc, block_rq_requeue,
 			probe_block_rq_requeue,
 			"data_len %u rw %d errors %d",
@@ -128,10 +128,10 @@ void probe_block_rq_complete(void *data, struct request_queue *q, struct request
 {
 	int rw = rq->cmd_flags & 0x03;
 
-	if (blk_discard_rq(rq))
-		rw |= (1 << BIO_RW_DISCARD);
+	if (rq->cmd_flags & REQ_DISCARD)
+		rw |= REQ_DISCARD;
 
-	if (blk_pc_request(rq)) {
+	if (rq->cmd_type == REQ_TYPE_BLOCK_PC) {
 		trace_mark_tp(block, rq_complete_pc, block_rq_complete,
 			probe_block_rq_complete,
 			"data_len %u rw %d errors %d",
-- 
1.7.0.4

