From 432558f026415d1e1d02563055290d4753376b8b Mon Sep 17 00:00:00 2001
From: Mathieu Desnoyers <mathieu.desnoyers@polymtl.ca>
Date: Fri, 29 Oct 2010 12:24:55 -0400
Subject: [PATCH 224/248] newer get_trace_clock may fail on arm

Signed-off-by: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
---
 ltt/ltt-tracer.c |    4 +++-
 1 files changed, 3 insertions(+), 1 deletions(-)

diff --git a/ltt/ltt-tracer.c b/ltt/ltt-tracer.c
index 8eae966..24a3a92 100644
--- a/ltt/ltt-tracer.c
+++ b/ltt/ltt-tracer.c
@@ -782,7 +782,9 @@ int ltt_trace_alloc(const char *trace_name)
 	kref_init(&trace->kref);
 	init_waitqueue_head(&trace->kref_wq);
 	trace->active = 0;
-	get_trace_clock();
+	err = get_trace_clock();
+	if (err)
+		goto traces_error;
 	trace->freq_scale = trace_clock_freq_scale();
 
 	if (!trace->transport) {
-- 
1.7.0.4

