From eb21b395a8bdd0d7e6f2c2d7ad30987e4ca78832 Mon Sep 17 00:00:00 2001
From: Mathieu Desnoyers <mathieu.desnoyers@polymtl.ca>
Date: Thu, 13 Jan 2011 16:42:51 -0500
Subject: [PATCH 226/248] Update net-trace probe for 2.6.37

Signed-off-by: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
---
 ltt/probes/net-trace.c |    4 ++--
 1 files changed, 2 insertions(+), 2 deletions(-)

diff --git a/ltt/probes/net-trace.c b/ltt/probes/net-trace.c
index 3124125..b4eafe0 100644
--- a/ltt/probes/net-trace.c
+++ b/ltt/probes/net-trace.c
@@ -19,7 +19,7 @@
 
 void probe_net_dev_xmit(void *_data, struct sk_buff *skb);
 
-DEFINE_MARKER_TP(net, dev_xmit, net_dev_xmit, probe_net_dev_xmit,
+DEFINE_MARKER_TP(net, dev_xmit, lttng_net_dev_xmit, probe_net_dev_xmit,
 	"skb %p protocol #n2u%hu");
 
 notrace void probe_net_dev_xmit(void *_data, struct sk_buff *skb)
@@ -37,7 +37,7 @@ notrace void probe_net_dev_xmit(void *_data, struct sk_buff *skb)
 
 void probe_net_dev_receive(void *_data, struct sk_buff *skb);
 
-DEFINE_MARKER_TP(net, dev_receive, net_dev_receive, probe_net_dev_receive,
+DEFINE_MARKER_TP(net, dev_receive, lttng_net_dev_receive, probe_net_dev_receive,
 	"skb %p protocol #n2u%hu");
 
 notrace void probe_net_dev_receive(void *_data, struct sk_buff *skb)
-- 
1.7.0.4

