From 9e7733f81b712b9763040cceaf244b0f4832ea39 Mon Sep 17 00:00:00 2001
From: Mathieu Desnoyers <mathieu.desnoyers@polymtl.ca>
Date: Wed, 16 Mar 2011 19:02:05 -0400
Subject: [PATCH 230/248] kvm-trace: fix warning (probe prototype mismatch)

Signed-off-by: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
---
 ltt/probes/kvm-trace.c |    6 +++---
 1 files changed, 3 insertions(+), 3 deletions(-)

diff --git a/ltt/probes/kvm-trace.c b/ltt/probes/kvm-trace.c
index 9a85f92..b611efc 100644
--- a/ltt/probes/kvm-trace.c
+++ b/ltt/probes/kvm-trace.c
@@ -59,12 +59,12 @@ void probe_kvm_apic (void *_data, unsigned int rw, unsigned int reg,
 		reg, val);
 }
 
-void probe_kvm_exit(void *_data, unsigned int exit_reason, struct kvm_vcpu *vcpu)
+void probe_kvm_exit(void *_data, unsigned int exit_reason, struct kvm_vcpu *vcpu, u32 isa)
 {
 	trace_mark_tp(kvm, kvm_exit, kvm_exit,
 		probe_kvm_exit,
-		"reason %d",
-		exit_reason);
+		"reason %d isa %u",
+		exit_reason, (unsigned int)isa);
 }
 
 void probe_kvm_inj_virq(void *_data, unsigned int irq)
-- 
1.7.0.4

