From 3323fb0155c7e0c691f356158107f91b037d2b82 Mon Sep 17 00:00:00 2001
From: Francis Giraldeau <francis.giraldeau@usherbrooke.ca>
Date: Wed, 18 May 2011 16:25:15 -0400
Subject: [PATCH 233/248] Add dup probe to lttng module

This patch provides instrumentation for dup system call. It records the
oldfd and newfd if the operation succeed. This is required to recover in
the trace tre fd that newfd reference to.

Signed-off-by: Francis Giraldeau <francis.giraldeau@usherbrooke.ca>
Signed-off-by: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
---
 ltt/probes/fs-trace.c |    6 ++++++
 1 files changed, 6 insertions(+), 0 deletions(-)

diff --git a/ltt/probes/fs-trace.c b/ltt/probes/fs-trace.c
index ccfc282..4635c88 100644
--- a/ltt/probes/fs-trace.c
+++ b/ltt/probes/fs-trace.c
@@ -153,6 +153,12 @@ void probe_fs_poll(void *_data, int fd)
 		"fd %d", fd);
 }
 
+void probe_fs_dup(void *_data, unsigned int oldfd, unsigned int newfd)
+{
+	trace_mark_tp(fs, dup, fs_dup, probe_fs_dup,
+		"oldfd %d newfd %d", oldfd, newfd);
+}
+
 MODULE_LICENSE("GPL and additional rights");
 MODULE_AUTHOR("Mathieu Desnoyers");
 MODULE_DESCRIPTION("FS Tracepoint Probes");
-- 
1.7.0.4

