From 64c0f41f1045f36713afc8c14d1ddf1116171054 Mon Sep 17 00:00:00 2001
From: Francis Giraldeau <francis.giraldeau@usherbrooke.ca>
Date: Wed, 18 May 2011 16:26:10 -0400
Subject: [PATCH 234/248] Fix typo in tracepoint name

This patch fix a type in tracepoint definition of socket_sendmsg and
socket_recvmsg, that was preventing to appear in the trace.

Signed-off-by: Francis Giraldeau <francis.giraldeau@usherbrooke.ca>
Signed-off-by: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
---
 ltt/probes/net-trace.c |    4 ++--
 1 files changed, 2 insertions(+), 2 deletions(-)

diff --git a/ltt/probes/net-trace.c b/ltt/probes/net-trace.c
index b4eafe0..7e7419a 100644
--- a/ltt/probes/net-trace.c
+++ b/ltt/probes/net-trace.c
@@ -172,7 +172,7 @@ void probe_socket_socketpair(void *_data, int family, int type, int protocol,
 void probe_socket_sendmsg(void *_data, struct socket *sock, struct msghdr *msg, size_t size,
 	int ret);
 
-DEFINE_MARKER_TP(net, socket_sendmsg, net_socket_sendmsg,
+DEFINE_MARKER_TP(net, socket_sendmsg, socket_sendmsg,
 	probe_socket_sendmsg,
 	"sock %p msg %p size %zu ret %d");
 
@@ -195,7 +195,7 @@ notrace void probe_socket_sendmsg(void *_data, struct socket *sock, struct msghd
 void probe_socket_recvmsg(void *_data, struct socket *sock, struct msghdr *msg, size_t size,
 	int flags, int ret);
 
-DEFINE_MARKER_TP(net, socket_recvmsg, net_socket_recvmsg,
+DEFINE_MARKER_TP(net, socket_recvmsg, socket_recvmsg,
 	probe_socket_recvmsg,
 	"sock %p msg %p size %zu flags %d ret %d");
 
-- 
1.7.0.4

