From 23b6c33ed70d0b5b61104d5ea4869c7a1b2f6d4d Mon Sep 17 00:00:00 2001
From: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
Date: Thu, 13 May 2010 19:27:22 -0400
Subject: [PATCH 236/391] lttng-instrumentation-net-fix-null-napi

LTTng instrumentation net fix NULL napi

Hi Mathieu, LTT devs,
  I found that I needed this patch to allow LTT 0.116 to play nice with
  v2.6.29.  I am not 100% sure that I am not papering over a legitimate
  issue w.r.t napi->dev being legally NULL.  But FWIW, this patch allowed
  me to get 0.116 up and running.  Note that I saw the same issue in 0.110
  against 29-rc8, but I didnt persue since the newer trees were available.
  I suspect the same patch would fix things there as well.

  FWIW: The devices in question seemed to be the built-in skb-queues that
  are used as a psuedo device when you use netif_rx().  In these cases
  dev == NULL (not sure if this is by design).  However, since this path
  is probably not that common unless you are using something like tuntap,
  I suspect it might be that this has just been a latent bug in LTT.

Regards,
-Greg

lttng: fix net events to check for null n->dev

Applies to 29-v0.116

Certain napi structures may have a NULL device, and thus will crash when
the net-tracer'er is installed.

From: Gregory Haskins <ghaskins@novell.com>
Signed-off-by: Gregory Haskins <ghaskins@novell.com>
Signed-off-by: Mathieu Desnoyers <mathieu.desnoyers@polymtl.ca>
---
 ltt/probes/net-trace.c |    6 +++---
 1 files changed, 3 insertions(+), 3 deletions(-)

diff --git a/ltt/probes/net-trace.c b/ltt/probes/net-trace.c
index 72397aa..db3f5c9 100644
--- a/ltt/probes/net-trace.c
+++ b/ltt/probes/net-trace.c
@@ -263,7 +263,7 @@ notrace void probe_net_napi_schedule(struct napi_struct *n)
 	data.f1 = (unsigned long)n;
 	data_len += sizeof(data.f1);
 	/* No need to align for strings */
-	strcpy(data.f2, n->dev->name);
+	strcpy(data.f2, n->dev ? n->dev->name : "<unk>");
 	data_len += strlen(data.f2) + 1;
 
 	marker = &GET_MARKER(net, napi_schedule);
@@ -286,7 +286,7 @@ notrace void probe_net_napi_poll(struct napi_struct *n)
 	data.f1 = (unsigned long)n;
 	data_len += sizeof(data.f1);
 	/* No need to align for strings */
-	strcpy(data.f2, n->dev->name);
+	strcpy(data.f2, n->dev ? n->dev->name : "<unk>");
 	data_len += strlen(data.f2) + 1;
 
 	marker = &GET_MARKER(net, napi_poll);
@@ -309,7 +309,7 @@ notrace void probe_net_napi_complete(struct napi_struct *n)
 	data.f1 = (unsigned long)n;
 	data_len += sizeof(data.f1);
 	/* No need to align for strings */
-	strcpy(data.f2, n->dev->name);
+	strcpy(data.f2, n->dev ? n->dev->name : "<unk>");
 	data_len += strlen(data.f2) + 1;
 
 	marker = &GET_MARKER(net, napi_complete);
-- 
1.6.5.2

