From 8976f4b20a7f8aa2cf56cbcf1d85abd0d9702f7a Mon Sep 17 00:00:00 2001
From: Yang Shi <yang.shi@windriver.com>
Date: Thu, 24 May 2012 15:25:40 -0700
Subject: [PATCH 237/248] lttng: modpost: Fix wrong symbol address for vmlinux symbol

For the symbol in vmlinux image, sym->st_value is target kernel address,
so disregard it for calculating vmlinux markers address.

Signed-off-by: Yang Shi <yang.shi@windriver.com>
---
 scripts/mod/modpost.c |    6 +++++-
 1 files changed, 5 insertions(+), 1 deletions(-)

diff --git a/scripts/mod/modpost.c b/scripts/mod/modpost.c
index 9cc7dda..00acd10 100644
--- a/scripts/mod/modpost.c
+++ b/scripts/mod/modpost.c
@@ -1732,7 +1732,11 @@ static void get_markers(struct elf_info *info, struct module *mod)
 		    sym->st_shndx == info->markers_strings_sec &&
 		    !strncmp(info->strtab + sym->st_name,
 			     "__mstrtab_", sizeof "__mstrtab_" - 1)) {
-			const char *name = strings + sym->st_value;
+			const char *name;
+			if (is_vmlinux(mod->name))
+				name = strings;
+			else
+				name = strings + sym->st_value;
 			const char *fmt = strchr(name, '\0') + 1;
 			char *line = NULL;
 			asprintf(&line, "%s\t%s\t%s\n", name, mod->name, fmt);
-- 
1.7.0.4

