From 2bbefa769470b3a29d2afa8977f76d21d2e5a700 Mon Sep 17 00:00:00 2001
From: Yang Shi <yang.shi@windriver.com>
Date: Thu, 24 May 2012 16:40:10 -0700
Subject: [PATCH 243/248] lttng: Replace _TRACE_RCU_H macro to _LTT_TRACE_RCU_H

Since _TRACE_RCU_H is already defined in trace/events/rcu.h, so change
lttng trace/rcu.h to adopt _LTT_TRACE_RCU_H.

Signed-off-by: Yang Shi <yang.shi@windriver.com>
---
 include/trace/rcu.h |    4 ++--
 1 files changed, 2 insertions(+), 2 deletions(-)

diff --git a/include/trace/rcu.h b/include/trace/rcu.h
index f551c2c..ed04b3d 100644
--- a/include/trace/rcu.h
+++ b/include/trace/rcu.h
@@ -1,5 +1,5 @@
-#ifndef _TRACE_RCU_H
-#define _TRACE_RCU_H
+#ifndef _LTT_TRACE_RCU_H
+#define _LTT_TRACE_RCU_H
 
 #include <linux/tracepoint.h>
 #include <linux/rcupdate.h>
-- 
1.7.0.4

