From e1770802c4dff983e8094510e2ac5c3ed0674095 Mon Sep 17 00:00:00 2001
From: Yang Shi <yang.shi@windriver.com>
Date: Tue, 29 May 2012 15:54:49 -0700
Subject: [PATCH 245/248] lttng-modules: Remove smp_lock.h

Due to commit 4ba8216cd90560bc402f52076f64d8546e8aefcb
[BKL: That's all, folks],
remove include of smp_lock.h.

Signed-off-by: Yang Shi <yang.shi@windriver.com>
---
 ltt/ltt-relay-lockless.c |    1 -
 ltt/ltt-relay-lockless.h |    1 -
 2 files changed, 0 insertions(+), 2 deletions(-)

diff --git a/ltt/ltt-relay-lockless.c b/ltt/ltt-relay-lockless.c
index db4362d..7a66ab6 100644
--- a/ltt/ltt-relay-lockless.c
+++ b/ltt/ltt-relay-lockless.c
@@ -44,7 +44,6 @@
 #include <linux/timer.h>
 #include <linux/sched.h>
 #include <linux/bitops.h>
-#include <linux/smp_lock.h>
 #include <linux/stat.h>
 #include <linux/cpu.h>
 #include <linux/idle.h>
diff --git a/ltt/ltt-relay-lockless.h b/ltt/ltt-relay-lockless.h
index 73c6a3d..06db6d9 100644
--- a/ltt/ltt-relay-lockless.h
+++ b/ltt/ltt-relay-lockless.h
@@ -49,7 +49,6 @@
 #include <linux/sched.h>
 #include <linux/bitops.h>
 #include <linux/fs.h>
-#include <linux/smp_lock.h>
 #include <linux/debugfs.h>
 #include <linux/stat.h>
 #include <linux/cpu.h>
-- 
1.7.0.4

