From f8ce2841f1e62d5ca4d0a0115d9be19b3c74f678 Mon Sep 17 00:00:00 2001
From: Yang Shi <yang.shi@windriver.com>
Date: Tue, 29 May 2012 16:07:15 -0700
Subject: [PATCH 247/248] lttng-modules: Remove non-exist block trace points

block_unplug_io/timer tracepoints are no longer existing, so remove them.

Signed-off-by: Yang Shi <yang.shi@windriver.com>
---
 ltt/probes/block-trace.c |   14 +++-----------
 1 files changed, 3 insertions(+), 11 deletions(-)

diff --git a/ltt/probes/block-trace.c b/ltt/probes/block-trace.c
index a01b6f4..84e41cd 100644
--- a/ltt/probes/block-trace.c
+++ b/ltt/probes/block-trace.c
@@ -8,6 +8,7 @@
  */
 
 #include <linux/module.h>
+#include <linux/marker.h>
 
 #include <trace/events/block.h>
 
@@ -259,20 +260,11 @@ void probe_block_plug(void *data, struct request_queue *q)
 			 MARK_NOARGS);
 }
 
-void probe_block_unplug_io(void *data, struct request_queue *q)
+void probe_block_unplug(void *data, struct request_queue *q)
 {
 	unsigned int pdu = q->rq.count[READ] + q->rq.count[WRITE];
 
-	trace_mark_tp(block, unplug_io, block_unplug_io, probe_block_unplug_io,
-			"pdu %u", pdu);
-}
-
-void probe_block_unplug_timer(void *data, struct request_queue *q)
-{
-	unsigned int pdu = q->rq.count[READ] + q->rq.count[WRITE];
-
-	trace_mark_tp(block, unplug_timer, block_unplug_timer,
-			probe_block_unplug_timer,
+	trace_mark_tp(block, unplug, block_unplug, probe_block_unplug,
 			"pdu %u", pdu);
 }
 
-- 
1.7.0.4

