From 86cbd02103ac618dd9f40ff6f4ae12f291fe3c78 Mon Sep 17 00:00:00 2001
From: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
Date: Thu, 13 May 2010 19:27:34 -0400
Subject: [PATCH 256/391] lttng-instrumentation-rcu-tree-fix
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

LTTng instrumentation fix tree RCU

I'm having problems building the most recent version of Lttng:
$ make i386_defconfig
$ make menuconfig
# Linux Trace Toolkit Next Generation (LTTng) = y
$ make
[...]
CC [M] ltt/probes/rcu-trace.o
ltt/probes/rcu-trace.c: In function ‘__check_tp_type’:
ltt/probes/rcu-trace.c:55: error: implicit declaration of function
‘register_trace_probe_rcu_tree_callback’
ltt/probes/rcu-trace.c: In function ‘probe_rcu_tree_callback’:
ltt/probes/rcu-trace.c:55: error: expected ‘,’ or ‘;’ before ‘head’
ltt/probes/rcu-trace.c:55: warning: passing argument 1 of
‘___mark_check_format’ from incompatible pointer type
ltt/probes/rcu-trace.c: In function ‘__check_tp_type’:
ltt/probes/rcu-trace.c:61: error: implicit declaration of function
‘register_trace_probe_rcu_tree_call_rcu’
ltt/probes/rcu-trace.c: In function ‘probe_rcu_tree_call_rcu’:
ltt/probes/rcu-trace.c:61: error: expected ‘,’ or ‘;’ before ‘head’
ltt/probes/rcu-trace.c:61: warning: passing argument 1 of
‘___mark_check_format’ from incompatible pointer type
ltt/probes/rcu-trace.c: In function ‘__check_tp_type’:
ltt/probes/rcu-trace.c:67: error: implicit declaration of function
‘register_trace_probe_rcu_tree_call_rcu_bh’
ltt/probes/rcu-trace.c: In function ‘probe_rcu_tree_call_rcu_bh’:
ltt/probes/rcu-trace.c:67: error: expected ‘,’ or ‘;’ before ‘head’
ltt/probes/rcu-trace.c:67: warning: passing argument 1 of
‘___mark_check_format’ from incompatible pointer type
make[2]: *** [ltt/probes/rcu-trace.o] Error 1
make[1]: *** [ltt/probes] Error 2
make: *** [ltt] Error 2

Signed-off-by: Mathieu Desnoyers <mathieu.desnoyers@polymtl.ca>
---
 ltt/probes/rcu-trace.c |    6 +++---
 1 files changed, 3 insertions(+), 3 deletions(-)

diff --git a/ltt/probes/rcu-trace.c b/ltt/probes/rcu-trace.c
index 6467f80..876d510 100644
--- a/ltt/probes/rcu-trace.c
+++ b/ltt/probes/rcu-trace.c
@@ -13,19 +13,19 @@
 #ifdef CONFIG_TREE_RCU
 void probe_rcu_tree_callback(struct rcu_head *head)
 {
-	trace_mark_tp(rcu_tree_callback, rcu_tree_callback,
+	trace_mark_tp(rcu, tree_callback, rcu_tree_callback,
 		probe_rcu_tree_callback, "func %p", head->func);
 }
 
 void probe_rcu_tree_call_rcu(struct rcu_head *head, unsigned long ip)
 {
-	trace_mark_tp(rcu_tree_call_rcu, rcu_tree_call_rcu,
+	trace_mark_tp(rcu, tree_call_rcu, rcu_tree_call_rcu,
 		probe_rcu_tree_call_rcu, "func %p ip 0x%lX", head->func, ip);
 }
 
 void probe_rcu_tree_call_rcu_bh(struct rcu_head *head, unsigned long ip)
 {
-	trace_mark_tp(rcu_tree_call_rcu_bh, rcu_tree_call_rcu_bh,
+	trace_mark_tp(rcu, tree_call_rcu_bh, rcu_tree_call_rcu_bh,
 		probe_rcu_tree_call_rcu_bh, "func %p ip 0x%lX",
 		head->func, ip);
 }
-- 
1.6.5.2

