From 9c22e4d6a0b791dd814834a042cb6e8a363cf865 Mon Sep 17 00:00:00 2001
From: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
Date: Thu, 13 May 2010 19:27:50 -0400
Subject: [PATCH 285/391] lttng-relay-irqoff-fix-alloc

lttng relay irqoff fix alloc

Does not have real-life effect in irqoff because local_t and pointer size are
the same.

Signed-off-by: Mathieu Desnoyers <mathieu.desnoyers@polymtl.ca>
---
 ltt/ltt-relay-irqoff.c |    4 ++--
 1 files changed, 2 insertions(+), 2 deletions(-)

diff --git a/ltt/ltt-relay-irqoff.c b/ltt/ltt-relay-irqoff.c
index bab874f..096072d 100644
--- a/ltt/ltt-relay-irqoff.c
+++ b/ltt/ltt-relay-irqoff.c
@@ -860,7 +860,7 @@ static int ltt_relay_create_buffer(struct ltt_trace_struct *trace,
 		return -ENOMEM;
 
 	ltt_buf->commit_count =
-		kzalloc_node(ALIGN(sizeof(ltt_buf->commit_count) * n_subbufs,
+		kzalloc_node(ALIGN(sizeof(*ltt_buf->commit_count) * n_subbufs,
 				   1 << INTERNODE_CACHE_SHIFT),
 			GFP_KERNEL, cpu_to_node(cpu));
 	if (!ltt_buf->commit_count) {
@@ -870,7 +870,7 @@ static int ltt_relay_create_buffer(struct ltt_trace_struct *trace,
 
 #ifdef CONFIG_LTT_VMCORE
 	ltt_buf->commit_seq =
-		kzalloc_node(ALIGN(sizeof(ltt_buf->commit_seq) * n_subbufs,
+		kzalloc_node(ALIGN(sizeof(*ltt_buf->commit_seq) * n_subbufs,
 				   1 << INTERNODE_CACHE_SHIFT),
 			GFP_KERNEL, cpu_to_node(cpu));
 	if (!ltt_buf->commit_seq) {
-- 
1.6.5.2

