From bdcaae6c69548b5947adf45cca67237db6414ec6 Mon Sep 17 00:00:00 2001
From: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
Date: Thu, 13 May 2010 19:28:02 -0400
Subject: [PATCH 311/391] lttng-fix-relay-vfs

lttng fix relay vfs

Fix file close bug: should not delete structures there.

Signed-off-by: Mathieu Desnoyers <mathieu.desnoyers@polymtl.ca>
---
 ltt/ltt-relay-vfs.c |    3 ---
 1 files changed, 0 insertions(+), 3 deletions(-)

diff --git a/ltt/ltt-relay-vfs.c b/ltt/ltt-relay-vfs.c
index 802d191..daa6e74 100644
--- a/ltt/ltt-relay-vfs.c
+++ b/ltt/ltt-relay-vfs.c
@@ -55,10 +55,7 @@ end:
 static int ltt_release(struct inode *inode, struct file *file)
 {
 	struct ltt_chanbuf *buf = inode->i_private;
-	struct ltt_chan *chan = container_of(buf->a.chan, struct ltt_chan, a);
 
-	ltt_ascii_remove(chan);
-	debugfs_remove(buf->a.dentry);
 	ltt_chanbuf_release_read(buf);
 
 	return 0;
-- 
1.6.5.2

