From ac48c74d6265562441e2c808820e45e8172d6803 Mon Sep 17 00:00:00 2001
From: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
Date: Thu, 13 May 2010 19:28:03 -0400
Subject: [PATCH 313/391] lttng-fix-kref

LTTng fix kref

Fix kref problems in lttng (came in during refactoring).

Signed-off-by: Mathieu Desnoyers <mathieu.desnoyers@polymtl.ca>
---
 ltt/ltt-tracer.c |   15 ++++++++-------
 1 files changed, 8 insertions(+), 7 deletions(-)

diff --git a/ltt/ltt-tracer.c b/ltt/ltt-tracer.c
index 179256f..4e7b223 100644
--- a/ltt/ltt-tracer.c
+++ b/ltt/ltt-tracer.c
@@ -491,6 +491,8 @@ void ltt_release_trace(struct kref *kref)
 {
 	struct ltt_trace *trace = container_of(kref, struct ltt_trace, kref);
 
+	trace->ops->remove_dirs(trace);
+	module_put(trace->transport->owner);
 	ltt_channels_trace_free(trace->channels, trace->nr_channels);
 	kfree(trace);
 }
@@ -906,9 +908,11 @@ int ltt_trace_alloc(const char *trace_name)
 	return 0;
 
 create_channel_error:
-	for (chan--; chan >= 0; chan--)
+	for (chan--; chan >= 0; chan--) {
 		if (trace->channels[chan].active)
-			trace->ops->remove_channel(&trace->channels[chan].a.kref);
+			kref_put(&trace->channels[chan].a.kref,
+				 trace->ops->remove_channel);
+	}
 	trace->ops->remove_dirs(trace);
 
 dirs_error:
@@ -1010,13 +1014,10 @@ static void __ltt_trace_destroy(struct ltt_trace *trace)
 	for (i = 0; i < trace->nr_channels; i++) {
 		chan = &trace->channels[i];
 		if (chan->active)
-			trace->ops->remove_channel(&chan->a.kref);
+			kref_put(&chan->a.kref,
+				 trace->ops->remove_channel);
 	}
 
-	trace->ops->remove_dirs(trace);
-
-	module_put(trace->transport->owner);
-
 	/*
 	 * Wait for lttd readers to release the files, therefore making sure
 	 * the last subbuffers have been read.
-- 
1.6.5.2

