From 8b9ad2884b76a4d3cde965d20c67c3644114d57a Mon Sep 17 00:00:00 2001
From: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
Date: Thu, 13 May 2010 19:28:26 -0400
Subject: [PATCH 356/391] lttng-transport-lockless-add-timer-on-works-with-del-timer-sync

lttng transport lockless add timer on works with del timer sync

Needed for RT kernel.

Signed-off-by: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
---
 ltt/ltt-relay-lockless.c |   16 ++--------------
 1 files changed, 2 insertions(+), 14 deletions(-)

diff --git a/ltt/ltt-relay-lockless.c b/ltt/ltt-relay-lockless.c
index d8b1a9c..6836150 100644
--- a/ltt/ltt-relay-lockless.c
+++ b/ltt/ltt-relay-lockless.c
@@ -444,9 +444,7 @@ static void switch_buffer(unsigned long data)
 	if (atomic_long_read(&buf->active_readers))
 		ltt_force_switch(buf, FORCE_ACTIVE);
 
-	del_timer(&buf->switch_timer);
-	buf->switch_timer.expires += chan->switch_timer_interval;
-	add_timer_on(&buf->switch_timer, smp_processor_id());
+	mod_timer_pinned(&buf->switch_timer, chan->switch_timer_interval);
 }
 
 static void ltt_chanbuf_start_switch_timer(struct ltt_chanbuf *buf)
@@ -480,16 +478,6 @@ void ltt_chan_start_switch_timer(struct ltt_chan *chan)
 		ltt_chanbuf_start_switch_timer(buf);
 	}
 }
-/*
- * Cannot use del_timer_sync with add_timer_on, so use an IPI to locally
- * delete the timer.
- */
-static void stop_switch_timer_ipi(void *info)
-{
-	struct ltt_chanbuf *buf = (struct ltt_chanbuf *)info;
-
-	del_timer(&buf->switch_timer);
-}
 
 static void ltt_chanbuf_stop_switch_timer(struct ltt_chanbuf *buf)
 {
@@ -498,7 +486,7 @@ static void ltt_chanbuf_stop_switch_timer(struct ltt_chanbuf *buf)
 	if (!chan->switch_timer_interval)
 		return;
 
-	smp_call_function(stop_switch_timer_ipi, buf, 1);
+	del_timer_sync(&buf->switch_timer);
 }
 
 /*
-- 
1.6.5.2

