From fecf69ea0a93b349adf526de7f40ea32b8cd14f3 Mon Sep 17 00:00:00 2001
From: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
Date: Thu, 13 May 2010 19:28:42 -0400
Subject: [PATCH 387/391] lttng-ascii-fix-buggy-error-handling

Fix ltt_relay_iter_open_channel to cleanup correctly on error.

From: Balaji Rao <balaji@raobalaji.com>
Signed-off-by: Balaji Rao <balaji@raobalaji.com>
Signed-off-by: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
---
 ltt/ltt-ascii.c |    7 +++++--
 1 files changed, 5 insertions(+), 2 deletions(-)

diff --git a/ltt/ltt-ascii.c b/ltt/ltt-ascii.c
index e67b521..76da1cd 100644
--- a/ltt/ltt-ascii.c
+++ b/ltt/ltt-ascii.c
@@ -424,8 +424,11 @@ static int ltt_relay_iter_open_channel(struct ltt_relay_iter *iter,
 		}
 		update_cpu_iter(citer, citer->hdr_offset);
 	}
-	if (!iter->nr_refs)
-		return -ENODATA; /* no data available */
+	if (!iter->nr_refs) {
+		ret = -ENODATA;
+		goto error;
+	}
+
 	return 0;
 
 error:
-- 
1.6.5.2

