From e7efded85fd373a98c4831f2219b3a475e2db02b Mon Sep 17 00:00:00 2001
From: Bruce Ashfield <bruce.ashfield@windriver.com>
Date: Sun, 30 May 2010 04:40:09 -0400
Subject: [PATCH 391/391] lttng: add slub.h include

Signed-off-by: Bruce Ashfield <bruce.ashfield@windriver.com>
---
 kernel/trace/power-traces.c |    1 +
 1 files changed, 1 insertions(+), 0 deletions(-)

diff --git a/kernel/trace/power-traces.c b/kernel/trace/power-traces.c
index a22582a..9f4f565 100644
--- a/kernel/trace/power-traces.c
+++ b/kernel/trace/power-traces.c
@@ -9,6 +9,7 @@
 #include <linux/workqueue.h>
 #include <linux/sched.h>
 #include <linux/module.h>
+#include <linux/slab.h>
 
 #define CREATE_TRACE_POINTS
 #include <trace/events/power.h>
-- 
1.6.5.2

