From 82411d5c36a78e840c20721df67a728659191e3d Mon Sep 17 00:00:00 2001
From: Bruce Ashfield <bruce.ashfield@windriver.com>
Date: Wed, 8 Oct 2008 09:40:56 -0400
Subject: [PATCH] Fix non-x86 build failure due to missing cpu variable

We need to explicitly indicate the cpu number on non-x86
when allocating relay buffers. So let's pass it as a
parameter to the function itself.

Signed-off-by: Bruce Ashfield <bruce.ashfield@windriver.com>
---
 ltt/ltt-relay-alloc.c |    4 ++--
 1 files changed, 2 insertions(+), 2 deletions(-)

diff --git a/ltt/ltt-relay-alloc.c b/ltt/ltt-relay-alloc.c
index f1ecf26..05fe3ab 100644
--- a/ltt/ltt-relay-alloc.c
+++ b/ltt/ltt-relay-alloc.c
@@ -31,7 +31,7 @@ static LIST_HEAD(relay_channels);
  *	@buf: the buffer struct
  *	@size: total size of the buffer
  */
-static int relay_alloc_buf(struct rchan_buf *buf, size_t *size)
+static int relay_alloc_buf(int cpu, struct rchan_buf *buf, size_t *size)
 {
 	unsigned int i, n_pages;
 	struct buf_page *buf_page, *n;
@@ -90,7 +90,7 @@ static struct rchan_buf *relay_create_buf(struct rchan *chan, int cpu)
 		return NULL;
 
  	buf->cpu = cpu;
-	ret = relay_alloc_buf(buf, &chan->alloc_size);
+	ret = relay_alloc_buf(cpu, buf, &chan->alloc_size);
 	if (ret)
 		goto free_buf;
 
-- 
1.6.0.90.g436ed

