From ec5243b8e2da5eacecd75b870b834e0866c8bc35 Mon Sep 17 00:00:00 2001
From: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
Date: Thu, 13 May 2010 19:27:34 -0400
Subject: [PATCH 257/391] lttng-kprobes-fix-module-disabled-support

Fix LTTng Kprobes support when modules are disabled

Change some Kconfig dependencies to allow the kernel to build when LTTng is
enabled and module support is not. Without this patch, the kernel build breaks
because LTT_KPROBES selects KPROBES but the latter depends on module support.

Signed-off-by: Benjamin Poirier <benjamin.poirier@polytml.ca>
Signed-off-by: Mathieu Desnoyers <mathieu.desnoyers@polymtl.ca>
---
 ltt/Kconfig |    4 +---
 1 files changed, 1 insertions(+), 3 deletions(-)

diff --git a/ltt/Kconfig b/ltt/Kconfig
index dc85246..eec424c 100644
--- a/ltt/Kconfig
+++ b/ltt/Kconfig
@@ -183,12 +183,10 @@ config LTT_STATEDUMP
 
 config LTT_KPROBES
 	tristate "Linux Trace Toolkit Kprobes Support"
-	depends on HAVE_KPROBES
 	depends on LTT_TRACE_CONTROL
 	depends on LTT_FAST_SERIALIZE
 	depends on LTT_STATEDUMP
-	select KPROBES
-	select KALLSYMS
+	depends on KPROBES
 	default y
 	help
 	  Allows connecting the LTTng tracer on kprobes using simple debugfs
-- 
1.6.5.2

