From 707840d79a844ea7e2fee869e759177031cd2b86 Mon Sep 17 00:00:00 2001
From: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
Date: Thu, 13 May 2010 19:28:33 -0400
Subject: [PATCH 368/391] lttng-relay-vfs-fix-flight-prefix

lttng relay vfs fix flight prefix

LTTng refactoring dropped the flight- file prefix for flight recorder channels.

Signed-off-by: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
---
 ltt/ltt-relay-vfs.c |    5 ++++-
 1 files changed, 4 insertions(+), 1 deletions(-)

diff --git a/ltt/ltt-relay-vfs.c b/ltt/ltt-relay-vfs.c
index 10ea078..6bd30a4 100644
--- a/ltt/ltt-relay-vfs.c
+++ b/ltt/ltt-relay-vfs.c
@@ -212,7 +212,10 @@ int ltt_chanbuf_create_file(const char *filename, struct dentry *parent,
 		ret = -ENOMEM;
 		goto end;
 	}
-	snprintf(tmpname, NAME_MAX, "%s_%d", chan->a.filename, buf->a.cpu);
+
+	snprintf(tmpname, NAME_MAX, "%s%s_%d",
+		 chan->overwrite ? LTT_FLIGHT_PREFIX : "",
+		 chan->a.filename, buf->a.cpu);
 
 	buf->a.dentry = debugfs_create_file(tmpname, mode, parent, buf,
 					    &ltt_file_operations);
-- 
1.6.5.2

