From fb8dc22b71edab65d647076e16d7603c357d4b99 Mon Sep 17 00:00:00 2001
From: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
Date: Thu, 13 May 2010 19:28:29 -0400
Subject: [PATCH 362/391] lttng-transport-lockless-add-timer-on-works-with-del-timer-sync-fix

lttng transport lockless add timer on works with del timer sync

I noticed that switch_buffer was not being called at intervals as it
should be.

From: Balaji Rao <balaji@raobalaji.com>
Signed-off-by: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
---
 ltt/ltt-relay-lockless.c |    3 ++-
 1 files changed, 2 insertions(+), 1 deletions(-)

diff --git a/ltt/ltt-relay-lockless.c b/ltt/ltt-relay-lockless.c
index 6836150..0d1c064 100644
--- a/ltt/ltt-relay-lockless.c
+++ b/ltt/ltt-relay-lockless.c
@@ -444,7 +444,8 @@ static void switch_buffer(unsigned long data)
 	if (atomic_long_read(&buf->active_readers))
 		ltt_force_switch(buf, FORCE_ACTIVE);
 
-	mod_timer_pinned(&buf->switch_timer, chan->switch_timer_interval);
+	mod_timer_pinned(&buf->switch_timer,
+			 jiffies + chan->switch_timer_interval);
 }
 
 static void ltt_chanbuf_start_switch_timer(struct ltt_chanbuf *buf)
-- 
1.6.5.2

