From 67b0ee73a813e283ec852008c3dfee8b3d1d7c85 Mon Sep 17 00:00:00 2001
From: Michel Thebeau <michel.thebeau@windriver.com>
Date: Tue, 24 Aug 2010 12:48:48 -0700
Subject: [PATCH] lttng: LTT_ASCII depends on TRACER, RELAY_LOCKLESS

Prevents configuration of LTT_ASCII=y without LTT_TRACER=y.
    LTT_ASCII is made to depend on LTT_RELAY_LOCKLESS.

Signed-off-by: Michel Thebeau <michel.thebeau@windriver.com>
---
 ltt/Kconfig |    3 ++-
 1 files changed, 2 insertions(+), 1 deletions(-)

diff --git a/ltt/Kconfig b/ltt/Kconfig
index f81f4ec..fb0b05b 100644
--- a/ltt/Kconfig
+++ b/ltt/Kconfig
@@ -202,9 +202,10 @@ config LTT_FTRACE
 	  surroundings of a problem has been identified in a prior trace.
 
 config LTT_ASCII
-	bool "Linux Trace Toolkit Ascii Output (EXPERIMENTAL)"
+	bool "Linux Trace Toolkit Ascii Output (EXPERIMENTAL)" if LTT_TRACER = "y"
 	depends on EXPERIMENTAL
 	depends on LTT_TRACER
+        depends on LTT_RELAY_LOCKLESS
 	depends on LTT_SERIALIZE
 	default n
 	help
-- 
1.6.5.2

