From d63b5afedb2051da4c71c8205acefc62a096cef9 Mon Sep 17 00:00:00 2001
From: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
Date: Thu, 13 May 2010 19:27:49 -0400
Subject: [PATCH 284/391] lttng-relay-locked-fix-alloc

lttng relay locked fix alloc

Does not have real-life effect in irqoff because long and pointer size are the
same.

Signed-off-by: Mathieu Desnoyers <mathieu.desnoyers@polymtl.ca>
---
 ltt/ltt-relay-locked.c |    4 ++--
 1 files changed, 2 insertions(+), 2 deletions(-)

diff --git a/ltt/ltt-relay-locked.c b/ltt/ltt-relay-locked.c
index 71ceb2d..a579986 100644
--- a/ltt/ltt-relay-locked.c
+++ b/ltt/ltt-relay-locked.c
@@ -813,7 +813,7 @@ static int ltt_relay_create_buffer(struct ltt_trace_struct *trace,
 		return -ENOMEM;
 
 	ltt_buf->commit_count =
-		kzalloc_node(ALIGN(sizeof(ltt_buf->commit_count) * n_subbufs,
+		kzalloc_node(ALIGN(sizeof(*ltt_buf->commit_count) * n_subbufs,
 				   1 << INTERNODE_CACHE_SHIFT),
 			GFP_KERNEL, cpu_to_node(cpu));
 	if (!ltt_buf->commit_count) {
@@ -823,7 +823,7 @@ static int ltt_relay_create_buffer(struct ltt_trace_struct *trace,
 
 #ifdef CONFIG_LTT_VMCORE
 	ltt_buf->commit_seq =
-		kzalloc_node(ALIGN(sizeof(ltt_buf->commit_seq) * n_subbufs,
+		kzalloc_node(ALIGN(sizeof(*ltt_buf->commit_seq) * n_subbufs,
 				   1 << INTERNODE_CACHE_SHIFT),
 			GFP_KERNEL, cpu_to_node(cpu));
 	if (!ltt_buf->commit_seq) {
-- 
1.6.5.2

