From 5d9ca69d3e606b2074b41a162c7ba87a09067062 Mon Sep 17 00:00:00 2001
From: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
Date: Thu, 13 May 2010 19:28:12 -0400
Subject: [PATCH 333/391] lttng-ext4-probes-disable
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

lttng-ext4-probes-disable.patch

2.6.32-rc:

In file included from ltt/probes/../../fs/ext4/ext4_jbd2.h:20,
                 from ltt/probes/../../fs/ext4/mballoc.h:23,
                 from ltt/probes/ext4-trace.c:18:
ltt/probes/../../fs/ext4/ext4.h:1017: error: expected identifier or ‘(’ before ‘{’ token
  CC      fs/autofs4/init.o
make[2]: *** [ltt/probes/ext4-trace.o] Error 1
make[1]: *** [ltt/probes] Error 2
make: *** [ltt] Error 2
make: *** Waiting for unfinished jobs....

Signed-off-by: Mathieu Desnoyers <mathieu.desnoyers@polymtl.ca>
---
 ltt/probes/Makefile |   12 ++++++------
 1 files changed, 6 insertions(+), 6 deletions(-)

diff --git a/ltt/probes/Makefile b/ltt/probes/Makefile
index 56c4fff..adf4ddf 100644
--- a/ltt/probes/Makefile
+++ b/ltt/probes/Makefile
@@ -31,12 +31,12 @@ endif
 obj-$(CONFIG_LTT_TRACEPROBES)	+= jbd2-trace.o
 endif
 
-ifdef CONFIG_EXT4_FS
-ifdef CONFIG_FTRACE
-CFLAGS_REMOVE_ext4-trace.o = -pg
-endif
-obj-$(CONFIG_LTT_TRACEPROBES)	+= ext4-trace.o
-endif
+#ifdef CONFIG_EXT4_FS
+#ifdef CONFIG_FTRACE
+#CFLAGS_REMOVE_ext4-trace.o = -pg
+#endif
+#obj-$(CONFIG_LTT_TRACEPROBES)	+= ext4-trace.o
+#endif
 
 ifdef CONFIG_BLOCK
 ifdef CONFIG_FTRACE
-- 
1.6.5.2

