From ebc016c88865d07ee125185d4ebd8ad24598373f Mon Sep 17 00:00:00 2001
From: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
Date: Thu, 13 May 2010 19:28:01 -0400
Subject: [PATCH 307/391] lttng-relay-flush-only-if-reader-present

LTTng relay periodic flush only if reader present

Perform periodic flush only if reader if present for a buffer.
Note that we leave the flush for CPUs entering idle even if no reader is present
because the reader currently has no mean to actively ask for a flush of a remote
buffer. Not sure we'd want to neither, because sending an IPI from a reader to
an idle cpu would wake up the idle cpu and disturb the workload.

Signed-off-by: Mathieu Desnoyers <mathieu.desnoyers@polymtl.ca>
---
 ltt/ltt-relay-lockless.c |    6 +++++-
 1 files changed, 5 insertions(+), 1 deletions(-)

diff --git a/ltt/ltt-relay-lockless.c b/ltt/ltt-relay-lockless.c
index dd3915e..6de815c 100644
--- a/ltt/ltt-relay-lockless.c
+++ b/ltt/ltt-relay-lockless.c
@@ -441,7 +441,11 @@ static void switch_buffer(unsigned long data)
 	struct ltt_chanbuf *buf = (struct ltt_chanbuf *)data;
 	struct ltt_chan *chan = container_of(buf->a.chan, struct ltt_chan, a);
 
-	ltt_force_switch(buf, FORCE_ACTIVE);
+	/*
+	 * Only flush buffers periodically if readers are active.
+	 */
+	if (atomic_long_read(&buf->active_readers))
+		ltt_force_switch(buf, FORCE_ACTIVE);
 
 	buf->switch_timer.expires += chan->switch_timer_interval;
 	add_timer_on(&buf->switch_timer, smp_processor_id());
-- 
1.6.5.2

