From a8f209048bd4a5596e6d73afd2613d7a673f036f Mon Sep 17 00:00:00 2001
From: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
Date: Thu, 13 May 2010 19:27:36 -0400
Subject: [PATCH 260/391] lttng-instrumentation-net-fix-address-field-byte-order

LTTng instrumentation net: Fix address field byte order

The address field should be recorded in network byte order, as defined in
struct in_ifaddr.

From: Benjamin Poirier <benjamin.poirier@polytml.ca>
Signed-off-by: Benjamin Poirier <benjamin.poirier@polytml.ca>
Signed-off-by: Mathieu Desnoyers <mathieu.desnoyers@polymtl.ca>
---
 ltt/ltt-statedump.c    |    4 ++--
 ltt/probes/net-trace.c |    4 ++--
 2 files changed, 4 insertions(+), 4 deletions(-)

diff --git a/ltt/ltt-statedump.c b/ltt/ltt-statedump.c
index e5100e3..b9c115f 100644
--- a/ltt/ltt-statedump.c
+++ b/ltt/ltt-statedump.c
@@ -96,14 +96,14 @@ static void ltt_enumerate_device(struct ltt_probe_private_data *call_data,
 				__trace_mark(0, netif_state,
 					network_ipv4_interface,
 					call_data,
-					"name %s address #4u%lu up %d",
+					"name %s address #n4u%lu up %d",
 					dev->name,
 					(unsigned long)ifa->ifa_address, 0);
 			in_dev_put(in_dev);
 		}
 	} else
 		__trace_mark(0, netif_state, network_ip_interface,
-			call_data, "name %s address #4u%lu up %d",
+			call_data, "name %s address #n4u%lu up %d",
 			dev->name, 0UL, 0);
 }
 
diff --git a/ltt/probes/net-trace.c b/ltt/probes/net-trace.c
index db3f5c9..ca829f1 100644
--- a/ltt/probes/net-trace.c
+++ b/ltt/probes/net-trace.c
@@ -19,7 +19,7 @@
 void probe_net_dev_xmit(struct sk_buff *skb);
 
 DEFINE_MARKER_TP(net, dev_xmit, net_dev_xmit, probe_net_dev_xmit,
-	"skb %p protocol #2u%hu");
+	"skb %p protocol #n2u%hu");
 
 notrace void probe_net_dev_xmit(struct sk_buff *skb)
 {
@@ -37,7 +37,7 @@ notrace void probe_net_dev_xmit(struct sk_buff *skb)
 void probe_net_dev_receive(struct sk_buff *skb);
 
 DEFINE_MARKER_TP(net, dev_receive, net_dev_receive, probe_net_dev_receive,
-	"skb %p protocol #2u%hu");
+	"skb %p protocol #n2u%hu");
 
 notrace void probe_net_dev_receive(struct sk_buff *skb)
 {
-- 
1.6.5.2

