From 6e00fb4825d6a0bd1620030c18b39412f591ef79 Mon Sep 17 00:00:00 2001
From: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
Date: Thu, 13 May 2010 19:27:13 -0400
Subject: [PATCH 218/391] lttng-instrumentation-panic-probes

LTTng instrumentation panic probes

By this patch, people can trace there events by marker interface.
LTTng can also trace there events.

From: Zhao Lei <zhaolei@cn.fujitsu.com>
Signed-off-by: Zhao Lei <zhaolei@cn.fujitsu.com>
Signed-off-by: Mathieu Desnoyers <mathieu.desnoyers@polymtl.ca>
---
 ltt/probes/kernel-trace.c |   21 +++++++++++++++++++++
 1 files changed, 21 insertions(+), 0 deletions(-)

diff --git a/ltt/probes/kernel-trace.c b/ltt/probes/kernel-trace.c
index 9f12b95..cb9a797 100644
--- a/ltt/probes/kernel-trace.c
+++ b/ltt/probes/kernel-trace.c
@@ -422,6 +422,27 @@ void probe_kernel_module_load(struct module *mod)
 }
 #endif
 
+void probe_kernel_panic(const char *fmt, va_list args)
+{
+	char info[64];
+	vsnprintf(info, sizeof(info), fmt, args);
+	trace_mark_tp(kernel, panic, kernel_panic, probe_kernel_panic,
+		"info %s", info);
+}
+
+void probe_kernel_kernel_kexec(struct kimage *image)
+{
+	trace_mark_tp(kernel, kernel_kexec, kernel_kernel_kexec,
+		probe_kernel_kernel_kexec, "image %p", image);
+}
+
+void probe_kernel_crash_kexec(struct kimage *image, struct pt_regs *regs)
+{
+	trace_mark_tp(kernel, crash_kexec, kernel_crash_kexec,
+		probe_kernel_crash_kexec, "image %p ip %p", image,
+		regs ? (void *)instruction_pointer(regs) : NULL);
+}
+
 /* kernel_page_fault_entry specialized tracepoint probe */
 
 void probe_kernel_page_fault_entry(struct pt_regs *regs, int trapnr,
-- 
1.6.5.2

