From 9750398e8403e043aff40d0898b460e2dfe69e26 Mon Sep 17 00:00:00 2001
From: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
Date: Thu, 13 May 2010 19:28:37 -0400
Subject: [PATCH 377/391] lttng-kref-wait-sleep

lttng kref wait sleep

Sleep with msleep() rather than busy loop when waiting for readers to release
the file handles. This should work better on UP, especially for !PREEMPT
configs.

Also add this scheme to ltt-ascii.

Signed-off-by: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
---
 ltt/ltt-ascii.c       |   14 +++++++++++++-
 ltt/ltt-relay-alloc.c |    3 ++-
 2 files changed, 15 insertions(+), 2 deletions(-)

diff --git a/ltt/ltt-ascii.c b/ltt/ltt-ascii.c
index eeb7033..e67b521 100644
--- a/ltt/ltt-ascii.c
+++ b/ltt/ltt-ascii.c
@@ -45,6 +45,7 @@
 #include <linux/debugfs.h>
 #include <linux/module.h>
 #include <linux/string.h>
+#include <linux/delay.h>
 #include <linux/slab.h>
 #include <linux/cpu.h>
 #include <linux/fs.h>
@@ -530,7 +531,18 @@ EXPORT_SYMBOL_GPL(ltt_ascii_create);
 
 void ltt_ascii_remove(struct ltt_chan *chan)
 {
-	debugfs_remove(chan->a.ascii_dentry);
+	struct dentry *dentry;
+
+	dentry = dget(chan->a.ascii_dentry);
+	debugfs_remove(dentry);
+	/* TODO: wait / wakeup instead */
+	/*
+	 * Wait for every reference to the dentry to be gone,
+	 * except us.
+	 */
+	while (atomic_read(&dentry->d_count) != 1)
+		msleep(100);
+	dput(dentry);
 }
 EXPORT_SYMBOL_GPL(ltt_ascii_remove);
 
diff --git a/ltt/ltt-relay-alloc.c b/ltt/ltt-relay-alloc.c
index 3bd2eb2..e16b065 100644
--- a/ltt/ltt-relay-alloc.c
+++ b/ltt/ltt-relay-alloc.c
@@ -16,6 +16,7 @@
 #include <linux/mm.h>
 #include <linux/cpu.h>
 #include <linux/bitops.h>
+#include <linux/delay.h>
 #include <linux/ltt-tracer.h>
 
 #include "ltt-relay-select.h"	/* for cpu hotplug */
@@ -360,7 +361,7 @@ void ltt_chan_alloc_remove_files(struct ltt_chan_alloc *chan)
 		 * except us.
 		 */
 		while (atomic_read(&dentry->d_count) != 1)
-			cpu_relax();
+			msleep(100);
 		dput(dentry);
 	}
 }
-- 
1.6.5.2

