From adb75e6abd3d327f41f79670e47f963bf9990fc2 Mon Sep 17 00:00:00 2001
From: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
Date: Thu, 13 May 2010 19:27:57 -0400
Subject: [PATCH 300/390] lttng-relay-periodic-flush-deferrable

LTTng relay periodic flush deferrable timer

Make the periodic flush timer deferrable, so we don't wake up the systems going
in deep sleep mode.

Signed-off-by: Mathieu Desnoyers <mathieu.desnoyers@polymtl.ca>
---
 ltt/ltt-relay-lockless.c |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/ltt/ltt-relay-lockless.c b/ltt/ltt-relay-lockless.c
index b74f144..259ab6c 100644
--- a/ltt/ltt-relay-lockless.c
+++ b/ltt/ltt-relay-lockless.c
@@ -452,7 +452,7 @@ static void ltt_chanbuf_start_switch_timer(struct ltt_chanbuf *buf)
 	if (!chan->switch_timer_interval)
 		return;
 
-	init_timer(&buf->switch_timer);
+	init_timer_deferrable(&buf->switch_timer);
 	buf->switch_timer.function = switch_buffer;
 	buf->switch_timer.expires = jiffies + chan->switch_timer_interval;
 	buf->switch_timer.data = (unsigned long)buf;
-- 
1.6.5.2

