From 8c1b9b44a14727fce899cb902d52322051bdd8b3 Mon Sep 17 00:00:00 2001
From: auto commit <unknown@unknown>
Date: Tue, 15 Jul 2008 14:57:10 -0400
Subject: [PATCH] auto_msg: importing lttng-instrumentation-lockdep.patch

This is an automatic import of patch lttng-instrumentation-lockdep.patch, no headers were
detected and a default message was constructed
---
 kernel/lockdep-trace.h |   32 ++++++++++++++++++++++++++++++++
 1 files changed, 32 insertions(+), 0 deletions(-)
 create mode 100644 kernel/lockdep-trace.h

diff --git a/kernel/lockdep-trace.h b/kernel/lockdep-trace.h
new file mode 100644
index 0000000..2475de2
--- /dev/null
+++ b/kernel/lockdep-trace.h
@@ -0,0 +1,32 @@
+#ifndef _LOCKDEP_TRACE_H
+#define _LOCKDEP_TRACE_H
+
+#include <linux/tracepoint.h>
+
+/*
+ * lockdep tracing must be very careful with respect to reentrancy.
+ *
+ * It should not use immediate values for activation because it involves
+ * traps called when the code patching is done.
+ */
+DEFINE_TRACE(kernel_hardirqs_on,
+	TPPROTO(void *retaddr),
+	TPARGS(retaddr));
+DEFINE_TRACE(kernel_hardirqs_off,
+	TPPROTO(void *retaddr),
+	TPARGS(retaddr));
+DEFINE_TRACE(kernel_softirqs_on,
+	TPPROTO(void *retaddr),
+	TPARGS(retaddr));
+DEFINE_TRACE(kernel_softirqs_off,
+	TPPROTO(void *retaddr),
+	TPARGS(retaddr));
+DEFINE_TRACE(kernel_lock_acquire,
+	TPPROTO(unsigned long retaddr, unsigned int subclass,
+		struct lockdep_map *lock, int trylock),
+	TPARGS(retaddr, subclass, lock, trylock));
+DEFINE_TRACE(kernel_lock_release,
+	TPPROTO(unsigned long retaddr, struct lockdep_map *lock, int nested),
+	TPARGS(retaddr, lock, nested));
+
+#endif
-- 
1.5.5.1

