From 69858d2985144bf6862a47d20579500fd63ed5ba Mon Sep 17 00:00:00 2001
From: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
Date: Thu, 13 May 2010 19:27:32 -0400
Subject: [PATCH 253/390] lttng-use-deferrable-timers

lttng tracer: use deferrable timer

Good for low-power consumption.

Signed-off-by: Mathieu Desnoyers <mathieu.desnoyers@polymtl.ca>
---
 ltt/ltt-tracer.c |    1 +
 1 files changed, 1 insertions(+), 0 deletions(-)

diff --git a/ltt/ltt-tracer.c b/ltt/ltt-tracer.c
index 61fa207..2bf5219 100644
--- a/ltt/ltt-tracer.c
+++ b/ltt/ltt-tracer.c
@@ -1272,6 +1272,7 @@ int __init ltt_init(void)
 {
 	/* Make sure no page fault can be triggered by this module */
 	vmalloc_sync_all();
+	init_timer_deferrable(&ltt_async_wakeup_timer);
 	return 0;
 }
 
-- 
1.6.5.2

