From 2a3196c94b2b312680da0389c893e64f28a94d58 Mon Sep 17 00:00:00 2001
From: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
Date: Thu, 13 May 2010 19:27:16 -0400
Subject: [PATCH 223/391] lttng-prefetch

LTTng add prefetch

- Add commit count prefetch
- Use RCU list without prefetch, because there is usually only few elements to
  the list and prefetch slows down tracing.

Signed-off-by: Mathieu Desnoyers <mathieu.desnoyers@polymtl.ca>
---
 ltt/ltt-relay-lockless.h  |    3 +++
 ltt/ltt-type-serializer.c |    7 +++++--
 2 files changed, 8 insertions(+), 2 deletions(-)

diff --git a/ltt/ltt-relay-lockless.h b/ltt/ltt-relay-lockless.h
index 9074b71..f6b21b7 100644
--- a/ltt/ltt-relay-lockless.h
+++ b/ltt/ltt-relay-lockless.h
@@ -180,6 +180,9 @@ static __inline__ int ltt_relay_try_reserve(
 	*o_old = *o_begin;
 
 	*tsc = trace_clock_read64();
+
+	prefetch(&ltt_buf->commit_count[SUBBUF_INDEX(*o_begin, rchan)]);
+
 	if (last_tsc_overflow(ltt_buf, *tsc))
 		*rflags = LTT_RFLAG_ID_SIZE_TSC;
 
diff --git a/ltt/ltt-type-serializer.c b/ltt/ltt-type-serializer.c
index 34dc21a..eda69fb 100644
--- a/ltt/ltt-type-serializer.c
+++ b/ltt/ltt-type-serializer.c
@@ -47,8 +47,11 @@ notrace void _ltt_specialized_trace(const struct marker *mdata,
 	eID = mdata->event_id;
 	chan_index = mdata->channel_id;
 
-	/* Iterate on each trace */
-	list_for_each_entry_rcu(trace, &ltt_traces.head, list) {
+	/*
+	 * Iterate on each trace, typically small number of active traces,
+	 * list iteration with prefetch is usually slower.
+	 */
+	__list_for_each_entry_rcu(trace, &ltt_traces.head, list) {
 		if (unlikely(!trace->active))
 			continue;
 		if (unlikely(!ltt_run_filter(trace, eID)))
-- 
1.6.5.2

