From 84128c036a88df903fab94c8eea3642cdd78e965 Mon Sep 17 00:00:00 2001
From: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
Date: Thu, 13 May 2010 19:26:26 -0400
Subject: [PATCH 136/390] lttng-instrumentation/lttng-use-kernel-trace-thread-flag-api

LTTng use kernel trace thread flag api

Make LTTng aware of syscall tracing enable.

Signed-off-by: Mathieu Desnoyers <mathieu.desnoyers@polymtl.ca>
---
 ltt/ltt-tracer.c |    2 ++
 1 files changed, 2 insertions(+), 0 deletions(-)

diff --git a/ltt/ltt-tracer.c b/ltt/ltt-tracer.c
index e9c926d..d5ae09a 100644
--- a/ltt/ltt-tracer.c
+++ b/ltt/ltt-tracer.c
@@ -821,6 +821,7 @@ int ltt_trace_alloc(const char *trace_name)
 	if (list_empty(&ltt_traces.head)) {
 		mod_timer(&ltt_async_wakeup_timer,
 				jiffies + LTT_PERCPU_TIMER_INTERVAL);
+		set_kernel_trace_flag_all_tasks();
 	}
 	list_add_rcu(&trace->list, &ltt_traces.head);
 	synchronize_sched();
@@ -892,6 +893,7 @@ static int _ltt_trace_destroy(struct ltt_trace_struct	*trace)
 	list_del_rcu(&trace->list);
 	synchronize_sched();
 	if (list_empty(&ltt_traces.head)) {
+		clear_kernel_trace_flag_all_tasks();
 		/*
 		 * We stop the asynchronous delivery of reader wakeup, but
 		 * we must make one last check for reader wakeups pending
-- 
1.6.5.2

