From 7c5dafedc86ea77241900ab20f5a10547c6b004e Mon Sep 17 00:00:00 2001
From: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
Date: Thu, 13 May 2010 19:27:23 -0400
Subject: [PATCH 237/390] lttng-relay-fix-type

LTTng relay fix type

Fix 32-bit build long -> size_t types.

Signed-off-by: Mathieu Desnoyers <mathieu.desnoyers@polymtl.ca>
---
 ltt/ltt-relay-irqoff.h   |    2 +-
 ltt/ltt-relay-locked.h   |    2 +-
 ltt/ltt-relay-lockless.h |    2 +-
 ltt/ltt-serialize.c      |    2 +-
 4 files changed, 4 insertions(+), 4 deletions(-)

diff --git a/ltt/ltt-relay-irqoff.h b/ltt/ltt-relay-irqoff.h
index c6b09c2..281c349 100644
--- a/ltt/ltt-relay-irqoff.h
+++ b/ltt/ltt-relay-irqoff.h
@@ -183,7 +183,7 @@ static __inline__ int ltt_relay_try_reserve(
 		size_t data_size,
 		u64 *tsc, unsigned int *rflags, int largest_align,
 		long *o_begin, long *o_end, long *o_old,
-		long *before_hdr_pad, long *size)
+		size_t *before_hdr_pad, size_t *size)
 {
 	*o_begin = local_read(&ltt_buf->offset);
 	*o_old = *o_begin;
diff --git a/ltt/ltt-relay-locked.h b/ltt/ltt-relay-locked.h
index 7e0221d..6a6adb5 100644
--- a/ltt/ltt-relay-locked.h
+++ b/ltt/ltt-relay-locked.h
@@ -172,7 +172,7 @@ static __inline__ int ltt_relay_try_reserve(
 		size_t data_size,
 		u64 *tsc, unsigned int *rflags, int largest_align,
 		long *o_begin, long *o_end, long *o_old,
-		long *before_hdr_pad, long *size)
+		size_t *before_hdr_pad, size_t *size)
 {
 	*o_begin = ltt_buf->offset;
 	*o_old = *o_begin;
diff --git a/ltt/ltt-relay-lockless.h b/ltt/ltt-relay-lockless.h
index c394dc9..1970dfa 100644
--- a/ltt/ltt-relay-lockless.h
+++ b/ltt/ltt-relay-lockless.h
@@ -181,7 +181,7 @@ static __inline__ int ltt_relay_try_reserve(
 		size_t data_size,
 		u64 *tsc, unsigned int *rflags, int largest_align,
 		long *o_begin, long *o_end, long *o_old,
-		long *before_hdr_pad, long *size)
+		size_t *before_hdr_pad, size_t *size)
 {
 	*o_begin = local_read(&ltt_buf->offset);
 	*o_old = *o_begin;
diff --git a/ltt/ltt-serialize.c b/ltt/ltt-serialize.c
index a31a98f..925b769 100644
--- a/ltt/ltt-serialize.c
+++ b/ltt/ltt-serialize.c
@@ -629,7 +629,7 @@ size_t ltt_serialize_printf(struct rchan_buf *buf, unsigned long buf_offset,
 	char outfmt[4] = "%";
 	size_t outpos = 0;
 	size_t len;
-	unsigned long msgpos = buf_offset;
+	size_t msgpos = buf_offset;
 
 	for (; *fmt ; ++fmt) {
 		switch (*fmt) {
-- 
1.6.5.2

