From c15c2f3ae30ed97918964b3d87884ba351aced66 Mon Sep 17 00:00:00 2001
From: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
Date: Thu, 13 May 2010 19:27:36 -0400
Subject: [PATCH 259/390] markers-fix-documentation

Update the documentation for DEFINE_MARKER_TP and GET_MARKER.

From: Benjamin Poirier <benjamin.poirier@polytml.ca>
Signed-off-by: Benjamin Poirier <benjamin.poirier@polytml.ca>
Signed-off-by: Mathieu Desnoyers <mathieu.desnoyers@polymtl.ca>
---
 Documentation/markers.txt |    6 +++---
 1 files changed, 3 insertions(+), 3 deletions(-)

diff --git a/Documentation/markers.txt b/Documentation/markers.txt
index 951a2f3..e25df7c 100644
--- a/Documentation/markers.txt
+++ b/Documentation/markers.txt
@@ -83,12 +83,12 @@ with tracepoint probes in a scheme like this :
 
 void probe_tracepoint_name(unsigned int arg1, struct task_struct *tsk);
 
-DEFINE_MARKER_TP(marker_eventname, tracepoint_name, probe_tracepoint_name,
-	"arg1 %u pid %d");
+DEFINE_MARKER_TP(marker_channel, marker_eventname, tracepoint_name,
+	probe_tracepoint_name, "arg1 %u pid %d");
 
 notrace void probe_tracepoint_name(unsigned int arg1, struct task_struct *tsk)
 {
-	struct marker *marker = &GET_MARKER(kernel_irq_entry);
+	struct marker *marker = &GET_MARKER(marker_channel, marker_eventname);
 	/* write data to trace buffers ... */
 }
 
-- 
1.6.5.2

