From 9d9bb4a6d9d8b9a3bcc40e896aa275e25c985bd1 Mon Sep 17 00:00:00 2001
From: Paul Gortmaker <paul.gortmaker@windriver.com>
Date: Fri, 3 May 2013 19:01:03 -0400
Subject: [PATCH] modpost: silence compile warnings caused by LTT.
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

LTT support reverts a mainline commit that deletes a block of
modpost code; specifically:

 ---
 Subject: [PATCH] markers-revert-modpost-removal

 markers revert modpost removal

 commit a8773769d1a1e08d0ca15f890515401ab3860637
 Author: Wenji Huang <wenji.huang@oracle.com>
 Date:   Mon Nov 16 13:49:55 2009 +0800

     Kbuild: clear marker out of modpost

     Remove the unnecessary functions and variables.

 Signed-off-by: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
 ---

However, the restored code causes this warning with any recent compiler:

scripts/mod/modpost.c: In function ‘get_markers’:
scripts/mod/modpost.c:1742:12: warning: ignoring return value of ‘asprintf’, declared with attribute warn_unused_result [-Wunused-result]
scripts/mod/modpost.c: In function ‘add_marker’:
scripts/mod/modpost.c:2168:10: warning: ignoring return value of ‘asprintf’, declared with attribute warn_unused_result [-Wunused-result]

It does a check on the pointer directly below, which is dubious, since
TFM says the pointer is undefined on ret<0, i.e. failure.  So the warning
was pointing to an actual bug in the old code.

Signed-off-by: Paul Gortmaker <paul.gortmaker@windriver.com>

diff --git a/scripts/mod/modpost.c b/scripts/mod/modpost.c
index 96104e2..3ae3fd6 100644
--- a/scripts/mod/modpost.c
+++ b/scripts/mod/modpost.c
@@ -1735,7 +1735,10 @@ static void get_markers(struct elf_info *info, struct module *mod)
 			const char *name = strings + sym->st_value;
 			const char *fmt = strchr(name, '\0') + 1;
 			char *line = NULL;
-			asprintf(&line, "%s\t%s\t%s\n", name, mod->name, fmt);
+			int r;
+			r = asprintf(&line, "%s\t%s\t%s\n", name, mod->name, fmt);
+			if (r < 0)
+				NOFAIL(NULL);
 			NOFAIL(line);
 			mod->markers[n++] = line;
 		}
@@ -2161,7 +2164,10 @@ static void write_dump(const char *fname)
 static void add_marker(struct module *mod, const char *name, const char *fmt)
 {
 	char *line = NULL;
-	asprintf(&line, "%s\t%s\t%s\n", name, mod->name, fmt);
+	int r;
+	r = asprintf(&line, "%s\t%s\t%s\n", name, mod->name, fmt);
+	if (r < 0)
+		NOFAIL(NULL);
 	NOFAIL(line);
 
 	mod->markers = NOFAIL(realloc(mod->markers, ((mod->nmarkers + 1) *
-- 
1.8.3.1

