From 8c9be0ac4335a5662905c12bfc31c132f2bd9efd Mon Sep 17 00:00:00 2001
From: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
Date: Thu, 13 May 2010 19:27:33 -0400
Subject: [PATCH 254/391] modpost-should-learn-about-discard-section

modprost should learn about __discard section

Fixes :

WARNING: ltt/ltt-kprobes.o (__discard): unexpected non-allocatable section.
Did you forget to use "ax"/"aw" in a .S file?
Note that for example <linux/init.h> contains
section definitions for use in .S files.

Signed-off-by: Mathieu Desnoyers <mathieu.desnoyers@polymtl.ca>
---
 scripts/mod/modpost.c |    1 +
 1 files changed, 1 insertions(+), 0 deletions(-)

diff --git a/scripts/mod/modpost.c b/scripts/mod/modpost.c
index 2092361..6aa753a 100644
--- a/scripts/mod/modpost.c
+++ b/scripts/mod/modpost.c
@@ -743,6 +743,7 @@ static const char *section_white_list[] =
 	".note*",
 	".got*",
 	".toc*",
+	"__discard",
 	NULL
 };
 
-- 
1.6.5.2

