From 272657cf46493fc478f9cb799cbe6f6d62b5acf2 Mon Sep 17 00:00:00 2001
From: Mathieu Desnoyers <mathieu.desnoyers@efficios.com>
Date: Thu, 13 May 2010 19:28:34 -0400
Subject: [PATCH 371/391] rcu-init-null-marker

---
 kernel/marker.c |    4 ++--
 1 files changed, 2 insertions(+), 2 deletions(-)

diff --git a/kernel/marker.c b/kernel/marker.c
index b72d1e0..7f94b4a 100644
--- a/kernel/marker.c
+++ b/kernel/marker.c
@@ -292,7 +292,7 @@ marker_entry_add_probe(struct marker_entry *entry,
 			GFP_KERNEL);
 	if (new == NULL)
 		return ERR_PTR(-ENOMEM);
-	INIT_RCU_HEAD(&new->rcu);
+	rcu_head_init(&new->rcu);
 	if (!old)
 		new->c[0] = entry->single;
 	else
@@ -361,7 +361,7 @@ marker_entry_remove_probe(struct marker_entry *entry,
 			      GFP_KERNEL);
 		if (new == NULL)
 			return ERR_PTR(-ENOMEM);
-		INIT_RCU_HEAD(&new->rcu);
+		rcu_head_init(&new->rcu);
 		for (i = 0; old->c[i].func; i++)
 			if ((probe && old->c[i].func != probe) ||
 			    old->c[i].probe_private != probe_private)
-- 
1.6.5.2

