From f797a8d42499e6ef0f1148d0391898aff46bf106 Mon Sep 17 00:00:00 2001
From: auto commit <unknown@unknown>
Date: Tue, 15 Jul 2008 15:19:28 -0400
Subject: [PATCH] auto_msg: importing restore-x86-nr-syscalls.patch

This is an automatic import of patch restore-x86-nr-syscalls.patch, no headers were
detected and a default message was constructed
---
 include/asm-x86/unistd_32.h |    2 ++
 1 files changed, 2 insertions(+), 0 deletions(-)

diff --git a/include/asm-x86/unistd_32.h b/include/asm-x86/unistd_32.h
index 8317d94..194211e 100644
--- a/include/asm-x86/unistd_32.h
+++ b/include/asm-x86/unistd_32.h
@@ -333,6 +333,8 @@
 #define __NR_timerfd_settime	325
 #define __NR_timerfd_gettime	326
 
+#define NR_syscalls		327
+
 #ifdef __KERNEL__
 
 #define __ARCH_WANT_IPC_PARSE_VERSION
-- 
1.5.5.1

