From: Bruce Ashfield <bruce.ashfield@windriver.com>
Subject: [PATCH]  yaffs2: fix basic compilations errors

Due to chagnes in filesystem handling, yaffs2 no longer builds. 
These changes allow it to compile (with warnings), but no runtime
checks are made.

Signed-off-by Bruce Ashfield <bruce.ashfield@windriver.com>
---

 b/fs/yaffs2/yaffs_fs.c |   14 +++++++++-----
 1 file changed, 9 insertions(+), 5 deletions(-)

--- a/fs/yaffs2/yaffs_fs.c.orig
+++ b/fs/yaffs2/yaffs_fs.c
@@ -286,8 +286,12 @@
 
 static struct super_operations yaffs_super_ops = {
 	.statfs = yaffs_statfs,
-	.read_inode = yaffs_read_inode,
-	.put_inode = yaffs_put_inode,
+	/* removed by commit 12debc4248a4a7f1873e47cda2cdd7faca80b099
+         * .read_inode = yaffs_read_inode,
+         */
+        /* removed by commit 33dcdac2df54e66c447ae03f58c95c7251aa5649
+         * .put_inode = yaffs_put_inode,
+         */
 	.put_super = yaffs_put_super,
 	.delete_inode = yaffs_delete_inode,
 	.clear_inode = yaffs_clear_inode,
@@ -848,7 +852,7 @@
 	T(YAFFS_TRACE_OS,
 	  (KERN_DEBUG "yaffs_get_inode for object %d\n", obj->objectId));
 
-	inode = iget(sb, obj->objectId);
+	inode = iget_locked(sb, obj->objectId);
 
 	/* NB Side effect: iget calls back to yaffs_read_inode(). */
 	/* iget also increments the inode's i_count */
@@ -2247,7 +2251,7 @@
 	/* Install the proc_fs entry */
 	my_proc_entry = create_proc_entry("yaffs",
 					       S_IRUGO | S_IFREG,
-					       &proc_root);
+					       &proc_root_kcore);
 
 	if (my_proc_entry) {
 		my_proc_entry->write_proc = yaffs_proc_write;
@@ -2293,7 +2297,7 @@
 	T(YAFFS_TRACE_ALWAYS, ("yaffs " __DATE__ " " __TIME__
 			       " removing. \n"));
 
-	remove_proc_entry("yaffs", &proc_root);
+	remove_proc_entry("yaffs", &proc_root_kcore);
 
 	fsinst = fs_to_install;
 
