From 6150be01745a552940406324e1d856cd16c67a3c Mon Sep 17 00:00:00 2001
From: Zumeng Chen <zumeng.chen@windriver.com>
Date: Mon, 16 Aug 2010 14:43:33 +0800
Subject: [PATCH 2/2] yaffs2: fix semaphore self-blocked

The root cause is that there is not proper semaphore using at yaffs_GrossLock
in the following scenario:
     umount                       yaffs_BackgroundThread
        |                                  |
  yaffs_put_super                if(kthread_should_stop())
        |                                  |--Yes?-> break;
  yaffs_GrossLock(Got)                   no|
        |                            yaffs_GrossLock(Blocked)
       ...                       yaffs_UpdateDirtyDirectories
 yaffs_BackgroundStop(waiting bg's stop)   |
        |                            yaffs_GrossUnLock
   yaffs_GrossUnLock                   msleep(500)

As described in yaffs_BackgroundThread head comments, it should be stopped
before umount. But currently it was called in yaffs_put_super after having
got yaffs_GrossLock as follows:
       2054         yaffs_GrossLock(dev); /* get the GrossLock */
       2055
       2056         yaffs_FlushSuperBlock(sb,1);/* will do what bg does */
       2057
       2058         if (yaffs_DeviceToContext(dev)->putSuperFunc)
       2059                 yaffs_DeviceToContext(dev)->putSuperFunc(sb);
       2060
       2062         yaffs_BackgroundStop(dev);
And what yaffs_BackgroundThread do will be done in yaffs_FlushSuperBlock.
So we call yaffs_BackgroundStop before yaffs_GrossLock in yaffs_put_super,
and yaffs_BackgroundStop will wait until yaffs_BackgroundThread exits.
Then we are safe.

Signed-Off-By: Zumeng Chen <zumeng.chen@windriver.com>
---
 fs/yaffs2/yaffs_fs.c |    4 ++--
 1 files changed, 2 insertions(+), 2 deletions(-)

diff --git a/fs/yaffs2/yaffs_fs.c b/fs/yaffs2/yaffs_fs.c
index 69b739b..302cfb8 100644
--- a/fs/yaffs2/yaffs_fs.c
+++ b/fs/yaffs2/yaffs_fs.c
@@ -2051,6 +2051,8 @@ static void yaffs_put_super(struct super_block *sb)
 {
 	yaffs_Device *dev = yaffs_SuperToDevice(sb);
 
+	yaffs_BackgroundStop(dev);
+
 	T(YAFFS_TRACE_OS, ("yaffs_put_super\n"));
 
 	yaffs_GrossLock(dev);
@@ -2060,8 +2062,6 @@ static void yaffs_put_super(struct super_block *sb)
 	if (yaffs_DeviceToContext(dev)->putSuperFunc)
 		yaffs_DeviceToContext(dev)->putSuperFunc(sb);
 
-	yaffs_BackgroundStop(dev);
-
 	yaffs_Deinitialise(dev);
 
 	yaffs_GrossUnlock(dev);
-- 
1.6.5.2

