From 7a6e712bd0edef07a386beccc75ce91c7f09f0cb Mon Sep 17 00:00:00 2001
From: Zumeng Chen <zumeng.chen@windriver.com>
Date: Mon, 16 Aug 2010 15:13:58 +0800
Subject: [PATCH 1/2] yaffs2: fix ugly strcopy and kmalloc

After having checked all the related codes, it's obvious no sense
when len = 0. And there are many scenarios call yaffs_CloneString
as follows:
	yaffs_CloneString(_Y(""));
So bypass it or consider kmalloc failure when len =0. :)

Signed-Off-By: Zumeng Chen <zumeng.chen@windriver.com>
---
 fs/yaffs2/yaffs_guts.c |   10 ++++++----
 1 files changed, 6 insertions(+), 4 deletions(-)

diff --git a/fs/yaffs2/yaffs_guts.c b/fs/yaffs2/yaffs_guts.c
index 523bfb7..0276c77 100644
--- a/fs/yaffs2/yaffs_guts.c
+++ b/fs/yaffs2/yaffs_guts.c
@@ -2404,10 +2404,12 @@ static YCHAR *yaffs_CloneString(const YCHAR *str)
 		str = _Y("");
 
 	len = yaffs_strnlen(str,YAFFS_MAX_ALIAS_LENGTH);
-	newStr = YMALLOC((len + 1) * sizeof(YCHAR));
-	if (newStr){
-		yaffs_strncpy(newStr, str,len);
-		newStr[len] = 0;
+	if(len != 0) {
+		newStr = YMALLOC((len + 1) * sizeof(YCHAR));
+		if (newStr){
+			yaffs_strncpy(newStr, str,len);
+			newStr[len] = 0;
+		}
 	}
 	return newStr;
 
-- 
1.6.5.2

