From 2a09325e58e71145616f93243b4588f888136a11 Mon Sep 17 00:00:00 2001
From: huoysh <huoysh@actions-semi.com>
Date: Fri, 25 Dec 2015 16:24:30 +0800
Subject: [PATCH 49/62] usb: BUGFIX: fix BUG00464386 & BUG00465179 :aotg udc
 connect pc problems: 1. when plug out won't exit usb
 ui; 2. command exit usb device mode cause hangup;

commit 343452fdabb177e3a5eb2b43117f88bc07337a55 from
https://github.com/xapp-le/kernel.git

Change-Id: I4febc697a9e5cacd2ca19f04de73d963bd43938d
---
 drivers/usb/aotg/aotg.h     |    1 +
 drivers/usb/aotg/aotg_mon.c |    6 ++++++
 2 files changed, 7 insertions(+), 0 deletions(-)

diff --git a/drivers/usb/aotg/aotg.h b/drivers/usb/aotg/aotg.h
index ab6849d..bc88751 100755
--- a/drivers/usb/aotg/aotg.h
+++ b/drivers/usb/aotg/aotg.h
@@ -50,6 +50,7 @@ extern struct of_device_id aotg_of_match[];
 extern int aotg_probe(struct platform_device *pdev);
 extern int aotg_remove(struct platform_device *pdev);
 void aotg_clk_enable(int id, int is_enable);
+int aotg_udc_register(int id);
 
 
 #endif
\ No newline at end of file
diff --git a/drivers/usb/aotg/aotg_mon.c b/drivers/usb/aotg/aotg_mon.c
index d707647..b07f81f 100755
--- a/drivers/usb/aotg/aotg_mon.c
+++ b/drivers/usb/aotg/aotg_mon.c
@@ -263,6 +263,9 @@ void aotg_uhost_mon_init(struct work_struct *w)
 		wake_lock_init(&aotg_uhost_mon0->aotg_wake_lock, WAKE_LOCK_SUSPEND, "aotg_wake_lock0");
 		printk("start mon 0 ......\n");
 		mod_timer(&aotg_uhost_mon0->hotplug_timer, jiffies + msecs_to_jiffies(10000));
+        	if (aotg_udc_enable[0]){
+			aotg_udc_register(0);
+		}
 	}
 	if (port_host_plug_detect[1]) {
 		aotg_uhost_mon1 = aotg_uhost_mon_alloc();
@@ -281,6 +284,9 @@ void aotg_uhost_mon_init(struct work_struct *w)
 		wake_lock_init(&aotg_uhost_mon1->aotg_wake_lock, WAKE_LOCK_SUSPEND, "aotg_wake_lock1");
 		printk("start mon 1 ......\n");
 		mod_timer(&aotg_uhost_mon1->hotplug_timer, jiffies + msecs_to_jiffies(1000));
+    		if (aotg_udc_enable[1]){
+			aotg_udc_register(1);
+		}
 	}
 
 	return;
-- 
1.7.5.4

