From c7ad18601ab05c338a32b4d3ec1baf04417a5994 Mon Sep 17 00:00:00 2001
Message-Id: <c7ad18601ab05c338a32b4d3ec1baf04417a5994.1383338101.git.Jim.Somerville@windriver.com>
From: Jim Somerville <Jim.Somerville@windriver.com>
Date: Fri, 1 Nov 2013 16:32:27 -0400
Subject: [PATCH 1/1] gpio include rt.h

Signed-off-by: Jim Somerville <Jim.Somerville@windriver.com>
---
 drivers/gpio/gpio-pca953x.c |    1 +
 1 files changed, 1 insertions(+), 0 deletions(-)

diff --git a/drivers/gpio/gpio-pca953x.c b/drivers/gpio/gpio-pca953x.c
index 23b821d..50ed3b6 100644
--- a/drivers/gpio/gpio-pca953x.c
+++ b/drivers/gpio/gpio-pca953x.c
@@ -27,6 +27,7 @@
 
 #include <linux/io.h>
 #include <linux/acpi.h>
+#include <linux/sched/rt.h>
 #define PCA953X_INPUT		0
 #define PCA953X_OUTPUT		1
 #define PCA953X_INVERT		2
-- 
1.7.4.1

