From df9cf08bd4c8334582a8d38995a17921dc970854 Mon Sep 17 00:00:00 2001
From: Jim Somerville <Jim.Somerville@windriver.com>
Date: Fri, 1 Nov 2013 16:32:27 -0400
Subject: [PATCH] gpio include rt.h

Signed-off-by: Jim Somerville <Jim.Somerville@windriver.com>

diff --git a/drivers/gpio/gpio-pca953x.c b/drivers/gpio/gpio-pca953x.c
index d3f3e8f..2956e3c 100644
--- a/drivers/gpio/gpio-pca953x.c
+++ b/drivers/gpio/gpio-pca953x.c
@@ -15,6 +15,7 @@
 #include <linux/init.h>
 #include <linux/gpio.h>
 #include <linux/interrupt.h>
+#include <linux/sched/rt.h>
 #include <linux/irq.h>
 #include <linux/i2c.h>
 #include <linux/i2c/pca953x.h>
-- 
1.8.4.1

