From eb460c64c3f549802d8fa194c5469cece31de92e Mon Sep 17 00:00:00 2001
From: Jiang Fang <jiang.xx.fang@gmail.com>
Date: Sat, 2 Feb 2013 14:13:42 -0800
Subject: [PATCH] rcu: Fix spacing problem

commit b5b393601dbce0bce3f0717f29e6c8d1cf0295da upstream.

Signed-off-by: Jiang Fang <jiang.xx.fang@gmail.com>
Signed-off-by: Paul E. McKenney <paulmck@linux.vnet.ibm.com>
Signed-off-by: Paul Gortmaker <paul.gortmaker@windriver.com>

diff --git a/kernel/rcutree.c b/kernel/rcutree.c
index ed1db61..3390fda 100644
--- a/kernel/rcutree.c
+++ b/kernel/rcutree.c
@@ -3362,7 +3362,7 @@ void __init rcu_init(void)
 	rcu_init_one(&rcu_sched_state, &rcu_sched_data);
 	rcu_init_one(&rcu_bh_state, &rcu_bh_data);
 	__rcu_init_preempt();
-	 open_softirq(RCU_SOFTIRQ, rcu_process_callbacks);
+	open_softirq(RCU_SOFTIRQ, rcu_process_callbacks);
 
 	/*
 	 * We don't need protection against CPU-hotplug here because
-- 
1.8.3.1

