From 80ffc50d60587caf0f6a6cdfd22819de8f7435b3 Mon Sep 17 00:00:00 2001
From: "Paul E. McKenney" <paulmck@linux.vnet.ibm.com>
Date: Mon, 28 Oct 2013 08:57:29 -0700
Subject: [PATCH] rcutorture: Move checkarg to functions.sh

commit 32caccb8f47c032df55e77102ea43c627f1ca507 upstream.

Signed-off-by: Paul E. McKenney <paulmck@linux.vnet.ibm.com>
Cc: Greg KH <gregkh@linuxfoundation.org>
Signed-off-by: Paul Gortmaker <paul.gortmaker@windriver.com>

diff --git a/tools/testing/selftests/rcutorture/bin/functions.sh b/tools/testing/selftests/rcutorture/bin/functions.sh
index 963b6f0..3052f1b 100644
--- a/tools/testing/selftests/rcutorture/bin/functions.sh
+++ b/tools/testing/selftests/rcutorture/bin/functions.sh
@@ -28,6 +28,30 @@ bootparam_hotplug_cpu () {
 	echo "$1" | grep -q "rcutorture\.onoff_"
 }
 
+# checkarg --argname argtype $# arg mustmatch cannotmatch
+#
+# Checks the specified argument "arg" against the mustmatch and cannotmatch
+# patterns.
+checkarg () {
+	if test $3 -le 1
+	then
+		echo $1 needs argument $2 matching \"$5\"
+		usage
+	fi
+	if echo "$4" | grep -q -e "$5"
+	then
+		:
+	else
+		echo $1 $2 \"$4\" must match \"$5\"
+		usage
+	fi
+	if echo "$4" | grep -q -e "$6"
+	then
+		echo $1 $2 \"$4\" must not match \"$6\"
+		usage
+	fi
+}
+
 # configfrag_boot_params bootparam-string config-fragment-file
 #
 # Adds boot parameters from the .boot file, if any.
diff --git a/tools/testing/selftests/rcutorture/bin/kvm.sh b/tools/testing/selftests/rcutorture/bin/kvm.sh
index 0783ec9..7ba375e 100644
--- a/tools/testing/selftests/rcutorture/bin/kvm.sh
+++ b/tools/testing/selftests/rcutorture/bin/kvm.sh
@@ -32,6 +32,7 @@ args="$*"
 
 dur=30
 KVM="`pwd`/tools/testing/selftests/rcutorture"; export KVM
+PATH=${KVM}/bin:$PATH; export PATH
 builddir="${KVM}/b1"
 RCU_INITRD="$KVM/initrd"; export RCU_INITRD
 RCU_KMAKE_ARG=""; export RCU_KMAKE_ARG
@@ -40,6 +41,8 @@ configs=""
 ds=`date +%Y.%m.%d-%H:%M:%S`
 kversion=""
 
+. functions.sh
+
 usage () {
 	echo "Usage: $scriptname optional arguments:"
 	echo "       --bootargs kernel-boot-arguments"
@@ -60,27 +63,6 @@ usage () {
 	exit 1
 }
 
-# checkarg --argname argtype $# arg mustmatch cannotmatch
-checkarg () {
-	if test $3 -le 1
-	then
-		echo $1 needs argument $2 matching \"$5\"
-		usage
-	fi
-	if echo "$4" | grep -q -e "$5"
-	then
-		:
-	else
-		echo $1 $2 \"$4\" must match \"$5\"
-		usage
-	fi
-	if echo "$4" | grep -q -e "$6"
-	then
-		echo $1 $2 \"$4\" must not match \"$6\"
-		usage
-	fi
-}
-
 while test $# -gt 0
 do
 	case "$1" in
@@ -164,7 +146,6 @@ do
 	shift
 done
 
-PATH=${KVM}/bin:$PATH; export PATH
 CONFIGFRAG=${KVM}/configs; export CONFIGFRAG
 KVPATH=${CONFIGFRAG}/$kversion; export KVPATH
 
-- 
1.8.2.3

