From b16089096947a34a62baed8e135911e1ac92736e Mon Sep 17 00:00:00 2001
From: Paul Gortmaker <paul.gortmaker@windriver.com>
Date: Thu, 19 Sep 2013 12:13:20 -0400
Subject: [PATCH] softirq: notify when masking messages

Signed-off-by: Paul Gortmaker <paul.gortmaker@windriver.com>

diff --git a/kernel/softirq.c b/kernel/softirq.c
index 541adf3a3f7b..9ed2d398cd7f 100644
--- a/kernel/softirq.c
+++ b/kernel/softirq.c
@@ -97,6 +97,7 @@ static inline void softirq_clr_runner(unsigned int sirq)
  *
  * This code is called with interrupts disabled.
  */
+#define PENDING_MAX_MSG 10
 void softirq_check_pending_idle(void)
 {
 	static int rate_limit;
@@ -104,7 +105,7 @@ void softirq_check_pending_idle(void)
 	u32 warnpending;
 	int i;
 
-	if (rate_limit >= 10)
+	if (rate_limit >= PENDING_MAX_MSG)
 		return;
 
 	warnpending = local_softirq_pending() & SOFTIRQ_STOP_IDLE_MASK;
@@ -132,6 +133,8 @@ void softirq_check_pending_idle(void)
 		printk(KERN_ERR "NOHZ: local_softirq_pending %02x\n",
 		       warnpending);
 		rate_limit++;
+		if (ratelimit == PENDING_MAX_MSG)
+			pr_warn("NOHZ: masking local_softirq_pending messages\n ");
 	}
 }
 # else
-- 
1.9.0

