From 8b7b0a0aa35676fca1eea456392951eaf4484661 Mon Sep 17 00:00:00 2001
From: Ohad Ben-Cohen <ohad@wizery.com>
Date: Tue, 19 Jun 2012 10:22:35 +0300
Subject: [PATCH] MAINTAINERS: add remoteproc's git

commit 6bb697b6b06041d2d0affc862b17b4b443a107e0 upstream.

Add remoteproc's git tree to the MAINTAINERS entry

Signed-off-by: Ohad Ben-Cohen <ohad@wizery.com>
Signed-off-by: Paul Barrette <paul.barrette@windriver.com>

diff --git a/MAINTAINERS b/MAINTAINERS
index e01e1b4..fb7246e 100644
--- a/MAINTAINERS
+++ b/MAINTAINERS
@@ -5693,6 +5693,7 @@ F:	include/linux/regmap.h
 
 REMOTE PROCESSOR (REMOTEPROC) SUBSYSTEM
 M:	Ohad Ben-Cohen <ohad@wizery.com>
+T:	git git://git.kernel.org/pub/scm/linux/kernel/git/ohad/remoteproc.git
 S:	Maintained
 F:	drivers/remoteproc/
 F:	Documentation/remoteproc.txt
-- 
1.7.9.5

