From 0571cf23a10e7894f4447256f38689603c62e16b Mon Sep 17 00:00:00 2001
From: Simon Horman <horms+renesas@verge.net.au>
Date: Tue, 20 Oct 2015 14:06:31 +0900
Subject: [PATCH 0322/2066] MAINTAINERS: Remove link to oss.renesas.com which
 is closed

upstream 804b0d7011ee464d8ca8c52dc38525572e526ab4 commit

There doesn't seem much value in keeping this link listed
in the MAINTAINERS file any more.

Signed-off-by: Simon Horman <horms+renesas@verge.net.au>
---
 MAINTAINERS |    1 -
 1 files changed, 0 insertions(+), 1 deletions(-)

diff --git a/MAINTAINERS b/MAINTAINERS
index 948c8e9..8f7bd64 100644
--- a/MAINTAINERS
+++ b/MAINTAINERS
@@ -1432,7 +1432,6 @@ ARM/SHMOBILE ARM ARCHITECTURE
 M:	Simon Horman <horms@verge.net.au>
 M:	Magnus Damm <magnus.damm@gmail.com>
 L:	linux-sh@vger.kernel.org
-W:	http://oss.renesas.com
 Q:	http://patchwork.kernel.org/project/linux-sh/list/
 T:	git git://git.kernel.org/pub/scm/linux/kernel/git/horms/renesas.git next
 S:	Supported
-- 
1.7.5.4

