From 56dcd9c5114182454c4e07a5d8ac8d7dc3896484 Mon Sep 17 00:00:00 2001
From: Geert Uytterhoeven <geert+renesas@glider.be>
Date: Thu, 14 Apr 2016 11:08:09 +0200
Subject: [PATCH 0757/2066] serial: doc: .(un)throttle() depends on hardware
 assisted flow control

upstream a3bedc3bdc6e640141157636e253d63279860e23 commit

Document that .throttle() and .unthrottle() are relevant only if
hardware assisted flow control is enabled.

Signed-off-by: Geert Uytterhoeven <geert+renesas@glider.be>
Signed-off-by: Jonathan Corbet <corbet@lwn.net>
---
 Documentation/serial/driver |    2 ++
 1 files changed, 2 insertions(+), 0 deletions(-)

diff --git a/Documentation/serial/driver b/Documentation/serial/driver
index 19f842c..15f7d8b 100644
--- a/Documentation/serial/driver
+++ b/Documentation/serial/driver
@@ -135,6 +135,7 @@ hardware.
 	Notify the serial driver that input buffers for the line discipline are
 	close to full, and it should somehow signal that no more characters
 	should be sent to the serial port.
+	This will be called only if hardware assisted flow control is enabled.
 
 	Locking: none.
 
@@ -142,6 +143,7 @@ hardware.
 	Notify the serial driver that characters can now be sent to the serial
 	port without fear of overrunning the input buffers of the line
 	disciplines.
+	This will be called only if hardware assisted flow control is enabled.
 
 	Locking: none.
 
-- 
1.7.5.4

