From 7aa10451db4c9b47d1f42eda9d9bfc387b106114 Mon Sep 17 00:00:00 2001
From: Geert Uytterhoeven <geert+renesas@glider.be>
Date: Mon, 14 Mar 2016 16:16:17 +0100
Subject: [PATCH 0736/2066] serial: doc: Correct return type of
 mctrl_gpio_to_gpiod()

upstream 4817ebb144ffa5a1a2bc84b89ed9655dbe6c4502 commit

Signed-off-by: Geert Uytterhoeven <geert+renesas@glider.be>
Signed-off-by: Jonathan Corbet <corbet@lwn.net>
---
 Documentation/serial/driver |    3 ++-
 1 files changed, 2 insertions(+), 1 deletions(-)

diff --git a/Documentation/serial/driver b/Documentation/serial/driver
index e73ca36..3517e71 100644
--- a/Documentation/serial/driver
+++ b/Documentation/serial/driver
@@ -460,7 +460,8 @@ mctrl_gpio_free(dev, gpios):
 	this function.
 
 mctrl_gpio_to_gpiod(gpios, gidx)
-	This returns the gpio structure associated to the modem line index.
+	This returns the gpio_desc structure associated to the modem line
+	index.
 
 mctrl_gpio_set(gpios, mctrl):
 	This will sets the gpios according to the mctrl state.
-- 
1.7.5.4

