From a6dbd445b484e272ff70b87cbc8979fbf7d1dfe2 Mon Sep 17 00:00:00 2001
From: Geert Uytterhoeven <geert+renesas@glider.be>
Date: Thu, 14 Apr 2016 11:08:10 +0200
Subject: [PATCH 0758/2066] serial: doc: .(un)throttle() are serialized by the
 tty layer

upstream 18717b06ee4897c2cc5e92783386c4b304b9fce9 commit

Document that .(un)throttle() are serialized with each other, and with
termios modification by the tty layer.

Reported-by: Peter Hurley <peter@hurleysoftware.com>
Signed-off-by: Geert Uytterhoeven <geert+renesas@glider.be>
Signed-off-by: Jonathan Corbet <corbet@lwn.net>
---
 Documentation/serial/driver |    6 ++++--
 1 files changed, 4 insertions(+), 2 deletions(-)

diff --git a/Documentation/serial/driver b/Documentation/serial/driver
index 15f7d8b..57a9e61 100644
--- a/Documentation/serial/driver
+++ b/Documentation/serial/driver
@@ -137,7 +137,8 @@ hardware.
 	should be sent to the serial port.
 	This will be called only if hardware assisted flow control is enabled.
 
-	Locking: none.
+	Locking: serialized with .unthrottle() and termios modification by the
+		 tty layer.
 
   unthrottle(port)
 	Notify the serial driver that characters can now be sent to the serial
@@ -145,7 +146,8 @@ hardware.
 	disciplines.
 	This will be called only if hardware assisted flow control is enabled.
 
-	Locking: none.
+	Locking: serialized with .throttle() and termios modification by the
+		 tty layer.
 
   send_xchar(port,ch)
 	Transmit a high priority character, even if the port is stopped.
-- 
1.7.5.4

