From 14c50af26925b327c4a0892f97e3dd467f12dae4 Mon Sep 17 00:00:00 2001
From: Geert Uytterhoeven <geert+renesas@glider.be>
Date: Thu, 14 Apr 2016 11:08:11 +0200
Subject: [PATCH 0759/2066] serial: doc: .break_ctl() may sleep

upstream 101ca9ccdd30c46afc5c9a30e7acdcbeaa330fa1 commit

break_ctl() is not called from any sort of atomic context, so there is no
problem with it sleeping.

Reported-by: Peter Hurley <peter@hurleysoftware.com>
Signed-off-by: Geert Uytterhoeven <geert+renesas@glider.be>
Signed-off-by: Jonathan Corbet <corbet@lwn.net>
---
 Documentation/serial/driver |    1 -
 1 files changed, 0 insertions(+), 1 deletions(-)

diff --git a/Documentation/serial/driver b/Documentation/serial/driver
index 57a9e61..d7cec83 100644
--- a/Documentation/serial/driver
+++ b/Documentation/serial/driver
@@ -187,7 +187,6 @@ hardware.
 	ctl.
 
 	Locking: caller holds port->mutex
-	This call must not sleep
 
   startup(port)
 	Grab any interrupt resources and initialise any low level driver
-- 
1.7.5.4

