From a83fed1aecba2fe05720ff1c9fa0a109c3c4dfaf Mon Sep 17 00:00:00 2001
From: Paul Gortmaker <paul.gortmaker@windriver.com>
Date: Thu, 18 Nov 2010 17:52:46 -0500
Subject: [PATCH] Linux 2.6.34.8

Signed-off-by: Paul Gortmaker <paul.gortmaker@windriver.com>
---
 Makefile |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/Makefile b/Makefile
index 290f058..9f54405 100644
--- a/Makefile
+++ b/Makefile
@@ -1,7 +1,7 @@
 VERSION = 2
 PATCHLEVEL = 6
 SUBLEVEL = 34
-EXTRAVERSION = .7
+EXTRAVERSION = .8
 NAME = Sheep on Meth
 
 # *DOCUMENTATION*
-- 
1.7.0.4

