From dd33c32c2b8451962a29806a98834ae475676b22 Mon Sep 17 00:00:00 2001
From: Paul Gortmaker <paul.gortmaker@windriver.com>
Date: Wed, 17 Nov 2010 19:13:58 -0500
Subject: [PATCH] Linux 2.6.34.10

Signed-off-by: Paul Gortmaker <paul.gortmaker@windriver.com>
---
 Makefile |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/Makefile b/Makefile
index 6300664..833008c 100644
--- a/Makefile
+++ b/Makefile
@@ -1,7 +1,7 @@
 VERSION = 2
 PATCHLEVEL = 6
 SUBLEVEL = 34
-EXTRAVERSION = .9
+EXTRAVERSION = .10
 NAME = Sheep on Meth
 
 # *DOCUMENTATION*
-- 
1.7.0.4

