From f32d19c5a4919b9557a0e9db8dcd153e8898118d Mon Sep 17 00:00:00 2001
From: Paul Gortmaker <paul.gortmaker@windriver.com>
Date: Wed, 1 Dec 2010 15:45:53 -0500
Subject: [PATCH] Linux 2.6.34.12

Signed-off-by: Paul Gortmaker <paul.gortmaker@windriver.com>
---
 Makefile |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/Makefile b/Makefile
index 3889155..eb9038f 100644
--- a/Makefile
+++ b/Makefile
@@ -1,7 +1,7 @@
 VERSION = 2
 PATCHLEVEL = 6
 SUBLEVEL = 34
-EXTRAVERSION = .11
+EXTRAVERSION = .12
 NAME = Sheep on Meth
 
 # *DOCUMENTATION*
-- 
1.7.3.GIT

