From 9a68559b4b325fb823661f0e39eb3ad7d674f994 Mon Sep 17 00:00:00 2001
From: Paul Gortmaker <paul.gortmaker@windriver.com>
Date: Wed, 17 Nov 2010 22:13:58 -0500
Subject: [PATCH] Linux 2.6.34.11

Signed-off-by: Paul Gortmaker <paul.gortmaker@windriver.com>
---
 Makefile |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/Makefile b/Makefile
index 833008c..3889155 100644
--- a/Makefile
+++ b/Makefile
@@ -1,7 +1,7 @@
 VERSION = 2
 PATCHLEVEL = 6
 SUBLEVEL = 34
-EXTRAVERSION = .10
+EXTRAVERSION = .11
 NAME = Sheep on Meth
 
 # *DOCUMENTATION*
-- 
1.7.0.4

