From 586b37180b0206464d418067cf8b7d2567468f9c Mon Sep 17 00:00:00 2001
From: Paul Gortmaker <paul.gortmaker@windriver.com>
Date: Wed, 6 Jul 2011 10:17:28 -0400
Subject: [PATCH] stable: update 34-longterm maintainer entry

Per Greg's advice, taken from a random discussion on LKML:

   > Ok - hmmm maintainers in 2.6.35.13 still lists you - but I'll look
   > at a later kernel maintainers file.

   It's not really obvious who is in charge, the -longterm maintainer
   really should put their name there.  I think for .35 it's Andi Kleen.

Signed-off-by: Paul Gortmaker <paul.gortmaker@windriver.com>
Cc: Greg Kroah-Hartman <greg@kroah.com>

diff --git a/MAINTAINERS b/MAINTAINERS
index 502b380..2777bf2 100644
--- a/MAINTAINERS
+++ b/MAINTAINERS
@@ -5304,9 +5304,8 @@ M:	Jan-Benedict Glaw <jbglaw@lug-owl.de>
 S:	Maintained
 F:	arch/alpha/kernel/srm_env.c
 
-STABLE BRANCH
-M:	Greg Kroah-Hartman <greg@kroah.com>
-M:	Chris Wright <chrisw@sous-sol.org>
+STABLE BRANCH (2.6.34 only)
+M:	Paul Gortmaker <paul.gortmaker@windriver.com>
 L:	stable@kernel.org
 S:	Maintained
 
-- 
1.7.6

