From ddb7951e4b9be1e24b617c63c9650adddee85000 Mon Sep 17 00:00:00 2001
From: Jason Wessel <jason.wessel@windriver.com>
Date: Mon, 22 Sep 2008 07:14:02 -0500
Subject: [PATCH] Add ability to unoptimize module compiles

This patch adds in the ability to unoptimize a single kernel module to
make source stepping more linear and not optimize out variables that
the developer might like to inspect.  It used with adding
COPTIMIZE=-O0 to the build line.

Signed-off-by: Jason Wessel <jason.wessel@windriver.com>
---
 Makefile |    7 +++++--
 1 files changed, 5 insertions(+), 2 deletions(-)

diff --git a/Makefile b/Makefile
index cab896a..c7f3dd0 100644
--- a/Makefile
+++ b/Makefile
@@ -510,10 +510,13 @@ endif # $(dot-config)
 all: vmlinux
 
 ifdef CONFIG_CC_OPTIMIZE_FOR_SIZE
-KBUILD_CFLAGS	+= -Os
+COPTIMIZE	+= -Os
 else
-KBUILD_CFLAGS	+= -O2
+COPTIMIZE	+= -O2
 endif
+# COPTIMIZE may be overridden on the make command line with
+# 	make ... COPTIMIZE=""
+KBUILD_CFLAGS	+= $(COPTIMIZE)
 
 include $(srctree)/arch/$(SRCARCH)/Makefile
 
-- 
1.5.5.1

