From ff429ae628993c1987c3e1ad9d8e9f48c54d1b47 Mon Sep 17 00:00:00 2001
From: Shinya Kuribayashi <shinya.kuribayashi@necel.com>
Date: Tue, 10 Nov 2009 02:41:31 -0800
Subject: [PATCH 1/4] MIPS: nec_emma3pf: emma3pf.h: Style cleanups

* Remove useless REGBASE.
* Switch around "base address (foo_BASE)" and "offset" in EMMA_foo.
* No logical changes.

Signed-off-by: Shinya Kuribayashi <shinya.kuribayashi@necel.com>
---
 include/asm-mips/emma/emma3pf.h |  149 +++++++++++++++++++--------------------
 1 files changed, 72 insertions(+), 77 deletions(-)

diff --git a/include/asm-mips/emma/emma3pf.h b/include/asm-mips/emma/emma3pf.h
index 4c372f8..9e48d0b 100644
--- a/include/asm-mips/emma/emma3pf.h
+++ b/include/asm-mips/emma/emma3pf.h
@@ -39,86 +39,81 @@
 #ifndef __ASM_EMMA_EMMA3PF_H
 #define __ASM_EMMA_EMMA3PF_H
 
-/*
- * EMMA3PF registers
- */
-#define REGBASE 0x10000000
-
-#define EMMA_BHIF_BASE		(0x0000+REGBASE)
-#define EMMA_ETHER_BASE		(0x012000+REGBASE)
-#define EMMA_PBRD_BASE		(0x100000+REGBASE)
-#define EMMA_TIMER_BASE		(0x110000+REGBASE)
-#define EMMA_PCI_BASE		(0x200000+REGBASE)
-#define EMMA_SATA_PORT_BASE	(0x204000+REGBASE)
+#define EMMA_BHIF_BASE			0x10000000
+#define EMMA_ETHER_BASE			0x10012000
+#define EMMA_PBRD_BASE			0x10100000
+#define EMMA_TIMER_BASE			0x10110000
+#define EMMA_PCI_BASE			0x10200000
+#define EMMA_SATA_PORT_BASE		0x10204000
 
-#define EMMA_BHIF_STRAP_0	(0x0010+EMMA_BHIF_BASE)
-#define EMMA_BHIF_CLKSEL	(0x001c+EMMA_BHIF_BASE)
-#define EMMA_BHIF_ALTERNATE0	(0x0028+EMMA_BHIF_BASE)
-#define EMMA_BHIF_ALTERNATE1	(0x0018+EMMA_BHIF_BASE)
-#define EMMA_BHIF_INT_ST_0	(0x0030+EMMA_BHIF_BASE)
-#define EMMA_BHIF_INT_ST_1	(0x0034+EMMA_BHIF_BASE)
-#define EMMA_BHIF_INT_ST_2	(0x0038+EMMA_BHIF_BASE)
-#define EMMA_BHIF_INT_EN_0	(0x0040+EMMA_BHIF_BASE)
-#define EMMA_BHIF_INT_EN_1	(0x0044+EMMA_BHIF_BASE)
-#define EMMA_BHIF_INT_EN_2	(0x0048+EMMA_BHIF_BASE)
-#define EMMA_BHIF_SUB_INT_ST_0	(0x0100+EMMA_BHIF_BASE)
-#define EMMA_BHIF_SUB_INT_ST_1	(0x0104+EMMA_BHIF_BASE)
-#define EMMA_BHIF_SUB_INT_ST_2	(0x0108+EMMA_BHIF_BASE)
-#define EMMA_BHIF_SUB_INT_EN_0	(0x0110+EMMA_BHIF_BASE)
-#define EMMA_BHIF_SUB_INT_EN_1	(0x0114+EMMA_BHIF_BASE)
-#define EMMA_BHIF_SUB_INT_EN_2	(0x0118+EMMA_BHIF_BASE)
-#define EMMA_BHIF_SW_INT	(0x01a0+EMMA_BHIF_BASE)
-#define EMMA_BHIF_SW_INT_EN	(0x01a4+EMMA_BHIF_BASE)
-#define EMMA_BHIF_SW_INT_CLR	(0x01a8+EMMA_BHIF_BASE)
+#define EMMA_BHIF_STRAP_0		(EMMA_BHIF_BASE + 0x0010)
+#define EMMA_BHIF_CLKSEL		(EMMA_BHIF_BASE + 0x001c)
+#define EMMA_BHIF_ALTERNATE0		(EMMA_BHIF_BASE + 0x0028)
+#define EMMA_BHIF_ALTERNATE1		(EMMA_BHIF_BASE + 0x0018)
+#define EMMA_BHIF_INT_ST_0		(EMMA_BHIF_BASE + 0x0030)
+#define EMMA_BHIF_INT_ST_1		(EMMA_BHIF_BASE + 0x0034)
+#define EMMA_BHIF_INT_ST_2		(EMMA_BHIF_BASE + 0x0038)
+#define EMMA_BHIF_INT_EN_0		(EMMA_BHIF_BASE + 0x0040)
+#define EMMA_BHIF_INT_EN_1		(EMMA_BHIF_BASE + 0x0044)
+#define EMMA_BHIF_INT_EN_2		(EMMA_BHIF_BASE + 0x0048)
+#define EMMA_BHIF_SUB_INT_ST_0		(EMMA_BHIF_BASE + 0x0100)
+#define EMMA_BHIF_SUB_INT_ST_1		(EMMA_BHIF_BASE + 0x0104)
+#define EMMA_BHIF_SUB_INT_ST_2		(EMMA_BHIF_BASE + 0x0108)
+#define EMMA_BHIF_SUB_INT_EN_0		(EMMA_BHIF_BASE + 0x0110)
+#define EMMA_BHIF_SUB_INT_EN_1		(EMMA_BHIF_BASE + 0x0114)
+#define EMMA_BHIF_SUB_INT_EN_2		(EMMA_BHIF_BASE + 0x0118)
+#define EMMA_BHIF_SW_INT		(EMMA_BHIF_BASE + 0x01a0)
+#define EMMA_BHIF_SW_INT_EN		(EMMA_BHIF_BASE + 0x01a4)
+#define EMMA_BHIF_SW_INT_CLR		(EMMA_BHIF_BASE + 0x01a8)
 
-#define EMMA_PUART0_BASE	(0x1000+EMMA_PBRD_BASE)
-#define EMMA_PUART1_BASE	(0x2000+EMMA_PBRD_BASE)
-#define EMMA_PIIC0_BASE		(0x4000+EMMA_PBRD_BASE)
-#define EMMA_PIIC1_BASE		(0x5000+EMMA_PBRD_BASE)
-#define EMMA_PIIC2_BASE		(0x6000+EMMA_PBRD_BASE)
+#define EMMA_PUART0_BASE		(EMMA_PBRD_BASE + 0x1000)
+#define EMMA_PUART1_BASE		(EMMA_PBRD_BASE + 0x2000)
+#define EMMA_PIIC0_BASE			(EMMA_PBRD_BASE + 0x4000)
+#define EMMA_PIIC1_BASE			(EMMA_PBRD_BASE + 0x5000)
+#define EMMA_PIIC2_BASE			(EMMA_PBRD_BASE + 0x6000)
 
-#define EMMA_GPIO0_BASE		(0x0c00+EMMA_TIMER_BASE)
-#define EMMA_GPIO0_DIR		(0xc0+EMMA_GPIO0_BASE)
-#define EMMA_GPIO0_INT		(0xc8+EMMA_GPIO0_BASE)
-#define EMMA_GPIO0_INT_CPUMEN	(0xd0+EMMA_GPIO0_BASE)
-#define EMMA_GPIO0_INT_CPUSEN	(0xd8+EMMA_GPIO0_BASE)
-#define EMMA_GPIO0_INT_MODE	(0xe0+EMMA_GPIO0_BASE)
-#define EMMA_GPIO0_INT_CND_A	(0xe8+EMMA_GPIO0_BASE)
-#define EMMA_GPIO0_INT_CND_B	(0xf0+EMMA_GPIO0_BASE)
+#define EMMA_GPIO0_BASE			(EMMA_TIMER_BASE + 0x0c00)
+#define EMMA_GPIO0_DIR			(EMMA_GPIO0_BASE + 0xc0)
+#define EMMA_GPIO0_INT			(EMMA_GPIO0_BASE + 0xc8)
+#define EMMA_GPIO0_INT_CPUMEN		(EMMA_GPIO0_BASE + 0xd0)
+#define EMMA_GPIO0_INT_CPUSEN		(EMMA_GPIO0_BASE + 0xd8)
+#define EMMA_GPIO0_INT_MODE		(EMMA_GPIO0_BASE + 0xe0)
+#define EMMA_GPIO0_INT_CND_A		(EMMA_GPIO0_BASE + 0xe8)
+#define EMMA_GPIO0_INT_CND_B		(EMMA_GPIO0_BASE + 0xf0)
 
-#define EMMA_GPIO1_BASE		(0x0d00+EMMA_TIMER_BASE)
-#define EMMA_GPIO1_DIR		(0xc0+EMMA_GPIO1_BASE)
-#define EMMA_GPIO1_INT		(0xc8+EMMA_GPIO1_BASE)
-#define EMMA_GPIO1_INT_CPUMEN	(0xd0+EMMA_GPIO1_BASE)
-#define EMMA_GPIO1_INT_CPUSEN	(0xd8+EMMA_GPIO1_BASE)
-#define EMMA_GPIO1_INT_MODE	(0xe0+EMMA_GPIO1_BASE)
-#define EMMA_GPIO1_INT_CND_A	(0xe8+EMMA_GPIO1_BASE)
-#define EMMA_GPIO1_INT_CND_B	(0xf0+EMMA_GPIO1_BASE)
+#define EMMA_GPIO1_BASE			(EMMA_TIMER_BASE + 0x0d00)
+#define EMMA_GPIO1_DIR			(EMMA_GPIO1_BASE + 0xc0)
+#define EMMA_GPIO1_INT			(EMMA_GPIO1_BASE + 0xc8)
+#define EMMA_GPIO1_INT_CPUMEN		(EMMA_GPIO1_BASE + 0xd0)
+#define EMMA_GPIO1_INT_CPUSEN		(EMMA_GPIO1_BASE + 0xd8)
+#define EMMA_GPIO1_INT_MODE		(EMMA_GPIO1_BASE + 0xe0)
+#define EMMA_GPIO1_INT_CND_A		(EMMA_GPIO1_BASE + 0xe8)
+#define EMMA_GPIO1_INT_CND_B		(EMMA_GPIO1_BASE + 0xf0)
 
-#define EMMA_GPIO_INT_ST	EMMA_GPIO0_INT
-#define EMMA_GPIO_INT_MASK	EMMA_GPIO0_INT_CPUMEN
-#define EMMA_GPIO_SUB_INT_MASK	EMMA_GPIO0_INT_CPUSEN
+#define EMMA_GPIO_INT_ST		EMMA_GPIO0_INT
+#define EMMA_GPIO_INT_MASK		EMMA_GPIO0_INT_CPUMEN
+#define EMMA_GPIO_SUB_INT_MASK		EMMA_GPIO0_INT_CPUSEN
 #define EMMA_GPIO_INT_ST_OFFSET		(EMMA_GPIO0_INT - EMMA_GPIO1_INT)
 #define EMMA_GPIO_INT_MASK_OFFSET	EMMA_GPIO_INT_ST_OFFSET
 
-#define EMMA_PCI_CONTROL	(0x0000+EMMA_PCI_BASE)
-#define EMMA_PCI_ARBIT_CTR	(0x0004+EMMA_PCI_BASE)
-#define EMMA_PCI_IWIN0_CTR	(0x0010+EMMA_PCI_BASE)
-#define EMMA_PCI_IWIN1_CTR	(0x0014+EMMA_PCI_BASE)
-#define EMMA_PCI_INIT_ESWP	(0x0018+EMMA_PCI_BASE)
-#define EMMA_PCI_INT		(0x0020+EMMA_PCI_BASE)
-#define EMMA_PCI_INT_EN		(0x0024+EMMA_PCI_BASE)
-#define EMMA_PCI_TWIN_CTR	(0x0030+EMMA_PCI_BASE)
-#define EMMA_PCI_TWIN_BADR	(0x0034+EMMA_PCI_BASE)
-#define EMMA_PCI_TWIN0_DADR	(0x0038+EMMA_PCI_BASE)
-#define EMMA_PCI_TWIN1_DADR	(0x003c+EMMA_PCI_BASE)
+#define EMMA_PCI_CONTROL		(EMMA_PCI_BASE + 0x0000)
+#define EMMA_PCI_ARBIT_CTR		(EMMA_PCI_BASE + 0x0004)
+#define EMMA_PCI_IWIN0_CTR		(EMMA_PCI_BASE + 0x0010)
+#define EMMA_PCI_IWIN1_CTR		(EMMA_PCI_BASE + 0x0014)
+#define EMMA_PCI_INIT_ESWP		(EMMA_PCI_BASE + 0x0018)
+#define EMMA_PCI_INT			(EMMA_PCI_BASE + 0x0020)
+#define EMMA_PCI_INT_EN			(EMMA_PCI_BASE + 0x0024)
+#define EMMA_PCI_TWIN_CTR		(EMMA_PCI_BASE + 0x0030)
+#define EMMA_PCI_TWIN_BADR		(EMMA_PCI_BASE + 0x0034)
+#define EMMA_PCI_TWIN0_DADR		(EMMA_PCI_BASE + 0x0038)
+#define EMMA_PCI_TWIN1_DADR		(EMMA_PCI_BASE + 0x003c)
 
-#define EMMA_IDE_BRIDGE		(0x0100+EMMA_PCI_BASE)
-#define EMMA_IDE_CONFIG		(0x0200+EMMA_PCI_BASE)
-#define EMMA_IDE_PRIMARY	(0x0600+EMMA_PCI_BASE)
+#define EMMA_IDE_BRIDGE			(EMMA_PCI_BASE + 0x0100)
+#define EMMA_IDE_CONFIG			(EMMA_PCI_BASE + 0x0200)
+#define EMMA_IDE_PRIMARY		(EMMA_PCI_BASE + 0x0600)
 
-#define EMMA_OHCI_BASE		(0x1000+EMMA_PCI_BASE)
-#define EMMA_EHCI_BASE		(0x0f00+EMMA_PCI_BASE)
+#define EMMA_OHCI_BASE			(EMMA_PCI_BASE + 0x1000)
+#define EMMA_EHCI_BASE			(EMMA_PCI_BASE + 0x0f00)
 
 /*
  *  Memory map (physical address)
-- 
1.6.5.2

