From 04686195473897073ddb279642f984f6ed9a3b44 Mon Sep 17 00:00:00 2001
From: Shinya Kuribayashi <shinya.kuribayashi@necel.com>
Date: Tue, 10 Nov 2009 02:42:15 -0800
Subject: [PATCH 3/4] MIPS: nec_emma3pf: emma3pf.h: Introduce EMMA_FLIP_BASE

Introduce EMMA_FLIP_BASE as a base address of FLIP bus bridge.

* Remove EMMA_IDE_BRIDGE as there is not such register set.
* Redefine EMMA_IDE_foo and EMMA_xHCI_BASE using EMMA_FLIP_BASE, not
  using EMMA_PCI_BASE.  This is because IDE and USB controllers are
  internally connected to the system via FLIP bus bridge.
* EMMA_IDE_CONFIG offset was wrong; now fixed.

Signed-off-by: Shinya Kuribayashi <shinya.kuribayashi@necel.com>
---
 include/asm-mips/emma/emma3pf.h |   10 +++++-----
 1 files changed, 5 insertions(+), 5 deletions(-)

diff --git a/include/asm-mips/emma/emma3pf.h b/include/asm-mips/emma/emma3pf.h
index 13c6f40..db48040 100644
--- a/include/asm-mips/emma/emma3pf.h
+++ b/include/asm-mips/emma/emma3pf.h
@@ -44,6 +44,7 @@
 #define EMMA_PBRD_BASE			0x10100000
 #define EMMA_TIMER_BASE			0x10110000
 #define EMMA_PCI_BASE			0x10200000
+#define EMMA_FLIP_BASE			0x10200100
 #define EMMA_SATA_PORT_BASE		0x10204000
 
 #define EMMA_BHIF_STRAP_0		(EMMA_BHIF_BASE + 0x0010)
@@ -108,12 +109,11 @@
 #define EMMA_PCI_TWIN0_DADR		(EMMA_PCI_BASE + 0x0038)
 #define EMMA_PCI_TWIN1_DADR		(EMMA_PCI_BASE + 0x003c)
 
-#define EMMA_IDE_BRIDGE			(EMMA_PCI_BASE + 0x0100)
-#define EMMA_IDE_CONFIG			(EMMA_PCI_BASE + 0x0200)
-#define EMMA_IDE_PRIMARY		(EMMA_PCI_BASE + 0x0600)
+#define EMMA_IDEP_CONFIG_BASE	(EMMA_FLIP_BASE + 0x0200) /* 0x1020.0300 */
+#define EMMA_IDEP_BASE		(EMMA_FLIP_BASE + 0x0500) /* 0x1020.0600 */
 
-#define EMMA_OHCI_BASE			(EMMA_PCI_BASE + 0x1000)
-#define EMMA_EHCI_BASE			(EMMA_PCI_BASE + 0x0f00)
+#define EMMA_EHCI_BASE		(EMMA_FLIP_BASE + 0x0e00) /* 0x1020.0F00 */
+#define EMMA_OHCI_BASE		(EMMA_FLIP_BASE + 0x0f00) /* 0x1020.1000 */
 
 /*
  *  Memory map (physical address)
-- 
1.6.5.2

