From bf59043924bc903f8e425303c252490903a19f85 Mon Sep 17 00:00:00 2001
From: Yong Zhang <yong.zhang@windriver.com>
Date: Thu, 9 Dec 2010 16:41:42 +0800
Subject: [PATCH] mips: make virt_to_phys() be aware of address in CKSEG0 in 64bit mode

virt_to_phys() assume all virtual address residents in XKPHYS area,
but the CKSEG0 adress is also used for compatibility. Let virt_to_phys()
be aware of it.

Signed-off-by: Yong Zhang <yong.zhang@windriver.com>
---
 include/asm-mips/io.h |    4 ++++
 1 files changed, 4 insertions(+), 0 deletions(-)

diff --git a/include/asm-mips/io.h b/include/asm-mips/io.h
index 501a40b..c300589 100644
--- a/include/asm-mips/io.h
+++ b/include/asm-mips/io.h
@@ -116,6 +116,10 @@ static inline void set_io_port_base(unsigned long base)
  */
 static inline unsigned long virt_to_phys(volatile const void *address)
 {
+#ifdef CONFIG_64BIT
+	if ((unsigned long)address >= CKSEG0)
+		return (unsigned long)address - CKSEG0 + PHYS_OFFSET;
+#endif
 	return (unsigned long)address - PAGE_OFFSET + PHYS_OFFSET;
 }
 
-- 
1.7.0.4

