From 9dbcce90248f5c158a8c90fbe3967cf270b26b94 Mon Sep 17 00:00:00 2001
From: Ralf Baechle <ralf@linux-mips.org>
Date: Mon, 12 Oct 2009 23:20:20 +0200
Subject: [PATCH 097/117] MIPS: 1004K: Enable SPRAM support.

commit 9dbcce90248f5c158a8c90fbe3967cf270b26b94 from linux-mips

Signed-off-by: Ralf Baechle <ralf@linux-mips.org>
---
 arch/mips/kernel/spram.c |    1 +
 1 files changed, 1 insertions(+), 0 deletions(-)

diff --git a/arch/mips/kernel/spram.c b/arch/mips/kernel/spram.c
index 6ddb507..71b4737 100644
--- a/arch/mips/kernel/spram.c
+++ b/arch/mips/kernel/spram.c
@@ -208,6 +208,7 @@ __cpuinit void spram_config(void)
 	case CPU_24K:
 	case CPU_34K:
 	case CPU_74K:
+	case CPU_1004K:
 		config0 = read_c0_config();
 		/* FIXME: addresses are Malta specific */
 		if (config0 & (1<<24)) {
-- 
1.7.0

