From 7dc60198e06835f226829a635f140bdc837b6250 Mon Sep 17 00:00:00 2001
From: Alexey Dobriyan <adobriyan@gmail.com>
Date: Wed, 12 Aug 2009 23:59:27 +0400
Subject: [PATCH 083/117] MIPS: Lasat: Fix compilation

commit 7dc60198e06835f226829a635f140bdc837b6250 from linux-mips

Header needed for current_cpu_data which expands to smp_processor_id().
However, linux/smp.h can't be included into asm/cpu-info.h due to horrible
circular dependencies, so plug it here.

Signed-off-by: Alexey Dobriyan <adobriyan@gmail.com>
Signed-off-by: Ralf Baechle <ralf@linux-mips.org>
(cherry picked from commit b19d0d2b2a3a11d25884d833f63ed040fb3247ae)
---
 include/asm-mips/lasat/lasat.h |    1 +
 1 files changed, 1 insertions(+), 0 deletions(-)

diff --git a/include/asm-mips/lasat/lasat.h b/include/asm-mips/lasat/lasat.h
index caeba1e..a1ada1c 100644
--- a/include/asm-mips/lasat/lasat.h
+++ b/include/asm-mips/lasat/lasat.h
@@ -227,6 +227,7 @@ extern void lasat_write_eeprom_info(void);
  * It is used for the bit-banging rtc and eeprom drivers */
 
 #include <linux/delay.h>
+#include <linux/smp.h>
 
 /* calculating with the slowest board with 100 MHz clock */
 #define LASAT_100_DIVIDER 20
-- 
1.7.0

