From 3970aba898e77e43283103bf89fccb1d22d61168 Mon Sep 17 00:00:00 2001
From: Ralf Baechle <ralf@linux-mips.org>
Date: Fri, 28 Aug 2009 12:29:58 +0100
Subject: [PATCH 084/117] MIPS: Fix potencial build error in <asm/delay.h>

commit 3970aba898e77e43283103bf89fccb1d22d61168 from linux-mips

<asm/delay.h> will break if HZ isn't defined.  In 2.6.26 and later we're
usually lucky ...

Signed-off-by: Ralf Baechle <ralf@linux-mips.org>
(cherry picked from commit 46b788f201b56c08ccfa2a38262d069f5f40cb0e)
---
 include/asm-mips/delay.h |    2 ++
 1 files changed, 2 insertions(+), 0 deletions(-)

diff --git a/include/asm-mips/delay.h b/include/asm-mips/delay.h
index d2d8949..e7cd782 100644
--- a/include/asm-mips/delay.h
+++ b/include/asm-mips/delay.h
@@ -11,6 +11,8 @@
 #ifndef _ASM_DELAY_H
 #define _ASM_DELAY_H
 
+#include <linux/param.h>
+
 extern void __delay(unsigned int loops);
 extern void __ndelay(unsigned int ns);
 extern void __udelay(unsigned int us);
-- 
1.7.0

