From 93a8521ddaf00d9df33277c86679fa3b07cb304c Mon Sep 17 00:00:00 2001
From: SangeethaRao <sangeetha.rao@intel.com>
Date: Fri, 22 May 2015 19:03:33 -0500
Subject: [PATCH 033/213] fs/Kconfig: Updated for 3.19 build

Signed-off-by: SangeethaRao <sangeetha.rao@intel.com>
---
 fs/Kconfig | 1 +
 1 file changed, 1 insertion(+)

diff --git a/fs/Kconfig b/fs/Kconfig
index 81e6291..9794b877 100644
--- a/fs/Kconfig
+++ b/fs/Kconfig
@@ -310,5 +310,6 @@ endif # NETWORK_FILESYSTEMS
 
 source "fs/nls/Kconfig"
 source "fs/dlm/Kconfig"
+source "fs/vmfs/Kconfig"
 
 endmenu
-- 
1.9.1

