From 522506d05044e10f2adcd78202bf07e29e15969d Mon Sep 17 00:00:00 2001
From: SangeethaRao <sangeetha.rao@intel.com>
Date: Fri, 22 May 2015 19:46:17 -0500
Subject: [PATCH 034/213] pci: Updated Kconfig to build correct PCIe driver for
 X9/X7

Signed-off-by: SangeethaRao <sangeetha.rao@intel.com>
---
 drivers/pci/host/Kconfig | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/drivers/pci/host/Kconfig b/drivers/pci/host/Kconfig
index a2c9f84..3f8d12f 100644
--- a/drivers/pci/host/Kconfig
+++ b/drivers/pci/host/Kconfig
@@ -3,7 +3,7 @@ menu "PCI host controller drivers"
 
 config PCI_AXXIA
 	bool "AXXIA PCIe controller"
-	depends on ARCH_AXXIA
+	depends on ARCH_AXXIA && ARM64
 	default ARCH_AXXIA
 	help
 	 Enables support for the Axxia PCIe controller. There are three PEI
-- 
1.9.1

