From 5d279bac07f613e34678a1c1c6acb5d0bae53f2d Mon Sep 17 00:00:00 2001
From: Charlie Paul <cpaul.windriver@gmail.com>
Date: Thu, 2 Apr 2015 11:03:26 -0700
Subject: [PATCH 069/213] arch/arm/mach-axxia: Removed axxia_circular_queue

Signed-off-by: Charlie Paul <cpaul.windriver@gmail.com>
---
 arch/arm/mach-axxia/Makefile | 1 -
 1 file changed, 1 deletion(-)

diff --git a/arch/arm/mach-axxia/Makefile b/arch/arm/mach-axxia/Makefile
index b665db8..73392a4 100644
--- a/arch/arm/mach-axxia/Makefile
+++ b/arch/arm/mach-axxia/Makefile
@@ -9,7 +9,6 @@ obj-y					+= ssp-gpio.o
 obj-y					+= timers.o
 obj-y					+= pci.o
 obj-y					+= ddr_retention.o ddr_shutdown.o
-obj-y					+= axxia_circular_queue.o
 obj-$(CONFIG_SMP)			+= platsmp.o headsmp.o
 obj-$(CONFIG_ARCH_AXXIA_GIC)		+= axxia-gic.o
 obj-$(CONFIG_HOTPLUG_CPU)		+= hotplug.o lsi_power_management.o
-- 
1.9.1

