From ad71d9a1726765fcdd74c2283f10b92aadfa7c49 Mon Sep 17 00:00:00 2001
From: Anders Berg <anders.berg@intel.com>
Date: Fri, 6 Mar 2015 14:46:57 +0100
Subject: [PATCH 070/213] misc: lsi-ncr: Only use AMP lock on PPC platforms.

Signed-off-by: Anders Berg <anders.berg@intel.com>
---
 arch/arm/mach-axxia/axxia.c | 1 -
 1 file changed, 1 deletion(-)

diff --git a/arch/arm/mach-axxia/axxia.c b/arch/arm/mach-axxia/axxia.c
index 07ac3b6..5f3de12 100644
--- a/arch/arm/mach-axxia/axxia.c
+++ b/arch/arm/mach-axxia/axxia.c
@@ -211,7 +211,6 @@ void __init axxia_dt_init(void)
 			     axxia_auxdata_lookup, NULL);
 	pm_power_off = NULL; /* TBD */
 
-	ncr_init();
 	axxia_ddr_retention_init();
 
 	axxia_pcie_init();
-- 
1.9.1

