From 3298dcf8f88cd2f9b196f140a361221f81df8e79 Mon Sep 17 00:00:00 2001
From: John Jacques <john.jacques@intel.com>
Date: Thu, 10 Sep 2015 19:38:00 -0500
Subject: [PATCH 099/213] axxia: Build the Axxia MDIO Driver for PHY Access

Signed-off-by: John Jacques <john.jacques@intel.com>
---
 drivers/net/ethernet/lsi/Makefile | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/drivers/net/ethernet/lsi/Makefile b/drivers/net/ethernet/lsi/Makefile
index c981d39..5ebc246 100644
--- a/drivers/net/ethernet/lsi/Makefile
+++ b/drivers/net/ethernet/lsi/Makefile
@@ -1,6 +1,6 @@
 # Makefile for the LSI FEMAC network interface.
 
-#obj-$(CONFIG_NET_LSI_MDIO) += lsi-mdio.o
-#obj-$(CONFIG_NET_LSI_FEMAC) += lsi-femac.o
+obj-$(CONFIG_NET_LSI_MDIO) += lsi-mdio.o
+obj-$(CONFIG_NET_LSI_FEMAC) += lsi-femac.o
 obj-$(CONFIG_LSI_NET) += lsi_acp_mdio.o
 obj-$(CONFIG_LSI_NET) += lsi_acp_net.o
-- 
1.9.1

