From 9c900f74b45d68b22ab032cbd5c2be2762735c0e Mon Sep 17 00:00:00 2001
From: John Jacques <john.jacques@intel.com>
Date: Fri, 8 Jan 2016 17:51:37 +0200
Subject: [PATCH 121/213] pmu: Fix Compiler Warnings

Signed-off-by: John Jacques <john.jacques@intel.com>
Signed-off-by: Bruce Ashfield <bruce.ashfield@windriver.com>
---
 include/linux/pmu.h | 1 +
 1 file changed, 1 insertion(+)

diff --git a/include/linux/pmu.h b/include/linux/pmu.h
index 99b400b..f11044d 100644
--- a/include/linux/pmu.h
+++ b/include/linux/pmu.h
@@ -8,6 +8,7 @@
 #ifndef _LINUX_PMU_H
 #define _LINUX_PMU_H
 
+#include <linux/adb.h>
 #include <uapi/linux/pmu.h>
 
 
-- 
1.9.1

