From 6e1d12f4dc5afb6c8ea0beb7ec206f77a633b36a Mon Sep 17 00:00:00 2001
From: John Jacques <john.jacques@intel.com>
Date: Thu, 16 Mar 2017 14:16:48 -0500
Subject: [PATCH 04/17] axxia: Remove Debug Prints

Signed-off-by: John Jacques <john.jacques@intel.com>
---
 drivers/misc/axxia-mdio.c |    3 ---
 1 files changed, 0 insertions(+), 3 deletions(-)

diff --git a/drivers/misc/axxia-mdio.c b/drivers/misc/axxia-mdio.c
index acaca53..1607c77 100644
--- a/drivers/misc/axxia-mdio.c
+++ b/drivers/misc/axxia-mdio.c
@@ -92,8 +92,6 @@ axxia_mdio_write(struct mii_bus *bus, int mii_id, int regnum, u16 value)
 	void __iomem *base = bus_to_regs(bus);
 	u32 ctrl;
 
-	printk("%s:%d - base=0x%p\n", __FILE__, __LINE__, base);
-
 	/* Wait for mdio_busy (control) to be clear. */
 	while ((readl(base + MDIO_CONTROL) & CONTROL_BUSY) != 0)
 		cpu_relax();
@@ -153,7 +151,6 @@ axxia_mdio_probe(struct platform_device *pdev)
 	priv->bus->write = axxia_mdio_write,
 	priv->bus->priv  = priv;
 	snprintf(priv->bus->id, MII_BUS_ID_SIZE, pdev->name);
-	printk("%s:%d - pdev->name=%s\n", __FILE__, __LINE__, pdev->name);
 
 	priv->base = devm_ioremap_resource(&pdev->dev, res);
 
-- 
1.7.5.4

