From 1299970a961c248dc70e6ef084df8e8c2cdcd9a0 Mon Sep 17 00:00:00 2001
From: John Jacques <john.jacques@intel.com>
Date: Fri, 17 Mar 2017 13:39:14 -0500
Subject: [PATCH 05/17] drivers/misc: Fix a Wrong Address in the Axxia NCR
 Driver

Signed-off-by: John Jacques <john.jacques@intel.com>
---
 drivers/misc/lsi-ncr.c |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/drivers/misc/lsi-ncr.c b/drivers/misc/lsi-ncr.c
index a7ba966..9885e0a 100644
--- a/drivers/misc/lsi-ncr.c
+++ b/drivers/misc/lsi-ncr.c
@@ -914,7 +914,7 @@ ncr_init(void)
 	} else if (of_find_compatible_node(NULL, NULL, "lsi,axc6732")) {
 		pr_debug("Using AXC6700 Addresses\n");
 		nca = ioremap(0x8020000000ULL, 0x20000);
-		apb2ser0 = ioremap(0x8080000000ULL, 0x400000);
+		apb2ser0 = ioremap(0x8002000000ULL, 0x400000);
 		is_6700 = 1;
 		nca_big_endian = 0; /* The 6700 NCA is LE */
 	} else {
-- 
1.7.5.4

