From 2a3006d5da0d144f9a8f039a445c62eac1e9b5a3 Mon Sep 17 00:00:00 2001
From: Daniel Vetter <daniel.vetter@ffwll.ch>
Date: Sun, 8 May 2016 18:20:53 +0200
Subject: [PATCH 3109/4706] drm/i915: Update DRIVER_DATE to 20160508

commit 2a55135c67e8eedea232f2c595698235a9ff92d4 upstream

Signed-off-by: Daniel Vetter <daniel.vetter@ffwll.ch>
---
 drivers/gpu/drm/i915/i915_drv.h |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/drivers/gpu/drm/i915/i915_drv.h b/drivers/gpu/drm/i915/i915_drv.h
index 1ff4835..b1c16da 100644
--- a/drivers/gpu/drm/i915/i915_drv.h
+++ b/drivers/gpu/drm/i915/i915_drv.h
@@ -66,7 +66,7 @@
 
 #define DRIVER_NAME		"i915"
 #define DRIVER_DESC		"Intel Graphics"
-#define DRIVER_DATE		"20160425"
+#define DRIVER_DATE		"20160508"
 
 #undef WARN_ON
 /* Many gcc seem to no see through this and fall over :( */
-- 
1.7.5.4

