From 0769f2209c3e39ecd9a357932982c8fe32cc5d7e Mon Sep 17 00:00:00 2001
From: Daniel Vetter <daniel.vetter@ffwll.ch>
Date: Mon, 11 Apr 2016 20:20:18 +0200
Subject: [PATCH 2752/4706] drm/i915: Update DRIVER_DATE to 20160411

commit ba3150ac3876acd082307f142597d3482107facc upstream

Signed-off-by: Daniel Vetter <daniel.vetter@ffwll.ch>
---
 drivers/gpu/drm/i915/i915_drv.h |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/drivers/gpu/drm/i915/i915_drv.h b/drivers/gpu/drm/i915/i915_drv.h
index 45c5f8a..3fc8726 100644
--- a/drivers/gpu/drm/i915/i915_drv.h
+++ b/drivers/gpu/drm/i915/i915_drv.h
@@ -60,7 +60,7 @@
 
 #define DRIVER_NAME		"i915"
 #define DRIVER_DESC		"Intel Graphics"
-#define DRIVER_DATE		"20160330"
+#define DRIVER_DATE		"20160411"
 
 #undef WARN_ON
 /* Many gcc seem to no see through this and fall over :( */
-- 
1.7.5.4

