From feb4597e9a847f5dac76e7cefd692873d0d5ecb8 Mon Sep 17 00:00:00 2001
From: "Subhransu S. Prusty" <subhransu.s.prusty@intel.com>
Date: Tue, 29 Sep 2015 13:56:10 +0530
Subject: [PATCH 4679/4706] hda: Add hduadio support to DEVTABLE

commit c21689b99ea192c4debd06935f68fa539062fc22 from
git://git.yoctoproject.org/linux-yocto-4.1

It adds two entries one for legacy hda devices and other one for
extended hda devices.

Signed-off-by: Subhransu S. Prusty <subhransu.s.prusty@intel.com>

Conflicts:
	scripts/mod/devicetable-offsets.c
	scripts/mod/file2alias.c

Change-Id: I3efc65e9e4d064c319b751f7c92663e8dd3b27fb
---
 scripts/mod/file2alias.c |   16 ++++++++++++++++
 1 files changed, 16 insertions(+), 0 deletions(-)

diff --git a/scripts/mod/file2alias.c b/scripts/mod/file2alias.c
index 5b96206..0ae7785 100644
--- a/scripts/mod/file2alias.c
+++ b/scripts/mod/file2alias.c
@@ -1280,6 +1280,22 @@ static bool sym_is(const char *name, unsigned namelen, const char *symbol)
 	return memcmp(name, symbol, namelen) == 0;
 }
 
+static int do_ehda_entry(const char *filename,
+			  void *symval, char *alias)
+{
+	DEF_FIELD(symval, hda_device_id, vendor_id);
+	DEF_FIELD(symval, hda_device_id, rev_id);
+
+	strcpy(alias, "ehdaudio:");
+	ADD(alias, "v", vendor_id != HDA_ID_ANY, vendor_id);
+	ADD(alias, "r", rev_id != HDA_ID_ANY, rev_id);
+
+	add_wildcard(alias);
+	return 1;
+
+}
+ADD_TO_DEVTABLE("ehdaudio", hda_device_id, do_hda_entry);
+
 static void do_table(void *symval, unsigned long size,
 		     unsigned long id_size,
 		     const char *device_id,
-- 
1.7.5.4

