From c1cc5f4ecd13607743e4c1cc8510ddf318b6e3fa Mon Sep 17 00:00:00 2001
From: Tomas Winkler <tomas.winkler@intel.com>
Date: Wed, 14 Sep 2016 11:24:37 +0300
Subject: [PATCH 27/68] Revert "tpm_crb: use wait_for_tpm_stat() in runtime
 PM"

commit b094a95c8cd99d965fa63908ffcd3bf777228a1e from
git://git.yoctoproject.org/linux-yocto-4.1

This reverts commit 664fe6950d981df232a32c3b091be49db8e54390.
---
 drivers/char/tpm/tpm_crb.c |   18 ++++--------------
 1 files changed, 4 insertions(+), 14 deletions(-)

diff --git a/drivers/char/tpm/tpm_crb.c b/drivers/char/tpm/tpm_crb.c
index d9447a9..105fad0 100644
--- a/drivers/char/tpm/tpm_crb.c
+++ b/drivers/char/tpm/tpm_crb.c
@@ -68,8 +68,6 @@ struct crb_control_area {
 
 enum crb_status {
 	CRB_DRV_STS_COMPLETE	= BIT(0),
-	CRB_DRV_STS_READY	= BIT(1),
-	CRB_DRV_STS_IDLE	= BIT(2),
 };
 
 enum crb_flags {
@@ -95,9 +93,9 @@ static int __maybe_unused crb_pause(struct tpm_chip *chip)
 
 	req = ioread32(&priv->cca->req);
 	iowrite32(cpu_to_le32(req | CRB_CTRL_REQ_GO_IDLE), &priv->cca->req);
+	msleep(chip->timeout_c);
 
-	if (wait_for_tpm_stat(chip, CRB_DRV_STS_IDLE, TPM2_TIMEOUT_C, NULL,
-			      false)) {
+	if (ioread32(&priv->cca->req) & CRB_CTRL_REQ_GO_IDLE) {
 		dev_warn(&chip->dev, "goIdle timed out\n");
 		return -ETIME;
 	}
@@ -115,9 +113,9 @@ static int __maybe_unused crb_resume(struct tpm_chip *chip)
 
 	req = ioread32(&priv->cca->req);
 	iowrite32(cpu_to_le32(req | CRB_CTRL_REQ_CMD_READY), &priv->cca->req);
+	msleep(chip->timeout_c);
 
-	if (wait_for_tpm_stat(chip, CRB_DRV_STS_READY, TPM2_TIMEOUT_C, NULL,
-			      false)) {
+	if (ioread32(&priv->cca->req) & CRB_CTRL_REQ_CMD_READY) {
 		dev_warn(&chip->dev, "cmdReady timed out\n");
 		return -ETIME;
 	}
@@ -136,14 +134,6 @@ static u8 crb_status(struct tpm_chip *chip)
 	    CRB_START_INVOKE)
 		sts |= CRB_DRV_STS_COMPLETE;
 
-	if ((ioread32(&priv->cca->req) & CRB_CTRL_REQ_CMD_READY) !=
-	    CRB_CTRL_REQ_CMD_READY)
-		sts |= CRB_DRV_STS_READY;
-
-	if ((ioread32(&priv->cca->req) & CRB_CTRL_REQ_GO_IDLE) !=
-	    CRB_CTRL_REQ_GO_IDLE)
-		sts |= CRB_DRV_STS_IDLE;
-
 	return sts;
 }
 
-- 
1.7.5.4

