From 12c48c5a0aa87efe36c0830da8e6f966e51da106 Mon Sep 17 00:00:00 2001
From: Guoqing Jiang <Guoqing.Jiang@windriver.com>
Date: Fri, 1 Apr 2011 17:50:23 +0800
Subject: [PATCH 14/21] TI 816x: add MAX_GPTIMER_ID definition

Add MAX_GPTIMER_ID definition since different number of GPTIMERs
on ti 816x.

This patch refers to below commit in arago git tree.
git://arago-project.org/git/projects/linux-omap3.git

dc9fa4 ti816x: cleanups

Integrated-by: Guoqing Jiang <Guoqing.Jiang@windriver.com>
---
 arch/arm/mach-omap2/timer-gp.c |    4 ++++
 1 files changed, 4 insertions(+), 0 deletions(-)

diff --git a/arch/arm/mach-omap2/timer-gp.c b/arch/arm/mach-omap2/timer-gp.c
index 74fbed8..cbb93e1 100644
--- a/arch/arm/mach-omap2/timer-gp.c
+++ b/arch/arm/mach-omap2/timer-gp.c
@@ -41,7 +41,11 @@
 #include <asm/localtimer.h>
 
 /* MAX_GPTIMER_ID: number of GPTIMERs on the chip */
+#ifndef CONFIG_ARCH_TI816X
 #define MAX_GPTIMER_ID		12
+#else
+#define MAX_GPTIMER_ID		8
+#endif
 
 static struct omap_dm_timer *gptimer;
 static struct clock_event_device clockevent_gpt;
-- 
1.7.0.2

