From 0ca4b96b60620a6ad80ffbcd06702485199cd437 Mon Sep 17 00:00:00 2001
From: Guoqing Jiang <Guoqing.Jiang@windriver.com>
Date: Fri, 1 Apr 2011 17:48:21 +0800
Subject: [PATCH 13/21] TI 816x: add ti 816x serial port support

Add ti 816x serial port support in common file, and new macros
are created for 816x.

This patch refers to below commits in arago git tree.
git://arago-project.org/git/projects/linux-omap3.git

b1da24 ti816x: Skip LSR check when accessing RX register as DLL
e1ea60 ti816x: Initial support for TI8168
dc9fa4 ti816x: cleanups

Integrated-by: Guoqing Jiang <Guoqing.Jiang@windriver.com>
---
 arch/arm/mach-omap2/serial.c             |   11 +++++++----
 arch/arm/plat-omap/include/plat/serial.h |    8 ++++++++
 2 files changed, 15 insertions(+), 4 deletions(-)

diff --git a/arch/arm/mach-omap2/serial.c b/arch/arm/mach-omap2/serial.c
index 3f98a92..3144ac2 100644
--- a/arch/arm/mach-omap2/serial.c
+++ b/arch/arm/mach-omap2/serial.c
@@ -612,9 +612,12 @@ static unsigned int serial_in_override(struct uart_port *up, int offset)
 {
 	if (UART_RX == offset) {
 		unsigned int lsr;
-		lsr = __serial_read_reg(up, UART_LSR);
-		if (!(lsr & UART_LSR_DR))
-			return -EPERM;
+		if (!cpu_is_ti816x() ||
+		    !(__serial_read_reg(up, UART_LCR) & UART_LCR_DLAB)) {
+			lsr = __serial_read_reg(up, UART_LSR);
+			if (!(lsr & UART_LSR_DR))
+				return -EPERM;
+		}
 	}
 
 	return __serial_read_reg(up, offset);
@@ -753,7 +756,7 @@ void __init omap_serial_init_port(int port)
 	 * omap3xxx: Never read empty UART fifo on UARTs
 	 * with IP rev >=0x52
 	 */
-	if (cpu_is_omap44xx()) {
+	if (cpu_is_omap44xx() || cpu_is_ti816x()) {
 		uart->p->serial_in = serial_in_override;
 		uart->p->serial_out = serial_out_override;
 	} else if ((serial_read_reg(uart->p, UART_OMAP_MVER) & 0xFF)
diff --git a/arch/arm/plat-omap/include/plat/serial.h b/arch/arm/plat-omap/include/plat/serial.h
index 83dce4c..685273c 100644
--- a/arch/arm/plat-omap/include/plat/serial.h
+++ b/arch/arm/plat-omap/include/plat/serial.h
@@ -37,6 +37,11 @@
 #define OMAP4_UART3_BASE	0x48020000
 #define OMAP4_UART4_BASE	0x4806e000
 
+/* TI816X serial ports */
+#define TI816X_UART1_BASE	0x48020000
+#define TI816X_UART2_BASE	0x48022000
+#define TI816X_UART3_BASE	0x48024000
+
 /* External port on Zoom2/3 */
 #define ZOOM_UART_BASE		0x10000000
 #define ZOOM_UART_VIRT		0xfb000000
@@ -67,6 +72,9 @@
 #define OMAP4UART2		OMAP2UART2
 #define OMAP4UART3		43
 #define OMAP4UART4		44
+#define TI816XUART1		81
+#define TI816XUART2		82
+#define TI816XUART3		83
 #define ZOOM_UART		95		/* Only on zoom2/3 */
 
 /* This is only used by 8250.c for omap1510 */
-- 
1.7.0.2

