From 9cc97e26dade0bc072b42060bb7221c3e20639c9 Mon Sep 17 00:00:00 2001
From: Guoqing Jiang <Guoqing.Jiang@windriver.com>
Date: Fri, 1 Apr 2011 17:39:50 +0800
Subject: [PATCH 06/21] TI 816x: add ti 816x debug support in debug-macro.S

Add ti 816x debug support in debug-macro.S.

This patch refers to below commit in arago git tree.
git://arago-project.org/git/projects/linux-omap3.git

e1ea60 ti816x: Initial support for TI8168

Integrated-by: Guoqing Jiang <Guoqing.Jiang@windriver.com>
---
 arch/arm/mach-omap2/include/mach/debug-macro.S |   16 ++++++++++++++++
 1 files changed, 16 insertions(+), 0 deletions(-)

diff --git a/arch/arm/mach-omap2/include/mach/debug-macro.S b/arch/arm/mach-omap2/include/mach/debug-macro.S
index 4a63a2e..1c4a09b 100644
--- a/arch/arm/mach-omap2/include/mach/debug-macro.S
+++ b/arch/arm/mach-omap2/include/mach/debug-macro.S
@@ -45,7 +45,11 @@ omap_uart_lsr:	.word	0
 		tst	\rx, #1			@ MMU enabled?
 		moveq	\rx, #0x48000000	@ physical base address
 		movne	\rx, #0xfa000000	@ virtual base
+#ifdef CONFIG_ARCH_TI816X
+		orr	\rx, \rx, #0x00020000	@ uart1 on ti816x
+#else
 		orr	\rx, \rx, #0x0006a000	@ uart1 on omap2/3/4
+#endif
 		ldrb	\rx, [\rx, #(UART_SCR << OMAP_PORT_SHIFT)] @ scratchpad
 
 		/* Select the UART to use based on the UART1 scratchpad value */
@@ -65,6 +69,12 @@ omap_uart_lsr:	.word	0
 		beq	34f			@ configure OMAP3UART4
 		cmp	\rx, #OMAP4UART4	@ only on 44xx
 		beq	44f			@ configure OMAP4UART4
+		cmp	\rx, #TI816XUART1	@ ti816x UART offsets different
+		beq	81f			@ configure UART1
+		cmp	\rx, #TI816XUART2	@ ti816x UART offsets different
+		beq	82f			@ configure UART2
+		cmp	\rx, #TI816XUART3	@ ti816x UART offsets different
+		beq	83f			@ configure UART3
 		cmp	\rx, #ZOOM_UART		@ only on zoom2/3
 		beq	95f			@ configure ZOOM_UART
 
@@ -87,6 +97,12 @@ omap_uart_lsr:	.word	0
 		b	98f
 44:		mov	\rx, #UART_OFFSET(OMAP4_UART4_BASE)
 		b	98f
+81:		mov	\rx, #UART_OFFSET(TI816X_UART1_BASE)
+		b	98f
+82:		mov	\rx, #UART_OFFSET(TI816X_UART2_BASE)
+		b	98f
+83:		mov	\rx, #UART_OFFSET(TI816X_UART3_BASE)
+		b	98f
 95:		mov	\rx, #ZOOM_UART_BASE
 		ldr	\tmp, =omap_uart_phys
 		str	\rx, [\tmp, #0]
-- 
1.7.0.2

