From 089a530a48ab17df7403f9180cc2ef8f20257f7a Mon Sep 17 00:00:00 2001
From: Guoqing Jiang <Guoqing.Jiang@windriver.com>
Date: Fri, 1 Apr 2011 17:54:02 +0800
Subject: [PATCH 18/21] TI 816x: add sram initialize handle code

Add sram initialize handle code for ti 816x since sram part
is not currently supported.

This patch refers to below commit in arago git tree.
git://arago-project.org/git/projects/linux-omap3.git

a9090c ti816x: Incorporate changes as per review comments

Integrated-by: Guoqing Jiang <Guoqing.Jiang@windriver.com>
---
 arch/arm/plat-omap/sram.c |    4 ++++
 1 files changed, 4 insertions(+), 0 deletions(-)

diff --git a/arch/arm/plat-omap/sram.c b/arch/arm/plat-omap/sram.c
index 51f4dfb..2ddbf38 100644
--- a/arch/arm/plat-omap/sram.c
+++ b/arch/arm/plat-omap/sram.c
@@ -439,6 +439,10 @@ static inline int omap34xx_sram_init(void)
 
 int __init omap_sram_init(void)
 {
+	/* TODO */
+	if (cpu_is_ti816x())
+		return 0;
+
 	omap_detect_sram();
 	omap_map_sram();
 
-- 
1.7.0.2

