From b780db9fc20a3da2534224df43f51f9a2eb87e2f Mon Sep 17 00:00:00 2001
From: Guoqing Jiang <Guoqing.Jiang@windriver.com>
Date: Fri, 1 Apr 2011 17:51:21 +0800
Subject: [PATCH 15/21] TI 816x: add platform compile option for 816x support

Add platform compile option for 816x support due to different dma
implementation.

This patch refers to below commit in arago git tree.
git://arago-project.org/git/projects/linux-omap3.git

f62937d6ec8616ea0861d0a21f909c67675322fe
ti816x: sdma2edma wrapper support - sdma binary (plat-omap/dma.o) should not be included for ti816x - sdma2edma wrapper should include edma header

Integrated-by: Guoqing Jiang <Guoqing.Jiang@windriver.com>
---
 arch/arm/plat-omap/Makefile |    6 ++++++
 1 files changed, 6 insertions(+), 0 deletions(-)

diff --git a/arch/arm/plat-omap/Makefile b/arch/arm/plat-omap/Makefile
index 98f0191..f85e196 100644
--- a/arch/arm/plat-omap/Makefile
+++ b/arch/arm/plat-omap/Makefile
@@ -3,8 +3,13 @@
 #
 
 # Common support
+ifeq ($(CONFIG_ARCH_TI816X),y)
+obj-y := common.o sram.o clock.o devices.o mux.o gpio.o \
+	 usb.o fb.o io.o edma.o
+else
 obj-y := common.o sram.o clock.o devices.o dma.o mux.o gpio.o \
 	 usb.o fb.o io.o
+endif
 obj-m :=
 obj-n :=
 obj-  :=
@@ -15,6 +20,7 @@ obj-$(CONFIG_ARCH_OMAP16XX) += ocpi.o
 # omap_device support (OMAP2+ only at the moment)
 obj-$(CONFIG_ARCH_OMAP2) += omap_device.o
 obj-$(CONFIG_ARCH_OMAP3) += omap_device.o
+obj-$(CONFIG_ARCH_TI816X) += sdma2edma.o
 
 obj-$(CONFIG_OMAP_MCBSP) += mcbsp.o
 obj-$(CONFIG_OMAP_IOMMU) += iommu.o iovmm.o
-- 
1.7.0.2

