From 68b429c5defa3fae3fcf2d0bad5aaa741b9d988d Mon Sep 17 00:00:00 2001
From: Guoqing Jiang <Guoqing.Jiang@windriver.com>
Date: Fri, 1 Apr 2011 17:38:29 +0800
Subject: [PATCH 04/21] TI 816X: Add support for TI816X GPMC

Add support for TI816X GPMC.

This patch refers to below commit in arago git tree.
git://arago-project.org/git/projects/linux-omap3.git

3ddf7e ti816x: NAND, JFFS2 validated on EVM

Integrated-by: Guoqing Jiang <Guoqing.Jiang@windriver.com>
---
 arch/arm/mach-omap2/gpmc.c |    3 +++
 1 files changed, 3 insertions(+), 0 deletions(-)

diff --git a/arch/arm/mach-omap2/gpmc.c b/arch/arm/mach-omap2/gpmc.c
index 5bc3ca0..79ce799 100644
--- a/arch/arm/mach-omap2/gpmc.c
+++ b/arch/arm/mach-omap2/gpmc.c
@@ -519,6 +519,9 @@ void __init gpmc_init(void)
 	} else if (cpu_is_omap44xx()) {
 		ck = "gpmc_ck";
 		l = OMAP44XX_GPMC_BASE;
+	} else if (cpu_is_ti816x()) {
+		ck = "gpmc_fck";
+		l = OMAP44XX_GPMC_BASE;
 	}
 
 	if (WARN_ON(!ck))
-- 
1.7.0.2

