From 24d11fec827acf1ca36f596277ac9648fff603ba Mon Sep 17 00:00:00 2001
From: Guoqing Jiang <Guoqing.Jiang@windriver.com>
Date: Fri, 1 Apr 2011 17:59:24 +0800
Subject: [PATCH 19/21] TI 816x: add 816x support in plat-omap header files

Add 816x support in plat-omap header files such as
macro definitions and ti 816x specific handle codes etc.

This patch refers to below commits in arago git tree.
git://arago-project.org/git/projects/linux-omap3.git

5092ce ahci: Move the ahci related platform file to the Linux Include directory.
69be15 TI816X: Audio support
78ac6e ti816x:Audio related fixes
af0617 ti816x: Prepare for multi-omap configuration
1c0570 ti816x: Add platform data for PCIe RC
e1ea60 ti816x: Initial support for TI8168
656a57 ti816x: Provide default definition for cpu_is_ti816x()
ff2ce5 ti816x: mcspi PIO mode support validated on ti8168 evm

Integrated-by: Guoqing Jiang <Guoqing.Jiang@windriver.com>
---
 arch/arm/plat-omap/include/plat/ahci_platform.h |   29 +++++++++
 arch/arm/plat-omap/include/plat/asp.h           |   75 +++++++++++++++++++++++
 arch/arm/plat-omap/include/plat/control.h       |   27 ++++++++-
 arch/arm/plat-omap/include/plat/cpu.h           |   15 ++++-
 arch/arm/plat-omap/include/plat/dma.h           |   11 +++
 arch/arm/plat-omap/include/plat/hardware.h      |    1 +
 arch/arm/plat-omap/include/plat/ti816x.h        |   46 ++++++++++++++
 arch/arm/plat-omap/include/plat/ti81xx_ram.h    |   37 +++++++++++
 arch/arm/plat-omap/include/plat/uncompress.h    |    7 ++
 9 files changed, 246 insertions(+), 2 deletions(-)
 create mode 100644 arch/arm/plat-omap/include/plat/ahci_platform.h
 create mode 100644 arch/arm/plat-omap/include/plat/asp.h
 create mode 100644 arch/arm/plat-omap/include/plat/ti816x.h
 create mode 100755 arch/arm/plat-omap/include/plat/ti81xx_ram.h

diff --git a/arch/arm/plat-omap/include/plat/ahci_platform.h b/arch/arm/plat-omap/include/plat/ahci_platform.h
new file mode 100644
index 0000000..f7dd576
--- /dev/null
+++ b/arch/arm/plat-omap/include/plat/ahci_platform.h
@@ -0,0 +1,29 @@
+/*
+ * AHCI SATA platform driver
+ *
+ * Copyright 2004-2005  Red Hat, Inc.
+ *   Jeff Garzik <jgarzik@pobox.com>
+ * Copyright 2010  MontaVista Software, LLC.
+ *   Anton Vorontsov <avorontsov@ru.mvista.com>
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2, or (at your option)
+ * any later version.
+ */
+
+#ifndef _AHCI_PLATFORM_H
+#define _AHCI_PLATFORM_H
+
+struct device;
+struct ata_port_info;
+
+struct ahci_platform_data {
+	int (*init)(struct device *dev);
+	void (*exit)(struct device *dev);
+	const struct ata_port_info *ata_port_info;
+	unsigned int force_port_map;
+	unsigned int mask_port_map;
+};
+
+#endif /* _AHCI_PLATFORM_H */
diff --git a/arch/arm/plat-omap/include/plat/asp.h b/arch/arm/plat-omap/include/plat/asp.h
new file mode 100644
index 0000000..40435d7
--- /dev/null
+++ b/arch/arm/plat-omap/include/plat/asp.h
@@ -0,0 +1,75 @@
+/*
+ * <plat/asp.h> - TI816X Audio Serial Port support
+ * Based on <arch/arm/mach-davinci/include/mach/asp.h>
+ */
+#ifndef __ASM_ARCH_TI816X_ASP_H
+#define __ASM_ARCH_TI816X_ASP_H
+
+#include <mach/irqs.h>
+#include <mach/edma.h>
+
+/* Bases of ti816x register banks */
+#define TI816X_ASP0_BASE	0x48038000
+#define TI816X_ASP1_BASE	0x4803C000
+#define TI816X_ASP2_BASE	0x48050000
+
+/* EDMA channels of ti816x McASP0 */
+#define	TI816X_DMA_MCASP0_AXEVT	8
+#define	TI816X_DMA_MCASP0_AREVT	9
+
+/* EDMA channels of ti816x McASP1 */
+#define	TI816X_DMA_MCASP1_AXEVT	10
+#define	TI816X_DMA_MCASP1_AREVT	11
+
+/* EDMA channels of ti816x McASP2 */
+#define	TI816X_DMA_MCASP2_AXEVT	12
+#define	TI816X_DMA_MCASP2_AREVT	13
+
+/* Interrupts */
+#define TI816X_ASP0_RX_INT	TI816X_IRQ_MCASP0_RX
+#define TI816X_ASP0_TX_INT	TI816X_IRQ_MCASP0_TX
+#define TI816X_ASP1_RX_INT	TI816X_IRQ_MCASP1_RX
+#define TI816X_ASP1_TX_INT	TI816X_IRQ_MCASP1_TX
+#define TI816X_ASP2_RX_INT	TI816X_IRQ_MCASP2_RX
+#define TI816X_ASP2_TX_INT	TI816X_IRQ_MCASP2_TX
+
+struct snd_platform_data {
+	u32 tx_dma_offset;
+	u32 rx_dma_offset;
+	enum dma_event_q eventq_no;	/* event queue number */
+	unsigned int codec_fmt;
+	/*
+	 * Allowing this is more efficient and eliminates left and right swaps
+	 * caused by underruns, but will swap the left and right channels
+	 * when compared to previous behavior.
+	 */
+	unsigned enable_channel_combine:1;
+	unsigned sram_size_playback;
+	unsigned sram_size_capture;
+
+	/* McASP specific fields */
+	int tdm_slots;
+	u8 op_mode;
+	u8 num_serializer;
+	u8 *serial_dir;
+	u8 version;
+	u8 txnumevt;
+	u8 rxnumevt;
+};
+
+enum {
+	MCASP_VERSION_1 = 0,	/* DM646x */
+	MCASP_VERSION_2,	/* DA8xx/OMAPL1x */
+	MCASP_VERSION_3,	/* TODO: VB__Verify the version for TI816x */
+};
+
+#define INACTIVE_MODE	0
+#define TX_MODE		1
+#define RX_MODE		2
+
+#define DAVINCI_MCASP_IIS_MODE	0 /* Driver code needs to change modified */
+#define DAVINCI_MCASP_DIT_MODE	1
+
+void ti816x_register_mcasp(int id, struct snd_platform_data *pdata);
+
+#endif /* __ASM_ARCH_TI816X_ASP_H */
diff --git a/arch/arm/plat-omap/include/plat/control.h b/arch/arm/plat-omap/include/plat/control.h
index 330cae9..f3432e2 100644
--- a/arch/arm/plat-omap/include/plat/control.h
+++ b/arch/arm/plat-omap/include/plat/control.h
@@ -233,6 +233,24 @@
  * that should be added.
  */
 
+/* TI816X spefic control submodules */
+#define TI816X_CONTROL_OCPCONF		0x000
+#define TI816X_CONTROL_DEVBOOT		0x040
+#define TI816X_CONTROL_DEVCONF		0x600
+
+/* TI816X CONTROL_DEVBOOT register offsets */
+#define TI816X_CONTROL_STATUS		(TI816X_CONTROL_DEVBOOT + 0x000)
+#define TI816X_CONTROL_BOOTSTAT		(TI816X_CONTROL_DEVBOOT + 0x004)
+
+/* TI816X CONTROL_DEVCONF register offsets */
+#define TI816X_CONTROL_DEVICE_ID	(TI816X_CONTROL_DEVCONF + 0x000)
+#define TI816X_CONTROL_MAC_ID0_LO	(TI816X_CONTROL_DEVCONF + 0x030)
+#define TI816X_CONTROL_MAC_ID0_HI	(TI816X_CONTROL_DEVCONF + 0x034)
+#define TI816X_CONTROL_MAC_ID1_LO	(TI816X_CONTROL_DEVCONF + 0x038)
+#define TI816X_CONTROL_MAC_ID1_HI	(TI816X_CONTROL_DEVCONF + 0x03c)
+#define TI816X_CONTROL_PCIE_CFG		(TI816X_CONTROL_DEVCONF + 0x040)
+/* TODO: Add other ti816x registers... */
+
 /*
  * Control module register bit defines - these should eventually go into
  * their own regbits file.  Some of these will be complicated, depending
@@ -315,6 +333,13 @@
 #define AM35XX_HECC_SW_RST		BIT(3)
 #define AM35XX_VPFE_PCLK_SW_RST		BIT(4)
 
+/* TI816X CONTROL_PCIE_CFG bits */
+#define TI816X_PCIE_DEVTYPE_SHIFT	0
+#define TI816X_PCIE_DEVTYPE_MASK	(0x3 << TI816X_PCIE_DEVTYPE_SHIFT)
+#define TI816X_PCIE_DEVTYPE_RC		(0x2 << TI816X_PCIE_DEVTYPE_SHIFT)
+#define TI816X_PCIE_CFGPLL_SHIFT	16
+#define TI816X_PCIE_PLLMUX_25X		(0x01C9 << TI816X_PCIE_CFGPLL_SHIFT)
+
 /*
  * CONTROL OMAP STATUS register to identify OMAP3 features
  */
@@ -350,7 +375,7 @@
 
 
 #ifndef __ASSEMBLY__
-#ifdef CONFIG_ARCH_OMAP2PLUS
+#if defined(CONFIG_ARCH_OMAP2PLUS) || defined(CONFIG_ARCH_TI816X)
 extern void __iomem *omap_ctrl_base_get(void);
 extern u8 omap_ctrl_readb(u16 offset);
 extern u16 omap_ctrl_readw(u16 offset);
diff --git a/arch/arm/plat-omap/include/plat/cpu.h b/arch/arm/plat-omap/include/plat/cpu.h
index 2b502bf..fef4b13 100644
--- a/arch/arm/plat-omap/include/plat/cpu.h
+++ b/arch/arm/plat-omap/include/plat/cpu.h
@@ -89,6 +89,7 @@ unsigned int omap_rev(void);
  * cpu_is_omap243x():	True for OMAP2430
  * cpu_is_omap343x():	True for OMAP3430
  * cpu_is_omap443x():	True for OMAP4430
+ * cpu_is_ti816x():	True for TI8168
  */
 #define GET_OMAP_CLASS	(omap_rev() & 0xff)
 
@@ -112,12 +113,14 @@ IS_OMAP_CLASS(16xx, 0x16)
 IS_OMAP_CLASS(24xx, 0x24)
 IS_OMAP_CLASS(34xx, 0x34)
 IS_OMAP_CLASS(44xx, 0x44)
+IS_OMAP_CLASS(81xx, 0x81)
 
 IS_OMAP_SUBCLASS(242x, 0x242)
 IS_OMAP_SUBCLASS(243x, 0x243)
 IS_OMAP_SUBCLASS(343x, 0x343)
 IS_OMAP_SUBCLASS(363x, 0x363)
 IS_OMAP_SUBCLASS(443x, 0x443)
+IS_OMAP_SUBCLASS(816x, 0x816)
 
 #define cpu_is_omap7xx()		0
 #define cpu_is_omap15xx()		0
@@ -129,6 +132,7 @@ IS_OMAP_SUBCLASS(443x, 0x443)
 #define cpu_is_omap343x()		0
 #define cpu_is_omap44xx()		0
 #define cpu_is_omap443x()		0
+#define cpu_is_ti816x()			0
 
 #if defined(MULTI_OMAP1)
 # if defined(CONFIG_ARCH_OMAP730)
@@ -355,11 +359,16 @@ IS_OMAP_TYPE(3517, 0x3517)
 # define cpu_is_omap443x()		is_omap443x()
 # endif
 
+# if defined(CONFIG_ARCH_TI816X)
+# undef cpu_is_ti816x
+# define cpu_is_ti816x()		is_omap816x()
+# endif
+
 /* Macros to detect if we have OMAP1 or OMAP2 */
 #define cpu_class_is_omap1()	(cpu_is_omap7xx() || cpu_is_omap15xx() || \
 				cpu_is_omap16xx())
 #define cpu_class_is_omap2()	(cpu_is_omap24xx() || cpu_is_omap34xx() || \
-				cpu_is_omap44xx())
+				cpu_is_omap44xx() || cpu_is_ti816x())
 
 /* Various silicon revisions for omap2 */
 #define OMAP242X_CLASS		0x24200024
@@ -392,6 +401,9 @@ IS_OMAP_TYPE(3517, 0x3517)
 #define OMAP443X_CLASS		0x44300044
 #define OMAP4430_REV_ES1_0	0x44300044
 
+#define TI816X_CLASS		0x81600081
+#define TI8168_REV_ES1_0	0x81600081
+
 /*
  * Silicon revisions
  */
@@ -472,6 +484,7 @@ OMAP_REV_FUNCTIONS(3_1)
 #define CHIP_IS_OMAP4430ES1		(1 << 8)
 #define CHIP_IS_OMAP3630ES1_1           (1 << 9)
 #define CHIP_IS_OMAP3630ES1_2           (1 << 10)
+#define CHIP_IS_TI816X			(1 << 11)
 
 #define CHIP_IS_OMAP24XX		(CHIP_IS_OMAP2420 | CHIP_IS_OMAP2430)
 
diff --git a/arch/arm/plat-omap/include/plat/dma.h b/arch/arm/plat-omap/include/plat/dma.h
index 105bdb5..80e4e82 100644
--- a/arch/arm/plat-omap/include/plat/dma.h
+++ b/arch/arm/plat-omap/include/plat/dma.h
@@ -271,6 +271,16 @@
 #define OMAP24XX_DMA_MCBSP1_RX		32	/* S_DMA_31 */
 #define OMAP24XX_DMA_MCBSP2_TX		33	/* S_DMA_32 */
 #define OMAP24XX_DMA_MCBSP2_RX		34	/* S_DMA_33 */
+#ifdef CONFIG_ARCH_TI816X
+#define OMAP24XX_DMA_SPI1_TX0		16	/* E_DMA_16 */
+#define OMAP24XX_DMA_SPI1_RX0		17	/* E_DMA_17 */
+#define OMAP24XX_DMA_SPI1_TX1		18	/* E_DMA_18 */
+#define OMAP24XX_DMA_SPI1_RX1		19	/* E_DMA_19 */
+#define OMAP24XX_DMA_SPI1_TX2		20	/* E_DMA_20 */
+#define OMAP24XX_DMA_SPI1_RX2		21	/* E_DMA_21 */
+#define OMAP24XX_DMA_SPI1_TX3		22	/* E_DMA_22 */
+#define OMAP24XX_DMA_SPI1_RX3		23	/* E_DMA_23 */
+#else
 #define OMAP24XX_DMA_SPI1_TX0		35	/* S_DMA_34 */
 #define OMAP24XX_DMA_SPI1_RX0		36	/* S_DMA_35 */
 #define OMAP24XX_DMA_SPI1_TX1		37	/* S_DMA_36 */
@@ -279,6 +289,7 @@
 #define OMAP24XX_DMA_SPI1_RX2		40	/* S_DMA_39 */
 #define OMAP24XX_DMA_SPI1_TX3		41	/* S_DMA_40 */
 #define OMAP24XX_DMA_SPI1_RX3		42	/* S_DMA_41 */
+#endif
 #define OMAP24XX_DMA_SPI2_TX0		43	/* S_DMA_42 */
 #define OMAP24XX_DMA_SPI2_RX0		44	/* S_DMA_43 */
 #define OMAP24XX_DMA_SPI2_TX1		45	/* S_DMA_44 */
diff --git a/arch/arm/plat-omap/include/plat/hardware.h b/arch/arm/plat-omap/include/plat/hardware.h
index d5b26ad..e87efe1 100644
--- a/arch/arm/plat-omap/include/plat/hardware.h
+++ b/arch/arm/plat-omap/include/plat/hardware.h
@@ -286,5 +286,6 @@
 #include <plat/omap24xx.h>
 #include <plat/omap34xx.h>
 #include <plat/omap44xx.h>
+#include <plat/ti816x.h>
 
 #endif	/* __ASM_ARCH_OMAP_HARDWARE_H */
diff --git a/arch/arm/plat-omap/include/plat/ti816x.h b/arch/arm/plat-omap/include/plat/ti816x.h
new file mode 100644
index 0000000..df41d9f
--- /dev/null
+++ b/arch/arm/plat-omap/include/plat/ti816x.h
@@ -0,0 +1,46 @@
+/*
+ * arch/arm/plat-omap/include/plat/ti816x.h
+ *
+ * This file contains the address data for various ti816x modules.
+ *
+ * Copyright (C) 2010 Texas Instruments, Inc. - http://www.ti.com/
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License as
+ * published by the Free Software Foundation version 2.
+ *
+ * This program is distributed "as is" WITHOUT ANY WARRANTY of any
+ * kind, whether express or implied; without even the implied warranty
+ * of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ */
+
+#ifndef __ASM_ARCH_TI816X_H
+#define __ASM_ARCH_TI816X_H
+
+#define L3_TI816X_BASE		0x44000000
+#define L4_FAST_TI816X_BASE	0x4a000000
+#define L4_SLOW_TI816X_BASE	0x48000000
+
+#define TI816X_SCM_BASE		0x48140000
+#define TI816X_CTRL_BASE	TI816X_SCM_BASE
+#define TI816X_PRCM_BASE	0x48180000
+
+#define TI816X_ARM_INTC_BASE	0x48200000
+
+#define TI816X_GPMC_BASE	0x50000000
+
+#define TI816X_PCIE_REG_BASE    0x51000000
+#define TI816X_PCIE_MEM_BASE    0x20000000
+#define TI816X_PCIE_IO_BASE     0x40000000	/* Using 3MB reserved space */
+
+#define TI816X_USBSS_BASE	0x47400000
+#define TI816X_USBSS_LEN	0xFFF
+#define TI816X_USB0_BASE	0x47401000
+#define TI816X_USB1_BASE	0x47401800
+#define TI816X_USB_CPPIDMA_BASE	0x47402000
+#define TI816X_USB_CPPIDMA_LEN	0x5FFF
+
+#define TI816X_SATA_BASE	0x4A140000
+
+#endif /* __ASM_ARCH_TI816X_H */
diff --git a/arch/arm/plat-omap/include/plat/ti81xx_ram.h b/arch/arm/plat-omap/include/plat/ti81xx_ram.h
new file mode 100755
index 0000000..4ccea2d
--- /dev/null
+++ b/arch/arm/plat-omap/include/plat/ti81xx_ram.h
@@ -0,0 +1,37 @@
+/*
+ *
+ * Copyright (C) 2009 Texas Instruments Inc.
+ * Author: Yihe Hu  <yihehu@ti.com>
+ *
+ *
+ * Some code and ideas taken from TI OMAP2 Platforms
+ * by Tomi Valkeinen.
+ *
+ * This program is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License as published by the
+ * Free Software Foundation; either version 2 of the License, or (at your
+ * option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License along
+ * with this program; if not, write to the Free Software Foundation, Inc.,
+ * 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
+ */
+
+#ifndef __ARCH_ARM_TI81XX_RAM_H__
+#define __ARCH_ARM_TI81XX_RAM_H__
+
+#include <asm/types.h>
+
+extern int ti81xx_vram_add_region(unsigned long paddr, size_t size);
+extern int ti81xx_vram_free(unsigned long paddr, void *vaddr, size_t size);
+extern int ti81xx_vram_alloc(int mtype, size_t size, unsigned long *paddr);
+extern void ti81xx_set_sdram_vram(u32 size, u32 start);
+extern void ti81xx_vram_get_info(unsigned long *vram,
+				 unsigned long *free_vram,
+				 unsigned long *largest_free_block);
+#endif
diff --git a/arch/arm/plat-omap/include/plat/uncompress.h b/arch/arm/plat-omap/include/plat/uncompress.h
index 81d9ec5..df9408f 100644
--- a/arch/arm/plat-omap/include/plat/uncompress.h
+++ b/arch/arm/plat-omap/include/plat/uncompress.h
@@ -97,6 +97,10 @@ static inline void flush(void)
 	_DEBUG_LL_ENTRY(mach, OMAP2_UART1_BASE, OMAP_PORT_SHIFT,	\
 		ZOOM_UART_BASE, ZOOM_PORT_SHIFT, ZOOM_UART)
 
+#define DEBUG_LL_TI816X(p, mach)					\
+	_DEBUG_LL_ENTRY(mach, TI816X_UART1_BASE, OMAP_PORT_SHIFT,	\
+		TI816X_UART##p##_BASE, OMAP_PORT_SHIFT,	TI816XUART##p)
+
 static inline void __arch_decomp_setup(unsigned long arch_id)
 {
 	int port = 0;
@@ -162,6 +166,9 @@ static inline void __arch_decomp_setup(unsigned long arch_id)
 		DEBUG_LL_ZOOM(omap_zoom2);
 		DEBUG_LL_ZOOM(omap_zoom3);
 
+		/* TI8168 base boards using UART3 */
+		DEBUG_LL_TI816X(3, ti8168evm);
+
 	} while (0);
 }
 
-- 
1.7.0.2

