From 7433323b57bd98c68a57239c2df4c1b4563a36ba Mon Sep 17 00:00:00 2001
From: Sergio Aguirre <saaguirre@ti.com>
Date: Sun, 13 Sep 2009 15:23:31 +0300
Subject: [PATCH 6/9] V4L: Int if: Add support for private sensor ioctls

This patch adds the following internal private sensor ioctls:
- VIDIOC_INT_PRIV_G_PIXCLK
	Retrieves current pixel clock of the sensor.

- VIDIOC_INT_PRIV_G_ACTIVESIZE
	Retrieves current sensor selected basesize.

- VIDIOC_INT_PRIV_G_FULLSIZE
	Retrieves current sensor maximum size.

[Picked from http://arago-project.org/git/projects/linux-omap3.git]

Signed-off-by: Sergio Aguirre <saaguirre@ti.com>
Signed-off-by: Vaibhav Hiremath <hvaibhav@ti.com>
Integrated-by: Wang Hui <hui.wang@windriver.com>
---
 include/media/v4l2-int-device.h |    9 +++++++++
 1 files changed, 9 insertions(+), 0 deletions(-)

diff --git a/include/media/v4l2-int-device.h b/include/media/v4l2-int-device.h
index 81f4863..16c3897 100644
--- a/include/media/v4l2-int-device.h
+++ b/include/media/v4l2-int-device.h
@@ -231,6 +231,12 @@ enum v4l2_int_ioctl_num {
 	 *
 	 */
 	vidioc_int_priv_start_num = 2000,
+	/* VIDIOC_INT_PRIV_G_PIXCLK */
+	vidioc_int_priv_g_pixclk_num,
+	/* VIDIOC_INT_PRIV_G_ACTIVESIZE */
+	vidioc_int_priv_g_activesize_num,
+	/* VIDIOC_INT_PRIV_G_FULLSIZE */
+	vidioc_int_priv_g_fullsize_num,
 };
 
 /*
@@ -308,4 +314,7 @@ V4L2_INT_WRAPPER_0(reset);
 V4L2_INT_WRAPPER_0(init);
 V4L2_INT_WRAPPER_1(g_chip_ident, int, *);
 
+V4L2_INT_WRAPPER_1(priv_g_pixclk, u32, *);
+V4L2_INT_WRAPPER_1(priv_g_activesize, struct v4l2_pix_format, *);
+V4L2_INT_WRAPPER_1(priv_g_fullsize, struct v4l2_pix_format, *);
 #endif
-- 
1.6.5.2

