From 504561fae05e263dbcb554099792cd965e86445e Mon Sep 17 00:00:00 2001
From: Shaveta Leekha <shaveta@freescale.com>
Date: Tue, 1 Mar 2016 15:54:48 +0530
Subject: [PATCH 1149/1383] ls1043ardb: Add Kernel ITS file for UEFI

  This patch adds Kernel.its file for linux
  to boot on UEFI on LS1043ARDB platform

Signed-off-by: Shaveta Leekha <shaveta@freescale.com>
[Original patch taken from QorIQ-SDK-V2.0-20160527-yocto]
Signed-off-by: Yanjiang Jin <yanjiang.jin@windriver.com>
---
 kernel-ls1043a-uefi-rdb.its | 57 +++++++++++++++++++++++++++++++++++++++++++++
 1 file changed, 57 insertions(+)
 create mode 100644 kernel-ls1043a-uefi-rdb.its

diff --git a/kernel-ls1043a-uefi-rdb.its b/kernel-ls1043a-uefi-rdb.its
new file mode 100644
index 0000000..b39f3de
--- /dev/null
+++ b/kernel-ls1043a-uefi-rdb.its
@@ -0,0 +1,57 @@
+/*
+ * Copyright (C) 2015, Freescale Semiconductor
+ *
+ * Mingkai Hu <Mingkai.hu@freescale.com>
+ * Hou Zhiqiang <B48286@freescale.com>
+ *
+ * This file is licensed under the terms of the GNU General Public
+ * License version 2.  This program is licensed "as is" without any
+ * warranty of any kind, whether express or implied.
+ */
+
+/dts-v1/;
+
+/ {
+	description = "Image file for the LS1043A Linux Kernel";
+	#address-cells = <1>;
+
+	images {
+		kernel@1 {
+			description = "ARM64 Linux kernel";
+			/* UEFI doesn't have a decompression unit yet */
+			data = /incbin/("./arch/arm64/boot/Image");
+			type = "kernel";
+			arch = "arm64";
+			os = "linux";
+			compression = "none";
+			load = <0x80080000>;
+			entry = <0x80080000>;
+		};
+		fdt@1 {
+			description = "Flattened Device Tree blob";
+			data = /incbin/("./arch/arm64/boot/dts/freescale/fsl-ls1043a-uefi-rdb.dtb");
+			type = "flat_dt";
+			arch = "arm64";
+			compression = "none";
+			load = <0x90000000>;
+		};
+		ramdisk@1 {
+			description = "LS1043 Ramdisk";
+			data = /incbin/("./fsl-image-core-ls1043ardb.ext2.gz");
+			type = "ramdisk";
+			arch = "arm64";
+			os = "linux";
+			compression = "none";
+		};
+	};
+
+	configurations {
+		default = "config@1";
+		config@1 {
+			description = "Boot Linux kernel";
+			kernel = "kernel@1";
+			fdt = "fdt@1";
+			ramdisk = "ramdisk@1";
+		};
+	};
+};
-- 
2.8.1

