From 4d684754933cf3bed7f16a1239960f0c65d62cb4 Mon Sep 17 00:00:00 2001
From: Catalin Enache <catalin.enache@windriver.com>
Date: Tue, 4 Feb 2014 11:50:52 +0200
Subject: [PATCH 68/70] Fix WM8940 playback issue

Signed-off-by: Catalin Enache <catalin.enache@windriver.com>
---
 sound/soc/msm/mdm9615.c |   20 +++++++++++++++++++-
 1 files changed, 19 insertions(+), 1 deletions(-)

diff --git a/sound/soc/msm/mdm9615.c b/sound/soc/msm/mdm9615.c
index 29acde9..2effae1 100644
--- a/sound/soc/msm/mdm9615.c
+++ b/sound/soc/msm/mdm9615.c
@@ -2696,6 +2696,9 @@ static int mdm9615_sec_auxpcm_startup(struct snd_pcm_substream *substream)
 static int mdm9615_ar7_sec_auxpcm_startup(struct snd_pcm_substream *substream)
 {
 	int ret = 0;
+	struct snd_soc_pcm_runtime *rtd = substream->private_data;
+	struct snd_soc_codec *codec = rtd->codec;
+
 	pr_info("%s\n", __func__ );
 	pr_debug("%s(): substream = %s\n", __func__, substream->name);
 	
@@ -2711,9 +2714,24 @@ static int mdm9615_ar7_sec_auxpcm_startup(struct snd_pcm_substream *substream)
 		msm9615_config_sif_mux(MSM_SIF_FUNC_PCM);
 		msm9615_config_port_select();
 	}
+
+	mdm9615_ar7_enable_codec_ext_clk(codec, 1, true);
+
 	return 0;
 }
 
+static void mdm9615_ar7_sec_auxpcm_shutdown(struct snd_pcm_substream *substream)
+{
+	struct snd_soc_pcm_runtime *rtd = substream->private_data;
+	struct snd_soc_codec *codec = rtd->codec;
+
+	pr_debug("%s(): substream = %s\n", __func__, substream->name);
+	if (atomic_dec_return(&msm9615_sec_auxpcm_ref) == 0)
+		mdm9615_sec_aux_pcm_free_gpios();
+
+	mdm9615_ar7_enable_codec_ext_clk(codec, 0, true);
+}
+
 #endif
 
 /* SWISTOP */
@@ -2745,7 +2763,7 @@ static struct snd_soc_ops mdm9615_auxpcm_be_ops = {
 #if defined(CONFIG_SIERRA)
 static struct snd_soc_ops mdm9615_ar7_sec_auxpcm_be_ops = {
 	.startup = mdm9615_ar7_sec_auxpcm_startup,
-	.shutdown = mdm9615_sec_auxpcm_shutdown,
+	.shutdown = mdm9615_ar7_sec_auxpcm_shutdown,
 };
 #endif
 /* SWISTOP */
-- 
1.7.5.4

