From 604298610586d975c19cd56510362921153d2b4d Mon Sep 17 00:00:00 2001
From: auto commit <unknown@unknown>
Date: Fri, 24 Oct 2008 12:22:33 -0700
Subject: [PATCH] msg: importing 0158-cavium-remove-obsolete-mc146818rtc.patch

This is an automatic import of patch 0158-cavium-remove-obsolete-mc146818rtc.patch, no headers were
detected and a default message was constructed
---
 include/asm-mips/mach-cavium-octeon/mc146818rtc.h |   14 --------------
 1 files changed, 0 insertions(+), 14 deletions(-)
 delete mode 100644 include/asm-mips/mach-cavium-octeon/mc146818rtc.h

diff --git a/include/asm-mips/mach-cavium-octeon/mc146818rtc.h b/include/asm-mips/mach-cavium-octeon/mc146818rtc.h
deleted file mode 100644
index f49ece1..0000000
--- a/include/asm-mips/mach-cavium-octeon/mc146818rtc.h
+++ /dev/null
@@ -1,14 +0,0 @@
-/*
- * This file is subject to the terms and conditions of the GNU General Public
- * License.  See the file "COPYING" in the main directory of this archive
- * for more details.
- *
- * Copyright (C) 2004 Cavium Networks
- */
-#ifndef __ASM_MACH_CAVIUM_OCTEON_MC146818RTC_H__
-#define __ASM_MACH_CAVIUM_OCTEON_MC146818RTC_H__
-
-#define RTC_IRQ 88
-#include <asm/mach-generic/mc146818rtc.h>
-
-#endif
-- 
1.5.5.1

