From 40995abaaaa351401977781cbc80477bf4313b44 Mon Sep 17 00:00:00 2001
From: auto commit <unknown@unknown>
Date: Fri, 24 Oct 2008 12:22:35 -0700
Subject: [PATCH] cavium: barrier handling ops are in barrier.h

Move the syncw for the errata to barrier.h where it is used.

Signed-off-by: Tomaso Paoletti <tpaoletti@caviumnetworks.com>
Signed-off-by: Paul Gortmaker <Paul.Gortmaker@windriver.com>
---
 .../mach-cavium-octeon/cpu-feature-overrides.h     |    6 ------
 1 files changed, 0 insertions(+), 6 deletions(-)

diff --git a/include/asm-mips/mach-cavium-octeon/cpu-feature-overrides.h b/include/asm-mips/mach-cavium-octeon/cpu-feature-overrides.h
index f1ecf81..ee3b9d1 100644
--- a/include/asm-mips/mach-cavium-octeon/cpu-feature-overrides.h
+++ b/include/asm-mips/mach-cavium-octeon/cpu-feature-overrides.h
@@ -32,12 +32,6 @@
 #define cpu_has_saa             octeon_has_saa()
 #define cpu_has_mips64r2        1
 #define cpu_has_counter         1
-/* We actually use two syncw instructions in a row when we need a write memory
-   barrier. This is because the CN3XXX series of Octeons have errata Core-401.
-   This can cause a single syncw to not enforce ordering under very rare
-   conditions. Even if it is rare, better safe than sorry */
-#define OCTEON_SYNCW_STR ".set push\n.set arch=octeon\nsyncw\nsyncw\n.set pop\n"
-
 #define ARCH_HAS_READ_CURRENT_TIMER 1
 #define ARCH_HAS_IRQ_PER_CPU 1
 #define ARCH_HAS_SPINLOCK_PREFETCH 1
-- 
1.5.5.1

