From 56c03995e74662cf597808551aa98cbcb55ab69e Mon Sep 17 00:00:00 2001
From: Wally Gleemer <Wally.Gleemer@windriver.com>
Date: Tue, 25 Mar 2008 16:21:37 -0700
Subject: [PATCH] cavium: dma coherency isn't a problem

The Cavium Octeon has fully coherent DMA but doesn't use
CONFIG_DMA_COHERENT.

Signed-off-by: Wally Gleemer <Wally.Gleemer@windriver.com>
---
 include/asm-mips/mach-generic/dma-coherence.h |    7 +++++++
 1 files changed, 7 insertions(+), 0 deletions(-)

diff --git a/include/asm-mips/mach-generic/dma-coherence.h b/include/asm-mips/mach-generic/dma-coherence.h
index 76e04e7..bdafa8c 100644
--- a/include/asm-mips/mach-generic/dma-coherence.h
+++ b/include/asm-mips/mach-generic/dma-coherence.h
@@ -40,6 +40,13 @@ static inline int plat_device_is_coherent(struct device *dev)
 #ifdef CONFIG_DMA_NONCOHERENT
 	return 0;
 #endif
+#ifdef CONFIG_CPU_CAVIUM_OCTEON
+	/* The Cavium Octeon has fully coherent DMA but doesn't use
+		CONFIG_DMA_COHERENT so it can replace the normally DMA
+		mapping functions with ones supporting 64bit mappings
+		specific to Octeon */
+	return 1;
+#endif
 }
 
 #endif /* __ASM_MACH_GENERIC_DMA_COHERENCE_H */
-- 
1.5.5.1

