From bfcd9d34bf835c70b4e08f292164d849982070fb Mon Sep 17 00:00:00 2001
From: Wally Gleemer <Wally.Gleemer@windriver.com>
Date: Thu, 13 Mar 2008 20:59:46 -0700
Subject: [PATCH] cavium: use builtin prefetch

Cavium has selected to go with using the builtin prefetch() vs. using
the assembly one in processor.h

Signed-off-by: Wally Gleemer <Wally.Gleemer@windriver.com>
---
 include/asm-mips/processor.h |    5 +++++
 1 files changed, 5 insertions(+), 0 deletions(-)

diff --git a/include/asm-mips/processor.h b/include/asm-mips/processor.h
index d6589ec..834f044 100644
--- a/include/asm-mips/processor.h
+++ b/include/asm-mips/processor.h
@@ -303,12 +303,17 @@ unsigned long get_wchan(struct task_struct *p);
 
 static inline void prefetch(const void *addr)
 {
+#ifdef CONFIG_CPU_CAVIUM_OCTEON
+	/* Use the GCC builtin so the compiler can reorder instructions */
+	__builtin_prefetch(addr);
+#else
 	__asm__ __volatile__(
 	"	.set	mips4		\n"
 	"	pref	%0, (%1)	\n"
 	"	.set	mips0		\n"
 	:
 	: "i" (Pref_Load), "r" (addr));
+#endif
 }
 
 #endif
-- 
1.5.5.1

