From c48f81ded51d7c191b1dcb253486ef84241d0df4 Mon Sep 17 00:00:00 2001
From: Wally Gleemer <Wally.Gleemer@windriver.com>
Date: Thu, 13 Mar 2008 21:11:44 -0700
Subject: [PATCH] cavium: CONFIG_FAST_ACCESS_TO_THREAD_POINTER on stackframe

Stick the Cavium specific FAST_ACCESS_TO_THREAD_POINTER value on the
stackframe for MIPS (if enabled).

Signed-off-by: Wally Gleemer <Wally.Gleemer@windriver.com>
---
 include/asm-mips/stackframe.h |    3 +++
 1 files changed, 3 insertions(+), 0 deletions(-)

diff --git a/include/asm-mips/stackframe.h b/include/asm-mips/stackframe.h
index 92bde3d..d5b624c 100644
--- a/include/asm-mips/stackframe.h
+++ b/include/asm-mips/stackframe.h
@@ -404,6 +404,9 @@
 		.macro	RESTORE_SP_AND_RET
 		LONG_L	sp, PT_R29(sp)
 		.set	mips3
+#ifdef CONFIG_FAST_ACCESS_TO_THREAD_POINTER
+		LONG_L	k0, FAST_ACCESS_THREAD_OFFSET($0) /* K0 = thread pointer */
+#endif
 		eret
 		.set	mips0
 		.endm
-- 
1.5.5.1

