From 5ae31da1dcb17346c725806e9bedbda5dd640dd2 Mon Sep 17 00:00:00 2001
From: Wally Gleemer <Wally.Gleemer@windriver.com>
Date: Tue, 25 Mar 2008 16:21:41 -0700
Subject: [PATCH] cavium: use CPHYSADDR for __pa_symbol

If you don't use CPHYSADDR for __pa_symbol on 64 bit cavium,
the bootmem handling goes awry and you get silent boot death.

Signed-off-by: Wally Gleemer <Wally.Gleemer@windriver.com>
---
 include/asm-mips/page.h |    4 ++++
 1 files changed, 4 insertions(+), 0 deletions(-)

diff --git a/include/asm-mips/page.h b/include/asm-mips/page.h
index 25796fd..3658a09 100644
--- a/include/asm-mips/page.h
+++ b/include/asm-mips/page.h
@@ -161,7 +161,11 @@ typedef struct { unsigned long pgprot; } pgprot_t;
     ((unsigned long)(x) - PAGE_OFFSET + PHYS_OFFSET)
 #endif
 #define __va(x)		((void *)((unsigned long)(x) + PAGE_OFFSET - PHYS_OFFSET))
+#if (defined (CONFIG_64BIT) && defined(CONFIG_CPU_CAVIUM_OCTEON))
+#define __pa_symbol(x)	CPHYSADDR(x)
+#else
 #define __pa_symbol(x)	__pa(RELOC_HIDE((unsigned long)(x), 0))
+#endif
 
 #define pfn_to_kaddr(pfn)	__va((pfn) << PAGE_SHIFT)
 
-- 
1.5.5.1

