From 1fcb01d47113677dccdcd371d461c50b3e58d68b Mon Sep 17 00:00:00 2001
From: Wally Gleemer <Wally.Gleemer@windriver.com>
Date: Thu, 13 Mar 2008 22:35:36 -0700
Subject: [PATCH] cavium: extra sync in various compare/exchange ops

Add in extra sync for Cavium to ensure data integrity on
various compare and exchange operations.

Signed-off-by: Wally Gleemer <Wally.Gleemer@windriver.com>
---
 include/asm-mips/cmpxchg.h |    1 +
 include/asm-mips/system.h  |    2 ++
 2 files changed, 3 insertions(+), 0 deletions(-)

diff --git a/include/asm-mips/cmpxchg.h b/include/asm-mips/cmpxchg.h
index 4a812c3..b07cd50 100644
--- a/include/asm-mips/cmpxchg.h
+++ b/include/asm-mips/cmpxchg.h
@@ -35,6 +35,7 @@
 		: "memory");						\
 	} else if (cpu_has_llsc) {					\
 		__asm__ __volatile__(					\
+		OCTEON_SYNCW_STR					\
 		"	.set	push				\n"	\
 		"	.set	noat				\n"	\
 		"	.set	mips3				\n"	\
diff --git a/include/asm-mips/system.h b/include/asm-mips/system.h
index a944eda..5f6f2b4 100644
--- a/include/asm-mips/system.h
+++ b/include/asm-mips/system.h
@@ -101,6 +101,7 @@ static inline unsigned long __xchg_u32(volatile int * m, unsigned int val)
 		unsigned long dummy;
 
 		__asm__ __volatile__(
+		OCTEON_SYNCW_STR
 		"	.set	mips3					\n"
 		"1:	ll	%0, %3			# xchg_u32	\n"
 		"	.set	mips0					\n"
@@ -151,6 +152,7 @@ static inline __u64 __xchg_u64(volatile __u64 * m, __u64 val)
 		unsigned long dummy;
 
 		__asm__ __volatile__(
+		OCTEON_SYNCW_STR
 		"	.set	mips3					\n"
 		"1:	lld	%0, %3			# xchg_u64	\n"
 		"	move	%2, %z4					\n"
-- 
1.5.5.1

