From 764fc71892e61ed916466c2c8c26179744566908 Mon Sep 17 00:00:00 2001
From: Yongli He <yongli.he@windriver.com>
Date: Fri, 8 Aug 2008 11:46:13 +0800
Subject: [PATCH] memmon fix ppc64 tlb flush

bring back flush tlb pending only for memmon.

Signed-off-by: Yongli he   <yongli.he@windriver.com>
diff --git a/mm/memmon.c b/mm/memmon.c
index 57b2834..2271edd 100644
--- a/mm/memmon.c
+++ b/mm/memmon.c
@@ -11,8 +11,21 @@
 #include <asm/timex.h>
 
 
-#ifdef __PPC64__
+#ifdef CONFIG_PPC64
 #define BATCH_FLUSH
+static inline void flush_tlb_pending()
+{
+	struct ppc64_tlb_batch *tlbbatch = &get_cpu_var(ppc64_tlb_batch);
+
+	/* If there's a TLB batch pending, then we must flush it because the
+	 * pages are going to be freed and we really don't want to have a CPU
+	 * access a freed page because it has a stale TLB
+	 */
+	if (tlbbatch->index)
+		__flush_tlb_pending(tlbbatch);
+
+        put_cpu_var(ppc64_tlb_batch);
+}
 #endif
 
 /* The memmon "sister" page flags works like this:
