From fa008a5522fe9ddd8ba71b94bada89953f4ea1e1 Mon Sep 17 00:00:00 2001
From: Yue Tao <Yue.Tao@windriver.com>
Date: Wed, 19 Jun 2013 15:22:46 +0800
Subject: [PATCH] Remove redefinition

Signed-off-by: Yue Tao <Yue.Tao@windriver.com>
---
 drivers/gpu/drm/i915/i915_gem_execbuffer.c |    3 ---
 1 files changed, 0 insertions(+), 3 deletions(-)

diff --git a/drivers/gpu/drm/i915/i915_gem_execbuffer.c b/drivers/gpu/drm/i915/i915_gem_execbuffer.c
index c26d242..785ad60 100644
--- a/drivers/gpu/drm/i915/i915_gem_execbuffer.c
+++ b/drivers/gpu/drm/i915/i915_gem_execbuffer.c
@@ -709,9 +709,6 @@ validate_exec_list(struct drm_i915_gem_exec_object2 *exec,
 	int relocs_total = 0;
 	int relocs_max = INT_MAX / sizeof(struct drm_i915_gem_relocation_entry);
 
-	int relocs_total = 0;
-	int relocs_max = INT_MAX / sizeof(struct drm_i915_gem_relocation_entry);
-
 	for (i = 0; i < count; i++) {
 		char __user *ptr = (char __user *)(uintptr_t)exec[i].relocs_ptr;
 		int length; /* limited by fault_in_pages_readable() */
-- 
1.7.5.4

