From 22c2a456a59fb38a56ce675168c607ec54fa1038 Mon Sep 17 00:00:00 2001
From: Daniel Vetter <daniel.vetter@ffwll.ch>
Date: Fri, 19 Jun 2015 21:17:42 +0200
Subject: [PATCH 0699/1077] drm/i915: Update DRIVER_DATE to 20150619

commit fbb35c1981012c72adff8f09f09005a8900a6dfd upstream.

Signed-off-by: Daniel Vetter <daniel.vetter@ffwll.ch>
Signed-off-by: Pengyu Ma <pengyu.ma@windriver.com>
---
 drivers/gpu/drm/i915/i915_drv.h | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/drivers/gpu/drm/i915/i915_drv.h b/drivers/gpu/drm/i915/i915_drv.h
index 6c1f37a..fb7fc91 100644
--- a/drivers/gpu/drm/i915/i915_drv.h
+++ b/drivers/gpu/drm/i915/i915_drv.h
@@ -56,7 +56,7 @@
 
 #define DRIVER_NAME		"i915"
 #define DRIVER_DESC		"Intel Graphics"
-#define DRIVER_DATE		"20150522"
+#define DRIVER_DATE		"20150619"
 
 #undef WARN_ON
 /* Many gcc seem to no see through this and fall over :( */
-- 
2.0.2

