From 1fb96a9c7fd64d8e10e4ed0d1627573c76ad84eb Mon Sep 17 00:00:00 2001
From: Daniel Vetter <daniel.vetter@ffwll.ch>
Date: Fri, 3 Jul 2015 15:53:23 +0200
Subject: [PATCH 0856/1077] drm/i915: Update DRIVER_DATE to 20150703

commit a3d1d001c9c583c79b1d9ab20f0a24333e35a3f8 upstream.

Signed-off-by: Daniel Vetter <daniel.vetter@ffwll.ch>
Signed-off-by: Pengyu Ma <pengyu.ma@windriver.com>
---
 drivers/gpu/drm/i915/i915_drv.h | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/drivers/gpu/drm/i915/i915_drv.h b/drivers/gpu/drm/i915/i915_drv.h
index 25d787f..c06bd77 100644
--- a/drivers/gpu/drm/i915/i915_drv.h
+++ b/drivers/gpu/drm/i915/i915_drv.h
@@ -56,7 +56,7 @@
 
 #define DRIVER_NAME		"i915"
 #define DRIVER_DESC		"Intel Graphics"
-#define DRIVER_DATE		"20150619"
+#define DRIVER_DATE		"20150703"
 
 #undef WARN_ON
 /* Many gcc seem to no see through this and fall over :( */
-- 
2.0.2

