From c3dca86a387e72c581e6e87278b7989faf71ad2d Mon Sep 17 00:00:00 2001
From: Daniel Vetter <daniel.vetter@ffwll.ch>
Date: Fri, 17 Jul 2015 22:24:32 +0200
Subject: [PATCH 0965/1077] drm/i915: Update DRIVER_DATE to 20150717

commit e0548f1979bfee900fb0671a5dd3a2f217dce5df upstream.

Signed-off-by: Daniel Vetter <daniel.vetter@ffwll.ch>
Signed-off-by: Pengyu Ma <pengyu.ma@windriver.com>
---
 drivers/gpu/drm/i915/i915_drv.h | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/drivers/gpu/drm/i915/i915_drv.h b/drivers/gpu/drm/i915/i915_drv.h
index 7d505a7..b1f93b6 100644
--- a/drivers/gpu/drm/i915/i915_drv.h
+++ b/drivers/gpu/drm/i915/i915_drv.h
@@ -56,7 +56,7 @@
 
 #define DRIVER_NAME		"i915"
 #define DRIVER_DESC		"Intel Graphics"
-#define DRIVER_DATE		"20150703"
+#define DRIVER_DATE		"20150717"
 
 #undef WARN_ON
 /* Many gcc seem to no see through this and fall over :( */
-- 
2.0.2

