From 5ef7dd86327e1e0990bd599db8d3fdece300f232 Mon Sep 17 00:00:00 2001
From: Alex Deucher <alexander.deucher@amd.com>
Date: Wed, 17 Apr 2013 16:27:40 -0400
Subject: [PATCH 553/886] drm/radeon/si: fix typo in function name

commit ba19031a80950aa50bc5d09b57b689972c015179 upstream

Signed-off-by: Alex Deucher <alexander.deucher@amd.com>
---
 drivers/gpu/drm/radeon/si.c |    4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/drivers/gpu/drm/radeon/si.c b/drivers/gpu/drm/radeon/si.c
index ae218b5..e68e2fd 100644
--- a/drivers/gpu/drm/radeon/si.c
+++ b/drivers/gpu/drm/radeon/si.c
@@ -4787,7 +4787,7 @@ static void si_init_gfx_cgpg(struct radeon_device *rdev)
 	WREG32(RLC_AUTO_PG_CTRL, tmp);
 }
 
-static u32 get_cu_active_bitmap(struct radeon_device *rdev, u32 se, u32 sh)
+static u32 si_get_cu_active_bitmap(struct radeon_device *rdev, u32 se, u32 sh)
 {
 	u32 mask = 0, tmp, tmp1;
 	int i;
@@ -4822,7 +4822,7 @@ static void si_init_ao_cu_mask(struct radeon_device *rdev)
 			cu_bitmap = 0;
 			counter  = 0;
 			for (k = 0; k < rdev->config.si.max_cu_per_sh; k++) {
-				if (get_cu_active_bitmap(rdev, i, j) & mask) {
+				if (si_get_cu_active_bitmap(rdev, i, j) & mask) {
 					if (counter < 2)
 						cu_bitmap |= mask;
 					counter++;
-- 
1.7.9.5

