From 310cb12bbce614cbe92d8bed2d00686127e17a61 Mon Sep 17 00:00:00 2001
From: Daniel Vetter <daniel.vetter@ffwll.ch>
Date: Fri, 22 Aug 2014 22:39:37 +0200
Subject: [PATCH 1133/1369] drm/i915: Update DRIVER_DATE to 20140822

commit c28135481428d0674fcc1da0740ed3f4343df5b2 upstream.

Signed-off-by: Daniel Vetter <daniel.vetter@ffwll.ch>
Signed-off-by: Pengyu Ma <pengyu.ma@windriver.com>
---
 drivers/gpu/drm/i915/i915_drv.h |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/drivers/gpu/drm/i915/i915_drv.h b/drivers/gpu/drm/i915/i915_drv.h
index 910035f..ed6d778 100644
--- a/drivers/gpu/drm/i915/i915_drv.h
+++ b/drivers/gpu/drm/i915/i915_drv.h
@@ -54,7 +54,7 @@
 
 #define DRIVER_NAME		"i915"
 #define DRIVER_DESC		"Intel Graphics"
-#define DRIVER_DATE		"20140808"
+#define DRIVER_DATE		"20140822"
 
 enum pipe {
 	INVALID_PIPE = -1,
-- 
1.7.5.4

