From 1b2428de018422b916c983bef1c9fc65556a6ca8 Mon Sep 17 00:00:00 2001
From: Yanjiang Jin <yanjiang.jin@windriver.com>
Date: Wed, 9 Apr 2014 14:26:15 +0800
Subject: [PATCH] cav-octeon3: set IRQF_PERCPU flag for CIB handler

Set IRQF_PERCPU flag for CIB handler.
CIB is a sub-layer of the central-interrupt unit (CIU) in CN7XXX.
It contains some devices' special interrupts, such as

CIU_CIB_USBDRD(0..1)_RAW(0)
CIU_CIB_SATA_RAW(0)

It needs IRQF_PERCPU flag, else system often hangs
during udevd if USB or SATA disk is connected.
---
 arch/mips/cavium-octeon/octeon-irq.c |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/arch/mips/cavium-octeon/octeon-irq.c b/arch/mips/cavium-octeon/octeon-irq.c
index 53acb5f..42d4f89 100644
--- a/arch/mips/cavium-octeon/octeon-irq.c
+++ b/arch/mips/cavium-octeon/octeon-irq.c
@@ -2756,7 +2756,7 @@ static int __init octeon_irq_init_cib(struct device_node *ciu_node,
 		return -ENOMEM;
 	}
 
-	r = request_irq(parent_irq, octeon_irq_cib_handler, IRQF_NO_THREAD,
+	r = request_irq(parent_irq, octeon_irq_cib_handler, IRQF_PERCPU,
 			"cib", cib_domain);
 	if (r) {
 		pr_err("request_irq cib failed %d\n", r);
-- 
1.7.5.4

