From debd054db89d36531abd497ef51c36c32becc022 Mon Sep 17 00:00:00 2001
From: Yue Tao <yue.tao@windriver.com>
Date: Fri, 16 May 2014 18:05:02 -0700
Subject: [PATCH 2/2] Remove useless codes.

Signed-off-by: Yue Tao <yue.tao@windriver.com>
---
 drivers/i2c/busses/i2c-xlp9xx.c |    3 ---
 1 files changed, 0 insertions(+), 3 deletions(-)

diff --git a/drivers/i2c/busses/i2c-xlp9xx.c b/drivers/i2c/busses/i2c-xlp9xx.c
index d9203d3..5bf41b2 100644
--- a/drivers/i2c/busses/i2c-xlp9xx.c
+++ b/drivers/i2c/busses/i2c-xlp9xx.c
@@ -42,7 +42,6 @@
 #include <linux/i2c.h>
 #include <linux/io.h>
 #include <linux/platform_device.h>
-#include <linux/of_i2c.h>
 
 #include <asm/netlogic/haldefs.h>
 #include <asm/netlogic/xlp-hal/iomap.h>
@@ -474,8 +473,6 @@ static int xlp9xx_i2c_probe(struct platform_device *pdev)
 
 	xlp9xx_i2c_init(priv);
 
-	of_i2c_register_devices(&priv->adap);
-
 	return 0;
 }
 
-- 
1.7.5.4

