From b1f2885c5dc8c9066ffd2a01648e167b9cb3a609 Mon Sep 17 00:00:00 2001
From: auto commit <unknown@unknown>
Date: Wed, 24 Aug 2011 12:53:38 -0700
Subject: [PATCH 092/238] Cavium: MDIO bus API update

Source: Local development

Update mdiobus_read/mdiobus_write API to account for change between
2.6.32 and 2.6.34 kernels.

Signed-off-by: Phil Staub <Phil.Staub@windriver.com>
---
 simple_exec/cvmx-mdio.h |    4 ++--
 1 files changed, 2 insertions(+), 2 deletions(-)

diff --git a/simple_exec/cvmx-mdio.h b/simple_exec/cvmx-mdio.h
index 8fdfdc1..2759556 100644
--- a/simple_exec/cvmx-mdio.h
+++ b/simple_exec/cvmx-mdio.h
@@ -388,7 +388,7 @@ static inline int cvmx_mdio_read(int bus_id, int phy_id, int location)
 	if (bus == NULL)
 		return -1;
 
-	rv = mdiobus_read(bus, phy_id, location);
+	rv = mdiobus_read(bus, phy_id, 0, location);
 
 	if (rv < 0)
 		return -1;
@@ -440,7 +440,7 @@ static inline int cvmx_mdio_write(int bus_id, int phy_id, int location, int val)
 	if (bus == NULL)
 		return -1;
 
-	rv = mdiobus_write(bus, phy_id, location, (u16)val);
+	rv = mdiobus_write(bus, phy_id, 0, location, (u16)val);
 
 	if (rv < 0)
 		return -1;
-- 
1.7.0

