From b4294757cee13832e55dc43ad069399099996489 Mon Sep 17 00:00:00 2001
From: Phil Staub <Phil.Staub@windriver.com>
Date: Wed, 3 Aug 2011 09:50:19 -0700
Subject: [PATCH 090/238] Cavium: Remove cvmx-tlb.c from usermode builds

Source: Local development after discussion with Cavium regarding
Simple Exec build failures.

cvmx-tlb.c must be removed from usermode builds for the simple exec
applications to build.

Signed-off-by: Phil Staub <Phil.Staub@windriver.com>
---
 simple_exec/cvmx.mk |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/simple_exec/cvmx.mk b/simple_exec/cvmx.mk
index 78e76f9..3b58280 100644
--- a/simple_exec/cvmx.mk
+++ b/simple_exec/cvmx.mk
@@ -119,7 +119,6 @@ OBJS_$(d)  :=  \
 	$(OBJ_DIR)/cvmx-sysinfo.o \
 	$(OBJ_DIR)/cvmx-thunder.o \
 	$(OBJ_DIR)/cvmx-tim.o \
-	$(OBJ_DIR)/cvmx-tlb.o \
 	$(OBJ_DIR)/cvmx-tra.o \
 	$(OBJ_DIR)/cvmx-twsi.o \
 	$(OBJ_DIR)/cvmx-usb.o \
@@ -135,6 +134,7 @@ OBJS_$(d)  +=  \
 	$(OBJ_DIR)/cvmx-app-init-linux.o
 else
 OBJS_$(d)  +=  \
+	$(OBJ_DIR)/cvmx-tlb.o \
 	$(OBJ_DIR)/cvmx-debug.o \
 	$(OBJ_DIR)/cvmx-debug-handler.o \
 	$(OBJ_DIR)/cvmx-debug-remote.o \
-- 
1.7.0

