From 42aa458b9de4ae8bb557ddaa930f7540a66dc06e Mon Sep 17 00:00:00 2001
From: Benjamin Walsh <benjamin.walsh@windriver.com>
Date: Sat, 14 Feb 2009 02:56:05 -0500
Subject: [PATCH] kexec: fix for emer_atca6101

IOMMU is forced off when the amount of memory is less than 1GB. This
prevents front panel network interface in the crash kernel on ATCA6101
from functionning. Always force the IOMMU on in these conditions.

Signed-off by: Benjamin Walsh <benjamin.walsh@windriver.com>
---
 arch/powerpc/sysdev/dart_iommu.c |    6 ++++++
 1 files changed, 6 insertions(+), 0 deletions(-)

diff --git a/arch/powerpc/sysdev/dart_iommu.c b/arch/powerpc/sysdev/dart_iommu.c
index 89639ec..7e4b7d6 100644
--- a/arch/powerpc/sysdev/dart_iommu.c
+++ b/arch/powerpc/sysdev/dart_iommu.c
@@ -404,6 +404,12 @@ void __init alloc_dart_table(void)
 	 * (i.e. Airport Extreme) have 30 bit address range limits.
 	 */
 
+#if (defined(CONFIG_CRASH_DUMP) && defined(CONFIG_PPC_MAPLE))
+	/* IOMMU is needed, normally using 128MB of memory, force it on */
+	iommu_is_off = 0;
+	iommu_force_on = 1;
+#endif
+
 	if (iommu_is_off)
 		return;
 
-- 
1.6.0.3

