From f14277433e5769a7b0d5e17ae2a3b29634aee077 Mon Sep 17 00:00:00 2001
From: Paul Gortmaker <paul.gortmaker@windriver.com>
Date: Wed, 28 May 2014 10:24:00 -0400
Subject: [PATCH] bnx2x: use csum fcn directly since tcp_v6_check isnt exported
 in 3.4

Signed-off-by: Paul Gortmaker <paul.gortmaker@windriver.com>

diff --git a/drivers/net/ethernet/broadcom/bnx2x/bnx2x_cmn.c b/drivers/net/ethernet/broadcom/bnx2x/bnx2x_cmn.c
index f3f4d1439e4b..e677b4515fcb 100644
--- a/drivers/net/ethernet/broadcom/bnx2x/bnx2x_cmn.c
+++ b/drivers/net/ethernet/broadcom/bnx2x/bnx2x_cmn.c
@@ -521,8 +521,15 @@ static void bnx2x_gro_ipv6_csum(struct bnx2x *bp, struct sk_buff *skb)
 	skb_set_transport_header(skb, sizeof(struct ipv6hdr));
 	th = tcp_hdr(skb);
 
+/* 3.4.x doesn't export tcp_v6_check .... so use csum directly. */
+#if 0
 	th->check = ~tcp_v6_check(skb->len - skb_transport_offset(skb),
 				  &iph->saddr, &iph->daddr, 0);
+#else
+	th->check = ~csum_ipv6_magic(&iph->saddr, &iph->daddr,
+				     skb->len - skb_transport_offset(skb),
+				     IPPROTO_TCP, 0);
+#endif
 }
 #endif
 
-- 
1.9.3

