From 509499c41550cc28f35aa0a7cc5fff662d164f26 Mon Sep 17 00:00:00 2001
From: Bruce Ashfield <bruce.ashfield@windriver.com>
Date: Fri, 22 Mar 2013 15:00:50 -0400
Subject: [PATCH] ovp: add localversion-ovp to indicate ovp kernel version

Signed-off-by: Bruce Ashfield <bruce.ashfield@windriver.com>
---
 localversion-ovp |    1 +
 1 file changed, 1 insertion(+)
 create mode 100644 localversion-ovp

diff --git a/localversion-ovp b/localversion-ovp
new file mode 100644
index 0000000..d4e02ea
--- /dev/null
+++ b/localversion-ovp
@@ -0,0 +1 @@
+-ovp1
-- 
1.7.10.4

