From 8ab236faf66832d9aec760bbe0aa51b4072b71ab Mon Sep 17 00:00:00 2001
From: Paul Gortmaker <paul.gortmaker@windriver.com>
Date: Sun, 28 May 2017 16:43:34 -0400
Subject: [PATCH 82/82] x86/kvm: fix missing prototype causing build error

Signed-off-by: Paul Gortmaker <paul.gortmaker@windriver.com>
---
 arch/x86/kvm/x86.c |    4 ++++
 1 files changed, 4 insertions(+), 0 deletions(-)

diff --git a/arch/x86/kvm/x86.c b/arch/x86/kvm/x86.c
index 8b51c28..8ebe29d 100644
--- a/arch/x86/kvm/x86.c
+++ b/arch/x86/kvm/x86.c
@@ -120,6 +120,10 @@ static int auto_inject_virq;
 module_param(inject_enabled, bool, S_IRUGO | S_IWUSR);
 #endif
 
+#ifdef CONFIG_VIRT_INT_LAT_HOST
+static void virq_inject_time(struct kvm_vcpu *vcpu);
+#endif
+
 #ifdef CONFIG_VIRT_INT_LOST_HOST
 /*
  * By default, we start injecting a quater of the maximum
-- 
1.7.5.4

