From 550a32912d130800631f514498f9bd25a9e0e0a0 Mon Sep 17 00:00:00 2001
From: Vinod Koul <vinod.koul@linux.intel.com>
Date: Tue, 25 Sep 2012 09:57:36 +0530
Subject: [PATCH 026/142] of: dma- fix build break for !CONFIG_OF

commit 4c26bc601d20fa67eefcb27477feda130c10790e upstream

Signed-off-by: Vinod Koul <vinod.koul@linux.intel.com>
Signed-off-by: fupan li <fupan.li@windriver.com>
---
 include/linux/of_dma.h |   29 ++++++++++++++++++++++++++++-
 1 files changed, 28 insertions(+), 1 deletions(-)

diff --git a/include/linux/of_dma.h b/include/linux/of_dma.h
index 337823d..667c19f 100644
--- a/include/linux/of_dma.h
+++ b/include/linux/of_dma.h
@@ -32,14 +32,41 @@ struct of_dma_filter_info {
 	dma_filter_fn	filter_fn;
 };
 
+#ifdef CONFIG_OF
 extern int of_dma_controller_register(struct device_node *np,
 		struct dma_chan *(*of_dma_xlate)
 		(struct of_phandle_args *, struct of_dma *),
 		void *data);
-extern void of_dma_controller_free(struct device_node *np);
+extern int of_dma_controller_free(struct device_node *np);
 extern struct dma_chan *of_dma_request_slave_channel(struct device_node *np,
 						     char *name);
 extern struct dma_chan *of_dma_simple_xlate(struct of_phandle_args *dma_spec,
 		struct of_dma *ofdma);
+#else
+static int of_dma_controller_register(struct device_node *np,
+		struct dma_chan *(*of_dma_xlate)
+		(struct of_phandle_args *, struct of_dma *),
+		void *data)
+{
+	return -ENODEV;
+}
+
+static void of_dma_controller_free(struct device_node *np)
+{
+}
+
+static struct dma_chan *of_dma_request_slave_channel(struct device_node *np,
+						     char *name)
+{
+	return NULL;
+}
+
+static struct dma_chan *of_dma_simple_xlate(struct of_phandle_args *dma_spec,
+		struct of_dma *ofdma)
+{
+	return NULL;
+}
+
+#endif
 
 #endif /* __LINUX_OF_DMA_H */
-- 
1.7.5.4

