From 83fb34bae084fc9d427a2026c4364da41cc3dad5 Mon Sep 17 00:00:00 2001
From: Mike Isely <isely@pobox.com>
Date: Mon, 20 Feb 2012 02:26:06 -0300
Subject: [PATCH 090/252] pvrusb2: Implement querystd for videodev_ioctl2

commit 0927ee67dd59f715f2a6c796a86a0eda9ea2e7b2 upstream

Signed-off-by: Mike Isely <isely@pobox.com>
Signed-off-by: Mauro Carvalho Chehab <mchehab@redhat.com>
Signed-off-by: fupan li <fupan.li@windriver.com>
---
 drivers/media/video/pvrusb2/pvrusb2-v4l2.c |    9 +++++++++
 1 files changed, 9 insertions(+), 0 deletions(-)

diff --git a/drivers/media/video/pvrusb2/pvrusb2-v4l2.c b/drivers/media/video/pvrusb2/pvrusb2-v4l2.c
index 83bc467..3669090 100644
--- a/drivers/media/video/pvrusb2/pvrusb2-v4l2.c
+++ b/drivers/media/video/pvrusb2/pvrusb2-v4l2.c
@@ -205,6 +205,14 @@ int pvr2_s_std(struct file *file, void *priv, v4l2_std_id *std)
 		pvr2_hdw_get_ctrl_by_id(hdw, PVR2_CID_STDCUR), *std);
 }
 
+static int pvr2_querystd(struct file *file, void *priv, v4l2_std_id *std)
+{
+	struct pvr2_v4l2_fh *fh = file->private_data;
+	struct pvr2_hdw *hdw = fh->channel.mc_head->hdw;
+
+	return pvr2_hdw_get_detected_std(hdw, std);
+}
+
 static int pvr2_enum_input(struct file *file, void *priv, struct v4l2_input *vi)
 {
 	struct pvr2_v4l2_fh *fh = file->private_data;
@@ -844,6 +852,7 @@ static const struct v4l2_ioctl_ops pvr2_ioctl_ops = {
 	.vidioc_g_tuner			    = pvr2_g_tuner,
 	.vidioc_g_std			    = pvr2_g_std,
 	.vidioc_s_std			    = pvr2_s_std,
+	.vidioc_querystd		    = pvr2_querystd,
 	.vidioc_log_status		    = pvr2_log_status,
 	.vidioc_enum_fmt_vid_cap	    = pvr2_enum_fmt_vid_cap,
 	.vidioc_g_fmt_vid_cap		    = pvr2_g_fmt_vid_cap,
-- 
1.7.5.4

