From 58bc7d9d7cb794b8ca371fabe7a2d0aabc1954d9 Mon Sep 17 00:00:00 2001
From: Kevin Hao <kexin.hao@windriver.com>
Date: Mon, 21 Oct 2013 11:12:40 +0800
Subject: [PATCH 4/5] ocf: pass the correct size argument to memset

Signed-off-by: Kevin Hao <kexin.hao@windriver.com>
---
 crypto/ocf/cryptosoft.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/crypto/ocf/cryptosoft.c b/crypto/ocf/cryptosoft.c
index aa2383d..7b5e938 100644
--- a/crypto/ocf/cryptosoft.c
+++ b/crypto/ocf/cryptosoft.c
@@ -320,7 +320,7 @@ execute_later(void (fn)(void *), void *arg)
 
 	w = (execute_later_t *) kmalloc(sizeof(execute_later_t), SLAB_ATOMIC);
 	if (w) {
-		memset(w, '\0', sizeof(w));
+		memset(w, '\0', sizeof(*w));
 		w->func = fn;
 		w->arg = arg;
 #if LINUX_VERSION_CODE >= KERNEL_VERSION(2,6,20)
-- 
1.8.1.2

