From 011c2c11cd9dc619e17ecd4cfb5cb670b889addf Mon Sep 17 00:00:00 2001
From: auto commit <unknown@unknown>
Date: Fri, 24 Oct 2008 12:22:36 -0700
Subject: [PATCH] cavium: fix type on read_current_timer

read_current_timer() had the wrong type an the wrong return value.

Signed-off-by: Paul Gortmaker <Paul.Gortmaker@windriver.com>
---
 .../mach-cavium-octeon/cpu-feature-overrides.h     |    4 ++--
 1 files changed, 2 insertions(+), 2 deletions(-)

diff --git a/include/asm-mips/mach-cavium-octeon/cpu-feature-overrides.h b/include/asm-mips/mach-cavium-octeon/cpu-feature-overrides.h
index ee3b9d1..6c5cb4b 100644
--- a/include/asm-mips/mach-cavium-octeon/cpu-feature-overrides.h
+++ b/include/asm-mips/mach-cavium-octeon/cpu-feature-overrides.h
@@ -38,14 +38,14 @@
 #define spin_lock_prefetch(x) prefetch(x)
 #define PREFETCH_STRIDE 128
 
-static inline unsigned long read_current_timer(unsigned long *result)
+static inline int read_current_timer(unsigned long *result)
 {
 	asm volatile ("rdhwr %0,$31\n"
 #ifndef CONFIG_64BIT
 		      "sll %0, 0\n"
 #endif
 		      :"=r" (*result));
-	return *result;
+	return 0;
 }
 
 static inline int octeon_has_saa(void)
-- 
1.5.5.1

