From 1a7ee3862eaec7491cd5f775db69699c247869d4 Mon Sep 17 00:00:00 2001
From: auto commit <unknown@unknown>
Date: Fri, 24 Oct 2008 12:22:39 -0700
Subject: [PATCH] _msg: importing 0163-cavium-fix-dma-octeon-without-PCI.patch

This is an automatic import of patch 0163-cavium-fix-dma-octeon-without-PCI.patch, no headers were
detected and a default message was constructed
---
 arch/mips/cavium-octeon/dma-octeon.c |    2 ++
 1 files changed, 2 insertions(+), 0 deletions(-)

diff --git a/arch/mips/cavium-octeon/dma-octeon.c b/arch/mips/cavium-octeon/dma-octeon.c
index f52fe9b..2b3f63a 100644
--- a/arch/mips/cavium-octeon/dma-octeon.c
+++ b/arch/mips/cavium-octeon/dma-octeon.c
@@ -23,7 +23,9 @@
 #include <linux/io.h>
 
 #include "hal.h"
+#ifdef CONFIG_PCI
 #include "pci-common.h"
+#endif
 
 #define BAR2_PCI_ADDRESS 0x8000000000ul
 
-- 
1.5.5.1

