From 34c3d9d75ffca1b71c8317ab606a375fcd0ed19a Mon Sep 17 00:00:00 2001
From: auto commit <unknown@unknown>
Date: Fri, 24 Oct 2008 12:22:39 -0700
Subject: [PATCH] auto_msg: importing 0164-cavium-fix-pci-console-init.patch

This is an automatic import of patch 0164-cavium-fix-pci-console-init.patch, no headers were
detected and a default message was constructed
---
 arch/mips/cavium-octeon/setup.c |    6 ++++--
 1 files changed, 4 insertions(+), 2 deletions(-)

diff --git a/arch/mips/cavium-octeon/setup.c b/arch/mips/cavium-octeon/setup.c
index 5befd72..6d27bfa 100644
--- a/arch/mips/cavium-octeon/setup.c
+++ b/arch/mips/cavium-octeon/setup.c
@@ -41,8 +41,9 @@ extern void cvmx_interrupt_rsl_enable(void);
 
 extern struct plat_smp_ops octeon_smp_ops;
 extern void octeon_user_io_init(void);
+#ifdef CONFIG_PCI
 extern void pci_console_init(const char *arg);
-
+#endif
 extern void putDebugChar(char ch);
 
 #ifdef CONFIG_CAVIUM_OCTEON_BOOTBUS_COMPACT_FLASH
@@ -224,9 +225,10 @@ void __init prom_init(void)
 		ebt3000_cf_enable_dma();
 #endif
 
-
+#ifdef CONFIG_PCI
 	if (strstr(arcs_cmdline, "console=pci"))
 		pci_console_init(strstr(arcs_cmdline, "console=pci") + 8);
+#endif
 
 	if (strstr(arcs_cmdline, "console=") == NULL) {
 #ifdef CONFIG_GDB_CONSOLE
-- 
1.5.5.1

