From b0e228935788c4c4066d3e59b0136405687e8fa5 Mon Sep 17 00:00:00 2001
From: auto commit <unknown@unknown>
Date: Fri, 24 Oct 2008 12:22:39 -0700
Subject: [PATCH] cavium: define MIPS_IRQ_BASE

We could take this from mach-generic, but we already have an irq.h so
add it in there.

Signed-off-by: Tomaso Paoletti <tpaoletti@caviumnetworks.com>
Signed-off-by: Paul Gortmaker <Paul.Gortmaker@windriver.com>
---
 include/asm-mips/mach-cavium-octeon/irq.h |    1 +
 1 files changed, 1 insertions(+), 0 deletions(-)

diff --git a/include/asm-mips/mach-cavium-octeon/irq.h b/include/asm-mips/mach-cavium-octeon/irq.h
index 43d5b68..3dc0e0b 100644
--- a/include/asm-mips/mach-cavium-octeon/irq.h
+++ b/include/asm-mips/mach-cavium-octeon/irq.h
@@ -9,6 +9,7 @@
 #define __OCTEON_IRQ_H__
 
 #define NR_IRQS OCTEON_IRQ_LAST
+#define MIPS_CPU_IRQ_BASE 0
 
 /* 0 - 7 represent the 8 MIPS standard interrupt sources */
 #define OCTEON_IRQ_SW0          0
-- 
1.5.5.1

