From b451dab25b19d9e554b53c3b1eb34ddae67e2e3b Mon Sep 17 00:00:00 2001
From: auto commit <unknown@unknown>
Date: Fri, 24 Oct 2008 12:22:43 -0700
Subject: [PATCH] cavium: add syncw to bitops.h

Add in extra syncw to ensure data integrity on Cavium CPUs for selected
bitops like test_and_set, test_and_clear, etc.

Signed-off-by: Tomaso Paoletti <tpaoletti@caviumnetworks.com>
Signed-off-by: Paul Gortmaker <Paul.Gortmaker@windriver.com>
---
 include/asm-mips/bitops.h |    6 ++++--
 1 files changed, 4 insertions(+), 2 deletions(-)

diff --git a/include/asm-mips/bitops.h b/include/asm-mips/bitops.h
index ac7ebe4..a2a2106 100644
--- a/include/asm-mips/bitops.h
+++ b/include/asm-mips/bitops.h
@@ -282,6 +282,7 @@ static inline int test_and_set_bit(unsigned long nr,
 		__asm__ __volatile__(
 		"	.set	push					\n"
 		"	.set	noreorder				\n"
+		OCTEON_SYNCW_STR
 		"	.set	mips3					\n"
 		"1:	" __LL "%0, %1		# test_and_set_bit	\n"
 		"	or	%2, %0, %3				\n"
@@ -438,6 +439,7 @@ static inline int test_and_clear_bit(unsigned long nr,
 		__asm__ __volatile__(
 		"	.set	push					\n"
 		"	.set	noreorder				\n"
+		OCTEON_SYNCW_STR
 		"	.set	mips3					\n"
 		"1:	" __LL	"%0, %1		# test_and_clear_bit	\n"
 		"	or	%2, %0, %3				\n"
@@ -509,6 +511,7 @@ static inline int test_and_change_bit(unsigned long nr,
 		__asm__ __volatile__(
 		"	.set	push					\n"
 		"	.set	noreorder				\n"
+		OCTEON_SYNCW_STR
 		"	.set	mips3					\n"
 		"1:	" __LL	"%0, %1		# test_and_change_bit	\n"
 		"	xor	%2, %0, %3				\n"
@@ -542,7 +545,6 @@ static inline int test_and_change_bit(unsigned long nr,
 }
 
 #include <asm-generic/bitops/non-atomic.h>
-
 /*
  * __clear_bit_unlock - Clears a bit in memory
  * @nr: Bit to clear
@@ -619,7 +621,7 @@ static inline int fls(int word)
 	return 32 - word;
 }
 
-#if defined(CONFIG_64BIT) && defined(CONFIG_CPU_MIPS64)
+#if defined(CONFIG_64BIT) && (defined(CONFIG_CPU_MIPS64) || defined(CONFIG_CPU_CAVIUM_OCTEON))
 static inline int fls64(__u64 word)
 {
 	__asm__("dclz %0, %1" : "=r" (word) : "r" (word));
-- 
1.5.5.1

