From f9ac6703bdb02eb96ab8df94470b46f921da8f8e Mon Sep 17 00:00:00 2001
From: auto commit <unknown@unknown>
Date: Fri, 24 Oct 2008 12:22:45 -0700
Subject: [PATCH] cavium: udelay optimization for 64bit Cavium CPU.

Use Octeon-specific 64-bit cycle counter to avoid overflow in udelay().

Signed-off-by: Tomaso Paoletti <tpaoletti@caviumnetworks.com>
Signed-off-by: Paul Gortmaker <Paul.Gortmaker@windriver.com>
---
 include/asm-mips/delay.h |   35 +++++++++++++++++++++++++++++++++++
 1 files changed, 35 insertions(+), 0 deletions(-)

diff --git a/include/asm-mips/delay.h b/include/asm-mips/delay.h
index b0bccd2..d0742cc 100644
--- a/include/asm-mips/delay.h
+++ b/include/asm-mips/delay.h
@@ -16,6 +16,7 @@
 
 #include <asm/compiler.h>
 #include <asm/war.h>
+#include <asm/time.h>
 
 static inline void __delay(unsigned long loops)
 {
@@ -48,6 +49,38 @@ static inline void __delay(unsigned long loops)
 		: "0" (loops), "r" (1));
 }
 
+#ifdef CONFIG_CPU_CAVIUM_OCTEON
+
+/*
+ * Replacement for what was being done in get_cycles()
+ * which is being phased out.  If the custom udelay()
+ * goes away, then the octeon_get_cycles should just
+ * revert to using read_c0_count()
+ */
+static inline unsigned long octeon_get_cycles(void)
+{
+	unsigned long result;
+	asm volatile ("rdhwr %0,$31\n"
+#ifndef CONFIG_64BIT
+			"sll %0,0\n"
+#endif
+			: "=r" (result));
+	return result;
+}
+
+/* Implementation of udelay, which unlike the default will actually delay
+    the correct amount of time. Among the other problems with the default
+    spin loop implementation, it overflows at 1ms */
+static inline void udelay(unsigned long usecs)
+{
+	unsigned long cycles = octeon_get_cycles();
+	unsigned long end_cycles = cycles + usecs * (mips_hpt_frequency/1000000);
+	do {
+		cycles = octeon_get_cycles();
+	} while (cycles < end_cycles);
+}
+
+#else
 
 /*
  * Division by multiplication: you don't have to worry about
@@ -109,4 +142,6 @@ static inline void __udelay(unsigned long usecs, unsigned long lpj)
 #define MAX_UDELAY_MS	(1000 / HZ)
 #endif
 
+#endif /* CONFIG_CPU_CAVIUM_OCTEON */
+
 #endif /* _ASM_DELAY_H */
-- 
1.5.5.1

