From f6bb8bffa94fe17cc18269406959c85358775932 Mon Sep 17 00:00:00 2001
From: auto commit <unknown@unknown>
Date: Mon, 27 Oct 2008 15:57:28 -0400
Subject: [PATCH] [PATCH] 0073-cavium-boot-CPU-may-not-be-proc-0.patch

From 8e7af4580ffe6763f32021389c28092027e4ef9a Mon Sep 17 00:00:00 2001
Subject: [PATCH] 0073-cavium-boot-CPU-may-not-be-proc-0.patch
---
 arch/mips/kernel/smp.c |    2 ++
 1 files changed, 2 insertions(+), 0 deletions(-)

diff --git a/arch/mips/kernel/smp.c b/arch/mips/kernel/smp.c
index 4410f17..b214c89 100644
--- a/arch/mips/kernel/smp.c
+++ b/arch/mips/kernel/smp.c
@@ -191,12 +191,14 @@ void __init smp_prepare_cpus(unsigned int max_cpus)
 /* preload SMP state for boot cpu */
 void __devinit smp_prepare_boot_cpu(void)
 {
+#ifndef CONFIG_CPU_CAVIUM_OCTEON
 	/*
 	 * This assumes that bootup is always handled by the processor
 	 * with the logic and physical number 0.
 	 */
 	__cpu_number_map[0] = 0;
 	__cpu_logical_map[0] = 0;
+#endif
 	cpu_set(0, phys_cpu_present_map);
 	cpu_set(0, cpu_online_map);
 	cpu_set(0, cpu_callin_map);
-- 
1.5.5.1

