From ec543b7f214ab2b5b4913d4565eb602032dd3b1c Mon Sep 17 00:00:00 2001
From: auto commit <unknown@unknown>
Date: Fri, 24 Oct 2008 12:23:03 -0700
Subject: [PATCH] sg: importing 1000-cavium-disable-section-per-function.patch

This is an automatic import of patch 1000-cavium-disable-section-per-function.patch, no headers were
detected and a default message was constructed
---
 arch/mips/Makefile |    3 ++-
 1 files changed, 2 insertions(+), 1 deletions(-)

diff --git a/arch/mips/Makefile b/arch/mips/Makefile
index 08bd1e2..102b29a 100644
--- a/arch/mips/Makefile
+++ b/arch/mips/Makefile
@@ -15,7 +15,8 @@
 KBUILD_DEFCONFIG := ip22_defconfig
 
 ifndef CONFIG_FTRACE
-cflags-y := -ffunction-sections
+#cflags-y := -ffunction-sections
+cflags-y :=
 endif
 
 #
-- 
1.5.5.1

