From fb4d2da2d50286547f4d9541760a6177efbd5077 Mon Sep 17 00:00:00 2001
From: auto commit <unknown@unknown>
Date: Fri, 24 Oct 2008 12:23:04 -0700
Subject: [PATCH] to_msg: importing 0001-cavium-cf-change-to-IRQF_SHARED.patch

This is an automatic import of patch 0001-cavium-cf-change-to-IRQF_SHARED.patch, no headers were
detected and a default message was constructed
---
 arch/mips/cavium-octeon/ebt3000_cf.c |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/arch/mips/cavium-octeon/ebt3000_cf.c b/arch/mips/cavium-octeon/ebt3000_cf.c
index 667610c..9f1e9e3 100644
--- a/arch/mips/cavium-octeon/ebt3000_cf.c
+++ b/arch/mips/cavium-octeon/ebt3000_cf.c
@@ -671,7 +671,7 @@ static int ebt3000cf_identify(struct cf_device *cf)
 				/* Request the interrupt */
 				cvmx_mio_boot_dma_int_enx_t dma_int_en;
 				request_irq(OCTEON_IRQ_BOOTDMA, interruptDma0Done,
-					    SA_SHIRQ, "MIO_DMA", cf);
+					    IRQF_SHARED, "MIO_DMA", cf);
     
 				/* Enable interrupts for DMA completion */
 				dma_int_en.u64 = cvmx_read_csr(CVMX_MIO_BOOT_DMA_INT_ENX(cf->dma_channel));
-- 
1.5.5.1

