From ebbeab8f73fd341c38e07387d7bf0b21dbb91577 Mon Sep 17 00:00:00 2001
From: auto commit <unknown@unknown>
Date: Fri, 24 Oct 2008 12:23:06 -0700
Subject: [PATCH] o_msg: importing 0001-cavium-PCI-change-to-IRQF_SHARED.patch

This is an automatic import of patch 0001-cavium-PCI-change-to-IRQF_SHARED.patch, no headers were
detected and a default message was constructed
---
 arch/mips/cavium-octeon/msi.c |   11 ++++++-----
 1 files changed, 6 insertions(+), 5 deletions(-)

diff --git a/arch/mips/cavium-octeon/msi.c b/arch/mips/cavium-octeon/msi.c
index 3b9ab3f..dcc3e6f 100644
--- a/arch/mips/cavium-octeon/msi.c
+++ b/arch/mips/cavium-octeon/msi.c
@@ -233,17 +233,18 @@ static irqreturn_t octeon_msi_interrupt(int cpl, void *dev_id)
  */
 int octeon_msi_initialize(void)
 {
+	int r;
 	if (octeon_has_feature(OCTEON_FEATURE_PCIE)) {
-		request_irq(OCTEON_IRQ_PCI_MSI0, octeon_msi_interrupt, SA_SHIRQ,
+		r = request_irq(OCTEON_IRQ_PCI_MSI0, octeon_msi_interrupt, IRQF_SHARED,
 			    "MSI[0:63]", octeon_msi_interrupt);
 	} else if (octeon_is_pci_host()) {
-		request_irq(OCTEON_IRQ_PCI_MSI0, octeon_msi_interrupt, SA_SHIRQ,
+		r = request_irq(OCTEON_IRQ_PCI_MSI0, octeon_msi_interrupt, IRQF_SHARED,
 			    "MSI[0:15]", octeon_msi_interrupt);
-		request_irq(OCTEON_IRQ_PCI_MSI1, octeon_msi_interrupt, SA_SHIRQ,
+		r += request_irq(OCTEON_IRQ_PCI_MSI1, octeon_msi_interrupt, IRQF_SHARED,
 			    "MSI[16:31]", octeon_msi_interrupt);
-		request_irq(OCTEON_IRQ_PCI_MSI2, octeon_msi_interrupt, SA_SHIRQ,
+		r += request_irq(OCTEON_IRQ_PCI_MSI2, octeon_msi_interrupt, IRQF_SHARED,
 			    "MSI[32:47]", octeon_msi_interrupt);
-		request_irq(OCTEON_IRQ_PCI_MSI3, octeon_msi_interrupt, SA_SHIRQ,
+		r += request_irq(OCTEON_IRQ_PCI_MSI3, octeon_msi_interrupt, IRQF_SHARED,
 			    "MSI[48:63]", octeon_msi_interrupt);
 	}
 	return 0;
-- 
1.5.5.1

