From 2095e933ac83979ed41d25c97e91cbc0d1b99c47 Mon Sep 17 00:00:00 2001
From: auto commit <unknown@unknown>
Date: Fri, 24 Oct 2008 12:23:06 -0700
Subject: [PATCH] auto_msg: importing 0001-cavium-hook-pci-console.patch

This is an automatic import of patch 0001-cavium-hook-pci-console.patch, no headers were
detected and a default message was constructed
---
 arch/mips/cavium-octeon/Makefile |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/arch/mips/cavium-octeon/Makefile b/arch/mips/cavium-octeon/Makefile
index 5128f61..e14b0df 100644
--- a/arch/mips/cavium-octeon/Makefile
+++ b/arch/mips/cavium-octeon/Makefile
@@ -20,7 +20,7 @@ obj-y += console.o
 
 obj-$(CONFIG_SMP)                     += smp.o
 obj-$(CONFIG_CAVIUM_OCTEON_BOOTBUS_COMPACT_FLASH) += ebt3000_cf.o
-obj-$(CONFIG_PCI)                     += pci-common.o
+obj-$(CONFIG_PCI)                     += pci-common.o pci_console.o
 obj-$(CONFIG_PCI)                     += pci.o pci_chips.o i8259.o
 obj-$(CONFIG_PCI_MSI)                 += msi.o
 obj-$(CONFIG_PCI)                     += pcie.o
-- 
1.5.5.1

