From 0ae39804cb941c52a3b9601442ea84f934472c1b Mon Sep 17 00:00:00 2001
From: auto commit <unknown@unknown>
Date: Fri, 24 Oct 2008 12:23:07 -0700
Subject: [PATCH] auto_msg: importing 0001-1.8.0-cavium-cleanup-watchdog.patch

This is an automatic import of patch 0001-1.8.0-cavium-cleanup-watchdog.patch, no headers were
detected and a default message was constructed
---
 arch/mips/cavium-octeon/watchdog.c |   13 +++++++------
 1 files changed, 7 insertions(+), 6 deletions(-)

diff --git a/arch/mips/cavium-octeon/watchdog.c b/arch/mips/cavium-octeon/watchdog.c
index cef1199..1a72bfb 100644
--- a/arch/mips/cavium-octeon/watchdog.c
+++ b/arch/mips/cavium-octeon/watchdog.c
@@ -16,8 +16,9 @@
 #include <linux/delay.h>
 #include "hal.h"
 
-int timeout = 0;		/* A zero value is replaced with the max
-				   timeout */
+extern void octeon_watchdog_nmi(void);
+
+int timeout;		/* A zero value is replaced with the max timeout */
 module_param(timeout, int, 0444);
 MODULE_PARM_DESC(timeout,
 		 " Watchdog timeout in milliseconds. First timeout causes a poke, \n"
@@ -35,7 +36,9 @@ MODULE_PARM_DESC(timeout,
 static irqreturn_t watchdog_poke_irq(int cpl, void *dev_id)
 {
 	/* We're alive, poke the watchdog */
-	// printk("Poking watchdog on core %d\n", cvmx_get_core_num());
+	/*
+	printk("Poking watchdog on core %d\n", cvmx_get_core_num());
+	*/
 	cvmx_write_csr(CVMX_CIU_PP_POKEX(cvmx_get_core_num()), 1);
 	return IRQ_HANDLED;
 }
@@ -56,8 +59,7 @@ static inline void uart_write_byte(int uart_index, uint8_t ch)
 	/* Spin until there is room */
 	do {
 		lsrval.u64 = cvmx_read_csr(CVMX_MIO_UARTX_LSR(uart_index));
-	}
-	while (lsrval.s.thre == 0);
+	} while (lsrval.s.thre == 0);
 
 	/* Write the byte */
 	cvmx_write_csr(CVMX_MIO_UARTX_THR(uart_index), ch);
@@ -218,7 +220,6 @@ static int __init watchdog_init(void)
 {
 	uint64_t threshold;
 	int i;
-	extern void octeon_watchdog_nmi(void);
 
 	/* Watchdog time expiration length = The 16 bits of LEN represent the
 	   most significant bits of a 24 bit decrementer that decrements every
-- 
1.5.5.1

