From 9f785fd806628a701e4c0ac0443e26a598ae0498 Mon Sep 17 00:00:00 2001
From: auto commit <unknown@unknown>
Date: Fri, 24 Oct 2008 12:23:07 -0700
Subject: [PATCH] importing 0001-cavium-watchdog-change-to-IRQF_SHARED.patch

This is an automatic import of patch 0001-cavium-watchdog-change-to-IRQF_SHARED.patch, no headers were
detected and a default message was constructed
---
 arch/mips/cavium-octeon/watchdog.c |    5 +++--
 1 files changed, 3 insertions(+), 2 deletions(-)

diff --git a/arch/mips/cavium-octeon/watchdog.c b/arch/mips/cavium-octeon/watchdog.c
index 1a72bfb..df28eba 100644
--- a/arch/mips/cavium-octeon/watchdog.c
+++ b/arch/mips/cavium-octeon/watchdog.c
@@ -198,8 +198,9 @@ static void setup_interrupt(void *arg)
 	int core = cvmx_get_core_num();
 	unsigned long threshold = (unsigned long) arg;
 
-	request_irq(OCTEON_IRQ_WDOG0 + core, watchdog_poke_irq, SA_SHIRQ,
-		    "watchdog", watchdog_poke_irq);
+	if (request_irq(OCTEON_IRQ_WDOG0 + core, watchdog_poke_irq, IRQF_SHARED,
+		    "watchdog", watchdog_poke_irq)) {
+	}
 
 	/* Poke the watchdog to clear out its state */
 	cvmx_write_csr(CVMX_CIU_PP_POKEX(core), 1);
-- 
1.5.5.1

