From 41dc81acfe996fc56a38be814460cf92431327e7 Mon Sep 17 00:00:00 2001
From: auto commit <unknown@unknown>
Date: Fri, 24 Oct 2008 12:23:07 -0700
Subject: [PATCH] auto_msg: importing 0001-cavium-watchdog-upgrade.patch

This is an automatic import of patch 0001-cavium-watchdog-upgrade.patch, no headers were
detected and a default message was constructed
---
 arch/mips/cavium-octeon/watchdog.c |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/arch/mips/cavium-octeon/watchdog.c b/arch/mips/cavium-octeon/watchdog.c
index df28eba..5331f70 100644
--- a/arch/mips/cavium-octeon/watchdog.c
+++ b/arch/mips/cavium-octeon/watchdog.c
@@ -245,7 +245,7 @@ static int __init watchdog_init(void)
 	}
 	cvmx_write_csr(CVMX_MIO_BOOT_LOC_CFGX(0), 0x81fc0000);
 
-	on_each_cpu(setup_interrupt, (void *) threshold, 1, 0);
+	on_each_cpu(setup_interrupt, (void *) threshold, 0);
 	printk("Octeon watchdog driver loaded with a timeout of %d ms.\n",
 	       timeout);
 	return 0;
-- 
1.5.5.1

