From c7a2ef5518603748d640bbacfb913979e502f13e Mon Sep 17 00:00:00 2001
From: auto commit <unknown@unknown>
Date: Fri, 24 Oct 2008 12:23:07 -0700
Subject: [PATCH] auto_msg: importing 0001-cavium-hook-watchdog.patch

This is an automatic import of patch 0001-cavium-hook-watchdog.patch, no headers were
detected and a default message was constructed
---
 arch/mips/cavium-octeon/Makefile |    1 +
 1 files changed, 1 insertions(+), 0 deletions(-)

diff --git a/arch/mips/cavium-octeon/Makefile b/arch/mips/cavium-octeon/Makefile
index e14b0df..8c29038 100644
--- a/arch/mips/cavium-octeon/Makefile
+++ b/arch/mips/cavium-octeon/Makefile
@@ -24,6 +24,7 @@ obj-$(CONFIG_PCI)                     += pci-common.o pci_console.o
 obj-$(CONFIG_PCI)                     += pci.o pci_chips.o i8259.o
 obj-$(CONFIG_PCI_MSI)                 += msi.o
 obj-$(CONFIG_PCI)                     += pcie.o
+obj-$(CONFIG_CAVIUM_OCTEON_WATCHDOG)  += watchdog.o watchdog_nmi.o
 
 clean:
 	rm -f *.o
-- 
1.5.5.1

