From 3a1434074e9d7ccc7bae72e9d42969434aea2fcc Mon Sep 17 00:00:00 2001
From: auto commit <unknown@unknown>
Date: Fri, 24 Oct 2008 12:23:08 -0700
Subject: [PATCH] porting 0001-cavium-trace-buffer-change-to-IRQF_SHARED.patch

This is an automatic import of patch 0001-cavium-trace-buffer-change-to-IRQF_SHARED.patch, no headers were
detected and a default message was constructed
---
 arch/mips/cavium-octeon/octeon-tra.c |    5 +++--
 1 files changed, 3 insertions(+), 2 deletions(-)

diff --git a/arch/mips/cavium-octeon/octeon-tra.c b/arch/mips/cavium-octeon/octeon-tra.c
index 7ae14cd..c95fb4b 100644
--- a/arch/mips/cavium-octeon/octeon-tra.c
+++ b/arch/mips/cavium-octeon/octeon-tra.c
@@ -152,8 +152,9 @@ static int __init octeon_tra_init(void)
 	cvmx_tra_trig_setup(1, filter, source_filter, trig0_did, address, address_mask);
 
 	/* Hook up to the trace buffer interrupt so we know when a trigger happens */
-	request_irq(OCTEON_IRQ_TRACE, octeon_tra_interrupt, SA_SHIRQ,
-		"Trace buffer", octeon_tra_interrupt);
+	if (request_irq(OCTEON_IRQ_TRACE, octeon_tra_interrupt, SA_SHIRQ,
+			"Trace buffer", octeon_tra_interrupt))
+		pr_err("Octeon TRA: failed to reserve TRACE IRQ\n");
 
 	/* All setup is complete. Enable the trace buffer */
 	cvmx_tra_enable(1);
-- 
1.5.5.1

