From 8d30cba68c582498f7b3519536c7411be61e302e Mon Sep 17 00:00:00 2001
From: auto commit <unknown@unknown>
Date: Fri, 24 Oct 2008 12:23:08 -0700
Subject: [PATCH] o_msg: importing 0001-cavium-trace-buffer-improvements.patch

This is an automatic import of patch 0001-cavium-trace-buffer-improvements.patch, no headers were
detected and a default message was constructed
---
 arch/mips/cavium-octeon/octeon-tra.c |    8 ++++----
 1 files changed, 4 insertions(+), 4 deletions(-)

diff --git a/arch/mips/cavium-octeon/octeon-tra.c b/arch/mips/cavium-octeon/octeon-tra.c
index c95fb4b..7325069 100644
--- a/arch/mips/cavium-octeon/octeon-tra.c
+++ b/arch/mips/cavium-octeon/octeon-tra.c
@@ -58,13 +58,13 @@ static irqreturn_t octeon_tra_interrupt(int cpl, void *dev_id)
 
 	/* We can optionally stop the other cores */
 	if (OCTEON_TRA_DUMP_CORES_ON_INTERRUPT) {
-		printk("Octeon Trace Buffer Dumping Core state\n");
+		pr_info("Octeon Trace Buffer Dumping Core state\n");
 		on_each_cpu(octeon_tra_dump_regs, NULL, 0, 1);
 	}
 
-	printk("Octeon Trace Buffer Start\n");
+	pr_info("Octeon Trace Buffer Start\n");
 	cvmx_tra_display();
-	printk("Octeon Trace Buffer End\n");
+	pr_info("Octeon Trace Buffer End\n");
 
 	/* Restart the trace buffer */
 	cvmx_tra_enable(1);
@@ -158,7 +158,7 @@ static int __init octeon_tra_init(void)
 
 	/* All setup is complete. Enable the trace buffer */
 	cvmx_tra_enable(1);
-	printk("Octeon TRA driver loaded.\n");
+	pr_notice("Octeon TRA driver loaded.\n");
 	return 0;
 }
 
-- 
1.5.5.1

