From 588deeed9cda8b896c9bc50e29ddceaedcb6d5cb Mon Sep 17 00:00:00 2001
From: auto commit <unknown@unknown>
Date: Fri, 24 Oct 2008 12:23:08 -0700
Subject: [PATCH] auto_msg: importing 0152-cavium-update-ethernet-mgmt.patch

This is an automatic import of patch 0152-cavium-update-ethernet-mgmt.patch, no headers were
detected and a default message was constructed
---
 arch/mips/cavium-octeon/ethernet-mgmt-port.c |    5 +++--
 1 files changed, 3 insertions(+), 2 deletions(-)

diff --git a/arch/mips/cavium-octeon/ethernet-mgmt-port.c b/arch/mips/cavium-octeon/ethernet-mgmt-port.c
index 752f89d..f6ad76f 100644
--- a/arch/mips/cavium-octeon/ethernet-mgmt-port.c
+++ b/arch/mips/cavium-octeon/ethernet-mgmt-port.c
@@ -326,8 +326,9 @@ static int __init ethernet_mgmt_port_init(void)
 		/* Register an IRQ hander for to receive interrupts */
 		dev->irq =
 			(priv->port == 0) ? OCTEON_IRQ_MII0 : OCTEON_IRQ_MII1;
-		request_irq(dev->irq, do_interrupt, SA_SHIRQ, dev->name,
-			    dev);
+		if (request_irq(dev->irq, do_interrupt, IRQF_SHARED, dev->name,
+			    dev))
+			pr_err("ethernet-mgmt: failed to assign interrupt %d\n", dev->irq);
 
 		/* Interrupt every single RX packet */
 		mix_irhwm.u64 = 0;
-- 
1.5.5.1

