From 35abf8a6e94598dde806c67ec1f40772c7860bcb Mon Sep 17 00:00:00 2001
From: auto commit <unknown@unknown>
Date: Fri, 24 Oct 2008 12:23:09 -0700
Subject: [PATCH] auto_msg: importing 0160-cavium-link-ethernet-mgmt.patch

This is an automatic import of patch 0160-cavium-link-ethernet-mgmt.patch, no headers were
detected and a default message was constructed
---
 arch/mips/cavium-octeon/Makefile |    1 +
 1 files changed, 1 insertions(+), 0 deletions(-)

diff --git a/arch/mips/cavium-octeon/Makefile b/arch/mips/cavium-octeon/Makefile
index c15c23f..7bca072 100644
--- a/arch/mips/cavium-octeon/Makefile
+++ b/arch/mips/cavium-octeon/Makefile
@@ -26,6 +26,7 @@ obj-$(CONFIG_PCI_MSI)                 += msi.o
 obj-$(CONFIG_PCI)                     += pcie.o
 obj-$(CONFIG_CAVIUM_OCTEON_WATCHDOG)  += watchdog.o watchdog_nmi.o
 obj-$(CONFIG_CAVIUM_OCTEON_TRA)       += octeon-tra.o
+obj-$(CONFIG_CAVIUM_OCTEON_MGMT_PORT_ETHERNET) += ethernet-mgmt-port.o
 
 clean:
 	rm -f *.o
-- 
1.5.5.1

