From b21f9ade2a4a0eb458c64b982fff533969ccc689 Mon Sep 17 00:00:00 2001
From: auto commit <unknown@unknown>
Date: Fri, 24 Oct 2008 12:23:10 -0700
Subject: [PATCH] auto_msg: importing 0153-cavium-update-ethernet-pow.patch

This is an automatic import of patch 0153-cavium-update-ethernet-pow.patch, no headers were
detected and a default message was constructed
---
 arch/mips/cavium-octeon/ethernet-pow.c |    5 +++--
 1 files changed, 3 insertions(+), 2 deletions(-)

diff --git a/arch/mips/cavium-octeon/ethernet-pow.c b/arch/mips/cavium-octeon/ethernet-pow.c
index 5a5bccb..c9bfb80 100644
--- a/arch/mips/cavium-octeon/ethernet-pow.c
+++ b/arch/mips/cavium-octeon/ethernet-pow.c
@@ -696,8 +696,9 @@ static int __init ethernet_pow_init(void)
 	}
 
 	/* Register an IRQ hander for to receive POW interrupts */
-	request_irq(OCTEON_IRQ_WORKQ0 + receive_group, do_interrupt, SA_SHIRQ,
-		    "POW Ethernet", global_device);
+	if (request_irq(OCTEON_IRQ_WORKQ0 + receive_group, do_interrupt, IRQF_SHARED,
+			"POW Ethernet", global_device)) {
+	}
 
 	/* Enable POW interrupt when our port has at least one packet */
 	cvmx_write_csr(OCTEON_POW_WQ_INT_THRX(receive_group), 0x1001);
-- 
1.5.5.1

