From d438bf4583a5e7e11283f81dcc0800bdc23338d8 Mon Sep 17 00:00:00 2001
From: auto commit <unknown@unknown>
Date: Fri, 24 Oct 2008 12:23:10 -0700
Subject: [PATCH] sg: importing 0161-cavium-link-ethernet-pow-into-build.patch

This is an automatic import of patch 0161-cavium-link-ethernet-pow-into-build.patch, no headers were
detected and a default message was constructed
---
 arch/mips/cavium-octeon/Makefile |    1 +
 1 files changed, 1 insertions(+), 0 deletions(-)

diff --git a/arch/mips/cavium-octeon/Makefile b/arch/mips/cavium-octeon/Makefile
index 7bca072..c250a87 100644
--- a/arch/mips/cavium-octeon/Makefile
+++ b/arch/mips/cavium-octeon/Makefile
@@ -27,6 +27,7 @@ obj-$(CONFIG_PCI)                     += pcie.o
 obj-$(CONFIG_CAVIUM_OCTEON_WATCHDOG)  += watchdog.o watchdog_nmi.o
 obj-$(CONFIG_CAVIUM_OCTEON_TRA)       += octeon-tra.o
 obj-$(CONFIG_CAVIUM_OCTEON_MGMT_PORT_ETHERNET) += ethernet-mgmt-port.o
+obj-$(CONFIG_CAVIUM_OCTEON_POW_ONLY_ETHERNET) += ethernet-pow.o
 
 clean:
 	rm -f *.o
-- 
1.5.5.1

