From 9ce85dbe87acc9c509a451cf183a035ab4cb1d06 Mon Sep 17 00:00:00 2001
From: auto commit <unknown@unknown>
Date: Fri, 24 Oct 2008 12:23:11 -0700
Subject: [PATCH] auto_msg: importing 0138-cavium-fix-printk_ratelimit-2.patch

This is an automatic import of patch 0138-cavium-fix-printk_ratelimit-2.patch, no headers were
detected and a default message was constructed
---
 drivers/net/cavium-ethernet/ethernet-util.h |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/drivers/net/cavium-ethernet/ethernet-util.h b/drivers/net/cavium-ethernet/ethernet-util.h
index 2073d16..a63da44 100644
--- a/drivers/net/cavium-ethernet/ethernet-util.h
+++ b/drivers/net/cavium-ethernet/ethernet-util.h
@@ -43,7 +43,7 @@
 * OF USE OR PERFORMANCE OF THE SOFTWARE LIES WITH YOU.
 *************************************************************************/
 
-#define DEBUGPRINT(format, ...) do{if (__printk_ratelimit(HZ, 10)) printk(format, ##__VA_ARGS__);} while (0)
+#define DEBUGPRINT(format, ...) do{if (printk_ratelimit()) printk(format, ##__VA_ARGS__);} while (0)
 
 /**
  * Given a packet data address, return a pointer to the
-- 
1.5.5.1

