From 529a21d79d8bc1ce5c7504cba5a356b6958d190a Mon Sep 17 00:00:00 2001
From: auto commit <unknown@unknown>
Date: Fri, 24 Oct 2008 12:23:13 -0700
Subject: [PATCH] importing 0013-cavium-update-use-of-smp_call_function.patch

This is an automatic import of patch 0013-cavium-update-use-of-smp_call_function.patch, no headers were
detected and a default message was constructed
---
 arch/mips/oprofile/op_model_cavium_octeon.c |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/arch/mips/oprofile/op_model_cavium_octeon.c b/arch/mips/oprofile/op_model_cavium_octeon.c
index 5ab89f0..d96cea0 100644
--- a/arch/mips/oprofile/op_model_cavium_octeon.c
+++ b/arch/mips/oprofile/op_model_cavium_octeon.c
@@ -120,7 +120,7 @@ static int octeon_init(void)
 	if (result == 0) {
 		irq_desc_t *desc = irq_desc + 6;
 		smp_call_function((void (*)(void *)) desc->chip->enable,
-				  (void *) (long) 6, 0, 1);
+				  (void *) (long) 6, 1);
 	}
 #endif
 	return result;
-- 
1.5.5.1

