From f3ec9803e389b272699cf4091c91b0dd0ac95104 Mon Sep 17 00:00:00 2001
From: auto commit <unknown@unknown>
Date: Fri, 24 Oct 2008 12:23:15 -0700
Subject: [PATCH] o_msg: importing 0169-cavium-hook-temporary-scratchpad.patch

This is an automatic import of patch 0169-cavium-hook-temporary-scratchpad.patch, no headers were
detected and a default message was constructed
---
 arch/mips/cavium-octeon/Kconfig |   29 +++++++++++++++++++++++++++++
 1 files changed, 29 insertions(+), 0 deletions(-)

diff --git a/arch/mips/cavium-octeon/Kconfig b/arch/mips/cavium-octeon/Kconfig
index a6997a7..52adb04 100644
--- a/arch/mips/cavium-octeon/Kconfig
+++ b/arch/mips/cavium-octeon/Kconfig
@@ -98,6 +98,35 @@ config REPLACE_EMULATED_ACCESS_TO_THREAD_POINTER
 		cause programs to fail, but makes it easy to stop gdb on the
 		instruction.
 
+config TEMPORARY_SCRATCHPAD_FOR_KERNEL
+	int "Number of 64-bit words in CVMSEG LM reserved for temporary use by the kernel"
+	depends on CAVIUM_OCTEON_SPECIFIC_OPTIONS
+	range 0 2
+	default 1
+	help
+	  For Mips, performance-critical kernel routines (like the TLB
+	  miss handlers) can normally only use registers K0 and K1 ($26
+	  and $27) from the main register file. This option allocates
+	  space in CVMSEG LM for the same function. This can make the
+	  kernel routines run faster. A side effect of
+	  this is that the kernel will trash the CVMSEG LM locations,
+	  which are placed as high as possible in CVMSEG LM space, but
+	  below the space allocated for the FAST_ACCESS_TO_THREAD_POINTER
+	  option. Like FAST_ACCESS_TO_THREAD_POINTER, these CVMSEG
+          locations are not available to user applications.
+
+config CAVIUM_OCTEON_CVMSEG_SIZE
+	int "Number of L1 cache lines reserved for CVMSEG memory"
+	depends on CAVIUM_OCTEON_SPECIFIC_OPTIONS
+	range 0 54
+	default 1
+	help
+          CVMSEG LM is a segment that accesses portions of the dcache as a
+	  local memory; the larger CVMSEG is, the smaller the cache is.
+	  This selects the size of CVMSEG LM, which is in cache blocks. The
+	  legally range is from zero to 54 cache blocks (i.e. CVMSEG LM is
+	  between zero and 6192 bytes).
+
 config CAVIUM_OCTEON_LOCK_L2
 	bool "Lock often used kernel code in the L2"
 	depends on CAVIUM_OCTEON_SPECIFIC_OPTIONS
-- 
1.5.5.1

