From db384413e01a195d001d0aa9055db02f289aa37c Mon Sep 17 00:00:00 2001
From: auto commit <unknown@unknown>
Date: Fri, 24 Oct 2008 12:23:16 -0700
Subject: [PATCH] cavium: make FPU EMU stats a nop

Collecting these stats generates cross CPU traffic for nothing
as they aren't exported anywhere.

Signed-off-by: Tomaso Paoletti <tpaoletti@caviumnetworks.com>
Signed-off-by: Paul Gortmaker <Paul.Gortmaker@windriver.com>
---
 include/asm-mips/fpu_emulator.h |    5 +++++
 1 files changed, 5 insertions(+), 0 deletions(-)

diff --git a/include/asm-mips/fpu_emulator.h b/include/asm-mips/fpu_emulator.h
index 2731c38..4917cbd 100644
--- a/include/asm-mips/fpu_emulator.h
+++ b/include/asm-mips/fpu_emulator.h
@@ -33,5 +33,10 @@ struct mips_fpu_emulator_stats {
 };
 
 extern struct mips_fpu_emulator_stats fpuemustats;
+#ifdef CONFIG_MIPS_FPU_EMULATOR_STATS_ENABLED
+#define FPUEMUSTATS_INC(field) fpuemustats.field++
+#else
+#define FPUEMUSTATS_INC(field) do {} while (0)
+#endif
 
 #endif /* _ASM_FPU_EMULATOR_H */
-- 
1.5.5.1

