From 1b7518f974a65b089f43a5a02fa7ca371a5d2435 Mon Sep 17 00:00:00 2001
From: auto commit <unknown@unknown>
Date: Fri, 24 Oct 2008 12:23:17 -0700
Subject: [PATCH] cavium: add temporary scratchpad for kernel

Signed-off-by: Tomaso Paoletti <tpaoletti@caviumnetworks.com>
---
 include/asm-mips/mipsregs.h |   19 ++++++++++++++++++-
 1 files changed, 18 insertions(+), 1 deletions(-)

diff --git a/include/asm-mips/mipsregs.h b/include/asm-mips/mipsregs.h
index 7fc936a..86a2b20 100644
--- a/include/asm-mips/mipsregs.h
+++ b/include/asm-mips/mipsregs.h
@@ -582,9 +582,26 @@
  * from userspace. Octeon uses a 64bit location in CVMSEG to store the thread pointer
  * for quick access.
  */
+#define TOP_OF_CVMSEG_STORE        (CONFIG_CAVIUM_OCTEON_CVMSEG_SIZE*128-32768)
 #ifdef CONFIG_FAST_ACCESS_TO_THREAD_POINTER
-#define FAST_ACCESS_THREAD_OFFSET       CONFIG_CAVIUM_OCTEON_CVMSEG_SIZE*128-8-32768
+#define FAST_ACCESS_THREAD_OFFSET       (TOP_OF_CVMSEG_STORE-8)
 #define FAST_ACCESS_THREAD_REGISTER     *(unsigned long *)(FAST_ACCESS_THREAD_OFFSET)
+#else
+#define FAST_ACCESS_THREAD_OFFSET       TOP_OF_CVMSEG_STORE
+#endif
+
+/*
+ * These defines are used on Octeon to give the kernel access to CVMSEG locations to
+ * improve performance. The kernel may use this space much like it normally uses
+ * K0 and K1 in the main register file - as a temporary store.
+ *
+ * Put it as high as possible in CVMSEG space, but below THREAD_POINTER
+ */
+#ifdef CONFIG_TEMPORARY_SCRATCHPAD_FOR_KERNEL
+#define TEMPORARY_SCRATCHPAD_FOR_KERNEL_OFFSET(i)    (FAST_ACCESS_THREAD_OFFSET-(8*(CONFIG_TEMPORARY_SCRATCHPAD_FOR_KERNEL-(i))))
+#define TEMPORARY_SCRATCHPAD_FOR_KERNEL_REGISTER(i)  *(unsigned long)(TEMPORARY_SCRATCHPAD_FOR_KERNEL_OFFSET(i))
+#else
+#define TEMPORARY_SCRATCHPAD_FOR_KERNEL_OFFSET(i)    FAST_ACCESS_THREAD_OFFSET
 #endif
 
 #ifndef __ASSEMBLY__
-- 
1.5.5.1

