From 72d3333b97e59001ab1184e7581657d1fee2d5d1 Mon Sep 17 00:00:00 2001
From: Greg Moffatt <greg.moffatt@windriver.com>
Date: Mon, 6 Jul 2009 08:12:20 -0400
Subject: [PATCH] cavium/ethernet: Fix driver to support IPv6 PING

Change the cavium ethernet driver so that it correctly calculates the
skb->tail to be either an absolute address or offset-based depending on
how the kernel is configured.  Currently it is only absolute based and
this causes problems with utilities that use the skb->tail address
for calculations.  Specifically, this was seen in testing as IPv6 pings
(via ICMP6) were failing due to neighbour association requests.
---
 drivers/net/cavium-ethernet/ethernet-rx.c |    4 ++++
 1 files changed, 4 insertions(+), 0 deletions(-)

diff --git a/drivers/net/cavium-ethernet/ethernet-rx.c b/drivers/net/cavium-ethernet/ethernet-rx.c
index 8ce7ed0..f31d409 100644
--- a/drivers/net/cavium-ethernet/ethernet-rx.c
+++ b/drivers/net/cavium-ethernet/ethernet-rx.c
@@ -261,7 +261,11 @@ void cvm_oct_tasklet_rx(unsigned long unused)
 			skb->data = skb->head + work->packet_ptr.s.addr - cvmx_ptr_to_phys(skb->head);
 			CVMX_PREFETCH(skb->data, 0);
 			skb->len = work->len;
+#ifdef NET_SKBUFF_DATA_USES_OFFSET 
+			skb->tail = skb->data + skb->len - skb->head;
+#else
 			skb->tail = skb->data + skb->len;
+#endif /* NET_SKBUFF_DATA_USES_OFFSET */
 			packet_not_copied = 1;
 		} else {
 
-- 
1.6.3.1

