From 855441dd1c8e5bb3b509cc4b264c8e064dbbc172 Mon Sep 17 00:00:00 2001
From: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
Date: Thu, 17 Jul 2014 15:58:15 -0700
Subject: [PATCH] Linux 3.10.49

commit d02dae430d1f41f2c1dc8592f68569f90b86f6d9 in linux-stable

[PG: temporarily drop 3.10.49 rtmutex changes due to conflicts w/ -rt]
Signed-off-by: Paul Gortmaker <paul.gortmaker@windriver.com>

diff --git a/Makefile b/Makefile
index f7e5680740f9..b8b8d33eab55 100644
--- a/Makefile
+++ b/Makefile
@@ -1,6 +1,6 @@
 VERSION = 3
 PATCHLEVEL = 10
-SUBLEVEL = 48
+SUBLEVEL = 49
 EXTRAVERSION =
 NAME = TOSSUG Baby Fish
 
-- 
2.0.1

