From 128501b73454a0780518d319e8ffbb379aed5bb2 Mon Sep 17 00:00:00 2001
From: Paul Gortmaker <paul.gortmaker@windriver.com>
Date: Thu, 24 Feb 2011 16:08:51 -0500
Subject: [PATCH] fork: cosmetic relocation of lockdep_tasklist_lock_is_held to match upstream

Signed-off-by: Paul Gortmaker <paul.gortmaker@windriver.com>
---
 kernel/fork.c |   16 ++++++++--------
 1 files changed, 8 insertions(+), 8 deletions(-)

diff --git a/kernel/fork.c b/kernel/fork.c
index a8d5c35..e95adda 100644
--- a/kernel/fork.c
+++ b/kernel/fork.c
@@ -96,14 +96,6 @@ __cacheline_aligned DEFINE_RWLOCK(tasklist_lock);  /* outer */
 #endif
 EXPORT_SYMBOL(tasklist_lock);
 
-/*
- * Delayed mmdrop. In the PREEMPT_RT case we
- * dont want to do this from the scheduling
- * context.
- */
-static DEFINE_PER_CPU(struct task_struct *, desched_task);
-static DEFINE_PER_CPU(struct list_head, delayed_drop_list);
-
 #ifdef CONFIG_PROVE_RCU
 int lockdep_tasklist_lock_is_held(void)
 {
@@ -112,6 +104,14 @@ int lockdep_tasklist_lock_is_held(void)
 EXPORT_SYMBOL_GPL(lockdep_tasklist_lock_is_held);
 #endif /* #ifdef CONFIG_PROVE_RCU */
 
+/*
+ * Delayed mmdrop. In the PREEMPT_RT case we
+ * dont want to do this from the scheduling
+ * context.
+ */
+static DEFINE_PER_CPU(struct task_struct *, desched_task);
+static DEFINE_PER_CPU(struct list_head, delayed_drop_list);
+
 int nr_processes(void)
 {
 	int cpu;
-- 
1.7.4

