From a1f8f163aac08e42a939c8c80104915e99b038e4 Mon Sep 17 00:00:00 2001
From: Michel Thebeau <michel.thebeau@windriver.com>
Date: Mon, 29 Nov 2010 12:58:31 -0500
Subject: [PATCH] Remove duplicate NR_syscalls for preempt-rt

Signed-off-by: Michel Thebeau <michel.thebeau@windriver.com>
---
 include/asm-x86/unistd_64.h |    2 --
 1 files changed, 0 insertions(+), 2 deletions(-)

diff --git a/include/asm-x86/unistd_64.h b/include/asm-x86/unistd_64.h
index d45adb7..1932a00 100644
--- a/include/asm-x86/unistd_64.h
+++ b/include/asm-x86/unistd_64.h
@@ -660,8 +660,6 @@ __SYSCALL(__NR_marker, sys_marker)
 #define __NR_trace				296
 __SYSCALL(__NR_trace, sys_trace)
 
-#define NR_syscalls				297
-
 #ifndef __NO_STUBS
 #define __ARCH_WANT_OLD_READDIR
 #define __ARCH_WANT_OLD_STAT
-- 
1.7.0.4

