From baf4db0274fbe9469f38dd2022a1ae51f0315aee Mon Sep 17 00:00:00 2001
From: Paul Gortmaker <paul.gortmaker@windriver.com>
Date: Sat, 19 Feb 2011 15:23:33 -0500
Subject: [PATCH] rt: delete duplicated definition of TASK_STATE_MAX

Signed-off-by: Paul Gortmaker <paul.gortmaker@windriver.com>
---
 include/linux/sched.h |    2 --
 1 files changed, 0 insertions(+), 2 deletions(-)

diff --git a/include/linux/sched.h b/include/linux/sched.h
index 705b903..b5e82d0 100644
--- a/include/linux/sched.h
+++ b/include/linux/sched.h
@@ -255,8 +255,6 @@ extern struct mutex kernel_sem;
   TASK_STATE_16 TASK_STATE_32 TASK_STATE_64 TASK_STATE_128 TASK_STATE_256 \
   TASK_STATE_512
 
-#define TASK_STATE_MAX		1024
-
 extern char ___assert_task_state[1 - 2*!!(
 		sizeof(TASK_STATE_TO_CHAR_STR)-1 != ilog2(TASK_STATE_MAX)+1)];
 
-- 
1.7.4

