From 5ee941448f1e9092ee970977be568f3daba7d313 Mon Sep 17 00:00:00 2001
From: Paul Gortmaker <paul.gortmaker@windriver.com>
Date: Wed, 2 Mar 2011 15:18:02 -0500
Subject: [PATCH] RT doesnt touch this file; restore it to what it was

Signed-off-by: Paul Gortmaker <paul.gortmaker@windriver.com>
---
 drivers/net/mlx4/mlx4.h |    1 -
 1 files changed, 0 insertions(+), 1 deletions(-)

diff --git a/drivers/net/mlx4/mlx4.h b/drivers/net/mlx4/mlx4.h
index 13343e8..bc72d6e 100644
--- a/drivers/net/mlx4/mlx4.h
+++ b/drivers/net/mlx4/mlx4.h
@@ -40,7 +40,6 @@
 #include <linux/mutex.h>
 #include <linux/radix-tree.h>
 #include <linux/timer.h>
-#include <linux/semaphore.h>
 #include <linux/workqueue.h>
 
 #include <linux/mlx4/device.h>
-- 
1.7.4

