From fe22856079ef6df079f19ab6f343757d5b49f0ff Mon Sep 17 00:00:00 2001
From: Paul Gortmaker <paul.gortmaker@windriver.com>
Date: Tue, 4 Nov 2008 15:29:36 -0500
Subject: [PATCH] tracepoint: add for sched_resched_task

preempt_rt has a tracepoint for sched_resched_task, but since LTT owns
this file, we let LTT add the file 1st (instead of RT) and then add
the tracepoint.

Signed-off-by: Paul Gortmaker <paul.gortmaker@windriver.com>
---
 include/trace/sched.h |    3 +++
 1 files changed, 3 insertions(+), 0 deletions(-)

diff --git a/include/trace/sched.h b/include/trace/sched.h
index e9f2b26..20b965d 100644
--- a/include/trace/sched.h
+++ b/include/trace/sched.h
@@ -41,6 +41,9 @@ DEFINE_TRACE(sched_process_fork,
 DEFINE_TRACE(sched_signal_send,
 	TPPROTO(int sig, struct task_struct *p),
 	TPARGS(sig, p));
+DEFINE_TRACE(sched_resched_task,
+	TPPROTO(struct task_struct *task, int cpu),
+	TPARGS(task, cpu));
 DEFINE_TRACE(sched_kthread_create,
 	TPPROTO(void *fn, int pid),
 	TPARGS(fn, pid));
-- 
1.5.5.1

