From 52d14f7d8d5aacfdd194d5b9817126b281c46028 Mon Sep 17 00:00:00 2001
From: Catalin Enache <catalin.enache@windriver.com>
Date: Mon, 24 Feb 2014 15:08:30 +0200
Subject: [PATCH 63/66] hsic: override enable HSIC flag provided by boottoapp

Signed-off-by: Catalin Enache <catalin.enache@windriver.com>
---
 drivers/usb/host/ehci-hcd.c |   15 ++++++++++++---
 1 files changed, 12 insertions(+), 3 deletions(-)

diff --git a/drivers/usb/host/ehci-hcd.c b/drivers/usb/host/ehci-hcd.c
index 0c0a719..5da651b 100644
--- a/drivers/usb/host/ehci-hcd.c
+++ b/drivers/usb/host/ehci-hcd.c
@@ -1610,16 +1610,25 @@ static int __init ehci_hcd_init(void)
 	for (i = 0; i < ARRAY_SIZE(plat_drivers); i++) {
 /* SWISTART */
 #ifdef CONFIG_SIERRA_AR7
+#ifndef HSIC_PLATFORM_DRIVER
 		if(!((bsreadboottoappflag()& BCBOOTAPPFLAG_HSIC_ENABLE_M)))
 		{
-			pr_info("%s():HSIC host is disabe", __func__ );
+			pr_info("%s():HSIC host is disabled", __func__ );
 			if(strcmp(plat_drivers[i]->driver.name, hsichostname )== 0)
 			{
 				continue;
 			}
 		}
-#endif
-
+#else /* HSIC_PLATFORM_DRIVER */
+		if(strcmp(plat_drivers[i]->driver.name, hsichostname )== 0)
+		{
+			pr_info("%s(): HSIC host: boottoapp says '%s' should be %s",
+					__func__, hsichostname,
+					((bsreadboottoappflag()& BCBOOTAPPFLAG_HSIC_ENABLE_M)) ? "enabled" : "disabled");
+			continue;
+		}
+#endif /* HSIC_PLATFORM_DRIVER */
+#endif /* CONFIG_SIERRA_AR7 */
 /* SWISTOP */
 		retval = platform_driver_register(plat_drivers[i]);
 		if (retval) {
-- 
1.7.5.4

