From 4f03ba5b0ba68d992c5c65a7bcef52ce24985547 Mon Sep 17 00:00:00 2001
From: Catalin Enache <catalin.enache@windriver.com>
Date: Mon, 20 Jan 2014 14:47:27 +0200
Subject: [PATCH 61/66] Remove compilation error

Signed-off-by: Catalin Enache <catalin.enache@windriver.com>
---
 arch/arm/mach-msm/timer.c |    2 ++
 1 files changed, 2 insertions(+), 0 deletions(-)

diff --git a/arch/arm/mach-msm/timer.c b/arch/arm/mach-msm/timer.c
index dd3ac4c..0eb99ae 100644
--- a/arch/arm/mach-msm/timer.c
+++ b/arch/arm/mach-msm/timer.c
@@ -1168,7 +1168,9 @@ static void __init msm_timer_init(void)
 	if (is_smp()) {
 		__raw_writel(1,
 			msm_clocks[MSM_CLOCK_DGT].regbase + TIMER_ENABLE);
+#if 0			
 		set_delay_fn(read_current_timer_delay_loop);
+#endif		
 	}
 #endif
 #endif
-- 
1.7.5.4

