From 7dc2e07b8c425c6ab7171f79d8475db3434cdb6a Mon Sep 17 00:00:00 2001
From: Daniel Vetter <daniel.vetter@ffwll.ch>
Date: Mon, 26 Dec 2016 16:48:25 +0100
Subject: [PATCH 2294/4301] drm/i915: Update DRIVER_DATE to 20161226

commit f061ff077e657b27d7e24e31b1da484b2376348d upstream

Signed-off-by: Daniel Vetter <daniel.vetter@ffwll.ch>
---
 drivers/gpu/drm/i915/i915_drv.h |    4 ++--
 1 files changed, 2 insertions(+), 2 deletions(-)

diff --git a/drivers/gpu/drm/i915/i915_drv.h b/drivers/gpu/drm/i915/i915_drv.h
index fdc60c0..46f3af6 100644
--- a/drivers/gpu/drm/i915/i915_drv.h
+++ b/drivers/gpu/drm/i915/i915_drv.h
@@ -78,8 +78,8 @@
 
 #define DRIVER_NAME		"i915"
 #define DRIVER_DESC		"Intel Graphics"
-#define DRIVER_DATE		"20161205"
-#define DRIVER_TIMESTAMP	1480926326
+#define DRIVER_DATE		"20161226"
+#define DRIVER_TIMESTAMP	1482767304
 
 #undef WARN_ON
 /* Many gcc seem to no see through this and fall over :( */
-- 
1.7.5.4

