From d82fd66e971ea0b5382ed273b95c18ae21f8f4ce Mon Sep 17 00:00:00 2001
From: Daniel Vetter <daniel.vetter@ffwll.ch>
Date: Mon, 9 Jan 2017 10:12:02 +0100
Subject: [PATCH 2627/4301] drm/i915: Update DRIVER_DATE to 20170109

commit 5d799acdd057e4f10fdd09ade22028c83f829f3e upstream

Signed-off-by: Daniel Vetter <daniel.vetter@ffwll.ch>
---
 drivers/gpu/drm/i915/i915_drv.h |    4 ++--
 1 files changed, 2 insertions(+), 2 deletions(-)

diff --git a/drivers/gpu/drm/i915/i915_drv.h b/drivers/gpu/drm/i915/i915_drv.h
index d99e76d..4ee5c65 100644
--- a/drivers/gpu/drm/i915/i915_drv.h
+++ b/drivers/gpu/drm/i915/i915_drv.h
@@ -80,8 +80,8 @@
 
 #define DRIVER_NAME		"i915"
 #define DRIVER_DESC		"Intel Graphics"
-#define DRIVER_DATE		"20161226"
-#define DRIVER_TIMESTAMP	1482767304
+#define DRIVER_DATE		"20170109"
+#define DRIVER_TIMESTAMP	1483953121
 
 #undef WARN_ON
 /* Many gcc seem to no see through this and fall over :( */
-- 
1.7.5.4

