From 608b05597af04e91398a5783265d0f82f471ede9 Mon Sep 17 00:00:00 2001
From: Daniel Vetter <daniel.vetter@ffwll.ch>
Date: Mon, 5 Dec 2016 09:25:26 +0100
Subject: [PATCH 2087/4301] drm/i915: Update DRIVER_DATE to 20161205

commit ce6612d6844701a5b5a162f201103240c07abd9e upstream

Signed-off-by: Daniel Vetter <daniel.vetter@ffwll.ch>
---
 drivers/gpu/drm/i915/i915_drv.h |    4 ++--
 1 files changed, 2 insertions(+), 2 deletions(-)

diff --git a/drivers/gpu/drm/i915/i915_drv.h b/drivers/gpu/drm/i915/i915_drv.h
index d4d9d86..15d6d7d 100644
--- a/drivers/gpu/drm/i915/i915_drv.h
+++ b/drivers/gpu/drm/i915/i915_drv.h
@@ -78,8 +78,8 @@
 
 #define DRIVER_NAME		"i915"
 #define DRIVER_DESC		"Intel Graphics"
-#define DRIVER_DATE		"20161121"
-#define DRIVER_TIMESTAMP	1479717903
+#define DRIVER_DATE		"20161205"
+#define DRIVER_TIMESTAMP	1480926326
 
 #undef WARN_ON
 /* Many gcc seem to no see through this and fall over :( */
-- 
1.7.5.4

