From 8698dbca40f842bdfd9e7f2f033e21ba7cb93616 Mon Sep 17 00:00:00 2001
From: Yue Tao <Yue.Tao@windriver.com>
Date: Wed, 10 Oct 2018 15:32:38 +0800
Subject: [PATCH] The part of upstream commit 447ae316

Signed-off-by: Yue Tao <Yue.Tao@windriver.com>
---
 drivers/gpu/drm/i915/intel_lpe_audio.c |    1 +
 1 file changed, 1 insertion(+)

diff --git a/drivers/gpu/drm/i915/intel_lpe_audio.c b/drivers/gpu/drm/i915/intel_lpe_audio.c
index 7a5b41b..427d350 100644
--- a/drivers/gpu/drm/i915/intel_lpe_audio.c
+++ b/drivers/gpu/drm/i915/intel_lpe_audio.c
@@ -63,6 +63,7 @@
 #include <linux/acpi.h>
 #include <linux/device.h>
 #include <linux/pci.h>
+#include <linux/irq.h>
 
 #include "i915_drv.h"
 #include <linux/delay.h>
-- 
1.7.9.5

