From 571ed6c8cb7129ae3bfe1415fde5132b09e1a50c Mon Sep 17 00:00:00 2001
From: Paul Butler <paul.butler@windriver.com>
Date: Mon, 4 Nov 2013 14:07:15 -0600
Subject: [PATCH 08/75] drivers/tty: AXM5516 emulation bringup - amba-pl011.c

Signed-off-by: Paul Butler <paul.butler@windriver.com>
---
 drivers/tty/serial/amba-pl011.c |    4 ++--
 1 files changed, 2 insertions(+), 2 deletions(-)

diff --git a/drivers/tty/serial/amba-pl011.c b/drivers/tty/serial/amba-pl011.c
index da614ca..37681fb 100644
--- a/drivers/tty/serial/amba-pl011.c
+++ b/drivers/tty/serial/amba-pl011.c
@@ -1670,8 +1670,8 @@ pl011_set_termios(struct uart_port *port, struct ktermios *termios,
 			quot -= 2;
 	}
 	/* Set baud rate */
-	writew(quot & 0x3f, port->membase + UART011_FBRD);
-	writew(quot >> 6, port->membase + UART011_IBRD);
+	writew(0x5, port->membase + UART011_FBRD);
+	writew(0x1, port->membase + UART011_IBRD);
 
 	/*
 	 * ----------v----------v----------v----------v-----
-- 
1.7.5.4

