From 2800e51e5fe711a6db4674b429622e44e3f01c55 Mon Sep 17 00:00:00 2001
From: Paul Butler <paul.butler@windriver.com>
Date: Mon, 4 Nov 2013 14:10:05 -0600
Subject: [PATCH 11/75] drivers/net: AXM5516 emulation bringup - Makefile

Signed-off-by: Paul Butler <paul.butler@windriver.com>
---
 drivers/net/ethernet/lsi/Makefile |    1 +
 1 files changed, 1 insertions(+), 0 deletions(-)

diff --git a/drivers/net/ethernet/lsi/Makefile b/drivers/net/ethernet/lsi/Makefile
index 2754e5b..a48a239 100644
--- a/drivers/net/ethernet/lsi/Makefile
+++ b/drivers/net/ethernet/lsi/Makefile
@@ -1,3 +1,4 @@
 # Makefile for the LSI FEMAC network interface.
 
+obj-$(CONFIG_LSI_NET) += lsi_acp_mdio.o
 obj-$(CONFIG_LSI_NET) += lsi_acp_net.o
-- 
1.7.5.4

