From 75a42591317e178047924dc1046a984389032afd Mon Sep 17 00:00:00 2001
From: Bruce Ashfield <bruce.ashfield@windriver.com>
Date: Mon, 2 Feb 2009 11:59:24 -0500
Subject: [PATCH 1/4] kconfig: OCD hooks

Allows the Wind River OCD tool to get module symbols
for module debug. It will also disable any potential manipulation
of the debug control register, which is generally done by further
selecting CONFIG_BDI_SWITCH.

Signed-off-by: Bruce Ashfield <bruce.ashfield@windriver.com>
Signed-off-by: Yang Shi <yang.shi@windriver.com>
---
 lib/Kconfig.debug | 13 ++++++++++++-
 1 file changed, 12 insertions(+), 1 deletion(-)

diff --git a/lib/Kconfig.debug b/lib/Kconfig.debug
index 74fdc5c..e27c9c2 100644
--- a/lib/Kconfig.debug
+++ b/lib/Kconfig.debug
@@ -25,6 +25,16 @@ config DEFAULT_MESSAGE_LOGLEVEL
 	  that are auditing their logs closely may want to set it to a lower
 	  priority.
 
+config WR_OCD_DEBUG
+	bool "WR_OCD_DEBUG: Debugging with WindRiver System OCD"
+	select DEBUG_INFO
+	select BDI_SWITCH
+	help
+	  If you say Y here, WindRiver OCD tool can get module symbols
+	  for module debug. It will also disable any potential manipulation
+	  of the debug control register, which is generally done by further
+	  selecting CONFIG_BDI_SWITCH. If unsure, say N.
+
 config ENABLE_WARN_DEPRECATED
 	bool "Enable __deprecated logic"
 	default y
@@ -709,7 +719,8 @@ config DEBUG_BUGVERBOSE
 
 config DEBUG_INFO
 	bool "Compile the kernel with debug info"
-	depends on DEBUG_KERNEL
+	depends on DEBUG_KERNEL || WR_OCD_DEBUG
+	default y if WR_OCD_DEBUG
 	help
           If you say Y here the resulting kernel image will include
 	  debugging info resulting in a larger kernel image.
-- 
1.8.4.93.g57e4c17

