From 10a622493d7f9343e8b4118031ff0c21a27cc4e9 Mon Sep 17 00:00:00 2001
From: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
Date: Thu, 31 Jul 2014 14:55:39 -0700
Subject: [PATCH] Linux 3.10.51


diff --git a/Makefile b/Makefile
index ffd8082dceb2..f9f6ee59c61a 100644
--- a/Makefile
+++ b/Makefile
@@ -1,6 +1,6 @@
 VERSION = 3
 PATCHLEVEL = 10
-SUBLEVEL = 50
+SUBLEVEL = 51
 EXTRAVERSION =
 NAME = TOSSUG Baby Fish
 
-- 
2.0.1

