From 5499f752802f5ebb11f4440dcf707c605bc5a9d0 Mon Sep 17 00:00:00 2001
From: Zumeng Chen <zumeng.chen@windriver.com>
Date: Sun, 26 Jan 2014 16:50:26 +0800
Subject: [PATCH 1115/1115] arm: omap2+: set SOC_AM33XX as no default yes

No soc should be set yes by default in a common system,
so 'default y' is removed from SOC_AM33XX config.

Signed-off-by; Zumeng Chen <zumeng.chen@windriver.com>
---
 arch/arm/mach-omap2/Kconfig |    1 -
 1 files changed, 0 insertions(+), 1 deletions(-)

diff --git a/arch/arm/mach-omap2/Kconfig b/arch/arm/mach-omap2/Kconfig
index 2141acb..bda4cc8 100644
--- a/arch/arm/mach-omap2/Kconfig
+++ b/arch/arm/mach-omap2/Kconfig
@@ -67,7 +67,6 @@ config SOC_OMAP5
 config SOC_AM33XX
 	bool "TI AM33XX"
 	depends on ARCH_MULTI_V7
-	default y
 	select ARM_CPU_SUSPEND if PM
 	select COMMON_CLK
 	select CPU_V7
-- 
1.7.5.4

