From de0985690cb8e386263139d557af62c6f19a7e1d Mon Sep 17 00:00:00 2001
From: Michel Thebeau <michel.thebeau@windriver.com>
Date: Wed, 4 Dec 2013 10:51:42 -0500
Subject: [PATCH] kgit: creating baseline state

Signed-off-by: Michel Thebeau <michel.thebeau@windriver.com>
---
 .gitignore |    6 +++---
 1 file changed, 3 insertions(+), 3 deletions(-)

diff --git a/.gitignore b/.gitignore
index 3b8b9b33be38..355799998e9d 100644
--- a/.gitignore
+++ b/.gitignore
@@ -64,11 +64,11 @@ include/generated
 arch/*/include/generated
 
 # stgit generated dirs
-patches-*
+#patches-*
 
 # quilt's files
-patches
-series
+#patches
+#series
 
 # cscope files
 cscope.*
-- 
1.7.10.4

