From bae9392c616b13563f8b12e646be4e1b7cfd74b5 Mon Sep 17 00:00:00 2001
From: Tomi Valkeinen <tomi.valkeinen@ti.com>
Date: Fri, 3 May 2013 11:35:43 +0300
Subject: [PATCH 0769/1115] OMAPDSS: omapdss.h: add owner field to
 omap_dss_device

Add struct module *owner field to omap_dss_device, which points to the
module containing the ops for this omap_dss_device. This will be used to
manage the ref count for the module.

Signed-off-by: Tomi Valkeinen <tomi.valkeinen@ti.com>
(cherry picked from commit 4f3e44ea07eafb5252e8414ae4f75f749e1729de)
Signed-off-by: Zumeng Chen <zumeng.chen@windriver.com>
---
 include/video/omapdss.h |    2 ++
 1 files changed, 2 insertions(+), 0 deletions(-)

diff --git a/include/video/omapdss.h b/include/video/omapdss.h
index 8d99e68..0d800e0 100644
--- a/include/video/omapdss.h
+++ b/include/video/omapdss.h
@@ -709,6 +709,8 @@ struct omapdss_dsi_ops {
 struct omap_dss_device {
 	struct device *dev;
 
+	struct module *owner;
+
 	struct list_head panel_list;
 
 	/* alias in the form of "display%d" */
-- 
1.7.5.4

