From 9392a3ba5b9555aa1e63469db732204ebaadb8f7 Mon Sep 17 00:00:00 2001
From: Tomi Valkeinen <tomi.valkeinen@ti.com>
Date: Fri, 24 May 2013 13:19:30 +0300
Subject: [PATCH 0460/1115] OMAPDSS: DVI: Add ops

Add "ops" style method for using DVI functionality.

Ops style calls will allow us to have arbitrarily long display
pipelines, where each entity can call ops in the previous display
entity.

Signed-off-by: Tomi Valkeinen <tomi.valkeinen@ti.com>
(cherry picked from commit 7700c2d4f79c423f29a5c2c10ca5a9b9c8c5c60f)
Signed-off-by: Zumeng Chen <zumeng.chen@windriver.com>
---
 include/video/omapdss.h |   18 ++++++++++++++++++
 1 files changed, 18 insertions(+), 0 deletions(-)

diff --git a/include/video/omapdss.h b/include/video/omapdss.h
index 99a67d0..633a885 100644
--- a/include/video/omapdss.h
+++ b/include/video/omapdss.h
@@ -611,6 +611,23 @@ struct omapdss_sdi_ops {
 	void (*set_datapairs)(struct omap_dss_device *dssdev, int datapairs);
 };
 
+struct omapdss_dvi_ops {
+	int (*connect)(struct omap_dss_device *dssdev,
+			struct omap_dss_device *dst);
+	void (*disconnect)(struct omap_dss_device *dssdev,
+			struct omap_dss_device *dst);
+
+	int (*enable)(struct omap_dss_device *dssdev);
+	void (*disable)(struct omap_dss_device *dssdev);
+
+	int (*check_timings)(struct omap_dss_device *dssdev,
+			struct omap_video_timings *timings);
+	void (*set_timings)(struct omap_dss_device *dssdev,
+			struct omap_video_timings *timings);
+	void (*get_timings)(struct omap_dss_device *dssdev,
+			struct omap_video_timings *timings);
+};
+
 struct omap_dss_device {
 	/* old device, to be removed */
 	struct device old_dev;
@@ -684,6 +701,7 @@ struct omap_dss_device {
 	union {
 		const struct omapdss_dpi_ops *dpi;
 		const struct omapdss_sdi_ops *sdi;
+		const struct omapdss_dvi_ops *dvi;
 	} ops;
 
 	/* helper variable for driver suspend/resume */
-- 
1.7.5.4

