From 88a58ac2d4db0adfcbd9f9e604998deff186cf73 Mon Sep 17 00:00:00 2001
From: Quanyang Wang <quanyang.wang@windriver.com>
Date: Wed, 22 Jun 2016 10:43:42 +0000
Subject: [PATCH 1/5] arm: axxia: delete redudant lsi-femac driver in Kconfig

Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
---
 drivers/net/ethernet/lsi/Kconfig |    8 --------
 1 files changed, 0 insertions(+), 8 deletions(-)

diff --git a/drivers/net/ethernet/lsi/Kconfig b/drivers/net/ethernet/lsi/Kconfig
index 408e643..c6d67f0 100644
--- a/drivers/net/ethernet/lsi/Kconfig
+++ b/drivers/net/ethernet/lsi/Kconfig
@@ -21,14 +21,6 @@ config NET_LSI_MDIO
 
 	  If your board uses an external PHY connected to FEMAC, say Y.
 
-config NET_LSI_FEMAC
-	tristate "LSI FEMAC Ethernet controller"
-	select LSI_MDIO
-	default y
-	---help---
-	  Say Y here if you want to use the built-in 10/100 Fast ethernet
-	  controller (FEMAC) on the AXM55xx devices.
-
 endif # NET_VENDOR_LSI
 
 
-- 
1.7.5.4

