From 501c3bbd30d12532688d1d64c524794fe7cc7003 Mon Sep 17 00:00:00 2001
From: Qi Hou <qi.hou@windriver.com>
Date: Tue, 23 Feb 2016 10:58:20 +0800
Subject: [PATCH] fixed ERROR while compiling bsp of axxiaarm with DEBUG_LL
 enabled.

When ERROR, the output likes below:
axxiaarm/kernel-source/arch/arm/mach-axxia/include/mach/debug-macro.S:21:38:
fatal error: asm/hardware/debug-pl01x.S: No such file or directory ...

fixed this compiling error by correcting the include of PL01X debug.

The PL01X debug file named "debug-pl01x.S" has been moved from arch/arm/include/asm/hardware/
to arch/arm/include/debug/ and been renamed as pl01x.S, but the include of it in
arch/arm/kernel/debug.S in code of bsp of axxiaarm hasn't be amended.
So here, I just corrected including path of it to "#include <debug/pl01x.S>".
---
 arch/arm/mach-axxia/include/mach/debug-macro.S |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/arch/arm/mach-axxia/include/mach/debug-macro.S b/arch/arm/mach-axxia/include/mach/debug-macro.S
index 1a28f4a..eebcc1f 100644
--- a/arch/arm/mach-axxia/include/mach/debug-macro.S
+++ b/arch/arm/mach-axxia/include/mach/debug-macro.S
@@ -18,4 +18,4 @@
 	ldr	\rv, =AXXIA_DEBUG_UART_VIRT
 	.endm
 
-#include <asm/hardware/debug-pl01x.S>
+#include <debug/pl01x.S>
-- 
1.7.5.4

