From 92ca7478a716d8756fcfeb145319bcf7e4ef80d3 Mon Sep 17 00:00:00 2001
From: SangeethaRao <sangeetha.rao@intel.com>
Date: Fri, 22 May 2015 19:03:33 -0500
Subject: [PATCH 42/86] fs/Kconfig: Updated for 3.19 build

Signed-off-by: SangeethaRao <sangeetha.rao@intel.com>
---
 fs/Kconfig |    1 +
 1 files changed, 1 insertions(+), 0 deletions(-)

diff --git a/fs/Kconfig b/fs/Kconfig
index 9d207b1..3b6bba2 100644
--- a/fs/Kconfig
+++ b/fs/Kconfig
@@ -279,5 +279,6 @@ endif # NETWORK_FILESYSTEMS
 
 source "fs/nls/Kconfig"
 source "fs/dlm/Kconfig"
+source "fs/vmfs/Kconfig"
 
 endmenu
-- 
1.7.5.4

