From b60d1daf373883b823ee0201cce47e239c9d2a45 Mon Sep 17 00:00:00 2001
From: Soren Brinkmann <soren.brinkmann@xilinx.com>
Date: Thu, 1 May 2014 09:36:40 -0700
Subject: [PATCH 363/793] ARM: zynq: Leave L2 cache enabled during suspend

Signed-off-by: Soren Brinkmann <soren.brinkmann@xilinx.com>
Signed-off-by: Michal Simek <michal.simek@xilinx.com>
---
 arch/arm/mach-zynq/pm.c |    2 --
 1 files changed, 0 insertions(+), 2 deletions(-)

diff --git a/arch/arm/mach-zynq/pm.c b/arch/arm/mach-zynq/pm.c
index 44e3f9c..26e2949 100644
--- a/arch/arm/mach-zynq/pm.c
+++ b/arch/arm/mach-zynq/pm.c
@@ -188,9 +188,7 @@ static int zynq_pm_enter(suspend_state_t suspend_state)
 	switch (suspend_state) {
 	case PM_SUSPEND_STANDBY:
 	case PM_SUSPEND_MEM:
-		outer_disable();
 		cpu_suspend(0, zynq_pm_suspend);
-		outer_resume();
 		break;
 	default:
 		return -EINVAL;
-- 
1.7.5.4

