From ec5c63b8534bdfe5f56159ed6f4e1bca08c0723c Mon Sep 17 00:00:00 2001
From: Yue Tao <Yue.Tao@windriver.com>
Date: Fri, 1 May 2015 17:47:06 +0800
Subject: [PATCH] driver:gpio-zynq.c remove all usage of gpio_remove retval in
 driver.

Signed-off-by: Yue Tao <Yue.Tao@windriver.com>
---
 drivers/gpio/gpio-zynq.c |    6 ++----
 1 files changed, 2 insertions(+), 4 deletions(-)

diff --git a/drivers/gpio/gpio-zynq.c b/drivers/gpio/gpio-zynq.c
index 1f5fdfc..7709949 100644
--- a/drivers/gpio/gpio-zynq.c
+++ b/drivers/gpio/gpio-zynq.c
@@ -599,15 +599,13 @@ static int zynq_gpio_probe(struct platform_device *pdev)
 	gpio->clk = devm_clk_get(&pdev->dev, NULL);
 	if (IS_ERR(gpio->clk)) {
 		dev_err(&pdev->dev, "input clock not found.\n");
-		if (gpiochip_remove(chip))
-			dev_err(&pdev->dev, "Failed to remove gpio chip\n");
+		gpiochip_remove(chip);
 		return PTR_ERR(gpio->clk);
 	}
 	ret = clk_prepare_enable(gpio->clk);
 	if (ret) {
 		dev_err(&pdev->dev, "Unable to enable clock.\n");
-		if (gpiochip_remove(chip))
-			dev_err(&pdev->dev, "Failed to remove gpio chip\n");
+		gpiochip_remove(chip);
 		return ret;
 	}
 
-- 
1.7.5.4

