From ae92585789e20e42747450a9894dbcd80b04144c Mon Sep 17 00:00:00 2001
From: Michal Simek <michal.simek@xilinx.com>
Date: Fri, 21 Aug 2015 10:52:03 +0200
Subject: [PATCH 554/793] irqchip: Enable Xilinx intc for ZynqMP

Signed-off-by: Michal Simek <michal.simek@xilinx.com>
---
 drivers/irqchip/Kconfig |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/drivers/irqchip/Kconfig b/drivers/irqchip/Kconfig
index 0775642..529d0645 100644
--- a/drivers/irqchip/Kconfig
+++ b/drivers/irqchip/Kconfig
@@ -201,7 +201,7 @@ config VERSATILE_FPGA_IRQ_NR
 
 config XILINX_INTC
 	bool "Xilinx Interrupt Controller (IP core)"
-	depends on ARCH_ZYNQ
+	depends on ARCH_ZYNQ || ARCH_ZYNQMP
 	select IRQ_DOMAIN
 	help
 	  Support for the Xilinx Interrupt Controller which can be used
-- 
1.7.5.4

