From 2fe31b62e160ef9a22ac2e2b6bd5f68bf69af620 Mon Sep 17 00:00:00 2001
From: Fabian Vogt <fabian@ritter-vogt.de>
Date: Mon, 16 Nov 2015 19:00:18 +0100
Subject: [PATCH 294/793] nand: pl35x: Fix unreachable code

Bug: (x == 1) && (x == 2) is always false.

Signed-off-by: Fabian Vogt <fabian@ritter-vogt.de>
Reviewed-by: Punnaiah Choudary Kalluri <punnaia@xilinx.com>
Signed-off-by: Michal Simek <michal.simek@xilinx.com>
---
 drivers/mtd/nand/pl35x_nand.c |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/drivers/mtd/nand/pl35x_nand.c b/drivers/mtd/nand/pl35x_nand.c
index 406f1f0..a7eb32b 100644
--- a/drivers/mtd/nand/pl35x_nand.c
+++ b/drivers/mtd/nand/pl35x_nand.c
@@ -726,7 +726,7 @@ static void pl35x_nand_cmd_function(struct mtd_info *mtd, unsigned int command,
 	else
 		end_cmd = curr_cmd->end_cmd;
 
-	if ((command == NAND_CMD_READ0) && (command == NAND_CMD_SEQIN))
+	if (command == NAND_CMD_READ0 || command == NAND_CMD_SEQIN)
 		addrcycles = xnand->row_addr_cycles + xnand->col_addr_cycles;
 	else if (command == NAND_CMD_ERASE1)
 		addrcycles = xnand->row_addr_cycles;
-- 
1.7.5.4

