From d8d098d7e2edeae755379b265e4b7f26f85528c0 Mon Sep 17 00:00:00 2001
From: Michal Simek <michal.simek@xilinx.com>
Date: Tue, 4 Feb 2014 15:34:45 +0100
Subject: [PATCH 234/793] spi: zynq: Change zynq spi compat string

Signed-off-by: Michal Simek <michal.simek@xilinx.com>
---
 drivers/spi/spi-zynq-qspi.c |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/drivers/spi/spi-zynq-qspi.c b/drivers/spi/spi-zynq-qspi.c
index bbf83ac..8d984d6 100644
--- a/drivers/spi/spi-zynq-qspi.c
+++ b/drivers/spi/spi-zynq-qspi.c
@@ -1239,7 +1239,7 @@ static int zynq_qspi_remove(struct platform_device *pdev)
 MODULE_ALIAS("platform:" DRIVER_NAME);
 
 static struct of_device_id zynq_qspi_of_match[] = {
-	{ .compatible = "xlnx,ps7-qspi-1.00.a", },
+	{ .compatible = "xlnx,zynq-qspi-1.00.a", },
 	{ /* end of table */}
 };
 MODULE_DEVICE_TABLE(of, zynq_qspi_of_match);
-- 
1.7.5.4

