From 290d9b2ef072454de2e26aa73c007f78c672834c Mon Sep 17 00:00:00 2001
From: Prathap M S <msprathap@ti.com>
Date: Mon, 2 Sep 2013 12:05:23 +0530
Subject: [PATCH 067/285] video: da8xx-fb: Add API to register wait for vsync
 callback

This patch adds APIs to register and unregister wait for vsync callback.
This is derived from commit id 2d44302545da24fd22912d964102bc31a7489e97
This commit id was part of 3.2 kernel sources.

Signed-off-by: Prathap M S <msprathap@ti.com>
[Kevin: The original patch taken from TI
ti-sdk-am437x-evm-07.01.00.00-Linux-x86-Install.bin, part of the
original patch was already merged into wrlinux]
Signed-off-by: Kevin Hao <kexin.hao@windriver.com>
---
 include/video/da8xx-fb.h |    4 ++++
 1 files changed, 4 insertions(+), 0 deletions(-)

diff --git a/include/video/da8xx-fb.h b/include/video/da8xx-fb.h
index 2a7ea6f..cc0f7c5 100644
--- a/include/video/da8xx-fb.h
+++ b/include/video/da8xx-fb.h
@@ -110,5 +110,9 @@ void da8xx_register_encoder(struct da8xx_encoder *encoder);
 void da8xx_unregister_encoder(struct da8xx_encoder *encoder);
 
 
+typedef void (*vsync_callback_t)(void *arg);
+int register_vsync_cb(vsync_callback_t handler, void *arg, int idx);
+int unregister_vsync_cb(vsync_callback_t handler, void *arg, int idx);
+
 #endif  /* ifndef DA8XX_FB_H */
 
-- 
1.7.5.4

