From f83ad4ff4a374999e7dfecac3edf8e4255f5dd71 Mon Sep 17 00:00:00 2001
From: Fugang Duan <b38611@freescale.com>
Date: Mon, 4 Aug 2014 10:46:11 +0800
Subject: [PATCH 0919/1074] ENGR00329822-02 dma: sdma: optimizate the sdma
 script for uart

Optimizate the sdma script for uart.

Signed-off-by: Robin Gong <b38343@freescale.com>
Signed-off-by: Fugang Duan <B38611@freescale.com>
[Original patch taken from git://git.freescale.com/imx/linux-2.6-imx.git]
Signed-off-by: Biyao Zhai <biyao.zhai@windriver.com>
---
 firmware/imx/sdma/sdma-imx6q.bin.ihex |  114 +++++++++++++++++----------------
 1 files changed, 58 insertions(+), 56 deletions(-)

diff --git a/firmware/imx/sdma/sdma-imx6q.bin.ihex b/firmware/imx/sdma/sdma-imx6q.bin.ihex
index 2511daa..c97e61f 100644
--- a/firmware/imx/sdma/sdma-imx6q.bin.ihex
+++ b/firmware/imx/sdma/sdma-imx6q.bin.ihex
@@ -1,15 +1,15 @@
 :1000000053444D4101000000010000001C000000AD
-:1000100028000000BC000000100800008202000060
-:10002000FFFFFFFF00000000FFFFFFFFA90400002B
+:1000100028000000BC000000280800008202000048
+:10002000FFFFFFFF00000000FFFFFFFFFFFFFFFFDC
 :10003000FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFD0
-:10004000FFFFFFFFFFFFFFFF6A1A0000FFFFFFFF38
-:10005000EB020000BB180000FFFFFFFFBC1A00000E
+:10004000FFFFFFFFFFFFFFFF6A1A00004B040000E5
+:10005000EB020000BB180000FFFFFFFFC81A000002
 :10006000FFFFFFFFC0030000FFFFFFFFFFFFFFFFD9
 :10007000FFFFFFFFAB020000FFFFFFFF7B0300005D
 :10008000FFFFFFFFFFFFFFFF4C0400006E040000B6
 :10009000FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF70
 :1000A000000000000018000062180000161A00008E
-:1000B0006B1B0000061B0000A21B0000E3C1DB5706
+:1000B000771B0000121B0000AE1B0000E3C1DB57E2
 :1000C000E35FE357F352016A8F00D500017D8D0095
 :1000D000A005EB5D7804037D79042C7D367C7904E2
 :1000E0001F7CEE56000F6006057D0965437E0A629F
@@ -88,55 +88,57 @@
 :100570002652096A097C286A077F0000DB57004D74
 :10058000057DFAC1DB571B9A77C254040AC2189A38
 :10059000E3C1DB57F352056AFB568E02941AC36A15
-:1005A000C8626902247D941EC36ED36EC86248027D
-:1005B000C86A9426981EC36ED36EC8624C02C86A7D
-:1005C0009826C36E981EC36EC8629826C36E6002DA
-:1005D000097CC8626E02247D096A1E7F0125004DD8
-:1005E000257D849A286A187F04627AC2B89AE36EDD
-:1005F0008F00D805017D8D00A005C8626E02107DB8
-:10060000096A0A7F0120F97C286A067F0000004DF4
-:100610000D7DFAC1DB576E9A070004620C6AB59A29
-:10062000286AFA7F04627AC258045404286AF47F64
-:100630000AC26B9AD9C1E3C1DB57F352056AFB5674
-:100640008E02941A025269021D7D941E06524802BF
-:10065000065A9426981E06524C02065A9826981E50
-:10066000065260020A7C982606526E02237D096AB1
-:100670001D7F0125004D247DD19A286A177F0462D1
-:100680007AC2029B8F00D805017D8D00A00506521D
-:100690006E02107D096A0A7F0120F97C286A067FB4
-:1006A0000000004D0D7DFAC1DB57C19A07000462BE
-:1006B0000C6AFF9A286AFA7F04627AC258045404CA
-:1006C000286AF47F0AC2BE9A4ADBDB57F352056AF6
-:1006D000C7698F00D500017D8D00A0057804037DDA
-:1006E00079041C7D157C79041E7CEE56C862287E38
-:1006F0006006027D10021202096A227F1202096A54
-:100700001F7F1202096A1C7F03200048EF7C399B7F
-:10071000150015000278C8620B6A389B1500027834
-:10072000C8620A6A389B0278C862096A097C286A2A
-:10073000077F004D077DD3520108025800040A9B31
-:1007400077C25404D35201080258010462DB079BAC
-:10075000700B11031353DB5F0A07D3588B00FB5E4A
-:100760003CC1157DC05AC85B4EC1117C8803F05D49
-:10077000FF0D1105FF1DBC05004D047D619B0807A1
-:10078000DB5F0A078B00C352CB5359C1EC7D0002DB
-:10079000529B016E0B612F7E0B622D7E0B632B7EB5
-:1007A0000C0D1704170417049D04081DCC05017CCB
-:1007B0000C0DD16A000F4207C86FDD6F1C7F8E00E1
-:1007C0009D0001680B67177ED56B04080278C86F1F
-:1007D0001207117C0B670F7E04080278C86F12079E
-:1007E0000A7CDD6F087FD169010FC86FDD6F037F61
-:1007F00001010004779B0700FF680C680002779BEB
-:10080000E7DBDB57F352056AFB52C76A8F00D5005E
-:10081000017D8D00A0057804037D79041C7D157C85
-:1008200079041E7CEE56C862287E6006027D1002A6
-:100830001202096A227F1202096A1F7F1202096AE4
-:100840001C7F03200048EF7CD69B15001500027822
-:10085000C8620B6AD59B15000278C8620A6AD59BEC
-:100860000278C862096A097C286A077F004D077D03
-:10087000D352010802580004A69B77C25404D352F5
-:10088000010802580104FFDBA39B700B11031353F3
-:10089000DB5F0A07D3588B00FB5E3CC1157DC05A55
-:1008A000C85B4EC1117C8803F05DFF0D1105FF1D73
-:1008B000BC05004D047DFE9B0807DB5F0A078B002B
-:0C08C000C352CB5359C1EC7D0002EF9BEA
+:1005A000C8626902267D941EC36ED36EC86248027B
+:1005B000C86A9426981EC36EC8629826C36E6002ED
+:1005C0000E7D981EC36EC8626C02037D9826C36EB2
+:1005D000AB9A4C02D36EC86A9826C36EBB9AC862A7
+:1005E0006E02017CB59A096A1A7F0125004D217DB2
+:1005F0007C9AE36E8F00D805017D8D00C8626E0283
+:10060000127D096A0C7F01250120F87CDB57004D23
+:10061000107D286A047F0000FAC1DB576E9A07003C
+:1006200004620C6AB89A286AFA7F04627AC2580493
+:100630005404FF081100FF18BC00CD00017CAB9AE8
+:10064000286AED7F04627AC20AC26B9AD9C1E3C1FB
+:10065000DB57F352056AFB568E02941A0252690266
+:100660001D7D941E06524802065A9426981E065274
+:100670004C02065A9826981E065260020A7C98265A
+:1006800006526E02237D096A1D7F0125004D247DDF
+:10069000DD9A286A177F04627AC20E9B8F00D80504
+:1006A000017D8D00A00506526E02107D096A0A7F49
+:1006B0000120F97C286A067F0000004D0D7DFAC1FB
+:1006C000DB57CD9A070004620C6A0B9B286AFA7FFD
+:1006D00004627AC258045404286AF47F0AC2CA9A8F
+:1006E00056DBDB57F352056AC7698F00D500017DE1
+:1006F0008D00A0057804037D79041C7D157C7904A8
+:100700001E7CEE56C862287E6006027D1002120230
+:10071000096A227F1202096A1F7F1202096A1C7F7E
+:1007200003200048EF7C459B150015000278C86245
+:100730000B6A449B15000278C8620A6A449B0278DF
+:10074000C862096A097C286A077F004D077DD35279
+:10075000010802580004169B77C25404D3520108C2
+:10076000025801046EDB139B700B11031353DB5F04
+:100770000A07D3588B00FB5E3CC1157DC05AC85B8D
+:100780004EC1117C8803F05DFF0D1105FF1DBC05F6
+:10079000004D047D6D9B0807DB5F0A078B00C35289
+:1007A000CB5359C1EC7D00025E9B016E0B612F7E25
+:1007B0000B622D7E0B632B7E0C0D170417041704A0
+:1007C0009D04081DCC05017C0C0DD16A000F420769
+:1007D000C86FDD6F1C7F8E009D0001680B67177E60
+:1007E000D56B04080278C86F1207117C0B670F7E67
+:1007F00004080278C86F12070A7CDD6F087FD16990
+:10080000010FC86FDD6F037F01010004839B0700A8
+:10081000FF680C680002839BF3DBDB57F352056A29
+:10082000FB52C76A8F00D500017D8D00A0057804BA
+:10083000037D79041C7D157C79041E7CEE56C8620C
+:10084000287E6006027D10021202096A227F1202CF
+:10085000096A1F7F1202096A1C7F03200048EF7C8F
+:10086000E29B150015000278C8620B6AE19B150037
+:100870000278C8620A6AE19B0278C862096A097C48
+:10088000286A077F004D077DD352010802580004F3
+:10089000B29B77C25404D3520108025801040BDC06
+:1008A000AF9B700B11031353DB5F0A07D3588B0008
+:1008B000FB5E3CC1157DC05AC85B4EC1117C8803EC
+:1008C000F05DFF0D1105FF1DBC05004D047D0A9C68
+:1008D0000807DB5F0A078B00C352CB5359C1EC7D7D
+:0408E0000002FB9B7C
 :00000001FF
-- 
1.7.5.4

