From 1b4bd1877d93d59580d210290f581ba899e554f0 Mon Sep 17 00:00:00 2001
From: Bin Jiang <Bin.Jiang@windriver.com>
Date: Thu, 6 Jan 2011 18:32:49 +0800
Subject: [PATCH 28/28] Fix: on fsl_p4080, kernel fails to build when configuring with 'CONFIG_SMP=n'

When compiling with "CONFIG_SMP=n", definition of function 'get_hard_smp_processor_id'
can't be found and then induce the error.
<asm/smp.h> rather than <linux/smp.h> should be included because the UP definition of
'get_hard_smp_processor_id' is in <asm/smp.h>, and <linux/smp.h> only includes
<asm/smp.h> when 'CONFIG_SMP=y'.
---
 drivers/hwalloc/bman_config.c  |    4 ----
 drivers/hwalloc/bman_private.h |    3 +++
 drivers/hwqueue/qman_config.c  |    4 ----
 drivers/hwqueue/qman_private.h |    3 +++
 4 files changed, 6 insertions(+), 8 deletions(-)

diff --git a/drivers/hwalloc/bman_config.c b/drivers/hwalloc/bman_config.c
index b31f265..74b321c 100644
--- a/drivers/hwalloc/bman_config.c
+++ b/drivers/hwalloc/bman_config.c
@@ -30,10 +30,6 @@
  * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
  */
 
-#ifndef CONFIG_SMP
-#include <linux/smp.h>	/* get_hard_smp_processor_id() */
-#endif
-
 #include "bman_private.h"
 
 /* Last updated for v00.79 of the BG */
diff --git a/drivers/hwalloc/bman_private.h b/drivers/hwalloc/bman_private.h
index bf36060..8b2c5d2 100644
--- a/drivers/hwalloc/bman_private.h
+++ b/drivers/hwalloc/bman_private.h
@@ -30,6 +30,9 @@
  * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
  */
 
+#ifndef CONFIG_SMP
+#include <asm/smp.h>    /* get_hard_smp_processor_id() */
+#endif
 #include "bman_sys.h"
 #include <linux/fsl_bman.h>
 
diff --git a/drivers/hwqueue/qman_config.c b/drivers/hwqueue/qman_config.c
index c3baa03..8381b29 100644
--- a/drivers/hwqueue/qman_config.c
+++ b/drivers/hwqueue/qman_config.c
@@ -30,10 +30,6 @@
  * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
  */
 
-#ifndef CONFIG_SMP
-#include <linux/smp.h>	/* get_hard_smp_processor_id() */
-#endif
-
 #include "qman_private.h"
 
 /* Last updated for v00.800 of the BG */
diff --git a/drivers/hwqueue/qman_private.h b/drivers/hwqueue/qman_private.h
index d85e62b..d158889 100644
--- a/drivers/hwqueue/qman_private.h
+++ b/drivers/hwqueue/qman_private.h
@@ -30,6 +30,9 @@
  * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
  */
 
+#ifndef CONFIG_SMP
+#include <asm/smp.h>    /* get_hard_smp_processor_id() */
+#endif
 #include "qman_sys.h"
 #include <linux/fsl_qman.h>
 
-- 
1.6.5.2

