From 04092ae83e853f32444d5414d729adea0fc2c56f Mon Sep 17 00:00:00 2001
From: Hui Wang <Hui.Wang@windriver.com>
Date: Tue, 1 Mar 2011 18:58:17 +0800
Subject: [PATCH 181/233] fsl_p1022_p1013: make lbc suspend safe for other platforms

Signed-off-by: Hui Wang <Hui.Wang@windriver.com>
---
 arch/powerpc/sysdev/fsl_lbc.c |    5 ++++-
 1 files changed, 4 insertions(+), 1 deletions(-)

diff --git a/arch/powerpc/sysdev/fsl_lbc.c b/arch/powerpc/sysdev/fsl_lbc.c
index a172f00..142dfc7 100644
--- a/arch/powerpc/sysdev/fsl_lbc.c
+++ b/arch/powerpc/sysdev/fsl_lbc.c
@@ -385,8 +385,11 @@ static int __devexit fsl_lbc_ctrl_remove(struct of_device *ofdev)
 #ifdef CONFIG_SUSPEND
 #ifdef CONFIG_P1022_DS
 #define COUNT_OF_BANK_P1022	(8)
-#endif
 #define COUNT_OF_BANKS COUNT_OF_BANK_P1022
+#else
+#define COUNT_OF_BANKS (8)
+#endif
+
 
 /* save lbc registers */
 static int fsl_lbc_suspend(struct of_device *ofdev, pm_message_t state)
-- 
1.7.0.4

