From 85db4a4d527ebb1fe69b8ae73c315858cdd04936 Mon Sep 17 00:00:00 2001
From: Bruce Ashfield <bruce.ashfield@windriver.com>
Date: Mon, 20 Dec 2010 22:31:45 -0500
Subject: [PATCH] match: set CONFIG_PATTERN_MATCHING default to n

Signed-off-by: Bruce Ashfield <bruce.ashfield@windriver.com>
---
 drivers/match/Kconfig |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/drivers/match/Kconfig b/drivers/match/Kconfig
index b0b40ea..e59cc8e 100644
--- a/drivers/match/Kconfig
+++ b/drivers/match/Kconfig
@@ -2,7 +2,7 @@ menu "Pattern matching support"
 
 menuconfig PATTERN_MATCHING
 	bool "Pattern matching h/w support"
-	default y
+	default n
 	---help---
 	  Say Y here to see options for pattern matching drivers.
 	  This option alone does not add any kernel code.
-- 
1.6.5.2

