From 698542e6edcbde152026bad612a8a35078819e5f Mon Sep 17 00:00:00 2001
From: Andrew Liu <shengping.liu@windriver.com>
Date: Tue, 22 Feb 2011 15:21:22 +0800
Subject: [PATCH 13/28] mpc8xxx_gpio: add qoriq-gpio node and its probe code.

Patch taken from FSL vendor SDK 2.2.

In dts, add qoriq gpio node and its reference in mdio node.

Add code to probe qoriq gpio and add it during boot up.

Integrated-by: Andrew Liu <shengping.liu@windriver.com>

[drop p4080ds_36b.dts's change, because this file is platform level.]

Integrated-by: Hui Wang <Hui.Wang@windriver.com>
---
 arch/powerpc/sysdev/mpc8xxx_gpio.c |    3 +++
 1 files changed, 3 insertions(+), 0 deletions(-)

diff --git a/arch/powerpc/sysdev/mpc8xxx_gpio.c b/arch/powerpc/sysdev/mpc8xxx_gpio.c
index 6478eb1..ed9f72c 100644
--- a/arch/powerpc/sysdev/mpc8xxx_gpio.c
+++ b/arch/powerpc/sysdev/mpc8xxx_gpio.c
@@ -186,6 +186,9 @@ static int __init mpc8xxx_add_gpiochips(void)
 	for_each_compatible_node(np, NULL, "fsl,mpc8610-gpio")
 		mpc8xxx_add_controller(np);
 
+	for_each_compatible_node(np, NULL, "fsl,qoriq-gpio")
+		mpc8xxx_add_controller(np);
+
 	return 0;
 }
 arch_initcall(mpc8xxx_add_gpiochips);
-- 
1.6.5.2

