From 611823d93c80e7f4722f48e14958419f44233670 Mon Sep 17 00:00:00 2001
From: Andrew Liu <shengping.liu@windriver.com>
Date: Wed, 4 Aug 2010 14:03:47 +0800
Subject: [PATCH 152/252] Simplfy code in dpa_fd_length calculation and dpa_fd_offset

It is from FSL vendor SDK 2.x:
  Simplfy-code-for-dpa_fd_length-calculation.patch
  Simply-code-in-dpa_fd_offset.patch

Make code more concise and clear.

Integrated-by: Andrew Liu <shengping.liu@windriver.com>
---
 drivers/net/dpa/dpa.c |   23 ++++-------------------
 1 files changed, 4 insertions(+), 19 deletions(-)

diff --git a/drivers/net/dpa/dpa.c b/drivers/net/dpa/dpa.c
index 8fb24a4..a068414 100644
--- a/drivers/net/dpa/dpa.c
+++ b/drivers/net/dpa/dpa.c
@@ -421,34 +421,19 @@ struct dpa_fd {
 static inline ssize_t __const __must_check __attribute__((nonnull))
 dpa_fd_length(const struct qm_fd *fd)
 {
-	switch (fd->format) {
-	case qm_fd_contig:
-	case qm_fd_sg:
+	if ((fd->format & 0x3) == 0)
 		return fd->length20;
-	case qm_fd_contig_big:
-	case qm_fd_sg_big:
+	else
 		return fd->length29;
-	case qm_fd_compound:
-		return fd->cong_weight;
-	}
-	BUG();
-	return -EINVAL;
 }
 
 static inline ssize_t __const __must_check __attribute__((nonnull))
 dpa_fd_offset(const struct qm_fd *fd)
 {
-	switch (fd->format) {
-	case qm_fd_contig:
-	case qm_fd_sg:
+	if ((fd->format == qm_fd_contig) || (fd->format == qm_fd_sg))
 		return fd->offset;
-	case qm_fd_contig_big:
-	case qm_fd_sg_big:
-	case qm_fd_compound:
+	else
 		return 0;
-	}
-	BUG();
-	return -EINVAL;
 }
 
 static int __must_check __attribute__((nonnull))
-- 
1.6.5.2

