From 21085b36cde6716ce1e58fa044303de9ef6332b6 Mon Sep 17 00:00:00 2001
From: Yue Tao <yue.tao@windriver.com>
Date: Sun, 26 May 2013 23:04:34 -0700
Subject: [PATCH 2/7] Modified the SPI initialization

Signed-off-by: Feng Feng <feng.feng@windriver.com>
---
 drivers/spi2/busses/comcerto_spi.c |    3 +++
 1 files changed, 3 insertions(+), 0 deletions(-)

diff --git a/drivers/spi2/busses/comcerto_spi.c b/drivers/spi2/busses/comcerto_spi.c
index 020af0d..1c6bba6 100644
--- a/drivers/spi2/busses/comcerto_spi.c
+++ b/drivers/spi2/busses/comcerto_spi.c
@@ -556,6 +556,9 @@ static struct platform_device comcerto_spi_device = {
  */
 static int __init comcerto_spi_init(void)
 {
+	writel(readl(COMCERTO_GPIO_PIN_SELECT_REG) & ~(SPI_BUS | SPI_SSN), COMCERTO_GPIO_PIN_SELECT_REG);
+	__raw_writel(0x4, COMCERTO_GPIO_INT_CFG_REG);
+
 	if (driver_register(&comcerto_spi_driver)) {
 		printk(KERN_INFO "comcerto_spi: error registering driver\n");
 		goto err0;
-- 
1.7.0

