From 08073ad2ef0e2da8ac45467af1e69e20d887ca2a Mon Sep 17 00:00:00 2001
From: Laurent Pinchart <laurent.pinchart@ideasonboard.com>
Date: Wed, 26 Mar 2014 04:01:45 +0100
Subject: [PATCH 701/827] media: entity: Add has_route entity operation

This patch comes from:
  https://github.com/Xilinx/linux-xlnx.git

The optional operation can be used by entities to report whether two
pads are internally connected.

Signed-off-by: Laurent Pinchart <laurent.pinchart@ideasonboard.com>
Signed-off-by: Michal Simek <michal.simek@xilinx.com>
(cherry picked from commit ce0239f7972b2c512e48d2b1c3781fd8d1c37480)
Signed-off-by: Zumeng Chen <zumeng.chen@windriver.com>
---
 include/media/media-entity.h | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/include/media/media-entity.h b/include/media/media-entity.h
index 70282f5..a1eb400 100644
--- a/include/media/media-entity.h
+++ b/include/media/media-entity.h
@@ -52,12 +52,17 @@ struct media_pad {
  * @link_validate:	Return whether a link is valid from the entity point of
  *			view. The media_entity_pipeline_start() function
  *			validates all links by calling this operation. Optional.
+ * @has_route:		Return whether a route exists inside the entity between
+ *			two given pads. Optional. If the operation isn't
+ *			implemented all pads will be considered as connected.
  */
 struct media_entity_operations {
 	int (*link_setup)(struct media_entity *entity,
 			  const struct media_pad *local,
 			  const struct media_pad *remote, u32 flags);
 	int (*link_validate)(struct media_link *link);
+	bool (*has_route)(struct media_entity *entity, unsigned int pad0,
+			  unsigned int pad1);
 };
 
 struct media_entity {
-- 
2.9.3

