From dd9cb5161ae6936d9d8677f616656da7f27e7ef3 Mon Sep 17 00:00:00 2001
From: Zumeng Chen <zumeng.chen@windriver.com>
Date: Thu, 27 Oct 2016 14:23:49 +0800
Subject: [PATCH 828/830] Revert "can: xilinx: add workaround for the possible
 dts mismatch"

This reverts commit 4fb7f35df5443ec002eee546ed8abdb0bb6b1daf.
---
 drivers/net/can/xilinx_can.c | 12 +-----------
 1 file changed, 1 insertion(+), 11 deletions(-)

diff --git a/drivers/net/can/xilinx_can.c b/drivers/net/can/xilinx_can.c
index 982b3c1..a89f6ee 100644
--- a/drivers/net/can/xilinx_can.c
+++ b/drivers/net/can/xilinx_can.c
@@ -1208,14 +1208,12 @@ static int xcan_open(struct net_device *ndev)
 	struct xcan_priv *priv = netdev_priv(ndev);
 	int ret;
 
-#ifdef CONFIG_PM_AUTOSLEEP
 	ret = pm_runtime_get_sync(priv->dev);
 	if (ret < 0) {
 		netdev_err(ndev, "%s: pm_runtime_get failed(%d)\n",
 				__func__, ret);
 		return ret;
 	}
-#endif
 
 	ret = request_irq(ndev->irq, xcan_interrupt, priv->irq_flags,
 			ndev->name, ndev);
@@ -1253,9 +1251,7 @@ err_candev:
 err_irq:
 	free_irq(ndev->irq, ndev);
 err:
-#ifdef CONFIG_PM_AUTOSLEEP
 	pm_runtime_put(priv->dev);
-#endif
 
 	return ret;
 }
@@ -1277,9 +1273,7 @@ static int xcan_close(struct net_device *ndev)
 	close_candev(ndev);
 
 	can_led_event(ndev, CAN_LED_EVENT_STOP);
-#ifdef CONFIG_PM_AUTOSLEEP
 	pm_runtime_put(priv->dev);
-#endif
 
 	return 0;
 }
@@ -1296,7 +1290,6 @@ static int xcan_get_berr_counter(const struct net_device *ndev,
 					struct can_berr_counter *bec)
 {
 	struct xcan_priv *priv = netdev_priv(ndev);
-#ifdef CONFIG_PM_AUTOSLEEP
 	int ret;
 
 	ret = pm_runtime_get_sync(priv->dev);
@@ -1305,15 +1298,12 @@ static int xcan_get_berr_counter(const struct net_device *ndev,
 				__func__, ret);
 		return ret;
 	}
-#endif
 
 	bec->txerr = priv->read_reg(priv, XCAN_ECR_OFFSET) & XCAN_ECR_TEC_MASK;
 	bec->rxerr = ((priv->read_reg(priv, XCAN_ECR_OFFSET) &
 			XCAN_ECR_REC_MASK) >> XCAN_ESR_REC_SHIFT);
 
-#ifdef CONFIG_PM_AUTOSLEEP
 	pm_runtime_put(priv->dev);
-#endif
 
 	return 0;
 }
@@ -1578,7 +1568,7 @@ static int xcan_probe(struct platform_device *pdev)
 
 	pm_runtime_put(&pdev->dev);
 
-	netdev_info(ndev, "reg_base=0x%p irq=%d clock=%d, tx fifo depth:%d\n",
+	netdev_dbg(ndev, "reg_base=0x%p irq=%d clock=%d, tx fifo depth:%d\n",
 			priv->reg_base, ndev->irq, priv->can.clock.freq,
 			priv->tx_max);
 
-- 
2.9.3

