From foo@baz Sun Feb 12 23:09:28 PST 2012
Date: Sun, 12 Feb 2012 23:09:28 -0800
To: Greg KH <gregkh@linuxfoundation.org>
From: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
Subject: LTSI Makefile addition

Change the extra version to have -ltsi to have a chance to realize what
kernel version we are using.

Signed-off-by: Greg Kroah-Hartman <gregkh@linuxfoundation.org>

---
 Makefile |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

--- a/Makefile
+++ b/Makefile
@@ -1,7 +1,7 @@
 VERSION = 3
 PATCHLEVEL = 4
 SUBLEVEL = 46
-EXTRAVERSION =
+EXTRAVERSION = -ltsi
 NAME = Saber-toothed Squirrel
 
 # *DOCUMENTATION*
