From 094a927c0cc5047cd850c5fb7d8ca78e94d324e8 Mon Sep 17 00:00:00 2001
From: Zumeng Chen <zumeng.chen@windriver.com>
Date: Mon, 25 Jul 2016 19:31:02 -0700
Subject: [PATCH 5/5] arm64: Add ls1043 uImage load address

Signed-off-by: Zumeng Chen <zumeng.chen@windriver.com>
---
 arch/arm64/boot/Makefile | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/arch/arm64/boot/Makefile b/arch/arm64/boot/Makefile
index f1ee036..5c6d861 100644
--- a/arch/arm64/boot/Makefile
+++ b/arch/arm64/boot/Makefile
@@ -33,6 +33,11 @@ ifeq ($(CONFIG_ARCH_AXXIA), y)
   UIMAGE_ENTRYADDR = 0x00080000
 endif
 
+ifeq ($(CONFIG_ARCH_LAYERSCAPE), y)
+UIMAGE_LOADADDR = 0x080080000
+UIMAGE_ENTRYADDR = 0x080080000
+endif
+
 check_for_multiple_loadaddr = \
 if [ $(words $(UIMAGE_LOADADDR)) -ne 1 ]; then \
 	echo 'multiple (or no) load addresses: $(UIMAGE_LOADADDR)'; \
-- 
2.7.4

