From 1e08ae44dea1b673682d03b071f4ca4c71f65a1c Mon Sep 17 00:00:00 2001
From: Ashok Kumar <ashoks@broadcom.com>
Date: Wed, 8 Apr 2015 12:57:33 -0400
Subject: kvm: change pci io port base for guest.

Signed-off-by: Ashok Kumar <ashoks@broadcom.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/pci/pci-xlp.c b/arch/mips/pci/pci-xlp.c
index 4fc5813..381babc 100644
--- a/arch/mips/pci/pci-xlp.c
+++ b/arch/mips/pci/pci-xlp.c
@@ -375,8 +375,18 @@ static int __init pcibios_init(void)
 		}
 	}
 
-	set_io_port_base(CKSEG1);
-	nlm_pci_controller.io_map_base = CKSEG1;
+	/* FIXME: for guest OS, let us set the ioport/iomap base address
+	 * at physical address 0x14000000. For qemu virtio emulation,
+	 * the device ioport address is from 0 to 0xffff. If io_map_base
+	 * is CKSEG1, the ioport address will collide with RAM address 0-0xffff.
+	 */
+	if (is_nlm_guest_os) {
+		set_io_port_base(CKSEG1 + 0x14000000);
+		nlm_pci_controller.io_map_base = CKSEG1 + 0x14000000;
+	} else {
+		set_io_port_base(CKSEG1);
+		nlm_pci_controller.io_map_base = CKSEG1;
+	}
 
 	register_pci_controller(&nlm_pci_controller);
 	pr_info("XLP PCIe Controller %pR%pR.\n", &nlm_pci_io_resource,
-- 
1.7.1

