From 05478f24aa834675a1e1c140f53c861c90f373f6 Mon Sep 17 00:00:00 2001
From: Anusha Nelluri <anushan@broadcom.com>
Date: Wed, 8 Apr 2015 13:05:02 -0400
Subject: kvm: Added device manager guest mode support

Signed-off-by: Anusha Nelluri <anushan@broadcom.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/include/asm/netlogic/kvm_para.h b/arch/mips/include/asm/netlogic/kvm_para.h
index 6910dda..01a45db 100644
--- a/arch/mips/include/asm/netlogic/kvm_para.h
+++ b/arch/mips/include/asm/netlogic/kvm_para.h
@@ -58,6 +58,10 @@ extern int is_nlm_guest_os;
 #define CMD_HALX_CDE_CLEANUP	0XF
 #define CMD_HALX_CMD_POE_ENQ_STORAGE	0x10
 
+/* Device Manager Guest Mode: sub cmds */
+#define CMD_DM_GUEST_INIT_SOC 0x11
+#define CMD_DM_GUEST_REQUEST  0x12
+
 static inline int do_hypcall_p1r1(int num, uint64_t param0, uint64_t *ret0)
 {
 	int	 ret;
diff --git a/arch/mips/kvm-netl/kvm.c b/arch/mips/kvm-netl/kvm.c
index e7cc1f5..d0aeda8 100644
--- a/arch/mips/kvm-netl/kvm.c
+++ b/arch/mips/kvm-netl/kvm.c
@@ -554,6 +554,7 @@ int kvm_arch_vcpu_ioctl_run(struct kvm_vcpu *vcpu, struct kvm_run *run)
 		run->mips_hypcall.cmd = vcpu->arch.mips_hypcall.cmd;
 		run->exit_reason = ret >> 24;
 #ifdef KVM_HYPERCALL_DEBUG
+		printk("KVM_EXIT_HC_HALX_NETSOC. vcpu id %d\n", vcpu->vcpu_id);
 		printk("KVM_EXIT_HC_HALX_NETSOC. run->mips_hypcall.args[0] = 0x%lx\n", run->mips_hypcall.args[0]);
 		printk("KVM_EXIT_HC_HALX_NETSOC. run->exit_reason = 0x%lx, run->mips_hypcall.cmd = %d\n", run->exit_reason, run->mips_hypcall.cmd);
 #endif
diff --git a/arch/mips/netlogic/kvm-netl/kvm_traps.c b/arch/mips/netlogic/kvm-netl/kvm_traps.c
index 7fe7d82..de0baac 100644
--- a/arch/mips/netlogic/kvm-netl/kvm_traps.c
+++ b/arch/mips/netlogic/kvm-netl/kvm_traps.c
@@ -505,7 +505,7 @@ static void process_hc(struct pt_regs *regs)
 		int i;
 		/* TODO: check guest_cpuid, make sure it in regs[3] */
 		/*int guest_cpuid = regs->regs[3];*/
-		int guest_cpuid = 0;
+		int guest_cpuid = regs->guest_cp0_ebase & 0x3ff;
 
 		kvm = ((struct kvm_vcpu *)regs->cp0_osscratch7)->kvm;
 #ifdef KVM_HYPERCALL_DEBUG
@@ -535,6 +535,7 @@ static void process_hc(struct pt_regs *regs)
 		sub_cmd = regs->regs[4];
 		gpa_to_hva((struct kvm_vcpu *)regs->cp0_osscratch7, gpa_addr, &hva_addr);
 #ifdef KVM_HYPERCALL_DEBUG
+		printk("KVM_HC_HALX_NETSOC: vcpu_id %d\n", vcpu->vcpu_id);
 		printk("Tentatively Leaving the guest ...gpa 0x%lx hva 0x%lx\n", gpa_addr, hva_addr);
 		printk("KVM_HC_HALX_NETSOC: sub_cmd = %d\n",  sub_cmd);
 #endif
-- 
1.7.1

