From 0a85cc04b4ad212c0a10cf0d59635727bf9185ab Mon Sep 17 00:00:00 2001
From: Nam Ninh <nam.ninh@windriver.com>
Date: Wed, 8 Apr 2015 14:29:44 -0400
Subject: Fix build warnings

Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/netlogic/lib/fmnlib/nlm_hal_fmn_config.c b/arch/mips/netlogic/lib/fmnlib/nlm_hal_fmn_config.c
index a4be5aa..4df616e 100644
--- a/arch/mips/netlogic/lib/fmnlib/nlm_hal_fmn_config.c
+++ b/arch/mips/netlogic/lib/fmnlib/nlm_hal_fmn_config.c
@@ -1311,7 +1311,7 @@ int nlm_hal_setup_outq(void *fdt, int node, int max_nodes)
 	const int q_ram_pages = 1;
 	const uint32_t q_ram_page_entries = 32; /* entries, not bytes */
 	const uint32_t q_ram_page_entries_9xx = 64; /* entries, not bytes */
-	int cnt =0, oq_num32pages = 0, oq_num32pages_stn , num_valid_queues;
+	int cnt =0, oq_num32pages = 0, oq_num32pages_stn , num_valid_queues = 0;
 	struct fmn_qsize_credit_config *fmn_q_config = nlm_node_cfg.fmn_cfg[node]->fmn_q_config;
 	int single_sae_vc = 0;
 
diff --git a/arch/mips/netlogic/lib/netlib/include/netsoc_nae.h b/arch/mips/netlogic/lib/netlib/include/netsoc_nae.h
index 745c6eb..6d3ae83 100644
--- a/arch/mips/netlogic/lib/netlib/include/netsoc_nae.h
+++ b/arch/mips/netlogic/lib/netlib/include/netsoc_nae.h
@@ -265,7 +265,7 @@ static uint32_t membar_fixup(uint32_t l)
 
 static inline void __netsoc_reset_nae(nae_t *nae)
 {
-        uint64_t bar0;
+        uint64_t bar0 = 0;
         int reset_bit = 9;
         /* Reset NAE */
 
diff --git a/arch/mips/netlogic/lib/netlib/src/netsoc_fdt.c b/arch/mips/netlogic/lib/netlib/src/netsoc_fdt.c
index da7fd45..9dbadfa 100644
--- a/arch/mips/netlogic/lib/netlib/src/netsoc_fdt.c
+++ b/arch/mips/netlogic/lib/netlib/src/netsoc_fdt.c
@@ -396,7 +396,7 @@ static int prepare_nae_shared_info(void *fdt, int domid, nae_t *nae_cfg)
 {
         char dom_node_str[128];
         int nodeoffset;
-        unsigned int *pval, mask = 0x0;
+        unsigned int *pval = NULL, mask = 0x0;
         int plen, i, j;
         struct nlm_nae_shinfo shinfo;
         static int shdomains[NLM_NAE_MAX_SHARED_DOMS], shdom_extracted = 0;
diff --git a/arch/mips/netlogic/lib/netlib/src/netsoc_nae.c b/arch/mips/netlogic/lib/netlib/src/netsoc_nae.c
index 41e10f5..c9cb9e4 100644
--- a/arch/mips/netlogic/lib/netlib/src/netsoc_nae.c
+++ b/arch/mips/netlogic/lib/netlib/src/netsoc_nae.c
@@ -1220,7 +1220,7 @@ void __netsoc_config_context_xoff_thr(nae_t *nae, net_port_t *netport, int thgrp
         /* xon & xoff threshold is based on the number of entries used
          send xon when rx_data_buffer_used <= xon threshold
          send xoff when rx_data_buffer_used >= xoff threshold */
-        int xoff, xon, val, reaction_len, inflight_len;
+        int xoff = 0, xon = 0, val, reaction_len, inflight_len;
 
 	if(netport->vlan_pri_en){
 		if(netport->iftype == XLAUI_IF){
@@ -4257,7 +4257,7 @@ int __netsoc_pcs_init(nae_t *nae)
 static int __netsoc_config_freein_fifo(nae_t *nae)
 {
 	uint32_t reg, size, spillsz, guest_id,parseq_th;
-        int start = 0, i,  th_hi, th_lo, minsz;
+        int start = 0, i,  th_hi, th_lo, minsz = 0;
         uint64_t spill_addr, spill_mem_addr, spill_mem_size;
         int freein_fifo_total_queues = nae->frin_total_queue;
         void *buf;
diff --git a/arch/mips/netlogic/lib/netlib/src/netsoc_parser.c b/arch/mips/netlogic/lib/netlib/src/netsoc_parser.c
index 8299120..28a2651 100644
--- a/arch/mips/netlogic/lib/netlib/src/netsoc_parser.c
+++ b/arch/mips/netlogic/lib/netlib/src/netsoc_parser.c
@@ -119,7 +119,7 @@ int __netsoc_config_enable_flow_table(nae_t *nae, parser_config_t *cfg_parser)
 	uint32_t val;
 	int port = 0;
 	int l2proto = 1;
-	uint32_t dest_vc, max_vc;
+	uint32_t dest_vc = 0, max_vc;
 	int i = 0, cpu = 0, shift = 0;
 	int word_num;
 
diff --git a/drivers/net/ethernet/broadcom/nae/xlpge_board.c b/drivers/net/ethernet/broadcom/nae/xlpge_board.c
index 47278e6..95c817d 100644
--- a/drivers/net/ethernet/broadcom/nae/xlpge_board.c
+++ b/drivers/net/ethernet/broadcom/nae/xlpge_board.c
@@ -45,7 +45,6 @@ void get_mac_address(char * const mac_addr)
 	struct eeprom_data *nlm_eeprom=NULL;
         unsigned char temp,buf_write[2],buf0_read[2],buf1_read[2];
         int if_mac_set=0,mac0_set=0, mac1_set=0;
-        int i,j;
 
         buf_write[0]= MAC_MAGIC_BYTE0;
         buf_write[1]= MAC_MAGIC_BYTE1;
diff --git a/drivers/net/ethernet/broadcom/nae/xlpge_nae.c b/drivers/net/ethernet/broadcom/nae/xlpge_nae.c
index d72429b..0696625 100644
--- a/drivers/net/ethernet/broadcom/nae/xlpge_nae.c
+++ b/drivers/net/ethernet/broadcom/nae/xlpge_nae.c
@@ -823,12 +823,12 @@ void nlm_xlp_nae_remove(void)
 
 static int nlm_xlp_port_poll(void *buff)
 {
-	uint32_t i,num_ports,maxnae,node,speed,duplex;
+	uint32_t i,maxnae,node,speed,duplex;
 	nae_t *nae_cfg;
 	static int arr[NLM_MAX_NODES][18];
 	maxnae = get_num_nae_pernode();
 	for (node = 0;node < NLM_MAX_NODES; node++) {
-		int num_nae, tmp, index=0;
+		int num_nae, index=0;
 		for(num_nae=0; num_nae<maxnae; num_nae++) {
 			nae_cfg = get_nae(node, num_nae);
 			if (nae_cfg == NULL)
@@ -984,11 +984,11 @@ static int p2p_desc_mem_init(void)
 
 static void nlm_enable_l3_l4_parser(nae_t *nae)
 {
-	nae_print(NAE_DBG_TRACE, "Enabling parser for nae node %d naeid %d\n",
-			nae->node, nae->nae_id);
 	l2_parser_config_t l2;
 	l3_parser_config_t l3;
 	l4_parser_config_t l4;
+	nae_print(NAE_DBG_TRACE, "Enabling parser for nae node %d naeid %d\n",
+			nae->node, nae->nae_id);
 
 	memset(&l2, 0, sizeof(l2));
 	memset(&l3, 0, sizeof(l3));
@@ -1151,11 +1151,10 @@ static int  nlm_xlp_nae_stop (struct net_device *dev)
  * nlm_xlp_change_rx_flags - called when promiscuous mode is enabled/disabled
  *
  ****************************************************************************/
-static void  nlm_xlp_change_rx_flags (struct net_device *dev)
+static void  nlm_xlp_change_rx_flags (struct net_device *dev, int flags)
 {
 	struct dev_data *priv = netdev_priv(dev);
         net_port_t* nae_port = priv->nae_port;
-        int reg_val=0;
         if ((dev->flags & IFF_PROMISC)){
                netsoc_promisc_enable(nae_port);}
 
diff --git a/drivers/net/ethernet/broadcom/nae/xlpge_rx.c b/drivers/net/ethernet/broadcom/nae/xlpge_rx.c
index b74309e..15df48f 100644
--- a/drivers/net/ethernet/broadcom/nae/xlpge_rx.c
+++ b/drivers/net/ethernet/broadcom/nae/xlpge_rx.c
@@ -72,8 +72,6 @@ void fill_alloc_failed(struct work_struct *work)
 	int cpu = smp_processor_id();
 	unsigned int frq_num, ret, abs_frq_num, buf_size;
 	struct sk_buff * skb;
-	int nodes = nlm_hal_get_num_nodes();
-	int nae_units = get_num_nae_pernode();
 	struct  dev_data *priv = container_of(work, struct dev_data, alloc_failed_work[cpu].work);
 	net_port_t *netport = priv->nae_port;	
 	nae_t* nae = netport->nae;
@@ -128,8 +126,8 @@ static inline int mac_refill_frin_one_buffer(struct net_device *dev, int cpu,
 	int nae_id = nae_cfg->nae_id;
 	int node = nae_cfg->node;
 	int hw_port_id = priv->hw_port_id; 
-	node_cpu = phys_cpu_to_log_map[node][nae_id][cpu];
 	int lcpu = smp_processor_id();
+	node_cpu = phys_cpu_to_log_map[node][nae_id][cpu];
 	
 	if (enable_jumbo)
 		if(truesize > NLM_RX_JUMBO_BUF_SIZE)
@@ -748,7 +746,9 @@ static inline void process_rx_packets(void *arg, int cpu, unsigned int src_id,
 	volatile u64 regval;
     struct skb_shared_hwtstamps *shhwtstamps;
 #endif
+#if 0
 	struct napi_struct *napi = (struct napi_struct *)arg;
+#endif
 
 	if(nae_family_9xx){
 		msg1 = msg2;	
-- 
1.7.1

