From 9a53066337fc1a057caf946c9dcc69fa807654ef Mon Sep 17 00:00:00 2001
From: Denis Oliver Kropp <dok@directfb.org>
Date: Fri, 1 Nov 2002 15:13:02 +0000
Subject: [PATCH] Use ANSI C member initialization.

---
 drivers/char/fusion/fusiondev.c |   14 +++++++-------
 1 files changed, 7 insertions(+), 7 deletions(-)

diff --git a/drivers/char/fusion/fusiondev.c b/drivers/char/fusion/fusiondev.c
index 023ac7d..2cae47e 100644
--- a/drivers/char/fusion/fusiondev.c
+++ b/drivers/char/fusion/fusiondev.c
@@ -142,16 +142,16 @@ fusion_ioctl (struct inode *inode, struct file *file,
 }
 
 static struct file_operations fusion_fops = {
-  owner:    THIS_MODULE,
-  ioctl:    fusion_ioctl,
-  open:     fusion_open,
-  release:  fusion_release,
+  .owner   = THIS_MODULE,
+  .ioctl   = fusion_ioctl,
+  .open    = fusion_open,
+  .release = fusion_release,
 };
 
 static struct miscdevice fusion_miscdev = {
-  minor:    FUSION_MINOR,
-  name:     "fusion",
-  fops:     &fusion_fops,
+  .minor   = FUSION_MINOR,
+  .name    = "fusion",
+  .fops    = &fusion_fops,
 };
 
 /******************************************************************************/
-- 
1.7.3.3

