From dcae75ab5434e946c8b2c2b643fa4ef4c27de833 Mon Sep 17 00:00:00 2001
From: Denis Oliver Kropp <dok@directfb.org>
Date: Sat, 2 Nov 2002 03:26:59 +0000
Subject: [PATCH] fusion_init/exit non-static now.

---
 drivers/char/fusion/fusiondev.c |    4 ++--
 1 files changed, 2 insertions(+), 2 deletions(-)

diff --git a/drivers/char/fusion/fusiondev.c b/drivers/char/fusion/fusiondev.c
index 6228429..9344c9f 100644
--- a/drivers/char/fusion/fusiondev.c
+++ b/drivers/char/fusion/fusiondev.c
@@ -185,7 +185,7 @@ static struct miscdevice fusion_miscdev = {
 
 /******************************************************************************/
 
-static int __init
+int __init
 fusion_init(void)
 {
   int ret;
@@ -224,7 +224,7 @@ fusion_init(void)
   return ret;
 }
 
-static void __exit
+void __exit
 fusion_exit(void)
 {
   misc_deregister (&fusion_miscdev);
-- 
1.7.3.3

