From 465f8bccf61645325ed2fce39d4f58cfa728f408 Mon Sep 17 00:00:00 2001
From: Denis Oliver Kropp <dok@directfb.org>
Date: Tue, 4 Feb 2003 20:03:14 +0000
Subject: [PATCH] Plugged a very small mem leak.

---
 drivers/char/fusion/reactor.c |    5 ++++-
 1 files changed, 4 insertions(+), 1 deletions(-)

diff --git a/drivers/char/fusion/reactor.c b/drivers/char/fusion/reactor.c
index 9115c72..2f2e48f 100644
--- a/drivers/char/fusion/reactor.c
+++ b/drivers/char/fusion/reactor.c
@@ -225,7 +225,10 @@ fusion_reactor_detach (int id, int fusion_id)
     }
 
   if (! --node->count)
-    fusion_list_remove (&reactor->nodes, &node->link);
+    {
+      fusion_list_remove (&reactor->nodes, &node->link);
+      kfree (node);
+    }
 
   unlock_reactor (reactor);
 
-- 
1.7.3.3

