From 2efd0ebf82858f2815101bb38d57071248efaeda Mon Sep 17 00:00:00 2001
From: Denis Oliver Kropp <dok@directfb.org>
Date: Mon, 3 Mar 2003 16:18:28 +0000
Subject: [PATCH] Don't free already freed message data during fusionee_destroy().

---
 drivers/char/fusion/fusionee.c |    1 -
 1 files changed, 0 insertions(+), 1 deletions(-)

diff --git a/drivers/char/fusion/fusionee.c b/drivers/char/fusion/fusionee.c
index 77dbafb..c50e8d2 100644
--- a/drivers/char/fusion/fusionee.c
+++ b/drivers/char/fusion/fusionee.c
@@ -385,7 +385,6 @@ fusionee_destroy (int id)
     {
       Message *message = (Message*) fusion_fifo_get (&fusionee->messages);
 
-      kfree (message->data);
       kfree (message);
     }
 
-- 
1.7.3.3

