From 58415d1f4099d5b371e3effde9492264cd9f5725 Mon Sep 17 00:00:00 2001
From: Andreas Hundt <andi@directfb.org>
Date: Tue, 8 Apr 2003 09:39:23 +0000
Subject: [PATCH] fix compilation for 2.5.xx

---
 drivers/char/fusion/fusiondev.c |   14 ++++++++++++++
 1 files changed, 14 insertions(+), 0 deletions(-)

diff --git a/drivers/char/fusion/fusiondev.c b/drivers/char/fusion/fusiondev.c
index fe35b8d..32af8ed 100644
--- a/drivers/char/fusion/fusiondev.c
+++ b/drivers/char/fusion/fusiondev.c
@@ -59,6 +59,19 @@ fusion_sleep_on(wait_queue_head_t *q, spinlock_t *lock)
 
   current->state = TASK_INTERRUPTIBLE;
 
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(2, 5, 0)
+  write_lock_irqsave (&q->lock,flags);
+  __add_wait_queue (q, &wait);
+  write_unlock (&q->lock);
+
+  spin_unlock (lock);
+
+  schedule();
+
+  write_lock_irq (&q->lock);
+  __remove_wait_queue (q, &wait);
+  write_unlock_irqrestore (&q->lock,flags);
+#else
   wq_write_lock_irqsave (&q->lock,flags);
   __add_wait_queue (q, &wait);
   wq_write_unlock (&q->lock);
@@ -70,6 +83,7 @@ fusion_sleep_on(wait_queue_head_t *q, spinlock_t *lock)
   wq_write_lock_irq (&q->lock);
   __remove_wait_queue (q, &wait);
   wq_write_unlock_irqrestore (&q->lock,flags);
+#endif
 }
 
 /******************************************************************************/
-- 
1.7.3.3

