From f5abf3b483ae0338b1875913b3af0af087ef655c Mon Sep 17 00:00:00 2001
From: Andreas Hundt <andi@directfb.org>
Date: Tue, 8 Apr 2003 10:00:39 +0000
Subject: [PATCH] forgot to include <linux/version.h>

---
 drivers/char/fusion/fusiondev.c |    1 +
 1 files changed, 1 insertions(+), 0 deletions(-)

diff --git a/drivers/char/fusion/fusiondev.c b/drivers/char/fusion/fusiondev.c
index 32af8ed..4c85bad 100644
--- a/drivers/char/fusion/fusiondev.c
+++ b/drivers/char/fusion/fusiondev.c
@@ -12,6 +12,7 @@
  *	2 of the License, or (at your option) any later version.
  */
  
+#include <linux/version.h>
 #include <linux/module.h>
 #include <linux/config.h>
 #include <linux/types.h>
-- 
1.7.3.3

