From 77f597949ace3ac8765aca2dd39acdb4a6609bc2 Mon Sep 17 00:00:00 2001
From: Denis Oliver Kropp <dok@directfb.org>
Date: Mon, 12 May 2003 16:48:18 +0000
Subject: [PATCH] Increased the purchase timeout to one second. Don't wait if property is purchased for more than one second.

Timeout of leasing stays at 0.1 sec, not waiting at all if property is
purchased for more than 0.1 sec.
---
 drivers/char/fusion/property.c |    4 ++--
 1 files changed, 2 insertions(+), 2 deletions(-)

diff --git a/drivers/char/fusion/property.c b/drivers/char/fusion/property.c
index ea00b38..c0a8541 100644
--- a/drivers/char/fusion/property.c
+++ b/drivers/char/fusion/property.c
@@ -273,14 +273,14 @@ fusion_property_purchase (int id, int fusion_id)
           switch (timeout)
             {
             case -1:
-              if (jiffies - property->purchase_stamp > HZ / 10)
+              if (jiffies - property->purchase_stamp > HZ)
                 {
                 case 0:
                   unlock_property (property);
                   return -EAGAIN;
                 }
               else
-                timeout = HZ / 10;
+                timeout = HZ;
 
               /* fall through */
 
-- 
1.7.3.3

