From 4f19deefd7c25514308fcc924de529868c2b830c Mon Sep 17 00:00:00 2001
From: Denis Oliver Kropp <dok@directfb.org>
Date: Mon, 12 May 2003 17:13:36 +0000
Subject: [PATCH] Fixed missing arguments.

---
 drivers/char/fusion/fusiondev.c |    6 +++---
 1 files changed, 3 insertions(+), 3 deletions(-)

diff --git a/drivers/char/fusion/fusiondev.c b/drivers/char/fusion/fusiondev.c
index a04a558..0cca194 100644
--- a/drivers/char/fusion/fusiondev.c
+++ b/drivers/char/fusion/fusiondev.c
@@ -285,19 +285,19 @@ fusion_ioctl (struct inode *inode, struct file *file,
       if (get_user (id, (int*) arg))
         return -EFAULT;
 
-      return fusion_ref_zero_lock (id);
+      return fusion_ref_zero_lock (id, fusion_id);
 
     case FUSION_REF_ZERO_TRYLOCK:
       if (get_user (id, (int*) arg))
         return -EFAULT;
 
-      return fusion_ref_zero_trylock (id);
+      return fusion_ref_zero_trylock (id, fusion_id);
 
     case FUSION_REF_UNLOCK:
       if (get_user (id, (int*) arg))
         return -EFAULT;
 
-      return fusion_ref_unlock (id);
+      return fusion_ref_unlock (id, fusion_id);
 
     case FUSION_REF_STAT:
       if (get_user (id, (int*) arg))
-- 
1.7.3.3

