From dd007a3184075950a59aa2ab6ccbec58e2f3b068 Mon Sep 17 00:00:00 2001
From: Denis Oliver Kropp <dok@directfb.org>
Date: Tue, 13 May 2003 13:17:16 +0000
Subject: [PATCH] Do yield() in fusion_property_cede() if property was purchased.

---
 drivers/char/fusion/property.c |    6 ++++++
 1 files changed, 6 insertions(+), 0 deletions(-)

diff --git a/drivers/char/fusion/property.c b/drivers/char/fusion/property.c
index c0a8541..12e6077 100644
--- a/drivers/char/fusion/property.c
+++ b/drivers/char/fusion/property.c
@@ -304,6 +304,7 @@ fusion_property_purchase (int id, int fusion_id)
 int
 fusion_property_cede (int id, int fusion_id)
 {
+  bool            purchased;
   FusionProperty *property = lock_property (id);
 
   if (!property)
@@ -315,6 +316,8 @@ fusion_property_cede (int id, int fusion_id)
       return -EIO;
     }
 
+  purchased = (property->state == FUSION_PROPERTY_PURCHASED);
+
   property->state     = FUSION_PROPERTY_AVAILABLE;
   property->fusion_id = 0;
 
@@ -322,6 +325,9 @@ fusion_property_cede (int id, int fusion_id)
 
   unlock_property (property);
 
+  if (purchased)
+    yield();
+
   return 0;
 }
 
-- 
1.7.3.3

