From 2dbda1ee2038bef74afcb881b0b279d3749451d1 Mon Sep 17 00:00:00 2001
From: Denis Oliver Kropp <dok@directfb.org>
Date: Tue, 24 Jun 2003 16:03:43 +0000
Subject: [PATCH] Show lock-holder (pid and fid) in /proc/fusion.

---
 drivers/char/fusion/skirmish.c |   12 +++++++++---
 1 files changed, 9 insertions(+), 3 deletions(-)

diff --git a/drivers/char/fusion/skirmish.c b/drivers/char/fusion/skirmish.c
index 3838215..7e6df63 100644
--- a/drivers/char/fusion/skirmish.c
+++ b/drivers/char/fusion/skirmish.c
@@ -63,9 +63,15 @@ skirmishs_read_proc(char *buf, char **start, off_t offset,
      fusion_list_foreach (l, dev->skirmish.list) {
           FusionSkirmish *skirmish = (FusionSkirmish*) l;
 
-          written += sprintf(buf+written, "(%5d) 0x%08x %s\n",
-                             skirmish->pid, skirmish->id,
-                             skirmish->lock_fid ? "(locked)" : "");
+          if (skirmish->lock_fid) {
+               written += sprintf(buf+written, "(%5d) 0x%08x (locked 0x%08x "
+                                  "%d)\n", skirmish->pid, skirmish->id,
+                                  skirmish->lock_fid, skirmish->lock_pid);
+          }
+          else {
+               written += sprintf(buf+written, "(%5d) 0x%08x\n",
+                                  skirmish->pid, skirmish->id);
+          }
           if (written < offset) {
                offset -= written;
                written = 0;
-- 
1.7.3.3

