From 0884131c8afed170afda9f4fe3055eb317c77fda Mon Sep 17 00:00:00 2001
From: Denis Oliver Kropp <dok@directfb.org>
Date: Mon, 30 Jun 2003 16:41:12 +0000
Subject: [PATCH] Fixed undeclared fusionees_lock, which didn't cause an error if spinlocks are not used.

---
 drivers/char/fusion/fusionee.c |    6 +++---
 1 files changed, 3 insertions(+), 3 deletions(-)

diff --git a/drivers/char/fusion/fusionee.c b/drivers/char/fusion/fusionee.c
index bae0921..bdcf18d 100644
--- a/drivers/char/fusion/fusionee.c
+++ b/drivers/char/fusion/fusionee.c
@@ -349,7 +349,7 @@ fusionee_kill (FusionDev *dev, int id, int target, int signal, int timeout_ms)
           if (timeout_ms) {
                switch (timeout) {
                     case 0:  /* timed out */
-                         spin_unlock (&fusionees_lock);
+                         spin_unlock (&dev->fusionee.lock);
                          return -ETIMEDOUT;
 
                     case -1: /* setup timeout */
@@ -372,7 +372,7 @@ fusionee_kill (FusionDev *dev, int id, int target, int signal, int timeout_ms)
                return -ERESTARTSYS;
      }
 
-     spin_unlock (&fusionees_lock);
+     spin_unlock (&dev->fusionee.lock);
 
      return 0;
 }
@@ -391,7 +391,7 @@ fusionee_destroy (FusionDev *dev, int id)
 
      wake_up_interruptible_all (&dev->fusionee.wait);
 
-     spin_unlock (&fusionees_lock);
+     spin_unlock (&dev->fusionee.lock);
 
 
      fusion_call_destroy_all (dev, id);
-- 
1.7.3.3

