From 6b42971a96db2c10e6681dacbcb8d49bd2a0445f Mon Sep 17 00:00:00 2001
From: Denis Oliver Kropp <dok@directfb.org>
Date: Mon, 7 Jul 2003 18:03:31 +0000
Subject: [PATCH] Removed call handler type definition.

---
 include/linux/fusion.h |   10 ++--------
 1 files changed, 2 insertions(+), 8 deletions(-)

diff --git a/include/linux/fusion.h b/include/linux/fusion.h
index b16be3c..20d2b9e 100644
--- a/include/linux/fusion.h
+++ b/include/linux/fusion.h
@@ -47,16 +47,10 @@ typedef struct {
 /*
  * Calling (synchronous RPC)
  */
-typedef int (*FusionCallHandler) (int   caller,   /* fusion id of the caller */
-                                  int   call_arg, /* optional call parameter */
-                                  void *call_ptr, /* optional call parameter */
-                                  void *ctx       /* optional handler context */
-                                  );
-
 typedef struct {
   int                call_id;   /* new call id returned */
 
-  FusionCallHandler  handler;   /* function pointer of handler to install */
+  void              *handler;   /* function pointer of handler to install */
   void              *ctx;       /* optional handler context */
 } FusionCallNew;
 
@@ -76,7 +70,7 @@ typedef struct {
 } FusionCallReturn;
 
 typedef struct {
-  FusionCallHandler  handler;   /* function pointer of handler to call */
+  void              *handler;   /* function pointer of handler to call */
   void              *ctx;       /* optional handler context */
 
   int                caller;    /* fusion id of the caller */
-- 
1.7.3.3

