From a1c6f411c8b5fd296f44b8f3cb02fdae405eaeca Mon Sep 17 00:00:00 2001
From: Denis Oliver Kropp <dok@directfb.org>
Date: Tue, 8 Jul 2003 13:42:16 +0000
Subject: [PATCH] Forgot to convert a FusionCallHandler to void*.

---
 drivers/char/fusion/call.c |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/drivers/char/fusion/call.c b/drivers/char/fusion/call.c
index 3f02d37..3bae945 100644
--- a/drivers/char/fusion/call.c
+++ b/drivers/char/fusion/call.c
@@ -48,7 +48,7 @@ typedef struct {
      int                pid;       /* owner pid */
      int                fusion_id; /* owner fusion id */
 
-     FusionCallHandler  handler;
+     void              *handler;
      void              *ctx;
 
      FusionLink          *executions;      /* prepending! */
-- 
1.7.3.3

