From 1f572d968f2405d7a2320e84a2c16f2aabcd0da8 Mon Sep 17 00:00:00 2001
From: Andreas Hundt <andi@directfb.org>
Date: Thu, 24 Jul 2003 00:13:33 +0000
Subject: [PATCH] fix typo ;-)

---
 drivers/char/fusion/fusiondev.c |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/drivers/char/fusion/fusiondev.c b/drivers/char/fusion/fusiondev.c
index b66ed9a..9d8c57b 100644
--- a/drivers/char/fusion/fusiondev.c
+++ b/drivers/char/fusion/fusiondev.c
@@ -679,7 +679,7 @@ fusion_exit(void)
      remove_proc_entry ("fusion", NULL);
 
 #if LINUX_VERSION_CODE >= KERNEL_VERSION(2, 6, 0)
-     devfs_remove ("sound");
+     devfs_remove ("fusion");
 #endif
 }
 
-- 
1.7.3.3

