From b9132e18a80dcdea67421a00d65a2777fa912d79 Mon Sep 17 00:00:00 2001
From: Andreas Hundt <andi@directfb.org>
Date: Thu, 24 Jul 2003 11:38:44 +0000
Subject: [PATCH] fix oops during initialization on 2.6.0-test1

---
 drivers/char/fusion/fusiondev.c |    8 ++++----
 1 files changed, 4 insertions(+), 4 deletions(-)

diff --git a/drivers/char/fusion/fusiondev.c b/drivers/char/fusion/fusiondev.c
index 9d8c57b..9483c15 100644
--- a/drivers/char/fusion/fusiondev.c
+++ b/drivers/char/fusion/fusiondev.c
@@ -632,15 +632,15 @@ fusion_init(void)
 {
      int ret;
 
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(2, 6, 0)
+     devfs_mk_dir("fusion");
+#endif
+
      ret = register_devices();
      if (ret)
           return ret;
 
      proc_fusion_dir = proc_mkdir ("fusion", NULL);
-
-#if LINUX_VERSION_CODE >= KERNEL_VERSION(2, 6, 0)
-     devfs_mk_dir("fusion");
-#endif
      
      return 0;
 }
-- 
1.7.3.3

