From b2b784f83005c2fea686bfb80cbdb252f8a64eb1 Mon Sep 17 00:00:00 2001
From: Denis Oliver Kropp <dok@directfb.org>
Date: Fri, 24 Oct 2003 09:50:48 +0000
Subject: [PATCH] More padding in /proc/fusion/*/fusionees.

---
 drivers/char/fusion/fusionee.c |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/drivers/char/fusion/fusionee.c b/drivers/char/fusion/fusionee.c
index 02fb314..44e894b 100644
--- a/drivers/char/fusion/fusionee.c
+++ b/drivers/char/fusion/fusionee.c
@@ -84,7 +84,7 @@ fusionees_read_proc(char *buf, char **start, off_t offset,
      fusion_list_foreach (l, dev->fusionee.list) {
           Fusionee *fusionee = (Fusionee*) l;
 
-          written += sprintf(buf+written, "(%5d) 0x%08x (%4d messages waiting, %5d received, %5d sent)\n",
+          written += sprintf(buf+written, "(%5d) 0x%08x (%4d messages waiting, %7d received, %7d sent)\n",
                              fusionee->pid, fusionee->id, fusionee->messages.count, fusionee->rcv_total, fusionee->snd_total);
           if (written < offset) {
                offset -= written;
-- 
1.7.3.3

