From 4f147ebf20c9723a57c5684aff9bbc8a3779dcbf Mon Sep 17 00:00:00 2001
From: Denis Oliver Kropp <dok@directfb.org>
Date: Mon, 16 Feb 2004 12:27:37 +0000
Subject: [PATCH] Use major 252.

---
 drivers/char/fusion/fusiondev.c |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/drivers/char/fusion/fusiondev.c b/drivers/char/fusion/fusiondev.c
index 49a0b16..3c1fd2e 100644
--- a/drivers/char/fusion/fusiondev.c
+++ b/drivers/char/fusion/fusiondev.c
@@ -42,7 +42,7 @@
 #endif
 
 #ifndef FUSION_MAJOR
-#define FUSION_MAJOR 253
+#define FUSION_MAJOR 252
 #endif
 
 MODULE_LICENSE("GPL");
-- 
1.7.3.3

