From 5d164ea5b6d2b769f2ea5ed451261c713b17326d Mon Sep 17 00:00:00 2001
From: Denis Oliver Kropp <dok@directfb.org>
Date: Sun, 7 Nov 2004 12:20:53 +0000
Subject: [PATCH] Fixed duplicate watch notification sent during cleanup.

---
 drivers/char/fusion/ref.c |    7 +++++--
 1 files changed, 5 insertions(+), 2 deletions(-)

diff --git a/drivers/char/fusion/ref.c b/drivers/char/fusion/ref.c
index 867da9b..36a451b 100644
--- a/drivers/char/fusion/ref.c
+++ b/drivers/char/fusion/ref.c
@@ -626,14 +626,17 @@ clear_local (FusionDev *dev, FusionRef *ref, int fusion_id)
 
      down (&ref->lock);
 
-     if (ref->locked == fusion_id)
+     if (ref->locked == fusion_id) {
           ref->locked = 0;
+          wake_up_interruptible_all (&ref->wait);
+     }
 
      fusion_list_foreach (l, ref->local_refs) {
           LocalRef *local = (LocalRef *) l;
 
           if (local->fusion_id == fusion_id) {
-               propagate_local( dev, ref, - local->refs );
+               if (local->refs)
+                    propagate_local( dev, ref, - local->refs );
 
                fusion_list_remove( &ref->local_refs, l );
 
-- 
1.7.3.3

