From 973994fffead25af052f0b4fd81c29ee6ca50a10 Mon Sep 17 00:00:00 2001
From: Ville Syrjala <syrjala@directfb.org>
Date: Fri, 26 Nov 2004 10:16:28 +0000
Subject: [PATCH] MINOR() -> iminor().

---
 drivers/char/fusion/fusiondev.c |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/drivers/char/fusion/fusiondev.c b/drivers/char/fusion/fusiondev.c
index 1818e29..10f3e2f 100644
--- a/drivers/char/fusion/fusiondev.c
+++ b/drivers/char/fusion/fusiondev.c
@@ -790,7 +790,7 @@ fusion_ioctl (struct inode *inode, struct file *file,
               unsigned int cmd, unsigned long arg)
 {
      int        id  = (int) file->private_data;
-     FusionDev *dev = fusion_devs[MINOR(inode->i_rdev)];
+     FusionDev *dev = fusion_devs[iminor(inode)];
 
      DEBUG( "fusion_ioctl (0x%08x)\n", cmd );
 
-- 
1.7.3.3

