From c91c73deae02335c8de94a39b1d9125fcad51c5c Mon Sep 17 00:00:00 2001
From: Ville Syrjala <syrjala@directfb.org>
Date: Fri, 26 Nov 2004 10:23:15 +0000
Subject: [PATCH] xtime is struct timespec on some kernels.

---
 drivers/char/fusion/entries.c |    5 +++++
 1 files changed, 5 insertions(+), 0 deletions(-)

diff --git a/drivers/char/fusion/entries.c b/drivers/char/fusion/entries.c
index db822b0..5a30a96 100644
--- a/drivers/char/fusion/entries.c
+++ b/drivers/char/fusion/entries.c
@@ -331,7 +331,12 @@ fusion_entry_lock( FusionEntries  *entries,
 
      /* Keep timestamp, but use the slightly
         inexact version to avoid performance impacts. */
+#ifdef _STRUCT_TIMESPEC
+     entry->last_lock.tv_sec = xtime.tv_sec;
+     entry->last_lock.tv_usec = xtime.tv_nsec / 1000;
+#else
      entry->last_lock = xtime;
+#endif
 
      /* Unlock entries. */
      up( &entries->lock );
-- 
1.7.3.3

