From 18cdf37d085698b2ed21c5471816036342a8da81 Mon Sep 17 00:00:00 2001
From: Denis Oliver Kropp <dok@directfb.org>
Date: Tue, 13 Sep 2005 14:32:25 +0000
Subject: [PATCH] Thanks to "Rob Shortt <rob@tvcentric.com>" for this patch fixing build with Linux 2.6.13 that introduced some changes to the device class API.

---
 drivers/char/fusion/fusiondev.c |   22 ++++++++++++++++++++++
 1 files changed, 22 insertions(+), 0 deletions(-)

diff --git a/drivers/char/fusion/fusiondev.c b/drivers/char/fusion/fusiondev.c
index 10f3e2f..4e4d979 100644
--- a/drivers/char/fusion/fusiondev.c
+++ b/drivers/char/fusion/fusiondev.c
@@ -68,8 +68,12 @@ static inline unsigned iminor(struct inode *inode)
 #endif
 
 #if LINUX_VERSION_CODE >= KERNEL_VERSION(2, 6, 2)
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(2, 6, 13)
+static struct class *fusion_class;
+#else
 static struct class_simple *fusion_class;
 #endif
+#endif
 
 /******************************************************************************/
 
@@ -844,7 +848,11 @@ register_devices(void)
      }
 
 #if LINUX_VERSION_CODE >= KERNEL_VERSION(2, 6, 2)
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(2, 6, 13)
+     fusion_class = class_create (THIS_MODULE, "fusion");
+#else
      fusion_class = class_simple_create (THIS_MODULE, "fusion");
+#endif
      if (IS_ERR(fusion_class)) {
           unregister_chrdev (FUSION_MAJOR, "fusion");
           return PTR_ERR(fusion_class);
@@ -855,10 +863,16 @@ register_devices(void)
 
      for (i=0; i<NUM_MINORS; i++) {
 #if LINUX_VERSION_CODE >= KERNEL_VERSION(2, 6, 2)
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(2, 6, 13)
+          class_device_create (fusion_class,
+                               MKDEV(FUSION_MAJOR, i),
+                               NULL, "fusion%d", i);
+#else
           class_simple_device_add (fusion_class,
                                    MKDEV(FUSION_MAJOR, i),
                                    NULL, "fusion%d", i);
 #endif
+#endif
 
           devfs_mk_cdev (MKDEV(FUSION_MAJOR, i),
                          S_IFCHR | S_IRUSR | S_IWUSR,
@@ -918,15 +932,23 @@ deregister_devices(void)
 
      for (i=0; i<NUM_MINORS; i++) {
 #if LINUX_VERSION_CODE >= KERNEL_VERSION(2, 6, 2)
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(2, 6, 13)
+          class_device_destroy (fusion_class, MKDEV(FUSION_MAJOR, i));
+#else
           class_simple_device_remove (MKDEV(FUSION_MAJOR, i));
 #endif
+#endif
 
           devfs_remove ("fusion/%d", i);
      }
 
 #if LINUX_VERSION_CODE >= KERNEL_VERSION(2, 6, 2)
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(2, 6, 13)
+     class_destroy (fusion_class);
+#else
      class_simple_destroy (fusion_class);
 #endif
+#endif
 
      devfs_remove ("fusion");
 }
-- 
1.7.3.3

