From 8df5110bccb9e714a2d02db7ccea296c565c19b2 Mon Sep 17 00:00:00 2001
From: Denis Oliver Kropp <dok@directfb.org>
Date: Sat, 29 Oct 2005 05:10:43 +0000
Subject: [PATCH] Make it build on 2.4, let's see if it works, too :)

---
 drivers/char/fusion/fusiondev.c |    6 ++++++
 1 files changed, 6 insertions(+), 0 deletions(-)

diff --git a/drivers/char/fusion/fusiondev.c b/drivers/char/fusion/fusiondev.c
index c78d0d0..1158eae 100644
--- a/drivers/char/fusion/fusiondev.c
+++ b/drivers/char/fusion/fusiondev.c
@@ -955,9 +955,15 @@ fusion_mmap( struct file           *file,
           SetPageReserved( virt_to_page(dev->shared_area) );
      }
 
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(2, 6, 0)
      return remap_pfn_range( vma, vma->vm_start,
                              virt_to_phys((void*)dev->shared_area) >> PAGE_SHIFT,
                              PAGE_SIZE, vma->vm_page_prot );
+#else
+     return io_remap_page_range( vma->vm_start,
+                                 virt_to_phys((void*)dev->shared_area),
+                                 PAGE_SIZE, vma->vm_page_prot );
+#endif
 }
 
 static struct file_operations fusion_fops = {
-- 
1.7.3.3

