From d7268aafd40881c5dd06634515602d69c817a12e Mon Sep 17 00:00:00 2001
From: Denis Oliver Kropp <dok@directfb.org>
Date: Sat, 29 Oct 2005 05:11:10 +0000
Subject: [PATCH] Added .tmp_versions.

---
 drivers/char/fusion/.cvsignore |    1 +
 1 files changed, 1 insertions(+), 0 deletions(-)

diff --git a/drivers/char/fusion/.cvsignore b/drivers/char/fusion/.cvsignore
index a4abad5..dbae693 100644
--- a/drivers/char/fusion/.cvsignore
+++ b/drivers/char/fusion/.cvsignore
@@ -3,3 +3,4 @@
 Makefile
 fusion.ko
 fusion.mod.c
+.tmp_versions
-- 
1.7.3.3

