From 8ad803c45eddc34b0142ad0cbcbe860bc362c419 Mon Sep 17 00:00:00 2001
From: Ville Syrjala <syrjala@directfb.org>
Date: Wed, 2 Nov 2005 10:31:42 +0000
Subject: [PATCH] Clear the reserved bit from shared_area before freeing it. Otherwise the kernel will complain and become tainted.

---
 drivers/char/fusion/fusiondev.c |    4 +++-
 1 files changed, 3 insertions(+), 1 deletions(-)

diff --git a/drivers/char/fusion/fusiondev.c b/drivers/char/fusion/fusiondev.c
index 1158eae..2eb133c 100644
--- a/drivers/char/fusion/fusiondev.c
+++ b/drivers/char/fusion/fusiondev.c
@@ -245,8 +245,10 @@ fusiondev_deinit (FusionDev *dev)
      fusion_ref_deinit (dev);
      fusionee_deinit (dev);
 
-     if (dev->shared_area)
+     if (dev->shared_area) {
+          ClearPageReserved( virt_to_page(dev->shared_area) );
           free_page( dev->shared_area );
+     }
 }
 
 /******************************************************************************/
-- 
1.7.3.3

