From 2c5c16a60d6d0ad99d72a102b0203b11489cf8c0 Mon Sep 17 00:00:00 2001
From: Ville Syrjala <syrjala@directfb.org>
Date: Wed, 2 Nov 2005 11:23:28 +0000
Subject: [PATCH] Made fusion_entry_wait() clear entry->lock_pid before calling fusion_sleep_on() thus making FUSION_ASSUME() happy.

---
 drivers/char/fusion/entries.c |    1 +
 1 files changed, 1 insertions(+), 0 deletions(-)

diff --git a/drivers/char/fusion/entries.c b/drivers/char/fusion/entries.c
index a6bfffa..895539e 100644
--- a/drivers/char/fusion/entries.c
+++ b/drivers/char/fusion/entries.c
@@ -379,6 +379,7 @@ fusion_entry_wait( FusionEntry *entry, long *timeout )
 
      entry->waiters++;
 
+     entry->lock_pid = 0;
      fusion_sleep_on( &entry->wait, &entry->lock, timeout );
 
      entry->waiters--;
-- 
1.7.3.3

