From 8b8dec9729232d10603f07a6e304e6706ae49ef1 Mon Sep 17 00:00:00 2001
From: Denis Oliver Kropp <dok@directfb.org>
Date: Thu, 29 Dec 2005 22:57:08 +0000
Subject: [PATCH] Builds with 2.6.15 now.

---
 drivers/char/fusion/fusiondev.c |   11 +++++++----
 1 files changed, 7 insertions(+), 4 deletions(-)

diff --git a/drivers/char/fusion/fusiondev.c b/drivers/char/fusion/fusiondev.c
index 2eb133c..4421d35 100644
--- a/drivers/char/fusion/fusiondev.c
+++ b/drivers/char/fusion/fusiondev.c
@@ -1007,17 +1007,20 @@ register_devices(void)
      devfs_mk_dir("fusion");
 
      for (i=0; i<NUM_MINORS; i++) {
-#if LINUX_VERSION_CODE >= KERNEL_VERSION(2, 6, 2)
-#if LINUX_VERSION_CODE >= KERNEL_VERSION(2, 6, 13)
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(2, 6, 15)
           class_device_create (fusion_class,
+                               NULL,
                                MKDEV(FUSION_MAJOR, i),
                                NULL, "fusion%d", i);
-#else
+#elif LINUX_VERSION_CODE >= KERNEL_VERSION(2, 6, 13)
+          class_device_create (fusion_class,
+                               MKDEV(FUSION_MAJOR, i),
+                               NULL, "fusion%d", i);
+#elif LINUX_VERSION_CODE >= KERNEL_VERSION(2, 6, 2)
           class_simple_device_add (fusion_class,
                                    MKDEV(FUSION_MAJOR, i),
                                    NULL, "fusion%d", i);
 #endif
-#endif
 
           devfs_mk_cdev (MKDEV(FUSION_MAJOR, i),
                          S_IFCHR | S_IRUSR | S_IWUSR,
-- 
1.7.3.3

