From 834bd0926bdcf8312276e54d9124cf89d5c19bf0 Mon Sep 17 00:00:00 2001
From: Denis Oliver Kropp <dok@directfb.org>
Date: Wed, 13 Sep 2006 13:02:23 +0000
Subject: [PATCH] Unlock entry before freeing it, thanks to Maxim Syrchin <syrchin@dev.rtsoft.ru>!

---
 drivers/char/fusion/entries.c |    3 +++
 1 files changed, 3 insertions(+), 0 deletions(-)

diff --git a/drivers/char/fusion/entries.c b/drivers/char/fusion/entries.c
index 2834bc7..e217415 100644
--- a/drivers/char/fusion/entries.c
+++ b/drivers/char/fusion/entries.c
@@ -260,6 +260,9 @@ fusion_entry_destroy_locked( FusionEntries  *entries,
      if (class->Destroy)
           class->Destroy( entry, entries->ctx );
 
+     /* Unlock the entry. */
+     up( &entry->lock );
+
      /* Deallocate the entry. */
      kfree( entry );
 }
-- 
1.7.3.3

