From ef63781fb7ace84f9b95aeb575fc8ac367b81325 Mon Sep 17 00:00:00 2001
From: Ville Syrjala <syrjala@directfb.org>
Date: Sat, 6 Jan 2007 13:03:48 +0000
Subject: [PATCH] Fix bool redefinition. Patch from Daniel Laird.

---
 drivers/char/fusion/types.h |    4 ++++
 1 files changed, 4 insertions(+), 0 deletions(-)

diff --git a/drivers/char/fusion/types.h b/drivers/char/fusion/types.h
index 843ce45..a0b1288 100644
--- a/drivers/char/fusion/types.h
+++ b/drivers/char/fusion/types.h
@@ -15,9 +15,13 @@
 #ifndef __FUSION__TYPES_H__
 #define __FUSION__TYPES_H__
 
+#include <linux/version.h>
+
+#if LINUX_VERSION_CODE <= KERNEL_VERSION(2, 6, 18)
 typedef enum {
   false = 0,
   true  = !false
 } bool;
+#endif
 
 #endif
-- 
1.7.3.3

