From 4187c2bc3131d7e576cd7786844ae457be75c31a Mon Sep 17 00:00:00 2001
From: Ville Syrjala <syrjala@directfb.org>
Date: Sat, 6 Jan 2007 13:19:37 +0000
Subject: [PATCH] lookup_fusionee() can't return a NULL fusionee on success.

---
 drivers/char/fusion/fusionee.c |   15 +++++----------
 1 files changed, 5 insertions(+), 10 deletions(-)

diff --git a/drivers/char/fusion/fusionee.c b/drivers/char/fusion/fusionee.c
index 6a43912..5a99151 100644
--- a/drivers/char/fusion/fusionee.c
+++ b/drivers/char/fusion/fusionee.c
@@ -425,9 +425,6 @@ fusionee_poll (FusionDev *dev, FusionID id, struct file *file, poll_table * wait
      if (ret)
           return ret;
 
-     if (!fusionee)
-          return -EINVAL;
-
      if (fusionee->messages.count) {
           unlock_fusionee (fusionee);
 
@@ -574,17 +571,15 @@ lock_fusionee (FusionDev *dev, FusionID id, Fusionee **ret_fusionee)
      if (ret)
           return ret;
 
-     if (fusionee) {
-          fusion_list_move_to_front (&dev->fusionee.list, &fusionee->link);
-
-          if (down_interruptible (&fusionee->lock)) {
-               up (&dev->fusionee.lock);
-               return -EINTR;
-          }
+     fusion_list_move_to_front (&dev->fusionee.list, &fusionee->link);
 
+     if (down_interruptible (&fusionee->lock)) {
           up (&dev->fusionee.lock);
+          return -EINTR;
      }
 
+     up (&dev->fusionee.lock);
+
      *ret_fusionee = fusionee;
 
      return 0;
-- 
1.7.3.3

