From 22fc1ea59a821974443bfce942142c30e83de024 Mon Sep 17 00:00:00 2001
From: Ville Syrjala <syrjala@directfb.org>
Date: Sat, 6 Jan 2007 13:26:36 +0000
Subject: [PATCH] Fix a potential double up() on dev->fusionee.lock.

---
 drivers/char/fusion/fusionee.c |    6 +++---
 1 files changed, 3 insertions(+), 3 deletions(-)

diff --git a/drivers/char/fusion/fusionee.c b/drivers/char/fusion/fusionee.c
index 5a99151..4255347 100644
--- a/drivers/char/fusion/fusionee.c
+++ b/drivers/char/fusion/fusionee.c
@@ -460,8 +460,10 @@ fusionee_kill (FusionDev *dev, FusionID id, int target, int signal, int timeout_
                }
           }
 
-          if (!killed || timeout_ms < 0)
+          if (!killed || timeout_ms < 0) {
+               up (&dev->fusionee.lock);
                break;
+          }
 
           if (timeout_ms) {
                switch (timeout) {
@@ -489,8 +491,6 @@ fusionee_kill (FusionDev *dev, FusionID id, int target, int signal, int timeout_
                return -EINTR;
      }
 
-     up (&dev->fusionee.lock);
-
      return 0;
 }
 
-- 
1.7.3.3

