From 6c25b0029e71e9a0a12d67dcd82ba320b20b5245 Mon Sep 17 00:00:00 2001
From: Ville Syrjala <syrjala@directfb.org>
Date: Sat, 6 Jan 2007 13:36:58 +0000
Subject: [PATCH] Don't allow fusion_release() to fail. I'm not sure if it's a good idea but at least keeps my console working when XDirectFB crashes.

---
 drivers/char/fusion/fusiondev.c |    7 +++----
 1 files changed, 3 insertions(+), 4 deletions(-)

diff --git a/drivers/char/fusion/fusiondev.c b/drivers/char/fusion/fusiondev.c
index 7d326ed..2da2d68 100644
--- a/drivers/char/fusion/fusiondev.c
+++ b/drivers/char/fusion/fusiondev.c
@@ -332,15 +332,14 @@ fusion_open (struct inode *inode, struct file *file)
 static int
 fusion_release (struct inode *inode, struct file *file)
 {
-     int ret;
      int minor        = iminor(inode);
      size_t fusion_id = (size_t) file->private_data;
 
      DEBUG( "fusion_release( %p, %d )\n", file, atomic_read(&file->f_count) );
 
-     ret = fusionee_destroy (fusion_devs[minor], fusion_id);
-     if (ret)
-          return ret;
+     /* FIXME: is this a good idea? */
+     while (fusionee_destroy (fusion_devs[minor], fusion_id))
+          schedule ();
 
      down (&devs_lock);
 
-- 
1.7.3.3

