From 21b8c20af8b13809d17af22e54ab8e4ab4aa5755 Mon Sep 17 00:00:00 2001
From: Denis Oliver Kropp <dok@directfb.org>
Date: Sat, 20 Jan 2007 04:16:22 +0000
Subject: [PATCH] Fixed O_EXCL (master) with slaves already waiting.

---
 drivers/char/fusion/fusiondev.c |    6 ++++--
 1 files changed, 4 insertions(+), 2 deletions(-)

diff --git a/drivers/char/fusion/fusiondev.c b/drivers/char/fusion/fusiondev.c
index 15a574b..6a77a99 100644
--- a/drivers/char/fusion/fusiondev.c
+++ b/drivers/char/fusion/fusiondev.c
@@ -303,8 +303,10 @@ fusion_open (struct inode *inode, struct file *file)
           }
      }
      else if (file->f_flags & O_EXCL) {
-          up (&devs_lock);
-          return -EBUSY;
+          if (fusion_devs[minor]->fusionee.last_id) {
+               up (&devs_lock);
+               return -EBUSY;
+          }
      }
 
      ret = fusionee_new (fusion_devs[minor], !!(file->f_flags & O_APPEND), &fusionee);
-- 
1.7.3.3

