From 3f45815809624ed556debf38cd93f8852ce11bd1 Mon Sep 17 00:00:00 2001
From: Ville Syrjala <syrjala@directfb.org>
Date: Sun, 28 Jan 2007 23:31:00 +0000
Subject: [PATCH] Make fusionee_poll() return POLLERR on error. The return value is a bitmask so returning the negative return value from lock_fusionee() doesn't seem like good idea.

---
 drivers/char/fusion/fusionee.c |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/drivers/char/fusion/fusionee.c b/drivers/char/fusion/fusionee.c
index 72292a5..3bba64e 100644
--- a/drivers/char/fusion/fusionee.c
+++ b/drivers/char/fusion/fusionee.c
@@ -403,7 +403,7 @@ fusionee_poll (FusionDev   *dev,
 
      ret = lock_fusionee (dev, id, &fusionee);
      if (ret)
-          return ret;
+          return POLLERR;
 
      if (fusionee->messages.count) {
           unlock_fusionee (fusionee);
-- 
1.7.3.3

