From 309ec01ce7b358d6bc6811ac1cba441a1a50bfc2 Mon Sep 17 00:00:00 2001
From: Denis Oliver Kropp <dok@directfb.org>
Date: Mon, 29 Jan 2007 00:43:50 +0000
Subject: [PATCH] Don't include <asm/types.h>.

---
 include/linux/fusion.h |    1 -
 1 files changed, 0 insertions(+), 1 deletions(-)

diff --git a/include/linux/fusion.h b/include/linux/fusion.h
index f634e87..d9139fc 100644
--- a/include/linux/fusion.h
+++ b/include/linux/fusion.h
@@ -2,7 +2,6 @@
 #define __LINUX__FUSION_H__
 
 #include <asm/ioctl.h>
-#include <asm/types.h>
 
 /*
  * Fusion Kernel Device API Version
-- 
1.7.3.3

