From 2b836787f3a87c0ca6dedadf1466f4a22df04829 Mon Sep 17 00:00:00 2001
From: Denis Oliver Kropp <dok@directfb.org>
Date: Thu, 8 Mar 2007 13:02:04 +0000
Subject: [PATCH] In fusion_entry_wait() after calling fusion_sleep_on() first check for signals, then check for timeout.

In fusion_entry_notify() don't rely on valid "waiters" counter, but always
do the wake up on the queue.
---
 drivers/char/fusion/entries.c |    9 +++------
 1 files changed, 3 insertions(+), 6 deletions(-)

diff --git a/drivers/char/fusion/entries.c b/drivers/char/fusion/entries.c
index ad143f9..0f7c277 100644
--- a/drivers/char/fusion/entries.c
+++ b/drivers/char/fusion/entries.c
@@ -405,12 +405,12 @@ fusion_entry_wait( FusionEntry *entry, long *timeout )
 
      entry->waiters--;
 
-     if (timeout && !*timeout)
-          return -ETIMEDOUT;
-
      if (signal_pending(current))
           return -EINTR;
 
+     if (timeout && !*timeout)
+          return -ETIMEDOUT;
+
      ret = fusion_entry_lock( entries, id, false, &entry2 );
      switch (ret) {
           case -EINVAL:
@@ -430,9 +430,6 @@ fusion_entry_notify( FusionEntry *entry, bool all )
      FUSION_ASSERT( entry != NULL );
      FUSION_ASSUME( entry->lock_pid == current->pid );
 
-     if (!entry->waiters)
-          return;
-
      if (all)
           wake_up_interruptible_all( &entry->wait );
      else
-- 
1.7.3.3

