From 796cf73b2dd3215d80dfb8f7f50d262463d433c2 Mon Sep 17 00:00:00 2001
From: Denis Oliver Kropp <dok@directfb.org>
Date: Thu, 8 Mar 2007 13:02:47 +0000
Subject: [PATCH] Fix fusion_sleep_on() for 2.6 kernels, using prepare_to_wait() and finish_wait() etc.

---
 drivers/char/fusion/fusiondev.c |   30 ++++++++++++++++--------------
 1 files changed, 16 insertions(+), 14 deletions(-)

diff --git a/drivers/char/fusion/fusiondev.c b/drivers/char/fusion/fusiondev.c
index 1ca50cc..4c92e7c 100644
--- a/drivers/char/fusion/fusiondev.c
+++ b/drivers/char/fusion/fusiondev.c
@@ -88,31 +88,33 @@ static struct class_simple *fusion_class;
 
 /******************************************************************************/
 
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(2, 6, 0)
 void
 fusion_sleep_on(wait_queue_head_t *q, struct semaphore *lock, signed long *timeout)
 {
-     wait_queue_t wait;
+     DEFINE_WAIT(wait);
 
-     init_waitqueue_entry (&wait, current);
+     prepare_to_wait( q, &wait, TASK_INTERRUPTIBLE );
 
-     current->state = TASK_INTERRUPTIBLE;
-
-#if LINUX_VERSION_CODE >= KERNEL_VERSION(2, 6, 0)
-     spin_lock (&q->lock);
-     __add_wait_queue (q, &wait);
-     spin_unlock (&q->lock);
-
-     up (lock);
+     up( lock );
 
      if (timeout)
           *timeout = schedule_timeout(*timeout);
      else
           schedule();
 
-     spin_lock (&q->lock);
-     __remove_wait_queue (q, &wait);
-     spin_unlock (&q->lock);
+     finish_wait( q, &wait );
+}
 #else
+void
+fusion_sleep_on(wait_queue_head_t *q, struct semaphore *lock, signed long *timeout)
+{
+     wait_queue_t wait;
+
+     init_waitqueue_entry (&wait, current);
+
+     current->state = TASK_INTERRUPTIBLE;
+
      write_lock (&q->lock);
      __add_wait_queue (q, &wait);
      write_unlock (&q->lock);
@@ -127,8 +129,8 @@ fusion_sleep_on(wait_queue_head_t *q, struct semaphore *lock, signed long *timeo
      write_lock (&q->lock);
      __remove_wait_queue (q, &wait);
      write_unlock (&q->lock);
-#endif
 }
+#endif
 
 /******************************************************************************/
 
-- 
1.7.3.3

