From 98c476e9d70be8edeecf8be6fcd2665bc167ae4d Mon Sep 17 00:00:00 2001
From: Claudio Ciccani <klan@directfb.org>
Date: Tue, 15 May 2007 21:21:50 +0200
Subject: [PATCH] Workaround to make forked processes work with fusion: reference local refs on fork() instead of duplicating them.

---
 drivers/char/fusion/ref.c |    4 ++++
 1 files changed, 4 insertions(+), 0 deletions(-)

diff --git a/drivers/char/fusion/ref.c b/drivers/char/fusion/ref.c
index ca1f81a..be8bdc0 100644
--- a/drivers/char/fusion/ref.c
+++ b/drivers/char/fusion/ref.c
@@ -527,6 +527,7 @@ fork_local (FusionDev *dev, FusionRef *ref, FusionID fusion_id, FusionID from_id
 
           if (local->fusion_id == from_id) {
                if (local->refs) {
+#if 0
                     LocalRef *new_local;
 
                     new_local = kmalloc (sizeof(LocalRef), GFP_KERNEL);
@@ -541,6 +542,9 @@ fork_local (FusionDev *dev, FusionRef *ref, FusionID fusion_id, FusionID from_id
                     fusion_list_prepend( &ref->local_refs, &new_local->link );
 
                     propagate_local( dev, ref, local->refs );
+#else
+                    local->refs++;
+#endif
                }
                break;
           }
-- 
1.7.3.3

