From d5bd79b170b0a655700ba5e3cdfe94a21e6a4217 Mon Sep 17 00:00:00 2001
From: Denis Oliver Kropp <dok@shizo.zion.home>
Date: Wed, 30 May 2007 16:27:55 +0200
Subject: [PATCH] Fix typo in address range check.

---
 drivers/char/fusion/reactor.c |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/drivers/char/fusion/reactor.c b/drivers/char/fusion/reactor.c
index 549daa8..ffdec61 100644
--- a/drivers/char/fusion/reactor.c
+++ b/drivers/char/fusion/reactor.c
@@ -344,7 +344,7 @@ fusion_reactor_dispatch (FusionDev *dev, int id, int channel, Fusionee *fusionee
 
           if (!reactor->call_ptr && msg_size == sizeof(ptr) &&
               (unsigned long) ptr >= FUSION_SHM_BASE &&
-              (unsigned long) ptr < (FUSION_SHM_BASE+FUSION_SHM_BASE))
+              (unsigned long) ptr < (FUSION_SHM_BASE+FUSION_SHM_SIZE))
                dispatch->call_ptr = ptr;
           else
                dispatch->call_ptr = reactor->call_ptr;
-- 
1.7.3.3

