From 3939bae94b49a10dc5eb9bcb0504f6c450f6d699 Mon Sep 17 00:00:00 2001
From: Denis Oliver Kropp <dok@shizo.zion.home>
Date: Wed, 30 May 2007 16:53:12 +0200
Subject: [PATCH] Well, ~16k GB was a bit too much. Changed size to ~64 GB starting at 0x523000010000.

---
 drivers/char/fusion/shmpool.h |    4 ++--
 1 files changed, 2 insertions(+), 2 deletions(-)

diff --git a/drivers/char/fusion/shmpool.h b/drivers/char/fusion/shmpool.h
index 0bbb588..a3a67b6 100644
--- a/drivers/char/fusion/shmpool.h
+++ b/drivers/char/fusion/shmpool.h
@@ -19,8 +19,8 @@
 #include "types.h"
 
 #ifdef __x86_64__
-#define FUSION_SHM_BASE    0x500000010000     /* virtual base address */
-#define FUSION_SHM_SIZE    0x0FFFFFFEF000     /* size of virtual address space */
+#define FUSION_SHM_BASE    0x523000010000     /* virtual base address */
+#define FUSION_SHM_SIZE    0x000FFFFEF000     /* size of virtual address space */
 #else
 
 
-- 
1.7.3.3

