From 99fc9ad2b474a2e20578f865247d51000c59e409 Mon Sep 17 00:00:00 2001
From: Denis Oliver Kropp <dok@shizo.(none)>
Date: Sun, 3 Jun 2007 19:11:53 +0200
Subject: [PATCH] Send proper channel number in reactor message.

---
 drivers/char/fusion/reactor.c |    4 ++--
 1 files changed, 2 insertions(+), 2 deletions(-)

diff --git a/drivers/char/fusion/reactor.c b/drivers/char/fusion/reactor.c
index ffdec61..0a75232 100644
--- a/drivers/char/fusion/reactor.c
+++ b/drivers/char/fusion/reactor.c
@@ -362,12 +362,12 @@ fusion_reactor_dispatch (FusionDev *dev, int id, int channel, Fusionee *fusionee
                dispatch->count++;
 
                fusionee_send_message (dev, fusionee, node->fusion_id, FMT_REACTOR,
-                                      reactor->entry.id, 0, msg_size, msg_data,
+                                      reactor->entry.id, channel, msg_size, msg_data,
                                       dispatch_callback, dispatch, reactor->entry.id);
           }
           else
                fusionee_send_message (dev, fusionee, node->fusion_id, FMT_REACTOR,
-                                      reactor->entry.id, 0, msg_size, msg_data, NULL, NULL, 0);
+                                      reactor->entry.id, channel, msg_size, msg_data, NULL, NULL, 0);
      }
 
      if (dispatch && !dispatch->count) {
-- 
1.7.3.3

