From 8dda6c1cf5381f8fdc81e83593b2ab43e1216d1d Mon Sep 17 00:00:00 2001
From: Denis Oliver Kropp <dok@shizo.zion.home>
Date: Sun, 12 Aug 2007 16:46:30 +0200
Subject: [PATCH] Changed major id to 250. Dynamic majors please.

---
 drivers/char/fusion/fusiondev.c |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/drivers/char/fusion/fusiondev.c b/drivers/char/fusion/fusiondev.c
index 223e0da..46f3da7 100644
--- a/drivers/char/fusion/fusiondev.c
+++ b/drivers/char/fusion/fusiondev.c
@@ -57,7 +57,7 @@
 #endif
 
 #ifndef FUSION_MAJOR
-#define FUSION_MAJOR 252
+#define FUSION_MAJOR 250
 #endif
 
 MODULE_LICENSE("GPL");
-- 
1.7.3.3

