From 41adb473a9e30e3e581d0d9b2bc574d257e52967 Mon Sep 17 00:00:00 2001
From: Denis Oliver Kropp <dok@shizo.zion.home>
Date: Sun, 12 Aug 2007 23:25:50 +0200
Subject: [PATCH] Change maximum number of pre-acquisitions to remember from 32 to 256.

---
 drivers/char/fusion/skirmish.c |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/drivers/char/fusion/skirmish.c b/drivers/char/fusion/skirmish.c
index 334204a..6f09cbb 100644
--- a/drivers/char/fusion/skirmish.c
+++ b/drivers/char/fusion/skirmish.c
@@ -29,7 +29,7 @@
 #include "skirmish.h"
 
 
-#define MAX_PRE_ACQUISITIONS  32
+#define MAX_PRE_ACQUISITIONS  256
 
 
 typedef struct __FUSION_FusionSkirmish FusionSkirmish;
-- 
1.7.3.3

