From 3f7fd4288165290fa0395aa1339fef8103524c82 Mon Sep 17 00:00:00 2001
From: Denis Oliver Kropp <dok@shizo.zion.home>
Date: Tue, 9 Oct 2007 00:48:43 +0200
Subject: [PATCH] Changed version check for bool from <= 2.6.18 to < 2.6.18.

---
 drivers/char/fusion/types.h |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/drivers/char/fusion/types.h b/drivers/char/fusion/types.h
index 4ee9c90..703a905 100644
--- a/drivers/char/fusion/types.h
+++ b/drivers/char/fusion/types.h
@@ -17,7 +17,7 @@
 
 #include <linux/version.h>
 
-#if LINUX_VERSION_CODE <= KERNEL_VERSION(2, 6, 18)
+#if LINUX_VERSION_CODE < KERNEL_VERSION(2, 6, 18)
 typedef enum {
   false = 0,
   true  = !false
-- 
1.7.3.3

