From 7f9d07405718077958bf5fc743ad823fa09e6188 Mon Sep 17 00:00:00 2001
From: Denis Oliver Kropp <dok@shizo.zion.home>
Date: Tue, 9 Oct 2007 03:47:31 +0200
Subject: [PATCH] Added missing do {} while (0) around FUSION_ASSERT and FUSION_ASSUME.

---
 drivers/char/fusion/fusiondev.h |    4 ++--
 1 files changed, 2 insertions(+), 2 deletions(-)

diff --git a/drivers/char/fusion/fusiondev.h b/drivers/char/fusion/fusiondev.h
index c61c9f0..20d6826 100644
--- a/drivers/char/fusion/fusiondev.h
+++ b/drivers/char/fusion/fusiondev.h
@@ -20,8 +20,8 @@
 #include "entries.h"
 #include "list.h"
 
-#define FUSION_ASSERT(exp)    if (!(exp)) BUG()
-#define FUSION_ASSUME(exp)    if (!(exp)) printk( KERN_ERR "fusiondev: assumption '" #exp "' failed!\n" )
+#define FUSION_ASSERT(exp)    do { if (!(exp)) BUG(); } while (0)
+#define FUSION_ASSUME(exp)    do { if (!(exp)) printk( KERN_ERR "fusiondev: assumption '" #exp "' failed!\n" ); } while (0)
 
 struct __Fusion_FusionDev {
      int                    refs;
-- 
1.7.3.3

