From 90921a0d95983ff0cac2bc7b2a3abe26a1ddb71f Mon Sep 17 00:00:00 2001
From: Denis Oliver Kropp <dok@shizo.zion.home>
Date: Tue, 9 Oct 2007 03:55:07 +0200
Subject: [PATCH] Added fusionee_dispatcher_pid( fusion_id ) to look up the pid of the (last) thread reading from the device.

There should always be only one thread reading, but for safety there's a FUSION_ASSUME failing if the pid changes.
---
 drivers/char/fusion/fusionee.c |   33 +++++++++++++++++++++++++++++++++
 drivers/char/fusion/fusionee.h |    3 +++
 2 files changed, 36 insertions(+), 0 deletions(-)

diff --git a/drivers/char/fusion/fusionee.c b/drivers/char/fusion/fusionee.c
index 95707c3..87eaaa2 100644
--- a/drivers/char/fusion/fusionee.c
+++ b/drivers/char/fusion/fusionee.c
@@ -61,6 +61,8 @@ struct __Fusion_Fusionee {
      bool              force_slave;
 
      struct mm_struct *mm;
+
+     pid_t             dispatcher_pid;
 };
 
 typedef struct {
@@ -358,6 +360,11 @@ fusionee_get_messages (FusionDev *dev,
      if (down_interruptible (&fusionee->lock))
           return -EINTR;
 
+     if (fusionee->dispatcher_pid)
+          FUSION_ASSUME( fusionee->dispatcher_pid == current->pid );
+
+     fusionee->dispatcher_pid = current->pid;
+
      prev_msgs = fusionee->prev_msgs;
 
      fusion_fifo_reset( &fusionee->prev_msgs );
@@ -610,6 +617,32 @@ fusionee_id( const Fusionee *fusionee )
      return fusionee->id;
 }
 
+pid_t
+fusionee_dispatcher_pid( FusionDev *dev,
+                         FusionID   fusion_id )
+{
+     FusionLink *l;
+     int         ret = -EINVAL;
+
+     down (&dev->fusionee.lock);
+
+     fusion_list_foreach (l, dev->fusionee.list) {
+          Fusionee *fusionee = (Fusionee *) l;
+
+          if (fusionee->id == fusion_id) {
+               /* FIXME: wait for it? */
+               FUSION_ASSUME( fusionee->dispatcher_pid != 0 );
+
+               ret = fusionee->dispatcher_pid;
+               break;
+          }
+     }
+
+     up (&dev->fusionee.lock);
+
+     return ret;
+}
+
 /******************************************************************************/
 
 static int
diff --git a/drivers/char/fusion/fusionee.h b/drivers/char/fusion/fusionee.h
index 0fdb598..c30035c 100644
--- a/drivers/char/fusion/fusionee.h
+++ b/drivers/char/fusion/fusionee.h
@@ -76,4 +76,7 @@ void fusionee_destroy      (FusionDev         *dev,
 
 FusionID fusionee_id( const Fusionee *fusionee );
 
+pid_t    fusionee_dispatcher_pid( FusionDev *dev,
+                                  FusionID   fusion_id );
+
 #endif
-- 
1.7.3.3

