From e5849518b3731a3c5500a9355a6eeada19a372cd Mon Sep 17 00:00:00 2001
From: Ville Syrjala <syrjala@sci.fi>
Date: Sat, 1 Dec 2007 21:18:59 +0200
Subject: [PATCH] Use current_kernel_time() instead of xtime

xtime is no longer exported in 2.6.24 so use current_kernel_time() instead.
---
 drivers/char/fusion/entries.c |    6 ++++++
 1 files changed, 6 insertions(+), 0 deletions(-)

diff --git a/drivers/char/fusion/entries.c b/drivers/char/fusion/entries.c
index d354a69..7469710 100644
--- a/drivers/char/fusion/entries.c
+++ b/drivers/char/fusion/entries.c
@@ -318,6 +318,9 @@ fusion_entry_lock( FusionEntries  *entries,
                    FusionEntry   **ret_entry )
 {
      FusionEntry *entry;
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(2, 6, 24)
+     struct timespec xtime;
+#endif
 
      FUSION_ASSERT( entries != NULL );
      FUSION_ASSERT( ret_entry != NULL );
@@ -355,6 +358,9 @@ fusion_entry_lock( FusionEntries  *entries,
      /* Keep timestamp, but use the slightly
         inexact version to avoid performance impacts. */
 #if LINUX_VERSION_CODE >= KERNEL_VERSION(2, 6, 0) && defined _STRUCT_TIMESPEC
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(2, 6, 24)
+     xtime = current_kernel_time();
+#endif
      entry->last_lock.tv_sec = xtime.tv_sec;
      entry->last_lock.tv_usec = xtime.tv_nsec / 1000;
 #else
-- 
1.7.3.3

