From 5a8c24072ecc04c6d88c73c19e0714b755b2624b Mon Sep 17 00:00:00 2001
From: Denis Oliver Kropp <dok@shizo.zion.home>
Date: Thu, 11 Oct 2007 11:59:22 +0200
Subject: [PATCH] Don't build with yield() by default, but require FUSION_SKIRMISH_YIELD to be defined.

---
 drivers/char/fusion/skirmish.c |    2 ++
 1 files changed, 2 insertions(+), 0 deletions(-)

diff --git a/drivers/char/fusion/skirmish.c b/drivers/char/fusion/skirmish.c
index baa77ce..6a08014 100644
--- a/drivers/char/fusion/skirmish.c
+++ b/drivers/char/fusion/skirmish.c
@@ -376,9 +376,11 @@ fusion_skirmish_dismiss (FusionDev *dev, int id, int fusion_id)
 
      fusion_skirmish_unlock( skirmish );
 
+#ifdef FUSION_SKIRMISH_YIELD
      /* Locked > 20 ms ? */
      if (lock_jiffies > HZ/50)// && current->policy == SCHED_NORMAL)
           yield();
+#endif
 
      return 0;
 }
-- 
1.7.3.3

