From 2f73c02b55b59645007a5cba9e5f73deaab59a9c Mon Sep 17 00:00:00 2001
From: Denis Oliver Kropp <dok@shizo.directfb.home>
Date: Sat, 26 Jan 2008 03:04:38 +0100
Subject: [PATCH] REACTOR: Added missing memset() and kfree() in reactor_attach(). Thanks to Amit Mahajan and Niels Roest!

Without memset() the new counters for the channels are not initialized.

The kfree() was missing in the code that reallocated the array for more channels.
---
 drivers/char/fusion/reactor.c |   14 +++++++++-----
 1 files changed, 9 insertions(+), 5 deletions(-)

diff --git a/drivers/char/fusion/reactor.c b/drivers/char/fusion/reactor.c
index 3d59cb8..a911a55 100644
--- a/drivers/char/fusion/reactor.c
+++ b/drivers/char/fusion/reactor.c
@@ -183,6 +183,8 @@ fusion_reactor_attach (FusionDev *dev, int id, int channel, FusionID fusion_id)
                return -ENOMEM;
           }
 
+          memset( node->counts, 0, sizeof(node->counts) );
+
           node->num_counts = ncount;
           node->fusion_id  = fusion_id;
 
@@ -203,6 +205,8 @@ fusion_reactor_attach (FusionDev *dev, int id, int channel, FusionID fusion_id)
                memcpy( counts, node->counts, sizeof(int) * node->num_counts );
                memset( counts + node->num_counts, 0, sizeof(int) * (ncount - node->num_counts) );
 
+               kfree( node->counts );
+
                node->counts     = counts;
                node->num_counts = ncount;
           }
@@ -363,13 +367,13 @@ fusion_reactor_dispatch (FusionDev *dev, int id, int channel, Fusionee *fusionee
           if (dispatch) {
                dispatch->count++;
 
-               fusionee_send_message (dev, fusionee, node->fusion_id, FMT_REACTOR,
-                                      reactor->entry.id, channel, msg_size, msg_data,
-                                      dispatch_callback, dispatch, reactor->entry.id);
+               ret = fusionee_send_message (dev, fusionee, node->fusion_id, FMT_REACTOR,
+								    reactor->entry.id, channel, msg_size, msg_data,
+								    dispatch_callback, dispatch, reactor->entry.id);
           }
           else
-               fusionee_send_message (dev, fusionee, node->fusion_id, FMT_REACTOR,
-                                      reactor->entry.id, channel, msg_size, msg_data, NULL, NULL, 0);
+               ret = fusionee_send_message (dev, fusionee, node->fusion_id, FMT_REACTOR,
+								    reactor->entry.id, channel, msg_size, msg_data, NULL, NULL, 0);
      }
 
      if (dispatch && !dispatch->count) {
-- 
1.7.3.3

