From b9f300c50a03ab00c817c1b2389c91027460ac43 Mon Sep 17 00:00:00 2001
From: Denis Oliver Kropp <dok@shizo.directfb.home>
Date: Wed, 26 Mar 2008 11:37:10 +0100
Subject: [PATCH] REACTOR: Fixed call to memset() in reactor_attach(). Thanks to Peter Deckmyn!

---
 drivers/char/fusion/reactor.c |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/drivers/char/fusion/reactor.c b/drivers/char/fusion/reactor.c
index a911a55..6805f0f 100644
--- a/drivers/char/fusion/reactor.c
+++ b/drivers/char/fusion/reactor.c
@@ -183,7 +183,7 @@ fusion_reactor_attach (FusionDev *dev, int id, int channel, FusionID fusion_id)
                return -ENOMEM;
           }
 
-          memset( node->counts, 0, sizeof(node->counts) );
+          memset( node->counts, 0, sizeof(int) * ncount );
 
           node->num_counts = ncount;
           node->fusion_id  = fusion_id;
-- 
1.7.3.3

