From bc07b578baa78404c6ccb608281cb62a57e8f145 Mon Sep 17 00:00:00 2001
From: Denis Oliver Kropp <dok@cyto.(none)>
Date: Sat, 19 Jul 2008 18:12:23 +0200
Subject: [PATCH] Include <linux/mutex.h> instead of <linux/semaphore.h>, probably not working with all kernels...

---
 drivers/char/fusion/entries.h |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/drivers/char/fusion/entries.h b/drivers/char/fusion/entries.h
index 7025f94..2ca5ae0 100644
--- a/drivers/char/fusion/entries.h
+++ b/drivers/char/fusion/entries.h
@@ -15,7 +15,7 @@
 #ifndef __FUSION__ENTRIES_H__
 #define __FUSION__ENTRIES_H__
 
-#include <linux/semaphore.h>
+#include <linux/mutex.h>
 
 #include "types.h"
 #include "list.h"
-- 
1.7.3.3

