From 99e081dfdb489a0172c47cffe8e5f14c189951c0 Mon Sep 17 00:00:00 2001
From: Denis Oliver Kropp <dok@debian.(none)>
Date: Wed, 3 Sep 2008 00:45:43 +0900
Subject: [PATCH] Changed maximum version for bool definition to 2.6.18 to make it build on such kernels.

---
 drivers/char/fusion/types.h |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/drivers/char/fusion/types.h b/drivers/char/fusion/types.h
index 703a905..22ced8c 100644
--- a/drivers/char/fusion/types.h
+++ b/drivers/char/fusion/types.h
@@ -17,7 +17,7 @@
 
 #include <linux/version.h>
 
-#if LINUX_VERSION_CODE < KERNEL_VERSION(2, 6, 18)
+#if LINUX_VERSION_CODE < KERNEL_VERSION(2, 6, 19)
 typedef enum {
   false = 0,
   true  = !false
-- 
1.7.3.3

