From 5d55a296052653b9dd88083a81e40828eac18af5 Mon Sep 17 00:00:00 2001
From: Denis Oliver Kropp <dok@cyto.(none)>
Date: Wed, 10 Sep 2008 18:21:59 +0200
Subject: [PATCH] Align pool base on a 64K boundary.

---
 drivers/char/fusion/shmpool.c |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/drivers/char/fusion/shmpool.c b/drivers/char/fusion/shmpool.c
index c53a090..4295889 100644
--- a/drivers/char/fusion/shmpool.c
+++ b/drivers/char/fusion/shmpool.c
@@ -115,7 +115,7 @@ fusion_shmpool_construct( FusionEntry *entry,
      shmpool->max_size  = poolnew->max_size;
      shmpool->addr_base = poolnew->addr_base = (void*) addr_base;
 
-     addr_base += PAGE_ALIGN(poolnew->max_size) + PAGE_SIZE; /* fence page */
+     addr_base = (addr_base + PAGE_ALIGN(poolnew->max_size) + PAGE_SIZE + 0xffff) & ~0xffff;
 
      shmpool->addr_entry = add_addr_entry( addr_base );
 
-- 
1.7.3.3

