From d5365eab18d376f86ab4af72c4dec2f98df455a1 Mon Sep 17 00:00:00 2001
From: Niels Roest <niels@directfb.org>
Date: Mon, 15 Sep 2008 15:22:46 +0200
Subject: [PATCH] made include order consistent. Was needed for new 2.6.26.5 kernel.

---
 drivers/char/fusion/property.c |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/drivers/char/fusion/property.c b/drivers/char/fusion/property.c
index 8180669..4059b0f 100644
--- a/drivers/char/fusion/property.c
+++ b/drivers/char/fusion/property.c
@@ -27,8 +27,8 @@
 
 #include <linux/fusion.h>
 
-#include "entries.h"
 #include "fusiondev.h"
+#include "entries.h"
 #include "fusionee.h"
 #include "list.h"
 #include "property.h"
-- 
1.7.3.3

