From a044fe7014fe82681fe77b10a43206b2ff079a74 Mon Sep 17 00:00:00 2001
From: Niels Roest <niels@directfb.org>
Date: Thu, 18 Sep 2008 10:52:53 +0200
Subject: [PATCH] new feature: automatically unlock transferred skirmishes If you expect that a callee will lock a skirmish, call you and thereby transferring that skirmish to you, you may already lock it; it will be unlocked automatically when the call is due.

---
 drivers/char/fusion/skirmish.c |    2 ++
 1 files changed, 2 insertions(+), 0 deletions(-)

diff --git a/drivers/char/fusion/skirmish.c b/drivers/char/fusion/skirmish.c
index 1cbb4a6..a5da11b 100644
--- a/drivers/char/fusion/skirmish.c
+++ b/drivers/char/fusion/skirmish.c
@@ -687,6 +687,8 @@ fusion_skirmish_transfer_all (FusionDev *dev,
                skirmish->lock_fid   = 0;
                skirmish->lock_pid   = 0;
                skirmish->lock_count = 0;
+
+               wake_up_interruptible_all (&skirmish->entry.wait);
           }
 
           up (&skirmish->entry.lock);
-- 
1.7.3.3

