From 71d0973eeadd2997beda9a7b57506d7f0c15ac24 Mon Sep 17 00:00:00 2001
From: Denis Oliver Kropp <dok@cyto.(none)>
Date: Sat, 18 Oct 2008 02:47:01 +0200
Subject: [PATCH] Remove obsolete proto type.

---
 drivers/char/fusion/fifo.h |    4 +---
 1 files changed, 1 insertions(+), 3 deletions(-)

diff --git a/drivers/char/fusion/fifo.h b/drivers/char/fusion/fifo.h
index 23b6919..3fa83b9 100644
--- a/drivers/char/fusion/fifo.h
+++ b/drivers/char/fusion/fifo.h
@@ -30,8 +30,6 @@ void        fusion_fifo_put   (FusionFifo *fifo,
 
 FusionLink *fusion_fifo_get   (FusionFifo *fifo);
 
-int         fusion_fifo_count (FusionFifo *fifo);
-
 
 static inline void
 fusion_fifo_reset( FusionFifo *fifo )
@@ -40,5 +38,5 @@ fusion_fifo_reset( FusionFifo *fifo )
      fifo->first = fifo->last = NULL;
 }
 
-#endif /* __FUSION__LIST_H__ */
+#endif /* __FUSION__FIFO_H__ */
 
-- 
1.7.3.3

