From 358a8da544b4652062cb1cad0bb53464fbdc4e4e Mon Sep 17 00:00:00 2001
From: Denis Oliver Kropp <dok@cyto.(none)>
Date: Sat, 18 Oct 2008 02:48:06 +0200
Subject: [PATCH] Flush previous messages before waiting for new ones which only applies to the unused blocking mode.

---
 drivers/char/fusion/fusionee.c |   18 ++++++++++--------
 1 files changed, 10 insertions(+), 8 deletions(-)

diff --git a/drivers/char/fusion/fusionee.c b/drivers/char/fusion/fusionee.c
index ddb26fb..7e91306 100644
--- a/drivers/char/fusion/fusionee.c
+++ b/drivers/char/fusion/fusionee.c
@@ -373,22 +373,24 @@ fusionee_get_messages (FusionDev *dev,
 
      while (!fusionee->messages.count) {
           if (!block) {
-               unlock_fusionee (fusionee);
+               unlock_fusionee( fusionee );
                flush_messages( dev, &prev_msgs );
                return -EAGAIN;
           }
 
-          fusion_sleep_on (&fusionee->wait, &fusionee->lock, 0);
-
-          if (signal_pending(current)) {
+          if (prev_msgs.count) {
+               unlock_fusionee( fusionee );
                flush_messages( dev, &prev_msgs );
-               return -EINTR;
           }
+          else {
+               fusion_sleep_on( &fusionee->wait, &fusionee->lock, 0 );
 
-          if (down_interruptible (&fusionee->lock)) {
-               flush_messages( dev, &prev_msgs );
-               return -EINTR;
+               if (signal_pending( current ))
+                    return -EINTR;
           }
+
+          if (down_interruptible (&fusionee->lock))
+               return -EINTR;
      }
 
      while (fusionee->messages.count) {
-- 
1.7.3.3

