From 7f7ad4358de32f6fd55d412646625bbfb3619799 Mon Sep 17 00:00:00 2001
From: Niels Roest <niels@directfb.org>
Date: Mon, 19 Jan 2009 16:30:58 +0100
Subject: [PATCH] small change on advise of dok, fixes other cases too.

---
 drivers/char/fusion/fusionee.c |    5 +----
 1 files changed, 1 insertions(+), 4 deletions(-)

diff --git a/drivers/char/fusion/fusionee.c b/drivers/char/fusion/fusionee.c
index 780b874..e79f33b 100644
--- a/drivers/char/fusion/fusionee.c
+++ b/drivers/char/fusion/fusionee.c
@@ -521,14 +521,11 @@ fusionee_poll (FusionDev   *dev,
 
      fusion_fifo_reset( &fusionee->prev_msgs );
 
-     /* if we have previous messages, fusionee_wait_processing() can be waiting */
-     if (prev_msgs.count)
-          wake_up_interruptible_all (&fusionee->wait);
-
      unlock_fusionee (fusionee);
 
      flush_messages( dev, &prev_msgs );
 
+     wake_up_all (&fusionee->wait);
 
      poll_wait (file, &fusionee->wait, wait);
 
-- 
1.7.3.3

