From b93efebe61dfaf87d7843c8c300aee6aff26d978 Mon Sep 17 00:00:00 2001
From: Niels Roest <niels@directfb.org>
Date: Tue, 3 Mar 2009 19:58:19 +0100
Subject: [PATCH] make macros less arch-dependent

---
 drivers/char/fusion/shmpool.h |   17 ++++++++---------
 1 files changed, 8 insertions(+), 9 deletions(-)

diff --git a/drivers/char/fusion/shmpool.h b/drivers/char/fusion/shmpool.h
index ec0e371..46977ce 100644
--- a/drivers/char/fusion/shmpool.h
+++ b/drivers/char/fusion/shmpool.h
@@ -18,19 +18,18 @@
 #include "fusiondev.h"
 #include "types.h"
 
-#if (sizeof(void*) == 8)
-#define FUSION_SHM_BASE    0x523000010000	/* virtual base address */
-#define FUSION_SHM_SIZE    0x000FFFFEF000	/* size of virtual address space */
-#else
-
 #ifdef __mips__
-#define FUSION_SHM_BASE    0x50010000	/* virtual base address */
+#define FUSION_SHM_BASE_32	0x50010000	/* virtual base address */
 #else
-#define FUSION_SHM_BASE    0x20010000	/* virtual base address */
+#define FUSION_SHM_BASE_32	0x20010000	/* virtual base address */
 #endif
+#define FUSION_SHM_SIZE_32	0x1FFEF000	/* size of virtual address space */
 
-#define FUSION_SHM_SIZE    0x1FFEF000	/* size of virtual address space */
-#endif
+#define FUSION_SHM_BASE_64	0x523000010000	/* virtual base address */
+#define FUSION_SHM_SIZE_64	0x000FFFFEF000	/* size of virtual address space */
+
+#define FUSION_SHM_BASE (sizeof(void*)==4 ? (FUSION_SHM_BASE_32) : (FUSION_SHM_BASE_64))
+#define FUSION_SHM_SIZE (sizeof(void*)==4 ? (FUSION_SHM_SIZE_32) : (FUSION_SHM_SIZE_64))
 
 /* module init/cleanup */
 
-- 
1.7.3.3

