From b288eca82add87a27e71bd5202004d8500377476 Mon Sep 17 00:00:00 2001
From: Niels Roest <niels@directfb.org>
Date: Mon, 9 Mar 2009 14:59:33 +0100
Subject: [PATCH] call_return: check serial only for API >= 4.0 (and: remove /proc/fusion/0/call duplicates)

---
 drivers/char/fusion/call.c |    7 +++----
 1 files changed, 3 insertions(+), 4 deletions(-)

diff --git a/drivers/char/fusion/call.c b/drivers/char/fusion/call.c
index b935b59..4c8ca91 100644
--- a/drivers/char/fusion/call.c
+++ b/drivers/char/fusion/call.c
@@ -112,9 +112,8 @@ fusion_call_print(FusionEntry * entry, void *ctx, struct seq_file *p)
 	if (call->executions)
 		idle = ((FusionCallExecution *) call->executions)->executed;
 
-	seq_printf(p, "(%5d) 0x%08x (%d calls) %s",
-		   call->entry.pid, call->entry.id, call->count,
-		   idle ? "idle" : "executing");
+	seq_printf(p, "(%d calls) %s",
+		   call->count, idle ? "idle" : "executing");
 
 	fusion_list_foreach(e, call->executions) {
 		FusionCallExecution *exec = (FusionCallExecution *) e;
@@ -255,7 +254,7 @@ fusion_call_return(FusionDev * dev, int fusion_id, FusionCallReturn * call_ret)
 	FusionLink *l;
 	FusionCall *call;
 
-	if (!call_ret->serial)
+	if ( (dev->api.major >= 4) && (call_ret->serial == 0) )
 		return -EOPNOTSUPP;
 
 	/* Lookup and lock call. */
-- 
1.7.3.3

