From aafa7ccebc42765a4a1f000c2d54edf18f179d67 Mon Sep 17 00:00:00 2001
From: Niels Roest <niels@directfb.org>
Date: Thu, 12 Mar 2009 20:00:44 +0100
Subject: [PATCH] upgrade provided API to 8.1.0

---
 include/linux/fusion.h |    5 +++++
 1 files changed, 5 insertions(+), 0 deletions(-)

diff --git a/include/linux/fusion.h b/include/linux/fusion.h
index 1170a56..18437dc 100644
--- a/include/linux/fusion.h
+++ b/include/linux/fusion.h
@@ -17,6 +17,11 @@
 
 #include <asm/ioctl.h>
 
+/* Fusion supports all API versions up to this version */
+#define FUSION_API_MAJOR_PROVIDED 8
+#define FUSION_API_MINOR_PROVIDED 1
+#define FUSION_API_MICRO_PROVIDED 0
+
 /*
  * Fusion Kernel Device API Version
  * Default behaviour: 3.3 -> DirectFB 1.0.x
-- 
1.7.3.3

