From 5e7633a2ea02429a6843a1862433daef3cd20ae2 Mon Sep 17 00:00:00 2001
From: Niels Roest <niels@directfb.org>
Date: Mon, 16 Nov 2009 19:01:03 +0100
Subject: [PATCH] bug - spurious message disappearance fixed

Observed when killing an application and sending
a message at the same time: the interruptability of the sending
permitted the message to get "lost". Fixed by making this
transition uninterruptable.
---
 drivers/char/fusion/fusionee.c |    8 ++------
 1 files changed, 2 insertions(+), 6 deletions(-)

diff --git a/drivers/char/fusion/fusionee.c b/drivers/char/fusion/fusionee.c
index afdee51..4e4046d 100644
--- a/drivers/char/fusion/fusionee.c
+++ b/drivers/char/fusion/fusionee.c
@@ -294,10 +294,7 @@ fusionee_send_message(FusionDev * dev,
 	if (ret)
 		return ret;
 
-	if (down_interruptible(&fusionee->lock)) {
-		up(&dev->fusionee.lock);
-		return -EINTR;
-	}
+	down(&fusionee->lock);
 
 	if (sender && sender != fusionee) {
 		if (down_interruptible(&sender->lock)) {
@@ -703,8 +700,7 @@ lookup_fusionee(FusionDev * dev, FusionID id, Fusionee ** ret_fusionee)
 {
 	FusionLink *l;
 
-	if (down_interruptible(&dev->fusionee.lock))
-		return -EINTR;
+	down(&dev->fusionee.lock);
 
 	fusion_list_foreach(l, dev->fusionee.list) {
 		Fusionee *fusionee = (Fusionee *) l;
-- 
1.7.3.3

