From 510e376054ab5e926710b4d456358969a88c5e38 Mon Sep 17 00:00:00 2001
From: Niels Roest <niels@directfb.org>
Date: Mon, 16 Nov 2009 19:02:57 +0100
Subject: [PATCH] sh4: clashing address ranges. Fixed by adding SH4-constant.

---
 drivers/char/fusion/shmpool.h |    4 +++-
 1 files changed, 3 insertions(+), 1 deletions(-)

diff --git a/drivers/char/fusion/shmpool.h b/drivers/char/fusion/shmpool.h
index 4f10d58..8feebeb 100644
--- a/drivers/char/fusion/shmpool.h
+++ b/drivers/char/fusion/shmpool.h
@@ -18,8 +18,10 @@
 #include "fusiondev.h"
 #include "types.h"
 
-#ifdef __mips__
+#if defined(__mips__)
 #define FUSION_SHM_BASE_32	0x50010000	/* virtual base address */
+#elif defined(CONFIG_CPU_SH4)
+#define FUSION_SHM_BASE_32	0x30010000	/* virtual base address */
 #else
 #define FUSION_SHM_BASE_32	0x20010000	/* virtual base address */
 #endif
-- 
1.7.3.3

