From 79e9c59d3548aa7591abc6a88efc44676805459b Mon Sep 17 00:00:00 2001
From: Niels Roest <niels@directfb.org>
Date: Mon, 23 Nov 2009 14:03:25 +0100
Subject: [PATCH] skirmish: fix for transferring transferred skirmishes

This behaviour was initially asserted on, but it is now
simply disallowed to transfer transferred skirmishes,
since the ASSERT will hit in some cases in normal circumstances.
---
 drivers/char/fusion/skirmish.c |    3 +--
 1 files changed, 1 insertions(+), 2 deletions(-)

diff --git a/drivers/char/fusion/skirmish.c b/drivers/char/fusion/skirmish.c
index b9c22a2..f1f48f9 100644
--- a/drivers/char/fusion/skirmish.c
+++ b/drivers/char/fusion/skirmish.c
@@ -678,8 +678,7 @@ fusion_skirmish_transfer_all(FusionDev * dev,
 
 		down(&skirmish->entry.lock);
 
-		if (skirmish->lock_pid == from_pid) {
-			FUSION_ASSERT(skirmish->transfer_to == 0);
+		if ( (skirmish->lock_pid == from_pid) && (skirmish->transfer_to == 0) ) {
 			FUSION_ASSERT(skirmish->transfer_from == 0);
 			FUSION_ASSERT(skirmish->transfer_from_pid == 0);
 			FUSION_ASSERT(skirmish->transfer_count == 0);
-- 
1.7.3.3

