From f4a38e53b7ef1356b87042318efeadb91617f4be Mon Sep 17 00:00:00 2001
From: Denis Oliver Kropp <dok@cyto.(none)>
Date: Sun, 28 Mar 2010 01:28:34 +0100
Subject: [PATCH] fusiondev: Include <linux/sched.h>

---
 drivers/char/fusion/fusiondev.c |    1 +
 1 files changed, 1 insertions(+), 0 deletions(-)

diff --git a/drivers/char/fusion/fusiondev.c b/drivers/char/fusion/fusiondev.c
index 93cd93e..2307e9e 100644
--- a/drivers/char/fusion/fusiondev.c
+++ b/drivers/char/fusion/fusiondev.c
@@ -31,6 +31,7 @@
 
 #include <linux/proc_fs.h>
 #include <linux/poll.h>
+#include <linux/sched.h>
 #include <linux/init.h>
 #include <asm/io.h>
 #include <asm/uaccess.h>
-- 
1.7.3.3

