From 3c4d488f594454c0889cacd3ad3f338b6387fff8 Mon Sep 17 00:00:00 2001
From: Andreas Shimokawa <andi@directfb.org>
Date: Thu, 30 Sep 2010 18:14:15 +0200
Subject: [PATCH] In recent 2.6 kernels rcu_read_lock() and rcu_read_unlock() are no longer macros but static inline functions, changed the way of checking for them. This should fix "fusion: Unknown symbol tasklist_lock" while loading the fusion module.

---
 drivers/char/fusion/fusionee.c |    5 +++--
 1 files changed, 3 insertions(+), 2 deletions(-)

diff --git a/drivers/char/fusion/fusionee.c b/drivers/char/fusion/fusionee.c
index a60e93c..ef31f69 100644
--- a/drivers/char/fusion/fusionee.c
+++ b/drivers/char/fusion/fusionee.c
@@ -562,11 +562,12 @@ fusionee_kill(FusionDev * dev,
 			if (f != fusionee && (!target || target == f->id)) {
 				struct task_struct *p;
 
-#ifdef rcu_read_lock
+#if defined(CONFIG_TREE_RCU) || defined(CONFIG_TREE_PREEMPT_RCU) || defined(CONFIG_TINY_RCU) || defined(rcu_read_lock)
 				rcu_read_lock();
 #else
 				read_lock(&tasklist_lock);
 #endif
+
 #ifdef for_each_task		/* 2.4 */
 				for_each_task(p) {
 #else /* for >= 2.6.0 & redhat WS EL3 w/ 2.4 kernel */
@@ -582,7 +583,7 @@ fusionee_kill(FusionDev * dev,
 					}
 				}
 
-#ifdef rcu_read_unlock
+#if defined(CONFIG_TREE_RCU) || defined(CONFIG_TREE_PREEMPT_RCU) || defined(CONFIG_TINY_RCU) || defined(rcu_read_unlock)
 				rcu_read_unlock();
 #else
 				read_unlock(&tasklist_lock);
-- 
1.7.3.3

