From 0f22c264b05cc341b8ac526ba3c073e37c309882 Mon Sep 17 00:00:00 2001
From: Denis Oliver Kropp <dok@directfb.org>
Date: Tue, 26 Oct 2010 18:44:16 +0200
Subject: [PATCH] fusionee: convert down_interruptible to down in lock_fusionee

---
 drivers/char/fusion/fusionee.c |    5 +----
 1 files changed, 1 insertions(+), 4 deletions(-)

diff --git a/drivers/char/fusion/fusionee.c b/drivers/char/fusion/fusionee.c
index 7cf2f0a..d187853 100644
--- a/drivers/char/fusion/fusionee.c
+++ b/drivers/char/fusion/fusionee.c
@@ -708,10 +708,7 @@ static int lock_fusionee(FusionDev * dev, FusionID id, Fusionee ** ret_fusionee)
 
 	fusion_list_move_to_front(&dev->fusionee.list, &fusionee->link);
 
-	if (down_interruptible(&fusionee->lock)) {
-		up(&dev->fusionee.lock);
-		return -EINTR;
-	}
+	down(&fusionee->lock);
 
 	up(&dev->fusionee.lock);
 
-- 
1.7.3.3

