From 32503a8cd920fb07fddfeeeed332d7f30326dea9 Mon Sep 17 00:00:00 2001
From: Yue Tao <Yue.Tao@windriver.com>
Date: Tue, 29 Mar 2011 11:21:18 +0800
Subject: [PATCH] Including the asm-offsets.h head file

The macro __NR_syscall_max is defined in asm-offsets.h file for
x86_64 arch along with preempt_rt kernel type.
---
 kernel/auditsc.c |    1 +
 1 files changed, 1 insertions(+), 0 deletions(-)

diff --git a/kernel/auditsc.c b/kernel/auditsc.c
index a981d34..14892b3 100644
--- a/kernel/auditsc.c
+++ b/kernel/auditsc.c
@@ -57,6 +57,7 @@
 #include <linux/time.h>
 #include <linux/netlink.h>
 #include <linux/compiler.h>
+#include <asm/asm-offsets.h>
 #include <asm/unistd.h>
 #include <linux/security.h>
 #include <linux/list.h>
-- 
1.7.0.4

