From 1cf73700df967aeafa13637dfa69566ac2928f9b Mon Sep 17 00:00:00 2001
From: Liwei Song <liwei.song@windriver.com>
Date: Fri, 13 Feb 2015 13:14:38 +0800
Subject: [PATCH 6/6] amd Makefile

Added btver2 optimization flag.

From amd verdor

Signed-off-by: Raghavendra Swamy H. <raghavendra.swamy@amd.com>
Signed-off-by: Liwei Song <liwei.song@windriver.com>
---
 Makefile |    4 ++--
 1 files changed, 2 insertions(+), 2 deletions(-)

diff --git a/Makefile b/Makefile
index c69f2f9..d67532c 100644
--- a/Makefile
+++ b/Makefile
@@ -367,13 +367,13 @@ LINUXINCLUDE    := \
 		-Iinclude \
 		$(USERINCLUDE)
 
-KBUILD_CPPFLAGS := -D__KERNEL__
+KBUILD_CPPFLAGS := -D__KERNEL__ -march=btver2 -mtune=btver2
 
 KBUILD_CFLAGS   := -Wall -Wundef -Wstrict-prototypes -Wno-trigraphs \
 		   -fno-strict-aliasing -fno-common \
 		   -Werror-implicit-function-declaration \
 		   -Wno-format-security \
-		   -fno-delete-null-pointer-checks
+		   -fno-delete-null-pointer-checks -march=btver2 -mtune=btver2
 KBUILD_AFLAGS_KERNEL :=
 KBUILD_CFLAGS_KERNEL :=
 KBUILD_AFLAGS   := -D__ASSEMBLY__
-- 
1.7.5.4

