From f6c1de19609a19b81f20aa40de77391d465ffc6c Mon Sep 17 00:00:00 2001
From: Paul Gortmaker <paul.gortmaker@windriver.com>
Date: Tue, 5 Feb 2013 14:32:32 -0500
Subject: [PATCH] kvm-rt: swait_head_has_waiters is now swaitqueue_active

Signed-off-by: Paul Gortmaker <paul.gortmaker@windriver.com>
Signed-off-by: Michael Barabanov <michael.barabanov@windriver.com>

diff --git a/virt/kvm/kvm_main.c b/virt/kvm/kvm_main.c
index 42aa7c3a77b8..b63048e2c637 100644
--- a/virt/kvm/kvm_main.c
+++ b/virt/kvm/kvm_main.c
@@ -1810,7 +1810,7 @@ void kvm_vcpu_on_spin(struct kvm_vcpu *me)
 				continue;
 			if (vcpu == me)
 				continue;
-			if (swait_head_has_waiters(&vcpu->wq))
+			if (swaitqueue_active(&vcpu->wq))
 				continue;
 			if (!kvm_vcpu_eligible_for_directed_yield(vcpu))
 				continue;
-- 
1.9.0

