From d7893f4a1ba9a23fd7261b954364ffb0f6a7c290 Mon Sep 17 00:00:00 2001
From: Li Zhou <li.zhou@windriver.com>
Date: Mon, 18 Apr 2016 11:10:14 +0800
Subject: [PATCH] locking-selftest: change some rwlock tests' expected results
 for rt kernel

Change some rwlock related tests' expected results because rt kernel's
rwlock works in a different way from standard kernel. rwlocks which are
implemented as busy loops on the mainline kernel, are sleeping mutexes
in the rt kernel.

Signed-off-by: Li Zhou <li.zhou@windriver.com>
---
 lib/locking-selftest.c |   16 +++++++++++++++-
 1 files changed, 15 insertions(+), 1 deletions(-)

diff --git a/lib/locking-selftest.c b/lib/locking-selftest.c
index 7fb623d..36b0b36 100644
--- a/lib/locking-selftest.c
+++ b/lib/locking-selftest.c
@@ -1859,7 +1859,7 @@ void locking_selftest(void)
 
 	init_shared_classes();
 	debug_locks_silent = !debug_locks_verbose;
-
+#ifndef CONFIG_PREEMPT_RT_FULL
 	DO_TESTCASE_6R("A-A deadlock", AA);
 	DO_TESTCASE_6R("A-B-B-A deadlock", ABBA);
 	DO_TESTCASE_6R("A-B-B-C-C-A deadlock", ABBCCA);
@@ -1868,6 +1868,16 @@ void locking_selftest(void)
 	DO_TESTCASE_6R("A-B-C-D-B-D-D-A deadlock", ABCDBDDA);
 	DO_TESTCASE_6R("A-B-C-D-B-C-D-A deadlock", ABCDBCDA);
 	DO_TESTCASE_6("double unlock", double_unlock);
+#else
+	DO_TESTCASE_6("A-A deadlock", AA);
+	DO_TESTCASE_6("A-B-B-A deadlock", ABBA);
+	DO_TESTCASE_6("A-B-B-C-C-A deadlock", ABBCCA);
+	DO_TESTCASE_6("A-B-C-A-B-C deadlock", ABCABC);
+	DO_TESTCASE_6("A-B-B-C-C-D-D-A deadlock", ABBCCDDA);
+	DO_TESTCASE_6("A-B-C-D-B-D-D-A deadlock", ABCDBDDA);
+	DO_TESTCASE_6("A-B-C-D-B-C-D-A deadlock", ABCDBCDA);
+	DO_TESTCASE_6R("double unlock", double_unlock);
+#endif
 	DO_TESTCASE_6("initialize held", init_held);
 	DO_TESTCASE_6_SUCCESS("bad unlock order", bad_unlock_order);
 
@@ -1881,7 +1891,11 @@ void locking_selftest(void)
 
 	print_testname("recursive read-lock #2");
 	printk("             |");
+#ifndef CONFIG_PREEMPT_RT_FULL
 	dotest(rlock_AA1B, SUCCESS, LOCKTYPE_RWLOCK);
+#else
+	dotest(rlock_AA1B, FAILURE, LOCKTYPE_RWLOCK);
+#endif
 	printk("             |");
 	dotest(rsem_AA1B, FAILURE, LOCKTYPE_RWSEM);
 	printk("\n");
-- 
1.7.5.4

