From 83933856fe5825838a20a98530b734998ad65e8e Mon Sep 17 00:00:00 2001
From: fupan li <fupan.li@windriver.com>
Date: Fri, 14 Nov 2014 09:20:25 +0800
Subject: [PATCH 04/15] set DMA_API_DEBUG feature depending on !PREEMPT_RT_FULL

Since DMA_API_DEBUG feature will causes a performance
degradation and a spinlock error as below, so disable
it on RT kernel.

[  161.937566] BUG: sleeping function called from invalid context at kernel/locking/rtmutex.c:905
[  161.937568] in_atomic(): 0, irqs_disabled(): 1, pid: 964, name: dpdk_nic_bind.p
[  161.937572] CPU: 1 PID: 964 Comm: dpdk_nic_bind.p Tainted: G           O 3.14.22-rt9-WR7.0.0.0_preempt-rt #1
[  161.937574] Hardware name: Intel Corporation Broadwell Client platform/WhiteTip Mountain 1, BIOS BDW-E1R1.86C.0080.R01.1406120446 06/12/2014
[  161.937579]  0000000000000000 ffff88023090fd00 ffffffff819ec195 0000000000000000
[  161.937582]  ffff88023090fd18 ffffffff8107edd3 ffffffff821a89d0 ffff88023090fd30
[  161.937584]  ffffffff819f2ba0 ffffffff821a89c0 ffff88023090fda0 ffffffff81407381
[  161.937585] Call Trace:
[  161.937595]  [<ffffffff819ec195>] dump_stack+0x4e/0x7a
[  161.937602]  [<ffffffff8107edd3>] __might_sleep+0xe3/0x160
[  161.937607]  [<ffffffff819f2ba0>] rt_spin_lock+0x20/0x50
[  161.937611]  [<ffffffff81407381>] dma_debug_device_change+0x51/0x1d0
[  161.937617]  [<ffffffff8104d856>] ? unpin_current_cpu+0x16/0x70
[  161.937622]  [<ffffffff810822c0>] ? migrate_enable+0xb0/0x1b0
[  161.937626]  [<ffffffff8107762d>] notifier_call_chain+0x4d/0x70
[  161.937629]  [<ffffffff810779a7>] __blocking_notifier_call_chain+0x47/0x60
[  161.937632]  [<ffffffff810779d6>] blocking_notifier_call_chain+0x16/0x20
[  161.937639]  [<ffffffff81572220>] __device_release_driver+0xd0/0xf0
[  161.937643]  [<ffffffff81572263>] device_release_driver+0x23/0x30
[  161.937646]  [<ffffffff81570fed>] unbind_store+0xed/0x150
[  161.937650]  [<ffffffff81570375>] drv_attr_store+0x25/0x40
[  161.937655]  [<ffffffff811f67da>] sysfs_kf_write+0x3a/0x50
[  161.937660]  [<ffffffff811fa502>] kernfs_fop_write+0xd2/0x140
[  161.937666]  [<ffffffff811882f6>] vfs_write+0xb6/0x1b0
[  161.937671]  [<ffffffff81188c4d>] SyS_write+0x4d/0xc0
[  161.937676]  [<ffffffff819f3dd6>] system_call_fastpath+0x1a/0x1f
[  161.945131] igb_uio 0000:02:00.0: uio device registered with irq 12

Signed-off-by: fupan li <fupan.li@windriver.com>
---
 lib/Kconfig.debug | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/lib/Kconfig.debug b/lib/Kconfig.debug
index cc6911c..7f7cf78 100644
--- a/lib/Kconfig.debug
+++ b/lib/Kconfig.debug
@@ -1878,7 +1878,7 @@ config BUILD_DOCSRC
 
 config DMA_API_DEBUG
 	bool "Enable debugging of DMA-API usage"
-	depends on HAVE_DMA_API_DEBUG
+	depends on HAVE_DMA_API_DEBUG && !PREEMPT_RT_FULL
 	help
 	  Enable this option to debug the use of the DMA API by device drivers.
 	  With this option you will be able to detect common bugs in device
-- 
2.0.2

