From 0b4e777c205eeb418382cb2f6351b38c2cefa8c2 Mon Sep 17 00:00:00 2001
From: Yue Tao <yue.tao@windriver.com>
Date: Tue, 11 Apr 2017 02:53:55 -0700
Subject: [PATCH] The timer_stats_account_hrtimer has been removed by commit

commit dfb4357da6ddbdf57d583ba64361c9d792b0e0b1 upstream

Signed-off-by: Yue Tao <yue.tao@windriver.com>
---
 kernel/time/hrtimer.c |    1 -
 1 files changed, 0 insertions(+), 1 deletions(-)

diff --git a/kernel/time/hrtimer.c b/kernel/time/hrtimer.c
index 11f8e47..df5e0f5 100644
--- a/kernel/time/hrtimer.c
+++ b/kernel/time/hrtimer.c
@@ -1397,7 +1397,6 @@ static void hrtimer_rt_run_pending(void)
 			 */
 			debug_hrtimer_deactivate(timer);
 			__remove_hrtimer(timer, base, HRTIMER_STATE_CALLBACK, 0);
-			timer_stats_account_hrtimer(timer);
 			fn = timer->function;
 
 			raw_spin_unlock_irq(&cpu_base->lock);
-- 
1.7.5.4

