From e12364707fb174ee96d6eb2d36ee73c2ed43aeb2 Mon Sep 17 00:00:00 2001
From: Michel Thebeau <michel.thebeau@windriver.com>
Date: Thu, 25 Nov 2010 11:29:01 -0500
Subject: [PATCH] Define NR_syscalls for audit patch

The patch "[audit] fix kernel panic due to invalid syscall" uses
NR_syscalls but this is not defined.

Signed-off-by: Michel Thebeau <michel.thebeau@windriver.com>
---
 include/asm-x86/unistd_64.h |    2 ++
 1 files changed, 2 insertions(+), 0 deletions(-)

diff --git a/include/asm-x86/unistd_64.h b/include/asm-x86/unistd_64.h
index 9bc435a..9c13345 100644
--- a/include/asm-x86/unistd_64.h
+++ b/include/asm-x86/unistd_64.h
@@ -658,6 +658,8 @@ __SYSCALL(__NR_marker, sys_marker)
 #define __NR_trace				296
 __SYSCALL(__NR_trace, sys_trace)
 
+#define NR_syscalls				297
+
 #ifndef __NO_STUBS
 #define __ARCH_WANT_OLD_READDIR
 #define __ARCH_WANT_OLD_STAT
-- 
1.7.0.4

