From 0ecec192373f90a4441f0bd98e4d69f091f250fe Mon Sep 17 00:00:00 2001
From: Bruce Ashfield <bruce.ashfield@windriver.com>
Date: Mon, 2 Apr 2012 14:17:51 -0400
Subject: [PATCH] perf: temp 3.4 build / cross compile fixups

Signed-off-by: Bruce Ashfield <bruce.ashfield@windriver.com>
---
 tools/perf/Makefile |   10 +++++-----
 1 files changed, 5 insertions(+), 5 deletions(-)

diff --git a/tools/perf/Makefile b/tools/perf/Makefile
index 4f4754a..2c90e6b 100644
--- a/tools/perf/Makefile
+++ b/tools/perf/Makefile
@@ -80,9 +80,9 @@ ifeq ($(ARCH),x86_64)
 endif
 
 # Treat warnings as errors unless directed not to
-ifneq ($(WERROR),0)
-	CFLAGS_WERROR := -Werror
-endif
+#ifneq ($(WERROR),0)
+#	CFLAGS_WERROR := -Werror
+#endif
 
 #
 # Include saner warnings here, which can catch bugs:
@@ -234,8 +234,8 @@ endif
 
 export PERL_PATH
 
-FLEX = $(CROSS_COMPILE)flex
-BISON= $(CROSS_COMPILE)bison
+FLEX = flex
+BISON= bison
 
 event-parser:
 	$(QUIET_BISON)$(BISON) -v util/parse-events.y -d -o $(OUTPUT)util/parse-events-bison.c
-- 
1.7.5.4

