From 62642eaea8b4b549efbff2d2d0516c79c1ab80bf Mon Sep 17 00:00:00 2001
From: Michel Thebeau <michel.thebeau@windriver.com>
Date: Fri, 19 Aug 2011 09:48:46 -0400
Subject: [PATCH] revoke: ignore unimplemented revokeat syscall

The revokeat system call is specifically for some cgl kernels.  It does not
need to be implemented.

Signed-off-by: Michel Thebeau <michel.thebeau@windriver.com>
---
 scripts/checksyscalls.sh |    3 +++
 1 files changed, 3 insertions(+), 0 deletions(-)

diff --git a/scripts/checksyscalls.sh b/scripts/checksyscalls.sh
index 66ad375..7ada84e 100755
--- a/scripts/checksyscalls.sh
+++ b/scripts/checksyscalls.sh
@@ -185,6 +185,9 @@ cat << EOF
 #define __IGNORE_vfork		/* clone */
 #define __IGNORE_wait4		/* waitid */
 
+/* revokeat has limited scope */
+#define __IGNORE_revokeat
+
 /* sync_file_range had a stupid ABI. Allow sync_file_range2 instead */
 #ifdef __NR_sync_file_range2
 #define __IGNORE_sync_file_range
-- 
1.7.0.4

