From 6afd531d0bbbe657597a1935bafd5e6a0ee2d20a Mon Sep 17 00:00:00 2001
From: Bruce Ashfield <bruce.ashfield@windriver.com>
Date: Sun, 27 Mar 2011 23:44:21 -0400
Subject: [PATCH] vbi: only include linux.h for cert hyp builds

Signed-off-by: Bruce Ashfield <bruce.ashfield@windriver.com>
---
 include/vbi/stats.h |    2 ++
 1 files changed, 2 insertions(+), 0 deletions(-)

diff --git a/include/vbi/stats.h b/include/vbi/stats.h
index af3e365..616da29 100644
--- a/include/vbi/stats.h
+++ b/include/vbi/stats.h
@@ -3,8 +3,10 @@
 
 #ifndef _ASMLANGUAGE
 
+#ifdef CONFIG_WRHV_CERT
 /* header file required by the printk, which is used in VBISTAT_VERBOSE */
 #include <linux/kernel.h>
+#endif
 
 /* api statistic structure */
 struct vbi_api_stats{
-- 
1.6.5.2

