From 95524a7d589aac2044951f611d1df30b3400cec6 Mon Sep 17 00:00:00 2001
From: Zumeng Chen <zumeng.chen@windriver.com>
Date: Mon, 31 Jul 2017 15:23:25 +0800
Subject: [PATCH 13/13] arm: ls1021a: avoid the meaningless wakeup resources
 settings.

Directly return if the device node is null since it does not have any
rcpm-wakeup node at all in dts. Actually ttyS0 is not allowed to be a
wakeup resource according to ls1021a SoC reference manual.

Signed-off-by Zumeng Chen <zumeng.chen@windriver.com>
---
 arch/arm/mach-imx/pm-ls1.c |    3 +++
 1 files changed, 3 insertions(+), 0 deletions(-)

diff --git a/arch/arm/mach-imx/pm-ls1.c b/arch/arm/mach-imx/pm-ls1.c
index 486f21b..acd3fb2 100644
--- a/arch/arm/mach-imx/pm-ls1.c
+++ b/arch/arm/mach-imx/pm-ls1.c
@@ -403,6 +403,9 @@ static void ls1_set_power_except(struct device *dev, int on)
 	 * the ippdexpcr0 register, and the third is the value of
 	 * the ippdexpcr1 register.
 	 */
+	if (!dev->of_node)
+		return;
+
 	ret = of_property_read_u32_array(dev->of_node, "rcpm-wakeup",
 						value, SLEEP_ARRAY_SIZE);
 	if (ret) {
-- 
1.7.5.4

