From 263ac5bc04f53aa04efc55a5e22101b814e969c0 Mon Sep 17 00:00:00 2001
From: Bartlomiej Zolnierkiewicz <b.zolnierkie@samsung.com>
Date: Mon, 17 Mar 2014 14:06:57 +0100
Subject: [PATCH 0475/1543] ata: ahci_sunxi: make ahci_sunxi_resume() static

commit 1bf9d885658cbee1bc8e4324d0e27b02b1540d58 upstream

Signed-off-by: Bartlomiej Zolnierkiewicz <b.zolnierkie@samsung.com>
Signed-off-by: Tejun Heo <tj@kernel.org>
Acked-by: Hans de Goede <hdegoede@redhat.com>
---
 drivers/ata/ahci_sunxi.c |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/drivers/ata/ahci_sunxi.c b/drivers/ata/ahci_sunxi.c
index d1bf3f7..0af6cb8 100644
--- a/drivers/ata/ahci_sunxi.c
+++ b/drivers/ata/ahci_sunxi.c
@@ -197,7 +197,7 @@ disable_resources:
 }
 
 #ifdef CONFIG_PM_SLEEP
-int ahci_sunxi_resume(struct device *dev)
+static int ahci_sunxi_resume(struct device *dev)
 {
 	struct ata_host *host = dev_get_drvdata(dev);
 	struct ahci_host_priv *hpriv = host->private_data;
-- 
1.7.5.4

