From 58f7161f1396e47b1fa0f44aa1a742bf259b3875 Mon Sep 17 00:00:00 2001
From: Yong Gan <b45748@freescale.com>
Date: Wed, 13 Aug 2014 14:51:34 +0800
Subject: [PATCH 0957/1543] ENGR00325693 [#1318] Add eglSwapInterval support
 in Wayland client

commit c04a6106f18be3d39321b7f58fb0329bef47d7a7 from
git://git.freescale.com/imx/linux-2.6-imx.git

Add new API gcoOS_SetSwapIntervalEx.

Date: Aug 04, 2014
Signed-off-by Yong Gan <yong.gan@freescale.com>

(cherry picked from commit 971632a7fb6d0744ccac563bcdbf6a4decf4e0a1)
---
 .../gpu-viv/hal/kernel/inc/gc_hal_eglplatform.h    |   11 ++++++++++-
 1 files changed, 10 insertions(+), 1 deletions(-)

diff --git a/drivers/mxc/gpu-viv/hal/kernel/inc/gc_hal_eglplatform.h b/drivers/mxc/gpu-viv/hal/kernel/inc/gc_hal_eglplatform.h
index 690c957..411e9f8 100644
--- a/drivers/mxc/gpu-viv/hal/kernel/inc/gc_hal_eglplatform.h
+++ b/drivers/mxc/gpu-viv/hal/kernel/inc/gc_hal_eglplatform.h
@@ -74,7 +74,8 @@ typedef struct _gcsWL_EGL_DISPLAY
    struct wl_display* wl_display;
    struct wl_viv* wl_viv;
    struct wl_registry *registry;
-   struct wl_event_queue    *wl_queue;
+   struct wl_event_queue *wl_queue;
+   gctINT swapInterval;
 } gcsWL_EGL_DISPLAY;
 
 typedef struct _gcsWL_EGL_BUFFER_INFO
@@ -87,6 +88,7 @@ typedef struct _gcsWL_EGL_BUFFER_INFO
    gcePOOL pool;
    gctUINT bytes;
    gcoSURF surface;
+   gcoSURF attached_surface;
    gctINT32 invalidate;
    gctBOOL locked;
 } gcsWL_EGL_BUFFER_INFO;
@@ -294,6 +296,13 @@ gcoOS_SetSwapInterval(
 );
 
 gceSTATUS
+gcoOS_SetSwapIntervalEx(
+    IN HALNativeDisplayType Display,
+    IN gctINT Interval,
+    IN gctPOINTER localDisplay
+);
+
+gceSTATUS
 gcoOS_GetSwapInterval(
     IN HALNativeDisplayType Display,
     IN gctINT_PTR Min,
-- 
1.7.5.4

