From 48cc2225b0685f848b860ca55f863dd54b62f657 Mon Sep 17 00:00:00 2001
From: Li Jun <b47624@freescale.com>
Date: Wed, 25 Jun 2014 15:28:45 +0800
Subject: [PATCH 1401/1543] ENGR00319720-9 Documentation: usb: chipidea:
 Update test procedure for HNP polling support

commit be647b3bdca7fd7bff7736183cf4aedb411a1d23 from
git://git.freescale.com/imx/linux-2.6-imx.git

Update HNP test procedure as HNP polling is supported.

Signed-off-by: Li Jun <b47624@freescale.com>
---
 Documentation/usb/chipidea.txt |    9 ++++-----
 1 files changed, 4 insertions(+), 5 deletions(-)

diff --git a/Documentation/usb/chipidea.txt b/Documentation/usb/chipidea.txt
index 3f848c1..95d9170 100644
--- a/Documentation/usb/chipidea.txt
+++ b/Documentation/usb/chipidea.txt
@@ -26,14 +26,13 @@ cat /sys/kernel/debug/ci_hdrc.0/registers
    On B-device:
    echo 1 > /sys/bus/platform/devices/ci_hdrc.0/inputs/b_bus_req
 
-   if HNP polling is not supported, also need:
-   On A-device:
-   echo 0 > /sys/bus/platform/devices/ci_hdrc.0/inputs/a_bus_req
-
    B-device should take host role and enumrate A-device.
 
 4) A-device switch back to host.
-   On B-device:
+   On A-device:
+   echo 1 > /sys/bus/platform/devices/ci_hdrc.0/inputs/a_bus_req
+
+   or, on B-device:
    echo 0 > /sys/bus/platform/devices/ci_hdrc.0/inputs/b_bus_req
 
    A-device should switch back to host and enumrate B-device.
-- 
1.7.5.4

