From db5e7ef71a86b002fa3d1f94a7ec4df7b92777e5 Mon Sep 17 00:00:00 2001
From: Li Jun <B47624@freescale.com>
Date: Sun, 28 Sep 2014 13:55:01 +0800
Subject: [PATCH 1449/1543] MLK-9618-1 doc: usb: chipidea: select gadget
 drivers for otg compliance test

commit 134822a344dad59a86bdfbb3b1177cf26096f664 from
git://git.freescale.com/imx/linux-2.6-imx.git

This patch adds guide for selecting available gadget drivers for otg and EH
compliance tests.

Acked-by: Peter Chen <peter.chen@freescale.com>
Signed-off-by: Li Jun <b47624@freescale.com>
(cherry picked from commit 520cac9e4fe938887dd45b5b4df6c8e35e125a59)
---
 Documentation/usb/chipidea.txt |    8 ++++++++
 1 files changed, 8 insertions(+), 0 deletions(-)

diff --git a/Documentation/usb/chipidea.txt b/Documentation/usb/chipidea.txt
index 95d9170..4e42821 100644
--- a/Documentation/usb/chipidea.txt
+++ b/Documentation/usb/chipidea.txt
@@ -69,6 +69,14 @@ cat /sys/kernel/debug/ci_hdrc.0/registers
 "On-The-Go and Embedded Host Supplement to the USB Revision 2.0 Specification
 July 27, 2012 Revision 2.0 version 1.1a"
 
+1.4 OTG compliance test
+----------------------
+Only below 3 popular gadget drivers are declared to be USB OTG and EH 2.0
+compliant(with otg descriptor comply with USB OTG and EH 2.0 as a peripheral):
+- mass storage
+- ether
+- serial
+
 2. How to enable USB as system wakeup source
 -----------------------------------
 Below is the example for how to enable USB as system wakeup source
-- 
1.7.5.4

