From 4bd15b5383667edfcd08e701204d2f5f6b66cf62 Mon Sep 17 00:00:00 2001
From: Kumar Gala <galak@kernel.crashing.org>
Date: Fri, 5 Feb 2010 13:25:37 -0600
Subject: [PATCH] Add padding when we allocate the skb for DPA_BP_HEAD

The additional padding for DPA_BP_HEAD prevents the skb_cow call in
the IP forward path from having to an extra copy.
[Kevin: Original patch taken from Freescale p4080 SDK 2.0.3 tar
package, apply cleanly to kernel 2.6.27]
Integrated-by: Kevin Hao <kexin.hao@windriver.com>
---
 drivers/net/dpa/dpa.c |    4 ++--
 1 files changed, 2 insertions(+), 2 deletions(-)

diff --git a/drivers/net/dpa/dpa.c b/drivers/net/dpa/dpa.c
index 39c0d33..e5550c5 100644
--- a/drivers/net/dpa/dpa.c
+++ b/drivers/net/dpa/dpa.c
@@ -1500,7 +1500,7 @@ static void __hot _dpa_rx(struct net_device		*net_dev,
 		size = dpa_fd_length(&dpa_fd->fd);
 
 	if (skb == NULL) {
-		skb = __netdev_alloc_skb(net_dev, NET_IP_ALIGN + size, GFP_DPA);
+		skb = __netdev_alloc_skb(net_dev, DPA_BP_HEAD + NET_IP_ALIGN + size, GFP_DPA);
 		if (unlikely(skb == NULL)) {
 			if (netif_msg_rx_err(priv))
 				cpu_netdev_err(net_dev, "%s:%hu:%s(): "
@@ -1513,7 +1513,7 @@ static void __hot _dpa_rx(struct net_device		*net_dev,
 		}
 	}
 
-	skb_reserve(skb, NET_IP_ALIGN);
+	skb_reserve(skb, NET_IP_ALIGN+DPA_BP_HEAD);
 
 	/* Fill the SKB */
 	if (dpa_fd->fd.format == qm_fd_sg)
-- 
1.6.0.4

