From d475ac9d1632fa7910e51fb942b62103260eafa2 Mon Sep 17 00:00:00 2001
From: Andrew Liu <shengping.liu@windriver.com>
Date: Wed, 11 Aug 2010 17:46:16 +0800
Subject: [PATCH] Remove kernel call trace in preempt_rt env

When smp_processor_id are invoked with preempt enabled,
will get a call trace. Replacing smp_processor_id with
raw_smp_processor_id will fix  this issue.

Acked-by: Daniel Stenberg <daniel.stenberg@windriver.com>
Integrated-by: Andrew Liu <shengping.liu@windriver.com>
---
 .../user/env/linux/kernel/2.6/xx/xx_linux.c        |    4 ++--
 1 files changed, 2 insertions(+), 2 deletions(-)

diff --git a/drivers/net/dpa/NetCommSw/user/env/linux/kernel/2.6/xx/xx_linux.c b/drivers/net/dpa/NetCommSw/user/env/linux/kernel/2.6/xx/xx_linux.c
index ff985bd..655bf61 100644
--- a/drivers/net/dpa/NetCommSw/user/env/linux/kernel/2.6/xx/xx_linux.c
+++ b/drivers/net/dpa/NetCommSw/user/env/linux/kernel/2.6/xx/xx_linux.c
@@ -202,7 +202,7 @@ void XX_Print(char *str, ...)
 #ifdef CONFIG_SMP
     if (vsnprintf (buf, BUF_SIZE, str, args) >= BUF_SIZE)
         printk(KERN_WARNING "Illegal string to print!\n    more than %d characters.\n\tString was not printed completelly.\n", BUF_SIZE);
-    printk (KERN_CRIT "cpu%d/%d: %s",hard_smp_processor_id(), smp_processor_id(), buf);
+    printk(KERN_CRIT "cpu%d/%d: %s", hard_smp_processor_id(), raw_smp_processor_id(), buf);
 #else
     vprintk(str, args);
 #endif /* CONFIG_SMP */
@@ -220,7 +220,7 @@ void XX_Fprint(void *file, char *str, ...)
 #ifdef CONFIG_SMP
     if (vsnprintf (buf, BUF_SIZE, str, args) >= BUF_SIZE)
         printk(KERN_WARNING "Illegal string to print!\n    more than %d characters.\n\tString was not printed completelly.\n", BUF_SIZE);
-    printk (KERN_CRIT "cpu%d/%d: %s",hard_smp_processor_id(), smp_processor_id(), buf);
+    printk(KERN_CRIT "cpu%d/%d: %s", hard_smp_processor_id(), raw_smp_processor_id(), buf);
 #else
     vprintk(str, args);
 #endif /* CONFIG_SMP */
-- 
1.6.0.3

