From 6e2b74e1ae52632300996beacb1f7e548af58818 Mon Sep 17 00:00:00 2001
From: Kumar Gala <galak@kernel.crashing.org>
Date: Tue, 9 Feb 2010 10:02:26 -0600
Subject: [PATCH] simplfy code for dpa_fd_length calculation

[Kevin: Original patch taken form Freescale moto-patches-2.tar.gz tar
package, apply cleanly to kernel 2.6.27]
Integrated-by: Kevin Hao <kexin.hao@windriver.com>
---
 drivers/net/dpa/dpa.c |   12 ++----------
 1 files changed, 2 insertions(+), 10 deletions(-)

diff --git a/drivers/net/dpa/dpa.c b/drivers/net/dpa/dpa.c
index 64b7431..c8f587b 100644
--- a/drivers/net/dpa/dpa.c
+++ b/drivers/net/dpa/dpa.c
@@ -623,18 +623,10 @@ struct dpa_fd {
 static inline ssize_t __const __must_check __attribute__((nonnull))
 dpa_fd_length(const struct qm_fd *fd)
 {
-	switch (fd->format) {
-	case qm_fd_contig:
-	case qm_fd_sg:
+	if ((fd->format & 0x3) == 0)
 		return fd->length20;
-	case qm_fd_contig_big:
-	case qm_fd_sg_big:
+	else
 		return fd->length29;
-	case qm_fd_compound:
-		return fd->cong_weight;
-	}
-	BUG();
-	return -EINVAL;
 }
 
 static inline ssize_t __const __must_check __attribute__((nonnull))
-- 
1.6.0.4

