From 57979375d86a94dcfb5b7b361a980d4637ecc60a Mon Sep 17 00:00:00 2001
From: Jack Tan <jack.tan@windriver.com>
Date: Fri, 11 Jul 2008 14:04:40 +0800
Subject: [PATCH] nec emma3p 8250 fixup


Signed-off-by: Jack Tan <jack.tan@windriver.com>
---
 drivers/serial/8250.c |    5 +++++
 1 files changed, 5 insertions(+), 0 deletions(-)

diff --git a/drivers/serial/8250.c b/drivers/serial/8250.c
index 59f09c7..e9b61d7 100644
--- a/drivers/serial/8250.c
+++ b/drivers/serial/8250.c
@@ -1468,6 +1468,11 @@ static irqreturn_t serial8250_interrupt(int irq, void *dev_id)
 		up = list_entry(l, struct uart_8250_port, list);
 
 		iir = serial_in(up, UART_IIR);
+#if defined(CONFIG_EMMA3P)
+       /* Dummy Read for EMMA UART IIR busy */
+		if ((iir & 0x07) == 0x07)
+			serial_in(up, 31);
+#endif
 		if (!(iir & UART_IIR_NO_INT)) {
 			serial8250_handle_port(up);
 
-- 
1.5.5.4

