From ea73656aad11d6c09105fa1c140fef51e0caf00d Mon Sep 17 00:00:00 2001
From: Yang Shi <yang.shi@windriver.com>
Date: Thu, 14 Jun 2012 11:01:18 -0700
Subject: [PATCH 13/13] ck: ck2-version

Integrating and porting from:
  http://ck.kolivas.org/patches/3.0/3.4/3.4-ck2/patches/

[Add -ck version in inconspicuous place where it will merge relatively easily
with later kernel versions.

-ck]

Signed-off-by: Yang Shi <yang.shi@windriver.com>
---
 Makefile |    4 ++++
 1 files changed, 4 insertions(+), 0 deletions(-)

diff --git a/Makefile b/Makefile
index 96f1ff3..53eeb8a 100644
--- a/Makefile
+++ b/Makefile
@@ -10,6 +10,10 @@ NAME = Saber-toothed Squirrel
 # Comments in this file are targeted only to the developer, do not
 # expect to learn how to build the kernel reading this file.
 
+CKVERSION = -ck2
+CKNAME = BFS Powered
+EXTRAVERSION := $(EXTRAVERSION)$(CKVERSION)
+
 # Do not:
 # o  use make's built-in rules and variables
 #    (this increases performance and avoids hard-to-debug behaviour);
-- 
1.7.5.4

