From 090e0233da761086b2382384f969210ecf85b612 Mon Sep 17 00:00:00 2001
From: Zumeng Chen <zumeng.chen@windriver.com>
Date: Fri, 9 Apr 2010 12:47:01 +0800
Subject: [PATCH] BFS: wind river fixups

Signed-off-by: Bruce Ashfield <bruce.ashfield@windriver.com>
---
 ltt/ltt-statedump.c |    4 ----
 1 files changed, 0 insertions(+), 4 deletions(-)

diff --git a/ltt/ltt-statedump.c b/ltt/ltt-statedump.c
index a387078..3161bc1 100644
--- a/ltt/ltt-statedump.c
+++ b/ltt/ltt-statedump.c
@@ -271,10 +271,6 @@ ltt_enumerate_process_states(struct ltt_probe_private_data *call_data)
 		else if (t->exit_state == EXIT_DEAD)
 			status = LTTNG_DEAD;
 		else if (t->state == TASK_RUNNING) {
-			/* Is this a forked child that has not run yet? */
-			if (list_empty(&t->rt.run_list))
-				status = LTTNG_WAIT_FORK;
-			else
 				/*
 				 * All tasks are considered as wait_cpu;
 				 * the viewer will sort out if the task was
-- 
1.6.0.4

