From a08bc359afd525ac67e7419a9f786eac7e9fada2 Mon Sep 17 00:00:00 2001
From: Siva Pochiraju <sivap@netlogicmicro.com>
Date: Wed, 7 Jul 2010 13:58:11 -0700
Subject: [PATCH 0629/1532] Added shared-mem and heap-size for netos purposes

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
---
 dts/xlp.dts | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/dts/xlp.dts b/dts/xlp.dts
index 883338c..704a4b0 100644
--- a/dts/xlp.dts
+++ b/dts/xlp.dts
@@ -1,4 +1,5 @@
 /* XLP8XX Device Tree Source 
+ *
  */
 
 /dts-v1/;
@@ -14,6 +15,8 @@
 
 		dom@0 {
 			device_type = "domain";
+			heap-size = <0x0 0x4000000>;
+			shared-mem = <0x0 0x80000000 0x0 0x20000000>;
 			cpu {
 				onlinemask = <0xff>;
 			};
-- 
1.9.1

