From 0a73c78d9bacb9d1d2e9d13956c4ad042c48743b Mon Sep 17 00:00:00 2001
From: Siva Pochiraju <sivap@netlogicmicro.com>
Date: Fri, 16 Jul 2010 16:27:53 -0700
Subject: [PATCH 0632/1532] Fixed compiler warning

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
---
 libfdt/fdt.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/libfdt/fdt.c b/libfdt/fdt.c
index 2acaec5..bbbda5c 100644
--- a/libfdt/fdt.c
+++ b/libfdt/fdt.c
@@ -79,7 +79,7 @@ const void *fdt_offset_ptr(const void *fdt, int offset, int len)
 	const char *p;
 
 	if (fdt_version(fdt) >= 0x11)
-		if (((offset + len) < offset)
+		if ((((unsigned int) offset + (unsigned int) len) < (unsigned int) offset)
 		    || ((offset + len) > fdt_size_dt_struct(fdt)))
 			return NULL;
 
-- 
1.9.1

