From b3aeeece4179e652d8f5c8b6eabe745bfcc8efaf Mon Sep 17 00:00:00 2001
From: Siva Pochiraju <sivap@netlogicmicro.com>
Date: Sat, 7 Aug 2010 13:57:13 -0700
Subject: [PATCH 0634/1532] Added rx_vc and fb_vc for cpu

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
---
 dts/xlp.dts | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/dts/xlp.dts b/dts/xlp.dts
index e20e90f..1dae99c 100644
--- a/dts/xlp.dts
+++ b/dts/xlp.dts
@@ -19,6 +19,8 @@
 			#size-cells = <1>;
 			cpu {
 				onlinemask = <0xff>;
+				nae-rx-vc = <0>;
+				nae-fb-vc = <3>;
 			};
 			uart {
 				id = <0>;
-- 
1.9.1

