From 086ea809e7bd33ad06b054b983679e318cbebced Mon Sep 17 00:00:00 2001
From: Venu Vadapalli <vvadapalli@netlogicmicro.com>
Date: Wed, 1 Sep 2010 23:01:22 -0700
Subject: [PATCH 0650/1532] Decrease heap size for NetOS domains and added Xen
 configuration files (these will be autogenerated from FDT files in future)

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
---
 dts/netos.dts               |  8 ++++----
 dts/sysconfig-xen-dom-1.cfg | 25 +++++++++++++++++++++++++
 dts/sysconfig-xen-dom-2.cfg | 24 ++++++++++++++++++++++++
 dts/sysconfig-xen.dts       | 10 +++++-----
 4 files changed, 58 insertions(+), 9 deletions(-)
 create mode 100644 dts/sysconfig-xen-dom-1.cfg
 create mode 100644 dts/sysconfig-xen-dom-2.cfg

diff --git a/dts/netos.dts b/dts/netos.dts
index 2a359a4..5db03b9 100644
--- a/dts/netos.dts
+++ b/dts/netos.dts
@@ -14,7 +14,8 @@
 		hypervisor-name = "Xen";
 		cpu_possible_map = <0xff>;
 		is_dom0_hvm = <0>;
-		bootargs = "dom0_loadaddr=0x09000000 dom0_cpumask=0x03 -- ";
+		bootargs = "dom0_loadaddr=0x09000000 dom0_cpumask=0x01 -- ";
+		domain_heap = <0x80000000 0x20000000>;
 	};
 
 	doms {
@@ -34,14 +35,13 @@
 			app-loadaddr = "0x60000000";
 
 			cpu {
-				onlinemask = <0x03>;
+				onlinemask = <0x01>;
 			};
 			uart {
 				id = <0>;
-				sharedcfg = <0x1f000000>;
 			};
 			memory {
-				reg = <0x30000000 0x10000000>;  // 128M @ 768M
+				reg = <0x30000000 0x08000000>;  // 128M @ 768M
 			};
 		};
 	};
diff --git a/dts/sysconfig-xen-dom-1.cfg b/dts/sysconfig-xen-dom-1.cfg
new file mode 100644
index 0000000..c458b30
--- /dev/null
+++ b/dts/sysconfig-xen-dom-1.cfg
@@ -0,0 +1,25 @@
+#  -*- mode: python; -*-
+
+kernel = "/tmp/xen/netos.elf"
+
+# Initial memory allocation (in megabytes) for the new domain.
+#
+# WARNING: Creating a domain with insufficient memory may cause out of
+#          memory errors. The domain needs enough memory to boot kernel
+#          and modules. Allocating less than 32MBs is not recommended.
+memory = 32
+
+# A name for your domain. All domains must have different names.
+name = "NetOS@1"
+
+# List of which CPUS this domain is allowed to use, default Xen picks
+#cpus = ""         # leave to Xen to pick
+#cpus = "0"        # all vcpus run on CPU0
+#cpus = "0-3,5,^1" # all vcpus run on cpus 0,2,3,5
+#cpus = ["3", "2"] # VCPU0 runs on CPU1, VCPU1 runs on CPU2
+cpus = ["128", "64"] # VCPU0 runs on CPU1
+
+# Number of Virtual CPUS to use, default is 1
+vcpus = 2
+
+
diff --git a/dts/sysconfig-xen-dom-2.cfg b/dts/sysconfig-xen-dom-2.cfg
new file mode 100644
index 0000000..bcc687f
--- /dev/null
+++ b/dts/sysconfig-xen-dom-2.cfg
@@ -0,0 +1,24 @@
+#  -*- mode: python; -*-
+
+kernel = "/tmp/xen/netos.elf"
+
+# Initial memory allocation (in megabytes) for the new domain.
+#
+# WARNING: Creating a domain with insufficient memory may cause out of
+#          memory errors. The domain needs enough memory to boot kernel
+#          and modules. Allocating less than 32MBs is not recommended.
+memory = 32
+
+# A name for your domain. All domains must have different names.
+name = "NetOS@2"
+
+# List of which CPUS this domain is allowed to use, default Xen picks
+#cpus = ""         # leave to Xen to pick
+#cpus = "0"        # all vcpus run on CPU0
+#cpus = "0-3,5,^1" # all vcpus run on cpus 0,2,3,5
+#cpus = ["3", "2"] # VCPU0 runs on CPU1, VCPU1 runs on CPU2
+cpus = ["32", "16"] # VCPU0 runs on CPU1
+
+# Number of Virtual CPUS to use, default is 1
+vcpus = 2
+
diff --git a/dts/sysconfig-xen.dts b/dts/sysconfig-xen.dts
index 459b945..bd098d8 100644
--- a/dts/sysconfig-xen.dts
+++ b/dts/sysconfig-xen.dts
@@ -46,7 +46,7 @@
 			device_type = "domain";
 			os = "netos";
 
-			heap-size = <0x0 0x4000000>;                   // 64M
+			heap-size = <0x0 0x2000000>;                   // 32M
 			shared-mem = <0x0 0x40000000 0x0 0x08000000>;  // 128M @ 1GB
 
 			#address-cells = <1>;
@@ -63,24 +63,24 @@
 				sharedcfg = <0x1f000000>;
 			};
 			memory {
-				reg = <0x30000000 0x10000000>;  // 256M @ 768M
+				reg = <0x30000000 0x08000000>;  // 128M @ 768M
 			};
 		};
 		dom@2 {
 			device_type = "domain";
 			os = "netos";
 
-			heap-size = <0x0 0x4000000>;                   // 64M
+			heap-size = <0x0 0x2000000>;                   // 32M
 			shared-mem = <0x0 0x48000000 0x0 0x08000000>;  // 128M @ 1.128GB
 
 			#address-cells = <1>;
 			#size-cells = <1>;
 
-			app-bootcmd = "ampload";
+			app-bootcmd = "smpload";
 			app-loadaddr = "0x60000000";
 
 			cpu {
-				onlinemask = <0x04>;
+				onlinemask = <0x30>;
 			};
 			uart {
 				id = <0>;
-- 
1.9.1

