From 7ae79268b6e9d819771d0063bcfe2ba4f34c9c44 Mon Sep 17 00:00:00 2001
From: Zi Shen Lim <zlim@netlogicmicro.com>
Date: Wed, 15 Sep 2010 15:00:45 -0700
Subject: [PATCH 0653/1532] Remove deprecated dts directory

These system configuration files now reside in the
'sysconfig' repository.
[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
---
 dts/netos.dts     |  52 ---------
 dts/sysconfig.dts | 257 -------------------------------------------
 dts/tiny.dts      | 153 --------------------------
 dts/xlp.dts       |  40 -------
 dts/xlr.dts       | 320 ------------------------------------------------------
 5 files changed, 822 deletions(-)
 delete mode 100644 dts/netos.dts
 delete mode 100644 dts/sysconfig.dts
 delete mode 100644 dts/tiny.dts
 delete mode 100644 dts/xlp.dts
 delete mode 100644 dts/xlr.dts

diff --git a/dts/netos.dts b/dts/netos.dts
deleted file mode 100644
index 5db03b9..0000000
--- a/dts/netos.dts
+++ /dev/null
@@ -1,52 +0,0 @@
-/* XLP8XX Device Tree Source 
- *
- */
-
-/dts-v1/;
-/ {
-	model = "MIPS,XLP8XX";
-	compatible = "NETL,XLP8XX_A0";
-
-	#address-cells = <1>;
-	#size-cells = <1>;
-
-	hypervisor {
-		hypervisor-name = "Xen";
-		cpu_possible_map = <0xff>;
-		is_dom0_hvm = <0>;
-		bootargs = "dom0_loadaddr=0x09000000 dom0_cpumask=0x01 -- ";
-		domain_heap = <0x80000000 0x20000000>;
-	};
-
-	doms {
-		#address-cells = <1>;
-		#size-cells = <1>;
-		dom@0 {
-			device_type = "domain";
-			os = "netos";
-
-			heap-size = <0x0 0x4000000>; // 64M
-			shared-mem = <0x0 0x40000000 0x0 0x08000000>; // 128M @ 1GB
-
-			#address-cells = <1>;
-			#size-cells = <1>;
-
-			app-bootcmd = "ampload";
-			app-loadaddr = "0x60000000";
-
-			cpu {
-				onlinemask = <0x01>;
-			};
-			uart {
-				id = <0>;
-			};
-			memory {
-				reg = <0x30000000 0x08000000>;  // 128M @ 768M
-			};
-		};
-	};
-
-	chosen {
-		bootargs = " xlp_noi2c console=ttyS0,115200 rdinit=/sbin/init";
-	};
-};
diff --git a/dts/sysconfig.dts b/dts/sysconfig.dts
deleted file mode 100644
index a2226a5..0000000
--- a/dts/sysconfig.dts
+++ /dev/null
@@ -1,257 +0,0 @@
-/* XLP8XX Device Tree Source
- *
- */
-
-/dts-v1/;
-/ {
-	model = "MIPS,XLP8XX";
-	compatible = "NETL,XLP8XX_A0";
-
-	hypervisor {
-
-		hypervisor-name = "Xen";
-
-		cpu_possible_map = <0xff>;
-
-		is_dom0_hvm = <1>;
-
-		bootargs = "dom0_loadaddr=0x09000000 dom0_cpumask=0x0f -- ";
-
-		uart {
-			/* Shared ports, not owner */
-			port@0 {};
-		};
-
-		memory {
-			map = <&mem_hypervisor>;
-		};
-	};
-
-	dom@0 {
-		os = "linux";
-
-		bootargs = " xlp_noi2c console=ttyS0,115200 rdinit=/sbin/init";
-
-		cpu {
-			set = <&cpu_dom0>;
-		};
-
-		uart {
-			/* owner, could be shared - read sharedcfg in sysconfig */
-			port@0 { owner; };
-			port@1 { owner; };
-		};
-
-		memory {
-			map = <&mem_dom0_lowmem &mem_dom0_highmem>;
-		};
-
-		pic { };
-
-		nae@mode0 {
-			/* Shared ports, not owner */
-			port@0 {};
-			port@1 {};
-			port@2 {};
-			port@3 {};
-		};
-		nae@mode1 {
-			/* owner, could be shared - read sharedcfg in sysconfig */
-			port@0 { owner;	};
-			port@1 { owner;	};
-		};
-	};
-
-	dom@1 {
-		os = "netos";
-
-		heap-size = <0x0 0x4000000>;
-		shared-mem = <&mem_shared>;
-
-		#address-cells = <1>;
-		#size-cells = <1>;
-
-		cpu {
-			set = <&cpu_dom1>;
-		};
-		uart {
-			/* Shared ports, not owner */
-			port@0 {};
-			port@1 {};
-		};
-		memory {
-			map = <&mem_dom1>;
-		};
-		nae@mode0 {
-			/* owner, could be shared - read sharedcfg in sysconfig */
-			port@0 { owner;	};
-			port@1 { owner;	};
-			port@2 { owner;	};
-			port@3 { owner;	};
-		};
-		nae@mode1 {
-			/* Shared ports, not owner */
-			port@0 {};
-			port@1 {};
-		};
-	};
-
-	soc {
-
-		cpu {
-			#mask-cells = <1>;
-
-			cpu_dom0: set@0 {
-				mask = <0x0f>;
-			};
-			cpu_dom1: set@1 {
-				mask = <0xf0>;
-			};
-		};
-
-		uart {
-			port@0 {
-				baud-rate = <115200>;
-				sharedcfg { memory = <&mem_uart0>; };
-			};
-			port@1 {
-				baud-rate = <115200>;
-				sharedcfg { memory = <&mem_uart1>; };
-			};
-		};
-
-		memory {
-			#address-cells = <1>;
-			#size-cells = <1>;
-
-			mem_expvec: chunk@0 {
-				range = <0x00000000 0x00100000>;	// 1M @ 0M
-				sharedcfg {};
-			};
-			mem_syscfgdb: chunk@1 {
-				range = <0x00100000 0x00100000>;	// 1M @ 1M
-				sharedcfg {};
-			};
-			mem_hypervisor: chunk@2 {
-				range = <0x00100000 0x03E000000>;	// 63M @ 2M
-				sharedcfg {};
-			};
-			mem_dom0_lowmem: chunk@3 {
-				range = <0x04000000 0x0C000000>;	// 192M at 64M
-				sharedcfg {};
-			};
-			mem_uart0: chunk@4 {
-				range = <0x1F000000 0x00400000>;  	// 4M @ 496M
-				sharedcfg {};
-			};
-			mem_uart1: chunk@5 {
-				range = <0x1F400000 0x00400000>;  	// 4M @ 500M
-				sharedcfg {};
-			};
-			mem_dom0_highmem: chunk@6 {
-				range = <0x20000000 0x20000000>;  	// 512M at 512M
-				sharedcfg {};
-			};
-			mem_dom1: chunk@7 {
-				range = <0x30000000 0x10000000>;  	// 256M @ 768M
-				sharedcfg {};
-			};
-			mem_shared: chunk@8 {
-				#address-cells = <2>;
-				#size-cells = <2>;
-
-				shared-mem = <0x0 0x80000000 0x0 0x20000000>; // 512M @ 2G
-				sharedcfg {};
-			};
-		};
-
-		nae@mode0 {
-			#index-cells = <1>;
-			#value-cells = <1>;
-
-                        num-ports = <4>;
-                        start-port-id = <0>;
-                        num-port-regs = <4>;
-			port@0 {
-
-	                        regs = <
-        	                        0x35 0xd
-                	                0x1f 0xd
-                        	        0x8a 0xd
-                                	0x8a 0xd
-	                        >;
-
-				sharedcfg { };
-			};
-                        port@1 {
-
-				regs = <
-	                                0x35 0x35
-        	                        0x1f 0x35
-                	                0x8a 0x35
-                        	        0x8a 0x1f
-	                        >;
-
-				sharedcfg { };
-			};
-                        port@2 {
-
-				regs = <
-	                                0x35 0x8a
-        	                        0x1f 0x8a
-                	                0x8a 0x8
-                        	        0x8a 0x8
-	                        >;
-
-				sharedcfg { };
-			};
-                        port@3 {
-
-				regs = <
-	                                0x35 0x14
-        	                        0x1f 0x14
-                	                0x8a 0x14
-                        	        0x8a 0x14
-	                        >;
-
-				sharedcfg { };
-			};
-		};
-		nae@mode1 {
-			#index-cells = <1>;
-			#value-cells = <1>;
-
-                        num-ports = <2>;
-                        start-port-id = <0>;
-                        num-port-regs = <2>;
-			port@0 {
-
-	                        regs = <
-        	                        0x35 0xd
-                	                0x1f 0xd
-                        	        0x8a 0xd
-                                	0x8a 0xd
-	                        >;
-
-				sharedcfg { };
-
-			};
-                        port@1 {
-
-				regs = <
-	                                0x35 0x35
-        	                        0x1f 0x35
-                	                0x8a 0x35
-                        	        0x8a 0x1f
-	                        >;
-
-				sharedcfg { };
-			};
-		};
-
-	};
-
-	board {
-	};
-
-};
diff --git a/dts/tiny.dts b/dts/tiny.dts
deleted file mode 100644
index 390541b..0000000
--- a/dts/tiny.dts
+++ /dev/null
@@ -1,153 +0,0 @@
-/* XLP8XX Device Tree Source 
- *
- */
-
-/dts-v1/;
-/ {
-	model = "MIPS,XLP8XX";
-	compatible = "NETL,XLP8XX_A0";
-	#address-cells = <1>;
-	#size-cells = <1>;
-	doms {
-		#address-cells = <1>;
-		#size-cells = <1>;
-		dom@0 {
-			device_type = "domain";
-			heap-size = <0x0 0x4000000>;
-			shared-mem = <0x0 0x80000000 0x0 0x20000000>;
-			#address-cells = <1>;
-			#size-cells = <1>;
-			cpu {
-				onlinemask = <0xf>;
-				nae-rx-vc = <0>;
-				nae-fb-vc = <3>;
-			};
-			uart {
-				id = <0>;
-			};
-			memory {
-				reg = <0x00100000 0x0FF00000	// 255M at 1M
-					   0x20000000 0x20000000>;  // 512M at 512M
-			};
-			pic {
-			};
-		};
-	};
-
-	chosen {
-		bootargs = "xlp_noi2c console=ttyS0,115200 rdinit=/sbin/init";
-		nae-cfg {
-			num-ports = <4>;
-			start-port-id = <0>;
-			num-port-regs = <5>;
-			ucore {
-				src@1 {
-					num-opcodes = <52>;
-					mask = <0xffff>;
-					opcodes = <
-						0x3c1c0010 
-						0x279cf808 
-						0x3c020010 
-						0x2442f808 
-						0x3c030010 
-						0x2463f808 
-						0xac400000 
-						0x0043082b 
-						0x1420fffd 
-						0x24420004 
-						0x3c1d0010 
-						0x27bdf9c0 
-						0x27a50020 
-						0x27a60028 
-						0xaca00000 
-						0xacc00000 
-						0x0c000029 
-						0x00002021 
-						0x1000ffff 
-						0x00000000 
-						0x3c020000 
-						0x8c420104 
-						0x27bdffe0 
-						0x2403ffff 
-						0xafbf001c 
-						0xafb10018 
-						0x10430009 
-						0xafb00014 
-						0x3c100000 
-						0x26100104 
-						0x2411ffff 
-						0x0040f809 
-						0x2610fffc 
-						0x8e020000 
-						0x1451fffc 
-						0x00000000 
-						0x8fbf001c 
-						0x8fb10018 
-						0x8fb00014 
-						0x03e00008 
-						0x27bd0020 
-						0x00001821 
-						0x34088004 
-						0x8d060000 
-						0x30c5000f 
-						0x00052280 
-						0x34820001 
-						0x34088030 
-						0xad020000 
-						0x34088000 
-						0x0800002a 
-						0xad030000 
-					>;
-				};
-			};
-
-			port0 {
-				num-free-desc = <10>;
-				cfg-regs = <
-					0x35 0xd
-					0x1f 0x35
-					0x12 0x28
-					0x21 0x14
-					0x28 0x400
-				>;
-			};
-			port1 {
-				num-free-desc = <10>;
-				cfg-regs = <
-					0x35 0xd
-					0x1f 0x35
-					0x12 0x8
-					0x21 0x14
-					0x28 0x401
-				>;
-			};
-			port2 {
-				num-free-desc = <10>;
-				cfg-regs = <
-					0x35 0xd
-					0x1f 0x1f
-					0x12 0x8
-					0x21 0x14
-					0x28 0x402
-				>;
-			};
-			port3 {
-				num-free-desc = <10>;
-				cfg-regs = <
-					0x35 0xd
-					0x1f 0x12
-					0x12 0x8
-					0x21 0x14
-					0x28 0x403
-				>;
-			};
-		};
-		fmn{
-			spill-base-hi   = <0x00000000>;
-			spill-base-low  = <0xE0000000>;
-			spill-size  	= <0x1000000>;
-			tx-vc       	= <524>;
-			credit      	= <50>;
-		};
-	};
-};
diff --git a/dts/xlp.dts b/dts/xlp.dts
deleted file mode 100644
index 9606782..0000000
--- a/dts/xlp.dts
+++ /dev/null
@@ -1,40 +0,0 @@
-/* XLP8XX Device Tree Source 
- *
- */
-
-/dts-v1/;
-/ {
-	model = "MIPS,XLP8XX";
-	compatible = "NETL,XLP8XX_A0";
-	#address-cells = <1>;
-	#size-cells = <1>;
-	doms {
-		#address-cells = <1>;
-		#size-cells = <1>;
-		dom@0 {
-			device_type = "domain";
-			heap-size = <0x0 0x4000000>;
-			shared-mem = <0x0 0x80000000 0x0 0x20000000>;
-			#address-cells = <1>;
-			#size-cells = <1>;
-			cpu {
-				onlinemask = <0xf>;
-				nae-rx-vc = <0>;
-				nae-fb-vc = <1>;
-			};
-			uart {
-				id = <0>;
-			};
-			memory {
-				reg = <0x00100000 0x0FF00000	// 255M at 1M
-					   0x20000000 0x20000000>;  // 512M at 512M
-			};
-			pic {
-			};
-		};
-	};
-
-	chosen {
-		bootargs = "xlp_noi2c console=ttyS0,115200 rdinit=/sbin/init";
-	};
-};
diff --git a/dts/xlr.dts b/dts/xlr.dts
deleted file mode 100644
index 34a227a..0000000
--- a/dts/xlr.dts
+++ /dev/null
@@ -1,320 +0,0 @@
-/* XLR,XLS Device Tree Source
- *
- */
-
-/dts-v1/;
-/ {
-	model = "MIPS,XLR(S)";
-	compatible = "NETL,XLR";
-	#address-cells = <1>;
-	#size-cells = <1>;
-	doms {
-		#address-cells = <1>;
-		#size-cells = <1>;
-		dom@0 {
-			device_type = "domain";
-			heap-size = <0x0 0x4000000>;
-			shared-mem = <0x0 0x80000000 0x0 0x20000000>;
-			#address-cells = <1>;
-			#size-cells = <1>;
-			cpu {
-				onlinemask = <0xffffffff>;
-			};
-			uart {
-				id = <0>;
-			};
-			memory {
-				reg = <0x00100000 0x0FF00000	// 255M at 1M
-					   0x20000000 0x20000000>;  // 512M at 512M
-			};
-			pic {
-			};
-		};
-		dom@1 {
-			device_type = "domain";
-			cpu {
-				onlinemask = <0xffff0000>;
-			};
-			uart {
-				id = <1>;
-			};
-			memory {
-			};
-		};
-		dom@2 {
-			device_type = "domain";
-			cpu {
-			};
-			uart {
-			};
-			memory {
-			};
-		};
-		dom@3 {
-			device_type = "domain";
-			cpu {
-			};
-			uart {
-			};
-			memory {
-			};
-		};
-		dom@4 {
-			device_type = "domain";
-			cpu {
-			};
-			uart {
-			};
-			memory {
-			};
-		};
-		dom@5 {
-			device_type = "domain";
-			cpu {
-			};
-			uart {
-			};
-			memory {
-			};
-		};
-		dom@6 {
-			device_type = "domain";
-			cpu {
-			};
-			uart {
-			};
-			memory {
-			};
-		};
-		dom@7 {
-			device_type = "domain";
-			cpu {
-			};
-			uart {
-			};
-			memory {
-			};
-		};
-		dom@8 {
-			device_type = "domain";
-			cpu {
-			};
-			uart {
-			};
-			memory {
-			};
-		};
-		dom@9 {
-			device_type = "domain";
-			cpu {
-			};
-			uart {
-			};
-			memory {
-			};
-		};
-		dom@10 {
-			device_type = "domain";
-			cpu {
-			};
-			uart {
-			};
-			memory {
-			};
-		};
-		dom@11 {
-			device_type = "domain";
-			cpu {
-			};
-			uart {
-			};
-			memory {
-			};
-		};
-		dom@12 {
-			device_type = "domain";
-			cpu {
-			};
-			uart {
-			};
-			memory {
-			};
-		};
-		dom@13 {
-			device_type = "domain";
-			cpu {
-			};
-			uart {
-			};
-			memory {
-			};
-		};
-		dom@14 {
-			device_type = "domain";
-			cpu {
-			};
-			uart {
-			};
-			memory {
-			};
-		};
-		dom@15 {
-			device_type = "domain";
-			cpu {
-			};
-			uart {
-			};
-			memory {
-			};
-		};
-		dom@16 {
-			device_type = "domain";
-			cpu {
-			};
-			uart {
-			};
-			memory {
-			};
-		};
-		dom@17 {
-			device_type = "domain";
-			cpu {
-			};
-			uart {
-			};
-			memory {
-			};
-		};
-		dom@18 {
-			device_type = "domain";
-			cpu {
-			};
-			uart {
-			};
-			memory {
-			};
-		};
-		dom@19 {
-			device_type = "domain";
-			cpu {
-			};
-			uart {
-			};
-			memory {
-			};
-		};
-		dom@20 {
-			device_type = "domain";
-			cpu {
-			};
-			uart {
-			};
-			memory {
-			};
-		};
-		dom@21 {
-			device_type = "domain";
-			cpu {
-			};
-			uart {
-			};
-			memory {
-			};
-		};
-		dom@23 {
-			device_type = "domain";
-			cpu {
-			};
-			uart {
-			};
-			memory {
-			};
-		};
-		dom@24 {
-			device_type = "domain";
-			cpu {
-			};
-			uart {
-			};
-			memory {
-			};
-		};
-		dom@25 {
-			device_type = "domain";
-			cpu {
-			};
-			uart {
-			};
-			memory {
-			};
-		};
-		dom@26 {
-			device_type = "domain";
-			cpu {
-			};
-			uart {
-			};
-			memory {
-			};
-		};
-		dom@27 {
-			device_type = "domain";
-			cpu {
-			};
-			uart {
-			};
-			memory {
-			};
-		};
-		dom@28 {
-			device_type = "domain";
-			cpu {
-			};
-			uart {
-			};
-			memory {
-			};
-		};
-		dom@29 {
-			device_type = "domain";
-			cpu {
-			};
-			uart {
-			};
-			memory {
-			};
-		};
-		dom@30 {
-			device_type = "domain";
-			cpu {
-			};
-			uart {
-			};
-			memory {
-			};
-		};
-		dom@31 {
-			device_type = "domain";
-			cpu {
-			};
-			uart {
-			};
-			memory {
-			};
-		};
-	};
-
-	chosen {
-		bootargs = "console=ttyS0,115200 rdinit=/sbin/init";
-		bootenv = "board=arizona"
-	};
-
-	system {
-		iobase = <0xbef00000>;
-		boardmajor = <2>;
-		boardminor = <1>;
-		macaddress = [ 00 0f 30 00 09 6a ];
-		cpufrequency = <1190000000>;
-		globalshmem = <0x800ff000>;
-	};
-};
-- 
1.9.1

