From bc20fad33aff9e9ea56ed356bc295487915185e2 Mon Sep 17 00:00:00 2001
From: Hareesh R <hareesh@netlogicmicro.com>
Date: Thu, 8 Dec 2011 15:50:19 +0530
Subject: [PATCH 0661/1532] Disabled debug prints

[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
---
 cryptolib/cryptoapi.c | 7 ++++---
 1 file changed, 4 insertions(+), 3 deletions(-)

diff --git a/cryptolib/cryptoapi.c b/cryptolib/cryptoapi.c
index ec5deb6..502ca55 100644
--- a/cryptolib/cryptoapi.c
+++ b/cryptolib/cryptoapi.c
@@ -50,7 +50,7 @@
 #include <nlm_hal_fmn.h>
 #include "cryptodev.h"
 #define CRYPTO_API_INCLUDE
-#define NLM_CRYPTO_DEBUG_EN
+//#define NLM_CRYPTO_DEBUG_EN
 #include "nlmcrypto.h"
 
 /**
@@ -637,9 +637,10 @@ int nlm_crypto_do_op(nlm_crypto_ctx_t *ctxt, enum nlm_crypto_op_type_t optype,
 		etype = 0;
 	}
 
-	nlm_dbg_print("Do op : cpu %d msg0 = %016llx msg1 %016llx rx_vc %d\n", cpu,
+	nlm_dbg_print("Do op : cpu %d msg0 = %016llx msg1 %016llx rx_vc %d dst %d\n", cpu,
 				(unsigned long long)msg0, (unsigned long long)msg1,
-				ctx->rx_vc);
+				ctx->rx_vc, dstvc_base);
+
 	if(ctx->mode == NLM_CRYPTO_MODE_SYNC_EXLVC) {
 		nlm_dbg_print("Mode is exclusive\n");
 
-- 
1.9.1

