From c5d5dcce438812e84991cd50218f112a35a78dee Mon Sep 17 00:00:00 2001
From: "P. Sadik" <psadik@broadcom.com>
Date: Wed, 8 May 2013 10:42:00 +0530
Subject: [PATCH 0686/1532] fdtlib: exclude temporary '*.mod.c' files from
 userspace source list

The *.mod.c files are created during Linux Kernel module compilation.
They should not be used when compiled for userspace.
[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
---
 libfdt/Makefile | 11 +++++------
 1 file changed, 5 insertions(+), 6 deletions(-)

diff --git a/libfdt/Makefile b/libfdt/Makefile
index 6a2220f..c514dc5 100644
--- a/libfdt/Makefile
+++ b/libfdt/Makefile
@@ -3,11 +3,10 @@ FDTLIB_DIR = $(SDK_ROOT)/libraries/fdt/libfdt
 
 include $(SDK_ROOT)/linux-userspace/Makefile.mk
 
-SRC = $(addprefix $(FDTLIB_DIR)/,$(shell cd $(FDTLIB_DIR) && ls *.c))
-SRC += $(addprefix $(FDTLIB_DIR)/contrib/,$(shell cd $(FDTLIB_DIR)/contrib && ls *.c))
+SRC = $(addprefix $(FDTLIB_DIR)/,$(shell cd $(FDTLIB_DIR) && find * -name '*.c' ! -name '*.mod.c' -print))
 
 FDTLIB_OBJS = $(patsubst %.c,%.o,$(SRC))
-	 
+
 FDT_LIB = libfdt.a
 
 include $(SDK_ROOT)/linux-userspace/Makefile.mk
@@ -19,10 +18,10 @@ HAL_INSTALL_PATH=$(SDK_ROOT)/libraries/install
 endif
 
 $(FDT_LIB): $(FDTLIB_OBJS)
-	echo "Build fdtlib library"
+	@echo "Build fdtlib library"
 	$(AR) -rcsP $(FDT_LIB) $(FDTLIB_OBJS)
 
-all: $(FDT_LIB) 
+all: $(FDT_LIB)
 
 clean:
 	$(RM) *.a $(FDTLIB_OBJS)
@@ -32,4 +31,4 @@ install:
 	cp -r $(FDTLIB_DIR)/libfdt.h  $(HAL_INSTALL_PATH)/include/
 	cp -r $(FDT_LIB) $(HAL_INSTALL_PATH)/lib/
 
-.PHONY: all clean install $(FDT_LIB) 
+.PHONY: all clean install $(FDT_LIB)
-- 
1.9.1

