From f870dbad8a5112c0ee2b05d2a4ae0dfb49b5bcc4 Mon Sep 17 00:00:00 2001
From: Jiang Lu <lu.jiang@windriver.com>
Date: Thu, 7 Aug 2014 12:36:35 +0800
Subject: [PATCH 10/14] netlogic:enable SELECT_MEMORY_MODEL for xlp

Signed-off-by: Jiang Lu <lu.jiang@windriver.com>
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
---
 arch/mips/netlogic/Kconfig |    4 ++++
 1 files changed, 4 insertions(+), 0 deletions(-)

diff --git a/arch/mips/netlogic/Kconfig b/arch/mips/netlogic/Kconfig
index 34288ad..5880e79 100644
--- a/arch/mips/netlogic/Kconfig
+++ b/arch/mips/netlogic/Kconfig
@@ -1,9 +1,13 @@
 if NLM_XLP_BOARD || NLM_XLR_BOARD
 
+config ARCH_SELECT_MEMORY_MODEL
+	bool
+
 if NLM_XLP_BOARD
 config DT_XLP_EVP
 	bool "Built-in device tree for XLP EVP boards"
 	default y
+	select ARCH_SELECT_MEMORY_MODEL
 	help
 	  Add an FDT blob for XLP EVP boards into the kernel.
 	  This DTB will be used if the firmware does not pass in a DTB
-- 
1.7.5.4

