From aa1469373b2705b9e9806154b4d6a7b51023cd07 Mon Sep 17 00:00:00 2001
From: wurui <wurui@actions-semi.com>
Date: Thu, 8 Oct 2015 16:06:41 +0800
Subject: [PATCH 48/62] wakeup: FIXBUG: fix BUG00397226, do not set vbus&wall
 as wakesource when there is no battery

commit 6e44254de16fe865e005d59a3afdbc689c47c283 from
https://github.com/xapp-le/kernel.git

Change-Id: I935f4378d6e8d55e6f82d310d02fe643484f9c8a

Conflicts:
	arch/arm/mach-owl/pm-owl.c
---
 arch/arm/mach-owl/pm-owl.c |   31 ++++++++++++++++++++++++++++---
 1 files changed, 28 insertions(+), 3 deletions(-)

diff --git a/arch/arm/mach-owl/pm-owl.c b/arch/arm/mach-owl/pm-owl.c
index 64eca13..a13e57e 100755
--- a/arch/arm/mach-owl/pm-owl.c
+++ b/arch/arm/mach-owl/pm-owl.c
@@ -133,6 +133,26 @@ void system_powerdown_withgpio(void)
 }
 EXPORT_SYMBOL_GPL(system_powerdown_withgpio);
 
+bool owl_pm_is_battery_connected(void)
+{
+	int ret;
+	struct power_supply *psy;
+	union power_supply_propval val;
+
+	/* Check if battery is known */
+	psy = power_supply_get_by_name("battery");
+	if (psy) {
+		ret = power_supply_get_property(psy, POWER_SUPPLY_PROP_PRESENT, &val);
+		if (!ret && val.intval) {
+			pr_info("find battery connect\n");
+			return true;
+		}
+	}
+	
+	pr_info("can not find battery connect\n");
+	return false;
+}
+
 void owl_pm_halt(void)
 {
 	int deep_pwrdn, wakeup_src;
@@ -144,9 +164,10 @@ void owl_pm_halt(void)
 	/* default sleep mode and wakeup source */
 	/* DO NOT add HARD_SWITCH source, we are here just because no batt switch. */
 	wakeup_src = OWL_PMIC_WAKEUP_SRC_RESET |
-		OWL_PMIC_WAKEUP_SRC_ONOFF_LONG/* |
-		OWL_PMIC_WAKEUP_SRC_WALL_IN |
-		OWL_PMIC_WAKEUP_SRC_VBUS_IN*/;
+		OWL_PMIC_WAKEUP_SRC_ONOFF_LONG;
+	
+	if(owl_pm_is_battery_connected())
+		wakeup_src |= OWL_PMIC_WAKEUP_SRC_WALL_IN | OWL_PMIC_WAKEUP_SRC_VBUS_IN;
 
 	/* if wall/usb is connect, cannot enter S4, only can enter S3 */
 	deep_pwrdn = 0;
@@ -824,6 +845,10 @@ static void owl_set_wakeup_source(void)
 		}
 	}
 
+	if(!owl_pm_is_battery_connected())
+		wakeup_src &= ~(OWL_PMIC_WAKEUP_SRC_WALL_IN | OWL_PMIC_WAKEUP_SRC_WALL_OUT
+				| OWL_PMIC_WAKEUP_SRC_VBUS_IN | OWL_PMIC_WAKEUP_SRC_VBUS_OUT);
+
 	/* add onoff long press wakeup source enabled,
 	 *    when the adapter is plugged in and the battery is full.
 	 */
-- 
1.7.5.4

