From 754b536d41432c407e79bfe2295a0ec79b3980ce Mon Sep 17 00:00:00 2001
From: Paul Gortmaker <paul.gortmaker@windriver.com>
Date: Wed, 24 Feb 2010 22:10:38 -0500
Subject: [PATCH] MILS shim: Remove CONFIG_WRHV_MILS from MILS specific file

This file is already MILS specifc, so it doesn't need a
CONFIG_WRHV_MILS trigger within.

Signed-off-by: Paul Gortmaker <paul.gortmaker@windriver.com>
---
 include/vbi/vdk_syscalls.h |    4 +---
 1 files changed, 1 insertions(+), 3 deletions(-)

diff --git a/include/vbi/vdk_syscalls.h b/include/vbi/vdk_syscalls.h
index 731e7a7..ff02a79 100644
--- a/include/vbi/vdk_syscalls.h
+++ b/include/vbi/vdk_syscalls.h
@@ -14,9 +14,7 @@
  *
  */
 
-
-
-#if !defined(_VDK_SYSCALLS_H) && defined(CONFIG_WRHV_MILS)
+#ifndef _VDK_SYSCALLS_H
 #define _VDK_SYSCALLS_H
 
 /*
-- 
1.7.0

