From d019853722e92893c58d936b519d199e25156008 Mon Sep 17 00:00:00 2001
From: Paul Gortmaker <paul.gortmaker@windriver.com>
Date: Fri, 26 Feb 2010 16:32:20 -0500
Subject: [PATCH] MILS vdk: UINT32 --> unsigned long

This is a cherry pick of a chunk of change that Keith is
currently using on the MILS team.  It will be folded into
the shim commit on the next rebase and disappear.

Signed-off-by: Paul Gortmaker <paul.gortmaker@windriver.com>
---
 include/vbi/vdk_interface.h |   30 +++++++++++++++---------------
 1 files changed, 15 insertions(+), 15 deletions(-)

diff --git a/include/vbi/vdk_interface.h b/include/vbi/vdk_interface.h
index bdb5c8f..714b57b 100644
--- a/include/vbi/vdk_interface.h
+++ b/include/vbi/vdk_interface.h
@@ -203,25 +203,25 @@ typedef enum                            /* queuing port protocol     */
 
 typedef struct portConfig
 {
-    UINT32              id;                    /* unique port ID                   */
-    unsigned char     * buf;                   /* port payload buffer vaddr        */
-    UINT32              bufAttr;               /* port payload buffer attribute    */
-    char                name[MAX_NAME_LENGTH]; /* port name                        */
-    UINT32              vbId;                  /* parent vb                        */
-    UINT32              moduleId;              /* parent module                    */
-    UINT32              inChannel;             /* in channel                       */
-    UINT32              outChannel;            /* out channel                      */
-    PORT_DIRECTION_TYPE direction;             /* SOURCE or DESTINATION            */
-    PORT_MODE_TYPE      mode;                  /* QUEUING or SAMPLING              */
-    PORT_PROTOCOL_TYPE  protocol;              /* RECEIVER_DISCARD or SENDER_BLOCK */
-    UINT32              msgSize;               /* max. message size (PAGE_ALIGNED) */
-    UINT32              numMsgs;               /* max. capacity for queuing ports  */
+    unsigned long id;			/* unique port ID                   */
+    unsigned char *buf;			/* port payload buffer vaddr        */
+    unsigned long bufAttr;		/* port payload buffer attribute    */
+    char name[MAX_NAME_LENGTH];		/* port name                        */
+    unsigned long vbId;			/* parent vb                        */
+    unsigned long moduleId;		/* parent module                    */
+    unsigned long inChannel;		/* in channel                       */
+    unsigned long outChannel;		/* out channel                      */
+    PORT_DIRECTION_TYPE direction;	/* SOURCE or DESTINATION            */
+    PORT_MODE_TYPE mode;		/* QUEUING or SAMPLING              */
+    PORT_PROTOCOL_TYPE protocol;	/* RECEIVER_DISCARD or SENDER_BLOCK */
+    unsigned long msgSize;		/* max. message size (PAGE_ALIGNED) */
+    unsigned long numMsgs;		/* max. capacity for queuing ports  */
 } PORT_CONFIG;
 
 typedef struct partPortConfig
 {
-    UINT32              numPorts;
-    PORT_CONFIG         portInfo[];
+    unsigned long numPorts;
+    PORT_CONFIG portInfo[];
 } PART_PORT_CONFIG;
 
 typedef struct portConfig VB_PORT_INFO;
-- 
1.7.0

