From 07fe50f60dc83690bb13a6fb028e2a174454a6d8 Mon Sep 17 00:00:00 2001
From: Paul Gortmaker <paul.gortmaker@windriver.com>
Date: Tue, 23 Mar 2010 15:07:49 -0400
Subject: [PATCH] fix demux of MILS syscalls

Signed-off-by: Paul Gortmaker <paul.gortmaker@windriver.com>
---
 include/vbi/syscall.h |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/include/vbi/syscall.h b/include/vbi/syscall.h
index a697305..f70ace2 100644
--- a/include/vbi/syscall.h
+++ b/include/vbi/syscall.h
@@ -14,10 +14,10 @@
  *
  */
 
-#ifndef _VBI_SYSCALL_H
 #ifdef CONFIG_WRHV_MILS
 #include <vbi/vdk_syscalls.h>
 #else
+#ifndef _VBI_SYSCALL_H
 #define _VBI_SYSCALL_H
 #include <linux/linkage.h>
 #include <vbi/types.h>
-- 
1.7.0

