From b80abe127b9e0ca860a620775fb59b72245abd92 Mon Sep 17 00:00:00 2001
From: Paul Gortmaker <paul.gortmaker@windriver.com>
Date: Tue, 23 Mar 2010 16:39:15 -0400
Subject: [PATCH] MILS: fix compile warning on missing VBI 2.0 structs

In file included from kernel/vbi/shmem.c:42:
include/vbi/private.h:40: warning: 'struct vbi_msg_ctl' declared inside parameter list
include/vbi/private.h:40: warning: its scope is only this definition or declaration, which is probably not what you want
include/vbi/private.h:40: warning: 'struct vbi_msg_info' declared inside parameter list

MILS doesn't use these, so insert an empty struct which will
cause a compile failure if anyone actually tries to use them.

Signed-off-by: Paul Gortmaker <paul.gortmaker@windriver.com>
---
 include/vbi/vdk_syscalls.h |   10 ++++++++++
 1 files changed, 10 insertions(+), 0 deletions(-)

diff --git a/include/vbi/vdk_syscalls.h b/include/vbi/vdk_syscalls.h
index d752b24..6a3deee 100644
--- a/include/vbi/vdk_syscalls.h
+++ b/include/vbi/vdk_syscalls.h
@@ -385,6 +385,16 @@ struct vbi_vtlb_control
 	struct vbi_vtlb_cr3_cache cr3_cache[VBI_VTLB_OP_CR3_CACHE_ENTRIES];
 };
 
+struct vbi_msg_ctl
+{
+	/* Unused by MILS */
+};
+
+struct vbi_msg_info
+{
+	/* Unused by MILS */
+};
+
 /*
  * Control structure used by vbi_vb_mgmt for commands memory read, memory write,
  *  register read, and register write.
-- 
1.7.0

