From 7e4af4fb8407a20302f3c549c5417e1e0061d785 Mon Sep 17 00:00:00 2001
From: Paul Gortmaker <paul.gortmaker@windriver.com>
Date: Fri, 26 Mar 2010 12:11:29 -0400
Subject: [PATCH] vdk_port: tag as GPL

As discussed with Keith, this has to be GPL because it is
required for vdk_interface.h which is used all through the
guest.  There is no IP here in basic declarations anyway.

This will be folded into the VDK shim, so that this commit
log won't ever even be seen externally.

Signed-off-by: Paul Gortmaker <paul.gortmaker@windriver.com>
---
 include/vbi/vdk_port.h |   16 +++++++++++-----
 1 files changed, 11 insertions(+), 5 deletions(-)

diff --git a/include/vbi/vdk_port.h b/include/vbi/vdk_port.h
index 9a28cc9..156d47c 100644
--- a/include/vbi/vdk_port.h
+++ b/include/vbi/vdk_port.h
@@ -1,11 +1,17 @@
-/* vdk_port.h - port facility header file */
-
 /*
+ * vdk_interface.h - virtual board interface header file
+ *
  * Copyright (c) 2008-2010 Wind River Systems, Inc.
  *
- * The right to copy, distribute, modify or otherwise make use
- * of this software may be licensed only pursuant to the terms
- * of an applicable Wind River license agreement.
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 2 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+ * See the GNU General Public License for more details.
+ *
  */
 
 #ifndef _VDK_PORT_H
-- 
1.7.0

