From 947b0267a2f3e3a855aa9d32329183f0278a58c7 Mon Sep 17 00:00:00 2001
From: Paul Gortmaker <paul.gortmaker@windriver.com>
Date: Fri, 26 Mar 2010 12:13:32 -0400
Subject: [PATCH] more compat definitions for MILS

Signed-off-by: Paul Gortmaker <paul.gortmaker@windriver.com>
---
 include/vbi/compat.h |    7 ++++++-
 1 files changed, 6 insertions(+), 1 deletions(-)

diff --git a/include/vbi/compat.h b/include/vbi/compat.h
index af32025..95e6c6d 100644
--- a/include/vbi/compat.h
+++ b/include/vbi/compat.h
@@ -233,6 +233,11 @@
 #define vdkPortBufRelease	vdk_port_buf_release
 #define vdkScheduleSet		vdk_schedule_set
 #define vdkScheduleNameToId	vdk_schedule_name_to_id
-#endif
+
+#define PORT_BUF_CTRL_DESC	struct portBufCtrlDesc
+#define PORT_STATUS		struct portStatus
+#define PORT_HY_INFO		struct portHyInfo
+
+#endif	/* MILS */
 
 #endif  /* _VBI_COMPAT_H */
-- 
1.7.0

