From 7e08225f52f548caf01b8469317fb1d5f6ac74af Mon Sep 17 00:00:00 2001
From: Paul Gortmaker <paul.gortmaker@windriver.com>
Date: Fri, 26 Mar 2010 12:14:07 -0400
Subject: [PATCH] MILS: fix port status and port config in interface struct.

One was referencing an undefined struct, and the other
was hard to follow due to the overzealous use of
typedefs on structs.

This only impacts anything trying to use the port fields.

Signed-off-by: Paul Gortmaker <paul.gortmaker@windriver.com>
---
 include/vbi/vdk_interface.h |   10 ++++++++--
 1 files changed, 8 insertions(+), 2 deletions(-)

diff --git a/include/vbi/vdk_interface.h b/include/vbi/vdk_interface.h
index f417a17..d5fbf28 100644
--- a/include/vbi/vdk_interface.h
+++ b/include/vbi/vdk_interface.h
@@ -291,7 +291,7 @@ struct vb_status
     /* Location in vbPhyscial memory of the virtual interrupt controller */
 
     struct vbInterruptController *vbInterruptController;  /* 10: */
-    struct vb_port_status *portStatus;             /* 11: port status */
+    struct portStatus *portStatus;             /* 11: port status */
 
     struct vb_log_status secAuditLogStatus; /* 12-16: Security Audit Event log status */
     struct vb_log_status safetyLogStatus;   /* 17-21: Safety Critical Event log status */
@@ -477,7 +477,13 @@ struct vb_config
     unsigned int  cacheMode;
 
     unsigned int      numPorts;  /* number of ports defined for the VB */
-    struct vbPortInfo *portInfo;  /* port configuration and status */
+    /*
+     * MILS vbInterface.h has the following:
+     *		typedef struct portConfig VB_PORT_INFO
+     *			... and later...
+     *		VB_PORT_INFO * portInfo; 
+     */
+    struct portConfig *portInfo;  /* port configuration and status */
 
     struct vb_log_cfg secAuditLogCfg;
     struct vb_log_cfg safetyLogCfg;
-- 
1.7.0

