From a93830456f14c957f0658387e46e50c1aac594f4 Mon Sep 17 00:00:00 2001
From: Paul Gortmaker <paul.gortmaker@windriver.com>
Date: Fri, 26 Mar 2010 16:15:16 -0400
Subject: [PATCH] MILS: namelen sync, fix proto definitions.

Two fixes for drivers being compiled out of tree.  One is
a mixup of fcn prototypes, the other a wrong NAMELEN value.
The NAMELEN one still exists as a redefinition in the standalone
driver, but we can't remove it from the kernel, so it will have
to be removed from there instead.

Signed-off-by: Paul Gortmaker <paul.gortmaker@windriver.com>
---
 include/vbi/vdk_port.h     |    2 +-
 include/vbi/vdk_syscalls.h |    4 ++--
 2 files changed, 3 insertions(+), 3 deletions(-)

diff --git a/include/vbi/vdk_port.h b/include/vbi/vdk_port.h
index 156d47c..2621ff7 100644
--- a/include/vbi/vdk_port.h
+++ b/include/vbi/vdk_port.h
@@ -53,7 +53,7 @@ typedef enum                            /* queuing port protocol     */
 #define PORT_Q_AVAILABLE			4
 #define PORT_Q_DEST_UNAVAILABLE			9
 
-#define MAX_NAME_LENGTH 80
+#define MAX_NAME_LENGTH 30
 
 struct portConfig
 {
diff --git a/include/vbi/vdk_syscalls.h b/include/vbi/vdk_syscalls.h
index 6a3deee..d4a2a06 100644
--- a/include/vbi/vdk_syscalls.h
+++ b/include/vbi/vdk_syscalls.h
@@ -437,8 +437,8 @@ extern int vbi_disable_vmmu (unsigned vmmu_num);
 
 extern asmlinkage int vbi_virt_to_phys (void * pVirtAddr, void *pPaddr);
 extern asmlinkage int vbi_int_controller_done(unsigned level);
-extern asmlinkage void vbi_vcore_irq_lock(void);
-extern asmlinkage int32_t vbi_vcore_irq_unlock(void);
+extern asmlinkage int32_t vbi_vcore_irq_lock(void);
+extern asmlinkage void vbi_vcore_irq_unlock(void);
 
 static inline int32_t vbi_hy_ioctl(unsigned ioctl, void *arg1, void *arg2, \
 		void *arg3, void *arg4)
-- 
1.7.0

