From 8e7da2e613bf3f71f8e42f03d9d5261a6cad3cbe Mon Sep 17 00:00:00 2001
From: Paul Gortmaker <paul.gortmaker@windriver.com>
Date: Wed, 25 Nov 2009 18:41:02 -0500
Subject: [PATCH 1/7] compat: no need to include any other files.

There is no need for the compatibility defines to require the
main vbi.h defines.  If people are using that file, they will
be including it themselves.

Signed-off-by: Paul Gortmaker <paul.gortmaker@windriver.com>
---
 include/vbi/compat.h |    2 --
 1 files changed, 0 insertions(+), 2 deletions(-)

diff --git a/include/vbi/compat.h b/include/vbi/compat.h
index af8e721..a069ad1 100644
--- a/include/vbi/compat.h
+++ b/include/vbi/compat.h
@@ -26,8 +26,6 @@
 #ifndef _VBI_COMPAT_H
 #define _VBI_COMPAT_H
 
-#include <vbi/vbi.h>
-
 /*
  * These were particularly confusing, since there are similarly named
  * macros used to map the bits within these structs.  Eeech.
-- 
1.6.5.2

