From 9b940b79fb8875c6475e72c909de3cdf1b5ca8da Mon Sep 17 00:00:00 2001
From: Paul Gortmaker <paul.gortmaker@windriver.com>
Date: Wed, 2 Dec 2009 14:21:27 -0500
Subject: [PATCH 17/43] core_set: delete macros for basic integer ops

What was the vbiCoreSet_t is actually just an uint32_t,
and so having macros for testing and setting to zero, etc,
is overkill.

Fortunately there are no users, so no real impact here.

Signed-off-by: Paul Gortmaker <paul.gortmaker@windriver.com>
---
 include/vbi/core_set.h |   39 ---------------------------------------
 1 files changed, 0 insertions(+), 39 deletions(-)
 delete mode 100644 include/vbi/core_set.h

diff --git a/include/vbi/core_set.h b/include/vbi/core_set.h
deleted file mode 100644
index 18d8a34..0000000
--- a/include/vbi/core_set.h
+++ /dev/null
@@ -1,39 +0,0 @@
-/*
- * core_set.h - macros to manipulate uint32_t type
- *
- * Copyright (c) 2009 Wind River Systems, Inc.
- *
- * This program is free software; you can redistribute it and/or modify
- * it under the terms of the GNU General Public License version 2 as
- * published by the Free Software Foundation.
- *
- * This program is distributed in the hope that it will be useful,
- * but WITHOUT ANY WARRANTY; without even the implied warranty of
- * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
- * See the GNU General Public License for more details.
- *
- */
-
-#ifndef _VBI_CORE_SET_H
-#define _VBI_CORE_SET_H
-
-#ifndef _ASMLANGUAGE
-
-/* set of macros to manipulate the coreset_t type */
-
-#define CORESET_COPY(coreSetDst, coreSetSrc)  (coreSetDst = coreSetSrc)
-#define CORESET_ISZERO(coreset)         (coreset == 0 ? TRUE : FALSE)
-#define CORESET_SET(coreset, n)         (coreset |= (1 << n))
-#define CORESET_ISSET(coreset, n)       (coreset & (1 << n))
-#define CORESET_CLR(coreset, n)         (coreset &= ~(1 << n))
-#define CORESET_ZERO(coreset)           (coreset = 0)
-
-#define CORESET_FIRST_SET(coreset)                ((int)coreset & (int)-coreset)
-#define CORESET_FIRST_INDEX(coreset)              (FFS_LSB (coreset) - 1)
-#define CORESET_ADD(coresetAugend,coresetAddend)   (coresetAugend |= coresetAddend)
-#define CORESET_SUB(coresetMinuend,coresetSubtrahend)      \
-        (coresetMinuend &= ~coresetSubtrahend)
-
-#endif
-
-#endif  /* _VBI_CORE_SET_H */
-- 
1.6.5.2

