From 9c070a6faf5cc8df2f2b230eb2ae295bb3d36001 Mon Sep 17 00:00:00 2001
From: Paul Gortmaker <paul.gortmaker@windriver.com>
Date: Sun, 28 Mar 2010 14:09:09 -0400
Subject: [PATCH] enable 8548 wrhv builds for !MILS

Signed-off-by: Paul Gortmaker <paul.gortmaker@windriver.com>
---
 arch/powerpc/platforms/85xx/Kconfig |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/arch/powerpc/platforms/85xx/Kconfig b/arch/powerpc/platforms/85xx/Kconfig
index bf0217b..75f2da8 100644
--- a/arch/powerpc/platforms/85xx/Kconfig
+++ b/arch/powerpc/platforms/85xx/Kconfig
@@ -124,7 +124,7 @@ config WRHV_8572
 
 config WRHV_SBC8548
 	bool "WindRiver SBC8548 (HV)"
-	depends on WRHV_MILS
+	depends on WRHV
 	help
 	  This option enables support for the Wind River SBC8548
 	  Hypervisor enabled board.
-- 
1.7.0

