From 50a6b9d5dc95d6bf3df68ea60333b5beced0263e Mon Sep 17 00:00:00 2001
From: Paul Gortmaker <paul.gortmaker@windriver.com>
Date: Wed, 2 Dec 2009 19:51:13 -0500
Subject: [PATCH 29/43] interface: XML config file name is not relevent

The guest OS has no interest in, nor any possible use for the
name of the XML that the hypervisor was configured with.  A
simple grep of the guest code shows it isn't used either.

Signed-off-by: Paul Gortmaker <paul.gortmaker@windriver.com>
---
 include/vbi/interface.h |    4 ----
 1 files changed, 0 insertions(+), 4 deletions(-)

diff --git a/include/vbi/interface.h b/include/vbi/interface.h
index fc360c0..07668ba 100644
--- a/include/vbi/interface.h
+++ b/include/vbi/interface.h
@@ -117,10 +117,6 @@ identity mapped.
 #define VB_DEBUG_MSG(fmt, args...)
 #endif
 
-
-/* Configuration Filename for the hypervisor this must not change */
-#define VB_WRHV_CONFIGURATION_FILE "wrhvConfig.xml"
-
 /* Hard limits for now */
 #define VB_MAX_VIRTUAL_BOARDS		1024
 #define VB_MAX_BUSES			1024
-- 
1.6.5.2

