From 4db33c171365fb71fd363951bedaef07129407bd Mon Sep 17 00:00:00 2001
From: Paul Gortmaker <paul.gortmaker@windriver.com>
Date: Thu, 3 Dec 2009 15:26:47 -0500
Subject: [PATCH 42/43] intPendingType: add a compat for it

The sed renaming for intPending fixed this one automatically,
but it didn't add a compat for it.

Signed-off-by: Paul Gortmaker <paul.gortmaker@windriver.com>
---
 include/vbi/compat.h |    3 ++-
 1 files changed, 2 insertions(+), 1 deletions(-)

diff --git a/include/vbi/compat.h b/include/vbi/compat.h
index ffb5dd2..8f9aebf 100644
--- a/include/vbi/compat.h
+++ b/include/vbi/compat.h
@@ -91,7 +91,6 @@
 #define sizeIn			size_in
 #define sizeOut			size_out
 
-#define intPending		irq_pend
 #define resetPC			reset_pc
 #define bootCount		boot_count
 #define boardName		board_name
@@ -108,6 +107,8 @@
 #define intName			irq_name
 #define intNumber		irq_num
 #define intDirection		irq_dir
+#define intPending		irq_pend
+#define intPendingType		irq_pend_type
 
 /*
  * function calls, etc.
-- 
1.6.5.2

