From 51ecd1590e937495e93e2d24e6ab7119dbf999f6 Mon Sep 17 00:00:00 2001
From: Paul Gortmaker <paul.gortmaker@windriver.com>
Date: Mon, 14 Dec 2009 18:12:03 -0500
Subject: [PATCH 3/8] io_apic.h: don't include unneeded header files.

There is nothing in this file that needs any other header
file, so don't include random junk.

Signed-off-by: Paul Gortmaker <paul.gortmaker@windriver.com>
---
 include/vbi/io_apic.h |    3 ---
 1 files changed, 0 insertions(+), 3 deletions(-)

diff --git a/include/vbi/io_apic.h b/include/vbi/io_apic.h
index 4fa63c1..ed8d454 100644
--- a/include/vbi/io_apic.h
+++ b/include/vbi/io_apic.h
@@ -17,9 +17,6 @@
 #ifndef	_VBI_IO_APIC_H
 #define	_VBI_IO_APIC_H
 
-#include <wrhv.h>
-#include <sys/types.h>
-
 #ifndef	_ASMLANGUAGE
 
 /* data structures */
-- 
1.6.5.2

