From 01d83890982ced0f6bae287d108e967192aec623 Mon Sep 17 00:00:00 2001
From: Paul Gortmaker <paul.gortmaker@windriver.com>
Date: Mon, 14 Dec 2009 18:13:04 -0500
Subject: [PATCH 4/8] io_apic.h: fix compile failure

This file has some extern decl that rely on "HY_CTX", but nothing in
the VBI reference defines the HY_CTX structure, so compilation will
fail, with an error message like:

error: expected ')' before '*' token

Further investigation reveals that all the extern lines are for
functions that are nowhere to be found, so it looks like they all can
be deleted.

Signed-off-by: Paul Gortmaker <paul.gortmaker@windriver.com>
---
 include/vbi/io_apic.h |    8 --------
 1 files changed, 0 insertions(+), 8 deletions(-)

diff --git a/include/vbi/io_apic.h b/include/vbi/io_apic.h
index ed8d454..f38cd72 100644
--- a/include/vbi/io_apic.h
+++ b/include/vbi/io_apic.h
@@ -98,14 +98,6 @@ typedef struct vioapic
 	VIOAPIC_ENTRY   entry [VIOAPIC_MAX_REDTABLE_ENTRIES];
 } VIOAPIC;
 
-/* external declarations */
-
-extern int vIoapicInterrupt (HY_CTX * ctx, uint32_t globalInt);
-extern void vIoapicInit (VB_INFO * info);
-extern uint32_t vIoapicGetVector (HY_CTX * ctx, uint32_t irq);
-extern uint32_t vIoapicGetMask (HY_CTX * ctx, uint32_t irq);
-extern void vIoapicSetMask (HY_CTX * ctx, uint32_t irq, uint32_t mask);
-
 #endif /* _ASMLANGUAGE */
 
 #endif	/* _VBI_IO_APIC_H */
-- 
1.6.5.2

