From 71fa32a7393ec7a1d730e0bfdcc088e3de1ba4be Mon Sep 17 00:00:00 2001
From: Paul Gortmaker <paul.gortmaker@windriver.com>
Date: Mon, 14 Dec 2009 16:18:20 -0500
Subject: [PATCH 1/8] io_apic.h: delete reference to nonexistent file

There is no such vbMgr file in the VBI reference implementation,
so delete the include of it.

Signed-off-by: Paul Gortmaker <paul.gortmaker@windriver.com>
---
 include/vbi/io_apic.h |    1 -
 1 files changed, 0 insertions(+), 1 deletions(-)

diff --git a/include/vbi/io_apic.h b/include/vbi/io_apic.h
index 3c80dc9..4fa63c1 100644
--- a/include/vbi/io_apic.h
+++ b/include/vbi/io_apic.h
@@ -19,7 +19,6 @@
 
 #include <wrhv.h>
 #include <sys/types.h>
-#include <vbMgr.h>
 
 #ifndef	_ASMLANGUAGE
 
-- 
1.6.5.2

