From caf061f6c3f8a8477f7e00c1045fc73738b72924 Mon Sep 17 00:00:00 2001
From: Paul Gortmaker <paul.gortmaker@windriver.com>
Date: Mon, 22 Mar 2010 14:16:42 -0400
Subject: [PATCH] MILS: map VDK syscalls onto linux-like names

This will be merged into the shim commit at the next rebase.

Signed-off-by: Paul Gortmaker <paul.gortmaker@windriver.com>
---
 include/vbi/compat.h |   21 +++++++++++++++++++++
 1 files changed, 21 insertions(+), 0 deletions(-)

diff --git a/include/vbi/compat.h b/include/vbi/compat.h
index fa1fe7d..af32025 100644
--- a/include/vbi/compat.h
+++ b/include/vbi/compat.h
@@ -214,4 +214,25 @@
 #define vbiVioapicVectorGet	vbi_get_vioapic_vec
 #define vbiVioapicVectorSet	vbi_set_vioapic_vec
 
+#ifdef CONFIG_WRHV_MILS
+/*
+ * MILS/VDK mappings
+ */
+#define vdkIntDisable		vbi_vcore_irq_lock
+#define vdkIntEnable		vbi_vcore_irq_unlock
+#define vdkIntControllerDone	vbi_int_controller_done
+/*
+ * The following are not strictly a part of VBI 2.0 nomenclature
+ */
+#define vdkSecAuditEventInject	vdk_sec_audit_event_inject
+#define vdkSecAuditEventCollect	vdk_sec_audit_event_collect
+#define vdkSafeCritEventInject	vdk_safe_crit_event_inject
+#define vdkSafeCritEventCollect	vdk_safe_crit_event_collect
+#define vdkPortSend		vdk_port_send
+#define vdkPortGetRecvBuf	vdk_port_get_recv_buf
+#define vdkPortBufRelease	vdk_port_buf_release
+#define vdkScheduleSet		vdk_schedule_set
+#define vdkScheduleNameToId	vdk_schedule_name_to_id
+#endif
+
 #endif  /* _VBI_COMPAT_H */
-- 
1.7.0

