From d20905e4c19caacfe6eeba29c6a96a9d3f42ebee Mon Sep 17 00:00:00 2001
From: Paul Gortmaker <paul.gortmaker@windriver.com>
Date: Thu, 3 Dec 2009 16:18:25 -0500
Subject: [PATCH 43/43] MILS: get rid of typedefs that clash with compat.h

The compat.h was intended for 3rd party additions to the kernel,
and not for exporters of VBI/VDK functionality.  Since it is
being used this way, the typedef compats bump into each other.

Deleting them from the VDK code is the fix; then the compat.h
ones can do their job.

Signed-off-by: Paul Gortmaker <paul.gortmaker@windriver.com>
---
 include/vbi/vdk_syscalls.h |   20 ++++++++++----------
 1 files changed, 10 insertions(+), 10 deletions(-)

diff --git a/include/vbi/vdk_syscalls.h b/include/vbi/vdk_syscalls.h
index 7948728..1c14717 100644
--- a/include/vbi/vdk_syscalls.h
+++ b/include/vbi/vdk_syscalls.h
@@ -318,7 +318,7 @@ have mapped a NULL / no-op to HY_SYSCALL(23) */
 
 /* statistics structure returned by VBI_HYIOCTL_GETSTATS ioctl */
 
-typedef struct
+struct vbi_ctx_stats
 {
 	unsigned long ctx_type;		/* type of context (user, supv, etc) */
 	unsigned long switchin;		/* number of times switched in       */
@@ -330,33 +330,33 @@ typedef struct
 	unsigned long tsCtxExcInH;	/* timestamp: exception entry (high) */
 	unsigned long tsCtxExcInL;	/* timestamp: exception entry (low)  */
 	unsigned long reset;		/* number of times context reset     */
-} VBI_CTX_STATS;
+};
 
 /* VTLB operation structures (x86 specific) */
-typedef struct vbi_vtlb_op
+struct vbi_vtlb_op
 {
 	uint32_t op;		/* VTLB operation id */
 	uint32_t arg1;		/* VTLB operation arg 1 */
 	uint32_t arg2;		/* VTLB operation arg 2 */
 	uint32_t arg3;		/* VTLB operation arg 3 */
-} VBI_VTLB_OP;
+};
 
-typedef struct vbi_vtlb_cr3_cache
+struct vbi_vtlb_cr3_cache
 {
 	uint32_t guest_cr3;	/* Guest CR3 register */
 	uint32_t host_cr3;	/* Host CR3 register */
-} VBI_VTLB_CR3_CACHE;
+};
 
-typedef struct vbi_vtlb_control
+struct vbi_vtlb_control
 {
 	uint32_t size;		/* VTLB Control structure size */
 	uint32_t mode;		/* VTLB module */
 	uint32_t vtlb_ops_ix;	/* VTLB operation index */
-	VBI_VTLB_OP vtlb_ops[VBI_VTLB_OP_MAX_OPS];	/* VTLB ops array */
+	struct vbi_vtlb_op vtlb_ops[VBI_VTLB_OP_MAX_OPS];	/* VTLB ops array */
 	uint32_t cr3_cache_ix;	/* CR3 cache index */
 				/* cr3 cache*/
-	VBI_VTLB_CR3_CACHE cr3_cache[VBI_VTLB_OP_CR3_CACHE_ENTRIES];
-} VBI_VTLB_CONTROL;
+	struct vbi_vtlb_cr3_cache cr3_cache[VBI_VTLB_OP_CR3_CACHE_ENTRIES];
+};
 
 /*
  * Control structure used by vbi_vb_mgmt for commands memory read, memory write,
-- 
1.6.5.2

