From 926dc349af2c476764bd9a2116dd2f076920088e Mon Sep 17 00:00:00 2001
From: Paul Gortmaker <paul.gortmaker@windriver.com>
Date: Thu, 26 Nov 2009 14:26:07 -0500
Subject: [PATCH 2/7] remove unused debugging code

Signed-off-by: Paul Gortmaker <paul.gortmaker@windriver.com>
---
 arch/x86/kernel/head_32.S |    5 -----
 1 files changed, 0 insertions(+), 5 deletions(-)

diff --git a/arch/x86/kernel/head_32.S b/arch/x86/kernel/head_32.S
index be13892..e66c361 100644
--- a/arch/x86/kernel/head_32.S
+++ b/arch/x86/kernel/head_32.S
@@ -84,11 +84,6 @@ INIT_MAP_BEYOND_END = BOOTBITMAP_SIZE + (PAGE_TABLE_SIZE + ALLOCATOR_SLOP)*PAGE_
 .section .text.head,"ax",@progbits
 ENTRY(startup_32)
 
-/*
-	mov $0x65, %al
-	mov $0x2f8, %dx
-	outb %al, %dx
-*/
 #ifndef CONFIG_WRHV
 	/* test KEEP_SEGMENTS flag to see if the bootloader is asking
 		us to not reload segments */
-- 
1.6.5.2

