From c371f471ba34beaffa6e7dffc64e5b1fd7e25d21 Mon Sep 17 00:00:00 2001
From: Paul Gortmaker <paul.gortmaker@windriver.com>
Date: Thu, 26 Nov 2009 20:11:16 -0500
Subject: [PATCH 5/7] v2009.11.20: fix prototypes of various show routines

Integrate the pre Black Monday release

[cf: 02b,18nov09,mmi  fix XXXRegsDisplay() prototype
     02a,13nov09,mmi  fix Show routines prototypes ]

Signed-off-by: Paul Gortmaker <paul.gortmaker@windriver.com>
---
 include/vbi/vbi.h |   10 +++++-----
 1 files changed, 5 insertions(+), 5 deletions(-)

diff --git a/include/vbi/vbi.h b/include/vbi/vbi.h
index ea2fbd7..3c6047c 100644
--- a/include/vbi/vbi.h
+++ b/include/vbi/vbi.h
@@ -539,11 +539,11 @@ extern vbiStatus_t vbi_guest_phys_to_phys(vbiGuestPhysAddr_t gaddr,
 				vbiPhysAddr_t  *paddr);
 
 extern void vbi_disp_vioapic(void);
-extern void vbi_show_shmem(struct vb_config *config);
-extern void vbi_show_stat(struct vb_status *p);
-extern void vbi_show_ctrl(struct vb_control *p);
-extern void vbi_show_cfg(struct vb_config *p);
-extern void vbi_show_mem(struct vb_config *config);
+extern void vbi_show_shmem(void);
+extern void vbi_show_stat(void);
+extern void vbi_show_ctrl(void);
+extern void vbi_show_cfg(void);
+extern void vbi_show_mem(void);
 extern void vbi_disp_status_regs(void);
 extern void vbi_disp_ctrl_regs(void);
 
-- 
1.6.5.2

