From dbf60538be89c1d7eb94f078be9d5837a0fe8220 Mon Sep 17 00:00:00 2001
From: Paul Gortmaker <paul.gortmaker@windriver.com>
Date: Thu, 5 Nov 2009 17:40:55 -0500
Subject: [PATCH 11/22] v2009.10.16: add vbiCoreSet_t

Integrate addition from Oct16 drop

[cf: 01f,25sep09,mmi  add vbiCore_t type]

Signed-off-by: Paul Gortmaker <paul.gortmaker@windriver.com>
---
 include/vbi/types.h |    1 +
 1 files changed, 1 insertions(+), 0 deletions(-)

diff --git a/include/vbi/types.h b/include/vbi/types.h
index cd9cc1e..b58fcb9 100644
--- a/include/vbi/types.h
+++ b/include/vbi/types.h
@@ -48,6 +48,7 @@ typedef int32_t vbiIntState_t;
 typedef int32_t vbiMemAttr_t;
 typedef int32_t vbiVector_t;
 typedef uint32_t vbiIrq_t;
+typedef uint32_t vbiCoreSet_t;
 
 #endif	/* _ASMLANGUAGE */
 
-- 
1.6.5.2

