From a663ab1b83bfb9fb57b77a4f2a213e9b85950fbc Mon Sep 17 00:00:00 2001
From: Paul Gortmaker <paul.gortmaker@windriver.com>
Date: Thu, 26 Nov 2009 19:34:40 -0500
Subject: [PATCH 6/7] v2009.11.20: fix prototypes of irq core unlock

Integrate the pre Black Monday release

[cf: 02u,12nov09,mmi  update vbiIntVcoreUnlock() prototype]

Signed-off-by: Paul Gortmaker <paul.gortmaker@windriver.com>
---
 include/vbi/syscall.h |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/include/vbi/syscall.h b/include/vbi/syscall.h
index 1488f68..20fcf55 100644
--- a/include/vbi/syscall.h
+++ b/include/vbi/syscall.h
@@ -132,7 +132,7 @@ extern asmlinkage vbiStatus_t  vbi_get_mem_attr(void *vaddr, vbiMemAttr_t * attr
  * START: New APIs introduced for vbi 2.0  
  */
 
-extern asmlinkage vbiStatus_t vbi_vcore_irq_unlock(void);
+extern asmlinkage void vbi_vcore_irq_unlock(void);
 extern asmlinkage vbiIntState_t vbi_vcore_irq_lock(void);
 extern vbiStatus_t vbi_update_text_cache(void *saddr, size_t size);
 extern vbiStatus_t  vbi_set_exc_base(void * excTblBase);
-- 
1.6.5.2

