From 2a999051db5f9323c09f4141d1e4d8e8f9f29cb4 Mon Sep 17 00:00:00 2001
From: Paul Gortmaker <paul.gortmaker@windriver.com>
Date: Thu, 5 Nov 2009 17:40:55 -0500
Subject: [PATCH 09/22] v2009.10.16: add vbiVioapicVcoreIntSend prototype

Integrate addition from Oct16 drop

[cf: 01z,25sep09,mmi  add vbiVioapicVcoreIntSend (irq, bitmap)]

Signed-off-by: Paul Gortmaker <paul.gortmaker@windriver.com>
---
 include/vbi/vbi.h |    3 ++-
 1 files changed, 2 insertions(+), 1 deletions(-)

diff --git a/include/vbi/vbi.h b/include/vbi/vbi.h
index 94016fd..cdec452 100644
--- a/include/vbi/vbi.h
+++ b/include/vbi/vbi.h
@@ -531,7 +531,8 @@ extern vbiStatus_t vbiVioapicIntSend(vbiIrq_t irq, uint32_t filter,
 extern vbiStatus_t vbiVioapicIntSend(vbiIrq_t irq, uint32_t filter, 
 				     vbiVb_t vb);
 extern vbiStatus_t vbiVioapicIntRedirect(vbiIrq_t irq, vbiCore_t tCore);
-
+extern vbiStatus_t vbiVioapicVcoreIntSend (vbiIrq_t irq, vbiCoreSet_t vcoreSet, 
+					  uint32_t options);
 extern vbiStatus_t vbiGuestDmaAddrGet(vbiGuestPhysAddr_t gaddr,
 				vbiPhysAddr_t  *paddr);
 extern vbiStatus_t vbiGuestPhysToPhysAddr(vbiGuestPhysAddr_t gaddr,
-- 
1.6.5.2

