From 82cdaa2cd68fca6ee620e144bd149f2f8ed86d5d Mon Sep 17 00:00:00 2001
From: Paul Gortmaker <paul.gortmaker@windriver.com>
Date: Thu, 5 Nov 2009 17:40:55 -0500
Subject: [PATCH 08/22] v2009.10.16: VCORE_SEND and smp reset fix

Integrate fixes from Oct16 drop

[cf: 02x,05oct09,mmi  add VBI_IOAPICIOCTL_VCORE_SEND, and
 02w,01oct09,mpk  fix smp reset issue]

Signed-off-by: Paul Gortmaker <paul.gortmaker@windriver.com>
---
 include/vbi/syscalls.h |    6 ++++++
 1 files changed, 6 insertions(+), 0 deletions(-)

diff --git a/include/vbi/syscalls.h b/include/vbi/syscalls.h
index 42a6da9..b4a47d2 100644
--- a/include/vbi/syscalls.h
+++ b/include/vbi/syscalls.h
@@ -150,6 +150,7 @@
 #define VBI_IOAPICIOCTL_UNMASK	    1	/* unmask v io apic interrupt src */
 #define VBI_IOAPICIOCTL_SEND	    2	/* inject a v io apic interrupt */
 #define VBI_IOAPICIOCTL_EOI	    3	/* end of interrupt acknowledge */
+#define VBI_IOAPICIOCTL_VCORE_SEND  4	/* Send an irq to a core	*/
 /* remove this after vbControl is working, we don't need a hypercall for it: */
 #define VBI_IOAPICIOCTL_MASK	 10	/* mask v io apic interrupt src */
 
@@ -160,6 +161,10 @@
 #define VBI_IOAPICSEND_UNICAST	 3	/* send interrupt to a only one vb*/
 #define VBI_IOAPICSEND_NONE	 4	/* ignore this call	       */
 
+/* sent interrupt to a virtual flags */
+#define VBI_IOAPICSEND_VCORE_NONE 0	/* there no option available for now */
+
+
 /* hyCtxctl system call supported operations */
 #define VBI_CTXCTL_IDLE	 	1	/* make current context idle	*/
 
@@ -199,6 +204,7 @@
 #define VBI_ERR_INVALID_ARG         -102   /* General error */
 
 /* vbMgmt reset macros */
+#define VBI_VBMGMT_RESET_NON_CORE0	    0x00000000     /* APs dont clear mem */
 #define VBI_VBMGMT_RESET_DOWNLOAD	 0x00000001 /* disable ELF reloading */
 #define VBI_VBMGMT_RESET_CLEAR		 0x00000002 /* do not clear memory */
 #define VBI_VBMGMT_RESET_AND_START_CORE0 0x00000004
-- 
1.6.5.2

