From 11215b0de7fc397b7cb8dbb4fe7d3b99ef4e097b Mon Sep 17 00:00:00 2001
From: Paul Gortmaker <paul.gortmaker@windriver.com>
Date: Thu, 26 Nov 2009 20:11:16 -0500
Subject: [PATCH 4/7] v2009.11.13: add compatibility mode descriptors

Integrate Friday the 13th release

[cf: 03t,26oct09,gws  add compatibility mode descriptors]

Signed-off-by: Paul Gortmaker <paul.gortmaker@windriver.com>
---
 include/asm-x86/sys_vbi.h |   25 +++++++++++++------------
 1 files changed, 13 insertions(+), 12 deletions(-)

diff --git a/include/asm-x86/sys_vbi.h b/include/asm-x86/sys_vbi.h
index 8381d04..687bfa3 100644
--- a/include/asm-x86/sys_vbi.h
+++ b/include/asm-x86/sys_vbi.h
@@ -306,18 +306,19 @@ CPU_PRIVATE extern ctx_t excMgrId;	/* exception mananger */
 #define TSS_FIRST_CTX_ENTRY	HY_MAX_CPU
 
 /* GDT assignments */
-
-#define	CS_SYS_GDT_ENTRY	1	/* entry  1: code for the supervisor */
-#define	DS_SYS_GDT_ENTRY	2	/* entry  2: data for the supervisor */
-#define	CS_EXC_GDT_ENTRY	3	/* entry  3: code for exceptions     */
-#define	CS_INT_GDT_ENTRY	4	/* entry  4: code for interrupts     */
-#define	CS_CTX_GDT_ENTRY	5	/* entry  5: code for the user task  */
-#define	DS_CTX_GDT_ENTRY	6	/* entry  6: data for the user task  */
-#define	HY_TSS_GDT_ENTRY	7	/* entry  7: task state (hypervisor) */
-					/* entry  7+cpu: system call gate    */
-
-#ifdef	LP64
-/* In 64 bits mode, a system segment entry requires to GDT entry slots */
+#define	CS_SYS_GDT_ENTRY	1	/* code for the supervisor */
+#define	DS_SYS_GDT_ENTRY	2	/* data for the supervisor */
+#define	CS_RMD_GDT_ENTRY	3	/* code for real-mode supv */
+#define	DS_RMD_GDT_ENTRY	4	/* data for real-mode supv */
+#define	CS_32C_GDT_ENTRY	5	/* code for 32bit compatible */
+#define	CS_EXC_GDT_ENTRY	6	/* code for exceptions     */
+#define	CS_INT_GDT_ENTRY	7	/* code for interrupts     */
+#define	CS_CTX_GDT_ENTRY	8	/* code for the user task  */
+#define	DS_CTX_GDT_ENTRY	9	/* data for the user task  */
+#define	HY_TSS_GDT_ENTRY	10	/* task state (hypervisor) */
+					/* 10+cpu: system call gate */
+#ifdef LP64
+/* In 64 bits mode, a system segment entry requires two GDT entry slots */
 
 #define	HY_SYS_SEG_ENTRY_SIZE	2
 #else	/* LP64 */
-- 
1.6.5.2

