From 1014d5eb7f7de32a21df269b96684c5f8f6a6ed1 Mon Sep 17 00:00:00 2001
From: Paul Gortmaker <paul.gortmaker@windriver.com>
Date: Thu, 5 Nov 2009 17:40:55 -0500
Subject: [PATCH 06/22] v2009.10.16: reg_vbi update

Integrate updated list of CPUs from Oct16 drop

[cf: 02a,01oct09,md   add MSR_PLATFORM_INFO MSR]

Signed-off-by: Paul Gortmaker <paul.gortmaker@windriver.com>
---
 include/asm-x86/reg_vbi.h |   43 ++++++++++++++++++++++++++++++++++++++++++-
 1 files changed, 42 insertions(+), 1 deletions(-)

diff --git a/include/asm-x86/reg_vbi.h b/include/asm-x86/reg_vbi.h
index 2d597b4..d8f65ff 100644
--- a/include/asm-x86/reg_vbi.h
+++ b/include/asm-x86/reg_vbi.h
@@ -127,7 +127,7 @@ typedef union
 #define CPUID_FAMILY_EXTENDED	15	/* family: Extended		*/
 #define CPUID_FAMILY_PENTIUM4	 0	/* extended family: PENTIUM4	*/
 
-/* Pentium PRO family models */
+/* Intel family models */
 
 #define CPUID_MODEL_PPRO 	 1 	/* model: Pentium Pro		*/
 #define CPUID_MODEL_P2_OVER 	 2 	/* model: Pentium II Overdrive	*/
@@ -143,6 +143,12 @@ typedef union
 #define CPUID_MODEL_PENTIUMM_0D	13 	/* model: Pentium M, model 0D	*/
 #define CPUID_MODEL_CORE 	14 	/* model: Core Duo/Solo		*/
 #define CPUID_MODEL_CORE2	15	/* model: Core2 Duo/Quad/Xeon	*/
+#define CPUID_MODEL_I7		26	/* model: Core i7		*/
+#define CPUID_MODEL_NEHALEM1	30	/* model: Core Architecture 	*/
+#define CPUID_MODEL_NEHALEM2	32	/* model: Core Architecture 	*/
+#define CPUID_MODEL_NEHALEM3	46	/* model: Core Architecture 	*/
+#define CPUID_MODEL_WESTMERE1	37	/* model: next generation	*/
+#define CPUID_MODEL_WESTMERE2	44	/* model: next generation	*/
 
 
 /* CPUID fields in the EBX register when EAX=1 */
@@ -366,6 +372,35 @@ typedef struct mtrr		/* MTRR */
 
 typedef MTRR *		MTRR_ID;
 
+/* PLATFORM_INFO MSR fields */
+
+typedef union
+    {
+    struct
+	{
+	uint64_t reserved1:8;
+	uint64_t ratio_max:8;	/* maximum non-turbo TSC ratio	*/
+	uint64_t reserved2:24;
+	uint64_t ratio_min:8;	/* minimum non-turbo TSC ratio	*/
+	uint64_t reserved3:16;
+	} field;
+    uint64_t value;
+    } PLATFORM_INFO;
+
+/* MSR_FLEX_RATIO MSR fields */
+
+typedef union
+    {
+    struct
+	{
+	uint64_t reserved1:8;
+	uint64_t ratio:8;	/* current non-turbo TSC ratio	*/
+	uint64_t enabled:1;	/* flex ratio enabled		*/
+	uint64_t reserved2:47;
+	} field;
+    uint64_t value;
+    } FLEX_RATIO;
+
 /*
  * Structure of a global descriptor table (GDT) entry
  * (common to 32 & 64 bits)
@@ -1402,6 +1437,12 @@ extern void	x86GetCpuid (uint32_t operation,
 
 #define	MSR_FSB_FREQ		0x00cd
 
+/* MSR, Nehalem */
+
+#define	MSR_PLATFORM_INFO	0x00ce
+#define MSR_FLEX_RATIO		0x0194
+#define MSR_CORE_THREAD_COUNT	0x0035
+
 /* MSR, VMX specific */
 
 #define	MSR_VMX_FEATURE			0x003A
-- 
1.6.5.2

