From 4032b373c5ea26eaabe3ead64c6f44b604e1ab99 Mon Sep 17 00:00:00 2001
From: Paul Gortmaker <paul.gortmaker@windriver.com>
Date: Mon, 19 Oct 2009 14:21:00 -0400
Subject: [PATCH 23/23] vbi: remove CPU related type cruft

The whole CPU type related stuff was used to demux on the
particular arch for the build.  But we already have this
for free now, with the include files in sane places.  So
delete all the old demux poo.

Signed-off-by: Paul Gortmaker <paul.gortmaker@windriver.com>
---
 include/vbi/arch.h  |   20 +-------------------
 include/vbi/types.h |    4 +---
 2 files changed, 2 insertions(+), 22 deletions(-)

diff --git a/include/vbi/arch.h b/include/vbi/arch.h
index d47d0b6..f7118b8 100644
--- a/include/vbi/arch.h
+++ b/include/vbi/arch.h
@@ -24,27 +24,9 @@ modification history
 #define __VBI_PDP_ENDIAN    3412  /* LSB first in word, MSW first in long */
 
 /* include the appropriate arch specific header file */
-
-#if (CPU == MIPSI64R2)
-# include <mips64/vbiMips64Arch.h>
-#elif (CPU == PPC85XX) || (CPU == PPC32) || (CPU == PPCE500MC)
-#ifdef CONFIG_WRHV
-#include <vbi/ppc/vbiPpcArch.h>
-#else
-# include <ppc/vbiPpcArch.h>
-#endif /* CONFIG_WRHV */
-#elif (CPU == PENTIUM) || (CPU == PENTIUM4)
-#ifdef CONFIG_WRHV
-# include "x86/vbiX86Arch.h"
-#else
-# include <x86/vbiX86Arch.h>
-#endif /* CONFIG_WRHV */
-#else
-# error "no arch header file included"
-#endif
+#include <asm/arch_vbi.h>
 
 /* ensure the arch has specified a byte ordering */
-
 #if !defined(__VBI_BYTE_ORDER)
 # error: Architecture has not defined a byte order!
 #endif
diff --git a/include/vbi/types.h b/include/vbi/types.h
index 6daba40..d064a7e 100644
--- a/include/vbi/types.h
+++ b/include/vbi/types.h
@@ -22,6 +22,7 @@ modification history
 #ifndef __INCvbiTypesh
 #define __INCvbiTypesh
 
+/* The CPU specifics here should be relocated to arch dirs */
 #ifdef CONFIG_WRHV
 #ifdef CONFIG_X86
 #define CPU PENTIUM
@@ -33,9 +34,6 @@ modification history
 #define _ASMLANGUAGE
 #endif /* __ASSEMBLY__ */
 #endif /* CONFIG_PPC */
-#include "support/sys/cpu_types.h"
-#else
-#include <ctype.h>
 #endif /* CONFIG_WRHV */
 
 #ifdef __cplusplus
-- 
1.6.3.3

