From fb1bb400f8a8d16d0267f5a104e894d53a373d10 Mon Sep 17 00:00:00 2001
From: Paul Gortmaker <paul.gortmaker@windriver.com>
Date: Fri, 20 Nov 2009 14:08:10 -0500
Subject: [PATCH 80/80] vbi: remove network <---> host translation macros

We already have these in linux; we don't need yet another
copy floating around.

Signed-off-by: Paul Gortmaker <paul.gortmaker@windriver.com>
---
 include/asm-x86/sys_vbi.h |   18 ------------------
 1 files changed, 0 insertions(+), 18 deletions(-)

diff --git a/include/asm-x86/sys_vbi.h b/include/asm-x86/sys_vbi.h
index 790c9aa..8381d04 100644
--- a/include/asm-x86/sys_vbi.h
+++ b/include/asm-x86/sys_vbi.h
@@ -655,24 +655,6 @@ CPU_PRIVATE extern ctx_t excMgrId;	/* exception mananger */
 #define	VMXON_REGION_NPAGES	1	/* number of 4KB pages for VMXON */
 #define	VMCS_REGION_NPAGES	1	/* number of 4KB pages for VMCS */
 
-/* network <-> host translations (for use by romfs) */
-
-#define ntohl(x)        ((((x) & 0x000000ff) << 24) | \
-                         (((x) & 0x0000ff00) <<  8) | \
-                         (((x) & 0x00ff0000) >>  8) | \
-                         (((x) & 0xff000000) >> 24))
-
-#define htonl(x)        ((((x) & 0x000000ff) << 24) | \
-                         (((x) & 0x0000ff00) <<  8) | \
-                         (((x) & 0x00ff0000) >>  8) | \
-                         (((x) & 0xff000000) >> 24))
-
-#define ntohs(x)        ((((x) & 0x00ff) << 8) | \
-                         (((x) & 0xff00) >> 8))
-
-#define htons(x)        ((((x) & 0x00ff) << 8) | \
-                         (((x) & 0xff00) >> 8))
-
 /* context virtual address start for the VB */
 
 #define CTX_VADDR_START 		0x0
-- 
1.6.5.2

