From b01852799298648298d745fd6cbdb4e08ff703bb Mon Sep 17 00:00:00 2001
From: Paul Gortmaker <paul.gortmaker@windriver.com>
Date: Thu, 5 Nov 2009 19:00:42 -0500
Subject: [PATCH 10/22] vbi: introduce core_set.h

It appears that other vbi updates may make this file required.
This is the cleaned up version.

[cf: 01a,09jun09,mmi  written]

Signed-off-by: Paul Gortmaker <paul.gortmaker@windriver.com>
---
 include/vbi/core_set.h |   39 +++++++++++++++++++++++++++++++++++++++
 1 files changed, 39 insertions(+), 0 deletions(-)
 create mode 100644 include/vbi/core_set.h

diff --git a/include/vbi/core_set.h b/include/vbi/core_set.h
new file mode 100644
index 0000000..b629dfc
--- /dev/null
+++ b/include/vbi/core_set.h
@@ -0,0 +1,39 @@
+/*
+ * core_set.h - macros to manipulate vbiCoreSet_t type
+ *
+ * Copyright (c) 2009 Wind River Systems, Inc.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 2 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+ * See the GNU General Public License for more details.
+ *
+ */
+
+#ifndef _VBI_CORE_SET_H
+#define _VBI_CORE_SET_H
+
+#ifndef _ASMLANGUAGE
+
+/* set of macros to manipulate the coreset_t type */
+
+#define CORESET_COPY(coreSetDst, coreSetSrc)  (coreSetDst = coreSetSrc)
+#define CORESET_ISZERO(coreset)         (coreset == 0 ? TRUE : FALSE)
+#define CORESET_SET(coreset, n)         (coreset |= (1 << n))
+#define CORESET_ISSET(coreset, n)       (coreset & (1 << n))
+#define CORESET_CLR(coreset, n)         (coreset &= ~(1 << n))
+#define CORESET_ZERO(coreset)           (coreset = 0)
+
+#define CORESET_FIRST_SET(coreset)                ((int)coreset & (int)-coreset)
+#define CORESET_FIRST_INDEX(coreset)              (FFS_LSB (coreset) - 1)
+#define CORESET_ADD(coresetAugend,coresetAddend)   (coresetAugend |= coresetAddend)
+#define CORESET_SUB(coresetMinuend,coresetSubtrahend)      \
+        (coresetMinuend &= ~coresetSubtrahend)
+
+#endif
+
+#endif  /* _VBI_CORE_SET_H */
-- 
1.6.5.2

