From fb024913a66a7d3e00b4a5446e891f59d78209a0 Mon Sep 17 00:00:00 2001
From: Paul Gortmaker <paul.gortmaker@windriver.com>
Date: Tue, 1 Dec 2009 21:12:18 -0500
Subject: [PATCH 59/59] vbi: whitespace cleanup; no real changes here

Signed-off-by: Paul Gortmaker <paul.gortmaker@windriver.com>
---
 arch/powerpc/kernel/vbi/wrhv.c |   18 ++++++++----------
 1 files changed, 8 insertions(+), 10 deletions(-)

diff --git a/arch/powerpc/kernel/vbi/wrhv.c b/arch/powerpc/kernel/vbi/wrhv.c
index a27d6f2..fbe5e44 100644
--- a/arch/powerpc/kernel/vbi/wrhv.c
+++ b/arch/powerpc/kernel/vbi/wrhv.c
@@ -510,11 +510,10 @@ void __init wrhv_time_init_cont(void)
 
 
 /* arch/powerpc/mm/fault.c */
-void wrhv_vmmu_restore (void)
+void wrhv_vmmu_restore(void)
 {
 	/*
-	 * this function is called by the end of page fault handling to reinstall
-	 * the vmmu
+	 * called by the end of page fault handling to reinstall the vmmu
 	 */
 	wr_control->vmmu0 = wr_status->vmmu0;
 	wr_control->vmmu1 = wr_status->vmmu1;
@@ -601,7 +600,8 @@ void __init wrhv_MMU_init(void)
 	btext_unmap();
 #endif
 
-	/* we enables the mmu here without having to do this from the caller
+	/*
+	 * we enable the mmu here without having to do this from the caller
 	 * (which is in assembly world)
 	 */
 	vb_context_mmu_on(0, swapper_pg_dir, PAGE_SIZE, 0);
@@ -758,10 +758,8 @@ int __init wrhv_gfar_of_init(void)
 			of_node_put(mdio);
 		}
 
-		ret =
-		    platform_device_add_data(gfar_dev, &gfar_data,
-					     sizeof(struct
-						    gianfar_platform_data));
+		ret = platform_device_add_data(gfar_dev, &gfar_data,
+					sizeof(struct gianfar_platform_data));
 		if (ret)
 			goto unreg;
 	}
@@ -939,8 +937,8 @@ int wrhv_kgdb_arch_handle_exception(int vector, int signo, int err_code,
 			wr_control->vb_control_regs.dbcr0 |= (DBCR0_IC | DBCR0_IDM);
 
 			/*
-			 * Set Debug Enable bit in hypervisor's emulated MSR
-			 * which is restored on return from exception
+			 * Set Debug Enable bit in hypervisor's emulated MSR which
+			 * is restored on return from exception
 			 */
 			wr_control->vb_control_regs.emsr |= MSR_DE;
 #else
-- 
1.6.5.2

