From a5cc73c2d60bdf34d644bf350bf99a8a156de412 Mon Sep 17 00:00:00 2001
From: Paul Gortmaker <paul.gortmaker@windriver.com>
Date: Tue, 3 Nov 2009 22:36:50 -0500
Subject: [PATCH 05/80] vbi: newIntDisable -> next_irq_disable

for i in `git grep -l newIntDisable` ; do echo $i ; mv $i $i~ ; cat $i~ |sed 's/newIntDisable/next_irq_disable/g' > $i ; done

Signed-off-by: Paul Gortmaker <paul.gortmaker@windriver.com>
---
 include/vbi/interface.h |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/include/vbi/interface.h b/include/vbi/interface.h
index 44aae33..5fb13ed 100644
--- a/include/vbi/interface.h
+++ b/include/vbi/interface.h
@@ -309,7 +309,7 @@ struct vb_control
 #endif
 
 	uint32_t irq_disable;		/*  0: -1 => Disable all interrupts */
-	uint32_t newIntDisable;		/*  1: New value for ->irq_disable */
+	uint32_t next_irq_disable;		/*  1: New value for ->irq_disable */
 
 					/*  2: VMMU 0 table for ctxt switch */
 	VB_ALIGN_FIELD_64 (void *vmmu0, pad1);
-- 
1.6.5.2

