From 9a28f89682e65e941738d58ccf46283923a860cd Mon Sep 17 00:00:00 2001
From: Paul Gortmaker <paul.gortmaker@windriver.com>
Date: Wed, 2 Dec 2009 13:49:51 -0500
Subject: [PATCH 13/43] vbi: vbiIntState_t --> int32_t

for i in `git grep -l vbiIntState_t` ; do mv $i $i~ ; cat $i~|sed 's/vbiIntState_t/int32_t/' > $i ; done

Signed-off-by: Paul Gortmaker <paul.gortmaker@windriver.com>
---
 include/vbi/compat.h  |    1 +
 include/vbi/syscall.h |    2 +-
 include/vbi/types.h   |    1 -
 3 files changed, 2 insertions(+), 2 deletions(-)

diff --git a/include/vbi/compat.h b/include/vbi/compat.h
index 3717811..fb81708 100644
--- a/include/vbi/compat.h
+++ b/include/vbi/compat.h
@@ -52,6 +52,7 @@
 #define vbiCtx_t		int32_t
 #define vbiPhysAddr_t		uint64_t
 #define vbiGuestPhysAddr_t	void *
+#define vbiIntState_t		int32_t
 
 /*
  * struct fields and similar
diff --git a/include/vbi/syscall.h b/include/vbi/syscall.h
index 571261b..6c7e738 100644
--- a/include/vbi/syscall.h
+++ b/include/vbi/syscall.h
@@ -133,7 +133,7 @@ extern asmlinkage int32_t  vbi_get_mem_attr(void *vaddr, vbiMemAttr_t * attr);
  */
 
 extern asmlinkage void vbi_vcore_irq_unlock(void);
-extern asmlinkage vbiIntState_t vbi_vcore_irq_lock(void);
+extern asmlinkage int32_t vbi_vcore_irq_lock(void);
 extern int32_t vbi_update_text_cache(void *saddr, size_t size);
 extern int32_t vbi_set_exc_base(void *excTblBase);
 
diff --git a/include/vbi/types.h b/include/vbi/types.h
index 31bb40b..0824a88 100644
--- a/include/vbi/types.h
+++ b/include/vbi/types.h
@@ -34,7 +34,6 @@
 
 #ifndef _ASMLANGUAGE
 
-typedef int32_t vbiIntState_t;
 typedef int32_t vbiMemAttr_t;
 typedef int32_t vbiVector_t;
 typedef uint32_t vbiCoreSet_t;
-- 
1.6.5.2

