From a1ea0027c85673973c12baf6b4987345abbc2329 Mon Sep 17 00:00:00 2001
From: Paul Gortmaker <paul.gortmaker@windriver.com>
Date: Wed, 28 Oct 2009 20:50:03 -0400
Subject: [PATCH 02/26] vbi: introduce a compat.h file

The desire is to ensure all vbi users are using linux-like
calls into the virtual board infrastructure; however there
may be those who are carrying forward legacy code that isn't
easily changed.  Those users should be able to get legacy
mapping by including this file.

It will start off without exposing any compat defines, and
they will be added as they are removed from the global scope
of other files.

Signed-off-by: Paul Gortmaker <paul.gortmaker@windriver.com>
---
 include/vbi/compat.h |   31 +++++++++++++++++++++++++++++++
 1 files changed, 31 insertions(+), 0 deletions(-)
 create mode 100644 include/vbi/compat.h

diff --git a/include/vbi/compat.h b/include/vbi/compat.h
new file mode 100644
index 0000000..db1279d
--- /dev/null
+++ b/include/vbi/compat.h
@@ -0,0 +1,31 @@
+/*
+ * compat.h - virtual board compatibility support definitions
+ *
+ * Copyright (c) 2009 Wind River Systems, Inc.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 2 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+ * See the GNU General Public License for more details.
+ *
+ */
+
+/*
+ * This introduces some of the non-linux like things into scope for
+ * those VBI users who have an expectation that they will be present.
+ * None of the default linux VBI code needs to include this file.
+ * Only those source files which expect default/legacy VBI API
+ * content need to include this file.
+ */
+
+
+#ifndef _VBI_COMPAT_H
+#define _VBI_COMPAT_H
+
+#include <vbi/vbi.h>
+
+#endif  /* _VBI_COMPAT_H */
-- 
1.6.5.2

