From 88bc65b5d8ba48025c40757b9af6f697f2a1d5e1 Mon Sep 17 00:00:00 2001
From: Paul Gortmaker <paul.gortmaker@windriver.com>
Date: Thu, 29 Oct 2009 23:23:24 -0400
Subject: [PATCH 14/26] vbi: kill off VB_WRHV_NAME typedef

Signed-off-by: Paul Gortmaker <paul.gortmaker@windriver.com>
---
 include/vbi/interface.h |   11 +++++------
 1 files changed, 5 insertions(+), 6 deletions(-)

diff --git a/include/vbi/interface.h b/include/vbi/interface.h
index 4713662..6ad4298 100644
--- a/include/vbi/interface.h
+++ b/include/vbi/interface.h
@@ -130,7 +130,6 @@ identity mapped.
 #define VB_NAMELEN		64
 
 #ifndef	_ASMLANGUAGE
-typedef int8_t VB_WRHV_NAME[VB_NAMELEN];
 
 #define VB_MAX_BOOTLINE_LENGTH		256
 
@@ -447,7 +446,7 @@ struct vb_status {
 /* Configuration information for interrupts */
 struct vb_int_info
 {
-	VB_WRHV_NAME intName;	/* the name of this interrupt */
+	char intName[VB_NAMELEN];	/* the name of this interrupt */
 	uint16_t intDirection;	/* interrupt direction: INPUT_INT, OUTPUT_INT */
 	uint16_t intNumber;	/* the local VB interrupt number */
 	uint32_t intCore;	/* the receiving core for this incomming int */
@@ -456,7 +455,7 @@ struct vb_int_info
 /* Information about a shared memory region */
 struct vb_sm_info
 {
-	VB_WRHV_NAME name;	/* the name of the shared memory region */
+	char name[VB_NAMELEN];	/* the name of the shared memory region */
 				/* the vbphysical address of the region */
 	VB_ALIGN_FIELD_64 (void *addr, pad1);
 	uint32_t length;	/* the length in bytes of the region */
@@ -469,12 +468,12 @@ struct vb_sm_info
 /* Information for the memory map info */
 struct vb_mem_info
 {
-	VB_WRHV_NAME name;	/* the name of the memory region */
+	char name[VB_NAMELEN];	/* the name of the memory region */
 				/* the vbphysical address of the region */
 	VB_ALIGN_FIELD_64 (void *addr, pad1);
 	uint32_t length;	/* the length in bytes of the region */
 	uint32_t attr;		/* the MMU attributes of the region */
-	VB_WRHV_NAME type;		/* the type of the region */
+	char type[VB_NAMELEN];		/* the type of the region */
 };
 
 
@@ -523,7 +522,7 @@ struct vb_config
 
 	int32_t supervisoryMode;/* board is running in privileged mode */
 
-	VB_WRHV_NAME boardName;	/* the name of this board */
+	char boardName[VB_NAMELEN];	/* the name of this board */
 	uint32_t boardType;	/* the type of this board */
 	uint32_t bootCount;	/* number of times this core has booted */
 
-- 
1.6.5.2

