From 5171b7e286d2871ef8657ba46be141ce6870819d Mon Sep 17 00:00:00 2001
From: Paul Gortmaker <paul.gortmaker@windriver.com>
Date: Wed, 11 Nov 2009 21:09:29 -0500
Subject: [PATCH 01/80] vbi: add more compat extensions

Extend the legacy name mapping to include more functions and
similar that are now going to use linux like naming.
Anyone assuming the legacy naming scheme is still being
used in linux won't get broken code if they include this file.

Signed-off-by: Paul Gortmaker <paul.gortmaker@windriver.com>
---
 include/vbi/compat.h |  115 ++++++++++++++++++++++++++++++++++++++++++++++++++
 1 files changed, 115 insertions(+), 0 deletions(-)

diff --git a/include/vbi/compat.h b/include/vbi/compat.h
index c55867a..af8e721 100644
--- a/include/vbi/compat.h
+++ b/include/vbi/compat.h
@@ -48,4 +48,119 @@
 
 #define VB_TIMESTAMP		uint64_t
 
+/*
+ * struct fields and similar
+ */
+#define memoryAliasSize		mem_alias_size
+#define memoryAliasAddress	mem_alias_addr
+#define physicalMemorySize	phys_mem_size
+
+#define numInt		num_int
+#define numSm		num_sm
+#define numMem		num_mem
+
+#define intPending	int_pend
+#define resetPC		reset_pc
+#define bootCount	boot_count
+#define boardName	board_name
+#define boardType	board_type
+
+#define tickTimerFrequency	tick_freq
+#define tickCount		tick_count
+#define timeStampFrequency	stamp_freq
+
+#define intDisable		irq_disable
+#define newIntDisable		next_irq_disable
+#define oldIntDisable		prev_irq_disable
+#define intName			irq_name
+#define intNumber		irq_num
+#define intDirection		irq_dir
+
+/*
+ * function calls, etc.
+ */
+#define vbiVbSuspend		vbi_vb_suspend
+#define vbiVbResume		vbi_vb_resume
+#define vbiVbRestart		vbi_vb_restart
+#define vbiVbReset		vbi_vb_reset
+
+#define vbiVbRegisterRead	vbi_vb_read_reg
+#define vbiVbRegisterWrite	vbi_vb_write_reg
+#define vbiVbMemoryRead		vbi_vb_read_mem
+#define vbiVbMemoryWrite	vbi_vb_write_mem
+
+#define vbiVcoreIntRed_op	vbi_vcore_irq_redirect
+#define vbiDebugShellStart	vbi_shell_start_debug
+#define vbiIoapicOp		vbi_io_apic_op
+#define vbiIoapicIoctl		vbi_io_apic_ioctl
+#define vbiHyIoctl		vbi_hy_ioctl
+#define vbiCtxctl		vbi_ctx_ctl
+#define vbiVbMgmt		vbi_vb_mgmt
+
+#define vbiSend			vbi_send
+#define vbiReceive		vbi_receive
+#define vbiReply		vbi_reply
+#define vbiPanic		vbi_panic
+#define vbiPs			vbi_ps
+#define vbiKputs		vbi_kputs
+#define vbiKputc		vbi_kputc
+
+#define vbiIcacheFlush		vbi_flush_icache
+#define vbiDcacheFlush		vbi_flush_dcache
+#define vbiCacheTextUpdate	vbi_update_text_cache
+#define vbiTlbFlush		vbi_flush_tlb
+#define vbiVtlbOp		vbi_vtlb_op
+
+#define vbiIntVCoreUnlock	vbi_vcore_irq_unlock
+#define vbiIntVCoreLock		vbi_vcore_irq_lock
+#define vbiIntVCoreStateGet	vbi_vcore_irq_state
+#define vbiExcBaseSet		vbi_set_exc_base
+#define vbiExcOffsetsSet	vbi_set_exc_offset
+#define vbiExcOffsetsGet	vbi_get_exc_offset
+#define vbiMemAttrSet		vbi_set_mem_attr
+#define vbiMemAttrGet		vbi_get_mem_attr
+
+#define vbiVmmuConfig		vbi_config_vmmu
+#define vbiVmmuEnable		vbi_enable_vmmu
+#define vbiVmmuDisable		vbi_disable_vmmu
+#define vbiVmmuTlbLoad		vbi_tlb_load_vmmu
+#define vbiVmmuTlbFlush		vbi_tlb_flush_vmmu
+
+#define vbiNsOp			vbi_ns_op
+#define vbiNsLookup		vbi_ns_lookup
+#define vbiNsRegister		vbi_ns_register
+#define vbiNsUnRegister		vbi_ns_unregister
+
+#define vbiConfigShow		vbi_show_cfg
+#define vbiControlShow		vbi_show_ctrl
+#define vbiCtrlRegsDisplay	vbi_disp_ctrl_regs
+#define vbiInterruptsShow	vbi_show_irq
+#define vbiMemoryShow		vbi_show_mem
+#define vbiSharedMemoryShow	vbi_show_shmem
+#define vbiStatusShow		vbi_show_stat
+#define vbiStsRegsDisplay	vbi_disp_status_regs
+
+#define vbiExcStub		vbi_exc_stub
+#define vbiGuestDmaAddrGet	vbi_get_guest_dma_addr
+#define vbiGuestPhysToPhysAddr	vbi_guest_phys_to_phys
+#define vbiIdle			vbi_idle
+#define vbiInit			vbi_init
+#define vbiMmuAttrSet		vbi_set_mmu_attr
+
+#define vbiCorePrvMemFind	vbi_find_core_prv_mem
+#define vbiIntVecFind		vbi_find_irq
+#define vbiMemRegionFind	vbi_find_mem
+#define vbiShmemRegionFind	vbi_find_shmem
+
+#define vbiVioapicAddrGet	vbi_get_vioapic_addr
+#define vbiVioapicDisplay	vbi_disp_vioapic
+#define vbiVioapicIntAck	vbi_ack_vioapic_irq
+#define vbiVioapicIntMask	vbi_mask_vioapic_irq
+#define vbiVioapicIntRedirect	vbi_redir_vioapic_irq
+#define vbiVioapicIntSend	vbi_send_vioapic_irq
+#define vbiVioapicIntUnmask	vbi_unmask_vioapic_irq
+#define vbiVioapicVcoreIntSend	vbi_send_vcore_vioapic_irq
+#define vbiVioapicVectorGet	vbi_get_vioapic_vec
+#define vbiVioapicVectorSet	vbi_set_vioapic_vec
+
 #endif  /* _VBI_COMPAT_H */
-- 
1.6.5.2

