From d8ecf8a60b74fdad065bbd6fa3dae5b6c41a8bf3 Mon Sep 17 00:00:00 2001
From: Paul Gortmaker <paul.gortmaker@windriver.com>
Date: Thu, 5 Nov 2009 19:43:22 -0500
Subject: [PATCH 13/22] vbi: use cpu_types from vbi reference implementation

The reference implementation had values for the CPU types
so we might as well be consistent with that by using them.
Also delete the unrequired WRHV config check wrapper.

Signed-off-by: Paul Gortmaker <paul.gortmaker@windriver.com>
---
 include/vbi/cpu_types.h |   76 +++++++++++++++++++++++++++++++++++++++++++++++
 include/vbi/types.h     |   18 ++++-------
 2 files changed, 83 insertions(+), 11 deletions(-)
 create mode 100644 include/vbi/cpu_types.h

diff --git a/include/vbi/cpu_types.h b/include/vbi/cpu_types.h
new file mode 100644
index 0000000..dbceb30
--- /dev/null
+++ b/include/vbi/cpu_types.h
@@ -0,0 +1,76 @@
+/*
+ * Copyright 2007-2009 Wind River Systems, Inc.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 2 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+ * See the GNU General Public License for more details.
+ *
+ */
+
+#ifndef _VBI_CPU_TYPES_H
+#define _VBI_CPU_TYPES_H
+
+/* define the cpu types that we currently support */
+#define PPC85XX		1
+#define PENTIUM		2
+#define ARM1136		3
+#define MIPSI64R2	4
+#define PPCE500MC	5
+
+#ifndef _ASMLANGUAGE
+
+/* MIPS64R2 specific types */
+#if (CPU == MIPSI64R2)
+
+#define CPU_MIPS_32BIT ((CPU==_VX_MIPS32) || \
+			    (CPU==MIPSI2) || \
+			    (CPU==MIPSI32) || \
+			    (CPU==MIPSI32R2))
+
+#define CPU_MIPS_64BIT ((CPU==MIPS64) || \
+			    (CPU==MIPSI3) || \
+			    (CPU==MIPSI64) || \
+			    (CPU==MIPSI64R2))
+
+typedef unsigned long	INSTR;		/* 32 bit word-aligned instructions */
+
+#if (CPU_MIPS_32BIT)
+typedef unsigned int	_RType;		/* registers are 32 bits */
+#elif (CPU_MIPS_64BIT)
+typedef unsigned long	_RType;		/* registers are 64 bits */
+#else /* CPU_MIPSxxBIT */
+#error "Invalid CPU value"
+#endif
+
+#endif /* CPU == MIPSI64R2 */
+
+/* PowerPC specific types */
+#if (CPU == PPC85XX) || (CPU == PPCE500MC)
+typedef unsigned long	INSTR;		/* 32 bit word-aligned instructions */
+typedef unsigned int	_RType;		/* register type */
+#endif
+
+/* Pentium specific types */
+#if (CPU == PENTIUM)
+typedef unsigned char	INSTR;		/* char instructions */
+					/* register type */
+#if defined (LP64)
+typedef unsigned long	_RType;
+#else
+typedef unsigned int	_RType;
+#endif
+#endif /* CPU == PENTIUM */
+
+#if (CPU == ARM1136)
+typedef unsigned long	INSTR;		/* 32 bit word-aligned instructions */
+typedef unsigned int	_RType;		/* register type */
+#endif
+
+#endif /* _ASMLANGUAGE */
+
+#endif /* _VBI_CPU_TYPES_H */
diff --git a/include/vbi/types.h b/include/vbi/types.h
index b58fcb9..c112a6a 100644
--- a/include/vbi/types.h
+++ b/include/vbi/types.h
@@ -14,27 +14,23 @@
  *
  */
 
-
 #ifndef _VBI_TYPES_H
 #define _VBI_TYPES_H
 
-/* The CPU specifics here should be relocated to arch dirs */
-
-/* Currently supported CPUs */
-#define PENTIUM		1
-#define PPC85XX		2
+#ifdef __ASSEMBLY__
+#define _ASMLANGUAGE
+#endif /* __ASSEMBLY__ */
 
-#ifdef CONFIG_WRHV
+/* The CPU specifics here should be relocated to arch dirs */
 #ifdef CONFIG_X86
 #define CPU PENTIUM
 #endif /* CONFIG_X86 */
+
 #ifdef CONFIG_PPC
 #define CPU PPC85XX   /* Currently all PPC treated the same in vbi */
-#ifdef __ASSEMBLY__
-#define _ASMLANGUAGE
-#endif /* __ASSEMBLY__ */
 #endif /* CONFIG_PPC */
-#endif /* CONFIG_WRHV */
+
+#include <vbi/cpu_types.h>	/* for PENTIUM, PPC85XX, etc */
 
 #ifndef _ASMLANGUAGE
 
-- 
1.6.5.2

