From 9b4378ec5722d223c3f372f654accadda4c5082a Mon Sep 17 00:00:00 2001
From: Paul Gortmaker <paul.gortmaker@windriver.com>
Date: Wed, 14 Oct 2009 18:41:58 -0400
Subject: [PATCH 01/27] vbi: the kernel is not libc

None of this crap is being used, and the libc-like content
doesn't belong in the kernel anyway.

Signed-off-by: Paul Gortmaker <paul.gortmaker@windriver.com>
---
 include/vbi/support/ctype.h         |   96 ----------
 include/vbi/support/macros.h        |   32 ---
 include/vbi/support/stdio.h         |  239 -----------------------
 include/vbi/support/stdlib.h        |  109 -----------
 include/vbi/support/string.h        |   76 --------
 include/vbi/support/sys/arch.h      |   49 -----
 include/vbi/support/sys/cpu_types.h |   82 --------
 include/vbi/support/sys/regs.h      |   35 ----
 include/vbi/support/sys/types.h     |  359 -----------------------------------
 9 files changed, 0 insertions(+), 1077 deletions(-)
 delete mode 100644 include/vbi/support/ctype.h
 delete mode 100644 include/vbi/support/macros.h
 delete mode 100644 include/vbi/support/stdio.h
 delete mode 100644 include/vbi/support/stdlib.h
 delete mode 100644 include/vbi/support/string.h
 delete mode 100644 include/vbi/support/sys/arch.h
 delete mode 100644 include/vbi/support/sys/cpu_types.h
 delete mode 100644 include/vbi/support/sys/regs.h
 delete mode 100644 include/vbi/support/sys/types.h

diff --git a/include/vbi/support/ctype.h b/include/vbi/support/ctype.h
deleted file mode 100644
index 7d31ff3..0000000
--- a/include/vbi/support/ctype.h
+++ /dev/null
@@ -1,96 +0,0 @@
-/* ctype.h - ANSI standard ctype functions header */
-
-/* Copyright 1992 Wind River Systems, Inc. */
-
-/*
-modification history
---------------------
-01b,09dec08,???  fix for inclusion by assembly files
-01a,29feb08,md   written based on VxWorks version 01k
-*/
-
-#ifndef __INCctypeh
-#define __INCctypeh
-
-#ifdef __cplusplus
-extern "C" {
-#endif
-
-#ifdef CONFIG_WRHV
-#include <linux/types.h>
-#include "sys/cpu_types.h"
-#else
-#include <sys/types.h>
-#endif /* CONFIG_WRHV */
-
-#ifndef _ASMLANGUAGE
-
-extern int      isalnum(int __c);
-extern int      isalpha(int __c);
-extern int      iscntrl(int __c);
-extern int      isdigit(int __c);
-extern int      isgraph(int __c);
-extern int      islower(int __c);
-extern int      isprint(int __c);
-extern int      ispunct(int __c);
-extern int      isspace(int __c);
-extern int      isupper(int __c);
-extern int      isxdigit(int __c);
-extern int      tolower(int __c);
-extern int      toupper(int __c);
-
-extern  const unsigned char *__ctype;
-extern  const unsigned char *__ctype_tolower;
-extern  const unsigned char *__ctype_toupper;
-
-#endif /* _ASMLANGUAGE */
-
-#define _C_UPPER         0x1
-#define _C_LOWER         0x2
-#define _C_NUMBER        0x4
-#define _C_WHITE_SPACE   0x8
-#define _C_PUNCT         0x10
-#define _C_CONTROL       0x20
-#define _C_HEX_NUMBER    0x40
-#define _C_B             0x80
-
-#define __isalpha(c)    (__ctype[(unsigned char)(c)] & (_C_UPPER | _C_LOWER))
-#define __isupper(c)    (__ctype[(unsigned char)(c)] & (_C_UPPER))
-#define __islower(c)    (__ctype[(unsigned char)(c)] & (_C_LOWER))
-#define __isdigit(c)    (__ctype[(unsigned char)(c)] & (_C_NUMBER))
-#define __isxdigit(c)   (__ctype[(unsigned char)(c)] & (_C_HEX_NUMBER))
-#define __isspace(c)    (__ctype[(unsigned char)(c)] & (_C_WHITE_SPACE | _C_CONTROL))
-#define __ispunct(c)    (__ctype[(unsigned char)(c)] & (_C_PUNCT))
-#define __isalnum(c)    (__ctype[(unsigned char)(c)] & (_C_UPPER | _C_LOWER | _C_NUMBER))
-#define __isprint(c)    (__ctype[(unsigned char)(c)] & (_C_PUNCT | _C_UPPER | _C_LOWER | \
-				       _C_WHITE_SPACE | _C_NUMBER))
-#define __isgraph(c)    (__ctype[(unsigned char)(c)] & (_C_PUNCT | _C_UPPER | _C_LOWER | \
-				       _C_NUMBER))
-#define __iscntrl(c)    (__ctype[(unsigned char)(c)] & (_C_CONTROL | _C_B))
-#define __toupper(c)    (__ctype_toupper[(unsigned char)(c)])
-#define __tolower(c)    (__ctype_tolower[(unsigned char)(c)])
-
-#ifndef __cplusplus
-#define isalpha(c)      __isalpha((unsigned char)(c))
-#define isupper(c)      __isupper((unsigned char)(c))
-#define islower(c)      __islower((unsigned char)(c))
-#define isdigit(c)      __isdigit((unsigned char)(c))
-#define isxdigit(c)     __isxdigit((unsigned char)(c))
-#define isspace(c)      __isspace((unsigned char)(c))
-#define ispunct(c)      __ispunct((unsigned char)(c))
-#define isalnum(c)      __isalnum((unsigned char)(c))
-#define isprint(c)      __isprint((unsigned char)(c))
-#define isgraph(c)      __isgraph((unsigned char)(c))
-#define iscntrl(c)      __iscntrl((unsigned char)(c))
-#define toupper(c)      __toupper((unsigned char)(c))
-#define tolower(c)      __tolower((unsigned char)(c))
-#endif
-
-#define isascii(c)	((unsigned char)(c) <= 0177)
-#define toascii(c)	((unsigned char)(c) & 0177)
-
-#ifdef __cplusplus
-}
-#endif
-
-#endif /* __INCctypeh */
diff --git a/include/vbi/support/macros.h b/include/vbi/support/macros.h
deleted file mode 100644
index 2dea8d3..0000000
--- a/include/vbi/support/macros.h
+++ /dev/null
@@ -1,32 +0,0 @@
-/* macros.h - razor miscellaneous C macro definitions */
-
-/* Copyright 2007 Wind River Systems, Inc. */
-
-/*
-modification history
---------------------
-01b,09apr09,mpk  hypervisor 64 bit changes to support mips cavium-octeon bsp
-01a,05oct07,md   written
-*/
-
-#ifndef __INCmacrosh
-#define __INCmacrosh
-
-/* macro for rounding values */
-
-#define ROUND_UP(x, align)	(((long) (x) + (align - 1)) & ~(align - 1))
-#define ROUND_DOWN(x, align)	((long)(x) & ~(align - 1))
-#define ALIGNED(x, align)	(((long)(x) & (align - 1)) == 0)
-
-/* macros for calcualating structure stuff */
-
-#define OFFSET(structure, member)	/* byte offset of member in structure*/\
-		((int) &(((structure *) 0) -> member))
-
-#define MEMBER_SIZE(structure, member)	/* size of a member of a structure */\
-		(sizeof (((structure *) 0) -> member))
-
-#define NELEMENTS(array)		/* number of elements in an array */ \
-		(sizeof (array) / sizeof ((array) [0]))
-
-#endif  /* __INCmacrosh */
diff --git a/include/vbi/support/stdio.h b/include/vbi/support/stdio.h
deleted file mode 100644
index 090c42d..0000000
--- a/include/vbi/support/stdio.h
+++ /dev/null
@@ -1,239 +0,0 @@
-/* stdio.h - stdio header file */
-
-/* Copyright 1992-2003 Wind River Systems, Inc. */
-/* Copyright (c) 1990 The Regents of the University of California */
-
-/*
-modification history
---------------------
-01b,27nov07,foo  add our prototypes
-01a,04oct07,md   written based on VxWorks 02o version
-*/
-
-#ifndef __INCstdioh
-#define __INCstdioh
-
-#ifdef __cplusplus
-extern "C" {
-#endif
-
-#include <sys/types.h>
-
-#if defined(__STDC__) || defined(__cplusplus)
-#include <stdarg.h>
-#endif /* __STDC__ */
-
-/* types */
-
-struct __sbuf 			/* stdio buffers */
-    {
-    uchar_t * _base;		/* base address of {std,unget,line} buffer */
-    int	      _size;		/* size of the buffer */
-    };
-
-typedef	struct __sFILE
-    {
-    uchar_t *		_p;		/* current position in (some) buffer */
-    int			_r;		/* read space left for getc() */
-    int			_w;		/* write space left for putc() */
-    short		_flags;		/* flags, below;this FILE is free if 0*/
-    short		_file;		/* fileno, if Unix descriptor, else -1*/
-    struct __sbuf	_bf;		/* buffer (at least 1 byte,if !NULL) */
-    int			_lbfsize;	/* 0 or -_bf._size, for inline putc */
-    struct __sbuf	_ub;		/* ungetc buffer */
-    uchar_t *		_up;		/* old _p if _p is doing ungetc data */
-    int			_ur;		/* old _r if _r counting ungetc data */
-    uchar_t		_ubuf[3];	/* guarantee an ungetc() buffer */
-    uchar_t		_nbuf[1];	/* guarantee a getc() buffer */
-    struct __sbuf	_lb;		/* buffer for fgetline() */
-    int			_blksize;	/* stat.st_blksize (may be!=_bf._size)*/
-    int			_offset;	/* current lseek offset */
-    int			taskId;		/* task that owns this file pointer */
-    } FILE;
-
-/* __SRD and __SWR are never simultaneously asserted */
-
-#define	__SLBF		0x0001		/* line buffered */
-#define	__SNBF		0x0002		/* unbuffered */
-#define	__SRD		0x0004		/* OK to read */
-#define	__SWR		0x0008		/* OK to write */
-#define	__SWRNBF	(__SWR|__SNBF)	/* write unbuffered */
-#define	__SRW		0x0010		/* open for reading & writing */
-#define	__SEOF		0x0020		/* found EOF */
-#define	__SERR		0x0040		/* found error */
-#define	__SMBF		0x0080		/* _buf is from malloc */
-#define	__SAPP		0x0100		/* fdopen()ed in append mode */
-#define	__SSTR		0x0200		/* this is an sprintf/snprintf string */
-#define	__SOPT		0x0400		/* do fseek() optimisation */
-#define	__SNPT		0x0800		/* do not do fseek() optimisation */
-#define	__SOFF		0x1000		/* set iff _offset is in fact correct */
-#define	__SMOD		0x2000		/* true => fgetline modified _p text */
-
-#define	_IOFBF	0		/* setvbuf should set fully buffered */
-#define	_IOLBF	1		/* setvbuf should set line buffered */
-#define	_IONBF	2		/* setvbuf should set unbuffered */
-
-#define	BUFSIZ	_PARM_BUFSIZ	/* size of buffer used by setbuf */
-#define	BUFSIZE	BUFSIZ
-
-#ifndef EOF
-#define	EOF	(-1)
-#endif  /* EOF */
-
-#define FOPEN_MAX       _PARM_FOPEN_MAX
-#define FILENAME_MAX    _PARM_FILENAME_MAX
-
-#define	L_tmpnam	_PARM_L_tmpnam
-#define	TMP_MAX		_PARM_TMP_MAX
-
-#ifndef SEEK_SET
-#define	SEEK_SET	0	/* set file offset to offset */
-#endif
-
-#ifndef SEEK_CUR
-#define	SEEK_CUR	1	/* set file offset to current plus offset */
-#endif
-
-#ifndef SEEK_END
-#define	SEEK_END	2	/* set file offset to EOF plus offset */
-#endif
-
-#if _EXTENSION_POSIX_1003
-
-#define	L_cuserid	_PARM_L_cuserid
-#define	L_ctermid	_PARM_L_ctermid
-#define STREAM_MAX      FOPEN_MAX
-
-#endif /* _EXTENSION_POSIX_1003 */
-
-extern void	clearerr (FILE *);
-extern int	fclose (FILE *);
-extern int	feof (FILE *);
-extern int	ferror (FILE *);
-extern int	fflush (FILE *);
-extern int	fgetc (FILE *);
-extern int	fgetpos (FILE *, fpos_t *);
-extern char *	fgets (char *, size_t, FILE *);
-extern FILE *	fopen (const char *, const char *);
-extern int	fprintf (FILE *, const char *, ...);
-extern int	fputc (int, FILE *);
-extern int	fputs (const char *, FILE *);
-extern int	fread (void *, size_t, size_t, FILE *);
-extern FILE *	freopen (const char *, const char *, FILE *);
-extern int	fscanf (FILE *, const char *, ...);
-extern int	fseek (FILE *, long, int);
-extern int	fsetpos (FILE *, const fpos_t *);
-extern long	ftell (FILE *);
-extern int	fwrite (const void *, size_t, size_t, FILE *);
-extern int	getc (FILE *);
-extern int	getchar (void);
-extern char *	gets (char *);
-extern void	perror (const char *);
-extern int	printf (const char *, ...);
-extern int	putc (int, FILE *);
-extern int	putchar (int);
-extern int	puts (const char *);
-extern int	remove (const char *);
-extern int	rename  (const char *, const char *);
-extern void	rewind (FILE *);
-extern int	scanf (const char *, ...);
-extern void	setbuf (FILE *, char *);
-extern int	setvbuf (FILE *, char *, int, size_t);
-extern int	sprintf (char *, const char *, ...);
-extern int 	snprintf (char *, size_t, const char *, ...);
-extern int	sscanf (const char *, const char *, ...);
-extern FILE *	tmpfile (void);
-extern char *	tmpnam (char *);
-extern int	ungetc (int, FILE *);
-extern int	vfprintf (FILE *, const char *, va_list);
-extern int	vprintf (const char *, va_list);
-extern int	vsprintf (char *, const char *, va_list);
-extern int	vsnprintf (char *, size_t, const char *, va_list); 
-extern int	__srget (FILE *);		/* for macro definition below */
-extern int	__swbuf (int, FILE *);		/* for macro definition below */
-
-/* _EXTENSION_POSIX_1003 */
-
-extern FILE *	fdopen (int, const char *);
-extern int	fileno (FILE *);
-
-/* _EXTENSION_POSIX_1003 */
-
-/* WRS stdio functions declarations */
-
-#if _EXTENSION_WRS			/* undef for ANSI */
-
-extern int	fdprintf (int fd, const char *fmt, ...);
-extern int	vfdprintf (int fd, const char *fmt, va_list ap);
-extern int	printErr (const char *fmt, ...);
-extern int	getw (FILE *);
-extern int	putw (int, FILE *);
-extern void	setbuffer (FILE *, char *, int);
-extern int	setlinebuf (FILE *);
-extern FILE *   stdioFp (int std);
-extern STATUS   stdioShow (FILE * fp, int level);
-extern STATUS   stdioShowInit (void);
-extern STATUS   stdioInit (void);
-
-#endif /* _EXTENSION_WRS */
-
-
-/* Definition of stdin, stdout, stderr */
-
-extern FILE **	__stdin(void);		/* returns current task's stdin */
-extern FILE **	__stdout(void);		/* returns current task's stdout */
-extern FILE **	__stderr(void);		/* returns current task's stderr */
-
-#define stdin	(*__stdin())
-#define stdout	(*__stdout())
-#define stderr	(*__stderr())
-
-
-/* definition of stdio macros */
-
-#define	__sfileno(p)	((HANDLE_VERIFY(p, handleTypeFile) != OK) ? (-1) : \
-			((p)->_file))
-
-#define	__sgetc(p)	((HANDLE_VERIFY(p, handleTypeFile) != OK) ? (EOF) : \
-			((--(p)->_r < 0) ? (__srget(p)) : ((int)(*(p)->_p++))))
-
-#define	__sputc(c, p) 	((HANDLE_VERIFY(p, handleTypeFile) != OK) ? (EOF) : \
-			 (--(p)->_w < 0 ?				\
-			     (p)->_w >= (p)->_lbfsize ?			\
-				 (*(p)->_p = (c)), *(p)->_p != '\n' ?	\
-				     (int)*(p)->_p++ :			\
-				     __swbuf('\n', p) :			\
-				 __swbuf((int)(c), p) : 		\
-			     (*(p)->_p = (c), (int)*(p)->_p++)))
-
-#define	__sfeof(p)	((HANDLE_VERIFY(p, handleTypeFile) != OK) ? \
-			 (FALSE) : (((p)->_flags & __SEOF) != 0))
-
-#define	__sferror(p)	((HANDLE_VERIFY(p, handleTypeFile) != OK) ? \
-			 (FALSE) : (((p)->_flags & __SERR) != 0))
-
-#define	__sclearerr(p)	((void)((HANDLE_VERIFY(p, handleTypeFile) != OK) ? \
-				(0) : (((p)->_flags &= ~(__SERR|__SEOF)))))
-
-
-#define	getchar()	__sgetc(stdin)
-#define	getc(p)		__sgetc(p)
-#define	putchar(c)	(__sputc(c, (stdout)))
-#define	putc(c,p)	__sputc(c, p)
-#define	feof(p)		__sfeof(p)
-#define	ferror(p)	__sferror(p)
-#define	clearerr(p)	__sclearerr(p)
-
-#if _EXTENSION_POSIX_1003		/* undef for ANSI */
-#define	fileno(p)	__sfileno(p)
-#endif
-
-/* Razor extensions */
-extern int oprintf (FUNCPTR, int, const char *, ...);
-extern int kprintf (const char *, ...);
-
-#ifdef __cplusplus
-}
-#endif
-
-#endif /* __INCstdioh */
diff --git a/include/vbi/support/stdlib.h b/include/vbi/support/stdlib.h
deleted file mode 100644
index 2fdfc65..0000000
--- a/include/vbi/support/stdlib.h
+++ /dev/null
@@ -1,109 +0,0 @@
-/* stdlib.h - razor C standard library definitions */
-
-/* Copyright 2007 Wind River Systems, Inc. */
-
-/*
-modification history
---------------------
-01a,04oct07,md   written
-*/
-
-#ifndef __INCstdlibh
-#define __INCstdlibh
-
-#ifdef CONFIG_WRHV
-#include <linux/types.h>
-#else
-#include <sys/types.h>
-#endif /* CONFIG_WRHV */
-
-#ifdef __cplusplus
-extern "C" {
-#endif /* __cplusplus */
-
-
-#ifndef NULL
-#define NULL	0
-#endif
-
-#define EXIT_FAILURE 1
-#define EXIT_SUCCESS 0
-#define RAND_MAX 32767
-#define MB_CUR_MAX 1
-
-#define	OK	0
-#define	ERROR	(-1)
-
-#undef	TRUE
-#define	TRUE	1
-#undef	FALSE
-#define	FALSE	0
-
-#define EOS	'\0'	/* C string terminator */
-
-#ifndef	_ASMLANGUAGE
-
-extern double atof(const char *);
-extern int atoi(const char *);
-extern long atol(const char *);
-extern double strtod(const char *, char **);
-extern long strtol(const char *, char **, int);
-extern unsigned long strtoul(const char *, char **, int);
-extern int rand(void);
-extern int rand_r(unsigned int *);
-extern void srand(unsigned int);
-extern void *calloc(size_t, size_t);
-extern void free(void *);
-extern void *malloc(size_t);
-extern void *realloc(void *, size_t);
-#ifdef __VEC__
-extern void *vec_calloc(size_t, size_t);
-extern void vec_free(void *);
-extern void *vec_malloc(size_t);
-extern void *vec_realloc(void *, size_t);
-#endif
-
-extern void abort(void);
-extern int atexit(void (*)(void));
-extern void exit(int);
-extern char *getenv(const char *);
-extern int system(const char *);
-
-extern void *bsearch(const void *, const void *, size_t, size_t, int (*)(const void *, const void *));
-extern void qsort(void *, size_t, size_t, int (*)(const void *, const void *));
-
-extern int abs(int);
-extern div_t div(int, int);
-extern long labs(long);
-extern ldiv_t ldiv(long, long);
-extern int mblen(const char *, size_t);
-extern int mbtowc(wchar_t *, const char *, size_t);
-extern int wctomb(char *, wchar_t);
-extern size_t mbstowcs(wchar_t *, const char *, size_t);
-extern size_t wcstombs(char *, const wchar_t *, size_t);
-
-#if !defined(__STRICT_ANSI__) || defined(__cplusplus)
-extern	long		a64l(const char *);
-extern	char *		l64a(long l);
-extern	double		drand48(void);
-extern	double		erand48(unsigned short *);
-extern	long		irand48(unsigned short);
-extern	long		jrand48(unsigned short *);
-extern	long		krand48(unsigned short *, unsigned short);
-extern	void		lcong48(unsigned short *);
-extern	long		lrand48(void);
-extern	long		mrand48(void);
-extern	long		nrand48(unsigned short *);
-extern	void		srand48(long);
-extern	unsigned short *seed48(unsigned short *);
-extern	char *		getpass(const char *);
-extern	int		putenv(char *);
-#endif /* __STRICT_ANSI__ */
-
-#endif /* _ASMLANGUAGE */
-
-#ifdef __cplusplus
-}
-#endif /* __cplusplus */
-
-#endif  /* __INCstdlibh */
diff --git a/include/vbi/support/string.h b/include/vbi/support/string.h
deleted file mode 100644
index cfd372b..0000000
--- a/include/vbi/support/string.h
+++ /dev/null
@@ -1,76 +0,0 @@
-/* string.h - string library header file */
-
-/* Copyright 2007-2009 Wind River Systems, Inc. */
-
-/*
-modification history
---------------------
-01b,21jul09,md   use size_t in prototypes
-01a,03oct07,md   written.
-*/
-
-#ifndef __INCstringh
-#define __INCstringh
-
-#ifdef __cplusplus
-extern "C" {
-#endif
-
-#include "sys/types.h"
-
-#define	_EXTENSION_WRS	1
-
-extern void *	memchr (const void *__s, int __c, size_t __n);
-extern int 	memcmp (const void *__s1, const void *__s2, size_t __n);
-extern void *	memcpy (void *__s1, const void *__s2, size_t __n);
-extern void *	memmove (void *__s1, const void *__s2, size_t __n);
-extern void *	memset (void *__s, int __c, size_t __n);
-
-extern char *	strcat (char *__s1, const char *__s2);
-extern char *	strchr (const char *__s, int __c);
-extern int 	strcmp (const char *__s1, const char *__s2);
-extern int 	strcoll (const char *__s1, const char *__s2);
-extern char *	strcpy (char *__s1, const char *__s2);
-extern size_t 	strcspn (const char *__s1, const char *__s2);
-extern char *   strdup (const char * str);
-extern size_t 	strlen (const char *__s);
-extern char *	strncat (char *__s1, const char *__s2, size_t __n);
-extern int 	strncmp (const char *__s1, const char *__s2, size_t __n);
-extern char *	strncpy (char *__s1, const char *__s2, size_t __n);
-extern char *	strpbrk (const char *__s1, const char *__s2);
-extern char *	strrchr (const char *__s, int __c);
-extern size_t 	strspn (const char *__s1, const char *__s2);
-extern char *	strstr (const char *__s1, const char *__s2);
-extern char *	strtok (char *__s, const char *__sep);
-extern size_t 	strxfrm (char *__s1, const char *__s2, size_t __n);
-extern char *	strerror(int __errcode);
-
-#if _EXTENSION_POSIX_REENTRANT		/* undef for ANSI */
-extern char *	strtok_r (char *__s, const char *__sep, char **__ppLast);
-#endif
-
-#if _EXTENSION_WRS		 	/* undef for ANSI */
-extern int	strerror_r (int __errcode, char *__buf);
-extern void 	bcopy (const char *source, char *dest, size_t nbytes);
-extern void 	bcopyBytes (char *source, char *dest, size_t nbytes);
-extern void 	bcopyWords (char *source, char *dest, size_t nwords);
-extern void 	bcopyLongs (char *source, char *dest, size_t nlongs);
-extern void 	bfill (char *buf, size_t nbytes, int ch);
-extern void 	bfillBytes (char *buf, size_t nbytes, int ch);
-extern void 	bzero (char *buffer, size_t nbytes);
-extern int 	bcmp (char *buf1, char *buf2, size_t nbytes);
-extern void 	binvert (char *buf, size_t nbytes);
-extern void 	bswap (char *buf1, char *buf2, size_t nbytes);
-extern void 	uswab (char *source, char *destination, size_t nbytes);
-extern void 	swab (char *source, char *dest, size_t nbytes);
-extern char *	index (const char *s, int c);
-extern char *	rindex (const char *s, int c);
-extern size_t	strlcat (char *dst, const char *src, size_t siz);
-extern size_t   strlcpy (char *dst, const char *src, size_t siz);
-#endif
-
-#ifdef __cplusplus
-}
-#endif
-
-#endif /* __INCstringh */
diff --git a/include/vbi/support/sys/arch.h b/include/vbi/support/sys/arch.h
deleted file mode 100644
index 9223b9e..0000000
--- a/include/vbi/support/sys/arch.h
+++ /dev/null
@@ -1,49 +0,0 @@
-/* sys/arch.h - hypervisor architecture specific definitions */
-
-/* Copyright 2007-2009 Wind River Systems, Inc. */
-
-/*
-modification history
---------------------
-01g,06aug09,md   include wrhv arch specific file
-01f,15jul09,dtr  Add CPU PPCE500MC support.
-01e,20may09,md   add endian order type defines
-01d,17mar09,mpk  OCTEON port from drc
-01c,16jul08,gws  adding ARM cpu
-01b,04oct07,md   add cpu defines
-01a,06jul07,md   written.
-*/
-
-#ifndef __INCsysArchh
-#define __INCsysArchh
-
-#include <sys/cpu_types.h>
-
-/* endian types selectable by the arch specific header files */
-
-#define	__LITTLE_ENDIAN	1234	/* least-significant byte first		*/
-#define	__BIG_ENDIAN	4321	/* most-significant byte first		*/
-#define	__PDP_ENDIAN	3412	/* LSB first in word, MSW first in long	*/
-
-/* include the appropriate arch specific header file */
-
-#if (CPU == MIPSI64R2)
-# include <sys/mips64/arch.h>
-#elif (CPU == PPC85XX)
-# include <sys/ppc/arch.h>
-#elif (CPU == PPCE500MC)
-# include <sys/ppc/arch.h>
-#elif (CPU == PENTIUM)
-# include <sys/x86/arch.h>
-# include <sys/x86/wrhvArch.h>
-#elif (CPU == ARM1136)
-# include <sys/arm1136/arch.h>
-#endif
-
-/* ensure the arch has specified a byte ordering */
-
-#if !defined(__BYTE_ORDER)
-#error: Architecture has not defined a byte order!
-#endif
-
-#endif /* __INCsysArchh */
diff --git a/include/vbi/support/sys/cpu_types.h b/include/vbi/support/sys/cpu_types.h
deleted file mode 100644
index 5e4d4c9..0000000
--- a/include/vbi/support/sys/cpu_types.h
+++ /dev/null
@@ -1,82 +0,0 @@
-/* sys/cpu_types.h - hypervisor cpu architecture types */
-
-/* Copyright 2007-2009 Wind River Systems, Inc. */
-
-/*
-modification history
---------------------
-01g,15jul09,dtr  Add CPU PPCE500MC support.
-01f,04aug09,md   add _RType for each cpu arch
-01e,17mar09,mpk  OCTEON port from drc
-01d,16jul08,gws  adding ARM cpu
-01c,29feb08,md   move cpu specific stuff from types.h
-01b,09oct07,md   don't define if we're included by vxWorks
-01a,04oct07,md   written.
-*/
-
-#ifndef __INCsysCpuTypesh
-#define __INCsysCpuTypesh
-
-/* define the cpu types that we currently support */
-
-#define	PPC85XX		1
-#define	PENTIUM		2
-#define	ARM1136		3
-#define MIPSI64R2       4
-#define	PPCE500MC	5
-
-#ifndef _ASMLANGUAGE
-
-/* MIPS64R2 specific types */
-
-#if (CPU == MIPSI64R2)
-
-#define CPU_MIPS_32BIT ((CPU==_VX_MIPS32) || \
-			    (CPU==MIPSI2) || \
-			    (CPU==MIPSI32) || \
-			    (CPU==MIPSI32R2))
-
-#define CPU_MIPS_64BIT ((CPU==MIPS64) || \
-			    (CPU==MIPSI3) || \
-			    (CPU==MIPSI64) || \
-			    (CPU==MIPSI64R2))
-
-typedef unsigned long	INSTR;		/* 32 bit word-aligned instructions */
-
-#if	(CPU_MIPS_32BIT)
-typedef unsigned int	_RType;		/* registers are 32 bits */
-#elif	(CPU_MIPS_64BIT)
-typedef unsigned long	_RType;		/* registers are 64 bits */
-#else	/* CPU_MIPSxxBIT */
-#error "Invalid CPU value"
-#endif
-
-#endif /* CPU == MIPSI64R2 */
-
-/* PowerPC specific types */
-
-#if (CPU == PPC85XX) || (CPU == PPCE500MC)
-typedef unsigned long	INSTR;		/* 32 bit word-aligned instructions */
-typedef unsigned int	_RType;		/* register type */
-#endif
-
-/* Pentium specific types */
-
-#if (CPU == PENTIUM)
-typedef unsigned char	INSTR;		/* char instructions */
-					/* register type */
-#if defined (LP64)
-typedef unsigned long	_RType;
-#else
-typedef unsigned int	_RType;
-#endif
-#endif /* CPU == PENTIUM */
-
-#if (CPU == ARM1136)
-typedef unsigned long	INSTR;		/* 32 bit word-aligned instructions */
-typedef unsigned int	_RType;		/* register type */
-#endif
-
-#endif /* _ASMLANGUAGE */
-
-#endif /* __INCsysCpuTypesh */
diff --git a/include/vbi/support/sys/regs.h b/include/vbi/support/sys/regs.h
deleted file mode 100644
index 98ac7aa..0000000
--- a/include/vbi/support/sys/regs.h
+++ /dev/null
@@ -1,35 +0,0 @@
-/* sys/regs.h - hypervisor context registers */
-
-/* Copyright 2007 Wind River Systems, Inc. */
-
-/*
-modification history
---------------------
-01f,15jul09,dtr  Support CPU PPCE500MC.
-01e,17mar09,mpk  OCTEON port from drc
-01d,16jul08,gws  adding ARM cpu
-01c,04oct07,md   use local PowerPC register definitions
-01b,16jul07,md   use local x86 register definitions
-01a,06jul07,md   written.
-*/
-
-#ifndef __INCsysRegsh
-#define __INCsysRegsh
-
-#include <sys/cpu_types.h>
-
-/* include the appropriate arch specific header file */
-
-#if (CPU == MIPSI64R2)
-# include <sys/mips64/regs.h>
-#elif (CPU == PPC85XX)
-# include <sys/ppc/regs.h>
-#elif (CPU == PPCE500MC)
-# include <sys/ppc/regs.h>
-#elif (CPU == PENTIUM)
-# include <sys/x86/regs.h>
-#elif (CPU == ARM1136)
-# include <sys/arm1136/regs.h>
-#endif
-
-#endif /* __INCsysRegsh */
diff --git a/include/vbi/support/sys/types.h b/include/vbi/support/sys/types.h
deleted file mode 100644
index f316eea..0000000
--- a/include/vbi/support/sys/types.h
+++ /dev/null
@@ -1,359 +0,0 @@
-/* sys/types.h - hypervisor data types */
-
-/* Copyright 2007 Wind River Systems, Inc. */
-
-/*
-modification history
---------------------
-01i,09sep09,mmi  remove vb_t type but instead use the definition in vbiTypes.h
-01h,29jun09,mmi  add vb_t
-01g,18may09,mpk  n32 changes
-01f,11may09,md   add hypervisor specific types
-01e,03apr09,mpk  build mips64 hypervisor using mips-wrs-linux-gnu compiler
-01d,18mar09,md   add pointer types
-01c,29feb08,md   add math type defines
-01b,09oct07,md   fix types when included from gcc
-01a,03oct07,md   written.
-*/
-
-#ifndef __INCsysTypesh
-#define __INCsysTypesh
-
-#ifdef CONFIG_WRHV
-#include "cpu_types.h"
-#else
-#include <sys/cpu_types.h>
-#endif
-
-/* don't define the types if we're building an assembler file */
-
-#if !defined(_ASMLANGUAGE)
-
-/* standard variable types */
-
-#if defined(__STDC__) || defined(__cplusplus)
- typedef signed char int8_t;
-#else
- typedef char int8_t;
-#endif
-typedef unsigned char uint8_t;
-typedef short int16_t;
-typedef unsigned short uint16_t;
-
-#if defined (LP64)
-typedef int 			fpos_t;
-typedef int 			wchar_t;
-typedef long unsigned int 	size_t;
-typedef int 			int32_t;
-typedef unsigned int 		uint32_t;
-typedef long 			int64_t;
-typedef unsigned long		uint64_t;
-#else
-typedef long 			fpos_t;
-typedef unsigned short   	wchar_t;
-typedef unsigned int 		size_t;
-typedef long 			int32_t;
-typedef unsigned long		uint32_t;
-typedef long long 		int64_t;
-typedef unsigned long long 	uint64_t;
-#endif
-
-typedef long 			intptr_t;
-typedef unsigned long   	uintptr_t;
-
-typedef struct {
-    int quot;
-    int rem;
-} div_t;
-
-typedef struct {
-    long quot;
-    long rem;
-} ldiv_t;
-
-typedef unsigned char u_char;
-typedef unsigned short u_short;
-typedef unsigned int u_int;
-typedef unsigned long u_long;
-
-typedef unsigned char uchar_t;
-typedef unsigned short ushort_t;
-typedef unsigned int uint_t;
-typedef unsigned long ulong_t;
-
-/* hypervisor specific types */
-
-typedef uint32_t ctx_t;		/* context id handle */
-typedef uint64_t timeout_t;	/* timeout specification */
-typedef int32_t  status_t;	/* operation/function status */
-typedef uint32_t clock_t;	/* clock rate specification */
-
-/* file system related types */
-
-#if defined (LP64)
-typedef unsigned long  dev_t;
-typedef unsigned long  ino_t;
-typedef signed   int   mode_t;
-typedef unsigned long  nlink_t;
-typedef unsigned short uid_t;
-typedef unsigned short gid_t;
-typedef unsigned long  time_t;
-typedef signed   long  blksize_t;
-typedef unsigned long  blkcnt_t;
-typedef signed   long  off_t;
-typedef signed   long  daddr_t;
-#else
-typedef unsigned int dev_t;
-typedef unsigned int ino_t;
-typedef unsigned int mode_t;
-typedef unsigned int nlink_t;
-typedef unsigned int uid_t;
-typedef unsigned int gid_t;
-typedef unsigned int time_t;
-typedef unsigned int blksize_t;
-typedef unsigned int blkcnt_t;
-typedef unsigned int off_t;
-typedef unsigned int daddr_t;
-#endif
-
-#ifdef LP64
-#define LONG_FMT "0x%llx"
-#else
-#define LONG_FMT "0x%lx"
-#endif
-
-/* VxWorks compatibility for importing source code */
-
-typedef unsigned int	UINT;
-typedef unsigned char	UINT8;
-typedef unsigned short	UINT16;
-typedef unsigned int	UINT32;
-typedef unsigned long	ULONG;
-typedef unsigned short	USHORT;
-typedef unsigned char	UCHAR;
-typedef signed char	INT8;
-typedef signed short	INT16;
-typedef signed int	INT32;
-typedef signed long	LONG;
-typedef signed short	SHORT;
-typedef int		BOOL;
-typedef int		STATUS;
-#ifdef __cplusplus
-typedef void (*VOIDFUNCPTR) (...);
-typedef int (*FUNCPTR) (...);
-#else
-typedef void (*VOIDFUNCPTR) ();
-typedef int (*FUNCPTR) ();
-#endif
-
-#define	IMPORT	extern
-#define	LOCAL	static
-#define	FAST
-
-#define _WRS_INLINE	static __inline__
-
-#endif /* _ASMLANGUAGE */
-
-/*
- * stuff for limits.h
- */
-#ifndef _ARCH_MB_LEN_MAX
-#define _ARCH_MB_LEN_MAX		1
-#endif
-
-#ifndef _ARCH_CHAR_BIT
-#define _ARCH_CHAR_BIT			8
-#endif
-
-#ifndef _ARCH_CHAR_MAX
-#define _ARCH_CHAR_MAX			127
-#endif
-
-#ifndef _ARCH_CHAR_MIN
-#define _ARCH_CHAR_MIN			(-127-1)
-#endif
-
-#ifndef _ARCH_SHRT_MAX
-#define _ARCH_SHRT_MAX			32767
-#endif
-
-#ifndef _ARCH_SHRT_MIN
-#define _ARCH_SHRT_MIN			(-32767-1)
-#endif
-
-#ifndef _ARCH_INT_MAX
-#define _ARCH_INT_MAX			2147483647
-#endif
-
-#ifndef _ARCH_INT_MIN
-#define _ARCH_INT_MIN			(-2147483647-1)
-#endif
-
-#ifndef _ARCH_LONG_MAX
-#define _ARCH_LONG_MAX			2147483647
-#endif
-
-#ifndef _ARCH_LONG_MIN
-#define _ARCH_LONG_MIN			(-2147483647-1)
-#endif
-
-#ifndef _ARCH_SCHAR_MAX
-#define _ARCH_SCHAR_MAX			127
-#endif
-
-#ifndef _ARCH_SCHAR_MIN
-#define _ARCH_SCHAR_MIN			(-127-1)
-#endif
-
-#ifndef _ARCH_UCHAR_MAX
-#define _ARCH_UCHAR_MAX			255
-#endif
-
-#ifndef _ARCH_USHRT_MAX
-#define _ARCH_USHRT_MAX			65535
-#endif
-
-#ifndef _ARCH_UINT_MAX
-#ifdef __STDC__
-#define _ARCH_UINT_MAX			4294967295u
-#else
-#define _ARCH_UINT_MAX			4294967295
-#endif
-#endif
-
-#ifndef _ARCH_ULONG_MAX
-#ifdef __STDC__
-#define _ARCH_ULONG_MAX			4294967295u
-#else
-#define _ARCH_ULONG_MAX			4294967295
-#endif
-#endif
-
-/*
- * stuff for float.h
- */
-#ifndef _ARCH_FLT_RADIX
-#define _ARCH_FLT_RADIX			2
-#endif
-
-#ifndef _ARCH_FLT_MANT_DIG
-#define _ARCH_FLT_MANT_DIG		24
-#endif
-
-#ifndef _ARCH_FLT_DIG
-#define _ARCH_FLT_DIG			7	/* not correct in ANSI spec.s */
-#endif
-
-#ifndef _ARCH_FLT_ROUNDS
-#define _ARCH_FLT_ROUNDS		1
-#endif
-
-#ifndef _ARCH_FLT_EPSILON
-#define _ARCH_FLT_EPSILON		1.19209290e-07F
-#endif
-
-#ifndef _ARCH_FLT_MIN_EXP
-#define _ARCH_FLT_MIN_EXP		(-125)
-#endif
-
-#ifndef _ARCH_FLT_MIN
-#define _ARCH_FLT_MIN			1.17549435e-38F
-#endif
-
-#ifndef _ARCH_FLT_MIN_10_EXP
-#define _ARCH_FLT_MIN_10_EXP		(-37)
-#endif
-
-#ifndef _ARCH_FLT_MAX_EXP
-#define _ARCH_FLT_MAX_EXP		128
-#endif
-
-#ifndef _ARCH_FLT_MAX
-#define _ARCH_FLT_MAX			3.40282347e+38F
-#endif
-
-#ifndef _ARCH_FLT_MAX_10_EXP
-#define _ARCH_FLT_MAX_10_EXP		38
-#endif
-
-#ifndef _ARCH_DBL_MANT_DIG
-#define _ARCH_DBL_MANT_DIG		53
-#endif
-
-#ifndef _ARCH_DBL_DIG
-#define _ARCH_DBL_DIG			15
-#endif
-
-#ifndef _ARCH_DBL_EPSILON
-#define _ARCH_DBL_EPSILON		2.2204460492503131e-16
-#endif
-
-#ifndef _ARCH_DBL_MIN_EXP
-#define _ARCH_DBL_MIN_EXP		(-1021)
-#endif
-
-#ifndef _ARCH_DBL_MIN
-#define _ARCH_DBL_MIN			2.2250738585072014e-308
-#endif
-
-#ifndef _ARCH_DBL_MIN_10_EXP
-#define _ARCH_DBL_MIN_10_EXP		(-307)
-#endif
-
-#ifndef _ARCH_DBL_MAX_EXP
-#define _ARCH_DBL_MAX_EXP		1024
-#endif
-
-#ifndef _ARCH_DBL_MAX
-#define _ARCH_DBL_MAX			1.7976931348623157e+308
-#endif
-
-#ifndef _ARCH_DBL_MAX_10_EXP
-#define _ARCH_DBL_MAX_10_EXP		308
-#endif
-
-#ifndef _ARCH_LDBL_MANT_DIG
-#define _ARCH_LDBL_MANT_DIG		53
-#endif
-
-#ifndef _ARCH_LDBL_DIG
-#define _ARCH_LDBL_DIG			15
-#endif
-
-#ifndef _ARCH_LDBL_EPSILON
-#define _ARCH_LDBL_EPSILON		2.2204460492503131e-16L
-#endif
-
-#ifndef _ARCH_LDBL_MIN_EXP
-#define _ARCH_LDBL_MIN_EXP		(-1021)
-#endif
-
-#ifndef _ARCH_LDBL_MIN
-#define _ARCH_LDBL_MIN			2.2250738585072014e-308L
-#endif
-
-#ifndef _ARCH_LDBL_MIN_10_EXP
-#define _ARCH_LDBL_MIN_10_EXP		(-307)
-#endif
-
-#ifndef _ARCH_LDBL_MAX_EXP
-#define _ARCH_LDBL_MAX_EXP		1024
-#endif
-
-#ifndef _ARCH_LDBL_MAX
-#define _ARCH_LDBL_MAX			1.7976931348623157e+308L
-#endif
-
-#ifndef _ARCH_LDBL_MAX_10_EXP
-#define _ARCH_LDBL_MAX_10_EXP		308
-#endif
-
-/*
- * stuff for math.h
- */
-#ifndef _ARCH_HUGH_VAL
-#define _ARCH_HUGH_VAL			_ARCH_DBL_MAX
-#endif
-
-
-#endif /* __INCsysTypesh */
-- 
1.6.3.3

