From 8f2e9652aee3b31166dc54e195c2c5e93fcd977d Mon Sep 17 00:00:00 2001
From: Paul Gortmaker <paul.gortmaker@windriver.com>
Date: Thu, 29 Oct 2009 19:06:05 -0400
Subject: [PATCH 05/26] vbi: cosmetic clean up of private.h

Delete the useless defines by simply placing them in the
relevant lines of code directly.

Signed-off-by: Paul Gortmaker <paul.gortmaker@windriver.com>
---
 include/vbi/private.h |   15 ++++++---------
 1 files changed, 6 insertions(+), 9 deletions(-)

diff --git a/include/vbi/private.h b/include/vbi/private.h
index c609b34..0b1a719 100644
--- a/include/vbi/private.h
+++ b/include/vbi/private.h
@@ -17,9 +17,6 @@
 #ifndef _VBI_PRIVATE_H
 #define _VBI_PRIVATE_H
 
-#define _WRS_INLINE     static __inline__
-#define EOS     '\0'    /* C string terminator */
-#define LINKAGE asmlinkage
 #include <linux/types.h>
 #include <linux/linkage.h>
 #include <vbi/syscalls.h>
@@ -28,14 +25,14 @@
 
 #ifndef _ASMLANGUAGE
 
-_WRS_INLINE vbiStatus_t _vbiStrncmp(const char * s1, const char * s2, size_t n)
+static __inline__ vbiStatus_t _vbiStrncmp(const char * s1, const char * s2, size_t n)
 {                                        
 	if (n == 0)
 		return (0);
  
 	while (*s1++ == *s2++)
 	{
- 		if ((s1 [-1] == EOS) || (--n == 0))
+ 		if ((s1 [-1] == '\0') || (--n == 0))
  		    return (0);
 	}
  
@@ -43,17 +40,17 @@ _WRS_INLINE vbiStatus_t _vbiStrncmp(const char * s1, const char * s2, size_t n)
 }
 
 /* private ioapic operator */
-extern LINKAGE vbiStatus_t vbiIoapicOp (uint32_t ioctl, vbiIrq_t irq, uint32_t filter, vbiVb_t vbId);
+extern asmlinkage vbiStatus_t vbiIoapicOp (uint32_t ioctl, vbiIrq_t irq, uint32_t filter, vbiVb_t vbId);
 
 /* private interrupt redirection operator */
-extern LINKAGE vbiStatus_t vbiVcoreIntRed_op(vbiIrq_t irq, vbiCore_t CoreId);
+extern asmlinkage vbiStatus_t vbiVcoreIntRed_op(vbiIrq_t irq, vbiCore_t CoreId);
 
 /* Name service private operator */
-extern LINKAGE vbiStatus_t vbiNsOp(uint32_t op, char* name, uint32_t rev, 
+extern asmlinkage vbiStatus_t vbiNsOp(uint32_t op, char* name, uint32_t rev, 
 			VBI_NS_HANDLE *handle);
 
 /* Message receive private operator */
-extern LINKAGE vbiStatus_t vbiReceiveOp(void *rmsg, uint32_t rlen, 
+extern asmlinkage vbiStatus_t vbiReceiveOp(void *rmsg, uint32_t rlen, 
 			VBI_MSG_INFO *info, VBI_MSG_CTL *ctl);                                   
 
 #endif
-- 
1.6.5.2

