From e4a9bf4bab495ec640db894549979e8120bc8e01 Mon Sep 17 00:00:00 2001
From: Paul Gortmaker <paul.gortmaker@windriver.com>
Date: Thu, 3 Dec 2009 13:58:16 -0500
Subject: [PATCH 35/43] VBI_MEM_CTL --> struct vbi_mem_ctl

Signed-off-by: Paul Gortmaker <paul.gortmaker@windriver.com>
---
 include/vbi/syscall.h |    4 ++--
 1 files changed, 2 insertions(+), 2 deletions(-)

diff --git a/include/vbi/syscall.h b/include/vbi/syscall.h
index 0507ba5..00b280d 100644
--- a/include/vbi/syscall.h
+++ b/include/vbi/syscall.h
@@ -144,8 +144,8 @@ extern asmlinkage int32_t vbi_vb_restart(uint32_t id, int32_t core);
 extern asmlinkage int32_t vbi_vb_resume(uint32_t id, int32_t core);
 
 /* read remote vb's memory */
-extern asmlinkage int32_t vbi_vb_read_mem(VBI_MEM_CTL *memCtl, uint32_t targetBoard);
-extern asmlinkage int32_t vbi_vb_write_mem(VBI_MEM_CTL *memCtl, uint32_t targetBoard);
+extern asmlinkage int32_t vbi_vb_read_mem(struct vbi_mem_ctl *memCtl, uint32_t targetBoard);
+extern asmlinkage int32_t vbi_vb_write_mem(struct vbi_mem_ctl *memCtl, uint32_t targetBoard);
 extern asmlinkage void vbi_shell_start_debug(uint32_t  flags);
 
 /* read remote vb's registers */
-- 
1.6.5.2

