From 29df36a6ebd5c3243d3e8f532c393a7b78727ee9 Mon Sep 17 00:00:00 2001
From: Paul Gortmaker <paul.gortmaker@windriver.com>
Date: Wed, 18 Nov 2009 10:07:25 -0500
Subject: [PATCH 35/80] vbiExcOffsetsSet --> vbi_set_exc_offset

Signed-off-by: Paul Gortmaker <paul.gortmaker@windriver.com>
---
 arch/powerpc/kernel/vbi/syscalls.S |   10 +++++-----
 1 files changed, 5 insertions(+), 5 deletions(-)

diff --git a/arch/powerpc/kernel/vbi/syscalls.S b/arch/powerpc/kernel/vbi/syscalls.S
index ddeac6d..fbddc95 100644
--- a/arch/powerpc/kernel/vbi/syscalls.S
+++ b/arch/powerpc/kernel/vbi/syscalls.S
@@ -52,7 +52,7 @@ FUNC_EXPORT(vbi_vcore_irq_unlock)
 FUNC_EXPORT(vbi_vcore_irq_lock)
 FUNC_EXPORT(vbi_vcore_irq_state)
 FUNC_EXPORT(vbi_set_exc_base)
-FUNC_EXPORT(vbiExcOffsetsSet)
+FUNC_EXPORT(vbi_set_exc_offset)
 FUNC_EXPORT(vbiExcOffsetsGet)
 FUNC_EXPORT(vbiMemAttrSet)
 FUNC_EXPORT(vbiMemAttrGet)
@@ -507,7 +507,7 @@ baseAlignError:
 FUNC_END(vbi_set_exc_base)
 
 /*
- * vbiExcOffsetsSet - Set the exceptions vector offsetcs for a virtual core
+ * vbi_set_exc_offset - Set the exceptions vector offsetcs for a virtual core
  *
  * This routine changes the exceptions vector table offset addresses for the
  * active virtual core. By default they follow the standard PPC exception
@@ -523,7 +523,7 @@ FUNC_END(vbi_set_exc_base)
  * started up and before it's added to the scheduler for the guest OS.
  *
  */
-FUNC_LABEL(vbiExcOffsetsSet)
+FUNC_LABEL(vbi_set_exc_offset)
 	/* set command for the ioctl */
 	mr      r4, r3
 	li	r3, VBI_HYIOCTL_EXCOFFSETS_SET
@@ -533,14 +533,14 @@ FUNC_LABEL(vbiExcOffsetsSet)
 	ori r0, r0, LO(VBI_SYS_hyIoctl)
 	HCALL
 	blr
-FUNC_END(vbiExcOffsetsSet)
+FUNC_END(vbi_set_exc_offset)
 
 /*
  * vbiExcOffsetsGet - Get the exceptions vector offsetcs for a virtual core
  *
  * This routine gets the exceptions vector table offset addresses for the
  * active virtual core. By default they follow the standard PPC exception
- * offsets. vbiExcOffsetsSet will enable the offsets to be changed for the
+ * offsets. vbi_set_exc_offset will enable the offsets to be changed for the
  * virtual board. The offset table uses an index based on IVOR number.
  *
  */
-- 
1.6.5.2

