From 0edd2c367e33ed3a6dd848d9e30c12bd0823dd83 Mon Sep 17 00:00:00 2001
From: Paul Gortmaker <paul.gortmaker@windriver.com>
Date: Wed, 18 Nov 2009 12:34:57 -0500
Subject: [PATCH 52/80] vbiExcStub --> vbi_exc_stub

Signed-off-by: Paul Gortmaker <paul.gortmaker@windriver.com>
---
 include/vbi/vbi.h |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/include/vbi/vbi.h b/include/vbi/vbi.h
index 48c7506..43afa05 100644
--- a/include/vbi/vbi.h
+++ b/include/vbi/vbi.h
@@ -517,7 +517,7 @@ extern struct vb_status *wrhvVbStatus;
 extern struct vb_control *wrhvVbControl;
 extern int32_t wrhvVbIntNested;
 extern void vbiInit(struct vb_config *config);
-extern void vbiExcStub(void);
+extern void vbi_exc_stub(void);
 extern void ASSERT_FN(const char *, const char *, int);
 extern void vbiIdle(uint64_t);
 extern void *vbiVioapicAddrGet(void);
-- 
1.6.5.2

