From dfa1d8cd84dcb1561c989750dbfac3c1d32f3e74 Mon Sep 17 00:00:00 2001
From: Paul Gortmaker <paul.gortmaker@windriver.com>
Date: Wed, 18 Nov 2009 10:07:24 -0500
Subject: [PATCH 33/80] vbiIntVCoreStateGet --> vbi_vcore_irq_state

Signed-off-by: Paul Gortmaker <paul.gortmaker@windriver.com>
---
 arch/powerpc/kernel/vbi/syscalls.S |    8 ++++----
 1 files changed, 4 insertions(+), 4 deletions(-)

diff --git a/arch/powerpc/kernel/vbi/syscalls.S b/arch/powerpc/kernel/vbi/syscalls.S
index d8f7a50..e7a8728 100644
--- a/arch/powerpc/kernel/vbi/syscalls.S
+++ b/arch/powerpc/kernel/vbi/syscalls.S
@@ -50,7 +50,7 @@ FUNC_EXPORT(vbi_kputs)
 FUNC_EXPORT(vbi_kputc)
 FUNC_EXPORT(vbi_vcore_irq_unlock)
 FUNC_EXPORT(vbi_vcore_irq_lock)
-FUNC_EXPORT(vbiIntVCoreStateGet)
+FUNC_EXPORT(vbi_vcore_irq_state)
 FUNC_EXPORT(vbiExcBaseSet)
 FUNC_EXPORT(vbiExcOffsetsSet)
 FUNC_EXPORT(vbiExcOffsetsGet)
@@ -457,19 +457,19 @@ FUNC_END(vbi_vcore_irq_lock)
 
 /*
  *
- * vbiIntVCoreStateGet - Get interrupts state for running core
+ * vbi_vcore_irq_state - Get interrupts state for running core
  *
  * This routine returns the interrupts state for the calling core. This is C
  * wrapper function for VBI_INT_VCORE_STATE_GET(). If interrupts are locked it
  * returns TRUE if interrupts are locked otherwise FALSE.
  *
  */
-FUNC_LABEL(vbiIntVCoreStateGet)
+FUNC_LABEL(vbi_vcore_irq_state)
 	lis	p2, HIADJ(wrhvVbControl)
 	lwz	p2, LO(wrhvVbControl)(p2)
 	lwz	p0, VB_CONTROL_INT_DISABLE(p2)
 	blr
-FUNC_END(vbiIntVCoreStateGet)
+FUNC_END(vbi_vcore_irq_state)
 
 /*
  * vbiExcBaseSet - Set the exceptions vector table base for a virtual core
-- 
1.6.5.2

