From 2e6d77253879a89be294ccfaa7f18c181e954a8e Mon Sep 17 00:00:00 2001
From: Paul Gortmaker <paul.gortmaker@windriver.com>
Date: Wed, 18 Nov 2009 12:35:00 -0500
Subject: [PATCH 57/80] vbiInterruptsShow --> vbi_show_irq

Signed-off-by: Paul Gortmaker <paul.gortmaker@windriver.com>
---
 kernel/vbi/show.c |    6 +++---
 1 files changed, 3 insertions(+), 3 deletions(-)

diff --git a/kernel/vbi/show.c b/kernel/vbi/show.c
index 510f3fc..89c8f30 100644
--- a/kernel/vbi/show.c
+++ b/kernel/vbi/show.c
@@ -80,14 +80,14 @@ void vbiMemoryShow(struct vb_config *config)
 }
 
 /*
- * vbiInterruptsShow - print information about a board's
+ * vbi_show_irq - print information about a board's
  * interrupts configuration
  *
  * This routine traverse the interrupts descriptors table for a given
  * board and displays the information on hypervisor console.
  *
  */
-static void vbiInterruptsShow(struct vb_config *config)
+static void vbi_show_irq(struct vb_config *config)
 {
 	struct vb_int_info *p = config->interruptConfiguration;
 	int num_int = config->num_ints;
@@ -185,5 +185,5 @@ void vbi_show_cfg(struct vb_config *p)
 
 	vbiMemoryShow(p);
 	vbiSharedMemoryShow(p);
-	vbiInterruptsShow(p);
+	vbi_show_irq(p);
 }
-- 
1.6.5.2

