From 425eb028dc6bacd31754df7e0335e57014ca2acd Mon Sep 17 00:00:00 2001
From: Paul Gortmaker <paul.gortmaker@windriver.com>
Date: Wed, 2 Dec 2009 13:56:40 -0500
Subject: [PATCH 14/43] vbiMemAttr_t --> int32_t

for i in `git grep -l vbiMemAttr_t` ; do mv $i $i~ ; cat $i~|sed 's/vbiMemAttr_t/int32_t/' > $i ; done

Signed-off-by: Paul Gortmaker <paul.gortmaker@windriver.com>
---
 include/vbi/compat.h  |    1 +
 include/vbi/syscall.h |    4 ++--
 include/vbi/types.h   |    1 -
 3 files changed, 3 insertions(+), 3 deletions(-)

diff --git a/include/vbi/compat.h b/include/vbi/compat.h
index fb81708..aa1965c 100644
--- a/include/vbi/compat.h
+++ b/include/vbi/compat.h
@@ -53,6 +53,7 @@
 #define vbiPhysAddr_t		uint64_t
 #define vbiGuestPhysAddr_t	void *
 #define vbiIntState_t		int32_t
+#define vbiMemAttr_t		int32_t
 
 /*
  * struct fields and similar
diff --git a/include/vbi/syscall.h b/include/vbi/syscall.h
index 6c7e738..dc2339f 100644
--- a/include/vbi/syscall.h
+++ b/include/vbi/syscall.h
@@ -125,8 +125,8 @@ extern int32_t vbi_tlb_flush_vmmu(struct vmmuConfig * config, void *addr, size_t
 
 /* Prior to vbi 2.0 these api were vbi_set_mmu_attr/Get */
 
-extern asmlinkage int32_t  vbi_set_mem_attr(void *vaddr, size_t len, vbiMemAttr_t attr);
-extern asmlinkage int32_t  vbi_get_mem_attr(void *vaddr, vbiMemAttr_t * attr);
+extern asmlinkage int32_t  vbi_set_mem_attr(void *vaddr, size_t len, int32_t attr);
+extern asmlinkage int32_t  vbi_get_mem_attr(void *vaddr, int32_t * attr);
 
 /*
  * START: New APIs introduced for vbi 2.0  
diff --git a/include/vbi/types.h b/include/vbi/types.h
index 0824a88..f36cac3 100644
--- a/include/vbi/types.h
+++ b/include/vbi/types.h
@@ -34,7 +34,6 @@
 
 #ifndef _ASMLANGUAGE
 
-typedef int32_t vbiMemAttr_t;
 typedef int32_t vbiVector_t;
 typedef uint32_t vbiCoreSet_t;
 
-- 
1.6.5.2

