From ee80cd4a1b7f4e723f0cf7345481f924f36fbd23 Mon Sep 17 00:00:00 2001
From: Paul Gortmaker <paul.gortmaker@windriver.com>
Date: Wed, 18 Nov 2009 12:35:05 -0500
Subject: [PATCH 64/80] vbiNsUnRegister --> vbi_ns_unregister

Signed-off-by: Paul Gortmaker <paul.gortmaker@windriver.com>
---
 include/vbi/syscall.h |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/include/vbi/syscall.h b/include/vbi/syscall.h
index 9f9e70e..89da136 100644
--- a/include/vbi/syscall.h
+++ b/include/vbi/syscall.h
@@ -118,7 +118,7 @@ extern vbiStatus_t vbi_config_vmmu(struct vmmuConfig * config);
 extern vbiStatus_t vbi_enable_vmmu(uint32_t  vmmu_num);
 extern vbiStatus_t vbi_disable_vmmu(uint32_t vmmu_num);
 extern vbiStatus_t vbi_ns_register(char  *name, uint32_t  revision);
-extern vbiStatus_t vbiNsUnRegister(char *name, uint32_t  revision);
+extern vbiStatus_t vbi_ns_unregister(char *name, uint32_t  revision);
 extern vbiStatus_t vbi_ns_lookup(char *name, uint32_t  rev, VBI_NS_HANDLE *pHandle);
 extern vbiStatus_t vbi_tlb_flush_vmmu(struct vmmuConfig * config, void *addr, size_t len);
 
-- 
1.6.5.2

