From 304d4acfc1bbee49e899caad11c653c9f3250b9f Mon Sep 17 00:00:00 2001
From: Paul Gortmaker <paul.gortmaker@windriver.com>
Date: Wed, 18 Nov 2009 10:07:31 -0500
Subject: [PATCH 42/80] vbiVmmuTlbLoad --> vbi_tlb_load_vmmu

Signed-off-by: Paul Gortmaker <paul.gortmaker@windriver.com>
---
 arch/powerpc/kernel/vbi/syscalls.S |    8 ++++----
 1 files changed, 4 insertions(+), 4 deletions(-)

diff --git a/arch/powerpc/kernel/vbi/syscalls.S b/arch/powerpc/kernel/vbi/syscalls.S
index 6e1448d..1a9d673 100644
--- a/arch/powerpc/kernel/vbi/syscalls.S
+++ b/arch/powerpc/kernel/vbi/syscalls.S
@@ -59,7 +59,7 @@ FUNC_EXPORT(vbi_get_mem_attr)
 FUNC_EXPORT(vbi_config_vmmu)
 FUNC_EXPORT(vbi_enable_vmmu)
 FUNC_EXPORT(vbi_disable_vmmu)
-FUNC_EXPORT(vbiVmmuTlbLoad)
+FUNC_EXPORT(vbi_tlb_load_vmmu)
 FUNC_EXPORT(vbiVmmuTlbFlush)
 FUNC_EXPORT(vbiVbMgmt)
 FUNC_EXPORT(vbiNsOp)
@@ -611,18 +611,18 @@ FUNC_END(vbi_disable_vmmu)
 
 /*
  *
- * vbiVmmuTlbLoad - load a TLB entry into the virtual MMU
+ * vbi_tlb_load_vmmu - load a TLB entry into the virtual MMU
  *
  * This system call loads the TLB entries for the specified address range into
  * the virtual MMU.
  *
  */
-FUNC_LABEL(vbiVmmuTlbLoad)
+FUNC_LABEL(vbi_tlb_load_vmmu)
 	lis r0, HI(VBI_SYS_vmmu_tlbload)
 	ori r0, r0, LO(VBI_SYS_vmmu_tlbload)
 	HCALL
 	blr
-FUNC_END(vbiVmmuTlbLoad)
+FUNC_END(vbi_tlb_load_vmmu)
 
 
 /*
-- 
1.6.5.2

