diff --git a/arch/x86/kernel/pci-dma.c b/arch/x86/kernel/pci-dma.c
index 87d4d69..1fec06f 100644
--- a/arch/x86/kernel/pci-dma.c
+++ b/arch/x86/kernel/pci-dma.c
@@ -3,6 +3,11 @@
 #include <linux/bootmem.h>
 #include <linux/pci.h>
 
+#ifdef CONFIG_WRHV
+#include <vbi/interface.h>
+#include <vbi/vbi.h>
+#endif
+
 #include <asm/proto.h>
 #include <asm/dma.h>
 #include <asm/iommu.h>
@@ -336,6 +341,18 @@ dma_alloc_coherent(struct device *dev, size_t size, dma_addr_t *dma_handle,
 
 		memset(memory, 0, size);
 		if (!mmu) {
+#ifdef CONFIG_WRHV
+			u64 paddr;
+
+			if (vbi_get_guest_dma_addr((void *)bus, &paddr) == 0)
+				*dma_handle = (dma_addr_t)paddr;
+			else {
+				free_pages((unsigned long)memory,
+				   			get_order(size));
+				return NULL;
+			}
+#else
 			*dma_handle = bus;
+#endif
 			return memory;
 		}
diff --git a/arch/x86/kernel/pci-nommu.c b/arch/x86/kernel/pci-nommu.c
index 3f91f71..b00fbbe 100644
--- a/arch/x86/kernel/pci-nommu.c
+++ b/arch/x86/kernel/pci-nommu.c
@@ -7,6 +7,12 @@
 #include <linux/dma-mapping.h>
 #include <linux/scatterlist.h>
 
+#ifdef CONFIG_WRHV
+#include <vbi/interface.h>
+#include <vbi/vbi.h>
+#endif
+
+
 #include <asm/iommu.h>
 #include <asm/processor.h>
 #include <asm/dma.h>
@@ -63,6 +69,18 @@ static int nommu_map_sg(struct device *hwdev, struct scatterlist *sg,
 
 	for_each_sg(sg, s, nents, i) {
 		BUG_ON(!sg_page(s));
+#ifdef CONFIG_WRHV
+		{
+			u64 paddr;
+			dma_addr_t addr = sg_phys(s);
+
+			if (vbi_get_guest_dma_addr((void *)addr, &paddr) == 0)
+				s->dma_address = (dma_addr_t)paddr;
+			else
+				s->dma_address = -1;
+		}
+#else
 		s->dma_address = sg_phys(s);
+#endif
 		if (!check_addr("map_sg", hwdev, s->dma_address, s->length))
 			return 0;
diff --git a/arch/x86/mach-default/setup.c b/arch/x86/mach-default/setup.c
index 3d31783..36854ab 100644
--- a/arch/x86/mach-default/setup.c
+++ b/arch/x86/mach-default/setup.c
@@ -9,6 +9,7 @@
 #include <asm/arch_hooks.h>
 #include <asm/e820.h>
 #include <asm/setup.h>
+#include <asm/wrhv.h>	/* for TIMER_INT_NUM, used for WRHV and !WRHV */
 
 #ifdef CONFIG_HOTPLUG_CPU
 #define DEFAULT_SEND_IPI	(1)
@@ -132,7 +133,7 @@ void __init time_init_hook(void)
 	}
 
 	irq0.mask = cpumask_of_cpu(0);
-	setup_irq(0, &irq0);
+	setup_irq(TIMER_INT_NUM, &irq0);
 }
 
 #ifdef CONFIG_MCA
