From f3e57d431f1b47196cffa3627faea043ea958793 Mon Sep 17 00:00:00 2001
From: Weiwei Wang <weiwei.wang@windriver.com>
Date: Mon, 19 Oct 2009 10:21:48 +0800
Subject: [PATCH 4/4] x86: switch serial port into irq mode

Currently serial port works in poll mode, intput overrun
error always appears. So switch to irq mode to fix it

Signed-off-by: Weiwei Wang <weiwei.wang@windriver.com>
---
 include/asm-x86/serial.h |    4 ++--
 1 files changed, 2 insertions(+), 2 deletions(-)

diff --git a/include/asm-x86/serial.h b/include/asm-x86/serial.h
index dd26c58..b6f4ce8 100644
--- a/include/asm-x86/serial.h
+++ b/include/asm-x86/serial.h
@@ -21,8 +21,8 @@
 
 #define SERIAL_PORT_DFNS			\
 	/* UART CLK   PORT IRQ     FLAGS        */			\
-	{ 0, BASE_BAUD, 0x3F8, 0, STD_COM_FLAGS },	/* ttyS0 */	\
-	{ 0, BASE_BAUD, 0x2F8, 0, STD_COM_FLAGS },	/* ttyS1 */	\
+	{ 0, BASE_BAUD, 0x3F8, 4, STD_COM_FLAGS },	/* ttyS0 */	\
+	{ 0, BASE_BAUD, 0x2F8, 3, STD_COM_FLAGS },	/* ttyS1 */	\
 	{ 0, BASE_BAUD, 0x3E8, 4, STD_COM_FLAGS },	/* ttyS2 */	\
 	{ 0, BASE_BAUD, 0x2E8, 3, STD_COM4_FLAGS },	/* ttyS3 */	\
 	{ 0, BASE_BAUD, 0x220, 0, STD_COM_FLAGS },      /* ttyS4 */     \
-- 
1.6.3.3

