From 45e5a1d866baba2101255b29d3206f27fd2be390 Mon Sep 17 00:00:00 2001
From: Paul Gortmaker <paul.gortmaker@windriver.com>
Date: Thu, 5 Nov 2009 15:13:15 -0500
Subject: [PATCH 02/22] x86: Fix sys_vbi.h includes.

This file isn't being used, and so its include paths were never
updated to things that map into real files.

Signed-off-by: Paul Gortmaker <paul.gortmaker@windriver.com>
---
 include/asm-x86/sys_vbi.h |    5 ++---
 1 files changed, 2 insertions(+), 3 deletions(-)

diff --git a/include/asm-x86/sys_vbi.h b/include/asm-x86/sys_vbi.h
index 9ff0c08..d9f11e2 100644
--- a/include/asm-x86/sys_vbi.h
+++ b/include/asm-x86/sys_vbi.h
@@ -17,9 +17,8 @@
 #ifndef __INCx86Archh
 #define __INCx86Archh
 
-#include <sys/regs.h>
-#include <sys/x86/vmx.h>
-#include <interface.h>
+#include <asm/reg_vbi.h>
+#include <vbi/interface.h>
 
 #ifdef LP64
 #define _WRS_PORTABLE_bLib
-- 
1.6.5.2

