diff --git a/arch/x86/kernel/setup.c b/arch/x86/kernel/setup.c
index 8d5feb2..814f7f9 100644
--- a/arch/x86/kernel/setup.c
+++ b/arch/x86/kernel/setup.c
@@ -102,6 +102,9 @@
 #include <asm/percpu.h>
 #include <asm/topology.h>
 #include <asm/apicdef.h>
+#ifdef CONFIG_WRHV
+#include <asm/wrhv.h>
+#endif
 #ifdef CONFIG_X86_64
 #include <asm/numa_64.h>
 #endif
@@ -617,6 +629,9 @@ static struct dmi_system_id __initdata bad_bios_dmi_table[] = {
 
 void __init setup_arch(char **cmdline_p)
 {
+#ifdef CONFIG_WRHV
+	wrhv_boot_config();
+#endif
 #ifdef CONFIG_X86_32
 	memcpy(&boot_cpu_data, &new_cpu_data, sizeof(new_cpu_data));
 	visws_early_detect();
@@ -696,6 +710,9 @@ void __init setup_arch(char **cmdline_p)
 	strlcpy(command_line, boot_command_line, COMMAND_LINE_SIZE);
 	*cmdline_p = command_line;
 
+#ifdef CONFIG_WRHV
+	wrhv_init();
+#endif
 	parse_early_param();
 
 #ifdef CONFIG_X86_64
@@ -902,6 +918,11 @@ void __init setup_arch(char **cmdline_p)
 
 	e820_setup_gap();
+
+#ifdef CONFIG_WRHV
+	wrhv_calibrate_smp_cpus();
+#endif
+
 #ifdef CONFIG_VT
 #if defined(CONFIG_VGA_CONSOLE)
 	if (!efi_enabled || (efi_mem_type(0xa0000) != EFI_CONVENTIONAL_MEMORY))
