From 4ec068c1c7c6487d45d15c0c7d6835d3fddb9a55 Mon Sep 17 00:00:00 2001
From: "Yadi.hu" <yadi.hu@windriver.com>
Date: Tue, 17 Jun 2014 19:32:34 -0700
Subject: [PATCH] PPC: LSI ACP Elpaso support 476fpe errata by default

Signed-off-by: Yadi.hu <yadi.hu@windriver.com>
---
 arch/powerpc/platforms/44x/Kconfig |    2 ++
 1 files changed, 2 insertions(+), 0 deletions(-)

diff --git a/arch/powerpc/platforms/44x/Kconfig b/arch/powerpc/platforms/44x/Kconfig
index fafa41b..d50bbf2 100644
--- a/arch/powerpc/platforms/44x/Kconfig
+++ b/arch/powerpc/platforms/44x/Kconfig
@@ -204,6 +204,8 @@ config ELPASO
 	depends on (44x || 40x)
 	default n
 	select ACP
+	select 476FPE
+	select 476FPE_ERR46
 	select PPC_CLOCK
 	select DEFAULT_UIMAGE
 
-- 
1.7.5.4

