From af9b0a7fac14de305ff36c3216a069291c1b734c Mon Sep 17 00:00:00 2001
From: Paul Gortmaker <paul.gortmaker@windriver.com>
Date: Fri, 5 Sep 2008 00:22:19 -0400
Subject: [PATCH] patch rcu-preempt-boost-default.patch

---
 kernel/Kconfig.preempt |    1 +
 1 files changed, 1 insertions(+), 0 deletions(-)

diff --git a/kernel/Kconfig.preempt b/kernel/Kconfig.preempt
index e8b6489..d86f0a7 100644
--- a/kernel/Kconfig.preempt
+++ b/kernel/Kconfig.preempt
@@ -69,6 +69,7 @@ config PREEMPT_RCU
 config PREEMPT_RCU_BOOST
 	bool "Enable priority boosting of RCU read-side critical sections"
 	depends on PREEMPT_RCU
+	default y if PREEMPT_RT
 	help
 	  This option permits priority boosting of RCU read-side critical
 	  sections tat have been preempted and a RT process is waiting
-- 
1.6.0.90.g436ed

