From e8f4b13bc1e287cf72d64daf3771458bccab47a1 Mon Sep 17 00:00:00 2001
From: Paul Gortmaker <paul.gortmaker@windriver.com>
Date: Tue, 30 Sep 2008 16:54:46 -0400
Subject: [PATCH] patch rcu-preempt-hotplug-hackaround.patch

---
 kernel/sched.c |    1 -
 1 files changed, 0 insertions(+), 1 deletions(-)

diff --git a/kernel/sched.c b/kernel/sched.c
index b0aed91..f138be0 100644
--- a/kernel/sched.c
+++ b/kernel/sched.c
@@ -7676,7 +7676,6 @@ static void detach_destroy_domains(const cpumask_t *cpu_map)
 
 	for_each_cpu_mask_nr(i, *cpu_map)
 		cpu_attach_domain(NULL, &def_root_domain, i);
-	synchronize_sched();
 	arch_destroy_sched_domains(cpu_map, &tmpmask);
 }
 
-- 
1.6.0.90.g436ed

