From cf5499fb329f3d6b2255e0b3087b3ea0ece4bb7e Mon Sep 17 00:00:00 2001
From: Paul Gortmaker <paul.gortmaker@windriver.com>
Date: Tue, 30 Sep 2008 16:57:25 -0400
Subject: [PATCH] patch preempt-realtime-ftrace-disable-ftraced.patch

---
 init/main.c |    3 +++
 1 files changed, 3 insertions(+), 0 deletions(-)

diff --git a/init/main.c b/init/main.c
index a62eaaf..d6ca9bf 100644
--- a/init/main.c
+++ b/init/main.c
@@ -340,6 +340,9 @@ static int __init rdinit_setup(char *str)
 {
 	unsigned int i;
 
+#ifdef CONFIG_PREEMPT_RT
+	ftrace_disable_daemon();
+#endif
 	ramdisk_execute_command = str;
 	/* See "auto" comment in init_setup */
 	for (i = 1; i < MAX_INIT_ARGS; i++)
-- 
1.6.0.90.g436ed

