From 6630cf562f58ca2f8f322961dbf2f31f9d64b755 Mon Sep 17 00:00:00 2001
From: Paul Gortmaker <paul.gortmaker@windriver.com>
Date: Tue, 30 Sep 2008 20:02:51 -0400
Subject: [PATCH] patch remove-check-pgt-cache-calls.patch

---
 arch/x86/kernel/process_32.c |    1 -
 1 files changed, 0 insertions(+), 1 deletions(-)

diff --git a/arch/x86/kernel/process_32.c b/arch/x86/kernel/process_32.c
index 0843d7f..13c21c0 100644
--- a/arch/x86/kernel/process_32.c
+++ b/arch/x86/kernel/process_32.c
@@ -132,7 +132,6 @@ void cpu_idle(void)
 		tick_nohz_stop_sched_tick(1);
 		while (!need_resched()) {
 
-			check_pgt_cache();
 			rmb();
 
 			if (rcu_pending(cpu))
-- 
1.6.0.90.g436ed

