From 5dd219995a84bf3712b4fc5023e7275d8ba13195 Mon Sep 17 00:00:00 2001
From: Paul Gortmaker <paul.gortmaker@windriver.com>
Date: Tue, 30 Sep 2008 16:57:58 -0400
Subject: [PATCH] patch preempt-irqs-i386-idle-poll-loop-fix.patch

---
 arch/x86/kernel/process.c |    4 +++-
 1 files changed, 3 insertions(+), 1 deletions(-)

diff --git a/arch/x86/kernel/process.c b/arch/x86/kernel/process.c
index 876e918..36e133e 100644
--- a/arch/x86/kernel/process.c
+++ b/arch/x86/kernel/process.c
@@ -185,7 +185,9 @@ static void mwait_idle(void)
 static void poll_idle(void)
 {
 	local_irq_enable();
-	cpu_relax();
+	do {
+		cpu_relax();
+	} while (!need_resched());
 }
 
 /*
-- 
1.6.0.90.g436ed

