From 79151e03b4dc96033d07084e0919938939a1cc30 Mon Sep 17 00:00:00 2001
From: Paul Gortmaker <paul.gortmaker@windriver.com>
Date: Fri, 5 Sep 2008 00:23:08 -0400
Subject: [PATCH] patch preempt-realtime-sound.patch

---
 sound/core/pcm_lib.c |    1 +
 1 files changed, 1 insertions(+), 0 deletions(-)

diff --git a/sound/core/pcm_lib.c b/sound/core/pcm_lib.c
index 1533f03..68dc1b7 100644
--- a/sound/core/pcm_lib.c
+++ b/sound/core/pcm_lib.c
@@ -29,6 +29,7 @@
 #include <sound/pcm_params.h>
 #include <sound/timer.h>
 
+#include <linux/ftrace.h>
 /*
  * fill ring buffer with silence
  * runtime->silence_start: starting pointer to silence area
-- 
1.6.0.90.g436ed

