From 5767e5ae7bf6b15bb334857c7598818ba42b4b9a Mon Sep 17 00:00:00 2001
From: Paul Gortmaker <paul.gortmaker@windriver.com>
Date: Tue, 30 Sep 2008 16:58:49 -0400
Subject: [PATCH] patch preempt-realtime-supress-rtc-printk.patch

---
 drivers/char/rtc.c |    2 ++
 1 files changed, 2 insertions(+), 0 deletions(-)

diff --git a/drivers/char/rtc.c b/drivers/char/rtc.c
index b1b9dd7..283267e 100644
--- a/drivers/char/rtc.c
+++ b/drivers/char/rtc.c
@@ -1379,10 +1379,12 @@ static void rtc_dropped_irq(unsigned long data)
 
 	spin_unlock_irq(&rtc_lock);
 
+#ifndef CONFIG_PREEMPT_RT
 	if (printk_ratelimit()) {
 		printk(KERN_WARNING "rtc: lost some interrupts at %ldHz.\n",
 			freq);
 	}
+#endif
 
 	/* Now we have new data */
 	wake_up_interruptible(&rtc_wait);
-- 
1.6.0.90.g436ed

