From 6f26664e1c33eabcbc09cb5092bfe13b2201c68d Mon Sep 17 00:00:00 2001
From: Paul Gortmaker <paul.gortmaker@windriver.com>
Date: Fri, 5 Sep 2008 00:23:15 -0400
Subject: [PATCH] patch panic-dont-stop-box.patch

---
 kernel/panic.c |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/kernel/panic.c b/kernel/panic.c
index d717270..0b37928 100644
--- a/kernel/panic.c
+++ b/kernel/panic.c
@@ -96,7 +96,7 @@ NORET_TYPE void panic(const char * fmt, ...)
 	 * unfortunately means it may not be hardened to work in a panic
 	 * situation.
 	 */
-	smp_send_stop();
+//	smp_send_stop();
 #endif
 
 	atomic_notifier_call_chain(&panic_notifier_list, 0, buf);
-- 
1.6.0.90.g436ed

