From 0709427547617a874f10d9f7cfdaccb72526414f Mon Sep 17 00:00:00 2001
From: Ingo Molnar <mingo@elte.hu>
Date: Fri, 5 Sep 2008 00:23:16 -0400
Subject: [PATCH] core: make asm/irq_regs.h available on every platform
 the softlockup detector would like to use get_irq_regs(), so generalize
 the availability on every Linux architecture.

(it is fine for an architecture to always return NULL to get_irq_regs(),
which it does by default.)

Signed-off-by: Ingo Molnar <mingo@elte.hu>
---
 include/asm-arm26/irq_regs.h |    1 +
 1 files changed, 1 insertions(+), 0 deletions(-)
 create mode 100644 include/asm-arm26/irq_regs.h

diff --git a/include/asm-arm26/irq_regs.h b/include/asm-arm26/irq_regs.h
new file mode 100644
index 0000000..3dd9c0b
--- /dev/null
+++ b/include/asm-arm26/irq_regs.h
@@ -0,0 +1 @@
+#include <asm-generic/irq_regs.h>
-- 
1.6.0.90.g436ed

