From b5dadf0f348b314ee0155b3a96f27589be0ed0ee Mon Sep 17 00:00:00 2001
From: Paul Gortmaker <paul.gortmaker@windriver.com>
Date: Fri, 5 Sep 2008 00:23:22 -0400
Subject: [PATCH] patch radix-concurrent-lockdep.patch

---
 lib/radix-tree.c |   24 +++++++++++++++++++++++-
 1 files changed, 23 insertions(+), 1 deletions(-)

diff --git a/lib/radix-tree.c b/lib/radix-tree.c
index f3fcd14..fa64877 100644
--- a/lib/radix-tree.c
+++ b/lib/radix-tree.c
@@ -79,6 +79,26 @@ static unsigned long height_to_maxindex[RADIX_TREE_MAX_PATH + 1] __read_mostly;
 #ifdef CONFIG_RADIX_TREE_CONCURRENT
 static struct lock_class_key radix_node_class[RADIX_TREE_MAX_PATH];
 #endif
+#ifdef CONFIG_DEBUG_LOCK_ALLOC
+static const char *radix_node_key_string[RADIX_TREE_MAX_PATH] = {
+	"radix-node-00",
+	"radix-node-01",
+	"radix-node-02",
+	"radix-node-03",
+	"radix-node-04",
+	"radix-node-05",
+	"radix-node-06",
+	"radix-node-07",
+	"radix-node-08",
+	"radix-node-09",
+	"radix-node-10",
+	"radix-node-11",
+	"radix-node-12",
+	"radix-node-13",
+	"radix-node-14",
+	"radix-node-15",
+};
+#endif
 
 #ifdef CONFIG_RADIX_TREE_OPTIMISTIC
 static DEFINE_PER_CPU(unsigned long[RADIX_TREE_MAX_PATH+1], optimistic_histogram);
@@ -282,7 +302,9 @@ radix_tree_node_alloc(struct radix_tree_root *root, int height)
 	BUG_ON(radix_tree_is_indirect_ptr(ret));
 #ifdef CONFIG_RADIX_TREE_CONCURRENT
 	spin_lock_init(&ret->lock);
-	lockdep_set_class(&ret->lock, &radix_node_class[height]);
+	lockdep_set_class_and_name(&ret->lock,
+			&radix_node_class[height],
+			radix_node_key_string[height]);
 #endif
 	ret->height = height;
 	return ret;
-- 
1.6.0.90.g436ed

