From 8e6560d40d7f5b80a800766030a67a1fc19cc020 Mon Sep 17 00:00:00 2001
From: Paul Gortmaker <paul.gortmaker@windriver.com>
Date: Fri, 5 Sep 2008 00:23:23 -0400
Subject: [PATCH] As it is currently written, sys_select checks its return code to convert
 ERESTARTNOHAND to EINTR.  However, the check is within an if (tvp) clause, and
 so if select is called from userspace with a NULL timeval, then it is possible
 for the ERESTARTNOHAND errno to leak into userspace, which is incorrect.  This
 patch moves that check outside of the conditional, and prevents the errno leak.

Thanks & Regards
Neil

Signed-Off-By: Neil Horman <nhorman@tuxdriver.com>


 fs/select.c |   18 +++++-------------
 1 file changed, 5 insertions(+), 13 deletions(-)


Index: linux-2.6.26.3-rt6/fs/select.c
===================================================================
---
 fs/select.c |   18 +++++-------------
 1 files changed, 5 insertions(+), 13 deletions(-)

diff --git a/fs/select.c b/fs/select.c
index da0e882..8f6d906 100644
--- a/fs/select.c
+++ b/fs/select.c
@@ -408,20 +408,12 @@ asmlinkage long sys_select(int n, fd_set __user *inp, fd_set __user *outp,
 		rtv.tv_sec = timeout;
 		if (timeval_compare(&rtv, &tv) >= 0)
 			rtv = tv;
-		if (copy_to_user(tvp, &rtv, sizeof(rtv))) {
-sticky:
-			/*
-			 * If an application puts its timeval in read-only
-			 * memory, we don't want the Linux-specific update to
-			 * the timeval to cause a fault after the select has
-			 * completed successfully. However, because we're not
-			 * updating the timeval, we can't restart the system
-			 * call.
-			 */
-			if (ret == -ERESTARTNOHAND)
-				ret = -EINTR;
-		}
+		if (copy_to_user(tvp, &rtv, sizeof(rtv)))
+			return -EFAULT;
 	}
+sticky:
+	if (ret == -ERESTARTNOHAND)
+		ret = -EINTR;
 
 	return ret;
 }
-- 
1.6.0.90.g436ed

