From 4a7d0166fb10474d51e280e6902003e516feda4f Mon Sep 17 00:00:00 2001
From: Paul Gortmaker <paul.gortmaker@windriver.com>
Date: Tue, 30 Sep 2008 16:59:32 -0400
Subject: [PATCH] patch highmem-redo-mainline.patch

---
 mm/highmem.c |    8 ++++++++
 1 files changed, 8 insertions(+), 0 deletions(-)

diff --git a/mm/highmem.c b/mm/highmem.c
index 1a390da..d4e8870 100644
--- a/mm/highmem.c
+++ b/mm/highmem.c
@@ -214,6 +214,14 @@ static unsigned long pkmap_insert(struct page *page)
 	return vaddr;
 }
 
+/*
+ * Flush all unused kmap mappings in order to remove stray mappings.
+ */
+void kmap_flush_unused(void)
+{
+	WARN_ON_ONCE(1);
+}
+
  void *kmap_high(struct page *page)
 {
 	unsigned long vaddr;
-- 
1.6.0.90.g436ed

