From d9714f976df83ccbceab2d98ddd88d2ba0ea0fa8 Mon Sep 17 00:00:00 2001
From: Paul Gortmaker <paul.gortmaker@windriver.com>
Date: Tue, 30 Sep 2008 16:59:57 -0400
Subject: [PATCH] patch preempt-rt-no-slub.patch

---
 init/Kconfig |    1 +
 1 files changed, 1 insertions(+), 0 deletions(-)

diff --git a/init/Kconfig b/init/Kconfig
index 8b57f46..d8f79ec 100644
--- a/init/Kconfig
+++ b/init/Kconfig
@@ -758,6 +758,7 @@ config SLAB
 
 config SLUB
 	bool "SLUB (Unqueued Allocator)"
+	depends on !PREEMPT_RT
 	help
 	   SLUB is a slab allocator that minimizes cache line usage
 	   instead of managing queues of cached objects (SLAB approach).
-- 
1.6.0.90.g436ed

