From 199a1fa97cc75a5d46bf6e15ec50da563ee15c3e Mon Sep 17 00:00:00 2001
From: Paul Gortmaker <paul.gortmaker@windriver.com>
Date: Fri, 5 Sep 2008 00:23:40 -0400
Subject: [PATCH] patch printk-dont-bug-on-sched.patch

---
 kernel/rtmutex.c |    9 +++++++++
 1 files changed, 9 insertions(+), 0 deletions(-)

diff --git a/kernel/rtmutex.c b/kernel/rtmutex.c
index 97b7205..8938919 100644
--- a/kernel/rtmutex.c
+++ b/kernel/rtmutex.c
@@ -14,6 +14,7 @@
 #include <linux/module.h>
 #include <linux/sched.h>
 #include <linux/timer.h>
+#include <linux/hardirq.h>
 #include <linux/semaphore.h>
 
 #include "rtmutex_common.h"
@@ -636,6 +637,9 @@ rt_spin_lock_fastlock(struct rt_mutex *lock,
 	/* Temporary HACK! */
 	if (!current->in_printk)
 		might_sleep();
+	else if (in_atomic() || irqs_disabled())
+		/* don't grab locks for printk in atomic */
+		return;
 
 	if (likely(rt_mutex_cmpxchg(lock, NULL, current)))
 		rt_mutex_deadlock_account_lock(lock, current);
@@ -647,6 +651,11 @@ static inline void
 rt_spin_lock_fastunlock(struct rt_mutex *lock,
 			void  (*slowfn)(struct rt_mutex *lock))
 {
+	/* Temporary HACK! */
+	if (current->in_printk && (in_atomic() || irqs_disabled()))
+		/* don't grab locks for printk in atomic */
+		return;
+
 	if (likely(rt_mutex_cmpxchg(lock, current, NULL)))
 		rt_mutex_deadlock_account_unlock(current);
 	else
-- 
1.6.0.90.g436ed

