From e6d66b8d9d37a07ad325d038a41e12f1536b2e5d Mon Sep 17 00:00:00 2001
From: Paul Gortmaker <paul.gortmaker@windriver.com>
Date: Fri, 5 Sep 2008 00:23:40 -0400
Subject: [PATCH] patch user-no-irq-disable.patch

---
 kernel/user.c |    4 ++--
 1 files changed, 2 insertions(+), 2 deletions(-)

diff --git a/kernel/user.c b/kernel/user.c
index 22fc905..ec6ee9e 100644
--- a/kernel/user.c
+++ b/kernel/user.c
@@ -289,14 +289,14 @@ static void remove_user_sysfs_dir(struct work_struct *w)
 	 */
 	uids_mutex_lock();
 
-	local_irq_save(flags);
+	local_irq_save_nort(flags);
 
 	if (atomic_dec_and_lock(&up->__count, &uidhash_lock)) {
 		uid_hash_remove(up);
 		remove_user = 1;
 		spin_unlock_irqrestore(&uidhash_lock, flags);
 	} else {
-		local_irq_restore(flags);
+		local_irq_restore_nort(flags);
 	}
 
 	if (!remove_user)
-- 
1.6.0.90.g436ed

