From 8fd2d582eb712d907a638eb0733f485bb9ce2820 Mon Sep 17 00:00:00 2001
From: Paul Gortmaker <paul.gortmaker@windriver.com>
Date: Fri, 5 Sep 2008 00:23:41 -0400
Subject: [PATCH] patch local_irq_save_nort-in-swap.patch

---
 mm/swap.c |    4 ++--
 1 files changed, 2 insertions(+), 2 deletions(-)

diff --git a/mm/swap.c b/mm/swap.c
index 5ff1e2d..0b53830 100644
--- a/mm/swap.c
+++ b/mm/swap.c
@@ -290,9 +290,9 @@ static void drain_cpu_pagevecs(int cpu)
 		unsigned long flags;
 
 		/* No harm done if a racing interrupt already did this */
-		local_irq_save(flags);
+		local_irq_save_nort(flags);
 		pagevec_move_tail(pvec);
-		local_irq_restore(flags);
+		local_irq_restore_nort(flags);
 	}
 	swap_per_cpu_unlock(lru_rotate_pvecs, cpu);
 }
-- 
1.6.0.90.g436ed

