From f2b84a392b8462be4a9a9e422c0796e45b59dfcb Mon Sep 17 00:00:00 2001
From: Paul Gortmaker <paul.gortmaker@windriver.com>
Date: Fri, 5 Sep 2008 00:23:47 -0400
Subject: [PATCH] Hack fix to prevent a bunch of warnings about using smp_processor_id
 in preemptable code. The real fix would be to pass down a cpu variable
 and have an RT sleeping spinlock to protect it.

Signed-off-by: Steven Rostedt <srostedt@redhat.com>
---
 lib/radix-tree.c |   10 ++++++++++
 1 files changed, 10 insertions(+), 0 deletions(-)

diff --git a/lib/radix-tree.c b/lib/radix-tree.c
index fa64877..90d2304 100644
--- a/lib/radix-tree.c
+++ b/lib/radix-tree.c
@@ -105,10 +105,20 @@ static DEFINE_PER_CPU(unsigned long[RADIX_TREE_MAX_PATH+1], optimistic_histogram
 
 static void optimistic_hit(unsigned long height)
 {
+	unsigned long flags;
+
 	if (height > RADIX_TREE_MAX_PATH)
 		height = RADIX_TREE_MAX_PATH;
 
+	/*
+	 * HACK:
+	 *  On PREEMPT_RT this protects the percpu var.
+	 *  We really need to fix this to pass in cpu var protected
+	 *  with a RT sleeping spinlock.
+	 */
+	local_irq_save(flags);
 	__get_cpu_var(optimistic_histogram)[height]++;
+	local_irq_restore(flags);
 }
 
 #ifdef CONFIG_PROC_FS
-- 
1.6.0.90.g436ed

