From 7725811e31a4fb9c05c1892a37ac057f7f857bfb Mon Sep 17 00:00:00 2001
From: Paul Gortmaker <paul.gortmaker@windriver.com>
Date: Tue, 30 Sep 2008 17:01:26 -0400
Subject: [PATCH] Protect use of smp_processor_id in genhd.h with preempt disable.

Signed-off-by: Steven Rostedt <srostedt@redhat.com>
---
 include/linux/genhd.h |    6 +++++-
 1 files changed, 5 insertions(+), 1 deletions(-)

diff --git a/include/linux/genhd.h b/include/linux/genhd.h
index 21d325d..6150331 100644
--- a/include/linux/genhd.h
+++ b/include/linux/genhd.h
@@ -192,7 +192,11 @@ static inline void disk_stat_set_all(struct gendisk *gendiskp, int value)	{
 }		
 
 #define __part_stat_add(part, field, addnd)				\
-	(per_cpu_ptr(part->dkstats, smp_processor_id())->field += addnd)
+do {									\
+	preempt_disable();						\
+	(per_cpu_ptr(part->dkstats, smp_processor_id())->field += addnd); \
+	preempt_enable();						\
+} while (0)
 
 #define __all_stat_add(gendiskp, part, field, addnd, sector)	\
 ({								\
-- 
1.6.0.90.g436ed

