From 542ad8539ca2f495e7ac1b923880d3da235c31e8 Mon Sep 17 00:00:00 2001
From: Paul Gortmaker <paul.gortmaker@windriver.com>
Date: Fri, 5 Sep 2008 00:23:58 -0400
Subject: [PATCH] patch rwlock-slowunlock-mutex-fix.patch

---
 kernel/rtmutex.c |   10 ++++++++++
 1 files changed, 10 insertions(+), 0 deletions(-)

diff --git a/kernel/rtmutex.c b/kernel/rtmutex.c
index 502bd9e..50b0850 100644
--- a/kernel/rtmutex.c
+++ b/kernel/rtmutex.c
@@ -1794,6 +1794,16 @@ rt_read_slowunlock(struct rw_mutex *rwm, int mtx)
 
 	wakeup_next_waiter(mutex, savestate);
 
+	/*
+	 * If we woke up a reader but the lock is already held by readers
+	 * we need to set the mutex owner to RT_RW_READER, since the
+	 * wakeup_next_waiter set it to the pending reader.
+	 */
+	if (reader_count) {
+		WARN_ON(waiter->write_lock);
+		rt_mutex_set_owner(mutex, RT_RW_READER, 0);
+	}
+
 	if (rt_mutex_has_waiters(mutex)) {
 		waiter = rt_mutex_top_waiter(mutex);
 		rwm->prio = waiter->task->prio;
-- 
1.6.0.90.g436ed

