From a97b7fc595e7b467666ccb28aae769cc4fe05ea2 Mon Sep 17 00:00:00 2001
From: Steven Rostedt <srostedt@redhat.com>
Date: Fri, 5 Sep 2008 00:24:00 -0400
Subject: [PATCH] Subject: rwlock: fix torture test to handle non-rt

Signed-off-by: Steven Rostedt <srostedt@redhat.com>
---
 kernel/rwlock_torture.c |    2 ++
 1 files changed, 2 insertions(+), 0 deletions(-)

diff --git a/kernel/rwlock_torture.c b/kernel/rwlock_torture.c
index 0c4b44d..2820815 100644
--- a/kernel/rwlock_torture.c
+++ b/kernel/rwlock_torture.c
@@ -570,7 +570,9 @@ static int run_test(unsigned long time)
 		run_two_locks(time, read);
 	}
 
+#ifdef CONFIG_PREEMPT_RT
 	WARN_ON_ONCE(current->reader_lock_count);
+#endif
 
 	return ret;
 }
-- 
1.6.0.90.g436ed

