From 8ee0355e34fe42bec832314272bb0cdfd5d6a37f Mon Sep 17 00:00:00 2001
From: Paul Gortmaker <paul.gortmaker@windriver.com>
Date: Fri, 5 Sep 2008 00:24:04 -0400
Subject: [PATCH] Signed-off-by: Steven Rostedt <srostedt@redhat.com>

---
 lib/locking-selftest.c |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/lib/locking-selftest.c b/lib/locking-selftest.c
index 938d048..f0766fe 100644
--- a/lib/locking-selftest.c
+++ b/lib/locking-selftest.c
@@ -948,7 +948,7 @@ static void dotest(void (*testcase_fn)(void), int expected, int lockclass_mask)
 	unsigned long saved_preempt_count = preempt_count();
 	int expected_failure = 0;
 #if defined(CONFIG_DEBUG_PREEMPT) && defined(CONFIG_DEBUG_RT_MUTEXES)
-        int saved_lock_count = current->lock_count;
+        atomic_t saved_lock_count = current->lock_count;
 #endif
 
 	WARN_ON(irqs_disabled());
-- 
1.6.0.90.g436ed

