From 1008df5f3fe792f4c72ea69db675c5a2fa33f7c7 Mon Sep 17 00:00:00 2001
From: Paul Gortmaker <paul.gortmaker@windriver.com>
Date: Tue, 30 Sep 2008 17:03:09 -0400
Subject: [PATCH] Compile fix for when profiling is off

prof_pid will be an unresolved symbol if you don't enable
profiling, and the final link of the kernel will fail. Make
the sysctl conditional on the appropriate Kconfig option.

[merge me into preempt-realtime-debug-sysctl.patch]
---
 kernel/sysctl.c |    2 ++
 1 files changed, 2 insertions(+), 0 deletions(-)

diff --git a/kernel/sysctl.c b/kernel/sysctl.c
index 91a482a..284e599 100644
--- a/kernel/sysctl.c
+++ b/kernel/sysctl.c
@@ -378,6 +378,7 @@ static struct ctl_table kern_table[] = {
 		.proc_handler	= &proc_dointvec,
 	},
 #endif
+#ifdef CONFIG_PROFILING
 	{
 		.ctl_name	= CTL_UNNUMBERED,
 		.procname	= "prof_pid",
@@ -386,6 +387,7 @@ static struct ctl_table kern_table[] = {
 		.mode		= 0644,
 		.proc_handler	= &proc_dointvec,
 	},
+#endif
 #ifdef CONFIG_PREEMPT
 	{
 		.ctl_name	= CTL_UNNUMBERED,
-- 
1.6.0.90.g436ed

