From fe62e83d1c3fa67639ab96599435e0f5d949715b Mon Sep 17 00:00:00 2001
From: Paul Gortmaker <paul.gortmaker@windriver.com>
Date: Tue, 30 Sep 2008 15:16:33 -0400
Subject: [PATCH] Date: Wed, 24 Sep 2008 21:53:13 -0400
 Subject: [PATCH] compat_rwsem: fix compile failure for when PREEMPT_RT is disabled.
 If you disable PRREMPT_RT, you will get a compile failure on a missing
 compat prototype.  This change should be considered for folding into
 rt-mutex-i386.patch

Signed-off-by: Paul Gortmaker <paul.gortmaker@windriver.com>
---
 include/asm-x86/rwsem.h |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/include/asm-x86/rwsem.h b/include/asm-x86/rwsem.h
index b60eb18..4e4f94a 100644
--- a/include/asm-x86/rwsem.h
+++ b/include/asm-x86/rwsem.h
@@ -257,7 +257,7 @@ static inline int rwsem_atomic_update(int delta, struct compat_rw_semaphore *sem
 	return tmp + delta;
 }
 
-static inline int rwsem_is_locked(struct compat_rw_semaphore *sem)
+static inline int compat_rwsem_is_locked(struct compat_rw_semaphore *sem)
 {
 	return (sem->count != 0);
 }
-- 
1.6.0.90.g436ed

