From e89cb0dd800b58494f5cc4534d0a98f0aef90d16 Mon Sep 17 00:00:00 2001
From: Ingo Molnar <mingo@elte.hu>
Date: Tue, 30 Sep 2008 19:43:44 -0400
Subject: [PATCH] add -rt extra-version add -rt extra-version.

Signed-off-by: Ingo Molnar <mingo@elte.hu>
Signed-off-by: Thomas Gleixner <tglx@linutronix.de>
Signed-off-by: Steven Rostedt <rostedt@goodmis.org>
---
 Makefile |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/Makefile b/Makefile
index df72336..026448c 100644
--- a/Makefile
+++ b/Makefile
@@ -1,7 +1,7 @@
 VERSION = 2
 PATCHLEVEL = 6
 SUBLEVEL = 27
-EXTRAVERSION = .2
+EXTRAVERSION = .2-26rt9
 NAME = Rotary Wombat
 
 # *DOCUMENTATION*
-- 
1.5.5.1

