From b16d1a6960f3e026462ed1dbb4cd1cabb9daaea4 Mon Sep 17 00:00:00 2001
From: Thomas Gleixner <tglx@linutronix.de>
Date: Sun, 17 Jul 2011 22:51:33 +0200
Subject: [PATCH] cond-resched-lock-rt-tweak.patch

Signed-off-by: Thomas Gleixner <tglx@linutronix.de>

diff --git a/include/linux/preempt_mask.h b/include/linux/preempt_mask.h
index 5cb25f17331a..795f9c154f72 100644
--- a/include/linux/preempt_mask.h
+++ b/include/linux/preempt_mask.h
@@ -83,7 +83,11 @@
 /*
  * The preempt_count offset after spin_lock()
  */
+#ifdef CONFIG_PREEMPT_RT_FULL
+#define PREEMPT_LOCK_OFFSET	0
+#else
 #define PREEMPT_LOCK_OFFSET	PREEMPT_DISABLE_OFFSET
+#endif
 
 /*
  * The preempt_count offset needed for things like:
-- 
2.1.4

