From a315590018d104d3df0d0945ee47c3e8e570cad3 Mon Sep 17 00:00:00 2001
From: Thomas Gleixner <tglx@linutronix.de>
Date: Fri, 8 Jul 2011 16:35:35 +0200
Subject: [PATCH] epoll.patch

Signed-off-by: Thomas Gleixner <tglx@linutronix.de>

diff --git a/fs/eventpoll.c b/fs/eventpoll.c
index 33c9599c14e7..05bc433dcca8 100644
--- a/fs/eventpoll.c
+++ b/fs/eventpoll.c
@@ -488,12 +488,12 @@ static int ep_poll_wakeup_proc(void *priv, void *cookie, int call_nests)
  */
 static void ep_poll_safewake(wait_queue_head_t *wq)
 {
-	int this_cpu = get_cpu();
+	int this_cpu = get_cpu_light();
 
 	ep_call_nested(&poll_safewake_ncalls, EP_MAX_NESTS,
 		       ep_poll_wakeup_proc, NULL, wq, (void *) (long) this_cpu);
 
-	put_cpu();
+	put_cpu_light();
 }
 
 static void ep_remove_wait_queue(struct eppoll_entry *pwq)
-- 
2.0.1

