From: Steven Rostedt <srostedt@redhat.com>
Subject: ftrace: declare function_trace_stop if ftrace not defined

Signed-off-by: Steven Rostedt <srostedt@redhat.com>
---
 include/linux/ftrace.h |    3 +++
 1 files changed, 3 insertions(+), 0 deletions(-)

diff --git a/include/linux/ftrace.h b/include/linux/ftrace.h
index a77301b..e7b3050 100644
--- a/include/linux/ftrace.h
+++ b/include/linux/ftrace.h
@@ -118,6 +118,9 @@ static inline void ftrace_preempt_enable(int resched)
 # define ftrace_kill()				do { } while (0)
 # define __ftrace_kill()			do { } while (0)
 # define ftrace_kill_atomic()			do { } while (0)
+# define function_trace_stop			0
+# define ftrace_stop()				do { } while (0)
+# define ftrace_start()				do { } while (0)
 #endif /* CONFIG_FTRACE */
 
 #ifdef CONFIG_DYNAMIC_FTRACE
