From 1a514356a51ac7fbd2210888b64c6d8bd9225334 Mon Sep 17 00:00:00 2001
From: Mike Galbraith <bitbucket@online.de>
Date: Thu, 25 Apr 2013 14:36:14 +0200
Subject: [PATCH] hrtimer: fix hrtimer free zone build bug

Link: http://lkml.kernel.org/r/1366893374.6139.5.camel@marge.simpson.net

Cc: Thomas Gleixner <tglx@linutronix.de>
Signed-off-by: Mike Galbraith <bitbucket@online.de>
Signed-off-by: Steven Rostedt <rostedt@goodmis.org>

diff --git a/kernel/hrtimer.c b/kernel/hrtimer.c
index 59c1de9..aeaa70b 100644
--- a/kernel/hrtimer.c
+++ b/kernel/hrtimer.c
@@ -1043,7 +1043,7 @@ int __hrtimer_start_range_ns(struct hrtimer *timer, ktime_t tim,
 	 * XXX send_remote_softirq() ?
 	 */
 	if (leftmost && new_base->cpu_base == &__get_cpu_var(hrtimer_bases)) {
-#ifdef CONFIG_PREEMPT_RT_BASE
+#if defined(CONFIG_PREEMPT_RT_BASE) && defined(CONFIG_HIGH_RES_TIMERS)
 again:
 		if (hrtimer_enqueue_reprogram(timer, new_base)) {
 			/*
-- 
1.8.4.1

