From ff06f4934ecfc2309ed1d5a6314d95a7e2c4185f Mon Sep 17 00:00:00 2001
From: Thomas Gleixner <tglx@linutronix.de>
Date: Wed, 14 Sep 2011 14:48:43 +0200
Subject: [PATCH] hrtimer-fix-reprogram-madness.patch

Signed-off-by: Thomas Gleixner <tglx@linutronix.de>

diff --git a/kernel/hrtimer.c b/kernel/hrtimer.c
index df62d7b2dd77..6f0711ae32d7 100644
--- a/kernel/hrtimer.c
+++ b/kernel/hrtimer.c
@@ -1335,7 +1335,11 @@ static void hrtimer_rt_reprogram(int restart, struct hrtimer *timer,
 		if (!enqueue_hrtimer(timer, base))
 			return;
 
-		if (hrtimer_reprogram(timer, base))
+#ifndef CONFIG_HIGH_RES_TIMERS
+	}
+#else
+		if (base->cpu_base->hres_active &&
+		    hrtimer_reprogram(timer, base))
 			goto requeue;
 
 	} else if (hrtimer_active(timer)) {
@@ -1344,6 +1348,7 @@ static void hrtimer_rt_reprogram(int restart, struct hrtimer *timer,
 		 * the event device.
 		 */
 		if (&timer->node == base->active.next &&
+		    base->cpu_base->hres_active &&
 		    hrtimer_reprogram(timer, base))
 			goto requeue;
 	}
@@ -1356,6 +1361,7 @@ requeue:
 	 */
 	__remove_hrtimer(timer, base, timer->state, 0);
 	list_add_tail(&timer->cb_entry, &base->expired);
+#endif
 }
 
 /*
-- 
2.0.1

