From 48a804c567880605245c05c1c4ad91a4e86fd46c Mon Sep 17 00:00:00 2001
From: Thomas Gleixner <tglx@linutronix.de>
Date: Wed, 12 Aug 2009 12:16:59 +0200
Subject: [PATCH] hwlat: Move wakequeue wake up out of irq disable region

commit 5fc485f6341c177e9d5f641c6fb9e33ca465aa84 in tip.

Signed-off-by: Thomas Gleixner <tglx@linutronix.de>
Signed-off-by: Paul Gortmaker <paul.gortmaker@windriver.com>
---
 drivers/misc/hwlat_detector.c |    4 ++--
 1 files changed, 2 insertions(+), 2 deletions(-)

diff --git a/drivers/misc/hwlat_detector.c b/drivers/misc/hwlat_detector.c
index e02d8e1..be6553f 100644
--- a/drivers/misc/hwlat_detector.c
+++ b/drivers/misc/hwlat_detector.c
@@ -262,8 +262,6 @@ static int get_sample(void *unused)
 		/* Keep a running maximum ever recorded hardware latency */
 		if (sample > data.max_sample)
 			data.max_sample = sample;
-
-		wake_up(&data.wq); /* wake up reader(s) */
 	}
 
 	ret = 0;
@@ -301,6 +299,8 @@ static int kthread_fn(void *unused)
 			goto err_out;
 		}
 
+		wake_up(&data.wq); /* wake up reader(s) */
+
 		interval = data.sample_window - data.sample_width;
 		do_div(interval, USEC_PER_MSEC); /* modifies interval value */
 
-- 
1.7.0.4

