From d01b08a935c7b0776233ad46a069869ecd54e69f Mon Sep 17 00:00:00 2001
From: Paul Gortmaker <paul.gortmaker@windriver.com>
Date: Tue, 23 Sep 2014 11:29:51 -0400
Subject: [PATCH] Linux 3.10.55-rt58 REBASE

Signed-off-by: Paul Gortmaker <paul.gortmaker@windriver.com>

diff --git a/localversion-rt b/localversion-rt
index 08b3e75841ad..c06cc4356292 100644
--- a/localversion-rt
+++ b/localversion-rt
@@ -1 +1 @@
--rt14
+-rt58
-- 
1.9.2

