From edcbe7c031c252e502a2bd3f66dccea4c9370f64 Mon Sep 17 00:00:00 2001
From: "Steven Rostedt (Red Hat)" <rostedt@goodmis.org>
Date: Wed, 11 Mar 2015 21:51:28 -0400
Subject: [PATCH] Linux 3.14.38-rt36 REBASE

commit fc3e663810917d73f15351dae0386caf819e4be9 upstream.


diff --git a/localversion-rt b/localversion-rt
new file mode 100644
index 000000000000..22940344b6c0
--- /dev/null
+++ b/localversion-rt
@@ -0,0 +1 @@
+-rt36
-- 
2.3.4

