From 0a27ad2767f7c530759faa14d4474f001d37cadb Mon Sep 17 00:00:00 2001
From: "Steven Rostedt (Red Hat)" <rostedt@goodmis.org>
Date: Mon, 7 Mar 2016 15:03:04 -0500
Subject: [PATCH 335/335] Linux 4.1.19-rt22 REBASE

---
 localversion-rt | 1 +
 1 file changed, 1 insertion(+)
 create mode 100644 localversion-rt

diff --git a/localversion-rt b/localversion-rt
new file mode 100644
index 000000000000..c29508d21914
--- /dev/null
+++ b/localversion-rt
@@ -0,0 +1 @@
+-rt22
-- 
2.5.0

