From 6ddab7ca2ad6488649f435d8ec0f83c5810fb009 Mon Sep 17 00:00:00 2001
From: "Steven Rostedt (Red Hat)" <rostedt@goodmis.org>
Date: Tue, 7 May 2013 09:38:22 -0400
Subject: [PATCH] Linux 3.4.57-rt72 REBASE

---
 localversion-rt | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/localversion-rt b/localversion-rt
index 700c857..2c95a3c 100644
--- a/localversion-rt
+++ b/localversion-rt
@@ -1 +1 @@
--rt8
+-rt72
-- 
1.8.3.1

