From b7ae940e49f1f5ff2ad128f47cf27fb743fe503b Mon Sep 17 00:00:00 2001
From: Sebastian Andrzej Siewior <bigeasy@linutronix.de>
Date: Tue, 22 Dec 2015 16:54:04 +0100
Subject: [PATCH] Linux 4.1.15-rt17 REBASE

Signed-off-by: Sebastian Andrzej Siewior <bigeasy@linutronix.de>

diff --git a/localversion-rt b/localversion-rt
new file mode 100644
index 000000000000..1e584b47c987
--- /dev/null
+++ b/localversion-rt
@@ -0,0 +1 @@
+-rt17
-- 
2.5.0

