From 26f384e2ed7aff092ede4c8800c1a44dc9620641 Mon Sep 17 00:00:00 2001
From: "Steven Rostedt (Red Hat)" <rostedt@goodmis.org>
Date: Thu, 17 Jul 2014 11:09:48 -0400
Subject: [PATCH] Linux 3.10.64-rt68 REBASE

---
 localversion-rt | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/localversion-rt b/localversion-rt
index 08b3e75841ad..df2c52c58e02 100644
--- a/localversion-rt
+++ b/localversion-rt
@@ -1 +1 @@
--rt14
+-rt68
-- 
2.1.0

