From 2bbe91ffd10dda791b20ea51934221d3541124c7 Mon Sep 17 00:00:00 2001
From: Paul Gortmaker <paul.gortmaker@windriver.com>
Date: Wed, 17 Sep 2014 18:20:32 -0400
Subject: [PATCH] Linux 3.4.103-rt127 REBASE

Signed-off-by: Paul Gortmaker <paul.gortmaker@windriver.com>

diff --git a/localversion-rt b/localversion-rt
index 700c857efd9b..4de2714cd138 100644
--- a/localversion-rt
+++ b/localversion-rt
@@ -1 +1 @@
--rt8
+-rt127
-- 
2.0.1

