From 35238790f65238c27fe972c4b27f4b2eb052ddd5 Mon Sep 17 00:00:00 2001
From: Thomas Gleixner <tglx@linutronix.de>
Date: Fri, 24 Jun 2011 18:40:37 +0200
Subject: [PATCH 084/309] local-var.patch

Signed-off-by: Thomas Gleixner <tglx@linutronix.de>
---
 include/linux/percpu.h | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/include/linux/percpu.h b/include/linux/percpu.h
index e3817d2441b6..0eb64302fc0e 100644
--- a/include/linux/percpu.h
+++ b/include/linux/percpu.h
@@ -49,6 +49,11 @@
 	preempt_enable();				\
 } while (0)
 
+#define get_local_var(var)	get_cpu_var(var)
+#define put_local_var(var)	put_cpu_var(var)
+#define get_local_ptr(var)	get_cpu_ptr(var)
+#define put_local_ptr(var)	put_cpu_ptr(var)
+
 /* minimum unit size, also is the maximum supported allocation size */
 #define PCPU_MIN_UNIT_SIZE		PFN_ALIGN(32 << 10)
 
-- 
1.8.1.2

