From dcf814a353029641603f3d5462fb60a36e103bd0 Mon Sep 17 00:00:00 2001
From: Thomas Gleixner <tglx@linutronix.de>
Date: Fri, 8 Jul 2011 20:25:16 +0200
Subject: [PATCH] Add localversion for -RT release

Signed-off-by: Thomas Gleixner <tglx@linutronix.de>

diff --git a/localversion-rt b/localversion-rt
new file mode 100644
index 000000000000..6f206be67cd2
--- /dev/null
+++ b/localversion-rt
@@ -0,0 +1 @@
+-rt1
-- 
2.5.0

