From 7c0a873f73080fd3117df6917a296c09f049b63c Mon Sep 17 00:00:00 2001
From: Paul Gortmaker <paul.gortmaker@windriver.com>
Date: Mon, 24 Jan 2011 20:35:30 -0500
Subject: [PATCH] Makefile: mark the tree as RT WIP

Signed-off-by: Paul Gortmaker <paul.gortmaker@windriver.com>
---
 Makefile |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/Makefile b/Makefile
index 64b1dd6..70e7a48 100644
--- a/Makefile
+++ b/Makefile
@@ -1,7 +1,7 @@
 VERSION = 2
 PATCHLEVEL = 6
 SUBLEVEL = 34
-EXTRAVERSION = .10
+EXTRAVERSION = .10-rt
 NAME = Sheep on Meth
 
 # *DOCUMENTATION*
-- 
1.7.0.2

