From b276129c853027fce83095dc01592d14eb9540dd Mon Sep 17 00:00:00 2001
From: Thomas Gleixner <tglx@linutronix.de>
Date: Mon, 12 Aug 2013 09:16:09 +0200
Subject: [PATCH] mm: disable slab on rt


diff --git a/init/Kconfig b/init/Kconfig
index 75669bb34fc6..289091c4863f 100644
--- a/init/Kconfig
+++ b/init/Kconfig
@@ -1541,6 +1541,7 @@ choice
 
 config SLAB
 	bool "SLAB"
+	depends on !PREEMPT_RT_FULL
 	help
 	  The regular slab allocator that is established and known to work
 	  well in all environments. It organizes cache hot objects in
diff --git a/mm/slab.h b/mm/slab.h
index 4d6d836247dd..fc3c0976e664 100644
--- a/mm/slab.h
+++ b/mm/slab.h
@@ -247,7 +247,11 @@ static inline struct kmem_cache *cache_from_obj(struct kmem_cache *s, void *x)
  * The slab lists for all objects.
  */
 struct kmem_cache_node {
+#ifdef CONFIG_SLAB
 	spinlock_t list_lock;
+#else
+	raw_spinlock_t list_lock;
+#endif
 
 #ifdef CONFIG_SLAB
 	struct list_head slabs_partial;	/* partial list first, better asm code */
-- 
2.1.0

