From cd8c9730e6271bcf7851265c5f5e3858b902ef40 Mon Sep 17 00:00:00 2001
From: Thomas Gleixner <tglx@linutronix.de>
Date: Thu, 21 Jul 2011 16:47:49 +0200
Subject: [PATCH] mm-page-alloc-fix.patch

Signed-off-by: Thomas Gleixner <tglx@linutronix.de>

diff --git a/mm/page_alloc.c b/mm/page_alloc.c
index 7e80bfbc96e5..97cd54e4e75c 100644
--- a/mm/page_alloc.c
+++ b/mm/page_alloc.c
@@ -2327,8 +2327,8 @@ __alloc_pages_direct_compact(gfp_t gfp_mask, unsigned int order,
 		struct page *page;
 
 		/* Page migration frees to the PCP lists but we want merging */
-		drain_pages(get_cpu());
-		put_cpu();
+		drain_pages(get_cpu_light());
+		put_cpu_light();
 
 		page = get_page_from_freelist(gfp_mask, nodemask,
 				order, zonelist, high_zoneidx,
-- 
2.1.2

