From 33992e239717c5c3d6fb02b816cbc23030fcbf4c Mon Sep 17 00:00:00 2001
From: Corey Minyard <cminyard@mvista.com>
Date: Fri, 20 Sep 2013 13:02:29 -0500
Subject: [PATCH] mm/slab: Don't free locks twice in cpuup_canceled

Signed-off-by: Corey Minyard <cminyard@cavium.com>
Signed-off-by: Sebastian Andrzej Siewior <bigeasy@linutronix.de>

diff --git a/mm/slab.c b/mm/slab.c
index fa94c0b04265..1e9330f43f46 100644
--- a/mm/slab.c
+++ b/mm/slab.c
@@ -1276,7 +1276,6 @@ static void __cpuinit cpuup_canceled(long cpu)
 			free_block(cachep, nc->entry, nc->avail, node);
 
 		if (!cpumask_empty(mask)) {
-			local_spin_unlock_irq(slab_lock, &n->list_lock);
 			unlock_l3_and_free_delayed(&n->list_lock);
 			goto free_array_cache;
 		}
@@ -1291,7 +1290,6 @@ static void __cpuinit cpuup_canceled(long cpu)
 		alien = n->alien;
 		n->alien = NULL;
 
-		local_spin_unlock_irq(slab_lock, &n->list_lock);
 		unlock_l3_and_free_delayed(&n->list_lock);
 
 		kfree(shared);
-- 
2.1.0

