From 20a6dd0786f59c6100bd4a14fe8f9b70e41280e3 Mon Sep 17 00:00:00 2001
From: Thomas Gleixner <tglx@linutronix.de>
Date: Mon, 1 Oct 2012 17:12:35 +0100
Subject: [PATCH] net: Use get_cpu_light() in ip_send_unicast_reply()

Signed-off-by: Thomas Gleixner <tglx@linutronix.de>

diff --git a/net/ipv4/ip_output.c b/net/ipv4/ip_output.c
index 22fa05e041ea..0baf2220a250 100644
--- a/net/ipv4/ip_output.c
+++ b/net/ipv4/ip_output.c
@@ -1508,7 +1508,8 @@ void ip_send_unicast_reply(struct net *net, struct sk_buff *skb, __be32 daddr,
 	if (IS_ERR(rt))
 		return;
 
-	inet = &get_cpu_var(unicast_sock);
+	get_cpu_light();
+	inet = &__get_cpu_var(unicast_sock);
 
 	inet->tos = arg->tos;
 	sk = &inet->sk;
@@ -1537,7 +1538,7 @@ void ip_send_unicast_reply(struct net *net, struct sk_buff *skb, __be32 daddr,
 		ip_push_pending_frames(sk, &fl4);
 	}
 out:
-	put_cpu_var(unicast_sock);
+	put_cpu_light();
 
 	ip_rt_put(rt);
 }
-- 
2.0.1

