From 208b1c926ed88a43661cbc0122bacd3f6a3919c9 Mon Sep 17 00:00:00 2001
From: Thomas Gleixner <tglx@linutronix.de>
Date: Sun, 17 Jul 2011 16:29:27 +0200
Subject: [PATCH] net-netif_rx_ni-migrate-disable.patch

Signed-off-by: Thomas Gleixner <tglx@linutronix.de>

diff --git a/net/core/dev.c b/net/core/dev.c
index c4e2ae5f67fa..3fb67572d286 100644
--- a/net/core/dev.c
+++ b/net/core/dev.c
@@ -2965,11 +2965,11 @@ int netif_rx_ni(struct sk_buff *skb)
 {
 	int err;
 
-	preempt_disable();
+	migrate_disable();
 	err = netif_rx(skb);
 	if (local_softirq_pending())
 		thread_do_softirq();
-	preempt_enable();
+	migrate_enable();
 
 	return err;
 }
-- 
2.0.1

