From 672d5bfd6b86fe545d29b08e42522c4bddd2458a Mon Sep 17 00:00:00 2001
From: Ingo Molnar <mingo@elte.hu>
Date: Fri, 3 Jul 2009 08:30:11 -0500
Subject: [PATCH] net: loopback: realtime fix

commit 56c1457848714e720b1ac800656400a5ce61ca4f in tip.

Signed-off-by: Ingo Molnar <mingo@elte.hu>
Signed-off-by: Thomas Gleixner <tglx@linutronix.de>
Signed-off-by: Paul Gortmaker <paul.gortmaker@windriver.com>
---
 drivers/net/loopback.c |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/drivers/net/loopback.c b/drivers/net/loopback.c
index 377144f..6ad108f 100644
--- a/drivers/net/loopback.c
+++ b/drivers/net/loopback.c
@@ -88,7 +88,7 @@ static netdev_tx_t loopback_xmit(struct sk_buff *skb,
 
 	skb->protocol = eth_type_trans(skb, dev);
 	len = skb->len;
-	res = netif_rx(skb);
+	res = netif_rx_ni(skb);
 
 	pcpu_lstats = (void __percpu __force *)dev->ml_priv;
 	lb_stats = per_cpu_ptr(pcpu_lstats, xmit_get_cpu());
-- 
1.7.0.4

