From 316f9a4c0e464804033f032ace8ed5151c345d7d Mon Sep 17 00:00:00 2001
From: Ingo Molnar <mingo@elte.hu>
Date: Fri, 3 Jul 2009 08:29:53 -0500
Subject: [PATCH] net: Reduce preempt disabled region

commit 5a90027b0f8c9b872a7f51a4d5bbc63f5a8716cb in tip.

Signed-off-by: Ingo Molnar <mingo@elte.hu>
Signed-off-by: Thomas Gleixner <tglx@linutronix.de>
Signed-off-by: Paul Gortmaker <paul.gortmaker@windriver.com>
---
 net/core/dev.c |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/net/core/dev.c b/net/core/dev.c
index 740e71f..b11a306 100644
--- a/net/core/dev.c
+++ b/net/core/dev.c
@@ -2239,8 +2239,8 @@ int netif_rx_ni(struct sk_buff *skb)
 {
 	int err;
 
-	preempt_disable();
 	err = netif_rx(skb);
+	preempt_disable();
 	if (local_softirq_pending())
 		do_softirq();
 	preempt_enable();
-- 
1.7.0.4

