From 1de24ca18f44c8c3e2cd6dfc3f78cae6dd9bd943 Mon Sep 17 00:00:00 2001
From: Thomas Gleixner <tglx@linutronix.de>
Date: Thu, 7 Jun 2012 07:46:55 -0400
Subject: [PATCH] panic-disable-random-on-rt


diff --git a/kernel/panic.c b/kernel/panic.c
index 10b7685997c3..c93d13504228 100644
--- a/kernel/panic.c
+++ b/kernel/panic.c
@@ -371,9 +371,11 @@ static u64 oops_id;
 
 static int init_oops_id(void)
 {
+#ifndef CONFIG_PREEMPT_RT_FULL
 	if (!oops_id)
 		get_random_bytes(&oops_id, sizeof(oops_id));
 	else
+#endif
 		oops_id++;
 
 	return 0;
-- 
2.0.1

