From 81c0643963ab3aefdc8a45098dc2c4edd279f1e8 Mon Sep 17 00:00:00 2001
From: Thomas Gleixner <tglx@linutronix.de>
Date: Wed, 13 Nov 2013 11:21:05 -0500
Subject: [PATCH] panic: skip get_random_bytes for RT_FULL in init_oops_id


diff --git a/kernel/panic.c b/kernel/panic.c
index 167ec097ce8b..90b669058d58 100644
--- a/kernel/panic.c
+++ b/kernel/panic.c
@@ -363,9 +363,11 @@ static u64 oops_id;
 
 static int init_oops_id(void)
 {
+#ifndef CONFIG_PREEMPT_RT_FULL
 	if (!oops_id)
 		get_random_bytes(&oops_id, sizeof(oops_id));
 	else
+#endif
 		oops_id++;
 
 	return 0;
-- 
2.1.0

