From c5a0c8451e06ee6d2a5613ba414d886a8dbe1059 Mon Sep 17 00:00:00 2001
From: Thomas Gleixner <tglx@linutronix.de>
Date: Fri, 31 Oct 2014 10:54:18 -0400
Subject: [PATCH] panic: skip get_random_bytes for RT_FULL in init_oops_id


diff --git a/kernel/panic.c b/kernel/panic.c
index 6d6300375090..d7eb2453a18f 100644
--- a/kernel/panic.c
+++ b/kernel/panic.c
@@ -368,9 +368,11 @@ static u64 oops_id;
 
 static int init_oops_id(void)
 {
+#ifndef CONFIG_PREEMPT_RT_FULL
 	if (!oops_id)
 		get_random_bytes(&oops_id, sizeof(oops_id));
 	else
+#endif
 		oops_id++;
 
 	return 0;
-- 
2.1.2

