From 2eedcaf5d2234cf29d5c910caf91d643b1e230c7 Mon Sep 17 00:00:00 2001
From: Thomas Gleixner <tglx@linutronix.de>
Date: Mon, 14 Nov 2011 10:57:54 +0100
Subject: [PATCH] rcu-more-fallout.patch

Signed-off-by: Thomas Gleixner <tglx@linutronix.de>

diff --git a/kernel/rcutiny.c b/kernel/rcutiny.c
index a0714a51b6d7..c3f39bb59e73 100644
--- a/kernel/rcutiny.c
+++ b/kernel/rcutiny.c
@@ -373,6 +373,7 @@ void call_rcu_sched(struct rcu_head *head, void (*func)(struct rcu_head *rcu))
 }
 EXPORT_SYMBOL_GPL(call_rcu_sched);
 
+#ifndef CONFIG_PREEMPT_RT_FULL
 /*
  * Post an RCU bottom-half callback to be invoked after any subsequent
  * quiescent state.
@@ -382,3 +383,4 @@ void call_rcu_bh(struct rcu_head *head, void (*func)(struct rcu_head *rcu))
 	__call_rcu(head, func, &rcu_bh_ctrlblk);
 }
 EXPORT_SYMBOL_GPL(call_rcu_bh);
+#endif
-- 
1.9.0

