Subject: rtmutex-tglx-is-a-moron.patch
From: Thomas Gleixner <tglx@linutronix.de>
Date: Wed, 10 Dec 2008 12:36:37 +0100

Signed-off-by: Thomas Gleixner <tglx@linutronix.de>
---
 kernel/rtmutex.c |    5 +++--
 1 file changed, 3 insertions(+), 2 deletions(-)

Index: linux-2.6.26.8-rt12/kernel/rtmutex.c
===================================================================
--- linux-2.6.26.8-rt12.orig/kernel/rtmutex.c	2008-12-19 23:31:08.000000000 -0500
+++ linux-2.6.26.8-rt12/kernel/rtmutex.c	2008-12-19 23:31:08.000000000 -0500
@@ -2137,12 +2137,13 @@ static int rt_mutex_adjust_readers(struc
 		int chain_walk = 0;
 
 		task = rls->task;
-		get_task_struct(task);
 
 		spin_lock(&task->pi_lock);
 		__rt_mutex_adjust_prio(task);
-		if (task->pi_blocked_on)
+		if (task->pi_blocked_on) {
 			chain_walk = 1;
+			get_task_struct(task);
+		}
 		spin_unlock(&task->pi_lock);
 
 		/*
