From 874c50a0e9292318d2a7762827f465b532ccfc06 Mon Sep 17 00:00:00 2001
From: Thomas Gleixner <tglx@linutronix.de>
Date: Wed, 9 Jan 2013 23:03:29 +0100
Subject: [PATCH] sched: Init idle->on_rq in init_idle()

Cc: stable-rt@vger.kernel.org
Signed-off-by: Thomas Gleixner <tglx@linutronix.de>
Signed-off-by: Steven Rostedt <rostedt@goodmis.org>

diff --git a/kernel/sched/core.c b/kernel/sched/core.c
index 2920f78df850..bb15035b13b0 100644
--- a/kernel/sched/core.c
+++ b/kernel/sched/core.c
@@ -5322,6 +5322,7 @@ void __cpuinit init_idle(struct task_struct *idle, int cpu)
 	rcu_read_unlock();
 
 	rq->curr = rq->idle = idle;
+	idle->on_rq = 1;
 #if defined(CONFIG_SMP)
 	idle->on_cpu = 1;
 #endif
-- 
2.0.1

