From: Steven Rostedt <srostedt@redhat.com>
Subject: trace_point: add trace point to resched_task

Signed-off-by: Steven Rostedt <srostedt@redhat.com>
diff --git a/kernel/sched.c b/kernel/sched.c
index 0b0a0f8..d1a38d9 100644
--- a/kernel/sched.c
+++ b/kernel/sched.c
@@ -1335,6 +1335,8 @@ static void resched_task(struct task_struct *p)
 
 	set_tsk_thread_flag(p, TIF_NEED_RESCHED);
 
+	trace_sched_resched_task(p, task_cpu(p));
+
 	cpu = task_cpu(p);
 	if (cpu == smp_processor_id())
 		return;
