From 261117583a0e67f7f2b411c6d91a9bc3575e1be6 Mon Sep 17 00:00:00 2001
From: Ingo Molnar <mingo@elte.hu>
Date: Fri, 3 Jul 2009 08:30:01 -0500
Subject: [PATCH] serial: 8250: Call flush_to_ldisc when the irq is threaded

Signed-off-by: Ingo Molnar <mingo@elte.hu>
---
 drivers/tty/tty_buffer.c | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/drivers/tty/tty_buffer.c b/drivers/tty/tty_buffer.c
index cf78d1985cd8..e6b38a17f26e 100644
--- a/drivers/tty/tty_buffer.c
+++ b/drivers/tty/tty_buffer.c
@@ -362,7 +362,11 @@ void tty_schedule_flip(struct tty_port *port)
 	struct tty_bufhead *buf = &port->buf;
 
 	buf->tail->commit = buf->tail->used;
+#ifndef CONFIG_PREEMPT_RT_FULL
 	schedule_work(&buf->work);
+#else
+	schedule_work(&buf->work);
+#endif
 }
 EXPORT_SYMBOL(tty_schedule_flip);
 
-- 
1.8.1.2

