From 73e1541d5922d60697a79f39df9062f3ec2e5cc1 Mon Sep 17 00:00:00 2001
From: Yue Tao <Yue.Tao@windriver.com>
Date: Fri, 28 Jul 2017 09:08:53 +0800
Subject: [PATCH] The timer_stats_account_hrtimer has been removed by the fix
 of CVE-2017-5967

Signed-off-by: Yue Tao <Yue.Tao@windriver.com>
---
 kernel/hrtimer.c |    1 -
 1 files changed, 0 insertions(+), 1 deletions(-)

diff --git a/kernel/hrtimer.c b/kernel/hrtimer.c
index 40fdf2d..daf1509 100644
--- a/kernel/hrtimer.c
+++ b/kernel/hrtimer.c
@@ -1438,7 +1438,6 @@ static void hrtimer_rt_run_pending(void)
 			 */
 			debug_hrtimer_deactivate(timer);
 			__remove_hrtimer(timer, base, HRTIMER_STATE_CALLBACK, 0);
-			timer_stats_account_hrtimer(timer);
 			fn = timer->function;
 
 			raw_spin_unlock_irq(&cpu_base->lock);
-- 
1.7.5.4

