From 7c4ffc9f1e89606262657d67df9df500b4f08968 Mon Sep 17 00:00:00 2001
From: Thomas Gleixner <tglx@linutronix.de>
Date: Fri, 3 Jul 2009 08:44:36 -0500
Subject: [PATCH] trace: disable hw-branch tracer on -rt

commit 520d22385df681e3417065185b86065e2532f928 in tip.

Signed-off-by: Thomas Gleixner <tglx@linutronix.de>
Signed-off-by: Paul Gortmaker <paul.gortmaker@windriver.com>
---
 kernel/trace/Kconfig |    1 +
 1 files changed, 1 insertions(+), 0 deletions(-)

diff --git a/kernel/trace/Kconfig b/kernel/trace/Kconfig
index 43fe88c..5807cff 100644
--- a/kernel/trace/Kconfig
+++ b/kernel/trace/Kconfig
@@ -376,6 +376,7 @@ config STACK_TRACER
 
 config HW_BRANCH_TRACER
 	depends on HAVE_HW_BRANCH_TRACER
+	depends on !PREEMPT_RT
 	bool "Trace hw branches"
 	select GENERIC_TRACER
 	help
-- 
1.7.0.4

