From e855cab924c91477f661849bc455a367bd1ad07e Mon Sep 17 00:00:00 2001
From: Thomas Gleixner <tglx@linutronix.de>
Date: Sun, 17 Jul 2011 21:42:26 +0200
Subject: [PATCH] workqueue-use-get-cpu-light.patch

Signed-off-by: Thomas Gleixner <tglx@linutronix.de>

diff --git a/kernel/workqueue.c b/kernel/workqueue.c
index 5c5c3cf1263a..4d21bfdc1637 100644
--- a/kernel/workqueue.c
+++ b/kernel/workqueue.c
@@ -1078,8 +1078,8 @@ int queue_work(struct workqueue_struct *wq, struct work_struct *work)
 {
 	int ret;
 
-	ret = queue_work_on(get_cpu(), wq, work);
-	put_cpu();
+	ret = queue_work_on(get_cpu_light(), wq, work);
+	put_cpu_light();
 
 	return ret;
 }
-- 
2.0.1

