From cd4ac9b7a1c80cfcd9b8e76a1f6855a0329d982f Mon Sep 17 00:00:00 2001
From: Thomas Gleixner <tglx@linutronix.de>
Date: Tue, 16 Mar 2010 13:52:03 -0700
Subject: [PATCH] x86: Fix iomap_atomic with highmem=y again

commit cd4ac9b7a1c80cfcd9b8e76a1f6855a0329d982f in tip.

This whole highmem mess is racking my nerves !

Signed-off-by: Thomas Gleixner <tglx@linutronix.de>

diff --git a/arch/x86/Kconfig b/arch/x86/Kconfig
index f039584..ec97339 100644
--- a/arch/x86/Kconfig
+++ b/arch/x86/Kconfig
@@ -2081,7 +2081,7 @@ endmenu
 
 config HAVE_ATOMIC_IOMAP
 	def_bool y
-	depends on X86_32 && !PREEMPT_RT
+	depends on X86_32 && (!PREEMPT_RT || HIGHMEM)
 
 source "net/Kconfig"
 
-- 
1.7.1.1

