From e89fab2bf44afccaa48ffb8ae33c03f4f3dcd208 Mon Sep 17 00:00:00 2001
From: Thomas Gleixner <tglx@linutronix.de>
Date: Wed, 17 Mar 2010 15:47:14 -0700
Subject: [PATCH] x86: Fix the highmem fallout really

commit e89fab2bf44afccaa48ffb8ae33c03f4f3dcd208 in tip.

Bah, forgot to fix iomap_32.c along with the Kconfig

Signed-off-by: Thomas Gleixner <tglx@linutronix.de>

diff --git a/arch/x86/mm/iomap_32.c b/arch/x86/mm/iomap_32.c
index c889bad..d4fe51a 100644
--- a/arch/x86/mm/iomap_32.c
+++ b/arch/x86/mm/iomap_32.c
@@ -97,8 +97,7 @@ void __kunmap_atomic(void *kvaddr, enum km_type type)
 }
 EXPORT_SYMBOL(__kunmap_atomic);
 
-
-#ifndef CONFIG_PREEMPT_RT
+#if !defined(CONFIG_PREEMPT_RT) || defined(CONFIG_HIGHMEM)
 
 # ifndef CONFIG_HIGHMEM
 #  define kmap_atomic_prot_pfn(pfn, type, prot) \
-- 
1.7.1.1

