From f32e8221cca6cc5401b247e63f8a331209d68bdd Mon Sep 17 00:00:00 2001
From: Bruce Ashfield <bruce.ashfield@windriver.com> 
Date: Tue, 10 Jun 2008 21:58:35 -0400
Subject: [PATCH] Ensures that the image targets don't get build in parallel

Signed-off-by: Bruce Ashfield <bruce.ashfield@windriver.com>
---
 arch/arm/Makefile |    3 +++
 1 files changed, 3 insertions(+), 0 deletions(-)

diff --git a/arch/arm/Makefile b/arch/arm/Makefile
index 40a0b44..9481807 100644
--- a/arch/arm/Makefile
+++ b/arch/arm/Makefile
@@ -230,6 +230,9 @@ maketools: include/linux/version.h include/asm-arm/.arch FORCE
 # Convert bzImage to zImage
 bzImage: zImage
 
+# These targets cannot be built in parallel
+.NOTPARALLEL: zImage Image xipImage bootpImage uImage
+
 zImage Image xipImage bootpImage uImage: vmlinux
 	$(Q)$(MAKE) $(build)=$(boot) MACHINE=$(MACHINE) $(boot)/$@
 
-- 
1.5.5.1

