From d4abcab55fb4d0ea9b92759f6abdcc2e6117f88c Mon Sep 17 00:00:00 2001
From: Bruce Ashfield <bruce.ashfield@windriver.com>
Date: Thu, 19 Jun 2008 11:03:07 -0400
Subject: [PATCH] exclude wrs directory from distclean processing

The wrs directory has some files (notably files of size 0) that are
removed by Kbuild's distclean processing. Add wrs to the exclude 
list to prevent recursion into wrs.

Signed-off-by: Bruce Ashfield <bruce.ashfield@windriver.com>

---
 Makefile |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/Makefile b/Makefile
index 2b95120..038e39a 100644
--- a/Makefile
+++ b/Makefile
@@ -355,7 +355,7 @@ export MODVERDIR := $(if $(KBUILD_EXTMOD),$(firstword $(KBUILD_EXTMOD))/).tmp_ve
 
 # Files to ignore in find ... statements
 
-RCS_FIND_IGNORE := \( -name SCCS -o -name BitKeeper -o -name .svn -o -name CVS -o -name .pc -o -name .hg -o -name .git \) -prune -o
+RCS_FIND_IGNORE := \( -name SCCS -o -name BitKeeper -o -name .svn -o -name CVS -o -name .pc -o -name .hg -o -name .git -o -name wrs \) -prune -o
 export RCS_TAR_IGNORE := --exclude SCCS --exclude BitKeeper --exclude .svn --exclude CVS --exclude .pc --exclude .hg --exclude .git
 
 # ===========================================================================
-- 
1.5.5.1

