From 20e0f7949e28acc2643dc28a5b472b7bbf28bfc8 Mon Sep 17 00:00:00 2001
From: Paul Gortmaker <paul.gortmaker@windriver.com>
Date: Sun, 25 Jan 2009 17:58:09 -0500
Subject: [PATCH 2/8] kallsyms: mask trivial warning

Newer HOSTCC will complain about various stdio fcns because
they are now declared with "warn_unused_return" -- trivially
assign the return value so we aren't overwhelmed with these
once the kernel build kicks off.

Signed-off-by: Paul Gortmaker <paul.gortmaker@windriver.com>
---
 scripts/kallsyms.c |    3 ++-
 1 files changed, 2 insertions(+), 1 deletions(-)

diff --git a/scripts/kallsyms.c b/scripts/kallsyms.c
index 86c3896..e7d83fe 100644
--- a/scripts/kallsyms.c
+++ b/scripts/kallsyms.c
@@ -108,8 +108,9 @@ static int read_symbol(FILE *in, struct sym_entry *s)
 	rc = fscanf(in, "%llx %c %499s\n", &s->addr, &stype, str);
 	if (rc != 3) {
 		if (rc != EOF) {
+			char *c;
 			/* skip line */
-			fgets(str, 500, in);
+			c = fgets(str, 500, in);
 		}
 		return -1;
 	}
-- 
1.6.5.2

