From c9dfad6a04b10cee65a9a2bc2339bb3a509c75e4 Mon Sep 17 00:00:00 2001
From: Paul Gortmaker <paul.gortmaker@windriver.com>
Date: Mon, 12 Apr 2010 09:57:10 +0800
Subject: [PATCH 7/8] kbuild: no git IDs from random parent trees

git will walk upwards looking for a .git to parse. Ensure
that we don't fold in an ID unless it is from a .git relating to this
actual tree.

Signed-off-by: Paul Gortmaker <paul.gortmaker@windriver.com>
---
 scripts/setlocalversion |    4 ++++
 1 files changed, 4 insertions(+), 0 deletions(-)

diff --git a/scripts/setlocalversion b/scripts/setlocalversion
index 46989b8..3e75808 100755
--- a/scripts/setlocalversion
+++ b/scripts/setlocalversion
@@ -17,6 +17,10 @@ usage() {
 cd "${1:-.}" || usage
 
 # Check for git and a git repo.
+if [ ! -d .git ]; then
+	exit 0
+fi
+
 if head=`git rev-parse --verify --short HEAD 2>/dev/null`; then
 
 	# If we are at a tagged commit (like "v2.6.30-rc6"), we ignore it,
-- 
1.6.5.2

