From 19933a599e4be8c9c84dc58948e4052938a93056 Mon Sep 17 00:00:00 2001
From: Bruce Ashfield <bruce.ashfield@windriver.com>
Date: Mon, 12 Apr 2010 09:57:10 +0800
Subject: [PATCH 8/8] kbuild: Allow quilt patches directory to be ignored

If the variable QUILT_PC is non-zero add it to the list
of directories / files to be ignored.

Signed-off-by: Bruce Ashfield <bruce.ashfield@windriver.com>
---
 Makefile |    5 ++++-
 1 files changed, 4 insertions(+), 1 deletions(-)

diff --git a/Makefile b/Makefile
index 9250f06..1720385 100644
--- a/Makefile
+++ b/Makefile
@@ -374,7 +374,10 @@ export MODVERDIR := $(if $(KBUILD_EXTMOD),$(firstword $(KBUILD_EXTMOD))/).tmp_ve
 
 # Files to ignore in find ... statements
 
-RCS_FIND_IGNORE := \( -name SCCS -o -name BitKeeper -o -name .svn -o -name CVS -o -name .pc -o -name .hg -o -name .git -o -name wrs \) -prune -o
+ifneq (,${QUILT_PC})
+	QUILT_IGNORE := -o -name ${QUILT_PC}
+endif
+RCS_FIND_IGNORE := \( -name SCCS -o -name BitKeeper -o -name .svn -o -name CVS -o -name .pc ${QUILT_IGNORE} -o -name .hg -o -name .git -o -name wrs \) -prune -o
 export RCS_TAR_IGNORE := --exclude SCCS --exclude BitKeeper --exclude .svn --exclude CVS --exclude .pc --exclude .hg --exclude .git
 
 # ===========================================================================
-- 
1.6.5.2

