From 64543217119dfadd0a9e2bc53a3e4457a3388f0a Mon Sep 17 00:00:00 2001
From: Bruce Ashfield <bruce.ashfield@windriver.com>
Date: Sun, 15 Feb 2009 20:23:26 -0500
Subject: [PATCH] kbuild: add Wind River creates files to the ignore list

Signed-off-by: Bruce Ashfield <bruce.ashfield@windriver.com>
---
 .gitignore |    2 ++
 1 files changed, 2 insertions(+), 0 deletions(-)

diff --git a/.gitignore b/.gitignore
index 36d8bd7..defd068 100644
--- a/.gitignore
+++ b/.gitignore
@@ -64,3 +64,5 @@ ncscope.*
 *.orig
 *~
 \#*#
+LDAT_LINUX_PATCHES
+wrs
-- 
1.6.0.3

