From f7048040365cda15e3aaff7d5fbeb80a768695e6 Mon Sep 17 00:00:00 2001
From: Bruce Ashfield <bruce.ashfield@windriver.com>
Date: Sun, 15 Feb 2009 20:23:26 -0500
Subject: [PATCH 3/8] kbuild: add Wind River creates files to the ignore list

Signed-off-by: Bruce Ashfield <bruce.ashfield@windriver.com>
---
 .gitignore |    2 ++
 1 files changed, 2 insertions(+), 0 deletions(-)

diff --git a/.gitignore b/.gitignore
index 52dca6f..5fcb216 100644
--- a/.gitignore
+++ b/.gitignore
@@ -76,3 +76,5 @@ GTAGS
 *.orig
 *~
 \#*#
+/LDAT_LINUX_PATCHES
+/wrs
-- 
1.6.5.2

