From 693dbbb55f7e3e5651354c45c6f0b177a83e611d Mon Sep 17 00:00:00 2001
From: auto commit <unknown@unknown>
Date: Tue, 10 Jun 2008 23:19:54 -0400
Subject: [PATCH] , git will walk upwards looking for a .git to parse. Ensure

that we don't fold in an ID unless it is from a .git relating to this
actual tree.

Signed-off-by: Paul Gortmaker <paul.gortmaker@windriver.com>
---
 scripts/setlocalversion |    4 ++++
 1 files changed, 4 insertions(+), 0 deletions(-)

diff --git a/scripts/setlocalversion b/scripts/setlocalversion
index 1c1bdaf..3aa57f1 100755
--- a/scripts/setlocalversion
+++ b/scripts/setlocalversion
@@ -9,6 +9,10 @@ usage() {
 cd "${1:-.}" || usage
 
 # Check for git and a git repo.
+if [ ! -d .git ]; then
+	exit 0
+fi
+
 if head=`git rev-parse --verify HEAD 2>/dev/null`; then
 	# Do we have an untagged version?
 	if git name-rev --tags HEAD | grep -E '^HEAD[[:space:]]+(.*~[0-9]*|undefined)$' > /dev/null; then
-- 
1.5.5.1

