From bd9edd8647ddadd82bf349e7b8476be425763194 Mon Sep 17 00:00:00 2001
From: Robert Yang <liezhi.yang@windriver.com>
Date: Fri, 13 Sep 2013 22:01:02 -0400
Subject: [PATCH] scripts/Makefile.headersinst: install headers from scratch
 file

If headers_install is executed from a deep/long directory structure, the
shell's maximum argument length can be execeeded, which breaks the operation
with:

| make[2]: execvp: /bin/sh: Argument list too long
| make[2]: ***

By dumping the input files to a scratch file and using xargs to read the
input list from the scratch file, we can avoid blowing out the maximum
argument size and install headers in a long path name environment.

Robert:
Use "$(foreach ,,$(shell echo))" rather than "echo" to make it work when
len(TMPDIR) <= 410.

Upstream-Status: Inappropriate [oe-specific]

Signed-off-by: Bruce Ashfield <bruce.ashfield@windriver.com>
Signed-off-by: Robert Yang <liezhi.yang@windriver.com>

CQID: WIND00443051

Signed-off-by: Kai Kang <kai.kang@windriver.com>
---
 scripts/Makefile.headersinst | 9 ++++++++-
 1 file changed, 8 insertions(+), 1 deletion(-)

diff --git a/scripts/Makefile.headersinst b/scripts/Makefile.headersinst
index 182084d..30b7463 100644
--- a/scripts/Makefile.headersinst
+++ b/scripts/Makefile.headersinst
@@ -72,7 +72,7 @@ printdir = $(patsubst $(INSTALL_HDR_PATH)/%/,%,$(dir $@))
 quiet_cmd_install = INSTALL $(printdir) ($(words $(all-files))\
                             file$(if $(word 2, $(all-files)),s))
       cmd_install = \
-        $(CONFIG_SHELL) $< $(installdir) $(input-files); \
+        xargs $(CONFIG_SHELL) $< $(installdir) < $(INSTALL_HDR_PATH)/.input-files; \
         for F in $(wrapper-files); do                                   \
                 echo "\#include <asm-generic/$$F>" > $(installdir)/$$F;    \
         done;                                                           \
@@ -98,10 +98,17 @@ __headersinst: $(subdirs) $(install-file)
 	@:
 
 targets += $(install-file)
+# Use the $(foreach ,,$(shell echo)) to avoid the "Argument list too
+# long" error, and use the "$(shell echo)" rather than "echo" to keep
+# the order.
 $(install-file): scripts/headers_install.sh $(input-files) FORCE
 	$(if $(unwanted),$(call cmd,remove),)
 	$(if $(wildcard $(dir $@)),,$(shell mkdir -p $(dir $@)))
+	$(shell echo -n > $(INSTALL_HDR_PATH)/.input-files)
+	$(foreach f,$(input-files),$(shell echo -n "$(f) " >> \
+			$(INSTALL_HDR_PATH)/.input-files))
 	$(call if_changed,install)
+	@rm $(INSTALL_HDR_PATH)/.input-files
 
 else
 __headerscheck: $(subdirs) $(check-file)
-- 
1.8.4.93.g57e4c17

