Allow quilt patches directory to be ignored

If the variable QUILT_PC is non-zero add it to the list
of directories / files to be ignored.

Signed-off-by: Bruce Ashfield <bruce.ashfield@windriver.com>
diff --git a/Makefile b/Makefile
index 4693592..9d7aba2 100644
--- a/Makefile
+++ b/Makefile
@@ -363,7 +363,10 @@ export MODVERDIR := $(if $(KBUILD_EXTMOD),$(firstword $(KBUILD_EXTMOD))/).tmp_ve
 
 # Files to ignore in find ... statements
 
-RCS_FIND_IGNORE := \( -name SCCS -o -name BitKeeper -o -name .svn -o -name CVS -o -name .pc -o -name .hg -o -name .git -o -name wrs \) -prune -o
+ifneq (,${QUILT_PC})
+	QUILT_IGNORE := -o -name ${QUILT_PC}
+endif
+RCS_FIND_IGNORE := \( -name SCCS -o -name BitKeeper -o -name .svn -o -name CVS -o -name .pc ${QUILT_IGNORE} -o -name .hg -o -name .git -o -name wrs \) -prune -o
 export RCS_TAR_IGNORE := --exclude SCCS --exclude BitKeeper --exclude .svn --exclude CVS --exclude .pc --exclude .hg --exclude .git
 
 # ===========================================================================
