From 8230677ce3c07341b8b31dd4534b4764ddc6748e Mon Sep 17 00:00:00 2001
From: Robin Gong <b38343@freescale.com>
Date: Fri, 25 Dec 2015 10:02:44 +0800
Subject: [PATCH 1265/1691] MLK-12076-2: firmare: imx: sdma: update ecspi tx
 script for ERR008517

commit 8230677ce3c07341b8b31dd4534b4764ddc6748e from
git://git.freescale.com/imx/linux-2.6-imx.git imx_4.1.15_1.0.0_ga

Since ERR008517(TKT238285) fixed on i.mx6ul, we need change script to
compatible all i.mx6 chips.

Signed-off-by: Robin Gong <b38343@freescale.com>
(cherry picked from commit 6970e2b497621e3f2fc4005d06bb8ca76d234074)
---
 firmware/imx/sdma/sdma-imx6q.bin.ihex | 262 +++++++++++++++++-----------------
 1 file changed, 131 insertions(+), 131 deletions(-)

diff --git a/firmware/imx/sdma/sdma-imx6q.bin.ihex b/firmware/imx/sdma/sdma-imx6q.bin.ihex
index 5275e15..112bec4 100644
--- a/firmware/imx/sdma/sdma-imx6q.bin.ihex
+++ b/firmware/imx/sdma/sdma-imx6q.bin.ihex
@@ -1,139 +1,139 @@
-:1000000053444D4103000000020000001C000000AA
-:1000100029000000C0000000D2070000820200009A
+:1000000053444D4103000000030000001C000000A9
+:1000100029000000C0000000D80700008202000094
 :10002000FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFE0
 :10003000FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFD0
-:10004000FFFFFFFFFFFFFFFF3F1A0000FFFFFFFF63
-:10005000EB020000B1180000FFFFFFFF9D1A000037
+:10004000FFFFFFFFFFFFFFFF421A0000FFFFFFFF60
+:10005000EB020000B4180000FFFFFFFFA01A000031
 :10006000FFFFFFFFC0030000FFFFFFFFFFFFFFFFD9
 :10007000FFFFFFFFAB020000FFFFFFFF7B0300005D
 :10008000FFFFFFFFFFFFFFFF4C0400006E040000B6
 :10009000FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF70
-:1000A000FFFFFFFF0018000058180000EB190000C8
-:1000B0004C1B0000E71A0000831B00000018000022
+:1000A000FFFFFFFF001800005B180000EE190000C2
+:1000B0004F1B0000EA1A0000861B00000018000019
 :1000C000E3C1DB57F352016AFB52D36AFB521C1A9D
-:1000D000C36AE8621102FF3A3008D002337C8F0015
-:1000E000D500017D8D00A005EB5D7804037D7904CA
-:1000F0001C7D207C7904157CEE566006057D096523
-:10010000317E0A622F7E28980A622C7E09652A7E3B
-:1001100012051205AD02C86A257F03200048ED7C58
-:10012000409802780962C86A3F98150002780A620E
-:10013000C86A3F981500150002780B62C86A127CE5
-:10014000DF6D107FFB52041AC36AD36AC862042AA7
-:10015000C86AFB52D36A0000EB55004D077DFAC117
-:10016000DB5706980700CC680C6813C20AC20198D6
-:10017000D9C1E3C1DB57E35FE357F352216A8F0034
-:10018000D500017D8D00A005EB5DFB567804037D55
-:1001900079042A7D317C7904207C700B1103EB53A8
-:1001A000000F6003057D0965377E0A62357E7C9805
-:1001B0000A62327E0965307E12051205AD026007C3
-:1001C000027C065A8498265A277F011F0320004884
-:1001D000E87C700B11031353A5981500047809628D
-:1001E000065A0962265AA4981500150004780A6276
-:1001F000065A0A62265AA4981500150015000478BC
-:100200000B62065A0B62265A077C0000EB55004D24
-:10021000067DFAC1E3575F9807000C6813C20AC253
-:100220005C98700B110313536C07017CD9C1FB5E02
-:100230008A066B07017CD9C1F35EDB59D3588F0165
-:1002400010010F398B003CC12B7DC05AC85B4EC1D9
-:10025000277C88038906E35CFF0D1105FF1DBC05A3
-:100260003E07004D187D700811007E07097D7D074F
-:10027000027D2852DC98F852DB54BC02CC02097C87
-:100280007C07027D2852E598F852D354BC02CC0278
-:10029000097D0004D3988B00C052C85359C1D67D44
-:1002A0000002C398FF08BF007F07157D8804D500B2
-:1002B000017D8D00A005EB5D8F0212021202FF3A54
-:1002C000DA05027C3E071099A402DD02027D3E079A
-:1002D00010995E071099EB55004D027C9805EB5D77
-:1002E000F352FB546A07387D6C07017D5F996B07F9
-:1002F0007D7C6907047D6807027D010E279993586C
-:10030000D600017D8E009355A005935DA0066D0774
-:10031000067D6E070B7D02780255045D3C99150041
-:1003200004780255045D2255045D3C99150004785B
-:100330000255045D0255245D1D7C004E087C690752
-:10034000037D0255177E4699045D147F89069350FC
-:100350000048017D1F99D299150006780255045D69
-:100360004F070255245D2F07017CD29917006F07B4
-:10037000017C012093559D000700D9D9EB986907AE
-:10038000047D6807027D010E6D999358D600017DAA
-:100390008E009355A005935DA0066D07077D6E073F
-:1003A0000D7DD36C02780255C86D85991500D36C0C
-:1003B00004780255C86D2255C86D859915000578D9
-:1003C000D76C0255C86D0255C86D0F7C004E087C75
-:1003D0006907037D0255097E8F99C86D067F8906DE
-:1003E00093500048017D6599D299CC996907047DA5
-:1003F0006807027D010EA4999358D600017D8E00F6
-:100400009355A005935DA0066D07077D6E070D7DD2
-:10041000C36A0278C865045DBC9915000578C76A8F
-:10042000C865045DC865045DBC991500C36A04789D
-:10043000C865045DC865245D0F7C004E087C6907B3
-:10044000037DC865097EC699045D067F89069350C1
-:100450000048017D9C99D29993559D000700FF6C3F
-:10046000D9D9EB980000E354EB55004D017CEB9893
-:10047000D398E354EB55FF0A1102FF1A7F07027C61
-:10048000A005E6999D008C05BA05A0051002BA04E6
-:10049000AD0454040600D9C1E3C1DB57F352056A29
-:1004A0008F00D500017D8D00A005FB567804037DEB
-:1004B0007904297D1F7C79042E7CE35D700D110584
-:1004C000ED55000F6007027D0652079A2652337ED3
-:1004D0006005027D10021202096A2D7F1202096A6C
-:1004E0002A7F1202096A277F011F03200048EA7C45
-:1004F000E355329A150015001500047806520B6A70
-:1005000026520B6A319A15001500047806520A6AC1
-:1005100026520A6A319A150004780652096A265250
-:10052000096A097C286A077F0000DB57004D057DBA
-:10053000FAC1DB57F09977C254040AC2ED99E3C1BE
-:10054000DB57F352056AFB568E02941AC36AC862DF
-:100550006902267D941EC36ED36EC8624802C86AC3
-:100560009426981EC36EC8629826C36E60020E7DE4
-:10057000981EC36EC8626C02037D9826C36E809A73
-:100580004C02D36EC86A9826C36E909AC8626E02F7
-:10059000017C8A9A096A1A7F0125004D217D519AB2
-:1005A000E36E8F00D805017D8D00C8626E02127D5A
-:1005B000096A0C7F01250120F87CDB57004D107D76
-:1005C000286A047F0000FAC1DB57439A07000462DF
-:1005D0000C6A8D9A286AFA7F04627AC2580454041D
-:1005E000FF081100FF18BC00CD00017C809A286A2A
-:1005F000ED7F04627AC20AC2409AD9C1E3C1DB57D7
-:10060000F352056AFB568E02941A025269021D7D4E
-:10061000941E06524802065A9426981E06524C0210
-:10062000065A9826981E065260020A7C98260652A0
-:100630006E02237D096A1D7F0125004D247DB29A3B
-:10064000286A177F04627AC2E39A8F00D805017D79
-:100650008D00A00506526E02107D096A0A7F0120F6
-:10066000F97C286A067F0000004D0D7DFAC1DB573A
-:10067000A29A070004620C6AE09A286AFA7F046270
-:100680007AC258045404286AF47F0AC29F9A2BDB6A
-:10069000DB57F352056AC7698F00D500017D8D00D5
-:1006A000A0057804037D79041C7D157C79041E7CEB
-:1006B000EE56C862287E6006027D10021202096AA8
-:1006C000227F1202096A1F7F1202096A1C7F03201F
-:1006D0000048EF7C1A9B150015000278C8620B6A6F
-:1006E000199B15000278C8620A6A199B0278C862D1
-:1006F000096A097C286A077F004D077DD3520108EB
-:1007000002580004EB9A77C25404D35201080258ED
-:10071000010443DBE89A700B11031353DB5F0A07F4
-:10072000D3588B00FB5E3CC1157DC05AC85B4EC1DF
-:10073000117C8803F05DFF0D1105FF1DBC05004D08
-:10074000047D429B0807DB5F0A078B00C352CB5333
-:1007500059C1EC7D0002339B016E0B612F7E0B6251
-:100760002D7E0B632B7E0C0D1704170417049D04BC
-:10077000081DCC05017C0C0DD16A000F4207C86F23
-:10078000DD6F1C7F8E009D0001680B67177ED56BA7
-:1007900004080278C86F1207117C0B670F7E0408EB
-:1007A0000278C86F12070A7CDD6F087FD169010FDC
-:1007B000C86FDD6F037F01010004589B0700FF68CD
-:1007C0000C680002589BC8DBDB57F352056AFB52EA
-:1007D000C76A8F00D500017D8D00A0057804037DD8
-:1007E00079041C7D157C79041E7CEE56C862287E37
-:1007F0006006027D10021202096A227F1202096A53
-:100800001F7F1202096A1C7F03200048EF7CB79B00
-:10081000150015000278C8620B6AB69B15000278B5
-:10082000C8620A6AB69B0278C862096A097C286AAB
-:10083000077F004D077DD352010802580004879BB3
-:1008400077C25404D352010802580104E0DB849BB0
-:10085000700B11031353DB5F0A07D3588B00FB5E49
-:100860003CC1157DC05AC85B4EC1117C8803F05D48
-:10087000FF0D1105FF1DBC05004D047DDF9B080722
-:10088000DB5F0A078B00C352CB5359C1EC7D0002DA
-:02089000D09BFB
+:1000D000C36AE8621102FF3A3008D0023F7C8F0009
+:1000E0003F00D500017D8D00A005EB5D7804037D08
+:1000F00079041C7D207C7904157CEE566006057D14
+:100100000965337E0A62317E29980A622E7E09656E
+:100110002C7E12051205AD02C86A277F0320004815
+:10012000ED7C419802780962C86A4098150002780F
+:100130000A62C86A40981500150002780B62C86A06
+:10014000147CDF6D127F7F07097DFB52041AC36A9E
+:10015000D36AC862042AC86AFB52D36A0000EB550E
+:10016000004D077DFAC1DB5706980700CC680C6884
+:1001700013C20AC20198D9C1E3C1DB57E35FE35759
+:10018000F352216A8F00D500017D8D00A005EB5D43
+:10019000FB567804037D79042A7D317C7904207C28
+:1001A000700B1103EB53000F6003057D0965377E6B
+:1001B0000A62357E7F980A62327E0965307E1205BA
+:1001C0001205AD026007027C065A8798265A277FDF
+:1001D000011F03200048E87C700B11031353A898FB
+:1001E000150004780962065A0962265AA798150074
+:1001F000150004780A62065A0A62265AA798150062
+:100200001500150004780B62065A0B62265A077C0B
+:100210000000EB55004D067DFAC1E35762980700D8
+:100220000C6813C20AC25F98700B110313536C075A
+:10023000017CD9C1FB5E8A066B07017CD9C1F35EE4
+:10024000DB59D3588F0110010F398B003CC12B7D36
+:10025000C05AC85B4EC1277C88038906E35CFF0D4A
+:100260001105FF1DBC053E07004D187D70081100EB
+:100270007E07097D7D07027D2852DF98F852DB5406
+:10028000BC02CC02097C7C07027D2852E898F85217
+:10029000D354BC02CC02097D0004D6988B00C05216
+:1002A000C85359C1D67D0002C698FF08BF007F071A
+:1002B000157D8804D500017D8D00A005EB5D8F02C2
+:1002C00012021202FF3ADA05027C3E071399A402D9
+:1002D000DD02027D3E0713995E071399EB55004D31
+:1002E000027C9805EB5DF352FB546A07387D6C077E
+:1002F000017D62996B077D7C6907047D6807027D3B
+:10030000010E2A999358D600017D8E009355A005C1
+:10031000935DA0066D07067D6E070B7D0278025582
+:10032000045D3F99150004780255045D2255045D73
+:100330003F99150004780255045D0255245D1D7C2B
+:10034000004E087C6907037D0255177E4999045DBC
+:10035000147F890693500048017D2299D599150094
+:1003600006780255045D4F070255245D2F07017C76
+:10037000D59917006F07017C012093559D00070058
+:10038000DCD9EE986907047D6807027D010E70993B
+:100390009358D600017D8E009355A005935DA0066D
+:1003A0006D07077D6E070D7DD36C02780255C86D11
+:1003B00088991500D36C04780255C86D2255C86D14
+:1003C000889915000578D76C0255C86D0255C86D1F
+:1003D0000F7C004E087C6907037D0255097E9299C7
+:1003E000C86D067F890693500048017D6899D599AC
+:1003F000CF996907047D6807027D010EA79993587C
+:10040000D600017D8E009355A005935DA0066D0773
+:10041000077D6E070D7DC36A0278C865045DBF99CC
+:1004200015000578C76AC865045DC865045DBF9995
+:100430001500C36A0478C865045DC865245D0F7C37
+:10044000004E087C6907037DC865097EC999045D73
+:10045000067F890693500048017D9F99D599935551
+:100460009D000700FF6CDCD9EE980000E354EB55CB
+:10047000004D017CEE98D698E354EB55FF0A11022B
+:10048000FF1A7F07027CA005E9999D008C05BA053B
+:10049000A0051002BA04AD0454040600D9C1E3C19A
+:1004A000DB57F352056A8F00D500017D8D00A00552
+:1004B000FB567804037D7904297D1F7C79042E7C0A
+:1004C000E35D700D1105ED55000F6007027D0652CA
+:1004D0000A9A2652337E6005027D10021202096AD2
+:1004E0002D7F1202096A2A7F1202096A277F011FE3
+:1004F00003200048EA7CE355359A150015001500E5
+:10050000047806520B6A26520B6A349A15001500BD
+:10051000047806520A6A26520A6A349A1500047848
+:100520000652096A2652096A097C286A077F000078
+:10053000DB57004D057DFAC1DB57F39977C25404B0
+:100540000AC2F099E3C1DB57F352056AFB568E02EB
+:10055000941AC36AC8626902267D941EC36ED36E64
+:10056000C8624802C86A9426981EC36EC86298265C
+:10057000C36E60020E7D981EC36EC8626C02037D5E
+:100580009826C36E839A4C02D36EC86A9826C36EAF
+:10059000939AC8626E02017C8D9A096A1A7F0125BE
+:1005A000004D217D549AE36E8F00D805017D8D00AA
+:1005B000C8626E02127D096A0C7F01250120F87C59
+:1005C000DB57004D107D286A047F0000FAC1DB571D
+:1005D000469A070004620C6A909A286AFA7F0462BD
+:1005E0007AC258045404FF081100FF18BC00CD0063
+:1005F000017C839A286AED7F04627AC20AC2439A18
+:10060000D9C1E3C1DB57F352056AFB568E02941A37
+:10061000025269021D7D941E06524802065A942613
+:10062000981E06524C02065A9826981E06526002E0
+:100630000A7C982606526E02237D096A1D7F0125D9
+:10064000004D247DB59A286A177F04627AC2E69A23
+:100650008F00D805017D8D00A00506526E02107D29
+:10066000096A0A7F0120F97C286A067F0000004D94
+:100670000D7DFAC1DB57A59A070004620C6AE39A64
+:10068000286AFA7F04627AC258045404286AF47F04
+:100690000AC2A29A2EDBDB57F352056AC7698F00A4
+:1006A000D500017D8D00A0057804037D79041C7DB3
+:1006B000157C79041E7CEE56C862287E6006027D99
+:1006C00010021202096A227F1202096A1F7F1202B7
+:1006D000096A1C7F03200048EF7C1D9B1500150054
+:1006E0000278C8620B6A1C9B15000278C8620A6A0D
+:1006F0001C9B0278C862096A097C286A077F004D42
+:10070000077DD352010802580004EE9A77C25404C0
+:10071000D35201080258010446DBEB9A700B110317
+:100720001353DB5F0A07D3588B00FB5E3CC1157D7A
+:10073000C05AC85B4EC1117C8803F05DFF0D1105E6
+:10074000FF1DBC05004D047D459B0807DB5F0A07C4
+:100750008B00C352CB5359C1EC7D0002369B016E16
+:100760000B612F7E0B622D7E0B632B7E0C0D17040D
+:10077000170417049D04081DCC05017C0C0DD16ADB
+:10078000000F4207C86FDD6F1C7F8E009D0001685F
+:100790000B67177ED56B04080278C86F1207117CAF
+:1007A0000B670F7E04080278C86F12070A7CDD6FA2
+:1007B000087FD169010FC86FDD6F037F010100045D
+:1007C0005B9B0700FF680C6800025B9BCBDBDB5781
+:1007D000F352056AFB52C76A8F00D500017D8D0078
+:1007E000A0057804037D79041C7D157C79041E7CAA
+:1007F000EE56C862287E6006027D10021202096A67
+:10080000227F1202096A1F7F1202096A1C7F0320DD
+:100810000048EF7CBA9B150015000278C8620B6A8D
+:10082000B99B15000278C8620A6AB99B0278C8624F
+:10083000096A097C286A077F004D077DD3520108A9
+:10084000025800048A9B77C25404D352010802580C
+:100850000104E3DB879B700B11031353DB5F0A0773
+:10086000D3588B00FB5E3CC1157DC05AC85B4EC19E
+:10087000117C8803F05DFF0D1105FF1DBC05004DC7
+:10088000047DE29B0807DB5F0A078B00C352CB5352
+:0808900059C1EC7D0002D39B6D
 :00000001FF
-- 
1.9.1

