From 245f7fd3746beb06db158510d426b75762491712 Mon Sep 17 00:00:00 2001
From: zumeng chen <zchen@pek-cc-pb02l.(none)>
Date: Fri, 20 Jun 2008 19:55:16 +0800
Subject: [PATCH 1/1] EXAR XR17C152 init

This patch addes pci exar init routine into
8250_pci.c, which comes from wrlinux-2.0 bsp.
---
 drivers/serial/8250_pci.c |   61 ++++++++++++++++++++++++++++++++++++++++++++-
 1 files changed, 60 insertions(+), 1 deletions(-)

diff --git a/drivers/serial/8250_pci.c b/drivers/serial/8250_pci.c
index 788c355..6d14d0f 100644
--- a/drivers/serial/8250_pci.c
+++ b/drivers/serial/8250_pci.c
@@ -564,6 +564,53 @@ pci_timedia_setup(struct serial_private *priv, struct pciserial_board *board,
 
 	return setup_port(priv, port, bar, offset, board->reg_shift);
 }
+/*
+ *
+ * EXAR XR17C152 setup routine
+ *
+ */
+static int __devinit pci_exar_init(struct pci_dev *dev)
+{
+	u8 *p;
+	u8 reg;
+	u16 i; 
+
+	if ((pci_resource_flags(dev, 0) & IORESOURCE_MEM) == 0) {
+		moan_device("no memory in bar 0", dev);
+		return 0;
+	}
+
+	p = ioremap(pci_resource_start(dev, 0), 0x80);
+	if (p == NULL)
+		return -ENOMEM;
+
+	/*
+	 * soft reset uart channels  (make sure in 16550 mode)
+	 */ 
+	writeb(0x03, (unsigned long)p + 0x8A);
+
+	/*
+	 * Set baud clock prescaler to divide by 4 mode 
+	 */
+
+	for ( i = 0; i <= 0x200; i += 0x200 ) {
+	  /* set EFR Bit 4 */
+	  reg = readb( (unsigned long)p + i + 0x09 );
+	  writeb( (reg|0x10), ((unsigned long)p + i + 0x09));
+
+	  /* set MCR Bit 7 */
+	  reg = readb( (unsigned long)p + i + 0x04 );
+	  writeb( (reg|0x80), ((unsigned long)p + i + 0x04) );
+	  
+	  /* clear EFR Bit 4 */
+	  reg = readb( (unsigned long)p + i + 0x09 );
+	  writeb( (reg&0xef), ((unsigned long)p + i + 0x09) );
+	}
+
+	iounmap(p);
+
+	return 0;
+}
 
 /*
  * Some Titan cards are also a little weird
@@ -1007,6 +1054,18 @@ static struct pci_serial_quirk pci_serial_quirks[] __refdata = {
 		.init		= pci_netmos_init,
 		.setup		= pci_default_setup,
 	},
+
+	/* 
+	 * EXAR XR17C152
+	 */
+	{
+	  .vendor         = PCI_VENDOR_ID_EXAR,
+	  .device         = PCI_DEVICE_ID_EXAR_XR17C152,
+	  .subvendor      = PCI_ANY_ID,
+	  .subdevice      = PCI_ANY_ID,
+	  .init           = pci_exar_init,
+	  .setup          = pci_default_setup,
+	},
 	/*
 	 * Default "match everything" terminator entry
 	 */
@@ -1679,7 +1738,7 @@ static struct pciserial_board pci_boards[] __devinitdata = {
 	[pbn_exar_XR17C152] = {
 		.flags		= FL_BASE0,
 		.num_ports	= 2,
-		.base_baud	= 921600,
+		.base_baud	= 230400,
 		.uart_offset	= 0x200,
 	},
 	[pbn_exar_XR17C154] = {
-- 
1.5.5.1.67.gbdb8

