From ba53df0127a9ab86901ea2de3b67d4c1a2c3614d Mon Sep 17 00:00:00 2001
From: auto commit <unknown@unknown>
Date: Tue, 22 Jul 2008 16:25:46 -0400
Subject: [PATCH] auto_msg: importing squashfs-fixups.patch

This is an automatic import of patch squashfs-fixups.patch, no headers were
detected and a default message was constructed
---
 fs/squashfs/inode.c |    5 +++--
 1 files changed, 3 insertions(+), 2 deletions(-)

diff --git a/fs/squashfs/inode.c b/fs/squashfs/inode.c
index 911ec11..9594fc4 100644
--- a/fs/squashfs/inode.c
+++ b/fs/squashfs/inode.c
@@ -83,7 +83,8 @@ static struct super_operations squashfs_export_super_ops = {
 	.destroy_inode = squashfs_destroy_inode,
 	.statfs = squashfs_statfs,
 	.put_super = squashfs_put_super,
-	.read_inode = vfs_read_inode
+	/* removed by commit 12debc4248a4a7f1873e47cda2cdd7faca80b099 */
+	/* .read_inode = vfs_read_inode */
 };
 
 static struct export_operations squashfs_export_ops = {
@@ -647,7 +648,7 @@ static void vfs_read_inode(struct inode *i)
 static struct dentry *squashfs_get_parent(struct dentry *child)
 {
 	struct inode *i = child->d_inode;
-	struct inode *parent = iget(i->i_sb, SQUASHFS_I(i)->u.s2.parent_inode);
+	struct inode *parent = iget_locked(i->i_sb, SQUASHFS_I(i)->u.s2.parent_inode);
 	struct dentry *rv;
 
 	TRACE("Entered squashfs_get_parent\n");
-- 
1.5.5.1

