From 594024bd9af2d332d9a2a7d22a53a77e6a21b141 Mon Sep 17 00:00:00 2001
From: Xulin Sun <xulin.sun@windriver.com>
Date: Mon, 2 Feb 2015 14:28:28 +0800
Subject: [PATCH 2/4] crypto:talitos: pass the same dev to
 dma_{map,unmap}_single

This patch refers to another caam commit
"crypto: caam: pass the same dev to dma_{map,unmap}_single".

In the current code, two struct device are used to map and umap dma
memories respectively. One is embedded in platform_device, and the
other is the one embedded in net_device. The content of these devices
are the same, but it will trigger a warning when CONFIG_DMA_API_DEBUG
is enabled:

talitos ffe30000.crypto: DMA-API: device driver tries to free DMA memory it has not allocated [device address=0x000000000588d028] [size=64 bytes]
WARNING: at linux-windriver/3.10-r0/linux/lib/dma-debug.c:877
Modules linked in:
CPU: 1 PID: 87 Comm: irq/58-talitos Not tainted 3.10.62-ltsi-WR6.0.0.0_cgl #50
task: c58ba7c0 ti: effe6000 task.ti: c58c0000
NIP: c0440624 LR: c0440624 CTR: c0490758
REGS: effe7d60 TRAP: 0700 Not tainted (3.10.62-ltsi-WR6.0.0.0_cgl)
MSR: 00021000 <CE,ME> CR: 22822088 XER: 20000000

GPR00: c0440624 effe7e10 c58ba7c0 00000091 00000001 c004c0e0 00000000 00000000
GPR08: c0b60000 00000000 c0b5c734 000001a3 22822084 00000000 c0b55080 00000000
GPR16: 00000000 efa15c00 c0b50000 00000000 c09d1e5c 00000000 c0b5d088 00000000
GPR24: 00021000 c10004c8 c0ae6794 c100f340 effe7e58 ffbf48c8 00000000 0588d028
NIP [c0440624] check_unmap+0x7a4/0x9b0
LR [c0440624] check_unmap+0x7a4/0x9b0
Call Trace:
[effe7e10] [c0440624] check_unmap+0x7a4/0x9b0 (unreliable)
[effe7e50] [c04408a8] debug_dma_unmap_page+0x78/0x8c
[effe7ed0] [c06620a8] flush_channel+0x194/0x284
[effe7f20] [c0662218] talitos_done_ch1_3+0x80/0x120
[effe7f40] [c069db40] net_rx_action+0x158/0x290
[effe7f90] [c0054064] __do_softirq+0x124/0x2c4
[effe7ff0] [c000ef60] call_do_softirq+0x14/0x24
[c58c1e20] [c00056cc] do_softirq+0xec/0x118
[c58c1e40] [c0053cf0] local_bh_enable+0xf8/0x110
[c58c1e50] [c00f51d8] irq_forced_thread_fn+0x5c/0x74
[c58c1e70] [c00f5514] irq_thread+0x124/0x154
[c58c1eb0] [c007693c] kthread+0xb4/0xb8
[c58c1f40] [c0011230] ret_from_kernel_thread+0x5c/0x64

Singed-off-by: Xulin Sun <xulin.sun@windriver.com>
---
 drivers/crypto/talitos.c |    1 +
 1 files changed, 1 insertions(+), 0 deletions(-)

diff --git a/drivers/crypto/talitos.c b/drivers/crypto/talitos.c
index 7520aa6..636e2ec 100644
--- a/drivers/crypto/talitos.c
+++ b/drivers/crypto/talitos.c
@@ -315,6 +315,7 @@ static int talitos_done_##name(struct napi_struct *napi, int budget)	\
 	int budget_per_ch, work_done = 0;				\
 	unsigned long flags;						\
 									\
+	dev = priv->dev;						\
 	budget_per_ch = budget / num_ch;				\
 	if (ch_done_mask & 1)						\
 		work_done += flush_channel(dev, 0, 0, 0, budget_per_ch);\
-- 
1.7.5.4

