From 6232cdf9db467a04b9e0bc94ffb6fe50c37ec8bd Mon Sep 17 00:00:00 2001
From: Yanjiang Jin <yanjiang.jin@windriver.com>
Date: Wed, 25 Feb 2015 10:32:18 +0800
Subject: [PATCH 09/10] crypto: caamalg_qi: fix DMA_MAP direction problem

Fix DMA_MAP direction problem for caamalg_qi to avoid the below calltrace:

platform caam_qi.0: DMA-API: device driver frees DMA memory with different direction [device address=0x00000000f765c000] [size=0 bytes] [mapped with DMA_TO_DEVICE] [unmapped with DMA_BIDIRECTIONAL]
------------[ cut here ]------------
WARNING: at lib/dma-debug.c:928
Modules linked in:
CPU: 1 PID: 0 Comm: swapper/1 Not tainted 3.10.62-ltsi-WR6.0.0.0_standard #187
task: c0000000f90c5500 ti: c0000000fffd0000 task.ti: c0000000f9194000
NIP: c0000000004f5a2c LR: c0000000004f5a28 CTR: c00000000055a160
REGS: c0000000fffd34a0 TRAP: 0700   Not tainted  (3.10.62-ltsi-WR6.0.0.0_standard)
MSR: 0000000080029000 <CE,EE,ME>  CR: 24008082  XER: 00000000
SOFTE: 0

004f5a28 c0000000fffd3720 c0000000012af360 00000000000000c5
00000000 0000000000000000 c00000000125f1c8 000000000000018f
00000060 0000000000000000 0000000000000000 c0000000012d85e8
0000018f c000000001fff780 c0000000011ac928 00000000fa337000
fffedcbe c0000000012b9080 0000000000000000 0000000000000001
fffd0000 ffffffffffffffff c00000000154d3c0 0000000000000002
00000000 c0000000f7511810 6db6db6db6db6db7 0000000000000001
01562710 c000000001565b80 c0000000fffd3830 c0000000f9427840
NIP [c0000000004f5a2c] .check_unmap+0x39c/0x9c0
LR [c0000000004f5a28] .check_unmap+0x398/0x9c0
Call Trace:
[c0000000fffd3720] [c0000000004f5a28] .check_unmap+0x398/0x9c0 (unreliable)
[c0000000fffd37c0] [c0000000004f62b4] .debug_dma_unmap_sg+0x114/0x180
[c0000000fffd3910] [c000000000831824] .aead_unmap+0xa4/0x6d0
[c0000000fffd39f0] [c00000000083022c] .aead_done+0x7c/0xd0
[c0000000fffd3bc0] [c000000000819de8] .caam_rsp_fq_dqrr_cb+0x188/0x250
[c0000000fffd3c50] [c00000000085fed4] .qman_p_poll_dqrr+0x1d4/0x2e0
[c0000000fffd3d10] [c000000000819670] .caam_qi_poll+0x30/0x80
[c0000000fffd3da0] [c0000000008be56c] .net_rx_action+0x1cc/0x330
[c0000000fffd3e80] [c00000000007276c] .__do_softirq+0x19c/0x3d0
[c0000000fffd3f90] [c000000000017054] .call_do_softirq+0x14/0x24
[c0000000f91978e0] [c000000000005fe8] .do_softirq+0x118/0x150
[c0000000f9197970] [c000000000072c54] .irq_exit+0x124/0x140
[c0000000f91979f0] [c000000000005ac4] .do_IRQ+0x184/0x370
[c0000000f9197aa0] [c00000000001b93c] exc_0x500_common+0xfc/0x100
---
 drivers/crypto/caam/caamalg_qi.c |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/drivers/crypto/caam/caamalg_qi.c b/drivers/crypto/caam/caamalg_qi.c
index 785b86d..1937b3c 100644
--- a/drivers/crypto/caam/caamalg_qi.c
+++ b/drivers/crypto/caam/caamalg_qi.c
@@ -957,7 +957,7 @@ static void aead_unmap(struct device *dev,
 	int ivsize = crypto_aead_ivsize(aead);
 
 	dma_unmap_sg_chained(dev, req->assoc, edesc->assoc_nents,
-			     DMA_BIDIRECTIONAL, edesc->assoc_chained);
+			     DMA_TO_DEVICE, edesc->assoc_chained);
 
 	caam_unmap(dev, req->src, req->dst,
 		   edesc->src_nents, edesc->src_chained, edesc->dst_nents,
-- 
1.7.5.4

