From 7356143af4b550bdfdd7ffd80275ac6d28b630e0 Mon Sep 17 00:00:00 2001
From: Yanjiang Jin <yanjiang.jin@windriver.com>
Date: Thu, 9 Apr 2015 16:27:28 +0800
Subject: [PATCH] fsl: caam: Call irq_dispose_mapping() only when it is the
 last handler

Call irq_dispose_mapping() only when this is the last handler,
else system would report the below warnings during kexec.

remove_proc_entry: removing non-empty directory 'irq/45', leaking at least 'fffe34000.jr'
------------[ cut here ]------------
WARNING: at fs/proc/generic.c:562
Modules linked in:
CPU: 0 PID: 550 Comm: reboot Not tainted 3.10.62-ltsi-WR6.0.0.0_standard #40
task: c7e92780 ti: c5ca4000 task.ti: c5ca4000
NIP: c021ee0c LR: c021ee0c CTR: c057ed5c
REGS: c5ca5bb0 TRAP: 0700   Not tainted  (3.10.62-ltsi-WR6.0.0.0_standard)
MSR: 00029000 <CE,EE,ME>  CR: 22088f84  XER: 20000000

GPR00: c021ee0c c5ca5c60 c7e92780 00000059 c7e92e08 00000006 c0c22d90 c1a223ec
GPR08: 00406e16 c0b9d934 00000000 000001d6 42088f82 1001abf0 00000000 100f0000
GPR16: 100f0000 00000001 10010000 00000000 00000001 00000001 00000000 00000000
GPR24: 00000000 c0bc7494 c5ca5ca8 c5ca5ca8 c61bc3f1 c62298a8 00000002 c61bc380
NIP [c021ee0c] remove_proc_entry+0x198/0x1a0
LR [c021ee0c] remove_proc_entry+0x198/0x1a0
Call Trace:
[c5ca5c60] [c021ee0c] remove_proc_entry+0x198/0x1a0 (unreliable)
[c5ca5ca0] [c0103a9c] unregister_irq_proc+0xc0/0xd4
[c5ca5cd0] [c00fcb64] free_desc+0x48/0x88
[c5ca5cf0] [c00fcc00] irq_free_descs+0x5c/0xc4
[c5ca5d10] [c06c13bc] caam_jr_remove+0x94/0xe0
[c5ca5d30] [c04fc890] __device_release_driver+0x74/0x120
[c5ca5d40] [c04fc974] device_release_driver+0x38/0x50
[c5ca5d50] [c04fc06c] bus_remove_device+0xfc/0x150
[c5ca5d70] [c04f8554] device_del+0x12c/0x1f4
[c5ca5d90] [c04f8640] device_unregister+0x24/0x58
[c5ca5da0] [c06bf080] caam_remove+0x64/0x2f8
[c5ca5de0] [c04fa400] device_shutdown+0xfc/0x1c8
[c5ca5e10] [c006c450] kernel_restart+0x24/0x98
[c5ca5e20] [c006c704] SyS_reboot+0x214/0x238
[c5ca5f40] [c0011f24] ret_from_syscall+0x0/0x3c
---
 drivers/crypto/caam/jr.c |    7 ++++++-
 1 files changed, 6 insertions(+), 1 deletions(-)

diff --git a/drivers/crypto/caam/jr.c b/drivers/crypto/caam/jr.c
index eea4bc8..0f46ec1 100644
--- a/drivers/crypto/caam/jr.c
+++ b/drivers/crypto/caam/jr.c
@@ -118,7 +118,12 @@ static int caam_jr_remove(struct platform_device *pdev)
 	ret = caam_jr_shutdown(jrdev);
 	if (ret)
 		dev_err(jrdev, "Failed to shut down job ring\n");
-	irq_dispose_mapping(jrpriv->irq);
+
+	/* Call irq_dispose_mapping() only when this is the last handler.
+	 * The better check point should be within irq_dispose_mapping().
+	*/
+	if (!irq_to_desc(jrpriv->irq)->action)
+		irq_dispose_mapping(jrpriv->irq);
 
 	return ret;
 }
-- 
1.7.5.4

