From bd58dc5d0bca2eb84846cbee31569e92cd7c416d Mon Sep 17 00:00:00 2001
From: Sandeep Paulraj <s-paulraj@ti.com>
Date: Mon, 12 Mar 2012 18:28:14 -0400
Subject: [PATCH 160/256] firmware: add support for keystone QMSS

The keystone Queue Manager Subsystem needs firmware to be
downloaded to the PDSP for features such as accumulator
support to function.
While in big endian mode qmss_pdsp_acc48_be_1_0_2_0.fw.ihex needs to be
downloaded to the PDSP.
While in little endian mode qmss_pdsp_acc48_le_1_0_2_0.fw.ihex needs to be
downloaded to the PDSP.

Signed-off-by: Sandeep Paulraj <s-paulraj@ti.com>
(cherry picked from commit a3177759848a7986664e7f14e420e2c2da76d349)
Signed-off-by: Zumeng Chen <zumeng.chen@windriver.com>
---
 firmware/Makefile                                  |    4 +-
 .../keystone/qmss_pdsp_acc48_be_1_0_2_0.fw.ihex    |   83 ++++++++++++++++++++
 .../keystone/qmss_pdsp_acc48_le_1_0_2_0.fw.ihex    |   83 ++++++++++++++++++++
 3 files changed, 169 insertions(+), 1 deletions(-)
 create mode 100644 firmware/keystone/qmss_pdsp_acc48_be_1_0_2_0.fw.ihex
 create mode 100644 firmware/keystone/qmss_pdsp_acc48_le_1_0_2_0.fw.ihex

diff --git a/firmware/Makefile b/firmware/Makefile
index 03e0a71..addd41d 100644
--- a/firmware/Makefile
+++ b/firmware/Makefile
@@ -97,7 +97,9 @@ fw-shipped-$(CONFIG_TEHUTI) += tehuti/bdx.bin
 fw-shipped-$(CONFIG_TIGON3) += tigon/tg3.bin tigon/tg3_tso.bin \
 			       tigon/tg3_tso5.bin
 fw-shipped-$(CONFIG_TI_KEYSTONE) += keystone/sa_pdsp0_1_0_2_1.fw \
-				    keystone/sa_pdsp1_1_0_2_1.fw
+				    keystone/sa_pdsp1_1_0_2_1.fw \
+				    keystone/qmss_pdsp_acc48_le_1_0_2_0.fw \
+				    keystone/qmss_pdsp_acc48_be_1_0_2_0.fw
 fw-shipped-$(CONFIG_TYPHOON) += 3com/typhoon.bin
 fw-shipped-$(CONFIG_USB_EMI26) += emi26/loader.fw emi26/firmware.fw \
 				  emi26/bitstream.fw
diff --git a/firmware/keystone/qmss_pdsp_acc48_be_1_0_2_0.fw.ihex b/firmware/keystone/qmss_pdsp_acc48_be_1_0_2_0.fw.ihex
new file mode 100644
index 0000000..941be9d
--- /dev/null
+++ b/firmware/keystone/qmss_pdsp_acc48_be_1_0_2_0.fw.ihex
@@ -0,0 +1,83 @@
+:10000000E017112480220481E003802480220081F3
+:100010008287FF2E65FF002445FF002425FF002472
+:1000200005FF0024032F002444000024E0004024A6
+:10003000E1000024E0E004058120E080FEE0006FA4
+:100040006464001D03FF1FC9FFFF1F1F6464001FBE
+:10005000030301010203306903200024862000937A
+:100060003A2607C918268069030630616603002412
+:100070000050002107062061E006200503C3E0D000
+:100080006604002400500021C3C3E01C00260021A8
+:10009000FCE206CEE2E2061C23060610E023040979
+:1000A000E1230209E0E0E100E0E040018720E09286
+:1000B00064647F11028B00519E2E0123660100248F
+:1000C000005000211E268169E00020240389E060A1
+:1000D00066060024005000210306306166030024F8
+:1000E0000050002107062061E006200503C3E0C898
+:1000F0006605002400500021C3C3E01E0042002119
+:10010000FCE206D6E2E2061E6A0000248B00002410
+:10011000CB000024034A01D1024A00C96A6A011FC8
+:10012000E0060409E1060209E0E0E100E0E0400148
+:100130008720E0826601002400500021660200242E
+:1001400026000024862000819EFF00232300002437
+:10015000E0440F1103E00F699EFF002300540021CB
+:1001600004232061E023200534C3E0C8005D0021A2
+:1001700032E223C864647F11E0230409E123020909
+:10018000E0E0E100EFE040018720EF922B6A02D12E
+:10019000E08B010120E0C950E089050BE0E0020995
+:1001A0008125E090074A05D10FE189C8CC898910E3
+:1001B0009E960023E08B010117E0C95000790021D1
+:1001C000EDE1E710CCED012707CC2051EDEDCC1C83
+:1001D000CCCC89009E960023E08B0101FAE0C96E29
+:1001E00000850021066407C9054A01C9024A00D1F9
+:1001F000036A01D1CB8A8A106A6A011F0E6A01C99B
+:1002000004CB00510C6400C9CBCB01050ACB0069BB
+:10021000098B0051E123000104E12061E1E12005A7
+:1002200080230C91008B00218023089102E0E1D013
+:100230009EDD00238720EF8282231F4F04231F515E
+:1002400023230101C323306700100021230303107F
+:10025000C0233067001000218E008424CE0B0024C0
+:10026000E0230209E0E02300E0E00609EEEEE00012
+:100270008C8B01018CC98C043E8C0051004A0C11FE
+:100280001C0004512F0008518C8C0419E0CC040987
+:10029000E0E00C019026E09036F000510C8C0171EA
+:1002A0009126E09002F100698C010024088C027113
+:1002B0009226E09002F200698C020024048C037103
+:1002C0009326E09002F300698C030024E18B02097D
+:1002D000024A04C9E1E10401008C020910CEE1EEFA
+:1002E0008B8B8C006464071FE28C045700DC0021B8
+:1002F0008C8C0219E0CC0409E0E008019066E090E3
+:100300001CF10051D0CCCC10048C01519266E090CD
+:1003100002F300698C010024D2CCCC10E18B0309DC
+:10032000024A04C9E1E10801008C030910CEE1EEA4
+:100330008B8B8C006464071FCE8C025700DC00217D
+:100340008C8C0119E0CC040990E6E09009F300518F
+:10035000D2CCCC10E18B0409024A04C9E1E11001BE
+:1003600090EEE1E08B8B01016464071F009C00218B
+:1003700000009E2094008424D40B0024E023020972
+:10038000E0E02300E0E00609F4F4E000E04A020BBC
+:10039000E0E00311E1E00201044A04C9E08B8B10A4
+:1003A000803400E100EF00219087FF2E000100243F
+:1003B0000000E108C08BE10810D4C0EEF5E8E810B9
+:1003C000036A00C9E0C9E108F5F5E000E08B01012E
+:1003D00096E0E108D63090240044F013E0000027B6
+:1003E000E1E0060994A8E1806A6A021F4444E01E25
+:1003F00001E014010123802C00009E2077440F119E
+:10040000577701272C5720517777571CE05706095B
+:1004100094A8E090FBF61ED7FA96006FE15704010E
+:100420000944E1D04444E11EE1F40405F4F5040577
+:10043000F5E1E11096040024D630902494A8E080E1
+:10044000000001214444E11C4444571C015714019D
+:100450003701202CE0370409E1370209E0E0E10030
+:10046000EFE040018720EF926A6A021D6A6A011577
+:100470008B0000246A6A011D044A01D1034A00C9A5
+:10048000CB8A8A106A6A011F8720EF82E037000159
+:10049000E101002405E02061E0E02005E1E1E00861
+:1004A0008123048100000121E1E1E0088123008132
+:1004B0000000012100009E20DE9E9E10E12300012D
+:1004C00004E12061E1E1200580230C910035012148
+:1004D00080230891FAE0E1D60C6A02D1E0440F11C2
+:1004E00009E00F699EFF0023E0440F11FEE00F5763
+:1004F000E0230409E1230209E0E0E100E0E040013B
+:100500008720E0929EDD00239EFF0023E0440F1130
+:0C051000FEE0006F9EDEDE1000009E206A
+:00000001FF
diff --git a/firmware/keystone/qmss_pdsp_acc48_le_1_0_2_0.fw.ihex b/firmware/keystone/qmss_pdsp_acc48_le_1_0_2_0.fw.ihex
new file mode 100644
index 0000000..bc920c4
--- /dev/null
+++ b/firmware/keystone/qmss_pdsp_acc48_le_1_0_2_0.fw.ihex
@@ -0,0 +1,83 @@
+:10000000E017112480220481E003802480220081F3
+:100010008287FF2E05FF002425FF002445FF0024D2
+:1000200065FF0024632F002424000024E000402406
+:10003000E1000024E0E004058120E080FEE0006FA4
+:100040000404001D03FF1FC9FFFF1F1F0404001F3E
+:1000500063630101026330696320002486200093FA
+:100060003A2607C918268069030630616603002412
+:100070000050002107062061E00620050383E0D040
+:1000800066040024005000218383E01C0026002128
+:10009000FCE206CEE2E2061C43060610E043040939
+:1000A000E1430209E0E0E100E0E040018720E09266
+:1000B00004047F11028B00519E2E0123660100244F
+:1000C000005000211E268169E00020240389E060A1
+:1000D00066060024005000210306306166030024F8
+:1000E0000050002107062061E00620050383E0C8D8
+:1000F00066050024005000218383E01E0042002199
+:10010000FCE206D6E2E2061E6A0000248B00002410
+:10011000CB000024034A01D1024A00C96A6A011FC8
+:10012000E0060409E1060209E0E0E100E0E0400148
+:100130008720E0826601002400500021660200242E
+:1001400026000024862000819EFF00234300002417
+:10015000E0240F1103E00F699EFF002300540021EB
+:1001600004432061E04320053483E0C8005D0021A2
+:1001700032E243C804047F11E0430409E143020969
+:10018000E0E0E100EFE040018720EF922B6A02D12E
+:10019000E08B010120E0C950E089050BE0E0020995
+:1001A0008125E090074A05D10FE189C88C89891023
+:1001B0009E960023E08B010117E0C95000790021D1
+:1001C000EDE1E7108CED0127078C2051EDED8C1C43
+:1001D0008C8C89009E960023E08B0101FAE0C96EA9
+:1001E00000850021060407C9054A01C9024A00D159
+:1001F000036A01D1CB8A8A106A6A011F0E6A01C99B
+:1002000004CB00510C0400C9CBCB01050ACB00691B
+:10021000098B0051E143000104E12061E1E1200587
+:1002200080230C91008B00218023089102E0E1D013
+:100230009EDD00238720EF8282431F4F04431F511E
+:1002400043430101C343306700100021436363103F
+:10025000C0433067001000218E008424CE0B0024A0
+:10026000E0430209E0E04300E0E00609EEEEE000D2
+:10027000CC8B0101CCC9CC043ECC0051004A0C11FE
+:100280001C0004512F000851CCCC0419E08C040947
+:10029000E0E00C019026E09036F000510CCC0171AA
+:1002A0009126E09002F10069CC01002408CC027193
+:1002B0009226E09002F20069CC02002404CC037183
+:1002C0009326E09002F30069CC030024E18B02093D
+:1002D000024A04C9E1E1040100CC020910CEE1EEBA
+:1002E0008B8BCC000404071FE2CC045700DC0021F8
+:1002F000CCCC0219E08C0409E0E008019066E090A3
+:100300001CF10051D08C8C1004CC01519266E0900D
+:1003100002F30069CC010024D28C8C10E18B03091C
+:10032000024A04C9E1E1080100CC030910CEE1EE64
+:100330008B8BCC000404071FCECC025700DC0021BD
+:10034000CCCC0119E08C040990E6E09009F300514F
+:10035000D28C8C10E18B0409024A04C9E1E110013E
+:1003600090EEE1E08B8B01010404071F009C00214B
+:1003700000009E2094008424D40B0024E043020952
+:10038000E0E04300E0E00609F4F4E000E04A020B9C
+:10039000E0E00311E1E00201044A04C9E08B8B10A4
+:1003A000803400E100EF00219087FF2E000100243F
+:1003B0000000E108C08BE10810D4C0EEF5E8E810B9
+:1003C000036A00C9E0C9E108F5F5E000E08B01012E
+:1003D00096E0E108D63090240024F013E0000027D6
+:1003E000E1E0060994A8E1806A6A021F2424E01E65
+:1003F00001E014010143802C00009E2017240F11FE
+:10040000371701272C3720511717371CE0370609FB
+:1004100094A8E090FBF61ED7FA96006FE13704012E
+:100420000924E1D02424E11EE1F40405F4F50405D7
+:10043000F5E1E11096040024D630902494A8E080E1
+:10044000000001212424E11C2424371C013714015D
+:100450005701202CE0570409E1570209E0E0E100D0
+:10046000EFE040018720EF926A6A021D6A6A011577
+:100470008B0000246A6A011D044A01D1034A00C9A5
+:10048000CB8A8A106A6A011F8720EF82E057000139
+:10049000E101002405E02061E0E02005E1E1E00861
+:1004A0008123048100000121E1E1E0088123008132
+:1004B0000000012100009E20DE9E9E10E14300010D
+:1004C00004E12061E1E1200580230C910035012148
+:1004D00080230891FAE0E1D60C6A02D1E0240F11E2
+:1004E00009E00F699EFF0023E0240F11FEE00F5783
+:1004F000E0430409E1430209E0E0E100E0E04001FB
+:100500008720E0929EDD00239EFF0023E0240F1150
+:0C051000FEE0006F9EDEDE1000009E206A
+:00000001FF
-- 
1.7.5.4

