From 6f2f6a27030064cf103c5ef668b17b758ba11d75 Mon Sep 17 00:00:00 2001
From: Hao Zhang <hzhang@ti.com>
Date: Mon, 28 Jul 2014 16:25:27 -0400
Subject: [PATCH 177/256] firmware: keystone: pa: Update PA firmware to
 version 03.00.00.10

Update PA firmware to revision 3.0.0.10. This update applies to
Keystone II K2E and K2L SoCs.

Signed-off-by: Hao Zhang <hzhang@ti.com>
(cherry picked from commit 07636f8bbaeb7b5e00ae05d7082015cc65451567)
Signed-off-by: Zumeng Chen <zumeng.chen@windriver.com>
---
 firmware/keystone/pa_eg0_pdsp0.fw.ihex  |    4 ++--
 firmware/keystone/pa_eg0_pdsp1.fw.ihex  |    4 ++--
 firmware/keystone/pa_eg0_pdsp2.fw.ihex  |    4 ++--
 firmware/keystone/pa_eg1_pdsp0.fw.ihex  |    4 ++--
 firmware/keystone/pa_eg2_pdsp0.fw.ihex  |    4 ++--
 firmware/keystone/pa_in0_pdsp0.fw.ihex  |    4 ++--
 firmware/keystone/pa_in0_pdsp1.fw.ihex  |    4 ++--
 firmware/keystone/pa_in1_pdsp0.fw.ihex  |    4 ++--
 firmware/keystone/pa_in1_pdsp1.fw.ihex  |    4 ++--
 firmware/keystone/pa_in2_pdsp0.fw.ihex  |    4 ++--
 firmware/keystone/pa_in3_pdsp0.fw.ihex  |    4 ++--
 firmware/keystone/pa_in4_pdsp0.fw.ihex  |    4 ++--
 firmware/keystone/pa_in4_pdsp1.fw.ihex  |    4 ++--
 firmware/keystone/pa_post_pdsp0.fw.ihex |    4 ++--
 firmware/keystone/pa_post_pdsp1.fw.ihex |    4 ++--
 15 files changed, 30 insertions(+), 30 deletions(-)

diff --git a/firmware/keystone/pa_eg0_pdsp0.fw.ihex b/firmware/keystone/pa_eg0_pdsp0.fw.ihex
index ccd009c..c635e01 100644
--- a/firmware/keystone/pa_eg0_pdsp0.fw.ihex
+++ b/firmware/keystone/pa_eg0_pdsp0.fw.ihex
@@ -1,5 +1,5 @@
-:1000000021000300BABE0600030000092300CB9EB6
-:1000100024000982240300C2810437822EFF8582D6
+:1000000021000300BABE06000300000A2300CB9EB5
+:1000100024000A82240300C2810437822EFF8582D5
 :100020008104A482240001E281002482C900FF002F
 :10003000D100FF002EFF819D24000A1D8100177D45
 :10004000C9067D1BC9077D02D110FF00D110FF1822
diff --git a/firmware/keystone/pa_eg0_pdsp1.fw.ihex b/firmware/keystone/pa_eg0_pdsp1.fw.ihex
index c6f2783..3d7ad1f 100644
--- a/firmware/keystone/pa_eg0_pdsp1.fw.ihex
+++ b/firmware/keystone/pa_eg0_pdsp1.fw.ihex
@@ -1,4 +1,4 @@
-:1000000021015E00BABE060103000009C907FF0016
+:1000000021015E00BABE06010300000AC907FF0015
 :10001000911007C01F0F8080811007C010D7D794A0
 :100020006900661381182787510008089109076243
 :10003000110F020211C008001200020281090762BA
@@ -87,7 +87,7 @@
 :1005600021015A00240003E581043B852EFF81848C
 :1005700081003B84210173002306209E2EFF85828B
 :100580008104A482240001E281002482C900FF00CA
-:10059000D100FF0024000B7D2400005D24000982AF
+:10059000D100FF0024000B7D2400005D24000A82AE
 :1005A000240300C2810437822EFF87808110F780E8
 :1005B00091003B8451006406570164E2240002E587
 :1005C00081043B852400006481001B04CF04FFF9F3
diff --git a/firmware/keystone/pa_eg0_pdsp2.fw.ihex b/firmware/keystone/pa_eg0_pdsp2.fw.ihex
index dd9f6ba..09186e7 100644
--- a/firmware/keystone/pa_eg0_pdsp2.fw.ihex
+++ b/firmware/keystone/pa_eg0_pdsp2.fw.ihex
@@ -1,4 +1,4 @@
-:1000000021028900BABE060203000009C907FF00E9
+:1000000021028900BABE06020300000AC907FF00E8
 :10001000911007C01D0F8080811007C010D7D794A2
 :100020006900661381182787510008089109076243
 :10003000110F020211C008001200020281090762BA
@@ -162,7 +162,7 @@
 :100A1000240003E581043B852EFF818481003B8413
 :100A200021029E002305D79E2EFF85828104A48289
 :100A3000240001E281002482C900FF00D100FF00F0
-:100A400024000C7D2400005D24000982240300C2E0
+:100A400024000C7D2400005D24000A82240300C2DF
 :100A5000810437822EFF87808110F78091003B84CC
 :100A600051006406570164E4240002E581043B85DB
 :100A70002400006481001B04CF04FFF9910024814D
diff --git a/firmware/keystone/pa_eg1_pdsp0.fw.ihex b/firmware/keystone/pa_eg1_pdsp0.fw.ihex
index f83cd7f..4ee72a8 100644
--- a/firmware/keystone/pa_eg1_pdsp0.fw.ihex
+++ b/firmware/keystone/pa_eg1_pdsp0.fw.ihex
@@ -1,4 +1,4 @@
-:100000002100F600BABE070003000009C907FF007F
+:100000002100F600BABE07000300000AC907FF007E
 :10001000911007C01D0F8080811007C010D7D794A2
 :100020006900661381182787510008089109076243
 :10003000110F020211C008001200020281090762BA
@@ -61,7 +61,7 @@
 :1003C0002100F200240003E581043B852EFF818497
 :1003D00081003B8421010B002304449E2EFF858273
 :1003E0008104A482240001E281002482C900FF006C
-:1003F000D100FF0024000D7D2400005D240009824F
+:1003F000D100FF0024000D7D2400005D24000A824E
 :10040000240300C2810437822EFF87808110F78089
 :1004100091003B8451006406570164E4240002E526
 :1004200081043B852400006481001B04CF04FFF994
diff --git a/firmware/keystone/pa_eg2_pdsp0.fw.ihex b/firmware/keystone/pa_eg2_pdsp0.fw.ihex
index e40da55..9962103 100644
--- a/firmware/keystone/pa_eg2_pdsp0.fw.ihex
+++ b/firmware/keystone/pa_eg2_pdsp0.fw.ihex
@@ -1,4 +1,4 @@
-:100000002101ED00BABE080003000009C907FF0086
+:100000002101ED00BABE08000300000AC907FF0085
 :10001000911007C01D0F8080811007C010D7D794A2
 :100020006900661381182787510008089109076243
 :10003000110F020211C008001200020281090762BA
@@ -123,7 +123,7 @@
 :1007A000240003E581043B852EFF818481003B8486
 :1007B0002102020023053B9E2EFF85828104A48234
 :1007C000240001E281002482C900FF00D100FF0063
-:1007D00024000E7D2400005D24000982240300C251
+:1007D00024000E7D2400005D24000A82240300C250
 :1007E000810437822EFF87808110F78091003B843F
 :1007F00051006406570164E4240002E581043B854E
 :100800002400006481001B04CF04FFF991002481BF
diff --git a/firmware/keystone/pa_in0_pdsp0.fw.ihex b/firmware/keystone/pa_in0_pdsp0.fw.ihex
index 2f62beb..8cc6c3a 100644
--- a/firmware/keystone/pa_in0_pdsp0.fw.ihex
+++ b/firmware/keystone/pa_in0_pdsp0.fw.ihex
@@ -1,11 +1,11 @@
-:1000000021001D00BABE0000030000092EFFA780DA
+:1000000021001D00BABE00000300000A2EFFA780D9
 :100010008900F3808950F380240000E52401F0E199
 :100020002EFF87882400066880E5F4880110E5E546
 :100030006EE1E5FE2EFF938285A0738224000262AA
 :10004000240002422400022281A033822400406561
 :100050002400404581AC33852400026981C0338986
 :100060002400026A81C4338A240000E581F03385CC
-:10007000209E000023026A9E24000982240300C2FD
+:10007000209E000023026A9E24000A82240300C2FC
 :10008000810437822EFF85828104A482240001E24C
 :1000900081002482C900FF00D100FF002300039EDD
 :1000A0002EFF819D2400001D8100177DC91FFF05C3
diff --git a/firmware/keystone/pa_in0_pdsp1.fw.ihex b/firmware/keystone/pa_in0_pdsp1.fw.ihex
index 9ed1503..28e7cf7 100644
--- a/firmware/keystone/pa_in0_pdsp1.fw.ihex
+++ b/firmware/keystone/pa_in0_pdsp1.fw.ihex
@@ -1,5 +1,5 @@
-:1000000021000300BABE0001030000092302559E2F
-:1000100024000982240300C2810437822EFF8582D6
+:1000000021000300BABE00010300000A2302559E2E
+:1000100024000A82240300C2810437822EFF8582D5
 :100020008104A482240001E281002482C900FF002F
 :10003000D100FF002EFF819D2400011D8100177D4E
 :10004000C903FF07910C248111F85D5D12015D5D0C
diff --git a/firmware/keystone/pa_in1_pdsp0.fw.ihex b/firmware/keystone/pa_in1_pdsp0.fw.ihex
index 3343019..76834ca 100644
--- a/firmware/keystone/pa_in1_pdsp0.fw.ihex
+++ b/firmware/keystone/pa_in1_pdsp0.fw.ihex
@@ -1,5 +1,5 @@
-:1000000021000300BABE0002030000092302509E33
-:1000100024000982240300C2810437822EFF8582D6
+:1000000021000300BABE00020300000A2302509E32
+:1000100024000A82240300C2810437822EFF8582D5
 :100020008104A482240001E281002482C900FF002F
 :10003000D100FF002EFF819D2400021D8100177D4D
 :10004000C901FF1324000680510280032400A4810B
diff --git a/firmware/keystone/pa_in1_pdsp1.fw.ihex b/firmware/keystone/pa_in1_pdsp1.fw.ihex
index 01b8eef..5251574 100644
--- a/firmware/keystone/pa_in1_pdsp1.fw.ihex
+++ b/firmware/keystone/pa_in1_pdsp1.fw.ihex
@@ -1,5 +1,5 @@
-:1000000021000300BABE0101030000092302299E5A
-:1000100024000982240300C2810437822EFF8582D6
+:1000000021000300BABE01010300000A2302299E59
+:1000100024000A82240300C2810437822EFF8582D5
 :100020008104A482240001E281002482C900FF002F
 :10003000D100FF002EFF819D2400031D8100177D4C
 :10004000C9067D21C9077D02D110FF00D110FF1E16
diff --git a/firmware/keystone/pa_in2_pdsp0.fw.ihex b/firmware/keystone/pa_in2_pdsp0.fw.ihex
index 5e9dff0..9853012 100644
--- a/firmware/keystone/pa_in2_pdsp0.fw.ihex
+++ b/firmware/keystone/pa_in2_pdsp0.fw.ihex
@@ -1,5 +1,5 @@
-:1000000021000300BABE0200030000092302259E5E
-:1000100024000982240300C2810437822EFF8582D6
+:1000000021000300BABE02000300000A2302259E5D
+:1000100024000A82240300C2810437822EFF8582D5
 :100020008104A482240001E281002482C900FF002F
 :10003000D100FF002EFF819D2400041D8100177D4B
 :10004000C9067D21C9077D02D110FF00D110FF1E16
diff --git a/firmware/keystone/pa_in3_pdsp0.fw.ihex b/firmware/keystone/pa_in3_pdsp0.fw.ihex
index 8b3e959..88b764b 100644
--- a/firmware/keystone/pa_in3_pdsp0.fw.ihex
+++ b/firmware/keystone/pa_in3_pdsp0.fw.ihex
@@ -1,5 +1,5 @@
-:1000000021000300BABE0300030000092302569E2C
-:1000100024000982240300C2810437822EFF8582D6
+:1000000021000300BABE03000300000A2302569E2B
+:1000100024000A82240300C2810437822EFF8582D5
 :100020008104A482240001E281002482C900FF002F
 :10003000D100FF002EFF819D2400051D8100177D4A
 :10004000C903FF07910C248111F85D5D12015D5D0C
diff --git a/firmware/keystone/pa_in4_pdsp0.fw.ihex b/firmware/keystone/pa_in4_pdsp0.fw.ihex
index 7f47585..83ab213 100644
--- a/firmware/keystone/pa_in4_pdsp0.fw.ihex
+++ b/firmware/keystone/pa_in4_pdsp0.fw.ihex
@@ -1,5 +1,5 @@
-:1000000021000300BABE0401030000092302559E2B
-:1000100024000982240300C2810437822EFF8582D6
+:1000000021000300BABE04010300000A2302559E2A
+:1000100024000A82240300C2810437822EFF8582D5
 :100020008104A482240001E281002482C900FF002F
 :10003000D100FF002EFF819D2400061D8100177D49
 :10004000C901FF1324000680510680032400A48107
diff --git a/firmware/keystone/pa_in4_pdsp1.fw.ihex b/firmware/keystone/pa_in4_pdsp1.fw.ihex
index 38db6a5..1113658 100644
--- a/firmware/keystone/pa_in4_pdsp1.fw.ihex
+++ b/firmware/keystone/pa_in4_pdsp1.fw.ihex
@@ -1,7 +1,7 @@
-:1000000021000300BABE0401030000092300E79E9B
+:1000000021000300BABE04010300000A2300E79E9A
 :100010002EFF85828104A482240001E281002482D3
 :10002000C900FF00D100FF002EFF819D2400073D85
-:100030008100175D24000982240300C281043782F5
+:100030008100175D24000A82240300C281043782F4
 :10004000C904FF00910024810101E1E181002481C4
 :100050002E808F862E84098EC90FCA0424001004B6
 :100060002F000384210076002480169F24000804BA
diff --git a/firmware/keystone/pa_post_pdsp0.fw.ihex b/firmware/keystone/pa_post_pdsp0.fw.ihex
index a5ade9f..373a4ba 100644
--- a/firmware/keystone/pa_post_pdsp0.fw.ihex
+++ b/firmware/keystone/pa_post_pdsp0.fw.ihex
@@ -1,7 +1,7 @@
-:1000000021000300BABE05000300000923043D9E41
+:1000000021000300BABE05000300000A23043D9E40
 :100010002EFF85828104A482240001E281002482D3
 :10002000C900FF00D100FF002400087D2400005D0E
-:1000300024000982240300C2810437822EFF8780B6
+:1000300024000A82240300C2810437822EFF8780B5
 :100040008110F78024000F6E105D5D4E51006E0F21
 :1000500009064E8E01108ECF908E00CF01014E4EBC
 :10006000690F4E022400004E05016E6E562F0FF8E8
diff --git a/firmware/keystone/pa_post_pdsp1.fw.ihex b/firmware/keystone/pa_post_pdsp1.fw.ihex
index 3114f63..e069bc1 100644
--- a/firmware/keystone/pa_post_pdsp1.fw.ihex
+++ b/firmware/keystone/pa_post_pdsp1.fw.ihex
@@ -1,7 +1,7 @@
-:1000000021000300BABE0501030000092302BA9EC5
+:1000000021000300BABE05010300000A2302BA9EC4
 :100010002EFF85828104A482240001E281002482D3
 :10002000C900FF00D100FF002400097D2400005D0D
-:1000300024000982240300C2810437822EFF8780B6
+:1000300024000A82240300C2810437822EFF8780B5
 :100040008110F780C904FF00910024810101E1E1E2
 :10005000810024812EFF83942EFF80DD2E808F86E9
 :100060002E84098EC90FCA04240010042F000384B3
-- 
1.7.5.4

