From ad2a09dcdd118a9b4649cb3b05a0d0bdb40814a0 Mon Sep 17 00:00:00 2001
From: Zumeng Chen <zumeng.chen@windriver.com>
Date: Fri, 17 Mar 2017 15:04:48 +0800
Subject: [PATCH] grsecurity: mark these varibles with kmemleak_ignore()

These vaiables will never be freed, and get caught by kmemleak
with the following:

unreferenced object 0xebd34fc0 (size 512):
  comm "swapper/0", pid 1, jiffies 4294938265 (age 295.490s)
  hex dump (first 32 bytes):
    00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00  ................
    00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00  ................
  backtrace:
    [<c0128f68>] kmem_cache_alloc_trace+0x198/0x27c
    [<c09b3444>] 0xc09b3444
    [<c0987ef4>] 0xc0987ef4
    [<c06ea5ec>] kernel_init+0x1c/0xf4
    [<c000e3b8>] ret_from_fork+0x14/0x20
    [<ffffffff>] 0xffffffff

Signed-off-by: Zumeng Chen <zumeng.chen@windriver.com>
---
 grsecurity/grsec_init.c |    6 ++++++
 1 files changed, 6 insertions(+), 0 deletions(-)

diff --git a/grsecurity/grsec_init.c b/grsecurity/grsec_init.c
index 4ed9e7d..c67cdac 100644
--- a/grsecurity/grsec_init.c
+++ b/grsecurity/grsec_init.c
@@ -127,7 +127,13 @@ grsecurity_init(void)
 	gr_system_salt = kmalloc(GR_SALT_LEN, GFP_KERNEL);
 	gr_system_sum = kmalloc(GR_SHA_LEN, GFP_KERNEL);
 
+	kmemleak_ignore(gr_usermode);
+	kmemleak_ignore(gr_system_salt);
+	kmemleak_ignore(gr_system_sum);
 	if (!gr_usermode || !gr_system_salt || !gr_system_sum) {
+		kfree(gr_usermode);
+		kfree(gr_system_salt);
+		kfree(gr_system_sum);
 		panic("Unable to allocate grsecurity authentication structure");
 		return;
 	}
-- 
1.7.5.4

