From 449611fac3ed413cdf7278fd6e7a69c255ba8574 Mon Sep 17 00:00:00 2001
From: Zumeng Chen <zumeng.chen@windriver.com>
Date: Wed, 16 Nov 2011 14:06:38 +0800
Subject: [PATCH 01/18] fs: pramfs: Fix the right printk information for pramfs

Upstream ID 876f8222bb922fbac7be819dabd8c62713dbb19a
  git://pramfs.git.sourceforge.net/gitroot/pramfs/pramfs
  From: Marco <marco@ncc-1701-f.site>

Integrated-by: Zumeng Chen <zumeng.chen@windriver.com>
---
 fs/pramfs/super.c |    9 ++++-----
 1 files changed, 4 insertions(+), 5 deletions(-)

diff --git a/fs/pramfs/super.c b/fs/pramfs/super.c
index 9eaa8ba..9a5317a 100644
--- a/fs/pramfs/super.c
+++ b/fs/pramfs/super.c
@@ -108,7 +108,7 @@ static loff_t pram_max_size(int bits)
 	if (res > MAX_LFS_FILESIZE)
 		res = MAX_LFS_FILESIZE;
 
-	pram_info("max file size %llu bytes", res);
+	pram_info("max file size %llu bytes\n", res);
 	return res;
 }
 
@@ -277,7 +277,7 @@ static int pram_parse_options(char *options, struct pram_sb_info *sbi, bool remo
 #else
 		case Opt_user_xattr:
 		case Opt_nouser_xattr:
-			pram_info("(no)user_xattr options not supported");
+			pram_info("(no)user_xattr options not supported\n");
 			break;
 #endif
 #ifdef CONFIG_PRAMFS_POSIX_ACL
@@ -290,7 +290,7 @@ static int pram_parse_options(char *options, struct pram_sb_info *sbi, bool remo
 #else
 		case Opt_acl:
 		case Opt_noacl:
-			pram_info("(no)acl options not supported");
+			pram_info("(no)acl options not supported\n");
 			break;
 #endif
 		case Opt_xip:
@@ -300,7 +300,7 @@ static int pram_parse_options(char *options, struct pram_sb_info *sbi, bool remo
 			set_opt(sbi->s_mount_opt, XIP);
 			break;
 #else
-			pram_info("xip option not supported");
+			pram_info("xip option not supported\n");
 			break;
 #endif
 		default: {
@@ -601,7 +601,6 @@ static int pram_fill_super(struct super_block *sb, void *data, int silent)
 		goto out;
 	}
 	super = pram_get_super(sb);
-	root_i = pram_get_inode(sb, PRAM_ROOT_INO);
 
 #ifdef CONFIG_PRAMFS_TEST
 	if (!first_pram_super)
-- 
1.7.0.4

