From b506389c54cceaab1e2b9d2af6c7caf159aa74ca Mon Sep 17 00:00:00 2001
From: Zumeng Chen <zumeng.chen@windriver.com>
Date: Wed, 16 Nov 2011 14:06:41 +0800
Subject: [PATCH 04/18] fs: pramfs: Used less checking __copy_{from|to}_user functions

Upstream ID: 7d85e7ad686152869a1e44db02948a47a76a7dae
  git://pramfs.git.sourceforge.net/gitroot/pramfs/pramfs
  From: Marco <marco@ncc-1701-f.site>

Integrated-by: Zumeng Chen <zumeng.chen@windriver.com>
---
 fs/pramfs/file.c |   12 ++++++------
 1 files changed, 6 insertions(+), 6 deletions(-)

diff --git a/fs/pramfs/file.c b/fs/pramfs/file.c
index b22d1f3..0b26f7a 100644
--- a/fs/pramfs/file.c
+++ b/fs/pramfs/file.c
@@ -38,7 +38,7 @@ static size_t __pram_iov_copy_from(char *vaddr,
 		int copy = min(bytes, iov->iov_len - base);
 
 		base = 0;
-		left = copy_from_user(vaddr, buf, copy);
+		left = __copy_from_user(vaddr, buf, copy);
 		copied += copy;
 		bytes -= copy;
 		vaddr += copy;
@@ -60,7 +60,7 @@ static size_t __pram_iov_copy_to(char *vaddr,
 		int copy = min(bytes, iov->iov_len - base);
 
 		base = 0;
-		left = copy_to_user(buf, vaddr, copy);
+		left = __copy_to_user(buf, vaddr, copy);
 		copied += copy;
 		bytes -= copy;
 		vaddr += copy;
@@ -79,7 +79,7 @@ static size_t pram_iov_copy_from(void *to, struct iov_iter *i, size_t bytes)
 	if (likely(i->nr_segs == 1)) {
 		int left;
 		char __user *buf = i->iov->iov_base + i->iov_offset;
-		left = copy_from_user(to, buf, bytes);
+		left = __copy_from_user(to, buf, bytes);
 		copied = bytes - left;
 	} else {
 		copied = __pram_iov_copy_from(to, i->iov, i->iov_offset, bytes);
@@ -95,7 +95,7 @@ static size_t pram_iov_copy_to(void *from, struct iov_iter *i, size_t bytes)
 	if (likely(i->nr_segs == 1)) {
 		int left;
 		char __user *buf = i->iov->iov_base + i->iov_offset;
-		left = copy_to_user(buf, from, bytes);
+		left = __copy_to_user(buf, from, bytes);
 		copied = bytes - left;
 	} else {
 		copied = __pram_iov_copy_to(from, i->iov, i->iov_offset, bytes);
@@ -113,7 +113,7 @@ static size_t __pram_clear_user(const struct iovec *iov, size_t base, size_t byt
 		int clear = min(bytes, iov->iov_len - base);
 
 		base = 0;
-		left = clear_user(buf, clear);
+		left = __clear_user(buf, clear);
 		claened += clear;
 		bytes -= clear;
 		iov++;
@@ -131,7 +131,7 @@ static size_t pram_clear_user(struct iov_iter *i, size_t bytes)
 	if (likely(i->nr_segs == 1)) {
 		int left;
 		char __user *buf = i->iov->iov_base + i->iov_offset;
-		left = clear_user(buf, bytes);
+		left = __clear_user(buf, bytes);
 		clear = bytes - left;
 	} else {
 		clear = __pram_clear_user(i->iov, i->iov_offset, bytes);
-- 
1.7.0.4

