From 51502ebc245fbcb19528f4c692bfe01d38bdeee4 Mon Sep 17 00:00:00 2001
From: Zumeng Chen <zumeng.chen@windriver.com>
Date: Wed, 16 Nov 2011 14:06:47 +0800
Subject: [PATCH 10/18] fs: pramfs: Added compat ioctl support

Upstream ID: 1b8d92852e7dd89c5b8e55f6421f18abb0a96ca1
  git://pramfs.git.sourceforge.net/gitroot/pramfs/pramfs
  From: Marco <marco@ncc-1701-f.site>

Integrated-by: Zumeng Chen <zumeng.chen@windriver.com>
---
 fs/pramfs/dir.c   |    3 +++
 fs/pramfs/file.c  |    6 ++++++
 fs/pramfs/ioctl.c |   23 +++++++++++++++++++++++
 fs/pramfs/pram.h  |    4 ++++
 4 files changed, 36 insertions(+), 0 deletions(-)

diff --git a/fs/pramfs/dir.c b/fs/pramfs/dir.c
index 8dc2b29..cf0bcba 100644
--- a/fs/pramfs/dir.c
+++ b/fs/pramfs/dir.c
@@ -202,4 +202,7 @@ struct file_operations pram_dir_operations = {
 	.readdir	= pram_readdir,
 	.fsync		= noop_fsync,
 	.unlocked_ioctl	= pram_ioctl,
+#ifdef CONFIG_COMPAT
+	.compat_ioctl	= pram_compat_ioctl,
+#endif
 };
diff --git a/fs/pramfs/file.c b/fs/pramfs/file.c
index fe83132..db36c00 100644
--- a/fs/pramfs/file.c
+++ b/fs/pramfs/file.c
@@ -276,6 +276,9 @@ struct file_operations pram_file_operations = {
 	.fsync		= noop_fsync,
 	.check_flags	= pram_check_flags,
 	.unlocked_ioctl	= pram_ioctl,
+#ifdef CONFIG_COMPAT
+	.compat_ioctl	= pram_compat_ioctl,
+#endif
 };
 
 #ifdef CONFIG_PRAMFS_XIP
@@ -287,6 +290,9 @@ struct file_operations pram_xip_file_operations = {
 	.open		= generic_file_open,
 	.fsync		= noop_fsync,
 	.unlocked_ioctl	= pram_ioctl,
+#ifdef CONFIG_COMPAT
+	.compat_ioctl	= pram_compat_ioctl,
+#endif
 };
 #endif
 
diff --git a/fs/pramfs/ioctl.c b/fs/pramfs/ioctl.c
index de7f523..092cbe6 100644
--- a/fs/pramfs/ioctl.c
+++ b/fs/pramfs/ioctl.c
@@ -96,3 +96,26 @@ flags_out:
 		return -ENOTTY;
 	}
 }
+
+#ifdef CONFIG_COMPAT
+long pram_compat_ioctl(struct file *file, unsigned int cmd, unsigned long arg)
+{
+	switch (cmd) {
+	case FS_IOC32_GETFLAGS:
+		cmd = FS_IOC_GETFLAGS;
+		break;
+	case FS_IOC32_SETFLAGS:
+		cmd = FS_IOC_SETFLAGS;
+		break;
+	case FS_IOC32_GETVERSION:
+		cmd = FS_IOC_GETVERSION;
+		break;
+	case FS_IOC32_SETVERSION:
+		cmd = FS_IOC_SETVERSION;
+		break;
+	default:
+		return -ENOIOCTLCMD;
+	}
+	return pram_ioctl(file, cmd, (unsigned long) compat_ptr(arg));
+}
+#endif
diff --git a/fs/pramfs/pram.h b/fs/pramfs/pram.h
index d39ae14..918f550 100644
--- a/fs/pramfs/pram.h
+++ b/fs/pramfs/pram.h
@@ -115,6 +115,10 @@ extern void pram_get_inode_flags(struct inode *inode, struct pram_inode *pi);
 
 /* ioctl.c */
 extern long pram_ioctl(struct file *filp, unsigned int cmd, unsigned long arg);
+#ifdef CONFIG_COMPAT
+extern long pram_compat_ioctl(struct file *file, unsigned int cmd,
+			      unsigned long arg);
+#endif
 
 /* super.c */
 #ifdef CONFIG_PRAMFS_TEST
-- 
1.7.0.4

