From 2c3d2eaf2544f1e4ed81951e172b5b5fe671330d Mon Sep 17 00:00:00 2001
From: Zumeng Chen <zumeng.chen@windriver.com>
Date: Wed, 16 Nov 2011 14:06:53 +0800
Subject: [PATCH 16/18] fs: pramfs: enable tristate for pramfs

Upstream ID: de3ceaf54d0f5b15986eb5cdde0a2b9087924a11
  git://pramfs.git.sourceforge.net/gitroot/pramfs/pramfs
  From Marco <marco@ncc-1701-f.site>

Integrated-by: Zumeng Chen <zumeng.chen@windriver.com>
---
 fs/pramfs/Kconfig |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/fs/pramfs/Kconfig b/fs/pramfs/Kconfig
index c1c286b..0a2c54a 100644
--- a/fs/pramfs/Kconfig
+++ b/fs/pramfs/Kconfig
@@ -1,5 +1,5 @@
 config PRAMFS
-	bool "Persistent and Protected RAM file system support"
+	tristate "Persistent and Protected RAM file system support"
 	depends on HAS_IOMEM && EXPERIMENTAL
 	select CRC16
 	help
-- 
1.7.0.4

