From 09b179c5511a2521987e59db92883169ff203853 Mon Sep 17 00:00:00 2001
From: Zumeng Chen <zumeng.chen@windriver.com>
Date: Wed, 16 Nov 2011 14:06:43 +0800
Subject: [PATCH 06/18] fs: pramfs: Added fallocate support

Upstream ID 8868d72fc259b3922fa7b208f1f2317fe4486cad
  git://pramfs.git.sourceforge.net/gitroot/pramfs/pramfs
  From: Marco <marco@ncc-1701-f.site>

Integrated-by: Zumeng Chen <zumeng.chen@windriver.com>
---
 fs/pramfs/file.c  |    1 +
 fs/pramfs/inode.c |   44 ++++++++++++++++++++++++++++++++++++++++++++
 fs/pramfs/pram.h  |    6 ++++--
 3 files changed, 49 insertions(+), 2 deletions(-)

diff --git a/fs/pramfs/file.c b/fs/pramfs/file.c
index 7e6e3ad..1b522de 100644
--- a/fs/pramfs/file.c
+++ b/fs/pramfs/file.c
@@ -297,4 +297,5 @@ struct inode_operations pram_file_inode_operations = {
 #endif
 	.setattr	= pram_notify_change,
 	.check_acl	= pram_check_acl,
+	.fallocate	= pram_fallocate,
 };
diff --git a/fs/pramfs/inode.c b/fs/pramfs/inode.c
index 72f080f..a3e53eb 100644
--- a/fs/pramfs/inode.c
+++ b/fs/pramfs/inode.c
@@ -20,6 +20,7 @@
 #include <linux/module.h>
 #include <linux/mpage.h>
 #include <linux/backing-dev.h>
+#include <linux/falloc.h>
 #include "pram.h"
 #include "xattr.h"
 #include "xip.h"
@@ -718,6 +719,49 @@ int pram_notify_change(struct dentry *dentry, struct iattr *attr)
 	return error;
 }
 
+long pram_fallocate(struct inode *inode, int mode, loff_t offset, loff_t len)
+{
+	long ret = 0;
+	unsigned long blocknr, blockoff;
+	int num_blocks, blocksize_mask;
+	loff_t new_size;
+
+	/* preallocation to directories is currently not supported */
+	if (S_ISDIR(inode->i_mode))
+		return -ENODEV;
+
+	/* keep size option not supported */
+	if (mode & FALLOC_FL_KEEP_SIZE)
+		return -EOPNOTSUPP;
+
+	mutex_lock(&inode->i_mutex);
+	mutex_lock(&PRAM_I(inode)->truncate_mutex);
+	new_size = len + offset;
+	if (new_size > inode->i_size) {
+		ret = inode_newsize_ok(inode, new_size);
+		if (ret)
+			goto out;
+	}
+
+	blocksize_mask = (1 << inode->i_sb->s_blocksize_bits) - 1;
+	offset += inode->i_size;
+	blocknr = offset >> inode->i_sb->s_blocksize_bits;
+	blockoff = offset & blocksize_mask;
+	num_blocks = (blockoff + len + blocksize_mask) >>
+						inode->i_sb->s_blocksize_bits;
+	ret = pram_alloc_blocks(inode, blocknr, num_blocks);
+	if (ret)
+		goto out;
+	inode->i_mtime = inode->i_ctime = CURRENT_TIME_SEC;
+	if (new_size > inode->i_size)
+		inode->i_size = new_size;
+	ret = pram_update_inode(inode);
+ out:
+	mutex_unlock(&PRAM_I(inode)->truncate_mutex);
+	mutex_unlock(&inode->i_mutex);
+	return ret;
+}
+
 struct address_space_operations pram_aops = {
 	.readpage	= pram_readpage,
 	.direct_IO	= pram_direct_IO,
diff --git a/fs/pramfs/pram.h b/fs/pramfs/pram.h
index 074f350..4e4d331 100644
--- a/fs/pramfs/pram.h
+++ b/fs/pramfs/pram.h
@@ -53,8 +53,8 @@ extern int pram_mmap(struct file *file, struct vm_area_struct *vma);
 
 #define clear_opt(o, opt)	(o &= ~PRAM_MOUNT_##opt)
 #define set_opt(o, opt)		(o |= PRAM_MOUNT_##opt)
-#define test_opt(sb, opt)	(((struct pram_sb_info *)sb->s_fs_info)->s_mount_opt & \
-				 PRAM_MOUNT_##opt)
+#define test_opt(sb, opt) \
+	(((struct pram_sb_info *)sb->s_fs_info)->s_mount_opt & PRAM_MOUNT_##opt)
 
 /* balloc.c */
 extern void pram_init_bitmap(struct super_block *sb);
@@ -82,6 +82,8 @@ extern int pram_update_inode(struct inode *inode);
 extern int pram_write_inode(struct inode *inode, struct writeback_control *wbc);
 extern void pram_dirty_inode(struct inode *inode);
 extern int pram_notify_change(struct dentry *dentry, struct iattr *attr);
+extern long pram_fallocate(struct inode *inode, int mode, loff_t offset,
+			  loff_t len);
 
 
 /* super.c */
-- 
1.7.0.4

