From f4696a1db312ba60ae1fb6bc87255b2a4a41360d Mon Sep 17 00:00:00 2001
From: Zumeng Chen <zumeng.chen@windriver.com>
Date: Wed, 16 Nov 2011 14:06:49 +0800
Subject: [PATCH 12/18] fs: pramfs: Added support for splice system call

Upstream ID: 51dad1d20a5f2556fc7fad4e813a1246d4041275
  git://pramfs.git.sourceforge.net/gitroot/pramfs/pramfs
  From: Marco <marco@ncc-1701-f.site>

Integrated-by: Zumeng Chen <zumeng.chen@windriver.com>
---
 fs/pramfs/file.c |    1 +
 1 files changed, 1 insertions(+), 0 deletions(-)

diff --git a/fs/pramfs/file.c b/fs/pramfs/file.c
index db36c00..41b3c63 100644
--- a/fs/pramfs/file.c
+++ b/fs/pramfs/file.c
@@ -276,6 +276,7 @@ struct file_operations pram_file_operations = {
 	.fsync		= noop_fsync,
 	.check_flags	= pram_check_flags,
 	.unlocked_ioctl	= pram_ioctl,
+	.splice_read	= generic_file_splice_read,
 #ifdef CONFIG_COMPAT
 	.compat_ioctl	= pram_compat_ioctl,
 #endif
-- 
1.7.0.4

