From ab63d7179b377db455c4d79c742ab4062403feb9 Mon Sep 17 00:00:00 2001
From: Yang Shi <yang.shi@windriver.com>
Date: Wed, 10 Sep 2014 13:30:08 -0700
Subject: [PATCH 4/4] ktap: Make ns2usecs definition valid for module build

Signed-off-by: Yang Shi <yang.shi@windriver.com>
---
 kernel/trace/ktap/kp_transport.c | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/kernel/trace/ktap/kp_transport.c b/kernel/trace/ktap/kp_transport.c
index 521aa76..f922b9e 100644
--- a/kernel/trace/ktap/kp_transport.c
+++ b/kernel/trace/ktap/kp_transport.c
@@ -119,12 +119,16 @@ static void trace_consume(struct trace_iterator *iter)
 			    &iter->lost_events);
 }
 
+#ifdef CONFIG_KTAP_MODULE
 unsigned long long ns2usecs(cycle_t nsec)
 {
 	nsec += 500;
 	do_div(nsec, 1000);
 	return nsec;
 }
+#else
+extern unsigned long long ns2usecs(cycle_t nsec);
+#endif
 
 static int trace_print_timestamp(struct trace_iterator *iter)
 {
-- 
2.0.2

