From 739a1fd54519c2d383ea4433f7cd6bf1c3487d02 Mon Sep 17 00:00:00 2001
From: Yue Tao <Yue.Tao@windriver.com>
Date: Tue, 18 Dec 2012 18:23:09 +0800
Subject: [PATCH] Add Macro CSD_PORT_ID for P5040

Signed-off-by: Yue Tao <Yue.Tao@windriver.com>
---
 arch/powerpc/sysdev/fsl_pamu.c |    2 ++
 1 files changed, 2 insertions(+), 0 deletions(-)

diff --git a/arch/powerpc/sysdev/fsl_pamu.c b/arch/powerpc/sysdev/fsl_pamu.c
index 538ac5f..0d54764 100644
--- a/arch/powerpc/sysdev/fsl_pamu.c
+++ b/arch/powerpc/sysdev/fsl_pamu.c
@@ -352,6 +352,8 @@ irqreturn_t pamu_av_isr(int irq, void *arg)
 #define CSD_PORT_ID	0xFFF80000
 #elif CONFIG_P5020_DS || CONFIG_SBCP4080
 #define CSD_PORT_ID	0xFC000000
+#elif CONFIG_P5040_DS
+#define CSD_PORT_ID     0xFF800000 
 #endif
 
 enum law_size {
-- 
1.7.0

