From 99bf35507e27c334b8df9cce6d82385bd491c54d Mon Sep 17 00:00:00 2001
From: Kevin Hao <kexin.hao@windriver.com>
Date: Tue, 27 Mar 2012 16:33:02 +0800
Subject: [PATCH 21/25] netdev/dpa: replace {hard_}smp_processor_id with
 {get_hard,raw}_smp_processor_id in debug code

The {hard_}smp_processor_id are used in the dpa debug
code to get the processor id. Since most of these codes run
with preempt enabled, it will cause the following warning.
Just replace them with the {get_hard,raw}_smp_processor_id
to fix this issue.

BUG: using smp_processor_id() in preemptible [00000000] code: fmc/733
caller is E500_GetId+0x10/0x20
Call Trace:
[eae97d80] [c00072cc] show_stack+0x44/0x160 (unreliable)
[eae97db0] [c035fb50] debug_smp_processor_id+0xdc/0xec
[eae97dd0] [c04af35c] E500_GetId+0x10/0x20
[eae97de0] [c048cae4] FM_PORT_DeletePCD+0x258/0x3e0
[eae97e30] [c04aae84] LnxwrpFmPortIOCTL+0x4fc/0xf40
[eae97e60] [c04a93b8] fm_ioctl+0x104/0x140
[eae97e80] [c011a8b4] vfs_ioctl+0x3c/0xec
[eae97ea0] [c011ab2c] do_vfs_ioctl+0x80/0x758
[eae97f10] [c011b298] sys_ioctl+0x94/0x108
[eae97f40] [c00105dc] ret_from_syscall+0x0/0x4
---
 .../src/xx/integrations/P4080/xx_integration.c     |    2 +-
 drivers/net/dpa/NetCommSw/src/xx/xx_linux.c        |    2 +-
 2 files changed, 2 insertions(+), 2 deletions(-)

diff --git a/drivers/net/dpa/NetCommSw/src/xx/integrations/P4080/xx_integration.c b/drivers/net/dpa/NetCommSw/src/xx/integrations/P4080/xx_integration.c
index 71ed0e5..56e33e6 100644
--- a/drivers/net/dpa/NetCommSw/src/xx/integrations/P4080/xx_integration.c
+++ b/drivers/net/dpa/NetCommSw/src/xx/integrations/P4080/xx_integration.c
@@ -107,7 +107,7 @@
 /*Forced to introduce due to PRINT_FMT_PARAMS define*/
 uint32_t E500_GetId(void)
 {
-    return smp_processor_id();
+    return raw_smp_processor_id();
 }
 
 void * XX_GetMemPartitionBase(int memPartitionId)
diff --git a/drivers/net/dpa/NetCommSw/src/xx/xx_linux.c b/drivers/net/dpa/NetCommSw/src/xx/xx_linux.c
index 5f51336..041d48e 100644
--- a/drivers/net/dpa/NetCommSw/src/xx/xx_linux.c
+++ b/drivers/net/dpa/NetCommSw/src/xx/xx_linux.c
@@ -183,7 +183,7 @@ void XX_Print(char *str, ...)
 #ifdef CONFIG_SMP
     if (vsnprintf (buf, BUF_SIZE, str, args) >= BUF_SIZE)
         printk(KERN_WARNING "Illegal string to print!\n    more than %d characters.\n\tString was not printed completelly.\n", BUF_SIZE);
-    printk (KERN_CRIT "cpu%d/%d: %s",hard_smp_processor_id(), smp_processor_id(), buf);
+    printk(KERN_CRIT "cpu%d/%d: %s", get_hard_smp_processor_id(raw_smp_processor_id()), raw_smp_processor_id(), buf);
 #else
     vprintk(str, args);
 #endif /* CONFIG_SMP */
-- 
1.7.9.7

