From 5aac43f5e1052f9f5c97f02bc36c50875466f26d Mon Sep 17 00:00:00 2001
From: Yanjiang Jin <yanjiang.jin@windriver.com>
Date: Thu, 24 Oct 2013 17:32:37 +0800
Subject: [PATCH] MIPS: add -fno-pic for recordmcount

Add -fno-pic during using recordmcount to construct a table of the locations of
calls to 'mcount',  else it will construct a wrong Kernel module table, then
we would get the below errors during installing some modules, such as tipc, ext4.

Reserved instruction in kernel code[#1]:
CPU: 3 PID: 616 Comm: modprobe Not tainted 3.10.10-WR6.0.0.0_standard #5
task: c0000000ff2ce900 ti: c0000000fe880000 task.ti: c0000000fe880000
$ 0   : 0000000000000000 ffffffffc16e0258 ffffffffe002c420 0000000000016c4f
$ 4   : 0000000000000001 0000000000000001 0000000000000001 0000000000000003
$ 8   : 0000000000000000 ffffffffd0450001 ffffffffd0450002 0000000000000000
$12   : c0000000fe88fb90 ffffffffc13fcd68 0000000000000001 c0000001005bffff
$16   : ffffffffe003c198 0000000000000001 0000000000000003 ffffffffd0450002
$20   : ffffffffd0450001 0000000000000001 0000000000000001 ffffffffc1920000
$24   : 0000000000010000 00000000f5c8079f
$28   : c0000000fe880000 c0000000fe88fb28 0000000000000000 ffffffffe002cb48
Hi    : 000000000002fece
Lo    : 000000000000ff9a
epc   : ffffffffe002c420 nametbl_find_seq+0x0/0x78 [tipc]
    Not tainted
ra    : ffffffffe002cb48 tipc_nametbl_insert_publ+0x68/0x5d0 [tipc]
Status: 7010f8e3	KX SX UX KERNEL EXL IE
Cause : 80800028
PrId  : 000c1201 (XLP208 Rev A1)
Modules linked in: tipc(+)
Process modprobe (pid: 616, threadinfo=c0000000fe880000)
Stack : 0000000000000001 ffffffffc1148d0c c0000000ff596040 ffffffffc1051594
	  ffffffffe0040000 ffffffffc16e0198 ffffffffe003c198 ffffffffd0450002
	  ffffffffe0040000 ffffffffd0450001 0000000000000001 0000000000000001
	  ffffffffe003ab38 ffffffffc1920000 ffffffffc1920000 ffffffffe002d9a8
	  0000000000000001 0000000000000001 0000000000000003 ffffffffc16e0198
	  c0000000fea64c00 ffffffffd0450001 c0000000fe88fc38 0000000000000003
	  ffffffffc16d5128 ffffffffe0030f70 ffffffffc16d5128 0000000000000000
	  ffffffffe0040000 ffffffffe0040000 ffffffffe0040000 ffffffffe002c2b4
	  0000000000000000 ffffffffe003c170 0000000100000001 00000001e003ab38
	  ffffffffc1920000 0000000000000000 ffffffffc16d5128 ffffffffe00600d4
	  ...
Call Trace:
[<ffffffffe002c420>] nametbl_find_seq+0x0/0x78 [tipc]
[<ffffffffe002cb48>] tipc_nametbl_insert_publ+0x68/0x5d0 [tipc]
[<ffffffffe002d9a8>] tipc_nametbl_publish+0xa0/0x118 [tipc]
[<ffffffffe0030f70>] tipc_publish+0xc0/0x110 [tipc]
[<ffffffffe002c2b4>] tipc_subscr_start+0xb4/0x108 [tipc]
[<ffffffffe00600d4>] tipc_init+0xd4/0x178 [tipc]
[<ffffffffc10004f0>] do_one_initcall+0xf0/0x140

Code: 67bd0040  00000000  00000000 <10000004> 10000004

Signed-off-by: Yanjiang Jin <yanjiang.jin@windriver.com>
---
 scripts/Makefile.build | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/scripts/Makefile.build b/scripts/Makefile.build
index d5d859c..a2afcd8 100644
--- a/scripts/Makefile.build
+++ b/scripts/Makefile.build
@@ -266,6 +266,9 @@ ifdef BUILD_C_RECORDMCOUNT
 ifeq ("$(origin RECORDMCOUNT_WARN)", "command line")
   RECORDMCOUNT_FLAGS = -w
 endif
+ifeq ($(ARCH),mips)
+RECORDMCOUNT_FLAGS += -fno-pic
+endif
 # Due to recursion, we must skip empty.o.
 # The empty.o file is created in the make process in order to determine
 #  the target endianness and word size. It is made before all other C
-- 
1.8.4.93.g57e4c17

