From 275f39a03437e1bf3fe6990f0bcb146119f37bc2 Mon Sep 17 00:00:00 2001
From: Jonas Rendel <Jonas.Rendel@windriver.com>
Date: Thu, 27 Nov 2008 17:27:54 +0100
Subject: [PATCH] added missing ANT specific SO_VR and SO_PERMVR to getsockopt

---
 net/core/sock.c |    9 ++++++++-
 1 files changed, 8 insertions(+), 1 deletions(-)

diff --git a/net/core/sock.c b/net/core/sock.c
index af3c8e6..a2c383b 100644
--- a/net/core/sock.c
+++ b/net/core/sock.c
@@ -856,7 +856,14 @@ int sock_getsockopt(struct socket *sock, int level, int optname,
 	case SO_MARK:
 		v.val = sk->sk_mark;
 		break;
-
+#ifdef CONFIG_INTERPEAK
+    case SO_VR:
+        v.val = sk->sk_vr;
+        break;
+    case SO_PERMVR:
+        v.val = (int)current->vr;
+        break;
+#endif
 	default:
 		return -ENOPROTOOPT;
 	}
-- 
1.5.6

