From e6a555146d39813e5f7ba24675f7dcf5934669bb Mon Sep 17 00:00:00 2001
From: Xufeng Zhang <xufeng.zhang@windriver.com>
Date: Tue, 6 Mar 2012 12:56:39 +0800
Subject: [PATCH] fix compile error for undefined pax_task_size

Signed-off-by: Xufeng Zhang <xufeng.zhang@windriver.com>
---
 mm/mremap.c |    6 ++++++
 1 files changed, 6 insertions(+), 0 deletions(-)

diff --git a/mm/mremap.c b/mm/mremap.c
index 0b805bd..42cddd5 100644
--- a/mm/mremap.c
+++ b/mm/mremap.c
@@ -319,6 +319,12 @@ static unsigned long mremap_to(unsigned long addr,
 	unsigned long ret = -EINVAL;
 	unsigned long charged = 0;
 	unsigned long map_flags;
+	unsigned long pax_task_size = TASK_SIZE;
+
+#ifdef CONFIG_PAX_SEGMEXEC
+        if (current->mm->pax_flags & MF_PAX_SEGMEXEC)
+                pax_task_size = SEGMEXEC_TASK_SIZE;
+#endif
 
 	if (new_addr & ~PAGE_MASK)
 		goto out;
-- 
1.7.0.4

