From 1870da76f0ef24ea5eea5a1ab65e75afd05c8a16 Mon Sep 17 00:00:00 2001
From: Li Wang <li.wang@windriver.com>
Date: Fri, 5 Jul 2013 13:54:06 +0800
Subject: [PATCH 6/8] scripts: compile warning: different variable type
 comparison

warning: comparison between signed and unsigned integer expressions [-Wsign-compare]

Signed-off-by: Li Wang <li.wang@windriver.com>
---
 scripts/mod/file2alias.c                |    2 +-
 scripts/selinux/genheaders/genheaders.c |    3 ++-
 2 files changed, 3 insertions(+), 2 deletions(-)

diff --git a/scripts/mod/file2alias.c b/scripts/mod/file2alias.c
index a3119bd..dc95b24 100644
--- a/scripts/mod/file2alias.c
+++ b/scripts/mod/file2alias.c
@@ -816,7 +816,7 @@ ADD_TO_DEVTABLE("virtio", struct virtio_device_id, do_virtio_entry);
 static int do_vmbus_entry(const char *filename, struct hv_vmbus_device_id *id,
 			  char *alias)
 {
-	int i;
+	unsigned int i;
 	char guid_name[((sizeof(id->guid) + 1)) * 2];
 
 	for (i = 0; i < (sizeof(id->guid) * 2); i += 2)
diff --git a/scripts/selinux/genheaders/genheaders.c b/scripts/selinux/genheaders/genheaders.c
index 539855f..660e685 100644
--- a/scripts/selinux/genheaders/genheaders.c
+++ b/scripts/selinux/genheaders/genheaders.c
@@ -40,7 +40,8 @@ static char *stoupperx(const char *s)
 
 int main(int argc, char *argv[])
 {
-	int i, j, k;
+	int i;
+	unsigned int j, k;
 	int isids_len;
 	FILE *fout;
 	const char *needle = "SOCKET";
-- 
1.7.5.4

