From a9bce3f433a3bd3c76d3304750d8cd56731ec019 Mon Sep 17 00:00:00 2001
From: Li Wang <li.wang@windriver.com>
Date: Fri, 5 Jul 2013 13:40:14 +0800
Subject: [PATCH 2/8] scripts: compile warning: correct use setjmp() function

warning: variable 'xxx' might be clobbered by 'longjmp' or 'vfork' [-Wclobbered]

Signed-off-by: Li Wang <li.wang@windriver.com>
---
 scripts/recordmcount.c |    7 ++++++-
 1 files changed, 6 insertions(+), 1 deletions(-)

diff --git a/scripts/recordmcount.c b/scripts/recordmcount.c
index ee52cb8..b50a4af 100644
--- a/scripts/recordmcount.c
+++ b/scripts/recordmcount.c
@@ -67,6 +67,11 @@ fail_file(void)
 	longjmp(jmpenv, SJ_FAIL);
 }
 
+static int do_setjmp(void)
+{
+	return setjmp(jmpenv);
+}
+
 static void __attribute__((noreturn))
 succeed_file(void)
 {
@@ -432,7 +437,7 @@ main(int argc, char *argv[])
 	/* Process each file in turn, allowing deep failure. */
 	for (i = optind; i < argc; i++) {
 		char *file = argv[i];
-		int const sjval = setjmp(jmpenv);
+		int const sjval = do_setjmp();
 		int len;
 
 		/*
-- 
1.7.5.4

