From ac8cef6f7e0391124b7c978ac36000fdd5ecc383 Mon Sep 17 00:00:00 2001
From: Jianchuan Wang <jianchuan.wang@windriver.com>
Date: Wed, 26 Aug 2015 13:54:57 +0800
Subject: [PATCH] Omit to optimize vsprintf.c/kasprintf.c

Add "-O0" for vsprintf.c/kasprintf.c

Signed-off-by: Jianchuan Wang <jianchuan.wang@windriver.com>
---
 lib/Makefile | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/lib/Makefile b/lib/Makefile
index 6897b527581a..63d292697c6d 100644
--- a/lib/Makefile
+++ b/lib/Makefile
@@ -45,9 +45,13 @@ CFLAGS_kobject.o += -DDEBUG
 CFLAGS_kobject_uevent.o += -DDEBUG
 endif
 
+
 obj-$(CONFIG_DEBUG_INFO_REDUCED) += debug_info.o
 CFLAGS_debug_info.o += $(call cc-option, -femit-struct-debug-detailed=any)
 
+CFLAGS_vsprintf.o += -O0
+CFLAGS_asprintf.o += -O0
+
 obj-$(CONFIG_GENERIC_IOMAP) += iomap.o
 obj-$(CONFIG_GENERIC_PCI_IOMAP) += pci_iomap.o
 obj-$(CONFIG_HAS_IOMEM) += iomap_copy.o devres.o
-- 
2.1.0

