From 6a1e5cbb0fcffb4172e17ddd0ca5f9966ccaa8e4 Mon Sep 17 00:00:00 2001
From: Vaibhav Hiremath <hvaibhav@ti.com>
Date: Tue, 24 Jan 2012 11:24:18 +0530
Subject: [PATCH 148/609] ARM: OMAP: am335xevm: Fix pruss clock alias error

Original commit: 646170e70065e66d8d5fa112e9eb23617d38851e

  From git://arago-project.org/git/projects/linux-am33x.git
  And ti-sdk-am335x-evm-05.05.00.00-Linux-x86-Install image

Due to changes in clock tree the name of the clock has
changed, so TEMP create new alias entry for pruss.

NOTE: Revisit this commit/patch, based on final clock tree.
---
 arch/arm/mach-omap2/board-am335xevm.c |    6 ++++--
 1 files changed, 4 insertions(+), 2 deletions(-)

diff --git a/arch/arm/mach-omap2/board-am335xevm.c b/arch/arm/mach-omap2/board-am335xevm.c
index cd6096e..f7f296d 100644
--- a/arch/arm/mach-omap2/board-am335xevm.c
+++ b/arch/arm/mach-omap2/board-am335xevm.c
@@ -2135,10 +2135,12 @@ static void __init am335x_evm_init(void)
 	omap_board_config_size = ARRAY_SIZE(am335x_evm_config);
 	/* Create an alias for icss clock */
 	if (clk_add_alias("pruss", NULL, "icss_uart_gclk", NULL))
-		pr_err("failed to create an alias: icss_uart_gclk --> pruss\n");
+		pr_warn("failed to create an alias: icss_uart_gclk --> pruss\n");
+	if (clk_add_alias("pruss", NULL, "icss_fck", NULL))
+		pr_warn("failed to create an alias: icss_fck --> pruss\n");
 	/* Create an alias for gfx/sgx clock */
 	if (clk_add_alias("sgx_ck", NULL, "gfx_fclk", NULL))
-		pr_err("failed to create an alias: gfx_fclk --> sgx_ck\n");
+		pr_warn("failed to create an alias: gfx_fclk --> sgx_ck\n");
 }
 
 static void __init am335x_evm_map_io(void)
-- 
1.7.5.4

