From 1938e7ed36942e9ee0b373626341bfab89437c71 Mon Sep 17 00:00:00 2001
From: Ajay Kumar Gupta <ajay.gupta@ti.com>
Date: Tue, 6 Mar 2012 13:21:48 +0530
Subject: [PATCH 437/609] usb: musb: cppi41dma: use transparent mode for
 g_mass_stoarge

Original commit: 0da671ccc0813470e116e5e5eebba5b404e753f2

  From git://arago-project.org/git/projects/linux-am33x.git
  And ti-sdk-am335x-evm-05.05.00.00-Linux-x86-Install image

USB MSC as a device need to use transparent mode of CPPI4.1 DMA
and the same was getting done for g_file_storage. Updated it
for g_mass_storage also.

Merge to:
commit e141f0a5ab310a5d5ec9a937070577c7e5b0f4a7
musb:cppi41: use GRNDIS mode for rx-dma except file-storage gadget
---
 drivers/usb/musb/cppi41_dma.c |    3 ++-
 1 files changed, 2 insertions(+), 1 deletions(-)

diff --git a/drivers/usb/musb/cppi41_dma.c b/drivers/usb/musb/cppi41_dma.c
index ca23458..dc73253 100644
--- a/drivers/usb/musb/cppi41_dma.c
+++ b/drivers/usb/musb/cppi41_dma.c
@@ -765,7 +765,8 @@ static unsigned cppi41_next_rx_segment(struct cppi41_channel *rx_ch)
 
 		pkt_len = rx_ch->pkt_size;
 		mode = USB_GENERIC_RNDIS_MODE;
-		if (!strcmp(gadget_driver->driver.name, "g_file_storage")) {
+		if (!strcmp(gadget_driver->driver.name, "g_file_storage") ||
+			!strcmp(gadget_driver->driver.name, "g_mass_storage")) {
 			if (cppi->inf_mode && length > pkt_len) {
 				pkt_len = 0;
 				length = length - rx_ch->pkt_size;
-- 
1.7.5.4

