From a2399d0455cc887a00731fe4d829aa4d4ccdbcea Mon Sep 17 00:00:00 2001
From: Ajay Kumar Gupta <ajay.gupta@ti.com>
Date: Fri, 8 Jun 2012 16:36:05 +0530
Subject: [PATCH 458/609] usb: musb: ti81xx: fix wrong timer check during
 suspend

Original commit: 597b432fda2942c7303f43f67cc4aa5899e00934

  From git://arago-project.org/git/projects/linux-am33x.git
  And ti-sdk-am335x-evm-05.05.00.00-Linux-x86-Install image

Corrected the wrong check done for musb->otg_workaround instead of
musb->otg_timer for deleting musb->otg_timer. So far no issue is
seen due to this but fixing it anyways.
---
 drivers/usb/musb/ti81xx.c |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/drivers/usb/musb/ti81xx.c b/drivers/usb/musb/ti81xx.c
index 6b0eb9e..d0e7202 100644
--- a/drivers/usb/musb/ti81xx.c
+++ b/drivers/usb/musb/ti81xx.c
@@ -1505,7 +1505,7 @@ static void ti81xx_save_context(struct ti81xx_glue *glue)
 			musb->en_otgw_timer = 1;
 		}
 
-		if (timer_pending(&musb->otg_workaround) &&
+		if (timer_pending(&musb->otg_timer) &&
 					is_otg_enabled(musb)) {
 			del_timer_sync(&musb->otg_timer);
 			musb->en_otg_timer = 1;
-- 
1.7.5.4

