From 8abc92bdec5129fac8eb2c4a2b4c5d00ce7b4eae Mon Sep 17 00:00:00 2001
From: Ajay Kumar Gupta <ajay.gupta@ti.com>
Date: Wed, 4 Jul 2012 10:47:14 +0530
Subject: [PATCH 459/609] usb: musb: gadget: fix kernel crash after rmmod

Original commit: ef17643d4f1a61276f51a08163d74e3c4f597efe

  From git://arago-project.org/git/projects/linux-am33x.git
  And ti-sdk-am335x-evm-05.05.00.00-Linux-x86-Install image

musb->gadget_driver was not getting reset to NULL after the gadget
driver is removed. This leads to kernel panic while doing proc read
when it tries to print the gadget driver name.

Fixing the same by resetting the musb->gadget_driver to NULL when
gadget driver is removed.
---
 drivers/usb/musb/musb_gadget.c |    5 +++--
 1 files changed, 3 insertions(+), 2 deletions(-)

diff --git a/drivers/usb/musb/musb_gadget.c b/drivers/usb/musb/musb_gadget.c
index 7fd1799..153a7ec 100644
--- a/drivers/usb/musb/musb_gadget.c
+++ b/drivers/usb/musb/musb_gadget.c
@@ -584,7 +584,7 @@ void musb_g_tx(struct musb *musb, u8 epnum)
 			 * on SMP systems. Reselect the INDEX to be sure
 			 * we are reading/modifying the right registers
 			 */
-			musb_ep_select(mbase, epnum);
+			musb_ep_select(musb, mbase, epnum);
 			req = musb_ep->desc ? next_request(musb_ep) : NULL;
 			if (!req) {
 				dev_dbg(musb->controller, "%s idle now\n",
@@ -995,7 +995,7 @@ void musb_g_rx(struct musb *musb, u8 epnum)
 		 * on SMP systems. Reselect the INDEX to be sure
 		 * we are reading/modifying the right registers
 		 */
-		musb_ep_select(mbase, epnum);
+		musb_ep_select(musb, mbase, epnum);
 
 		req = next_request(musb_ep);
 		if (!req)
@@ -2072,6 +2072,7 @@ static int musb_gadget_stop(struct usb_gadget *g,
 	if (!is_otg_enabled(musb))
 		musb_stop(musb);
 
+	musb->gadget_driver = NULL;
 	pm_runtime_put(musb->controller);
 
 	return 0;
-- 
1.7.5.4

