From ec759ff78e67437f410a9c017ab9def175103b23 Mon Sep 17 00:00:00 2001
From: Zumeng Chen <zumeng.chen@windriver.com>
Date: Tue, 18 Sep 2012 10:30:01 +0800
Subject: [PATCH 596/609] musb: pm runtime compatible to 3530evm

Signed-off-by: Zumeng Chen <zumeng.chen@windriver.com>
---
 drivers/usb/musb/musb_gadget.c |   14 ++++++++++++--
 1 files changed, 12 insertions(+), 2 deletions(-)

diff --git a/drivers/usb/musb/musb_gadget.c b/drivers/usb/musb/musb_gadget.c
index 153a7ec..cc38d20 100644
--- a/drivers/usb/musb/musb_gadget.c
+++ b/drivers/usb/musb/musb_gadget.c
@@ -1207,7 +1207,9 @@ static int musb_gadget_disable(struct usb_ep *ep)
 	epnum = musb_ep->current_epnum;
 	epio = musb->endpoints[epnum].regs;
 
+#ifdef  CONFIG_USB_MUSB_OMAP2PLUS_GLUE
 	pm_runtime_get_sync(musb->controller);
+#endif
 	spin_lock_irqsave(&musb->lock, flags);
 	musb_ep_select(musb, musb->mregs, epnum);
 
@@ -1231,8 +1233,9 @@ static int musb_gadget_disable(struct usb_ep *ep)
 	nuke(musb_ep, -ESHUTDOWN);
 
 	spin_unlock_irqrestore(&(musb->lock), flags);
+#ifdef  CONFIG_USB_MUSB_OMAP2PLUS_GLUE
 	pm_runtime_put(musb->controller);
-
+#endif
 	schedule_work(&musb->irq_work);
 
 	dev_dbg(musb->controller, "%s\n", musb_ep->end_point.name);
@@ -1290,10 +1293,13 @@ void musb_ep_restart(struct musb *musb, struct musb_request *req)
 		req->tx ? "TX/IN" : "RX/OUT",
 		&req->request, req->request.length, req->epnum);
 
+#ifdef  CONFIG_USB_MUSB_OMAP2PLUS_GLUE
 	pm_runtime_get_sync(musb->controller);
+#endif
 	musb_ep_select(musb, musb->mregs, req->epnum);
+#ifdef  CONFIG_USB_MUSB_OMAP2PLUS_GLUE
 	pm_runtime_put(musb->controller);
-
+#endif
 	if (req->tx)
 		txstate(musb, req);
 	else
@@ -1542,7 +1548,9 @@ static void musb_gadget_fifo_flush(struct usb_ep *ep)
 	unsigned long	flags;
 	u16		csr, int_txe;
 
+#ifdef  CONFIG_USB_MUSB_OMAP2PLUS_GLUE
 	pm_runtime_get_sync(musb->controller);
+#endif
 	mbase = musb->mregs;
 
 	spin_lock_irqsave(&musb->lock, flags);
@@ -1576,7 +1584,9 @@ static void musb_gadget_fifo_flush(struct usb_ep *ep)
 	/* re-enable interrupt */
 	musb_writew(mbase, MUSB_INTRTXE, int_txe);
 	spin_unlock_irqrestore(&musb->lock, flags);
+#ifdef  CONFIG_USB_MUSB_OMAP2PLUS_GLUE
 	pm_runtime_put(musb->controller);
+#endif
 }
 
 static const struct usb_ep_ops musb_ep_ops = {
-- 
1.7.5.4

