From 578155fe125ec286cbcd5ebde20bd86809557e9e Mon Sep 17 00:00:00 2001
From: Liang Li <liang.li@windriver.com>
Date: Wed, 9 Mar 2011 16:40:37 +0800
Subject: [PATCH 1/2] WRHV: Cavium: enable mips GOS use ttyS1 on octeon board

Signed-off-by: Liang Li <liang.li@windriver.com>
---
 arch/mips/kernel/vbi/wrhv.c |    4 ++++
 1 files changed, 4 insertions(+), 0 deletions(-)

diff --git a/arch/mips/kernel/vbi/wrhv.c b/arch/mips/kernel/vbi/wrhv.c
index 265784f..6495750 100644
--- a/arch/mips/kernel/vbi/wrhv.c
+++ b/arch/mips/kernel/vbi/wrhv.c
@@ -170,6 +170,10 @@ void __init octeon_boot_info_init(void)
  	octeon_boot_descriptor_struct.board_type = CVMX_BOARD_TYPE_CUST_WRHV_OCTEON;
  	octeon_boot_descriptor_struct.cvmx_desc_vaddr = (uint64_t) &octeon_bootinfo_struct;
  
+	if (strstr(octeon_boot_descriptor_struct.argv[0], "console=ttyS1"))
+		octeon_boot_descriptor_struct.flags |=
+				 OCTEON_BL_FLAG_CONSOLE_UART1;
+
  	fw_arg3 = (unsigned long) &octeon_boot_descriptor_struct;
 }
 #endif /* CONFIG_CPU_CAVIUM_OCTEON */
-- 
1.6.5.2

