From b803c2d083bf4a38e9459367378a62dccab17fcb Mon Sep 17 00:00:00 2001
From: Gregory Haskins <ghaskins@novell.com>
Date: Fri, 28 Aug 2009 11:32:51 -0400
Subject: [PATCH 112/119] KVM: check mm in xinterface

Signed-off-by: Gregory Haskins <ghaskins@novell.com>
---
 virt/kvm/xinterface.c |    4 ++--
 1 files changed, 2 insertions(+), 2 deletions(-)

diff --git a/virt/kvm/xinterface.c b/virt/kvm/xinterface.c
index c02c441..f6db9f0 100644
--- a/virt/kvm/xinterface.c
+++ b/virt/kvm/xinterface.c
@@ -206,7 +206,7 @@ xinterface_copy_to(struct kvm_xinterface *intf, unsigned long gpa,
 	if (kthread)
 		use_mm(_intf->mm);
 
-	if (kthread || _intf->task == current)
+	if (kthread || _intf->mm == current->mm)
 		n = copy_to_user((void *)dst, src, n);
 	else
 		n = _slow_copy_to_user(_intf, dst, src, n);
@@ -280,7 +280,7 @@ xinterface_copy_from(struct kvm_xinterface *intf, void *dst,
 	if (kthread)
 		use_mm(_intf->mm);
 
-	if (kthread || _intf->task == current)
+	if (kthread || _intf->mm == current->mm)
 		n = copy_from_user(dst, (void *)src, n);
 	else
 		n = _slow_copy_from_user(_intf, dst, src, n);
-- 
1.6.5.2

