From 3eb8b8103c0c65466501dd5d46be649d5580c69e Mon Sep 17 00:00:00 2001
From: Gregory Haskins <ghaskins@novell.com>
Date: Tue, 22 Sep 2009 23:26:45 -0400
Subject: [PATCH 053/119] vbus: add zc support to the kvm-connector

Signed-off-by: Gregory Haskins <ghaskins@novell.com>
---
 kernel/vbus/kvm.c |   10 ++++++++++
 1 files changed, 10 insertions(+), 0 deletions(-)

diff --git a/kernel/vbus/kvm.c b/kernel/vbus/kvm.c
index 56e2c56..90a8531 100644
--- a/kernel/vbus/kvm.c
+++ b/kernel/vbus/kvm.c
@@ -205,6 +205,15 @@ _memctx_mm_get(struct vbus_memctx *ctx)
 	return mm;
 }
 
+static unsigned long
+_memctx_sg_map(struct vbus_memctx *ctx, struct scatterlist *sgl, int nelems)
+{
+	struct _memctx *_memctx = to_memctx(ctx);
+	struct kvm_xinterface *kvm = _memctx->vkvm->kvm;
+
+	return kvm->ops->sgmap(kvm, sgl, nelems, 0);
+}
+
 static void
 _memctx_release(struct vbus_memctx *ctx)
 {
@@ -218,6 +227,7 @@ static struct vbus_memctx_ops _memctx_ops = {
 	.copy_to   = &_memctx_copy_to,
 	.copy_from = &_memctx_copy_from,
 	.mm_get    = &_memctx_mm_get,
+	.sg_map    = &_memctx_sg_map,
 	.release   = &_memctx_release,
 };
 
-- 
1.6.5.2

