From d274c2b7f012ac987c14fb402a0e693644408b6a Mon Sep 17 00:00:00 2001
From: Gregory Haskins <ghaskins@novell.com>
Date: Mon, 28 Sep 2009 16:45:44 -0400
Subject: [PATCH 057/119] vbus: export vbus_find() routine

Signed-off-by: Gregory Haskins <ghaskins@novell.com>
---
 kernel/vbus/core.c |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/kernel/vbus/core.c b/kernel/vbus/core.c
index cfe96a8..606f179 100644
--- a/kernel/vbus/core.c
+++ b/kernel/vbus/core.c
@@ -585,8 +585,8 @@ out:
 	mutex_unlock(&vbus_root.lock);
 
 	return bus;
-
 }
+EXPORT_SYMBOL_GPL(vbus_find);
 
 struct vbus_root vbus_root;
 
-- 
1.6.5.2

