From fde46626c1313250868279e88d050b35b397d458 Mon Sep 17 00:00:00 2001
From: Gregory Haskins <ghaskins@novell.com>
Date: Tue, 17 Nov 2009 17:15:02 -0500
Subject: [PATCH 085/119] vbus: fix fastcall-channel leaks in kvm-connector

Signed-off-by: Gregory Haskins <ghaskins@novell.com>
---
 kernel/vbus/connectors/kvm.c |    8 ++++----
 1 files changed, 4 insertions(+), 4 deletions(-)

diff --git a/kernel/vbus/connectors/kvm.c b/kernel/vbus/connectors/kvm.c
index d4ff89b..e0cee4b 100644
--- a/kernel/vbus/connectors/kvm.c
+++ b/kernel/vbus/connectors/kvm.c
@@ -1297,19 +1297,19 @@ _fcc_assign_ioctl(struct vbus_kvm *vkvm,
 		return PTR_ERR(_shm);
 	}
 
+	_fcc->id   = args->vector;
 	_fcc->shm  = &_shm->shm;
 	_fcc->desc = (struct vbus_pci_fastcall_desc *)_fcc->shm->ptr;
 
 	mutex_lock(&vkvm->lock);
 
 	ret = radix_tree_insert(&vkvm->fastcalls.channels.radix,
-				args->vector, _fcc);
+				_fcc->id, _fcc);
 	if (ret < 0) {
 		vbus_shm_put(_fcc->shm);
 		kfree(_fcc);
-	}
-
-	list_add_tail(&_fcc->list, &vkvm->fastcalls.channels.list);
+	} else
+		list_add_tail(&_fcc->list, &vkvm->fastcalls.channels.list);
 
 	mutex_unlock(&vkvm->lock);
 
-- 
1.6.5.2

