From 897c2b4f857b77b980fb0803f7e466549c833f8e Mon Sep 17 00:00:00 2001
From: Gregory Haskins <ghaskins@novell.com>
Date: Tue, 17 Nov 2009 11:16:05 -0500
Subject: [PATCH 083/119] vbus: fix eventfd leak in kvm connector

Signed-off-by: Gregory Haskins <ghaskins@novell.com>
---
 kernel/vbus/connectors/kvm.c |    1 -
 1 files changed, 0 insertions(+), 1 deletions(-)

diff --git a/kernel/vbus/connectors/kvm.c b/kernel/vbus/connectors/kvm.c
index 5d895d0..d006137 100644
--- a/kernel/vbus/connectors/kvm.c
+++ b/kernel/vbus/connectors/kvm.c
@@ -1235,7 +1235,6 @@ _eventq_assign_ioctl(struct vbus_kvm *vkvm,
 	if (ret < 0)
 		goto fail;
 
-	get_file(input);
 	fd_install(fd, input); /* cannot fail after installing the fd */
 
 	eventq->eventfd = output;
-- 
1.6.5.2

