From b01d3baac9d238761e17940af6a42c2720a2ed04 Mon Sep 17 00:00:00 2001
From: Gregory Haskins <ghaskins@novell.com>
Date: Mon, 23 Nov 2009 12:17:12 -0500
Subject: [PATCH 086/119] vbus: make driver code properly declared as GPL

Signed-off-by: Gregory Haskins <ghaskins@novell.com>
---
 lib/ioq.c        |    4 ++++
 lib/shm_signal.c |    4 ++++
 2 files changed, 8 insertions(+), 0 deletions(-)

diff --git a/lib/ioq.c b/lib/ioq.c
index a6c8664..d5e57be 100644
--- a/lib/ioq.c
+++ b/lib/ioq.c
@@ -25,6 +25,10 @@
 #include <linux/bitops.h>
 #include <linux/module.h>
 
+MODULE_AUTHOR("Gregory Haskins");
+MODULE_LICENSE("GPL");
+MODULE_VERSION("1");
+
 #ifndef NULL
 #define NULL 0
 #endif
diff --git a/lib/shm_signal.c b/lib/shm_signal.c
index 3ea8c66..552f601 100644
--- a/lib/shm_signal.c
+++ b/lib/shm_signal.c
@@ -25,6 +25,10 @@
 #include <linux/shm_signal.h>
 #include <linux/workqueue.h>
 
+MODULE_AUTHOR("Gregory Haskins");
+MODULE_LICENSE("GPL");
+MODULE_VERSION("1");
+
 int shm_signal_enable(struct shm_signal *s, int flags)
 {
 	struct shm_signal_irq *irq = &s->desc->irq[s->locale];
-- 
1.6.5.2

