From 26a637e1f49416851a5785f136e4a8eba285d1df Mon Sep 17 00:00:00 2001
From: Gregory Haskins <ghaskins@novell.com>
Date: Wed, 26 Aug 2009 20:31:27 -0400
Subject: [PATCH 028/119] venettap: fix decode() return value to properly detect errors

Signed-off-by: Gregory Haskins <ghaskins@novell.com>
---
 kernel/vbus/devices/venet-tap.c |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/kernel/vbus/devices/venet-tap.c b/kernel/vbus/devices/venet-tap.c
index 2b81662..8a8fd3a 100644
--- a/kernel/vbus/devices/venet-tap.c
+++ b/kernel/vbus/devices/venet-tap.c
@@ -543,7 +543,7 @@ venettap_rx(struct venettap *priv)
 	 */
 	while (iter.desc->sown) {
 		struct sk_buff *skb = NULL;
-		size_t len;
+		int len;
 
 		len = rx_ops->decode(priv,
 				     (void *)iter.desc->ptr,
-- 
1.6.5.2

