From 606831b59c88be0d71e50ba426172d24b96a8dce Mon Sep 17 00:00:00 2001
From: Gregory Haskins <ghaskins@novell.com>
Date: Tue, 25 Aug 2009 20:34:56 -0400
Subject: [PATCH 026/119] venettap: fix burstthresh related compiler warning

Signed-off-by: Gregory Haskins <ghaskins@novell.com>
---
 kernel/vbus/devices/venet-tap.c |    1 -
 1 files changed, 0 insertions(+), 1 deletions(-)

diff --git a/kernel/vbus/devices/venet-tap.c b/kernel/vbus/devices/venet-tap.c
index 8a67624..2b81662 100644
--- a/kernel/vbus/devices/venet-tap.c
+++ b/kernel/vbus/devices/venet-tap.c
@@ -1560,7 +1560,6 @@ burstthresh_store(struct vbus_device *dev, struct vbus_device_attribute *attr,
 {
 	struct venettap *priv = vdev_to_priv(dev);
 	int val = -1;
-	int ret = 0;
 
 	if (count > 0)
 		sscanf(buf, "%d", &val);
-- 
1.6.5.2

