From de4439e080f10d08776662be295a855d5852101a Mon Sep 17 00:00:00 2001
From: Gregory Haskins <ghaskins@novell.com>
Date: Wed, 26 Aug 2009 20:54:24 -0400
Subject: [PATCH 030/119] venettap: verify max GSO size

Signed-off-by: Gregory Haskins <ghaskins@novell.com>
---
 kernel/vbus/devices/venet-tap.c |   19 ++++++++++++-------
 1 files changed, 12 insertions(+), 7 deletions(-)

diff --git a/kernel/vbus/devices/venet-tap.c b/kernel/vbus/devices/venet-tap.c
index 48bb9dc..6689e4d 100644
--- a/kernel/vbus/devices/venet-tap.c
+++ b/kernel/vbus/devices/venet-tap.c
@@ -343,13 +343,18 @@ venettap_sg_decode(struct venettap *priv, void *ptr, int len)
 	ret = ctx->ops->copy_from(ctx, vsg, ptr, len);
 	BUG_ON(ret);
 
-	/*
-	 * Non GSO type packets should be constrained by the MTU setting
-	 * on the host
-	 */
-	if (!(vsg->flags & VENET_SG_FLAG_GSO)
-	    && (vsg->len > (priv->netif.dev->mtu + ETH_HLEN)))
-		return -1;
+	if (vsg->flags & VENET_SG_FLAG_GSO) {
+		/* GSO packets shall not exceed 64k frames */
+		if (vsg->len > 65536)
+			return -1;
+
+	} else
+		/*
+		 * Non GSO type packets should be constrained by the MTU setting
+		 * on the host
+		 */
+		if (vsg->len > (priv->netif.dev->mtu + ETH_HLEN))
+			return -1;
 
 	return vsg->len;
 }
-- 
1.6.5.2

