From 1e9996a8b6d0045dd47e9e2bd79cbe4f411371e5 Mon Sep 17 00:00:00 2001
From: Ming Liu <ming.liu@windriver.com>
Date: Wed, 27 Jul 2011 16:54:36 +0800
Subject: [PATCH] common_pc:topology: fix macro miss error

While CONFIG_NUMA and CONFIG_X86_32 are enabled by `make menuconfig`,
we will miss `node_to_cpumask_ptr` macro which used by `node_to_first_cpu`
lower, that causes compiling error.

This patch adds `node_to_cpumask_ptr` and `node_to_cpumask_ptr_next` into
linux/include/asm-x86/topology.h to solve that problem, these two macros
are copied from linux/include/asm-generic/topology.h.

Signed-off-by: Ming Liu <ming.liu@windriver.com>
---
 include/asm-x86/topology.h |    8 ++++++++
 1 files changed, 8 insertions(+), 0 deletions(-)

diff --git a/include/asm-x86/topology.h b/include/asm-x86/topology.h
index 90ac771..bb1b53d 100644
--- a/include/asm-x86/topology.h
+++ b/include/asm-x86/topology.h
@@ -68,6 +68,14 @@ static inline cpumask_t node_to_cpumask(int node)
 	return node_to_cpumask_map[node];
 }
 
+/* returns pointer to cpumask for specified node */
+#define	node_to_cpumask_ptr(v, node) 					\
+		cpumask_t _##v = node_to_cpumask(node);			\
+		const cpumask_t *v = &_##v
+
+#define node_to_cpumask_ptr_next(v, node)				\
+			  _##v = node_to_cpumask(node)
+
 #else /* CONFIG_X86_64 */
 
 /* Mappings between node number and cpus on that node. */
-- 
1.7.0.4

