From af74f58ef6ffdc68cc77d314ad31bd6a322309d3 Mon Sep 17 00:00:00 2001
From: Bruce Ashfield <bruce.ashfield@windriver.com>
Date: Tue, 12 Jun 2012 16:51:14 -0400
Subject: [PATCH] routerstationpro: lib/kobject_uevent: switch to
 uevent_sock_mutex

-stable commit: 377c2f4aa985f937d77a6bb9b938b4deda2dc282 removed
the old sequence_lock and replaced it with uevent_sock_mutex.
Switching to the new lock to fix the resulting build error.

Signed-off-by: Bruce Ashfield <bruce.ashfield@windriver.com>
---
 lib/kobject_uevent.c |    4 ++--
 1 files changed, 2 insertions(+), 2 deletions(-)

diff --git a/lib/kobject_uevent.c b/lib/kobject_uevent.c
index c7cb2ed..df64823 100644
--- a/lib/kobject_uevent.c
+++ b/lib/kobject_uevent.c
@@ -54,9 +54,9 @@ u64 uevent_next_seqnum(void)
 {
 	u64 seq;
 
-	spin_lock(&sequence_lock);
+	mutex_lock(&uevent_sock_mutex);
 	seq = ++uevent_seqnum;
-	spin_unlock(&sequence_lock);
+	mutex_unlock(&uevent_sock_mutex);
 
 	return seq;
 }
-- 
1.7.5.4

