From 3d65107e71e6ec41b90326800512338c2c4e6c6f Mon Sep 17 00:00:00 2001
From: jiaxi liu <jiaxi.liu@windriver.com>
Date: Wed, 29 Sep 2010 16:05:28 +0800
Subject: [PATCH] Fix typo in e1000_remove

'unregister_netdev' called twice from e1000_remove. this patch
delete one line.
---
 drivers/net/e1000e/netdev.c |    1 -
 1 files changed, 0 insertions(+), 1 deletions(-)

diff --git a/drivers/net/e1000e/netdev.c b/drivers/net/e1000e/netdev.c
index 1b41536..52ba830 100644
--- a/drivers/net/e1000e/netdev.c
+++ b/drivers/net/e1000e/netdev.c
@@ -5333,7 +5333,6 @@ static void __devexit e1000_remove(struct pci_dev *pdev)
 	 */
 	e1000_release_hw_control(adapter);
 
-	unregister_netdev(netdev);
 
 	if (!e1000_check_reset_block(&adapter->hw))
 		e1000_phy_hw_reset(&adapter->hw);
-- 
1.7.0.4

