From 9c96ceac3ee86ae600bf2af1bba373cfa9b8978f Mon Sep 17 00:00:00 2001
From: Yue Tao <yue.tao@windriver.com>
Date: Sun, 15 Jun 2014 23:50:53 -0700
Subject: [PATCH] Miss header file when porting upstream f77b86d7

Signed-off-by: Yue Tao <yue.tao@windriver.com>
---
 net/wireless/reg.c |    1 +
 1 files changed, 1 insertions(+), 0 deletions(-)

diff --git a/net/wireless/reg.c b/net/wireless/reg.c
index 3ff410e..7e6471a 100644
--- a/net/wireless/reg.c
+++ b/net/wireless/reg.c
@@ -53,6 +53,7 @@
 #include <linux/nl80211.h>
 #include <linux/platform_device.h>
 #include <linux/moduleparam.h>
+#include <linux/rtnetlink.h>
 #include <net/cfg80211.h>
 #include "core.h"
 #include "reg.h"
-- 
1.7.5.4

