From a3ca5814c070bbd6166f59df14892775a03f0d99 Mon Sep 17 00:00:00 2001
From: Robin Gong <b38343@freescale.com>
Date: Wed, 11 Mar 2015 15:20:51 +0800
Subject: [PATCH 0179/1221] MLK-10404-3: firmware: imx: sdma: use XCH mode in
 escspi tx script

Use XCH mode in ecspi tx script to workaround TKT238285.

Signed-off-by: Robin Gong <b38343@freescale.com>
(cherry picked from commit 97edb3bedaa1c314f628a930447b7626d188acc5)
[zou:Original patch taken from
git.freescale.com/imx/fsl-arm-yocto-bsp.git-b imx-3.14.52-1.1.0_ga]
Signed-off-by: zou cao <cao.zou@windriver.com>
---
 firmware/imx/sdma/sdma-imx6q.bin.ihex |  235 +++++++++++++++---------------
 firmware/imx/sdma/sdma-imx7d.bin.ihex |  263 +++++++++++++++++----------------
 2 files changed, 250 insertions(+), 248 deletions(-)

diff --git a/firmware/imx/sdma/sdma-imx6q.bin.ihex b/firmware/imx/sdma/sdma-imx6q.bin.ihex
index 9f1977c..9ecea5f 100755
--- a/firmware/imx/sdma/sdma-imx6q.bin.ihex
+++ b/firmware/imx/sdma/sdma-imx6q.bin.ihex
@@ -1,130 +1,131 @@
 :1000000053444D4103000000010000001C000000AB
-:1000100029000000C00000004C0700008202000020
+:1000100029000000C00000005E070000820200000E
 :10002000FFFFFFFF00000000FFFFFFFFFFFFFFFFDC
 :10003000FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFD0
-:10004000FFFFFFFFFFFFFFFFFC1900004B04000054
-:10005000EB020000A8180000FFFFFFFF5A1A000083
+:10004000FFFFFFFFFFFFFFFF051A00004B0400004A
+:10005000EB020000B1180000FFFFFFFF631A000071
 :10006000FFFFFFFFC0030000FFFFFFFFFFFFFFFFD9
 :10007000FFFFFFFFAB020000FFFFFFFF7B0300005D
 :10008000FFFFFFFFFFFFFFFF4C0400006E040000B6
 :10009000FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF70
-:1000A00000000000001800004F180000A819000010
-:1000B000091B0000A41A0000401B000000180000EB
+:1000A000000000000018000058180000B1190000FE
+:1000B000121B0000AD1A0000491B000000180000D0
 :1000C000E3C1DB57F352016AFB52D36AFB521C1A9D
 :1000D000C36AE8621102FF3A3008D002337C8F0015
 :1000E000D500017D8D00A005EB5D7804037D7904CA
 :1000F0001C7D207C7904157CEE566006057D096523
-:10010000287E0A62267E28980A62237E0965217E5F
-:1001100012051205AD02C86A1C7F03200048ED7C61
+:10010000317E0A622F7E28980A622C7E09652A7E3B
+:1001100012051205AD02C86A257F03200048ED7C58
 :10012000409802780962C86A3F98150002780A620E
-:10013000C86A3F981500150002780B62C86A097CEE
-:10014000DF6D077F0000EB55004D077DFAC1DB57DF
-:1001500006980700CC680C6813C20AC20198D9C17E
-:10016000E3C1DB57E35FE357F352216A8F00D50009
-:10017000017D8D00A005EB5DFB567804037D7904BD
-:100180002A7D317C7904207C700B1103EB53000F26
-:100190006003057D0965377E0A62357E73980A62C1
-:1001A000327E0965307E12051205AD026007027CC1
-:1001B000065A7B98265A277F011F03200048E87CB7
-:1001C000700B110313539C98150004780962065AAA
-:1001D0000962265A9B981500150004780A62065A8F
-:1001E0000A62265A9B9815001500150004780B62C8
-:1001F000065A0B62265A077C0000EB55004D067D1F
-:10020000FAC1E357569807000C6813C20AC2539804
-:10021000700B110313536C07017CD9C1FB5E8A0676
-:100220006B07017CD9C1F35EDB59D3588F011001F4
-:100230000F398B003CC12B7DC05AC85B4EC1277C57
-:1002400088038906E35CFF0D1105FF1DBC053E0711
-:10025000004D187D700811007E07097D7D07027D25
-:100260002852D398F852DB54BC02CC02097C7C079C
-:10027000027D2852DC98F852D354BC02CC02097D8E
-:100280000004CA988B00C052C85359C1D67D0002E1
-:10029000BA98FF08BF007F07157D8804D500017D4F
-:1002A0008D00A005EB5D8F0212021202FF3ADA0503
-:1002B000027C3E070799A402DD02027D3E070799F2
-:1002C0005E070799EB55004D027C9805EB5DF352F4
-:1002D000FB546A07267D6C07017D44996B07577CA8
-:1002E0006907047D6807027D010E1E999358D600A8
-:1002F000017D8E009355A005935DA00602780255FE
-:10030000045D1D7C004E087C6907037D0255177E45
-:100310002B99045D147F890693500048017D16993E
-:100320008F99150006780255045D4F070255245D2C
-:100330002F07017C8F9917006F07017C01209355CF
-:100340009D00070096D9E298D36C6907047D680781
-:10035000027D010E53999358D600017D8E0093556E
-:10036000A005935DA00602780255C86D0F7C004E73
-:10037000087C6907037D0255097E6099C86D067F78
-:10038000890693500048017D4B998F998999C36ADA
-:100390006907047D6807027D010E76999358D6009F
-:1003A000017D8E009355A005935DA0060278C86577
-:1003B000045D0F7C004E087C6907037DC865097EDB
-:1003C0008399045D067F890693500048017D6E99EC
-:1003D0008F9993559D000700FF6C96D9E298000015
-:1003E000E354EB55004D017CE298CA98E354EB5579
-:1003F000FF0A1102FF1A7F07027CA005A3999D0046
-:100400008C05BA05A0051002BA04AD045404060018
-:10041000D9C1E3C1DB57F352056A8F00D500017DD6
-:100420008D00A005FB567804037D7904297D1F7C8F
-:1004300079042E7CE35D700D1105ED55000F60070A
-:10044000027D0652C4992652337E6005027D100259
-:100450001202096A2D7F1202096A2A7F1202096AB2
-:10046000277F011F03200048EA7CE355EF99150020
-:1004700015001500047806520B6A26520B6AEE9995
-:1004800015001500047806520A6A26520A6AEE9987
-:10049000150004780652096A2652096A097C286AFE
-:1004A000077F0000DB57004D057DFAC1DB57AD9992
-:1004B00077C254040AC2AA99E3C1DB57F352056A12
-:1004C000FB568E02941AC36AC8626902267D941E86
-:1004D000C36ED36EC8624802C86A9426981EC36E63
-:1004E000C8629826C36E60020E7D981EC36EC862F5
-:1004F0006C02037D9826C36E3D9A4C02D36EC86A87
-:100500009826C36E4D9AC8626E02017C479A096AAA
-:100510001A7F0125004D217D0E9AE36E8F00D805CC
-:10052000017D8D00C8626E02127D096A0C7F012573
-:100530000120F87CDB57004D107D286A047F000005
-:10054000FAC1DB57009A070004620C6A4A9A286ACB
-:10055000FA7F04627AC258045404FF081100FF189D
-:10056000BC00CD00017C3D9A286AED7F04627AC20E
-:100570000AC2FD99D9C1E3C1DB57F352056AFB56A4
-:100580008E02941A025269021D7D941E0652480280
-:10059000065A9426981E06524C02065A9826981E11
-:1005A000065260020A7C982606526E02237D096A72
-:1005B0001D7F0125004D247D6F9A286A177F0462F4
-:1005C0007AC2A09A8F00D805017D8D00A005065241
-:1005D0006E02107D096A0A7F0120F97C286A067F75
-:1005E0000000004D0D7DFAC1DB575F9A07000462E1
-:1005F0000C6A9D9A286AFA7F04627AC258045404ED
-:10060000286AF47F0AC25C9AE8DADB57F352056A7B
-:10061000C7698F00D500017D8D00A0057804037D9A
-:1006200079041C7D157C79041E7CEE56C862287EF8
-:100630006006027D10021202096A227F1202096A14
-:100640001F7F1202096A1C7F03200048EF7CD79AA3
-:10065000150015000278C8620B6AD69A1500027858
-:10066000C8620A6AD69A0278C862096A097C286A4E
-:10067000077F004D077DD352010802580004A89A55
-:1006800077C25404D35201080258010400DBA59A32
-:10069000700B11031353DB5F0A07D3588B00FB5E0B
-:1006A0003CC1157DC05AC85B4EC1117C8803F05D0A
-:1006B000FF0D1105FF1DBC05004D047DFF9A0807C5
-:1006C000DB5F0A078B00C352CB5359C1EC7D00029C
-:1006D000F09A016E0B612F7E0B622D7E0B632B7ED9
-:1006E0000C0D1704170417049D04081DCC05017C8C
-:1006F0000C0DD16A000F4207C86FDD6F1C7F8E00A2
-:100700009D0001680B67177ED56B04080278C86FDF
-:100710001207117C0B670F7E04080278C86F12075E
-:100720000A7CDD6F087FD169010FC86FDD6F037F21
-:1007300001010004159B0700FF680C680002159B6F
-:1007400085DBDB57F352056AFB52C76A8F00D50081
-:10075000017D8D00A0057804037D79041C7D157C46
-:1007600079041E7CEE56C862287E6006027D100267
-:100770001202096A227F1202096A1F7F1202096AA5
-:100780001C7F03200048EF7C749B15001500027845
-:10079000C8620B6A739B15000278C8620A6A739B71
-:1007A0000278C862096A097C286A077F004D077DC4
-:1007B000D352010802580004449B77C25404D35218
-:1007C0000108025801049DDB419B700B1103135378
-:1007D000DB5F0A07D3588B00FB5E3CC1157DC05A16
-:1007E000C85B4EC1117C8803F05DFF0D1105FF1D34
-:1007F000BC05004D047D9C9B0807DB5F0A078B004E
-:0C080000C352CB5359C1EC7D00028D9B0C
+:10013000C86A3F981500150002780B62C86A127CE5
+:10014000DF6D107FFB52041AC36AD36AC862042AA7
+:10015000C86AFB52D36A0000EB55004D077DFAC117
+:10016000DB5706980700CC680C6813C20AC20198D6
+:10017000D9C1E3C1DB57E35FE357F352216A8F0034
+:10018000D500017D8D00A005EB5DFB567804037D55
+:1001900079042A7D317C7904207C700B1103EB53A8
+:1001A000000F6003057D0965377E0A62357E7C9805
+:1001B0000A62327E0965307E12051205AD026007C3
+:1001C000027C065A8498265A277F011F0320004884
+:1001D000E87C700B11031353A5981500047809628D
+:1001E000065A0962265AA4981500150004780A6276
+:1001F000065A0A62265AA4981500150015000478BC
+:100200000B62065A0B62265A077C0000EB55004D24
+:10021000067DFAC1E3575F9807000C6813C20AC253
+:100220005C98700B110313536C07017CD9C1FB5E02
+:100230008A066B07017CD9C1F35EDB59D3588F0165
+:1002400010010F398B003CC12B7DC05AC85B4EC1D9
+:10025000277C88038906E35CFF0D1105FF1DBC05A3
+:100260003E07004D187D700811007E07097D7D074F
+:10027000027D2852DC98F852DB54BC02CC02097C87
+:100280007C07027D2852E598F852D354BC02CC0278
+:10029000097D0004D3988B00C052C85359C1D67D44
+:1002A0000002C398FF08BF007F07157D8804D500B2
+:1002B000017D8D00A005EB5D8F0212021202FF3A54
+:1002C000DA05027C3E071099A402DD02027D3E079A
+:1002D00010995E071099EB55004D027C9805EB5D77
+:1002E000F352FB546A07267D6C07017D4D996B071D
+:1002F000577C6907047D6807027D010E2799935892
+:10030000D600017D8E009355A005935DA00602786E
+:100310000255045D1D7C004E087C6907037D025573
+:10032000177E3499045D147F890693500048017D3F
+:100330001F999899150006780255045D4F070255DC
+:10034000245D2F07017C989917006F07017C01201D
+:1003500093559D0007009FD9EB98D36C6907047DE6
+:100360006807027D010E5C999358D600017D8E00CE
+:100370009355A005935DA00602780255C86D0F7CC9
+:10038000004E087C6907037D0255097E6999C86D96
+:10039000067F890693500048017D54999899929957
+:1003A000C36A6907047D6807027D010E7F9993582F
+:1003B000D600017D8E009355A005935DA0060278BE
+:1003C000C865045D0F7C004E087C6907037DC86525
+:1003D000097E8C99045D067F890693500048017D53
+:1003E0007799989993559D000700FF6C9FD9EB98DA
+:1003F0000000E354EB55004D017CEB98D398E35497
+:10040000EB55FF0A1102FF1A7F07027CA005AC9989
+:100410009D008C05BA05A0051002BA04AD04540471
+:100420000600D9C1E3C1DB57F352056A8F00D5003E
+:10043000017D8D00A005FB567804037D7904297D9C
+:100440001F7C79042E7CE35D700D1105ED55000FC6
+:100450006007027D0652CD992652337E6005027DEB
+:1004600010021202096A2D7F1202096A2A7F120203
+:10047000096A277F011F03200048EA7CE355F899A9
+:10048000150015001500047806520B6A26520B6AF7
+:10049000F79915001500047806520A6A26520A6A6E
+:1004A000F799150004780652096A2652096A097CF0
+:1004B000286A077F0000DB57004D057DFAC1DB5736
+:1004C000B69977C254040AC2B399E3C1DB57F35219
+:1004D000056AFB568E02941AC36AC8626902267DB9
+:1004E000941EC36ED36EC8624802C86A9426981ED2
+:1004F000C36EC8629826C36E60020E7D981EC36EDE
+:10050000C8626C02037D9826C36E469A4C02D36E75
+:10051000C86A9826C36E569AC8626E02017C509AC9
+:10052000096A1A7F0125004D217D179AE36E8F001D
+:10053000D805017D8D00C8626E02127D096A0C7FAC
+:1005400001250120F87CDB57004D107D286A047FCF
+:100550000000FAC1DB57099A070004620C6A539A3B
+:10056000286AFA7F04627AC258045404FF08110012
+:10057000FF18BC00CD00017C469A286AED7F04621A
+:100580007AC20AC2069AD9C1E3C1DB57F352056A9F
+:10059000FB568E02941A025269021D7D941E065269
+:1005A0004802065A9426981E06524C02065A98266D
+:1005B000981E065260020A7C982606526E02237D1F
+:1005C000096A1D7F0125004D247D789A286A177FCE
+:1005D00004627AC2A99A8F00D805017D8D00A0051A
+:1005E00006526E02107D096A0A7F0120F97C286A92
+:1005F000067F0000004D0D7DFAC1DB57689A0700A9
+:1006000004620C6AA69A286AFA7F04627AC25804C5
+:100610005404286AF47F0AC2659AF1DADB57F35270
+:10062000056AC7698F00D500017D8D00A00578049B
+:10063000037D79041C7D157C79041E7CEE56C8620E
+:10064000287E6006027D10021202096A227F1202D1
+:10065000096A1F7F1202096A1C7F03200048EF7C91
+:10066000E09A150015000278C8620B6ADF9A15003F
+:100670000278C8620A6ADF9A0278C862096A097C4D
+:10068000286A077F004D077DD352010802580004F5
+:10069000B19A77C25404D35201080258010409DB0D
+:1006A000AE9A700B11031353DB5F0A07D3588B000C
+:1006B000FB5E3CC1157DC05AC85B4EC1117C8803EE
+:1006C000F05DFF0D1105FF1DBC05004D047D089B6D
+:1006D0000807DB5F0A078B00C352CB5359C1EC7D7F
+:1006E0000002F99A016E0B612F7E0B622D7E0B6367
+:1006F0002B7E0C0D1704170417049D04081DCC0550
+:10070000017C0C0DD16A000F4207C86FDD6F1C7FA2
+:100710008E009D0001680B67177ED56B0408027878
+:10072000C86F1207117C0B670F7E04080278C86F30
+:1007300012070A7CDD6F087FD169010FC86FDD6F7A
+:10074000037F010100041E9B0700FF680C68000284
+:100750001E9B8EDBDB57F352056AFB52C76A8F0084
+:10076000D500017D8D00A0057804037D79041C7DF2
+:10077000157C79041E7CEE56C862287E6006027DD8
+:1007800010021202096A227F1202096A1F7F1202F6
+:10079000096A1C7F03200048EF7C7D9B1500150033
+:1007A0000278C8620B6A7C9B15000278C8620A6AEC
+:1007B0007C9B0278C862096A097C286A077F004D21
+:1007C000077DD3520108025800044D9B77C25404A0
+:1007D000D352010802580104A6DB4A9B700B110397
+:1007E0001353DB5F0A07D3588B00FB5E3CC1157DBA
+:1007F000C05AC85B4EC1117C8803F05DFF0D110526
+:10080000FF1DBC05004D047DA59B0807DB5F0A07A3
+:0E0810008B00C352CB5359C1EC7D0002969B66
 :00000001FF
diff --git a/firmware/imx/sdma/sdma-imx7d.bin.ihex b/firmware/imx/sdma/sdma-imx7d.bin.ihex
index 5b16d2e..91a6711 100644
--- a/firmware/imx/sdma/sdma-imx7d.bin.ihex
+++ b/firmware/imx/sdma/sdma-imx7d.bin.ihex
@@ -1,144 +1,145 @@
 :1000000053444D4104000000010000001C000000AA
-:100010002A000000C400000026080000840200003E
+:100010002A000000C400000038080000840200002C
 :10002000FFFFFFFF00000000FFFFFFFFFFFFFFFFDC
 :10003000FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFD0
-:10004000FFFFFFFFFFFFFFFFFC190000FFFFFFFFA7
-:10005000ED020000A8180000FFFFFFFF5A1A000081
+:10004000FFFFFFFFFFFFFFFF051A0000FFFFFFFF9D
+:10005000ED020000B1180000FFFFFFFF631A00006F
 :10006000FFFFFFFFC2030000FFFFFFFFFFFFFFFFD7
 :10007000FFFFFFFFAD020000FFFFFFFF7D03000059
 :10008000FFFFFFFFFFFFFFFF4E04000070040000B2
 :10009000FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF70
-:1000A00000000000001800004F180000A819000010
-:1000B000091B0000A41A0000401B000000180000EB
-:1000C000A61B0000E5C1DB57F352016AFB52D36A5D
+:1000A000000000000018000058180000B1190000FE
+:1000B000121B0000AD1A0000491B000000180000D0
+:1000C000AF1B0000E5C1DB57F352016AFB52D36A54
 :1000D000FB521C1AC36AE8621102FF3A3008D002D0
 :1000E000337C8F00D500017D8D00A005EB5D780489
 :1000F000037D79041C7D207C7904157CEE56600616
-:10010000057D0965287E0A62267E28980A62237E7C
-:100110000965217E12051205AD02C86A1C7F032005
+:10010000057D0965317E0A622F7E28980A622C7E61
+:1001100009652A7E12051205AD02C86A257F0320F3
 :100120000048ED7C409802780962C86A3F98150043
 :1001300002780A62C86A3F981500150002780B62BF
-:10014000C86A097CDF6D077F0000EB55004D077D15
-:10015000FCC1DB5706980700CC680C6815C20CC2BE
-:100160000198D9C1E5C1DB57E35FE357F352216A38
-:100170008F00D500017D8D00A005EB5DFB56780456
-:10018000037D79042A7D317C7904207C700B110376
-:10019000EB53000F6003057D0965377E0A62357EEB
-:1001A00073980A62327E0965307E12051205AD022F
-:1001B0006007027C065A7B98265A277F011F03207E
-:1001C0000048E87C700B110313539C9815000478C9
-:1001D0000962065A0962265A9B9815001500047890
-:1001E0000A62065A0A62265A9B98150015001500E5
-:1001F00004780B62065A0B62265A077C0000EB5506
-:10020000004D067DFCC1E357569807000C6815C2E7
-:100210000CC25398700B110313536C07017CD9C1A6
-:10022000FB5E8A066B07017CD9C1F35EDB59D358AC
-:100230008F0110010F398B003CC12B7DC05AC85B68
-:100240004EC1277C88038906E35CFF0D1105FF1D65
-:10025000BC053E07004D187D700811007E07097D22
-:100260007D07027D2852D398F852DB54BC02CC02A1
-:10027000097C7C07027D2852DC98F852D354BC02DA
-:10028000CC02097D0004CA988B00C052C85359C1E2
-:10029000D67D0002BA98FF08BF007F07157D88044D
-:1002A000D500017D8D00A005EB5D8F0212021202C8
-:1002B000FF3ADA05027C3E070799A402DD02027DBF
-:1002C0003E0707995E070799EB55004D027C98059C
-:1002D000EB5DF352FB546A07267D6C07017D449960
-:1002E0006B07577C6907047D6807027D010E1E9924
-:1002F0009358D600017D8E009355A005935DA0060E
-:1003000002780255045D1D7C004E087C6907037D60
-:100310000255177E2B99045D147F8906935000487F
-:10032000017D16998F99150006780255045D4F07D7
-:100330000255245D2F07017C8F9917006F07017C00
-:10034000012093559D00070096D9E298D36C690768
-:10035000047D6807027D010E53999358D600017DF4
-:100360008E009355A005935DA00602780255C86DD6
-:100370000F7C004E087C6907037D0255097E609959
-:10038000C86D067F890693500048017D4B998F996F
-:100390008999C36A6907047D6807027D010E769911
-:1003A0009358D600017D8E009355A005935DA0065D
-:1003B0000278C865045D0F7C004E087C6907037DE8
-:1003C000C865097E8399045D067F890693500048BD
-:1003D000017D6E998F9993559D000700FF6C96D90A
-:1003E000E2980000E354EB55004D017CE298CA9876
-:1003F000E354EB55FF0A1102FF1A7F07027CA005A8
-:10040000A3999D008C05BA05A0051002BA04AD049D
-:1004100054040600D9C1E5C1DB57F352056A8F00C9
-:10042000D500017D8D00A005FB567804037D79047D
-:10043000297D1F7C79042E7CE35D700D1105ED553F
-:10044000000F6007027D0652C4992652337E600574
-:10045000027D10021202096A2D7F1202096A2A7FA8
-:100460001202096A277F011F03200048EA7CE35536
-:10047000EF99150015001500047806520B6A2652F4
-:100480000B6AEE9915001500047806520A6A265286
-:100490000A6AEE99150004780652096A2652096A1A
-:1004A000097C286A077F0000DB57004D057DFCC1F1
-:1004B000DB57AD9979C254040CC2AA99E5C1DB5748
-:1004C000F352056AFB568E02941AC36AC862690227
-:1004D000267D941EC36ED36EC8624802C86A9426F5
-:1004E000981EC36EC8629826C36E60020E7D981E69
-:1004F000C36EC8626C02037D9826C36E3D9A4C029F
-:10050000D36EC86A9826C36E4D9AC8626E02017C8B
-:10051000479A096A1A7F0125004D217D0E9AE36EE4
-:100520008F00D805017D8D00C8626E02127D096AB8
-:100530000C7F01250120F87CDB57004D107D286AD7
-:10054000047F0000FCC1DB57009A070004620C6ABC
-:100550004A9A286AFA7F04627CC258045404FF084D
-:100560001100FF18BC00CD00017C3D9A286AED7F88
-:1005700004627CC20CC2FD99D9C1E5C1DB57F352BC
-:10058000056AFB568E02941A025269021D7D941E62
-:1005900006524802065A9426981E06524C02065AE3
-:1005A0009826981E065260020A7C982606526E0211
-:1005B000237D096A1D7F0125004D247D6F9A286ADD
-:1005C000177F04627CC2A09A8F00D805017D8D0040
-:1005D000A00506526E02107D096A0A7F0120F97C8F
-:1005E000286A067F0000004D0D7DFCC1DB575F9A35
-:1005F000070004620C6A9D9A286AFA7F04627CC232
-:1006000058045404286AF47F0CC25C9AE8DADB5779
-:10061000F352056AC7698F00D500017D8D00A005E2
-:100620007804037D79041C7D157C79041E7CEE56CC
-:10063000C862287E6006027D10021202096A227FCB
-:100640001202096A1F7F1202096A1C7F03200048F8
-:10065000EF7CD79A150015000278C8620B6AD69A0B
-:1006600015000278C8620A6AD69A0278C862096AD6
-:10067000097C286A077F004D077DD3520108025884
-:100680000004A89A79C25404D35201080258010404
-:1006900000DBA59A700B11031353DB5F0A07D358D5
-:1006A0008B00FB5E3CC1157DC05AC85B4EC1117CFE
-:1006B0008803F05DFF0D1105FF1DBC05004D047D95
-:1006C000FF9A0807DB5F0A078B00C352CB5359C15F
-:1006D000EC7D0002F09A016E0B612F7E0B622D7E85
-:1006E0000B632B7E0C0D1704170417049D04081DC3
-:1006F000CC05017C0C0DD16A000F4207C86FDD6F7D
-:100700001C7F8E009D0001680B67177ED56B040867
-:100710000278C86F1207117C0B670F7E04080278FD
-:10072000C86F12070A7CDD6F087FD169010FC86F9F
-:10073000DD6F037F01010004159B0700FF680C6853
-:100740000002159B85DBDB57F352056AFB52C76A33
-:100750008F00D500017D8D00A0057804037D79040C
-:100760001C7D157C79041E7CEE56C862287E6006CE
-:10077000027D10021202096A227F1202096A1F7F9B
-:100780001202096A1C7F03200048EF7C749B15004D
-:1007900015000278C8620B6A739B15000278C86264
-:1007A0000A6A739B0278C862096A097C286A077F13
-:1007B000004D077DD352010802580004449B79C2C2
-:1007C0005404D3520108025801049DDB419B700B75
-:1007D00011031353DB5F0A07D3588B00FB5E3CC148
-:1007E000157DC05AC85B4EC1117C8803F05DFF0DBA
-:1007F0001105FF1DBC05004D047D9C9B0807DB5FB8
-:100800000A078B00C352CB5359C1EC7D00028D9B6C
-:10081000F2DBDB57F352016AFB52D36A8F00D5003B
-:10082000017D8D00A005EB5D7804037D79041C7DBE
-:10083000207C7904157CEE566006057D09652A7ECC
-:100840000A62287EC59B0A62257E0965237E120501
-:100850001205AD02C86A1E7F03200048ED7CDD9BB7
-:1008600002780962C86ADC9B150002780A62C86ACD
-:10087000DC9B1500150002780B62C86A0B7CDF6DEB
-:10088000097F0000EB55004D097DD3520108025845
-:100890000004AC9B0700CC680C6815C2D352010859
-:1008A000025801040ADCA79B700B11031353DB5F92
-:1008B0000A07D3588B00FB5E3CC1157DC05AC85B4C
-:1008C0004EC1117C8803F05DFF0D1105FF1DBC05B5
-:1008D000004D047D099C0807DB5F0A078B00C352AB
-:0A08E000CB5359C1EC7D0002FA9BD6
+:10014000C86A127CDF6D107FFB52041AC36AD36A3F
+:10015000C862042AC86AFB52D36A0000EB55004DFE
+:10016000077DFCC1DB5706980700CC680C6815C2F8
+:100170000CC20198D9C1E5C1DB57E35FE357F352E5
+:10018000216A8F00D500017D8D00A005EB5DFB5637
+:100190007804037D79042A7D317C7904207C700BFE
+:1001A0001103EB53000F6003057D0965377E0A627A
+:1001B000357E7C980A62327E0965307E1205120512
+:1001C000AD026007027C065A8498265A277F011FD9
+:1001D00003200048E87C700B11031353A598150009
+:1001E00004780962065A0962265AA4981500150077
+:1001F00004780A62065A0A62265AA4981500150065
+:10020000150004780B62065A0B62265A077C000020
+:10021000EB55004D067DFCC1E3575F9807000C6865
+:1002200015C20CC25C98700B110313536C07017C50
+:10023000D9C1FB5E8A066B07017CD9C1F35EDB592D
+:10024000D3588F0110010F398B003CC12B7DC05A50
+:10025000C85B4EC1277C88038906E35CFF0D11054E
+:10026000FF1DBC053E07004D187D700811007E077C
+:10027000097D7D07027D2852DC98F852DB54BC02D0
+:10028000CC02097C7C07027D2852E598F852D354B1
+:10029000BC02CC02097D0004D3988B00C052C85325
+:1002A00059C1D67D0002C398FF08BF007F07157DA6
+:1002B0008804D500017D8D00A005EB5D8F02120240
+:1002C0001202FF3ADA05027C3E071099A402DD0211
+:1002D000027D3E0710995E071099EB55004D027C98
+:1002E0009805EB5DF352FB546A07267D6C07017D90
+:1002F0004D996B07577C6907047D6807027D010EE5
+:1003000027999358D600017D8E009355A005935DE3
+:10031000A00602780255045D1D7C004E087C69072A
+:10032000037D0255177E3499045D147F890693502E
+:100330000048017D1F999899150006780255045DC3
+:100340004F070255245D2F07017C989917006F070E
+:10035000017C012093559D0007009FD9EB98D36C39
+:100360006907047D6807027D010E5C999358D600E9
+:10037000017D8E009355A005935DA006027802557D
+:10038000C86D0F7C004E087C6907037D0255097E0D
+:100390006999C86D067F890693500048017D54997C
+:1003A00098999299C36A6907047D6807027D010ED6
+:1003B0007F999358D600017D8E009355A005935DDB
+:1003C000A0060278C865045D0F7C004E087C6907B2
+:1003D000037DC865097E8C99045D067F890693506C
+:1003E0000048017D7799989993559D000700FF6C0F
+:1003F0009FD9EB980000E354EB55004D017CEB983E
+:10040000D398E354EB55FF0A1102FF1A7F07027CD1
+:10041000A005AC999D008C05BA05A0051002BA0490
+:10042000AD0454040600D9C1E5C1DB57F352056A97
+:100430008F00D500017D8D00A005FB567804037D5B
+:100440007904297D1F7C79042E7CE35D700D1105F4
+:10045000ED55000F6007027D0652CD992652337E7E
+:100460006005027D10021202096A2D7F1202096ADC
+:100470002A7F1202096A277F011F03200048EA7CB5
+:10048000E355F899150015001500047806520B6A1B
+:1004900026520B6AF79915001500047806520A6A6D
+:1004A00026520A6AF799150004780652096A2652FC
+:1004B000096A097C286A077F0000DB57004D057D2B
+:1004C000FCC1DB57B69979C254040CC2B399E5C19B
+:1004D000DB57F352056AFB568E02941AC36AC86250
+:1004E0006902267D941EC36ED36EC8624802C86A34
+:1004F0009426981EC36EC8629826C36E60020E7D55
+:10050000981EC36EC8626C02037D9826C36E469A1D
+:100510004C02D36EC86A9826C36E569AC8626E02A1
+:10052000017C509A096A1A7F0125004D217D179A96
+:10053000E36E8F00D805017D8D00C8626E02127DCA
+:10054000096A0C7F01250120F87CDB57004D107DE6
+:10055000286A047F0000FCC1DB57099A0700046287
+:100560000C6A539A286AFA7F04627CC258045404C5
+:10057000FF081100FF18BC00CD00017C469A286AD4
+:10058000ED7F04627CC20CC2069AD9C1E5C1DB577B
+:10059000F352056AFB568E02941A025269021D7DBF
+:1005A000941E06524802065A9426981E06524C0281
+:1005B000065A9826981E065260020A7C9826065211
+:1005C0006E02237D096A1D7F0125004D247D789AE6
+:1005D000286A177F04627CC2A99A8F00D805017D22
+:1005E0008D00A00506526E02107D096A0A7F012067
+:1005F000F97C286A067F0000004D0D7DFCC1DB57A9
+:10060000689A070004620C6AA69A286AFA7F046254
+:100610007CC258045404286AF47F0CC2659AF1DA4B
+:10062000DB57F352056AC7698F00D500017D8D0045
+:10063000A0057804037D79041C7D157C79041E7C5B
+:10064000EE56C862287E6006027D10021202096A18
+:10065000227F1202096A1F7F1202096A1C7F03208F
+:100660000048EF7CE09A150015000278C8620B6A1A
+:10067000DF9A15000278C8620A6ADF9A0278C862B7
+:10068000096A097C286A077F004D077DD35201085B
+:1006900002580004B19A79C25404D3520108025896
+:1006A000010409DBAE9A700B11031353DB5F0A07D9
+:1006B000D3588B00FB5E3CC1157DC05AC85B4EC150
+:1006C000117C8803F05DFF0D1105FF1DBC05004D79
+:1006D000047D089B0807DB5F0A078B00C352CB53DE
+:1006E00059C1EC7D0002F99A016E0B612F7E0B62FD
+:1006F0002D7E0B632B7E0C0D1704170417049D042D
+:10070000081DCC05017C0C0DD16A000F4207C86F93
+:10071000DD6F1C7F8E009D0001680B67177ED56B17
+:1007200004080278C86F1207117C0B670F7E04085B
+:100730000278C86F12070A7CDD6F087FD169010F4C
+:10074000C86FDD6F037F010100041E9B0700FF6877
+:100750000C6800021E9B8EDBDB57F352056AFB52CE
+:10076000C76A8F00D500017D8D00A0057804037D48
+:1007700079041C7D157C79041E7CEE56C862287EA7
+:100780006006027D10021202096A227F1202096AC3
+:100790001F7F1202096A1C7F03200048EF7C7D9BAB
+:1007A000150015000278C8620B6A7C9B1500027860
+:1007B000C8620A6A7C9B0278C862096A097C286A56
+:1007C000077F004D077DD3520108025800044D9B5E
+:1007D00079C25404D352010802580104A6DB4A9B93
+:1007E000700B11031353DB5F0A07D3588B00FB5EBA
+:1007F0003CC1157DC05AC85B4EC1117C8803F05DB9
+:10080000FF0D1105FF1DBC05004D047DA59B0807CC
+:10081000DB5F0A078B00C352CB5359C1EC7D00024A
+:10082000969BFBDBDB57F352016AFB52D36A8F00C6
+:10083000D500017D8D00A005EB5D7804037D790472
+:100840001C7D207C7904157CEE566006057D0965CB
+:100850002A7E0A62287ECE9B0A62257E0965237E57
+:1008600012051205AD02C86A1E7F03200048ED7C08
+:10087000E69B02780962C86AE59B150002780A6265
+:10088000C86AE59B1500150002780B62C86A0B7CEC
+:10089000DF6D097F0000EB55004D097DD352010843
+:1008A00002580004B59B0700CC680C6815C2D352EF
+:1008B00001080258010413DCB09B700B11031353A1
+:1008C000DB5F0A07D3588B00FB5E3CC1157DC05A25
+:1008D000C85B4EC1117C8803F05DFF0D1105FF1D43
+:1008E000BC05004D047D129C0807DB5F0A078B00E6
+:0C08F000C352CB5359C1EC7D0002039CA5
 :00000001FF
-- 
1.7.5.4

