From fde04648a1016d019bc7c82c6afde1434bb59b1f Mon Sep 17 00:00:00 2001
From: Robin Gong <b38343@freescale.com>
Date: Wed, 15 Jul 2015 17:57:55 +0800
Subject: [PATCH 0930/1221] MLK-11257: add dual fifo for p_2_p script

SSI need support for dual fifo both dest and source, update sdma
script firstly.

Signed-off-by: Robin Gong <b38343@freescale.com>
[zou:Original patch taken from
git.freescale.com/imx/fsl-arm-yocto-bsp.git-b imx-3.14.52-1.1.0_ga]
Signed-off-by: zou cao <cao.zou@windriver.com>
---
 firmware/imx/sdma/sdma-imx6q.bin.ihex |  190 +++++++++++++++++----------------
 1 files changed, 99 insertions(+), 91 deletions(-)

diff --git a/firmware/imx/sdma/sdma-imx6q.bin.ihex b/firmware/imx/sdma/sdma-imx6q.bin.ihex
index 9ecea5f..5275e15 100755
--- a/firmware/imx/sdma/sdma-imx6q.bin.ihex
+++ b/firmware/imx/sdma/sdma-imx6q.bin.ihex
@@ -1,15 +1,15 @@
-:1000000053444D4103000000010000001C000000AB
-:1000100029000000C00000005E070000820200000E
-:10002000FFFFFFFF00000000FFFFFFFFFFFFFFFFDC
+:1000000053444D4103000000020000001C000000AA
+:1000100029000000C0000000D2070000820200009A
+:10002000FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFE0
 :10003000FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFD0
-:10004000FFFFFFFFFFFFFFFF051A00004B0400004A
-:10005000EB020000B1180000FFFFFFFF631A000071
+:10004000FFFFFFFFFFFFFFFF3F1A0000FFFFFFFF63
+:10005000EB020000B1180000FFFFFFFF9D1A000037
 :10006000FFFFFFFFC0030000FFFFFFFFFFFFFFFFD9
 :10007000FFFFFFFFAB020000FFFFFFFF7B0300005D
 :10008000FFFFFFFFFFFFFFFF4C0400006E040000B6
 :10009000FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF70
-:1000A000000000000018000058180000B1190000FE
-:1000B000121B0000AD1A0000491B000000180000D0
+:1000A000FFFFFFFF0018000058180000EB190000C8
+:1000B0004C1B0000E71A0000831B00000018000022
 :1000C000E3C1DB57F352016AFB52D36AFB521C1A9D
 :1000D000C36AE8621102FF3A3008D002337C8F0015
 :1000E000D500017D8D00A005EB5D7804037D7904CA
@@ -44,88 +44,96 @@
 :1002B000017D8D00A005EB5D8F0212021202FF3A54
 :1002C000DA05027C3E071099A402DD02027D3E079A
 :1002D00010995E071099EB55004D027C9805EB5D77
-:1002E000F352FB546A07267D6C07017D4D996B071D
-:1002F000577C6907047D6807027D010E2799935892
-:10030000D600017D8E009355A005935DA00602786E
-:100310000255045D1D7C004E087C6907037D025573
-:10032000177E3499045D147F890693500048017D3F
-:100330001F999899150006780255045D4F070255DC
-:10034000245D2F07017C989917006F07017C01201D
-:1003500093559D0007009FD9EB98D36C6907047DE6
-:100360006807027D010E5C999358D600017D8E00CE
-:100370009355A005935DA00602780255C86D0F7CC9
-:10038000004E087C6907037D0255097E6999C86D96
-:10039000067F890693500048017D54999899929957
-:1003A000C36A6907047D6807027D010E7F9993582F
-:1003B000D600017D8E009355A005935DA0060278BE
-:1003C000C865045D0F7C004E087C6907037DC86525
-:1003D000097E8C99045D067F890693500048017D53
-:1003E0007799989993559D000700FF6C9FD9EB98DA
-:1003F0000000E354EB55004D017CEB98D398E35497
-:10040000EB55FF0A1102FF1A7F07027CA005AC9989
-:100410009D008C05BA05A0051002BA04AD04540471
-:100420000600D9C1E3C1DB57F352056A8F00D5003E
-:10043000017D8D00A005FB567804037D7904297D9C
-:100440001F7C79042E7CE35D700D1105ED55000FC6
-:100450006007027D0652CD992652337E6005027DEB
-:1004600010021202096A2D7F1202096A2A7F120203
-:10047000096A277F011F03200048EA7CE355F899A9
-:10048000150015001500047806520B6A26520B6AF7
-:10049000F79915001500047806520A6A26520A6A6E
-:1004A000F799150004780652096A2652096A097CF0
-:1004B000286A077F0000DB57004D057DFAC1DB5736
-:1004C000B69977C254040AC2B399E3C1DB57F35219
-:1004D000056AFB568E02941AC36AC8626902267DB9
-:1004E000941EC36ED36EC8624802C86A9426981ED2
-:1004F000C36EC8629826C36E60020E7D981EC36EDE
-:10050000C8626C02037D9826C36E469A4C02D36E75
-:10051000C86A9826C36E569AC8626E02017C509AC9
-:10052000096A1A7F0125004D217D179AE36E8F001D
-:10053000D805017D8D00C8626E02127D096A0C7FAC
-:1005400001250120F87CDB57004D107D286A047FCF
-:100550000000FAC1DB57099A070004620C6A539A3B
-:10056000286AFA7F04627AC258045404FF08110012
-:10057000FF18BC00CD00017C469A286AED7F04621A
-:100580007AC20AC2069AD9C1E3C1DB57F352056A9F
-:10059000FB568E02941A025269021D7D941E065269
-:1005A0004802065A9426981E06524C02065A98266D
-:1005B000981E065260020A7C982606526E02237D1F
-:1005C000096A1D7F0125004D247D789A286A177FCE
-:1005D00004627AC2A99A8F00D805017D8D00A0051A
-:1005E00006526E02107D096A0A7F0120F97C286A92
-:1005F000067F0000004D0D7DFAC1DB57689A0700A9
-:1006000004620C6AA69A286AFA7F04627AC25804C5
-:100610005404286AF47F0AC2659AF1DADB57F35270
-:10062000056AC7698F00D500017D8D00A00578049B
-:10063000037D79041C7D157C79041E7CEE56C8620E
-:10064000287E6006027D10021202096A227F1202D1
-:10065000096A1F7F1202096A1C7F03200048EF7C91
-:10066000E09A150015000278C8620B6ADF9A15003F
-:100670000278C8620A6ADF9A0278C862096A097C4D
-:10068000286A077F004D077DD352010802580004F5
-:10069000B19A77C25404D35201080258010409DB0D
-:1006A000AE9A700B11031353DB5F0A07D3588B000C
-:1006B000FB5E3CC1157DC05AC85B4EC1117C8803EE
-:1006C000F05DFF0D1105FF1DBC05004D047D089B6D
-:1006D0000807DB5F0A078B00C352CB5359C1EC7D7F
-:1006E0000002F99A016E0B612F7E0B622D7E0B6367
-:1006F0002B7E0C0D1704170417049D04081DCC0550
-:10070000017C0C0DD16A000F4207C86FDD6F1C7FA2
-:100710008E009D0001680B67177ED56B0408027878
-:10072000C86F1207117C0B670F7E04080278C86F30
-:1007300012070A7CDD6F087FD169010FC86FDD6F7A
-:10074000037F010100041E9B0700FF680C68000284
-:100750001E9B8EDBDB57F352056AFB52C76A8F0084
-:10076000D500017D8D00A0057804037D79041C7DF2
-:10077000157C79041E7CEE56C862287E6006027DD8
-:1007800010021202096A227F1202096A1F7F1202F6
-:10079000096A1C7F03200048EF7C7D9B1500150033
-:1007A0000278C8620B6A7C9B15000278C8620A6AEC
-:1007B0007C9B0278C862096A097C286A077F004D21
-:1007C000077DD3520108025800044D9B77C25404A0
-:1007D000D352010802580104A6DB4A9B700B110397
-:1007E0001353DB5F0A07D3588B00FB5E3CC1157DBA
-:1007F000C05AC85B4EC1117C8803F05DFF0D110526
-:10080000FF1DBC05004D047DA59B0807DB5F0A07A3
-:0E0810008B00C352CB5359C1EC7D0002969B66
+:1002E000F352FB546A07387D6C07017D5F996B07F9
+:1002F0007D7C6907047D6807027D010E279993586C
+:10030000D600017D8E009355A005935DA0066D0774
+:10031000067D6E070B7D02780255045D3C99150041
+:1003200004780255045D2255045D3C99150004785B
+:100330000255045D0255245D1D7C004E087C690752
+:10034000037D0255177E4699045D147F89069350FC
+:100350000048017D1F99D299150006780255045D69
+:100360004F070255245D2F07017CD29917006F07B4
+:10037000017C012093559D000700D9D9EB986907AE
+:10038000047D6807027D010E6D999358D600017DAA
+:100390008E009355A005935DA0066D07077D6E073F
+:1003A0000D7DD36C02780255C86D85991500D36C0C
+:1003B00004780255C86D2255C86D859915000578D9
+:1003C000D76C0255C86D0255C86D0F7C004E087C75
+:1003D0006907037D0255097E8F99C86D067F8906DE
+:1003E00093500048017D6599D299CC996907047DA5
+:1003F0006807027D010EA4999358D600017D8E00F6
+:100400009355A005935DA0066D07077D6E070D7DD2
+:10041000C36A0278C865045DBC9915000578C76A8F
+:10042000C865045DC865045DBC991500C36A04789D
+:10043000C865045DC865245D0F7C004E087C6907B3
+:10044000037DC865097EC699045D067F89069350C1
+:100450000048017D9C99D29993559D000700FF6C3F
+:10046000D9D9EB980000E354EB55004D017CEB9893
+:10047000D398E354EB55FF0A1102FF1A7F07027C61
+:10048000A005E6999D008C05BA05A0051002BA04E6
+:10049000AD0454040600D9C1E3C1DB57F352056A29
+:1004A0008F00D500017D8D00A005FB567804037DEB
+:1004B0007904297D1F7C79042E7CE35D700D110584
+:1004C000ED55000F6007027D0652079A2652337ED3
+:1004D0006005027D10021202096A2D7F1202096A6C
+:1004E0002A7F1202096A277F011F03200048EA7C45
+:1004F000E355329A150015001500047806520B6A70
+:1005000026520B6A319A15001500047806520A6AC1
+:1005100026520A6A319A150004780652096A265250
+:10052000096A097C286A077F0000DB57004D057DBA
+:10053000FAC1DB57F09977C254040AC2ED99E3C1BE
+:10054000DB57F352056AFB568E02941AC36AC862DF
+:100550006902267D941EC36ED36EC8624802C86AC3
+:100560009426981EC36EC8629826C36E60020E7DE4
+:10057000981EC36EC8626C02037D9826C36E809A73
+:100580004C02D36EC86A9826C36E909AC8626E02F7
+:10059000017C8A9A096A1A7F0125004D217D519AB2
+:1005A000E36E8F00D805017D8D00C8626E02127D5A
+:1005B000096A0C7F01250120F87CDB57004D107D76
+:1005C000286A047F0000FAC1DB57439A07000462DF
+:1005D0000C6A8D9A286AFA7F04627AC2580454041D
+:1005E000FF081100FF18BC00CD00017C809A286A2A
+:1005F000ED7F04627AC20AC2409AD9C1E3C1DB57D7
+:10060000F352056AFB568E02941A025269021D7D4E
+:10061000941E06524802065A9426981E06524C0210
+:10062000065A9826981E065260020A7C98260652A0
+:100630006E02237D096A1D7F0125004D247DB29A3B
+:10064000286A177F04627AC2E39A8F00D805017D79
+:100650008D00A00506526E02107D096A0A7F0120F6
+:10066000F97C286A067F0000004D0D7DFAC1DB573A
+:10067000A29A070004620C6AE09A286AFA7F046270
+:100680007AC258045404286AF47F0AC29F9A2BDB6A
+:10069000DB57F352056AC7698F00D500017D8D00D5
+:1006A000A0057804037D79041C7D157C79041E7CEB
+:1006B000EE56C862287E6006027D10021202096AA8
+:1006C000227F1202096A1F7F1202096A1C7F03201F
+:1006D0000048EF7C1A9B150015000278C8620B6A6F
+:1006E000199B15000278C8620A6A199B0278C862D1
+:1006F000096A097C286A077F004D077DD3520108EB
+:1007000002580004EB9A77C25404D35201080258ED
+:10071000010443DBE89A700B11031353DB5F0A07F4
+:10072000D3588B00FB5E3CC1157DC05AC85B4EC1DF
+:10073000117C8803F05DFF0D1105FF1DBC05004D08
+:10074000047D429B0807DB5F0A078B00C352CB5333
+:1007500059C1EC7D0002339B016E0B612F7E0B6251
+:100760002D7E0B632B7E0C0D1704170417049D04BC
+:10077000081DCC05017C0C0DD16A000F4207C86F23
+:10078000DD6F1C7F8E009D0001680B67177ED56BA7
+:1007900004080278C86F1207117C0B670F7E0408EB
+:1007A0000278C86F12070A7CDD6F087FD169010FDC
+:1007B000C86FDD6F037F01010004589B0700FF68CD
+:1007C0000C680002589BC8DBDB57F352056AFB52EA
+:1007D000C76A8F00D500017D8D00A0057804037DD8
+:1007E00079041C7D157C79041E7CEE56C862287E37
+:1007F0006006027D10021202096A227F1202096A53
+:100800001F7F1202096A1C7F03200048EF7CB79B00
+:10081000150015000278C8620B6AB69B15000278B5
+:10082000C8620A6AB69B0278C862096A097C286AAB
+:10083000077F004D077DD352010802580004879BB3
+:1008400077C25404D352010802580104E0DB849BB0
+:10085000700B11031353DB5F0A07D3588B00FB5E49
+:100860003CC1157DC05AC85B4EC1117C8803F05D48
+:10087000FF0D1105FF1DBC05004D047DDF9B080722
+:10088000DB5F0A078B00C352CB5359C1EC7D0002DA
+:02089000D09BFB
 :00000001FF
-- 
1.7.5.4

