From 611e4312215b39dc37a696c3db51293416ef7363 Mon Sep 17 00:00:00 2001
From: Robin Gong <b38343@freescale.com>
Date: Wed, 7 Jan 2015 18:07:25 +0800
Subject: [PATCH 0043/1221] MLK-10244-1 firmware: imx: sdma: add new sdma
 firmware for i.mx7d

New i.mx7d sdma firmware added.

Signed-off-by: Robin Gong <b38343@freescale.com>
(cherry picked from commit 6ad95ba538d2c0986770acdabb76af5f7dc029ae)
[zou:Original patch taken from
git.freescale.com/imx/fsl-arm-yocto-bsp.git-b imx-3.14.52-1.1.0_ga]
Signed-off-by: zou cao <cao.zou@windriver.com>
---
 firmware/Makefile                     |    2 +-
 firmware/imx/sdma/sdma-imx7d.bin.ihex |  144 +++++++++++++++++++++++++++++++++
 2 files changed, 145 insertions(+), 1 deletions(-)
 create mode 100644 firmware/imx/sdma/sdma-imx7d.bin.ihex

diff --git a/firmware/Makefile b/firmware/Makefile
index fc975f9..6f741de 100644
--- a/firmware/Makefile
+++ b/firmware/Makefile
@@ -61,7 +61,7 @@ fw-shipped-$(CONFIG_DRM_RADEON) += radeon/R100_cp.bin radeon/R200_cp.bin \
 				   radeon/RV770_pfp.bin radeon/RV770_me.bin \
 				   radeon/RV730_pfp.bin radeon/RV730_me.bin \
 				   radeon/RV710_pfp.bin radeon/RV710_me.bin
-fw-shipped-$(CONFIG_IMX_SDMA) += imx/sdma/sdma-imx6q.bin
+fw-shipped-$(CONFIG_IMX_SDMA) += imx/sdma/sdma-imx6q.bin imx/sdma/sdma-imx7d.bin
 fw-shipped-$(CONFIG_DVB_AV7110) += av7110/bootcode.bin
 fw-shipped-$(CONFIG_DVB_TTUSB_BUDGET) += ttusb-budget/dspbootcode.bin
 fw-shipped-$(CONFIG_E100) += e100/d101m_ucode.bin e100/d101s_ucode.bin \
diff --git a/firmware/imx/sdma/sdma-imx7d.bin.ihex b/firmware/imx/sdma/sdma-imx7d.bin.ihex
new file mode 100644
index 0000000..7bb0c30
--- /dev/null
+++ b/firmware/imx/sdma/sdma-imx7d.bin.ihex
@@ -0,0 +1,144 @@
+:1000000053444D4104000000010000001C000000AA
+:100010002A000000C4000000220800008402000042
+:10002000FFFFFFFF00000000FFFFFFFFFFFFFFFFDC
+:10003000FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFD0
+:10004000FFFFFFFFFFFFFFFFFA190000FFFFFFFFA9
+:10005000ED020000A8180000FFFFFFFF581A000083
+:10006000FFFFFFFFC2030000FFFFFFFFFFFFFFFFD7
+:10007000FFFFFFFFAD020000FFFFFFFF7D03000059
+:10008000FFFFFFFFFFFFFFFF4E04000070040000B2
+:10009000FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF70
+:1000A00000000000001800004F180000A619000012
+:1000B000071B0000A21A00003E1B000000180000F1
+:1000C000A41B0000E5C1DB57F352016AFB52D36A5F
+:1000D000FB521C1AC36AE8621102FF3A3008D002D0
+:1000E000337C8F00D500017D8D00A005EB5D780489
+:1000F000037D79041C7D207C7904157CEE56600616
+:10010000057D0965287E0A62267E28980A62237E7C
+:100110000965217E12051205AD02C86A1C7F032005
+:100120000048ED7C409802780962C86A3F98150043
+:1001300002780A62C86A3F981500150002780B62BF
+:10014000C86A097CDF6D077F0000EB55004D077D15
+:10015000FCC1DB5706980700CC680C6815C20CC2BE
+:100160000198D9C1E5C1DB57E35FE357F352216A38
+:100170008F00D500017D8D00A005EB5DFB56780456
+:10018000037D79042A7D317C7904207C700B110376
+:10019000EB53000F6003057D0965377E0A62357EEB
+:1001A00073980A62327E0965307E12051205AD022F
+:1001B0006007027C065A7B98265A277F011F03207E
+:1001C0000048E87C700B110313539C9815000478C9
+:1001D0000962065A0962265A9B9815001500047890
+:1001E0000A62065A0A62265A9B98150015001500E5
+:1001F00004780B62065A0B62265A077C0000EB5506
+:10020000004D067DFCC1E357569807000C6815C2E7
+:100210000CC25398700B110313536C07017CD9C1A6
+:10022000FB5E8A066B07017CD9C1F35EDB59D358AC
+:100230008F0110010F398B003CC12B7DC05AC85B68
+:100240004EC1277C88038906E35CFF0D1105FF1D65
+:10025000BC053E07004D187D700811007E07097D22
+:100260007D07027D2852D398F852DB54BC02CC02A1
+:10027000097C7C07027D2852DC98F852D354BC02DA
+:10028000CC02097D0004CA988B00C052C85359C1E2
+:10029000D67D0002BA98FF08BF007F07157D88044D
+:1002A000D500017D8D00A005EB5D8F0212021202C8
+:1002B000FF3ADA05027C3E070599A402DD02027DC1
+:1002C0003E0705995E070599EB559805EB5DF352DE
+:1002D000FB546A07267D6C07017D42996B07577CAA
+:1002E0006907047D6807027D010E1C999358D600AA
+:1002F000017D8E009355A005935DA00602780255FE
+:10030000045D1D7C004E087C6907037D0255177E45
+:100310002999045D147F890693500048017D149942
+:100320008D99150006780255045D4F070255245D2E
+:100330002F07017C8D9917006F07017C01209355D1
+:100340009D00070094D9E298D36C6907047D680783
+:10035000027D010E51999358D600017D8E00935570
+:10036000A005935DA00602780255C86D0F7C004E73
+:10037000087C6907037D0255097E5E99C86D067F7A
+:10038000890693500048017D49998D998799C36AE0
+:100390006907047D6807027D010E74999358D600A1
+:1003A000017D8E009355A005935DA0060278C86577
+:1003B000045D0F7C004E087C6907037DC865097EDB
+:1003C0008199045D067F890693500048017D6C99F0
+:1003D0008D9993559D000700FF6C94D9E298000019
+:1003E000E354EB55004D017CE298CA98E354EB5579
+:1003F000FF0A1102FF1A7F07027CA005A1999D0048
+:100400008C05BA05A0051002BA04AD045404060018
+:10041000D9C1E5C1DB57F352056A8F00D500017DD4
+:100420008D00A005FB567804037D7904297D1F7C8F
+:1004300079042E7CE35D700D1105ED55000F60070A
+:10044000027D0652C2992652337E6005027D10025B
+:100450001202096A2D7F1202096A2A7F1202096AB2
+:10046000277F011F03200048EA7CE355ED99150022
+:1004700015001500047806520B6A26520B6AEC9997
+:1004800015001500047806520A6A26520A6AEC9989
+:10049000150004780652096A2652096A097C286AFE
+:1004A000077F0000DB57004D057DFCC1DB57AB9992
+:1004B00079C254040CC2A899E5C1DB57F352056A0E
+:1004C000FB568E02941AC36AC8626902267D941E86
+:1004D000C36ED36EC8624802C86A9426981EC36E63
+:1004E000C8629826C36E60020E7D981EC36EC862F5
+:1004F0006C02037D9826C36E3B9A4C02D36EC86A89
+:100500009826C36E4B9AC8626E02017C459A096AAE
+:100510001A7F0125004D217D0C9AE36E8F00D805CE
+:10052000017D8D00C8626E02127D096A0C7F012573
+:100530000120F87CDB57004D107D286A047F000005
+:10054000FCC1DB57FE99070004620C6A489A286ACE
+:10055000FA7F04627CC258045404FF081100FF189B
+:10056000BC00CD00017C3B9A286AED7F04627CC20E
+:100570000CC2FB99D9C1E5C1DB57F352056AFB56A2
+:100580008E02941A025269021D7D941E0652480280
+:10059000065A9426981E06524C02065A9826981E11
+:1005A000065260020A7C982606526E02237D096A72
+:1005B0001D7F0125004D247D6D9A286A177F0462F6
+:1005C0007CC29E9A8F00D805017D8D00A005065241
+:1005D0006E02107D096A0A7F0120F97C286A067F75
+:1005E0000000004D0D7DFCC1DB575D9A07000462E1
+:1005F0000C6A9B9A286AFA7F04627CC258045404ED
+:10060000286AF47F0CC25A9AE6DADB57F352056A7D
+:10061000C7698F00D500017D8D00A0057804037D9A
+:1006200079041C7D157C79041E7CEE56C862287EF8
+:100630006006027D10021202096A227F1202096A14
+:100640001F7F1202096A1C7F03200048EF7CD59AA5
+:10065000150015000278C8620B6AD49A150002785A
+:10066000C8620A6AD49A0278C862096A097C286A50
+:10067000077F004D077DD352010802580004A69A57
+:1006800079C25404D352010802580104FEDAA39A35
+:10069000700B11031353DB5F0A07D3588B00FB5E0B
+:1006A0003CC1157DC05AC85B4EC1117C8803F05D0A
+:1006B000FF0D1105FF1DBC05004D047DFD9A0807C7
+:1006C000DB5F0A078B00C352CB5359C1EC7D00029C
+:1006D000EE9A016E0B612F7E0B622D7E0B632B7EDB
+:1006E0000C0D1704170417049D04081DCC05017C8C
+:1006F0000C0DD16A000F4207C86FDD6F1C7F8E00A2
+:100700009D0001680B67177ED56B04080278C86FDF
+:100710001207117C0B670F7E04080278C86F12075E
+:100720000A7CDD6F087FD169010FC86FDD6F037F21
+:1007300001010004139B0700FF680C680002139B73
+:1007400083DBDB57F352056AFB52C76A8F00D50083
+:10075000017D8D00A0057804037D79041C7D157C46
+:1007600079041E7CEE56C862287E6006027D100267
+:100770001202096A227F1202096A1F7F1202096AA5
+:100780001C7F03200048EF7C729B15001500027847
+:10079000C8620B6A719B15000278C8620A6A719B75
+:1007A0000278C862096A097C286A077F004D077DC4
+:1007B000D352010802580004429B79C25404D35218
+:1007C0000108025801049BDB3F9B700B110313537C
+:1007D000DB5F0A07D3588B00FB5E3CC1157DC05A16
+:1007E000C85B4EC1117C8803F05DFF0D1105FF1D34
+:1007F000BC05004D047D9A9B0807DB5F0A078B0050
+:10080000C352CB5359C1EC7D00028B9BF0DBDB570D
+:10081000F352016AFB52D36A8F00D500017D8D002F
+:10082000A005EB5D7804037D79041C7D207C7904B0
+:10083000157CEE566006057D09652A7E0A62287ED3
+:10084000C39B0A62257E0965237E12051205AD024F
+:10085000C86A1E7F03200048ED7CDB9B027809629A
+:10086000C86ADA9B150002780A62C86ADA9B15002A
+:10087000150002780B62C86A0B7CDF6D097F0000EF
+:10088000EB55004D097DD352010802580004AA9B84
+:100890000700CC680C6815C2D35201080258010445
+:1008A00008DCA59B700B11031353DB5F0A07D358B9
+:1008B0008B00FB5E3CC1157DC05AC85B4EC1117CEC
+:1008C0008803F05DFF0D1105FF1DBC05004D047D83
+:1008D000079C0807DB5F0A078B00C352CB5359C143
+:0608E000EC7D0002F89B14
+:00000001FF
-- 
1.7.5.4

