From 0f307762956ee427e6da4be5d169dcea2cb5c23d Mon Sep 17 00:00:00 2001
From: Li Jun <jun.li@freescale.com>
Date: Mon, 1 Jun 2015 10:34:42 +0800
Subject: [PATCH 0739/1221] MLK-11005-6 Documentation: usb: chipidea: add usb
 OTG ADP test

Add how to enable USB OTG ADP feature and test sequence.

Signed-off-by: Li Jun <jun.li@freescale.com>
[zou:Original patch taken from
git.freescale.com/imx/fsl-arm-yocto-bsp.git-b imx-3.14.52-1.1.0_ga]
Signed-off-by: zou cao <cao.zou@windriver.com>
---
 Documentation/usb/chipidea.txt |   13 +++++++++++++
 1 files changed, 13 insertions(+), 0 deletions(-)

diff --git a/Documentation/usb/chipidea.txt b/Documentation/usb/chipidea.txt
index 44ddaeb..faf16c5 100644
--- a/Documentation/usb/chipidea.txt
+++ b/Documentation/usb/chipidea.txt
@@ -88,6 +88,19 @@ end device by:
 	echo 1 or 0 > /sys/bus/platform/devices/ci_hdrc.0/inputs/hnp_enable
 If this input is 0, only OTG role switch by ID pin detection can be supported.
 
+1.6 Enable and test USB OTG ADP
+----------------------
+Enable:
+1) Enable OTG FSM as 1.1
+2) Add "adp-support" properity in dts for your hardware
+
+Test sequence:
+1) Connect 2 OTG devices, enable HNP by enable hnp_enable at 2 boards
+	echo 1 > /sys/bus/platform/devices/ci_hdrc.0/inputs/hnp_enable
+2) After A-device ends session(remove B-device and wait 10s, or a_bus_drop)
+3) Remove B-device, B-device will show "OTG device detached".
+4) Insert B-device, A-device will enumerate B-device.
+
 2. How to enable USB as system wakeup source
 -----------------------------------
 Below is the example for how to enable USB as system wakeup source
-- 
1.7.5.4

