From a494217ef5007b4206f5b52c857f1b88115d7603 Mon Sep 17 00:00:00 2001
From: Li Jun <jun.li@freescale.com>
Date: Fri, 3 Jul 2015 15:31:32 +0800
Subject: [PATCH 0892/1221] MLK-11123-32 Documentation: usb: chipidea: update
 for otg tests

After introduce otg features in dts, and remove hnp_enable, we should update
otg test document for how to enable HNP, SRP and ADP.

Signed-off-by: Li Jun <jun.li@freescale.com>
[zou:Original patch taken from
git.freescale.com/imx/fsl-arm-yocto-bsp.git-b imx-3.14.52-1.1.0_ga]
Signed-off-by: zou cao <cao.zou@windriver.com>
---
 Documentation/usb/chipidea.txt |   57 +++++++++++++++++-----------------------
 1 files changed, 24 insertions(+), 33 deletions(-)

diff --git a/Documentation/usb/chipidea.txt b/Documentation/usb/chipidea.txt
index faf16c5..c7c16e7 100644
--- a/Documentation/usb/chipidea.txt
+++ b/Documentation/usb/chipidea.txt
@@ -3,16 +3,14 @@
 To show how to demo OTG HNP and SRP functions via sys input files
 with 2 Freescale i.MX6Q sabre SD boards.
 
-1.1 How to enable OTG FSM in menuconfig
+1.1 How to enable OTG HNP and SRP
 ---------------------------------------
-Select CONFIG_USB_OTG_FSM, rebuild kernel Image and modules.
-If you want to check some internal variables for otg fsm,
-select CONFIG_USB_CHIPIDEA_DEBUG, there are 2 files which
-can show otg fsm variables and some controller registers value:
-cat /sys/kernel/debug/ci_hdrc.0/otg
-cat /sys/kernel/debug/ci_hdrc.0/registers
+Pass otg-rev to be 0x0200, remove srp-disable and hnp-disable in dts like below:
+dr_mode = "otg";
+otg-rev = <0x0200>;
+adp-disable;
 
-1.2 Test operations
+1.2 HNP and SRP test operations
 -------------------
 1) Power up 2 Freescale i.MX6Q sabre SD boards with gadget class driver loaded
    (e.g. g_mass_storage).
@@ -64,12 +62,22 @@ cat /sys/kernel/debug/ci_hdrc.0/registers
 
    A-device should resume usb bus and enumrate B-device.
 
-1.3 Reference document
+1.3 Enable and test USB OTG ADP
+----------------------
+Enable ADP in dts:
+1) Based on 1.2, remove adp-disable in dts.
+
+Test sequence:
+1) After A-device ends session(remove B-device and wait 10s, or a_bus_drop)
+2) Remove B-device, B-device will show "OTG device detached".
+3) Insert B-device, A-device will enumerate B-device.
+
+1.4 Reference document
 ----------------------
 "On-The-Go and Embedded Host Supplement to the USB Revision 2.0 Specification
 July 27, 2012 Revision 2.0 version 1.1a"
 
-1.4 OTG compliance test
+1.5 OTG compliance test
 ----------------------
 Only below 3 popular gadget drivers are declared to be USB OTG and EH 2.0
 compliant(with otg descriptor comply with USB OTG and EH 2.0 as a peripheral):
@@ -77,29 +85,12 @@ compliant(with otg descriptor comply with USB OTG and EH 2.0 as a peripheral):
 - ether
 - serial
 
-1.5 Runtime enable and disable full HNP
----------------------------------------
-One more input file is created:
-/sys/bus/platform/devices/ci_hdrc.0/inputs/hnp_enable
-
-Defaultly USB OTG full HNP(do role switch by SW protocol, no need exchange OTG
-plug) is not enabled, user can enable or disable it after disconnect from other
-end device by:
-	echo 1 or 0 > /sys/bus/platform/devices/ci_hdrc.0/inputs/hnp_enable
-If this input is 0, only OTG role switch by ID pin detection can be supported.
-
-1.6 Enable and test USB OTG ADP
-----------------------
-Enable:
-1) Enable OTG FSM as 1.1
-2) Add "adp-support" properity in dts for your hardware
-
-Test sequence:
-1) Connect 2 OTG devices, enable HNP by enable hnp_enable at 2 boards
-	echo 1 > /sys/bus/platform/devices/ci_hdrc.0/inputs/hnp_enable
-2) After A-device ends session(remove B-device and wait 10s, or a_bus_drop)
-3) Remove B-device, B-device will show "OTG device detached".
-4) Insert B-device, A-device will enumerate B-device.
+1.6 Debug
+If you want to check some internal variables for otg fsm,
+Select CONFIG_USB_CHIPIDEA_DEBUG, there are 2 files which
+can show otg fsm variables and some controller registers value:
+cat /sys/kernel/debug/ci_hdrc.0/otg
+cat /sys/kernel/debug/ci_hdrc.0/registers
 
 2. How to enable USB as system wakeup source
 -----------------------------------
-- 
1.7.5.4

