From 9073f53f50fba7db15cc6975908aa82a0a7c0445 Mon Sep 17 00:00:00 2001
From: Kosta Zertsekel <konszert@marvell.com>
Date: Mon, 29 Apr 2013 11:06:04 +0300
Subject: [PATCH 0606/1825] ALP: Remove release notes file

https://github.com/MISL-EBU-System-SW/misl-windriver.git linux-3.4.69-14t2-read
commit 7e7de7836c9364af426eb6d4a917fd3a37de41b6

Release notes file is managed outside the source code.

Signed-off-by: Kosta Zertsekel <konszert@marvell.com>
Change-Id: I57d3437ac6205d146debe7a31992efd72bff18a2
Reviewed-on: http://vgitil04.il.marvell.com:8080/1675
Reviewed-by: Star_Automation <star@marvell.com>
Reviewed-by: Eran Ben-Avi <benavi@marvell.com>
Tested-by: Eran Ben-Avi <benavi@marvell.com>
Signed-off-by: Zhong Hongbo <hongbo.zhong@windriver.com>
---
 ReleaseNotes |  264 ----------------------------------------------------------
 1 files changed, 0 insertions(+), 264 deletions(-)
 delete mode 100755 ReleaseNotes

diff --git a/ReleaseNotes b/ReleaseNotes
deleted file mode 100755
index 7a2f5c0..0000000
--- a/ReleaseNotes
+++ /dev/null
@@ -1,264 +0,0 @@
-
-Release notes for Armada-XP and Armada 370 LSP v3.2.y release 2012_Q4.3
-========================================================================
-Table of Contents
------------------
-
-1.	Requirements
-2.	Contents of this Release
-3.	Supported Features and Device Interfaces
-4.	Unsupported Interfaces
-5.	Changes List
-6.	How to Build
-7.	Specific UBoot configuration, Kernel command line params and tools
-8.	Known Issues
-
-1. Requirements 
-===============
-
-	Stable vanilla kernel base: v3.2.27 (referred to as kernel)
-	Supported SoCs: ArmadaXP A0/B0 and Armada370 A0/A1
-	Required SDK: Marvell SDK 6.0 (release compiled with SDK 6.0; arm-marvell-linux-gnueabi-gcc-4.6.2)
-		SDK 5.1.2 is also supported because it has the same toolchain as SDK 6.0
-	Tested with U-Boot:
-	    ArmadaXP: 2012-Q4.4 or higher (based on DENX vanilla u-boot 2011.12).
-	    Armada370: 1.1.3 or higher.
-	
-	
-2. Contents of This Release
-===========================
-
-	src/linux-<kernel>-<release>.zip   ==> Files modified or added on top of the linux vanilla kernel.
-
-	binaries.zip ==> compressed file that contain the following precompiled images:
-	
-		ArmadaXP SoC:
-			xp_v7smp:		==> SMP V7 mode with hardware cache coherency.
-							Optimized for routing (no Network filter).
-			xp_v7smp_nat:		==> SMP V7 mode with hardware cache coherency (with Network filter).
-			xp_v7smp_nas:       	==> SMP V7 mode with hardware cache coherency.
-							Optimized configuration for NAS platforms.
-			xp_v7smp_lpae:		==> SMP V7 with LPAE 40bit support (B0 only).
-			xp_v7smp_stat:		==> SMP V7 mode with hardware cache coherency and network statistics.
-			xp_v7smp_be8:		==> SMP V7 BE8 mode with hardware cache coherency.
-			xp_amp_g0/g1:       	==> AMP V7 mode, images for Group 0 & Group 1 (see README for further info).
-			
-		Armada370 SoC:
-			370_v7up:		==> UP V7 mode with hardware cache coherency
-							Optimized for routing (no Network filter).
-			370_v7up_nat:		==> UP V7 mode with hardware cache coherency (with Network filter).
-			370_v7up_nas:		==> UP V7 mode with hardware cache coherency.
-                                        		Optimized configuration for NAS platforms.
-			370_v7up_stat:		==> UP V7 mode with hardware cache coherency and network statistics.
-
-
-			
-Supported boards:
-	ArmadaXP:
-		1.	DB-78460-HE-BP Development Board for Armada-XP A0/B0 silicon
-		2.	RD-78460-SERVER-REV2 Board for Armada-XP A0/B0 silicon
-		3.	DB-784MP-GP General Purpose small form factor board for Armada-XP B0 silicon
-	Armada370:
-		1.	DB-88F6710-BP-DDR3 development board for Armada370 A0/A1 silicon
-		2.	DB-88F6710-PCAC PC adaptor card
-		3.	RD-A370 reference design board
-
-    
-3. Supported Features and Device Interfaces:
-==========================================
-- CPU Features: 
-    - V7 SMP (UP for A370)
-    - V7 architectural L2 cache
-    - Floating Point Unit (VFPv3)
-    - Thumb2 and ThumbEE instruction sets
-- Giga ports using the HAL based driver
-- SATA using the mainline LIBATA kernel driver
-- UART interfaces
-- SDIO
-- PCIe interfaces
-- I/O cache coherency
-- RTC
-- I2C interface
-- SPI flash
-- NAND flash
-- XOR DMA engines
-- BE support
-- Power Management (WFI, Idle, and Deep-Idle, CPU DFS)
-- Support for Marvell Vanir SAS/SATA3 controllers (provided separately on demand)
-AXP only:
-- 2 groups AMP support
-- LCD
-A370 only:
-- CESA
-- TDM
-- AUDIO
-
-4. Unsupported / UNTESTED Features:
-===================================
-- IDMA engines
-AXP only:
-- Cryptographic (CESA) engine.
-- TDM unit.
-
-5. Changes List
-================
-
-Changes from LSP 2012_Q4.1
---------------------------
-- PM: Suspend to RAM: 
-	- Added L2 flush in power-down routine
-	- Enabled suspend to RAM with indirect interrupt mode
-	- Added routines to save and restore the content of the DDR3 training memory space
-	- Added suspend to RAM support for GP board
-- PM: Added/Fixed suspend/resume support for: SDIO, sflash, PCIe, GPIO, XOR, USB, SMP
-- PM: Fixed bug in LCD initialization routine Enabled LCD suspend to RAM
-- PM: PCIe: Fix PM support in PCI Express.
-- PM: A370: Resolved compilation errors
-- NETA: Fix mvNetaPowerUp call during resume operation
-- NETA: Fix Switch and Gateway mode support
-- NETA: fix NAPI groups configuration in Kconfig
-- QD-DSDT: update switch driver version to 3.0D
-- AMP: Fixed bug on IPC network driver
-- KERNEL: mutex: Place lock in contended state after fastpath_lock failure
-- LPAE: LPAE Fixes:
-	- set CP15 registers to the same value as non-LPAE
-	- cpu_pj4bv7_set_pte_ext did not flush pte if not using 64KB page
-	- march=armv6 is for AFLAGS_proc-sheeva_pj4bv6.o
-	- add AFLAGS_proc-sheeva_pj4bv7lpae.o
-- MISC: Fixed hwmon Kconfig to support A370 and AXP
-- MISC: Updated defconfigs for A370 and AXP, added the XOR support
-- MISC: Fixed compilation errors related to MV_INCLUDE_GIG_ETH and MV_INCLUDE_SPI
-- MISC: A370: Fixed SPI, NAND and CESA functionalities
-- NAS: A370: updating NAS defconfig 
- 
-Changes from LSP 2012_Q4.0
---------------------------
-- Updated main defconfig file - removed XOR support and Vanir driver
-
-Changes from LSP 5.1.1
-----------------------
-- Removed Vanir driver from LSP
-- NAS: updating nas_init to version 3.4
-- NAS: updating NAS defconfig (2g split, removing netdma for splice)
-- NETA: Fix bug in tx_done timer invocation in SMP environment
-- SDK 5.1.2 is also supported as it has the same toolchain as SDK 6.0
-
-Changes from LSP 5.1.0
-----------------------
-- Added Armada370 support
-
-Changes from LSP 5.0.2
-----------------------
-- PM: Suspend to RAM – Initial support – Disabled by default
-- Kernel memcpy performance - reverts the memcpy acceleration code that was added for AXP Z1
-- LPAE: Fix IO window config when using more than 4GB DRAM
-- LPAE: ARM: 7488/1: mm: use 5 bits for swapfile type encoding
-- LPAE: ARM: 7487/1: mm: avoid setting nG bit for user mappings that aren't present
-- cpufreq support for ArmadaXP SoC – Disabled by default
-- AMP: Modified the global register used for AMP synchronization
-- NETA: Fix skb_recycle for kernel 3.2.27
-- NETA: avoid build warning by adding include <linux/module.h>
-- NETA: pppoe fixes for NFPv2
-- NETA: Check CPU validity on TRC_OUTPUT function
-- NETA: Fix sysfs command to read/write from PHY registers. All arguments for all commands in hex format
-- NFP: add ppp_sync support for NFPv2
-
-Changes from LSP 5.0.1
-----------------------
-- Fix bug in L2 replacement algorithm decoding
-- PM: Deep idle - Fix and update for kernel 3.2.x
-- NETA: Return arguments "port_mask" and "cpu_mask" to mvSysNetaInit() prototype
-
-Changes from LSP 5.0.0
-----------------------
-- Updated defconfig files - added XOR support and Vanir driver
-- Add CONFIG_IKCONFIG_PROC to defconfigs - allow reading the .config file from /proc/config.gz at runtime
-- Fix DRAM windows for mainline drivers
-- Disable window overlap testing
-- Add some statistics for the mm bounce subsystem
-- GP Board - Fix sata power GPIO
-- LPAE: Fix mvOs APIs to use dma_addr_t instead of U32 variables for phys-addr
-- LPAE: Fix address decoding init for mainline drivers
-
-Changes from LSP 3.2.0
-----------------------
-- Added Armada-XP revision B0 support
-- Added support for the RD-78460-SERVER-REV2 and DB-784MP-GP boards
-- Added LPAE Initial support
-- Updated hwmon threshold interrupt handling
-- Updated the error handler
-- AMP: Changed Global timer of second AMP image from 1 to co-exist with vxWorks
-- NAS: Updated nas_init to v3.3
-- SATA: Setting PHY speed according to SControl speed
-- NETA: Fix settings RX filtering mode and Multicast addresses for kernel 3.2.22
-- NETA: Fix power management support: 
-	- Add support for WoL
-	- Fix power management for down interfaces
-- NETA: Add platform device as parent to our network device
-- NETA: Fix possible bug in locking mechanism used for shared TXQs
-- NETA: Fix bugs in TXQs usage in multi core environment
-	- Fix TOS to TXQ mapping
-	- Fix TX_DONE processing for TXQs shared between different CPUs
-	
-Changes from LSP 3.0.6
-----------------------
-- Align the 3.2.x kernel feature list to the 2.1.1 release (based on 3.0.x kernel).
-- Enabling AMP.
-- Enabling NAND flash.
-- Enabling Perf.
-
-6. How To build
-===============
-
-    - Download the vanilla <kernel> sources from kernel.org or using the GIT. Assume file is linux-<kernel>.tar.bz2.
-	- Extract the vanilla kernel tarball (for example "tar -xjf linux-3.2.27.tar.bz2")
-	- Rename the linux-<kernel> folder to linux-<kernel>-<rel> (for example linux-3.2.27-axp-2012_Q4.0)
-	- extract the linux-<kernel>-<rel>.zip (for example "unzip linux-3.2.27-axp-2.1.1.zip"). This will 
-	  overruide part of the kernel files and will add the extra added files.
-Note: at this point you will be asked if you want to replace a certain file, choose A for all files that have such a conflict. 	
-    - cd to the kernel directory.
-	- Under the shell, issue the commands to set the environment:
-    'export ARCH=arm'
-    'export CROSS_COMPILE=./your/path/to/the/toolchain/on/your/disk/arm-mv7-linux-gnueabi-'
-	- Cleanup and old residuals
-    'make mrproper'
-	- Configure the kernel
-    config the kernel (example):
-        'make armada_xp_defconfig' ==> SMP ARMv7 mode
-	- Build the kernel
-		'make uImage'	
-    once the compilation is completed, The kernel uImage should be located under arch/arm/boot/ directory.
-
-7. Armada-XP specific Uboot configuration and Kernel command line parameters:
-=============================================================================
-7.1 Uboot environment variables:
---------------------------------
-The following U-Boot env. variables have direct impact on the Linux kernel,
-those variables can be changed using the U-Boot "setenv" command.
-
-usb0Mode:
-	Type: string, default "host"
-	Desc: usb 0 mode ("host" or "device");
-
-usb1Mode:
-	Type: string, default "host"
-	Desc: usb 1 mode ("host" or "device");
-
-l2rep:
-        Type: string, default "WayRR"
-        Desc: L2 replacement algorithm.
-        options: "WayRR" ==> Way round robin
-                 "LFSR" ==> Pseudo Random using LFSR
-                 "pLRU" ==> semi PLRU
-
-8. Known Issues
-===============
- - MSI/ MSI-X is not working when enabling the indirect (atomic) IRQ mode. It can also fail to init in SMP mode.
-  - MTU change under traffic can cause instability. Jumbo packets may be dropped.
- - SPI flash partitioning is not supported. The SPI flash is nOt recognized on the DB-784MP-GP board.
- - AMP with dual Linux may fail to boot on DB-MV78460-GP board.
- - Some features were not verified:
-   - LCD Controller
-   - AMP support
-   - 64K page table support
-   - LPAE mode
-- 
1.7.5.4

