From 98b320740c194385bf87e581e66e0d6053b75204 Mon Sep 17 00:00:00 2001
From: Yoni Farhadia <yonif@marvell.com>
Date: Thu, 7 Nov 2013 13:35:31 +0200
Subject: [PATCH 1081/1825] checkpatch.pl: ignore overline error in sysfs
 files

https://github.com/MISL-EBU-System-SW/misl-windriver.git linux-3.4.69-14t2-read
commit fd2817b2c05bfcda12e4708acfa752985d376795

	- If file name contains substring 'sysfs', then ignore overline errors
	- This is because sysfs help messages are ussually longer than 120 characters

Change-Id: I2d187297916ebfaec5239ee7c1cac80c9fa3f674
Signed-off-by: Yoni Farhadia <yonif@marvell.com>
Reviewed-on: http://vgitil04.il.marvell.com:8080/4326
Tested-by: Star_Automation <star@marvell.com>
Reviewed-by: Nadav Haklai <nadavh@marvell.com>
Signed-off-by: Zhong Hongbo <hongbo.zhong@windriver.com>
---
 scripts/checkpatch.pl |    5 +++--
 1 files changed, 3 insertions(+), 2 deletions(-)

diff --git a/scripts/checkpatch.pl b/scripts/checkpatch.pl
index 3ae7bc8..4061fb8 100755
--- a/scripts/checkpatch.pl
+++ b/scripts/checkpatch.pl
@@ -1768,10 +1768,11 @@ sub process {
 		    $rawline !~ /^.\s*\*\s*\@$Ident\s/ &&
 		    !($line =~ /^\+\s*$logFunctions\s*\(\s*(?:(KERN_\S+\s*|[^"]*))?"[X\t]*"\s*(?:|,|\)\s*;)\s*$/ ||
 		    $line =~ /^\+\s*"[^"]*"\s*(?:\s*|,|\)\s*;)\s*$/) &&
-		    $length > $LINELENGTH)
+		    $length > $LINELENGTH &&
+		    !($realfile =~ /sysfs/))
 		{
 			WARN("LONG_LINE",
-			     "line over LINELENGTH characters\n" . $herecurr);
+			     "line over " . $LINELENGTH . " characters\n" . $herecurr);
 		}
 
 # Check for user-visible strings broken across lines, which breaks the ability
-- 
1.7.5.4

