From c049eedf4e42a5752df45ba7446b4896efefc4a0 Mon Sep 17 00:00:00 2001
From: Victor Gu <xigu@marvell.com>
Date: Thu, 19 Jun 2014 14:19:15 +0800
Subject: [PATCH 1716/1825] fix: tools: voice build script issue for silicon
 labs

https://github.com/MISL-EBU-System-SW/misl-windriver.git linux-3.4.69-14t2-read
commit 7cfbce716a08dfd5851928ba450be1980fa4b095

    need to return 0 if case no release related parameters are inputted
Signed-off-by: Victor Gu <xigu@marvell.com>

Change-Id: I08fc910e71114f023300cc1ae7e2bbe9c77ff1b0
Reviewed-on: http://vgitil04.il.marvell.com:8080/8600
Reviewed-by: Yehuda Yitschak <yehuday@marvell.com>
Tested-by: Yehuda Yitschak <yehuday@marvell.com>
Signed-off-by: Zhong Hongbo <hongbo.zhong@windriver.com>
---
 tools/voice/build-silabs.sh |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/tools/voice/build-silabs.sh b/tools/voice/build-silabs.sh
index fd8b269..fb08801 100755
--- a/tools/voice/build-silabs.sh
+++ b/tools/voice/build-silabs.sh
@@ -54,7 +54,7 @@ cp mv_voice_tool-* $bin_path/.
 cd -
 
 if [ ! -n "$4" ]; then
-	leave 1
+	leave 0
 fi
 
 if [ ! -n "$5" ]; then
-- 
1.7.5.4

