From ae6ca660eed37e662ca5ba33411a0a15afa6de84 Mon Sep 17 00:00:00 2001
From: Victor Gu <xigu@marvell.com>
Date: Mon, 16 Jun 2014 15:48:53 +0800
Subject: [PATCH 1713/1825] fix: alp: tools: support make release for silicon
 labs SLIC build script

https://github.com/MISL-EBU-System-SW/misl-windriver.git linux-3.4.69-14t2-read
commit 52db3a01e8aae946b1ac7877177ebdb6ca4f2bd9

    There are five input parameters, the later two define the path for
SLIC release and version, in case they are not set, the script will ignore
making release.
Signed-off-by: Victor Gu <xigu@marvell.com>

Change-Id: I16a9db17e21001a1b799af46d77e5f0c09b1455f
Reviewed-on: http://vgitil04.il.marvell.com:8080/8540
Tested-by: Star_Automation <star@marvell.com>
Reviewed-by: Yehuda Yitschak <yehuday@marvell.com>
Signed-off-by: Zhong Hongbo <hongbo.zhong@windriver.com>
---
 tools/voice/build-silabs.sh |   47 +++++++++++++++++++++++++++++++++++++++++-
 1 files changed, 45 insertions(+), 2 deletions(-)

diff --git a/tools/voice/build-silabs.sh b/tools/voice/build-silabs.sh
index 3fb9b50..fd8b269 100755
--- a/tools/voice/build-silabs.sh
+++ b/tools/voice/build-silabs.sh
@@ -9,6 +9,7 @@ mod_path="${1}/kernel/slic"
 kernel_path=$2
 org_path=`pwd`
 bin_path=$3
+release_path="${4}/slic"
 
 # Create directory for build
 rm -rf /tmp/slic
@@ -25,7 +26,7 @@ fi
 cd silabs_proslic_api
 make release KDIR=$kernel_path
 if [ $? != 0 ]; then
-	echo "Failed to build Silabs release"
+	echo "Failed to build Silabs SLIC"
 	leave 1
 fi
 
@@ -39,7 +40,6 @@ mkdir -p $mod_path/silabs_proslic_api
 find ./ -name "*.ko" | xargs cp -t $mod_path/silabs_proslic_api
 cd -
 
-
 # Silabs voice tool
 cd $kernel_path
 cd tools/voice/silabs
@@ -53,5 +53,48 @@ mkdir -p $bin_path
 cp mv_voice_tool-* $bin_path/.
 cd -
 
+if [ ! -n "$4" ]; then
+	leave 1
+fi
+
+if [ ! -n "$5" ]; then
+	verion_str="silabs_proslic_api"
+else
+	verion_str="silabs_proslic_api-${5}"
+fi
+
+# Clear Silabs SLIC
+rm -rf /tmp/slic
+mkdir -p /tmp/slic
+cd /tmp/slic
+
+# Clone Silabs SLIC moduel again
+git clone git://vgitil04/slic/silabs_proslic_api
+if [ $? != 0 ]; then
+	echo "Failed to clone Silabs SLIC"
+	leave 1
+fi
+
+# Make internal version
+mkdir -p $release_path/silabs_proslic_api
+cd /tmp/slic
+mv silabs_proslic_api ${verion_str}
+tar czf ${verion_str}_internal.tgz ${verion_str}
+mv ${verion_str}_internal.tgz $release_path/silabs_proslic_api
+
+# make Silabs SLIC release version
+mv ${verion_str} silabs_proslic_api
+cd silabs_proslic_api
+make release KDIR=$kernel_path
+if [ $? != 0 ]; then
+	echo "Failed to build Silabs SLIC"
+	leave 1
+fi
+
+cd /tmp/slic
+mv silabs_proslic_api ${verion_str}
+tar czf ${verion_str}.tgz ${verion_str}
+mv ${verion_str}.tgz $release_path/silabs_proslic_api
+cd ${org_path}
 
 leave 0
-- 
1.7.5.4

