From 9f51e50ab1bc99caa6c19c925538f85a609ad803 Mon Sep 17 00:00:00 2001
From: Victor Gu <xigu@marvell.com>
Date: Thu, 19 Jun 2014 14:16:18 +0800
Subject: [PATCH 1717/1825] fix: tools: voice build script issue for zarlink

https://github.com/MISL-EBU-System-SW/misl-windriver.git linux-3.4.69-14t2-read
commit 65eb4a129a49c407ebd307fe3f8dd354687582f4

    need to return 0 if case no release related parameters are inputted
Signed-off-by: Victor Gu <xigu@marvell.com>

Change-Id: I99e44178c2c125447223bc0f2bcd188e6b290e0f
Reviewed-on: http://vgitil04.il.marvell.com:8080/8599
Reviewed-by: Yehuda Yitschak <yehuday@marvell.com>
Tested-by: Yehuda Yitschak <yehuday@marvell.com>
Signed-off-by: Zhong Hongbo <hongbo.zhong@windriver.com>
---
 tools/voice/build-zarlink.sh |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/tools/voice/build-zarlink.sh b/tools/voice/build-zarlink.sh
index e32cbb9..802d5e8 100755
--- a/tools/voice/build-zarlink.sh
+++ b/tools/voice/build-zarlink.sh
@@ -48,7 +48,7 @@ cp mv_voice_tool-* $bin_path/.
 cd -
 
 if [ ! -n "$4" ]; then
-	leave 1
+	leave 0
 fi
 
 if [ ! -n "$5" ]; then
-- 
1.7.5.4

