From 85cc0dda4ced9dabb4e4ee17f61e9479e3bb192f Mon Sep 17 00:00:00 2001
From: Victor Gu <xigu@marvell.com>
Date: Fri, 13 Jun 2014 00:00:35 +0800
Subject: [PATCH 1709/1825] fix: alp: tools: correct the Readme file for SLIC
 modules.

https://github.com/MISL-EBU-System-SW/misl-windriver.git linux-3.4.69-14t2-read
commit e3cf2de26705e3287fe15c4fee110c2b3d06cf0d

    Linux devices tal/vpapi/silabs will be created automatically by module
themselves, no need to creat them manually.
Signed-off-by: Victor Gu <xigu@marvell.com>

Change-Id: I16c64141049e29d34b281c4e2e310033e1c97535
Reviewed-on: http://vgitil04.il.marvell.com:8080/8515
Tested-by: Star_Automation <star@marvell.com>
Reviewed-by: Yehuda Yitschak <yehuday@marvell.com>
Signed-off-by: Zhong Hongbo <hongbo.zhong@windriver.com>
---
 tools/voice/lantiq-dfev/README    |    3 +--
 tools/voice/lantiq/kernel/README  |    2 +-
 tools/voice/silabs/README         |    8 ++------
 tools/voice/zarlink/kernel/README |    8 ++------
 4 files changed, 6 insertions(+), 15 deletions(-)

diff --git a/tools/voice/lantiq-dfev/README b/tools/voice/lantiq-dfev/README
index 874b6c8..e6d277b 100644
--- a/tools/voice/lantiq-dfev/README
+++ b/tools/voice/lantiq-dfev/README
@@ -92,8 +92,7 @@
 
 1. Copy all modules and mv_voice_tool to the target and run it in AMP mode.
 
-2. Make device nodes under /dev:
-	# mknod /dev/tal c 10 74
+2. Make device nodes under /dev(should be changed if conflicts with existed device):
 	# mknod /dev/sdd c 253 0
 
 3. Load drivers:
diff --git a/tools/voice/lantiq/kernel/README b/tools/voice/lantiq/kernel/README
index 54a62c8..4490fd3 100644
--- a/tools/voice/lantiq/kernel/README
+++ b/tools/voice/lantiq/kernel/README
@@ -43,7 +43,7 @@
 
 1. Copy all modules and mv_voice_tool to the target.
 
-2. Make device node under /dev:
+2. Make device node under /dev(should be changed if conflicts with existed device):
 	# mknod /dev/tapi c 253 0
 
 3. Load Lantiq drivers:
diff --git a/tools/voice/silabs/README b/tools/voice/silabs/README
index 57ba990..7dc1153 100644
--- a/tools/voice/silabs/README
+++ b/tools/voice/silabs/README
@@ -24,12 +24,8 @@
 
 1. Copy all modules and mv_voice_tool utilities to the target.
 
-2. Make device nodes under /dev:
-	# mknod /dev/tdm c 10 74
-	# mknod /dev/silabs c 10 73
-
-3. Load Silabs drivers (replace si3226x with proper SLIC type):
+2. Load Silabs drivers (replace si3226x with proper SLIC type):
 	# insmod silabs-si3226x.ko
 
-4. Run mv_voice_tool (replace si3226x with proper SLIC type):
+3. Run mv_voice_tool (replace si3226x with proper SLIC type):
 	# ./mv_voice_tool-si3226x
diff --git a/tools/voice/zarlink/kernel/README b/tools/voice/zarlink/kernel/README
index eea3d3e..c6900a1 100644
--- a/tools/voice/zarlink/kernel/README
+++ b/tools/voice/zarlink/kernel/README
@@ -24,12 +24,8 @@
 
 1. Copy all modules and mv_voice_tool utilities to the target.
 
-2. Make device nodes under /dev:
-	# mknod /dev/tdm c 10 74
-	# mknod /dev/vpapi c 10 73
-
-3. Load Zarlink drivers (replace ve880 with proper SLIC type):
+2. Load Zarlink drivers (replace ve880 with proper SLIC type):
 	# insmod zarlink-ve880.ko
 
-4. Run mv_voice_tool (replace ve880 with proper SLIC type):
+3. Run mv_voice_tool (replace ve880 with proper SLIC type):
 	# ./mv_voice_tool-ve880
-- 
1.7.5.4

