From 620349175987d7b919679e2132b1b540741ef2ba Mon Sep 17 00:00:00 2001
From: Kosta Zertsekel <konszert@marvell.com>
Date: Mon, 22 Jul 2013 15:01:03 +0300
Subject: [PATCH 0887/1825] Revert "Added component patches to .gitignore"

https://github.com/MISL-EBU-System-SW/misl-windriver.git linux-3.4.69-14t2-read
commit 13bac805464623bf332e601636d3f5e125b93feb

	This reverts commit 14deb6e5d6e624f031e97594ba799b3cd5630e90.
	Rationate: keep .gitignore file clean from repo quirks and fixes,
	because this file is used as-is in unified kernel git.

Signed-off-by: Kosta Zertsekel <konszert@marvell.com>
Change-Id: I5e2f72260ec2eb234322e7f41460fd06b9b4bbb3
Reviewed-on: http://vgitil04.il.marvell.com:8080/2758
Tested-by: Star_Automation <star@marvell.com>
Reviewed-by: Igor Patrik <igorp@marvell.com>
Signed-off-by: Zhong Hongbo <hongbo.zhong@windriver.com>
---
 .gitignore |   10 ----------
 1 files changed, 0 insertions(+), 10 deletions(-)

diff --git a/.gitignore b/.gitignore
index 7bd3ff9..ef96988 100644
--- a/.gitignore
+++ b/.gitignore
@@ -1,4 +1,3 @@
-
 #
 # NOTE! Don't add files that are generated in specific
 # subdirectories here. Add them in the ".gitignore" file
@@ -35,15 +34,6 @@ modules.builtin
 *.gcno
 
 #
-#
-#
-arch/arm/mach-armadaxp/armada_xp_family/
-arch/arm/mach-avantalp/avanta_lp_family/
-arch/arm/plat-armada/common/
-arch/arm/plat-armada/mv_hal/
-arch/arm/plat-armada/mv_drivers_lsp/mv_pp2/
-
-#
 # Top-level generic files
 #
 /tags
-- 
1.7.5.4

