From a69192752829d795ca23612bb12d15580a431ddf Mon Sep 17 00:00:00 2001
From: Eran Ben-Avi <benavi@marvell.com>
Date: Thu, 11 Jul 2013 17:57:01 +0300
Subject: [PATCH 0799/1825] silabs/zarlink: Added README files for
 mv_voice_tool.

https://github.com/MISL-EBU-System-SW/misl-windriver.git linux-3.4.69-14t2-read
commit 5e9af4ec4bc7e70e1a9abf9eb966226ce321a96b

	The added README files describe build process of mv_voice_tool with
	external driver packages.

Change-Id: I843a215f8b509e409c89d78405ed223a6612ba3d
Signed-off-by: Piotr Ziecik <kosmo@semihalf.com>
Signed-off-by: Eran Ben-Avi <benavi@marvell.com>
Signed-off-by: Zhong Hongbo <hongbo.zhong@windriver.com>
---
 tools/voice/silabs/README         |   40 +++++++++++++++++++++++++++++++++++++
 tools/voice/zarlink/kernel/README |   40 +++++++++++++++++++++++++++++++++++++
 2 files changed, 80 insertions(+), 0 deletions(-)
 create mode 100644 tools/voice/silabs/README
 create mode 100644 tools/voice/zarlink/kernel/README

diff --git a/tools/voice/silabs/README b/tools/voice/silabs/README
new file mode 100644
index 0000000..67a541c
--- /dev/null
+++ b/tools/voice/silabs/README
@@ -0,0 +1,40 @@
++-------------------------------------------------------------------------+
+| How to compile Silabs mv_voice_tool with Silabs external module package |
++-------------------------------------------------------------------------+
+
+1. Unpack, configure and build LSP. Make sure that the following
+   options are set:
+
+	CONFIG_SILABS_SLIC_SUPPORT=y
+	CONFIG_SILABS_SLIC_EXTERNAL_DRIVER=y
+	CONFIG_MV_TDM_PCM_CLK_8MHZ=y
+	CONFIG_MV_TDM_USE_DCO=y
+
+2. Build Silabs Drivers (shipped separately):
+
+	$ cd silabs-slic
+	$ make -C <path-to-lsp> ARCH=arm CROSS_COMPILE=arm-marvell-linux-gnueabi- \
+		M=$PWD
+	$ ls -l silabs-*/*.ko
+
+3. Build mv_voice_tool for Silabs SLICs
+	$ cd <path-to-lsp>/tools/voice/silabs
+	$ make ARCH=arm CROSS_COMPILE=arm-marvell-linux-gnueabi-		\
+		SILABS_SLIC=<path-to-silabs-slic>
+	$ ls -l mv_voice_tool-*
+
++---------------------------------------------+
+| How to run Silabs drivers and mv_voice_tool |
++---------------------------------------------+
+
+1. Copy all modules and mv_voice_tool utilities to the target.
+
+2. Make device nodes under /dev:
+	# mknod /dev/tdm c 10 74
+	# mknod /dev/silabs c 10 73
+
+3. Load Silabs drivers (replace si3226x with proper SLIC type):
+	# insmod silabs-si3226x.ko
+
+4. Run mv_voice_tool (replace si3226x with proper SLIC type):
+	# ./mv_voice_tool-si3226x
diff --git a/tools/voice/zarlink/kernel/README b/tools/voice/zarlink/kernel/README
new file mode 100644
index 0000000..951326e
--- /dev/null
+++ b/tools/voice/zarlink/kernel/README
@@ -0,0 +1,40 @@
++---------------------------------------------------------------------------+
+| How to compile Zarlink mv_voice_tool with Zarlink external module package |
++---------------------------------------------------------------------------+
+
+1. Unpack, configure and build LSP. Make sure that the following
+   options are set:
+
+	CONFIG_ZARLINK_SLIC_SUPPORT=y
+	CONFIG_ZARLINK_SLIC_EXTERNAL_DRIVER=y
+	CONFIG_MV_TDM_PCM_CLK_8MHZ=y
+	CONFIG_MV_TDM_USE_INTERNAL_PCLK_SOURCE=y
+
+2. Build Zarlink Drivers (shipped separately):
+
+	$ cd zarlink-vpapi
+	$ make -C <path-to-lsp> ARCH=arm CROSS_COMPILE=arm-marvell-linux-gnueabi- \
+		M=$PWD
+	$ ls -l zarlink-*/*.ko
+
+3. Build mv_voice_tool for Zarlink SLICs
+	$ cd <path-to-lsp>/tools/voice/zarlink/kernel
+	$ make ARCH=arm CROSS_COMPILE=arm-marvell-linux-gnueabi-		\
+		ZARLINK_VPAPI=<path-to-zarlink-vpapi>
+	$ ls -l mv_voice_tool-*
+
++----------------------------------------------+
+| How to run Zarlink drivers and mv_voice_tool |
++----------------------------------------------+
+
+1. Copy all modules and mv_voice_tool utilities to the target.
+
+2. Make device nodes under /dev:
+	# mknod /dev/tdm c 10 74
+	# mknod /dev/vpapi c 10 73
+
+3. Load Zarlink drivers (replace ve880 with proper SLIC type):
+	# insmod zarlink-ve880.ko
+
+4. Run mv_voice_tool (replace ve880 with proper SLIC type):
+	# ./mv_voice_tool-ve880
-- 
1.7.5.4

