From b53e36200ccfe81a523b408308a1c88da62a69e5 Mon Sep 17 00:00:00 2001
From: Eran Ben-Avi <benavi@marvell.com>
Date: Wed, 31 Jul 2013 19:26:25 +0300
Subject: [PATCH 0916/1825] tdm: alp: update tone test

https://github.com/MISL-EBU-System-SW/misl-windriver.git linux-3.4.69-14t2-read
commit 2dad218c363b75e4249886bcdb621d262c03d3cb

	- stop pcm channels for each frequency
	  in tone test

Change-Id: If216571aef52c8399954e28be5109aed7b67ec44
Signed-off-by: Eran Ben-Avi <benavi@marvell.com>
Reviewed-on: http://vgitil04.il.marvell.com:8080/2907
Tested-by: Star_Automation <star@marvell.com>
Reviewed-by: Omri Itach <omrii@marvell.com>
Reviewed-by: Igor Patrik <igorp@marvell.com>
Signed-off-by: Zhong Hongbo <hongbo.zhong@windriver.com>
---
 tools/voice/zarlink/kernel/mv_voice_tool.c |    6 ++++++
 1 files changed, 6 insertions(+), 0 deletions(-)

diff --git a/tools/voice/zarlink/kernel/mv_voice_tool.c b/tools/voice/zarlink/kernel/mv_voice_tool.c
index 02ab6bd..12ca296 100644
--- a/tools/voice/zarlink/kernel/mv_voice_tool.c
+++ b/tools/voice/zarlink/kernel/mv_voice_tool.c
@@ -944,6 +944,12 @@ static void sw_tone_test(int tdm_fd, unsigned char line_id)
 			/* Reload timeout */
 			timeout.tv_usec = TIMEOUT;
 		}
+
+		if (ioctl(tdm_fd, TDM_DEV_PCM_STOP, 0)) {
+			printf("Error, unable to stop pcm bus\n");
+			return;
+		}
+
 	}
 }
 
-- 
1.7.5.4

