From b671c67b5ab3d5b66f4f8538f1d8ddda673b2cd4 Mon Sep 17 00:00:00 2001
From: Eran Ben-Avi <benavi@marvell.com>
Date: Thu, 11 Jul 2013 17:50:41 +0300
Subject: [PATCH 0795/1825] tools: Update mv_voice_tool to compile among with
 external SLIC drivers.

https://github.com/MISL-EBU-System-SW/misl-windriver.git linux-3.4.69-14t2-read
commit 5800756d3e92872799b20ceb45d0f747da977f33

	The Zarlink (kernel) and Silabs tools were modified. The SLIC driver
	package location has to be pointed in ZARLINK_VPAPI and SLIABS_SLIC
	make variables respectively.

Change-Id: I68cec67abd83d7474fe0e104551dde2bd374f66a
Signed-off-by: Piotr Ziecik <kosmo@semihalf.com>
Signed-off-by: Eran Ben-Avi <benavi@marvell.com>
Signed-off-by: Zhong Hongbo <hongbo.zhong@windriver.com>
---
 tools/voice/silabs/Makefile         |   19 +++++++++++++++++++
 tools/voice/silabs/silabs_lib.h     |    2 +-
 tools/voice/zarlink/kernel/Makefile |   17 ++++++++++++++++-
 3 files changed, 36 insertions(+), 2 deletions(-)

diff --git a/tools/voice/silabs/Makefile b/tools/voice/silabs/Makefile
index 1832f85..e1cee06 100644
--- a/tools/voice/silabs/Makefile
+++ b/tools/voice/silabs/Makefile
@@ -20,11 +20,29 @@ AR := $(CROSS_COMPILE)ar
 #CFLAGS += -I$(KSRC)/tools/voice/
 
 ifeq ($(CONFIG_SILABS_SLIC_SUPPORT),y)
+ifeq ($(CONFIG_SILAB_SLIC_EXTERNAL_DRIVER),y)
+CFLAGS += -I$(KSRC)/arch/arm/plat-feroceon/mv_drivers_lsp/mv_phone/
+CFLAGS += -I$(SILABS_SLIC)/proslic_api/inc -I$(SILABS_SLIC)/proslic_api/custom
+CFLAGS += -I$(SILABS_SLIC)/silabs_dev
+
+all: mv_voice_tool-si3217x mv_voice_tool-si3226x
+
+mv_voice_tool-si3217x: mv_voice_tool.c silabs_lib.c
+	$(CC) $(CFLAGS) -DMV_TDM_USE_DCO -DSI3217x -o $@ $^
+
+mv_voice_tool-si3226x: mv_voice_tool.c silabs_lib.c
+	$(CC) $(CFLAGS) -DMV_TDM_USE_DCO -DSI3226x -o $@ $^
+
+clean:
+	rm -f mv_voice_tool-si3217x mv_voice_tool-si3226x
+
+else
 TARGETS += siapi_lib.o
 LIBS := libsiapi.a
 TOOL := mv_voice_tool.c
 CFLAGS += -I$(KSRC)/arch/arm/plat-feroceon/mv_hal/voiceband/slic/silabs/arch_marvell/
 CFLAGS += -I$(KSRC)/arch/arm/plat-feroceon/mv_drivers_lsp/mv_phone/
+CFLAGS += -I$(KSRC)/arch/arm/plat-feroceon/mv_drivers_lsp/mv_phone/slic/
 CFLAGS += -I$(KSRC)/arch/arm/plat-feroceon/mv_hal/voiceband/slic/silabs/inc
 CFLAGS += -I$(KSRC)/arch/arm/plat-feroceon/mv_hal/voiceband/slic/silabs/custom/
 CFLAGS += -I$(KSRC)/arch/arm/plat-feroceon/mv_drivers_lsp/mv_phone/test/
@@ -55,4 +73,5 @@ clean:
 	@for subdir in $(SUBDIRS); do \
                 (cd $$subdir && $(MAKE) $@) || exit 1; \
         done
+endif #CONFIG_SILAB_SLIC_EXTERNAL_DRIVER
 endif #CONFIG_SILABS_SLIC_SUPPORT
diff --git a/tools/voice/silabs/silabs_lib.h b/tools/voice/silabs/silabs_lib.h
index 6e4ab67..385b2cd 100644
--- a/tools/voice/silabs/silabs_lib.h
+++ b/tools/voice/silabs/silabs_lib.h
@@ -66,7 +66,7 @@ SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 #define _SILABS_LIB_H
 
 #include <stdbool.h>
-#include <slic/silabs_dev.h>
+#include <silabs_dev.h>
 
 int silabs_open_device(void);
 int silabs_close_device(void);
diff --git a/tools/voice/zarlink/kernel/Makefile b/tools/voice/zarlink/kernel/Makefile
index 1b500f5..286f157 100644
--- a/tools/voice/zarlink/kernel/Makefile
+++ b/tools/voice/zarlink/kernel/Makefile
@@ -41,6 +41,21 @@ ifeq ($(CONFIG_ZARLINK_SLIC_VE890),y)
  PROFILE = profile_89116.c
 endif
 
+ifeq ($(CONFIG_ZARLINK_SLIC_EXTERNAL_DRIVER),y)
+CFLAGS += -I$(ZARLINK_VPAPI)/api_lib/includes -I$(ZARLINK_VPAPI)/vp_hal -I$(ZARLINK_VPAPI)/vapi_dev
+
+all: mv_voice_tool-ve880 mv_voice_tool-ve890
+
+mv_voice_tool-ve880: mv_voice_tool.c vpapi_lib.c profile_88266.c
+	$(CC) $(CFLAGS) -DZARLINK_SLIC_VE880 -static -o $@ $^
+
+mv_voice_tool-ve890: mv_voice_tool.c vpapi_lib.c profile_89116.c
+	$(CC) $(CFLAGS) -DZARLINK_SLIC_VE890 -static -o $@ $^
+
+clean:
+	rm -f mv_voice_tool-ve880 mv_voice_tool-ve890
+
+else
 CFLAGS += -I$(KSRC)/arch/arm/plat-feroceon/mv_drivers_lsp/mv_phone/test/
 
 all:
@@ -56,5 +71,5 @@ clean:
 	@for subdir in $(SUBDIRS); do \
                 (cd $$subdir && $(MAKE) $@) || exit 1; \
         done
-
+endif #CONFIG_ZARLINK_SLIC_EXTERNAL_DRIVER
 endif #CONFIG_ZARLINK_SLIC_SUPPORT
-- 
1.7.5.4

