From e1fb76bb538ddd6cd440fa6b8f7ab31f03fb6c13 Mon Sep 17 00:00:00 2001
From: Eran Ben-Avi <benavi@marvell.com>
Date: Thu, 11 Jul 2013 18:05:38 +0300
Subject: [PATCH 0804/1825] tools/voice: Remove external module package build
 instructions from README files.

https://github.com/MISL-EBU-System-SW/misl-windriver.git linux-3.4.69-14t2-read
commit fd70ee8b07bebdbe83d11acb4bd2d38e3a98d7d2

	The instructions how to build external module packages are now placed
	in README files shipped with the packages.

Change-Id: Ibc1a4f0a8cf56266628679969f1eccc8421265e5
Signed-off-by: Piotr Ziecik <kosmo@semihalf.com>
Signed-off-by: Eran Ben-Avi <benavi@marvell.com>
Signed-off-by: Zhong Hongbo <hongbo.zhong@windriver.com>
---
 tools/voice/silabs/README         |    7 +------
 tools/voice/zarlink/kernel/README |    7 +------
 2 files changed, 2 insertions(+), 12 deletions(-)

diff --git a/tools/voice/silabs/README b/tools/voice/silabs/README
index 67a541c..57ba990 100644
--- a/tools/voice/silabs/README
+++ b/tools/voice/silabs/README
@@ -10,12 +10,7 @@
 	CONFIG_MV_TDM_PCM_CLK_8MHZ=y
 	CONFIG_MV_TDM_USE_DCO=y
 
-2. Build Silabs Drivers (shipped separately):
-
-	$ cd silabs-slic
-	$ make -C <path-to-lsp> ARCH=arm CROSS_COMPILE=arm-marvell-linux-gnueabi- \
-		M=$PWD
-	$ ls -l silabs-*/*.ko
+2. Build Silabs Drivers (shipped separately).
 
 3. Build mv_voice_tool for Silabs SLICs
 	$ cd <path-to-lsp>/tools/voice/silabs
diff --git a/tools/voice/zarlink/kernel/README b/tools/voice/zarlink/kernel/README
index 951326e..eea3d3e 100644
--- a/tools/voice/zarlink/kernel/README
+++ b/tools/voice/zarlink/kernel/README
@@ -10,12 +10,7 @@
 	CONFIG_MV_TDM_PCM_CLK_8MHZ=y
 	CONFIG_MV_TDM_USE_INTERNAL_PCLK_SOURCE=y
 
-2. Build Zarlink Drivers (shipped separately):
-
-	$ cd zarlink-vpapi
-	$ make -C <path-to-lsp> ARCH=arm CROSS_COMPILE=arm-marvell-linux-gnueabi- \
-		M=$PWD
-	$ ls -l zarlink-*/*.ko
+2. Build Zarlink Drivers (shipped separately).
 
 3. Build mv_voice_tool for Zarlink SLICs
 	$ cd <path-to-lsp>/tools/voice/zarlink/kernel
-- 
1.7.5.4

