From 484b3e690f0886bc6ff6514266298e200c33f13d Mon Sep 17 00:00:00 2001
From: Eran Ben-Avi <benavi@marvell.com>
Date: Thu, 11 Jul 2013 17:46:03 +0300
Subject: [PATCH 0792/1825] zarlink: Update VE880 and VE890 profile data.

https://github.com/MISL-EBU-System-SW/misl-windriver.git linux-3.4.69-14t2-read
commit ea458a20f82e021377659111baa45951912fba7d

	The profiles were generated by Profile Wizard Version 2.4.0.

Change-Id: I5a3dac0d285b029918277eeba263237b3830ecec
Signed-off-by: Piotr Ziecik <kosmo@semihalf.com>
Signed-off-by: Eran Ben-Avi <benavi@marvell.com>
Signed-off-by: Zhong Hongbo <hongbo.zhong@windriver.com>
---
 tools/voice/zarlink/kernel/mv_voice_tool.c |    2 +-
 tools/voice/zarlink/kernel/profile_88266.c |  148 +++++++---------------------
 tools/voice/zarlink/kernel/profile_88266.h |   31 +++----
 tools/voice/zarlink/kernel/profile_89116.c |   10 +-
 tools/voice/zarlink/kernel/profile_89116.h |    6 +-
 5 files changed, 56 insertions(+), 141 deletions(-)

diff --git a/tools/voice/zarlink/kernel/mv_voice_tool.c b/tools/voice/zarlink/kernel/mv_voice_tool.c
index bc40847..fe8dacb 100644
--- a/tools/voice/zarlink/kernel/mv_voice_tool.c
+++ b/tools/voice/zarlink/kernel/mv_voice_tool.c
@@ -81,7 +81,7 @@ SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 #define DEV_PROFILE		ABS_VBL_FLYBACK
 #define DC_COEFF		DC_22MA_CC
 #define AC_COEFF		AC_FXS_RF14_DEF
-#define WB_AC_COEFF		AC_FXS_RF14_WB_US
+#define WB_AC_COEFF		AC_FXS_RF14_DEF
 #define RING_PROFILE		RING_DEF
 
 #elif defined(ZARLINK_SLIC_VE890)
diff --git a/tools/voice/zarlink/kernel/profile_88266.c b/tools/voice/zarlink/kernel/profile_88266.c
index e9ab45c..866243f 100644
--- a/tools/voice/zarlink/kernel/profile_88266.c
+++ b/tools/voice/zarlink/kernel/profile_88266.c
@@ -1,28 +1,37 @@
 /*
- * profile_8266.c --
+ * profile_88266.c --
  *
  * This file contains profile data in byte format
  *
- *   This file was generated with Profile Wizard Version: P1.12.3
+ * Project Info --
+ *   File:   C:\Users\Kosmo\Documents\VE880.vpw
+ *   Type:   Le71HR0865G Line Module Project - VBL Reg ABS Power Supply and 8.192MHz PCLK
+ *   Date:   Thursday, April 18, 2013 12:59:41
+ *   Device: VE880 Le88266
+ *
+ *   This file was generated with Profile Wizard Version: P2.4.0
  */
 
 #include "profile_88266.h"
 
 
-/*** Device Parameters for VoicePort Device Family ***/
+/************** Device Parameters **************/
+
+/* Device Configuration Data */
 const VpProfileDataType ABS_VBL_FLYBACK[] =
 {
-  /* Device Profile */
- 0x00, 0xFF, 0x00, 0x1A, /* Profile Index and Length */
- 0x04, 0x00, /* Version and MPI Command */
+  /* Device Profile for VE880 Device Family */
+ 0x00, 0xFF, 0x00, 0x20, /* Profile Index and Length */
+ 0x05, 0x00, /* Version and MPI Command */
  0x20, 0x00, /* PCLK Frequency */
- 0x8A, 0x00, /* Device Config 1 and MCLK Correction */
+ 0x8A, 0x00, /* Device Config 1, Battery Control Mode */
  0x40, 0x02, /* Clock Slot and Max Events */
- 0x0A, 0x00, /* Device Tickrate */
- 0xE4, 0x00, 0x05, 0x00, /* Switching Regulator Parameters */
- 0x00, 0x00, /* RSVD */
- 0x22, 0x00, 0x55, 0x55, /* Linecard Parameters */
- 0xB2, 0x00, 0xB1, 0x00, 0xB0, 0x40  /* Regulator Timing Parameters */
+ 0x05, 0x00, /* Device Tickrate */
+ 0xE4, 0x00, 0x04, 0x00, /* Switching Regulator Parameters */
+ 0x1B, 0x05, /* Y-Voltage, Z-Voltage */
+ 0x00, 0x00, 0x00, 0x00, /* RSVD */
+ 0xB1, 0x00, 0xB1, 0x00, 0xB0, 0x40, /* Regulator Timing Parameters */
+ 0x2C, 0x40, 0x2C, 0x40, 0x2C, 0x40  /* FreeRun Timing Parameters */
 };
 
 /************** AC Filter Coefficients **************/
@@ -430,92 +439,13 @@ const VpProfileDataType AC_FXS_RF14_US_SS[] =
  0x72, 0xA5, 0x2B, 0x50, 0x96, 0x1D, 0x01, 0x00
 };
 
-/************** WideBand AC Filter Coefficients **************/
-/* AC FXS RF14 600 Ohm Wideband Coefficients */
-const VpProfileDataType AC_FXS_RF14_WB_US[] =
-{
-  /* AC Profile */
- 0xED, 0x00, 0xA9, 0x4C, 0x01, 0x49, 0xCA, 0xFC, 0x98, 0xCC, 0x2A, 0xA2,
- 0x3B, 0x98, 0xB3, 0x43, 0xA2, 0x3A, 0xAB, 0x9A, 0x2D, 0xAA, 0xB6, 0x9F,
- 0x01, 0x8A, 0x2E, 0x01, 0xC4, 0xA0, 0x2E, 0x23, 0x3A, 0x2E, 0x9F, 0x87,
- 0x4C, 0xCE, 0xB2, 0x27, 0x88, 0xDE, 0x20, 0xAA, 0xCA, 0x22, 0x32, 0x23,
- 0xAA, 0xA2, 0xB4, 0x2A, 0x87, 0x82, 0xA8, 0x71, 0x80, 0xA9, 0xF0, 0x50,
- 0x00, 0x86, 0xCB, 0x5E, 0x22, 0xBC, 0x12, 0xA3, 0xA9, 0xFA, 0xC6, 0x34,
- 0xE2, 0x9F, 0xBA, 0xF0, 0x96, 0x2E, 0x01, 0x00
-};
-
-/* AC FXS RF14 European Union Wideband Coefficients */
-const VpProfileDataType AC_FXS_RF14_WB_EU[] =
-{
-  /* AC Profile */
- 0xAA, 0x00, 0x9C, 0x4C, 0x01, 0x49, 0xCA, 0xE0, 0x98, 0xAA, 0xA3, 0x6B,
- 0x22, 0x43, 0xA2, 0xAC, 0xBF, 0x22, 0xA6, 0x9A, 0x01, 0x90, 0x01, 0x01,
- 0x90, 0x8A, 0x2E, 0x01, 0xBA, 0x10, 0xA2, 0x3A, 0xBA, 0xE9, 0xAA, 0xB5,
- 0x2A, 0xBC, 0xAF, 0x97, 0x88, 0xAB, 0xC0, 0x4A, 0xBE, 0xBA, 0x9F, 0xC3,
- 0xAF, 0x3A, 0x3F, 0xCD, 0xA7, 0x82, 0xC8, 0x71, 0x80, 0x69, 0xF0, 0x50,
- 0x00, 0x86, 0xBA, 0x52, 0xB2, 0xBA, 0x12, 0x42, 0x4A, 0x2A, 0x33, 0xBA,
- 0x33, 0x34, 0x4B, 0x40, 0x96, 0xB2, 0xA0, 0x00
-};
-
-/* AC FXS RF14 France Wideband Coefficients */
-const VpProfileDataType AC_FXS_RF14_WB_FR[] =
-{
-  /* AC Profile */
- 0x98, 0x00, 0x26, 0x4C, 0x01, 0x49, 0xCA, 0xD7, 0x98, 0x4F, 0x22, 0xB5,
- 0xB1, 0xC6, 0x23, 0x3A, 0xAB, 0x24, 0xB5, 0x9A, 0x01, 0x90, 0x01, 0x01,
- 0x90, 0x8A, 0x01, 0x11, 0x22, 0x10, 0x3A, 0x29, 0x3C, 0x29, 0x32, 0xB3,
- 0x4A, 0x2C, 0x2E, 0xB6, 0x88, 0xAC, 0xB0, 0x43, 0x37, 0xAB, 0x3E, 0xB4,
- 0x3E, 0x22, 0x37, 0x22, 0xAF, 0x82, 0xA8, 0x71, 0x80, 0xA9, 0xF0, 0x50,
- 0x00, 0x86, 0xAA, 0x52, 0xE2, 0x2B, 0x12, 0x62, 0x63, 0x3A, 0x24, 0xA2,
- 0x5C, 0x36, 0x39, 0xF0, 0x96, 0x2E, 0x01, 0x00
-};
-
-/* AC FXS RF14 German Wideband Coefficients */
-const VpProfileDataType AC_FXS_RF14_WB_DE[] =
-{
-  /* AC Profile */
- 0x27, 0x00, 0xD7, 0x4C, 0x01, 0x49, 0xCA, 0xDB, 0x98, 0xBA, 0xA1, 0xBA,
- 0x22, 0x2A, 0xF4, 0xAA, 0xAB, 0x22, 0x65, 0x9A, 0x01, 0x90, 0x01, 0x01,
- 0x90, 0x8A, 0x2E, 0x01, 0xA3, 0x10, 0x2B, 0x29, 0x53, 0xB9, 0xBD, 0xB3,
- 0x7B, 0xCC, 0xDB, 0xA6, 0x88, 0x2A, 0xB0, 0x2B, 0x9F, 0x2E, 0x3E, 0x22,
- 0xBE, 0xC3, 0x26, 0x35, 0x2F, 0x82, 0xB8, 0x71, 0x80, 0xAB, 0xF0, 0x50,
- 0x00, 0x86, 0xCA, 0x5A, 0xA2, 0xBA, 0x1F, 0x22, 0x3B, 0x2A, 0x23, 0x2B,
- 0x3A, 0x24, 0x4B, 0x40, 0x96, 0x22, 0x21, 0x00
-};
-
-/* AC FXS RF14 China Wideband Coefficients */
-const VpProfileDataType AC_FXS_RF14_WB_CN[] =
-{
-  /* AC Profile */
- 0xEF, 0x00, 0x46, 0x4C, 0x01, 0x49, 0xCA, 0xDA, 0x98, 0xB3, 0xB1, 0x23,
- 0x32, 0x64, 0x97, 0xAA, 0x2B, 0x4A, 0xA4, 0x9A, 0x01, 0x90, 0x01, 0x01,
- 0x90, 0x8A, 0x2E, 0x01, 0xCD, 0x10, 0x2B, 0x29, 0x22, 0x2A, 0xA2, 0x33,
- 0x4B, 0x4C, 0xBD, 0xF6, 0x88, 0x2A, 0xC0, 0x52, 0x97, 0xF2, 0x2E, 0xAA,
- 0x4E, 0xAC, 0xB5, 0x5B, 0xBE, 0x82, 0xA8, 0x71, 0x80, 0xA9, 0xF0, 0x50,
- 0x00, 0x86, 0x2A, 0x5A, 0xA2, 0xBA, 0x1B, 0x22, 0xCA, 0x24, 0xC3, 0x43,
- 0x42, 0xA4, 0x4B, 0x50, 0x96, 0x3C, 0x41, 0x00
-};
-
-/* AC FXS RF14 Australia Wideband Coefficients */
-const VpProfileDataType AC_FXS_RF14_WB_AU[] =
-{
-  /* AC Profile */
- 0xC2, 0x00, 0x2E, 0x4C, 0x01, 0x49, 0xCA, 0xF0, 0x98, 0xBA, 0xA1, 0xAD,
- 0xA3, 0xA2, 0x2C, 0x2D, 0xAB, 0x22, 0xD4, 0x9A, 0xCC, 0xA2, 0x97, 0x9F,
- 0x01, 0x8A, 0x2E, 0x01, 0x7A, 0x10, 0xBA, 0x29, 0x2A, 0x3A, 0xA2, 0x32,
- 0xB2, 0xAA, 0xF2, 0xA4, 0x88, 0xBA, 0xA0, 0x5C, 0x3F, 0x2C, 0xCE, 0xAA,
- 0x87, 0xC2, 0x66, 0xDB, 0x7E, 0x82, 0xA8, 0x71, 0x80, 0x39, 0x70, 0x50,
- 0x00, 0x86, 0xAA, 0x59, 0x81, 0x25, 0x1A, 0xA1, 0xBB, 0x2B, 0xB3, 0x2C,
- 0x46, 0xD5, 0xBC, 0x60, 0x96, 0xAA, 0xB1, 0x00
-};
-
 /************** DC Feed Parameters **************/
 
 /* 22mA  current feed */
 const VpProfileDataType DC_22MA_CC[] =
 {
  /* DC Profile */
- 0x00, 0x01, 0x01, 0x0B, 0x01, 0x08, 0xC2, 0x1A, 0x88, 0xD0, 0x05, 0xC6,
+ 0x00, 0x01, 0x01, 0x0B, 0x01, 0x08, 0xC2, 0x19, 0x84, 0xB0, 0x05, 0xC6,
  0x11, 0x04, 0x01
 };
 
@@ -524,10 +454,10 @@ const VpProfileDataType DC_22MA_CC[] =
 /* Default Rnging, 25Hz, Sinewave, 80Vpk */
 const VpProfileDataType RING_DEF[] =
 {
-  /* Sine, 25 Hz, 1.41 CF, 70.00 Vpk, 0.00 Bias */
+  /* Sine, 25 Hz, 1.41 CF, 80.00 Vpk, 0.00 Bias */
   /* Ringing Profile */
  0x00, 0x04, 0x00, 0x12, 0x00, 0x0E, 0xD2, 0x00, 0x00, 0x00, 0x00, 0x44,
- 0x3A, 0x0A, 0x00, 0x00, 0x00, 0x00, 0x68, 0x01, 0x00, 0x00
+ 0x42, 0x55, 0x00, 0x00, 0x00, 0x00, 0x68, 0x01, 0x00, 0x00
 };
 
 /* US Ringing, 20Hz, Sinewave, 80Vpk */
@@ -535,7 +465,7 @@ const VpProfileDataType RING_US[] =
 {
   /* Sine, 20 Hz, 1.41 CF, 80.00 Vpk, 0.00 Bias */
   /* Ringing Profile */
- 0x00, 0x04, 0x00, 0x12, 0x00, 0x0E, 0xD2, 0x00, 0x00, 0x00, 0x00, 0x36,
+ 0x00, 0x04, 0x00, 0x12, 0x00, 0x0E, 0xD2, 0x00, 0x00, 0x00, 0x00, 0x37,
  0x42, 0x55, 0x00, 0x00, 0x00, 0x00, 0x68, 0x01, 0x00, 0x00
 };
 
@@ -544,7 +474,7 @@ const VpProfileDataType RING_CA[] =
 {
   /* Sine, 20 Hz, 1.41 CF, 80.00 Vpk, 0.00 Bias */
   /* Ringing Profile */
- 0x00, 0x04, 0x00, 0x12, 0x00, 0x0E, 0xD2, 0x00, 0x00, 0x00, 0x00, 0x36,
+ 0x00, 0x04, 0x00, 0x12, 0x00, 0x0E, 0xD2, 0x00, 0x00, 0x00, 0x00, 0x37,
  0x42, 0x55, 0x00, 0x00, 0x00, 0x00, 0x68, 0x01, 0x00, 0x00
 };
 
@@ -553,7 +483,7 @@ const VpProfileDataType RING_FR[] =
 {
   /* Sine, 50 Hz, 1.41 CF, 80.00 Vpk, 0.00 Bias */
   /* Ringing Profile */
- 0x00, 0x04, 0x00, 0x12, 0x00, 0x0E, 0xD2, 0x00, 0x00, 0x00, 0x00, 0x88,
+ 0x00, 0x04, 0x00, 0x12, 0x00, 0x0E, 0xD2, 0x00, 0x00, 0x00, 0x00, 0x89,
  0x42, 0x55, 0x00, 0x00, 0x00, 0x00, 0x68, 0x01, 0x00, 0x00
 };
 
@@ -562,7 +492,7 @@ const VpProfileDataType RING_JP[] =
 {
   /* Sine, 16 Hz, 1.41 CF, 80.00 Vpk, 0.00 Bias */
   /* Ringing Profile */
- 0x00, 0x04, 0x00, 0x12, 0x00, 0x0E, 0xD2, 0x00, 0x00, 0x00, 0x00, 0x2B,
+ 0x00, 0x04, 0x00, 0x12, 0x00, 0x0E, 0xD2, 0x00, 0x00, 0x00, 0x00, 0x2C,
  0x42, 0x55, 0x00, 0x00, 0x00, 0x00, 0x68, 0x01, 0x00, 0x00
 };
 
@@ -571,7 +501,7 @@ const VpProfileDataType RING_KR[] =
 {
   /* Sine, 20 Hz, 1.41 CF, 80.00 Vpk, 0.00 Bias */
   /* Ringing Profile */
- 0x00, 0x04, 0x00, 0x12, 0x00, 0x0E, 0xD2, 0x00, 0x00, 0x00, 0x00, 0x36,
+ 0x00, 0x04, 0x00, 0x12, 0x00, 0x0E, 0xD2, 0x00, 0x00, 0x00, 0x00, 0x37,
  0x42, 0x55, 0x00, 0x00, 0x00, 0x00, 0x68, 0x01, 0x00, 0x00
 };
 
@@ -580,7 +510,7 @@ const VpProfileDataType RING_TW[] =
 {
   /* Sine, 20 Hz, 1.41 CF, 80.00 Vpk, 0.00 Bias */
   /* Ringing Profile */
- 0x00, 0x04, 0x00, 0x12, 0x00, 0x0E, 0xD2, 0x00, 0x00, 0x00, 0x00, 0x36,
+ 0x00, 0x04, 0x00, 0x12, 0x00, 0x0E, 0xD2, 0x00, 0x00, 0x00, 0x00, 0x37,
  0x42, 0x55, 0x00, 0x00, 0x00, 0x00, 0x68, 0x01, 0x00, 0x00
 };
 
@@ -589,7 +519,7 @@ const VpProfileDataType RING_HK[] =
 {
   /* Sine, 20 Hz, 1.41 CF, 80.00 Vpk, 0.00 Bias */
   /* Ringing Profile */
- 0x00, 0x04, 0x00, 0x12, 0x00, 0x0E, 0xD2, 0x00, 0x00, 0x00, 0x00, 0x36,
+ 0x00, 0x04, 0x00, 0x12, 0x00, 0x0E, 0xD2, 0x00, 0x00, 0x00, 0x00, 0x37,
  0x42, 0x55, 0x00, 0x00, 0x00, 0x00, 0x68, 0x01, 0x00, 0x00
 };
 
@@ -598,7 +528,7 @@ const VpProfileDataType RING_SG[] =
 {
   /* Sine, 24 Hz, 1.41 CF, 80.00 Vpk, 0.00 Bias */
   /* Ringing Profile */
- 0x00, 0x04, 0x00, 0x12, 0x00, 0x0E, 0xD2, 0x00, 0x00, 0x00, 0x00, 0x41,
+ 0x00, 0x04, 0x00, 0x12, 0x00, 0x0E, 0xD2, 0x00, 0x00, 0x00, 0x00, 0x42,
  0x42, 0x55, 0x00, 0x00, 0x00, 0x00, 0x68, 0x01, 0x00, 0x00
 };
 
@@ -607,7 +537,7 @@ const VpProfileDataType RING_AT[] =
 {
   /* Sine, 50 Hz, 1.41 CF, 80.00 Vpk, 0.00 Bias */
   /* Ringing Profile */
- 0x00, 0x04, 0x00, 0x12, 0x00, 0x0E, 0xD2, 0x00, 0x00, 0x00, 0x00, 0x88,
+ 0x00, 0x04, 0x00, 0x12, 0x00, 0x0E, 0xD2, 0x00, 0x00, 0x00, 0x00, 0x89,
  0x42, 0x55, 0x00, 0x00, 0x00, 0x00, 0x68, 0x01, 0x00, 0x00
 };
 
@@ -616,21 +546,13 @@ const VpProfileDataType RING_AU[] =
 {
   /* Sine, 20 Hz, 1.41 CF, 80.00 Vpk, 0.00 Bias */
   /* Ringing Profile */
- 0x00, 0x04, 0x00, 0x12, 0x00, 0x0E, 0xD2, 0x00, 0x00, 0x00, 0x00, 0x36,
+ 0x00, 0x04, 0x00, 0x12, 0x00, 0x0E, 0xD2, 0x00, 0x00, 0x00, 0x00, 0x37,
  0x42, 0x55, 0x00, 0x00, 0x00, 0x00, 0x68, 0x01, 0x00, 0x00
 };
 
-
 /************** Call Progress Tones **************/
 
 /************** Cadence Definitions **************/
-/* RING Cadence Profile */
-const VpProfileDataType RING_CAD_STD[] =
-{
-  /* Cadence Profile */
- 0x00, 0x08, 0x00, 0x0E, 0x00, 0x00, 0x00, 0x0A, 0x01, 0x07, 0x21, 0x90,
- 0x01, 0x00, 0x23, 0x20, 0x40, 0x00
-};
 
 /************** Caller ID **************/
 
@@ -641,4 +563,4 @@ int dc_profile_size = sizeof(DC_22MA_CC);
 int ac_profile_size = sizeof(AC_FXS_RF14_DEF);
 int ring_profile_size = sizeof(RING_DEF);
 
-/* end of file profile_8266.c */
+/* end of file profile_88266.c */
diff --git a/tools/voice/zarlink/kernel/profile_88266.h b/tools/voice/zarlink/kernel/profile_88266.h
index 1b14d76..4ea2a90 100644
--- a/tools/voice/zarlink/kernel/profile_88266.h
+++ b/tools/voice/zarlink/kernel/profile_88266.h
@@ -1,18 +1,19 @@
 /*
- * profile_8266.h --
+ * profile_88266.h --
  *
  * This header file exports the Profile data types
  *
  * Project Info --
- *   File:  C:\Documents and Settings\benavi\Desktop\tmp\NewProject.vpw
- *   Type:  Le71HR0865G Line Module Project - VBL Reg ABS Power Supply and 8.192MHz PCLK
- *   Date:  Thursday, February 25, 2010 16:58:41
+ *   File:   C:\Users\Kosmo\Documents\VE880.vpw
+ *   Type:   Le71HR0865G Line Module Project - VBL Reg ABS Power Supply and 8.192MHz PCLK
+ *   Date:   Thursday, April 18, 2013 12:59:41
+ *   Device: VE880 Le88266
  *
- *   This file was generated with Profile Wizard Version: P1.12.3
+ *   This file was generated with Profile Wizard Version: P2.4.0
  */
 
-#ifndef PROFILE_8266_H
-#define PROFILE_8266_H
+#ifndef PROFILE_88266_H
+#define PROFILE_88266_H
 
 #ifdef VP_API_TYPES_H
 #include "vp_api_types.h"
@@ -26,8 +27,7 @@ extern int ac_profile_size;
 extern int ring_profile_size;
 
 /************** Device Parameters **************/
-/* Device Profile */
-extern const VpProfileDataType ABS_VBL_FLYBACK[];
+extern const VpProfileDataType ABS_VBL_FLYBACK[];    /* Device Configuration Data */
 
 /************** AC Filter Coefficients **************/
 extern const VpProfileDataType AC_FXS_RF14_DEF[];    /* AC FXS RF14 600 Ohm Coefficients [Default} */
@@ -62,14 +62,6 @@ extern const VpProfileDataType AC_FXS_RF14_US_loaded[];/* AC FXS RF14 US 900//2.
 extern const VpProfileDataType AC_FXS_RF14_US_Nonloaded[];/* AC FXS RF14 US 900//2,16uF_800//(100+50nF) Coefficients */
 extern const VpProfileDataType AC_FXS_RF14_US_SS[];  /* AC FXS RF14 US 900//2.16uF Coefficients */
 
-/************** WideBand AC Filter Coefficients **************/
-extern const VpProfileDataType AC_FXS_RF14_WB_US[];  /* AC FXS RF14 600 Ohm Wideband Coefficients */
-extern const VpProfileDataType AC_FXS_RF14_WB_EU[];  /* AC FXS RF14 European Union Wideband Coefficients */
-extern const VpProfileDataType AC_FXS_RF14_WB_FR[];  /* AC FXS RF14 France Wideband Coefficients */
-extern const VpProfileDataType AC_FXS_RF14_WB_DE[];  /* AC FXS RF14 German Wideband Coefficients */
-extern const VpProfileDataType AC_FXS_RF14_WB_CN[];  /* AC FXS RF14 China Wideband Coefficients */
-extern const VpProfileDataType AC_FXS_RF14_WB_AU[];  /* AC FXS RF14 Australia Wideband Coefficients */
-
 /************** DC Feed Parameters **************/
 extern const VpProfileDataType DC_22MA_CC[];         /* 22mA  current feed */
 
@@ -89,9 +81,10 @@ extern const VpProfileDataType RING_AU[];            /* Australia Ringing 20Hz,
 /************** Call Progress Tones **************/
 
 /************** Cadence Definitions **************/
-extern const VpProfileDataType RING_CAD_STD[];
+
 /************** Caller ID **************/
 
 /************** Metering Profile **************/
 
-#endif /* PROFILE_8266_H */
+#endif /* PROFILE_88266_H */
+
diff --git a/tools/voice/zarlink/kernel/profile_89116.c b/tools/voice/zarlink/kernel/profile_89116.c
index d69e91a..fe09f67 100644
--- a/tools/voice/zarlink/kernel/profile_89116.c
+++ b/tools/voice/zarlink/kernel/profile_89116.c
@@ -4,12 +4,12 @@
  * This file contains profile data in byte format
  *
  * Project Info --
- *   File:   C:\Microsemi\Profile_Wizard_P2.1.2\890\NewProject116_new.vpw
+ *   File:   C:\Users\Kosmo\Documents\VE890.vpw
  *   Type:   VE890 Configuration for 100V Buck-Boost Power Supply, Lite Narrowband FXS Coefficients, and 8.192MHz PCLK
- *   Date:   Wednesday, May 02, 2012 12:15:05
+ *   Date:   Thursday, April 18, 2013 13:08:09
  *   Device: VE890 Le89316
  *
- *   This file was generated with Profile Wizard Version: P2.1.2
+ *   This file was generated with Profile Wizard Version: P2.4.0
  *
  * Project Comments --
  *  --------------------------------------------
@@ -124,7 +124,7 @@ const VpProfileDataType DEV_PROFILE_VE890_BB100V[] =
  0x20, 0x00, /* PCLK Frequency */
  0x8A, 0x00, /* Device Config 1 and MCLK Correction */
  0x40, 0x02, /* Clock Slot and Max Events */
- 0x0A, 0x00, /* Device Tickrate */
+ 0x05, 0x00, /* Device Tickrate */
  0xE4, 0x02, 0x44, 0x0A, /* Buck-Boost Switching Regulator Parameters */
  0x75, 0x00, 0x98, 0x40, 0xC0, 0x40, /* Regulator Timing Parameters */
  0x28, 0x40, 0x30, 0x40, 0x30, 0x40  /* FreeRun Timing Parameters */
@@ -1474,7 +1474,7 @@ const VpProfileDataType DC_FXS_VE890_BB100V_DEF[] =
 {
  /* DC Profile */
  0x00, 0x01, 0x00, 0x09, 0x02, 0x03, /* Header Information */
- 0xC6, 0x13, 0xC4, /* Feed Parameters */
+ 0xC6, 0x13, 0xC7, /* Feed Parameters */
  0x18, 0x84, /* Loop Supervision Parameters */
  0x04, /* Min Floor Voltage 25V */
  0x00  /* Hook Threshold Hysteresis */
diff --git a/tools/voice/zarlink/kernel/profile_89116.h b/tools/voice/zarlink/kernel/profile_89116.h
index b17ea1e..afc3f33 100644
--- a/tools/voice/zarlink/kernel/profile_89116.h
+++ b/tools/voice/zarlink/kernel/profile_89116.h
@@ -4,12 +4,12 @@
  * This header file exports the Profile data types
  *
  * Project Info --
- *   File:   C:\Microsemi\Profile_Wizard_P2.1.2\890\NewProject116_new.vpw
+ *   File:   C:\Users\Kosmo\Documents\VE890.vpw
  *   Type:   VE890 Configuration for 100V Buck-Boost Power Supply, Lite Narrowband FXS Coefficients, and 8.192MHz PCLK
- *   Date:   Wednesday, May 02, 2012 12:15:05
+ *   Date:   Thursday, April 18, 2013 13:08:09
  *   Device: VE890 Le89316
  *
- *   This file was generated with Profile Wizard Version: P2.1.2
+ *   This file was generated with Profile Wizard Version: P2.4.0
  *
  * Project Comments --
  *  --------------------------------------------
-- 
1.7.5.4

