From 0fe80fd0728f4b80bb7b4079dda57917bb427d49 Mon Sep 17 00:00:00 2001
From: Zumeng Chen <zumeng.chen@windriver.com>
Date: Tue, 21 Oct 2008 18:50:17 +0800
Subject: [PATCH 4/5] fix schedule_timeout in alloc_ioband_device

Implement finder grained locking to reduce the amount of
time with spinlocks held. Holding the spinlock for 
excessive time triggers schedule_timeout errors.

Signed-off-by Zumeng Chen <zumeng.chen@windriver.com>
---
 drivers/md/dm-ioband-ctl.c |    6 ++++--
 1 files changed, 4 insertions(+), 2 deletions(-)

diff --git a/drivers/md/dm-ioband-ctl.c b/drivers/md/dm-ioband-ctl.c
index a792620..18cec92 100644
--- a/drivers/md/dm-ioband-ctl.c
+++ b/drivers/md/dm-ioband-ctl.c
@@ -100,6 +100,7 @@ static struct ioband_device *alloc_ioband_device(char *name,
 			return dp;
 		}
 	}
+	spin_unlock_irqrestore(&ioband_devicelist_lock, flags);
 
 	/*
 	 * Prepare its own workqueue as generic_make_request() may
@@ -107,7 +108,6 @@ static struct ioband_device *alloc_ioband_device(char *name,
 	 */
 	new->g_ioband_wq = create_workqueue("kioband");
 	if (!new->g_ioband_wq) {
-		spin_unlock_irqrestore(&ioband_devicelist_lock, flags);
 		kfree(new);
 		return NULL;
 	}
@@ -133,9 +133,11 @@ static struct ioband_device *alloc_ioband_device(char *name,
 	init_waitqueue_head(&new->g_waitq);
 	init_waitqueue_head(&new->g_waitq_suspend);
 	init_waitqueue_head(&new->g_waitq_flush);
-	list_add_tail(&new->g_list, &ioband_device_list);
 
+	spin_lock_irqsave(&ioband_devicelist_lock, flags);
+	list_add_tail(&new->g_list, &ioband_device_list);
 	spin_unlock_irqrestore(&ioband_devicelist_lock, flags);
+
 	return new;
 }
 
-- 
1.5.5.1.67.gbdb8

