From 1b860cfd73a1f0682c853fb1089c7c618404bc2d Mon Sep 17 00:00:00 2001
From: Hui Wang <Hui.Wang@windriver.com>
Date: Tue, 1 Mar 2011 18:58:17 +0800
Subject: [PATCH 13/13] fsl_p1022_p1013: make lbc suspend safe for other platforms

Signed-off-by: Hui Wang <Hui.Wang@windriver.com>
---
 arch/powerpc/sysdev/fsl_lbc.c |    5 ++++-
 1 files changed, 4 insertions(+), 1 deletions(-)

diff --git a/arch/powerpc/sysdev/fsl_lbc.c b/arch/powerpc/sysdev/fsl_lbc.c
index 0ef4670..2951d36 100644
--- a/arch/powerpc/sysdev/fsl_lbc.c
+++ b/arch/powerpc/sysdev/fsl_lbc.c
@@ -340,8 +340,11 @@ err:
 #ifdef CONFIG_SUSPEND
 #ifdef CONFIG_P1022_DS
 #define COUNT_OF_BANK_P1022	(8)
-#endif
 #define COUNT_OF_BANKS COUNT_OF_BANK_P1022
+#else
+#define COUNT_OF_BANKS (8)
+#endif
+
 
 /* save lbc registers */
 static int fsl_lbc_suspend(struct of_device *ofdev, pm_message_t state)
-- 
1.6.5.2

