From 5d0e625b1ee02d9ace299c12f97cec04c6921b76 Mon Sep 17 00:00:00 2001
From: WRS Support <support@windriver.com>
Date: Fri, 2 Oct 2009 16:12:13 -0400
Subject: [PATCH 06/18] hypervisor: basic hooks in of hypervisor content

Put in place the Kconfig and Makefile hooks for WR HV/guest
content.
---
 init/Kconfig    |    2 ++
 kernel/Makefile |    1 +
 2 files changed, 3 insertions(+), 0 deletions(-)

diff --git a/init/Kconfig b/init/Kconfig
index eb77e8c..90ec2d2 100644
--- a/init/Kconfig
+++ b/init/Kconfig
@@ -23,6 +23,8 @@ config CONSTRUCTORS
 
 menu "General setup"
 
+source init/Kconfig.wrhv
+
 config EXPERIMENTAL
 	bool "Prompt for development and/or incomplete code/drivers"
 	---help---
diff --git a/kernel/Makefile b/kernel/Makefile
index 80b17ce..26977b1 100644
--- a/kernel/Makefile
+++ b/kernel/Makefile
@@ -39,6 +39,7 @@ obj-$(CONFIG_FUTEX) += futex.o
 ifeq ($(CONFIG_COMPAT),y)
 obj-$(CONFIG_FUTEX) += futex_compat.o
 endif
+obj-$(CONFIG_WRHV) += vbi/
 obj-$(CONFIG_RT_MUTEXES) += rtmutex.o
 obj-$(CONFIG_DEBUG_RT_MUTEXES) += rtmutex-debug.o
 obj-$(CONFIG_RT_MUTEX_TESTER) += rtmutex-tester.o
-- 
1.6.5.2

