From 9f6362fede8aa3fd65f759ca2996e21a61c9a5e9 Mon Sep 17 00:00:00 2001
From: Paul Gortmaker <paul.gortmaker@windriver.com>
Date: Mon, 19 Apr 2010 12:26:45 -0400
Subject: [PATCH] MILS: create separate CPU type for PPC604

The existing code was piggybacking off the definitions for
all PPC.  This introduces a new CPU type, to allow distinguising
it from E500 and similar for clarity.  There is no actual change
to generated code by this commit.

Signed-off-by: Paul Gortmaker <paul.gortmaker@windriver.com>
---
 include/vbi/cpu_types.h |    3 ++-
 include/vbi/types.h     |   10 +++++++---
 2 files changed, 9 insertions(+), 4 deletions(-)

diff --git a/include/vbi/cpu_types.h b/include/vbi/cpu_types.h
index dbceb30..75bf23d 100644
--- a/include/vbi/cpu_types.h
+++ b/include/vbi/cpu_types.h
@@ -21,6 +21,7 @@
 #define ARM1136		3
 #define MIPSI64R2	4
 #define PPCE500MC	5
+#define PPC604		6
 
 #ifndef _ASMLANGUAGE
 
@@ -50,7 +51,7 @@ typedef unsigned long	_RType;		/* registers are 64 bits */
 #endif /* CPU == MIPSI64R2 */
 
 /* PowerPC specific types */
-#if (CPU == PPC85XX) || (CPU == PPCE500MC)
+#if (CPU == PPC85XX) || (CPU == PPCE500MC) || (CPU == PPC604)
 typedef unsigned long	INSTR;		/* 32 bit word-aligned instructions */
 typedef unsigned int	_RType;		/* register type */
 #endif
diff --git a/include/vbi/types.h b/include/vbi/types.h
index b367c43..0d4b0aa 100644
--- a/include/vbi/types.h
+++ b/include/vbi/types.h
@@ -26,9 +26,13 @@
 #define CPU PENTIUM
 #endif /* CONFIG_X86 */
 
-#ifdef CONFIG_PPC
-#define CPU PPC85XX   /* Currently all PPC treated the same in vbi */
-#endif /* CONFIG_PPC */
+#ifdef CONFIG_E500
+#define CPU PPC85XX 
+#endif /* CONFIG_E500 */
+
+#ifdef CONFIG_6xx
+#define CPU PPC604
+#endif /* CONFIG_6xx */
 
 #include <vbi/cpu_types.h>	/* for PENTIUM, PPC85XX, etc */
 
-- 
1.6.5.2

