From a09c8cbd4ffbcda4b33621790e146430e9ba3601 Mon Sep 17 00:00:00 2001
From: Paul Gortmaker <paul.gortmaker@windriver.com>
Date: Fri, 15 Apr 2011 15:03:03 -0400
Subject: [PATCH] MIPS: fix compile failure for non-WRHV

Signed-off-by: Paul Gortmaker <paul.gortmaker@windriver.com>
---
 arch/mips/mm/fault.c |    2 ++
 1 files changed, 2 insertions(+), 0 deletions(-)

diff --git a/arch/mips/mm/fault.c b/arch/mips/mm/fault.c
index 49eb415..66e630e 100644
--- a/arch/mips/mm/fault.c
+++ b/arch/mips/mm/fault.c
@@ -286,6 +286,7 @@ vmalloc_fault:
 		return;
 	}
 #endif
+#ifdef CONFIG_WRHV
 wrhv_vmalloc_fault:
 	{
 		pgd_t *upd, *pgd;
@@ -303,4 +304,5 @@ wrhv_vmalloc_fault:
 
 		return;
 	}
+#endif
 }
-- 
1.7.0.4

