From d96b33bed8cd1f5b4daf9537e6b4e81afa84f6b8 Mon Sep 17 00:00:00 2001
From: Paul Gortmaker <paul.gortmaker@windriver.com>
Date: Thu, 15 Apr 2010 14:00:54 -0400
Subject: [PATCH] use get_pvr unconditionally

Signed-off-by: Paul Gortmaker <paul.gortmaker@windriver.com>
---
 arch/powerpc/kernel/traps.c |   10 +---------
 1 files changed, 1 insertions(+), 9 deletions(-)

diff --git a/arch/powerpc/kernel/traps.c b/arch/powerpc/kernel/traps.c
index b8250ab..64ca63c 100644
--- a/arch/powerpc/kernel/traps.c
+++ b/arch/powerpc/kernel/traps.c
@@ -789,15 +789,7 @@ static int emulate_instruction(struct pt_regs *regs)
 	/* Emulate the mfspr rD, PVR. */
 	if ((instword & INST_MFSPR_PVR_MASK) == INST_MFSPR_PVR) {
 		rd = (instword >> 21) & 0x1f;
-#ifndef CONFIG_WRHV
-		regs->gpr[rd] = mfspr(SPRN_PVR);
-#else
-		/*
-		 * PVR for wrhv hypervisor should be 0x80200000,
-		 * why is it 0x80200010?
-		 */
-		regs->gpr[rd] = 0x80200010;
-#endif
+		regs->gpr[rd] = get_pvr();
 		return 0;
 	}
 
-- 
1.6.5.2

