From 0c33c7eb3594b307af34ff16f7dc23beb2b89781 Mon Sep 17 00:00:00 2001
From: Paul Gortmaker <paul.gortmaker@windriver.com>
Date: Mon, 23 Aug 2010 13:15:34 -0400
Subject: [PATCH] vbi: block warnings related to VBI syscalls

The VBI interface is exported everywhere, but it is only available
on a limited subset of supported architectures and BSPs.  Ensure
warnings of the syscall audit exclude the VBI syscalls from the
normal checklist so that unsupported targets see no warnings.

Signed-off-by: Paul Gortmaker <paul.gortmaker@windriver.com>

diff --git a/scripts/checksyscalls.sh b/scripts/checksyscalls.sh
index 66ad375..bab1bc9 100755
--- a/scripts/checksyscalls.sh
+++ b/scripts/checksyscalls.sh
@@ -185,6 +185,11 @@ cat << EOF
 #define __IGNORE_vfork		/* clone */
 #define __IGNORE_wait4		/* waitid */

+/* the WR VBI sys for the guest OS are largely uninteresting to audit */
+#define __IGNORE_vbi_mem
+#define __IGNORE_vbi_activate_vb
+#define __IGNORE_vbi_control
+
 /* sync_file_range had a stupid ABI. Allow sync_file_range2 instead */
 #ifdef __NR_sync_file_range2
 #define __IGNORE_sync_file_range
Signed-off-by: Paul Gortmaker <paul.gortmaker@windriver.com>
---
 scripts/checksyscalls.sh |    5 +++++
 1 files changed, 5 insertions(+), 0 deletions(-)

diff --git a/scripts/checksyscalls.sh b/scripts/checksyscalls.sh
index 66ad375..bab1bc9 100755
--- a/scripts/checksyscalls.sh
+++ b/scripts/checksyscalls.sh
@@ -185,6 +185,11 @@ cat << EOF
 #define __IGNORE_vfork		/* clone */
 #define __IGNORE_wait4		/* waitid */
 
+/* the WR VBI for the guest OS are largely uninteresting */
+#define __IGNORE_vbi_mem
+#define __IGNORE_vbi_activate_vb
+#define __IGNORE_vbi_control
+
 /* sync_file_range had a stupid ABI. Allow sync_file_range2 instead */
 #ifdef __NR_sync_file_range2
 #define __IGNORE_sync_file_range
-- 
1.7.2.1

