From b83da589d2c1eab518a741fd4d0116d3a19e3331 Mon Sep 17 00:00:00 2001
From: Thomas Tai <thomas.tai@windriver.com>
Date: Fri, 7 Jan 2011 10:00:06 -0500
Subject: [PATCH] vbi: correct header file structure mismatch

Two version of intrDeviceChanneBuffer is defined in hypervisor.
One in wrhv-1.2/include/sys/wrhvDevCore.h and one in
wrhv-1.2/vbi/include/vbiPdc.h. The wrong version of the
structure is being checkin sometime in end of Dec 2010 causing
duart not working. Corrected structure declaration and add comments
to show the hyervisor header file location for reference.

Signed-off-by: Thomas Tai <thomas.tai@windriver.com>
---
 include/vbi/pdc.h |   26 ++++++++++++--------------
 1 files changed, 12 insertions(+), 14 deletions(-)

diff --git a/include/vbi/pdc.h b/include/vbi/pdc.h
index 044e805..a60d1e9 100644
--- a/include/vbi/pdc.h
+++ b/include/vbi/pdc.h
@@ -64,22 +64,20 @@ struct ioctlOp
 }
 #endif /* DEV_ASYNC_IOCTL */
 
+/* following structure is taken from hypervisor
+ * include/sys/wrhvDevCore.h
+ */
 struct intr_device_channel_buffer
 {
-	VB_ALIGN_FIELD_64 (void *rxBuf, pad1);
-	VB_ALIGN_FIELD_64 (size_t rxBufLen, pad2);
-	VB_ALIGN_FIELD_64 (void *txBuf, pad3);
-	VB_ALIGN_FIELD_64 (size_t txBufLen, pad4);
-	uint64_t rxBufRdPtr;
-	uint64_t rxBufWrPtr;
-	uint64_t txBufRdPtr;
-	uint64_t txBufWrPtr;
-#ifdef DEV_ASYNC_IOCTL
-	VB_ALIGN_FIELD_64 (struct ioctlOp *ioctlBuf, pad5);
-	VB_ALIGN_FIELD_64 (size_t ioctlBufLen, pad6);
-	uint64_t ioctlBufRdPtr;
-	uint64_t ioctlBufWrPtr;
-#endif
+	VB_ALIGN_FIELD_64 (void *rxBuf, pad1);      /* rx buffer to rx  */
+	VB_ALIGN_FIELD_64 (size_t rxBufLen, pad2);  /* rx buffer length */
+	VB_ALIGN_FIELD_64 (void *txBuf, pad3);      /* tx buffer to tx  */
+	VB_ALIGN_FIELD_64 (size_t txBufLen, pad4);  /* tx buffer length */
+	uint32_t rxBufRdPtr;                /* rx buffer read pointer  */
+	uint32_t rxBufWrPtr;                /* rx buffer write pointer */
+	uint32_t txBufRdPtr;                /* tx buffer read pointer  */
+	uint32_t txBufWrPtr;                /* tx buffer write pointer */
+	uint32_t intStatus;                 /* intStatus, or what intr */
 };
 
 #define SYS_PDC_REQUEST_OK		0
-- 
1.7.0.4

