From d34d31d2231bc307fa0cc16708c8254bea3b5081 Mon Sep 17 00:00:00 2001
From: Bruce Ashfield <bruce.ashfield@windriver.com>
Date: Sun, 20 Jun 2010 16:19:55 -0400
Subject: [PATCH] WRHV/ppc: create e500 common config option

Signed-off-by: Bruce Ashfield <bruce.ashfield@windriver.com>
---
 arch/powerpc/platforms/85xx/Kconfig |    3 +++
 1 files changed, 3 insertions(+), 0 deletions(-)

diff --git a/arch/powerpc/platforms/85xx/Kconfig b/arch/powerpc/platforms/85xx/Kconfig
index b86f143..2f8b0fd 100644
--- a/arch/powerpc/platforms/85xx/Kconfig
+++ b/arch/powerpc/platforms/85xx/Kconfig
@@ -157,6 +157,9 @@ config P4080_DS
 	help
 	  This option enables support for the P4080 DS board
  
+config WRHV_E500
+	bool
+
 endif # FSL_SOC_BOOKE
 
 config TQM85xx
-- 
1.6.0.3

