From 08f31ac3dc731f82fa7ab483672d7f3650507da9 Mon Sep 17 00:00:00 2001
From: Ming Liu <ming.liu@windriver.com>
Date: Fri, 13 Apr 2012 15:56:06 +0800
Subject: [PATCH] lsi_app3300: change register setting order of app3k_clk_enable

Serial port output was found delayed by about 30 seconds at boot time
after decompressing the kernel, this is caused by a wrong order register
setting in function 'app3k_clk_enable'.

TIMER_n_CONTROL register should be set '0' in the first then we
can write a value to TIMER_n_LOAD register.

Signed-off-by Ming Liu <ming.liu@windriver.com>
---
 arch/arm/mach-app/clk-app3k.c |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/arch/arm/mach-app/clk-app3k.c b/arch/arm/mach-app/clk-app3k.c
index d75e146..b92bebb 100644
--- a/arch/arm/mach-app/clk-app3k.c
+++ b/arch/arm/mach-app/clk-app3k.c
@@ -37,8 +37,8 @@ static int app3k_clk_enable(struct clk *clk)
 
 	value = (core_speed / clk->rate) - 1;
 
-	__raw_writel(value, (clk->enable_reg + TIMER_n_LOAD));
 	__raw_writel(control, (clk->enable_reg + TIMER_n_CONTROL));
+	__raw_writel(value, (clk->enable_reg + TIMER_n_LOAD));
 
 
 	control = (TIMER_n_CONTROL_ENABLE |
-- 
1.7.0.4

