From 9b30e82fb75afb161c83aced192ab907f3eab18a Mon Sep 17 00:00:00 2001
From: Paul Gortmaker <paul.gortmaker@windriver.com>
Date: Thu, 19 Sep 2013 12:13:20 -0400
Subject: [PATCH] softirq: notify when masking messages

Signed-off-by: Paul Gortmaker <paul.gortmaker@windriver.com>

diff --git a/kernel/time/tick-sched.c b/kernel/time/tick-sched.c
index 855513e..11b161f 100644
--- a/kernel/time/tick-sched.c
+++ b/kernel/time/tick-sched.c
@@ -737,6 +737,8 @@ static bool can_stop_idle_tick(int cpu, struct tick_sched *ts)
 					printk(" %s", softirq_to_name[vec_nr]);
 			printk("\n");
 			ratelimit++;
+			if (ratelimit == 10)
+				pr_warn("NOHZ: masking local_softirq_pending messages\n ");
 		}
 		return false;
 	}
-- 
1.8.3.1

