From 22f7a003e430be520f95a103429f0b5c762e4e7f Mon Sep 17 00:00:00 2001
From: Corey Minyard <cminyard@mvista.com>
Date: Fri, 6 Sep 2013 19:31:52 -0500
Subject: [PATCH 316/382] ftrace MIPS/recordmcount: Don't ignore the first function in a file

Based on SDK octeon3_3.10.

When scanning files for mcount values, the MIPS "fake mcount" detector
would always fail if the first offset was at the beginning of a file.
So for a leaf function, whose mcount offset is zero, was at the
beginning of a file, it would not detect that the next offset was
a fake one.

Initialize the previous offset to 1 and use that for the comparison
instead of using zero, as 1 is an invalid offset you will never see.

Signed-off-by: Corey Minyard <cminyard@mvista.com>
Signed-off-by: David Daney <david.daney@cavium.com>
Signed-off-by: Yanjiang Jin <yanjiang.jin@windriver.com>
---
 scripts/recordmcount.h |    4 ++--
 1 files changed, 2 insertions(+), 2 deletions(-)

diff --git a/scripts/recordmcount.h b/scripts/recordmcount.h
index 9d1421e..a75d91b 100644
--- a/scripts/recordmcount.h
+++ b/scripts/recordmcount.h
@@ -163,11 +163,11 @@ static int mcount_adjust = 0;
 
 static int MIPS_is_fake_mcount(Elf_Rel const *rp)
 {
-	static Elf_Addr old_r_offset;
+	static Elf_Addr old_r_offset = 1; /* Init to an invalid offset */
 	Elf_Addr current_r_offset = _w(rp->r_offset);
 	int is_fake;
 
-	is_fake = old_r_offset &&
+	is_fake = (old_r_offset != 1) &&
 		(current_r_offset - old_r_offset == MIPS_FAKEMCOUNT_OFFSET);
 	old_r_offset = current_r_offset;
 
-- 
1.7.0.4

