From d7e4f48864a0a671a28d1e0d569eb225189f8268 Mon Sep 17 00:00:00 2001
From: Stanley.Miao <stanley.miao@windriver.com>
Date: Mon, 18 Oct 2010 15:26:30 +0800
Subject: [PATCH 05/26] Comcerto: M83xxx: ARAM registers difinition.

Original codes came from Mindspeed's vendor drop sdk-comcerto-openwrt-6.0.

The base address of ARAM configuration registers is 0x100F0000. Now add all
registers offset in aram_ctrl.h.

Integrated-by: Stanley.Miao <stanley.miao@windriver.com>
---
 .../arm/mach-comcerto/include/mach/comcerto-1000.h |    1 +
 .../include/mach/comcerto-1000/aram_ctrl.h         |   76 ++++++++++++++++++++
 2 files changed, 77 insertions(+), 0 deletions(-)
 create mode 100644 arch/arm/mach-comcerto/include/mach/comcerto-1000/aram_ctrl.h

diff --git a/arch/arm/mach-comcerto/include/mach/comcerto-1000.h b/arch/arm/mach-comcerto/include/mach/comcerto-1000.h
index 05a2941..b222198 100644
--- a/arch/arm/mach-comcerto/include/mach/comcerto-1000.h
+++ b/arch/arm/mach-comcerto/include/mach/comcerto-1000.h
@@ -89,6 +89,7 @@
 
 #define APB_VADDR(x)            ((x) - COMCERTO_AHB_APB_BASE + APB_VADDR_BASE)
 
+#include <mach/comcerto-1000/aram_ctrl.h>
 #include <mach/comcerto-1000/clk-rst.h>
 #include <mach/comcerto-1000/timer.h>
 #include <mach/comcerto-1000/intr.h>
diff --git a/arch/arm/mach-comcerto/include/mach/comcerto-1000/aram_ctrl.h b/arch/arm/mach-comcerto/include/mach/comcerto-1000/aram_ctrl.h
new file mode 100644
index 0000000..5d411fd
--- /dev/null
+++ b/arch/arm/mach-comcerto/include/mach/comcerto-1000/aram_ctrl.h
@@ -0,0 +1,76 @@
+/*
+ *  arch/arm/mach-comcerto/include/mach/comcerto-1000/aram_ctrl.h
+ *
+ *  Copyright (C) 2008 Mindspeed Technologies, Inc.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
+ */
+
+#ifndef __ASM_ARCH_COMCERTO1000_ARAM_CTRL_H
+#define __ASM_ARCH_COMCERTO1000_ARAM_CTRL_H
+
+#define SRESET_CTRL		0x00
+#define AHB_PORT_PRTY		0x04
+#define FPOOL_PRTY		0x08
+#define PORTS_FPOOL_PRTY	0x0C
+#define FPOOL_A_BASE		0x10
+#define FPOOL_A_LENG		0x14
+#define FPOOL_B_BASE		0x18
+#define FPOOL_B_LENG		0x1C
+#define FPOOL_A_RDPTR		0x20
+#define FPOOL_A_WRPTR		0x24
+#define FPOOL_A_DEPTH		0x28
+#define FPOOL_B_RDPTR		0x2C
+#define FPOOL_B_WRPTR		0x30
+#define FPOOL_B_DEPTH		0x34
+#define PIPE_CTRL		0x38
+#define SENSEAMP_DLY		0x3C
+#define ALLOC_BUF_CNT		0x40
+#define INT_EN			0x44
+#define INT_STAT		0x48
+#define FPOOL_A_LTHRES		0x4C
+#define FPOOL_A_HTHRES		0x50
+#define FPOOL_B_LTHRES		0x54
+#define FPOOL_B_HTHRES		0x58
+#define ARAM_MBIST		0x5C
+#define USB_MBIST		0x60
+#define IPSEC_MBIST		0x64
+#define IPSEC_SENSEAMP_DLY0	0x68
+#define IPSEC_SENSEAMP_DLY1	0x6C
+
+
+
+#define OVR_A_INT_EN		(1 << 0)
+#define UDR_A_INT_EN		(1 << 1)
+#define HTHRES_A_INT_EN		(1 << 2)
+#define LTHRES_A_INT_EN		(1 << 3)
+
+#define OVR_B_INT_EN		(1 << 16)
+#define UDR_B_INT_EN		(1 << 17)
+#define HTHRES_B_INT_EN		(1 << 18)
+#define LTHRES_B_INT_EN		(1 << 19)
+
+#define OVR_A_INT_STAT		(1 << 0)
+#define UDR_A_INT_STAT		(1 << 1)
+#define HTHRES_A_INT_STAT	(1 << 2)
+#define LTHRES_A_INT_STAT	(1 << 3)
+
+#define OVR_B_INT_STAT		(1 << 16)
+#define UDR_B_INT_STAT		(1 << 17)
+#define HTHRES_B_INT_STAT	(1 << 18)
+#define LTHRES_B_INT_STAT	(1 << 19)
+
+
+#endif /* __ASM_ARCH_COMCERTO1000_ARAM_CTRL_H */
-- 
1.6.5.2

