From 0fd208da049bda41994ae1e39da493ce9e6eed6a Mon Sep 17 00:00:00 2001
From: Yue Tao <Yue.Tao@windriver.com>
Date: Fri, 3 Nov 2017 10:27:38 +0800
Subject: [PATCH] Clean up the strings introduced by a merging conflict

Signed-off-by: Yue Tao <Yue.Tao@windriver.com>
---
 arch/xtensa/kernel/setup.c |   14 --------------
 1 files changed, 0 insertions(+), 14 deletions(-)

diff --git a/arch/xtensa/kernel/setup.c b/arch/xtensa/kernel/setup.c
index 438875c..6c8823b 100644
--- a/arch/xtensa/kernel/setup.c
+++ b/arch/xtensa/kernel/setup.c
@@ -240,25 +240,11 @@ void * __init early_init_dt_alloc_memory_arch(u64 size, u64 align)
 	return __alloc_bootmem(size, align, 0);
 }
 
-<<<<<<< found
 void __init early_init_devtree(void *params)
 {
 	if (sysmem.nr_banks == 0)
 		dt_memory_scan = true;
 
-||||||| expected
-void cpu_reset(void)
-{
-#if XCHAL_HAVE_PTP_MMU
-	local_irq_disable();
-	/*
-=======
-void cpu_reset(void)
-{
-#if XCHAL_HAVE_PTP_MMU && IS_ENABLED(CONFIG_MMU)
-	local_irq_disable();
-	/*
->>>>>>> replacement
 	early_init_dt_scan(params);
 	of_scan_flat_dt(xtensa_dt_io_area, NULL);
 
-- 
1.7.5.4

