From e3763b91027052b967a103a14824ba7058b3cf9e Mon Sep 17 00:00:00 2001
From: Liming Wang <liming.wang@windriver.com>
Date: Fri, 11 Apr 2014 10:26:05 +0800
Subject: [PATCH 057/628] clocksource: cadence_ttc: Fix missing including file

Signed-off-by: Liming Wang <liming.wang@windriver.com>
---
 drivers/clocksource/cadence_ttc_timer.c |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/drivers/clocksource/cadence_ttc_timer.c b/drivers/clocksource/cadence_ttc_timer.c
index b2bb3a4b..25961ea 100644
--- a/drivers/clocksource/cadence_ttc_timer.c
+++ b/drivers/clocksource/cadence_ttc_timer.c
@@ -21,7 +21,7 @@
 #include <linux/of_address.h>
 #include <linux/of_irq.h>
 #include <linux/slab.h>
-#include <linux/sched_clock.h>
+#include <asm/sched_clock.h>
 
 /*
  * This driver configures the 2 16-bit count-up timers as follows:
-- 
1.7.5.4

