From 78a5a3a63b4b2d379c5319c68fae16228fb9a283 Mon Sep 17 00:00:00 2001
From: Theodore Ts'o <tytso@mit.edu>
Date: Sun, 22 Jul 2012 20:23:31 -0400
Subject: [PATCH] ext4: remove unused variable in ext4_update_super()

commit 8a9918497bcf5aaa8d45eb61c373605bc4e8c81f upstream.

Signed-off-by: "Theodore Ts'o" <tytso@mit.edu>
Signed-off-by: Paul Gortmaker <paul.gortmaker@windriver.com>

diff --git a/fs/ext4/resize.c b/fs/ext4/resize.c
index 231cacb..0dbeb39 100644
--- a/fs/ext4/resize.c
+++ b/fs/ext4/resize.c
@@ -1150,7 +1150,7 @@ static void ext4_update_super(struct super_block *sb,
 	struct ext4_new_group_data *group_data = flex_gd->groups;
 	struct ext4_sb_info *sbi = EXT4_SB(sb);
 	struct ext4_super_block *es = sbi->s_es;
-	int i, ret;
+	int i;
 
 	BUG_ON(flex_gd->count == 0 || group_data == NULL);
 	/*
-- 
1.8.1.2

