From 845d76e1bec2aaa63b40152697c508018a8b7b0a Mon Sep 17 00:00:00 2001
From: Mark Asselstine <mark.asselstine@windriver.com>
Date: Fri, 24 Jun 2011 14:36:17 -0400
Subject: [PATCH] Fix 'function declaration isn't a prototype' warning

Signed-off-by: Mark Asselstine <mark.asselstine@windriver.com>
---
 include/linux/kernel.h |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/include/linux/kernel.h b/include/linux/kernel.h
index 7e69c81..04ed39c 100644
--- a/include/linux/kernel.h
+++ b/include/linux/kernel.h
@@ -322,7 +322,7 @@ static inline bool printk_timed_ratelimit(unsigned long *caller_jiffies, \
 					  unsigned int interval_msec)	\
 		{ return false; }
 static inline void register_emit_crash_char(void (*fn)(char c)) {}
-static inline void unregister_emit_crash_char() {}
+static inline void unregister_emit_crash_char(void) {}
 
 /* No effect, but we still get type checking even in the !PRINTK case: */
 #define printk_once(x...) printk(x)
-- 
1.7.4.1

