From 2186692d733adfc38840bb40d56a81533accf328 Mon Sep 17 00:00:00 2001
From: Liming Wang <liming.wang@windriver.com>
Date: Wed, 26 Mar 2014 18:33:26 +0800
Subject: [PATCH 319/628] i2c: xi2cps: Remove unused including file

Signed-off-by: Liming Wang <liming.wang@windriver.com>
---
 drivers/i2c/busses/i2c-xilinx_ps.c |    1 -
 1 files changed, 0 insertions(+), 1 deletions(-)

diff --git a/drivers/i2c/busses/i2c-xilinx_ps.c b/drivers/i2c/busses/i2c-xilinx_ps.c
index c4edfb1..67a1001 100644
--- a/drivers/i2c/busses/i2c-xilinx_ps.c
+++ b/drivers/i2c/busses/i2c-xilinx_ps.c
@@ -43,7 +43,6 @@
 #include <linux/of_i2c.h>
 #include <linux/platform_device.h>
 #include <linux/slab.h>
-#include <linux/xilinx_devices.h>
 
 /*
  * Register Map
-- 
1.7.5.4

