From 844777ab868aed64950672c61bef447272712799 Mon Sep 17 00:00:00 2001
From: Li Wang <li.wang@windriver.com>
Date: Fri, 6 Jul 2012 16:42:13 +0800
Subject: [PATCH] jffs2: use cond_resched() instead of yield()

upstream commit: 3866f673ebd86e5be2533923f5c0aed91fe1669f

yield() has different semantics meanwhile and even causes RT-kernels to
BUG. Replace the only appearance left in jffs2.

[root@host]# BUG: cp:620 RT task yield()-ing!
Call Trace:
[eac39ac0] [c0008d78] show_stack+0x68/0x1d8 (unreliable)
[eac39b10] [c04a6cb4] dump_stack+0x2c/0x44
[eac39b20] [c04a4200] yield+0x68/0x84
[eac39b30] [c01d3e50]jffs2_erase_pending_blocks+0x288/0x934
[eac39bc0] [c01cc308] jffs2_do_reserve_space+0x378/0x414
[eac39c10] [c01cc5ec] jffs2_reserve_space+0x1a0/0x240
[eac39c50] [c01cf828] jffs2_write_inode_range+0x238/0x2e8
[eac39cc0] [c01c9044] jffs2_write_end+0x13c/0x32c
[eac39d10] [c00d0f34] generic_file_buffered_write+0x108/0x2a0
[eac39da0] [c00d3718] __generic_file_aio_write+0x2b4/0x4f8
[eac39e10] [c00d39dc] generic_file_aio_write+0x80/0x108
[eac39e40] [c011262c] do_sync_write+0xbc/0x124
[eac39ee0] [c01134d8] vfs_write+0xb4/0x1ac
[eac39f00] [c011376c] sys_write+0x60/0x128
[eac39f40] [c001364c] ret_from_syscall+0x0/0x4
--- Exception: c01 at 0x480dcc3c
LR = 0x100065cc

Signed-off-by: Wolfram Sang <w.sang@pengutronix.de>
Signed-off-by: Artem Bityutskiy <Artem.Bityutskiy@nokia.com>
Signed-off-by: David Woodhouse <David.Woodhouse@intel.com>
Integrated-by: Li Wang <li.wang@windriver.com>
---
 fs/jffs2/erase.c |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/fs/jffs2/erase.c b/fs/jffs2/erase.c
index b47679b..8517ce1 100644
--- a/fs/jffs2/erase.c
+++ b/fs/jffs2/erase.c
@@ -148,7 +148,7 @@ void jffs2_erase_pending_blocks(struct jffs2_sb_info *c, int count)
 		}
 
 		/* Be nice */
-		yield();
+		cond_resched();
 		mutex_lock(&c->erase_free_sem);
 		spin_lock(&c->erase_completion_lock);
 	}
-- 
1.7.0.5

