From d1e7ea624757bf953541b45be65c4df350af8d9c Mon Sep 17 00:00:00 2001
From: Yue Tao <yue.tao@windriver.com>
Date: Mon, 2 Jun 2014 01:32:46 -0700
Subject: [PATCH] Keep the asmregparm NULL definition to avoid build error.

Signed-off-by: Yue Tao <yue.tao@windriver.com>
---
 include/linux/linkage.h |    4 ++++
 1 files changed, 4 insertions(+), 0 deletions(-)

diff --git a/include/linux/linkage.h b/include/linux/linkage.h
index a500d6b..5126cce 100644
--- a/include/linux/linkage.h
+++ b/include/linux/linkage.h
@@ -14,6 +14,10 @@
 #define asmlinkage CPP_ASMLINKAGE
 #endif
 
+#ifndef asmregparm
+# define asmregparm
+#endif
+
 #define __page_aligned_data	__section(.data.page_aligned) __aligned(PAGE_SIZE)
 #define __page_aligned_bss	__section(.bss.page_aligned) __aligned(PAGE_SIZE)
 
-- 
1.7.0

