From 13639f569cc70236773d2a3edc5f615d6d51539d Mon Sep 17 00:00:00 2001
From: Jianchuan Wang <jianchuan.wang@windriver.com>
Date: Thu, 4 Feb 2016 22:11:15 -0500
Subject: [PATCH] Livedump: Compile error for ppc64

Signed-off-by: Jianchuan Wang <jianchuan.wang@windriver.com>
---
 arch/powerpc/kernel/entry_64.S |    6 +++---
 1 files changed, 3 insertions(+), 3 deletions(-)

diff --git a/arch/powerpc/kernel/entry_64.S b/arch/powerpc/kernel/entry_64.S
index 358911e..1b7d88f 100644
--- a/arch/powerpc/kernel/entry_64.S
+++ b/arch/powerpc/kernel/entry_64.S
@@ -366,21 +366,21 @@ _GLOBAL(ppc_ptrace)
 	bl	.save_nvgprs
 #endif
 	bl	.sys_ptrace
-	b	syscall_exit
+	b	.Lsyscall_exit
 
 _GLOBAL(ppc_exit_group)
 #ifdef CONFIG_LIVEDUMP
 	bl	.save_nvgprs
 #endif
 	bl	.sys_exit_group
-	b	syscall_exit
+	b	.Lsyscall_exit
 
 _GLOBAL(ppc_exit)
 #ifdef CONFIG_LIVEDUMP
 	bl	.save_nvgprs
 #endif
 	bl	.sys_exit
-	b	syscall_exit
+	b	.Lsyscall_exit
 
 _GLOBAL(ret_from_fork)
 	bl	schedule_tail
-- 
1.7.5.4

