From 16b472d41e4877bea00a3df7bf8fecfb48a93aad Mon Sep 17 00:00:00 2001
From: Li Wang <li.wang@windriver.com>
Date: Fri, 5 Jul 2013 13:29:30 +0800
Subject: [PATCH 1/8] ltt: compile warning: call function not match its
 declaration

warning: initialization from incompatible pointer type [enabled by default]
warning: passing argument 1 of 'xxx' from incompatible pointer type [enabled by default]
note: expected 'xxx' but argument is of type 'yyy'

Signed-off-by: Li Wang <li.wang@windriver.com>
---
 ltt/ltt-trace-control.c   |    4 ++--
 ltt/probes/block-trace.c  |    2 +-
 ltt/probes/kernel-trace.c |    2 +-
 3 files changed, 4 insertions(+), 4 deletions(-)

diff --git a/ltt/ltt-trace-control.c b/ltt/ltt-trace-control.c
index 12a3f19..4759a33 100644
--- a/ltt/ltt-trace-control.c
+++ b/ltt/ltt-trace-control.c
@@ -989,7 +989,7 @@ static const struct file_operations info_fops = {
 	.read = marker_info_read,
 };
 
-static int marker_mkdir(struct inode *dir, struct dentry *dentry, int mode)
+static int marker_mkdir(struct inode *dir, struct dentry *dentry, umode_t mode)
 {
 	struct dentry *marker_d, *enable_d, *info_d, *channel_d;
 	int ret;
@@ -1079,7 +1079,7 @@ const struct inode_operations channel_dir_opt = {
 	.rmdir = marker_rmdir,
 };
 
-static int channel_mkdir(struct inode *dir, struct dentry *dentry, int mode)
+static int channel_mkdir(struct inode *dir, struct dentry *dentry, umode_t mode)
 {
 	struct dentry *channel_d;
 	int ret;
diff --git a/ltt/probes/block-trace.c b/ltt/probes/block-trace.c
index 84e41cd..2f88179 100644
--- a/ltt/probes/block-trace.c
+++ b/ltt/probes/block-trace.c
@@ -260,7 +260,7 @@ void probe_block_plug(void *data, struct request_queue *q)
 			 MARK_NOARGS);
 }
 
-void probe_block_unplug(void *data, struct request_queue *q)
+void probe_block_unplug(void *data, struct request_queue *q, unsigned int depth, bool explicit)
 {
 	unsigned int pdu = q->rq.count[READ] + q->rq.count[WRITE];
 
diff --git a/ltt/probes/kernel-trace.c b/ltt/probes/kernel-trace.c
index 6920f6f..e04dcda 100644
--- a/ltt/probes/kernel-trace.c
+++ b/ltt/probes/kernel-trace.c
@@ -272,7 +272,7 @@ void probe_sched_migrate_task(void *_data, struct task_struct *p, int dest_cpu)
 		p->pid, p->state, dest_cpu);
 }
 
-void probe_sched_signal_send(void *_data, int sig, struct siginfo *info, struct task_struct *t)
+void probe_sched_signal_send(void *_data, int sig, struct siginfo *info, struct task_struct *t, int group, int result)
 {
 	trace_mark_tp(kernel, send_signal, signal_generate,
 		probe_sched_signal_send, "pid %d signal %d", t->pid, sig);
-- 
1.7.5.4

