From 214a7487b5ad33ea881531fac4ceb73a54693e96 Mon Sep 17 00:00:00 2001
From: Li Wang <li.wang@windriver.com>
Date: Fri, 5 Jul 2013 13:59:07 +0800
Subject: [PATCH 5/8] ltt: compile warning: code should be after variable
 declaration

warning: ISO C90 forbids mixed declarations and code [-Wdeclaration-after-statement]

Signed-off-by: Li Wang <li.wang@windriver.com>
---
 ltt/ltt-statedump.c |    3 +--
 1 files changed, 1 insertions(+), 2 deletions(-)

diff --git a/ltt/ltt-statedump.c b/ltt/ltt-statedump.c
index 47a0232..739fd08 100644
--- a/ltt/ltt-statedump.c
+++ b/ltt/ltt-statedump.c
@@ -303,12 +303,11 @@ static inline void list_interrupts(struct ltt_probe_private_data *call_data)
 	unsigned int irq;
 	unsigned long flags = 0;
 	struct irq_desc *desc;
-	struct irq_chip *chip;
 
 	/* needs irq_desc */
 	for_each_irq_desc(irq, desc) {
 		struct irqaction *action;
-		chip = irq_desc_get_chip(desc);
+		struct irq_chip *chip = irq_desc_get_chip(desc);
 
 		const char *irq_chip_name =
 			chip->name ? : "unnamed_irq_chip";
-- 
1.7.5.4

