From 490629d78f6a21e04b079554e89bc6d50b468027 Mon Sep 17 00:00:00 2001
From: Yue Tao <Yue.Tao@windriver.com>
Date: Thu, 12 Oct 2017 23:42:41 -0700
Subject: [PATCH] Miss header file while backporting upstream commit 197e7e52

Signed-off-by: Yue Tao <Yue.Tao@windriver.com>
---
 mm/migrate.c |    1 +
 1 files changed, 1 insertions(+), 0 deletions(-)

diff --git a/mm/migrate.c b/mm/migrate.c
index 6e93f25..848442a 100644
--- a/mm/migrate.c
+++ b/mm/migrate.c
@@ -35,6 +35,7 @@
 #include <linux/syscalls.h>
 #include <linux/hugetlb.h>
 #include <linux/gfp.h>
+#include <linux/ptrace.h>
 
 #include <asm/tlbflush.h>
 
-- 
1.7.5.4

