From c989c62d094702138759f02f8e60ee22fec922fe Mon Sep 17 00:00:00 2001
From: Zumeng Chen <zumeng.chen@windriver.com>
Date: Tue, 8 Jun 2010 10:28:06 +0800
Subject: [PATCH] perf: give the correct path to perl

Signed-off-by: Zumeng Chen <zumeng.chen@windriver.com>
---
 tools/perf/Makefile |    8 ++++----
 1 files changed, 4 insertions(+), 4 deletions(-)

diff --git a/tools/perf/Makefile b/tools/perf/Makefile
index bc0f670..3fe5e27 100644
--- a/tools/perf/Makefile
+++ b/tools/perf/Makefile
@@ -514,8 +514,8 @@ else
 endif
 
 ifndef NO_LIBPERL
-PERL_EMBED_LDOPTS = `perl -MExtUtils::Embed -e ldopts 2>/dev/null`
-PERL_EMBED_CCOPTS = `perl -MExtUtils::Embed -e ccopts 2>/dev/null`
+PERL_EMBED_LDOPTS = `$(HOST_BUILD_BIN_DIR)perl -MExtUtils::Embed -e ldopts 2>/dev/null`
+PERL_EMBED_CCOPTS = `$(HOST_BUILD_BIN_DIR)perl -MExtUtils::Embed -e ccopts 2>/dev/null`
 endif
 
 ifneq ($(shell sh -c "(echo '\#include <EXTERN.h>'; echo '\#include <perl.h>'; echo 'int main(void) { perl_alloc(); return 0; }') | $(CC) -x c - $(PERL_EMBED_CCOPTS) -o $(BITBUCKET) $(PERL_EMBED_LDOPTS) > /dev/null 2>&1 && echo y"), y)
@@ -527,8 +527,8 @@ else
 endif
 
 ifndef NO_LIBPYTHON
-PYTHON_EMBED_LDOPTS = `python-config --ldflags 2>/dev/null`
-PYTHON_EMBED_CCOPTS = `python-config --cflags 2>/dev/null`
+PYTHON_EMBED_LDOPTS = `$(HOST_BUILD_BIN_DIR)python-config --ldflags 2>/dev/null`
+PYTHON_EMBED_CCOPTS = `$(HOST_BUILD_BIN_DIR)python-config --cflags 2>/dev/null`
 endif
 
 ifneq ($(shell sh -c "(echo '\#include <Python.h>'; echo 'int main(void) { Py_Initialize(); return 0; }') | $(CC) -x c - $(PYTHON_EMBED_CCOPTS) -o /dev/null $(PYTHON_EMBED_LDOPTS) > /dev/null 2>&1 && echo y"), y)
-- 
1.6.5.2

