From 4bce83d2637b7d4842bacbb0ef4b196d54c090a9 Mon Sep 17 00:00:00 2001
From: Yue Tao <Yue.Tao@windriver.com>
Date: Wed, 10 Sep 2014 14:40:49 +0800
Subject: [PATCH 8/8] Remove the redundant parameter of kvm_gfn_to_hva_cache_init()

Signed-off-by: Yue Tao <Yue.Tao@windriver.com>
---
 arch/x86/kvm/lapic.c |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/arch/x86/kvm/lapic.c b/arch/x86/kvm/lapic.c
index 0b97b5c..8d420d6 100644
--- a/arch/x86/kvm/lapic.c
+++ b/arch/x86/kvm/lapic.c
@@ -1213,7 +1213,7 @@ int kvm_lapic_set_vapic_addr(struct kvm_vcpu *vcpu, gpa_t vapic_addr)
 	if (vapic_addr) {
 		if (kvm_gfn_to_hva_cache_init(vcpu->kvm,
 					&vcpu->arch.apic->vapic_cache,
-					vapic_addr, sizeof(u32)))
+					vapic_addr))
 			return -EINVAL;
 	}
 
-- 
1.7.0

