From e07853d484fa3d9a98822369f2b085e1c64e5c93 Mon Sep 17 00:00:00 2001
From: Bruce Ashfield <bruce.ashfield@windriver.com>
Date: Thu, 31 Jan 2013 09:13:10 -0500
Subject: [PATCH] Revert "kbuild: clean up samples directory"

This reverts commit fb68d4be95fef4aab82c2a97da93a898d532dd57.

When coupled with the commit [Samples: Clean up so modules and hostprogs build
 correctly], this commit causes the following warning:

* Makefile:1178: target `_clean_samples' given more than once in the same rule.

Removing samples from clean-dirs in the top level Makefile fixes the
issue.

Signed-off-by: Bruce Ashfield <bruce.ashfield@windriver.com>
---
 Makefile |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/Makefile b/Makefile
index 18124f6..3d7cea3 100644
--- a/Makefile
+++ b/Makefile
@@ -1172,7 +1172,7 @@ MRPROPER_FILES += .config .config.old .version .old_version             \
 #
 clean: rm-dirs  := $(CLEAN_DIRS)
 clean: rm-files := $(CLEAN_FILES)
-clean-dirs      := $(addprefix _clean_, . $(vmlinux-alldirs) Documentation samples)
+clean-dirs      := $(addprefix _clean_, . $(vmlinux-alldirs) Documentation)
 
 PHONY += $(clean-dirs) clean archclean
 $(clean-dirs):
-- 
1.7.0.4

