From a70bff95c605ac1e5a9bab4fdcbcd33fb36f7ef1 Mon Sep 17 00:00:00 2001
From: Yue Tao <yue.tao@windriver.com>
Date: Thu, 19 May 2016 02:25:57 -0700
Subject: [PATCH] revert commit "mm: remove free_area_cache"

Signed-off-by: Yue Tao <yue.tao@windriver.com>
---
 arch/x86/mm/mmap.c |    2 ++
 1 files changed, 2 insertions(+), 0 deletions(-)

diff --git a/arch/x86/mm/mmap.c b/arch/x86/mm/mmap.c
index 2cd89be..7da1b9a 100644
--- a/arch/x86/mm/mmap.c
+++ b/arch/x86/mm/mmap.c
@@ -109,8 +109,10 @@ void arch_pick_mmap_layout(struct mm_struct *mm)
 	if (mmap_is_legacy()) {
 		mm->mmap_base = mm->mmap_legacy_base;
 		mm->get_unmapped_area = arch_get_unmapped_area;
+		mm->unmap_area = arch_unmap_area;
 	} else {
 		mm->mmap_base = mmap_base(random_factor);
 		mm->get_unmapped_area = arch_get_unmapped_area_topdown;
+		mm->unmap_area = arch_unmap_area_topdown;
 	}
 }
-- 
1.7.5.4

