From 5424316a6f6272c94f2f3e6aad14f8a9b281838f Mon Sep 17 00:00:00 2001
From: Li Wang <li.wang@windriver.com>
Date: Fri, 5 Jul 2013 13:37:54 +0800
Subject: [PATCH 3/8] scripts: compile warning: ignoring return value of
 function

warning: ignoring return value of 'xxx', declared with attribute warn_unused_result [-Wunused-result]

Signed-off-by: Li Wang <li.wang@windriver.com>
---
 scripts/mod/modpost.c |    8 ++++----
 1 files changed, 4 insertions(+), 4 deletions(-)

diff --git a/scripts/mod/modpost.c b/scripts/mod/modpost.c
index 00acd10..2fd9279 100644
--- a/scripts/mod/modpost.c
+++ b/scripts/mod/modpost.c
@@ -1739,8 +1739,8 @@ static void get_markers(struct elf_info *info, struct module *mod)
 				name = strings + sym->st_value;
 			const char *fmt = strchr(name, '\0') + 1;
 			char *line = NULL;
-			asprintf(&line, "%s\t%s\t%s\n", name, mod->name, fmt);
-			NOFAIL(line);
+			if (asprintf(&line, "%s\t%s\t%s\n", name, mod->name, fmt) == -1)
+				NOFAIL(line);
 			mod->markers[n++] = line;
 		}
 }
@@ -2165,8 +2165,8 @@ static void write_dump(const char *fname)
 static void add_marker(struct module *mod, const char *name, const char *fmt)
 {
 	char *line = NULL;
-	asprintf(&line, "%s\t%s\t%s\n", name, mod->name, fmt);
-	NOFAIL(line);
+	if (asprintf(&line, "%s\t%s\t%s\n", name, mod->name, fmt) == -1)
+		NOFAIL(line);
 
 	mod->markers = NOFAIL(realloc(mod->markers, ((mod->nmarkers + 1) *
 						     sizeof mod->markers[0])));
-- 
1.7.5.4

