From 6813050f99082ac6cdfdff45e064996d32fa0029 Mon Sep 17 00:00:00 2001
From: Yue Tao <yue.tao@windriver.com>
Date: Wed, 25 Nov 2015 21:01:19 -0800
Subject: [PATCH] The udf_get_filename has been redefined by upstream 0e5cc9a4

Signed-off-by: Yue Tao <yue.tao@windriver.com>
---
 fs/udf/dir.c |    3 ++-
 1 files changed, 2 insertions(+), 1 deletions(-)

diff --git a/fs/udf/dir.c b/fs/udf/dir.c
index b3e93f5..84a8bf6 100644
--- a/fs/udf/dir.c
+++ b/fs/udf/dir.c
@@ -163,7 +163,8 @@ static int do_udf_readdir(struct inode *dir, struct file *filp,
 			struct kernel_lb_addr tloc = lelb_to_cpu(cfi.icb.extLocation);
 
 			iblock = udf_get_lb_pblock(dir->i_sb, &tloc, 0);
-			flen = udf_get_filename(dir->i_sb, nameptr, fname, lfi);
+			flen = udf_get_filename(dir->i_sb, nameptr, lfi, fname,
+						UDF_NAME_LEN);
 			dt_type = DT_UNKNOWN;
 		}
 
-- 
1.7.5.4

